// nodemon/lib/utils/log.js
var colour = require('./colour');
var bus = require('./bus');
var required = false;
var useColours = true;

var coding = ***REMOVED***
  log: 'black',
  info: 'yellow',
  status: 'green',
  detail: 'yellow',
  fail: 'red',
  error: 'red',
***REMOVED***;

function log(type, text) ***REMOVED***
  var msg = '[nodemon] ' + (text || '');

  if (useColours) ***REMOVED***
    msg = colour(coding[type], msg);
  ***REMOVED***

  // always push the message through our bus
  bus.emit('log', ***REMOVED*** type: type, message: text, colour: msg ***REMOVED***);

  // but if we're running on the command line, also echo out
  // question: should we actually just consume our own events?
  if (!required) ***REMOVED***
    if (type === 'error') ***REMOVED***
      console.error(msg);
    ***REMOVED*** else ***REMOVED***
      console.log(msg || '');
    ***REMOVED***
  ***REMOVED***
***REMOVED***

var Logger = function (r) ***REMOVED***
  if (!(this instanceof Logger)) ***REMOVED***
    return new Logger(r);
  ***REMOVED***
  this.required(r);
  return this;
***REMOVED***;

Object.keys(coding).forEach(function (type) ***REMOVED***
  Logger.prototype[type] = log.bind(null, type);
***REMOVED***);

// detail is for messages that are turned on during debug
Logger.prototype.detail = function (msg) ***REMOVED***
  if (this.debug) ***REMOVED***
    log('detail', msg);
  ***REMOVED***
***REMOVED***;

Logger.prototype.required = function (val) ***REMOVED***
  required = val;
***REMOVED***;

Logger.prototype.debug = false;
Logger.prototype._log = function (type, msg) ***REMOVED***
  if (required) ***REMOVED***
    bus.emit('log', ***REMOVED*** type: type, message: msg || '', colour: msg || '' ***REMOVED***);
  ***REMOVED*** else if (type === 'error') ***REMOVED***
    console.error(msg);
  ***REMOVED*** else ***REMOVED***
    console.log(msg || '');
  ***REMOVED***
***REMOVED***;

Object.defineProperty(Logger.prototype, 'useColours', ***REMOVED***
  set: function (val) ***REMOVED***
    useColours = val;
  ***REMOVED***,
  get: function () ***REMOVED***
    return useColours;
  ***REMOVED***,
***REMOVED***);

module.exports = Logger;
