// empower-core/build/empower-core.js
/**
 * Modules in this bundle
 * @license
 *
 * empower-core:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   author: Takuto Wada <takuto.wada@gmail.com>
 *   contributors: James Talmage
 *   homepage: http://github.com/twada/empower-core
 *   version: 0.6.2
 *
 * call-signature:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   author: James Talmage <james@talmage.io>
 *   maintainers: jamestalmage <james@talmage.io>, twada <takuto.wada@gmail.com>
 *   homepage: https://github.com/jamestalmage/call-signature#readme
 *   version: 0.0.2
 *
 * core-js:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   maintainers: zloirock <zloirock@zloirock.ru>
 *   homepage: https://github.com/zloirock/core-js#readme
 *   version: 2.4.1
 *
 * This header is generated by licensify (https://github.com/twada/licensify)
 */
(function(f)***REMOVED***if(typeof exports==="object"&&typeof module!=="undefined")***REMOVED***module.exports=f()***REMOVED***else if(typeof define==="function"&&define.amd)***REMOVED***define([],f)***REMOVED***else***REMOVED***var g;if(typeof window!=="undefined")***REMOVED***g=window***REMOVED***else if(typeof global!=="undefined")***REMOVED***g=global***REMOVED***else if(typeof self!=="undefined")***REMOVED***g=self***REMOVED***else***REMOVED***g=this***REMOVED***g.empowerCore = f()***REMOVED******REMOVED***)(function()***REMOVED***var define,module,exports;return (function e(t,n,r)***REMOVED***function s(o,u)***REMOVED***if(!n[o])***REMOVED***if(!t[o])***REMOVED***var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw (f.code="MODULE_NOT_FOUND", f)***REMOVED***var l=n[o]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;t[o][0].call(l.exports,function(e)***REMOVED***var n=t[o][1][e];return s(n?n:e)***REMOVED***,l,l.exports,e,t,n,r)***REMOVED***return n[o].exports***REMOVED***var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s***REMOVED***)(***REMOVED***1:[function(_dereq_,module,exports)***REMOVED***
/**
 * empower-core - Power Assert feature enhancer for assert function/object.
 *
 * https://github.com/twada/empower-core
 *
 * Copyright (c) 2013-2017 Takuto Wada
 * Licensed under the MIT license.
 *   https://github.com/twada/empower-core/blob/master/MIT-LICENSE.txt
 */
var create = _dereq_('core-js/library/fn/object/create');
var assign = _dereq_('core-js/library/fn/object/assign');
var defaultOptions = _dereq_('./lib/default-options');
var Decorator = _dereq_('./lib/decorator');
var define = _dereq_('./lib/define-properties');
var slice = Array.prototype.slice;

/**
 * Enhance Power Assert feature to assert function/object.
 * @param assert target assert function or object to enhance
 * @param options enhancement options
 * @return enhanced assert function/object
 */
function empowerCore (assert, options) ***REMOVED***
    var typeOfAssert = (typeof assert);
    var enhancedAssert;
    if ((typeOfAssert !== 'object' && typeOfAssert !== 'function') || assert === null) ***REMOVED***
        throw new TypeError('empower-core argument should be a function or object.');
    ***REMOVED***
    if (isEmpowered(assert)) ***REMOVED***
        return assert;
    ***REMOVED***
    switch (typeOfAssert) ***REMOVED***
    case 'function':
        enhancedAssert = empowerAssertFunction(assert, options);
        break;
    case 'object':
        enhancedAssert = empowerAssertObject(assert, options);
        break;
    default:
        throw new Error('Cannot be here');
    ***REMOVED***
    define(enhancedAssert, ***REMOVED*** _empowered: true ***REMOVED***);
    return enhancedAssert;
***REMOVED***

function empowerAssertObject (assertObject, options) ***REMOVED***
    var config = assign(defaultOptions(), options);
    var target = config.destructive ? assertObject : create(assertObject);
    var decorator = new Decorator(target, config);
    return assign(target, decorator.enhancement());
***REMOVED***

function empowerAssertFunction (assertFunction, options) ***REMOVED***
    var config = assign(defaultOptions(), options);
    if (config.destructive) ***REMOVED***
        throw new Error('cannot use destructive:true to function.');
    ***REMOVED***
    var decorator = new Decorator(assertFunction, config);
    var enhancement = decorator.enhancement();
    var powerAssert;
    if (typeof enhancement === 'function') ***REMOVED***
        powerAssert = function powerAssert () ***REMOVED***
            return enhancement.apply(null, slice.apply(arguments));
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        powerAssert = function powerAssert () ***REMOVED***
            return assertFunction.apply(null, slice.apply(arguments));
        ***REMOVED***;
    ***REMOVED***
    assign(powerAssert, assertFunction);
    return assign(powerAssert, enhancement);
***REMOVED***

function isEmpowered (assertObjectOrFunction) ***REMOVED***
    return assertObjectOrFunction._empowered;
***REMOVED***

empowerCore.defaultOptions = defaultOptions;
module.exports = empowerCore;

***REMOVED***,***REMOVED***"./lib/decorator":3,"./lib/default-options":4,"./lib/define-properties":5,"core-js/library/fn/object/assign":11,"core-js/library/fn/object/create":12***REMOVED***],2:[function(_dereq_,module,exports)***REMOVED***
'use strict';

var some = _dereq_('core-js/library/fn/array/some');
var map = _dereq_('core-js/library/fn/array/map');

function decorate (callSpec, decorator) ***REMOVED***
    var numArgsToCapture = callSpec.numArgsToCapture;

    return function decoratedAssert () ***REMOVED***
        var context, message, hasMessage = false;

        // see: https://github.com/twada/empower-core/pull/8#issue-127859465
        // see: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments
        var args = new Array(arguments.length);
        for(var i = 0; i < args.length; ++i) ***REMOVED***
            args[i] = arguments[i];
        ***REMOVED***

        if (numArgsToCapture === (args.length - 1)) ***REMOVED***
            message = args.pop();
            hasMessage = true;
        ***REMOVED***

        var invocation = ***REMOVED***
            thisObj: this,
            values: args,
            message: message,
            hasMessage: hasMessage
        ***REMOVED***;

        if (some(args, isCaptured)) ***REMOVED***
            invocation.values = map(args.slice(0, numArgsToCapture), function (arg) ***REMOVED***
                if (isNotCaptured(arg)) ***REMOVED***
                    return arg;
                ***REMOVED***
                if (!context) ***REMOVED***
                    context = ***REMOVED***
                        source: arg.source,
                        args: []
                    ***REMOVED***;
                ***REMOVED***
                context.args.push(***REMOVED***
                    value: arg.powerAssertContext.value,
                    events: arg.powerAssertContext.events
                ***REMOVED***);
                return arg.powerAssertContext.value;
            ***REMOVED***);

            return decorator.concreteAssert(callSpec, invocation, context);
        ***REMOVED*** else ***REMOVED***
            return decorator.concreteAssert(callSpec, invocation);
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

function isNotCaptured (value) ***REMOVED***
    return !isCaptured(value);
***REMOVED***

function isCaptured (value) ***REMOVED***
    return (typeof value === 'object') &&
        (value !== null) &&
        (typeof value.powerAssertContext !== 'undefined');
***REMOVED***

module.exports = decorate;

***REMOVED***,***REMOVED***"core-js/library/fn/array/map":9,"core-js/library/fn/array/some":10***REMOVED***],3:[function(_dereq_,module,exports)***REMOVED***
'use strict';

var forEach = _dereq_('core-js/library/fn/array/for-each');
var filter = _dereq_('core-js/library/fn/array/filter');
var map = _dereq_('core-js/library/fn/array/map');
var signature = _dereq_('call-signature');
var decorate = _dereq_('./decorate');
var keys = _dereq_('core-js/library/fn/object/keys');


function Decorator (receiver, config) ***REMOVED***
    this.receiver = receiver;
    this.config = config;
    this.onError = config.onError;
    this.onSuccess = config.onSuccess;
    this.signatures = map(config.patterns, parse);
    this.wrapOnlySignatures = map(config.wrapOnlyPatterns, parse);
***REMOVED***

Decorator.prototype.enhancement = function () ***REMOVED***
    var that = this;
    var container = this.container();
    var wrappedMethods = [];

    function attach(matcherSpec, enhanced) ***REMOVED***
        var matcher = matcherSpec.parsed;
        var methodName = detectMethodName(matcher.callee);
        if (typeof that.receiver[methodName] !== 'function' || wrappedMethods.indexOf(methodName) !== -1) ***REMOVED***
            return;
        ***REMOVED***
        var callSpec = ***REMOVED***
            thisObj: that.receiver,
            func: that.receiver[methodName],
            numArgsToCapture: numberOfArgumentsToCapture(matcherSpec),
            matcherSpec: matcherSpec,
            enhanced: enhanced
        ***REMOVED***;
        container[methodName] = callSpec.enhancedFunc = decorate(callSpec, that);
        wrappedMethods.push(methodName);
    ***REMOVED***

    forEach(filter(this.signatures, methodCall), function (matcher) ***REMOVED***
        attach(matcher, true);
    ***REMOVED***);

    forEach(filter(this.wrapOnlySignatures, methodCall), function (matcher) ***REMOVED***
        attach(matcher, false);
    ***REMOVED***);

    return container;
***REMOVED***;

Decorator.prototype.container = function () ***REMOVED***
    var basement = ***REMOVED******REMOVED***;
    if (typeof this.receiver === 'function') ***REMOVED***
        var candidates = filter(this.signatures, functionCall);
        var enhanced = true;
        if (candidates.length === 0) ***REMOVED***
            enhanced = false;
            candidates = filter(this.wrapOnlySignatures, functionCall);
        ***REMOVED***
        if (candidates.length === 1) ***REMOVED***
            var callSpec = ***REMOVED***
                thisObj: null,
                func: this.receiver,
                numArgsToCapture: numberOfArgumentsToCapture(candidates[0]),
                matcherSpec: candidates[0],
                enhanced: enhanced
            ***REMOVED***;
            basement = callSpec.enhancedFunc = decorate(callSpec, this);
        ***REMOVED***
    ***REMOVED***
    return basement;
***REMOVED***;

Decorator.prototype.concreteAssert = function (callSpec, invocation, context) ***REMOVED***
    var func = callSpec.func;
    var thisObj = this.config.bindReceiver ? callSpec.thisObj : invocation.thisObj;
    var enhanced = callSpec.enhanced;
    var args = invocation.values;
    var message = invocation.message;
    var matcherSpec = callSpec.matcherSpec;

    if (context && typeof this.config.modifyMessageBeforeAssert === 'function') ***REMOVED***
        message = this.config.modifyMessageBeforeAssert(***REMOVED***originalMessage: message, powerAssertContext: context***REMOVED***);
    ***REMOVED***
    args = args.concat(message);

    var data = ***REMOVED***
        thisObj: invocation.thisObj,
        assertionFunction: callSpec.enhancedFunc,
        originalMessage: message,
        defaultMessage: matcherSpec.defaultMessage,
        matcherSpec: matcherSpec,
        enhanced: enhanced,
        args: args
    ***REMOVED***;

    if (context) ***REMOVED***
        data.powerAssertContext = context;
    ***REMOVED***

    return this._callFunc(func, thisObj, args, data);
***REMOVED***;

// see: https://github.com/twada/empower-core/pull/8#issuecomment-173480982
Decorator.prototype._callFunc = function (func, thisObj, args, data) ***REMOVED***
    var ret;
    try ***REMOVED***
        ret = func.apply(thisObj, args);
    ***REMOVED*** catch (e) ***REMOVED***
        data.assertionThrew = true;
        data.error = e;
        return this.onError.call(thisObj, data);
    ***REMOVED***
    data.assertionThrew = false;
    data.returnValue = ret;
    return this.onSuccess.call(thisObj, data);
***REMOVED***;

function numberOfArgumentsToCapture (matcherSpec) ***REMOVED***
    var matcher = matcherSpec.parsed;
    var len = matcher.args.length;
    var lastArg;
    if (0 < len) ***REMOVED***
        lastArg = matcher.args[len - 1];
        if (lastArg.name === 'message' && lastArg.optional) ***REMOVED***
            len -= 1;
        ***REMOVED***
    ***REMOVED***
    return len;
***REMOVED***


function detectMethodName (callee) ***REMOVED***
    if (callee.type === 'MemberExpression') ***REMOVED***
        return callee.member;
    ***REMOVED***
    return null;
***REMOVED***


function functionCall (matcherSpec) ***REMOVED***
    return matcherSpec.parsed.callee.type === 'Identifier';
***REMOVED***


function methodCall (matcherSpec) ***REMOVED***
    return matcherSpec.parsed.callee.type === 'MemberExpression';
***REMOVED***

function parse(matcherSpec) ***REMOVED***
    if (typeof matcherSpec === 'string') ***REMOVED***
        matcherSpec = ***REMOVED***pattern: matcherSpec***REMOVED***;
    ***REMOVED***
    var ret = ***REMOVED******REMOVED***;
    forEach(keys(matcherSpec), function (key) ***REMOVED***
        ret[key] = matcherSpec[key];
    ***REMOVED***);
    ret.parsed = signature.parse(matcherSpec.pattern);
    return ret;
***REMOVED***


module.exports = Decorator;

***REMOVED***,***REMOVED***"./decorate":2,"call-signature":6,"core-js/library/fn/array/filter":7,"core-js/library/fn/array/for-each":8,"core-js/library/fn/array/map":9,"core-js/library/fn/object/keys":14***REMOVED***],4:[function(_dereq_,module,exports)***REMOVED***
'use strict';

module.exports = function defaultOptions () ***REMOVED***
    return ***REMOVED***
        destructive: false,
        bindReceiver: true,
        onError: onError,
        onSuccess: onSuccess,
        patterns: [
            'assert(value, [message])',
            'assert.ok(value, [message])',
            'assert.equal(actual, expected, [message])',
            'assert.notEqual(actual, expected, [message])',
            'assert.strictEqual(actual, expected, [message])',
            'assert.notStrictEqual(actual, expected, [message])',
            'assert.deepEqual(actual, expected, [message])',
            'assert.notDeepEqual(actual, expected, [message])',
            'assert.deepStrictEqual(actual, expected, [message])',
            'assert.notDeepStrictEqual(actual, expected, [message])'
        ],
        wrapOnlyPatterns: []
    ***REMOVED***;
***REMOVED***;

function onError (errorEvent) ***REMOVED***
    var e = errorEvent.error;
    if (errorEvent.powerAssertContext && /^AssertionError/.test(e.name)) ***REMOVED***
        e.powerAssertContext = errorEvent.powerAssertContext;
    ***REMOVED***
    throw e;
***REMOVED***

function onSuccess(successEvent) ***REMOVED***
    return successEvent.returnValue;
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],5:[function(_dereq_,module,exports)***REMOVED***
'use strict';

var defineProperty = _dereq_('core-js/library/fn/object/define-property');
var forEach = _dereq_('core-js/library/fn/array/for-each');
var keys = _dereq_('core-js/library/fn/object/keys');

module.exports = function defineProperties (obj, map) ***REMOVED***
    forEach(keys(map), function (name) ***REMOVED***
        defineProperty(obj, name, ***REMOVED***
            configurable: true,
            enumerable: false,
            value: map[name],
            writable: true
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***;

***REMOVED***,***REMOVED***"core-js/library/fn/array/for-each":8,"core-js/library/fn/object/define-property":13,"core-js/library/fn/object/keys":14***REMOVED***],6:[function(_dereq_,module,exports)***REMOVED***
'use strict';
module.exports.parse = parse;
module.exports.generate = generate;

// TODO(jamestalmage): Allow full range of identifier characters instead of just ASCII
//
// This will likely require a build step
//
// SPEC: http://www.ecma-international.org/ecma-262/5.1/#sec-7.6
//
// TOOLING:
//    https://github.com/mathiasbynens/regenerate
//    https://www.npmjs.com/package/regjsgen

var regex = /^\s*(?:([A-Za-z$_][A-Za-z0-9$_]*)\s*\.)?\s*([A-Za-z$_][A-Za-z0-9$_]*)\s*\(\s*((?:[A-Za-z$_][A-Za-z0-9$_]*)|(?:\[\s*[A-Za-z$_][A-Za-z0-9$_]*\s*]))?((?:\s*,\s*(?:(?:[A-Za-z$_][A-Za-z0-9$_]*)|(?:\[\s*[A-Za-z$_][A-Za-z0-9$_]*\s*])))+)?\s*\)\s*$/;

function parse(str) ***REMOVED***
	var match = regex.exec(str);
	if (!match) ***REMOVED***
		return null;
	***REMOVED***

	var callee;
	if (match[1]) ***REMOVED***
		callee = ***REMOVED***
			type: 'MemberExpression',
			object: match[1],
			member: match[2]
		***REMOVED***;
	***REMOVED*** else ***REMOVED***
		callee = ***REMOVED***
			type: 'Identifier',
			name: match[2]
		***REMOVED***;
	***REMOVED***

	var args = match[4] || '';
	args = args.split(',');
	if (match[3]) ***REMOVED***
		args[0] = match[3];
	***REMOVED***
	var trimmed = [];
	args.forEach(function (str) ***REMOVED***
		var optional = false;
		str = str.replace(/\s+/g, '');
		if (!str.length) ***REMOVED***
			return;
		***REMOVED***
		if (str.charAt(0) === '[' && str.charAt(str.length - 1) === ']') ***REMOVED***
			optional = true;
			str = str.substring(1, str.length - 1);
		***REMOVED***
		trimmed.push(***REMOVED***
			name: str,
			optional: optional
		***REMOVED***);
	***REMOVED***);

	return ***REMOVED***
		callee: callee,
		args: trimmed
	***REMOVED***;
***REMOVED***

function generate(parsed) ***REMOVED***
	var callee;
	if (parsed.callee.type === 'MemberExpression') ***REMOVED***
		callee = [
			parsed.callee.object,
			'.',
			parsed.callee.member
		];
	***REMOVED*** else ***REMOVED***
		callee = [parsed.callee.name];
	***REMOVED***
	return callee.concat([
		'(',
		parsed.args.map(function (arg) ***REMOVED***
			return arg.optional ? '[' + arg.name + ']' : arg.name;
		***REMOVED***).join(', '),
		')'
	]).join('');
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],7:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.filter');
module.exports = _dereq_('../../modules/_core').Array.filter;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.array.filter":59***REMOVED***],8:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.for-each');
module.exports = _dereq_('../../modules/_core').Array.forEach;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.array.for-each":60***REMOVED***],9:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.map');
module.exports = _dereq_('../../modules/_core').Array.map;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.array.map":61***REMOVED***],10:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.some');
module.exports = _dereq_('../../modules/_core').Array.some;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.array.some":62***REMOVED***],11:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.assign');
module.exports = _dereq_('../../modules/_core').Object.assign;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.object.assign":63***REMOVED***],12:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.create');
var $Object = _dereq_('../../modules/_core').Object;
module.exports = function create(P, D)***REMOVED***
  return $Object.create(P, D);
***REMOVED***;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.object.create":64***REMOVED***],13:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.define-property');
var $Object = _dereq_('../../modules/_core').Object;
module.exports = function defineProperty(it, key, desc)***REMOVED***
  return $Object.defineProperty(it, key, desc);
***REMOVED***;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.object.define-property":65***REMOVED***],14:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.keys');
module.exports = _dereq_('../../modules/_core').Object.keys;
***REMOVED***,***REMOVED***"../../modules/_core":22,"../../modules/es6.object.keys":66***REMOVED***],15:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(it)***REMOVED***
  if(typeof it != 'function')throw TypeError(it + ' is not a function!');
  return it;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],16:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object');
module.exports = function(it)***REMOVED***
  if(!isObject(it))throw TypeError(it + ' is not an object!');
  return it;
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-object":37***REMOVED***],17:[function(_dereq_,module,exports)***REMOVED***
// false -> Array#indexOf
// true  -> Array#includes
var toIObject = _dereq_('./_to-iobject')
  , toLength  = _dereq_('./_to-length')
  , toIndex   = _dereq_('./_to-index');
module.exports = function(IS_INCLUDES)***REMOVED***
  return function($this, el, fromIndex)***REMOVED***
    var O      = toIObject($this)
      , length = toLength(O.length)
      , index  = toIndex(fromIndex, length)
      , value;
    // Array#includes uses SameValueZero equality algorithm
    if(IS_INCLUDES && el != el)while(length > index)***REMOVED***
      value = O[index++];
      if(value != value)return true;
    // Array#toIndex ignores holes, Array#includes - not
    ***REMOVED*** else for(;length > index; index++)if(IS_INCLUDES || index in O)***REMOVED***
      if(O[index] === el)return IS_INCLUDES || index || 0;
    ***REMOVED*** return !IS_INCLUDES && -1;
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-index":51,"./_to-iobject":53,"./_to-length":54***REMOVED***],18:[function(_dereq_,module,exports)***REMOVED***
// 0 -> Array#forEach
// 1 -> Array#map
// 2 -> Array#filter
// 3 -> Array#some
// 4 -> Array#every
// 5 -> Array#find
// 6 -> Array#findIndex
var ctx      = _dereq_('./_ctx')
  , IObject  = _dereq_('./_iobject')
  , toObject = _dereq_('./_to-object')
  , toLength = _dereq_('./_to-length')
  , asc      = _dereq_('./_array-species-create');
module.exports = function(TYPE, $create)***REMOVED***
  var IS_MAP        = TYPE == 1
    , IS_FILTER     = TYPE == 2
    , IS_SOME       = TYPE == 3
    , IS_EVERY      = TYPE == 4
    , IS_FIND_INDEX = TYPE == 6
    , NO_HOLES      = TYPE == 5 || IS_FIND_INDEX
    , create        = $create || asc;
  return function($this, callbackfn, that)***REMOVED***
    var O      = toObject($this)
      , self   = IObject(O)
      , f      = ctx(callbackfn, that, 3)
      , length = toLength(self.length)
      , index  = 0
      , result = IS_MAP ? create($this, length) : IS_FILTER ? create($this, 0) : undefined
      , val, res;
    for(;length > index; index++)if(NO_HOLES || index in self)***REMOVED***
      val = self[index];
      res = f(val, index, O);
      if(TYPE)***REMOVED***
        if(IS_MAP)result[index] = res;            // map
        else if(res)switch(TYPE)***REMOVED***
          case 3: return true;                    // some
          case 5: return val;                     // find
          case 6: return index;                   // findIndex
          case 2: result.push(val);               // filter
        ***REMOVED*** else if(IS_EVERY)return false;          // every
      ***REMOVED***
    ***REMOVED***
    return IS_FIND_INDEX ? -1 : IS_SOME || IS_EVERY ? IS_EVERY : result;
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-species-create":20,"./_ctx":23,"./_iobject":35,"./_to-length":54,"./_to-object":55***REMOVED***],19:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object')
  , isArray  = _dereq_('./_is-array')
  , SPECIES  = _dereq_('./_wks')('species');

module.exports = function(original)***REMOVED***
  var C;
  if(isArray(original))***REMOVED***
    C = original.constructor;
    // cross-realm fallback
    if(typeof C == 'function' && (C === Array || isArray(C.prototype)))C = undefined;
    if(isObject(C))***REMOVED***
      C = C[SPECIES];
      if(C === null)C = undefined;
    ***REMOVED***
  ***REMOVED*** return C === undefined ? Array : C;
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-array":36,"./_is-object":37,"./_wks":58***REMOVED***],20:[function(_dereq_,module,exports)***REMOVED***
// 9.4.2.3 ArraySpeciesCreate(originalArray, length)
var speciesConstructor = _dereq_('./_array-species-constructor');

module.exports = function(original, length)***REMOVED***
  return new (speciesConstructor(original))(length);
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-species-constructor":19***REMOVED***],21:[function(_dereq_,module,exports)***REMOVED***
var toString = ***REMOVED******REMOVED***.toString;

module.exports = function(it)***REMOVED***
  return toString.call(it).slice(8, -1);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],22:[function(_dereq_,module,exports)***REMOVED***
var core = module.exports = ***REMOVED***version: '2.4.0'***REMOVED***;
if(typeof __e == 'number')__e = core; // eslint-disable-line no-undef
***REMOVED***,***REMOVED******REMOVED***],23:[function(_dereq_,module,exports)***REMOVED***
// optional / simple context binding
var aFunction = _dereq_('./_a-function');
module.exports = function(fn, that, length)***REMOVED***
  aFunction(fn);
  if(that === undefined)return fn;
  switch(length)***REMOVED***
    case 1: return function(a)***REMOVED***
      return fn.call(that, a);
    ***REMOVED***;
    case 2: return function(a, b)***REMOVED***
      return fn.call(that, a, b);
    ***REMOVED***;
    case 3: return function(a, b, c)***REMOVED***
      return fn.call(that, a, b, c);
    ***REMOVED***;
  ***REMOVED***
  return function(/* ...args */)***REMOVED***
    return fn.apply(that, arguments);
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_a-function":15***REMOVED***],24:[function(_dereq_,module,exports)***REMOVED***
// 7.2.1 RequireObjectCoercible(argument)
module.exports = function(it)***REMOVED***
  if(it == undefined)throw TypeError("Can't call method on  " + it);
  return it;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],25:[function(_dereq_,module,exports)***REMOVED***
// Thank's IE8 for his funny defineProperty
module.exports = !_dereq_('./_fails')(function()***REMOVED***
  return Object.defineProperty(***REMOVED******REMOVED***, 'a', ***REMOVED***get: function()***REMOVED*** return 7; ***REMOVED******REMOVED***).a != 7;
***REMOVED***);
***REMOVED***,***REMOVED***"./_fails":29***REMOVED***],26:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object')
  , document = _dereq_('./_global').document
  // in old IE typeof document.createElement is 'object'
  , is = isObject(document) && isObject(document.createElement);
module.exports = function(it)***REMOVED***
  return is ? document.createElement(it) : ***REMOVED******REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_global":30,"./_is-object":37***REMOVED***],27:[function(_dereq_,module,exports)***REMOVED***
// IE 8- don't enum bug keys
module.exports = (
  'constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf'
).split(',');
***REMOVED***,***REMOVED******REMOVED***],28:[function(_dereq_,module,exports)***REMOVED***
var global    = _dereq_('./_global')
  , core      = _dereq_('./_core')
  , ctx       = _dereq_('./_ctx')
  , hide      = _dereq_('./_hide')
  , PROTOTYPE = 'prototype';

var $export = function(type, name, source)***REMOVED***
  var IS_FORCED = type & $export.F
    , IS_GLOBAL = type & $export.G
    , IS_STATIC = type & $export.S
    , IS_PROTO  = type & $export.P
    , IS_BIND   = type & $export.B
    , IS_WRAP   = type & $export.W
    , exports   = IS_GLOBAL ? core : core[name] || (core[name] = ***REMOVED******REMOVED***)
    , expProto  = exports[PROTOTYPE]
    , target    = IS_GLOBAL ? global : IS_STATIC ? global[name] : (global[name] || ***REMOVED******REMOVED***)[PROTOTYPE]
    , key, own, out;
  if(IS_GLOBAL)source = name;
  for(key in source)***REMOVED***
    // contains in native
    own = !IS_FORCED && target && target[key] !== undefined;
    if(own && key in exports)continue;
    // export native or passed
    out = own ? target[key] : source[key];
    // prevent global pollution for namespaces
    exports[key] = IS_GLOBAL && typeof target[key] != 'function' ? source[key]
    // bind timers to global for call from export context
    : IS_BIND && own ? ctx(out, global)
    // wrap global constructors for prevent change them in library
    : IS_WRAP && target[key] == out ? (function(C)***REMOVED***
      var F = function(a, b, c)***REMOVED***
        if(this instanceof C)***REMOVED***
          switch(arguments.length)***REMOVED***
            case 0: return new C;
            case 1: return new C(a);
            case 2: return new C(a, b);
          ***REMOVED*** return new C(a, b, c);
        ***REMOVED*** return C.apply(this, arguments);
      ***REMOVED***;
      F[PROTOTYPE] = C[PROTOTYPE];
      return F;
    // make static versions for prototype methods
    ***REMOVED***)(out) : IS_PROTO && typeof out == 'function' ? ctx(Function.call, out) : out;
    // export proto methods to core.%CONSTRUCTOR%.methods.%NAME%
    if(IS_PROTO)***REMOVED***
      (exports.virtual || (exports.virtual = ***REMOVED******REMOVED***))[key] = out;
      // export proto methods to core.%CONSTRUCTOR%.prototype.%NAME%
      if(type & $export.R && expProto && !expProto[key])hide(expProto, key, out);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
// type bitmap
$export.F = 1;   // forced
$export.G = 2;   // global
$export.S = 4;   // static
$export.P = 8;   // proto
$export.B = 16;  // bind
$export.W = 32;  // wrap
$export.U = 64;  // safe
$export.R = 128; // real proto method for `library` 
module.exports = $export;
***REMOVED***,***REMOVED***"./_core":22,"./_ctx":23,"./_global":30,"./_hide":32***REMOVED***],29:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(exec)***REMOVED***
  try ***REMOVED***
    return !!exec();
  ***REMOVED*** catch(e)***REMOVED***
    return true;
  ***REMOVED***
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],30:[function(_dereq_,module,exports)***REMOVED***
// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
var global = module.exports = typeof window != 'undefined' && window.Math == Math
  ? window : typeof self != 'undefined' && self.Math == Math ? self : Function('return this')();
if(typeof __g == 'number')__g = global; // eslint-disable-line no-undef
***REMOVED***,***REMOVED******REMOVED***],31:[function(_dereq_,module,exports)***REMOVED***
var hasOwnProperty = ***REMOVED******REMOVED***.hasOwnProperty;
module.exports = function(it, key)***REMOVED***
  return hasOwnProperty.call(it, key);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],32:[function(_dereq_,module,exports)***REMOVED***
var dP         = _dereq_('./_object-dp')
  , createDesc = _dereq_('./_property-desc');
module.exports = _dereq_('./_descriptors') ? function(object, key, value)***REMOVED***
  return dP.f(object, key, createDesc(1, value));
***REMOVED*** : function(object, key, value)***REMOVED***
  object[key] = value;
  return object;
***REMOVED***;
***REMOVED***,***REMOVED***"./_descriptors":25,"./_object-dp":40,"./_property-desc":47***REMOVED***],33:[function(_dereq_,module,exports)***REMOVED***
module.exports = _dereq_('./_global').document && document.documentElement;
***REMOVED***,***REMOVED***"./_global":30***REMOVED***],34:[function(_dereq_,module,exports)***REMOVED***
module.exports = !_dereq_('./_descriptors') && !_dereq_('./_fails')(function()***REMOVED***
  return Object.defineProperty(_dereq_('./_dom-create')('div'), 'a', ***REMOVED***get: function()***REMOVED*** return 7; ***REMOVED******REMOVED***).a != 7;
***REMOVED***);
***REMOVED***,***REMOVED***"./_descriptors":25,"./_dom-create":26,"./_fails":29***REMOVED***],35:[function(_dereq_,module,exports)***REMOVED***
// fallback for non-array-like ES3 and non-enumerable old V8 strings
var cof = _dereq_('./_cof');
module.exports = Object('z').propertyIsEnumerable(0) ? Object : function(it)***REMOVED***
  return cof(it) == 'String' ? it.split('') : Object(it);
***REMOVED***;
***REMOVED***,***REMOVED***"./_cof":21***REMOVED***],36:[function(_dereq_,module,exports)***REMOVED***
// 7.2.2 IsArray(argument)
var cof = _dereq_('./_cof');
module.exports = Array.isArray || function isArray(arg)***REMOVED***
  return cof(arg) == 'Array';
***REMOVED***;
***REMOVED***,***REMOVED***"./_cof":21***REMOVED***],37:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(it)***REMOVED***
  return typeof it === 'object' ? it !== null : typeof it === 'function';
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],38:[function(_dereq_,module,exports)***REMOVED***
'use strict';
// 19.1.2.1 Object.assign(target, source, ...)
var getKeys  = _dereq_('./_object-keys')
  , gOPS     = _dereq_('./_object-gops')
  , pIE      = _dereq_('./_object-pie')
  , toObject = _dereq_('./_to-object')
  , IObject  = _dereq_('./_iobject')
  , $assign  = Object.assign;

// should work with symbols and should have deterministic property order (V8 bug)
module.exports = !$assign || _dereq_('./_fails')(function()***REMOVED***
  var A = ***REMOVED******REMOVED***
    , B = ***REMOVED******REMOVED***
    , S = Symbol()
    , K = 'abcdefghijklmnopqrst';
  A[S] = 7;
  K.split('').forEach(function(k)***REMOVED*** B[k] = k; ***REMOVED***);
  return $assign(***REMOVED******REMOVED***, A)[S] != 7 || Object.keys($assign(***REMOVED******REMOVED***, B)).join('') != K;
***REMOVED***) ? function assign(target, source)***REMOVED*** // eslint-disable-line no-unused-vars
  var T     = toObject(target)
    , aLen  = arguments.length
    , index = 1
    , getSymbols = gOPS.f
    , isEnum     = pIE.f;
  while(aLen > index)***REMOVED***
    var S      = IObject(arguments[index++])
      , keys   = getSymbols ? getKeys(S).concat(getSymbols(S)) : getKeys(S)
      , length = keys.length
      , j      = 0
      , key;
    while(length > j)if(isEnum.call(S, key = keys[j++]))T[key] = S[key];
  ***REMOVED*** return T;
***REMOVED*** : $assign;
***REMOVED***,***REMOVED***"./_fails":29,"./_iobject":35,"./_object-gops":42,"./_object-keys":44,"./_object-pie":45,"./_to-object":55***REMOVED***],39:[function(_dereq_,module,exports)***REMOVED***
// 19.1.2.2 / 15.2.3.5 Object.create(O [, Properties])
var anObject    = _dereq_('./_an-object')
  , dPs         = _dereq_('./_object-dps')
  , enumBugKeys = _dereq_('./_enum-bug-keys')
  , IE_PROTO    = _dereq_('./_shared-key')('IE_PROTO')
  , Empty       = function()***REMOVED*** /* empty */ ***REMOVED***
  , PROTOTYPE   = 'prototype';

// Create object with fake `null` prototype: use iframe Object with cleared prototype
var createDict = function()***REMOVED***
  // Thrash, waste and sodomy: IE GC bug
  var iframe = _dereq_('./_dom-create')('iframe')
    , i      = enumBugKeys.length
    , lt     = '<'
    , gt     = '>'
    , iframeDocument;
  iframe.style.display = 'none';
  _dereq_('./_html').appendChild(iframe);
  iframe.src = 'javascript:'; // eslint-disable-line no-script-url
  // createDict = iframe.contentWindow.Object;
  // html.removeChild(iframe);
  iframeDocument = iframe.contentWindow.document;
  iframeDocument.open();
  iframeDocument.write(lt + 'script' + gt + 'document.F=Object' + lt + '/script' + gt);
  iframeDocument.close();
  createDict = iframeDocument.F;
  while(i--)delete createDict[PROTOTYPE][enumBugKeys[i]];
  return createDict();
***REMOVED***;

module.exports = Object.create || function create(O, Properties)***REMOVED***
  var result;
  if(O !== null)***REMOVED***
    Empty[PROTOTYPE] = anObject(O);
    result = new Empty;
    Empty[PROTOTYPE] = null;
    // add "__proto__" for Object.getPrototypeOf polyfill
    result[IE_PROTO] = O;
  ***REMOVED*** else result = createDict();
  return Properties === undefined ? result : dPs(result, Properties);
***REMOVED***;

***REMOVED***,***REMOVED***"./_an-object":16,"./_dom-create":26,"./_enum-bug-keys":27,"./_html":33,"./_object-dps":41,"./_shared-key":48***REMOVED***],40:[function(_dereq_,module,exports)***REMOVED***
var anObject       = _dereq_('./_an-object')
  , IE8_DOM_DEFINE = _dereq_('./_ie8-dom-define')
  , toPrimitive    = _dereq_('./_to-primitive')
  , dP             = Object.defineProperty;

exports.f = _dereq_('./_descriptors') ? Object.defineProperty : function defineProperty(O, P, Attributes)***REMOVED***
  anObject(O);
  P = toPrimitive(P, true);
  anObject(Attributes);
  if(IE8_DOM_DEFINE)try ***REMOVED***
    return dP(O, P, Attributes);
  ***REMOVED*** catch(e)***REMOVED*** /* empty */ ***REMOVED***
  if('get' in Attributes || 'set' in Attributes)throw TypeError('Accessors not supported!');
  if('value' in Attributes)O[P] = Attributes.value;
  return O;
***REMOVED***;
***REMOVED***,***REMOVED***"./_an-object":16,"./_descriptors":25,"./_ie8-dom-define":34,"./_to-primitive":56***REMOVED***],41:[function(_dereq_,module,exports)***REMOVED***
var dP       = _dereq_('./_object-dp')
  , anObject = _dereq_('./_an-object')
  , getKeys  = _dereq_('./_object-keys');

module.exports = _dereq_('./_descriptors') ? Object.defineProperties : function defineProperties(O, Properties)***REMOVED***
  anObject(O);
  var keys   = getKeys(Properties)
    , length = keys.length
    , i = 0
    , P;
  while(length > i)dP.f(O, P = keys[i++], Properties[P]);
  return O;
***REMOVED***;
***REMOVED***,***REMOVED***"./_an-object":16,"./_descriptors":25,"./_object-dp":40,"./_object-keys":44***REMOVED***],42:[function(_dereq_,module,exports)***REMOVED***
exports.f = Object.getOwnPropertySymbols;
***REMOVED***,***REMOVED******REMOVED***],43:[function(_dereq_,module,exports)***REMOVED***
var has          = _dereq_('./_has')
  , toIObject    = _dereq_('./_to-iobject')
  , arrayIndexOf = _dereq_('./_array-includes')(false)
  , IE_PROTO     = _dereq_('./_shared-key')('IE_PROTO');

module.exports = function(object, names)***REMOVED***
  var O      = toIObject(object)
    , i      = 0
    , result = []
    , key;
  for(key in O)if(key != IE_PROTO)has(O, key) && result.push(key);
  // Don't enum bug & hidden keys
  while(names.length > i)if(has(O, key = names[i++]))***REMOVED***
    ~arrayIndexOf(result, key) || result.push(key);
  ***REMOVED***
  return result;
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-includes":17,"./_has":31,"./_shared-key":48,"./_to-iobject":53***REMOVED***],44:[function(_dereq_,module,exports)***REMOVED***
// 19.1.2.14 / 15.2.3.14 Object.keys(O)
var $keys       = _dereq_('./_object-keys-internal')
  , enumBugKeys = _dereq_('./_enum-bug-keys');

module.exports = Object.keys || function keys(O)***REMOVED***
  return $keys(O, enumBugKeys);
***REMOVED***;
***REMOVED***,***REMOVED***"./_enum-bug-keys":27,"./_object-keys-internal":43***REMOVED***],45:[function(_dereq_,module,exports)***REMOVED***
exports.f = ***REMOVED******REMOVED***.propertyIsEnumerable;
***REMOVED***,***REMOVED******REMOVED***],46:[function(_dereq_,module,exports)***REMOVED***
// most Object methods by ES6 should accept primitives
var $export = _dereq_('./_export')
  , core    = _dereq_('./_core')
  , fails   = _dereq_('./_fails');
module.exports = function(KEY, exec)***REMOVED***
  var fn  = (core.Object || ***REMOVED******REMOVED***)[KEY] || Object[KEY]
    , exp = ***REMOVED******REMOVED***;
  exp[KEY] = exec(fn);
  $export($export.S + $export.F * fails(function()***REMOVED*** fn(1); ***REMOVED***), 'Object', exp);
***REMOVED***;
***REMOVED***,***REMOVED***"./_core":22,"./_export":28,"./_fails":29***REMOVED***],47:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(bitmap, value)***REMOVED***
  return ***REMOVED***
    enumerable  : !(bitmap & 1),
    configurable: !(bitmap & 2),
    writable    : !(bitmap & 4),
    value       : value
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],48:[function(_dereq_,module,exports)***REMOVED***
var shared = _dereq_('./_shared')('keys')
  , uid    = _dereq_('./_uid');
module.exports = function(key)***REMOVED***
  return shared[key] || (shared[key] = uid(key));
***REMOVED***;
***REMOVED***,***REMOVED***"./_shared":49,"./_uid":57***REMOVED***],49:[function(_dereq_,module,exports)***REMOVED***
var global = _dereq_('./_global')
  , SHARED = '__core-js_shared__'
  , store  = global[SHARED] || (global[SHARED] = ***REMOVED******REMOVED***);
module.exports = function(key)***REMOVED***
  return store[key] || (store[key] = ***REMOVED******REMOVED***);
***REMOVED***;
***REMOVED***,***REMOVED***"./_global":30***REMOVED***],50:[function(_dereq_,module,exports)***REMOVED***
var fails = _dereq_('./_fails');

module.exports = function(method, arg)***REMOVED***
  return !!method && fails(function()***REMOVED***
    arg ? method.call(null, function()***REMOVED******REMOVED***, 1) : method.call(null);
  ***REMOVED***);
***REMOVED***;
***REMOVED***,***REMOVED***"./_fails":29***REMOVED***],51:[function(_dereq_,module,exports)***REMOVED***
var toInteger = _dereq_('./_to-integer')
  , max       = Math.max
  , min       = Math.min;
module.exports = function(index, length)***REMOVED***
  index = toInteger(index);
  return index < 0 ? max(index + length, 0) : min(index, length);
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-integer":52***REMOVED***],52:[function(_dereq_,module,exports)***REMOVED***
// 7.1.4 ToInteger
var ceil  = Math.ceil
  , floor = Math.floor;
module.exports = function(it)***REMOVED***
  return isNaN(it = +it) ? 0 : (it > 0 ? floor : ceil)(it);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],53:[function(_dereq_,module,exports)***REMOVED***
// to indexed object, toObject with fallback for non-array-like ES3 strings
var IObject = _dereq_('./_iobject')
  , defined = _dereq_('./_defined');
module.exports = function(it)***REMOVED***
  return IObject(defined(it));
***REMOVED***;
***REMOVED***,***REMOVED***"./_defined":24,"./_iobject":35***REMOVED***],54:[function(_dereq_,module,exports)***REMOVED***
// 7.1.15 ToLength
var toInteger = _dereq_('./_to-integer')
  , min       = Math.min;
module.exports = function(it)***REMOVED***
  return it > 0 ? min(toInteger(it), 0x1fffffffffffff) : 0; // pow(2, 53) - 1 == 9007199254740991
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-integer":52***REMOVED***],55:[function(_dereq_,module,exports)***REMOVED***
// 7.1.13 ToObject(argument)
var defined = _dereq_('./_defined');
module.exports = function(it)***REMOVED***
  return Object(defined(it));
***REMOVED***;
***REMOVED***,***REMOVED***"./_defined":24***REMOVED***],56:[function(_dereq_,module,exports)***REMOVED***
// 7.1.1 ToPrimitive(input [, PreferredType])
var isObject = _dereq_('./_is-object');
// instead of the ES6 spec version, we didn't implement @@toPrimitive case
// and the second argument - flag - preferred type is a string
module.exports = function(it, S)***REMOVED***
  if(!isObject(it))return it;
  var fn, val;
  if(S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it)))return val;
  if(typeof (fn = it.valueOf) == 'function' && !isObject(val = fn.call(it)))return val;
  if(!S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it)))return val;
  throw TypeError("Can't convert object to primitive value");
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-object":37***REMOVED***],57:[function(_dereq_,module,exports)***REMOVED***
var id = 0
  , px = Math.random();
module.exports = function(key)***REMOVED***
  return 'Symbol('.concat(key === undefined ? '' : key, ')_', (++id + px).toString(36));
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],58:[function(_dereq_,module,exports)***REMOVED***
var store      = _dereq_('./_shared')('wks')
  , uid        = _dereq_('./_uid')
  , Symbol     = _dereq_('./_global').Symbol
  , USE_SYMBOL = typeof Symbol == 'function';

var $exports = module.exports = function(name)***REMOVED***
  return store[name] || (store[name] =
    USE_SYMBOL && Symbol[name] || (USE_SYMBOL ? Symbol : uid)('Symbol.' + name));
***REMOVED***;

$exports.store = store;
***REMOVED***,***REMOVED***"./_global":30,"./_shared":49,"./_uid":57***REMOVED***],59:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export = _dereq_('./_export')
  , $filter = _dereq_('./_array-methods')(2);

$export($export.P + $export.F * !_dereq_('./_strict-method')([].filter, true), 'Array', ***REMOVED***
  // 22.1.3.7 / 15.4.4.20 Array.prototype.filter(callbackfn [, thisArg])
  filter: function filter(callbackfn /* , thisArg */)***REMOVED***
    return $filter(this, callbackfn, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-methods":18,"./_export":28,"./_strict-method":50***REMOVED***],60:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export  = _dereq_('./_export')
  , $forEach = _dereq_('./_array-methods')(0)
  , STRICT   = _dereq_('./_strict-method')([].forEach, true);

$export($export.P + $export.F * !STRICT, 'Array', ***REMOVED***
  // 22.1.3.10 / 15.4.4.18 Array.prototype.forEach(callbackfn [, thisArg])
  forEach: function forEach(callbackfn /* , thisArg */)***REMOVED***
    return $forEach(this, callbackfn, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-methods":18,"./_export":28,"./_strict-method":50***REMOVED***],61:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export = _dereq_('./_export')
  , $map    = _dereq_('./_array-methods')(1);

$export($export.P + $export.F * !_dereq_('./_strict-method')([].map, true), 'Array', ***REMOVED***
  // 22.1.3.15 / 15.4.4.19 Array.prototype.map(callbackfn [, thisArg])
  map: function map(callbackfn /* , thisArg */)***REMOVED***
    return $map(this, callbackfn, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-methods":18,"./_export":28,"./_strict-method":50***REMOVED***],62:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export = _dereq_('./_export')
  , $some   = _dereq_('./_array-methods')(3);

$export($export.P + $export.F * !_dereq_('./_strict-method')([].some, true), 'Array', ***REMOVED***
  // 22.1.3.23 / 15.4.4.17 Array.prototype.some(callbackfn [, thisArg])
  some: function some(callbackfn /* , thisArg */)***REMOVED***
    return $some(this, callbackfn, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-methods":18,"./_export":28,"./_strict-method":50***REMOVED***],63:[function(_dereq_,module,exports)***REMOVED***
// 19.1.3.1 Object.assign(target, source)
var $export = _dereq_('./_export');

$export($export.S + $export.F, 'Object', ***REMOVED***assign: _dereq_('./_object-assign')***REMOVED***);
***REMOVED***,***REMOVED***"./_export":28,"./_object-assign":38***REMOVED***],64:[function(_dereq_,module,exports)***REMOVED***
var $export = _dereq_('./_export')
// 19.1.2.2 / 15.2.3.5 Object.create(O [, Properties])
$export($export.S, 'Object', ***REMOVED***create: _dereq_('./_object-create')***REMOVED***);
***REMOVED***,***REMOVED***"./_export":28,"./_object-create":39***REMOVED***],65:[function(_dereq_,module,exports)***REMOVED***
var $export = _dereq_('./_export');
// 19.1.2.4 / 15.2.3.6 Object.defineProperty(O, P, Attributes)
$export($export.S + $export.F * !_dereq_('./_descriptors'), 'Object', ***REMOVED***defineProperty: _dereq_('./_object-dp').f***REMOVED***);
***REMOVED***,***REMOVED***"./_descriptors":25,"./_export":28,"./_object-dp":40***REMOVED***],66:[function(_dereq_,module,exports)***REMOVED***
// 19.1.2.14 Object.keys(O)
var toObject = _dereq_('./_to-object')
  , $keys    = _dereq_('./_object-keys');

_dereq_('./_object-sap')('keys', function()***REMOVED***
  return function keys(it)***REMOVED***
    return $keys(toObject(it));
  ***REMOVED***;
***REMOVED***);
***REMOVED***,***REMOVED***"./_object-keys":44,"./_object-sap":46,"./_to-object":55***REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[1])(1)
***REMOVED***);
