// tar/lib/pack.js
// pipe in an fstream, and it'll make a tarball.
// key-value pair argument is global extended header props.

module.exports = Pack

var EntryWriter = require("./entry-writer.js")
  , Stream = require("stream").Stream
  , path = require("path")
  , inherits = require("inherits")
  , GlobalHeaderWriter = require("./global-header-writer.js")
  , collect = require("fstream").collect
  , eof = new Buffer(512)

for (var i = 0; i < 512; i ++) eof[i] = 0

inherits(Pack, Stream)

function Pack (props) ***REMOVED***
  // console.error("-- p ctor")
  var me = this
  if (!(me instanceof Pack)) return new Pack(props)

  if (props) me._noProprietary = props.noProprietary
  else me._noProprietary = false

  me._global = props

  me.readable = true
  me.writable = true
  me._buffer = []
  // console.error("-- -- set current to null in ctor")
  me._currentEntry = null
  me._processing = false

  me._pipeRoot = null
  me.on("pipe", function (src) ***REMOVED***
    if (src.root === me._pipeRoot) return
    me._pipeRoot = src
    src.on("end", function () ***REMOVED***
      me._pipeRoot = null
    ***REMOVED***)
    me.add(src)
  ***REMOVED***)
***REMOVED***

Pack.prototype.addGlobal = function (props) ***REMOVED***
  // console.error("-- p addGlobal")
  if (this._didGlobal) return
  this._didGlobal = true

  var me = this
  GlobalHeaderWriter(props)
    .on("data", function (c) ***REMOVED***
      me.emit("data", c)
    ***REMOVED***)
    .end()
***REMOVED***

Pack.prototype.add = function (stream) ***REMOVED***
  if (this._global && !this._didGlobal) this.addGlobal(this._global)

  if (this._ended) return this.emit("error", new Error("add after end"))

  collect(stream)
  this._buffer.push(stream)
  this._process()
  this._needDrain = this._buffer.length > 0
  return !this._needDrain
***REMOVED***

Pack.prototype.pause = function () ***REMOVED***
  this._paused = true
  if (this._currentEntry) this._currentEntry.pause()
  this.emit("pause")
***REMOVED***

Pack.prototype.resume = function () ***REMOVED***
  this._paused = false
  if (this._currentEntry) this._currentEntry.resume()
  this.emit("resume")
  this._process()
***REMOVED***

Pack.prototype.end = function () ***REMOVED***
  this._ended = true
  this._buffer.push(eof)
  this._process()
***REMOVED***

Pack.prototype._process = function () ***REMOVED***
  var me = this
  if (me._paused || me._processing) ***REMOVED***
    return
  ***REMOVED***

  var entry = me._buffer.shift()

  if (!entry) ***REMOVED***
    if (me._needDrain) ***REMOVED***
      me.emit("drain")
    ***REMOVED***
    return
  ***REMOVED***

  if (entry.ready === false) ***REMOVED***
    // console.error("-- entry is not ready", entry)
    me._buffer.unshift(entry)
    entry.on("ready", function () ***REMOVED***
      // console.error("-- -- ready!", entry)
      me._process()
    ***REMOVED***)
    return
  ***REMOVED***

  me._processing = true

  if (entry === eof) ***REMOVED***
    // need 2 ending null blocks.
    me.emit("data", eof)
    me.emit("data", eof)
    me.emit("end")
    me.emit("close")
    return
  ***REMOVED***

  // Change the path to be relative to the root dir that was
  // added to the tarball.
  //
  // XXX This should be more like how -C works, so you can
  // explicitly set a root dir, and also explicitly set a pathname
  // in the tarball to use.  That way we can skip a lot of extra
  // work when resolving symlinks for bundled dependencies in npm.

  var root = path.dirname((entry.root || entry).path);
  if (me._global && me._global.fromBase && entry.root && entry.root.path) ***REMOVED***
    // user set 'fromBase: true' indicating tar root should be directory itself
    root = entry.root.path;
  ***REMOVED***

  var wprops = ***REMOVED******REMOVED***

  Object.keys(entry.props || ***REMOVED******REMOVED***).forEach(function (k) ***REMOVED***
    wprops[k] = entry.props[k]
  ***REMOVED***)

  if (me._noProprietary) wprops.noProprietary = true

  wprops.path = path.relative(root, entry.path || '')

  // actually not a matter of opinion or taste.
  if (process.platform === "win32") ***REMOVED***
    wprops.path = wprops.path.replace(/\\/g, "/")
  ***REMOVED***

  if (!wprops.type)
    wprops.type = 'Directory'

  switch (wprops.type) ***REMOVED***
    // sockets not supported
    case "Socket":
      return

    case "Directory":
      wprops.path += "/"
      wprops.size = 0
      break

    case "Link":
      var lp = path.resolve(path.dirname(entry.path), entry.linkpath)
      wprops.linkpath = path.relative(root, lp) || "."
      wprops.size = 0
      break

    case "SymbolicLink":
      var lp = path.resolve(path.dirname(entry.path), entry.linkpath)
      wprops.linkpath = path.relative(path.dirname(entry.path), lp) || "."
      wprops.size = 0
      break
  ***REMOVED***

  // console.error("-- new writer", wprops)
  // if (!wprops.type) ***REMOVED***
  //   // console.error("-- no type?", entry.constructor.name, entry)
  // ***REMOVED***

  // console.error("-- -- set current to new writer", wprops.path)
  var writer = me._currentEntry = EntryWriter(wprops)

  writer.parent = me

  // writer.on("end", function () ***REMOVED***
  //   // console.error("-- -- writer end", writer.path)
  // ***REMOVED***)

  writer.on("data", function (c) ***REMOVED***
    me.emit("data", c)
  ***REMOVED***)

  writer.on("header", function () ***REMOVED***
    Buffer.prototype.toJSON = function () ***REMOVED***
      return this.toString().split(/\0/).join(".")
    ***REMOVED***
    // console.error("-- -- writer header %j", writer.props)
    if (writer.props.size === 0) nextEntry()
  ***REMOVED***)
  writer.on("close", nextEntry)

  var ended = false
  function nextEntry () ***REMOVED***
    if (ended) return
    ended = true

    // console.error("-- -- writer close", writer.path)
    // console.error("-- -- set current to null", wprops.path)
    me._currentEntry = null
    me._processing = false
    me._process()
  ***REMOVED***

  writer.on("error", function (er) ***REMOVED***
    // console.error("-- -- writer error", writer.path)
    me.emit("error", er)
  ***REMOVED***)

  // if it's the root, then there's no need to add its entries,
  // or data, since they'll be added directly.
  if (entry === me._pipeRoot) ***REMOVED***
    // console.error("-- is the root, don't auto-add")
    writer.add = null
  ***REMOVED***

  entry.pipe(writer)
***REMOVED***

Pack.prototype.destroy = function () ***REMOVED******REMOVED***
Pack.prototype.write = function () ***REMOVED******REMOVED***
