// har-validator/lib/node4/async.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
***REMOVED***);
exports.validate = validate;
exports.afterRequest = afterRequest;
exports.beforeRequest = beforeRequest;
exports.browser = browser;
exports.cache = cache;
exports.content = content;
exports.cookie = cookie;
exports.creator = creator;
exports.entry = entry;
exports.har = har;
exports.header = header;
exports.log = log;
exports.page = page;
exports.pageTimings = pageTimings;
exports.postData = postData;
exports.query = query;
exports.request = request;
exports.response = response;
exports.timings = timings;

var _harSchema = require('har-schema');

var schemas = _interopRequireWildcard(_harSchema);

var _ajv = require('ajv');

var _ajv2 = _interopRequireDefault(_ajv);

var _error = require('./error');

var _error2 = _interopRequireDefault(_error);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

function _interopRequireWildcard(obj) ***REMOVED*** if (obj && obj.__esModule) ***REMOVED*** return obj; ***REMOVED*** else ***REMOVED*** var newObj = ***REMOVED******REMOVED***; if (obj != null) ***REMOVED*** for (var key in obj) ***REMOVED*** if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; ***REMOVED*** ***REMOVED*** newObj.default = obj; return newObj; ***REMOVED*** ***REMOVED***

var ajv = void 0;

function validate(name) ***REMOVED***
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : ***REMOVED******REMOVED***;
  var next = arguments[2];

  // validator config
  ajv = ajv || new _ajv2.default(***REMOVED***
    allErrors: true,
    schemas: schemas
  ***REMOVED***);

  var validate = ajv.getSchema(name + '.json');

  var valid = validate(data);

  // callback?
  if (typeof next === 'function') ***REMOVED***
    return next(!valid ? new _error2.default(validate.errors) : null, valid);
  ***REMOVED***

  return valid;
***REMOVED***

function afterRequest(data, next) ***REMOVED***
  return validate('afterRequest', data, next);
***REMOVED***

function beforeRequest(data, next) ***REMOVED***
  return validate('beforeRequest', data, next);
***REMOVED***

function browser(data, next) ***REMOVED***
  return validate('browser', data, next);
***REMOVED***

function cache(data, next) ***REMOVED***
  return validate('cache', data, next);
***REMOVED***

function content(data, next) ***REMOVED***
  return validate('content', data, next);
***REMOVED***

function cookie(data, next) ***REMOVED***
  return validate('cookie', data, next);
***REMOVED***

function creator(data, next) ***REMOVED***
  return validate('creator', data, next);
***REMOVED***

function entry(data, next) ***REMOVED***
  return validate('entry', data, next);
***REMOVED***

function har(data, next) ***REMOVED***
  return validate('har', data, next);
***REMOVED***

function header(data, next) ***REMOVED***
  return validate('header', data, next);
***REMOVED***

function log(data, next) ***REMOVED***
  return validate('log', data, next);
***REMOVED***

function page(data, next) ***REMOVED***
  return validate('page', data, next);
***REMOVED***

function pageTimings(data, next) ***REMOVED***
  return validate('pageTimings', data, next);
***REMOVED***

function postData(data, next) ***REMOVED***
  return validate('postData', data, next);
***REMOVED***

function query(data, next) ***REMOVED***
  return validate('query', data, next);
***REMOVED***

function request(data, next) ***REMOVED***
  return validate('request', data, next);
***REMOVED***

function response(data, next) ***REMOVED***
  return validate('response', data, next);
***REMOVED***

function timings(data, next) ***REMOVED***
  return validate('timings', data, next);
***REMOVED***
