// @google-cloud/common/src/index.js
/*!
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @type ***REMOVED***module:common/logger***REMOVED***
 * @private
 */
exports.logger = require('./logger.js');

/**
 * @type ***REMOVED***module:common/operation***REMOVED***
 * @private
 */
exports.Operation = require('./operation.js');

/**
 * @type ***REMOVED***module:common/paginator***REMOVED***
 * @private
 */
exports.paginator = require('./paginator.js');

/**
 * @type ***REMOVED***module:common/service***REMOVED***
 * @private
 */
exports.Service = require('./service.js');

/**
 * @type ***REMOVED***module:common/serviceObject***REMOVED***
 * @private
 */
exports.ServiceObject = require('./service-object.js');

/**
 * @type ***REMOVED***module:common/util***REMOVED***
 * @private
 */
exports.util = require('./util.js');
