// @google-cloud/common/src/service-object.js
/*!
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*!
 * @module common/service-object
 */

'use strict';

var arrify = require('arrify');
var exec = require('methmeth');
var extend = require('extend');
var is = require('is');

/**
 * @type ***REMOVED***module:common/util***REMOVED***
 * @private
 */
var util = require('./util.js');

/**
 * ServiceObject is a base class, meant to be inherited from by a "service
 * object," like a BigQuery dataset or Storage bucket.
 *
 * Most of the time, these objects share common functionality; they can be
 * created or deleted, and you can get or set their metadata.
 *
 * By inheriting from this class, a service object will be extended with these
 * shared behaviors. Note that any method can be overridden when the service
 * object requires specific behavior.
 *
 * @constructor
 * @alias module:common/service-object
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** config - Configuration object.
 * @param ***REMOVED***string***REMOVED*** config.baseUrl - The base URL to make API requests to.
 * @param ***REMOVED***string***REMOVED*** config.createMethod - The method which creates this object.
 * @param ***REMOVED***string=***REMOVED*** config.id - The identifier of the object. For example, the
 *     name of a Storage bucket or Pub/Sub topic.
 * @param ***REMOVED***object=***REMOVED*** config.methods - A map of each method name that should be
 *     inherited.
 * @param ***REMOVED***object***REMOVED*** config.methods[].reqOpts - Default request options for this
 *     particular method. A common use case is when `setMetadata` requires a
 *     `PUT` method to override the default `PATCH`.
 * @param ***REMOVED***object***REMOVED*** config.parent - The parent service instance. For example, an
 *     instance of Storage if the object is Bucket.
 */
function ServiceObject(config) ***REMOVED***
  var self = this;

  this.metadata = ***REMOVED******REMOVED***;

  this.baseUrl = config.baseUrl;
  this.parent = config.parent; // Parent class.
  this.id = config.id; // Name or ID (e.g. dataset ID, bucket name, etc.)
  this.createMethod = config.createMethod;
  this.methods = config.methods || ***REMOVED******REMOVED***;
  this.interceptors = [];
  this.Promise = this.parent.Promise;

  if (config.methods) ***REMOVED***
    var allMethodNames = Object.keys(ServiceObject.prototype);
    allMethodNames
      .filter(function(methodName) ***REMOVED***
        return (// All ServiceObjects need `request`.
          !/^request/.test(methodName) &&
          // The ServiceObject didn't redefine the method.
          self[methodName] === ServiceObject.prototype[methodName] &&
          // This method isn't wanted.
          !config.methods[methodName] );
      ***REMOVED***)
      .forEach(function(methodName) ***REMOVED***
        self[methodName] = undefined;
      ***REMOVED***);
  ***REMOVED***
***REMOVED***

/**
 * Create the object.
 *
 * @param ***REMOVED***object=***REMOVED*** options - Configuration object.
 * @param ***REMOVED***function***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***object***REMOVED*** callback.instance - The instance.
 * @param ***REMOVED***object***REMOVED*** callback.apiResponse - The full API response.
 */
ServiceObject.prototype.create = function(options, callback) ***REMOVED***
  var self = this;
  var args = [this.id];

  if (is.fn(options)) ***REMOVED***
    callback = options;
  ***REMOVED***

  if (is.object(options)) ***REMOVED***
    args.push(options);
  ***REMOVED***

  // Wrap the callback to return *this* instance of the object, not the newly-
  // created one.
  function onCreate(err, instance) ***REMOVED***
    var args = [].slice.call(arguments);

    if (!err) ***REMOVED***
      self.metadata = instance.metadata;
      args[1] = self; // replace the created `instance` with this one.
    ***REMOVED***

    callback.apply(null, args);
  ***REMOVED***

  args.push(onCreate);

  this.createMethod.apply(null, args);
***REMOVED***;

/**
 * Delete the object.
 *
 * @param ***REMOVED***function=***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***object***REMOVED*** callback.apiResponse - The full API response.
 */
ServiceObject.prototype.delete = function(callback) ***REMOVED***
  var methodConfig = this.methods.delete || ***REMOVED******REMOVED***;
  callback = callback || util.noop;

  var reqOpts = extend(
    ***REMOVED***
      method: 'DELETE',
      uri: '',
    ***REMOVED***,
    methodConfig.reqOpts
  );

  // The `request` method may have been overridden to hold any special behavior.
  // Ensure we call the original `request` method.
  ServiceObject.prototype.request.call(this, reqOpts, function(err, resp) ***REMOVED***
    callback(err, resp);
  ***REMOVED***);
***REMOVED***;

/**
 * Check if the object exists.
 *
 * @param ***REMOVED***function***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***boolean***REMOVED*** callback.exists - Whether the object exists or not.
 */
ServiceObject.prototype.exists = function(callback) ***REMOVED***
  this.get(function(err) ***REMOVED***
    if (err) ***REMOVED***
      if (err.code === 404) ***REMOVED***
        callback(null, false);
      ***REMOVED*** else ***REMOVED***
        callback(err);
      ***REMOVED***

      return;
    ***REMOVED***

    callback(null, true);
  ***REMOVED***);
***REMOVED***;

/**
 * Get the object if it exists. Optionally have the object created if an options
 * object is provided with `autoCreate: true`.
 *
 * @param ***REMOVED***object=***REMOVED*** config - The configuration object that will be used to
 *     create the object if necessary.
 * @param ***REMOVED***boolean***REMOVED*** config.autoCreate - Create the object if it doesn't already
 *     exist.
 * @param ***REMOVED***function***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***object***REMOVED*** callback.instance - The instance.
 * @param ***REMOVED***object***REMOVED*** callback.apiResponse - The full API response.
 */
ServiceObject.prototype.get = function(config, callback) ***REMOVED***
  var self = this;

  if (is.fn(config)) ***REMOVED***
    callback = config;
    config = ***REMOVED******REMOVED***;
  ***REMOVED***

  config = config || ***REMOVED******REMOVED***;

  var autoCreate = config.autoCreate && is.fn(this.create);
  delete config.autoCreate;

  function onCreate(err, instance, apiResponse) ***REMOVED***
    if (err) ***REMOVED***
      if (err.code === 409) ***REMOVED***
        self.get(config, callback);
        return;
      ***REMOVED***

      callback(err, null, apiResponse);
      return;
    ***REMOVED***

    callback(null, instance, apiResponse);
  ***REMOVED***

  this.getMetadata(function(err, metadata) ***REMOVED***
    if (err) ***REMOVED***
      if (err.code === 404 && autoCreate) ***REMOVED***
        var args = [];

        if (!is.empty(config)) ***REMOVED***
          args.push(config);
        ***REMOVED***

        args.push(onCreate);

        self.create.apply(self, args);
        return;
      ***REMOVED***

      callback(err, null, metadata);
      return;
    ***REMOVED***

    callback(null, self, metadata);
  ***REMOVED***);
***REMOVED***;

/**
 * Get the metadata of this object.
 *
 * @param ***REMOVED***function***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***object***REMOVED*** callback.metadata - The metadata for this object.
 * @param ***REMOVED***object***REMOVED*** callback.apiResponse - The full API response.
 */
ServiceObject.prototype.getMetadata = function(callback) ***REMOVED***
  var self = this;

  var methodConfig = this.methods.getMetadata || ***REMOVED******REMOVED***;

  var reqOpts = extend(
    ***REMOVED***
      uri: '',
    ***REMOVED***,
    methodConfig.reqOpts
  );

  // The `request` method may have been overridden to hold any special behavior.
  // Ensure we call the original `request` method.
  ServiceObject.prototype.request.call(this, reqOpts, function(err, resp) ***REMOVED***
    if (err) ***REMOVED***
      callback(err, null, resp);
      return;
    ***REMOVED***

    self.metadata = resp;

    callback(null, self.metadata, resp);
  ***REMOVED***);
***REMOVED***;

/**
 * Set the metadata for this object.
 *
 * @param ***REMOVED***object***REMOVED*** metadata - The metadata to set on this object.
 * @param ***REMOVED***function=***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***object***REMOVED*** callback.instance - The instance.
 * @param ***REMOVED***object***REMOVED*** callback.apiResponse - The full API response.
 */
ServiceObject.prototype.setMetadata = function(metadata, callback) ***REMOVED***
  var self = this;

  callback = callback || util.noop;

  var methodConfig = this.methods.setMetadata || ***REMOVED******REMOVED***;

  var reqOpts = extend(
    true,
    ***REMOVED***
      method: 'PATCH',
      uri: '',
      json: metadata,
    ***REMOVED***,
    methodConfig.reqOpts
  );

  // The `request` method may have been overridden to hold any special behavior.
  // Ensure we call the original `request` method.
  ServiceObject.prototype.request.call(this, reqOpts, function(err, resp) ***REMOVED***
    if (err) ***REMOVED***
      callback(err, resp);
      return;
    ***REMOVED***

    self.metadata = resp;

    callback(null, resp);
  ***REMOVED***);
***REMOVED***;

/**
 * Make an authenticated API request.
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** reqOpts - Request options that are passed to `request`.
 * @param ***REMOVED***string***REMOVED*** reqOpts.uri - A URI relative to the baseUrl.
 * @param ***REMOVED***function***REMOVED*** callback - The callback function passed to `request`.
 */
ServiceObject.prototype.request_ = function(reqOpts, callback) ***REMOVED***
  reqOpts = extend(true, ***REMOVED******REMOVED***, reqOpts);

  var isAbsoluteUrl = reqOpts.uri.indexOf('http') === 0;

  var uriComponents = [this.baseUrl, this.id || '', reqOpts.uri];

  if (isAbsoluteUrl) ***REMOVED***
    uriComponents.splice(0, uriComponents.indexOf(reqOpts.uri));
  ***REMOVED***

  reqOpts.uri = uriComponents
    .filter(exec('trim')) // Limit to non-empty strings.
    .map(function(uriComponent) ***REMOVED***
      var trimSlashesRegex = /^\/*|\/*$/g;
      return uriComponent.replace(trimSlashesRegex, '');
    ***REMOVED***)
    .join('/');

  var childInterceptors = arrify(reqOpts.interceptors_);
  var localInterceptors = [].slice.call(this.interceptors);

  reqOpts.interceptors_ = childInterceptors.concat(localInterceptors);

  if (!callback) ***REMOVED***
    return this.parent.requestStream(reqOpts);
  ***REMOVED***

  this.parent.request(reqOpts, callback);
***REMOVED***;

/**
 * Make an authenticated API request.
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** reqOpts - Request options that are passed to `request`.
 * @param ***REMOVED***string***REMOVED*** reqOpts.uri - A URI relative to the baseUrl.
 * @param ***REMOVED***function***REMOVED*** callback - The callback function passed to `request`.
 */
ServiceObject.prototype.request = function(reqOpts, callback) ***REMOVED***
  ServiceObject.prototype.request_.call(this, reqOpts, callback);
***REMOVED***;

/**
 * Make an authenticated API request.
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** reqOpts - Request options that are passed to `request`.
 * @param ***REMOVED***string***REMOVED*** reqOpts.uri - A URI relative to the baseUrl.
 */
ServiceObject.prototype.requestStream = function(reqOpts) ***REMOVED***
  return ServiceObject.prototype.request_.call(this, reqOpts);
***REMOVED***;

/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
util.promisifyAll(ServiceObject);

module.exports = ServiceObject;
