// @firebase/messaging/dist/esm/src/controllers/controller-interface.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';
import ***REMOVED*** ErrorFactory ***REMOVED*** from '@firebase/util';
import Errors from '../models/errors';
import TokenManager from '../models/token-manager';
import NOTIFICATION_PERMISSION from '../models/notification-permission';
var SENDER_ID_OPTION_NAME = 'messagingSenderId';
var ControllerInterface = /** @class */ (function () ***REMOVED***
    /**
     * An interface of the Messaging Service API
     * @param ***REMOVED***!firebase.app.App***REMOVED*** app
     */
    function ControllerInterface(app) ***REMOVED***
        var _this = this;
        this.errorFactory_ = new ErrorFactory('messaging', 'Messaging', Errors.map);
        if (!app.options[SENDER_ID_OPTION_NAME] ||
            typeof app.options[SENDER_ID_OPTION_NAME] !== 'string') ***REMOVED***
            throw this.errorFactory_.create(Errors.codes.BAD_SENDER_ID);
        ***REMOVED***
        this.messagingSenderId_ = app.options[SENDER_ID_OPTION_NAME];
        this.tokenManager_ = new TokenManager();
        this.app = app;
        this.INTERNAL = ***REMOVED******REMOVED***;
        this.INTERNAL.delete = function () ***REMOVED*** return _this.delete; ***REMOVED***;
    ***REMOVED***
    /**
     * @export
     * @return ***REMOVED***Promise<string> | Promise<null>***REMOVED*** Returns a promise that
     * resolves to an FCM token.
     */
    ControllerInterface.prototype.getToken = function () ***REMOVED***
        var _this = this;
        // Check with permissions
        var currentPermission = this.getNotificationPermission_();
        if (currentPermission !== NOTIFICATION_PERMISSION.granted) ***REMOVED***
            if (currentPermission === NOTIFICATION_PERMISSION.denied) ***REMOVED***
                return Promise.reject(this.errorFactory_.create(Errors.codes.NOTIFICATIONS_BLOCKED));
            ***REMOVED***
            // We must wait for permission to be granted
            return Promise.resolve(null);
        ***REMOVED***
        return this.getSWRegistration_().then(function (registration) ***REMOVED***
            return _this.tokenManager_
                .getSavedToken(_this.messagingSenderId_, registration)
                .then(function (token) ***REMOVED***
                if (token) ***REMOVED***
                    return token;
                ***REMOVED***
                return _this.tokenManager_.createToken(_this.messagingSenderId_, registration);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * This method deletes tokens that the token manager looks after and then
     * unregisters the push subscription if it exists.
     * @export
     * @param ***REMOVED***string***REMOVED*** token
     * @return ***REMOVED***Promise<void>***REMOVED***
     */
    ControllerInterface.prototype.deleteToken = function (token) ***REMOVED***
        var _this = this;
        return this.tokenManager_.deleteToken(token).then(function () ***REMOVED***
            return _this.getSWRegistration_()
                .then(function (registration) ***REMOVED***
                if (registration) ***REMOVED***
                    return registration.pushManager.getSubscription();
                ***REMOVED***
            ***REMOVED***)
                .then(function (subscription) ***REMOVED***
                if (subscription) ***REMOVED***
                    return subscription.unsubscribe();
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    ControllerInterface.prototype.getSWRegistration_ = function () ***REMOVED***
        throw this.errorFactory_.create(Errors.codes.SHOULD_BE_INHERITED);
    ***REMOVED***;
    //
    // The following methods should only be available in the window.
    //
    ControllerInterface.prototype.requestPermission = function () ***REMOVED***
        throw this.errorFactory_.create(Errors.codes.AVAILABLE_IN_WINDOW);
    ***REMOVED***;
    /**
     * @export
     * @param ***REMOVED***!ServiceWorkerRegistration***REMOVED*** registration
     */
    ControllerInterface.prototype.useServiceWorker = function (registration) ***REMOVED***
        throw this.errorFactory_.create(Errors.codes.AVAILABLE_IN_WINDOW);
    ***REMOVED***;
    /**
     * @export
     * @param ***REMOVED***!firebase.Observer|function(*)***REMOVED*** nextOrObserver
     * @param ***REMOVED***function(!Error)=***REMOVED*** optError
     * @param ***REMOVED***function()=***REMOVED*** optCompleted
     * @return ***REMOVED***!function()***REMOVED***
     */
    ControllerInterface.prototype.onMessage = function (nextOrObserver, optError, optCompleted) ***REMOVED***
        throw this.errorFactory_.create(Errors.codes.AVAILABLE_IN_WINDOW);
    ***REMOVED***;
    /**
     * @export
     * @param ***REMOVED***!firebase.Observer|function()***REMOVED*** nextOrObserver An observer object
     * or a function triggered on token refresh.
     * @param ***REMOVED***function(!Error)=***REMOVED*** optError Optional A function
     * triggered on token refresh error.
     * @param ***REMOVED***function()=***REMOVED*** optCompleted Optional function triggered when the
     * observer is removed.
     * @return ***REMOVED***!function()***REMOVED*** The unsubscribe function for the observer.
     */
    ControllerInterface.prototype.onTokenRefresh = function (nextOrObserver, optError, optCompleted) ***REMOVED***
        throw this.errorFactory_.create(Errors.codes.AVAILABLE_IN_WINDOW);
    ***REMOVED***;
    //
    // The following methods are used by the service worker only.
    //
    /**
     * @export
     * @param ***REMOVED***function(Object)***REMOVED*** callback
     */
    ControllerInterface.prototype.setBackgroundMessageHandler = function (callback) ***REMOVED***
        throw this.errorFactory_.create(Errors.codes.AVAILABLE_IN_SW);
    ***REMOVED***;
    //
    // The following methods are used by the service themselves and not exposed
    // publicly or not expected to be used by developers.
    //
    /**
     * This method is required to adhere to the Firebase interface.
     * It closes any currently open indexdb database connections.
     */
    ControllerInterface.prototype.delete = function () ***REMOVED***
        return this.tokenManager_.closeDatabase();
    ***REMOVED***;
    /**
     * Returns the current Notification Permission state.
     * @private
     * @return ***REMOVED***string***REMOVED*** The currenct permission state.
     */
    ControllerInterface.prototype.getNotificationPermission_ = function () ***REMOVED***
        return Notification.permission;
    ***REMOVED***;
    /**
     * @protected
     * @returns ***REMOVED***TokenManager***REMOVED***
     */
    ControllerInterface.prototype.getTokenManager = function () ***REMOVED***
        return this.tokenManager_;
    ***REMOVED***;
    return ControllerInterface;
***REMOVED***());
export default ControllerInterface;
