// snapdragon/lib/compiler.js
'use strict';

var use = require('use');
var define = require('define-property');
var debug = require('debug')('snapdragon:compiler');
var utils = require('./utils');

/**
 * Create a new `Compiler` with the given `options`.
 * @param ***REMOVED***Object***REMOVED*** `options`
 */

function Compiler(options, state) ***REMOVED***
  debug('initializing', __filename);
  this.options = utils.extend(***REMOVED***source: 'string'***REMOVED***, options);
  this.state = state || ***REMOVED******REMOVED***;
  this.compilers = ***REMOVED******REMOVED***;
  this.output = '';
  this.set('eos', function(node) ***REMOVED***
    return this.emit(node.val, node);
  ***REMOVED***);
  this.set('noop', function(node) ***REMOVED***
    return this.emit(node.val, node);
  ***REMOVED***);
  this.set('bos', function(node) ***REMOVED***
    return this.emit(node.val, node);
  ***REMOVED***);
  use(this);
***REMOVED***

/**
 * Prototype methods
 */

Compiler.prototype = ***REMOVED***

  /**
   * Throw an error message with details including the cursor position.
   * @param ***REMOVED***String***REMOVED*** `msg` Message to use in the Error.
   */

  error: function(msg, node) ***REMOVED***
    var pos = node.position || ***REMOVED***start: ***REMOVED***column: 0***REMOVED******REMOVED***;
    var message = this.options.source + ' column:' + pos.start.column + ': ' + msg;

    var err = new Error(message);
    err.reason = msg;
    err.column = pos.start.column;
    err.source = this.pattern;

    if (this.options.silent) ***REMOVED***
      this.errors.push(err);
    ***REMOVED*** else ***REMOVED***
      throw err;
    ***REMOVED***
  ***REMOVED***,

  /**
   * Define a non-enumberable property on the `Compiler` instance.
   *
   * ```js
   * compiler.define('foo', 'bar');
   * ```
   * @name .define
   * @param ***REMOVED***String***REMOVED*** `key` propery name
   * @param ***REMOVED***any***REMOVED*** `val` property value
   * @return ***REMOVED***Object***REMOVED*** Returns the Compiler instance for chaining.
   * @api public
   */

  define: function(key, val) ***REMOVED***
    define(this, key, val);
    return this;
  ***REMOVED***,

  /**
   * Emit `node.val`
   */

  emit: function(str, node) ***REMOVED***
    this.output += str;
    return str;
  ***REMOVED***,

  /**
   * Add a compiler `fn` with the given `name`
   */

  set: function(name, fn) ***REMOVED***
    this.compilers[name] = fn;
    return this;
  ***REMOVED***,

  /**
   * Get compiler `name`.
   */

  get: function(name) ***REMOVED***
    return this.compilers[name];
  ***REMOVED***,

  /**
   * Get the previous AST node.
   */

  prev: function(n) ***REMOVED***
    return this.ast.nodes[this.idx - (n || 1)] || ***REMOVED*** type: 'bos', val: '' ***REMOVED***;
  ***REMOVED***,

  /**
   * Get the next AST node.
   */

  next: function(n) ***REMOVED***
    return this.ast.nodes[this.idx + (n || 1)] || ***REMOVED*** type: 'eos', val: '' ***REMOVED***;
  ***REMOVED***,

  /**
   * Visit `node`.
   */

  visit: function(node, nodes, i) ***REMOVED***
    var fn = this.compilers[node.type];
    this.idx = i;

    if (typeof fn !== 'function') ***REMOVED***
      throw this.error('compiler "' + node.type + '" is not registered', node);
    ***REMOVED***
    return fn.call(this, node, nodes, i);
  ***REMOVED***,

  /**
   * Map visit over array of `nodes`.
   */

  mapVisit: function(nodes) ***REMOVED***
    if (!Array.isArray(nodes)) ***REMOVED***
      throw new TypeError('expected an array');
    ***REMOVED***
    var len = nodes.length;
    var idx = -1;
    while (++idx < len) ***REMOVED***
      this.visit(nodes[idx], nodes, idx);
    ***REMOVED***
    return this;
  ***REMOVED***,

  /**
   * Compile `ast`.
   */

  compile: function(ast, options) ***REMOVED***
    var opts = utils.extend(***REMOVED******REMOVED***, this.options, options);
    this.ast = ast;
    this.parsingErrors = this.ast.errors;
    this.output = '';

    // source map support
    if (opts.sourcemap) ***REMOVED***
      var sourcemaps = require('./source-maps');
      sourcemaps(this);
      this.mapVisit(this.ast.nodes);
      this.applySourceMaps();
      this.map = opts.sourcemap === 'generator' ? this.map : this.map.toJSON();
      return this;
    ***REMOVED***

    this.mapVisit(this.ast.nodes);
    return this;
  ***REMOVED***
***REMOVED***;

/**
 * Expose `Compiler`
 */

module.exports = Compiler;
