// snapdragon/lib/source-maps.js
'use strict';

var fs = require('fs');
var path = require('path');
var define = require('define-property');
var utils = require('./utils');

/**
 * Expose `mixin()`.
 * This code is based on `source-maps-support.js` in reworkcss/css
 * https://github.com/reworkcss/css/blob/master/lib/stringify/source-map-support.js
 * Copyright (c) 2012 TJ Holowaychuk <tj@vision-media.ca>
 */

module.exports = mixin;

/**
 * Mixin source map support into `compiler`.
 *
 * @param ***REMOVED***Object***REMOVED*** `compiler`
 * @api public
 */

function mixin(compiler) ***REMOVED***
  define(compiler, '_comment', compiler.comment);
  compiler.map = new utils.SourceMap.SourceMapGenerator();
  compiler.position = ***REMOVED*** line: 1, column: 1 ***REMOVED***;
  compiler.content = ***REMOVED******REMOVED***;
  compiler.files = ***REMOVED******REMOVED***;

  for (var key in exports) ***REMOVED***
    define(compiler, key, exports[key]);
  ***REMOVED***
***REMOVED***

/**
 * Update position.
 *
 * @param ***REMOVED***String***REMOVED*** str
 */

exports.updatePosition = function(str) ***REMOVED***
  var lines = str.match(/\n/g);
  if (lines) this.position.line += lines.length;
  var i = str.lastIndexOf('\n');
  this.position.column = ~i ? str.length - i : this.position.column + str.length;
***REMOVED***;

/**
 * Emit `str` with `position`.
 *
 * @param ***REMOVED***String***REMOVED*** str
 * @param ***REMOVED***Object***REMOVED*** [pos]
 * @return ***REMOVED***String***REMOVED***
 */

exports.emit = function(str, node) ***REMOVED***
  var position = node.position || ***REMOVED******REMOVED***;
  var source = position.source;
  if (source) ***REMOVED***
    if (position.filepath) ***REMOVED***
      source = utils.unixify(position.filepath);
    ***REMOVED***

    this.map.addMapping(***REMOVED***
      source: source,
      generated: ***REMOVED***
        line: this.position.line,
        column: Math.max(this.position.column - 1, 0)
      ***REMOVED***,
      original: ***REMOVED***
        line: position.start.line,
        column: position.start.column - 1
      ***REMOVED***
    ***REMOVED***);

    if (position.content) ***REMOVED***
      this.addContent(source, position);
    ***REMOVED***
    if (position.filepath) ***REMOVED***
      this.addFile(source, position);
    ***REMOVED***

    this.updatePosition(str);
    this.output += str;
  ***REMOVED***
  return str;
***REMOVED***;

/**
 * Adds a file to the source map output if it has not already been added
 * @param ***REMOVED***String***REMOVED*** `file`
 * @param ***REMOVED***Object***REMOVED*** `pos`
 */

exports.addFile = function(file, position) ***REMOVED***
  if (typeof position.content !== 'string') return;
  if (Object.prototype.hasOwnProperty.call(this.files, file)) return;
  this.files[file] = position.content;
***REMOVED***;

/**
 * Adds a content source to the source map output if it has not already been added
 * @param ***REMOVED***String***REMOVED*** `source`
 * @param ***REMOVED***Object***REMOVED*** `position`
 */

exports.addContent = function(source, position) ***REMOVED***
  if (typeof position.content !== 'string') return;
  if (Object.prototype.hasOwnProperty.call(this.content, source)) return;
  this.map.setSourceContent(source, position.content);
***REMOVED***;

/**
 * Applies any original source maps to the output and embeds the source file
 * contents in the source map.
 */

exports.applySourceMaps = function() ***REMOVED***
  Object.keys(this.files).forEach(function(file) ***REMOVED***
    var content = this.files[file];
    this.map.setSourceContent(file, content);

    if (this.options.inputSourcemaps === true) ***REMOVED***
      var originalMap = utils.sourceMapResolve.resolveSync(content, file, fs.readFileSync);
      if (originalMap) ***REMOVED***
        var map = new utils.SourceMap.SourceMapConsumer(originalMap.map);
        var relativeTo = originalMap.sourcesRelativeTo;
        this.map.applySourceMap(map, file, utils.unixify(path.dirname(relativeTo)));
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***, this);
***REMOVED***;

/**
 * Process comments, drops sourceMap comments.
 * @param ***REMOVED***Object***REMOVED*** node
 */

exports.comment = function(node) ***REMOVED***
  if (/^# sourceMappingURL=/.test(node.comment)) ***REMOVED***
    return this.emit('', node.position);
  ***REMOVED***
  return this._comment(node);
***REMOVED***;
