// snapdragon/lib/utils.js
'use strict';

/**
 * Module dependencies
 */

exports.extend = require('extend-shallow');
exports.SourceMap = require('source-map');
exports.sourceMapResolve = require('source-map-resolve');

/**
 * Convert backslash in the given string to forward slashes
 */

exports.unixify = function(fp) ***REMOVED***
  return fp.split(/\\+/).join('/');
***REMOVED***;

/**
 * Return true if `val` is a non-empty string
 *
 * @param ***REMOVED***String***REMOVED*** `str`
 * @return ***REMOVED***Boolean***REMOVED***
 */

exports.isString = function(str) ***REMOVED***
  return str && typeof str === 'string';
***REMOVED***;

/**
 * Cast `val` to an array
 * @return ***REMOVED***Array***REMOVED***
 */

exports.arrayify = function(val) ***REMOVED***
  if (typeof val === 'string') return [val];
  return val ? (Array.isArray(val) ? val : [val]) : [];
***REMOVED***;

/**
 * Get the last `n` element from the given `array`
 * @param ***REMOVED***Array***REMOVED*** `array`
 * @return ***REMOVED*******REMOVED***
 */

exports.last = function(arr, n) ***REMOVED***
  return arr[arr.length - (n || 1)];
***REMOVED***;
