// firebase-admin/lib/messaging/messaging-api-request.js
/*! firebase-admin v5.8.2 */
"use strict";
/*!
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var error_1 = require("../utils/error");
var api_request_1 = require("../utils/api-request");
var error_2 = require("../utils/error");
var validator = require("../utils/validator");
// FCM backend constants
var FIREBASE_MESSAGING_PORT = 443;
var FIREBASE_MESSAGING_TIMEOUT = 10000;
var FIREBASE_MESSAGING_HTTP_METHOD = 'POST';
var FIREBASE_MESSAGING_HEADERS = ***REMOVED***
    'Content-Type': 'application/json',
    'Sdk-Version': 'Node/Admin/5.8.2',
    access_token_auth: 'true',
***REMOVED***;
/**
 * Class that provides a mechanism to send requests to the Firebase Cloud Messaging backend.
 */
var FirebaseMessagingRequestHandler = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***FirebaseApp***REMOVED*** app The app used to fetch access tokens to sign API requests.
     * @constructor
     */
    function FirebaseMessagingRequestHandler(app) ***REMOVED***
        this.signedApiRequestHandler = new api_request_1.SignedApiRequestHandler(app);
    ***REMOVED***
    /**
     * @param ***REMOVED***Object***REMOVED*** response The response to check for errors.
     * @return ***REMOVED***string|null***REMOVED*** The error code if present; null otherwise.
     */
    FirebaseMessagingRequestHandler.getErrorCode = function (response) ***REMOVED***
        if (validator.isNonNullObject(response) && 'error' in response) ***REMOVED***
            if (typeof response.error === 'string') ***REMOVED***
                return response.error;
            ***REMOVED***
            else ***REMOVED***
                return response.error.message;
            ***REMOVED***
        ***REMOVED***
        return null;
    ***REMOVED***;
    /**
     * Invokes the request handler with the provided request data.
     *
     * @param ***REMOVED***string***REMOVED*** host The host to which to send the request.
     * @param ***REMOVED***string***REMOVED*** path The path to which to send the request.
     * @param ***REMOVED***Object***REMOVED*** requestData The request data.
     * @return ***REMOVED***Promise<Object>***REMOVED*** A promise that resolves with the response.
     */
    FirebaseMessagingRequestHandler.prototype.invokeRequestHandler = function (host, path, requestData) ***REMOVED***
        return this.signedApiRequestHandler.sendRequest(host, FIREBASE_MESSAGING_PORT, path, FIREBASE_MESSAGING_HTTP_METHOD, requestData, FIREBASE_MESSAGING_HEADERS, FIREBASE_MESSAGING_TIMEOUT).then(function (response) ***REMOVED***
            // Send non-JSON responses to the catch() below where they will be treated as errors.
            if (typeof response === 'string') ***REMOVED***
                return Promise.reject(***REMOVED***
                    error: response,
                    statusCode: 200,
                ***REMOVED***);
            ***REMOVED***
            // Check for backend errors in the response.
            var errorCode = FirebaseMessagingRequestHandler.getErrorCode(response);
            if (errorCode) ***REMOVED***
                return Promise.reject(***REMOVED***
                    error: response,
                    statusCode: 200,
                ***REMOVED***);
            ***REMOVED***
            // Return entire response.
            return response;
        ***REMOVED***)
            .catch(function (response) ***REMOVED***
            // Re-throw the error if it already has the proper format.
            if (response instanceof error_1.FirebaseError) ***REMOVED***
                throw response;
            ***REMOVED***
            else if (response.error instanceof error_1.FirebaseError) ***REMOVED***
                throw response.error;
            ***REMOVED***
            // Add special handling for non-JSON responses.
            if (typeof response.error === 'string') ***REMOVED***
                var error = void 0;
                switch (response.statusCode) ***REMOVED***
                    case 400:
                        error = error_2.MessagingClientErrorCode.INVALID_ARGUMENT;
                        break;
                    case 401:
                    case 403:
                        error = error_2.MessagingClientErrorCode.AUTHENTICATION_ERROR;
                        break;
                    case 500:
                        error = error_2.MessagingClientErrorCode.INTERNAL_ERROR;
                        break;
                    case 503:
                        error = error_2.MessagingClientErrorCode.SERVER_UNAVAILABLE;
                        break;
                    default:
                        // Treat non-JSON responses with unexpected status codes as unknown errors.
                        error = error_2.MessagingClientErrorCode.UNKNOWN_ERROR;
                ***REMOVED***
                throw new error_2.FirebaseMessagingError(***REMOVED***
                    code: error.code,
                    message: error.message + " Raw server response: \"" + response.error + "\". Status code: " +
                        (response.statusCode + "."),
                ***REMOVED***);
            ***REMOVED***
            // For JSON responses, map the server response to a client-side error.
            var errorCode = FirebaseMessagingRequestHandler.getErrorCode(response.error);
            throw error_2.FirebaseMessagingError.fromServerError(errorCode, /* message */ undefined, response.error);
        ***REMOVED***);
    ***REMOVED***;
    return FirebaseMessagingRequestHandler;
***REMOVED***());
exports.FirebaseMessagingRequestHandler = FirebaseMessagingRequestHandler;
