// firebase-admin/lib/messaging/messaging.js
/*! firebase-admin v5.8.2 */
"use strict";
/*!
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var deep_copy_1 = require("../utils/deep-copy");
var messaging_api_request_1 = require("./messaging-api-request");
var error_1 = require("../utils/error");
var utils = require("../utils");
var validator = require("../utils/validator");
// FCM endpoints
var FCM_SEND_HOST = 'fcm.googleapis.com';
var FCM_SEND_PATH = '/fcm/send';
var FCM_TOPIC_MANAGEMENT_HOST = 'iid.googleapis.com';
var FCM_TOPIC_MANAGEMENT_ADD_PATH = '/iid/v1:batchAdd';
var FCM_TOPIC_MANAGEMENT_REMOVE_PATH = '/iid/v1:batchRemove';
// Key renames for the messaging notification payload object.
var CAMELCASED_NOTIFICATION_PAYLOAD_KEYS_MAP = ***REMOVED***
    bodyLocArgs: 'body_loc_args',
    bodyLocKey: 'body_loc_key',
    clickAction: 'click_action',
    titleLocArgs: 'title_loc_args',
    titleLocKey: 'title_loc_key',
***REMOVED***;
// Key renames for the messaging options object.
var CAMELCASE_OPTIONS_KEYS_MAP = ***REMOVED***
    dryRun: 'dry_run',
    timeToLive: 'time_to_live',
    collapseKey: 'collapse_key',
    mutableContent: 'mutable_content',
    contentAvailable: 'content_available',
    restrictedPackageName: 'restricted_package_name',
***REMOVED***;
// Key renames for the MessagingDeviceResult object.
var MESSAGING_DEVICE_RESULT_KEYS_MAP = ***REMOVED***
    message_id: 'messageId',
    registration_id: 'canonicalRegistrationToken',
***REMOVED***;
// Key renames for the MessagingDevicesResponse object.
var MESSAGING_DEVICES_RESPONSE_KEYS_MAP = ***REMOVED***
    canonical_ids: 'canonicalRegistrationTokenCount',
    failure: 'failureCount',
    success: 'successCount',
    multicast_id: 'multicastId',
***REMOVED***;
// Key renames for the MessagingDeviceGroupResponse object.
var MESSAGING_DEVICE_GROUP_RESPONSE_KEYS_MAP = ***REMOVED***
    success: 'successCount',
    failure: 'failureCount',
    failed_registration_ids: 'failedRegistrationTokens',
***REMOVED***;
// Key renames for the MessagingTopicResponse object.
var MESSAGING_TOPIC_RESPONSE_KEYS_MAP = ***REMOVED***
    message_id: 'messageId',
***REMOVED***;
// Key renames for the MessagingConditionResponse object.
var MESSAGING_CONDITION_RESPONSE_KEYS_MAP = ***REMOVED***
    message_id: 'messageId',
***REMOVED***;
// Keys which are not allowed in the messaging data payload object.
exports.BLACKLISTED_DATA_PAYLOAD_KEYS = ['from'];
// Keys which are not allowed in the messaging options object.
exports.BLACKLISTED_OPTIONS_KEYS = [
    'condition', 'data', 'notification', 'registrationIds', 'registration_ids', 'to',
];
/**
 * Maps a raw FCM server response to a MessagingDevicesResponse object.
 *
 * @param ***REMOVED***Object***REMOVED*** response The raw FCM server response to map.
 *
 * @return ***REMOVED***MessagingDeviceGroupResponse***REMOVED*** The mapped MessagingDevicesResponse object.
 */
function mapRawResponseToDevicesResponse(response) ***REMOVED***
    // Rename properties on the server response
    utils.renameProperties(response, MESSAGING_DEVICES_RESPONSE_KEYS_MAP);
    if ('results' in response) ***REMOVED***
        response.results.forEach(function (messagingDeviceResult) ***REMOVED***
            utils.renameProperties(messagingDeviceResult, MESSAGING_DEVICE_RESULT_KEYS_MAP);
            // Map the FCM server's error strings to actual error objects.
            if ('error' in messagingDeviceResult) ***REMOVED***
                var newError = error_1.FirebaseMessagingError.fromServerError(messagingDeviceResult.error, /* message */ undefined, messagingDeviceResult.error);
                messagingDeviceResult.error = newError;
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***
    return response;
***REMOVED***
/**
 * Maps a raw FCM server response to a MessagingDeviceGroupResponse object.
 *
 * @param ***REMOVED***Object***REMOVED*** response The raw FCM server response to map.
 *
 * @return ***REMOVED***MessagingDeviceGroupResponse***REMOVED*** The mapped MessagingDeviceGroupResponse object.
 */
function mapRawResponseToDeviceGroupResponse(response) ***REMOVED***
    // Rename properties on the server response
    utils.renameProperties(response, MESSAGING_DEVICE_GROUP_RESPONSE_KEYS_MAP);
    // Add the 'failedRegistrationTokens' property if it does not exist on the response, which
    // it won't when the 'failureCount' property has a value of 0)
    response.failedRegistrationTokens = response.failedRegistrationTokens || [];
    return response;
***REMOVED***
/**
 * Maps a raw FCM server response to a MessagingTopicManagementResponse object.
 *
 * @param ***REMOVED***Object***REMOVED*** response The raw FCM server response to map.
 *
 * @return ***REMOVED***MessagingTopicManagementResponse***REMOVED*** The mapped MessagingTopicManagementResponse object.
 */
function mapRawResponseToTopicManagementResponse(response) ***REMOVED***
    // Add the success and failure counts.
    response.successCount = 0;
    response.failureCount = 0;
    var errors = [];
    if ('results' in response) ***REMOVED***
        response.results.forEach(function (tokenManagementResult, index) ***REMOVED***
            // Map the FCM server's error strings to actual error objects.
            if ('error' in tokenManagementResult) ***REMOVED***
                response.failureCount += 1;
                var newError = error_1.FirebaseMessagingError.fromServerError(tokenManagementResult.error, /* message */ undefined, tokenManagementResult.error);
                errors.push(***REMOVED***
                    index: index,
                    error: newError,
                ***REMOVED***);
            ***REMOVED***
            else ***REMOVED***
                response.successCount += 1;
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***
    delete response.results;
    response.errors = errors;
    return response;
***REMOVED***
/**
 * Internals of a Messaging instance.
 */
var MessagingInternals = /** @class */ (function () ***REMOVED***
    function MessagingInternals() ***REMOVED***
    ***REMOVED***
    /**
     * Deletes the service and its associated resources.
     *
     * @return ***REMOVED***Promise<()>***REMOVED*** An empty Promise that will be fulfilled when the service is deleted.
     */
    MessagingInternals.prototype.delete = function () ***REMOVED***
        // There are no resources to clean up.
        return Promise.resolve(undefined);
    ***REMOVED***;
    return MessagingInternals;
***REMOVED***());
/**
 * Messaging service bound to the provided app.
 */
var Messaging = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***Object***REMOVED*** app The app for this Messaging service.
     * @constructor
     */
    function Messaging(app) ***REMOVED***
        this.INTERNAL = new MessagingInternals();
        if (!validator.isNonNullObject(app) || !('options' in app)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_ARGUMENT, 'First argument passed to admin.messaging() must be a valid Firebase app instance.');
        ***REMOVED***
        this.appInternal = app;
        // Initialize messaging request handler with the app.
        this.messagingRequestHandler = new messaging_api_request_1.FirebaseMessagingRequestHandler(app);
    ***REMOVED***
    Object.defineProperty(Messaging.prototype, "app", ***REMOVED***
        /**
         * Returns the app associated with this Messaging instance.
         *
         * @return ***REMOVED***FirebaseApp***REMOVED*** The app associated with this Messaging instance.
         */
        get: function () ***REMOVED***
            return this.appInternal;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    /**
     * Sends an FCM message to a single device or an array of devices.
     *
     * @param ***REMOVED***string|string[]***REMOVED*** registrationTokenOrTokens The registration token or an array of
     *     registration tokens for the device(s) to which to send the message.
     * @param ***REMOVED***MessagingPayload***REMOVED*** payload The message payload.
     * @param ***REMOVED***MessagingOptions***REMOVED*** [options = ***REMOVED******REMOVED***] Optional options to alter the message.
     *
     * @return ***REMOVED***Promise<MessagingDevicesResponse|MessagingDeviceGroupResponse>***REMOVED*** A Promise fulfilled
     *     with the server's response after the message has been sent.
     */
    Messaging.prototype.sendToDevice = function (registrationTokenOrTokens, payload, options) ***REMOVED***
        var _this = this;
        if (options === void 0) ***REMOVED*** options = ***REMOVED******REMOVED***; ***REMOVED***
        // Validate the input argument types. Since these are common developer errors when getting
        // started, throw an error instead of returning a rejected promise.
        this.validateRegistrationTokensType(registrationTokenOrTokens, 'sendToDevice', error_1.MessagingClientErrorCode.INVALID_RECIPIENT);
        this.validateMessagingPayloadAndOptionsTypes(payload, options);
        return Promise.resolve()
            .then(function () ***REMOVED***
            // Validate the contents of the input arguments. Because we are now in a promise, any thrown
            // error will cause this method to return a rejected promise.
            _this.validateRegistrationTokens(registrationTokenOrTokens, 'sendToDevice', error_1.MessagingClientErrorCode.INVALID_RECIPIENT);
            var payloadCopy = _this.validateMessagingPayload(payload);
            var optionsCopy = _this.validateMessagingOptions(options);
            var request = deep_copy_1.deepCopy(payloadCopy);
            deep_copy_1.deepExtend(request, optionsCopy);
            if (validator.isString(registrationTokenOrTokens)) ***REMOVED***
                request.to = registrationTokenOrTokens;
            ***REMOVED***
            else ***REMOVED***
                request.registration_ids = registrationTokenOrTokens;
            ***REMOVED***
            return _this.messagingRequestHandler.invokeRequestHandler(FCM_SEND_HOST, FCM_SEND_PATH, request);
        ***REMOVED***)
            .then(function (response) ***REMOVED***
            // The sendToDevice() and sendToDeviceGroup() methods both set the `to` query parameter in
            // the underlying FCM request. If the provided registration token argument is actually a
            // valid notification key, the response from the FCM server will be a device group response.
            // If that is the case, we map the response to a MessagingDeviceGroupResponse.
            // See b/35394951 for more context.
            if ('multicast_id' in response) ***REMOVED***
                return mapRawResponseToDevicesResponse(response);
            ***REMOVED***
            else ***REMOVED***
                return mapRawResponseToDeviceGroupResponse(response);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Sends an FCM message to a device group.
     *
     * @param ***REMOVED***string***REMOVED*** notificationKey The notification key representing the device group to which to
     *     send the message.
     * @param ***REMOVED***MessagingPayload***REMOVED*** payload The message payload.
     * @param ***REMOVED***MessagingOptions***REMOVED*** [options = ***REMOVED******REMOVED***] Optional options to alter the message.
     *
     * @return ***REMOVED***Promise<MessagingDeviceGroupResponse|MessagingDevicesResponse>***REMOVED*** A Promise fulfilled
     *     with the server's response after the message has been sent.
     */
    Messaging.prototype.sendToDeviceGroup = function (notificationKey, payload, options) ***REMOVED***
        var _this = this;
        if (options === void 0) ***REMOVED*** options = ***REMOVED******REMOVED***; ***REMOVED***
        if (!validator.isNonEmptyString(notificationKey)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_RECIPIENT, 'Notification key provided to sendToDeviceGroup() must be a non-empty string.');
        ***REMOVED***
        else if (notificationKey.indexOf(':') !== -1) ***REMOVED***
            // It is possible the developer provides a registration token instead of a notification key
            // to this method. We can detect some of those cases by checking to see if the string contains
            // a colon. Not all registration tokens will contain a colon (only newer ones will), but no
            // notification keys will contain a colon, so we can use it as a rough heuristic.
            // See b/35394951 for more context.
            return Promise.reject(new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_RECIPIENT, 'Notification key provided to sendToDeviceGroup() has the format of a registration token. ' +
                'You should use sendToDevice() instead.'));
        ***REMOVED***
        // Validate the types of the payload and options arguments. Since these are common developer
        // errors, throw an error instead of returning a rejected promise.
        this.validateMessagingPayloadAndOptionsTypes(payload, options);
        return Promise.resolve()
            .then(function () ***REMOVED***
            // Validate the contents of the payload and options objects. Because we are now in a
            // promise, any thrown error will cause this method to return a rejected promise.
            var payloadCopy = _this.validateMessagingPayload(payload);
            var optionsCopy = _this.validateMessagingOptions(options);
            var request = deep_copy_1.deepCopy(payloadCopy);
            deep_copy_1.deepExtend(request, optionsCopy);
            request.to = notificationKey;
            return _this.messagingRequestHandler.invokeRequestHandler(FCM_SEND_HOST, FCM_SEND_PATH, request);
        ***REMOVED***)
            .then(function (response) ***REMOVED***
            // The sendToDevice() and sendToDeviceGroup() methods both set the `to` query parameter in
            // the underlying FCM request. If the provided notification key argument has an invalid
            // format (that is, it is either a registration token or some random string), the response
            // from the FCM server will default to a devices response (which we detect by looking for
            // the `multicast_id` property). If that is the case, we either throw an error saying the
            // provided notification key is invalid (if the message failed to send) or map the response
            // to a MessagingDevicesResponse (if the message succeeded).
            // See b/35394951 for more context.
            if ('multicast_id' in response) ***REMOVED***
                if (response.success === 0) ***REMOVED***
                    throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_RECIPIENT, 'Notification key provided to sendToDeviceGroup() is invalid.');
                ***REMOVED***
                else ***REMOVED***
                    return mapRawResponseToDevicesResponse(response);
                ***REMOVED***
            ***REMOVED***
            return mapRawResponseToDeviceGroupResponse(response);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Sends an FCM message to a topic.
     *
     * @param ***REMOVED***string***REMOVED*** topic The name of the topic to which to send the message.
     * @param ***REMOVED***MessagingPayload***REMOVED*** payload The message payload.
     * @param ***REMOVED***MessagingOptions***REMOVED*** [options = ***REMOVED******REMOVED***] Optional options to alter the message.
     *
     * @return ***REMOVED***Promise<MessagingTopicResponse>***REMOVED*** A Promise fulfilled with the server's response after
     *     the message has been sent.
     */
    Messaging.prototype.sendToTopic = function (topic, payload, options) ***REMOVED***
        var _this = this;
        if (options === void 0) ***REMOVED*** options = ***REMOVED******REMOVED***; ***REMOVED***
        // Validate the input argument types. Since these are common developer errors when getting
        // started, throw an error instead of returning a rejected promise.
        this.validateTopicType(topic, 'sendToTopic', error_1.MessagingClientErrorCode.INVALID_RECIPIENT);
        this.validateMessagingPayloadAndOptionsTypes(payload, options);
        // Prepend the topic with /topics/ if necessary.
        topic = this.normalizeTopic(topic);
        return Promise.resolve()
            .then(function () ***REMOVED***
            // Validate the contents of the payload and options objects. Because we are now in a
            // promise, any thrown error will cause this method to return a rejected promise.
            var payloadCopy = _this.validateMessagingPayload(payload);
            var optionsCopy = _this.validateMessagingOptions(options);
            _this.validateTopic(topic, 'sendToTopic', error_1.MessagingClientErrorCode.INVALID_RECIPIENT);
            var request = deep_copy_1.deepCopy(payloadCopy);
            deep_copy_1.deepExtend(request, optionsCopy);
            request.to = topic;
            return _this.messagingRequestHandler.invokeRequestHandler(FCM_SEND_HOST, FCM_SEND_PATH, request);
        ***REMOVED***)
            .then(function (response) ***REMOVED***
            // Rename properties on the server response
            utils.renameProperties(response, MESSAGING_TOPIC_RESPONSE_KEYS_MAP);
            return response;
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Sends an FCM message to a condition.
     *
     * @param ***REMOVED***string***REMOVED*** condition The condition to which to send the message.
     * @param ***REMOVED***MessagingPayload***REMOVED*** payload The message payload.
     * @param ***REMOVED***MessagingOptions***REMOVED*** [options = ***REMOVED******REMOVED***] Optional options to alter the message.
     *
     * @return ***REMOVED***Promise<MessagingConditionResponse>***REMOVED*** A Promise fulfilled with the server's response
     *     after the message has been sent.
     */
    Messaging.prototype.sendToCondition = function (condition, payload, options) ***REMOVED***
        var _this = this;
        if (options === void 0) ***REMOVED*** options = ***REMOVED******REMOVED***; ***REMOVED***
        if (!validator.isNonEmptyString(condition)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_RECIPIENT, 'Condition provided to sendToCondition() must be a non-empty string.');
        ***REMOVED***
        // Validate the types of the payload and options arguments. Since these are common developer
        // errors, throw an error instead of returning a rejected promise.
        this.validateMessagingPayloadAndOptionsTypes(payload, options);
        // The FCM server rejects conditions which are surrounded in single quotes. When the condition
        // is stringified over the wire, double quotes in it get converted to \" which the FCM server
        // does not properly handle. We can get around this by replacing internal double quotes with
        // single quotes.
        condition = condition.replace(/"/g, '\'');
        return Promise.resolve()
            .then(function () ***REMOVED***
            // Validate the contents of the payload and options objects. Because we are now in a
            // promise, any thrown error will cause this method to return a rejected promise.
            var payloadCopy = _this.validateMessagingPayload(payload);
            var optionsCopy = _this.validateMessagingOptions(options);
            var request = deep_copy_1.deepCopy(payloadCopy);
            deep_copy_1.deepExtend(request, optionsCopy);
            request.condition = condition;
            return _this.messagingRequestHandler.invokeRequestHandler(FCM_SEND_HOST, FCM_SEND_PATH, request);
        ***REMOVED***)
            .then(function (response) ***REMOVED***
            // Rename properties on the server response
            utils.renameProperties(response, MESSAGING_CONDITION_RESPONSE_KEYS_MAP);
            return response;
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Subscribes a single device or an array of devices to a topic.
     *
     * @param ***REMOVED***string|string[]***REMOVED*** registrationTokenOrTokens The registration token or an array of
     *     registration tokens to subscribe to the topic.
     * @param ***REMOVED***string***REMOVED*** topic The topic to which to subscribe.
     *
     * @return ***REMOVED***Promise<MessagingTopicManagementResponse>***REMOVED*** A Promise fulfilled with the parsed FCM
     *   server response.
     */
    Messaging.prototype.subscribeToTopic = function (registrationTokenOrTokens, topic) ***REMOVED***
        return this.sendTopicManagementRequest(registrationTokenOrTokens, topic, 'subscribeToTopic', FCM_TOPIC_MANAGEMENT_ADD_PATH);
    ***REMOVED***;
    /**
     * Unsubscribes a single device or an array of devices from a topic.
     *
     * @param ***REMOVED***string|string[]***REMOVED*** registrationTokenOrTokens The registration token or an array of
     *     registration tokens to unsubscribe from the topic.
     * @param ***REMOVED***string***REMOVED*** topic The topic to which to subscribe.
     *
     * @return ***REMOVED***Promise<MessagingTopicManagementResponse>***REMOVED*** A Promise fulfilled with the parsed FCM
     *   server response.
     */
    Messaging.prototype.unsubscribeFromTopic = function (registrationTokenOrTokens, topic) ***REMOVED***
        return this.sendTopicManagementRequest(registrationTokenOrTokens, topic, 'unsubscribeFromTopic', FCM_TOPIC_MANAGEMENT_REMOVE_PATH);
    ***REMOVED***;
    /**
     * Helper method which sends and handles topic subscription management requests.
     *
     * @param ***REMOVED***string|string[]***REMOVED*** registrationTokenOrTokens The registration token or an array of
     *     registration tokens to unsubscribe from the topic.
     * @param ***REMOVED***string***REMOVED*** topic The topic to which to subscribe.
     * @param ***REMOVED***string***REMOVED*** methodName The name of the original method called.
     * @param ***REMOVED***string***REMOVED*** path The endpoint path to use for the request.
     *
     * @return ***REMOVED***Promise<MessagingTopicManagementResponse>***REMOVED*** A Promise fulfilled with the parsed server
     *   response.
     */
    Messaging.prototype.sendTopicManagementRequest = function (registrationTokenOrTokens, topic, methodName, path) ***REMOVED***
        var _this = this;
        this.validateRegistrationTokensType(registrationTokenOrTokens, methodName);
        this.validateTopicType(topic, methodName);
        // Prepend the topic with /topics/ if necessary.
        topic = this.normalizeTopic(topic);
        return Promise.resolve()
            .then(function () ***REMOVED***
            // Validate the contents of the input arguments. Because we are now in a promise, any thrown
            // error will cause this method to return a rejected promise.
            _this.validateRegistrationTokens(registrationTokenOrTokens, methodName);
            _this.validateTopic(topic, methodName);
            // Ensure the registration token(s) input argument is an array.
            var registrationTokensArray = registrationTokenOrTokens;
            if (validator.isString(registrationTokenOrTokens)) ***REMOVED***
                registrationTokensArray = [registrationTokenOrTokens];
            ***REMOVED***
            var request = ***REMOVED***
                to: topic,
                registration_tokens: registrationTokensArray,
            ***REMOVED***;
            return _this.messagingRequestHandler.invokeRequestHandler(FCM_TOPIC_MANAGEMENT_HOST, path, request);
        ***REMOVED***)
            .then(function (response) ***REMOVED***
            return mapRawResponseToTopicManagementResponse(response);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Validates the types of the messaging payload and options. If invalid, an error will be thrown.
     *
     * @param ***REMOVED***MessagingPayload***REMOVED*** payload The messaging payload to validate.
     * @param ***REMOVED***MessagingOptions***REMOVED*** options The messaging options to validate.
     */
    Messaging.prototype.validateMessagingPayloadAndOptionsTypes = function (payload, options) ***REMOVED***
        // Validate the payload is an object
        if (!validator.isNonNullObject(payload)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, 'Messaging payload must be an object with at least one of the "data" or "notification" properties.');
        ***REMOVED***
        // Validate the options argument is an object
        if (!validator.isNonNullObject(options)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, 'Messaging options must be an object.');
        ***REMOVED***
    ***REMOVED***;
    /**
     * Validates the messaging payload. If invalid, an error will be thrown.
     *
     * @param ***REMOVED***MessagingPayload***REMOVED*** payload The messaging payload to validate.
     *
     * @return ***REMOVED***MessagingPayload***REMOVED*** A copy of the provided payload with whitelisted properties switched
     *     from camelCase to underscore_case.
     */
    Messaging.prototype.validateMessagingPayload = function (payload) ***REMOVED***
        var payloadCopy = deep_copy_1.deepCopy(payload);
        var payloadKeys = Object.keys(payloadCopy);
        var validPayloadKeys = ['data', 'notification'];
        var containsDataOrNotificationKey = false;
        payloadKeys.forEach(function (payloadKey) ***REMOVED***
            // Validate the payload does not contain any invalid keys
            if (validPayloadKeys.indexOf(payloadKey) === -1) ***REMOVED***
                throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, "Messaging payload contains an invalid \"" + payloadKey + "\" property. Valid properties are " +
                    "\"data\" and \"notification\".");
            ***REMOVED***
            else ***REMOVED***
                containsDataOrNotificationKey = true;
            ***REMOVED***
        ***REMOVED***);
        // Validate the payload contains at least one of the "data" and "notification" keys
        if (!containsDataOrNotificationKey) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, 'Messaging payload must contain at least one of the "data" or "notification" properties.');
        ***REMOVED***
        payloadKeys.forEach(function (payloadKey) ***REMOVED***
            var value = payloadCopy[payloadKey];
            // Validate each top-level key in the payload is an object
            if (!validator.isNonNullObject(value)) ***REMOVED***
                throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, "Messaging payload contains an invalid value for the \"" + payloadKey + "\" property. " +
                    "Value must be an object.");
            ***REMOVED***
            Object.keys(value).forEach(function (subKey) ***REMOVED***
                if (!validator.isString(value[subKey])) ***REMOVED***
                    // Validate all sub-keys have a string value
                    throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, "Messaging payload contains an invalid value for the \"" + payloadKey + "." + subKey + "\" " +
                        "property. Values must be strings.");
                ***REMOVED***
                else if (payloadKey === 'data' && /^google\./.test(subKey)) ***REMOVED***
                    // Validate the data payload does not contain keys which start with 'google.'.
                    throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, "Messaging payload contains the blacklisted \"data." + subKey + "\" property.");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
        // Validate the data payload object does not contain blacklisted properties
        if ('data' in payloadCopy) ***REMOVED***
            exports.BLACKLISTED_DATA_PAYLOAD_KEYS.forEach(function (blacklistedKey) ***REMOVED***
                if (blacklistedKey in payloadCopy.data) ***REMOVED***
                    throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_PAYLOAD, "Messaging payload contains the blacklisted \"data." + blacklistedKey + "\" property.");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
        // Convert whitelisted camelCase keys to underscore_case
        if ('notification' in payloadCopy) ***REMOVED***
            utils.renameProperties(payloadCopy.notification, CAMELCASED_NOTIFICATION_PAYLOAD_KEYS_MAP);
        ***REMOVED***
        return payloadCopy;
    ***REMOVED***;
    /**
     * Validates the messaging options. If invalid, an error will be thrown.
     *
     * @param ***REMOVED***MessagingOptions***REMOVED*** options The messaging options to validate.
     *
     * @return ***REMOVED***MessagingOptions***REMOVED*** A copy of the provided options with whitelisted properties switched
     *   from camelCase to underscore_case.
     */
    Messaging.prototype.validateMessagingOptions = function (options) ***REMOVED***
        var optionsCopy = deep_copy_1.deepCopy(options);
        // Validate the options object does not contain blacklisted properties
        exports.BLACKLISTED_OPTIONS_KEYS.forEach(function (blacklistedKey) ***REMOVED***
            if (blacklistedKey in optionsCopy) ***REMOVED***
                throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains the blacklisted \"" + blacklistedKey + "\" property.");
            ***REMOVED***
        ***REMOVED***);
        // Convert whitelisted camelCase keys to underscore_case
        utils.renameProperties(optionsCopy, CAMELCASE_OPTIONS_KEYS_MAP);
        // Validate the options object contains valid values for whitelisted properties
        if ('collapse_key' in optionsCopy && !validator.isNonEmptyString(optionsCopy.collapse_key)) ***REMOVED***
            var keyName = ('collapseKey' in options) ? 'collapseKey' : 'collapse_key';
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains an invalid value for the \"" + keyName + "\" property. Value must " +
                'be a non-empty string.');
        ***REMOVED***
        else if ('dry_run' in optionsCopy && !validator.isBoolean(optionsCopy.dry_run)) ***REMOVED***
            var keyName = ('dryRun' in options) ? 'dryRun' : 'dry_run';
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains an invalid value for the \"" + keyName + "\" property. Value must " +
                'be a boolean.');
        ***REMOVED***
        else if ('priority' in optionsCopy && !validator.isNonEmptyString(optionsCopy.priority)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, 'Messaging options contains an invalid value for the "priority" property. Value must ' +
                'be a non-empty string.');
        ***REMOVED***
        else if ('restricted_package_name' in optionsCopy &&
            !validator.isNonEmptyString(optionsCopy.restricted_package_name)) ***REMOVED***
            var keyName = ('restrictedPackageName' in options) ? 'restrictedPackageName' : 'restricted_package_name';
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains an invalid value for the \"" + keyName + "\" property. Value must " +
                'be a non-empty string.');
        ***REMOVED***
        else if ('time_to_live' in optionsCopy && !validator.isNumber(optionsCopy.time_to_live)) ***REMOVED***
            var keyName = ('timeToLive' in options) ? 'timeToLive' : 'time_to_live';
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains an invalid value for the \"" + keyName + "\" property. Value must " +
                'be a number.');
        ***REMOVED***
        else if ('content_available' in optionsCopy && !validator.isBoolean(optionsCopy.content_available)) ***REMOVED***
            var keyName = ('contentAvailable' in options) ? 'contentAvailable' : 'content_available';
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains an invalid value for the \"" + keyName + "\" property. Value must " +
                'be a boolean.');
        ***REMOVED***
        else if ('mutable_content' in optionsCopy && !validator.isBoolean(optionsCopy.mutable_content)) ***REMOVED***
            var keyName = ('mutableContent' in options) ? 'mutableContent' : 'mutable_content';
            throw new error_1.FirebaseMessagingError(error_1.MessagingClientErrorCode.INVALID_OPTIONS, "Messaging options contains an invalid value for the \"" + keyName + "\" property. Value must " +
                'be a boolean.');
        ***REMOVED***
        return optionsCopy;
    ***REMOVED***;
    /**
     * Validates the type of the provided registration token(s). If invalid, an error will be thrown.
     *
     * @param ***REMOVED***string|string[]***REMOVED*** registrationTokenOrTokens The registration token(s) to validate.
     * @param ***REMOVED***string***REMOVED*** method The method name to use in error messages.
     * @param ***REMOVED***ErrorInfo?***REMOVED*** [errorInfo] The error info to use if the registration tokens are invalid.
     */
    Messaging.prototype.validateRegistrationTokensType = function (registrationTokenOrTokens, methodName, errorInfo) ***REMOVED***
        if (errorInfo === void 0) ***REMOVED*** errorInfo = error_1.MessagingClientErrorCode.INVALID_ARGUMENT; ***REMOVED***
        if (!validator.isNonEmptyArray(registrationTokenOrTokens) &&
            !validator.isNonEmptyString(registrationTokenOrTokens)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(errorInfo, "Registration token(s) provided to " + methodName + "() must be a non-empty string or a " +
                'non-empty array.');
        ***REMOVED***
    ***REMOVED***;
    /**
     * Validates the provided registration tokens. If invalid, an error will be thrown.
     *
     * @param ***REMOVED***string|string[]***REMOVED*** registrationTokenOrTokens The registration token or an array of
     *     registration tokens to validate.
     * @param ***REMOVED***string***REMOVED*** method The method name to use in error messages.
     * @param ***REMOVED***errorInfo?***REMOVED*** [ErrorInfo] The error info to use if the registration tokens are invalid.
     */
    Messaging.prototype.validateRegistrationTokens = function (registrationTokenOrTokens, methodName, errorInfo) ***REMOVED***
        if (errorInfo === void 0) ***REMOVED*** errorInfo = error_1.MessagingClientErrorCode.INVALID_ARGUMENT; ***REMOVED***
        if (validator.isArray(registrationTokenOrTokens)) ***REMOVED***
            // Validate the array contains no more than 1,000 registration tokens.
            if (registrationTokenOrTokens.length > 1000) ***REMOVED***
                throw new error_1.FirebaseMessagingError(errorInfo, "Too many registration tokens provided in a single request to " + methodName + "(). Batch " +
                    'your requests to contain no more than 1,000 registration tokens per request.');
            ***REMOVED***
            // Validate the array contains registration tokens which are non-empty strings.
            registrationTokenOrTokens.forEach(function (registrationToken, index) ***REMOVED***
                if (!validator.isNonEmptyString(registrationToken)) ***REMOVED***
                    throw new error_1.FirebaseMessagingError(errorInfo, "Registration token provided to " + methodName + "() at index " + index + " must be a " +
                        'non-empty string.');
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * Validates the type of the provided topic. If invalid, an error will be thrown.
     *
     * @param ***REMOVED***string***REMOVED*** topic The topic to validate.
     * @param ***REMOVED***string***REMOVED*** method The method name to use in error messages.
     * @param ***REMOVED***ErrorInfo?***REMOVED*** [errorInfo] The error info to use if the topic is invalid.
     */
    Messaging.prototype.validateTopicType = function (topic, methodName, errorInfo) ***REMOVED***
        if (errorInfo === void 0) ***REMOVED*** errorInfo = error_1.MessagingClientErrorCode.INVALID_ARGUMENT; ***REMOVED***
        if (!validator.isNonEmptyString(topic)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(errorInfo, "Topic provided to " + methodName + "() must be a string which matches the format " +
                '"/topics/[a-zA-Z0-9-_.~%]+".');
        ***REMOVED***
    ***REMOVED***;
    /**
     * Validates the provided topic. If invalid, an error will be thrown.
     *
     * @param ***REMOVED***string***REMOVED*** topic The topic to validate.
     * @param ***REMOVED***string***REMOVED*** method The method name to use in error messages.
     * @param ***REMOVED***ErrorInfo?***REMOVED*** [errorInfo] The error info to use if the topic is invalid.
     */
    Messaging.prototype.validateTopic = function (topic, methodName, errorInfo) ***REMOVED***
        if (errorInfo === void 0) ***REMOVED*** errorInfo = error_1.MessagingClientErrorCode.INVALID_ARGUMENT; ***REMOVED***
        if (!validator.isTopic(topic)) ***REMOVED***
            throw new error_1.FirebaseMessagingError(errorInfo, "Topic provided to " + methodName + "() must be a string which matches the format " +
                '"/topics/[a-zA-Z0-9-_.~%]+".');
        ***REMOVED***
    ***REMOVED***;
    /**
     * Normalizes the provided topic name by prepending it with '/topics/', if necessary.
     *
     * @param ***REMOVED***string***REMOVED*** topic The topic name to normalize.
     *
     * @return ***REMOVED***string***REMOVED*** The normalized topic name.
     */
    Messaging.prototype.normalizeTopic = function (topic) ***REMOVED***
        if (!/^\/topics\//.test(topic)) ***REMOVED***
            topic = "/topics/" + topic;
        ***REMOVED***
        return topic;
    ***REMOVED***;
    return Messaging;
***REMOVED***());
exports.Messaging = Messaging;
;
