// protobufjs/ext/debug/index.js
"use strict";
var protobuf = require("../..");

/**
 * Debugging utility functions. Only present in debug builds.
 * @namespace
 */
var debug = protobuf.debug = module.exports = ***REMOVED******REMOVED***;

var codegen = protobuf.util.codegen;

var debugFnRe = /function ([^(]+)\(([^)]*)\) ***REMOVED***/g;

// Counts number of calls to any generated function
function codegen_debug() ***REMOVED***
    codegen_debug.supported = codegen.supported;
    codegen_debug.verbose = codegen.verbose;
    var gen = codegen.apply(null, Array.prototype.slice.call(arguments));
    gen.str = (function(str) ***REMOVED*** return function str_debug() ***REMOVED***
        return str.apply(null, Array.prototype.slice.call(arguments)).replace(debugFnRe, "function $1($2) ***REMOVED***\n\t$1.calls=($1.calls|0)+1");
    ***REMOVED***;***REMOVED***)(gen.str);
    return gen;
***REMOVED***

/**
 * Returns a list of unused types within the specified root.
 * @param ***REMOVED***NamespaceBase***REMOVED*** ns Namespace to search
 * @returns ***REMOVED***Type[]***REMOVED*** Unused types
 */
debug.unusedTypes = function unusedTypes(ns) ***REMOVED***

    /* istanbul ignore if */
    if (!(ns instanceof protobuf.Namespace))
        throw TypeError("ns must be a Namespace");

    /* istanbul ignore if */
    if (!ns.nested)
        return [];

    var unused = [];
    for (var names = Object.keys(ns.nested), i = 0; i < names.length; ++i) ***REMOVED***
        var nested = ns.nested[names[i]];
        if (nested instanceof protobuf.Type) ***REMOVED***
            var calls = (nested.encode.calls|0)
                      + (nested.decode.calls|0)
                      + (nested.verify.calls|0)
                      + (nested.toObject.calls|0)
                      + (nested.fromObject.calls|0);
            if (!calls)
                unused.push(nested);
        ***REMOVED*** else if (nested instanceof protobuf.Namespace)
            Array.prototype.push.apply(unused, unusedTypes(nested));
    ***REMOVED***
    return unused;
***REMOVED***;

/**
 * Enables debugging extensions.
 * @returns ***REMOVED***undefined***REMOVED***
 */
debug.enable = function enable() ***REMOVED***
    protobuf.util.codegen = codegen_debug;
***REMOVED***;

/**
 * Disables debugging extensions.
 * @returns ***REMOVED***undefined***REMOVED***
 */
debug.disable = function disable() ***REMOVED***
    protobuf.util.codegen = codegen;
***REMOVED***;
