// har-validator/lib/browser/promise.js
import * as schemas from 'har-schema';
import Ajv from 'ajv';
import HARError from './error';

let ajv;

export function validate(name, data = ***REMOVED******REMOVED***) ***REMOVED***
  // validator config
  ajv = ajv || new Ajv(***REMOVED***
    allErrors: true,
    schemas: schemas
  ***REMOVED***);

  let validate = ajv.getSchema(name + '.json');

  return new Promise((resolve, reject) => ***REMOVED***
    let valid = validate(data);

    !valid ? reject(new HARError(validate.errors)) : resolve(data);
  ***REMOVED***);
***REMOVED***

export function afterRequest(data) ***REMOVED***
  return validate('afterRequest', data);
***REMOVED***

export function beforeRequest(data) ***REMOVED***
  return validate('beforeRequest', data);
***REMOVED***

export function browser(data) ***REMOVED***
  return validate('browser', data);
***REMOVED***

export function cache(data) ***REMOVED***
  return validate('cache', data);
***REMOVED***

export function content(data) ***REMOVED***
  return validate('content', data);
***REMOVED***

export function cookie(data) ***REMOVED***
  return validate('cookie', data);
***REMOVED***

export function creator(data) ***REMOVED***
  return validate('creator', data);
***REMOVED***

export function entry(data) ***REMOVED***
  return validate('entry', data);
***REMOVED***

export function har(data) ***REMOVED***
  return validate('har', data);
***REMOVED***

export function header(data) ***REMOVED***
  return validate('header', data);
***REMOVED***

export function log(data) ***REMOVED***
  return validate('log', data);
***REMOVED***

export function page(data) ***REMOVED***
  return validate('page', data);
***REMOVED***

export function pageTimings(data) ***REMOVED***
  return validate('pageTimings', data);
***REMOVED***

export function postData(data) ***REMOVED***
  return validate('postData', data);
***REMOVED***

export function query(data) ***REMOVED***
  return validate('query', data);
***REMOVED***

export function request(data) ***REMOVED***
  return validate('request', data);
***REMOVED***

export function response(data) ***REMOVED***
  return validate('response', data);
***REMOVED***

export function timings(data) ***REMOVED***
  return validate('timings', data);
***REMOVED***
