// es5-ext/date/#/format.js
/* eslint id-length: "off" */

"use strict";

var pad  = require("../../number/#/pad")
  , date = require("../valid-date")
  , format;

format = require("../../string/format-method")(***REMOVED***
	Y: function () ***REMOVED***
		return String(this.getFullYear());
	***REMOVED***,
	y: function () ***REMOVED***
		return String(this.getFullYear()).slice(-2);
	***REMOVED***,
	m: function () ***REMOVED***
		return pad.call(this.getMonth() + 1, 2);
	***REMOVED***,
	d: function () ***REMOVED***
		return pad.call(this.getDate(), 2);
	***REMOVED***,
	H: function () ***REMOVED***
		return pad.call(this.getHours(), 2);
	***REMOVED***,
	M: function () ***REMOVED***
		return pad.call(this.getMinutes(), 2);
	***REMOVED***,
	S: function () ***REMOVED***
		return pad.call(this.getSeconds(), 2);
	***REMOVED***,
	L: function () ***REMOVED***
		return pad.call(this.getMilliseconds(), 3);
	***REMOVED***
***REMOVED***);

module.exports = function (pattern) ***REMOVED***
	return format.call(date(this), pattern);
***REMOVED***;
