// micromatch/lib/utils.js
'use strict';

var win32 = process && process.platform === 'win32';
var path = require('path');
var fileRe = require('filename-regex');
var utils = module.exports;

/**
 * Module dependencies
 */

utils.diff = require('arr-diff');
utils.unique = require('array-unique');
utils.braces = require('braces');
utils.brackets = require('expand-brackets');
utils.extglob = require('extglob');
utils.isExtglob = require('is-extglob');
utils.isGlob = require('is-glob');
utils.typeOf = require('kind-of');
utils.normalize = require('normalize-path');
utils.omit = require('object.omit');
utils.parseGlob = require('parse-glob');
utils.cache = require('regex-cache');

/**
 * Get the filename of a filepath
 *
 * @param ***REMOVED***String***REMOVED*** `string`
 * @return ***REMOVED***String***REMOVED***
 */

utils.filename = function filename(fp) ***REMOVED***
  var seg = fp.match(fileRe());
  return seg && seg[0];
***REMOVED***;

/**
 * Returns a function that returns true if the given
 * pattern is the same as a given `filepath`
 *
 * @param ***REMOVED***String***REMOVED*** `pattern`
 * @return ***REMOVED***Function***REMOVED***
 */

utils.isPath = function isPath(pattern, opts) ***REMOVED***
  opts = opts || ***REMOVED******REMOVED***;
  return function(fp) ***REMOVED***
    var unixified = utils.unixify(fp, opts);
    if(opts.nocase)***REMOVED***
      return pattern.toLowerCase() === unixified.toLowerCase();
    ***REMOVED***
    return pattern === unixified;
  ***REMOVED***;
***REMOVED***;

/**
 * Returns a function that returns true if the given
 * pattern contains a `filepath`
 *
 * @param ***REMOVED***String***REMOVED*** `pattern`
 * @return ***REMOVED***Function***REMOVED***
 */

utils.hasPath = function hasPath(pattern, opts) ***REMOVED***
  return function(fp) ***REMOVED***
    return utils.unixify(pattern, opts).indexOf(fp) !== -1;
  ***REMOVED***;
***REMOVED***;

/**
 * Returns a function that returns true if the given
 * pattern matches or contains a `filepath`
 *
 * @param ***REMOVED***String***REMOVED*** `pattern`
 * @return ***REMOVED***Function***REMOVED***
 */

utils.matchPath = function matchPath(pattern, opts) ***REMOVED***
  var fn = (opts && opts.contains)
    ? utils.hasPath(pattern, opts)
    : utils.isPath(pattern, opts);
  return fn;
***REMOVED***;

/**
 * Returns a function that returns true if the given
 * regex matches the `filename` of a file path.
 *
 * @param ***REMOVED***RegExp***REMOVED*** `re`
 * @return ***REMOVED***Boolean***REMOVED***
 */

utils.hasFilename = function hasFilename(re) ***REMOVED***
  return function(fp) ***REMOVED***
    var name = utils.filename(fp);
    return name && re.test(name);
  ***REMOVED***;
***REMOVED***;

/**
 * Coerce `val` to an array
 *
 * @param  ***REMOVED*******REMOVED*** val
 * @return ***REMOVED***Array***REMOVED***
 */

utils.arrayify = function arrayify(val) ***REMOVED***
  return !Array.isArray(val)
    ? [val]
    : val;
***REMOVED***;

/**
 * Normalize all slashes in a file path or glob pattern to
 * forward slashes.
 */

utils.unixify = function unixify(fp, opts) ***REMOVED***
  if (opts && opts.unixify === false) return fp;
  if (opts && opts.unixify === true || win32 || path.sep === '\\') ***REMOVED***
    return utils.normalize(fp, false);
  ***REMOVED***
  if (opts && opts.unescape === true) ***REMOVED***
    return fp ? fp.toString().replace(/\\(\w)/g, '$1') : '';
  ***REMOVED***
  return fp;
***REMOVED***;

/**
 * Escape/unescape utils
 */

utils.escapePath = function escapePath(fp) ***REMOVED***
  return fp.replace(/[\\.]/g, '\\$&');
***REMOVED***;

utils.unescapeGlob = function unescapeGlob(fp) ***REMOVED***
  return fp.replace(/[\\"']/g, '');
***REMOVED***;

utils.escapeRe = function escapeRe(str) ***REMOVED***
  return str.replace(/[-[\\$*+?.#^\s***REMOVED******REMOVED***(|)\]]/g, '\\$&');
***REMOVED***;

/**
 * Expose `utils`
 */

module.exports = utils;
