// ajv/lib/compile/formats.js
'use strict';

var util = require('./util');

var DATE = /^\d\d\d\d-(\d\d)-(\d\d)$/;
var DAYS = [0,31,29,31,30,31,30,31,31,30,31,30,31];
var TIME = /^(\d\d):(\d\d):(\d\d)(\.\d+)?(z|[+-]\d\d:\d\d)?$/i;
var HOSTNAME = /^[a-z0-9](?:[a-z0-9-]***REMOVED***0,61***REMOVED***[a-z0-9])?(?:\.[a-z0-9](?:[-0-9a-z]***REMOVED***0,61***REMOVED***[0-9a-z])?)*$/i;
var URI = /^(?:[a-z][a-z0-9+\-.]*:)?(?:\/?\/(?:(?:[a-z0-9\-._~!$&'()*+,;=:]|%[0-9a-f]***REMOVED***2***REMOVED***)*@)?(?:\[(?:(?:(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***6***REMOVED***|::(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***5***REMOVED***|(?:[0-9a-f]***REMOVED***1,4***REMOVED***)?::(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***4***REMOVED***|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***0,1***REMOVED***[0-9a-f]***REMOVED***1,4***REMOVED***)?::(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***3***REMOVED***|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***0,2***REMOVED***[0-9a-f]***REMOVED***1,4***REMOVED***)?::(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***2***REMOVED***|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***0,3***REMOVED***[0-9a-f]***REMOVED***1,4***REMOVED***)?::[0-9a-f]***REMOVED***1,4***REMOVED***:|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***0,4***REMOVED***[0-9a-f]***REMOVED***1,4***REMOVED***)?::)(?:[0-9a-f]***REMOVED***1,4***REMOVED***:[0-9a-f]***REMOVED***1,4***REMOVED***|(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.)***REMOVED***3***REMOVED***(?:25[0-5]|2[0-4]\d|[01]?\d\d?))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***0,5***REMOVED***[0-9a-f]***REMOVED***1,4***REMOVED***)?::[0-9a-f]***REMOVED***1,4***REMOVED***|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***0,6***REMOVED***[0-9a-f]***REMOVED***1,4***REMOVED***)?::)|[Vv][0-9a-f]+\.[a-z0-9\-._~!$&'()*+,;=:]+)\]|(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.)***REMOVED***3***REMOVED***(?:25[0-5]|2[0-4]\d|[01]?\d\d?)|(?:[a-z0-9\-._~!$&'()*+,;=]|%[0-9a-f]***REMOVED***2***REMOVED***)*)(?::\d*)?(?:\/(?:[a-z0-9\-._~!$&'()*+,;=:@]|%[0-9a-f]***REMOVED***2***REMOVED***)*)*|\/(?:(?:[a-z0-9\-._~!$&'()*+,;=:@]|%[0-9a-f]***REMOVED***2***REMOVED***)+(?:\/(?:[a-z0-9\-._~!$&'()*+,;=:@]|%[0-9a-f]***REMOVED***2***REMOVED***)*)*)?|(?:[a-z0-9\-._~!$&'()*+,;=:@]|%[0-9a-f]***REMOVED***2***REMOVED***)+(?:\/(?:[a-z0-9\-._~!$&'()*+,;=:@]|%[0-9a-f]***REMOVED***2***REMOVED***)*)*)(?:\?(?:[a-z0-9\-._~!$&'()*+,;=:@\/?]|%[0-9a-f]***REMOVED***2***REMOVED***)*)?(?:\#(?:[a-z0-9\-._~!$&'()*+,;=:@\/?]|%[0-9a-f]***REMOVED***2***REMOVED***)*)?$/i;
var UUID = /^(?:urn\:uuid\:)?[0-9a-f]***REMOVED***8***REMOVED***-(?:[0-9a-f]***REMOVED***4***REMOVED***-)***REMOVED***3***REMOVED***[0-9a-f]***REMOVED***12***REMOVED***$/i;
var JSON_POINTER = /^(?:\/(?:[^~\/]|~0|~1)*)*$|^\#(?:\/(?:[a-z0-9_\-\.!$&'()*+,;:=@]|%[0-9a-f]***REMOVED***2***REMOVED***|~0|~1)*)*$/i;
var RELATIVE_JSON_POINTER = /^(?:0|[1-9][0-9]*)(?:\#|(?:\/(?:[^~\/]|~0|~1)*)*)$/;


module.exports = formats;

function formats(mode) ***REMOVED***
  mode = mode == 'full' ? 'full' : 'fast';
  var formatDefs = util.copy(formats[mode]);
  for (var fName in formats.compare) ***REMOVED***
    formatDefs[fName] = ***REMOVED***
      validate: formatDefs[fName],
      compare: formats.compare[fName]
    ***REMOVED***;
  ***REMOVED***
  return formatDefs;
***REMOVED***


formats.fast = ***REMOVED***
  // date: http://tools.ietf.org/html/rfc3339#section-5.6
  date: /^\d\d\d\d-[0-1]\d-[0-3]\d$/,
  // date-time: http://tools.ietf.org/html/rfc3339#section-5.6
  time: /^[0-2]\d:[0-5]\d:[0-5]\d(?:\.\d+)?(?:z|[+-]\d\d:\d\d)?$/i,
  'date-time': /^\d\d\d\d-[0-1]\d-[0-3]\d[t\s][0-2]\d:[0-5]\d:[0-5]\d(?:\.\d+)?(?:z|[+-]\d\d:\d\d)$/i,
  // uri: https://github.com/mafintosh/is-my-json-valid/blob/master/formats.js
  uri: /^(?:[a-z][a-z0-9+-.]*)?(?:\:|\/)\/?[^\s]*$/i,
  // email (sources from jsen validator):
  // http://stackoverflow.com/questions/201323/using-a-regular-expression-to-validate-an-email-address#answer-8829363
  // http://www.w3.org/TR/html5/forms.html#valid-e-mail-address (search for 'willful violation')
  email: /^[a-z0-9.!#$%&'*+\/=?^_`***REMOVED***|***REMOVED***~-]+@[a-z0-9](?:[a-z0-9-]***REMOVED***0,61***REMOVED***[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]***REMOVED***0,61***REMOVED***[a-z0-9])?)*$/i,
  hostname: HOSTNAME,
  // optimized https://www.safaribooksonline.com/library/view/regular-expressions-cookbook/9780596802837/ch07s16.html
  ipv4: /^(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.)***REMOVED***3***REMOVED***(?:25[0-5]|2[0-4]\d|[01]?\d\d?)$/,
  // optimized http://stackoverflow.com/questions/53497/regular-expression-that-matches-valid-ipv6-addresses
  ipv6: /^\s*(?:(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***7***REMOVED***(?:[0-9a-f]***REMOVED***1,4***REMOVED***|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***6***REMOVED***(?::[0-9a-f]***REMOVED***1,4***REMOVED***|(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***)|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***5***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,2***REMOVED***)|:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***)|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***4***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,3***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)?:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***3***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,4***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,2***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***2***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,5***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,3***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***1***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,6***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,4***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?::(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,7***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,5***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:)))(?:%.+)?\s*$/i,
  regex: regex,
  // uuid: http://tools.ietf.org/html/rfc4122
  uuid: UUID,
  // JSON-pointer: https://tools.ietf.org/html/rfc6901
  // uri fragment: https://tools.ietf.org/html/rfc3986#appendix-A
  'json-pointer': JSON_POINTER,
  // relative JSON-pointer: http://tools.ietf.org/html/draft-luff-relative-json-pointer-00
  'relative-json-pointer': RELATIVE_JSON_POINTER
***REMOVED***;


formats.full = ***REMOVED***
  date: date,
  time: time,
  'date-time': date_time,
  uri: uri,
  email: /^[a-z0-9!#$%&'*+\/=?^_`***REMOVED***|***REMOVED***~-]+(?:\.[a-z0-9!#$%&''*+\/=?^_`***REMOVED***|***REMOVED***~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i,
  hostname: hostname,
  ipv4: /^(?:(?:25[0-5]|2[0-4]\d|[01]?\d\d?)\.)***REMOVED***3***REMOVED***(?:25[0-5]|2[0-4]\d|[01]?\d\d?)$/,
  ipv6: /^\s*(?:(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***7***REMOVED***(?:[0-9a-f]***REMOVED***1,4***REMOVED***|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***6***REMOVED***(?::[0-9a-f]***REMOVED***1,4***REMOVED***|(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***)|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***5***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,2***REMOVED***)|:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***)|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***4***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,3***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)?:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***3***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,4***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,2***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***2***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,5***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,3***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?:(?:[0-9a-f]***REMOVED***1,4***REMOVED***:)***REMOVED***1***REMOVED***(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,6***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,4***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:))|(?::(?:(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***1,7***REMOVED***)|(?:(?::[0-9a-f]***REMOVED***1,4***REMOVED***)***REMOVED***0,5***REMOVED***:(?:(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(?:\.(?:25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d))***REMOVED***3***REMOVED***))|:)))(?:%.+)?\s*$/i,
  regex: regex,
  uuid: UUID,
  'json-pointer': JSON_POINTER,
  'relative-json-pointer': RELATIVE_JSON_POINTER
***REMOVED***;


formats.compare = ***REMOVED***
  date: compareDate,
  time: compareTime,
  'date-time': compareDateTime
***REMOVED***;


function date(str) ***REMOVED***
  // full-date from http://tools.ietf.org/html/rfc3339#section-5.6
  var matches = str.match(DATE);
  if (!matches) return false;

  var month = +matches[1];
  var day = +matches[2];
  return month >= 1 && month <= 12 && day >= 1 && day <= DAYS[month];
***REMOVED***


function time(str, full) ***REMOVED***
  var matches = str.match(TIME);
  if (!matches) return false;

  var hour = matches[1];
  var minute = matches[2];
  var second = matches[3];
  var timeZone = matches[5];
  return hour <= 23 && minute <= 59 && second <= 59 && (!full || timeZone);
***REMOVED***


var DATE_TIME_SEPARATOR = /t|\s/i;
function date_time(str) ***REMOVED***
  // http://tools.ietf.org/html/rfc3339#section-5.6
  var dateTime = str.split(DATE_TIME_SEPARATOR);
  return dateTime.length == 2 && date(dateTime[0]) && time(dateTime[1], true);
***REMOVED***


function hostname(str) ***REMOVED***
  // https://tools.ietf.org/html/rfc1034#section-3.5
  // https://tools.ietf.org/html/rfc1123#section-2
  return str.length <= 255 && HOSTNAME.test(str);
***REMOVED***


var NOT_URI_FRAGMENT = /\/|\:/;
function uri(str) ***REMOVED***
  // http://jmrware.com/articles/2009/uri_regexp/URI_regex.html + optional protocol + required "."
  return NOT_URI_FRAGMENT.test(str) && URI.test(str);
***REMOVED***


function regex(str) ***REMOVED***
  try ***REMOVED***
    new RegExp(str);
    return true;
  ***REMOVED*** catch(e) ***REMOVED***
    return false;
  ***REMOVED***
***REMOVED***


function compareDate(d1, d2) ***REMOVED***
  if (!(d1 && d2)) return;
  if (d1 > d2) return 1;
  if (d1 < d2) return -1;
  if (d1 === d2) return 0;
***REMOVED***


function compareTime(t1, t2) ***REMOVED***
  if (!(t1 && t2)) return;
  t1 = t1.match(TIME);
  t2 = t2.match(TIME);
  if (!(t1 && t2)) return;
  t1 = t1[1] + t1[2] + t1[3] + (t1[4]||'');
  t2 = t2[1] + t2[2] + t2[3] + (t2[4]||'');
  if (t1 > t2) return 1;
  if (t1 < t2) return -1;
  if (t1 === t2) return 0;
***REMOVED***


function compareDateTime(dt1, dt2) ***REMOVED***
  if (!(dt1 && dt2)) return;
  dt1 = dt1.split(DATE_TIME_SEPARATOR);
  dt2 = dt2.split(DATE_TIME_SEPARATOR);
  var res = compareDate(dt1[0], dt2[0]);
  if (res === undefined) return;
  return res || compareTime(dt1[1], dt2[1]);
***REMOVED***
