// cross-spawn/lib/parse.js
'use strict';

var resolveCommand = require('./util/resolveCommand');
var hasEmptyArgumentBug = require('./util/hasEmptyArgumentBug');
var escapeArgument = require('./util/escapeArgument');
var escapeCommand = require('./util/escapeCommand');
var readShebang = require('./util/readShebang');

var isWin = process.platform === 'win32';
var skipShellRegExp = /\.(?:com|exe)$/i;

// Supported in Node >= 6 and >= 4.8
var supportsShellOption = parseInt(process.version.substr(1).split('.')[0], 10) >= 6 ||
 parseInt(process.version.substr(1).split('.')[0], 10) === 4 && parseInt(process.version.substr(1).split('.')[1], 10) >= 8;

function parseNonShell(parsed) ***REMOVED***
    var shebang;
    var needsShell;
    var applyQuotes;

    if (!isWin) ***REMOVED***
        return parsed;
    ***REMOVED***

    // Detect & add support for shebangs
    parsed.file = resolveCommand(parsed.command);
    parsed.file = parsed.file || resolveCommand(parsed.command, true);
    shebang = parsed.file && readShebang(parsed.file);

    if (shebang) ***REMOVED***
        parsed.args.unshift(parsed.file);
        parsed.command = shebang;
        needsShell = hasEmptyArgumentBug || !skipShellRegExp.test(resolveCommand(shebang) || resolveCommand(shebang, true));
    ***REMOVED*** else ***REMOVED***
        needsShell = hasEmptyArgumentBug || !skipShellRegExp.test(parsed.file);
    ***REMOVED***

    // If a shell is required, use cmd.exe and take care of escaping everything correctly
    if (needsShell) ***REMOVED***
        // Escape command & arguments
        applyQuotes = (parsed.command !== 'echo');  // Do not quote arguments for the special "echo" command
        parsed.command = escapeCommand(parsed.command);
        parsed.args = parsed.args.map(function (arg) ***REMOVED***
            return escapeArgument(arg, applyQuotes);
        ***REMOVED***);

        // Make use of cmd.exe
        parsed.args = ['/d', '/s', '/c', '"' + parsed.command + (parsed.args.length ? ' ' + parsed.args.join(' ') : '') + '"'];
        parsed.command = process.env.comspec || 'cmd.exe';
        parsed.options.windowsVerbatimArguments = true;  // Tell node's spawn that the arguments are already escaped
    ***REMOVED***

    return parsed;
***REMOVED***

function parseShell(parsed) ***REMOVED***
    var shellCommand;

    // If node supports the shell option, there's no need to mimic its behavior
    if (supportsShellOption) ***REMOVED***
        return parsed;
    ***REMOVED***

    // Mimic node shell option, see: https://github.com/nodejs/node/blob/b9f6a2dc059a1062776133f3d4fd848c4da7d150/lib/child_process.js#L335
    shellCommand = [parsed.command].concat(parsed.args).join(' ');

    if (isWin) ***REMOVED***
        parsed.command = typeof parsed.options.shell === 'string' ? parsed.options.shell : process.env.comspec || 'cmd.exe';
        parsed.args = ['/d', '/s', '/c', '"' + shellCommand + '"'];
        parsed.options.windowsVerbatimArguments = true;  // Tell node's spawn that the arguments are already escaped
    ***REMOVED*** else ***REMOVED***
        if (typeof parsed.options.shell === 'string') ***REMOVED***
            parsed.command = parsed.options.shell;
        ***REMOVED*** else if (process.platform === 'android') ***REMOVED***
            parsed.command = '/system/bin/sh';
        ***REMOVED*** else ***REMOVED***
            parsed.command = '/bin/sh';
        ***REMOVED***

        parsed.args = ['-c', shellCommand];
    ***REMOVED***

    return parsed;
***REMOVED***

// ------------------------------------------------

function parse(command, args, options) ***REMOVED***
    var parsed;

    // Normalize arguments, similar to nodejs
    if (args && !Array.isArray(args)) ***REMOVED***
        options = args;
        args = null;
    ***REMOVED***

    args = args ? args.slice(0) : [];  // Clone array to avoid changing the original
    options = options || ***REMOVED******REMOVED***;

    // Build our parsed object
    parsed = ***REMOVED***
        command: command,
        args: args,
        options: options,
        file: undefined,
        original: command,
    ***REMOVED***;

    // Delegate further parsing to shell or non-shell
    return options.shell ? parseShell(parsed) : parseNonShell(parsed);
***REMOVED***

module.exports = parse;
