// protobufjs/dist/minimal/protobuf.js
/*!
 * protobuf.js v6.8.4 (c) 2016, daniel wirtz
 * compiled thu, 04 jan 2018 21:51:39 utc
 * licensed under the bsd-3-clause license
 * see: https://github.com/dcodeio/protobuf.js for details
 */
(function(global,undefined)***REMOVED***"use strict";(function prelude(modules, cache, entries) ***REMOVED***

    // This is the prelude used to bundle protobuf.js for the browser. Wraps up the CommonJS
    // sources through a conflict-free require shim and is again wrapped within an iife that
    // provides a unified `global` and a minification-friendly `undefined` var plus a global
    // "use strict" directive so that minification can remove the directives of each module.

    function $require(name) ***REMOVED***
        var $module = cache[name];
        if (!$module)
            modules[name][0].call($module = cache[name] = ***REMOVED*** exports: ***REMOVED******REMOVED*** ***REMOVED***, $require, $module, $module.exports);
        return $module.exports;
    ***REMOVED***

    // Expose globally
    var protobuf = global.protobuf = $require(entries[0]);

    // Be nice to AMD
    if (typeof define === "function" && define.amd)
        define(["long"], function(Long) ***REMOVED***
            if (Long && Long.isLong) ***REMOVED***
                protobuf.util.Long = Long;
                protobuf.configure();
            ***REMOVED***
            return protobuf;
        ***REMOVED***);

    // Be nice to CommonJS
    if (typeof module === "object" && module && module.exports)
        module.exports = protobuf;

***REMOVED***)/* end of prelude */(***REMOVED***1:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = asPromise;

/**
 * Callback as used by ***REMOVED***@link util.asPromise***REMOVED***.
 * @typedef asPromiseCallback
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***Error|null***REMOVED*** error Error, if any
 * @param ***REMOVED***...****REMOVED*** params Additional arguments
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Returns a promise from a node-style callback function.
 * @memberof util
 * @param ***REMOVED***asPromiseCallback***REMOVED*** fn Function to call
 * @param ***REMOVED*******REMOVED*** ctx Function context
 * @param ***REMOVED***...****REMOVED*** params Function arguments
 * @returns ***REMOVED***Promise<*>***REMOVED*** Promisified function
 */
function asPromise(fn, ctx/*, varargs */) ***REMOVED***
    var params  = new Array(arguments.length - 1),
        offset  = 0,
        index   = 2,
        pending = true;
    while (index < arguments.length)
        params[offset++] = arguments[index++];
    return new Promise(function executor(resolve, reject) ***REMOVED***
        params[offset] = function callback(err/*, varargs */) ***REMOVED***
            if (pending) ***REMOVED***
                pending = false;
                if (err)
                    reject(err);
                else ***REMOVED***
                    var params = new Array(arguments.length - 1),
                        offset = 0;
                    while (offset < params.length)
                        params[offset++] = arguments[offset];
                    resolve.apply(null, params);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
        try ***REMOVED***
            fn.apply(ctx || null, params);
        ***REMOVED*** catch (err) ***REMOVED***
            if (pending) ***REMOVED***
                pending = false;
                reject(err);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***);
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],2:[function(require,module,exports)***REMOVED***
"use strict";

/**
 * A minimal base64 implementation for number arrays.
 * @memberof util
 * @namespace
 */
var base64 = exports;

/**
 * Calculates the byte length of a base64 encoded string.
 * @param ***REMOVED***string***REMOVED*** string Base64 encoded string
 * @returns ***REMOVED***number***REMOVED*** Byte length
 */
base64.length = function length(string) ***REMOVED***
    var p = string.length;
    if (!p)
        return 0;
    var n = 0;
    while (--p % 4 > 1 && string.charAt(p) === "=")
        ++n;
    return Math.ceil(string.length * 3) / 4 - n;
***REMOVED***;

// Base64 encoding table
var b64 = new Array(64);

// Base64 decoding table
var s64 = new Array(123);

// 65..90, 97..122, 48..57, 43, 47
for (var i = 0; i < 64;)
    s64[b64[i] = i < 26 ? i + 65 : i < 52 ? i + 71 : i < 62 ? i - 4 : i - 59 | 43] = i++;

/**
 * Encodes a buffer to a base64 encoded string.
 * @param ***REMOVED***Uint8Array***REMOVED*** buffer Source buffer
 * @param ***REMOVED***number***REMOVED*** start Source start
 * @param ***REMOVED***number***REMOVED*** end Source end
 * @returns ***REMOVED***string***REMOVED*** Base64 encoded string
 */
base64.encode = function encode(buffer, start, end) ***REMOVED***
    var parts = null,
        chunk = [];
    var i = 0, // output index
        j = 0, // goto index
        t;     // temporary
    while (start < end) ***REMOVED***
        var b = buffer[start++];
        switch (j) ***REMOVED***
            case 0:
                chunk[i++] = b64[b >> 2];
                t = (b & 3) << 4;
                j = 1;
                break;
            case 1:
                chunk[i++] = b64[t | b >> 4];
                t = (b & 15) << 2;
                j = 2;
                break;
            case 2:
                chunk[i++] = b64[t | b >> 6];
                chunk[i++] = b64[b & 63];
                j = 0;
                break;
        ***REMOVED***
        if (i > 8191) ***REMOVED***
            (parts || (parts = [])).push(String.fromCharCode.apply(String, chunk));
            i = 0;
        ***REMOVED***
    ***REMOVED***
    if (j) ***REMOVED***
        chunk[i++] = b64[t];
        chunk[i++] = 61;
        if (j === 1)
            chunk[i++] = 61;
    ***REMOVED***
    if (parts) ***REMOVED***
        if (i)
            parts.push(String.fromCharCode.apply(String, chunk.slice(0, i)));
        return parts.join("");
    ***REMOVED***
    return String.fromCharCode.apply(String, chunk.slice(0, i));
***REMOVED***;

var invalidEncoding = "invalid encoding";

/**
 * Decodes a base64 encoded string to a buffer.
 * @param ***REMOVED***string***REMOVED*** string Source string
 * @param ***REMOVED***Uint8Array***REMOVED*** buffer Destination buffer
 * @param ***REMOVED***number***REMOVED*** offset Destination offset
 * @returns ***REMOVED***number***REMOVED*** Number of bytes written
 * @throws ***REMOVED***Error***REMOVED*** If encoding is invalid
 */
base64.decode = function decode(string, buffer, offset) ***REMOVED***
    var start = offset;
    var j = 0, // goto index
        t;     // temporary
    for (var i = 0; i < string.length;) ***REMOVED***
        var c = string.charCodeAt(i++);
        if (c === 61 && j > 1)
            break;
        if ((c = s64[c]) === undefined)
            throw Error(invalidEncoding);
        switch (j) ***REMOVED***
            case 0:
                t = c;
                j = 1;
                break;
            case 1:
                buffer[offset++] = t << 2 | (c & 48) >> 4;
                t = c;
                j = 2;
                break;
            case 2:
                buffer[offset++] = (t & 15) << 4 | (c & 60) >> 2;
                t = c;
                j = 3;
                break;
            case 3:
                buffer[offset++] = (t & 3) << 6 | c;
                j = 0;
                break;
        ***REMOVED***
    ***REMOVED***
    if (j === 1)
        throw Error(invalidEncoding);
    return offset - start;
***REMOVED***;

/**
 * Tests if the specified string appears to be base64 encoded.
 * @param ***REMOVED***string***REMOVED*** string String to test
 * @returns ***REMOVED***boolean***REMOVED*** `true` if probably base64 encoded, otherwise false
 */
base64.test = function test(string) ***REMOVED***
    return /^(?:[A-Za-z0-9+/]***REMOVED***4***REMOVED***)*(?:[A-Za-z0-9+/]***REMOVED***2***REMOVED***==|[A-Za-z0-9+/]***REMOVED***3***REMOVED***=)?$/.test(string);
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],3:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = EventEmitter;

/**
 * Constructs a new event emitter instance.
 * @classdesc A minimal event emitter.
 * @memberof util
 * @constructor
 */
function EventEmitter() ***REMOVED***

    /**
     * Registered listeners.
     * @type ***REMOVED***Object.<string,*>***REMOVED***
     * @private
     */
    this._listeners = ***REMOVED******REMOVED***;
***REMOVED***

/**
 * Registers an event listener.
 * @param ***REMOVED***string***REMOVED*** evt Event name
 * @param ***REMOVED***function***REMOVED*** fn Listener
 * @param ***REMOVED*******REMOVED*** [ctx] Listener context
 * @returns ***REMOVED***util.EventEmitter***REMOVED*** `this`
 */
EventEmitter.prototype.on = function on(evt, fn, ctx) ***REMOVED***
    (this._listeners[evt] || (this._listeners[evt] = [])).push(***REMOVED***
        fn  : fn,
        ctx : ctx || this
    ***REMOVED***);
    return this;
***REMOVED***;

/**
 * Removes an event listener or any matching listeners if arguments are omitted.
 * @param ***REMOVED***string***REMOVED*** [evt] Event name. Removes all listeners if omitted.
 * @param ***REMOVED***function***REMOVED*** [fn] Listener to remove. Removes all listeners of `evt` if omitted.
 * @returns ***REMOVED***util.EventEmitter***REMOVED*** `this`
 */
EventEmitter.prototype.off = function off(evt, fn) ***REMOVED***
    if (evt === undefined)
        this._listeners = ***REMOVED******REMOVED***;
    else ***REMOVED***
        if (fn === undefined)
            this._listeners[evt] = [];
        else ***REMOVED***
            var listeners = this._listeners[evt];
            for (var i = 0; i < listeners.length;)
                if (listeners[i].fn === fn)
                    listeners.splice(i, 1);
                else
                    ++i;
        ***REMOVED***
    ***REMOVED***
    return this;
***REMOVED***;

/**
 * Emits an event by calling its listeners with the specified arguments.
 * @param ***REMOVED***string***REMOVED*** evt Event name
 * @param ***REMOVED***...****REMOVED*** args Arguments
 * @returns ***REMOVED***util.EventEmitter***REMOVED*** `this`
 */
EventEmitter.prototype.emit = function emit(evt) ***REMOVED***
    var listeners = this._listeners[evt];
    if (listeners) ***REMOVED***
        var args = [],
            i = 1;
        for (; i < arguments.length;)
            args.push(arguments[i++]);
        for (i = 0; i < listeners.length;)
            listeners[i].fn.apply(listeners[i++].ctx, args);
    ***REMOVED***
    return this;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],4:[function(require,module,exports)***REMOVED***
"use strict";

module.exports = factory(factory);

/**
 * Reads / writes floats / doubles from / to buffers.
 * @name util.float
 * @namespace
 */

/**
 * Writes a 32 bit float to a buffer using little endian byte order.
 * @name util.float.writeFloatLE
 * @function
 * @param ***REMOVED***number***REMOVED*** val Value to write
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Target buffer
 * @param ***REMOVED***number***REMOVED*** pos Target buffer offset
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Writes a 32 bit float to a buffer using big endian byte order.
 * @name util.float.writeFloatBE
 * @function
 * @param ***REMOVED***number***REMOVED*** val Value to write
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Target buffer
 * @param ***REMOVED***number***REMOVED*** pos Target buffer offset
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Reads a 32 bit float from a buffer using little endian byte order.
 * @name util.float.readFloatLE
 * @function
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Source buffer
 * @param ***REMOVED***number***REMOVED*** pos Source buffer offset
 * @returns ***REMOVED***number***REMOVED*** Value read
 */

/**
 * Reads a 32 bit float from a buffer using big endian byte order.
 * @name util.float.readFloatBE
 * @function
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Source buffer
 * @param ***REMOVED***number***REMOVED*** pos Source buffer offset
 * @returns ***REMOVED***number***REMOVED*** Value read
 */

/**
 * Writes a 64 bit double to a buffer using little endian byte order.
 * @name util.float.writeDoubleLE
 * @function
 * @param ***REMOVED***number***REMOVED*** val Value to write
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Target buffer
 * @param ***REMOVED***number***REMOVED*** pos Target buffer offset
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Writes a 64 bit double to a buffer using big endian byte order.
 * @name util.float.writeDoubleBE
 * @function
 * @param ***REMOVED***number***REMOVED*** val Value to write
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Target buffer
 * @param ***REMOVED***number***REMOVED*** pos Target buffer offset
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Reads a 64 bit double from a buffer using little endian byte order.
 * @name util.float.readDoubleLE
 * @function
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Source buffer
 * @param ***REMOVED***number***REMOVED*** pos Source buffer offset
 * @returns ***REMOVED***number***REMOVED*** Value read
 */

/**
 * Reads a 64 bit double from a buffer using big endian byte order.
 * @name util.float.readDoubleBE
 * @function
 * @param ***REMOVED***Uint8Array***REMOVED*** buf Source buffer
 * @param ***REMOVED***number***REMOVED*** pos Source buffer offset
 * @returns ***REMOVED***number***REMOVED*** Value read
 */

// Factory function for the purpose of node-based testing in modified global environments
function factory(exports) ***REMOVED***

    // float: typed array
    if (typeof Float32Array !== "undefined") (function() ***REMOVED***

        var f32 = new Float32Array([ -0 ]),
            f8b = new Uint8Array(f32.buffer),
            le  = f8b[3] === 128;

        function writeFloat_f32_cpy(val, buf, pos) ***REMOVED***
            f32[0] = val;
            buf[pos    ] = f8b[0];
            buf[pos + 1] = f8b[1];
            buf[pos + 2] = f8b[2];
            buf[pos + 3] = f8b[3];
        ***REMOVED***

        function writeFloat_f32_rev(val, buf, pos) ***REMOVED***
            f32[0] = val;
            buf[pos    ] = f8b[3];
            buf[pos + 1] = f8b[2];
            buf[pos + 2] = f8b[1];
            buf[pos + 3] = f8b[0];
        ***REMOVED***

        /* istanbul ignore next */
        exports.writeFloatLE = le ? writeFloat_f32_cpy : writeFloat_f32_rev;
        /* istanbul ignore next */
        exports.writeFloatBE = le ? writeFloat_f32_rev : writeFloat_f32_cpy;

        function readFloat_f32_cpy(buf, pos) ***REMOVED***
            f8b[0] = buf[pos    ];
            f8b[1] = buf[pos + 1];
            f8b[2] = buf[pos + 2];
            f8b[3] = buf[pos + 3];
            return f32[0];
        ***REMOVED***

        function readFloat_f32_rev(buf, pos) ***REMOVED***
            f8b[3] = buf[pos    ];
            f8b[2] = buf[pos + 1];
            f8b[1] = buf[pos + 2];
            f8b[0] = buf[pos + 3];
            return f32[0];
        ***REMOVED***

        /* istanbul ignore next */
        exports.readFloatLE = le ? readFloat_f32_cpy : readFloat_f32_rev;
        /* istanbul ignore next */
        exports.readFloatBE = le ? readFloat_f32_rev : readFloat_f32_cpy;

    // float: ieee754
    ***REMOVED***)(); else (function() ***REMOVED***

        function writeFloat_ieee754(writeUint, val, buf, pos) ***REMOVED***
            var sign = val < 0 ? 1 : 0;
            if (sign)
                val = -val;
            if (val === 0)
                writeUint(1 / val > 0 ? /* positive */ 0 : /* negative 0 */ 2147483648, buf, pos);
            else if (isNaN(val))
                writeUint(2143289344, buf, pos);
            else if (val > 3.4028234663852886e+38) // +-Infinity
                writeUint((sign << 31 | 2139095040) >>> 0, buf, pos);
            else if (val < 1.1754943508222875e-38) // denormal
                writeUint((sign << 31 | Math.round(val / 1.401298464324817e-45)) >>> 0, buf, pos);
            else ***REMOVED***
                var exponent = Math.floor(Math.log(val) / Math.LN2),
                    mantissa = Math.round(val * Math.pow(2, -exponent) * 8388608) & 8388607;
                writeUint((sign << 31 | exponent + 127 << 23 | mantissa) >>> 0, buf, pos);
            ***REMOVED***
        ***REMOVED***

        exports.writeFloatLE = writeFloat_ieee754.bind(null, writeUintLE);
        exports.writeFloatBE = writeFloat_ieee754.bind(null, writeUintBE);

        function readFloat_ieee754(readUint, buf, pos) ***REMOVED***
            var uint = readUint(buf, pos),
                sign = (uint >> 31) * 2 + 1,
                exponent = uint >>> 23 & 255,
                mantissa = uint & 8388607;
            return exponent === 255
                ? mantissa
                ? NaN
                : sign * Infinity
                : exponent === 0 // denormal
                ? sign * 1.401298464324817e-45 * mantissa
                : sign * Math.pow(2, exponent - 150) * (mantissa + 8388608);
        ***REMOVED***

        exports.readFloatLE = readFloat_ieee754.bind(null, readUintLE);
        exports.readFloatBE = readFloat_ieee754.bind(null, readUintBE);

    ***REMOVED***)();

    // double: typed array
    if (typeof Float64Array !== "undefined") (function() ***REMOVED***

        var f64 = new Float64Array([-0]),
            f8b = new Uint8Array(f64.buffer),
            le  = f8b[7] === 128;

        function writeDouble_f64_cpy(val, buf, pos) ***REMOVED***
            f64[0] = val;
            buf[pos    ] = f8b[0];
            buf[pos + 1] = f8b[1];
            buf[pos + 2] = f8b[2];
            buf[pos + 3] = f8b[3];
            buf[pos + 4] = f8b[4];
            buf[pos + 5] = f8b[5];
            buf[pos + 6] = f8b[6];
            buf[pos + 7] = f8b[7];
        ***REMOVED***

        function writeDouble_f64_rev(val, buf, pos) ***REMOVED***
            f64[0] = val;
            buf[pos    ] = f8b[7];
            buf[pos + 1] = f8b[6];
            buf[pos + 2] = f8b[5];
            buf[pos + 3] = f8b[4];
            buf[pos + 4] = f8b[3];
            buf[pos + 5] = f8b[2];
            buf[pos + 6] = f8b[1];
            buf[pos + 7] = f8b[0];
        ***REMOVED***

        /* istanbul ignore next */
        exports.writeDoubleLE = le ? writeDouble_f64_cpy : writeDouble_f64_rev;
        /* istanbul ignore next */
        exports.writeDoubleBE = le ? writeDouble_f64_rev : writeDouble_f64_cpy;

        function readDouble_f64_cpy(buf, pos) ***REMOVED***
            f8b[0] = buf[pos    ];
            f8b[1] = buf[pos + 1];
            f8b[2] = buf[pos + 2];
            f8b[3] = buf[pos + 3];
            f8b[4] = buf[pos + 4];
            f8b[5] = buf[pos + 5];
            f8b[6] = buf[pos + 6];
            f8b[7] = buf[pos + 7];
            return f64[0];
        ***REMOVED***

        function readDouble_f64_rev(buf, pos) ***REMOVED***
            f8b[7] = buf[pos    ];
            f8b[6] = buf[pos + 1];
            f8b[5] = buf[pos + 2];
            f8b[4] = buf[pos + 3];
            f8b[3] = buf[pos + 4];
            f8b[2] = buf[pos + 5];
            f8b[1] = buf[pos + 6];
            f8b[0] = buf[pos + 7];
            return f64[0];
        ***REMOVED***

        /* istanbul ignore next */
        exports.readDoubleLE = le ? readDouble_f64_cpy : readDouble_f64_rev;
        /* istanbul ignore next */
        exports.readDoubleBE = le ? readDouble_f64_rev : readDouble_f64_cpy;

    // double: ieee754
    ***REMOVED***)(); else (function() ***REMOVED***

        function writeDouble_ieee754(writeUint, off0, off1, val, buf, pos) ***REMOVED***
            var sign = val < 0 ? 1 : 0;
            if (sign)
                val = -val;
            if (val === 0) ***REMOVED***
                writeUint(0, buf, pos + off0);
                writeUint(1 / val > 0 ? /* positive */ 0 : /* negative 0 */ 2147483648, buf, pos + off1);
            ***REMOVED*** else if (isNaN(val)) ***REMOVED***
                writeUint(0, buf, pos + off0);
                writeUint(2146959360, buf, pos + off1);
            ***REMOVED*** else if (val > 1.7976931348623157e+308) ***REMOVED*** // +-Infinity
                writeUint(0, buf, pos + off0);
                writeUint((sign << 31 | 2146435072) >>> 0, buf, pos + off1);
            ***REMOVED*** else ***REMOVED***
                var mantissa;
                if (val < 2.2250738585072014e-308) ***REMOVED*** // denormal
                    mantissa = val / 5e-324;
                    writeUint(mantissa >>> 0, buf, pos + off0);
                    writeUint((sign << 31 | mantissa / 4294967296) >>> 0, buf, pos + off1);
                ***REMOVED*** else ***REMOVED***
                    var exponent = Math.floor(Math.log(val) / Math.LN2);
                    if (exponent === 1024)
                        exponent = 1023;
                    mantissa = val * Math.pow(2, -exponent);
                    writeUint(mantissa * 4503599627370496 >>> 0, buf, pos + off0);
                    writeUint((sign << 31 | exponent + 1023 << 20 | mantissa * 1048576 & 1048575) >>> 0, buf, pos + off1);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        exports.writeDoubleLE = writeDouble_ieee754.bind(null, writeUintLE, 0, 4);
        exports.writeDoubleBE = writeDouble_ieee754.bind(null, writeUintBE, 4, 0);

        function readDouble_ieee754(readUint, off0, off1, buf, pos) ***REMOVED***
            var lo = readUint(buf, pos + off0),
                hi = readUint(buf, pos + off1);
            var sign = (hi >> 31) * 2 + 1,
                exponent = hi >>> 20 & 2047,
                mantissa = 4294967296 * (hi & 1048575) + lo;
            return exponent === 2047
                ? mantissa
                ? NaN
                : sign * Infinity
                : exponent === 0 // denormal
                ? sign * 5e-324 * mantissa
                : sign * Math.pow(2, exponent - 1075) * (mantissa + 4503599627370496);
        ***REMOVED***

        exports.readDoubleLE = readDouble_ieee754.bind(null, readUintLE, 0, 4);
        exports.readDoubleBE = readDouble_ieee754.bind(null, readUintBE, 4, 0);

    ***REMOVED***)();

    return exports;
***REMOVED***

// uint helpers

function writeUintLE(val, buf, pos) ***REMOVED***
    buf[pos    ] =  val        & 255;
    buf[pos + 1] =  val >>> 8  & 255;
    buf[pos + 2] =  val >>> 16 & 255;
    buf[pos + 3] =  val >>> 24;
***REMOVED***

function writeUintBE(val, buf, pos) ***REMOVED***
    buf[pos    ] =  val >>> 24;
    buf[pos + 1] =  val >>> 16 & 255;
    buf[pos + 2] =  val >>> 8  & 255;
    buf[pos + 3] =  val        & 255;
***REMOVED***

function readUintLE(buf, pos) ***REMOVED***
    return (buf[pos    ]
          | buf[pos + 1] << 8
          | buf[pos + 2] << 16
          | buf[pos + 3] << 24) >>> 0;
***REMOVED***

function readUintBE(buf, pos) ***REMOVED***
    return (buf[pos    ] << 24
          | buf[pos + 1] << 16
          | buf[pos + 2] << 8
          | buf[pos + 3]) >>> 0;
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],5:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = inquire;

/**
 * Requires a module only if available.
 * @memberof util
 * @param ***REMOVED***string***REMOVED*** moduleName Module to require
 * @returns ***REMOVED***?Object***REMOVED*** Required module if available and not empty, otherwise `null`
 */
function inquire(moduleName) ***REMOVED***
    try ***REMOVED***
        var mod = eval("quire".replace(/^/,"re"))(moduleName); // eslint-disable-line no-eval
        if (mod && (mod.length || Object.keys(mod).length))
            return mod;
    ***REMOVED*** catch (e) ***REMOVED******REMOVED*** // eslint-disable-line no-empty
    return null;
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],6:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = pool;

/**
 * An allocator as used by ***REMOVED***@link util.pool***REMOVED***.
 * @typedef PoolAllocator
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***number***REMOVED*** size Buffer size
 * @returns ***REMOVED***Uint8Array***REMOVED*** Buffer
 */

/**
 * A slicer as used by ***REMOVED***@link util.pool***REMOVED***.
 * @typedef PoolSlicer
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***number***REMOVED*** start Start offset
 * @param ***REMOVED***number***REMOVED*** end End offset
 * @returns ***REMOVED***Uint8Array***REMOVED*** Buffer slice
 * @this ***REMOVED***Uint8Array***REMOVED***
 */

/**
 * A general purpose buffer pool.
 * @memberof util
 * @function
 * @param ***REMOVED***PoolAllocator***REMOVED*** alloc Allocator
 * @param ***REMOVED***PoolSlicer***REMOVED*** slice Slicer
 * @param ***REMOVED***number***REMOVED*** [size=8192] Slab size
 * @returns ***REMOVED***PoolAllocator***REMOVED*** Pooled allocator
 */
function pool(alloc, slice, size) ***REMOVED***
    var SIZE   = size || 8192;
    var MAX    = SIZE >>> 1;
    var slab   = null;
    var offset = SIZE;
    return function pool_alloc(size) ***REMOVED***
        if (size < 1 || size > MAX)
            return alloc(size);
        if (offset + size > SIZE) ***REMOVED***
            slab = alloc(SIZE);
            offset = 0;
        ***REMOVED***
        var buf = slice.call(slab, offset, offset += size);
        if (offset & 7) // align to 32 bit
            offset = (offset | 7) + 1;
        return buf;
    ***REMOVED***;
***REMOVED***

***REMOVED***,***REMOVED******REMOVED***],7:[function(require,module,exports)***REMOVED***
"use strict";

/**
 * A minimal UTF8 implementation for number arrays.
 * @memberof util
 * @namespace
 */
var utf8 = exports;

/**
 * Calculates the UTF8 byte length of a string.
 * @param ***REMOVED***string***REMOVED*** string String
 * @returns ***REMOVED***number***REMOVED*** Byte length
 */
utf8.length = function utf8_length(string) ***REMOVED***
    var len = 0,
        c = 0;
    for (var i = 0; i < string.length; ++i) ***REMOVED***
        c = string.charCodeAt(i);
        if (c < 128)
            len += 1;
        else if (c < 2048)
            len += 2;
        else if ((c & 0xFC00) === 0xD800 && (string.charCodeAt(i + 1) & 0xFC00) === 0xDC00) ***REMOVED***
            ++i;
            len += 4;
        ***REMOVED*** else
            len += 3;
    ***REMOVED***
    return len;
***REMOVED***;

/**
 * Reads UTF8 bytes as a string.
 * @param ***REMOVED***Uint8Array***REMOVED*** buffer Source buffer
 * @param ***REMOVED***number***REMOVED*** start Source start
 * @param ***REMOVED***number***REMOVED*** end Source end
 * @returns ***REMOVED***string***REMOVED*** String read
 */
utf8.read = function utf8_read(buffer, start, end) ***REMOVED***
    var len = end - start;
    if (len < 1)
        return "";
    var parts = null,
        chunk = [],
        i = 0, // char offset
        t;     // temporary
    while (start < end) ***REMOVED***
        t = buffer[start++];
        if (t < 128)
            chunk[i++] = t;
        else if (t > 191 && t < 224)
            chunk[i++] = (t & 31) << 6 | buffer[start++] & 63;
        else if (t > 239 && t < 365) ***REMOVED***
            t = ((t & 7) << 18 | (buffer[start++] & 63) << 12 | (buffer[start++] & 63) << 6 | buffer[start++] & 63) - 0x10000;
            chunk[i++] = 0xD800 + (t >> 10);
            chunk[i++] = 0xDC00 + (t & 1023);
        ***REMOVED*** else
            chunk[i++] = (t & 15) << 12 | (buffer[start++] & 63) << 6 | buffer[start++] & 63;
        if (i > 8191) ***REMOVED***
            (parts || (parts = [])).push(String.fromCharCode.apply(String, chunk));
            i = 0;
        ***REMOVED***
    ***REMOVED***
    if (parts) ***REMOVED***
        if (i)
            parts.push(String.fromCharCode.apply(String, chunk.slice(0, i)));
        return parts.join("");
    ***REMOVED***
    return String.fromCharCode.apply(String, chunk.slice(0, i));
***REMOVED***;

/**
 * Writes a string as UTF8 bytes.
 * @param ***REMOVED***string***REMOVED*** string Source string
 * @param ***REMOVED***Uint8Array***REMOVED*** buffer Destination buffer
 * @param ***REMOVED***number***REMOVED*** offset Destination offset
 * @returns ***REMOVED***number***REMOVED*** Bytes written
 */
utf8.write = function utf8_write(string, buffer, offset) ***REMOVED***
    var start = offset,
        c1, // character 1
        c2; // character 2
    for (var i = 0; i < string.length; ++i) ***REMOVED***
        c1 = string.charCodeAt(i);
        if (c1 < 128) ***REMOVED***
            buffer[offset++] = c1;
        ***REMOVED*** else if (c1 < 2048) ***REMOVED***
            buffer[offset++] = c1 >> 6       | 192;
            buffer[offset++] = c1       & 63 | 128;
        ***REMOVED*** else if ((c1 & 0xFC00) === 0xD800 && ((c2 = string.charCodeAt(i + 1)) & 0xFC00) === 0xDC00) ***REMOVED***
            c1 = 0x10000 + ((c1 & 0x03FF) << 10) + (c2 & 0x03FF);
            ++i;
            buffer[offset++] = c1 >> 18      | 240;
            buffer[offset++] = c1 >> 12 & 63 | 128;
            buffer[offset++] = c1 >> 6  & 63 | 128;
            buffer[offset++] = c1       & 63 | 128;
        ***REMOVED*** else ***REMOVED***
            buffer[offset++] = c1 >> 12      | 224;
            buffer[offset++] = c1 >> 6  & 63 | 128;
            buffer[offset++] = c1       & 63 | 128;
        ***REMOVED***
    ***REMOVED***
    return offset - start;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],8:[function(require,module,exports)***REMOVED***
"use strict";
var protobuf = exports;

/**
 * Build type, one of `"full"`, `"light"` or `"minimal"`.
 * @name build
 * @type ***REMOVED***string***REMOVED***
 * @const
 */
protobuf.build = "minimal";

// Serialization
protobuf.Writer       = require(16);
protobuf.BufferWriter = require(17);
protobuf.Reader       = require(9);
protobuf.BufferReader = require(10);

// Utility
protobuf.util         = require(15);
protobuf.rpc          = require(12);
protobuf.roots        = require(11);
protobuf.configure    = configure;

/* istanbul ignore next */
/**
 * Reconfigures the library according to the environment.
 * @returns ***REMOVED***undefined***REMOVED***
 */
function configure() ***REMOVED***
    protobuf.Reader._configure(protobuf.BufferReader);
    protobuf.util._configure();
***REMOVED***

// Configure serialization
protobuf.Writer._configure(protobuf.BufferWriter);
configure();

***REMOVED***,***REMOVED***"10":10,"11":11,"12":12,"15":15,"16":16,"17":17,"9":9***REMOVED***],9:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = Reader;

var util      = require(15);

var BufferReader; // cyclic

var LongBits  = util.LongBits,
    utf8      = util.utf8;

/* istanbul ignore next */
function indexOutOfRange(reader, writeLength) ***REMOVED***
    return RangeError("index out of range: " + reader.pos + " + " + (writeLength || 1) + " > " + reader.len);
***REMOVED***

/**
 * Constructs a new reader instance using the specified buffer.
 * @classdesc Wire format reader using `Uint8Array` if available, otherwise `Array`.
 * @constructor
 * @param ***REMOVED***Uint8Array***REMOVED*** buffer Buffer to read from
 */
function Reader(buffer) ***REMOVED***

    /**
     * Read buffer.
     * @type ***REMOVED***Uint8Array***REMOVED***
     */
    this.buf = buffer;

    /**
     * Read buffer position.
     * @type ***REMOVED***number***REMOVED***
     */
    this.pos = 0;

    /**
     * Read buffer length.
     * @type ***REMOVED***number***REMOVED***
     */
    this.len = buffer.length;
***REMOVED***

var create_array = typeof Uint8Array !== "undefined"
    ? function create_typed_array(buffer) ***REMOVED***
        if (buffer instanceof Uint8Array || Array.isArray(buffer))
            return new Reader(buffer);
        throw Error("illegal buffer");
    ***REMOVED***
    /* istanbul ignore next */
    : function create_array(buffer) ***REMOVED***
        if (Array.isArray(buffer))
            return new Reader(buffer);
        throw Error("illegal buffer");
    ***REMOVED***;

/**
 * Creates a new reader using the specified buffer.
 * @function
 * @param ***REMOVED***Uint8Array|Buffer***REMOVED*** buffer Buffer to read from
 * @returns ***REMOVED***Reader|BufferReader***REMOVED*** A ***REMOVED***@link BufferReader***REMOVED*** if `buffer` is a Buffer, otherwise a ***REMOVED***@link Reader***REMOVED***
 * @throws ***REMOVED***Error***REMOVED*** If `buffer` is not a valid buffer
 */
Reader.create = util.Buffer
    ? function create_buffer_setup(buffer) ***REMOVED***
        return (Reader.create = function create_buffer(buffer) ***REMOVED***
            return util.Buffer.isBuffer(buffer)
                ? new BufferReader(buffer)
                /* istanbul ignore next */
                : create_array(buffer);
        ***REMOVED***)(buffer);
    ***REMOVED***
    /* istanbul ignore next */
    : create_array;

Reader.prototype._slice = util.Array.prototype.subarray || /* istanbul ignore next */ util.Array.prototype.slice;

/**
 * Reads a varint as an unsigned 32 bit value.
 * @function
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.uint32 = (function read_uint32_setup() ***REMOVED***
    var value = 4294967295; // optimizer type-hint, tends to deopt otherwise (?!)
    return function read_uint32() ***REMOVED***
        value = (         this.buf[this.pos] & 127       ) >>> 0; if (this.buf[this.pos++] < 128) return value;
        value = (value | (this.buf[this.pos] & 127) <<  7) >>> 0; if (this.buf[this.pos++] < 128) return value;
        value = (value | (this.buf[this.pos] & 127) << 14) >>> 0; if (this.buf[this.pos++] < 128) return value;
        value = (value | (this.buf[this.pos] & 127) << 21) >>> 0; if (this.buf[this.pos++] < 128) return value;
        value = (value | (this.buf[this.pos] &  15) << 28) >>> 0; if (this.buf[this.pos++] < 128) return value;

        /* istanbul ignore if */
        if ((this.pos += 5) > this.len) ***REMOVED***
            this.pos = this.len;
            throw indexOutOfRange(this, 10);
        ***REMOVED***
        return value;
    ***REMOVED***;
***REMOVED***)();

/**
 * Reads a varint as a signed 32 bit value.
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.int32 = function read_int32() ***REMOVED***
    return this.uint32() | 0;
***REMOVED***;

/**
 * Reads a zig-zag encoded varint as a signed 32 bit value.
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.sint32 = function read_sint32() ***REMOVED***
    var value = this.uint32();
    return value >>> 1 ^ -(value & 1) | 0;
***REMOVED***;

/* eslint-disable no-invalid-this */

function readLongVarint() ***REMOVED***
    // tends to deopt with local vars for octet etc.
    var bits = new LongBits(0, 0);
    var i = 0;
    if (this.len - this.pos > 4) ***REMOVED*** // fast route (lo)
        for (; i < 4; ++i) ***REMOVED***
            // 1st..4th
            bits.lo = (bits.lo | (this.buf[this.pos] & 127) << i * 7) >>> 0;
            if (this.buf[this.pos++] < 128)
                return bits;
        ***REMOVED***
        // 5th
        bits.lo = (bits.lo | (this.buf[this.pos] & 127) << 28) >>> 0;
        bits.hi = (bits.hi | (this.buf[this.pos] & 127) >>  4) >>> 0;
        if (this.buf[this.pos++] < 128)
            return bits;
        i = 0;
    ***REMOVED*** else ***REMOVED***
        for (; i < 3; ++i) ***REMOVED***
            /* istanbul ignore if */
            if (this.pos >= this.len)
                throw indexOutOfRange(this);
            // 1st..3th
            bits.lo = (bits.lo | (this.buf[this.pos] & 127) << i * 7) >>> 0;
            if (this.buf[this.pos++] < 128)
                return bits;
        ***REMOVED***
        // 4th
        bits.lo = (bits.lo | (this.buf[this.pos++] & 127) << i * 7) >>> 0;
        return bits;
    ***REMOVED***
    if (this.len - this.pos > 4) ***REMOVED*** // fast route (hi)
        for (; i < 5; ++i) ***REMOVED***
            // 6th..10th
            bits.hi = (bits.hi | (this.buf[this.pos] & 127) << i * 7 + 3) >>> 0;
            if (this.buf[this.pos++] < 128)
                return bits;
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        for (; i < 5; ++i) ***REMOVED***
            /* istanbul ignore if */
            if (this.pos >= this.len)
                throw indexOutOfRange(this);
            // 6th..10th
            bits.hi = (bits.hi | (this.buf[this.pos] & 127) << i * 7 + 3) >>> 0;
            if (this.buf[this.pos++] < 128)
                return bits;
        ***REMOVED***
    ***REMOVED***
    /* istanbul ignore next */
    throw Error("invalid varint encoding");
***REMOVED***

/* eslint-enable no-invalid-this */

/**
 * Reads a varint as a signed 64 bit value.
 * @name Reader#int64
 * @function
 * @returns ***REMOVED***Long***REMOVED*** Value read
 */

/**
 * Reads a varint as an unsigned 64 bit value.
 * @name Reader#uint64
 * @function
 * @returns ***REMOVED***Long***REMOVED*** Value read
 */

/**
 * Reads a zig-zag encoded varint as a signed 64 bit value.
 * @name Reader#sint64
 * @function
 * @returns ***REMOVED***Long***REMOVED*** Value read
 */

/**
 * Reads a varint as a boolean.
 * @returns ***REMOVED***boolean***REMOVED*** Value read
 */
Reader.prototype.bool = function read_bool() ***REMOVED***
    return this.uint32() !== 0;
***REMOVED***;

function readFixed32_end(buf, end) ***REMOVED*** // note that this uses `end`, not `pos`
    return (buf[end - 4]
          | buf[end - 3] << 8
          | buf[end - 2] << 16
          | buf[end - 1] << 24) >>> 0;
***REMOVED***

/**
 * Reads fixed 32 bits as an unsigned 32 bit integer.
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.fixed32 = function read_fixed32() ***REMOVED***

    /* istanbul ignore if */
    if (this.pos + 4 > this.len)
        throw indexOutOfRange(this, 4);

    return readFixed32_end(this.buf, this.pos += 4);
***REMOVED***;

/**
 * Reads fixed 32 bits as a signed 32 bit integer.
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.sfixed32 = function read_sfixed32() ***REMOVED***

    /* istanbul ignore if */
    if (this.pos + 4 > this.len)
        throw indexOutOfRange(this, 4);

    return readFixed32_end(this.buf, this.pos += 4) | 0;
***REMOVED***;

/* eslint-disable no-invalid-this */

function readFixed64(/* this: Reader */) ***REMOVED***

    /* istanbul ignore if */
    if (this.pos + 8 > this.len)
        throw indexOutOfRange(this, 8);

    return new LongBits(readFixed32_end(this.buf, this.pos += 4), readFixed32_end(this.buf, this.pos += 4));
***REMOVED***

/* eslint-enable no-invalid-this */

/**
 * Reads fixed 64 bits.
 * @name Reader#fixed64
 * @function
 * @returns ***REMOVED***Long***REMOVED*** Value read
 */

/**
 * Reads zig-zag encoded fixed 64 bits.
 * @name Reader#sfixed64
 * @function
 * @returns ***REMOVED***Long***REMOVED*** Value read
 */

/**
 * Reads a float (32 bit) as a number.
 * @function
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.float = function read_float() ***REMOVED***

    /* istanbul ignore if */
    if (this.pos + 4 > this.len)
        throw indexOutOfRange(this, 4);

    var value = util.float.readFloatLE(this.buf, this.pos);
    this.pos += 4;
    return value;
***REMOVED***;

/**
 * Reads a double (64 bit float) as a number.
 * @function
 * @returns ***REMOVED***number***REMOVED*** Value read
 */
Reader.prototype.double = function read_double() ***REMOVED***

    /* istanbul ignore if */
    if (this.pos + 8 > this.len)
        throw indexOutOfRange(this, 4);

    var value = util.float.readDoubleLE(this.buf, this.pos);
    this.pos += 8;
    return value;
***REMOVED***;

/**
 * Reads a sequence of bytes preceeded by its length as a varint.
 * @returns ***REMOVED***Uint8Array***REMOVED*** Value read
 */
Reader.prototype.bytes = function read_bytes() ***REMOVED***
    var length = this.uint32(),
        start  = this.pos,
        end    = this.pos + length;

    /* istanbul ignore if */
    if (end > this.len)
        throw indexOutOfRange(this, length);

    this.pos += length;
    if (Array.isArray(this.buf)) // plain array
        return this.buf.slice(start, end);
    return start === end // fix for IE 10/Win8 and others' subarray returning array of size 1
        ? new this.buf.constructor(0)
        : this._slice.call(this.buf, start, end);
***REMOVED***;

/**
 * Reads a string preceeded by its byte length as a varint.
 * @returns ***REMOVED***string***REMOVED*** Value read
 */
Reader.prototype.string = function read_string() ***REMOVED***
    var bytes = this.bytes();
    return utf8.read(bytes, 0, bytes.length);
***REMOVED***;

/**
 * Skips the specified number of bytes if specified, otherwise skips a varint.
 * @param ***REMOVED***number***REMOVED*** [length] Length if known, otherwise a varint is assumed
 * @returns ***REMOVED***Reader***REMOVED*** `this`
 */
Reader.prototype.skip = function skip(length) ***REMOVED***
    if (typeof length === "number") ***REMOVED***
        /* istanbul ignore if */
        if (this.pos + length > this.len)
            throw indexOutOfRange(this, length);
        this.pos += length;
    ***REMOVED*** else ***REMOVED***
        do ***REMOVED***
            /* istanbul ignore if */
            if (this.pos >= this.len)
                throw indexOutOfRange(this);
        ***REMOVED*** while (this.buf[this.pos++] & 128);
    ***REMOVED***
    return this;
***REMOVED***;

/**
 * Skips the next element of the specified wire type.
 * @param ***REMOVED***number***REMOVED*** wireType Wire type received
 * @returns ***REMOVED***Reader***REMOVED*** `this`
 */
Reader.prototype.skipType = function(wireType) ***REMOVED***
    switch (wireType) ***REMOVED***
        case 0:
            this.skip();
            break;
        case 1:
            this.skip(8);
            break;
        case 2:
            this.skip(this.uint32());
            break;
        case 3:
            do ***REMOVED*** // eslint-disable-line no-constant-condition
                if ((wireType = this.uint32() & 7) === 4)
                    break;
                this.skipType(wireType);
            ***REMOVED*** while (true);
            break;
        case 5:
            this.skip(4);
            break;

        /* istanbul ignore next */
        default:
            throw Error("invalid wire type " + wireType + " at offset " + this.pos);
    ***REMOVED***
    return this;
***REMOVED***;

Reader._configure = function(BufferReader_) ***REMOVED***
    BufferReader = BufferReader_;

    var fn = util.Long ? "toLong" : /* istanbul ignore next */ "toNumber";
    util.merge(Reader.prototype, ***REMOVED***

        int64: function read_int64() ***REMOVED***
            return readLongVarint.call(this)[fn](false);
        ***REMOVED***,

        uint64: function read_uint64() ***REMOVED***
            return readLongVarint.call(this)[fn](true);
        ***REMOVED***,

        sint64: function read_sint64() ***REMOVED***
            return readLongVarint.call(this).zzDecode()[fn](false);
        ***REMOVED***,

        fixed64: function read_fixed64() ***REMOVED***
            return readFixed64.call(this)[fn](true);
        ***REMOVED***,

        sfixed64: function read_sfixed64() ***REMOVED***
            return readFixed64.call(this)[fn](false);
        ***REMOVED***

    ***REMOVED***);
***REMOVED***;

***REMOVED***,***REMOVED***"15":15***REMOVED***],10:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = BufferReader;

// extends Reader
var Reader = require(9);
(BufferReader.prototype = Object.create(Reader.prototype)).constructor = BufferReader;

var util = require(15);

/**
 * Constructs a new buffer reader instance.
 * @classdesc Wire format reader using node buffers.
 * @extends Reader
 * @constructor
 * @param ***REMOVED***Buffer***REMOVED*** buffer Buffer to read from
 */
function BufferReader(buffer) ***REMOVED***
    Reader.call(this, buffer);

    /**
     * Read buffer.
     * @name BufferReader#buf
     * @type ***REMOVED***Buffer***REMOVED***
     */
***REMOVED***

/* istanbul ignore else */
if (util.Buffer)
    BufferReader.prototype._slice = util.Buffer.prototype.slice;

/**
 * @override
 */
BufferReader.prototype.string = function read_string_buffer() ***REMOVED***
    var len = this.uint32(); // modifies pos
    return this.buf.utf8Slice(this.pos, this.pos = Math.min(this.pos + len, this.len));
***REMOVED***;

/**
 * Reads a sequence of bytes preceeded by its length as a varint.
 * @name BufferReader#bytes
 * @function
 * @returns ***REMOVED***Buffer***REMOVED*** Value read
 */

***REMOVED***,***REMOVED***"15":15,"9":9***REMOVED***],11:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = ***REMOVED******REMOVED***;

/**
 * Named roots.
 * This is where pbjs stores generated structures (the option `-r, --root` specifies a name).
 * Can also be used manually to make roots available accross modules.
 * @name roots
 * @type ***REMOVED***Object.<string,Root>***REMOVED***
 * @example
 * // pbjs -r myroot -o compiled.js ...
 *
 * // in another module:
 * require("./compiled.js");
 *
 * // in any subsequent module:
 * var root = protobuf.roots["myroot"];
 */

***REMOVED***,***REMOVED******REMOVED***],12:[function(require,module,exports)***REMOVED***
"use strict";

/**
 * Streaming RPC helpers.
 * @namespace
 */
var rpc = exports;

/**
 * RPC implementation passed to ***REMOVED***@link Service#create***REMOVED*** performing a service request on network level, i.e. by utilizing http requests or websockets.
 * @typedef RPCImpl
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***Method|rpc.ServiceMethod<Message<***REMOVED******REMOVED***>,Message<***REMOVED******REMOVED***>>***REMOVED*** method Reflected or static method being called
 * @param ***REMOVED***Uint8Array***REMOVED*** requestData Request data
 * @param ***REMOVED***RPCImplCallback***REMOVED*** callback Callback function
 * @returns ***REMOVED***undefined***REMOVED***
 * @example
 * function rpcImpl(method, requestData, callback) ***REMOVED***
 *     if (protobuf.util.lcFirst(method.name) !== "myMethod") // compatible with static code
 *         throw Error("no such method");
 *     asynchronouslyObtainAResponse(requestData, function(err, responseData) ***REMOVED***
 *         callback(err, responseData);
 *     ***REMOVED***);
 * ***REMOVED***
 */

/**
 * Node-style callback as used by ***REMOVED***@link RPCImpl***REMOVED***.
 * @typedef RPCImplCallback
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***Error|null***REMOVED*** error Error, if any, otherwise `null`
 * @param ***REMOVED***Uint8Array|null***REMOVED*** [response] Response data or `null` to signal end of stream, if there hasn't been an error
 * @returns ***REMOVED***undefined***REMOVED***
 */

rpc.Service = require(13);

***REMOVED***,***REMOVED***"13":13***REMOVED***],13:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = Service;

var util = require(15);

// Extends EventEmitter
(Service.prototype = Object.create(util.EventEmitter.prototype)).constructor = Service;

/**
 * A service method callback as used by ***REMOVED***@link rpc.ServiceMethod|ServiceMethod***REMOVED***.
 *
 * Differs from ***REMOVED***@link RPCImplCallback***REMOVED*** in that it is an actual callback of a service method which may not return `response = null`.
 * @typedef rpc.ServiceMethodCallback
 * @template TRes extends Message<TRes>
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***Error|null***REMOVED*** error Error, if any
 * @param ***REMOVED***TRes***REMOVED*** [response] Response message
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * A service method part of a ***REMOVED***@link rpc.Service***REMOVED*** as created by ***REMOVED***@link Service.create***REMOVED***.
 * @typedef rpc.ServiceMethod
 * @template TReq extends Message<TReq>
 * @template TRes extends Message<TRes>
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***TReq|Properties<TReq>***REMOVED*** request Request message or plain object
 * @param ***REMOVED***rpc.ServiceMethodCallback<TRes>***REMOVED*** [callback] Node-style callback called with the error, if any, and the response message
 * @returns ***REMOVED***Promise<Message<TRes>>***REMOVED*** Promise if `callback` has been omitted, otherwise `undefined`
 */

/**
 * Constructs a new RPC service instance.
 * @classdesc An RPC service as returned by ***REMOVED***@link Service#create***REMOVED***.
 * @exports rpc.Service
 * @extends util.EventEmitter
 * @constructor
 * @param ***REMOVED***RPCImpl***REMOVED*** rpcImpl RPC implementation
 * @param ***REMOVED***boolean***REMOVED*** [requestDelimited=false] Whether requests are length-delimited
 * @param ***REMOVED***boolean***REMOVED*** [responseDelimited=false] Whether responses are length-delimited
 */
function Service(rpcImpl, requestDelimited, responseDelimited) ***REMOVED***

    if (typeof rpcImpl !== "function")
        throw TypeError("rpcImpl must be a function");

    util.EventEmitter.call(this);

    /**
     * RPC implementation. Becomes `null` once the service is ended.
     * @type ***REMOVED***RPCImpl|null***REMOVED***
     */
    this.rpcImpl = rpcImpl;

    /**
     * Whether requests are length-delimited.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.requestDelimited = Boolean(requestDelimited);

    /**
     * Whether responses are length-delimited.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.responseDelimited = Boolean(responseDelimited);
***REMOVED***

/**
 * Calls a service method through ***REMOVED***@link rpc.Service#rpcImpl|rpcImpl***REMOVED***.
 * @param ***REMOVED***Method|rpc.ServiceMethod<TReq,TRes>***REMOVED*** method Reflected or static method
 * @param ***REMOVED***Constructor<TReq>***REMOVED*** requestCtor Request constructor
 * @param ***REMOVED***Constructor<TRes>***REMOVED*** responseCtor Response constructor
 * @param ***REMOVED***TReq|Properties<TReq>***REMOVED*** request Request message or plain object
 * @param ***REMOVED***rpc.ServiceMethodCallback<TRes>***REMOVED*** callback Service callback
 * @returns ***REMOVED***undefined***REMOVED***
 * @template TReq extends Message<TReq>
 * @template TRes extends Message<TRes>
 */
Service.prototype.rpcCall = function rpcCall(method, requestCtor, responseCtor, request, callback) ***REMOVED***

    if (!request)
        throw TypeError("request must be specified");

    var self = this;
    if (!callback)
        return util.asPromise(rpcCall, self, method, requestCtor, responseCtor, request);

    if (!self.rpcImpl) ***REMOVED***
        setTimeout(function() ***REMOVED*** callback(Error("already ended")); ***REMOVED***, 0);
        return undefined;
    ***REMOVED***

    try ***REMOVED***
        return self.rpcImpl(
            method,
            requestCtor[self.requestDelimited ? "encodeDelimited" : "encode"](request).finish(),
            function rpcCallback(err, response) ***REMOVED***

                if (err) ***REMOVED***
                    self.emit("error", err, method);
                    return callback(err);
                ***REMOVED***

                if (response === null) ***REMOVED***
                    self.end(/* endedByRPC */ true);
                    return undefined;
                ***REMOVED***

                if (!(response instanceof responseCtor)) ***REMOVED***
                    try ***REMOVED***
                        response = responseCtor[self.responseDelimited ? "decodeDelimited" : "decode"](response);
                    ***REMOVED*** catch (err) ***REMOVED***
                        self.emit("error", err, method);
                        return callback(err);
                    ***REMOVED***
                ***REMOVED***

                self.emit("data", response, method);
                return callback(null, response);
            ***REMOVED***
        );
    ***REMOVED*** catch (err) ***REMOVED***
        self.emit("error", err, method);
        setTimeout(function() ***REMOVED*** callback(err); ***REMOVED***, 0);
        return undefined;
    ***REMOVED***
***REMOVED***;

/**
 * Ends this service and emits the `end` event.
 * @param ***REMOVED***boolean***REMOVED*** [endedByRPC=false] Whether the service has been ended by the RPC implementation.
 * @returns ***REMOVED***rpc.Service***REMOVED*** `this`
 */
Service.prototype.end = function end(endedByRPC) ***REMOVED***
    if (this.rpcImpl) ***REMOVED***
        if (!endedByRPC) // signal end to rpcImpl
            this.rpcImpl(null, null, null);
        this.rpcImpl = null;
        this.emit("end").off();
    ***REMOVED***
    return this;
***REMOVED***;

***REMOVED***,***REMOVED***"15":15***REMOVED***],14:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = LongBits;

var util = require(15);

/**
 * Constructs new long bits.
 * @classdesc Helper class for working with the low and high bits of a 64 bit value.
 * @memberof util
 * @constructor
 * @param ***REMOVED***number***REMOVED*** lo Low 32 bits, unsigned
 * @param ***REMOVED***number***REMOVED*** hi High 32 bits, unsigned
 */
function LongBits(lo, hi) ***REMOVED***

    // note that the casts below are theoretically unnecessary as of today, but older statically
    // generated converter code might still call the ctor with signed 32bits. kept for compat.

    /**
     * Low bits.
     * @type ***REMOVED***number***REMOVED***
     */
    this.lo = lo >>> 0;

    /**
     * High bits.
     * @type ***REMOVED***number***REMOVED***
     */
    this.hi = hi >>> 0;
***REMOVED***

/**
 * Zero bits.
 * @memberof util.LongBits
 * @type ***REMOVED***util.LongBits***REMOVED***
 */
var zero = LongBits.zero = new LongBits(0, 0);

zero.toNumber = function() ***REMOVED*** return 0; ***REMOVED***;
zero.zzEncode = zero.zzDecode = function() ***REMOVED*** return this; ***REMOVED***;
zero.length = function() ***REMOVED*** return 1; ***REMOVED***;

/**
 * Zero hash.
 * @memberof util.LongBits
 * @type ***REMOVED***string***REMOVED***
 */
var zeroHash = LongBits.zeroHash = "\0\0\0\0\0\0\0\0";

/**
 * Constructs new long bits from the specified number.
 * @param ***REMOVED***number***REMOVED*** value Value
 * @returns ***REMOVED***util.LongBits***REMOVED*** Instance
 */
LongBits.fromNumber = function fromNumber(value) ***REMOVED***
    if (value === 0)
        return zero;
    var sign = value < 0;
    if (sign)
        value = -value;
    var lo = value >>> 0,
        hi = (value - lo) / 4294967296 >>> 0;
    if (sign) ***REMOVED***
        hi = ~hi >>> 0;
        lo = ~lo >>> 0;
        if (++lo > 4294967295) ***REMOVED***
            lo = 0;
            if (++hi > 4294967295)
                hi = 0;
        ***REMOVED***
    ***REMOVED***
    return new LongBits(lo, hi);
***REMOVED***;

/**
 * Constructs new long bits from a number, long or string.
 * @param ***REMOVED***Long|number|string***REMOVED*** value Value
 * @returns ***REMOVED***util.LongBits***REMOVED*** Instance
 */
LongBits.from = function from(value) ***REMOVED***
    if (typeof value === "number")
        return LongBits.fromNumber(value);
    if (util.isString(value)) ***REMOVED***
        /* istanbul ignore else */
        if (util.Long)
            value = util.Long.fromString(value);
        else
            return LongBits.fromNumber(parseInt(value, 10));
    ***REMOVED***
    return value.low || value.high ? new LongBits(value.low >>> 0, value.high >>> 0) : zero;
***REMOVED***;

/**
 * Converts this long bits to a possibly unsafe JavaScript number.
 * @param ***REMOVED***boolean***REMOVED*** [unsigned=false] Whether unsigned or not
 * @returns ***REMOVED***number***REMOVED*** Possibly unsafe number
 */
LongBits.prototype.toNumber = function toNumber(unsigned) ***REMOVED***
    if (!unsigned && this.hi >>> 31) ***REMOVED***
        var lo = ~this.lo + 1 >>> 0,
            hi = ~this.hi     >>> 0;
        if (!lo)
            hi = hi + 1 >>> 0;
        return -(lo + hi * 4294967296);
    ***REMOVED***
    return this.lo + this.hi * 4294967296;
***REMOVED***;

/**
 * Converts this long bits to a long.
 * @param ***REMOVED***boolean***REMOVED*** [unsigned=false] Whether unsigned or not
 * @returns ***REMOVED***Long***REMOVED*** Long
 */
LongBits.prototype.toLong = function toLong(unsigned) ***REMOVED***
    return util.Long
        ? new util.Long(this.lo | 0, this.hi | 0, Boolean(unsigned))
        /* istanbul ignore next */
        : ***REMOVED*** low: this.lo | 0, high: this.hi | 0, unsigned: Boolean(unsigned) ***REMOVED***;
***REMOVED***;

var charCodeAt = String.prototype.charCodeAt;

/**
 * Constructs new long bits from the specified 8 characters long hash.
 * @param ***REMOVED***string***REMOVED*** hash Hash
 * @returns ***REMOVED***util.LongBits***REMOVED*** Bits
 */
LongBits.fromHash = function fromHash(hash) ***REMOVED***
    if (hash === zeroHash)
        return zero;
    return new LongBits(
        ( charCodeAt.call(hash, 0)
        | charCodeAt.call(hash, 1) << 8
        | charCodeAt.call(hash, 2) << 16
        | charCodeAt.call(hash, 3) << 24) >>> 0
    ,
        ( charCodeAt.call(hash, 4)
        | charCodeAt.call(hash, 5) << 8
        | charCodeAt.call(hash, 6) << 16
        | charCodeAt.call(hash, 7) << 24) >>> 0
    );
***REMOVED***;

/**
 * Converts this long bits to a 8 characters long hash.
 * @returns ***REMOVED***string***REMOVED*** Hash
 */
LongBits.prototype.toHash = function toHash() ***REMOVED***
    return String.fromCharCode(
        this.lo        & 255,
        this.lo >>> 8  & 255,
        this.lo >>> 16 & 255,
        this.lo >>> 24      ,
        this.hi        & 255,
        this.hi >>> 8  & 255,
        this.hi >>> 16 & 255,
        this.hi >>> 24
    );
***REMOVED***;

/**
 * Zig-zag encodes this long bits.
 * @returns ***REMOVED***util.LongBits***REMOVED*** `this`
 */
LongBits.prototype.zzEncode = function zzEncode() ***REMOVED***
    var mask =   this.hi >> 31;
    this.hi  = ((this.hi << 1 | this.lo >>> 31) ^ mask) >>> 0;
    this.lo  = ( this.lo << 1                   ^ mask) >>> 0;
    return this;
***REMOVED***;

/**
 * Zig-zag decodes this long bits.
 * @returns ***REMOVED***util.LongBits***REMOVED*** `this`
 */
LongBits.prototype.zzDecode = function zzDecode() ***REMOVED***
    var mask = -(this.lo & 1);
    this.lo  = ((this.lo >>> 1 | this.hi << 31) ^ mask) >>> 0;
    this.hi  = ( this.hi >>> 1                  ^ mask) >>> 0;
    return this;
***REMOVED***;

/**
 * Calculates the length of this longbits when encoded as a varint.
 * @returns ***REMOVED***number***REMOVED*** Length
 */
LongBits.prototype.length = function length() ***REMOVED***
    var part0 =  this.lo,
        part1 = (this.lo >>> 28 | this.hi << 4) >>> 0,
        part2 =  this.hi >>> 24;
    return part2 === 0
         ? part1 === 0
           ? part0 < 16384
             ? part0 < 128 ? 1 : 2
             : part0 < 2097152 ? 3 : 4
           : part1 < 16384
             ? part1 < 128 ? 5 : 6
             : part1 < 2097152 ? 7 : 8
         : part2 < 128 ? 9 : 10;
***REMOVED***;

***REMOVED***,***REMOVED***"15":15***REMOVED***],15:[function(require,module,exports)***REMOVED***
"use strict";
var util = exports;

// used to return a Promise where callback is omitted
util.asPromise = require(1);

// converts to / from base64 encoded strings
util.base64 = require(2);

// base class of rpc.Service
util.EventEmitter = require(3);

// float handling accross browsers
util.float = require(4);

// requires modules optionally and hides the call from bundlers
util.inquire = require(5);

// converts to / from utf8 encoded strings
util.utf8 = require(7);

// provides a node-like buffer pool in the browser
util.pool = require(6);

// utility to work with the low and high bits of a 64 bit value
util.LongBits = require(14);

/**
 * An immuable empty array.
 * @memberof util
 * @type ***REMOVED***Array.<*>***REMOVED***
 * @const
 */
util.emptyArray = Object.freeze ? Object.freeze([]) : /* istanbul ignore next */ []; // used on prototypes

/**
 * An immutable empty object.
 * @type ***REMOVED***Object***REMOVED***
 * @const
 */
util.emptyObject = Object.freeze ? Object.freeze(***REMOVED******REMOVED***) : /* istanbul ignore next */ ***REMOVED******REMOVED***; // used on prototypes

/**
 * Whether running within node or not.
 * @memberof util
 * @type ***REMOVED***boolean***REMOVED***
 * @const
 */
util.isNode = Boolean(global.process && global.process.versions && global.process.versions.node);

/**
 * Tests if the specified value is an integer.
 * @function
 * @param ***REMOVED*******REMOVED*** value Value to test
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the value is an integer
 */
util.isInteger = Number.isInteger || /* istanbul ignore next */ function isInteger(value) ***REMOVED***
    return typeof value === "number" && isFinite(value) && Math.floor(value) === value;
***REMOVED***;

/**
 * Tests if the specified value is a string.
 * @param ***REMOVED*******REMOVED*** value Value to test
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the value is a string
 */
util.isString = function isString(value) ***REMOVED***
    return typeof value === "string" || value instanceof String;
***REMOVED***;

/**
 * Tests if the specified value is a non-null object.
 * @param ***REMOVED*******REMOVED*** value Value to test
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the value is a non-null object
 */
util.isObject = function isObject(value) ***REMOVED***
    return value && typeof value === "object";
***REMOVED***;

/**
 * Checks if a property on a message is considered to be present.
 * This is an alias of ***REMOVED***@link util.isSet***REMOVED***.
 * @function
 * @param ***REMOVED***Object***REMOVED*** obj Plain object or message instance
 * @param ***REMOVED***string***REMOVED*** prop Property name
 * @returns ***REMOVED***boolean***REMOVED*** `true` if considered to be present, otherwise `false`
 */
util.isset =

/**
 * Checks if a property on a message is considered to be present.
 * @param ***REMOVED***Object***REMOVED*** obj Plain object or message instance
 * @param ***REMOVED***string***REMOVED*** prop Property name
 * @returns ***REMOVED***boolean***REMOVED*** `true` if considered to be present, otherwise `false`
 */
util.isSet = function isSet(obj, prop) ***REMOVED***
    var value = obj[prop];
    if (value != null && obj.hasOwnProperty(prop)) // eslint-disable-line eqeqeq, no-prototype-builtins
        return typeof value !== "object" || (Array.isArray(value) ? value.length : Object.keys(value).length) > 0;
    return false;
***REMOVED***;

/**
 * Any compatible Buffer instance.
 * This is a minimal stand-alone definition of a Buffer instance. The actual type is that exported by node's typings.
 * @interface Buffer
 * @extends Uint8Array
 */

/**
 * Node's Buffer class if available.
 * @type ***REMOVED***Constructor<Buffer>***REMOVED***
 */
util.Buffer = (function() ***REMOVED***
    try ***REMOVED***
        var Buffer = util.inquire("buffer").Buffer;
        // refuse to use non-node buffers if not explicitly assigned (perf reasons):
        return Buffer.prototype.utf8Write ? Buffer : /* istanbul ignore next */ null;
    ***REMOVED*** catch (e) ***REMOVED***
        /* istanbul ignore next */
        return null;
    ***REMOVED***
***REMOVED***)();

// Internal alias of or polyfull for Buffer.from.
util._Buffer_from = null;

// Internal alias of or polyfill for Buffer.allocUnsafe.
util._Buffer_allocUnsafe = null;

/**
 * Creates a new buffer of whatever type supported by the environment.
 * @param ***REMOVED***number|number[]***REMOVED*** [sizeOrArray=0] Buffer size or number array
 * @returns ***REMOVED***Uint8Array|Buffer***REMOVED*** Buffer
 */
util.newBuffer = function newBuffer(sizeOrArray) ***REMOVED***
    /* istanbul ignore next */
    return typeof sizeOrArray === "number"
        ? util.Buffer
            ? util._Buffer_allocUnsafe(sizeOrArray)
            : new util.Array(sizeOrArray)
        : util.Buffer
            ? util._Buffer_from(sizeOrArray)
            : typeof Uint8Array === "undefined"
                ? sizeOrArray
                : new Uint8Array(sizeOrArray);
***REMOVED***;

/**
 * Array implementation used in the browser. `Uint8Array` if supported, otherwise `Array`.
 * @type ***REMOVED***Constructor<Uint8Array>***REMOVED***
 */
util.Array = typeof Uint8Array !== "undefined" ? Uint8Array /* istanbul ignore next */ : Array;

/**
 * Any compatible Long instance.
 * This is a minimal stand-alone definition of a Long instance. The actual type is that exported by long.js.
 * @interface Long
 * @property ***REMOVED***number***REMOVED*** low Low bits
 * @property ***REMOVED***number***REMOVED*** high High bits
 * @property ***REMOVED***boolean***REMOVED*** unsigned Whether unsigned or not
 */

/**
 * Long.js's Long class if available.
 * @type ***REMOVED***Constructor<Long>***REMOVED***
 */
util.Long = /* istanbul ignore next */ global.dcodeIO && /* istanbul ignore next */ global.dcodeIO.Long || util.inquire("long");

/**
 * Regular expression used to verify 2 bit (`bool`) map keys.
 * @type ***REMOVED***RegExp***REMOVED***
 * @const
 */
util.key2Re = /^true|false|0|1$/;

/**
 * Regular expression used to verify 32 bit (`int32` etc.) map keys.
 * @type ***REMOVED***RegExp***REMOVED***
 * @const
 */
util.key32Re = /^-?(?:0|[1-9][0-9]*)$/;

/**
 * Regular expression used to verify 64 bit (`int64` etc.) map keys.
 * @type ***REMOVED***RegExp***REMOVED***
 * @const
 */
util.key64Re = /^(?:[\\x00-\\xff]***REMOVED***8***REMOVED***|-?(?:0|[1-9][0-9]*))$/;

/**
 * Converts a number or long to an 8 characters long hash string.
 * @param ***REMOVED***Long|number***REMOVED*** value Value to convert
 * @returns ***REMOVED***string***REMOVED*** Hash
 */
util.longToHash = function longToHash(value) ***REMOVED***
    return value
        ? util.LongBits.from(value).toHash()
        : util.LongBits.zeroHash;
***REMOVED***;

/**
 * Converts an 8 characters long hash string to a long or number.
 * @param ***REMOVED***string***REMOVED*** hash Hash
 * @param ***REMOVED***boolean***REMOVED*** [unsigned=false] Whether unsigned or not
 * @returns ***REMOVED***Long|number***REMOVED*** Original value
 */
util.longFromHash = function longFromHash(hash, unsigned) ***REMOVED***
    var bits = util.LongBits.fromHash(hash);
    if (util.Long)
        return util.Long.fromBits(bits.lo, bits.hi, unsigned);
    return bits.toNumber(Boolean(unsigned));
***REMOVED***;

/**
 * Merges the properties of the source object into the destination object.
 * @memberof util
 * @param ***REMOVED***Object.<string,*>***REMOVED*** dst Destination object
 * @param ***REMOVED***Object.<string,*>***REMOVED*** src Source object
 * @param ***REMOVED***boolean***REMOVED*** [ifNotSet=false] Merges only if the key is not already set
 * @returns ***REMOVED***Object.<string,*>***REMOVED*** Destination object
 */
function merge(dst, src, ifNotSet) ***REMOVED*** // used by converters
    for (var keys = Object.keys(src), i = 0; i < keys.length; ++i)
        if (dst[keys[i]] === undefined || !ifNotSet)
            dst[keys[i]] = src[keys[i]];
    return dst;
***REMOVED***

util.merge = merge;

/**
 * Converts the first character of a string to lower case.
 * @param ***REMOVED***string***REMOVED*** str String to convert
 * @returns ***REMOVED***string***REMOVED*** Converted string
 */
util.lcFirst = function lcFirst(str) ***REMOVED***
    return str.charAt(0).toLowerCase() + str.substring(1);
***REMOVED***;

/**
 * Creates a custom error constructor.
 * @memberof util
 * @param ***REMOVED***string***REMOVED*** name Error name
 * @returns ***REMOVED***Constructor<Error>***REMOVED*** Custom error constructor
 */
function newError(name) ***REMOVED***

    function CustomError(message, properties) ***REMOVED***

        if (!(this instanceof CustomError))
            return new CustomError(message, properties);

        // Error.call(this, message);
        // ^ just returns a new error instance because the ctor can be called as a function

        Object.defineProperty(this, "message", ***REMOVED*** get: function() ***REMOVED*** return message; ***REMOVED*** ***REMOVED***);

        /* istanbul ignore next */
        if (Error.captureStackTrace) // node
            Error.captureStackTrace(this, CustomError);
        else
            Object.defineProperty(this, "stack", ***REMOVED*** value: (new Error()).stack || "" ***REMOVED***);

        if (properties)
            merge(this, properties);
    ***REMOVED***

    (CustomError.prototype = Object.create(Error.prototype)).constructor = CustomError;

    Object.defineProperty(CustomError.prototype, "name", ***REMOVED*** get: function() ***REMOVED*** return name; ***REMOVED*** ***REMOVED***);

    CustomError.prototype.toString = function toString() ***REMOVED***
        return this.name + ": " + this.message;
    ***REMOVED***;

    return CustomError;
***REMOVED***

util.newError = newError;

/**
 * Constructs a new protocol error.
 * @classdesc Error subclass indicating a protocol specifc error.
 * @memberof util
 * @extends Error
 * @template T extends Message<T>
 * @constructor
 * @param ***REMOVED***string***REMOVED*** message Error message
 * @param ***REMOVED***Object.<string,*>***REMOVED*** [properties] Additional properties
 * @example
 * try ***REMOVED***
 *     MyMessage.decode(someBuffer); // throws if required fields are missing
 * ***REMOVED*** catch (e) ***REMOVED***
 *     if (e instanceof ProtocolError && e.instance)
 *         console.log("decoded so far: " + JSON.stringify(e.instance));
 * ***REMOVED***
 */
util.ProtocolError = newError("ProtocolError");

/**
 * So far decoded message instance.
 * @name util.ProtocolError#instance
 * @type ***REMOVED***Message<T>***REMOVED***
 */

/**
 * A OneOf getter as returned by ***REMOVED***@link util.oneOfGetter***REMOVED***.
 * @typedef OneOfGetter
 * @type ***REMOVED***function***REMOVED***
 * @returns ***REMOVED***string|undefined***REMOVED*** Set field name, if any
 */

/**
 * Builds a getter for a oneof's present field name.
 * @param ***REMOVED***string[]***REMOVED*** fieldNames Field names
 * @returns ***REMOVED***OneOfGetter***REMOVED*** Unbound getter
 */
util.oneOfGetter = function getOneOf(fieldNames) ***REMOVED***
    var fieldMap = ***REMOVED******REMOVED***;
    for (var i = 0; i < fieldNames.length; ++i)
        fieldMap[fieldNames[i]] = 1;

    /**
     * @returns ***REMOVED***string|undefined***REMOVED*** Set field name, if any
     * @this Object
     * @ignore
     */
    return function() ***REMOVED*** // eslint-disable-line consistent-return
        for (var keys = Object.keys(this), i = keys.length - 1; i > -1; --i)
            if (fieldMap[keys[i]] === 1 && this[keys[i]] !== undefined && this[keys[i]] !== null)
                return keys[i];
    ***REMOVED***;
***REMOVED***;

/**
 * A OneOf setter as returned by ***REMOVED***@link util.oneOfSetter***REMOVED***.
 * @typedef OneOfSetter
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***string|undefined***REMOVED*** value Field name
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Builds a setter for a oneof's present field name.
 * @param ***REMOVED***string[]***REMOVED*** fieldNames Field names
 * @returns ***REMOVED***OneOfSetter***REMOVED*** Unbound setter
 */
util.oneOfSetter = function setOneOf(fieldNames) ***REMOVED***

    /**
     * @param ***REMOVED***string***REMOVED*** name Field name
     * @returns ***REMOVED***undefined***REMOVED***
     * @this Object
     * @ignore
     */
    return function(name) ***REMOVED***
        for (var i = 0; i < fieldNames.length; ++i)
            if (fieldNames[i] !== name)
                delete this[fieldNames[i]];
    ***REMOVED***;
***REMOVED***;

/**
 * Default conversion options used for ***REMOVED***@link Message#toJSON***REMOVED*** implementations.
 *
 * These options are close to proto3's JSON mapping with the exception that internal types like Any are handled just like messages. More precisely:
 *
 * - Longs become strings
 * - Enums become string keys
 * - Bytes become base64 encoded strings
 * - (Sub-)Messages become plain objects
 * - Maps become plain objects with all string keys
 * - Repeated fields become arrays
 * - NaN and Infinity for float and double fields become strings
 *
 * @type ***REMOVED***IConversionOptions***REMOVED***
 * @see https://developers.google.com/protocol-buffers/docs/proto3?hl=en#json
 */
util.toJSONOptions = ***REMOVED***
    longs: String,
    enums: String,
    bytes: String,
    json: true
***REMOVED***;

util._configure = function() ***REMOVED***
    var Buffer = util.Buffer;
    /* istanbul ignore if */
    if (!Buffer) ***REMOVED***
        util._Buffer_from = util._Buffer_allocUnsafe = null;
        return;
    ***REMOVED***
    // because node 4.x buffers are incompatible & immutable
    // see: https://github.com/dcodeIO/protobuf.js/pull/665
    util._Buffer_from = Buffer.from !== Uint8Array.from && Buffer.from ||
        /* istanbul ignore next */
        function Buffer_from(value, encoding) ***REMOVED***
            return new Buffer(value, encoding);
        ***REMOVED***;
    util._Buffer_allocUnsafe = Buffer.allocUnsafe ||
        /* istanbul ignore next */
        function Buffer_allocUnsafe(size) ***REMOVED***
            return new Buffer(size);
        ***REMOVED***;
***REMOVED***;

***REMOVED***,***REMOVED***"1":1,"14":14,"2":2,"3":3,"4":4,"5":5,"6":6,"7":7***REMOVED***],16:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = Writer;

var util      = require(15);

var BufferWriter; // cyclic

var LongBits  = util.LongBits,
    base64    = util.base64,
    utf8      = util.utf8;

/**
 * Constructs a new writer operation instance.
 * @classdesc Scheduled writer operation.
 * @constructor
 * @param ***REMOVED***function(*, Uint8Array, number)***REMOVED*** fn Function to call
 * @param ***REMOVED***number***REMOVED*** len Value byte length
 * @param ***REMOVED*******REMOVED*** val Value to write
 * @ignore
 */
function Op(fn, len, val) ***REMOVED***

    /**
     * Function to call.
     * @type ***REMOVED***function(Uint8Array, number, *)***REMOVED***
     */
    this.fn = fn;

    /**
     * Value byte length.
     * @type ***REMOVED***number***REMOVED***
     */
    this.len = len;

    /**
     * Next operation.
     * @type ***REMOVED***Writer.Op|undefined***REMOVED***
     */
    this.next = undefined;

    /**
     * Value to write.
     * @type ***REMOVED*******REMOVED***
     */
    this.val = val; // type varies
***REMOVED***

/* istanbul ignore next */
function noop() ***REMOVED******REMOVED*** // eslint-disable-line no-empty-function

/**
 * Constructs a new writer state instance.
 * @classdesc Copied writer state.
 * @memberof Writer
 * @constructor
 * @param ***REMOVED***Writer***REMOVED*** writer Writer to copy state from
 * @ignore
 */
function State(writer) ***REMOVED***

    /**
     * Current head.
     * @type ***REMOVED***Writer.Op***REMOVED***
     */
    this.head = writer.head;

    /**
     * Current tail.
     * @type ***REMOVED***Writer.Op***REMOVED***
     */
    this.tail = writer.tail;

    /**
     * Current buffer length.
     * @type ***REMOVED***number***REMOVED***
     */
    this.len = writer.len;

    /**
     * Next state.
     * @type ***REMOVED***State|null***REMOVED***
     */
    this.next = writer.states;
***REMOVED***

/**
 * Constructs a new writer instance.
 * @classdesc Wire format writer using `Uint8Array` if available, otherwise `Array`.
 * @constructor
 */
function Writer() ***REMOVED***

    /**
     * Current length.
     * @type ***REMOVED***number***REMOVED***
     */
    this.len = 0;

    /**
     * Operations head.
     * @type ***REMOVED***Object***REMOVED***
     */
    this.head = new Op(noop, 0, 0);

    /**
     * Operations tail
     * @type ***REMOVED***Object***REMOVED***
     */
    this.tail = this.head;

    /**
     * Linked forked states.
     * @type ***REMOVED***Object|null***REMOVED***
     */
    this.states = null;

    // When a value is written, the writer calculates its byte length and puts it into a linked
    // list of operations to perform when finish() is called. This both allows us to allocate
    // buffers of the exact required size and reduces the amount of work we have to do compared
    // to first calculating over objects and then encoding over objects. In our case, the encoding
    // part is just a linked list walk calling operations with already prepared values.
***REMOVED***

/**
 * Creates a new writer.
 * @function
 * @returns ***REMOVED***BufferWriter|Writer***REMOVED*** A ***REMOVED***@link BufferWriter***REMOVED*** when Buffers are supported, otherwise a ***REMOVED***@link Writer***REMOVED***
 */
Writer.create = util.Buffer
    ? function create_buffer_setup() ***REMOVED***
        return (Writer.create = function create_buffer() ***REMOVED***
            return new BufferWriter();
        ***REMOVED***)();
    ***REMOVED***
    /* istanbul ignore next */
    : function create_array() ***REMOVED***
        return new Writer();
    ***REMOVED***;

/**
 * Allocates a buffer of the specified size.
 * @param ***REMOVED***number***REMOVED*** size Buffer size
 * @returns ***REMOVED***Uint8Array***REMOVED*** Buffer
 */
Writer.alloc = function alloc(size) ***REMOVED***
    return new util.Array(size);
***REMOVED***;

// Use Uint8Array buffer pool in the browser, just like node does with buffers
/* istanbul ignore else */
if (util.Array !== Array)
    Writer.alloc = util.pool(Writer.alloc, util.Array.prototype.subarray);

/**
 * Pushes a new operation to the queue.
 * @param ***REMOVED***function(Uint8Array, number, *)***REMOVED*** fn Function to call
 * @param ***REMOVED***number***REMOVED*** len Value byte length
 * @param ***REMOVED***number***REMOVED*** val Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 * @private
 */
Writer.prototype._push = function push(fn, len, val) ***REMOVED***
    this.tail = this.tail.next = new Op(fn, len, val);
    this.len += len;
    return this;
***REMOVED***;

function writeByte(val, buf, pos) ***REMOVED***
    buf[pos] = val & 255;
***REMOVED***

function writeVarint32(val, buf, pos) ***REMOVED***
    while (val > 127) ***REMOVED***
        buf[pos++] = val & 127 | 128;
        val >>>= 7;
    ***REMOVED***
    buf[pos] = val;
***REMOVED***

/**
 * Constructs a new varint writer operation instance.
 * @classdesc Scheduled varint writer operation.
 * @extends Op
 * @constructor
 * @param ***REMOVED***number***REMOVED*** len Value byte length
 * @param ***REMOVED***number***REMOVED*** val Value to write
 * @ignore
 */
function VarintOp(len, val) ***REMOVED***
    this.len = len;
    this.next = undefined;
    this.val = val;
***REMOVED***

VarintOp.prototype = Object.create(Op.prototype);
VarintOp.prototype.fn = writeVarint32;

/**
 * Writes an unsigned 32 bit value as a varint.
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.uint32 = function write_uint32(value) ***REMOVED***
    // here, the call to this.push has been inlined and a varint specific Op subclass is used.
    // uint32 is by far the most frequently used operation and benefits significantly from this.
    this.len += (this.tail = this.tail.next = new VarintOp(
        (value = value >>> 0)
                < 128       ? 1
        : value < 16384     ? 2
        : value < 2097152   ? 3
        : value < 268435456 ? 4
        :                     5,
    value)).len;
    return this;
***REMOVED***;

/**
 * Writes a signed 32 bit value as a varint.
 * @function
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.int32 = function write_int32(value) ***REMOVED***
    return value < 0
        ? this._push(writeVarint64, 10, LongBits.fromNumber(value)) // 10 bytes per spec
        : this.uint32(value);
***REMOVED***;

/**
 * Writes a 32 bit value as a varint, zig-zag encoded.
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.sint32 = function write_sint32(value) ***REMOVED***
    return this.uint32((value << 1 ^ value >> 31) >>> 0);
***REMOVED***;

function writeVarint64(val, buf, pos) ***REMOVED***
    while (val.hi) ***REMOVED***
        buf[pos++] = val.lo & 127 | 128;
        val.lo = (val.lo >>> 7 | val.hi << 25) >>> 0;
        val.hi >>>= 7;
    ***REMOVED***
    while (val.lo > 127) ***REMOVED***
        buf[pos++] = val.lo & 127 | 128;
        val.lo = val.lo >>> 7;
    ***REMOVED***
    buf[pos++] = val.lo;
***REMOVED***

/**
 * Writes an unsigned 64 bit value as a varint.
 * @param ***REMOVED***Long|number|string***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 * @throws ***REMOVED***TypeError***REMOVED*** If `value` is a string and no long library is present.
 */
Writer.prototype.uint64 = function write_uint64(value) ***REMOVED***
    var bits = LongBits.from(value);
    return this._push(writeVarint64, bits.length(), bits);
***REMOVED***;

/**
 * Writes a signed 64 bit value as a varint.
 * @function
 * @param ***REMOVED***Long|number|string***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 * @throws ***REMOVED***TypeError***REMOVED*** If `value` is a string and no long library is present.
 */
Writer.prototype.int64 = Writer.prototype.uint64;

/**
 * Writes a signed 64 bit value as a varint, zig-zag encoded.
 * @param ***REMOVED***Long|number|string***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 * @throws ***REMOVED***TypeError***REMOVED*** If `value` is a string and no long library is present.
 */
Writer.prototype.sint64 = function write_sint64(value) ***REMOVED***
    var bits = LongBits.from(value).zzEncode();
    return this._push(writeVarint64, bits.length(), bits);
***REMOVED***;

/**
 * Writes a boolish value as a varint.
 * @param ***REMOVED***boolean***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.bool = function write_bool(value) ***REMOVED***
    return this._push(writeByte, 1, value ? 1 : 0);
***REMOVED***;

function writeFixed32(val, buf, pos) ***REMOVED***
    buf[pos    ] =  val         & 255;
    buf[pos + 1] =  val >>> 8   & 255;
    buf[pos + 2] =  val >>> 16  & 255;
    buf[pos + 3] =  val >>> 24;
***REMOVED***

/**
 * Writes an unsigned 32 bit value as fixed 32 bits.
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.fixed32 = function write_fixed32(value) ***REMOVED***
    return this._push(writeFixed32, 4, value >>> 0);
***REMOVED***;

/**
 * Writes a signed 32 bit value as fixed 32 bits.
 * @function
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.sfixed32 = Writer.prototype.fixed32;

/**
 * Writes an unsigned 64 bit value as fixed 64 bits.
 * @param ***REMOVED***Long|number|string***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 * @throws ***REMOVED***TypeError***REMOVED*** If `value` is a string and no long library is present.
 */
Writer.prototype.fixed64 = function write_fixed64(value) ***REMOVED***
    var bits = LongBits.from(value);
    return this._push(writeFixed32, 4, bits.lo)._push(writeFixed32, 4, bits.hi);
***REMOVED***;

/**
 * Writes a signed 64 bit value as fixed 64 bits.
 * @function
 * @param ***REMOVED***Long|number|string***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 * @throws ***REMOVED***TypeError***REMOVED*** If `value` is a string and no long library is present.
 */
Writer.prototype.sfixed64 = Writer.prototype.fixed64;

/**
 * Writes a float (32 bit).
 * @function
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.float = function write_float(value) ***REMOVED***
    return this._push(util.float.writeFloatLE, 4, value);
***REMOVED***;

/**
 * Writes a double (64 bit float).
 * @function
 * @param ***REMOVED***number***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.double = function write_double(value) ***REMOVED***
    return this._push(util.float.writeDoubleLE, 8, value);
***REMOVED***;

var writeBytes = util.Array.prototype.set
    ? function writeBytes_set(val, buf, pos) ***REMOVED***
        buf.set(val, pos); // also works for plain array values
    ***REMOVED***
    /* istanbul ignore next */
    : function writeBytes_for(val, buf, pos) ***REMOVED***
        for (var i = 0; i < val.length; ++i)
            buf[pos + i] = val[i];
    ***REMOVED***;

/**
 * Writes a sequence of bytes.
 * @param ***REMOVED***Uint8Array|string***REMOVED*** value Buffer or base64 encoded string to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.bytes = function write_bytes(value) ***REMOVED***
    var len = value.length >>> 0;
    if (!len)
        return this._push(writeByte, 1, 0);
    if (util.isString(value)) ***REMOVED***
        var buf = Writer.alloc(len = base64.length(value));
        base64.decode(value, buf, 0);
        value = buf;
    ***REMOVED***
    return this.uint32(len)._push(writeBytes, len, value);
***REMOVED***;

/**
 * Writes a string.
 * @param ***REMOVED***string***REMOVED*** value Value to write
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.string = function write_string(value) ***REMOVED***
    var len = utf8.length(value);
    return len
        ? this.uint32(len)._push(utf8.write, len, value)
        : this._push(writeByte, 1, 0);
***REMOVED***;

/**
 * Forks this writer's state by pushing it to a stack.
 * Calling ***REMOVED***@link Writer#reset|reset***REMOVED*** or ***REMOVED***@link Writer#ldelim|ldelim***REMOVED*** resets the writer to the previous state.
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.fork = function fork() ***REMOVED***
    this.states = new State(this);
    this.head = this.tail = new Op(noop, 0, 0);
    this.len = 0;
    return this;
***REMOVED***;

/**
 * Resets this instance to the last state.
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.reset = function reset() ***REMOVED***
    if (this.states) ***REMOVED***
        this.head   = this.states.head;
        this.tail   = this.states.tail;
        this.len    = this.states.len;
        this.states = this.states.next;
    ***REMOVED*** else ***REMOVED***
        this.head = this.tail = new Op(noop, 0, 0);
        this.len  = 0;
    ***REMOVED***
    return this;
***REMOVED***;

/**
 * Resets to the last state and appends the fork state's current write length as a varint followed by its operations.
 * @returns ***REMOVED***Writer***REMOVED*** `this`
 */
Writer.prototype.ldelim = function ldelim() ***REMOVED***
    var head = this.head,
        tail = this.tail,
        len  = this.len;
    this.reset().uint32(len);
    if (len) ***REMOVED***
        this.tail.next = head.next; // skip noop
        this.tail = tail;
        this.len += len;
    ***REMOVED***
    return this;
***REMOVED***;

/**
 * Finishes the write operation.
 * @returns ***REMOVED***Uint8Array***REMOVED*** Finished buffer
 */
Writer.prototype.finish = function finish() ***REMOVED***
    var head = this.head.next, // skip noop
        buf  = this.constructor.alloc(this.len),
        pos  = 0;
    while (head) ***REMOVED***
        head.fn(head.val, buf, pos);
        pos += head.len;
        head = head.next;
    ***REMOVED***
    // this.head = this.tail = null;
    return buf;
***REMOVED***;

Writer._configure = function(BufferWriter_) ***REMOVED***
    BufferWriter = BufferWriter_;
***REMOVED***;

***REMOVED***,***REMOVED***"15":15***REMOVED***],17:[function(require,module,exports)***REMOVED***
"use strict";
module.exports = BufferWriter;

// extends Writer
var Writer = require(16);
(BufferWriter.prototype = Object.create(Writer.prototype)).constructor = BufferWriter;

var util = require(15);

var Buffer = util.Buffer;

/**
 * Constructs a new buffer writer instance.
 * @classdesc Wire format writer using node buffers.
 * @extends Writer
 * @constructor
 */
function BufferWriter() ***REMOVED***
    Writer.call(this);
***REMOVED***

/**
 * Allocates a buffer of the specified size.
 * @param ***REMOVED***number***REMOVED*** size Buffer size
 * @returns ***REMOVED***Buffer***REMOVED*** Buffer
 */
BufferWriter.alloc = function alloc_buffer(size) ***REMOVED***
    return (BufferWriter.alloc = util._Buffer_allocUnsafe)(size);
***REMOVED***;

var writeBytesBuffer = Buffer && Buffer.prototype instanceof Uint8Array && Buffer.prototype.set.name === "set"
    ? function writeBytesBuffer_set(val, buf, pos) ***REMOVED***
        buf.set(val, pos); // faster than copy (requires node >= 4 where Buffers extend Uint8Array and set is properly inherited)
                           // also works for plain array values
    ***REMOVED***
    /* istanbul ignore next */
    : function writeBytesBuffer_copy(val, buf, pos) ***REMOVED***
        if (val.copy) // Buffer values
            val.copy(buf, pos, 0, val.length);
        else for (var i = 0; i < val.length;) // plain array values
            buf[pos++] = val[i++];
    ***REMOVED***;

/**
 * @override
 */
BufferWriter.prototype.bytes = function write_bytes_buffer(value) ***REMOVED***
    if (util.isString(value))
        value = util._Buffer_from(value, "base64");
    var len = value.length >>> 0;
    this.uint32(len);
    if (len)
        this._push(writeBytesBuffer, len, value);
    return this;
***REMOVED***;

function writeStringBuffer(val, buf, pos) ***REMOVED***
    if (val.length < 40) // plain js is faster for short strings (probably due to redundant assertions)
        util.utf8.write(val, buf, pos);
    else
        buf.utf8Write(val, pos);
***REMOVED***

/**
 * @override
 */
BufferWriter.prototype.string = function write_string_buffer(value) ***REMOVED***
    var len = Buffer.byteLength(value);
    this.uint32(len);
    if (len)
        this._push(writeStringBuffer, len, value);
    return this;
***REMOVED***;


/**
 * Finishes the write operation.
 * @name BufferWriter#finish
 * @function
 * @returns ***REMOVED***Buffer***REMOVED*** Finished buffer
 */

***REMOVED***,***REMOVED***"15":15,"16":16***REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[8])

***REMOVED***)(typeof window==="object"&&window||typeof self==="object"&&self||this);
