// @firebase/storage/dist/esm/src/tasksnapshot.js
var UploadTaskSnapshot = /** @class */ (function () ***REMOVED***
    function UploadTaskSnapshot(bytesTransferred, totalBytes, state, metadata, task, ref) ***REMOVED***
        this.bytesTransferred = bytesTransferred;
        this.totalBytes = totalBytes;
        this.state = state;
        this.metadata = metadata;
        this.task = task;
        this.ref = ref;
    ***REMOVED***
    Object.defineProperty(UploadTaskSnapshot.prototype, "downloadURL", ***REMOVED***
        get: function () ***REMOVED***
            if (this.metadata !== null) ***REMOVED***
                var urls = this.metadata['downloadURLs'];
                if (urls != null && urls[0] != null) ***REMOVED***
                    return urls[0];
                ***REMOVED***
                else ***REMOVED***
                    return null;
                ***REMOVED***
            ***REMOVED***
            else ***REMOVED***
                return null;
            ***REMOVED***
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    return UploadTaskSnapshot;
***REMOVED***());
export ***REMOVED*** UploadTaskSnapshot ***REMOVED***;
