// bytebuffer/dist/bytebuffer.js
/*
 Copyright 2013-2014 Daniel Wirtz <dcode@dcode.io>

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

/**
 * @license bytebuffer.js (c) 2015 Daniel Wirtz <dcode@dcode.io>
 * Backing buffer: ArrayBuffer, Accessor: Uint8Array
 * Released under the Apache License, Version 2.0
 * see: https://github.com/dcodeIO/bytebuffer.js for details
 */
(function(global, factory) ***REMOVED***

    /* AMD */ if (typeof define === 'function' && define["amd"])
        define(["long"], factory);
    /* CommonJS */ else if (typeof require === 'function' && typeof module === "object" && module && module["exports"])
        module['exports'] = (function() ***REMOVED***
            var Long; try ***REMOVED*** Long = require("long"); ***REMOVED*** catch (e) ***REMOVED******REMOVED***
            return factory(Long);
        ***REMOVED***)();
    /* Global */ else
        (global["dcodeIO"] = global["dcodeIO"] || ***REMOVED******REMOVED***)["ByteBuffer"] = factory(global["dcodeIO"]["Long"]);

***REMOVED***)(this, function(Long) ***REMOVED***
    "use strict";

    /**
     * Constructs a new ByteBuffer.
     * @class The swiss army knife for binary data in JavaScript.
     * @exports ByteBuffer
     * @constructor
     * @param ***REMOVED***number=***REMOVED*** capacity Initial capacity. Defaults to ***REMOVED***@link ByteBuffer.DEFAULT_CAPACITY***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @expose
     */
    var ByteBuffer = function(capacity, littleEndian, noAssert) ***REMOVED***
        if (typeof capacity === 'undefined')
            capacity = ByteBuffer.DEFAULT_CAPACITY;
        if (typeof littleEndian === 'undefined')
            littleEndian = ByteBuffer.DEFAULT_ENDIAN;
        if (typeof noAssert === 'undefined')
            noAssert = ByteBuffer.DEFAULT_NOASSERT;
        if (!noAssert) ***REMOVED***
            capacity = capacity | 0;
            if (capacity < 0)
                throw RangeError("Illegal capacity");
            littleEndian = !!littleEndian;
            noAssert = !!noAssert;
        ***REMOVED***

        /**
         * Backing ArrayBuffer.
         * @type ***REMOVED***!ArrayBuffer***REMOVED***
         * @expose
         */
        this.buffer = capacity === 0 ? EMPTY_BUFFER : new ArrayBuffer(capacity);

        /**
         * Uint8Array utilized to manipulate the backing buffer. Becomes `null` if the backing buffer has a capacity of `0`.
         * @type ***REMOVED***?Uint8Array***REMOVED***
         * @expose
         */
        this.view = capacity === 0 ? null : new Uint8Array(this.buffer);

        /**
         * Absolute read/write offset.
         * @type ***REMOVED***number***REMOVED***
         * @expose
         * @see ByteBuffer#flip
         * @see ByteBuffer#clear
         */
        this.offset = 0;

        /**
         * Marked offset.
         * @type ***REMOVED***number***REMOVED***
         * @expose
         * @see ByteBuffer#mark
         * @see ByteBuffer#reset
         */
        this.markedOffset = -1;

        /**
         * Absolute limit of the contained data. Set to the backing buffer's capacity upon allocation.
         * @type ***REMOVED***number***REMOVED***
         * @expose
         * @see ByteBuffer#flip
         * @see ByteBuffer#clear
         */
        this.limit = capacity;

        /**
         * Whether to use little endian byte order, defaults to `false` for big endian.
         * @type ***REMOVED***boolean***REMOVED***
         * @expose
         */
        this.littleEndian = littleEndian;

        /**
         * Whether to skip assertions of offsets and values, defaults to `false`.
         * @type ***REMOVED***boolean***REMOVED***
         * @expose
         */
        this.noAssert = noAssert;
    ***REMOVED***;

    /**
     * ByteBuffer version.
     * @type ***REMOVED***string***REMOVED***
     * @const
     * @expose
     */
    ByteBuffer.VERSION = "5.0.1";

    /**
     * Little endian constant that can be used instead of its boolean value. Evaluates to `true`.
     * @type ***REMOVED***boolean***REMOVED***
     * @const
     * @expose
     */
    ByteBuffer.LITTLE_ENDIAN = true;

    /**
     * Big endian constant that can be used instead of its boolean value. Evaluates to `false`.
     * @type ***REMOVED***boolean***REMOVED***
     * @const
     * @expose
     */
    ByteBuffer.BIG_ENDIAN = false;

    /**
     * Default initial capacity of `16`.
     * @type ***REMOVED***number***REMOVED***
     * @expose
     */
    ByteBuffer.DEFAULT_CAPACITY = 16;

    /**
     * Default endianess of `false` for big endian.
     * @type ***REMOVED***boolean***REMOVED***
     * @expose
     */
    ByteBuffer.DEFAULT_ENDIAN = ByteBuffer.BIG_ENDIAN;

    /**
     * Default no assertions flag of `false`.
     * @type ***REMOVED***boolean***REMOVED***
     * @expose
     */
    ByteBuffer.DEFAULT_NOASSERT = false;

    /**
     * A `Long` class for representing a 64-bit two's-complement integer value. May be `null` if Long.js has not been loaded
     *  and int64 support is not available.
     * @type ***REMOVED***?Long***REMOVED***
     * @const
     * @see https://github.com/dcodeIO/long.js
     * @expose
     */
    ByteBuffer.Long = Long || null;

    /**
     * @alias ByteBuffer.prototype
     * @inner
     */
    var ByteBufferPrototype = ByteBuffer.prototype;

    /**
     * An indicator used to reliably determine if an object is a ByteBuffer or not.
     * @type ***REMOVED***boolean***REMOVED***
     * @const
     * @expose
     * @private
     */
    ByteBufferPrototype.__isByteBuffer__;

    Object.defineProperty(ByteBufferPrototype, "__isByteBuffer__", ***REMOVED***
        value: true,
        enumerable: false,
        configurable: false
    ***REMOVED***);

    // helpers

    /**
     * @type ***REMOVED***!ArrayBuffer***REMOVED***
     * @inner
     */
    var EMPTY_BUFFER = new ArrayBuffer(0);

    /**
     * String.fromCharCode reference for compile-time renaming.
     * @type ***REMOVED***function(...number):string***REMOVED***
     * @inner
     */
    var stringFromCharCode = String.fromCharCode;

    /**
     * Creates a source function for a string.
     * @param ***REMOVED***string***REMOVED*** s String to read from
     * @returns ***REMOVED***function():number|null***REMOVED*** Source function returning the next char code respectively `null` if there are
     *  no more characters left.
     * @throws ***REMOVED***TypeError***REMOVED*** If the argument is invalid
     * @inner
     */
    function stringSource(s) ***REMOVED***
        var i=0; return function() ***REMOVED***
            return i < s.length ? s.charCodeAt(i++) : null;
        ***REMOVED***;
    ***REMOVED***

    /**
     * Creates a destination function for a string.
     * @returns ***REMOVED***function(number=):undefined|string***REMOVED*** Destination function successively called with the next char code.
     *  Returns the final string when called without arguments.
     * @inner
     */
    function stringDestination() ***REMOVED***
        var cs = [], ps = []; return function() ***REMOVED***
            if (arguments.length === 0)
                return ps.join('')+stringFromCharCode.apply(String, cs);
            if (cs.length + arguments.length > 1024)
                ps.push(stringFromCharCode.apply(String, cs)),
                    cs.length = 0;
            Array.prototype.push.apply(cs, arguments);
        ***REMOVED***;
    ***REMOVED***

    /**
     * Gets the accessor type.
     * @returns ***REMOVED***Function***REMOVED*** `Buffer` under node.js, `Uint8Array` respectively `DataView` in the browser (classes)
     * @expose
     */
    ByteBuffer.accessor = function() ***REMOVED***
        return Uint8Array;
    ***REMOVED***;
    /**
     * Allocates a new ByteBuffer backed by a buffer of the specified capacity.
     * @param ***REMOVED***number=***REMOVED*** capacity Initial capacity. Defaults to ***REMOVED***@link ByteBuffer.DEFAULT_CAPACITY***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED***
     * @expose
     */
    ByteBuffer.allocate = function(capacity, littleEndian, noAssert) ***REMOVED***
        return new ByteBuffer(capacity, littleEndian, noAssert);
    ***REMOVED***;

    /**
     * Concatenates multiple ByteBuffers into one.
     * @param ***REMOVED***!Array.<!ByteBuffer|!ArrayBuffer|!Uint8Array|string>***REMOVED*** buffers Buffers to concatenate
     * @param ***REMOVED***(string|boolean)=***REMOVED*** encoding String encoding if `buffers` contains a string ("base64", "hex", "binary",
     *  defaults to "utf8")
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order for the resulting ByteBuffer. Defaults
     *  to ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values for the resulting ByteBuffer. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** Concatenated ByteBuffer
     * @expose
     */
    ByteBuffer.concat = function(buffers, encoding, littleEndian, noAssert) ***REMOVED***
        if (typeof encoding === 'boolean' || typeof encoding !== 'string') ***REMOVED***
            noAssert = littleEndian;
            littleEndian = encoding;
            encoding = undefined;
        ***REMOVED***
        var capacity = 0;
        for (var i=0, k=buffers.length, length; i<k; ++i) ***REMOVED***
            if (!ByteBuffer.isByteBuffer(buffers[i]))
                buffers[i] = ByteBuffer.wrap(buffers[i], encoding);
            length = buffers[i].limit - buffers[i].offset;
            if (length > 0) capacity += length;
        ***REMOVED***
        if (capacity === 0)
            return new ByteBuffer(0, littleEndian, noAssert);
        var bb = new ByteBuffer(capacity, littleEndian, noAssert),
            bi;
        i=0; while (i<k) ***REMOVED***
            bi = buffers[i++];
            length = bi.limit - bi.offset;
            if (length <= 0) continue;
            bb.view.set(bi.view.subarray(bi.offset, bi.limit), bb.offset);
            bb.offset += length;
        ***REMOVED***
        bb.limit = bb.offset;
        bb.offset = 0;
        return bb;
    ***REMOVED***;

    /**
     * Tests if the specified type is a ByteBuffer.
     * @param ***REMOVED*******REMOVED*** bb ByteBuffer to test
     * @returns ***REMOVED***boolean***REMOVED*** `true` if it is a ByteBuffer, otherwise `false`
     * @expose
     */
    ByteBuffer.isByteBuffer = function(bb) ***REMOVED***
        return (bb && bb["__isByteBuffer__"]) === true;
    ***REMOVED***;
    /**
     * Gets the backing buffer type.
     * @returns ***REMOVED***Function***REMOVED*** `Buffer` under node.js, `ArrayBuffer` in the browser (classes)
     * @expose
     */
    ByteBuffer.type = function() ***REMOVED***
        return ArrayBuffer;
    ***REMOVED***;
    /**
     * Wraps a buffer or a string. Sets the allocated ByteBuffer's ***REMOVED***@link ByteBuffer#offset***REMOVED*** to `0` and its
     *  ***REMOVED***@link ByteBuffer#limit***REMOVED*** to the length of the wrapped data.
     * @param ***REMOVED***!ByteBuffer|!ArrayBuffer|!Uint8Array|string|!Array.<number>***REMOVED*** buffer Anything that can be wrapped
     * @param ***REMOVED***(string|boolean)=***REMOVED*** encoding String encoding if `buffer` is a string ("base64", "hex", "binary", defaults to
     *  "utf8")
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** A ByteBuffer wrapping `buffer`
     * @expose
     */
    ByteBuffer.wrap = function(buffer, encoding, littleEndian, noAssert) ***REMOVED***
        if (typeof encoding !== 'string') ***REMOVED***
            noAssert = littleEndian;
            littleEndian = encoding;
            encoding = undefined;
        ***REMOVED***
        if (typeof buffer === 'string') ***REMOVED***
            if (typeof encoding === 'undefined')
                encoding = "utf8";
            switch (encoding) ***REMOVED***
                case "base64":
                    return ByteBuffer.fromBase64(buffer, littleEndian);
                case "hex":
                    return ByteBuffer.fromHex(buffer, littleEndian);
                case "binary":
                    return ByteBuffer.fromBinary(buffer, littleEndian);
                case "utf8":
                    return ByteBuffer.fromUTF8(buffer, littleEndian);
                case "debug":
                    return ByteBuffer.fromDebug(buffer, littleEndian);
                default:
                    throw Error("Unsupported encoding: "+encoding);
            ***REMOVED***
        ***REMOVED***
        if (buffer === null || typeof buffer !== 'object')
            throw TypeError("Illegal buffer");
        var bb;
        if (ByteBuffer.isByteBuffer(buffer)) ***REMOVED***
            bb = ByteBufferPrototype.clone.call(buffer);
            bb.markedOffset = -1;
            return bb;
        ***REMOVED***
        if (buffer instanceof Uint8Array) ***REMOVED*** // Extract ArrayBuffer from Uint8Array
            bb = new ByteBuffer(0, littleEndian, noAssert);
            if (buffer.length > 0) ***REMOVED*** // Avoid references to more than one EMPTY_BUFFER
                bb.buffer = buffer.buffer;
                bb.offset = buffer.byteOffset;
                bb.limit = buffer.byteOffset + buffer.byteLength;
                bb.view = new Uint8Array(buffer.buffer);
            ***REMOVED***
        ***REMOVED*** else if (buffer instanceof ArrayBuffer) ***REMOVED*** // Reuse ArrayBuffer
            bb = new ByteBuffer(0, littleEndian, noAssert);
            if (buffer.byteLength > 0) ***REMOVED***
                bb.buffer = buffer;
                bb.offset = 0;
                bb.limit = buffer.byteLength;
                bb.view = buffer.byteLength > 0 ? new Uint8Array(buffer) : null;
            ***REMOVED***
        ***REMOVED*** else if (Object.prototype.toString.call(buffer) === "[object Array]") ***REMOVED*** // Create from octets
            bb = new ByteBuffer(buffer.length, littleEndian, noAssert);
            bb.limit = buffer.length;
            for (var i=0; i<buffer.length; ++i)
                bb.view[i] = buffer[i];
        ***REMOVED*** else
            throw TypeError("Illegal buffer"); // Otherwise fail
        return bb;
    ***REMOVED***;

    /**
     * Writes the array as a bitset.
     * @param ***REMOVED***Array<boolean>***REMOVED*** value Array of booleans to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `length` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED***
     * @expose
     */
    ByteBufferPrototype.writeBitSet = function(value, offset) ***REMOVED***
      var relative = typeof offset === 'undefined';
      if (relative) offset = this.offset;
      if (!this.noAssert) ***REMOVED***
        if (!(value instanceof Array))
          throw TypeError("Illegal BitSet: Not an array");
        if (typeof offset !== 'number' || offset % 1 !== 0)
            throw TypeError("Illegal offset: "+offset+" (not an integer)");
        offset >>>= 0;
        if (offset < 0 || offset + 0 > this.buffer.byteLength)
            throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
      ***REMOVED***

      var start = offset,
          bits = value.length,
          bytes = (bits >> 3),
          bit = 0,
          k;

      offset += this.writeVarint32(bits,offset);

      while(bytes--) ***REMOVED***
        k = (!!value[bit++] & 1) |
            ((!!value[bit++] & 1) << 1) |
            ((!!value[bit++] & 1) << 2) |
            ((!!value[bit++] & 1) << 3) |
            ((!!value[bit++] & 1) << 4) |
            ((!!value[bit++] & 1) << 5) |
            ((!!value[bit++] & 1) << 6) |
            ((!!value[bit++] & 1) << 7);
        this.writeByte(k,offset++);
      ***REMOVED***

      if(bit < bits) ***REMOVED***
        var m = 0; k = 0;
        while(bit < bits) k = k | ((!!value[bit++] & 1) << (m++));
        this.writeByte(k,offset++);
      ***REMOVED***

      if (relative) ***REMOVED***
        this.offset = offset;
        return this;
      ***REMOVED***
      return offset - start;
    ***REMOVED***

    /**
     * Reads a BitSet as an array of booleans.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `length` if omitted.
     * @returns ***REMOVED***Array<boolean>
     * @expose
     */
    ByteBufferPrototype.readBitSet = function(offset) ***REMOVED***
      var relative = typeof offset === 'undefined';
      if (relative) offset = this.offset;

      var ret = this.readVarint32(offset),
          bits = ret.value,
          bytes = (bits >> 3),
          bit = 0,
          value = [],
          k;

      offset += ret.length;

      while(bytes--) ***REMOVED***
        k = this.readByte(offset++);
        value[bit++] = !!(k & 0x01);
        value[bit++] = !!(k & 0x02);
        value[bit++] = !!(k & 0x04);
        value[bit++] = !!(k & 0x08);
        value[bit++] = !!(k & 0x10);
        value[bit++] = !!(k & 0x20);
        value[bit++] = !!(k & 0x40);
        value[bit++] = !!(k & 0x80);
      ***REMOVED***

      if(bit < bits) ***REMOVED***
        var m = 0;
        k = this.readByte(offset++);
        while(bit < bits) value[bit++] = !!((k >> (m++)) & 1);
      ***REMOVED***

      if (relative) ***REMOVED***
        this.offset = offset;
      ***REMOVED***
      return value;
    ***REMOVED***
    /**
     * Reads the specified number of bytes.
     * @param ***REMOVED***number***REMOVED*** length Number of bytes to read
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `length` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED***
     * @expose
     */
    ByteBufferPrototype.readBytes = function(length, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + length > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+length+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var slice = this.slice(offset, offset + length);
        if (relative) this.offset += length;
        return slice;
    ***REMOVED***;

    /**
     * Writes a payload of bytes. This is an alias of ***REMOVED***@link ByteBuffer#append***REMOVED***.
     * @function
     * @param ***REMOVED***!ByteBuffer|!ArrayBuffer|!Uint8Array|string***REMOVED*** source Data to write. If `source` is a ByteBuffer, its offsets
     *  will be modified according to the performed read operation.
     * @param ***REMOVED***(string|number)=***REMOVED*** encoding Encoding if `data` is a string ("base64", "hex", "binary", defaults to "utf8")
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeBytes = ByteBufferPrototype.append;

    // types/ints/int8

    /**
     * Writes an 8bit signed integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeInt8 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value |= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 1;
        var capacity0 = this.buffer.byteLength;
        if (offset > capacity0)
            this.resize((capacity0 *= 2) > offset ? capacity0 : offset);
        offset -= 1;
        this.view[offset] = value;
        if (relative) this.offset += 1;
        return this;
    ***REMOVED***;

    /**
     * Writes an 8bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#writeInt8***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeByte = ByteBufferPrototype.writeInt8;

    /**
     * Reads an 8bit signed integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readInt8 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 1 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+1+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = this.view[offset];
        if ((value & 0x80) === 0x80) value = -(0xFF - value + 1); // Cast to signed
        if (relative) this.offset += 1;
        return value;
    ***REMOVED***;

    /**
     * Reads an 8bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#readInt8***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readByte = ByteBufferPrototype.readInt8;

    /**
     * Writes an 8bit unsigned integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeUint8 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value >>>= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 1;
        var capacity1 = this.buffer.byteLength;
        if (offset > capacity1)
            this.resize((capacity1 *= 2) > offset ? capacity1 : offset);
        offset -= 1;
        this.view[offset] = value;
        if (relative) this.offset += 1;
        return this;
    ***REMOVED***;

    /**
     * Writes an 8bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#writeUint8***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeUInt8 = ByteBufferPrototype.writeUint8;

    /**
     * Reads an 8bit unsigned integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readUint8 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 1 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+1+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = this.view[offset];
        if (relative) this.offset += 1;
        return value;
    ***REMOVED***;

    /**
     * Reads an 8bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#readUint8***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `1` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readUInt8 = ByteBufferPrototype.readUint8;

    // types/ints/int16

    /**
     * Writes a 16bit signed integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` or `value` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.writeInt16 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value |= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 2;
        var capacity2 = this.buffer.byteLength;
        if (offset > capacity2)
            this.resize((capacity2 *= 2) > offset ? capacity2 : offset);
        offset -= 2;
        if (this.littleEndian) ***REMOVED***
            this.view[offset+1] = (value & 0xFF00) >>> 8;
            this.view[offset  ] =  value & 0x00FF;
        ***REMOVED*** else ***REMOVED***
            this.view[offset]   = (value & 0xFF00) >>> 8;
            this.view[offset+1] =  value & 0x00FF;
        ***REMOVED***
        if (relative) this.offset += 2;
        return this;
    ***REMOVED***;

    /**
     * Writes a 16bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#writeInt16***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` or `value` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.writeShort = ByteBufferPrototype.writeInt16;

    /**
     * Reads a 16bit signed integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.readInt16 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 2 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+2+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = 0;
        if (this.littleEndian) ***REMOVED***
            value  = this.view[offset  ];
            value |= this.view[offset+1] << 8;
        ***REMOVED*** else ***REMOVED***
            value  = this.view[offset  ] << 8;
            value |= this.view[offset+1];
        ***REMOVED***
        if ((value & 0x8000) === 0x8000) value = -(0xFFFF - value + 1); // Cast to signed
        if (relative) this.offset += 2;
        return value;
    ***REMOVED***;

    /**
     * Reads a 16bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#readInt16***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.readShort = ByteBufferPrototype.readInt16;

    /**
     * Writes a 16bit unsigned integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` or `value` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.writeUint16 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value >>>= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 2;
        var capacity3 = this.buffer.byteLength;
        if (offset > capacity3)
            this.resize((capacity3 *= 2) > offset ? capacity3 : offset);
        offset -= 2;
        if (this.littleEndian) ***REMOVED***
            this.view[offset+1] = (value & 0xFF00) >>> 8;
            this.view[offset  ] =  value & 0x00FF;
        ***REMOVED*** else ***REMOVED***
            this.view[offset]   = (value & 0xFF00) >>> 8;
            this.view[offset+1] =  value & 0x00FF;
        ***REMOVED***
        if (relative) this.offset += 2;
        return this;
    ***REMOVED***;

    /**
     * Writes a 16bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#writeUint16***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` or `value` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.writeUInt16 = ByteBufferPrototype.writeUint16;

    /**
     * Reads a 16bit unsigned integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.readUint16 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 2 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+2+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = 0;
        if (this.littleEndian) ***REMOVED***
            value  = this.view[offset  ];
            value |= this.view[offset+1] << 8;
        ***REMOVED*** else ***REMOVED***
            value  = this.view[offset  ] << 8;
            value |= this.view[offset+1];
        ***REMOVED***
        if (relative) this.offset += 2;
        return value;
    ***REMOVED***;

    /**
     * Reads a 16bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#readUint16***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `2` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @expose
     */
    ByteBufferPrototype.readUInt16 = ByteBufferPrototype.readUint16;

    // types/ints/int32

    /**
     * Writes a 32bit signed integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @expose
     */
    ByteBufferPrototype.writeInt32 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value |= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 4;
        var capacity4 = this.buffer.byteLength;
        if (offset > capacity4)
            this.resize((capacity4 *= 2) > offset ? capacity4 : offset);
        offset -= 4;
        if (this.littleEndian) ***REMOVED***
            this.view[offset+3] = (value >>> 24) & 0xFF;
            this.view[offset+2] = (value >>> 16) & 0xFF;
            this.view[offset+1] = (value >>>  8) & 0xFF;
            this.view[offset  ] =  value         & 0xFF;
        ***REMOVED*** else ***REMOVED***
            this.view[offset  ] = (value >>> 24) & 0xFF;
            this.view[offset+1] = (value >>> 16) & 0xFF;
            this.view[offset+2] = (value >>>  8) & 0xFF;
            this.view[offset+3] =  value         & 0xFF;
        ***REMOVED***
        if (relative) this.offset += 4;
        return this;
    ***REMOVED***;

    /**
     * Writes a 32bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#writeInt32***REMOVED***.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @expose
     */
    ByteBufferPrototype.writeInt = ByteBufferPrototype.writeInt32;

    /**
     * Reads a 32bit signed integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readInt32 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 4 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+4+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = 0;
        if (this.littleEndian) ***REMOVED***
            value  = this.view[offset+2] << 16;
            value |= this.view[offset+1] <<  8;
            value |= this.view[offset  ];
            value += this.view[offset+3] << 24 >>> 0;
        ***REMOVED*** else ***REMOVED***
            value  = this.view[offset+1] << 16;
            value |= this.view[offset+2] <<  8;
            value |= this.view[offset+3];
            value += this.view[offset  ] << 24 >>> 0;
        ***REMOVED***
        value |= 0; // Cast to signed
        if (relative) this.offset += 4;
        return value;
    ***REMOVED***;

    /**
     * Reads a 32bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#readInt32***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and advance ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readInt = ByteBufferPrototype.readInt32;

    /**
     * Writes a 32bit unsigned integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @expose
     */
    ByteBufferPrototype.writeUint32 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value >>>= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 4;
        var capacity5 = this.buffer.byteLength;
        if (offset > capacity5)
            this.resize((capacity5 *= 2) > offset ? capacity5 : offset);
        offset -= 4;
        if (this.littleEndian) ***REMOVED***
            this.view[offset+3] = (value >>> 24) & 0xFF;
            this.view[offset+2] = (value >>> 16) & 0xFF;
            this.view[offset+1] = (value >>>  8) & 0xFF;
            this.view[offset  ] =  value         & 0xFF;
        ***REMOVED*** else ***REMOVED***
            this.view[offset  ] = (value >>> 24) & 0xFF;
            this.view[offset+1] = (value >>> 16) & 0xFF;
            this.view[offset+2] = (value >>>  8) & 0xFF;
            this.view[offset+3] =  value         & 0xFF;
        ***REMOVED***
        if (relative) this.offset += 4;
        return this;
    ***REMOVED***;

    /**
     * Writes a 32bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#writeUint32***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @expose
     */
    ByteBufferPrototype.writeUInt32 = ByteBufferPrototype.writeUint32;

    /**
     * Reads a 32bit unsigned integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readUint32 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 4 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+4+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = 0;
        if (this.littleEndian) ***REMOVED***
            value  = this.view[offset+2] << 16;
            value |= this.view[offset+1] <<  8;
            value |= this.view[offset  ];
            value += this.view[offset+3] << 24 >>> 0;
        ***REMOVED*** else ***REMOVED***
            value  = this.view[offset+1] << 16;
            value |= this.view[offset+2] <<  8;
            value |= this.view[offset+3];
            value += this.view[offset  ] << 24 >>> 0;
        ***REMOVED***
        if (relative) this.offset += 4;
        return value;
    ***REMOVED***;

    /**
     * Reads a 32bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#readUint32***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***number***REMOVED*** Value read
     * @expose
     */
    ByteBufferPrototype.readUInt32 = ByteBufferPrototype.readUint32;

    // types/ints/int64

    if (Long) ***REMOVED***

        /**
         * Writes a 64bit signed integer.
         * @param ***REMOVED***number|!Long***REMOVED*** value Value to write
         * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
         * @expose
         */
        ByteBufferPrototype.writeInt64 = function(value, offset) ***REMOVED***
            var relative = typeof offset === 'undefined';
            if (relative) offset = this.offset;
            if (!this.noAssert) ***REMOVED***
                if (typeof value === 'number')
                    value = Long.fromNumber(value);
                else if (typeof value === 'string')
                    value = Long.fromString(value);
                else if (!(value && value instanceof Long))
                    throw TypeError("Illegal value: "+value+" (not an integer or Long)");
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + 0 > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
            ***REMOVED***
            if (typeof value === 'number')
                value = Long.fromNumber(value);
            else if (typeof value === 'string')
                value = Long.fromString(value);
            offset += 8;
            var capacity6 = this.buffer.byteLength;
            if (offset > capacity6)
                this.resize((capacity6 *= 2) > offset ? capacity6 : offset);
            offset -= 8;
            var lo = value.low,
                hi = value.high;
            if (this.littleEndian) ***REMOVED***
                this.view[offset+3] = (lo >>> 24) & 0xFF;
                this.view[offset+2] = (lo >>> 16) & 0xFF;
                this.view[offset+1] = (lo >>>  8) & 0xFF;
                this.view[offset  ] =  lo         & 0xFF;
                offset += 4;
                this.view[offset+3] = (hi >>> 24) & 0xFF;
                this.view[offset+2] = (hi >>> 16) & 0xFF;
                this.view[offset+1] = (hi >>>  8) & 0xFF;
                this.view[offset  ] =  hi         & 0xFF;
            ***REMOVED*** else ***REMOVED***
                this.view[offset  ] = (hi >>> 24) & 0xFF;
                this.view[offset+1] = (hi >>> 16) & 0xFF;
                this.view[offset+2] = (hi >>>  8) & 0xFF;
                this.view[offset+3] =  hi         & 0xFF;
                offset += 4;
                this.view[offset  ] = (lo >>> 24) & 0xFF;
                this.view[offset+1] = (lo >>> 16) & 0xFF;
                this.view[offset+2] = (lo >>>  8) & 0xFF;
                this.view[offset+3] =  lo         & 0xFF;
            ***REMOVED***
            if (relative) this.offset += 8;
            return this;
        ***REMOVED***;

        /**
         * Writes a 64bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#writeInt64***REMOVED***.
         * @param ***REMOVED***number|!Long***REMOVED*** value Value to write
         * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
         * @expose
         */
        ByteBufferPrototype.writeLong = ByteBufferPrototype.writeInt64;

        /**
         * Reads a 64bit signed integer.
         * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!Long***REMOVED***
         * @expose
         */
        ByteBufferPrototype.readInt64 = function(offset) ***REMOVED***
            var relative = typeof offset === 'undefined';
            if (relative) offset = this.offset;
            if (!this.noAssert) ***REMOVED***
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + 8 > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+8+") <= "+this.buffer.byteLength);
            ***REMOVED***
            var lo = 0,
                hi = 0;
            if (this.littleEndian) ***REMOVED***
                lo  = this.view[offset+2] << 16;
                lo |= this.view[offset+1] <<  8;
                lo |= this.view[offset  ];
                lo += this.view[offset+3] << 24 >>> 0;
                offset += 4;
                hi  = this.view[offset+2] << 16;
                hi |= this.view[offset+1] <<  8;
                hi |= this.view[offset  ];
                hi += this.view[offset+3] << 24 >>> 0;
            ***REMOVED*** else ***REMOVED***
                hi  = this.view[offset+1] << 16;
                hi |= this.view[offset+2] <<  8;
                hi |= this.view[offset+3];
                hi += this.view[offset  ] << 24 >>> 0;
                offset += 4;
                lo  = this.view[offset+1] << 16;
                lo |= this.view[offset+2] <<  8;
                lo |= this.view[offset+3];
                lo += this.view[offset  ] << 24 >>> 0;
            ***REMOVED***
            var value = new Long(lo, hi, false);
            if (relative) this.offset += 8;
            return value;
        ***REMOVED***;

        /**
         * Reads a 64bit signed integer. This is an alias of ***REMOVED***@link ByteBuffer#readInt64***REMOVED***.
         * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!Long***REMOVED***
         * @expose
         */
        ByteBufferPrototype.readLong = ByteBufferPrototype.readInt64;

        /**
         * Writes a 64bit unsigned integer.
         * @param ***REMOVED***number|!Long***REMOVED*** value Value to write
         * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
         * @expose
         */
        ByteBufferPrototype.writeUint64 = function(value, offset) ***REMOVED***
            var relative = typeof offset === 'undefined';
            if (relative) offset = this.offset;
            if (!this.noAssert) ***REMOVED***
                if (typeof value === 'number')
                    value = Long.fromNumber(value);
                else if (typeof value === 'string')
                    value = Long.fromString(value);
                else if (!(value && value instanceof Long))
                    throw TypeError("Illegal value: "+value+" (not an integer or Long)");
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + 0 > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
            ***REMOVED***
            if (typeof value === 'number')
                value = Long.fromNumber(value);
            else if (typeof value === 'string')
                value = Long.fromString(value);
            offset += 8;
            var capacity7 = this.buffer.byteLength;
            if (offset > capacity7)
                this.resize((capacity7 *= 2) > offset ? capacity7 : offset);
            offset -= 8;
            var lo = value.low,
                hi = value.high;
            if (this.littleEndian) ***REMOVED***
                this.view[offset+3] = (lo >>> 24) & 0xFF;
                this.view[offset+2] = (lo >>> 16) & 0xFF;
                this.view[offset+1] = (lo >>>  8) & 0xFF;
                this.view[offset  ] =  lo         & 0xFF;
                offset += 4;
                this.view[offset+3] = (hi >>> 24) & 0xFF;
                this.view[offset+2] = (hi >>> 16) & 0xFF;
                this.view[offset+1] = (hi >>>  8) & 0xFF;
                this.view[offset  ] =  hi         & 0xFF;
            ***REMOVED*** else ***REMOVED***
                this.view[offset  ] = (hi >>> 24) & 0xFF;
                this.view[offset+1] = (hi >>> 16) & 0xFF;
                this.view[offset+2] = (hi >>>  8) & 0xFF;
                this.view[offset+3] =  hi         & 0xFF;
                offset += 4;
                this.view[offset  ] = (lo >>> 24) & 0xFF;
                this.view[offset+1] = (lo >>> 16) & 0xFF;
                this.view[offset+2] = (lo >>>  8) & 0xFF;
                this.view[offset+3] =  lo         & 0xFF;
            ***REMOVED***
            if (relative) this.offset += 8;
            return this;
        ***REMOVED***;

        /**
         * Writes a 64bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#writeUint64***REMOVED***.
         * @function
         * @param ***REMOVED***number|!Long***REMOVED*** value Value to write
         * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
         * @expose
         */
        ByteBufferPrototype.writeUInt64 = ByteBufferPrototype.writeUint64;

        /**
         * Reads a 64bit unsigned integer.
         * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!Long***REMOVED***
         * @expose
         */
        ByteBufferPrototype.readUint64 = function(offset) ***REMOVED***
            var relative = typeof offset === 'undefined';
            if (relative) offset = this.offset;
            if (!this.noAssert) ***REMOVED***
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + 8 > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+8+") <= "+this.buffer.byteLength);
            ***REMOVED***
            var lo = 0,
                hi = 0;
            if (this.littleEndian) ***REMOVED***
                lo  = this.view[offset+2] << 16;
                lo |= this.view[offset+1] <<  8;
                lo |= this.view[offset  ];
                lo += this.view[offset+3] << 24 >>> 0;
                offset += 4;
                hi  = this.view[offset+2] << 16;
                hi |= this.view[offset+1] <<  8;
                hi |= this.view[offset  ];
                hi += this.view[offset+3] << 24 >>> 0;
            ***REMOVED*** else ***REMOVED***
                hi  = this.view[offset+1] << 16;
                hi |= this.view[offset+2] <<  8;
                hi |= this.view[offset+3];
                hi += this.view[offset  ] << 24 >>> 0;
                offset += 4;
                lo  = this.view[offset+1] << 16;
                lo |= this.view[offset+2] <<  8;
                lo |= this.view[offset+3];
                lo += this.view[offset  ] << 24 >>> 0;
            ***REMOVED***
            var value = new Long(lo, hi, true);
            if (relative) this.offset += 8;
            return value;
        ***REMOVED***;

        /**
         * Reads a 64bit unsigned integer. This is an alias of ***REMOVED***@link ByteBuffer#readUint64***REMOVED***.
         * @function
         * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
         * @returns ***REMOVED***!Long***REMOVED***
         * @expose
         */
        ByteBufferPrototype.readUInt64 = ByteBufferPrototype.readUint64;

    ***REMOVED*** // Long


    // types/floats/float32

    /*
     ieee754 - https://github.com/feross/ieee754

     The MIT License (MIT)

     Copyright (c) Feross Aboukhadijeh

     Permission is hereby granted, free of charge, to any person obtaining a copy
     of this software and associated documentation files (the "Software"), to deal
     in the Software without restriction, including without limitation the rights
     to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
     copies of the Software, and to permit persons to whom the Software is
     furnished to do so, subject to the following conditions:

     The above copyright notice and this permission notice shall be included in
     all copies or substantial portions of the Software.

     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
     AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
     THE SOFTWARE.
    */

    /**
     * Reads an IEEE754 float from a byte array.
     * @param ***REMOVED***!Array***REMOVED*** buffer
     * @param ***REMOVED***number***REMOVED*** offset
     * @param ***REMOVED***boolean***REMOVED*** isLE
     * @param ***REMOVED***number***REMOVED*** mLen
     * @param ***REMOVED***number***REMOVED*** nBytes
     * @returns ***REMOVED***number***REMOVED***
     * @inner
     */
    function ieee754_read(buffer, offset, isLE, mLen, nBytes) ***REMOVED***
        var e, m,
            eLen = nBytes * 8 - mLen - 1,
            eMax = (1 << eLen) - 1,
            eBias = eMax >> 1,
            nBits = -7,
            i = isLE ? (nBytes - 1) : 0,
            d = isLE ? -1 : 1,
            s = buffer[offset + i];

        i += d;

        e = s & ((1 << (-nBits)) - 1);
        s >>= (-nBits);
        nBits += eLen;
        for (; nBits > 0; e = e * 256 + buffer[offset + i], i += d, nBits -= 8) ***REMOVED******REMOVED***

        m = e & ((1 << (-nBits)) - 1);
        e >>= (-nBits);
        nBits += mLen;
        for (; nBits > 0; m = m * 256 + buffer[offset + i], i += d, nBits -= 8) ***REMOVED******REMOVED***

        if (e === 0) ***REMOVED***
            e = 1 - eBias;
        ***REMOVED*** else if (e === eMax) ***REMOVED***
            return m ? NaN : ((s ? -1 : 1) * Infinity);
        ***REMOVED*** else ***REMOVED***
            m = m + Math.pow(2, mLen);
            e = e - eBias;
        ***REMOVED***
        return (s ? -1 : 1) * m * Math.pow(2, e - mLen);
    ***REMOVED***

    /**
     * Writes an IEEE754 float to a byte array.
     * @param ***REMOVED***!Array***REMOVED*** buffer
     * @param ***REMOVED***number***REMOVED*** value
     * @param ***REMOVED***number***REMOVED*** offset
     * @param ***REMOVED***boolean***REMOVED*** isLE
     * @param ***REMOVED***number***REMOVED*** mLen
     * @param ***REMOVED***number***REMOVED*** nBytes
     * @inner
     */
    function ieee754_write(buffer, value, offset, isLE, mLen, nBytes) ***REMOVED***
        var e, m, c,
            eLen = nBytes * 8 - mLen - 1,
            eMax = (1 << eLen) - 1,
            eBias = eMax >> 1,
            rt = (mLen === 23 ? Math.pow(2, -24) - Math.pow(2, -77) : 0),
            i = isLE ? 0 : (nBytes - 1),
            d = isLE ? 1 : -1,
            s = value < 0 || (value === 0 && 1 / value < 0) ? 1 : 0;

        value = Math.abs(value);

        if (isNaN(value) || value === Infinity) ***REMOVED***
            m = isNaN(value) ? 1 : 0;
            e = eMax;
        ***REMOVED*** else ***REMOVED***
            e = Math.floor(Math.log(value) / Math.LN2);
            if (value * (c = Math.pow(2, -e)) < 1) ***REMOVED***
                e--;
                c *= 2;
            ***REMOVED***
            if (e + eBias >= 1) ***REMOVED***
                value += rt / c;
            ***REMOVED*** else ***REMOVED***
                value += rt * Math.pow(2, 1 - eBias);
            ***REMOVED***
            if (value * c >= 2) ***REMOVED***
                e++;
                c /= 2;
            ***REMOVED***

            if (e + eBias >= eMax) ***REMOVED***
                m = 0;
                e = eMax;
            ***REMOVED*** else if (e + eBias >= 1) ***REMOVED***
                m = (value * c - 1) * Math.pow(2, mLen);
                e = e + eBias;
            ***REMOVED*** else ***REMOVED***
                m = value * Math.pow(2, eBias - 1) * Math.pow(2, mLen);
                e = 0;
            ***REMOVED***
        ***REMOVED***

        for (; mLen >= 8; buffer[offset + i] = m & 0xff, i += d, m /= 256, mLen -= 8) ***REMOVED******REMOVED***

        e = (e << mLen) | m;
        eLen += mLen;
        for (; eLen > 0; buffer[offset + i] = e & 0xff, i += d, e /= 256, eLen -= 8) ***REMOVED******REMOVED***

        buffer[offset + i - d] |= s * 128;
    ***REMOVED***

    /**
     * Writes a 32bit float.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeFloat32 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number')
                throw TypeError("Illegal value: "+value+" (not a number)");
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 4;
        var capacity8 = this.buffer.byteLength;
        if (offset > capacity8)
            this.resize((capacity8 *= 2) > offset ? capacity8 : offset);
        offset -= 4;
        ieee754_write(this.view, value, offset, this.littleEndian, 23, 4);
        if (relative) this.offset += 4;
        return this;
    ***REMOVED***;

    /**
     * Writes a 32bit float. This is an alias of ***REMOVED***@link ByteBuffer#writeFloat32***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeFloat = ByteBufferPrototype.writeFloat32;

    /**
     * Reads a 32bit float.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***number***REMOVED***
     * @expose
     */
    ByteBufferPrototype.readFloat32 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 4 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+4+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = ieee754_read(this.view, offset, this.littleEndian, 23, 4);
        if (relative) this.offset += 4;
        return value;
    ***REMOVED***;

    /**
     * Reads a 32bit float. This is an alias of ***REMOVED***@link ByteBuffer#readFloat32***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `4` if omitted.
     * @returns ***REMOVED***number***REMOVED***
     * @expose
     */
    ByteBufferPrototype.readFloat = ByteBufferPrototype.readFloat32;

    // types/floats/float64

    /**
     * Writes a 64bit float.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeFloat64 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number')
                throw TypeError("Illegal value: "+value+" (not a number)");
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        offset += 8;
        var capacity9 = this.buffer.byteLength;
        if (offset > capacity9)
            this.resize((capacity9 *= 2) > offset ? capacity9 : offset);
        offset -= 8;
        ieee754_write(this.view, value, offset, this.littleEndian, 52, 8);
        if (relative) this.offset += 8;
        return this;
    ***REMOVED***;

    /**
     * Writes a 64bit float. This is an alias of ***REMOVED***@link ByteBuffer#writeFloat64***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.writeDouble = ByteBufferPrototype.writeFloat64;

    /**
     * Reads a 64bit float.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
     * @returns ***REMOVED***number***REMOVED***
     * @expose
     */
    ByteBufferPrototype.readFloat64 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 8 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+8+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var value = ieee754_read(this.view, offset, this.littleEndian, 52, 8);
        if (relative) this.offset += 8;
        return value;
    ***REMOVED***;

    /**
     * Reads a 64bit float. This is an alias of ***REMOVED***@link ByteBuffer#readFloat64***REMOVED***.
     * @function
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by `8` if omitted.
     * @returns ***REMOVED***number***REMOVED***
     * @expose
     */
    ByteBufferPrototype.readDouble = ByteBufferPrototype.readFloat64;


    // types/varints/varint32

    /**
     * Maximum number of bytes required to store a 32bit base 128 variable-length integer.
     * @type ***REMOVED***number***REMOVED***
     * @const
     * @expose
     */
    ByteBuffer.MAX_VARINT32_BYTES = 5;

    /**
     * Calculates the actual number of bytes required to store a 32bit base 128 variable-length integer.
     * @param ***REMOVED***number***REMOVED*** value Value to encode
     * @returns ***REMOVED***number***REMOVED*** Number of bytes required. Capped to ***REMOVED***@link ByteBuffer.MAX_VARINT32_BYTES***REMOVED***
     * @expose
     */
    ByteBuffer.calculateVarint32 = function(value) ***REMOVED***
        // ref: src/google/protobuf/io/coded_stream.cc
        value = value >>> 0;
             if (value < 1 << 7 ) return 1;
        else if (value < 1 << 14) return 2;
        else if (value < 1 << 21) return 3;
        else if (value < 1 << 28) return 4;
        else                      return 5;
    ***REMOVED***;

    /**
     * Zigzag encodes a signed 32bit integer so that it can be effectively used with varint encoding.
     * @param ***REMOVED***number***REMOVED*** n Signed 32bit integer
     * @returns ***REMOVED***number***REMOVED*** Unsigned zigzag encoded 32bit integer
     * @expose
     */
    ByteBuffer.zigZagEncode32 = function(n) ***REMOVED***
        return (((n |= 0) << 1) ^ (n >> 31)) >>> 0; // ref: src/google/protobuf/wire_format_lite.h
    ***REMOVED***;

    /**
     * Decodes a zigzag encoded signed 32bit integer.
     * @param ***REMOVED***number***REMOVED*** n Unsigned zigzag encoded 32bit integer
     * @returns ***REMOVED***number***REMOVED*** Signed 32bit integer
     * @expose
     */
    ByteBuffer.zigZagDecode32 = function(n) ***REMOVED***
        return ((n >>> 1) ^ -(n & 1)) | 0; // // ref: src/google/protobuf/wire_format_lite.h
    ***REMOVED***;

    /**
     * Writes a 32bit base 128 variable-length integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** this if `offset` is omitted, else the actual number of bytes written
     * @expose
     */
    ByteBufferPrototype.writeVarint32 = function(value, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value |= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var size = ByteBuffer.calculateVarint32(value),
            b;
        offset += size;
        var capacity10 = this.buffer.byteLength;
        if (offset > capacity10)
            this.resize((capacity10 *= 2) > offset ? capacity10 : offset);
        offset -= size;
        value >>>= 0;
        while (value >= 0x80) ***REMOVED***
            b = (value & 0x7f) | 0x80;
            this.view[offset++] = b;
            value >>>= 7;
        ***REMOVED***
        this.view[offset++] = value;
        if (relative) ***REMOVED***
            this.offset = offset;
            return this;
        ***REMOVED***
        return size;
    ***REMOVED***;

    /**
     * Writes a zig-zag encoded (signed) 32bit base 128 variable-length integer.
     * @param ***REMOVED***number***REMOVED*** value Value to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** this if `offset` is omitted, else the actual number of bytes written
     * @expose
     */
    ByteBufferPrototype.writeVarint32ZigZag = function(value, offset) ***REMOVED***
        return this.writeVarint32(ByteBuffer.zigZagEncode32(value), offset);
    ***REMOVED***;

    /**
     * Reads a 32bit base 128 variable-length integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***number|!***REMOVED***value: number, length: number***REMOVED******REMOVED*** The value read if offset is omitted, else the value read
     *  and the actual number of bytes read.
     * @throws ***REMOVED***Error***REMOVED*** If it's not a valid varint. Has a property `truncated = true` if there is not enough data available
     *  to fully decode the varint.
     * @expose
     */
    ByteBufferPrototype.readVarint32 = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 1 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+1+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var c = 0,
            value = 0 >>> 0,
            b;
        do ***REMOVED***
            if (!this.noAssert && offset > this.limit) ***REMOVED***
                var err = Error("Truncated");
                err['truncated'] = true;
                throw err;
            ***REMOVED***
            b = this.view[offset++];
            if (c < 5)
                value |= (b & 0x7f) << (7*c);
            ++c;
        ***REMOVED*** while ((b & 0x80) !== 0);
        value |= 0;
        if (relative) ***REMOVED***
            this.offset = offset;
            return value;
        ***REMOVED***
        return ***REMOVED***
            "value": value,
            "length": c
        ***REMOVED***;
    ***REMOVED***;

    /**
     * Reads a zig-zag encoded (signed) 32bit base 128 variable-length integer.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***number|!***REMOVED***value: number, length: number***REMOVED******REMOVED*** The value read if offset is omitted, else the value read
     *  and the actual number of bytes read.
     * @throws ***REMOVED***Error***REMOVED*** If it's not a valid varint
     * @expose
     */
    ByteBufferPrototype.readVarint32ZigZag = function(offset) ***REMOVED***
        var val = this.readVarint32(offset);
        if (typeof val === 'object')
            val["value"] = ByteBuffer.zigZagDecode32(val["value"]);
        else
            val = ByteBuffer.zigZagDecode32(val);
        return val;
    ***REMOVED***;

    // types/varints/varint64

    if (Long) ***REMOVED***

        /**
         * Maximum number of bytes required to store a 64bit base 128 variable-length integer.
         * @type ***REMOVED***number***REMOVED***
         * @const
         * @expose
         */
        ByteBuffer.MAX_VARINT64_BYTES = 10;

        /**
         * Calculates the actual number of bytes required to store a 64bit base 128 variable-length integer.
         * @param ***REMOVED***number|!Long***REMOVED*** value Value to encode
         * @returns ***REMOVED***number***REMOVED*** Number of bytes required. Capped to ***REMOVED***@link ByteBuffer.MAX_VARINT64_BYTES***REMOVED***
         * @expose
         */
        ByteBuffer.calculateVarint64 = function(value) ***REMOVED***
            if (typeof value === 'number')
                value = Long.fromNumber(value);
            else if (typeof value === 'string')
                value = Long.fromString(value);
            // ref: src/google/protobuf/io/coded_stream.cc
            var part0 = value.toInt() >>> 0,
                part1 = value.shiftRightUnsigned(28).toInt() >>> 0,
                part2 = value.shiftRightUnsigned(56).toInt() >>> 0;
            if (part2 == 0) ***REMOVED***
                if (part1 == 0) ***REMOVED***
                    if (part0 < 1 << 14)
                        return part0 < 1 << 7 ? 1 : 2;
                    else
                        return part0 < 1 << 21 ? 3 : 4;
                ***REMOVED*** else ***REMOVED***
                    if (part1 < 1 << 14)
                        return part1 < 1 << 7 ? 5 : 6;
                    else
                        return part1 < 1 << 21 ? 7 : 8;
                ***REMOVED***
            ***REMOVED*** else
                return part2 < 1 << 7 ? 9 : 10;
        ***REMOVED***;

        /**
         * Zigzag encodes a signed 64bit integer so that it can be effectively used with varint encoding.
         * @param ***REMOVED***number|!Long***REMOVED*** value Signed long
         * @returns ***REMOVED***!Long***REMOVED*** Unsigned zigzag encoded long
         * @expose
         */
        ByteBuffer.zigZagEncode64 = function(value) ***REMOVED***
            if (typeof value === 'number')
                value = Long.fromNumber(value, false);
            else if (typeof value === 'string')
                value = Long.fromString(value, false);
            else if (value.unsigned !== false) value = value.toSigned();
            // ref: src/google/protobuf/wire_format_lite.h
            return value.shiftLeft(1).xor(value.shiftRight(63)).toUnsigned();
        ***REMOVED***;

        /**
         * Decodes a zigzag encoded signed 64bit integer.
         * @param ***REMOVED***!Long|number***REMOVED*** value Unsigned zigzag encoded long or JavaScript number
         * @returns ***REMOVED***!Long***REMOVED*** Signed long
         * @expose
         */
        ByteBuffer.zigZagDecode64 = function(value) ***REMOVED***
            if (typeof value === 'number')
                value = Long.fromNumber(value, false);
            else if (typeof value === 'string')
                value = Long.fromString(value, false);
            else if (value.unsigned !== false) value = value.toSigned();
            // ref: src/google/protobuf/wire_format_lite.h
            return value.shiftRightUnsigned(1).xor(value.and(Long.ONE).toSigned().negate()).toSigned();
        ***REMOVED***;

        /**
         * Writes a 64bit base 128 variable-length integer.
         * @param ***REMOVED***number|Long***REMOVED*** value Value to write
         * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
         *  written if omitted.
         * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** `this` if offset is omitted, else the actual number of bytes written.
         * @expose
         */
        ByteBufferPrototype.writeVarint64 = function(value, offset) ***REMOVED***
            var relative = typeof offset === 'undefined';
            if (relative) offset = this.offset;
            if (!this.noAssert) ***REMOVED***
                if (typeof value === 'number')
                    value = Long.fromNumber(value);
                else if (typeof value === 'string')
                    value = Long.fromString(value);
                else if (!(value && value instanceof Long))
                    throw TypeError("Illegal value: "+value+" (not an integer or Long)");
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + 0 > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
            ***REMOVED***
            if (typeof value === 'number')
                value = Long.fromNumber(value, false);
            else if (typeof value === 'string')
                value = Long.fromString(value, false);
            else if (value.unsigned !== false) value = value.toSigned();
            var size = ByteBuffer.calculateVarint64(value),
                part0 = value.toInt() >>> 0,
                part1 = value.shiftRightUnsigned(28).toInt() >>> 0,
                part2 = value.shiftRightUnsigned(56).toInt() >>> 0;
            offset += size;
            var capacity11 = this.buffer.byteLength;
            if (offset > capacity11)
                this.resize((capacity11 *= 2) > offset ? capacity11 : offset);
            offset -= size;
            switch (size) ***REMOVED***
                case 10: this.view[offset+9] = (part2 >>>  7) & 0x01;
                case 9 : this.view[offset+8] = size !== 9 ? (part2       ) | 0x80 : (part2       ) & 0x7F;
                case 8 : this.view[offset+7] = size !== 8 ? (part1 >>> 21) | 0x80 : (part1 >>> 21) & 0x7F;
                case 7 : this.view[offset+6] = size !== 7 ? (part1 >>> 14) | 0x80 : (part1 >>> 14) & 0x7F;
                case 6 : this.view[offset+5] = size !== 6 ? (part1 >>>  7) | 0x80 : (part1 >>>  7) & 0x7F;
                case 5 : this.view[offset+4] = size !== 5 ? (part1       ) | 0x80 : (part1       ) & 0x7F;
                case 4 : this.view[offset+3] = size !== 4 ? (part0 >>> 21) | 0x80 : (part0 >>> 21) & 0x7F;
                case 3 : this.view[offset+2] = size !== 3 ? (part0 >>> 14) | 0x80 : (part0 >>> 14) & 0x7F;
                case 2 : this.view[offset+1] = size !== 2 ? (part0 >>>  7) | 0x80 : (part0 >>>  7) & 0x7F;
                case 1 : this.view[offset  ] = size !== 1 ? (part0       ) | 0x80 : (part0       ) & 0x7F;
            ***REMOVED***
            if (relative) ***REMOVED***
                this.offset += size;
                return this;
            ***REMOVED*** else ***REMOVED***
                return size;
            ***REMOVED***
        ***REMOVED***;

        /**
         * Writes a zig-zag encoded 64bit base 128 variable-length integer.
         * @param ***REMOVED***number|Long***REMOVED*** value Value to write
         * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
         *  written if omitted.
         * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** `this` if offset is omitted, else the actual number of bytes written.
         * @expose
         */
        ByteBufferPrototype.writeVarint64ZigZag = function(value, offset) ***REMOVED***
            return this.writeVarint64(ByteBuffer.zigZagEncode64(value), offset);
        ***REMOVED***;

        /**
         * Reads a 64bit base 128 variable-length integer. Requires Long.js.
         * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
         *  read if omitted.
         * @returns ***REMOVED***!Long|!***REMOVED***value: Long, length: number***REMOVED******REMOVED*** The value read if offset is omitted, else the value read and
         *  the actual number of bytes read.
         * @throws ***REMOVED***Error***REMOVED*** If it's not a valid varint
         * @expose
         */
        ByteBufferPrototype.readVarint64 = function(offset) ***REMOVED***
            var relative = typeof offset === 'undefined';
            if (relative) offset = this.offset;
            if (!this.noAssert) ***REMOVED***
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + 1 > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+1+") <= "+this.buffer.byteLength);
            ***REMOVED***
            // ref: src/google/protobuf/io/coded_stream.cc
            var start = offset,
                part0 = 0,
                part1 = 0,
                part2 = 0,
                b  = 0;
            b = this.view[offset++]; part0  = (b & 0x7F)      ; if ( b & 0x80                                                   ) ***REMOVED***
            b = this.view[offset++]; part0 |= (b & 0x7F) <<  7; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part0 |= (b & 0x7F) << 14; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part0 |= (b & 0x7F) << 21; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part1  = (b & 0x7F)      ; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part1 |= (b & 0x7F) <<  7; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part1 |= (b & 0x7F) << 14; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part1 |= (b & 0x7F) << 21; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part2  = (b & 0x7F)      ; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            b = this.view[offset++]; part2 |= (b & 0x7F) <<  7; if ((b & 0x80) || (this.noAssert && typeof b === 'undefined')) ***REMOVED***
            throw Error("Buffer overrun"); ***REMOVED******REMOVED******REMOVED******REMOVED******REMOVED******REMOVED******REMOVED******REMOVED******REMOVED******REMOVED***
            var value = Long.fromBits(part0 | (part1 << 28), (part1 >>> 4) | (part2) << 24, false);
            if (relative) ***REMOVED***
                this.offset = offset;
                return value;
            ***REMOVED*** else ***REMOVED***
                return ***REMOVED***
                    'value': value,
                    'length': offset-start
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***;

        /**
         * Reads a zig-zag encoded 64bit base 128 variable-length integer. Requires Long.js.
         * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
         *  read if omitted.
         * @returns ***REMOVED***!Long|!***REMOVED***value: Long, length: number***REMOVED******REMOVED*** The value read if offset is omitted, else the value read and
         *  the actual number of bytes read.
         * @throws ***REMOVED***Error***REMOVED*** If it's not a valid varint
         * @expose
         */
        ByteBufferPrototype.readVarint64ZigZag = function(offset) ***REMOVED***
            var val = this.readVarint64(offset);
            if (val && val['value'] instanceof Long)
                val["value"] = ByteBuffer.zigZagDecode64(val["value"]);
            else
                val = ByteBuffer.zigZagDecode64(val);
            return val;
        ***REMOVED***;

    ***REMOVED*** // Long


    // types/strings/cstring

    /**
     * Writes a NULL-terminated UTF8 encoded string. For this to work the specified string must not contain any NULL
     *  characters itself.
     * @param ***REMOVED***string***REMOVED*** str String to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  contained in `str` + 1 if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** this if offset is omitted, else the actual number of bytes written
     * @expose
     */
    ByteBufferPrototype.writeCString = function(str, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        var i,
            k = str.length;
        if (!this.noAssert) ***REMOVED***
            if (typeof str !== 'string')
                throw TypeError("Illegal str: Not a string");
            for (i=0; i<k; ++i) ***REMOVED***
                if (str.charCodeAt(i) === 0)
                    throw RangeError("Illegal str: Contains NULL-characters");
            ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        // UTF8 strings do not contain zero bytes in between except for the zero character, so:
        k = utfx.calculateUTF16asUTF8(stringSource(str))[1];
        offset += k+1;
        var capacity12 = this.buffer.byteLength;
        if (offset > capacity12)
            this.resize((capacity12 *= 2) > offset ? capacity12 : offset);
        offset -= k+1;
        utfx.encodeUTF16toUTF8(stringSource(str), function(b) ***REMOVED***
            this.view[offset++] = b;
        ***REMOVED***.bind(this));
        this.view[offset++] = 0;
        if (relative) ***REMOVED***
            this.offset = offset;
            return this;
        ***REMOVED***
        return k;
    ***REMOVED***;

    /**
     * Reads a NULL-terminated UTF8 encoded string. For this to work the string read must not contain any NULL characters
     *  itself.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  read if omitted.
     * @returns ***REMOVED***string|!***REMOVED***string: string, length: number***REMOVED******REMOVED*** The string read if offset is omitted, else the string
     *  read and the actual number of bytes read.
     * @expose
     */
    ByteBufferPrototype.readCString = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 1 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+1+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var start = offset,
            temp;
        // UTF8 strings do not contain zero bytes in between except for the zero character itself, so:
        var sd, b = -1;
        utfx.decodeUTF8toUTF16(function() ***REMOVED***
            if (b === 0) return null;
            if (offset >= this.limit)
                throw RangeError("Illegal range: Truncated data, "+offset+" < "+this.limit);
            b = this.view[offset++];
            return b === 0 ? null : b;
        ***REMOVED***.bind(this), sd = stringDestination(), true);
        if (relative) ***REMOVED***
            this.offset = offset;
            return sd();
        ***REMOVED*** else ***REMOVED***
            return ***REMOVED***
                "string": sd(),
                "length": offset - start
            ***REMOVED***;
        ***REMOVED***
    ***REMOVED***;

    // types/strings/istring

    /**
     * Writes a length as uint32 prefixed UTF8 encoded string.
     * @param ***REMOVED***string***REMOVED*** str String to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** `this` if `offset` is omitted, else the actual number of bytes written
     * @expose
     * @see ByteBuffer#writeVarint32
     */
    ByteBufferPrototype.writeIString = function(str, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof str !== 'string')
                throw TypeError("Illegal str: Not a string");
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var start = offset,
            k;
        k = utfx.calculateUTF16asUTF8(stringSource(str), this.noAssert)[1];
        offset += 4+k;
        var capacity13 = this.buffer.byteLength;
        if (offset > capacity13)
            this.resize((capacity13 *= 2) > offset ? capacity13 : offset);
        offset -= 4+k;
        if (this.littleEndian) ***REMOVED***
            this.view[offset+3] = (k >>> 24) & 0xFF;
            this.view[offset+2] = (k >>> 16) & 0xFF;
            this.view[offset+1] = (k >>>  8) & 0xFF;
            this.view[offset  ] =  k         & 0xFF;
        ***REMOVED*** else ***REMOVED***
            this.view[offset  ] = (k >>> 24) & 0xFF;
            this.view[offset+1] = (k >>> 16) & 0xFF;
            this.view[offset+2] = (k >>>  8) & 0xFF;
            this.view[offset+3] =  k         & 0xFF;
        ***REMOVED***
        offset += 4;
        utfx.encodeUTF16toUTF8(stringSource(str), function(b) ***REMOVED***
            this.view[offset++] = b;
        ***REMOVED***.bind(this));
        if (offset !== start + 4 + k)
            throw RangeError("Illegal range: Truncated data, "+offset+" == "+(offset+4+k));
        if (relative) ***REMOVED***
            this.offset = offset;
            return this;
        ***REMOVED***
        return offset - start;
    ***REMOVED***;

    /**
     * Reads a length as uint32 prefixed UTF8 encoded string.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  read if omitted.
     * @returns ***REMOVED***string|!***REMOVED***string: string, length: number***REMOVED******REMOVED*** The string read if offset is omitted, else the string
     *  read and the actual number of bytes read.
     * @expose
     * @see ByteBuffer#readVarint32
     */
    ByteBufferPrototype.readIString = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 4 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+4+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var start = offset;
        var len = this.readUint32(offset);
        var str = this.readUTF8String(len, ByteBuffer.METRICS_BYTES, offset += 4);
        offset += str['length'];
        if (relative) ***REMOVED***
            this.offset = offset;
            return str['string'];
        ***REMOVED*** else ***REMOVED***
            return ***REMOVED***
                'string': str['string'],
                'length': offset - start
            ***REMOVED***;
        ***REMOVED***
    ***REMOVED***;

    // types/strings/utf8string

    /**
     * Metrics representing number of UTF8 characters. Evaluates to `c`.
     * @type ***REMOVED***string***REMOVED***
     * @const
     * @expose
     */
    ByteBuffer.METRICS_CHARS = 'c';

    /**
     * Metrics representing number of bytes. Evaluates to `b`.
     * @type ***REMOVED***string***REMOVED***
     * @const
     * @expose
     */
    ByteBuffer.METRICS_BYTES = 'b';

    /**
     * Writes an UTF8 encoded string.
     * @param ***REMOVED***string***REMOVED*** str String to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** this if offset is omitted, else the actual number of bytes written.
     * @expose
     */
    ByteBufferPrototype.writeUTF8String = function(str, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var k;
        var start = offset;
        k = utfx.calculateUTF16asUTF8(stringSource(str))[1];
        offset += k;
        var capacity14 = this.buffer.byteLength;
        if (offset > capacity14)
            this.resize((capacity14 *= 2) > offset ? capacity14 : offset);
        offset -= k;
        utfx.encodeUTF16toUTF8(stringSource(str), function(b) ***REMOVED***
            this.view[offset++] = b;
        ***REMOVED***.bind(this));
        if (relative) ***REMOVED***
            this.offset = offset;
            return this;
        ***REMOVED***
        return offset - start;
    ***REMOVED***;

    /**
     * Writes an UTF8 encoded string. This is an alias of ***REMOVED***@link ByteBuffer#writeUTF8String***REMOVED***.
     * @function
     * @param ***REMOVED***string***REMOVED*** str String to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** this if offset is omitted, else the actual number of bytes written.
     * @expose
     */
    ByteBufferPrototype.writeString = ByteBufferPrototype.writeUTF8String;

    /**
     * Calculates the number of UTF8 characters of a string. JavaScript itself uses UTF-16, so that a string's
     *  `length` property does not reflect its actual UTF8 size if it contains code points larger than 0xFFFF.
     * @param ***REMOVED***string***REMOVED*** str String to calculate
     * @returns ***REMOVED***number***REMOVED*** Number of UTF8 characters
     * @expose
     */
    ByteBuffer.calculateUTF8Chars = function(str) ***REMOVED***
        return utfx.calculateUTF16asUTF8(stringSource(str))[0];
    ***REMOVED***;

    /**
     * Calculates the number of UTF8 bytes of a string.
     * @param ***REMOVED***string***REMOVED*** str String to calculate
     * @returns ***REMOVED***number***REMOVED*** Number of UTF8 bytes
     * @expose
     */
    ByteBuffer.calculateUTF8Bytes = function(str) ***REMOVED***
        return utfx.calculateUTF16asUTF8(stringSource(str))[1];
    ***REMOVED***;

    /**
     * Calculates the number of UTF8 bytes of a string. This is an alias of ***REMOVED***@link ByteBuffer.calculateUTF8Bytes***REMOVED***.
     * @function
     * @param ***REMOVED***string***REMOVED*** str String to calculate
     * @returns ***REMOVED***number***REMOVED*** Number of UTF8 bytes
     * @expose
     */
    ByteBuffer.calculateString = ByteBuffer.calculateUTF8Bytes;

    /**
     * Reads an UTF8 encoded string.
     * @param ***REMOVED***number***REMOVED*** length Number of characters or bytes to read.
     * @param ***REMOVED***string=***REMOVED*** metrics Metrics specifying what `length` is meant to count. Defaults to
     *  ***REMOVED***@link ByteBuffer.METRICS_CHARS***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  read if omitted.
     * @returns ***REMOVED***string|!***REMOVED***string: string, length: number***REMOVED******REMOVED*** The string read if offset is omitted, else the string
     *  read and the actual number of bytes read.
     * @expose
     */
    ByteBufferPrototype.readUTF8String = function(length, metrics, offset) ***REMOVED***
        if (typeof metrics === 'number') ***REMOVED***
            offset = metrics;
            metrics = undefined;
        ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (typeof metrics === 'undefined') metrics = ByteBuffer.METRICS_CHARS;
        if (!this.noAssert) ***REMOVED***
            if (typeof length !== 'number' || length % 1 !== 0)
                throw TypeError("Illegal length: "+length+" (not an integer)");
            length |= 0;
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var i = 0,
            start = offset,
            sd;
        if (metrics === ByteBuffer.METRICS_CHARS) ***REMOVED*** // The same for node and the browser
            sd = stringDestination();
            utfx.decodeUTF8(function() ***REMOVED***
                return i < length && offset < this.limit ? this.view[offset++] : null;
            ***REMOVED***.bind(this), function(cp) ***REMOVED***
                ++i; utfx.UTF8toUTF16(cp, sd);
            ***REMOVED***);
            if (i !== length)
                throw RangeError("Illegal range: Truncated data, "+i+" == "+length);
            if (relative) ***REMOVED***
                this.offset = offset;
                return sd();
            ***REMOVED*** else ***REMOVED***
                return ***REMOVED***
                    "string": sd(),
                    "length": offset - start
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED*** else if (metrics === ByteBuffer.METRICS_BYTES) ***REMOVED***
            if (!this.noAssert) ***REMOVED***
                if (typeof offset !== 'number' || offset % 1 !== 0)
                    throw TypeError("Illegal offset: "+offset+" (not an integer)");
                offset >>>= 0;
                if (offset < 0 || offset + length > this.buffer.byteLength)
                    throw RangeError("Illegal offset: 0 <= "+offset+" (+"+length+") <= "+this.buffer.byteLength);
            ***REMOVED***
            var k = offset + length;
            utfx.decodeUTF8toUTF16(function() ***REMOVED***
                return offset < k ? this.view[offset++] : null;
            ***REMOVED***.bind(this), sd = stringDestination(), this.noAssert);
            if (offset !== k)
                throw RangeError("Illegal range: Truncated data, "+offset+" == "+k);
            if (relative) ***REMOVED***
                this.offset = offset;
                return sd();
            ***REMOVED*** else ***REMOVED***
                return ***REMOVED***
                    'string': sd(),
                    'length': offset - start
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED*** else
            throw TypeError("Unsupported metrics: "+metrics);
    ***REMOVED***;

    /**
     * Reads an UTF8 encoded string. This is an alias of ***REMOVED***@link ByteBuffer#readUTF8String***REMOVED***.
     * @function
     * @param ***REMOVED***number***REMOVED*** length Number of characters or bytes to read
     * @param ***REMOVED***number=***REMOVED*** metrics Metrics specifying what `n` is meant to count. Defaults to
     *  ***REMOVED***@link ByteBuffer.METRICS_CHARS***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  read if omitted.
     * @returns ***REMOVED***string|!***REMOVED***string: string, length: number***REMOVED******REMOVED*** The string read if offset is omitted, else the string
     *  read and the actual number of bytes read.
     * @expose
     */
    ByteBufferPrototype.readString = ByteBufferPrototype.readUTF8String;

    // types/strings/vstring

    /**
     * Writes a length as varint32 prefixed UTF8 encoded string.
     * @param ***REMOVED***string***REMOVED*** str String to write
     * @param ***REMOVED***number=***REMOVED*** offset Offset to write to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***!ByteBuffer|number***REMOVED*** `this` if `offset` is omitted, else the actual number of bytes written
     * @expose
     * @see ByteBuffer#writeVarint32
     */
    ByteBufferPrototype.writeVString = function(str, offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof str !== 'string')
                throw TypeError("Illegal str: Not a string");
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var start = offset,
            k, l;
        k = utfx.calculateUTF16asUTF8(stringSource(str), this.noAssert)[1];
        l = ByteBuffer.calculateVarint32(k);
        offset += l+k;
        var capacity15 = this.buffer.byteLength;
        if (offset > capacity15)
            this.resize((capacity15 *= 2) > offset ? capacity15 : offset);
        offset -= l+k;
        offset += this.writeVarint32(k, offset);
        utfx.encodeUTF16toUTF8(stringSource(str), function(b) ***REMOVED***
            this.view[offset++] = b;
        ***REMOVED***.bind(this));
        if (offset !== start+k+l)
            throw RangeError("Illegal range: Truncated data, "+offset+" == "+(offset+k+l));
        if (relative) ***REMOVED***
            this.offset = offset;
            return this;
        ***REMOVED***
        return offset - start;
    ***REMOVED***;

    /**
     * Reads a length as varint32 prefixed UTF8 encoded string.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to read from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  read if omitted.
     * @returns ***REMOVED***string|!***REMOVED***string: string, length: number***REMOVED******REMOVED*** The string read if offset is omitted, else the string
     *  read and the actual number of bytes read.
     * @expose
     * @see ByteBuffer#readVarint32
     */
    ByteBufferPrototype.readVString = function(offset) ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 1 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+1+") <= "+this.buffer.byteLength);
        ***REMOVED***
        var start = offset;
        var len = this.readVarint32(offset);
        var str = this.readUTF8String(len['value'], ByteBuffer.METRICS_BYTES, offset += len['length']);
        offset += str['length'];
        if (relative) ***REMOVED***
            this.offset = offset;
            return str['string'];
        ***REMOVED*** else ***REMOVED***
            return ***REMOVED***
                'string': str['string'],
                'length': offset - start
            ***REMOVED***;
        ***REMOVED***
    ***REMOVED***;


    /**
     * Appends some data to this ByteBuffer. This will overwrite any contents behind the specified offset up to the appended
     *  data's length.
     * @param ***REMOVED***!ByteBuffer|!ArrayBuffer|!Uint8Array|string***REMOVED*** source Data to append. If `source` is a ByteBuffer, its offsets
     *  will be modified according to the performed read operation.
     * @param ***REMOVED***(string|number)=***REMOVED*** encoding Encoding if `data` is a string ("base64", "hex", "binary", defaults to "utf8")
     * @param ***REMOVED***number=***REMOVED*** offset Offset to append at. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     * @example A relative `<01 02>03.append(<04 05>)` will result in `<01 02 04 05>, 04 05|`
     * @example An absolute `<01 02>03.append(04 05>, 1)` will result in `<01 04>05, 04 05|`
     */
    ByteBufferPrototype.append = function(source, encoding, offset) ***REMOVED***
        if (typeof encoding === 'number' || typeof encoding !== 'string') ***REMOVED***
            offset = encoding;
            encoding = undefined;
        ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        if (!(source instanceof ByteBuffer))
            source = ByteBuffer.wrap(source, encoding);
        var length = source.limit - source.offset;
        if (length <= 0) return this; // Nothing to append
        offset += length;
        var capacity16 = this.buffer.byteLength;
        if (offset > capacity16)
            this.resize((capacity16 *= 2) > offset ? capacity16 : offset);
        offset -= length;
        this.view.set(source.view.subarray(source.offset, source.limit), offset);
        source.offset += length;
        if (relative) this.offset += length;
        return this;
    ***REMOVED***;

    /**
     * Appends this ByteBuffer's contents to another ByteBuffer. This will overwrite any contents at and after the
        specified offset up to the length of this ByteBuffer's data.
     * @param ***REMOVED***!ByteBuffer***REMOVED*** target Target ByteBuffer
     * @param ***REMOVED***number=***REMOVED*** offset Offset to append to. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  read if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     * @see ByteBuffer#append
     */
    ByteBufferPrototype.appendTo = function(target, offset) ***REMOVED***
        target.append(this, offset);
        return this;
    ***REMOVED***;

    /**
     * Enables or disables assertions of argument types and offsets. Assertions are enabled by default but you can opt to
     *  disable them if your code already makes sure that everything is valid.
     * @param ***REMOVED***boolean***REMOVED*** assert `true` to enable assertions, otherwise `false`
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.assert = function(assert) ***REMOVED***
        this.noAssert = !assert;
        return this;
    ***REMOVED***;

    /**
     * Gets the capacity of this ByteBuffer's backing buffer.
     * @returns ***REMOVED***number***REMOVED*** Capacity of the backing buffer
     * @expose
     */
    ByteBufferPrototype.capacity = function() ***REMOVED***
        return this.buffer.byteLength;
    ***REMOVED***;
    /**
     * Clears this ByteBuffer's offsets by setting ***REMOVED***@link ByteBuffer#offset***REMOVED*** to `0` and ***REMOVED***@link ByteBuffer#limit***REMOVED*** to the
     *  backing buffer's capacity. Discards ***REMOVED***@link ByteBuffer#markedOffset***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.clear = function() ***REMOVED***
        this.offset = 0;
        this.limit = this.buffer.byteLength;
        this.markedOffset = -1;
        return this;
    ***REMOVED***;

    /**
     * Creates a cloned instance of this ByteBuffer, preset with this ByteBuffer's values for ***REMOVED***@link ByteBuffer#offset***REMOVED***,
     *  ***REMOVED***@link ByteBuffer#markedOffset***REMOVED*** and ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** copy Whether to copy the backing buffer or to return another view on the same, defaults to `false`
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** Cloned instance
     * @expose
     */
    ByteBufferPrototype.clone = function(copy) ***REMOVED***
        var bb = new ByteBuffer(0, this.littleEndian, this.noAssert);
        if (copy) ***REMOVED***
            bb.buffer = new ArrayBuffer(this.buffer.byteLength);
            bb.view = new Uint8Array(bb.buffer);
        ***REMOVED*** else ***REMOVED***
            bb.buffer = this.buffer;
            bb.view = this.view;
        ***REMOVED***
        bb.offset = this.offset;
        bb.markedOffset = this.markedOffset;
        bb.limit = this.limit;
        return bb;
    ***REMOVED***;

    /**
     * Compacts this ByteBuffer to be backed by a ***REMOVED***@link ByteBuffer#buffer***REMOVED*** of its contents' length. Contents are the bytes
     *  between ***REMOVED***@link ByteBuffer#offset***REMOVED*** and ***REMOVED***@link ByteBuffer#limit***REMOVED***. Will set `offset = 0` and `limit = capacity` and
     *  adapt ***REMOVED***@link ByteBuffer#markedOffset***REMOVED*** to the same relative position if set.
     * @param ***REMOVED***number=***REMOVED*** begin Offset to start at, defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***
     * @param ***REMOVED***number=***REMOVED*** end Offset to end at, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.compact = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined') begin = this.offset;
        if (typeof end === 'undefined') end = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        if (begin === 0 && end === this.buffer.byteLength)
            return this; // Already compacted
        var len = end - begin;
        if (len === 0) ***REMOVED***
            this.buffer = EMPTY_BUFFER;
            this.view = null;
            if (this.markedOffset >= 0) this.markedOffset -= begin;
            this.offset = 0;
            this.limit = 0;
            return this;
        ***REMOVED***
        var buffer = new ArrayBuffer(len);
        var view = new Uint8Array(buffer);
        view.set(this.view.subarray(begin, end));
        this.buffer = buffer;
        this.view = view;
        if (this.markedOffset >= 0) this.markedOffset -= begin;
        this.offset = 0;
        this.limit = len;
        return this;
    ***REMOVED***;

    /**
     * Creates a copy of this ByteBuffer's contents. Contents are the bytes between ***REMOVED***@link ByteBuffer#offset***REMOVED*** and
     *  ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** begin Begin offset, defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** end End offset, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** Copy
     * @expose
     */
    ByteBufferPrototype.copy = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined') begin = this.offset;
        if (typeof end === 'undefined') end = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        if (begin === end)
            return new ByteBuffer(0, this.littleEndian, this.noAssert);
        var capacity = end - begin,
            bb = new ByteBuffer(capacity, this.littleEndian, this.noAssert);
        bb.offset = 0;
        bb.limit = capacity;
        if (bb.markedOffset >= 0) bb.markedOffset -= begin;
        this.copyTo(bb, 0, begin, end);
        return bb;
    ***REMOVED***;

    /**
     * Copies this ByteBuffer's contents to another ByteBuffer. Contents are the bytes between ***REMOVED***@link ByteBuffer#offset***REMOVED*** and
     *  ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @param ***REMOVED***!ByteBuffer***REMOVED*** target Target ByteBuffer
     * @param ***REMOVED***number=***REMOVED*** targetOffset Offset to copy to. Will use and increase the target's ***REMOVED***@link ByteBuffer#offset***REMOVED***
     *  by the number of bytes copied if omitted.
     * @param ***REMOVED***number=***REMOVED*** sourceOffset Offset to start copying from. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the
     *  number of bytes copied if omitted.
     * @param ***REMOVED***number=***REMOVED*** sourceLimit Offset to end copying from, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.copyTo = function(target, targetOffset, sourceOffset, sourceLimit) ***REMOVED***
        var relative,
            targetRelative;
        if (!this.noAssert) ***REMOVED***
            if (!ByteBuffer.isByteBuffer(target))
                throw TypeError("Illegal target: Not a ByteBuffer");
        ***REMOVED***
        targetOffset = (targetRelative = typeof targetOffset === 'undefined') ? target.offset : targetOffset | 0;
        sourceOffset = (relative = typeof sourceOffset === 'undefined') ? this.offset : sourceOffset | 0;
        sourceLimit = typeof sourceLimit === 'undefined' ? this.limit : sourceLimit | 0;

        if (targetOffset < 0 || targetOffset > target.buffer.byteLength)
            throw RangeError("Illegal target range: 0 <= "+targetOffset+" <= "+target.buffer.byteLength);
        if (sourceOffset < 0 || sourceLimit > this.buffer.byteLength)
            throw RangeError("Illegal source range: 0 <= "+sourceOffset+" <= "+this.buffer.byteLength);

        var len = sourceLimit - sourceOffset;
        if (len === 0)
            return target; // Nothing to copy

        target.ensureCapacity(targetOffset + len);

        target.view.set(this.view.subarray(sourceOffset, sourceLimit), targetOffset);

        if (relative) this.offset += len;
        if (targetRelative) target.offset += len;

        return this;
    ***REMOVED***;

    /**
     * Makes sure that this ByteBuffer is backed by a ***REMOVED***@link ByteBuffer#buffer***REMOVED*** of at least the specified capacity. If the
     *  current capacity is exceeded, it will be doubled. If double the current capacity is less than the required capacity,
     *  the required capacity will be used instead.
     * @param ***REMOVED***number***REMOVED*** capacity Required capacity
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.ensureCapacity = function(capacity) ***REMOVED***
        var current = this.buffer.byteLength;
        if (current < capacity)
            return this.resize((current *= 2) > capacity ? current : capacity);
        return this;
    ***REMOVED***;

    /**
     * Overwrites this ByteBuffer's contents with the specified value. Contents are the bytes between
     *  ***REMOVED***@link ByteBuffer#offset***REMOVED*** and ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @param ***REMOVED***number|string***REMOVED*** value Byte value to fill with. If given as a string, the first character is used.
     * @param ***REMOVED***number=***REMOVED*** begin Begin offset. Will use and increase ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  written if omitted. defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** end End offset, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     * @example `someByteBuffer.clear().fill(0)` fills the entire backing buffer with zeroes
     */
    ByteBufferPrototype.fill = function(value, begin, end) ***REMOVED***
        var relative = typeof begin === 'undefined';
        if (relative) begin = this.offset;
        if (typeof value === 'string' && value.length > 0)
            value = value.charCodeAt(0);
        if (typeof begin === 'undefined') begin = this.offset;
        if (typeof end === 'undefined') end = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof value !== 'number' || value % 1 !== 0)
                throw TypeError("Illegal value: "+value+" (not an integer)");
            value |= 0;
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        if (begin >= end)
            return this; // Nothing to fill
        while (begin < end) this.view[begin++] = value;
        if (relative) this.offset = begin;
        return this;
    ***REMOVED***;

    /**
     * Makes this ByteBuffer ready for a new sequence of write or relative read operations. Sets `limit = offset` and
     *  `offset = 0`. Make sure always to flip a ByteBuffer when all relative read or write operations are complete.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.flip = function() ***REMOVED***
        this.limit = this.offset;
        this.offset = 0;
        return this;
    ***REMOVED***;
    /**
     * Marks an offset on this ByteBuffer to be used later.
     * @param ***REMOVED***number=***REMOVED*** offset Offset to mark. Defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @throws ***REMOVED***TypeError***REMOVED*** If `offset` is not a valid number
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset` is out of bounds
     * @see ByteBuffer#reset
     * @expose
     */
    ByteBufferPrototype.mark = function(offset) ***REMOVED***
        offset = typeof offset === 'undefined' ? this.offset : offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        this.markedOffset = offset;
        return this;
    ***REMOVED***;
    /**
     * Sets the byte order.
     * @param ***REMOVED***boolean***REMOVED*** littleEndian `true` for little endian byte order, `false` for big endian
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.order = function(littleEndian) ***REMOVED***
        if (!this.noAssert) ***REMOVED***
            if (typeof littleEndian !== 'boolean')
                throw TypeError("Illegal littleEndian: Not a boolean");
        ***REMOVED***
        this.littleEndian = !!littleEndian;
        return this;
    ***REMOVED***;

    /**
     * Switches (to) little endian byte order.
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Defaults to `true`, otherwise uses big endian
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.LE = function(littleEndian) ***REMOVED***
        this.littleEndian = typeof littleEndian !== 'undefined' ? !!littleEndian : true;
        return this;
    ***REMOVED***;

    /**
     * Switches (to) big endian byte order.
     * @param ***REMOVED***boolean=***REMOVED*** bigEndian Defaults to `true`, otherwise uses little endian
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.BE = function(bigEndian) ***REMOVED***
        this.littleEndian = typeof bigEndian !== 'undefined' ? !bigEndian : false;
        return this;
    ***REMOVED***;
    /**
     * Prepends some data to this ByteBuffer. This will overwrite any contents before the specified offset up to the
     *  prepended data's length. If there is not enough space available before the specified `offset`, the backing buffer
     *  will be resized and its contents moved accordingly.
     * @param ***REMOVED***!ByteBuffer|string|!ArrayBuffer***REMOVED*** source Data to prepend. If `source` is a ByteBuffer, its offset will be
     *  modified according to the performed read operation.
     * @param ***REMOVED***(string|number)=***REMOVED*** encoding Encoding if `data` is a string ("base64", "hex", "binary", defaults to "utf8")
     * @param ***REMOVED***number=***REMOVED*** offset Offset to prepend at. Will use and decrease ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  prepended if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     * @example A relative `00<01 02 03>.prepend(<04 05>)` results in `<04 05 01 02 03>, 04 05|`
     * @example An absolute `00<01 02 03>.prepend(<04 05>, 2)` results in `04<05 02 03>, 04 05|`
     */
    ByteBufferPrototype.prepend = function(source, encoding, offset) ***REMOVED***
        if (typeof encoding === 'number' || typeof encoding !== 'string') ***REMOVED***
            offset = encoding;
            encoding = undefined;
        ***REMOVED***
        var relative = typeof offset === 'undefined';
        if (relative) offset = this.offset;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: "+offset+" (not an integer)");
            offset >>>= 0;
            if (offset < 0 || offset + 0 > this.buffer.byteLength)
                throw RangeError("Illegal offset: 0 <= "+offset+" (+"+0+") <= "+this.buffer.byteLength);
        ***REMOVED***
        if (!(source instanceof ByteBuffer))
            source = ByteBuffer.wrap(source, encoding);
        var len = source.limit - source.offset;
        if (len <= 0) return this; // Nothing to prepend
        var diff = len - offset;
        if (diff > 0) ***REMOVED*** // Not enough space before offset, so resize + move
            var buffer = new ArrayBuffer(this.buffer.byteLength + diff);
            var view = new Uint8Array(buffer);
            view.set(this.view.subarray(offset, this.buffer.byteLength), len);
            this.buffer = buffer;
            this.view = view;
            this.offset += diff;
            if (this.markedOffset >= 0) this.markedOffset += diff;
            this.limit += diff;
            offset += diff;
        ***REMOVED*** else ***REMOVED***
            var arrayView = new Uint8Array(this.buffer);
        ***REMOVED***
        this.view.set(source.view.subarray(source.offset, source.limit), offset - len);

        source.offset = source.limit;
        if (relative)
            this.offset -= len;
        return this;
    ***REMOVED***;

    /**
     * Prepends this ByteBuffer to another ByteBuffer. This will overwrite any contents before the specified offset up to the
     *  prepended data's length. If there is not enough space available before the specified `offset`, the backing buffer
     *  will be resized and its contents moved accordingly.
     * @param ***REMOVED***!ByteBuffer***REMOVED*** target Target ByteBuffer
     * @param ***REMOVED***number=***REMOVED*** offset Offset to prepend at. Will use and decrease ***REMOVED***@link ByteBuffer#offset***REMOVED*** by the number of bytes
     *  prepended if omitted.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     * @see ByteBuffer#prepend
     */
    ByteBufferPrototype.prependTo = function(target, offset) ***REMOVED***
        target.prepend(this, offset);
        return this;
    ***REMOVED***;
    /**
     * Prints debug information about this ByteBuffer's contents.
     * @param ***REMOVED***function(string)=***REMOVED*** out Output function to call, defaults to console.log
     * @expose
     */
    ByteBufferPrototype.printDebug = function(out) ***REMOVED***
        if (typeof out !== 'function') out = console.log.bind(console);
        out(
            this.toString()+"\n"+
            "-------------------------------------------------------------------\n"+
            this.toDebug(/* columns */ true)
        );
    ***REMOVED***;

    /**
     * Gets the number of remaining readable bytes. Contents are the bytes between ***REMOVED***@link ByteBuffer#offset***REMOVED*** and
     *  ***REMOVED***@link ByteBuffer#limit***REMOVED***, so this returns `limit - offset`.
     * @returns ***REMOVED***number***REMOVED*** Remaining readable bytes. May be negative if `offset > limit`.
     * @expose
     */
    ByteBufferPrototype.remaining = function() ***REMOVED***
        return this.limit - this.offset;
    ***REMOVED***;
    /**
     * Resets this ByteBuffer's ***REMOVED***@link ByteBuffer#offset***REMOVED***. If an offset has been marked through ***REMOVED***@link ByteBuffer#mark***REMOVED***
     *  before, `offset` will be set to ***REMOVED***@link ByteBuffer#markedOffset***REMOVED***, which will then be discarded. If no offset has been
     *  marked, sets `offset = 0`.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @see ByteBuffer#mark
     * @expose
     */
    ByteBufferPrototype.reset = function() ***REMOVED***
        if (this.markedOffset >= 0) ***REMOVED***
            this.offset = this.markedOffset;
            this.markedOffset = -1;
        ***REMOVED*** else ***REMOVED***
            this.offset = 0;
        ***REMOVED***
        return this;
    ***REMOVED***;
    /**
     * Resizes this ByteBuffer to be backed by a buffer of at least the given capacity. Will do nothing if already that
     *  large or larger.
     * @param ***REMOVED***number***REMOVED*** capacity Capacity required
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @throws ***REMOVED***TypeError***REMOVED*** If `capacity` is not a number
     * @throws ***REMOVED***RangeError***REMOVED*** If `capacity < 0`
     * @expose
     */
    ByteBufferPrototype.resize = function(capacity) ***REMOVED***
        if (!this.noAssert) ***REMOVED***
            if (typeof capacity !== 'number' || capacity % 1 !== 0)
                throw TypeError("Illegal capacity: "+capacity+" (not an integer)");
            capacity |= 0;
            if (capacity < 0)
                throw RangeError("Illegal capacity: 0 <= "+capacity);
        ***REMOVED***
        if (this.buffer.byteLength < capacity) ***REMOVED***
            var buffer = new ArrayBuffer(capacity);
            var view = new Uint8Array(buffer);
            view.set(this.view);
            this.buffer = buffer;
            this.view = view;
        ***REMOVED***
        return this;
    ***REMOVED***;
    /**
     * Reverses this ByteBuffer's contents.
     * @param ***REMOVED***number=***REMOVED*** begin Offset to start at, defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***
     * @param ***REMOVED***number=***REMOVED*** end Offset to end at, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.reverse = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined') begin = this.offset;
        if (typeof end === 'undefined') end = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        if (begin === end)
            return this; // Nothing to reverse
        Array.prototype.reverse.call(this.view.subarray(begin, end));
        return this;
    ***REMOVED***;
    /**
     * Skips the next `length` bytes. This will just advance
     * @param ***REMOVED***number***REMOVED*** length Number of bytes to skip. May also be negative to move the offset back.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** this
     * @expose
     */
    ByteBufferPrototype.skip = function(length) ***REMOVED***
        if (!this.noAssert) ***REMOVED***
            if (typeof length !== 'number' || length % 1 !== 0)
                throw TypeError("Illegal length: "+length+" (not an integer)");
            length |= 0;
        ***REMOVED***
        var offset = this.offset + length;
        if (!this.noAssert) ***REMOVED***
            if (offset < 0 || offset > this.buffer.byteLength)
                throw RangeError("Illegal length: 0 <= "+this.offset+" + "+length+" <= "+this.buffer.byteLength);
        ***REMOVED***
        this.offset = offset;
        return this;
    ***REMOVED***;

    /**
     * Slices this ByteBuffer by creating a cloned instance with `offset = begin` and `limit = end`.
     * @param ***REMOVED***number=***REMOVED*** begin Begin offset, defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** end End offset, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** Clone of this ByteBuffer with slicing applied, backed by the same ***REMOVED***@link ByteBuffer#buffer***REMOVED***
     * @expose
     */
    ByteBufferPrototype.slice = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined') begin = this.offset;
        if (typeof end === 'undefined') end = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        var bb = this.clone();
        bb.offset = begin;
        bb.limit = end;
        return bb;
    ***REMOVED***;
    /**
     * Returns a copy of the backing buffer that contains this ByteBuffer's contents. Contents are the bytes between
     *  ***REMOVED***@link ByteBuffer#offset***REMOVED*** and ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** forceCopy If `true` returns a copy, otherwise returns a view referencing the same memory if
     *  possible. Defaults to `false`
     * @returns ***REMOVED***!ArrayBuffer***REMOVED*** Contents as an ArrayBuffer
     * @expose
     */
    ByteBufferPrototype.toBuffer = function(forceCopy) ***REMOVED***
        var offset = this.offset,
            limit = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof offset !== 'number' || offset % 1 !== 0)
                throw TypeError("Illegal offset: Not an integer");
            offset >>>= 0;
            if (typeof limit !== 'number' || limit % 1 !== 0)
                throw TypeError("Illegal limit: Not an integer");
            limit >>>= 0;
            if (offset < 0 || offset > limit || limit > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+offset+" <= "+limit+" <= "+this.buffer.byteLength);
        ***REMOVED***
        // NOTE: It's not possible to have another ArrayBuffer reference the same memory as the backing buffer. This is
        // possible with Uint8Array#subarray only, but we have to return an ArrayBuffer by contract. So:
        if (!forceCopy && offset === 0 && limit === this.buffer.byteLength)
            return this.buffer;
        if (offset === limit)
            return EMPTY_BUFFER;
        var buffer = new ArrayBuffer(limit - offset);
        new Uint8Array(buffer).set(new Uint8Array(this.buffer).subarray(offset, limit), 0);
        return buffer;
    ***REMOVED***;

    /**
     * Returns a raw buffer compacted to contain this ByteBuffer's contents. Contents are the bytes between
     *  ***REMOVED***@link ByteBuffer#offset***REMOVED*** and ***REMOVED***@link ByteBuffer#limit***REMOVED***. This is an alias of ***REMOVED***@link ByteBuffer#toBuffer***REMOVED***.
     * @function
     * @param ***REMOVED***boolean=***REMOVED*** forceCopy If `true` returns a copy, otherwise returns a view referencing the same memory.
     *  Defaults to `false`
     * @returns ***REMOVED***!ArrayBuffer***REMOVED*** Contents as an ArrayBuffer
     * @expose
     */
    ByteBufferPrototype.toArrayBuffer = ByteBufferPrototype.toBuffer;

    /**
     * Converts the ByteBuffer's contents to a string.
     * @param ***REMOVED***string=***REMOVED*** encoding Output encoding. Returns an informative string representation if omitted but also allows
     *  direct conversion to "utf8", "hex", "base64" and "binary" encoding. "debug" returns a hex representation with
     *  highlighted offsets.
     * @param ***REMOVED***number=***REMOVED*** begin Offset to begin at, defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***
     * @param ***REMOVED***number=***REMOVED*** end Offset to end at, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***
     * @returns ***REMOVED***string***REMOVED*** String representation
     * @throws ***REMOVED***Error***REMOVED*** If `encoding` is invalid
     * @expose
     */
    ByteBufferPrototype.toString = function(encoding, begin, end) ***REMOVED***
        if (typeof encoding === 'undefined')
            return "ByteBufferAB(offset="+this.offset+",markedOffset="+this.markedOffset+",limit="+this.limit+",capacity="+this.capacity()+")";
        if (typeof encoding === 'number')
            encoding = "utf8",
            begin = encoding,
            end = begin;
        switch (encoding) ***REMOVED***
            case "utf8":
                return this.toUTF8(begin, end);
            case "base64":
                return this.toBase64(begin, end);
            case "hex":
                return this.toHex(begin, end);
            case "binary":
                return this.toBinary(begin, end);
            case "debug":
                return this.toDebug();
            case "columns":
                return this.toColumns();
            default:
                throw Error("Unsupported encoding: "+encoding);
        ***REMOVED***
    ***REMOVED***;

    // lxiv-embeddable

    /**
     * lxiv-embeddable (c) 2014 Daniel Wirtz <dcode@dcode.io>
     * Released under the Apache License, Version 2.0
     * see: https://github.com/dcodeIO/lxiv for details
     */
    var lxiv = function() ***REMOVED***
        "use strict";

        /**
         * lxiv namespace.
         * @type ***REMOVED***!Object.<string,*>***REMOVED***
         * @exports lxiv
         */
        var lxiv = ***REMOVED******REMOVED***;

        /**
         * Character codes for output.
         * @type ***REMOVED***!Array.<number>***REMOVED***
         * @inner
         */
        var aout = [
            65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80,
            81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 97, 98, 99, 100, 101, 102,
            103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118,
            119, 120, 121, 122, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 43, 47
        ];

        /**
         * Character codes for input.
         * @type ***REMOVED***!Array.<number>***REMOVED***
         * @inner
         */
        var ain = [];
        for (var i=0, k=aout.length; i<k; ++i)
            ain[aout[i]] = i;

        /**
         * Encodes bytes to base64 char codes.
         * @param ***REMOVED***!function():number|null***REMOVED*** src Bytes source as a function returning the next byte respectively `null` if
         *  there are no more bytes left.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Characters destination as a function successively called with each encoded char
         *  code.
         */
        lxiv.encode = function(src, dst) ***REMOVED***
            var b, t;
            while ((b = src()) !== null) ***REMOVED***
                dst(aout[(b>>2)&0x3f]);
                t = (b&0x3)<<4;
                if ((b = src()) !== null) ***REMOVED***
                    t |= (b>>4)&0xf;
                    dst(aout[(t|((b>>4)&0xf))&0x3f]);
                    t = (b&0xf)<<2;
                    if ((b = src()) !== null)
                        dst(aout[(t|((b>>6)&0x3))&0x3f]),
                        dst(aout[b&0x3f]);
                    else
                        dst(aout[t&0x3f]),
                        dst(61);
                ***REMOVED*** else
                    dst(aout[t&0x3f]),
                    dst(61),
                    dst(61);
            ***REMOVED***
        ***REMOVED***;

        /**
         * Decodes base64 char codes to bytes.
         * @param ***REMOVED***!function():number|null***REMOVED*** src Characters source as a function returning the next char code respectively
         *  `null` if there are no more characters left.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Bytes destination as a function successively called with the next byte.
         * @throws ***REMOVED***Error***REMOVED*** If a character code is invalid
         */
        lxiv.decode = function(src, dst) ***REMOVED***
            var c, t1, t2;
            function fail(c) ***REMOVED***
                throw Error("Illegal character code: "+c);
            ***REMOVED***
            while ((c = src()) !== null) ***REMOVED***
                t1 = ain[c];
                if (typeof t1 === 'undefined') fail(c);
                if ((c = src()) !== null) ***REMOVED***
                    t2 = ain[c];
                    if (typeof t2 === 'undefined') fail(c);
                    dst((t1<<2)>>>0|(t2&0x30)>>4);
                    if ((c = src()) !== null) ***REMOVED***
                        t1 = ain[c];
                        if (typeof t1 === 'undefined')
                            if (c === 61) break; else fail(c);
                        dst(((t2&0xf)<<4)>>>0|(t1&0x3c)>>2);
                        if ((c = src()) !== null) ***REMOVED***
                            t2 = ain[c];
                            if (typeof t2 === 'undefined')
                                if (c === 61) break; else fail(c);
                            dst(((t1&0x3)<<6)>>>0|t2);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

        /**
         * Tests if a string is valid base64.
         * @param ***REMOVED***string***REMOVED*** str String to test
         * @returns ***REMOVED***boolean***REMOVED*** `true` if valid, otherwise `false`
         */
        lxiv.test = function(str) ***REMOVED***
            return /^(?:[A-Za-z0-9+/]***REMOVED***4***REMOVED***)*(?:[A-Za-z0-9+/]***REMOVED***2***REMOVED***==|[A-Za-z0-9+/]***REMOVED***3***REMOVED***=)?$/.test(str);
        ***REMOVED***;

        return lxiv;
    ***REMOVED***();

    // encodings/base64

    /**
     * Encodes this ByteBuffer's contents to a base64 encoded string.
     * @param ***REMOVED***number=***REMOVED*** begin Offset to begin at, defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** end Offset to end at, defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @returns ***REMOVED***string***REMOVED*** Base64 encoded string
     * @throws ***REMOVED***RangeError***REMOVED*** If `begin` or `end` is out of bounds
     * @expose
     */
    ByteBufferPrototype.toBase64 = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined')
            begin = this.offset;
        if (typeof end === 'undefined')
            end = this.limit;
        begin = begin | 0; end = end | 0;
        if (begin < 0 || end > this.capacity || begin > end)
            throw RangeError("begin, end");
        var sd; lxiv.encode(function() ***REMOVED***
            return begin < end ? this.view[begin++] : null;
        ***REMOVED***.bind(this), sd = stringDestination());
        return sd();
    ***REMOVED***;

    /**
     * Decodes a base64 encoded string to a ByteBuffer.
     * @param ***REMOVED***string***REMOVED*** str String to decode
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** ByteBuffer
     * @expose
     */
    ByteBuffer.fromBase64 = function(str, littleEndian) ***REMOVED***
        if (typeof str !== 'string')
            throw TypeError("str");
        var bb = new ByteBuffer(str.length/4*3, littleEndian),
            i = 0;
        lxiv.decode(stringSource(str), function(b) ***REMOVED***
            bb.view[i++] = b;
        ***REMOVED***);
        bb.limit = i;
        return bb;
    ***REMOVED***;

    /**
     * Encodes a binary string to base64 like `window.btoa` does.
     * @param ***REMOVED***string***REMOVED*** str Binary string
     * @returns ***REMOVED***string***REMOVED*** Base64 encoded string
     * @see https://developer.mozilla.org/en-US/docs/Web/API/Window.btoa
     * @expose
     */
    ByteBuffer.btoa = function(str) ***REMOVED***
        return ByteBuffer.fromBinary(str).toBase64();
    ***REMOVED***;

    /**
     * Decodes a base64 encoded string to binary like `window.atob` does.
     * @param ***REMOVED***string***REMOVED*** b64 Base64 encoded string
     * @returns ***REMOVED***string***REMOVED*** Binary string
     * @see https://developer.mozilla.org/en-US/docs/Web/API/Window.atob
     * @expose
     */
    ByteBuffer.atob = function(b64) ***REMOVED***
        return ByteBuffer.fromBase64(b64).toBinary();
    ***REMOVED***;

    // encodings/binary

    /**
     * Encodes this ByteBuffer to a binary encoded string, that is using only characters 0x00-0xFF as bytes.
     * @param ***REMOVED***number=***REMOVED*** begin Offset to begin at. Defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** end Offset to end at. Defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @returns ***REMOVED***string***REMOVED*** Binary encoded string
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset > limit`
     * @expose
     */
    ByteBufferPrototype.toBinary = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined')
            begin = this.offset;
        if (typeof end === 'undefined')
            end = this.limit;
        begin |= 0; end |= 0;
        if (begin < 0 || end > this.capacity() || begin > end)
            throw RangeError("begin, end");
        if (begin === end)
            return "";
        var chars = [],
            parts = [];
        while (begin < end) ***REMOVED***
            chars.push(this.view[begin++]);
            if (chars.length >= 1024)
                parts.push(String.fromCharCode.apply(String, chars)),
                chars = [];
        ***REMOVED***
        return parts.join('') + String.fromCharCode.apply(String, chars);
    ***REMOVED***;

    /**
     * Decodes a binary encoded string, that is using only characters 0x00-0xFF as bytes, to a ByteBuffer.
     * @param ***REMOVED***string***REMOVED*** str String to decode
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** ByteBuffer
     * @expose
     */
    ByteBuffer.fromBinary = function(str, littleEndian) ***REMOVED***
        if (typeof str !== 'string')
            throw TypeError("str");
        var i = 0,
            k = str.length,
            charCode,
            bb = new ByteBuffer(k, littleEndian);
        while (i<k) ***REMOVED***
            charCode = str.charCodeAt(i);
            if (charCode > 0xff)
                throw RangeError("illegal char code: "+charCode);
            bb.view[i++] = charCode;
        ***REMOVED***
        bb.limit = k;
        return bb;
    ***REMOVED***;

    // encodings/debug

    /**
     * Encodes this ByteBuffer to a hex encoded string with marked offsets. Offset symbols are:
     * * `<` : offset,
     * * `'` : markedOffset,
     * * `>` : limit,
     * * `|` : offset and limit,
     * * `[` : offset and markedOffset,
     * * `]` : markedOffset and limit,
     * * `!` : offset, markedOffset and limit
     * @param ***REMOVED***boolean=***REMOVED*** columns If `true` returns two columns hex + ascii, defaults to `false`
     * @returns ***REMOVED***string|!Array.<string>***REMOVED*** Debug string or array of lines if `asArray = true`
     * @expose
     * @example `>00'01 02<03` contains four bytes with `limit=0, markedOffset=1, offset=3`
     * @example `00[01 02 03>` contains four bytes with `offset=markedOffset=1, limit=4`
     * @example `00|01 02 03` contains four bytes with `offset=limit=1, markedOffset=-1`
     * @example `|` contains zero bytes with `offset=limit=0, markedOffset=-1`
     */
    ByteBufferPrototype.toDebug = function(columns) ***REMOVED***
        var i = -1,
            k = this.buffer.byteLength,
            b,
            hex = "",
            asc = "",
            out = "";
        while (i<k) ***REMOVED***
            if (i !== -1) ***REMOVED***
                b = this.view[i];
                if (b < 0x10) hex += "0"+b.toString(16).toUpperCase();
                else hex += b.toString(16).toUpperCase();
                if (columns)
                    asc += b > 32 && b < 127 ? String.fromCharCode(b) : '.';
            ***REMOVED***
            ++i;
            if (columns) ***REMOVED***
                if (i > 0 && i % 16 === 0 && i !== k) ***REMOVED***
                    while (hex.length < 3*16+3) hex += " ";
                    out += hex+asc+"\n";
                    hex = asc = "";
                ***REMOVED***
            ***REMOVED***
            if (i === this.offset && i === this.limit)
                hex += i === this.markedOffset ? "!" : "|";
            else if (i === this.offset)
                hex += i === this.markedOffset ? "[" : "<";
            else if (i === this.limit)
                hex += i === this.markedOffset ? "]" : ">";
            else
                hex += i === this.markedOffset ? "'" : (columns || (i !== 0 && i !== k) ? " " : "");
        ***REMOVED***
        if (columns && hex !== " ") ***REMOVED***
            while (hex.length < 3*16+3)
                hex += " ";
            out += hex + asc + "\n";
        ***REMOVED***
        return columns ? out : hex;
    ***REMOVED***;

    /**
     * Decodes a hex encoded string with marked offsets to a ByteBuffer.
     * @param ***REMOVED***string***REMOVED*** str Debug string to decode (not be generated with `columns = true`)
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** ByteBuffer
     * @expose
     * @see ByteBuffer#toDebug
     */
    ByteBuffer.fromDebug = function(str, littleEndian, noAssert) ***REMOVED***
        var k = str.length,
            bb = new ByteBuffer(((k+1)/3)|0, littleEndian, noAssert);
        var i = 0, j = 0, ch, b,
            rs = false, // Require symbol next
            ho = false, hm = false, hl = false, // Already has offset (ho), markedOffset (hm), limit (hl)?
            fail = false;
        while (i<k) ***REMOVED***
            switch (ch = str.charAt(i++)) ***REMOVED***
                case '!':
                    if (!noAssert) ***REMOVED***
                        if (ho || hm || hl) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        ho = hm = hl = true;
                    ***REMOVED***
                    bb.offset = bb.markedOffset = bb.limit = j;
                    rs = false;
                    break;
                case '|':
                    if (!noAssert) ***REMOVED***
                        if (ho || hl) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        ho = hl = true;
                    ***REMOVED***
                    bb.offset = bb.limit = j;
                    rs = false;
                    break;
                case '[':
                    if (!noAssert) ***REMOVED***
                        if (ho || hm) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        ho = hm = true;
                    ***REMOVED***
                    bb.offset = bb.markedOffset = j;
                    rs = false;
                    break;
                case '<':
                    if (!noAssert) ***REMOVED***
                        if (ho) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        ho = true;
                    ***REMOVED***
                    bb.offset = j;
                    rs = false;
                    break;
                case ']':
                    if (!noAssert) ***REMOVED***
                        if (hl || hm) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        hl = hm = true;
                    ***REMOVED***
                    bb.limit = bb.markedOffset = j;
                    rs = false;
                    break;
                case '>':
                    if (!noAssert) ***REMOVED***
                        if (hl) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        hl = true;
                    ***REMOVED***
                    bb.limit = j;
                    rs = false;
                    break;
                case "'":
                    if (!noAssert) ***REMOVED***
                        if (hm) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                        hm = true;
                    ***REMOVED***
                    bb.markedOffset = j;
                    rs = false;
                    break;
                case ' ':
                    rs = false;
                    break;
                default:
                    if (!noAssert) ***REMOVED***
                        if (rs) ***REMOVED***
                            fail = true;
                            break;
                        ***REMOVED***
                    ***REMOVED***
                    b = parseInt(ch+str.charAt(i++), 16);
                    if (!noAssert) ***REMOVED***
                        if (isNaN(b) || b < 0 || b > 255)
                            throw TypeError("Illegal str: Not a debug encoded string");
                    ***REMOVED***
                    bb.view[j++] = b;
                    rs = true;
            ***REMOVED***
            if (fail)
                throw TypeError("Illegal str: Invalid symbol at "+i);
        ***REMOVED***
        if (!noAssert) ***REMOVED***
            if (!ho || !hl)
                throw TypeError("Illegal str: Missing offset or limit");
            if (j<bb.buffer.byteLength)
                throw TypeError("Illegal str: Not a debug encoded string (is it hex?) "+j+" < "+k);
        ***REMOVED***
        return bb;
    ***REMOVED***;

    // encodings/hex

    /**
     * Encodes this ByteBuffer's contents to a hex encoded string.
     * @param ***REMOVED***number=***REMOVED*** begin Offset to begin at. Defaults to ***REMOVED***@link ByteBuffer#offset***REMOVED***.
     * @param ***REMOVED***number=***REMOVED*** end Offset to end at. Defaults to ***REMOVED***@link ByteBuffer#limit***REMOVED***.
     * @returns ***REMOVED***string***REMOVED*** Hex encoded string
     * @expose
     */
    ByteBufferPrototype.toHex = function(begin, end) ***REMOVED***
        begin = typeof begin === 'undefined' ? this.offset : begin;
        end = typeof end === 'undefined' ? this.limit : end;
        if (!this.noAssert) ***REMOVED***
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        var out = new Array(end - begin),
            b;
        while (begin < end) ***REMOVED***
            b = this.view[begin++];
            if (b < 0x10)
                out.push("0", b.toString(16));
            else out.push(b.toString(16));
        ***REMOVED***
        return out.join('');
    ***REMOVED***;

    /**
     * Decodes a hex encoded string to a ByteBuffer.
     * @param ***REMOVED***string***REMOVED*** str String to decode
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** ByteBuffer
     * @expose
     */
    ByteBuffer.fromHex = function(str, littleEndian, noAssert) ***REMOVED***
        if (!noAssert) ***REMOVED***
            if (typeof str !== 'string')
                throw TypeError("Illegal str: Not a string");
            if (str.length % 2 !== 0)
                throw TypeError("Illegal str: Length not a multiple of 2");
        ***REMOVED***
        var k = str.length,
            bb = new ByteBuffer((k / 2) | 0, littleEndian),
            b;
        for (var i=0, j=0; i<k; i+=2) ***REMOVED***
            b = parseInt(str.substring(i, i+2), 16);
            if (!noAssert)
                if (!isFinite(b) || b < 0 || b > 255)
                    throw TypeError("Illegal str: Contains non-hex characters");
            bb.view[j++] = b;
        ***REMOVED***
        bb.limit = j;
        return bb;
    ***REMOVED***;

    // utfx-embeddable

    /**
     * utfx-embeddable (c) 2014 Daniel Wirtz <dcode@dcode.io>
     * Released under the Apache License, Version 2.0
     * see: https://github.com/dcodeIO/utfx for details
     */
    var utfx = function() ***REMOVED***
        "use strict";

        /**
         * utfx namespace.
         * @inner
         * @type ***REMOVED***!Object.<string,*>***REMOVED***
         */
        var utfx = ***REMOVED******REMOVED***;

        /**
         * Maximum valid code point.
         * @type ***REMOVED***number***REMOVED***
         * @const
         */
        utfx.MAX_CODEPOINT = 0x10FFFF;

        /**
         * Encodes UTF8 code points to UTF8 bytes.
         * @param ***REMOVED***(!function():number|null) | number***REMOVED*** src Code points source, either as a function returning the next code point
         *  respectively `null` if there are no more code points left or a single numeric code point.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Bytes destination as a function successively called with the next byte
         */
        utfx.encodeUTF8 = function(src, dst) ***REMOVED***
            var cp = null;
            if (typeof src === 'number')
                cp = src,
                src = function() ***REMOVED*** return null; ***REMOVED***;
            while (cp !== null || (cp = src()) !== null) ***REMOVED***
                if (cp < 0x80)
                    dst(cp&0x7F);
                else if (cp < 0x800)
                    dst(((cp>>6)&0x1F)|0xC0),
                    dst((cp&0x3F)|0x80);
                else if (cp < 0x10000)
                    dst(((cp>>12)&0x0F)|0xE0),
                    dst(((cp>>6)&0x3F)|0x80),
                    dst((cp&0x3F)|0x80);
                else
                    dst(((cp>>18)&0x07)|0xF0),
                    dst(((cp>>12)&0x3F)|0x80),
                    dst(((cp>>6)&0x3F)|0x80),
                    dst((cp&0x3F)|0x80);
                cp = null;
            ***REMOVED***
        ***REMOVED***;

        /**
         * Decodes UTF8 bytes to UTF8 code points.
         * @param ***REMOVED***!function():number|null***REMOVED*** src Bytes source as a function returning the next byte respectively `null` if there
         *  are no more bytes left.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Code points destination as a function successively called with each decoded code point.
         * @throws ***REMOVED***RangeError***REMOVED*** If a starting byte is invalid in UTF8
         * @throws ***REMOVED***Error***REMOVED*** If the last sequence is truncated. Has an array property `bytes` holding the
         *  remaining bytes.
         */
        utfx.decodeUTF8 = function(src, dst) ***REMOVED***
            var a, b, c, d, fail = function(b) ***REMOVED***
                b = b.slice(0, b.indexOf(null));
                var err = Error(b.toString());
                err.name = "TruncatedError";
                err['bytes'] = b;
                throw err;
            ***REMOVED***;
            while ((a = src()) !== null) ***REMOVED***
                if ((a&0x80) === 0)
                    dst(a);
                else if ((a&0xE0) === 0xC0)
                    ((b = src()) === null) && fail([a, b]),
                    dst(((a&0x1F)<<6) | (b&0x3F));
                else if ((a&0xF0) === 0xE0)
                    ((b=src()) === null || (c=src()) === null) && fail([a, b, c]),
                    dst(((a&0x0F)<<12) | ((b&0x3F)<<6) | (c&0x3F));
                else if ((a&0xF8) === 0xF0)
                    ((b=src()) === null || (c=src()) === null || (d=src()) === null) && fail([a, b, c ,d]),
                    dst(((a&0x07)<<18) | ((b&0x3F)<<12) | ((c&0x3F)<<6) | (d&0x3F));
                else throw RangeError("Illegal starting byte: "+a);
            ***REMOVED***
        ***REMOVED***;

        /**
         * Converts UTF16 characters to UTF8 code points.
         * @param ***REMOVED***!function():number|null***REMOVED*** src Characters source as a function returning the next char code respectively
         *  `null` if there are no more characters left.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Code points destination as a function successively called with each converted code
         *  point.
         */
        utfx.UTF16toUTF8 = function(src, dst) ***REMOVED***
            var c1, c2 = null;
            while (true) ***REMOVED***
                if ((c1 = c2 !== null ? c2 : src()) === null)
                    break;
                if (c1 >= 0xD800 && c1 <= 0xDFFF) ***REMOVED***
                    if ((c2 = src()) !== null) ***REMOVED***
                        if (c2 >= 0xDC00 && c2 <= 0xDFFF) ***REMOVED***
                            dst((c1-0xD800)*0x400+c2-0xDC00+0x10000);
                            c2 = null; continue;
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
                dst(c1);
            ***REMOVED***
            if (c2 !== null) dst(c2);
        ***REMOVED***;

        /**
         * Converts UTF8 code points to UTF16 characters.
         * @param ***REMOVED***(!function():number|null) | number***REMOVED*** src Code points source, either as a function returning the next code point
         *  respectively `null` if there are no more code points left or a single numeric code point.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Characters destination as a function successively called with each converted char code.
         * @throws ***REMOVED***RangeError***REMOVED*** If a code point is out of range
         */
        utfx.UTF8toUTF16 = function(src, dst) ***REMOVED***
            var cp = null;
            if (typeof src === 'number')
                cp = src, src = function() ***REMOVED*** return null; ***REMOVED***;
            while (cp !== null || (cp = src()) !== null) ***REMOVED***
                if (cp <= 0xFFFF)
                    dst(cp);
                else
                    cp -= 0x10000,
                    dst((cp>>10)+0xD800),
                    dst((cp%0x400)+0xDC00);
                cp = null;
            ***REMOVED***
        ***REMOVED***;

        /**
         * Converts and encodes UTF16 characters to UTF8 bytes.
         * @param ***REMOVED***!function():number|null***REMOVED*** src Characters source as a function returning the next char code respectively `null`
         *  if there are no more characters left.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Bytes destination as a function successively called with the next byte.
         */
        utfx.encodeUTF16toUTF8 = function(src, dst) ***REMOVED***
            utfx.UTF16toUTF8(src, function(cp) ***REMOVED***
                utfx.encodeUTF8(cp, dst);
            ***REMOVED***);
        ***REMOVED***;

        /**
         * Decodes and converts UTF8 bytes to UTF16 characters.
         * @param ***REMOVED***!function():number|null***REMOVED*** src Bytes source as a function returning the next byte respectively `null` if there
         *  are no more bytes left.
         * @param ***REMOVED***!function(number)***REMOVED*** dst Characters destination as a function successively called with each converted char code.
         * @throws ***REMOVED***RangeError***REMOVED*** If a starting byte is invalid in UTF8
         * @throws ***REMOVED***Error***REMOVED*** If the last sequence is truncated. Has an array property `bytes` holding the remaining bytes.
         */
        utfx.decodeUTF8toUTF16 = function(src, dst) ***REMOVED***
            utfx.decodeUTF8(src, function(cp) ***REMOVED***
                utfx.UTF8toUTF16(cp, dst);
            ***REMOVED***);
        ***REMOVED***;

        /**
         * Calculates the byte length of an UTF8 code point.
         * @param ***REMOVED***number***REMOVED*** cp UTF8 code point
         * @returns ***REMOVED***number***REMOVED*** Byte length
         */
        utfx.calculateCodePoint = function(cp) ***REMOVED***
            return (cp < 0x80) ? 1 : (cp < 0x800) ? 2 : (cp < 0x10000) ? 3 : 4;
        ***REMOVED***;

        /**
         * Calculates the number of UTF8 bytes required to store UTF8 code points.
         * @param ***REMOVED***(!function():number|null)***REMOVED*** src Code points source as a function returning the next code point respectively
         *  `null` if there are no more code points left.
         * @returns ***REMOVED***number***REMOVED*** The number of UTF8 bytes required
         */
        utfx.calculateUTF8 = function(src) ***REMOVED***
            var cp, l=0;
            while ((cp = src()) !== null)
                l += (cp < 0x80) ? 1 : (cp < 0x800) ? 2 : (cp < 0x10000) ? 3 : 4;
            return l;
        ***REMOVED***;

        /**
         * Calculates the number of UTF8 code points respectively UTF8 bytes required to store UTF16 char codes.
         * @param ***REMOVED***(!function():number|null)***REMOVED*** src Characters source as a function returning the next char code respectively
         *  `null` if there are no more characters left.
         * @returns ***REMOVED***!Array.<number>***REMOVED*** The number of UTF8 code points at index 0 and the number of UTF8 bytes required at index 1.
         */
        utfx.calculateUTF16asUTF8 = function(src) ***REMOVED***
            var n=0, l=0;
            utfx.UTF16toUTF8(src, function(cp) ***REMOVED***
                ++n; l += (cp < 0x80) ? 1 : (cp < 0x800) ? 2 : (cp < 0x10000) ? 3 : 4;
            ***REMOVED***);
            return [n,l];
        ***REMOVED***;

        return utfx;
    ***REMOVED***();

    // encodings/utf8

    /**
     * Encodes this ByteBuffer's contents between ***REMOVED***@link ByteBuffer#offset***REMOVED*** and ***REMOVED***@link ByteBuffer#limit***REMOVED*** to an UTF8 encoded
     *  string.
     * @returns ***REMOVED***string***REMOVED*** Hex encoded string
     * @throws ***REMOVED***RangeError***REMOVED*** If `offset > limit`
     * @expose
     */
    ByteBufferPrototype.toUTF8 = function(begin, end) ***REMOVED***
        if (typeof begin === 'undefined') begin = this.offset;
        if (typeof end === 'undefined') end = this.limit;
        if (!this.noAssert) ***REMOVED***
            if (typeof begin !== 'number' || begin % 1 !== 0)
                throw TypeError("Illegal begin: Not an integer");
            begin >>>= 0;
            if (typeof end !== 'number' || end % 1 !== 0)
                throw TypeError("Illegal end: Not an integer");
            end >>>= 0;
            if (begin < 0 || begin > end || end > this.buffer.byteLength)
                throw RangeError("Illegal range: 0 <= "+begin+" <= "+end+" <= "+this.buffer.byteLength);
        ***REMOVED***
        var sd; try ***REMOVED***
            utfx.decodeUTF8toUTF16(function() ***REMOVED***
                return begin < end ? this.view[begin++] : null;
            ***REMOVED***.bind(this), sd = stringDestination());
        ***REMOVED*** catch (e) ***REMOVED***
            if (begin !== end)
                throw RangeError("Illegal range: Truncated data, "+begin+" != "+end);
        ***REMOVED***
        return sd();
    ***REMOVED***;

    /**
     * Decodes an UTF8 encoded string to a ByteBuffer.
     * @param ***REMOVED***string***REMOVED*** str String to decode
     * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
     * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
     *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
     * @returns ***REMOVED***!ByteBuffer***REMOVED*** ByteBuffer
     * @expose
     */
    ByteBuffer.fromUTF8 = function(str, littleEndian, noAssert) ***REMOVED***
        if (!noAssert)
            if (typeof str !== 'string')
                throw TypeError("Illegal str: Not a string");
        var bb = new ByteBuffer(utfx.calculateUTF16asUTF8(stringSource(str), true)[1], littleEndian, noAssert),
            i = 0;
        utfx.encodeUTF16toUTF8(stringSource(str), function(b) ***REMOVED***
            bb.view[i++] = b;
        ***REMOVED***);
        bb.limit = i;
        return bb;
    ***REMOVED***;

    return ByteBuffer;
***REMOVED***);
