// call-signature/index.js
'use strict';
module.exports.parse = parse;
module.exports.generate = generate;

// TODO(jamestalmage): Allow full range of identifier characters instead of just ASCII
//
// This will likely require a build step
//
// SPEC: http://www.ecma-international.org/ecma-262/5.1/#sec-7.6
//
// TOOLING:
//    https://github.com/mathiasbynens/regenerate
//    https://www.npmjs.com/package/regjsgen

var regex = /^\s*(?:([A-Za-z$_][A-Za-z0-9$_]*)\s*\.)?\s*([A-Za-z$_][A-Za-z0-9$_]*)\s*\(\s*((?:[A-Za-z$_][A-Za-z0-9$_]*)|(?:\[\s*[A-Za-z$_][A-Za-z0-9$_]*\s*]))?((?:\s*,\s*(?:(?:[A-Za-z$_][A-Za-z0-9$_]*)|(?:\[\s*[A-Za-z$_][A-Za-z0-9$_]*\s*])))+)?\s*\)\s*$/;

function parse(str) ***REMOVED***
	var match = regex.exec(str);
	if (!match) ***REMOVED***
		return null;
	***REMOVED***

	var callee;
	if (match[1]) ***REMOVED***
		callee = ***REMOVED***
			type: 'MemberExpression',
			object: match[1],
			member: match[2]
		***REMOVED***;
	***REMOVED*** else ***REMOVED***
		callee = ***REMOVED***
			type: 'Identifier',
			name: match[2]
		***REMOVED***;
	***REMOVED***

	var args = match[4] || '';
	args = args.split(',');
	if (match[3]) ***REMOVED***
		args[0] = match[3];
	***REMOVED***
	var trimmed = [];
	args.forEach(function (str) ***REMOVED***
		var optional = false;
		str = str.replace(/\s+/g, '');
		if (!str.length) ***REMOVED***
			return;
		***REMOVED***
		if (str.charAt(0) === '[' && str.charAt(str.length - 1) === ']') ***REMOVED***
			optional = true;
			str = str.substring(1, str.length - 1);
		***REMOVED***
		trimmed.push(***REMOVED***
			name: str,
			optional: optional
		***REMOVED***);
	***REMOVED***);

	return ***REMOVED***
		callee: callee,
		args: trimmed
	***REMOVED***;
***REMOVED***

function generate(parsed) ***REMOVED***
	var callee;
	if (parsed.callee.type === 'MemberExpression') ***REMOVED***
		callee = [
			parsed.callee.object,
			'.',
			parsed.callee.member
		];
	***REMOVED*** else ***REMOVED***
		callee = [parsed.callee.name];
	***REMOVED***
	return callee.concat([
		'(',
		parsed.args.map(function (arg) ***REMOVED***
			return arg.optional ? '[' + arg.name + ']' : arg.name;
		***REMOVED***).join(', '),
		')'
	]).join('');
***REMOVED***
