// hawk/test/browser.js
// Load modules

var Url = require('url');
var Code = require('code');
var Hawk = require('../lib');
var Hoek = require('hoek');
var Lab = require('lab');
var Browser = require('../lib/browser');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.experiment;
var it = lab.test;
var expect = Code.expect;


describe('Browser', function () ***REMOVED***

    var credentialsFunc = function (id, callback) ***REMOVED***

        var credentials = ***REMOVED***
            id: id,
            key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
            algorithm: (id === '1' ? 'sha1' : 'sha256'),
            user: 'steve'
        ***REMOVED***;

        return callback(null, credentials);
    ***REMOVED***;

    it('should generate a bewit then successfully authenticate it', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?a=1&b=2',
            host: 'example.com',
            port: 80
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var bewit = Browser.client.bewit('http://example.com/resource/4?a=1&b=2', ***REMOVED*** credentials: credentials1, ttlSec: 60 * 60 * 24 * 365 * 100, ext: 'some-app-data' ***REMOVED***);
            req.url += '&bewit=' + bewit;

            Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, attributes) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(attributes.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('should generate a bewit then successfully authenticate it (no ext)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?a=1&b=2',
            host: 'example.com',
            port: 80
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var bewit = Browser.client.bewit('http://example.com/resource/4?a=1&b=2', ***REMOVED*** credentials: credentials1, ttlSec: 60 * 60 * 24 * 365 * 100 ***REMOVED***);
            req.url += '&bewit=' + bewit;

            Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, attributes) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('bewit()', function () ***REMOVED***

        it('returns a valid bewit value', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdca3NjeHdOUjJ0SnBQMVQxekRMTlBiQjVVaUtJVTl0T1NKWFRVZEc3WDloOD1ceGFuZHlhbmR6');
            done();
        ***REMOVED***);

        it('returns a valid bewit value (explicit HTTP port)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('http://example.com:8080/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdcaFpiSjNQMmNLRW80a3kwQzhqa1pBa1J5Q1p1ZWc0V1NOYnhWN3ZxM3hIVT1ceGFuZHlhbmR6');
            done();
        ***REMOVED***);

        it('returns a valid bewit value (explicit HTTPS port)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com:8043/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdcL2t4UjhwK0xSaTdvQTRnUXc3cWlxa3BiVHRKYkR4OEtRMC9HRUwvVytTUT1ceGFuZHlhbmR6');
            done();
        ***REMOVED***);

        it('returns a valid bewit value (null ext)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: null ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdcSUdZbUxnSXFMckNlOEN4dktQczRKbFdJQStValdKSm91d2dBUmlWaENBZz1c');
            done();
        ***REMOVED***);

        it('errors on invalid options', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', 4);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing uri', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid uri', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit(5, ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid credentials (id)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 3000, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing credentials', function (done) ***REMOVED***

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** ttlSec: 3000, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid credentials (key)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 3000, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid algorithm', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'hmac-sha-0'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing options', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'hmac-sha-0'
            ***REMOVED***;

            var bewit = Browser.client.bewit('https://example.com/somewhere/over/the/rainbow');
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (configuration)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data' ***REMOVED***).field;
            expect(req.authorization).to.exist();

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (node request)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
                expect(res.headers['server-authorization']).to.exist();

                expect(Browser.client.authenticate(res, credentials2, artifacts, ***REMOVED*** payload: 'some reply' ***REMOVED***)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (browserify)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***,
                    getHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
                expect(res.headers['server-authorization']).to.exist();

                expect(Browser.client.authenticate(res, credentials2, artifacts, ***REMOVED*** payload: 'some reply' ***REMOVED***)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (time offset)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', localtimeOffsetMsec: 100000 ***REMOVED***).field;
            expect(req.authorization).to.exist();

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 100000 ***REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (no server header options)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts);
                expect(res.headers['server-authorization']).to.exist();

                expect(Browser.client.authenticate(res, credentials2, artifacts)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (no server header)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                expect(Browser.client.authenticate(res, credentials2, artifacts)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header with stale ts and successfully authenticate on second call', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            Browser.utils.setNtpOffset(60 * 60 * 1000);
            var header = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data' ***REMOVED***);
            req.authorization = header.field;
            expect(req.authorization).to.exist();

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts2) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Stale timestamp');

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'www-authenticate': err.output.headers['WWW-Authenticate']
                    ***REMOVED***,
                    getResponseHeader: function (lookup) ***REMOVED***

                        return res.headers[lookup.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                expect(Browser.utils.getNtpOffset()).to.equal(60 * 60 * 1000);
                expect(Browser.client.authenticate(res, credentials2, header.artifacts)).to.equal(true);
                expect(Browser.utils.getNtpOffset()).to.equal(0);

                req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials2, ext: 'some-app-data' ***REMOVED***).field;
                expect(req.authorization).to.exist();

                Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials3, artifacts3) ***REMOVED***

                    expect(err).to.not.exist();
                    expect(credentials3.user).to.equal('steve');
                    expect(artifacts3.ext).to.equal('some-app-data');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header with stale ts and successfully authenticate on second call (manual localStorage)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var localStorage = new Browser.internals.LocalStorage();

            Browser.utils.setStorage(localStorage);

            Browser.utils.setNtpOffset(60 * 60 * 1000);
            var header = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data' ***REMOVED***);
            req.authorization = header.field;
            expect(req.authorization).to.exist();

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts2) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Stale timestamp');

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'www-authenticate': err.output.headers['WWW-Authenticate']
                    ***REMOVED***,
                    getResponseHeader: function (lookup) ***REMOVED***

                        return res.headers[lookup.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                expect(parseInt(localStorage.getItem('hawk_ntp_offset'))).to.equal(60 * 60 * 1000);
                expect(Browser.utils.getNtpOffset()).to.equal(60 * 60 * 1000);
                expect(Browser.client.authenticate(res, credentials2, header.artifacts)).to.equal(true);
                expect(Browser.utils.getNtpOffset()).to.equal(0);
                expect(parseInt(localStorage.getItem('hawk_ntp_offset'))).to.equal(0);

                req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials2, ext: 'some-app-data' ***REMOVED***).field;
                expect(req.authorization).to.exist();

                Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials3, artifacts3) ***REMOVED***

                    expect(err).to.not.exist();
                    expect(credentials3.user).to.equal('steve');
                    expect(artifacts3.ext).to.equal('some-app-data');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then fails to parse it (missing server header hash)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts);
                expect(res.headers['server-authorization']).to.exist();

                expect(Browser.client.authenticate(res, credentials2, artifacts, ***REMOVED*** payload: 'some reply' ***REMOVED***)).to.equal(false);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (with hash)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it then validate payload', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload('hola!', credentials2, artifacts)).to.be.true();
                expect(Hawk.server.authenticatePayload('hello!', credentials2, artifacts)).to.be.false();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (app)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', app: 'asd23ased' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(artifacts.app).to.equal('asd23ased');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (app, dlg)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', app: 'asd23ased', dlg: '23434szr3q4d' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(artifacts.app).to.equal('asd23ased');
                expect(artifacts.dlg).to.equal('23434szr3q4d');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then fail authentication due to bad hash', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** payload: 'byebye!' ***REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Bad payload hash');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header for one resource then fail to authenticate another', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Browser.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data' ***REMOVED***).field;
            req.url = '/something/else';

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(credentials2).to.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('client', function () ***REMOVED***

        describe('header()', function () ***REMOVED***

            it('returns a valid authorization header (sha1)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha1'
                ***REMOVED***;

                var header = Browser.client.header('http://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, ext: 'Bazinga!', timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about' ***REMOVED***).field;
                expect(header).to.equal('Hawk id="123456", ts="1353809207", nonce="Ygvqdz", hash="bsvY3IfUllw6V5rvk4tStEvpBhE=", ext="Bazinga!", mac="qbf1ZPG/r/e06F4ht+T77LXi5vw="');
                done();
            ***REMOVED***);

            it('returns a valid authorization header (sha256)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, ext: 'Bazinga!', timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***).field;
                expect(header).to.equal('Hawk id="123456", ts="1353809207", nonce="Ygvqdz", hash="2QfCt3GuY9HQnHWyWD3wX68ZOKbynqlfYmuO2ZBRqtY=", ext="Bazinga!", mac="q1CwFoSHzPZSkbIvl0oYlD+91rBUEvFk763nMjMndj8="');
                done();
            ***REMOVED***);

            it('returns a valid authorization header (empty payload)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha1'
                ***REMOVED***;

                var header = Browser.client.header('http://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, ext: 'Bazinga!', timestamp: 1353809207, nonce: 'Ygvqdz', payload: '' ***REMOVED***).field;
                expect(header).to.equal('Hawk id=\"123456\", ts=\"1353809207\", nonce=\"Ygvqdz\", hash=\"404ghL7K+hfyhByKKejFBRGgTjU=\", ext=\"Bazinga!\", mac=\"Bh1sj1DOfFRWOdi3ww52nLCJdBE=\"');
                done();
            ***REMOVED***);

            it('returns a valid authorization header (no ext)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***).field;
                expect(header).to.equal('Hawk id="123456", ts="1353809207", nonce="Ygvqdz", hash="2QfCt3GuY9HQnHWyWD3wX68ZOKbynqlfYmuO2ZBRqtY=", mac="HTgtd0jPI6E4izx8e4OHdO36q00xFCU0FolNq3RiCYs="');
                done();
            ***REMOVED***);

            it('returns a valid authorization header (null ext)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain', ext: null ***REMOVED***).field;
                expect(header).to.equal('Hawk id="123456", ts="1353809207", nonce="Ygvqdz", hash="2QfCt3GuY9HQnHWyWD3wX68ZOKbynqlfYmuO2ZBRqtY=", mac="HTgtd0jPI6E4izx8e4OHdO36q00xFCU0FolNq3RiCYs="');
                done();
            ***REMOVED***);

            it('returns a valid authorization header (uri object)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var uri = Browser.utils.parseUri('https://example.net/somewhere/over/the/rainbow');
                var header = Browser.client.header(uri, 'POST', ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***).field;
                expect(header).to.equal('Hawk id="123456", ts="1353809207", nonce="Ygvqdz", hash="2QfCt3GuY9HQnHWyWD3wX68ZOKbynqlfYmuO2ZBRqtY=", mac="HTgtd0jPI6E4izx8e4OHdO36q00xFCU0FolNq3RiCYs="');
                done();
            ***REMOVED***);

            it('errors on missing options', function (done) ***REMOVED***

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST');
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid argument type');
                done();
            ***REMOVED***);

            it('errors on empty uri', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('', 'POST', ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid argument type');
                done();
            ***REMOVED***);

            it('errors on invalid uri', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header(4, 'POST', ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid argument type');
                done();
            ***REMOVED***);

            it('errors on missing method', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', '', ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid argument type');
                done();
            ***REMOVED***);

            it('errors on invalid method', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 5, ***REMOVED*** credentials: credentials, timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid argument type');
                done();
            ***REMOVED***);

            it('errors on missing credentials', function (done) ***REMOVED***

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** ext: 'Bazinga!', timestamp: 1353809207 ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid credentials object');
                done();
            ***REMOVED***);

            it('errors on invalid credentials (id)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, ext: 'Bazinga!', timestamp: 1353809207 ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid credentials object');
                done();
            ***REMOVED***);

            it('errors on invalid credentials (key)', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    algorithm: 'sha256'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, ext: 'Bazinga!', timestamp: 1353809207 ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Invalid credentials object');
                done();
            ***REMOVED***);

            it('errors on invalid algorithm', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'hmac-sha-0'
                ***REMOVED***;

                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', ***REMOVED*** credentials: credentials, payload: 'something, anything!', ext: 'Bazinga!', timestamp: 1353809207 ***REMOVED***);
                expect(header.field).to.equal('');
                expect(header.err).to.equal('Unknown algorithm');
                done();
            ***REMOVED***);

            it('uses a pre-calculated payload hash', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: '2983d45yun89q',
                    algorithm: 'sha256'
                ***REMOVED***;

                var options = ***REMOVED*** credentials: credentials, ext: 'Bazinga!', timestamp: 1353809207, nonce: 'Ygvqdz', payload: 'something to write about', contentType: 'text/plain' ***REMOVED***;
                options.hash = Browser.crypto.calculatePayloadHash(options.payload, credentials.algorithm, options.contentType);
                var header = Browser.client.header('https://example.net/somewhere/over/the/rainbow', 'POST', options).field;
                expect(header).to.equal('Hawk id="123456", ts="1353809207", nonce="Ygvqdz", hash="2QfCt3GuY9HQnHWyWD3wX68ZOKbynqlfYmuO2ZBRqtY=", ext="Bazinga!", mac="q1CwFoSHzPZSkbIvl0oYlD+91rBUEvFk763nMjMndj8="');
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('authenticate()', function () ***REMOVED***

            it('skips tsm validation when missing ts', function (done) ***REMOVED***

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'www-authenticate': 'Hawk error="Stale timestamp"'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                var credentials = ***REMOVED***
                    id: '123456',
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    algorithm: 'sha256',
                    user: 'steve'
                ***REMOVED***;

                var artifacts = ***REMOVED***
                    ts: 1402135580,
                    nonce: 'iBRB6t',
                    method: 'GET',
                    resource: '/resource/4?filter=a',
                    host: 'example.com',
                    port: '8080',
                    ext: 'some-app-data'
                ***REMOVED***;

                expect(Browser.client.authenticate(res, credentials, artifacts)).to.equal(true);
                done();
            ***REMOVED***);

            it('returns false on invalid header', function (done) ***REMOVED***

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'server-authorization': 'Hawk mac="abc", bad="xyz"'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                expect(Browser.client.authenticate(res, ***REMOVED******REMOVED***)).to.equal(false);
                done();
            ***REMOVED***);

            it('returns false on invalid mac', function (done) ***REMOVED***

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain',
                        'server-authorization': 'Hawk mac="_IJRsMl/4oL+nn+vKoeVZPdCHXB4yJkNnBbTbHFZUYE=", hash="f9cDF/TDm7TkYRLnGwRMfeDzT6LixQVLvrIKhh0vgmM=", ext="response-specific"'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                var artifacts = ***REMOVED***
                    method: 'POST',
                    host: 'example.com',
                    port: '8080',
                    resource: '/resource/4?filter=a',
                    ts: '1362336900',
                    nonce: 'eb5S_L',
                    hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                    ext: 'some-app-data',
                    app: undefined,
                    dlg: undefined,
                    mac: 'BlmSe8K+pbKIb6YsZCnt4E1GrYvY1AaYayNR82dGpIk=',
                    id: '123456'
                ***REMOVED***;

                var credentials = ***REMOVED***
                    id: '123456',
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    algorithm: 'sha256',
                    user: 'steve'
                ***REMOVED***;

                expect(Browser.client.authenticate(res, credentials, artifacts)).to.equal(false);
                done();
            ***REMOVED***);

            it('returns true on ignoring hash', function (done) ***REMOVED***

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain',
                        'server-authorization': 'Hawk mac="XIJRsMl/4oL+nn+vKoeVZPdCHXB4yJkNnBbTbHFZUYE=", hash="f9cDF/TDm7TkYRLnGwRMfeDzT6LixQVLvrIKhh0vgmM=", ext="response-specific"'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                var artifacts = ***REMOVED***
                    method: 'POST',
                    host: 'example.com',
                    port: '8080',
                    resource: '/resource/4?filter=a',
                    ts: '1362336900',
                    nonce: 'eb5S_L',
                    hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                    ext: 'some-app-data',
                    app: undefined,
                    dlg: undefined,
                    mac: 'BlmSe8K+pbKIb6YsZCnt4E1GrYvY1AaYayNR82dGpIk=',
                    id: '123456'
                ***REMOVED***;

                var credentials = ***REMOVED***
                    id: '123456',
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    algorithm: 'sha256',
                    user: 'steve'
                ***REMOVED***;

                expect(Browser.client.authenticate(res, credentials, artifacts)).to.equal(true);
                done();
            ***REMOVED***);

            it('errors on invalid WWW-Authenticate header format', function (done) ***REMOVED***

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'www-authenticate': 'Hawk ts="1362346425875", tsm="PhwayS28vtnn3qbv0mqRBYSXebN/zggEtucfeZ620Zo=", x="Stale timestamp"'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                expect(Browser.client.authenticate(res, ***REMOVED******REMOVED***)).to.equal(false);
                done();
            ***REMOVED***);

            it('errors on invalid WWW-Authenticate header format', function (done) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '123456',
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    algorithm: 'sha256',
                    user: 'steve'
                ***REMOVED***;

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'www-authenticate': 'Hawk ts="1362346425875", tsm="hwayS28vtnn3qbv0mqRBYSXebN/zggEtucfeZ620Zo=", error="Stale timestamp"'
                    ***REMOVED***,
                    getResponseHeader: function (header) ***REMOVED***

                        return res.headers[header.toLowerCase()];
                    ***REMOVED***
                ***REMOVED***;

                expect(Browser.client.authenticate(res, credentials)).to.equal(false);
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('message()', function () ***REMOVED***

            it('generates an authorization then successfully parse it', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                    var auth = Browser.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                    expect(auth).to.exist();

                    Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                        expect(err).to.not.exist();
                        expect(credentials2.user).to.equal('steve');
                        done();
                    ***REMOVED***);
                ***REMOVED***);
            ***REMOVED***);

            it('generates an authorization using custom nonce/timestamp', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials, nonce: 'abc123', timestamp: 1398536270957 ***REMOVED***);
                    expect(auth).to.exist();
                    expect(auth.nonce).to.equal('abc123');
                    expect(auth.ts).to.equal(1398536270957);
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on missing host', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message(null, 8080, 'some message', ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on invalid host', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message(5, 8080, 'some message', ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on missing port', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message('example.com', 0, 'some message', ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on invalid port', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message('example.com', 'a', 'some message', ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on missing message', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message('example.com', 8080, undefined, ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on null message', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message('example.com', 8080, null, ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on invalid message', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var auth = Browser.client.message('example.com', 8080, 5, ***REMOVED*** credentials: credentials ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on missing credentials', function (done) ***REMOVED***

                var auth = Browser.client.message('example.com', 8080, 'some message', ***REMOVED******REMOVED***);
                expect(auth).to.not.exist();
                done();
            ***REMOVED***);

            it('errors on missing options', function (done) ***REMOVED***

                var auth = Browser.client.message('example.com', 8080, 'some message');
                expect(auth).to.not.exist();
                done();
            ***REMOVED***);

            it('errors on invalid credentials (id)', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var creds = Hoek.clone(credentials);
                    delete creds.id;
                    var auth = Browser.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: creds ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on invalid credentials (key)', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var creds = Hoek.clone(credentials);
                    delete creds.key;
                    var auth = Browser.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: creds ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('errors on invalid algorithm', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var creds = Hoek.clone(credentials);
                    creds.algorithm = 'blah';
                    var auth = Browser.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: creds ***REMOVED***);
                    expect(auth).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        describe('authenticateTimestamp()', function (done) ***REMOVED***

            it('validates a timestamp', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var tsm = Hawk.crypto.timestampMessage(credentials);
                    expect(Browser.client.authenticateTimestamp(tsm, credentials)).to.equal(true);
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('validates a timestamp without updating local time', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var offset = Browser.utils.getNtpOffset();
                    var tsm = Hawk.crypto.timestampMessage(credentials, 10000);
                    expect(Browser.client.authenticateTimestamp(tsm, credentials, false)).to.equal(true);
                    expect(offset).to.equal(Browser.utils.getNtpOffset());
                    done();
                ***REMOVED***);
            ***REMOVED***);

            it('detects a bad timestamp', function (done) ***REMOVED***

                credentialsFunc('123456', function (err, credentials) ***REMOVED***

                    var tsm = Hawk.crypto.timestampMessage(credentials);
                    tsm.ts = 4;
                    expect(Browser.client.authenticateTimestamp(tsm, credentials)).to.equal(false);
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('internals', function () ***REMOVED***

        describe('LocalStorage', function () ***REMOVED***

            it('goes through the full lifecycle', function (done) ***REMOVED***

                var storage = new Browser.internals.LocalStorage();
                expect(storage.length).to.equal(0);
                expect(storage.getItem('a')).to.equal(null);
                storage.setItem('a', 5);
                expect(storage.length).to.equal(1);
                expect(storage.key()).to.equal('a');
                expect(storage.key(0)).to.equal('a');
                expect(storage.getItem('a')).to.equal('5');
                storage.setItem('b', 'test');
                expect(storage.key()).to.equal('a');
                expect(storage.key(0)).to.equal('a');
                expect(storage.key(1)).to.equal('b');
                expect(storage.length).to.equal(2);
                expect(storage.getItem('b')).to.equal('test');
                storage.removeItem('a');
                expect(storage.length).to.equal(1);
                expect(storage.getItem('a')).to.equal(null);
                expect(storage.getItem('b')).to.equal('test');
                storage.clear();
                expect(storage.length).to.equal(0);
                expect(storage.getItem('a')).to.equal(null);
                expect(storage.getItem('b')).to.equal(null);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('utils', function () ***REMOVED***

        describe('setStorage()', function () ***REMOVED***

            it('sets storage for the first time', function (done) ***REMOVED***

                Browser.utils.storage = new Browser.internals.LocalStorage();        // Reset state

                expect(Browser.utils.storage.getItem('hawk_ntp_offset')).to.not.exist();
                Browser.utils.storage.setItem('test', '1');
                Browser.utils.setStorage(new Browser.internals.LocalStorage());
                expect(Browser.utils.storage.getItem('test')).to.not.exist();
                Browser.utils.storage.setItem('test', '2');
                expect(Browser.utils.storage.getItem('test')).to.equal('2');
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('setNtpOffset()', function (done) ***REMOVED***

            it('catches localStorage errors', ***REMOVED*** parallel: false ***REMOVED***, function (done) ***REMOVED***

                var orig = Browser.utils.storage.setItem;
                var consoleOrig = console.error;
                var count = 0;
                console.error = function () ***REMOVED***

                    if (count++ === 2) ***REMOVED***

                        console.error = consoleOrig;
                    ***REMOVED***
                ***REMOVED***;

                Browser.utils.storage.setItem = function () ***REMOVED***

                    Browser.utils.storage.setItem = orig;
                    throw new Error();
                ***REMOVED***;

                expect(function () ***REMOVED***

                    Browser.utils.setNtpOffset(100);
                ***REMOVED***).not.to.throw();

                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('parseAuthorizationHeader()', function (done) ***REMOVED***

            it('returns null on missing header', function (done) ***REMOVED***

                expect(Browser.utils.parseAuthorizationHeader()).to.equal(null);
                done();
            ***REMOVED***);

            it('returns null on bad header syntax (structure)', function (done) ***REMOVED***

                expect(Browser.utils.parseAuthorizationHeader('Hawk')).to.equal(null);
                done();
            ***REMOVED***);

            it('returns null on bad header syntax (parts)', function (done) ***REMOVED***

                expect(Browser.utils.parseAuthorizationHeader(' ')).to.equal(null);
                done();
            ***REMOVED***);

            it('returns null on bad scheme name', function (done) ***REMOVED***

                expect(Browser.utils.parseAuthorizationHeader('Basic asdasd')).to.equal(null);
                done();
            ***REMOVED***);

            it('returns null on bad attribute value', function (done) ***REMOVED***

                expect(Browser.utils.parseAuthorizationHeader('Hawk test="\t"', ['test'])).to.equal(null);
                done();
            ***REMOVED***);

            it('returns null on duplicated attribute', function (done) ***REMOVED***

                expect(Browser.utils.parseAuthorizationHeader('Hawk test="a", test="b"', ['test'])).to.equal(null);
                done();
            ***REMOVED***);
        ***REMOVED***);

        describe('parseUri()', function () ***REMOVED***

            it('returns empty object on invalid', function (done) ***REMOVED***

                var uri = Browser.utils.parseUri('ftp');
                expect(uri).to.deep.equal(***REMOVED*** host: '', port: '', resource: '' ***REMOVED***);
                done();
            ***REMOVED***);

            it('returns empty port when unknown scheme', function (done) ***REMOVED***

                var uri = Browser.utils.parseUri('ftp://example.com');
                expect(uri.port).to.equal('');
                done();
            ***REMOVED***);

            it('returns default port when missing', function (done) ***REMOVED***

                var uri = Browser.utils.parseUri('http://example.com');
                expect(uri.port).to.equal('80');
                done();
            ***REMOVED***);

            it('handles unusual characters correctly', function (done) ***REMOVED***

                var parts = ***REMOVED***
                    protocol: 'http+vnd.my-extension',
                    user: 'user!$&\'()*+,;=%40my-domain.com',
                    password: 'pass!$&\'()*+,;=%40:word',
                    hostname: 'foo-bar.com',
                    port: '99',
                    pathname: '/path/%40/!$&\'()*+,;=:@/',
                    query: 'query%40/!$&\'()*+,;=:@/?',
                    fragment: 'fragm%40/!$&\'()*+,;=:@/?'
                ***REMOVED***;

                parts.userInfo = parts.user + ':' + parts.password;
                parts.authority = parts.userInfo + '@' + parts.hostname + ':' + parts.port;
                parts.relative = parts.pathname + '?' + parts.query;
                parts.resource = parts.relative + '#' + parts.fragment;
                parts.source = parts.protocol + '://' + parts.authority + parts.resource;

                var uri = Browser.utils.parseUri(parts.source);
                expect(uri.host).to.equal('foo-bar.com');
                expect(uri.port).to.equal('99');
                expect(uri.resource).to.equal(parts.pathname + '?' + parts.query);
                done();
            ***REMOVED***);
        ***REMOVED***);

        var str = 'https://www.google.ca/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=url';
        var base64str = 'aHR0cHM6Ly93d3cuZ29vZ2xlLmNhL3dlYmhwP3NvdXJjZWlkPWNocm9tZS1pbnN0YW50Jmlvbj0xJmVzcHY9MiZpZT1VVEYtOCNxPXVybA';

        describe('base64urlEncode()', function () ***REMOVED***

            it('should base64 URL-safe decode a string', function (done) ***REMOVED***

                expect(Browser.utils.base64urlEncode(str)).to.equal(base64str);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
