// hawk/test/server.js
// Load modules

var Url = require('url');
var Code = require('code');
var Hawk = require('../lib');
var Hoek = require('hoek');
var Lab = require('lab');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.experiment;
var it = lab.test;
var expect = Code.expect;


describe('Server', function () ***REMOVED***

    var credentialsFunc = function (id, callback) ***REMOVED***

        var credentials = ***REMOVED***
            id: id,
            key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
            algorithm: (id === '1' ? 'sha1' : 'sha256'),
            user: 'steve'
        ***REMOVED***;

        return callback(null, credentials);
    ***REMOVED***;

    describe('authenticate()', function () ***REMOVED***

        it('parses a valid authentication header (sha1)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="1", ts="1353788437", nonce="k3j4h2", mac="zy79QQ5/EYFmQqutVnYb73gAc/U=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('parses a valid authentication header (sha256)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/1?b=1&a=2',
                host: 'example.com',
                port: 8000,
                authorization: 'Hawk id="dh37fgj492je", ts="1353832234", nonce="j4h3g2", mac="m8r1rHbXN6NgO+KIIhjO7sFRyd78RNGVUwehe8Cp2dU=", ext="some-app-data"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353832234000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('parses a valid authentication header (host override)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                headers: ***REMOVED***
                    host: 'example1.com:8080',
                    authorization: 'Hawk id="1", ts="1353788437", nonce="k3j4h2", mac="zy79QQ5/EYFmQqutVnYb73gAc/U=", ext="hello"'
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** host: 'example.com', localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('parses a valid authentication header (host port override)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                headers: ***REMOVED***
                    host: 'example1.com:80',
                    authorization: 'Hawk id="1", ts="1353788437", nonce="k3j4h2", mac="zy79QQ5/EYFmQqutVnYb73gAc/U=", ext="hello"'
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** host: 'example.com', port: 8080, localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('parses a valid authentication header (POST with payload)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'POST',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123456", ts="1357926341", nonce="1AwuJD", hash="qAiXIVv+yjDATneWxZP2YCTa9aHRgQdnH9b3Wc+o3dg=", ext="some-app-data", mac="UeYcj5UoTVaAWXNvJfLVia7kU3VabxCqrccXP8sUGC4="'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1357926341000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on missing hash', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/1?b=1&a=2',
                host: 'example.com',
                port: 8000,
                authorization: 'Hawk id="dh37fgj492je", ts="1353832234", nonce="j4h3g2", mac="m8r1rHbXN6NgO+KIIhjO7sFRyd78RNGVUwehe8Cp2dU=", ext="some-app-data"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** payload: 'body', localtimeOffsetMsec: 1353832234000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Missing required payload hash');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on a stale timestamp', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123456", ts="1362337299", nonce="UzmxSs", ext="some-app-data", mac="wnNUxchvvryMH2RxckTdZ/gY3ijzvccx4keVvELC61w="'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Stale timestamp');
                var header = err.output.headers['WWW-Authenticate'];
                var ts = header.match(/^Hawk ts\=\"(\d+)\"\, tsm\=\"([^\"]+)\"\, error=\"Stale timestamp\"$/);
                var now = Hawk.utils.now();
                expect(parseInt(ts[1], 10) * 1000).to.be.within(now - 1000, now + 1000);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'www-authenticate': header
                    ***REMOVED***
                ***REMOVED***;

                expect(Hawk.client.authenticate(res, credentials, artifacts)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on a replay', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="bXx7a7p1h9QYQNZ8x7QhvDQym8ACgab4m3lVSFn4DBw=", ext="hello"'
            ***REMOVED***;

            var memoryCache = ***REMOVED******REMOVED***;
            var options = ***REMOVED***
                localtimeOffsetMsec: 1353788437000 - Hawk.utils.now(),
                nonceFunc: function (key, nonce, ts, callback) ***REMOVED***

                    if (memoryCache[key + nonce]) ***REMOVED***
                        return callback(new Error());
                    ***REMOVED***

                    memoryCache[key + nonce] = true;
                    return callback();
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, options, function (err, credentials1, artifacts1) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials1.user).to.equal('steve');

                Hawk.server.authenticate(req, credentialsFunc, options, function (err, credentials2, artifacts2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.output.payload.message).to.equal('Invalid nonce');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('does not error on nonce collision if keys differ', function (done) ***REMOVED***

            var reqSteve = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="bXx7a7p1h9QYQNZ8x7QhvDQym8ACgab4m3lVSFn4DBw=", ext="hello"'
            ***REMOVED***;

            var reqBob = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="456", ts="1353788437", nonce="k3j4h2", mac="LXfmTnRzrLd9TD7yfH+4se46Bx6AHyhpM94hLCiNia4=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                var credentials = ***REMOVED***
                    '123': ***REMOVED***
                        id: id,
                        key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                        algorithm: (id === '1' ? 'sha1' : 'sha256'),
                        user: 'steve'
                    ***REMOVED***,
                    '456': ***REMOVED***
                        id: id,
                        key: 'xrunpaw3489ruxnpa98w4rxnwerxhqb98rpaxn39848',
                        algorithm: (id === '1' ? 'sha1' : 'sha256'),
                        user: 'bob'
                    ***REMOVED***
                ***REMOVED***;

                return callback(null, credentials[id]);
            ***REMOVED***;

            var memoryCache = ***REMOVED******REMOVED***;
            var options = ***REMOVED***
                localtimeOffsetMsec: 1353788437000 - Hawk.utils.now(),
                nonceFunc: function (key, nonce, ts, callback) ***REMOVED***

                    if (memoryCache[key + nonce]) ***REMOVED***
                        return callback(new Error());
                    ***REMOVED***

                    memoryCache[key + nonce] = true;
                    return callback();
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(reqSteve, credentialsFuncion, options, function (err, credentials1, artifacts1) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials1.user).to.equal('steve');

                Hawk.server.authenticate(reqBob, credentialsFuncion, options, function (err, credentials2, artifacts2) ***REMOVED***

                    expect(err).to.not.exist();
                    expect(credentials2.user).to.equal('bob');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an invalid authentication header: wrong scheme', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Basic asdasdasdasd'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an invalid authentication header: no scheme', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: '!@#'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Invalid header syntax');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an missing authorization header', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.isMissing).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an missing host header', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                headers: ***REMOVED***
                    authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Invalid Host header');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an missing authorization attribute (id)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Missing attributes');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an missing authorization attribute (ts)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Missing attributes');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an missing authorization attribute (nonce)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Missing attributes');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an missing authorization attribute (mac)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Missing attributes');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an unknown authorization attribute', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", x="3", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Unknown attribute: x');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an bad authorization header format', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123\\", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Bad header format');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an bad authorization attribute value', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="\t", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Bad attribute value: id');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an empty authorization attribute value', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Bad attribute value: id');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on duplicated authorization attribute key', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", id="456", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Duplicate attribute: id');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an invalid authorization header format', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk'
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Invalid header syntax');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an bad host header (missing host)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                headers: ***REMOVED***
                    host: ':8080',
                    authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Invalid Host header');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on an bad host header (pad port)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                headers: ***REMOVED***
                    host: 'example.com:something',
                    authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
                ***REMOVED***
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Invalid Host header');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on credentialsFunc error', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                return callback(new Error('Unknown user'));
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Unknown user');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on credentialsFunc error (with credentials)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                return callback(new Error('Unknown user'), ***REMOVED*** some: 'value' ***REMOVED***);
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Unknown user');
                expect(credentials.some).to.equal('value');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on missing credentials', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                return callback(null, null);
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Unknown credentials');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on invalid credentials (id)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                var credentials = ***REMOVED***
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    user: 'steve'
                ***REMOVED***;

                return callback(null, credentials);
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Invalid credentials');
                expect(err.output.payload.message).to.equal('An internal server error occurred');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on invalid credentials (key)', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                var credentials = ***REMOVED***
                    id: '23434d3q4d5345d',
                    user: 'steve'
                ***REMOVED***;

                return callback(null, credentials);
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Invalid credentials');
                expect(err.output.payload.message).to.equal('An internal server error occurred');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on unknown credentials algorithm', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcUyW6EEgUH4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                var credentials = ***REMOVED***
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    algorithm: 'hmac-sha-0',
                    user: 'steve'
                ***REMOVED***;

                return callback(null, credentials);
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.message).to.equal('Unknown algorithm');
                expect(err.output.payload.message).to.equal('An internal server error occurred');
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('errors on unknown bad mac', function (done) ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="123", ts="1353788437", nonce="k3j4h2", mac="/qwS4UjfVWMcU4jlr7T/wuKe3dKijvTvSos=", ext="hello"'
            ***REMOVED***;

            var credentialsFuncion = function (id, callback) ***REMOVED***

                var credentials = ***REMOVED***
                    key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                    algorithm: 'sha256',
                    user: 'steve'
                ***REMOVED***;

                return callback(null, credentials);
            ***REMOVED***;

            Hawk.server.authenticate(req, credentialsFuncion, ***REMOVED*** localtimeOffsetMsec: 1353788437000 - Hawk.utils.now() ***REMOVED***, function (err, credentials, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Bad mac');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('header()', function () ***REMOVED***

        it('generates header', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                ext: 'some-app-data',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('Hawk mac=\"n14wVJK4cOxAytPUMc5bPezQzuJGl5n7MYXhFQgEKsE=\", hash=\"f9cDF/TDm7TkYRLnGwRMfeDzT6LixQVLvrIKhh0vgmM=\", ext=\"response-specific\"');
            done();
        ***REMOVED***);

        it('generates header (empty payload)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                ext: 'some-app-data',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, artifacts, ***REMOVED*** payload: '', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('Hawk mac=\"i8/kUBDx0QF+PpCtW860kkV/fa9dbwEoe/FpGUXowf0=\", hash=\"q/t+NNAkQZNlq/aAD6PlexImwQTxwgT2MahfTa9XRLA=\", ext=\"response-specific\"');
            done();
        ***REMOVED***);

        it('generates header (pre calculated hash)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                ext: 'some-app-data',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var options = ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***;
            options.hash = Hawk.crypto.calculatePayloadHash(options.payload, credentials.algorithm, options.contentType);
            var header = Hawk.server.header(credentials, artifacts, options);
            expect(header).to.equal('Hawk mac=\"n14wVJK4cOxAytPUMc5bPezQzuJGl5n7MYXhFQgEKsE=\", hash=\"f9cDF/TDm7TkYRLnGwRMfeDzT6LixQVLvrIKhh0vgmM=\", ext=\"response-specific\"');
            done();
        ***REMOVED***);

        it('generates header (null ext)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: null ***REMOVED***);
            expect(header).to.equal('Hawk mac=\"6PrybJTJs20jsgBw5eilXpcytD8kUbaIKNYXL+6g0ns=\", hash=\"f9cDF/TDm7TkYRLnGwRMfeDzT6LixQVLvrIKhh0vgmM=\"');
            done();
        ***REMOVED***);

        it('errors on missing artifacts', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, null, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid artifacts', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, 5, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing credentials', function (done) ***REMOVED***

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                ext: 'some-app-data',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var header = Hawk.server.header(null, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid credentials (key)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                algorithm: 'sha256',
                user: 'steve'
            ***REMOVED***;

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                ext: 'some-app-data',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid algorithm', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
                algorithm: 'x',
                user: 'steve'
            ***REMOVED***;

            var artifacts = ***REMOVED***
                method: 'POST',
                host: 'example.com',
                port: '8080',
                resource: '/resource/4?filter=a',
                ts: '1398546787',
                nonce: 'xUwusx',
                hash: 'nJjkVtBE5Y/Bk38Aiokwn0jiJxt/0S2WRSUwWLCf5xk=',
                ext: 'some-app-data',
                mac: 'dvIvMThwi28J61Jc3P0ryAhuKpanU63GXdx6hkmQkJA=',
                id: '123456'
            ***REMOVED***;

            var header = Hawk.server.header(credentials, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
            expect(header).to.equal('');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('authenticateBewit()', function () ***REMOVED***

        it('errors on uri too long', function (done) ***REMOVED***

            var long = '/';
            for (var i = 0; i < 5000; ++i) ***REMOVED***
                long += 'x';
            ***REMOVED***

            var req = ***REMOVED***
                method: 'GET',
                url: long,
                host: 'example.com',
                port: 8080,
                authorization: 'Hawk id="1", ts="1353788437", nonce="k3j4h2", mac="zy79QQ5/EYFmQqutVnYb73gAc/U=", ext="hello"'
            ***REMOVED***;

            Hawk.server.authenticateBewit(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, bewit) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.statusCode).to.equal(400);
                expect(err.message).to.equal('Resource path exceeds max length');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('authenticateMessage()', function () ***REMOVED***

        it('errors on invalid authorization (ts)', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                delete auth.ts;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid authorization');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('errors on invalid authorization (nonce)', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                delete auth.nonce;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid authorization');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('errors on invalid authorization (hash)', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                delete auth.hash;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid authorization');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('errors with credentials', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, function (id, callback) ***REMOVED***

                    callback(new Error('something'), ***REMOVED*** some: 'value' ***REMOVED***);
                ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('something');
                    expect(credentials2.some).to.equal('value');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('errors on nonce collision', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED***
                    nonceFunc: function (key, nonce, ts, nonceCallback) ***REMOVED***

                        nonceCallback(true);
                    ***REMOVED***
                ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid nonce');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should generate an authorization then successfully parse it', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.not.exist();
                    expect(credentials2.user).to.equal('steve');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on mismatching host', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example1.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Bad mac');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on stale timestamp', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 100000 ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Stale timestamp');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('overrides timestampSkewSec', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1, localtimeOffsetMsec: 100000 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED*** timestampSkewSec: 500 ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on invalid authorization', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();
                delete auth.id;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid authorization');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on bad hash', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message1', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Bad message hash');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on nonce error', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED***
                    nonceFunc: function (key, nonce, ts, callback) ***REMOVED***

                        callback(new Error('kaboom'));
                    ***REMOVED***
                ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid nonce');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on credentials error', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback(new Error('kablooey'));
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('kablooey');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on missing credentials', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback();
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Unknown credentials');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on invalid credentials', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback(null, ***REMOVED******REMOVED***);
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid credentials');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on invalid credentials algorithm', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback(null, ***REMOVED*** key: '123', algorithm: '456' ***REMOVED***);
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Unknown algorithm');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail on missing host', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials) ***REMOVED***

                var auth = Hawk.client.message(null, 8080, 'some message', ***REMOVED*** credentials: credentials ***REMOVED***);
                expect(auth).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('should fail on missing credentials', function (done) ***REMOVED***

            var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED******REMOVED***);
            expect(auth).to.not.exist();
            done();
        ***REMOVED***);

        it('should fail on invalid algorithm', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials) ***REMOVED***

                var creds = Hoek.clone(credentials);
                creds.algorithm = 'blah';
                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: creds ***REMOVED***);
                expect(auth).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    describe('authenticatePayloadHash()', function () ***REMOVED***

        it('checks payload hash', function (done) ***REMOVED***

            expect(Hawk.server.authenticatePayloadHash('abcdefg', ***REMOVED*** hash: 'abcdefg' ***REMOVED***)).to.equal(true);
            expect(Hawk.server.authenticatePayloadHash('1234567', ***REMOVED*** hash: 'abcdefg' ***REMOVED***)).to.equal(false);
            done();
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
