// @firebase/firestore/dist/esm/src/core/timestamp.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** assert ***REMOVED*** from '../util/assert';
import ***REMOVED*** primitiveComparator ***REMOVED*** from '../util/misc';
// A RegExp matching ISO 8601 UTC timestamps with optional fraction.
var isoRegExp = new RegExp(/^\d***REMOVED***4***REMOVED***-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.(\d+))?Z$/);
var Timestamp = /** @class */ (function () ***REMOVED***
    function Timestamp(seconds, nanos) ***REMOVED***
        this.seconds = seconds;
        this.nanos = nanos;
        assert(nanos >= 0, 'timestamp nanoseconds out of range: ' + nanos);
        assert(nanos < 1e9, 'timestamp nanoseconds out of range' + nanos);
        // Midnight at the beginning of 1/1/1 is the earliest Firestore supports.
        assert(seconds >= -62135596800, 'timestamp seconds out of range: ' + seconds);
        // This will break in the year 10,000.
        assert(seconds < 253402300800, 'timestamp seconds out of range' + seconds);
    ***REMOVED***
    Timestamp.now = function () ***REMOVED***
        return Timestamp.fromEpochMilliseconds(Date.now());
    ***REMOVED***;
    Timestamp.fromDate = function (date) ***REMOVED***
        return Timestamp.fromEpochMilliseconds(date.getTime());
    ***REMOVED***;
    Timestamp.fromEpochMilliseconds = function (milliseconds) ***REMOVED***
        var seconds = Math.floor(milliseconds / 1000);
        var nanos = (milliseconds - seconds * 1000) * 1e6;
        return new Timestamp(seconds, nanos);
    ***REMOVED***;
    Timestamp.fromISOString = function (utc) ***REMOVED***
        // The date string can have higher precision (nanos) than the Date class
        // (millis), so we do some custom parsing here.
        // Parse the nanos right out of the string.
        var nanos = 0;
        var fraction = isoRegExp.exec(utc);
        assert(!!fraction, 'invalid timestamp: ' + utc);
        if (fraction[1]) ***REMOVED***
            // Pad the fraction out to 9 digits (nanos).
            var nanoStr = fraction[1];
            nanoStr = (nanoStr + '000000000').substr(0, 9);
            nanos = Number(nanoStr);
        ***REMOVED***
        // Parse the date to get the seconds.
        var date = new Date(utc);
        var seconds = Math.floor(date.getTime() / 1000);
        return new Timestamp(seconds, nanos);
    ***REMOVED***;
    Timestamp.prototype.toDate = function () ***REMOVED***
        return new Date(this.toEpochMilliseconds());
    ***REMOVED***;
    Timestamp.prototype.toEpochMilliseconds = function () ***REMOVED***
        return this.seconds * 1000 + this.nanos / 1e6;
    ***REMOVED***;
    Timestamp.prototype.compareTo = function (other) ***REMOVED***
        if (this.seconds === other.seconds) ***REMOVED***
            return primitiveComparator(this.nanos, other.nanos);
        ***REMOVED***
        return primitiveComparator(this.seconds, other.seconds);
    ***REMOVED***;
    Timestamp.prototype.isEqual = function (other) ***REMOVED***
        return other.seconds === this.seconds && other.nanos === this.nanos;
    ***REMOVED***;
    Timestamp.prototype.toString = function () ***REMOVED***
        return 'Timestamp(seconds=' + this.seconds + ', nanos=' + this.nanos + ')';
    ***REMOVED***;
    return Timestamp;
***REMOVED***());
export ***REMOVED*** Timestamp ***REMOVED***;
