// @google-cloud/common-grpc/src/operation.js
/*!
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*!
 * @module commonGrpc/operation
 */

'use strict';

var common = require('@google-cloud/common');
var modelo = require('modelo');

/**
 * @type ***REMOVED***module:commonGrpc/serviceObject***REMOVED***
 * @private
 */
var ServiceObject = require('./service-object.js');

/**
 * @type ***REMOVED***module:commonGrpc/service***REMOVED***
 * @private
 */
var Service = require('./service.js');

// jscs:disable maximumLineLength
/**
 * An Operation object allows you to interact with APIs that take longer to
 * process things.
 *
 * @constructor
 * @alias module:common/grpcOperation
 *
 * @param ***REMOVED***module:commonGrpc/service|module:commonGrpc/serviceObject***REMOVED*** parent - The
 *     parent object. This should be configured to use the longrunning.operation
 *     service.
 * @param ***REMOVED***string***REMOVED*** name - The operation name.
 */
// jscs:enable maximumLineLength
function Operation(parent, name) ***REMOVED***
  var methods = ***REMOVED***
    /**
     * Deletes an operation.
     */
    delete: ***REMOVED***
      protoOpts: ***REMOVED***
        service: 'Operations',
        method: 'deleteOperation',
      ***REMOVED***,
      reqOpts: ***REMOVED***
        name: name,
      ***REMOVED***,
    ***REMOVED***,

    /**
     * Checks to see if an operation exists.
     */
    exists: true,

    /**
     * Retrieves the operation.
     */
    get: true,

    /**
     * Retrieves metadata for the operation.
     */
    getMetadata: ***REMOVED***
      protoOpts: ***REMOVED***
        service: 'Operations',
        method: 'getOperation',
      ***REMOVED***,
      reqOpts: ***REMOVED***
        name: name,
      ***REMOVED***,
    ***REMOVED***,
  ***REMOVED***;

  var config = ***REMOVED***
    parent: parent,
    id: name,
    methods: methods,
  ***REMOVED***;

  common.Operation.call(this, config);
  ServiceObject.call(this, config);
***REMOVED***

modelo.inherits(Operation, ServiceObject, common.Operation);

/**
 * Cancel the operation.
 *
 * @param ***REMOVED***function=***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this
 *     request.
 * @param ***REMOVED***object***REMOVED*** callback.apiResponse - The full API response.
 */
Operation.prototype.cancel = function(callback) ***REMOVED***
  var protoOpts = ***REMOVED***
    service: 'Operations',
    method: 'cancelOperation',
  ***REMOVED***;

  var reqOpts = ***REMOVED***
    name: this.id,
  ***REMOVED***;

  this.request(protoOpts, reqOpts, callback || common.util.noop);
***REMOVED***;

/**
 * Poll for a status update. Execute the callback:
 *
 *   - callback(err): Operation failed
 *   - callback(): Operation incomplete
 *   - callback(null, metadata): Operation complete
 *
 * @private
 *
 * @param ***REMOVED***function***REMOVED*** callback
 */
Operation.prototype.poll_ = function(callback) ***REMOVED***
  this.getMetadata(function(err, resp) ***REMOVED***
    if (err || resp.error) ***REMOVED***
      callback(err || Service.decorateGrpcStatus_(resp.error));
      return;
    ***REMOVED***

    if (!resp.done) ***REMOVED***
      callback();
      return;
    ***REMOVED***

    callback(null, resp);
  ***REMOVED***);
***REMOVED***;

module.exports = Operation;
