// @google-cloud/common-grpc/src/service-object.js
/*!
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*!
 * @module commonGrpc/serviceObject
 */

'use strict';

var extend = require('extend');
var nodeutil = require('util');
var ServiceObject = require('@google-cloud/common').ServiceObject;
var util = require('@google-cloud/common').util;

/**
 * GrpcServiceObject is a base class, meant to be inherited from by a service
 * object that uses the gRPC protobuf API.
 *
 * @constructor
 * @alias module:common/grpc-service-object
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** config - Configuration object.
 */
function GrpcServiceObject(config) ***REMOVED***
  ServiceObject.call(this, config);
***REMOVED***

nodeutil.inherits(GrpcServiceObject, ServiceObject);

/**
 * Delete the object.
 *
 * @param ***REMOVED***function=***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 */
GrpcServiceObject.prototype.delete = function(callback) ***REMOVED***
  var protoOpts = this.methods.delete.protoOpts;
  var reqOpts = this.methods.delete.reqOpts;

  this.request(protoOpts, reqOpts, callback || util.noop);
***REMOVED***;

/**
 * Get the metadata of this object.
 *
 * @param ***REMOVED***function***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 * @param ***REMOVED***object***REMOVED*** callback.metadata - The metadata for this object.
 */
GrpcServiceObject.prototype.getMetadata = function(callback) ***REMOVED***
  var self = this;

  var protoOpts = this.methods.getMetadata.protoOpts;
  var reqOpts = this.methods.getMetadata.reqOpts;

  this.request(protoOpts, reqOpts, function(err, resp) ***REMOVED***
    if (err) ***REMOVED***
      callback(err, null, resp);
      return;
    ***REMOVED***

    self.metadata = resp;

    callback(null, self.metadata, resp);
  ***REMOVED***);
***REMOVED***;

/**
 * Set the metadata for this object.
 *
 * @param ***REMOVED***object***REMOVED*** metadata - The metadata to set on this object.
 * @param ***REMOVED***function=***REMOVED*** callback - The callback function.
 * @param ***REMOVED***?error***REMOVED*** callback.err - An error returned while making this request.
 */
GrpcServiceObject.prototype.setMetadata = function(metadata, callback) ***REMOVED***
  var protoOpts = this.methods.setMetadata.protoOpts;
  var reqOpts = extend(true, ***REMOVED******REMOVED***, this.methods.setMetadata.reqOpts, metadata);

  this.request(protoOpts, reqOpts, callback || util.noop);
***REMOVED***;

/**
 * Patch a request to the GrpcService object.
 *
 * @private
 */
GrpcServiceObject.prototype.request = function() ***REMOVED***
  return this.parent.request.apply(this.parent, arguments);
***REMOVED***;

/**
 * Patch a streaming request to the GrpcService object.
 *
 * @private
 */
GrpcServiceObject.prototype.requestStream = function() ***REMOVED***
  return this.parent.requestStream.apply(this.parent, arguments);
***REMOVED***;

/**
 * Patch a writable streaming request to the GrpcService object.
 *
 * @private
 */
GrpcServiceObject.prototype.requestWritableStream = function() ***REMOVED***
  return this.parent.requestWritableStream.apply(this.parent, arguments);
***REMOVED***;

/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
util.promisifyAll(GrpcServiceObject);

module.exports = GrpcServiceObject;
