// protobufjs/src/field.js
"use strict";
module.exports = Field;

// extends ReflectionObject
var ReflectionObject = require("./object");
((Field.prototype = Object.create(ReflectionObject.prototype)).constructor = Field).className = "Field";

var Enum  = require("./enum"),
    types = require("./types"),
    util  = require("./util");

var Type; // cyclic

var ruleRe = /^required|optional|repeated$/;

/**
 * Constructs a new message field instance. Note that ***REMOVED***@link MapField|map fields***REMOVED*** have their own class.
 * @name Field
 * @classdesc Reflected message field.
 * @extends FieldBase
 * @constructor
 * @param ***REMOVED***string***REMOVED*** name Unique name within its namespace
 * @param ***REMOVED***number***REMOVED*** id Unique id within its namespace
 * @param ***REMOVED***string***REMOVED*** type Value type
 * @param ***REMOVED***string|Object.<string,*>***REMOVED*** [rule="optional"] Field rule
 * @param ***REMOVED***string|Object.<string,*>***REMOVED*** [extend] Extended type if different from parent
 * @param ***REMOVED***Object.<string,*>***REMOVED*** [options] Declared options
 */

/**
 * Constructs a field from a field descriptor.
 * @param ***REMOVED***string***REMOVED*** name Field name
 * @param ***REMOVED***IField***REMOVED*** json Field descriptor
 * @returns ***REMOVED***Field***REMOVED*** Created field
 * @throws ***REMOVED***TypeError***REMOVED*** If arguments are invalid
 */
Field.fromJSON = function fromJSON(name, json) ***REMOVED***
    return new Field(name, json.id, json.type, json.rule, json.extend, json.options);
***REMOVED***;

/**
 * Not an actual constructor. Use ***REMOVED***@link Field***REMOVED*** instead.
 * @classdesc Base class of all reflected message fields. This is not an actual class but here for the sake of having consistent type definitions.
 * @exports FieldBase
 * @extends ReflectionObject
 * @constructor
 * @param ***REMOVED***string***REMOVED*** name Unique name within its namespace
 * @param ***REMOVED***number***REMOVED*** id Unique id within its namespace
 * @param ***REMOVED***string***REMOVED*** type Value type
 * @param ***REMOVED***string|Object.<string,*>***REMOVED*** [rule="optional"] Field rule
 * @param ***REMOVED***string|Object.<string,*>***REMOVED*** [extend] Extended type if different from parent
 * @param ***REMOVED***Object.<string,*>***REMOVED*** [options] Declared options
 */
function Field(name, id, type, rule, extend, options) ***REMOVED***

    if (util.isObject(rule)) ***REMOVED***
        options = rule;
        rule = extend = undefined;
    ***REMOVED*** else if (util.isObject(extend)) ***REMOVED***
        options = extend;
        extend = undefined;
    ***REMOVED***

    ReflectionObject.call(this, name, options);

    if (!util.isInteger(id) || id < 0)
        throw TypeError("id must be a non-negative integer");

    if (!util.isString(type))
        throw TypeError("type must be a string");

    if (rule !== undefined && !ruleRe.test(rule = rule.toString().toLowerCase()))
        throw TypeError("rule must be a string rule");

    if (extend !== undefined && !util.isString(extend))
        throw TypeError("extend must be a string");

    /**
     * Field rule, if any.
     * @type ***REMOVED***string|undefined***REMOVED***
     */
    this.rule = rule && rule !== "optional" ? rule : undefined; // toJSON

    /**
     * Field type.
     * @type ***REMOVED***string***REMOVED***
     */
    this.type = type; // toJSON

    /**
     * Unique field id.
     * @type ***REMOVED***number***REMOVED***
     */
    this.id = id; // toJSON, marker

    /**
     * Extended type if different from parent.
     * @type ***REMOVED***string|undefined***REMOVED***
     */
    this.extend = extend || undefined; // toJSON

    /**
     * Whether this field is required.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.required = rule === "required";

    /**
     * Whether this field is optional.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.optional = !this.required;

    /**
     * Whether this field is repeated.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.repeated = rule === "repeated";

    /**
     * Whether this field is a map or not.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.map = false;

    /**
     * Message this field belongs to.
     * @type ***REMOVED***Type|null***REMOVED***
     */
    this.message = null;

    /**
     * OneOf this field belongs to, if any,
     * @type ***REMOVED***OneOf|null***REMOVED***
     */
    this.partOf = null;

    /**
     * The field type's default value.
     * @type ***REMOVED*******REMOVED***
     */
    this.typeDefault = null;

    /**
     * The field's default value on prototypes.
     * @type ***REMOVED*******REMOVED***
     */
    this.defaultValue = null;

    /**
     * Whether this field's value should be treated as a long.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.long = util.Long ? types.long[type] !== undefined : /* istanbul ignore next */ false;

    /**
     * Whether this field's value is a buffer.
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.bytes = type === "bytes";

    /**
     * Resolved type if not a basic type.
     * @type ***REMOVED***Type|Enum|null***REMOVED***
     */
    this.resolvedType = null;

    /**
     * Sister-field within the extended type if a declaring extension field.
     * @type ***REMOVED***Field|null***REMOVED***
     */
    this.extensionField = null;

    /**
     * Sister-field within the declaring namespace if an extended field.
     * @type ***REMOVED***Field|null***REMOVED***
     */
    this.declaringField = null;

    /**
     * Internally remembers whether this field is packed.
     * @type ***REMOVED***boolean|null***REMOVED***
     * @private
     */
    this._packed = null;
***REMOVED***

/**
 * Determines whether this field is packed. Only relevant when repeated and working with proto2.
 * @name Field#packed
 * @type ***REMOVED***boolean***REMOVED***
 * @readonly
 */
Object.defineProperty(Field.prototype, "packed", ***REMOVED***
    get: function() ***REMOVED***
        // defaults to packed=true if not explicity set to false
        if (this._packed === null)
            this._packed = this.getOption("packed") !== false;
        return this._packed;
    ***REMOVED***
***REMOVED***);

/**
 * @override
 */
Field.prototype.setOption = function setOption(name, value, ifNotSet) ***REMOVED***
    if (name === "packed") // clear cached before setting
        this._packed = null;
    return ReflectionObject.prototype.setOption.call(this, name, value, ifNotSet);
***REMOVED***;

/**
 * Field descriptor.
 * @interface IField
 * @property ***REMOVED***string***REMOVED*** [rule="optional"] Field rule
 * @property ***REMOVED***string***REMOVED*** type Field type
 * @property ***REMOVED***number***REMOVED*** id Field id
 * @property ***REMOVED***Object.<string,*>***REMOVED*** [options] Field options
 */

/**
 * Extension field descriptor.
 * @interface IExtensionField
 * @extends IField
 * @property ***REMOVED***string***REMOVED*** extend Extended type
 */

/**
 * Converts this field to a field descriptor.
 * @returns ***REMOVED***IField***REMOVED*** Field descriptor
 */
Field.prototype.toJSON = function toJSON() ***REMOVED***
    return util.toObject([
        "rule"    , this.rule !== "optional" && this.rule || undefined,
        "type"    , this.type,
        "id"      , this.id,
        "extend"  , this.extend,
        "options" , this.options
    ]);
***REMOVED***;

/**
 * Resolves this field's type references.
 * @returns ***REMOVED***Field***REMOVED*** `this`
 * @throws ***REMOVED***Error***REMOVED*** If any reference cannot be resolved
 */
Field.prototype.resolve = function resolve() ***REMOVED***

    if (this.resolved)
        return this;

    if ((this.typeDefault = types.defaults[this.type]) === undefined) ***REMOVED*** // if not a basic type, resolve it
        this.resolvedType = (this.declaringField ? this.declaringField.parent : this.parent).lookupTypeOrEnum(this.type);
        if (this.resolvedType instanceof Type)
            this.typeDefault = null;
        else // instanceof Enum
            this.typeDefault = this.resolvedType.values[Object.keys(this.resolvedType.values)[0]]; // first defined
    ***REMOVED***

    // use explicitly set default value if present
    if (this.options && this.options["default"] != null) ***REMOVED***
        this.typeDefault = this.options["default"];
        if (this.resolvedType instanceof Enum && typeof this.typeDefault === "string")
            this.typeDefault = this.resolvedType.values[this.typeDefault];
    ***REMOVED***

    // remove unnecessary options
    if (this.options) ***REMOVED***
        if (this.options.packed === true || this.options.packed !== undefined && this.resolvedType && !(this.resolvedType instanceof Enum))
            delete this.options.packed;
        if (!Object.keys(this.options).length)
            this.options = undefined;
    ***REMOVED***

    // convert to internal data type if necesssary
    if (this.long) ***REMOVED***
        this.typeDefault = util.Long.fromNumber(this.typeDefault, this.type.charAt(0) === "u");

        /* istanbul ignore else */
        if (Object.freeze)
            Object.freeze(this.typeDefault); // long instances are meant to be immutable anyway (i.e. use small int cache that even requires it)

    ***REMOVED*** else if (this.bytes && typeof this.typeDefault === "string") ***REMOVED***
        var buf;
        if (util.base64.test(this.typeDefault))
            util.base64.decode(this.typeDefault, buf = util.newBuffer(util.base64.length(this.typeDefault)), 0);
        else
            util.utf8.write(this.typeDefault, buf = util.newBuffer(util.utf8.length(this.typeDefault)), 0);
        this.typeDefault = buf;
    ***REMOVED***

    // take special care of maps and repeated fields
    if (this.map)
        this.defaultValue = util.emptyObject;
    else if (this.repeated)
        this.defaultValue = util.emptyArray;
    else
        this.defaultValue = this.typeDefault;

    // ensure proper value on prototype
    if (this.parent instanceof Type)
        this.parent.ctor.prototype[this.name] = this.defaultValue;

    return ReflectionObject.prototype.resolve.call(this);
***REMOVED***;

/**
 * Decorator function as returned by ***REMOVED***@link Field.d***REMOVED*** and ***REMOVED***@link MapField.d***REMOVED*** (TypeScript).
 * @typedef FieldDecorator
 * @type ***REMOVED***function***REMOVED***
 * @param ***REMOVED***Object***REMOVED*** prototype Target prototype
 * @param ***REMOVED***string***REMOVED*** fieldName Field name
 * @returns ***REMOVED***undefined***REMOVED***
 */

/**
 * Field decorator (TypeScript).
 * @name Field.d
 * @function
 * @param ***REMOVED***number***REMOVED*** fieldId Field id
 * @param ***REMOVED***"double"|"float"|"int32"|"uint32"|"sint32"|"fixed32"|"sfixed32"|"int64"|"uint64"|"sint64"|"fixed64"|"sfixed64"|"string"|"bool"|"bytes"|Object***REMOVED*** fieldType Field type
 * @param ***REMOVED***"optional"|"required"|"repeated"***REMOVED*** [fieldRule="optional"] Field rule
 * @param ***REMOVED***T***REMOVED*** [defaultValue] Default value
 * @returns ***REMOVED***FieldDecorator***REMOVED*** Decorator function
 * @template T extends number | number[] | Long | Long[] | string | string[] | boolean | boolean[] | Uint8Array | Uint8Array[] | Buffer | Buffer[]
 */
Field.d = function decorateField(fieldId, fieldType, fieldRule, defaultValue) ***REMOVED***

    // submessage: decorate the submessage and use its name as the type
    if (typeof fieldType === "function")
        fieldType = util.decorateType(fieldType).name;

    // enum reference: create a reflected copy of the enum and keep reuseing it
    else if (fieldType && typeof fieldType === "object")
        fieldType = util.decorateEnum(fieldType).name;

    return function fieldDecorator(prototype, fieldName) ***REMOVED***
        util.decorateType(prototype.constructor)
            .add(new Field(fieldName, fieldId, fieldType, fieldRule, ***REMOVED*** "default": defaultValue ***REMOVED***));
    ***REMOVED***;
***REMOVED***;

/**
 * Field decorator (TypeScript).
 * @name Field.d
 * @function
 * @param ***REMOVED***number***REMOVED*** fieldId Field id
 * @param ***REMOVED***Constructor<T>|string***REMOVED*** fieldType Field type
 * @param ***REMOVED***"optional"|"required"|"repeated"***REMOVED*** [fieldRule="optional"] Field rule
 * @returns ***REMOVED***FieldDecorator***REMOVED*** Decorator function
 * @template T extends Message<T>
 * @variation 2
 */
// like Field.d but without a default value

Field._configure = function configure(Type_) ***REMOVED***
    Type = Type_;
***REMOVED***;
