// protobufjs/src/message.js
"use strict";
module.exports = Message;

var util = require("./util/minimal");

/**
 * Constructs a new message instance.
 * @classdesc Abstract runtime message.
 * @constructor
 * @param ***REMOVED***Properties<T>***REMOVED*** [properties] Properties to set
 * @template T extends object
 */
function Message(properties) ***REMOVED***
    // not used internally
    if (properties)
        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
            this[keys[i]] = properties[keys[i]];
***REMOVED***

/**
 * Reference to the reflected type.
 * @name Message.$type
 * @type ***REMOVED***Type***REMOVED***
 * @readonly
 */

/**
 * Reference to the reflected type.
 * @name Message#$type
 * @type ***REMOVED***Type***REMOVED***
 * @readonly
 */

/*eslint-disable valid-jsdoc*/

/**
 * Creates a new message of this type using the specified properties.
 * @param ***REMOVED***Object.<string,*>***REMOVED*** [properties] Properties to set
 * @returns ***REMOVED***Message<T>***REMOVED*** Message instance
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.create = function create(properties) ***REMOVED***
    return this.$type.create(properties);
***REMOVED***;

/**
 * Encodes a message of this type.
 * @param ***REMOVED***T|Object.<string,*>***REMOVED*** message Message to encode
 * @param ***REMOVED***Writer***REMOVED*** [writer] Writer to use
 * @returns ***REMOVED***Writer***REMOVED*** Writer
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.encode = function encode(message, writer) ***REMOVED***
    return this.$type.encode(message, writer);
***REMOVED***;

/**
 * Encodes a message of this type preceeded by its length as a varint.
 * @param ***REMOVED***T|Object.<string,*>***REMOVED*** message Message to encode
 * @param ***REMOVED***Writer***REMOVED*** [writer] Writer to use
 * @returns ***REMOVED***Writer***REMOVED*** Writer
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.encodeDelimited = function encodeDelimited(message, writer) ***REMOVED***
    return this.$type.encodeDelimited(message, writer);
***REMOVED***;

/**
 * Decodes a message of this type.
 * @name Message.decode
 * @function
 * @param ***REMOVED***Reader|Uint8Array***REMOVED*** reader Reader or buffer to decode
 * @returns ***REMOVED***T***REMOVED*** Decoded message
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.decode = function decode(reader) ***REMOVED***
    return this.$type.decode(reader);
***REMOVED***;

/**
 * Decodes a message of this type preceeded by its length as a varint.
 * @name Message.decodeDelimited
 * @function
 * @param ***REMOVED***Reader|Uint8Array***REMOVED*** reader Reader or buffer to decode
 * @returns ***REMOVED***T***REMOVED*** Decoded message
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.decodeDelimited = function decodeDelimited(reader) ***REMOVED***
    return this.$type.decodeDelimited(reader);
***REMOVED***;

/**
 * Verifies a message of this type.
 * @name Message.verify
 * @function
 * @param ***REMOVED***Object.<string,*>***REMOVED*** message Plain object to verify
 * @returns ***REMOVED***string|null***REMOVED*** `null` if valid, otherwise the reason why it is not
 */
Message.verify = function verify(message) ***REMOVED***
    return this.$type.verify(message);
***REMOVED***;

/**
 * Creates a new message of this type from a plain object. Also converts values to their respective internal types.
 * @param ***REMOVED***Object.<string,*>***REMOVED*** object Plain object
 * @returns ***REMOVED***T***REMOVED*** Message instance
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.fromObject = function fromObject(object) ***REMOVED***
    return this.$type.fromObject(object);
***REMOVED***;

/**
 * Creates a plain object from a message of this type. Also converts values to other types if specified.
 * @param ***REMOVED***T***REMOVED*** message Message instance
 * @param ***REMOVED***IConversionOptions***REMOVED*** [options] Conversion options
 * @returns ***REMOVED***Object.<string,*>***REMOVED*** Plain object
 * @template T extends Message<T>
 * @this Constructor<T>
 */
Message.toObject = function toObject(message, options) ***REMOVED***
    return this.$type.toObject(message, options);
***REMOVED***;

/**
 * Converts this message to JSON.
 * @returns ***REMOVED***Object.<string,*>***REMOVED*** JSON object
 */
Message.prototype.toJSON = function toJSON() ***REMOVED***
    return this.$type.toObject(this, util.toJSONOptions);
***REMOVED***;

/*eslint-enable valid-jsdoc*/
