// protobufjs/src/method.js
"use strict";
module.exports = Method;

// extends ReflectionObject
var ReflectionObject = require("./object");
((Method.prototype = Object.create(ReflectionObject.prototype)).constructor = Method).className = "Method";

var util = require("./util");

/**
 * Constructs a new service method instance.
 * @classdesc Reflected service method.
 * @extends ReflectionObject
 * @constructor
 * @param ***REMOVED***string***REMOVED*** name Method name
 * @param ***REMOVED***string|undefined***REMOVED*** type Method type, usually `"rpc"`
 * @param ***REMOVED***string***REMOVED*** requestType Request message type
 * @param ***REMOVED***string***REMOVED*** responseType Response message type
 * @param ***REMOVED***boolean|Object.<string,*>***REMOVED*** [requestStream] Whether the request is streamed
 * @param ***REMOVED***boolean|Object.<string,*>***REMOVED*** [responseStream] Whether the response is streamed
 * @param ***REMOVED***Object.<string,*>***REMOVED*** [options] Declared options
 */
function Method(name, type, requestType, responseType, requestStream, responseStream, options) ***REMOVED***

    /* istanbul ignore next */
    if (util.isObject(requestStream)) ***REMOVED***
        options = requestStream;
        requestStream = responseStream = undefined;
    ***REMOVED*** else if (util.isObject(responseStream)) ***REMOVED***
        options = responseStream;
        responseStream = undefined;
    ***REMOVED***

    /* istanbul ignore if */
    if (!(type === undefined || util.isString(type)))
        throw TypeError("type must be a string");

    /* istanbul ignore if */
    if (!util.isString(requestType))
        throw TypeError("requestType must be a string");

    /* istanbul ignore if */
    if (!util.isString(responseType))
        throw TypeError("responseType must be a string");

    ReflectionObject.call(this, name, options);

    /**
     * Method type.
     * @type ***REMOVED***string***REMOVED***
     */
    this.type = type || "rpc"; // toJSON

    /**
     * Request type.
     * @type ***REMOVED***string***REMOVED***
     */
    this.requestType = requestType; // toJSON, marker

    /**
     * Whether requests are streamed or not.
     * @type ***REMOVED***boolean|undefined***REMOVED***
     */
    this.requestStream = requestStream ? true : undefined; // toJSON

    /**
     * Response type.
     * @type ***REMOVED***string***REMOVED***
     */
    this.responseType = responseType; // toJSON

    /**
     * Whether responses are streamed or not.
     * @type ***REMOVED***boolean|undefined***REMOVED***
     */
    this.responseStream = responseStream ? true : undefined; // toJSON

    /**
     * Resolved request type.
     * @type ***REMOVED***Type|null***REMOVED***
     */
    this.resolvedRequestType = null;

    /**
     * Resolved response type.
     * @type ***REMOVED***Type|null***REMOVED***
     */
    this.resolvedResponseType = null;
***REMOVED***

/**
 * Method descriptor.
 * @interface IMethod
 * @property ***REMOVED***string***REMOVED*** [type="rpc"] Method type
 * @property ***REMOVED***string***REMOVED*** requestType Request type
 * @property ***REMOVED***string***REMOVED*** responseType Response type
 * @property ***REMOVED***boolean***REMOVED*** [requestStream=false] Whether requests are streamed
 * @property ***REMOVED***boolean***REMOVED*** [responseStream=false] Whether responses are streamed
 * @property ***REMOVED***Object.<string,*>***REMOVED*** [options] Method options
 */

/**
 * Constructs a method from a method descriptor.
 * @param ***REMOVED***string***REMOVED*** name Method name
 * @param ***REMOVED***IMethod***REMOVED*** json Method descriptor
 * @returns ***REMOVED***Method***REMOVED*** Created method
 * @throws ***REMOVED***TypeError***REMOVED*** If arguments are invalid
 */
Method.fromJSON = function fromJSON(name, json) ***REMOVED***
    return new Method(name, json.type, json.requestType, json.responseType, json.requestStream, json.responseStream, json.options);
***REMOVED***;

/**
 * Converts this method to a method descriptor.
 * @returns ***REMOVED***IMethod***REMOVED*** Method descriptor
 */
Method.prototype.toJSON = function toJSON() ***REMOVED***
    return util.toObject([
        "type"           , this.type !== "rpc" && /* istanbul ignore next */ this.type || undefined,
        "requestType"    , this.requestType,
        "requestStream"  , this.requestStream,
        "responseType"   , this.responseType,
        "responseStream" , this.responseStream,
        "options"        , this.options
    ]);
***REMOVED***;

/**
 * @override
 */
Method.prototype.resolve = function resolve() ***REMOVED***

    /* istanbul ignore if */
    if (this.resolved)
        return this;

    this.resolvedRequestType = this.parent.lookupType(this.requestType);
    this.resolvedResponseType = this.parent.lookupType(this.responseType);

    return ReflectionObject.prototype.resolve.call(this);
***REMOVED***;
