// protobufjs/src/parse.js
"use strict";
module.exports = parse;

parse.filename = null;
parse.defaults = ***REMOVED*** keepCase: false ***REMOVED***;

var tokenize  = require("./tokenize"),
    Root      = require("./root"),
    Type      = require("./type"),
    Field     = require("./field"),
    MapField  = require("./mapfield"),
    OneOf     = require("./oneof"),
    Enum      = require("./enum"),
    Service   = require("./service"),
    Method    = require("./method"),
    types     = require("./types"),
    util      = require("./util");

var base10Re    = /^[1-9][0-9]*$/,
    base10NegRe = /^-?[1-9][0-9]*$/,
    base16Re    = /^0[x][0-9a-fA-F]+$/,
    base16NegRe = /^-?0[x][0-9a-fA-F]+$/,
    base8Re     = /^0[0-7]+$/,
    base8NegRe  = /^-?0[0-7]+$/,
    numberRe    = /^(?![eE])[0-9]*(?:\.[0-9]*)?(?:[eE][+-]?[0-9]+)?$/,
    nameRe      = /^[a-zA-Z_][a-zA-Z_0-9]*$/,
    typeRefRe   = /^(?:\.?[a-zA-Z_][a-zA-Z_0-9]*)+$/,
    fqTypeRefRe = /^(?:\.[a-zA-Z][a-zA-Z_0-9]*)+$/;

/**
 * Result object returned from ***REMOVED***@link parse***REMOVED***.
 * @interface IParserResult
 * @property ***REMOVED***string|undefined***REMOVED*** package Package name, if declared
 * @property ***REMOVED***string[]|undefined***REMOVED*** imports Imports, if any
 * @property ***REMOVED***string[]|undefined***REMOVED*** weakImports Weak imports, if any
 * @property ***REMOVED***string|undefined***REMOVED*** syntax Syntax, if specified (either `"proto2"` or `"proto3"`)
 * @property ***REMOVED***Root***REMOVED*** root Populated root instance
 */

/**
 * Options modifying the behavior of ***REMOVED***@link parse***REMOVED***.
 * @interface IParseOptions
 * @property ***REMOVED***boolean***REMOVED*** [keepCase=false] Keeps field casing instead of converting to camel case
 */

/**
 * Parses the given .proto source and returns an object with the parsed contents.
 * @param ***REMOVED***string***REMOVED*** source Source contents
 * @param ***REMOVED***Root***REMOVED*** root Root to populate
 * @param ***REMOVED***IParseOptions***REMOVED*** [options] Parse options. Defaults to ***REMOVED***@link parse.defaults***REMOVED*** when omitted.
 * @returns ***REMOVED***IParserResult***REMOVED*** Parser result
 * @property ***REMOVED***string***REMOVED*** filename=null Currently processing file name for error reporting, if known
 * @property ***REMOVED***IParseOptions***REMOVED*** defaults Default ***REMOVED***@link IParseOptions***REMOVED***
 */
function parse(source, root, options) ***REMOVED***
    /* eslint-disable callback-return */
    if (!(root instanceof Root)) ***REMOVED***
        options = root;
        root = new Root();
    ***REMOVED***
    if (!options)
        options = parse.defaults;

    var tn = tokenize(source),
        next = tn.next,
        push = tn.push,
        peek = tn.peek,
        skip = tn.skip,
        cmnt = tn.cmnt;

    var head = true,
        pkg,
        imports,
        weakImports,
        syntax,
        isProto3 = false;

    var ptr = root;

    var applyCase = options.keepCase ? function(name) ***REMOVED*** return name; ***REMOVED*** : util.camelCase;

    /* istanbul ignore next */
    function illegal(token, name, insideTryCatch) ***REMOVED***
        var filename = parse.filename;
        if (!insideTryCatch)
            parse.filename = null;
        return Error("illegal " + (name || "token") + " '" + token + "' (" + (filename ? filename + ", " : "") + "line " + tn.line + ")");
    ***REMOVED***

    function readString() ***REMOVED***
        var values = [],
            token;
        do ***REMOVED***
            /* istanbul ignore if */
            if ((token = next()) !== "\"" && token !== "'")
                throw illegal(token);

            values.push(next());
            skip(token);
            token = peek();
        ***REMOVED*** while (token === "\"" || token === "'");
        return values.join("");
    ***REMOVED***

    function readValue(acceptTypeRef) ***REMOVED***
        var token = next();
        switch (token) ***REMOVED***
            case "'":
            case "\"":
                push(token);
                return readString();
            case "true": case "TRUE":
                return true;
            case "false": case "FALSE":
                return false;
        ***REMOVED***
        try ***REMOVED***
            return parseNumber(token, /* insideTryCatch */ true);
        ***REMOVED*** catch (e) ***REMOVED***

            /* istanbul ignore else */
            if (acceptTypeRef && typeRefRe.test(token))
                return token;

            /* istanbul ignore next */
            throw illegal(token, "value");
        ***REMOVED***
    ***REMOVED***

    function readRanges(target, acceptStrings) ***REMOVED***
        var token, start;
        do ***REMOVED***
            if (acceptStrings && ((token = peek()) === "\"" || token === "'"))
                target.push(readString());
            else
                target.push([ start = parseId(next()), skip("to", true) ? parseId(next()) : start ]);
        ***REMOVED*** while (skip(",", true));
        skip(";");
    ***REMOVED***

    function parseNumber(token, insideTryCatch) ***REMOVED***
        var sign = 1;
        if (token.charAt(0) === "-") ***REMOVED***
            sign = -1;
            token = token.substring(1);
        ***REMOVED***
        switch (token) ***REMOVED***
            case "inf": case "INF": case "Inf":
                return sign * Infinity;
            case "nan": case "NAN": case "Nan": case "NaN":
                return NaN;
            case "0":
                return 0;
        ***REMOVED***
        if (base10Re.test(token))
            return sign * parseInt(token, 10);
        if (base16Re.test(token))
            return sign * parseInt(token, 16);
        if (base8Re.test(token))
            return sign * parseInt(token, 8);

        /* istanbul ignore else */
        if (numberRe.test(token))
            return sign * parseFloat(token);

        /* istanbul ignore next */
        throw illegal(token, "number", insideTryCatch);
    ***REMOVED***

    function parseId(token, acceptNegative) ***REMOVED***
        switch (token) ***REMOVED***
            case "max": case "MAX": case "Max":
                return 536870911;
            case "0":
                return 0;
        ***REMOVED***

        /* istanbul ignore if */
        if (!acceptNegative && token.charAt(0) === "-")
            throw illegal(token, "id");

        if (base10NegRe.test(token))
            return parseInt(token, 10);
        if (base16NegRe.test(token))
            return parseInt(token, 16);

        /* istanbul ignore else */
        if (base8NegRe.test(token))
            return parseInt(token, 8);

        /* istanbul ignore next */
        throw illegal(token, "id");
    ***REMOVED***

    function parsePackage() ***REMOVED***

        /* istanbul ignore if */
        if (pkg !== undefined)
            throw illegal("package");

        pkg = next();

        /* istanbul ignore if */
        if (!typeRefRe.test(pkg))
            throw illegal(pkg, "name");

        ptr = ptr.define(pkg);
        skip(";");
    ***REMOVED***

    function parseImport() ***REMOVED***
        var token = peek();
        var whichImports;
        switch (token) ***REMOVED***
            case "weak":
                whichImports = weakImports || (weakImports = []);
                next();
                break;
            case "public":
                next();
                // eslint-disable-line no-fallthrough
            default:
                whichImports = imports || (imports = []);
                break;
        ***REMOVED***
        token = readString();
        skip(";");
        whichImports.push(token);
    ***REMOVED***

    function parseSyntax() ***REMOVED***
        skip("=");
        syntax = readString();
        isProto3 = syntax === "proto3";

        /* istanbul ignore if */
        if (!isProto3 && syntax !== "proto2")
            throw illegal(syntax, "syntax");

        skip(";");
    ***REMOVED***

    function parseCommon(parent, token) ***REMOVED***
        switch (token) ***REMOVED***

            case "option":
                parseOption(parent, token);
                skip(";");
                return true;

            case "message":
                parseType(parent, token);
                return true;

            case "enum":
                parseEnum(parent, token);
                return true;

            case "service":
                parseService(parent, token);
                return true;

            case "extend":
                parseExtension(parent, token);
                return true;
        ***REMOVED***
        return false;
    ***REMOVED***

    function ifBlock(obj, fnIf, fnElse) ***REMOVED***
        var trailingLine = tn.line;
        if (obj) ***REMOVED***
            obj.comment = cmnt(); // try block-type comment
            obj.filename = parse.filename;
        ***REMOVED***
        if (skip("***REMOVED***", true)) ***REMOVED***
            var token;
            while ((token = next()) !== "***REMOVED***")
                fnIf(token);
            skip(";", true);
        ***REMOVED*** else ***REMOVED***
            if (fnElse)
                fnElse();
            skip(";");
            if (obj && typeof obj.comment !== "string")
                obj.comment = cmnt(trailingLine); // try line-type comment if no block
        ***REMOVED***
    ***REMOVED***

    function parseType(parent, token) ***REMOVED***

        /* istanbul ignore if */
        if (!nameRe.test(token = next()))
            throw illegal(token, "type name");

        var type = new Type(token);
        ifBlock(type, function parseType_block(token) ***REMOVED***
            if (parseCommon(type, token))
                return;

            switch (token) ***REMOVED***

                case "map":
                    parseMapField(type, token);
                    break;

                case "required":
                case "optional":
                case "repeated":
                    parseField(type, token);
                    break;

                case "oneof":
                    parseOneOf(type, token);
                    break;

                case "extensions":
                    readRanges(type.extensions || (type.extensions = []));
                    break;

                case "reserved":
                    readRanges(type.reserved || (type.reserved = []), true);
                    break;

                default:
                    /* istanbul ignore if */
                    if (!isProto3 || !typeRefRe.test(token))
                        throw illegal(token);

                    push(token);
                    parseField(type, "optional");
                    break;
            ***REMOVED***
        ***REMOVED***);
        parent.add(type);
    ***REMOVED***

    function parseField(parent, rule, extend) ***REMOVED***
        var type = next();
        if (type === "group") ***REMOVED***
            parseGroup(parent, rule);
            return;
        ***REMOVED***

        /* istanbul ignore if */
        if (!typeRefRe.test(type))
            throw illegal(type, "type");

        var name = next();

        /* istanbul ignore if */
        if (!nameRe.test(name))
            throw illegal(name, "name");

        name = applyCase(name);
        skip("=");

        var field = new Field(name, parseId(next()), type, rule, extend);
        ifBlock(field, function parseField_block(token) ***REMOVED***

            /* istanbul ignore else */
            if (token === "option") ***REMOVED***
                parseOption(field, token);
                skip(";");
            ***REMOVED*** else
                throw illegal(token);

        ***REMOVED***, function parseField_line() ***REMOVED***
            parseInlineOptions(field);
        ***REMOVED***);
        parent.add(field);

        // JSON defaults to packed=true if not set so we have to set packed=false explicity when
        // parsing proto2 descriptors without the option, where applicable. This must be done for
        // all known packable types and anything that could be an enum (= is not a basic type).
        if (!isProto3 && field.repeated && (types.packed[type] !== undefined || types.basic[type] === undefined))
            field.setOption("packed", false, /* ifNotSet */ true);
    ***REMOVED***

    function parseGroup(parent, rule) ***REMOVED***
        var name = next();

        /* istanbul ignore if */
        if (!nameRe.test(name))
            throw illegal(name, "name");

        var fieldName = util.lcFirst(name);
        if (name === fieldName)
            name = util.ucFirst(name);
        skip("=");
        var id = parseId(next());
        var type = new Type(name);
        type.group = true;
        var field = new Field(fieldName, id, name, rule);
        field.filename = parse.filename;
        ifBlock(type, function parseGroup_block(token) ***REMOVED***
            switch (token) ***REMOVED***

                case "option":
                    parseOption(type, token);
                    skip(";");
                    break;

                case "required":
                case "optional":
                case "repeated":
                    parseField(type, token);
                    break;

                /* istanbul ignore next */
                default:
                    throw illegal(token); // there are no groups with proto3 semantics
            ***REMOVED***
        ***REMOVED***);
        parent.add(type)
              .add(field);
    ***REMOVED***

    function parseMapField(parent) ***REMOVED***
        skip("<");
        var keyType = next();

        /* istanbul ignore if */
        if (types.mapKey[keyType] === undefined)
            throw illegal(keyType, "type");

        skip(",");
        var valueType = next();

        /* istanbul ignore if */
        if (!typeRefRe.test(valueType))
            throw illegal(valueType, "type");

        skip(">");
        var name = next();

        /* istanbul ignore if */
        if (!nameRe.test(name))
            throw illegal(name, "name");

        skip("=");
        var field = new MapField(applyCase(name), parseId(next()), keyType, valueType);
        ifBlock(field, function parseMapField_block(token) ***REMOVED***

            /* istanbul ignore else */
            if (token === "option") ***REMOVED***
                parseOption(field, token);
                skip(";");
            ***REMOVED*** else
                throw illegal(token);

        ***REMOVED***, function parseMapField_line() ***REMOVED***
            parseInlineOptions(field);
        ***REMOVED***);
        parent.add(field);
    ***REMOVED***

    function parseOneOf(parent, token) ***REMOVED***

        /* istanbul ignore if */
        if (!nameRe.test(token = next()))
            throw illegal(token, "name");

        var oneof = new OneOf(applyCase(token));
        ifBlock(oneof, function parseOneOf_block(token) ***REMOVED***
            if (token === "option") ***REMOVED***
                parseOption(oneof, token);
                skip(";");
            ***REMOVED*** else ***REMOVED***
                push(token);
                parseField(oneof, "optional");
            ***REMOVED***
        ***REMOVED***);
        parent.add(oneof);
    ***REMOVED***

    function parseEnum(parent, token) ***REMOVED***

        /* istanbul ignore if */
        if (!nameRe.test(token = next()))
            throw illegal(token, "name");

        var enm = new Enum(token);
        ifBlock(enm, function parseEnum_block(token) ***REMOVED***
          switch(token) ***REMOVED***
            case "option":
              parseOption(enm, token);
              skip(";");
              break;

            case "reserved":
              readRanges(enm.reserved || (enm.reserved = []), true);
              break;

            default:
              parseEnumValue(enm, token);
          ***REMOVED***
        ***REMOVED***);
        parent.add(enm);
    ***REMOVED***

    function parseEnumValue(parent, token) ***REMOVED***

        /* istanbul ignore if */
        if (!nameRe.test(token))
            throw illegal(token, "name");

        skip("=");
        var value = parseId(next(), true),
            dummy = ***REMOVED******REMOVED***;
        ifBlock(dummy, function parseEnumValue_block(token) ***REMOVED***

            /* istanbul ignore else */
            if (token === "option") ***REMOVED***
                parseOption(dummy, token); // skip
                skip(";");
            ***REMOVED*** else
                throw illegal(token);

        ***REMOVED***, function parseEnumValue_line() ***REMOVED***
            parseInlineOptions(dummy); // skip
        ***REMOVED***);
        parent.add(token, value, dummy.comment);
    ***REMOVED***

    function parseOption(parent, token) ***REMOVED***
        var isCustom = skip("(", true);

        /* istanbul ignore if */
        if (!typeRefRe.test(token = next()))
            throw illegal(token, "name");

        var name = token;
        if (isCustom) ***REMOVED***
            skip(")");
            name = "(" + name + ")";
            token = peek();
            if (fqTypeRefRe.test(token)) ***REMOVED***
                name += token;
                next();
            ***REMOVED***
        ***REMOVED***
        skip("=");
        parseOptionValue(parent, name);
    ***REMOVED***

    function parseOptionValue(parent, name) ***REMOVED***
        if (skip("***REMOVED***", true)) ***REMOVED*** // ***REMOVED*** a: "foo" b ***REMOVED*** c: "bar" ***REMOVED*** ***REMOVED***
            do ***REMOVED***
                /* istanbul ignore if */
                if (!nameRe.test(token = next()))
                    throw illegal(token, "name");

                if (peek() === "***REMOVED***")
                    parseOptionValue(parent, name + "." + token);
                else ***REMOVED***
                    skip(":");
                    if (peek() === "***REMOVED***")
                        parseOptionValue(parent, name + "." + token);
                    else
                        setOption(parent, name + "." + token, readValue(true));
                ***REMOVED***
            ***REMOVED*** while (!skip("***REMOVED***", true));
        ***REMOVED*** else
            setOption(parent, name, readValue(true));
        // Does not enforce a delimiter to be universal
    ***REMOVED***

    function setOption(parent, name, value) ***REMOVED***
        if (parent.setOption)
            parent.setOption(name, value);
    ***REMOVED***

    function parseInlineOptions(parent) ***REMOVED***
        if (skip("[", true)) ***REMOVED***
            do ***REMOVED***
                parseOption(parent, "option");
            ***REMOVED*** while (skip(",", true));
            skip("]");
        ***REMOVED***
        return parent;
    ***REMOVED***

    function parseService(parent, token) ***REMOVED***

        /* istanbul ignore if */
        if (!nameRe.test(token = next()))
            throw illegal(token, "service name");

        var service = new Service(token);
        ifBlock(service, function parseService_block(token) ***REMOVED***
            if (parseCommon(service, token))
                return;

            /* istanbul ignore else */
            if (token === "rpc")
                parseMethod(service, token);
            else
                throw illegal(token);
        ***REMOVED***);
        parent.add(service);
    ***REMOVED***

    function parseMethod(parent, token) ***REMOVED***
        var type = token;

        /* istanbul ignore if */
        if (!nameRe.test(token = next()))
            throw illegal(token, "name");

        var name = token,
            requestType, requestStream,
            responseType, responseStream;

        skip("(");
        if (skip("stream", true))
            requestStream = true;

        /* istanbul ignore if */
        if (!typeRefRe.test(token = next()))
            throw illegal(token);

        requestType = token;
        skip(")"); skip("returns"); skip("(");
        if (skip("stream", true))
            responseStream = true;

        /* istanbul ignore if */
        if (!typeRefRe.test(token = next()))
            throw illegal(token);

        responseType = token;
        skip(")");

        var method = new Method(name, type, requestType, responseType, requestStream, responseStream);
        ifBlock(method, function parseMethod_block(token) ***REMOVED***

            /* istanbul ignore else */
            if (token === "option") ***REMOVED***
                parseOption(method, token);
                skip(";");
            ***REMOVED*** else
                throw illegal(token);

        ***REMOVED***);
        parent.add(method);
    ***REMOVED***

    function parseExtension(parent, token) ***REMOVED***

        /* istanbul ignore if */
        if (!typeRefRe.test(token = next()))
            throw illegal(token, "reference");

        var reference = token;
        ifBlock(null, function parseExtension_block(token) ***REMOVED***
            switch (token) ***REMOVED***

                case "required":
                case "repeated":
                case "optional":
                    parseField(parent, token, reference);
                    break;

                default:
                    /* istanbul ignore if */
                    if (!isProto3 || !typeRefRe.test(token))
                        throw illegal(token);
                    push(token);
                    parseField(parent, "optional", reference);
                    break;
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    var token;
    while ((token = next()) !== null) ***REMOVED***
        switch (token) ***REMOVED***

            case "package":

                /* istanbul ignore if */
                if (!head)
                    throw illegal(token);

                parsePackage();
                break;

            case "import":

                /* istanbul ignore if */
                if (!head)
                    throw illegal(token);

                parseImport();
                break;

            case "syntax":

                /* istanbul ignore if */
                if (!head)
                    throw illegal(token);

                parseSyntax();
                break;

            case "option":

                /* istanbul ignore if */
                if (!head)
                    throw illegal(token);

                parseOption(ptr, token);
                skip(";");
                break;

            default:

                /* istanbul ignore else */
                if (parseCommon(ptr, token)) ***REMOVED***
                    head = false;
                    continue;
                ***REMOVED***

                /* istanbul ignore next */
                throw illegal(token);
        ***REMOVED***
    ***REMOVED***

    parse.filename = null;
    return ***REMOVED***
        "package"     : pkg,
        "imports"     : imports,
         weakImports  : weakImports,
         syntax       : syntax,
         root         : root
    ***REMOVED***;
***REMOVED***

/**
 * Parses the given .proto source and returns an object with the parsed contents.
 * @name parse
 * @function
 * @param ***REMOVED***string***REMOVED*** source Source contents
 * @param ***REMOVED***IParseOptions***REMOVED*** [options] Parse options. Defaults to ***REMOVED***@link parse.defaults***REMOVED*** when omitted.
 * @returns ***REMOVED***IParserResult***REMOVED*** Parser result
 * @property ***REMOVED***string***REMOVED*** filename=null Currently processing file name for error reporting, if known
 * @property ***REMOVED***IParseOptions***REMOVED*** defaults Default ***REMOVED***@link IParseOptions***REMOVED***
 * @variation 2
 */
