// protobufjs/src/writer_buffer.js
"use strict";
module.exports = BufferWriter;

// extends Writer
var Writer = require("./writer");
(BufferWriter.prototype = Object.create(Writer.prototype)).constructor = BufferWriter;

var util = require("./util/minimal");

var Buffer = util.Buffer;

/**
 * Constructs a new buffer writer instance.
 * @classdesc Wire format writer using node buffers.
 * @extends Writer
 * @constructor
 */
function BufferWriter() ***REMOVED***
    Writer.call(this);
***REMOVED***

/**
 * Allocates a buffer of the specified size.
 * @param ***REMOVED***number***REMOVED*** size Buffer size
 * @returns ***REMOVED***Buffer***REMOVED*** Buffer
 */
BufferWriter.alloc = function alloc_buffer(size) ***REMOVED***
    return (BufferWriter.alloc = util._Buffer_allocUnsafe)(size);
***REMOVED***;

var writeBytesBuffer = Buffer && Buffer.prototype instanceof Uint8Array && Buffer.prototype.set.name === "set"
    ? function writeBytesBuffer_set(val, buf, pos) ***REMOVED***
        buf.set(val, pos); // faster than copy (requires node >= 4 where Buffers extend Uint8Array and set is properly inherited)
                           // also works for plain array values
    ***REMOVED***
    /* istanbul ignore next */
    : function writeBytesBuffer_copy(val, buf, pos) ***REMOVED***
        if (val.copy) // Buffer values
            val.copy(buf, pos, 0, val.length);
        else for (var i = 0; i < val.length;) // plain array values
            buf[pos++] = val[i++];
    ***REMOVED***;

/**
 * @override
 */
BufferWriter.prototype.bytes = function write_bytes_buffer(value) ***REMOVED***
    if (util.isString(value))
        value = util._Buffer_from(value, "base64");
    var len = value.length >>> 0;
    this.uint32(len);
    if (len)
        this._push(writeBytesBuffer, len, value);
    return this;
***REMOVED***;

function writeStringBuffer(val, buf, pos) ***REMOVED***
    if (val.length < 40) // plain js is faster for short strings (probably due to redundant assertions)
        util.utf8.write(val, buf, pos);
    else
        buf.utf8Write(val, pos);
***REMOVED***

/**
 * @override
 */
BufferWriter.prototype.string = function write_string_buffer(value) ***REMOVED***
    var len = Buffer.byteLength(value);
    this.uint32(len);
    if (len)
        this._push(writeStringBuffer, len, value);
    return this;
***REMOVED***;


/**
 * Finishes the write operation.
 * @name BufferWriter#finish
 * @function
 * @returns ***REMOVED***Buffer***REMOVED*** Finished buffer
 */
