// @firebase/database/dist/esm/src/core/util/libs/parser.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** Path ***REMOVED*** from '../Path';
import ***REMOVED*** RepoInfo ***REMOVED*** from '../../RepoInfo';
import ***REMOVED*** warnIfPageIsSecure, fatal ***REMOVED*** from '../util';
/**
 * @param ***REMOVED***!string***REMOVED*** pathString
 * @return ***REMOVED***string***REMOVED***
 */
function decodePath(pathString) ***REMOVED***
    var pathStringDecoded = '';
    var pieces = pathString.split('/');
    for (var i = 0; i < pieces.length; i++) ***REMOVED***
        if (pieces[i].length > 0) ***REMOVED***
            var piece = pieces[i];
            try ***REMOVED***
                piece = decodeURIComponent(piece.replace(/\+/g, ' '));
            ***REMOVED***
            catch (e) ***REMOVED*** ***REMOVED***
            pathStringDecoded += '/' + piece;
        ***REMOVED***
    ***REMOVED***
    return pathStringDecoded;
***REMOVED***
/**
 *
 * @param ***REMOVED***!string***REMOVED*** dataURL
 * @return ***REMOVED******REMOVED***repoInfo: !RepoInfo, path: !Path***REMOVED******REMOVED***
 */
export var parseRepoInfo = function (dataURL) ***REMOVED***
    var parsedUrl = parseURL(dataURL), namespace = parsedUrl.subdomain;
    if (parsedUrl.domain === 'firebase') ***REMOVED***
        fatal(parsedUrl.host +
            ' is no longer supported. ' +
            'Please use <YOUR FIREBASE>.firebaseio.com instead');
    ***REMOVED***
    // Catch common error of uninitialized namespace value.
    if ((!namespace || namespace == 'undefined') &&
        parsedUrl.domain !== 'localhost') ***REMOVED***
        fatal('Cannot parse Firebase url. Please use https://<YOUR FIREBASE>.firebaseio.com');
    ***REMOVED***
    if (!parsedUrl.secure) ***REMOVED***
        warnIfPageIsSecure();
    ***REMOVED***
    var webSocketOnly = parsedUrl.scheme === 'ws' || parsedUrl.scheme === 'wss';
    return ***REMOVED***
        repoInfo: new RepoInfo(parsedUrl.host, parsedUrl.secure, namespace, webSocketOnly),
        path: new Path(parsedUrl.pathString)
    ***REMOVED***;
***REMOVED***;
/**
 *
 * @param ***REMOVED***!string***REMOVED*** dataURL
 * @return ***REMOVED******REMOVED***host: string, port: number, domain: string, subdomain: string, secure: boolean, scheme: string, pathString: string***REMOVED******REMOVED***
 */
export var parseURL = function (dataURL) ***REMOVED***
    // Default to empty strings in the event of a malformed string.
    var host = '', domain = '', subdomain = '', pathString = '';
    // Always default to SSL, unless otherwise specified.
    var secure = true, scheme = 'https', port = 443;
    // Don't do any validation here. The caller is responsible for validating the result of parsing.
    if (typeof dataURL === 'string') ***REMOVED***
        // Parse scheme.
        var colonInd = dataURL.indexOf('//');
        if (colonInd >= 0) ***REMOVED***
            scheme = dataURL.substring(0, colonInd - 1);
            dataURL = dataURL.substring(colonInd + 2);
        ***REMOVED***
        // Parse host and path.
        var slashInd = dataURL.indexOf('/');
        if (slashInd === -1) ***REMOVED***
            slashInd = dataURL.length;
        ***REMOVED***
        host = dataURL.substring(0, slashInd);
        pathString = decodePath(dataURL.substring(slashInd));
        // If we have a port, use scheme for determining if it's secure.
        colonInd = host.indexOf(':');
        if (colonInd >= 0) ***REMOVED***
            secure = scheme === 'https' || scheme === 'wss';
            port = parseInt(host.substring(colonInd + 1), 10);
        ***REMOVED***
        else ***REMOVED***
            colonInd = dataURL.length;
        ***REMOVED***
        var parts = host.split('.');
        if (parts.length === 3) ***REMOVED***
            // Normalize namespaces to lowercase to share storage / connection.
            domain = parts[1];
            subdomain = parts[0].toLowerCase();
        ***REMOVED***
        else if (parts.length === 2) ***REMOVED***
            domain = parts[0];
        ***REMOVED***
        else if (parts[0].slice(0, colonInd).toLowerCase() === 'localhost') ***REMOVED***
            domain = 'localhost';
        ***REMOVED***
    ***REMOVED***
    return ***REMOVED***
        host: host,
        port: port,
        domain: domain,
        subdomain: subdomain,
        secure: secure,
        scheme: scheme,
        pathString: pathString
    ***REMOVED***;
***REMOVED***;
