// espurify/build/espurify.js
/**
 * Modules in this bundle
 * @license
 *
 * espurify:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   author: Takuto Wada <takuto.wada@gmail.com>
 *   homepage: https://github.com/estools/espurify
 *   version: 1.7.0
 *
 * core-js:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   maintainers: zloirock <zloirock@zloirock.ru>
 *   homepage: https://github.com/zloirock/core-js#readme
 *   version: 2.4.1
 *
 * This header is generated by licensify (https://github.com/twada/licensify)
 */
(function(f)***REMOVED***if(typeof exports==="object"&&typeof module!=="undefined")***REMOVED***module.exports=f()***REMOVED***else if(typeof define==="function"&&define.amd)***REMOVED***define([],f)***REMOVED***else***REMOVED***var g;if(typeof window!=="undefined")***REMOVED***g=window***REMOVED***else if(typeof global!=="undefined")***REMOVED***g=global***REMOVED***else if(typeof self!=="undefined")***REMOVED***g=self***REMOVED***else***REMOVED***g=this***REMOVED***g.espurify = f()***REMOVED******REMOVED***)(function()***REMOVED***var define,module,exports;return (function e(t,n,r)***REMOVED***function s(o,u)***REMOVED***if(!n[o])***REMOVED***if(!t[o])***REMOVED***var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw (f.code="MODULE_NOT_FOUND", f)***REMOVED***var l=n[o]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;t[o][0].call(l.exports,function(e)***REMOVED***var n=t[o][1][e];return s(n?n:e)***REMOVED***,l,l.exports,e,t,n,r)***REMOVED***return n[o].exports***REMOVED***var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s***REMOVED***)(***REMOVED***1:[function(_dereq_,module,exports)***REMOVED***
/**
 * espurify - Clone new AST without extra properties
 * 
 * https://github.com/estools/espurify
 *
 * Copyright (c) 2014-2017 Takuto Wada
 * Licensed under the MIT license.
 *   https://github.com/estools/espurify/blob/master/MIT-LICENSE.txt
 */
'use strict';

var createWhitelist = _dereq_('./lib/create-whitelist');
var cloneWithWhitelist = _dereq_('./lib/clone-ast');

function createCloneFunction (options) ***REMOVED***
    return cloneWithWhitelist(createWhitelist(options));
***REMOVED***

var espurify = createCloneFunction();
espurify.customize = createCloneFunction;
espurify.cloneWithWhitelist = cloneWithWhitelist;
module.exports = espurify;

***REMOVED***,***REMOVED***"./lib/clone-ast":3,"./lib/create-whitelist":4***REMOVED***],2:[function(_dereq_,module,exports)***REMOVED***
module.exports = ***REMOVED***
    ArrayExpression: ['type', 'elements'],
    ArrayPattern: ['type', 'elements'],
    ArrowFunctionExpression: ['type', 'id', 'params', 'body', 'generator', 'expression', 'async'],
    AssignmentExpression: ['type', 'operator', 'left', 'right'],
    AssignmentPattern: ['type', 'left', 'right'],
    AwaitExpression: ['type', 'argument'],
    BinaryExpression: ['type', 'operator', 'left', 'right'],
    BlockStatement: ['type', 'body'],
    BreakStatement: ['type', 'label'],
    CallExpression: ['type', 'callee', 'arguments'],
    CatchClause: ['type', 'param', 'guard', 'body'],
    ClassBody: ['type', 'body'],
    ClassDeclaration: ['type', 'id', 'superClass', 'body'],
    ClassExpression: ['type', 'id', 'superClass', 'body'],
    ConditionalExpression: ['type', 'test', 'consequent', 'alternate'],
    ContinueStatement: ['type', 'label'],
    DebuggerStatement: ['type'],
    DoWhileStatement: ['type', 'body', 'test'],
    EmptyStatement: ['type'],
    ExportAllDeclaration: ['type', 'source'],
    ExportDefaultDeclaration: ['type', 'declaration'],
    ExportNamedDeclaration: ['type', 'declaration', 'specifiers', 'source'],
    ExportSpecifier: ['type', 'exported', 'local'],
    ExpressionStatement: ['type', 'expression'],
    ForInStatement: ['type', 'left', 'right', 'body'],
    ForOfStatement: ['type', 'left', 'right', 'body'],
    ForStatement: ['type', 'init', 'test', 'update', 'body'],
    FunctionDeclaration: ['type', 'id', 'params', 'body', 'generator', 'async'],
    FunctionExpression: ['type', 'id', 'params', 'body', 'generator', 'async'],
    Identifier: ['type', 'name'],
    IfStatement: ['type', 'test', 'consequent', 'alternate'],
    ImportDeclaration: ['type', 'specifiers', 'source'],
    ImportDefaultSpecifier: ['type', 'local'],
    ImportNamespaceSpecifier: ['type', 'local'],
    ImportSpecifier: ['type', 'imported', 'local'],
    LabeledStatement: ['type', 'label', 'body'],
    Literal: ['type', 'value', 'regex'],
    LogicalExpression: ['type', 'operator', 'left', 'right'],
    MemberExpression: ['type', 'object', 'property', 'computed'],
    MetaProperty: ['type', 'meta', 'property'],
    MethodDefinition: ['type', 'key', 'value', 'kind', 'computed', 'static'],
    NewExpression: ['type', 'callee', 'arguments'],
    ObjectExpression: ['type', 'properties'],
    ObjectPattern: ['type', 'properties'],
    Program: ['type', 'body', 'sourceType'],
    Property: ['type', 'key', 'value', 'kind', 'method', 'shorthand', 'computed'],
    RestElement: ['type', 'argument'],
    ReturnStatement: ['type', 'argument'],
    SequenceExpression: ['type', 'expressions'],
    SpreadElement: ['type', 'argument'],
    Super: ['type'],
    SwitchCase: ['type', 'test', 'consequent'],
    SwitchStatement: ['type', 'discriminant', 'cases', 'lexical'],
    TaggedTemplateExpression: ['type', 'tag', 'quasi'],
    TemplateElement: ['type', 'tail', 'value'],
    TemplateLiteral: ['type', 'quasis', 'expressions'],
    ThisExpression: ['type'],
    ThrowStatement: ['type', 'argument'],
    TryStatement: ['type', 'block', 'handler', 'finalizer'],
    UnaryExpression: ['type', 'operator', 'prefix', 'argument'],
    UpdateExpression: ['type', 'operator', 'argument', 'prefix'],
    VariableDeclaration: ['type', 'declarations', 'kind'],
    VariableDeclarator: ['type', 'id', 'init'],
    WhileStatement: ['type', 'test', 'body'],
    WithStatement: ['type', 'object', 'body'],
    YieldExpression: ['type', 'argument', 'delegate']
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],3:[function(_dereq_,module,exports)***REMOVED***
'use strict';

var isArray = _dereq_('core-js/library/fn/array/is-array');
var objectKeys = _dereq_('core-js/library/fn/object/keys');
var indexOf = _dereq_('core-js/library/fn/array/index-of');
var reduce = _dereq_('core-js/library/fn/array/reduce');

module.exports = function cloneWithWhitelist (astWhiteList) ***REMOVED***
    var whitelist = reduce(objectKeys(astWhiteList), function (props, key) ***REMOVED***
        var propNames = astWhiteList[key];
        var prepend = (indexOf(propNames, 'type') === -1) ? ['type'] : [];
        props[key] = prepend.concat(propNames);
        return props;
    ***REMOVED***, ***REMOVED******REMOVED***);

    function cloneNodeOrObject (obj) ***REMOVED***
        var props = obj.type ? whitelist[obj.type] : null;
        if (props) ***REMOVED***
            return cloneNode(obj, props);
        ***REMOVED*** else ***REMOVED***
            return cloneObject(obj);
        ***REMOVED***
    ***REMOVED***

    function cloneArray (ary) ***REMOVED***
        var i = ary.length, clone = [];
        while (i--) ***REMOVED***
            clone[i] = cloneOf(ary[i]);
        ***REMOVED***
        return clone;
    ***REMOVED***

    function cloneNode (node, props) ***REMOVED***
        var i, len, key, clone = ***REMOVED******REMOVED***;
        for (i = 0, len = props.length; i < len; i += 1) ***REMOVED***
            key = props[i];
            if (node.hasOwnProperty(key)) ***REMOVED***
                clone[key] = cloneOf(node[key]);
            ***REMOVED***
        ***REMOVED***
        return clone;
    ***REMOVED***

    function cloneObject (obj) ***REMOVED***
        var props = objectKeys(obj);
        var i, len, key, clone = ***REMOVED******REMOVED***;
        for (i = 0, len = props.length; i < len; i += 1) ***REMOVED***
            key = props[i];
            clone[key] = cloneOf(obj[key]);
        ***REMOVED***
        return clone;
    ***REMOVED***

    function cloneOf (val) ***REMOVED***
        if (typeof val === 'object' && val !== null) ***REMOVED***
            if (val instanceof RegExp) ***REMOVED***
                return new RegExp(val);
            ***REMOVED*** else if (isArray(val)) ***REMOVED***
                return cloneArray(val);
            ***REMOVED*** else ***REMOVED***
                return cloneNodeOrObject(val);
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            return val;
        ***REMOVED***
    ***REMOVED***

    return cloneNodeOrObject;
***REMOVED***;

***REMOVED***,***REMOVED***"core-js/library/fn/array/index-of":5,"core-js/library/fn/array/is-array":6,"core-js/library/fn/array/reduce":7,"core-js/library/fn/object/keys":9***REMOVED***],4:[function(_dereq_,module,exports)***REMOVED***
'use strict';

var defaultProps = _dereq_('./ast-properties');
var objectKeys = _dereq_('core-js/library/fn/object/keys');
var assign = _dereq_('core-js/library/fn/object/assign');

module.exports = function createWhitelist (options) ***REMOVED***
    var opts = assign(***REMOVED******REMOVED***, options);
    var typeName, i, len;
    var keys = objectKeys(defaultProps);
    var result = ***REMOVED******REMOVED***;
    for (i = 0, len = keys.length; i < len; i += 1) ***REMOVED***
        typeName = keys[i];
        result[typeName] = defaultProps[typeName].concat(opts.extra);
    ***REMOVED***
    return result;
***REMOVED***;

***REMOVED***,***REMOVED***"./ast-properties":2,"core-js/library/fn/object/assign":8,"core-js/library/fn/object/keys":9***REMOVED***],5:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.index-of');
module.exports = _dereq_('../../modules/_core').Array.indexOf;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.index-of":48***REMOVED***],6:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.is-array');
module.exports = _dereq_('../../modules/_core').Array.isArray;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.is-array":49***REMOVED***],7:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.reduce');
module.exports = _dereq_('../../modules/_core').Array.reduce;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.reduce":50***REMOVED***],8:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.assign');
module.exports = _dereq_('../../modules/_core').Object.assign;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.object.assign":51***REMOVED***],9:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.keys');
module.exports = _dereq_('../../modules/_core').Object.keys;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.object.keys":52***REMOVED***],10:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(it)***REMOVED***
  if(typeof it != 'function')throw TypeError(it + ' is not a function!');
  return it;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],11:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object');
module.exports = function(it)***REMOVED***
  if(!isObject(it))throw TypeError(it + ' is not an object!');
  return it;
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-object":29***REMOVED***],12:[function(_dereq_,module,exports)***REMOVED***
// false -> Array#indexOf
// true  -> Array#includes
var toIObject = _dereq_('./_to-iobject')
  , toLength  = _dereq_('./_to-length')
  , toIndex   = _dereq_('./_to-index');
module.exports = function(IS_INCLUDES)***REMOVED***
  return function($this, el, fromIndex)***REMOVED***
    var O      = toIObject($this)
      , length = toLength(O.length)
      , index  = toIndex(fromIndex, length)
      , value;
    // Array#includes uses SameValueZero equality algorithm
    if(IS_INCLUDES && el != el)while(length > index)***REMOVED***
      value = O[index++];
      if(value != value)return true;
    // Array#toIndex ignores holes, Array#includes - not
    ***REMOVED*** else for(;length > index; index++)if(IS_INCLUDES || index in O)***REMOVED***
      if(O[index] === el)return IS_INCLUDES || index || 0;
    ***REMOVED*** return !IS_INCLUDES && -1;
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-index":41,"./_to-iobject":43,"./_to-length":44***REMOVED***],13:[function(_dereq_,module,exports)***REMOVED***
var aFunction = _dereq_('./_a-function')
  , toObject  = _dereq_('./_to-object')
  , IObject   = _dereq_('./_iobject')
  , toLength  = _dereq_('./_to-length');

module.exports = function(that, callbackfn, aLen, memo, isRight)***REMOVED***
  aFunction(callbackfn);
  var O      = toObject(that)
    , self   = IObject(O)
    , length = toLength(O.length)
    , index  = isRight ? length - 1 : 0
    , i      = isRight ? -1 : 1;
  if(aLen < 2)for(;;)***REMOVED***
    if(index in self)***REMOVED***
      memo = self[index];
      index += i;
      break;
    ***REMOVED***
    index += i;
    if(isRight ? index < 0 : length <= index)***REMOVED***
      throw TypeError('Reduce of empty array with no initial value');
    ***REMOVED***
  ***REMOVED***
  for(;isRight ? index >= 0 : length > index; index += i)if(index in self)***REMOVED***
    memo = callbackfn(memo, self[index], index, O);
  ***REMOVED***
  return memo;
***REMOVED***;
***REMOVED***,***REMOVED***"./_a-function":10,"./_iobject":27,"./_to-length":44,"./_to-object":45***REMOVED***],14:[function(_dereq_,module,exports)***REMOVED***
var toString = ***REMOVED******REMOVED***.toString;

module.exports = function(it)***REMOVED***
  return toString.call(it).slice(8, -1);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],15:[function(_dereq_,module,exports)***REMOVED***
var core = module.exports = ***REMOVED***version: '2.4.0'***REMOVED***;
if(typeof __e == 'number')__e = core; // eslint-disable-line no-undef
***REMOVED***,***REMOVED******REMOVED***],16:[function(_dereq_,module,exports)***REMOVED***
// optional / simple context binding
var aFunction = _dereq_('./_a-function');
module.exports = function(fn, that, length)***REMOVED***
  aFunction(fn);
  if(that === undefined)return fn;
  switch(length)***REMOVED***
    case 1: return function(a)***REMOVED***
      return fn.call(that, a);
    ***REMOVED***;
    case 2: return function(a, b)***REMOVED***
      return fn.call(that, a, b);
    ***REMOVED***;
    case 3: return function(a, b, c)***REMOVED***
      return fn.call(that, a, b, c);
    ***REMOVED***;
  ***REMOVED***
  return function(/* ...args */)***REMOVED***
    return fn.apply(that, arguments);
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_a-function":10***REMOVED***],17:[function(_dereq_,module,exports)***REMOVED***
// 7.2.1 RequireObjectCoercible(argument)
module.exports = function(it)***REMOVED***
  if(it == undefined)throw TypeError("Can't call method on  " + it);
  return it;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],18:[function(_dereq_,module,exports)***REMOVED***
// Thank's IE8 for his funny defineProperty
module.exports = !_dereq_('./_fails')(function()***REMOVED***
  return Object.defineProperty(***REMOVED******REMOVED***, 'a', ***REMOVED***get: function()***REMOVED*** return 7; ***REMOVED******REMOVED***).a != 7;
***REMOVED***);
***REMOVED***,***REMOVED***"./_fails":22***REMOVED***],19:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object')
  , document = _dereq_('./_global').document
  // in old IE typeof document.createElement is 'object'
  , is = isObject(document) && isObject(document.createElement);
module.exports = function(it)***REMOVED***
  return is ? document.createElement(it) : ***REMOVED******REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_global":23,"./_is-object":29***REMOVED***],20:[function(_dereq_,module,exports)***REMOVED***
// IE 8- don't enum bug keys
module.exports = (
  'constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf'
).split(',');
***REMOVED***,***REMOVED******REMOVED***],21:[function(_dereq_,module,exports)***REMOVED***
var global    = _dereq_('./_global')
  , core      = _dereq_('./_core')
  , ctx       = _dereq_('./_ctx')
  , hide      = _dereq_('./_hide')
  , PROTOTYPE = 'prototype';

var $export = function(type, name, source)***REMOVED***
  var IS_FORCED = type & $export.F
    , IS_GLOBAL = type & $export.G
    , IS_STATIC = type & $export.S
    , IS_PROTO  = type & $export.P
    , IS_BIND   = type & $export.B
    , IS_WRAP   = type & $export.W
    , exports   = IS_GLOBAL ? core : core[name] || (core[name] = ***REMOVED******REMOVED***)
    , expProto  = exports[PROTOTYPE]
    , target    = IS_GLOBAL ? global : IS_STATIC ? global[name] : (global[name] || ***REMOVED******REMOVED***)[PROTOTYPE]
    , key, own, out;
  if(IS_GLOBAL)source = name;
  for(key in source)***REMOVED***
    // contains in native
    own = !IS_FORCED && target && target[key] !== undefined;
    if(own && key in exports)continue;
    // export native or passed
    out = own ? target[key] : source[key];
    // prevent global pollution for namespaces
    exports[key] = IS_GLOBAL && typeof target[key] != 'function' ? source[key]
    // bind timers to global for call from export context
    : IS_BIND && own ? ctx(out, global)
    // wrap global constructors for prevent change them in library
    : IS_WRAP && target[key] == out ? (function(C)***REMOVED***
      var F = function(a, b, c)***REMOVED***
        if(this instanceof C)***REMOVED***
          switch(arguments.length)***REMOVED***
            case 0: return new C;
            case 1: return new C(a);
            case 2: return new C(a, b);
          ***REMOVED*** return new C(a, b, c);
        ***REMOVED*** return C.apply(this, arguments);
      ***REMOVED***;
      F[PROTOTYPE] = C[PROTOTYPE];
      return F;
    // make static versions for prototype methods
    ***REMOVED***)(out) : IS_PROTO && typeof out == 'function' ? ctx(Function.call, out) : out;
    // export proto methods to core.%CONSTRUCTOR%.methods.%NAME%
    if(IS_PROTO)***REMOVED***
      (exports.virtual || (exports.virtual = ***REMOVED******REMOVED***))[key] = out;
      // export proto methods to core.%CONSTRUCTOR%.prototype.%NAME%
      if(type & $export.R && expProto && !expProto[key])hide(expProto, key, out);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
// type bitmap
$export.F = 1;   // forced
$export.G = 2;   // global
$export.S = 4;   // static
$export.P = 8;   // proto
$export.B = 16;  // bind
$export.W = 32;  // wrap
$export.U = 64;  // safe
$export.R = 128; // real proto method for `library` 
module.exports = $export;
***REMOVED***,***REMOVED***"./_core":15,"./_ctx":16,"./_global":23,"./_hide":25***REMOVED***],22:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(exec)***REMOVED***
  try ***REMOVED***
    return !!exec();
  ***REMOVED*** catch(e)***REMOVED***
    return true;
  ***REMOVED***
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],23:[function(_dereq_,module,exports)***REMOVED***
// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
var global = module.exports = typeof window != 'undefined' && window.Math == Math
  ? window : typeof self != 'undefined' && self.Math == Math ? self : Function('return this')();
if(typeof __g == 'number')__g = global; // eslint-disable-line no-undef
***REMOVED***,***REMOVED******REMOVED***],24:[function(_dereq_,module,exports)***REMOVED***
var hasOwnProperty = ***REMOVED******REMOVED***.hasOwnProperty;
module.exports = function(it, key)***REMOVED***
  return hasOwnProperty.call(it, key);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],25:[function(_dereq_,module,exports)***REMOVED***
var dP         = _dereq_('./_object-dp')
  , createDesc = _dereq_('./_property-desc');
module.exports = _dereq_('./_descriptors') ? function(object, key, value)***REMOVED***
  return dP.f(object, key, createDesc(1, value));
***REMOVED*** : function(object, key, value)***REMOVED***
  object[key] = value;
  return object;
***REMOVED***;
***REMOVED***,***REMOVED***"./_descriptors":18,"./_object-dp":31,"./_property-desc":37***REMOVED***],26:[function(_dereq_,module,exports)***REMOVED***
module.exports = !_dereq_('./_descriptors') && !_dereq_('./_fails')(function()***REMOVED***
  return Object.defineProperty(_dereq_('./_dom-create')('div'), 'a', ***REMOVED***get: function()***REMOVED*** return 7; ***REMOVED******REMOVED***).a != 7;
***REMOVED***);
***REMOVED***,***REMOVED***"./_descriptors":18,"./_dom-create":19,"./_fails":22***REMOVED***],27:[function(_dereq_,module,exports)***REMOVED***
// fallback for non-array-like ES3 and non-enumerable old V8 strings
var cof = _dereq_('./_cof');
module.exports = Object('z').propertyIsEnumerable(0) ? Object : function(it)***REMOVED***
  return cof(it) == 'String' ? it.split('') : Object(it);
***REMOVED***;
***REMOVED***,***REMOVED***"./_cof":14***REMOVED***],28:[function(_dereq_,module,exports)***REMOVED***
// 7.2.2 IsArray(argument)
var cof = _dereq_('./_cof');
module.exports = Array.isArray || function isArray(arg)***REMOVED***
  return cof(arg) == 'Array';
***REMOVED***;
***REMOVED***,***REMOVED***"./_cof":14***REMOVED***],29:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(it)***REMOVED***
  return typeof it === 'object' ? it !== null : typeof it === 'function';
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],30:[function(_dereq_,module,exports)***REMOVED***
'use strict';
// 19.1.2.1 Object.assign(target, source, ...)
var getKeys  = _dereq_('./_object-keys')
  , gOPS     = _dereq_('./_object-gops')
  , pIE      = _dereq_('./_object-pie')
  , toObject = _dereq_('./_to-object')
  , IObject  = _dereq_('./_iobject')
  , $assign  = Object.assign;

// should work with symbols and should have deterministic property order (V8 bug)
module.exports = !$assign || _dereq_('./_fails')(function()***REMOVED***
  var A = ***REMOVED******REMOVED***
    , B = ***REMOVED******REMOVED***
    , S = Symbol()
    , K = 'abcdefghijklmnopqrst';
  A[S] = 7;
  K.split('').forEach(function(k)***REMOVED*** B[k] = k; ***REMOVED***);
  return $assign(***REMOVED******REMOVED***, A)[S] != 7 || Object.keys($assign(***REMOVED******REMOVED***, B)).join('') != K;
***REMOVED***) ? function assign(target, source)***REMOVED*** // eslint-disable-line no-unused-vars
  var T     = toObject(target)
    , aLen  = arguments.length
    , index = 1
    , getSymbols = gOPS.f
    , isEnum     = pIE.f;
  while(aLen > index)***REMOVED***
    var S      = IObject(arguments[index++])
      , keys   = getSymbols ? getKeys(S).concat(getSymbols(S)) : getKeys(S)
      , length = keys.length
      , j      = 0
      , key;
    while(length > j)if(isEnum.call(S, key = keys[j++]))T[key] = S[key];
  ***REMOVED*** return T;
***REMOVED*** : $assign;
***REMOVED***,***REMOVED***"./_fails":22,"./_iobject":27,"./_object-gops":32,"./_object-keys":34,"./_object-pie":35,"./_to-object":45***REMOVED***],31:[function(_dereq_,module,exports)***REMOVED***
var anObject       = _dereq_('./_an-object')
  , IE8_DOM_DEFINE = _dereq_('./_ie8-dom-define')
  , toPrimitive    = _dereq_('./_to-primitive')
  , dP             = Object.defineProperty;

exports.f = _dereq_('./_descriptors') ? Object.defineProperty : function defineProperty(O, P, Attributes)***REMOVED***
  anObject(O);
  P = toPrimitive(P, true);
  anObject(Attributes);
  if(IE8_DOM_DEFINE)try ***REMOVED***
    return dP(O, P, Attributes);
  ***REMOVED*** catch(e)***REMOVED*** /* empty */ ***REMOVED***
  if('get' in Attributes || 'set' in Attributes)throw TypeError('Accessors not supported!');
  if('value' in Attributes)O[P] = Attributes.value;
  return O;
***REMOVED***;
***REMOVED***,***REMOVED***"./_an-object":11,"./_descriptors":18,"./_ie8-dom-define":26,"./_to-primitive":46***REMOVED***],32:[function(_dereq_,module,exports)***REMOVED***
exports.f = Object.getOwnPropertySymbols;
***REMOVED***,***REMOVED******REMOVED***],33:[function(_dereq_,module,exports)***REMOVED***
var has          = _dereq_('./_has')
  , toIObject    = _dereq_('./_to-iobject')
  , arrayIndexOf = _dereq_('./_array-includes')(false)
  , IE_PROTO     = _dereq_('./_shared-key')('IE_PROTO');

module.exports = function(object, names)***REMOVED***
  var O      = toIObject(object)
    , i      = 0
    , result = []
    , key;
  for(key in O)if(key != IE_PROTO)has(O, key) && result.push(key);
  // Don't enum bug & hidden keys
  while(names.length > i)if(has(O, key = names[i++]))***REMOVED***
    ~arrayIndexOf(result, key) || result.push(key);
  ***REMOVED***
  return result;
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-includes":12,"./_has":24,"./_shared-key":38,"./_to-iobject":43***REMOVED***],34:[function(_dereq_,module,exports)***REMOVED***
// 19.1.2.14 / 15.2.3.14 Object.keys(O)
var $keys       = _dereq_('./_object-keys-internal')
  , enumBugKeys = _dereq_('./_enum-bug-keys');

module.exports = Object.keys || function keys(O)***REMOVED***
  return $keys(O, enumBugKeys);
***REMOVED***;
***REMOVED***,***REMOVED***"./_enum-bug-keys":20,"./_object-keys-internal":33***REMOVED***],35:[function(_dereq_,module,exports)***REMOVED***
exports.f = ***REMOVED******REMOVED***.propertyIsEnumerable;
***REMOVED***,***REMOVED******REMOVED***],36:[function(_dereq_,module,exports)***REMOVED***
// most Object methods by ES6 should accept primitives
var $export = _dereq_('./_export')
  , core    = _dereq_('./_core')
  , fails   = _dereq_('./_fails');
module.exports = function(KEY, exec)***REMOVED***
  var fn  = (core.Object || ***REMOVED******REMOVED***)[KEY] || Object[KEY]
    , exp = ***REMOVED******REMOVED***;
  exp[KEY] = exec(fn);
  $export($export.S + $export.F * fails(function()***REMOVED*** fn(1); ***REMOVED***), 'Object', exp);
***REMOVED***;
***REMOVED***,***REMOVED***"./_core":15,"./_export":21,"./_fails":22***REMOVED***],37:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(bitmap, value)***REMOVED***
  return ***REMOVED***
    enumerable  : !(bitmap & 1),
    configurable: !(bitmap & 2),
    writable    : !(bitmap & 4),
    value       : value
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],38:[function(_dereq_,module,exports)***REMOVED***
var shared = _dereq_('./_shared')('keys')
  , uid    = _dereq_('./_uid');
module.exports = function(key)***REMOVED***
  return shared[key] || (shared[key] = uid(key));
***REMOVED***;
***REMOVED***,***REMOVED***"./_shared":39,"./_uid":47***REMOVED***],39:[function(_dereq_,module,exports)***REMOVED***
var global = _dereq_('./_global')
  , SHARED = '__core-js_shared__'
  , store  = global[SHARED] || (global[SHARED] = ***REMOVED******REMOVED***);
module.exports = function(key)***REMOVED***
  return store[key] || (store[key] = ***REMOVED******REMOVED***);
***REMOVED***;
***REMOVED***,***REMOVED***"./_global":23***REMOVED***],40:[function(_dereq_,module,exports)***REMOVED***
var fails = _dereq_('./_fails');

module.exports = function(method, arg)***REMOVED***
  return !!method && fails(function()***REMOVED***
    arg ? method.call(null, function()***REMOVED******REMOVED***, 1) : method.call(null);
  ***REMOVED***);
***REMOVED***;
***REMOVED***,***REMOVED***"./_fails":22***REMOVED***],41:[function(_dereq_,module,exports)***REMOVED***
var toInteger = _dereq_('./_to-integer')
  , max       = Math.max
  , min       = Math.min;
module.exports = function(index, length)***REMOVED***
  index = toInteger(index);
  return index < 0 ? max(index + length, 0) : min(index, length);
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-integer":42***REMOVED***],42:[function(_dereq_,module,exports)***REMOVED***
// 7.1.4 ToInteger
var ceil  = Math.ceil
  , floor = Math.floor;
module.exports = function(it)***REMOVED***
  return isNaN(it = +it) ? 0 : (it > 0 ? floor : ceil)(it);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],43:[function(_dereq_,module,exports)***REMOVED***
// to indexed object, toObject with fallback for non-array-like ES3 strings
var IObject = _dereq_('./_iobject')
  , defined = _dereq_('./_defined');
module.exports = function(it)***REMOVED***
  return IObject(defined(it));
***REMOVED***;
***REMOVED***,***REMOVED***"./_defined":17,"./_iobject":27***REMOVED***],44:[function(_dereq_,module,exports)***REMOVED***
// 7.1.15 ToLength
var toInteger = _dereq_('./_to-integer')
  , min       = Math.min;
module.exports = function(it)***REMOVED***
  return it > 0 ? min(toInteger(it), 0x1fffffffffffff) : 0; // pow(2, 53) - 1 == 9007199254740991
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-integer":42***REMOVED***],45:[function(_dereq_,module,exports)***REMOVED***
// 7.1.13 ToObject(argument)
var defined = _dereq_('./_defined');
module.exports = function(it)***REMOVED***
  return Object(defined(it));
***REMOVED***;
***REMOVED***,***REMOVED***"./_defined":17***REMOVED***],46:[function(_dereq_,module,exports)***REMOVED***
// 7.1.1 ToPrimitive(input [, PreferredType])
var isObject = _dereq_('./_is-object');
// instead of the ES6 spec version, we didn't implement @@toPrimitive case
// and the second argument - flag - preferred type is a string
module.exports = function(it, S)***REMOVED***
  if(!isObject(it))return it;
  var fn, val;
  if(S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it)))return val;
  if(typeof (fn = it.valueOf) == 'function' && !isObject(val = fn.call(it)))return val;
  if(!S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it)))return val;
  throw TypeError("Can't convert object to primitive value");
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-object":29***REMOVED***],47:[function(_dereq_,module,exports)***REMOVED***
var id = 0
  , px = Math.random();
module.exports = function(key)***REMOVED***
  return 'Symbol('.concat(key === undefined ? '' : key, ')_', (++id + px).toString(36));
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],48:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export       = _dereq_('./_export')
  , $indexOf      = _dereq_('./_array-includes')(false)
  , $native       = [].indexOf
  , NEGATIVE_ZERO = !!$native && 1 / [1].indexOf(1, -0) < 0;

$export($export.P + $export.F * (NEGATIVE_ZERO || !_dereq_('./_strict-method')($native)), 'Array', ***REMOVED***
  // 22.1.3.11 / 15.4.4.14 Array.prototype.indexOf(searchElement [, fromIndex])
  indexOf: function indexOf(searchElement /*, fromIndex = 0 */)***REMOVED***
    return NEGATIVE_ZERO
      // convert -0 to +0
      ? $native.apply(this, arguments) || 0
      : $indexOf(this, searchElement, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-includes":12,"./_export":21,"./_strict-method":40***REMOVED***],49:[function(_dereq_,module,exports)***REMOVED***
// 22.1.2.2 / 15.4.3.2 Array.isArray(arg)
var $export = _dereq_('./_export');

$export($export.S, 'Array', ***REMOVED***isArray: _dereq_('./_is-array')***REMOVED***);
***REMOVED***,***REMOVED***"./_export":21,"./_is-array":28***REMOVED***],50:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export = _dereq_('./_export')
  , $reduce = _dereq_('./_array-reduce');

$export($export.P + $export.F * !_dereq_('./_strict-method')([].reduce, true), 'Array', ***REMOVED***
  // 22.1.3.18 / 15.4.4.21 Array.prototype.reduce(callbackfn [, initialValue])
  reduce: function reduce(callbackfn /* , initialValue */)***REMOVED***
    return $reduce(this, callbackfn, arguments.length, arguments[1], false);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-reduce":13,"./_export":21,"./_strict-method":40***REMOVED***],51:[function(_dereq_,module,exports)***REMOVED***
// 19.1.3.1 Object.assign(target, source)
var $export = _dereq_('./_export');

$export($export.S + $export.F, 'Object', ***REMOVED***assign: _dereq_('./_object-assign')***REMOVED***);
***REMOVED***,***REMOVED***"./_export":21,"./_object-assign":30***REMOVED***],52:[function(_dereq_,module,exports)***REMOVED***
// 19.1.2.14 Object.keys(O)
var toObject = _dereq_('./_to-object')
  , $keys    = _dereq_('./_object-keys');

_dereq_('./_object-sap')('keys', function()***REMOVED***
  return function keys(it)***REMOVED***
    return $keys(toObject(it));
  ***REMOVED***;
***REMOVED***);
***REMOVED***,***REMOVED***"./_object-keys":34,"./_object-sap":36,"./_to-object":45***REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[1])(1)
***REMOVED***);
