// browserify-cipher/browser.js
var ebtk = require('evp_bytestokey')
var aes = require('browserify-aes/browser')
var DES = require('browserify-des')
var desModes = require('browserify-des/modes')
var aesModes = require('browserify-aes/modes')
function createCipher (suite, password) ***REMOVED***
  var keyLen, ivLen
  suite = suite.toLowerCase()
  if (aesModes[suite]) ***REMOVED***
    keyLen = aesModes[suite].key
    ivLen = aesModes[suite].iv
  ***REMOVED*** else if (desModes[suite]) ***REMOVED***
    keyLen = desModes[suite].key * 8
    ivLen = desModes[suite].iv
  ***REMOVED*** else ***REMOVED***
    throw new TypeError('invalid suite type')
  ***REMOVED***
  var keys = ebtk(password, false, keyLen, ivLen)
  return createCipheriv(suite, keys.key, keys.iv)
***REMOVED***
function createDecipher (suite, password) ***REMOVED***
  var keyLen, ivLen
  suite = suite.toLowerCase()
  if (aesModes[suite]) ***REMOVED***
    keyLen = aesModes[suite].key
    ivLen = aesModes[suite].iv
  ***REMOVED*** else if (desModes[suite]) ***REMOVED***
    keyLen = desModes[suite].key * 8
    ivLen = desModes[suite].iv
  ***REMOVED*** else ***REMOVED***
    throw new TypeError('invalid suite type')
  ***REMOVED***
  var keys = ebtk(password, false, keyLen, ivLen)
  return createDecipheriv(suite, keys.key, keys.iv)
***REMOVED***

function createCipheriv (suite, key, iv) ***REMOVED***
  suite = suite.toLowerCase()
  if (aesModes[suite]) ***REMOVED***
    return aes.createCipheriv(suite, key, iv)
  ***REMOVED*** else if (desModes[suite]) ***REMOVED***
    return new DES(***REMOVED***
      key: key,
      iv: iv,
      mode: suite
    ***REMOVED***)
  ***REMOVED*** else ***REMOVED***
    throw new TypeError('invalid suite type')
  ***REMOVED***
***REMOVED***
function createDecipheriv (suite, key, iv) ***REMOVED***
  suite = suite.toLowerCase()
  if (aesModes[suite]) ***REMOVED***
    return aes.createDecipheriv(suite, key, iv)
  ***REMOVED*** else if (desModes[suite]) ***REMOVED***
    return new DES(***REMOVED***
      key: key,
      iv: iv,
      mode: suite,
      decrypt: true
    ***REMOVED***)
  ***REMOVED*** else ***REMOVED***
    throw new TypeError('invalid suite type')
  ***REMOVED***
***REMOVED***
exports.createCipher = exports.Cipher = createCipher
exports.createCipheriv = exports.Cipheriv = createCipheriv
exports.createDecipher = exports.Decipher = createDecipher
exports.createDecipheriv = exports.Decipheriv = createDecipheriv
function getCiphers () ***REMOVED***
  return Object.keys(desModes).concat(aes.getCiphers())
***REMOVED***
exports.listCiphers = exports.getCiphers = getCiphers
