// map-cache/index.js
/*!
 * map-cache <https://github.com/jonschlinkert/map-cache>
 *
 * Copyright (c) 2015, Jon Schlinkert.
 * Licensed under the MIT License.
 */

'use strict';

var hasOwn = Object.prototype.hasOwnProperty;

/**
 * Expose `MapCache`
 */

module.exports = MapCache;

/**
 * Creates a cache object to store key/value pairs.
 *
 * ```js
 * var cache = new MapCache();
 * ```
 *
 * @api public
 */

function MapCache(data) ***REMOVED***
  this.__data__ = data || ***REMOVED******REMOVED***;
***REMOVED***

/**
 * Adds `value` to `key` on the cache.
 *
 * ```js
 * cache.set('foo', 'bar');
 * ```
 *
 * @param ***REMOVED***String***REMOVED*** `key` The key of the value to cache.
 * @param ***REMOVED*******REMOVED*** `value` The value to cache.
 * @returns ***REMOVED***Object***REMOVED*** Returns the `Cache` object for chaining.
 * @api public
 */

MapCache.prototype.set = function mapSet(key, value) ***REMOVED***
  if (key !== '__proto__') ***REMOVED***
    this.__data__[key] = value;
  ***REMOVED***
  return this;
***REMOVED***;

/**
 * Gets the cached value for `key`.
 *
 * ```js
 * cache.get('foo');
 * //=> 'bar'
 * ```
 *
 * @param ***REMOVED***String***REMOVED*** `key` The key of the value to get.
 * @returns ***REMOVED*******REMOVED*** Returns the cached value.
 * @api public
 */

MapCache.prototype.get = function mapGet(key) ***REMOVED***
  return key === '__proto__' ? undefined : this.__data__[key];
***REMOVED***;

/**
 * Checks if a cached value for `key` exists.
 *
 * ```js
 * cache.has('foo');
 * //=> true
 * ```
 *
 * @param ***REMOVED***String***REMOVED*** `key` The key of the entry to check.
 * @returns ***REMOVED***Boolean***REMOVED*** Returns `true` if an entry for `key` exists, else `false`.
 * @api public
 */

MapCache.prototype.has = function mapHas(key) ***REMOVED***
  return key !== '__proto__' && hasOwn.call(this.__data__, key);
***REMOVED***;

/**
 * Removes `key` and its value from the cache.
 *
 * ```js
 * cache.del('foo');
 * ```
 * @title .del
 * @param ***REMOVED***String***REMOVED*** `key` The key of the value to remove.
 * @returns ***REMOVED***Boolean***REMOVED*** Returns `true` if the entry was removed successfully, else `false`.
 * @api public
 */

MapCache.prototype.del = function mapDelete(key) ***REMOVED***
  return this.has(key) && delete this.__data__[key];
***REMOVED***;
