// firebase-admin/lib/utils/validator.js
/*! firebase-admin v5.8.2 */
"use strict";
/*!
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var url = require("url");
/**
 * Validates that a value is an array.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is an array or not.
 */
function isArray(value) ***REMOVED***
    return value instanceof Array;
***REMOVED***
exports.isArray = isArray;
/**
 * Validates that a value is a non-empty array.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is a non-empty array or not.
 */
function isNonEmptyArray(value) ***REMOVED***
    return isArray(value) && value.length !== 0;
***REMOVED***
exports.isNonEmptyArray = isNonEmptyArray;
/**
 * Validates that a value is a boolean.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is a boolean or not.
 */
function isBoolean(value) ***REMOVED***
    return typeof value === 'boolean';
***REMOVED***
exports.isBoolean = isBoolean;
/**
 * Validates that a value is a number.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is a number or not.
 */
function isNumber(value) ***REMOVED***
    return typeof value === 'number' && !isNaN(value);
***REMOVED***
exports.isNumber = isNumber;
/**
 * Validates that a value is a string.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is a string or not.
 */
function isString(value) ***REMOVED***
    return typeof value === 'string';
***REMOVED***
exports.isString = isString;
/**
 * Validates that a value is a non-empty string.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is a non-empty string or not.
 */
function isNonEmptyString(value) ***REMOVED***
    return isString(value) && value !== '';
***REMOVED***
exports.isNonEmptyString = isNonEmptyString;
/**
 * Validates that a value is a nullable object.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is an object or not.
 */
function isObject(value) ***REMOVED***
    return typeof value === 'object' && !(value instanceof Array);
***REMOVED***
exports.isObject = isObject;
/**
 * Validates that a value is a non-null object.
 *
 * @param ***REMOVED***any***REMOVED*** value The value to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the value is a non-null object or not.
 */
function isNonNullObject(value) ***REMOVED***
    return isObject(value) && value !== null;
***REMOVED***
exports.isNonNullObject = isNonNullObject;
/**
 * Validates that a string is a valid Firebase Auth uid.
 *
 * @param ***REMOVED***any***REMOVED*** uid The string to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the string is a valid Firebase Auth uid.
 */
function isUid(uid) ***REMOVED***
    return typeof uid === 'string' && uid.length > 0 && uid.length <= 128;
***REMOVED***
exports.isUid = isUid;
/**
 * Validates that a string is a valid Firebase Auth password.
 *
 * @param ***REMOVED***any***REMOVED*** password The password string to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the string is a valid Firebase Auth password.
 */
function isPassword(password) ***REMOVED***
    // A password must be a string of at least 6 characters.
    return typeof password === 'string' && password.length >= 6;
***REMOVED***
exports.isPassword = isPassword;
/**
 * Validates that a string is a valid email.
 *
 * @param ***REMOVED***any***REMOVED*** email The string to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the string is valid email or not.
 */
function isEmail(email) ***REMOVED***
    if (typeof email !== 'string') ***REMOVED***
        return false;
    ***REMOVED***
    // There must at least one character before the @ symbol and another after.
    var re = /^[^@]+@[^@]+$/;
    return re.test(email);
***REMOVED***
exports.isEmail = isEmail;
/**
 * Validates that a string is a valid phone number.
 *
 * @param ***REMOVED***any***REMOVED*** phoneNumber The string to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the string is a valid phone number or not.
 */
function isPhoneNumber(phoneNumber) ***REMOVED***
    if (typeof phoneNumber !== 'string') ***REMOVED***
        return false;
    ***REMOVED***
    // Phone number validation is very lax here. Backend will enforce E.164
    // spec compliance and will normalize accordingly.
    // The phone number string must be non-empty and starts with a plus sign.
    var re1 = /^\+/;
    // The phone number string must contain at least one alphanumeric character.
    var re2 = /[\da-zA-Z]+/;
    return re1.test(phoneNumber) && re2.test(phoneNumber);
***REMOVED***
exports.isPhoneNumber = isPhoneNumber;
/**
 * Validates that a string is a valid web URL.
 *
 * @param ***REMOVED***any***REMOVED*** urlStr The string to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the string is valid web URL or not.
 */
function isURL(urlStr) ***REMOVED***
    if (typeof urlStr !== 'string') ***REMOVED***
        return false;
    ***REMOVED***
    // Lookup illegal characters.
    var re = /[^a-z0-9\:\/\?\#\[\]\@\!\$\&\'\(\)\*\+\,\;\=\.\-\_\~\%]/i;
    if (re.test(urlStr)) ***REMOVED***
        return false;
    ***REMOVED***
    try ***REMOVED***
        var uri = url.parse(urlStr);
        var scheme = uri.protocol;
        var slashes = uri.slashes;
        var hostname = uri.hostname;
        var pathname = uri.pathname;
        if ((scheme !== 'http:' && scheme !== 'https:') || !slashes) ***REMOVED***
            return false;
        ***REMOVED***
        // Validate hostname: Can contain letters, numbers, underscore and dashes separated by a dot.
        // Each zone must not start with a hyphen or underscore.
        if (!/^[a-zA-Z0-9]+[\w\-]*([\.]?[a-zA-Z0-9]+[\w\-]*)*$/.test(hostname)) ***REMOVED***
            return false;
        ***REMOVED***
        // Allow for pathnames: (/chars+)*
        // Where chars can be a combination of: a-z A-Z 0-9 - _ . ~ ! $ & ' ( ) * + , ; = : @ %
        var pathnameRe = /^(\/[\w\-\.\~\!\$\'\(\)\*\+\,\;\=\:\@\%]+)*$/;
        // Validate pathname.
        if (pathname &&
            pathname !== '/' &&
            !pathnameRe.test(pathname)) ***REMOVED***
            return false;
        ***REMOVED***
        // Allow any query string and hash as long as no invalid character is used.
    ***REMOVED***
    catch (e) ***REMOVED***
        return false;
    ***REMOVED***
    return true;
***REMOVED***
exports.isURL = isURL;
/**
 * Validates that the provided topic is a valid FCM topic name.
 *
 * @param ***REMOVED***any***REMOVED*** topic The topic to validate.
 * @return ***REMOVED***boolean***REMOVED*** Whether the provided topic is a valid FCM topic name.
 */
function isTopic(topic) ***REMOVED***
    if (typeof topic !== 'string') ***REMOVED***
        return false;
    ***REMOVED***
    var VALID_TOPIC_REGEX = /^(\/topics\/)?(private\/)?[a-zA-Z0-9-_.~%]+$/;
    return VALID_TOPIC_REGEX.test(topic);
***REMOVED***
exports.isTopic = isTopic;
