// @firebase/database/dist/esm/src/core/view/ViewCache.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** ChildrenNode ***REMOVED*** from '../snap/ChildrenNode';
import ***REMOVED*** CacheNode ***REMOVED*** from './CacheNode';
/**
 * Stores the data we have cached for a view.
 *
 * serverSnap is the cached server data, eventSnap is the cached event data (server data plus any local writes).
 *
 * @constructor
 */
var ViewCache = /** @class */ (function () ***REMOVED***
    /**
     *
     * @param ***REMOVED***!CacheNode***REMOVED*** eventCache_
     * @param ***REMOVED***!CacheNode***REMOVED*** serverCache_
     */
    function ViewCache(eventCache_, serverCache_) ***REMOVED***
        this.eventCache_ = eventCache_;
        this.serverCache_ = serverCache_;
    ***REMOVED***
    /**
     * @param ***REMOVED***!Node***REMOVED*** eventSnap
     * @param ***REMOVED***boolean***REMOVED*** complete
     * @param ***REMOVED***boolean***REMOVED*** filtered
     * @return ***REMOVED***!ViewCache***REMOVED***
     */
    ViewCache.prototype.updateEventSnap = function (eventSnap, complete, filtered) ***REMOVED***
        return new ViewCache(new CacheNode(eventSnap, complete, filtered), this.serverCache_);
    ***REMOVED***;
    /**
     * @param ***REMOVED***!Node***REMOVED*** serverSnap
     * @param ***REMOVED***boolean***REMOVED*** complete
     * @param ***REMOVED***boolean***REMOVED*** filtered
     * @return ***REMOVED***!ViewCache***REMOVED***
     */
    ViewCache.prototype.updateServerSnap = function (serverSnap, complete, filtered) ***REMOVED***
        return new ViewCache(this.eventCache_, new CacheNode(serverSnap, complete, filtered));
    ***REMOVED***;
    /**
     * @return ***REMOVED***!CacheNode***REMOVED***
     */
    ViewCache.prototype.getEventCache = function () ***REMOVED***
        return this.eventCache_;
    ***REMOVED***;
    /**
     * @return ***REMOVED***?Node***REMOVED***
     */
    ViewCache.prototype.getCompleteEventSnap = function () ***REMOVED***
        return this.eventCache_.isFullyInitialized()
            ? this.eventCache_.getNode()
            : null;
    ***REMOVED***;
    /**
     * @return ***REMOVED***!CacheNode***REMOVED***
     */
    ViewCache.prototype.getServerCache = function () ***REMOVED***
        return this.serverCache_;
    ***REMOVED***;
    /**
     * @return ***REMOVED***?Node***REMOVED***
     */
    ViewCache.prototype.getCompleteServerSnap = function () ***REMOVED***
        return this.serverCache_.isFullyInitialized()
            ? this.serverCache_.getNode()
            : null;
    ***REMOVED***;
    /**
     * @const
     * @type ***REMOVED***ViewCache***REMOVED***
     */
    ViewCache.Empty = new ViewCache(new CacheNode(ChildrenNode.EMPTY_NODE, 
    /*fullyInitialized=*/ false, 
    /*filtered=*/ false), new CacheNode(ChildrenNode.EMPTY_NODE, 
    /*fullyInitialized=*/ false, 
    /*filtered=*/ false));
    return ViewCache;
***REMOVED***());
export ***REMOVED*** ViewCache ***REMOVED***;
