// elliptic/lib/elliptic/curve/base.js
'use strict';

var BN = require('bn.js');
var elliptic = require('../../elliptic');
var utils = elliptic.utils;
var getNAF = utils.getNAF;
var getJSF = utils.getJSF;
var assert = utils.assert;

function BaseCurve(type, conf) ***REMOVED***
  this.type = type;
  this.p = new BN(conf.p, 16);

  // Use Montgomery, when there is no fast reduction for the prime
  this.red = conf.prime ? BN.red(conf.prime) : BN.mont(this.p);

  // Useful for many curves
  this.zero = new BN(0).toRed(this.red);
  this.one = new BN(1).toRed(this.red);
  this.two = new BN(2).toRed(this.red);

  // Curve configuration, optional
  this.n = conf.n && new BN(conf.n, 16);
  this.g = conf.g && this.pointFromJSON(conf.g, conf.gRed);

  // Temporary arrays
  this._wnafT1 = new Array(4);
  this._wnafT2 = new Array(4);
  this._wnafT3 = new Array(4);
  this._wnafT4 = new Array(4);

  // Generalized Greg Maxwell's trick
  var adjustCount = this.n && this.p.div(this.n);
  if (!adjustCount || adjustCount.cmpn(100) > 0) ***REMOVED***
    this.redN = null;
  ***REMOVED*** else ***REMOVED***
    this._maxwellTrick = true;
    this.redN = this.n.toRed(this.red);
  ***REMOVED***
***REMOVED***
module.exports = BaseCurve;

BaseCurve.prototype.point = function point() ***REMOVED***
  throw new Error('Not implemented');
***REMOVED***;

BaseCurve.prototype.validate = function validate() ***REMOVED***
  throw new Error('Not implemented');
***REMOVED***;

BaseCurve.prototype._fixedNafMul = function _fixedNafMul(p, k) ***REMOVED***
  assert(p.precomputed);
  var doubles = p._getDoubles();

  var naf = getNAF(k, 1);
  var I = (1 << (doubles.step + 1)) - (doubles.step % 2 === 0 ? 2 : 1);
  I /= 3;

  // Translate into more windowed form
  var repr = [];
  for (var j = 0; j < naf.length; j += doubles.step) ***REMOVED***
    var nafW = 0;
    for (var k = j + doubles.step - 1; k >= j; k--)
      nafW = (nafW << 1) + naf[k];
    repr.push(nafW);
  ***REMOVED***

  var a = this.jpoint(null, null, null);
  var b = this.jpoint(null, null, null);
  for (var i = I; i > 0; i--) ***REMOVED***
    for (var j = 0; j < repr.length; j++) ***REMOVED***
      var nafW = repr[j];
      if (nafW === i)
        b = b.mixedAdd(doubles.points[j]);
      else if (nafW === -i)
        b = b.mixedAdd(doubles.points[j].neg());
    ***REMOVED***
    a = a.add(b);
  ***REMOVED***
  return a.toP();
***REMOVED***;

BaseCurve.prototype._wnafMul = function _wnafMul(p, k) ***REMOVED***
  var w = 4;

  // Precompute window
  var nafPoints = p._getNAFPoints(w);
  w = nafPoints.wnd;
  var wnd = nafPoints.points;

  // Get NAF form
  var naf = getNAF(k, w);

  // Add `this`*(N+1) for every w-NAF index
  var acc = this.jpoint(null, null, null);
  for (var i = naf.length - 1; i >= 0; i--) ***REMOVED***
    // Count zeroes
    for (var k = 0; i >= 0 && naf[i] === 0; i--)
      k++;
    if (i >= 0)
      k++;
    acc = acc.dblp(k);

    if (i < 0)
      break;
    var z = naf[i];
    assert(z !== 0);
    if (p.type === 'affine') ***REMOVED***
      // J +- P
      if (z > 0)
        acc = acc.mixedAdd(wnd[(z - 1) >> 1]);
      else
        acc = acc.mixedAdd(wnd[(-z - 1) >> 1].neg());
    ***REMOVED*** else ***REMOVED***
      // J +- J
      if (z > 0)
        acc = acc.add(wnd[(z - 1) >> 1]);
      else
        acc = acc.add(wnd[(-z - 1) >> 1].neg());
    ***REMOVED***
  ***REMOVED***
  return p.type === 'affine' ? acc.toP() : acc;
***REMOVED***;

BaseCurve.prototype._wnafMulAdd = function _wnafMulAdd(defW,
                                                       points,
                                                       coeffs,
                                                       len,
                                                       jacobianResult) ***REMOVED***
  var wndWidth = this._wnafT1;
  var wnd = this._wnafT2;
  var naf = this._wnafT3;

  // Fill all arrays
  var max = 0;
  for (var i = 0; i < len; i++) ***REMOVED***
    var p = points[i];
    var nafPoints = p._getNAFPoints(defW);
    wndWidth[i] = nafPoints.wnd;
    wnd[i] = nafPoints.points;
  ***REMOVED***

  // Comb small window NAFs
  for (var i = len - 1; i >= 1; i -= 2) ***REMOVED***
    var a = i - 1;
    var b = i;
    if (wndWidth[a] !== 1 || wndWidth[b] !== 1) ***REMOVED***
      naf[a] = getNAF(coeffs[a], wndWidth[a]);
      naf[b] = getNAF(coeffs[b], wndWidth[b]);
      max = Math.max(naf[a].length, max);
      max = Math.max(naf[b].length, max);
      continue;
    ***REMOVED***

    var comb = [
      points[a], /* 1 */
      null, /* 3 */
      null, /* 5 */
      points[b] /* 7 */
    ];

    // Try to avoid Projective points, if possible
    if (points[a].y.cmp(points[b].y) === 0) ***REMOVED***
      comb[1] = points[a].add(points[b]);
      comb[2] = points[a].toJ().mixedAdd(points[b].neg());
    ***REMOVED*** else if (points[a].y.cmp(points[b].y.redNeg()) === 0) ***REMOVED***
      comb[1] = points[a].toJ().mixedAdd(points[b]);
      comb[2] = points[a].add(points[b].neg());
    ***REMOVED*** else ***REMOVED***
      comb[1] = points[a].toJ().mixedAdd(points[b]);
      comb[2] = points[a].toJ().mixedAdd(points[b].neg());
    ***REMOVED***

    var index = [
      -3, /* -1 -1 */
      -1, /* -1 0 */
      -5, /* -1 1 */
      -7, /* 0 -1 */
      0, /* 0 0 */
      7, /* 0 1 */
      5, /* 1 -1 */
      1, /* 1 0 */
      3  /* 1 1 */
    ];

    var jsf = getJSF(coeffs[a], coeffs[b]);
    max = Math.max(jsf[0].length, max);
    naf[a] = new Array(max);
    naf[b] = new Array(max);
    for (var j = 0; j < max; j++) ***REMOVED***
      var ja = jsf[0][j] | 0;
      var jb = jsf[1][j] | 0;

      naf[a][j] = index[(ja + 1) * 3 + (jb + 1)];
      naf[b][j] = 0;
      wnd[a] = comb;
    ***REMOVED***
  ***REMOVED***

  var acc = this.jpoint(null, null, null);
  var tmp = this._wnafT4;
  for (var i = max; i >= 0; i--) ***REMOVED***
    var k = 0;

    while (i >= 0) ***REMOVED***
      var zero = true;
      for (var j = 0; j < len; j++) ***REMOVED***
        tmp[j] = naf[j][i] | 0;
        if (tmp[j] !== 0)
          zero = false;
      ***REMOVED***
      if (!zero)
        break;
      k++;
      i--;
    ***REMOVED***
    if (i >= 0)
      k++;
    acc = acc.dblp(k);
    if (i < 0)
      break;

    for (var j = 0; j < len; j++) ***REMOVED***
      var z = tmp[j];
      var p;
      if (z === 0)
        continue;
      else if (z > 0)
        p = wnd[j][(z - 1) >> 1];
      else if (z < 0)
        p = wnd[j][(-z - 1) >> 1].neg();

      if (p.type === 'affine')
        acc = acc.mixedAdd(p);
      else
        acc = acc.add(p);
    ***REMOVED***
  ***REMOVED***
  // Zeroify references
  for (var i = 0; i < len; i++)
    wnd[i] = null;

  if (jacobianResult)
    return acc;
  else
    return acc.toP();
***REMOVED***;

function BasePoint(curve, type) ***REMOVED***
  this.curve = curve;
  this.type = type;
  this.precomputed = null;
***REMOVED***
BaseCurve.BasePoint = BasePoint;

BasePoint.prototype.eq = function eq(/*other*/) ***REMOVED***
  throw new Error('Not implemented');
***REMOVED***;

BasePoint.prototype.validate = function validate() ***REMOVED***
  return this.curve.validate(this);
***REMOVED***;

BaseCurve.prototype.decodePoint = function decodePoint(bytes, enc) ***REMOVED***
  bytes = utils.toArray(bytes, enc);

  var len = this.p.byteLength();

  // uncompressed, hybrid-odd, hybrid-even
  if ((bytes[0] === 0x04 || bytes[0] === 0x06 || bytes[0] === 0x07) &&
      bytes.length - 1 === 2 * len) ***REMOVED***
    if (bytes[0] === 0x06)
      assert(bytes[bytes.length - 1] % 2 === 0);
    else if (bytes[0] === 0x07)
      assert(bytes[bytes.length - 1] % 2 === 1);

    var res =  this.point(bytes.slice(1, 1 + len),
                          bytes.slice(1 + len, 1 + 2 * len));

    return res;
  ***REMOVED*** else if ((bytes[0] === 0x02 || bytes[0] === 0x03) &&
              bytes.length - 1 === len) ***REMOVED***
    return this.pointFromX(bytes.slice(1, 1 + len), bytes[0] === 0x03);
  ***REMOVED***
  throw new Error('Unknown point format');
***REMOVED***;

BasePoint.prototype.encodeCompressed = function encodeCompressed(enc) ***REMOVED***
  return this.encode(enc, true);
***REMOVED***;

BasePoint.prototype._encode = function _encode(compact) ***REMOVED***
  var len = this.curve.p.byteLength();
  var x = this.getX().toArray('be', len);

  if (compact)
    return [ this.getY().isEven() ? 0x02 : 0x03 ].concat(x);

  return [ 0x04 ].concat(x, this.getY().toArray('be', len)) ;
***REMOVED***;

BasePoint.prototype.encode = function encode(enc, compact) ***REMOVED***
  return utils.encode(this._encode(compact), enc);
***REMOVED***;

BasePoint.prototype.precompute = function precompute(power) ***REMOVED***
  if (this.precomputed)
    return this;

  var precomputed = ***REMOVED***
    doubles: null,
    naf: null,
    beta: null
  ***REMOVED***;
  precomputed.naf = this._getNAFPoints(8);
  precomputed.doubles = this._getDoubles(4, power);
  precomputed.beta = this._getBeta();
  this.precomputed = precomputed;

  return this;
***REMOVED***;

BasePoint.prototype._hasDoubles = function _hasDoubles(k) ***REMOVED***
  if (!this.precomputed)
    return false;

  var doubles = this.precomputed.doubles;
  if (!doubles)
    return false;

  return doubles.points.length >= Math.ceil((k.bitLength() + 1) / doubles.step);
***REMOVED***;

BasePoint.prototype._getDoubles = function _getDoubles(step, power) ***REMOVED***
  if (this.precomputed && this.precomputed.doubles)
    return this.precomputed.doubles;

  var doubles = [ this ];
  var acc = this;
  for (var i = 0; i < power; i += step) ***REMOVED***
    for (var j = 0; j < step; j++)
      acc = acc.dbl();
    doubles.push(acc);
  ***REMOVED***
  return ***REMOVED***
    step: step,
    points: doubles
  ***REMOVED***;
***REMOVED***;

BasePoint.prototype._getNAFPoints = function _getNAFPoints(wnd) ***REMOVED***
  if (this.precomputed && this.precomputed.naf)
    return this.precomputed.naf;

  var res = [ this ];
  var max = (1 << wnd) - 1;
  var dbl = max === 1 ? null : this.dbl();
  for (var i = 1; i < max; i++)
    res[i] = res[i - 1].add(dbl);
  return ***REMOVED***
    wnd: wnd,
    points: res
  ***REMOVED***;
***REMOVED***;

BasePoint.prototype._getBeta = function _getBeta() ***REMOVED***
  return null;
***REMOVED***;

BasePoint.prototype.dblp = function dblp(k) ***REMOVED***
  var r = this;
  for (var i = 0; i < k; i++)
    r = r.dbl();
  return r;
***REMOVED***;
