// elliptic/lib/elliptic/curve/mont.js
'use strict';

var curve = require('../curve');
var BN = require('bn.js');
var inherits = require('inherits');
var Base = curve.base;

var elliptic = require('../../elliptic');
var utils = elliptic.utils;

function MontCurve(conf) ***REMOVED***
  Base.call(this, 'mont', conf);

  this.a = new BN(conf.a, 16).toRed(this.red);
  this.b = new BN(conf.b, 16).toRed(this.red);
  this.i4 = new BN(4).toRed(this.red).redInvm();
  this.two = new BN(2).toRed(this.red);
  this.a24 = this.i4.redMul(this.a.redAdd(this.two));
***REMOVED***
inherits(MontCurve, Base);
module.exports = MontCurve;

MontCurve.prototype.validate = function validate(point) ***REMOVED***
  var x = point.normalize().x;
  var x2 = x.redSqr();
  var rhs = x2.redMul(x).redAdd(x2.redMul(this.a)).redAdd(x);
  var y = rhs.redSqrt();

  return y.redSqr().cmp(rhs) === 0;
***REMOVED***;

function Point(curve, x, z) ***REMOVED***
  Base.BasePoint.call(this, curve, 'projective');
  if (x === null && z === null) ***REMOVED***
    this.x = this.curve.one;
    this.z = this.curve.zero;
  ***REMOVED*** else ***REMOVED***
    this.x = new BN(x, 16);
    this.z = new BN(z, 16);
    if (!this.x.red)
      this.x = this.x.toRed(this.curve.red);
    if (!this.z.red)
      this.z = this.z.toRed(this.curve.red);
  ***REMOVED***
***REMOVED***
inherits(Point, Base.BasePoint);

MontCurve.prototype.decodePoint = function decodePoint(bytes, enc) ***REMOVED***
  return this.point(utils.toArray(bytes, enc), 1);
***REMOVED***;

MontCurve.prototype.point = function point(x, z) ***REMOVED***
  return new Point(this, x, z);
***REMOVED***;

MontCurve.prototype.pointFromJSON = function pointFromJSON(obj) ***REMOVED***
  return Point.fromJSON(this, obj);
***REMOVED***;

Point.prototype.precompute = function precompute() ***REMOVED***
  // No-op
***REMOVED***;

Point.prototype._encode = function _encode() ***REMOVED***
  return this.getX().toArray('be', this.curve.p.byteLength());
***REMOVED***;

Point.fromJSON = function fromJSON(curve, obj) ***REMOVED***
  return new Point(curve, obj[0], obj[1] || curve.one);
***REMOVED***;

Point.prototype.inspect = function inspect() ***REMOVED***
  if (this.isInfinity())
    return '<EC Point Infinity>';
  return '<EC Point x: ' + this.x.fromRed().toString(16, 2) +
      ' z: ' + this.z.fromRed().toString(16, 2) + '>';
***REMOVED***;

Point.prototype.isInfinity = function isInfinity() ***REMOVED***
  // XXX This code assumes that zero is always zero in red
  return this.z.cmpn(0) === 0;
***REMOVED***;

Point.prototype.dbl = function dbl() ***REMOVED***
  // http://hyperelliptic.org/EFD/g1p/auto-montgom-xz.html#doubling-dbl-1987-m-3
  // 2M + 2S + 4A

  // A = X1 + Z1
  var a = this.x.redAdd(this.z);
  // AA = A^2
  var aa = a.redSqr();
  // B = X1 - Z1
  var b = this.x.redSub(this.z);
  // BB = B^2
  var bb = b.redSqr();
  // C = AA - BB
  var c = aa.redSub(bb);
  // X3 = AA * BB
  var nx = aa.redMul(bb);
  // Z3 = C * (BB + A24 * C)
  var nz = c.redMul(bb.redAdd(this.curve.a24.redMul(c)));
  return this.curve.point(nx, nz);
***REMOVED***;

Point.prototype.add = function add() ***REMOVED***
  throw new Error('Not supported on Montgomery curve');
***REMOVED***;

Point.prototype.diffAdd = function diffAdd(p, diff) ***REMOVED***
  // http://hyperelliptic.org/EFD/g1p/auto-montgom-xz.html#diffadd-dadd-1987-m-3
  // 4M + 2S + 6A

  // A = X2 + Z2
  var a = this.x.redAdd(this.z);
  // B = X2 - Z2
  var b = this.x.redSub(this.z);
  // C = X3 + Z3
  var c = p.x.redAdd(p.z);
  // D = X3 - Z3
  var d = p.x.redSub(p.z);
  // DA = D * A
  var da = d.redMul(a);
  // CB = C * B
  var cb = c.redMul(b);
  // X5 = Z1 * (DA + CB)^2
  var nx = diff.z.redMul(da.redAdd(cb).redSqr());
  // Z5 = X1 * (DA - CB)^2
  var nz = diff.x.redMul(da.redISub(cb).redSqr());
  return this.curve.point(nx, nz);
***REMOVED***;

Point.prototype.mul = function mul(k) ***REMOVED***
  var t = k.clone();
  var a = this; // (N / 2) * Q + Q
  var b = this.curve.point(null, null); // (N / 2) * Q
  var c = this; // Q

  for (var bits = []; t.cmpn(0) !== 0; t.iushrn(1))
    bits.push(t.andln(1));

  for (var i = bits.length - 1; i >= 0; i--) ***REMOVED***
    if (bits[i] === 0) ***REMOVED***
      // N * Q + Q = ((N / 2) * Q + Q)) + (N / 2) * Q
      a = a.diffAdd(b, c);
      // N * Q = 2 * ((N / 2) * Q + Q))
      b = b.dbl();
    ***REMOVED*** else ***REMOVED***
      // N * Q = ((N / 2) * Q + Q) + ((N / 2) * Q)
      b = a.diffAdd(b, c);
      // N * Q + Q = 2 * ((N / 2) * Q + Q)
      a = a.dbl();
    ***REMOVED***
  ***REMOVED***
  return b;
***REMOVED***;

Point.prototype.mulAdd = function mulAdd() ***REMOVED***
  throw new Error('Not supported on Montgomery curve');
***REMOVED***;

Point.prototype.jumlAdd = function jumlAdd() ***REMOVED***
  throw new Error('Not supported on Montgomery curve');
***REMOVED***;

Point.prototype.eq = function eq(other) ***REMOVED***
  return this.getX().cmp(other.getX()) === 0;
***REMOVED***;

Point.prototype.normalize = function normalize() ***REMOVED***
  this.x = this.x.redMul(this.z.redInvm());
  this.z = this.curve.one;
  return this;
***REMOVED***;

Point.prototype.getX = function getX() ***REMOVED***
  // Normalize coordinates
  this.normalize();

  return this.x.fromRed();
***REMOVED***;
