// os-browserify/browser.js
exports.endianness = function () ***REMOVED*** return 'LE' ***REMOVED***;

exports.hostname = function () ***REMOVED***
    if (typeof location !== 'undefined') ***REMOVED***
        return location.hostname
    ***REMOVED***
    else return '';
***REMOVED***;

exports.loadavg = function () ***REMOVED*** return [] ***REMOVED***;

exports.uptime = function () ***REMOVED*** return 0 ***REMOVED***;

exports.freemem = function () ***REMOVED***
    return Number.MAX_VALUE;
***REMOVED***;

exports.totalmem = function () ***REMOVED***
    return Number.MAX_VALUE;
***REMOVED***;

exports.cpus = function () ***REMOVED*** return [] ***REMOVED***;

exports.type = function () ***REMOVED*** return 'Browser' ***REMOVED***;

exports.release = function () ***REMOVED***
    if (typeof navigator !== 'undefined') ***REMOVED***
        return navigator.appVersion;
    ***REMOVED***
    return '';
***REMOVED***;

exports.networkInterfaces
= exports.getNetworkInterfaces
= function () ***REMOVED*** return ***REMOVED******REMOVED*** ***REMOVED***;

exports.arch = function () ***REMOVED*** return 'javascript' ***REMOVED***;

exports.platform = function () ***REMOVED*** return 'browser' ***REMOVED***;

exports.tmpdir = exports.tmpDir = function () ***REMOVED***
    return '/tmp';
***REMOVED***;

exports.EOL = '\n';

exports.homedir = function () ***REMOVED***
	return '/'
***REMOVED***;
