// empower-core/lib/decorate.js
'use strict';

var some = require('core-js/library/fn/array/some');
var map = require('core-js/library/fn/array/map');

function decorate (callSpec, decorator) ***REMOVED***
    var numArgsToCapture = callSpec.numArgsToCapture;

    return function decoratedAssert () ***REMOVED***
        var context, message, hasMessage = false;

        // see: https://github.com/twada/empower-core/pull/8#issue-127859465
        // see: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#32-leaking-arguments
        var args = new Array(arguments.length);
        for(var i = 0; i < args.length; ++i) ***REMOVED***
            args[i] = arguments[i];
        ***REMOVED***

        if (numArgsToCapture === (args.length - 1)) ***REMOVED***
            message = args.pop();
            hasMessage = true;
        ***REMOVED***

        var invocation = ***REMOVED***
            thisObj: this,
            values: args,
            message: message,
            hasMessage: hasMessage
        ***REMOVED***;

        if (some(args, isCaptured)) ***REMOVED***
            invocation.values = map(args.slice(0, numArgsToCapture), function (arg) ***REMOVED***
                if (isNotCaptured(arg)) ***REMOVED***
                    return arg;
                ***REMOVED***
                if (!context) ***REMOVED***
                    context = ***REMOVED***
                        source: arg.source,
                        args: []
                    ***REMOVED***;
                ***REMOVED***
                context.args.push(***REMOVED***
                    value: arg.powerAssertContext.value,
                    events: arg.powerAssertContext.events
                ***REMOVED***);
                return arg.powerAssertContext.value;
            ***REMOVED***);

            return decorator.concreteAssert(callSpec, invocation, context);
        ***REMOVED*** else ***REMOVED***
            return decorator.concreteAssert(callSpec, invocation);
        ***REMOVED***
    ***REMOVED***;
***REMOVED***

function isNotCaptured (value) ***REMOVED***
    return !isCaptured(value);
***REMOVED***

function isCaptured (value) ***REMOVED***
    return (typeof value === 'object') &&
        (value !== null) &&
        (typeof value.powerAssertContext !== 'undefined');
***REMOVED***

module.exports = decorate;
