// empower-core/lib/default-options.js
'use strict';

module.exports = function defaultOptions () ***REMOVED***
    return ***REMOVED***
        destructive: false,
        bindReceiver: true,
        onError: onError,
        onSuccess: onSuccess,
        patterns: [
            'assert(value, [message])',
            'assert.ok(value, [message])',
            'assert.equal(actual, expected, [message])',
            'assert.notEqual(actual, expected, [message])',
            'assert.strictEqual(actual, expected, [message])',
            'assert.notStrictEqual(actual, expected, [message])',
            'assert.deepEqual(actual, expected, [message])',
            'assert.notDeepEqual(actual, expected, [message])',
            'assert.deepStrictEqual(actual, expected, [message])',
            'assert.notDeepStrictEqual(actual, expected, [message])'
        ],
        wrapOnlyPatterns: []
    ***REMOVED***;
***REMOVED***;

function onError (errorEvent) ***REMOVED***
    var e = errorEvent.error;
    if (errorEvent.powerAssertContext && /^AssertionError/.test(e.name)) ***REMOVED***
        e.powerAssertContext = errorEvent.powerAssertContext;
    ***REMOVED***
    throw e;
***REMOVED***

function onSuccess(successEvent) ***REMOVED***
    return successEvent.returnValue;
***REMOVED***
