// @firebase/database/dist/esm/src/core/operation/Operation.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** assert ***REMOVED*** from '@firebase/util';
/**
 *
 * @enum
 */
export var OperationType;
(function (OperationType) ***REMOVED***
    OperationType[OperationType["OVERWRITE"] = 0] = "OVERWRITE";
    OperationType[OperationType["MERGE"] = 1] = "MERGE";
    OperationType[OperationType["ACK_USER_WRITE"] = 2] = "ACK_USER_WRITE";
    OperationType[OperationType["LISTEN_COMPLETE"] = 3] = "LISTEN_COMPLETE";
***REMOVED***)(OperationType || (OperationType = ***REMOVED******REMOVED***));
/**
 * @param ***REMOVED***boolean***REMOVED*** fromUser
 * @param ***REMOVED***boolean***REMOVED*** fromServer
 * @param ***REMOVED***?string***REMOVED*** queryId
 * @param ***REMOVED***boolean***REMOVED*** tagged
 * @constructor
 */
var OperationSource = /** @class */ (function () ***REMOVED***
    function OperationSource(fromUser, fromServer, queryId, tagged) ***REMOVED***
        this.fromUser = fromUser;
        this.fromServer = fromServer;
        this.queryId = queryId;
        this.tagged = tagged;
        assert(!tagged || fromServer, 'Tagged queries must be from server.');
    ***REMOVED***
    /**
     * @const
     * @type ***REMOVED***!OperationSource***REMOVED***
     */
    OperationSource.User = new OperationSource(
    /*fromUser=*/ true, false, null, 
    /*tagged=*/ false);
    /**
     * @const
     * @type ***REMOVED***!OperationSource***REMOVED***
     */
    OperationSource.Server = new OperationSource(false, 
    /*fromServer=*/ true, null, 
    /*tagged=*/ false);
    /**
     * @param ***REMOVED***string***REMOVED*** queryId
     * @return ***REMOVED***!OperationSource***REMOVED***
     */
    OperationSource.forServerTaggedQuery = function (queryId) ***REMOVED***
        return new OperationSource(false, 
        /*fromServer=*/ true, queryId, 
        /*tagged=*/ true);
    ***REMOVED***;
    return OperationSource;
***REMOVED***());
export ***REMOVED*** OperationSource ***REMOVED***;
