// @firebase/database/dist/esm/src/core/operation/Overwrite.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** OperationType ***REMOVED*** from './Operation';
import ***REMOVED*** Path ***REMOVED*** from '../util/Path';
/**
 * @param ***REMOVED***!OperationSource***REMOVED*** source
 * @param ***REMOVED***!Path***REMOVED*** path
 * @param ***REMOVED***!Node***REMOVED*** snap
 * @constructor
 * @implements ***REMOVED***Operation***REMOVED***
 */
var Overwrite = /** @class */ (function () ***REMOVED***
    function Overwrite(source, path, snap) ***REMOVED***
        this.source = source;
        this.path = path;
        this.snap = snap;
        /** @inheritDoc */
        this.type = OperationType.OVERWRITE;
    ***REMOVED***
    Overwrite.prototype.operationForChild = function (childName) ***REMOVED***
        if (this.path.isEmpty()) ***REMOVED***
            return new Overwrite(this.source, Path.Empty, this.snap.getImmediateChild(childName));
        ***REMOVED***
        else ***REMOVED***
            return new Overwrite(this.source, this.path.popFront(), this.snap);
        ***REMOVED***
    ***REMOVED***;
    return Overwrite;
***REMOVED***());
export ***REMOVED*** Overwrite ***REMOVED***;
