// hawk/lib/server.js
// Load modules

var Boom = require('boom');
var Hoek = require('hoek');
var Cryptiles = require('cryptiles');
var Crypto = require('./crypto');
var Utils = require('./utils');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Hawk authentication

/*
   req:                 node's HTTP request object or an object as follows:

                        var request = ***REMOVED***
                            method: 'GET',
                            url: '/resource/4?a=1&b=2',
                            host: 'example.com',
                            port: 8080,
                            authorization: 'Hawk id="dh37fgj492je", ts="1353832234", nonce="j4h3g2", ext="some-app-ext-data", mac="6R4rV5iE+NPoym+WwjeHzjAGXUtLNIxmo1vpMofpLAE="'
                        ***REMOVED***;

   credentialsFunc:     required function to lookup the set of Hawk credentials based on the provided credentials id.
                        The credentials include the MAC key, MAC algorithm, and other attributes (such as username)
                        needed by the application. This function is the equivalent of verifying the username and
                        password in Basic authentication.

                        var credentialsFunc = function (id, callback) ***REMOVED***

                            // Lookup credentials in database
                            db.lookup(id, function (err, item) ***REMOVED***

                                if (err || !item) ***REMOVED***
                                    return callback(err);
                                ***REMOVED***

                                var credentials = ***REMOVED***
                                    // Required
                                    key: item.key,
                                    algorithm: item.algorithm,
                                    // Application specific
                                    user: item.user
                                ***REMOVED***;

                                return callback(null, credentials);
                            ***REMOVED***);
                        ***REMOVED***;

   options: ***REMOVED***

        hostHeaderName:        optional header field name, used to override the default 'Host' header when used
                               behind a cache of a proxy. Apache2 changes the value of the 'Host' header while preserving
                               the original (which is what the module must verify) in the 'x-forwarded-host' header field.
                               Only used when passed a node Http.ServerRequest object.

        nonceFunc:             optional nonce validation function. The function signature is function(key, nonce, ts, callback)
                               where 'callback' must be called using the signature function(err).

        timestampSkewSec:      optional number of seconds of permitted clock skew for incoming timestamps. Defaults to 60 seconds.
                               Provides a +/- skew which means actual allowed window is double the number of seconds.

        localtimeOffsetMsec:   optional local clock time offset express in a number of milliseconds (positive or negative).
                               Defaults to 0.

        payload:               optional payload for validation. The client calculates the hash value and includes it via the 'hash'
                               header attribute. The server always ensures the value provided has been included in the request
                               MAC. When this option is provided, it validates the hash value itself. Validation is done by calculating
                               a hash value over the entire payload (assuming it has already be normalized to the same format and
                               encoding used by the client to calculate the hash on request). If the payload is not available at the time
                               of authentication, the authenticatePayload() method can be used by passing it the credentials and
                               attributes.hash returned in the authenticate callback.

        host:                  optional host name override. Only used when passed a node request object.
        port:                  optional port override. Only used when passed a node request object.
    ***REMOVED***

    callback: function (err, credentials, artifacts) ***REMOVED*** ***REMOVED***
 */

exports.authenticate = function (req, credentialsFunc, options, callback) ***REMOVED***

    callback = Hoek.nextTick(callback);

    // Default options

    options.nonceFunc = options.nonceFunc || internals.nonceFunc;
    options.timestampSkewSec = options.timestampSkewSec || 60;                                                  // 60 seconds

    // Application time

    var now = Utils.now(options.localtimeOffsetMsec);                           // Measure now before any other processing

    // Convert node Http request object to a request configuration object

    var request = Utils.parseRequest(req, options);
    if (request instanceof Error) ***REMOVED***
        return callback(Boom.badRequest(request.message));
    ***REMOVED***

    // Parse HTTP Authorization header

    var attributes = Utils.parseAuthorizationHeader(request.authorization);
    if (attributes instanceof Error) ***REMOVED***
        return callback(attributes);
    ***REMOVED***

    // Construct artifacts container

    var artifacts = ***REMOVED***
        method: request.method,
        host: request.host,
        port: request.port,
        resource: request.url,
        ts: attributes.ts,
        nonce: attributes.nonce,
        hash: attributes.hash,
        ext: attributes.ext,
        app: attributes.app,
        dlg: attributes.dlg,
        mac: attributes.mac,
        id: attributes.id
    ***REMOVED***;

    // Verify required header attributes

    if (!attributes.id ||
        !attributes.ts ||
        !attributes.nonce ||
        !attributes.mac) ***REMOVED***

        return callback(Boom.badRequest('Missing attributes'), null, artifacts);
    ***REMOVED***

    // Fetch Hawk credentials

    credentialsFunc(attributes.id, function (err, credentials) ***REMOVED***

        if (err) ***REMOVED***
            return callback(err, credentials || null, artifacts);
        ***REMOVED***

        if (!credentials) ***REMOVED***
            return callback(Boom.unauthorized('Unknown credentials', 'Hawk'), null, artifacts);
        ***REMOVED***

        if (!credentials.key ||
            !credentials.algorithm) ***REMOVED***

            return callback(Boom.internal('Invalid credentials'), credentials, artifacts);
        ***REMOVED***

        if (Crypto.algorithms.indexOf(credentials.algorithm) === -1) ***REMOVED***
            return callback(Boom.internal('Unknown algorithm'), credentials, artifacts);
        ***REMOVED***

        // Calculate MAC

        var mac = Crypto.calculateMac('header', credentials, artifacts);
        if (!Cryptiles.fixedTimeComparison(mac, attributes.mac)) ***REMOVED***
            return callback(Boom.unauthorized('Bad mac', 'Hawk'), credentials, artifacts);
        ***REMOVED***

        // Check payload hash

        if (options.payload ||
            options.payload === '') ***REMOVED***

            if (!attributes.hash) ***REMOVED***
                return callback(Boom.unauthorized('Missing required payload hash', 'Hawk'), credentials, artifacts);
            ***REMOVED***

            var hash = Crypto.calculatePayloadHash(options.payload, credentials.algorithm, request.contentType);
            if (!Cryptiles.fixedTimeComparison(hash, attributes.hash)) ***REMOVED***
                return callback(Boom.unauthorized('Bad payload hash', 'Hawk'), credentials, artifacts);
            ***REMOVED***
        ***REMOVED***

        // Check nonce

        options.nonceFunc(credentials.key, attributes.nonce, attributes.ts, function (err) ***REMOVED***

            if (err) ***REMOVED***
                return callback(Boom.unauthorized('Invalid nonce', 'Hawk'), credentials, artifacts);
            ***REMOVED***

            // Check timestamp staleness

            if (Math.abs((attributes.ts * 1000) - now) > (options.timestampSkewSec * 1000)) ***REMOVED***
                var tsm = Crypto.timestampMessage(credentials, options.localtimeOffsetMsec);
                return callback(Boom.unauthorized('Stale timestamp', 'Hawk', tsm), credentials, artifacts);
            ***REMOVED***

            // Successful authentication

            return callback(null, credentials, artifacts);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***;


// Authenticate payload hash - used when payload cannot be provided during authenticate()

/*
    payload:        raw request payload
    credentials:    from authenticate callback
    artifacts:      from authenticate callback
    contentType:    req.headers['content-type']
*/

exports.authenticatePayload = function (payload, credentials, artifacts, contentType) ***REMOVED***

    var calculatedHash = Crypto.calculatePayloadHash(payload, credentials.algorithm, contentType);
    return Cryptiles.fixedTimeComparison(calculatedHash, artifacts.hash);
***REMOVED***;


// Authenticate payload hash - used when payload cannot be provided during authenticate()

/*
    calculatedHash: the payload hash calculated using Crypto.calculatePayloadHash()
    artifacts:      from authenticate callback
*/

exports.authenticatePayloadHash = function (calculatedHash, artifacts) ***REMOVED***

    return Cryptiles.fixedTimeComparison(calculatedHash, artifacts.hash);
***REMOVED***;


// Generate a Server-Authorization header for a given response

/*
    credentials: ***REMOVED******REMOVED***,                                        // Object received from authenticate()
    artifacts: ***REMOVED******REMOVED***                                           // Object received from authenticate(); 'mac', 'hash', and 'ext' - ignored
    options: ***REMOVED***
        ext: 'application-specific',                        // Application specific data sent via the ext attribute
        payload: '***REMOVED***"some":"payload"***REMOVED***',                      // UTF-8 encoded string for body hash generation (ignored if hash provided)
        contentType: 'application/json',                    // Payload content-type (ignored if hash provided)
        hash: 'U4MKKSmiVxk37JCCrAVIjV='                     // Pre-calculated payload hash
    ***REMOVED***
*/

exports.header = function (credentials, artifacts, options) ***REMOVED***

    // Prepare inputs

    options = options || ***REMOVED******REMOVED***;

    if (!artifacts ||
        typeof artifacts !== 'object' ||
        typeof options !== 'object') ***REMOVED***

        return '';
    ***REMOVED***

    artifacts = Hoek.clone(artifacts);
    delete artifacts.mac;
    artifacts.hash = options.hash;
    artifacts.ext = options.ext;

    // Validate credentials

    if (!credentials ||
        !credentials.key ||
        !credentials.algorithm) ***REMOVED***

        // Invalid credential object
        return '';
    ***REMOVED***

    if (Crypto.algorithms.indexOf(credentials.algorithm) === -1) ***REMOVED***
        return '';
    ***REMOVED***

    // Calculate payload hash

    if (!artifacts.hash &&
        (options.payload || options.payload === '')) ***REMOVED***

        artifacts.hash = Crypto.calculatePayloadHash(options.payload, credentials.algorithm, options.contentType);
    ***REMOVED***

    var mac = Crypto.calculateMac('response', credentials, artifacts);

    // Construct header

    var header = 'Hawk mac="' + mac + '"' +
                 (artifacts.hash ? ', hash="' + artifacts.hash + '"' : '');

    if (artifacts.ext !== null &&
        artifacts.ext !== undefined &&
        artifacts.ext !== '') ***REMOVED***                       // Other falsey values allowed

        header += ', ext="' + Hoek.escapeHeaderAttribute(artifacts.ext) + '"';
    ***REMOVED***

    return header;
***REMOVED***;


/*
 * Arguments and options are the same as authenticate() with the exception that the only supported options are:
 * 'hostHeaderName', 'localtimeOffsetMsec', 'host', 'port'
 */


//                       1     2             3           4
internals.bewitRegex = /^(\/.*)([\?&])bewit\=([^&$]*)(?:&(.+))?$/;


exports.authenticateBewit = function (req, credentialsFunc, options, callback) ***REMOVED***

    callback = Hoek.nextTick(callback);

    // Application time

    var now = Utils.now(options.localtimeOffsetMsec);

    // Convert node Http request object to a request configuration object

    var request = Utils.parseRequest(req, options);
    if (request instanceof Error) ***REMOVED***
        return callback(Boom.badRequest(request.message));
    ***REMOVED***

    // Extract bewit

    if (request.url.length > Utils.limits.maxMatchLength) ***REMOVED***
        return callback(Boom.badRequest('Resource path exceeds max length'));
    ***REMOVED***

    var resource = request.url.match(internals.bewitRegex);
    if (!resource) ***REMOVED***
        return callback(Boom.unauthorized(null, 'Hawk'));
    ***REMOVED***

    // Bewit not empty

    if (!resource[3]) ***REMOVED***
        return callback(Boom.unauthorized('Empty bewit', 'Hawk'));
    ***REMOVED***

    // Verify method is GET

    if (request.method !== 'GET' &&
        request.method !== 'HEAD') ***REMOVED***

        return callback(Boom.unauthorized('Invalid method', 'Hawk'));
    ***REMOVED***

    // No other authentication

    if (request.authorization) ***REMOVED***
        return callback(Boom.badRequest('Multiple authentications'));
    ***REMOVED***

    // Parse bewit

    var bewitString = Hoek.base64urlDecode(resource[3]);
    if (bewitString instanceof Error) ***REMOVED***
        return callback(Boom.badRequest('Invalid bewit encoding'));
    ***REMOVED***

    // Bewit format: id\exp\mac\ext ('\' is used because it is a reserved header attribute character)

    var bewitParts = bewitString.split('\\');
    if (bewitParts.length !== 4) ***REMOVED***
        return callback(Boom.badRequest('Invalid bewit structure'));
    ***REMOVED***

    var bewit = ***REMOVED***
        id: bewitParts[0],
        exp: parseInt(bewitParts[1], 10),
        mac: bewitParts[2],
        ext: bewitParts[3] || ''
    ***REMOVED***;

    if (!bewit.id ||
        !bewit.exp ||
        !bewit.mac) ***REMOVED***

        return callback(Boom.badRequest('Missing bewit attributes'));
    ***REMOVED***

    // Construct URL without bewit

    var url = resource[1];
    if (resource[4]) ***REMOVED***
        url += resource[2] + resource[4];
    ***REMOVED***

    // Check expiration

    if (bewit.exp * 1000 <= now) ***REMOVED***
        return callback(Boom.unauthorized('Access expired', 'Hawk'), null, bewit);
    ***REMOVED***

    // Fetch Hawk credentials

    credentialsFunc(bewit.id, function (err, credentials) ***REMOVED***

        if (err) ***REMOVED***
            return callback(err, credentials || null, bewit.ext);
        ***REMOVED***

        if (!credentials) ***REMOVED***
            return callback(Boom.unauthorized('Unknown credentials', 'Hawk'), null, bewit);
        ***REMOVED***

        if (!credentials.key ||
            !credentials.algorithm) ***REMOVED***

            return callback(Boom.internal('Invalid credentials'), credentials, bewit);
        ***REMOVED***

        if (Crypto.algorithms.indexOf(credentials.algorithm) === -1) ***REMOVED***
            return callback(Boom.internal('Unknown algorithm'), credentials, bewit);
        ***REMOVED***

        // Calculate MAC

        var mac = Crypto.calculateMac('bewit', credentials, ***REMOVED***
            ts: bewit.exp,
            nonce: '',
            method: 'GET',
            resource: url,
            host: request.host,
            port: request.port,
            ext: bewit.ext
        ***REMOVED***);

        if (!Cryptiles.fixedTimeComparison(mac, bewit.mac)) ***REMOVED***
            return callback(Boom.unauthorized('Bad mac', 'Hawk'), credentials, bewit);
        ***REMOVED***

        // Successful authentication

        return callback(null, credentials, bewit);
    ***REMOVED***);
***REMOVED***;


/*
 *  options are the same as authenticate() with the exception that the only supported options are:
 * 'nonceFunc', 'timestampSkewSec', 'localtimeOffsetMsec'
 */

exports.authenticateMessage = function (host, port, message, authorization, credentialsFunc, options, callback) ***REMOVED***

    callback = Hoek.nextTick(callback);

    // Default options

    options.nonceFunc = options.nonceFunc || internals.nonceFunc;
    options.timestampSkewSec = options.timestampSkewSec || 60;                                                  // 60 seconds

    // Application time

    var now = Utils.now(options.localtimeOffsetMsec);                       // Measure now before any other processing

    // Validate authorization

    if (!authorization.id ||
        !authorization.ts ||
        !authorization.nonce ||
        !authorization.hash ||
        !authorization.mac) ***REMOVED***

        return callback(Boom.badRequest('Invalid authorization'));
    ***REMOVED***

    // Fetch Hawk credentials

    credentialsFunc(authorization.id, function (err, credentials) ***REMOVED***

        if (err) ***REMOVED***
            return callback(err, credentials || null);
        ***REMOVED***

        if (!credentials) ***REMOVED***
            return callback(Boom.unauthorized('Unknown credentials', 'Hawk'));
        ***REMOVED***

        if (!credentials.key ||
            !credentials.algorithm) ***REMOVED***

            return callback(Boom.internal('Invalid credentials'), credentials);
        ***REMOVED***

        if (Crypto.algorithms.indexOf(credentials.algorithm) === -1) ***REMOVED***
            return callback(Boom.internal('Unknown algorithm'), credentials);
        ***REMOVED***

        // Construct artifacts container

        var artifacts = ***REMOVED***
            ts: authorization.ts,
            nonce: authorization.nonce,
            host: host,
            port: port,
            hash: authorization.hash
        ***REMOVED***;

        // Calculate MAC

        var mac = Crypto.calculateMac('message', credentials, artifacts);
        if (!Cryptiles.fixedTimeComparison(mac, authorization.mac)) ***REMOVED***
            return callback(Boom.unauthorized('Bad mac', 'Hawk'), credentials);
        ***REMOVED***

        // Check payload hash

        var hash = Crypto.calculatePayloadHash(message, credentials.algorithm);
        if (!Cryptiles.fixedTimeComparison(hash, authorization.hash)) ***REMOVED***
            return callback(Boom.unauthorized('Bad message hash', 'Hawk'), credentials);
        ***REMOVED***

        // Check nonce

        options.nonceFunc(credentials.key, authorization.nonce, authorization.ts, function (err) ***REMOVED***

            if (err) ***REMOVED***
                return callback(Boom.unauthorized('Invalid nonce', 'Hawk'), credentials);
            ***REMOVED***

            // Check timestamp staleness

            if (Math.abs((authorization.ts * 1000) - now) > (options.timestampSkewSec * 1000)) ***REMOVED***
                return callback(Boom.unauthorized('Stale timestamp'), credentials);
            ***REMOVED***

            // Successful authentication

            return callback(null, credentials);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***;


internals.nonceFunc = function (key, nonce, ts, nonceCallback) ***REMOVED***

    return nonceCallback();         // No validation
***REMOVED***;
