// lodash.isplainobject/index.js
/**
 * lodash (Custom Build) <https://lodash.com/>
 * Build: `lodash modularize exports="npm" -o ./`
 * Copyright jQuery Foundation and other contributors <https://jquery.org/>
 * Released under MIT license <https://lodash.com/license>
 * Based on Underscore.js 1.8.3 <http://underscorejs.org/LICENSE>
 * Copyright Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors
 */

/** `Object#toString` result references. */
var objectTag = '[object Object]';

/**
 * Checks if `value` is a host object in IE < 9.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a host object, else `false`.
 */
function isHostObject(value) ***REMOVED***
  // Many host objects are `Object` objects that can coerce to strings
  // despite having improperly defined `toString` methods.
  var result = false;
  if (value != null && typeof value.toString != 'function') ***REMOVED***
    try ***REMOVED***
      result = !!(value + '');
    ***REMOVED*** catch (e) ***REMOVED******REMOVED***
  ***REMOVED***
  return result;
***REMOVED***

/**
 * Creates a unary function that invokes `func` with its argument transformed.
 *
 * @private
 * @param ***REMOVED***Function***REMOVED*** func The function to wrap.
 * @param ***REMOVED***Function***REMOVED*** transform The argument transform.
 * @returns ***REMOVED***Function***REMOVED*** Returns the new function.
 */
function overArg(func, transform) ***REMOVED***
  return function(arg) ***REMOVED***
    return func(transform(arg));
  ***REMOVED***;
***REMOVED***

/** Used for built-in method references. */
var funcProto = Function.prototype,
    objectProto = Object.prototype;

/** Used to resolve the decompiled source of functions. */
var funcToString = funcProto.toString;

/** Used to check objects for own properties. */
var hasOwnProperty = objectProto.hasOwnProperty;

/** Used to infer the `Object` constructor. */
var objectCtorString = funcToString.call(Object);

/**
 * Used to resolve the
 * [`toStringTag`](http://ecma-international.org/ecma-262/7.0/#sec-object.prototype.tostring)
 * of values.
 */
var objectToString = objectProto.toString;

/** Built-in value references. */
var getPrototype = overArg(Object.getPrototypeOf, Object);

/**
 * Checks if `value` is object-like. A value is object-like if it's not `null`
 * and has a `typeof` result of "object".
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is object-like, else `false`.
 * @example
 *
 * _.isObjectLike(***REMOVED******REMOVED***);
 * // => true
 *
 * _.isObjectLike([1, 2, 3]);
 * // => true
 *
 * _.isObjectLike(_.noop);
 * // => false
 *
 * _.isObjectLike(null);
 * // => false
 */
function isObjectLike(value) ***REMOVED***
  return !!value && typeof value == 'object';
***REMOVED***

/**
 * Checks if `value` is a plain object, that is, an object created by the
 * `Object` constructor or one with a `[[Prototype]]` of `null`.
 *
 * @static
 * @memberOf _
 * @since 0.8.0
 * @category Lang
 * @param ***REMOVED*******REMOVED*** value The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** Returns `true` if `value` is a plain object, else `false`.
 * @example
 *
 * function Foo() ***REMOVED***
 *   this.a = 1;
 * ***REMOVED***
 *
 * _.isPlainObject(new Foo);
 * // => false
 *
 * _.isPlainObject([1, 2, 3]);
 * // => false
 *
 * _.isPlainObject(***REMOVED*** 'x': 0, 'y': 0 ***REMOVED***);
 * // => true
 *
 * _.isPlainObject(Object.create(null));
 * // => true
 */
function isPlainObject(value) ***REMOVED***
  if (!isObjectLike(value) ||
      objectToString.call(value) != objectTag || isHostObject(value)) ***REMOVED***
    return false;
  ***REMOVED***
  var proto = getPrototype(value);
  if (proto === null) ***REMOVED***
    return true;
  ***REMOVED***
  var Ctor = hasOwnProperty.call(proto, 'constructor') && proto.constructor;
  return (typeof Ctor == 'function' &&
    Ctor instanceof Ctor && funcToString.call(Ctor) == objectCtorString);
***REMOVED***

module.exports = isPlainObject;
