// lazy-cache/index.js
'use strict';

var set = require('set-getter');

/**
 * Cache results of the first function call to ensure only calling once.
 *
 * ```js
 * var utils = require('lazy-cache')(require);
 * // cache the call to `require('ansi-yellow')`
 * utils('ansi-yellow', 'yellow');
 * // use `ansi-yellow`
 * console.log(utils.yellow('this is yellow'));
 * ```
 *
 * @param  ***REMOVED***Function***REMOVED*** `fn` Function that will be called only once.
 * @return ***REMOVED***Function***REMOVED*** Function that can be called to get the cached function
 * @api public
 */

function lazyCache(requireFn) ***REMOVED***
  var cache = ***REMOVED******REMOVED***;

  return function proxy(name, alias) ***REMOVED***
    var key = alias;

    // camel-case the module `name` if `alias` is not defined
    if (typeof key !== 'string') ***REMOVED***
      key = camelcase(name);
    ***REMOVED***

    // create a getter to lazily invoke the module the first time it's called
    function getter() ***REMOVED***
      return cache[key] || (cache[key] = requireFn(name));
    ***REMOVED***

    // trip the getter if `process.env.UNLAZY` is defined
    if (unlazy(process.env)) ***REMOVED***
      getter();
    ***REMOVED***

    set(proxy, key, getter);
    return getter;
  ***REMOVED***;
***REMOVED***

/**
 * Return true if `process.env.LAZY` is true, or travis is running.
 */

function unlazy(env) ***REMOVED***
  return env.UNLAZY === 'true' || env.UNLAZY === true || env.TRAVIS;
***REMOVED***

/**
 * Camelcase the the given module `name`.
 */

function camelcase(str) ***REMOVED***
  if (str.length === 1) ***REMOVED***
    return str.toLowerCase();
  ***REMOVED***
  str = str.replace(/^[\W_]+|[\W_]+$/g, '').toLowerCase();
  return str.replace(/[\W_]+(\w|$)/g, function(_, ch) ***REMOVED***
    return ch.toUpperCase();
  ***REMOVED***);
***REMOVED***

/**
 * Expose `lazyCache`
 */

module.exports = lazyCache;
