// @firebase/firestore/dist/esm/src/remote/remote_event.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** SnapshotVersion ***REMOVED*** from '../core/snapshot_version';
import ***REMOVED*** documentKeySet ***REMOVED*** from '../model/collections';
import ***REMOVED*** emptyByteString ***REMOVED*** from '../platform/platform';
/**
 * An event from the RemoteStore. It is split into targetChanges (changes to the
 * state or the set of documents in our watched targets) and documentUpdates
 * (changes to the actual documents).
 */
var RemoteEvent = /** @class */ (function () ***REMOVED***
    function RemoteEvent(
        /**
         * The snapshot version this event brings us up to, or MIN if not set.
         */
        snapshotVersion, 
        /**
         * A map from target to changes to the target. See TargetChange.
         */
        targetChanges, 
        /**
         * A set of which documents have changed or been deleted, along with the
         * doc's new values (if not deleted).
         */
        documentUpdates) ***REMOVED***
        this.snapshotVersion = snapshotVersion;
        this.targetChanges = targetChanges;
        this.documentUpdates = documentUpdates;
    ***REMOVED***
    RemoteEvent.prototype.addDocumentUpdate = function (doc) ***REMOVED***
        this.documentUpdates = this.documentUpdates.insert(doc.key, doc);
    ***REMOVED***;
    RemoteEvent.prototype.handleExistenceFilterMismatch = function (targetId) ***REMOVED***
        /*
         * An existence filter mismatch will reset the query and we need to reset
         * the mapping to contain no documents and an empty resume token.
         *
         * Note:
         *   * The reset mapping is empty, specifically forcing the consumer of the
         *     change to forget all keys for this targetID;
         *   * The resume snapshot for this target must be reset
         *   * The target must be unacked because unwatching and rewatching
         *     introduces a race for changes.
         */
        this.targetChanges[targetId] = ***REMOVED***
            mapping: new ResetMapping(),
            snapshotVersion: SnapshotVersion.MIN,
            currentStatusUpdate: CurrentStatusUpdate.MarkNotCurrent,
            resumeToken: emptyByteString()
        ***REMOVED***;
    ***REMOVED***;
    return RemoteEvent;
***REMOVED***());
export ***REMOVED*** RemoteEvent ***REMOVED***;
/**
 * Represents an update to the current status of a target, either explicitly
 * having no new state, or the new value to set. Note "current" has special
 * meaning for in the RPC protocol that implies that a target is both up-to-date
 * and consistent with the rest of the watch stream.
 */
export var CurrentStatusUpdate;
(function (CurrentStatusUpdate) ***REMOVED***
    /** The current status is not affected and should not be modified. */
    CurrentStatusUpdate[CurrentStatusUpdate["None"] = 0] = "None";
    /** The target must be marked as no longer "current". */
    CurrentStatusUpdate[CurrentStatusUpdate["MarkNotCurrent"] = 1] = "MarkNotCurrent";
    /** The target must be marked as "current". */
    CurrentStatusUpdate[CurrentStatusUpdate["MarkCurrent"] = 2] = "MarkCurrent";
***REMOVED***)(CurrentStatusUpdate || (CurrentStatusUpdate = ***REMOVED******REMOVED***));
var EMPTY_KEY_SET = documentKeySet();
var ResetMapping = /** @class */ (function () ***REMOVED***
    function ResetMapping() ***REMOVED***
        this.docs = EMPTY_KEY_SET;
    ***REMOVED***
    Object.defineProperty(ResetMapping.prototype, "documents", ***REMOVED***
        get: function () ***REMOVED***
            return this.docs;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    ResetMapping.prototype.add = function (key) ***REMOVED***
        this.docs = this.docs.add(key);
    ***REMOVED***;
    ResetMapping.prototype.delete = function (key) ***REMOVED***
        this.docs = this.docs.delete(key);
    ***REMOVED***;
    ResetMapping.prototype.isEqual = function (other) ***REMOVED***
        return other !== null && this.docs.isEqual(other.docs);
    ***REMOVED***;
    return ResetMapping;
***REMOVED***());
export ***REMOVED*** ResetMapping ***REMOVED***;
var UpdateMapping = /** @class */ (function () ***REMOVED***
    function UpdateMapping() ***REMOVED***
        this.addedDocuments = EMPTY_KEY_SET;
        this.removedDocuments = EMPTY_KEY_SET;
    ***REMOVED***
    UpdateMapping.prototype.applyToKeySet = function (keys) ***REMOVED***
        var result = keys;
        this.addedDocuments.forEach(function (key) ***REMOVED*** return (result = result.add(key)); ***REMOVED***);
        this.removedDocuments.forEach(function (key) ***REMOVED*** return (result = result.delete(key)); ***REMOVED***);
        return result;
    ***REMOVED***;
    UpdateMapping.prototype.add = function (key) ***REMOVED***
        this.addedDocuments = this.addedDocuments.add(key);
        this.removedDocuments = this.removedDocuments.delete(key);
    ***REMOVED***;
    UpdateMapping.prototype.delete = function (key) ***REMOVED***
        this.addedDocuments = this.addedDocuments.delete(key);
        this.removedDocuments = this.removedDocuments.add(key);
    ***REMOVED***;
    UpdateMapping.prototype.isEqual = function (other) ***REMOVED***
        return (other !== null &&
            this.addedDocuments.isEqual(other.addedDocuments) &&
            this.removedDocuments.isEqual(other.removedDocuments));
    ***REMOVED***;
    return UpdateMapping;
***REMOVED***());
export ***REMOVED*** UpdateMapping ***REMOVED***;
