// power-assert-renderer-base/index.js
'use strict';

function BaseRenderer () ***REMOVED***
***REMOVED***

BaseRenderer.prototype.init = function (traversal) ***REMOVED***
    var _this = this;
    traversal.on('start', function (context) ***REMOVED***
        _this.onStart(context);
    ***REMOVED***);
    traversal.on('data', function (esNode) ***REMOVED***
        _this.onData(esNode);
    ***REMOVED***);
    traversal.on('end', function () ***REMOVED***
        _this.onEnd();
    ***REMOVED***);
***REMOVED***;

BaseRenderer.prototype.setWritable = function (writable) ***REMOVED***
    this.writable = writable;
***REMOVED***;

// API
BaseRenderer.prototype.onStart = function (context) ***REMOVED***
***REMOVED***;

// API
BaseRenderer.prototype.onData = function (esNode) ***REMOVED***
***REMOVED***;

// API
BaseRenderer.prototype.onEnd = function () ***REMOVED***
***REMOVED***;

// API
BaseRenderer.prototype.write = function (str) ***REMOVED***
    this.writable.write(str);
***REMOVED***;

module.exports = BaseRenderer;
