// create-ecdh/browser.js
var elliptic = require('elliptic');
var BN = require('bn.js');

module.exports = function createECDH(curve) ***REMOVED***
	return new ECDH(curve);
***REMOVED***;

var aliases = ***REMOVED***
	secp256k1: ***REMOVED***
		name: 'secp256k1',
		byteLength: 32
	***REMOVED***,
	secp224r1: ***REMOVED***
		name: 'p224',
		byteLength: 28
	***REMOVED***,
	prime256v1: ***REMOVED***
		name: 'p256',
		byteLength: 32
	***REMOVED***,
	prime192v1: ***REMOVED***
		name: 'p192',
		byteLength: 24
	***REMOVED***,
	ed25519: ***REMOVED***
		name: 'ed25519',
		byteLength: 32
	***REMOVED***,
	secp384r1: ***REMOVED***
		name: 'p384',
		byteLength: 48
	***REMOVED***,
	secp521r1: ***REMOVED***
		name: 'p521',
		byteLength: 66
	***REMOVED***
***REMOVED***;

aliases.p224 = aliases.secp224r1;
aliases.p256 = aliases.secp256r1 = aliases.prime256v1;
aliases.p192 = aliases.secp192r1 = aliases.prime192v1;
aliases.p384 = aliases.secp384r1;
aliases.p521 = aliases.secp521r1;

function ECDH(curve) ***REMOVED***
	this.curveType = aliases[curve];
	if (!this.curveType ) ***REMOVED***
		this.curveType = ***REMOVED***
			name: curve
		***REMOVED***;
	***REMOVED***
	this.curve = new elliptic.ec(this.curveType.name);
	this.keys = void 0;
***REMOVED***

ECDH.prototype.generateKeys = function (enc, format) ***REMOVED***
	this.keys = this.curve.genKeyPair();
	return this.getPublicKey(enc, format);
***REMOVED***;

ECDH.prototype.computeSecret = function (other, inenc, enc) ***REMOVED***
	inenc = inenc || 'utf8';
	if (!Buffer.isBuffer(other)) ***REMOVED***
		other = new Buffer(other, inenc);
	***REMOVED***
	var otherPub = this.curve.keyFromPublic(other).getPublic();
	var out = otherPub.mul(this.keys.getPrivate()).getX();
	return formatReturnValue(out, enc, this.curveType.byteLength);
***REMOVED***;

ECDH.prototype.getPublicKey = function (enc, format) ***REMOVED***
	var key = this.keys.getPublic(format === 'compressed', true);
	if (format === 'hybrid') ***REMOVED***
		if (key[key.length - 1] % 2) ***REMOVED***
			key[0] = 7;
		***REMOVED*** else ***REMOVED***
			key [0] = 6;
		***REMOVED***
	***REMOVED***
	return formatReturnValue(key, enc);
***REMOVED***;

ECDH.prototype.getPrivateKey = function (enc) ***REMOVED***
	return formatReturnValue(this.keys.getPrivate(), enc);
***REMOVED***;

ECDH.prototype.setPublicKey = function (pub, enc) ***REMOVED***
	enc = enc || 'utf8';
	if (!Buffer.isBuffer(pub)) ***REMOVED***
		pub = new Buffer(pub, enc);
	***REMOVED***
	this.keys._importPublic(pub);
	return this;
***REMOVED***;

ECDH.prototype.setPrivateKey = function (priv, enc) ***REMOVED***
	enc = enc || 'utf8';
	if (!Buffer.isBuffer(priv)) ***REMOVED***
		priv = new Buffer(priv, enc);
	***REMOVED***
	var _priv = new BN(priv);
	_priv = _priv.toString(16);
	this.keys._importPrivate(_priv);
	return this;
***REMOVED***;

function formatReturnValue(bn, enc, len) ***REMOVED***
	if (!Array.isArray(bn)) ***REMOVED***
		bn = bn.toArray();
	***REMOVED***
	var buf = new Buffer(bn);
	if (len && buf.length < len) ***REMOVED***
		var zeros = new Buffer(len - buf.length);
		zeros.fill(0);
		buf = Buffer.concat([zeros, buf]);
	***REMOVED***
	if (!enc) ***REMOVED***
		return buf;
	***REMOVED*** else ***REMOVED***
		return buf.toString(enc);
	***REMOVED***
***REMOVED***
