// protobufjs/cli/pbjs.js
/*
 Copyright 2013 Daniel Wirtz <dcode@dcode.io>

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
var ProtoBuf = require(__dirname+"/../index.js"),
    fs       = require("fs"),
    path     = require("path"),
    cli      = require("ascli")("pbjs"),
    yargs    = require("yargs"),
    util     = require("./pbjs/util.js"),
    glob     = require("glob"),
    pkg      = require("../package.json");

/**
 * pbjs namespace.
 * @exports pbjs
 * @namespace
 */
var pbjs = module.exports = ***REMOVED******REMOVED***;

/**
 * @alias pbjs/util
 */
pbjs.util = util;

/**
 * Source formats.
 * @type ***REMOVED***!Object.<string,!function(string,!Object.<string,*>)>***REMOVED***
 */
pbjs.sources = ***REMOVED******REMOVED***;
fs.readdirSync(__dirname+"/pbjs/sources").forEach(function(source) ***REMOVED***
    if (/\.js$/.test(source)) ***REMOVED***
        var src = require(__dirname + "/pbjs/sources/" + source);
        if (!src.exclude)
            pbjs.sources[source.substring(0, source.lastIndexOf("."))] = src;
    ***REMOVED***
***REMOVED***);

/**
 * Target formats.
 * @type ***REMOVED***!Object.<string,!function(!ProtoBuf.Builder,!Object.<string,*>)>***REMOVED***
 */
pbjs.targets = ***REMOVED******REMOVED***;
fs.readdirSync(__dirname+"/pbjs/targets").forEach(function(target) ***REMOVED***
    if (/\.js$/.test(target))
        pbjs.targets[target.substring(0, target.lastIndexOf("."))] = require(__dirname + "/pbjs/targets/" + target);
***REMOVED***);

/**
 * Status code: Operation successful
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_OK = 0;

/**
 * Status code: Displaying usage information
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_USAGE = 1;

/**
 * Status code: No such include directory
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_ERR_INCLUDE_DIR = 2;

/**
 * Status code: No such source format
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_ERR_SOURCE_FORMAT = 3;

/**
 * Status code: No such target format
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_ERR_TARGET_FORMAT = 4;

/**
 * Status code: No such namespace
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_ERR_NAMESPACE = 5;

/**
 * Status code: Illegal dependency
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_ERR_DEPENDENCY = 6;

/**
 * Status code: No matching source files
 * @type ***REMOVED***number***REMOVED***
 */
pbjs.STATUS_ERR_NOSOURCE = 7;

// Makes a table of available source or target formats
function mkOptions(obj) ***REMOVED***
    var str = '';
    Object.keys(obj).forEach(function(key) ***REMOVED***
        str += "\n   "+util.pad(key, 10)+" "+obj[key].description;
    ***REMOVED***);
    return str;
***REMOVED***

/**
 * Executes the program.
 * @param ***REMOVED***!Array.<string>***REMOVED*** argv Command line arguments
 * @returns ***REMOVED***number***REMOVED*** Status code
 */
pbjs.main = function(argv) ***REMOVED***
    var options = yargs
        .usage(cli("pb".white.bold+"js".green.bold, util.pad("ProtoBuf.js v"+pkg['version'], 31, true)+" "+pkg['homepage'].grey) + "\n" +
                    "CLI utility to convert between .proto and JSON syntax / to generate classes.\n\n" +
                    "Usage: ".white.bold+path.basename(argv[1]).green.bold+" <source files...> [options] [> outFile]")
        .help("help")
        .version(pkg["version"])
        .wrap(null)
        .options(***REMOVED***
            source: ***REMOVED***
                alias: "s",
                describe: "Specifies the source format. Valid formats are:\n" + mkOptions(pbjs.sources)+"\n"
            ***REMOVED***,
            target: ***REMOVED***
                alias: "t",
                describe: "Specifies the target format. Valid formats are:\n" + mkOptions(pbjs.targets)+"\n"
            ***REMOVED***,
            using: ***REMOVED***
                alias: "u",
                describe: "Specifies an option to apply to the volatile builder\nloading the source, e.g. convertFieldsToCamelCase.",
                type: "array"
            ***REMOVED***,
            min: ***REMOVED***
                alias: "m",
                describe: "Minifies the output.",
                default: false
            ***REMOVED***,
            path: ***REMOVED***
                alias: "p",
                describe: "Adds a directory to the include path."
            ***REMOVED***,
            legacy: ***REMOVED***
                alias: "l",
                describe: "Includes legacy descriptors from google/protobuf/ if\nexplicitly referenced.",
                default: false
            ***REMOVED***,
            quiet: ***REMOVED***
                alias: "q",
                describe: "Suppresses any informatory output to stderr.",
                default: false
            ***REMOVED***,
            out: ***REMOVED***
                alias: "o",
                describe: "Send output to file instead of stdout.",
            ***REMOVED***,
            use: ***REMOVED***
                alias: "i",
                describe: "Specifies an option to apply to the emitted builder\nutilized by your program, e.g. populateAccessors.",
                type: "array"
            ***REMOVED***,
            exports: ***REMOVED***
                alias: "e",
                describe: "Specifies the namespace to export. Defaults to export\nthe root namespace."
            ***REMOVED***,
            dependency: ***REMOVED***
                alias: "d",
                describe: "Library dependency to use when generating classes.\nDefaults to 'protobufjs' for CommonJS, 'protobuf' for\nAMD modules and 'dcodeIO.ProtoBuf' for classes."
            ***REMOVED***
        ***REMOVED***)
        .alias("help", "h")
        .alias("version", "v")
        .check(function (args) ***REMOVED***
            if (args.source && typeof pbjs.sources[args.source] !== "function") ***REMOVED***
                return "Unrecognized source format: '" + args.source + "'";
            ***REMOVED***

            if (args.target && typeof pbjs.targets[args.target] !== "function") ***REMOVED***
                return "Unrecognized target format: '" + args.target + "'";
            ***REMOVED***

            if (args._.length < 3) ***REMOVED***
                return "The filename to parse is required.";
            ***REMOVED***

            return true;
        ***REMOVED***)
        .parse(argv);

    var start = Date.now(),
        sourceFiles = options._.slice(2);

    // Expand glob expressions
    var sourceFilesExpand = [];
    for (var i=0; i<sourceFiles.length; ++i) ***REMOVED***
        var filename = sourceFiles[i],
            files = glob.sync(filename);
        if (files.length === 0) ***REMOVED***
            cli.fail("No matching source files: "+filename);
            return pbjs.STATUS_ERR_NOSOURCE;
        ***REMOVED***
        files.forEach(function(filename) ***REMOVED***
            if (sourceFilesExpand.indexOf(filename) === -1)
                sourceFilesExpand.push(filename);
        ***REMOVED***);
    ***REMOVED***
    sourceFiles = sourceFilesExpand;

    if (!options.target)
        options.target = "json";

    // Set up include paths
    var includePath = Array.isArray(options['path']) ? options['path'] : (typeof options['path'] === 'string' ? [options['path']] : []);
    sourceFiles.forEach(function (sourceFile) ***REMOVED***
        var dir = path.dirname(sourceFile);
        if (includePath.indexOf(dir) === -1) ***REMOVED***
            includePath.push(dir);
        ***REMOVED***
    ***REMOVED***);
    includePath.forEach(function(path) ***REMOVED*** // Verify that include paths actually exist
        if (!fs.existsSync(path)) ***REMOVED***
            if (!options.quiet)
                cli.fail("No such include directory: "+path);
            return pbjs.STATUS_ERR_INCLUDE_DIR;
        ***REMOVED***
    ***REMOVED***);
    options.path = includePath;

    // Detect source format if not specified, then verify
    if (typeof options.source !== 'string') ***REMOVED***
        var source = fs.readFileSync(sourceFiles[0]).toString("utf8").trim();
        if (source.substring(0,1) === "***REMOVED***")
            options.source = "json";
        else
            options.source = "proto";
    ***REMOVED***

    // Load the source files to a common builder
    var builder = pbjs.sources[options.source](sourceFiles, options);

    // Validate exports and dependency if set
    if (typeof options.exports !== 'undefined') ***REMOVED***
        if (!(builder.lookup(options.exports) instanceof ProtoBuf.Reflect.Namespace)) ***REMOVED***
            if (!options.quiet)
                cli.fail("No such export namespace: "+options.exports);
            return pbjs.STATUS_ERR_NAMESPACE;
        ***REMOVED***
        if (options.exports.charAt(0) === '.')
            options.exports = options.exports.substring(1);
    ***REMOVED***
    if (typeof options.dependency !== 'undefined')
        if (typeof options.dependency !== 'string' || !options.dependency) ***REMOVED***
            if (!options.quiet)
                cli.fail("Illegal dependency: "+options.dependency);
            return pbjs.STATUS_ERR_DEPENDENCY;
        ***REMOVED***

    // Perform target conversion
    if (!options.quiet)
        cli.error("\nProcessing: "+sourceFiles.join(", ")+" ...\n");
    var res = pbjs.targets[options.target](builder, options);
    if (options.out)***REMOVED***
        fs.writeFileSync(options.out, res);
    ***REMOVED***else
    process.stdout.write(res);
    if (!options.quiet)
        cli.error(""),
        cli.ok("Converted "+sourceFiles.length+" source files to "+options.target+" ("+res.length+" bytes, "+(Date.now()-start)+" ms)");

    return pbjs.STATUS_OK;
***REMOVED***;
