// hawk/test/uri.js
// Load modules

var Http = require('http');
var Url = require('url');
var Code = require('code');
var Hawk = require('../lib');
var Hoek = require('hoek');
var Lab = require('lab');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.experiment;
var it = lab.test;
var expect = Code.expect;


describe('Uri', function () ***REMOVED***

    var credentialsFunc = function (id, callback) ***REMOVED***

        var credentials = ***REMOVED***
            id: id,
            key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
            algorithm: (id === '1' ? 'sha1' : 'sha256'),
            user: 'steve'
        ***REMOVED***;

        return callback(null, credentials);
    ***REMOVED***;

    it('should generate a bewit then successfully authenticate it', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?a=1&b=2',
            host: 'example.com',
            port: 80
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var bewit = Hawk.uri.getBewit('http://example.com/resource/4?a=1&b=2', ***REMOVED*** credentials: credentials1, ttlSec: 60 * 60 * 24 * 365 * 100, ext: 'some-app-data' ***REMOVED***);
            req.url += '&bewit=' + bewit;

            Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, attributes) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(attributes.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('should generate a bewit then successfully authenticate it (no ext)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?a=1&b=2',
            host: 'example.com',
            port: 80
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var bewit = Hawk.uri.getBewit('http://example.com/resource/4?a=1&b=2', ***REMOVED*** credentials: credentials1, ttlSec: 60 * 60 * 24 * 365 * 100 ***REMOVED***);
            req.url += '&bewit=' + bewit;

            Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, attributes) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('should successfully authenticate a request (last param)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?a=1&b=2&bewit=MTIzNDU2XDQ1MTE0ODQ2MjFcMzFjMmNkbUJFd1NJRVZDOVkva1NFb2c3d3YrdEVNWjZ3RXNmOGNHU2FXQT1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.not.exist();
            expect(credentials.user).to.equal('steve');
            expect(attributes.ext).to.equal('some-app-data');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should successfully authenticate a request (first param)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MTE0ODQ2MjFcMzFjMmNkbUJFd1NJRVZDOVkva1NFb2c3d3YrdEVNWjZ3RXNmOGNHU2FXQT1cc29tZS1hcHAtZGF0YQ&a=1&b=2',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.not.exist();
            expect(credentials.user).to.equal('steve');
            expect(attributes.ext).to.equal('some-app-data');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should successfully authenticate a request (only param)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MTE0ODQ2NDFcZm1CdkNWT3MvcElOTUUxSTIwbWhrejQ3UnBwTmo4Y1VrSHpQd3Q5OXJ1cz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.not.exist();
            expect(credentials.user).to.equal('steve');
            expect(attributes.ext).to.equal('some-app-data');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on multiple authentication', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MTE0ODQ2NDFcZm1CdkNWT3MvcElOTUUxSTIwbWhrejQ3UnBwTmo4Y1VrSHpQd3Q5OXJ1cz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080,
            authorization: 'Basic asdasdasdasd'
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Multiple authentications');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on method other than GET', function (done) ***REMOVED***

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var req = ***REMOVED***
                method: 'POST',
                url: '/resource/4?filter=a',
                host: 'example.com',
                port: 8080
            ***REMOVED***;

            var exp = Math.floor(Hawk.utils.now() / 1000) + 60;
            var ext = 'some-app-data';
            var mac = Hawk.crypto.calculateMac('bewit', credentials1, ***REMOVED***
                timestamp: exp,
                nonce: '',
                method: req.method,
                resource: req.url,
                host: req.host,
                port: req.port,
                ext: ext
            ***REMOVED***);

            var bewit = credentials1.id + '\\' + exp + '\\' + mac + '\\' + ext;

            req.url += '&bewit=' + Hoek.base64urlEncode(bewit);

            Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, attributes) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Invalid method');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on invalid host header', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            headers: ***REMOVED***
                host: 'example.com:something'
            ***REMOVED***
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Invalid Host header');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on empty bewit', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Empty bewit');
            expect(err.isMissing).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on invalid bewit', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=*',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Invalid bewit encoding');
            expect(err.isMissing).to.not.exist();
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on missing bewit', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.not.exist();
            expect(err.isMissing).to.equal(true);
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on invalid bewit structure', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=abc',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Invalid bewit structure');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on empty bewit attribute', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=YVxcY1xk',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Missing bewit attributes');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on missing bewit id attribute', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=XDQ1NTIxNDc2MjJcK0JFbFhQMXhuWjcvd1Nrbm1ldGhlZm5vUTNHVjZNSlFVRHk4NWpTZVJ4VT1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Missing bewit attributes');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on expired access', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?a=1&b=2&bewit=MTIzNDU2XDEzNTY0MTg1ODNcWk1wZlMwWU5KNHV0WHpOMmRucTRydEk3NXNXTjFjeWVITTcrL0tNZFdVQT1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Access expired');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on credentials function error', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, function (id, callback) ***REMOVED***

            callback(Hawk.error.badRequest('Boom'));
        ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Boom');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on credentials function error with credentials', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, function (id, callback) ***REMOVED***

            callback(Hawk.error.badRequest('Boom'), ***REMOVED*** some: 'value' ***REMOVED***);
        ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Boom');
            expect(credentials.some).to.equal('value');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on null credentials function response', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, function (id, callback) ***REMOVED***

            callback(null, null);
        ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Unknown credentials');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on invalid credentials function response', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, function (id, callback) ***REMOVED***

            callback(null, ***REMOVED******REMOVED***);
        ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.equal('Invalid credentials');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on invalid credentials function response (unknown algorithm)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, function (id, callback) ***REMOVED***

            callback(null, ***REMOVED*** key: 'xxx', algorithm: 'xxx' ***REMOVED***);
        ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.message).to.equal('Unknown algorithm');
            done();
        ***REMOVED***);
    ***REMOVED***);

    it('should fail on expired access', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?bewit=MTIzNDU2XDQ1MDk5OTE3MTlcTUE2eWkwRWRwR0pEcWRwb0JkYVdvVDJrL0hDSzA1T0Y3MkhuZlVmVy96Zz1cc29tZS1hcHAtZGF0YQ',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        Hawk.uri.authenticate(req, function (id, callback) ***REMOVED***

            callback(null, ***REMOVED*** key: 'xxx', algorithm: 'sha256' ***REMOVED***);
        ***REMOVED***, ***REMOVED******REMOVED***, function (err, credentials, attributes) ***REMOVED***

            expect(err).to.exist();
            expect(err.output.payload.message).to.equal('Bad mac');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('getBewit()', function () ***REMOVED***

        it('returns a valid bewit value', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdca3NjeHdOUjJ0SnBQMVQxekRMTlBiQjVVaUtJVTl0T1NKWFRVZEc3WDloOD1ceGFuZHlhbmR6');
            done();
        ***REMOVED***);

        it('returns a valid bewit value (explicit port)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com:8080/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdcaFpiSjNQMmNLRW80a3kwQzhqa1pBa1J5Q1p1ZWc0V1NOYnhWN3ZxM3hIVT1ceGFuZHlhbmR6');
            done();
        ***REMOVED***);

        it('returns a valid bewit value (null ext)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: null ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdcSUdZbUxnSXFMckNlOEN4dktQczRKbFdJQStValdKSm91d2dBUmlWaENBZz1c');
            done();
        ***REMOVED***);

        it('returns a valid bewit value (parsed uri)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit(Url.parse('https://example.com/somewhere/over/the/rainbow'), ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('MTIzNDU2XDEzNTY0MjA3MDdca3NjeHdOUjJ0SnBQMVQxekRMTlBiQjVVaUtJVTl0T1NKWFRVZEc3WDloOD1ceGFuZHlhbmR6');
            done();
        ***REMOVED***);

        it('errors on invalid options', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', 4);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing uri', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('', ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid uri', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit(5, ***REMOVED*** credentials: credentials, ttlSec: 300, localtimeOffsetMsec: 1356420407232 - Hawk.utils.now(), ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid credentials (id)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                key: '2983d45yun89q',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 3000, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing credentials', function (done) ***REMOVED***

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** ttlSec: 3000, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid credentials (key)', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                algorithm: 'sha256'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 3000, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on invalid algorithm', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'hmac-sha-0'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow', ***REMOVED*** credentials: credentials, ttlSec: 300, ext: 'xandyandz' ***REMOVED***);
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);

        it('errors on missing options', function (done) ***REMOVED***

            var credentials = ***REMOVED***
                id: '123456',
                key: '2983d45yun89q',
                algorithm: 'hmac-sha-0'
            ***REMOVED***;

            var bewit = Hawk.uri.getBewit('https://example.com/somewhere/over/the/rainbow');
            expect(bewit).to.equal('');
            done();
        ***REMOVED***);
    ***REMOVED***);

    describe('authenticateMessage()', function () ***REMOVED***

        it('should generate an authorization then successfully parse it', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.not.exist();
                    expect(credentials2.user).to.equal('steve');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on mismatching host', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example1.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Bad mac');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on stale timestamp', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED*** localtimeOffsetMsec: 100000 ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Stale timestamp');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('overrides timestampSkewSec', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1, localtimeOffsetMsec: 100000 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED*** timestampSkewSec: 500 ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.not.exist();
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on invalid authorization', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();
                delete auth.id;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid authorization');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on bad hash', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message1', auth, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Bad message hash');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on nonce error', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, credentialsFunc, ***REMOVED***
                    nonceFunc: function (key, nonce, ts, callback) ***REMOVED***

                        callback(new Error('kaboom'));
                    ***REMOVED***
                ***REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid nonce');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on credentials error', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback(new Error('kablooey'));
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('kablooey');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on missing credentials', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback();
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Unknown credentials');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on invalid credentials', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback(null, ***REMOVED******REMOVED***);
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Invalid credentials');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail authorization on invalid credentials algorithm', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.exist();

                var errFunc = function (id, callback) ***REMOVED***

                    callback(null, ***REMOVED*** key: '123', algorithm: '456' ***REMOVED***);
                ***REMOVED***;

                Hawk.server.authenticateMessage('example.com', 8080, 'some message', auth, errFunc, ***REMOVED******REMOVED***, function (err, credentials2) ***REMOVED***

                    expect(err).to.exist();
                    expect(err.message).to.equal('Unknown algorithm');
                    done();
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);

        it('should fail on missing host', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var auth = Hawk.client.message(null, 8080, 'some message', ***REMOVED*** credentials: credentials1 ***REMOVED***);
                expect(auth).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);

        it('should fail on missing credentials', function (done) ***REMOVED***

            var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED******REMOVED***);
            expect(auth).to.not.exist();
            done();
        ***REMOVED***);

        it('should fail on invalid algorithm', function (done) ***REMOVED***

            credentialsFunc('123456', function (err, credentials1) ***REMOVED***

                var creds = Hoek.clone(credentials1);
                creds.algorithm = 'blah';
                var auth = Hawk.client.message('example.com', 8080, 'some message', ***REMOVED*** credentials: creds ***REMOVED***);
                expect(auth).to.not.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
