// fstream/lib/abstract.js
// the parent class for all fstreams.

module.exports = Abstract

var Stream = require('stream').Stream
var inherits = require('inherits')

function Abstract () ***REMOVED***
  Stream.call(this)
***REMOVED***

inherits(Abstract, Stream)

Abstract.prototype.on = function (ev, fn) ***REMOVED***
  if (ev === 'ready' && this.ready) ***REMOVED***
    process.nextTick(fn.bind(this))
  ***REMOVED*** else ***REMOVED***
    Stream.prototype.on.call(this, ev, fn)
  ***REMOVED***
  return this
***REMOVED***

Abstract.prototype.abort = function () ***REMOVED***
  this._aborted = true
  this.emit('abort')
***REMOVED***

Abstract.prototype.destroy = function () ***REMOVED******REMOVED***

Abstract.prototype.warn = function (msg, code) ***REMOVED***
  var self = this
  var er = decorate(msg, code, self)
  if (!self.listeners('warn')) ***REMOVED***
    console.error('%s %s\n' +
    'path = %s\n' +
    'syscall = %s\n' +
    'fstream_type = %s\n' +
    'fstream_path = %s\n' +
    'fstream_unc_path = %s\n' +
    'fstream_class = %s\n' +
    'fstream_stack =\n%s\n',
      code || 'UNKNOWN',
      er.stack,
      er.path,
      er.syscall,
      er.fstream_type,
      er.fstream_path,
      er.fstream_unc_path,
      er.fstream_class,
      er.fstream_stack.join('\n'))
  ***REMOVED*** else ***REMOVED***
    self.emit('warn', er)
  ***REMOVED***
***REMOVED***

Abstract.prototype.info = function (msg, code) ***REMOVED***
  this.emit('info', msg, code)
***REMOVED***

Abstract.prototype.error = function (msg, code, th) ***REMOVED***
  var er = decorate(msg, code, this)
  if (th) throw er
  else this.emit('error', er)
***REMOVED***

function decorate (er, code, self) ***REMOVED***
  if (!(er instanceof Error)) er = new Error(er)
  er.code = er.code || code
  er.path = er.path || self.path
  er.fstream_type = er.fstream_type || self.type
  er.fstream_path = er.fstream_path || self.path
  if (self._path !== self.path) ***REMOVED***
    er.fstream_unc_path = er.fstream_unc_path || self._path
  ***REMOVED***
  if (self.linkpath) ***REMOVED***
    er.fstream_linkpath = er.fstream_linkpath || self.linkpath
  ***REMOVED***
  er.fstream_class = er.fstream_class || self.constructor.name
  er.fstream_stack = er.fstream_stack ||
    new Error().stack.split(/\n/).slice(3).map(function (s) ***REMOVED***
      return s.replace(/^ ***REMOVED***4***REMOVED***at /, '')
    ***REMOVED***)

  return er
***REMOVED***
