// snapdragon/index.js
'use strict';

var Base = require('base');
var define = require('define-property');
var Compiler = require('./lib/compiler');
var Parser = require('./lib/parser');
var utils = require('./lib/utils');
var regexCache = ***REMOVED******REMOVED***;
var cache = ***REMOVED******REMOVED***;

/**
 * Create a new instance of `Snapdragon` with the given `options`.
 *
 * ```js
 * var snapdragon = new Snapdragon();
 * ```
 *
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @api public
 */

function Snapdragon(options) ***REMOVED***
  Base.call(this, null, options);
  this.options = utils.extend(***REMOVED***source: 'string'***REMOVED***, this.options);
  this.compiler = new Compiler(this.options);
  this.parser = new Parser(this.options);

  Object.defineProperty(this, 'compilers', ***REMOVED***
    get: function() ***REMOVED***
      return this.compiler.compilers;
    ***REMOVED***
  ***REMOVED***);

  Object.defineProperty(this, 'parsers', ***REMOVED***
    get: function() ***REMOVED***
      return this.parser.parsers;
    ***REMOVED***
  ***REMOVED***);

  Object.defineProperty(this, 'regex', ***REMOVED***
    get: function() ***REMOVED***
      return this.parser.regex;
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

/**
 * Inherit Base
 */

Base.extend(Snapdragon);

/**
 * Add a parser to `snapdragon.parsers` for capturing the given `type` using
 * the specified regex or parser function. A function is useful if you need
 * to customize how the token is created and/or have access to the parser
 * instance to check options, etc.
 *
 * ```js
 * snapdragon
 *   .capture('slash', /^\//)
 *   .capture('dot', function() ***REMOVED***
 *     var pos = this.position();
 *     var m = this.match(/^\./);
 *     if (!m) return;
 *     return pos(***REMOVED***
 *       type: 'dot',
 *       val: m[0]
 *     ***REMOVED***);
 *   ***REMOVED***);
 * ```
 * @param ***REMOVED***String***REMOVED*** `type`
 * @param ***REMOVED***RegExp|Function***REMOVED*** `regex`
 * @return ***REMOVED***Object***REMOVED*** Returns the parser instance for chaining
 * @api public
 */

Snapdragon.prototype.capture = function() ***REMOVED***
  return this.parser.capture.apply(this.parser, arguments);
***REMOVED***;

/**
 * Register a plugin `fn`.
 *
 * ```js
 * var snapdragon = new Snapdgragon([options]);
 * snapdragon.use(function() ***REMOVED***
 *   console.log(this);          //<= snapdragon instance
 *   console.log(this.parser);   //<= parser instance
 *   console.log(this.compiler); //<= compiler instance
 * ***REMOVED***);
 * ```
 * @param ***REMOVED***Object***REMOVED*** `fn`
 * @api public
 */

Snapdragon.prototype.use = function(fn) ***REMOVED***
  fn.call(this, this);
  return this;
***REMOVED***;

/**
 * Parse the given `str`.
 *
 * ```js
 * var snapdragon = new Snapdgragon([options]);
 * // register parsers
 * snapdragon.parser.use(function() ***REMOVED******REMOVED***);
 *
 * // parse
 * var ast = snapdragon.parse('foo/bar');
 * console.log(ast);
 * ```
 * @param ***REMOVED***String***REMOVED*** `str`
 * @param ***REMOVED***Object***REMOVED*** `options` Set `options.sourcemap` to true to enable source maps.
 * @return ***REMOVED***Object***REMOVED*** Returns an AST.
 * @api public
 */

Snapdragon.prototype.parse = function(str, options) ***REMOVED***
  this.options = utils.extend(***REMOVED******REMOVED***, this.options, options);
  var parsed = this.parser.parse(str, this.options);

  // add non-enumerable parser reference
  define(parsed, 'parser', this.parser);
  return parsed;
***REMOVED***;

/**
 * Compile the given `AST`.
 *
 * ```js
 * var snapdragon = new Snapdgragon([options]);
 * // register plugins
 * snapdragon.use(function() ***REMOVED******REMOVED***);
 * // register parser plugins
 * snapdragon.parser.use(function() ***REMOVED******REMOVED***);
 * // register compiler plugins
 * snapdragon.compiler.use(function() ***REMOVED******REMOVED***);
 *
 * // parse
 * var ast = snapdragon.parse('foo/bar');
 *
 * // compile
 * var res = snapdragon.compile(ast);
 * console.log(res.output);
 * ```
 * @param ***REMOVED***Object***REMOVED*** `ast`
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @return ***REMOVED***Object***REMOVED*** Returns an object with an `output` property with the rendered string.
 * @api public
 */

Snapdragon.prototype.compile = function(ast, options) ***REMOVED***
  this.options = utils.extend(***REMOVED******REMOVED***, this.options, options);
  var compiled = this.compiler.compile(ast, this.options);

  // add non-enumerable compiler reference
  define(compiled, 'compiler', this.compiler);
  return compiled;
***REMOVED***;

/**
 * Expose `Snapdragon`
 */

module.exports = Snapdragon;

/**
 * Expose `Parser` and `Compiler`
 */

module.exports.Compiler = Compiler;
module.exports.Parser = Parser;
