// @firebase/storage/dist/esm/src/implementation/args.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as errorsExports from './error';
import * as MetadataUtils from './metadata';
import * as type from './type';
/**
 * @param name Name of the function.
 * @param specs Argument specs.
 * @param passed The actual arguments passed to the function.
 * @throws ***REMOVED***fbs.Error***REMOVED*** If the arguments are invalid.
 */
export function validate(name, specs, passed) ***REMOVED***
    var minArgs = specs.length;
    var maxArgs = specs.length;
    for (var i = 0; i < specs.length; i++) ***REMOVED***
        if (specs[i].optional) ***REMOVED***
            minArgs = i;
            break;
        ***REMOVED***
    ***REMOVED***
    var validLength = minArgs <= passed.length && passed.length <= maxArgs;
    if (!validLength) ***REMOVED***
        throw errorsExports.invalidArgumentCount(minArgs, maxArgs, name, passed.length);
    ***REMOVED***
    for (var i = 0; i < passed.length; i++) ***REMOVED***
        try ***REMOVED***
            specs[i].validator(passed[i]);
        ***REMOVED***
        catch (e) ***REMOVED***
            if (e instanceof Error) ***REMOVED***
                throw errorsExports.invalidArgument(i, name, e.message);
            ***REMOVED***
            else ***REMOVED***
                throw errorsExports.invalidArgument(i, name, e);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***
/**
 * @struct
 */
var ArgSpec = /** @class */ (function () ***REMOVED***
    function ArgSpec(validator, opt_optional) ***REMOVED***
        var self = this;
        this.validator = function (p) ***REMOVED***
            if (self.optional && !type.isJustDef(p)) ***REMOVED***
                return;
            ***REMOVED***
            validator(p);
        ***REMOVED***;
        this.optional = !!opt_optional;
    ***REMOVED***
    return ArgSpec;
***REMOVED***());
export ***REMOVED*** ArgSpec ***REMOVED***;
export function and_(v1, v2) ***REMOVED***
    return function (p) ***REMOVED***
        v1(p);
        v2(p);
    ***REMOVED***;
***REMOVED***
export function stringSpec(opt_validator, opt_optional) ***REMOVED***
    function stringValidator(p) ***REMOVED***
        if (!type.isString(p)) ***REMOVED***
            throw 'Expected string.';
        ***REMOVED***
    ***REMOVED***
    var validator;
    if (opt_validator) ***REMOVED***
        validator = and_(stringValidator, opt_validator);
    ***REMOVED***
    else ***REMOVED***
        validator = stringValidator;
    ***REMOVED***
    return new ArgSpec(validator, opt_optional);
***REMOVED***
export function uploadDataSpec() ***REMOVED***
    function validator(p) ***REMOVED***
        var valid = p instanceof Uint8Array ||
            p instanceof ArrayBuffer ||
            (type.isNativeBlobDefined() && p instanceof Blob);
        if (!valid) ***REMOVED***
            throw 'Expected Blob or File.';
        ***REMOVED***
    ***REMOVED***
    return new ArgSpec(validator);
***REMOVED***
export function metadataSpec(opt_optional) ***REMOVED***
    return new ArgSpec(MetadataUtils.metadataValidator, opt_optional);
***REMOVED***
export function nonNegativeNumberSpec() ***REMOVED***
    function validator(p) ***REMOVED***
        var valid = type.isNumber(p) && p >= 0;
        if (!valid) ***REMOVED***
            throw 'Expected a number 0 or greater.';
        ***REMOVED***
    ***REMOVED***
    return new ArgSpec(validator);
***REMOVED***
export function looseObjectSpec(opt_validator, opt_optional) ***REMOVED***
    function validator(p) ***REMOVED***
        var isLooseObject = p === null || (type.isDef(p) && p instanceof Object);
        if (!isLooseObject) ***REMOVED***
            throw 'Expected an Object.';
        ***REMOVED***
        if (opt_validator !== undefined && opt_validator !== null) ***REMOVED***
            opt_validator(p);
        ***REMOVED***
    ***REMOVED***
    return new ArgSpec(validator, opt_optional);
***REMOVED***
export function nullFunctionSpec(opt_optional) ***REMOVED***
    function validator(p) ***REMOVED***
        var valid = p === null || type.isFunction(p);
        if (!valid) ***REMOVED***
            throw 'Expected a Function.';
        ***REMOVED***
    ***REMOVED***
    return new ArgSpec(validator, opt_optional);
***REMOVED***
