// @firebase/storage/dist/esm/src/implementation/error.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** configOption ***REMOVED*** from './constants';
var FirebaseStorageError = /** @class */ (function () ***REMOVED***
    function FirebaseStorageError(code, message) ***REMOVED***
        this.code_ = prependCode(code);
        this.message_ = 'Firebase Storage: ' + message;
        this.serverResponse_ = null;
        this.name_ = 'FirebaseError';
    ***REMOVED***
    FirebaseStorageError.prototype.codeProp = function () ***REMOVED***
        return this.code;
    ***REMOVED***;
    FirebaseStorageError.prototype.codeEquals = function (code) ***REMOVED***
        return prependCode(code) === this.codeProp();
    ***REMOVED***;
    FirebaseStorageError.prototype.serverResponseProp = function () ***REMOVED***
        return this.serverResponse_;
    ***REMOVED***;
    FirebaseStorageError.prototype.setServerResponseProp = function (serverResponse) ***REMOVED***
        this.serverResponse_ = serverResponse;
    ***REMOVED***;
    Object.defineProperty(FirebaseStorageError.prototype, "name", ***REMOVED***
        get: function () ***REMOVED***
            return this.name_;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseStorageError.prototype, "code", ***REMOVED***
        get: function () ***REMOVED***
            return this.code_;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseStorageError.prototype, "message", ***REMOVED***
        get: function () ***REMOVED***
            return this.message_;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseStorageError.prototype, "serverResponse", ***REMOVED***
        get: function () ***REMOVED***
            return this.serverResponse_;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    return FirebaseStorageError;
***REMOVED***());
export ***REMOVED*** FirebaseStorageError ***REMOVED***;
export var errors = ***REMOVED******REMOVED***;
export var Code = ***REMOVED***
    // Shared between all platforms
    UNKNOWN: 'unknown',
    OBJECT_NOT_FOUND: 'object-not-found',
    BUCKET_NOT_FOUND: 'bucket-not-found',
    PROJECT_NOT_FOUND: 'project-not-found',
    QUOTA_EXCEEDED: 'quota-exceeded',
    UNAUTHENTICATED: 'unauthenticated',
    UNAUTHORIZED: 'unauthorized',
    RETRY_LIMIT_EXCEEDED: 'retry-limit-exceeded',
    INVALID_CHECKSUM: 'invalid-checksum',
    CANCELED: 'canceled',
    // JS specific
    INVALID_EVENT_NAME: 'invalid-event-name',
    INVALID_URL: 'invalid-url',
    INVALID_DEFAULT_BUCKET: 'invalid-default-bucket',
    NO_DEFAULT_BUCKET: 'no-default-bucket',
    CANNOT_SLICE_BLOB: 'cannot-slice-blob',
    SERVER_FILE_WRONG_SIZE: 'server-file-wrong-size',
    NO_DOWNLOAD_URL: 'no-download-url',
    INVALID_ARGUMENT: 'invalid-argument',
    INVALID_ARGUMENT_COUNT: 'invalid-argument-count',
    APP_DELETED: 'app-deleted',
    INVALID_ROOT_OPERATION: 'invalid-root-operation',
    INVALID_FORMAT: 'invalid-format',
    INTERNAL_ERROR: 'internal-error'
***REMOVED***;
export function prependCode(code) ***REMOVED***
    return 'storage/' + code;
***REMOVED***
export function unknown() ***REMOVED***
    var message = 'An unknown error occurred, please check the error payload for ' +
        'server response.';
    return new FirebaseStorageError(Code.UNKNOWN, message);
***REMOVED***
export function objectNotFound(path) ***REMOVED***
    return new FirebaseStorageError(Code.OBJECT_NOT_FOUND, "Object '" + path + "' does not exist.");
***REMOVED***
export function bucketNotFound(bucket) ***REMOVED***
    return new FirebaseStorageError(Code.BUCKET_NOT_FOUND, "Bucket '" + bucket + "' does not exist.");
***REMOVED***
export function projectNotFound(project) ***REMOVED***
    return new FirebaseStorageError(Code.PROJECT_NOT_FOUND, "Project '" + project + "' does not exist.");
***REMOVED***
export function quotaExceeded(bucket) ***REMOVED***
    return new FirebaseStorageError(Code.QUOTA_EXCEEDED, "Quota for bucket '" +
        bucket +
        "' exceeded, please view quota on " +
        'https://firebase.google.com/pricing/.');
***REMOVED***
export function unauthenticated() ***REMOVED***
    var message = 'User is not authenticated, please authenticate using Firebase ' +
        'Authentication and try again.';
    return new FirebaseStorageError(Code.UNAUTHENTICATED, message);
***REMOVED***
export function unauthorized(path) ***REMOVED***
    return new FirebaseStorageError(Code.UNAUTHORIZED, "User does not have permission to access '" + path + "'.");
***REMOVED***
export function retryLimitExceeded() ***REMOVED***
    return new FirebaseStorageError(Code.RETRY_LIMIT_EXCEEDED, 'Max retry time for operation exceeded, please try again.');
***REMOVED***
export function invalidChecksum(path, checksum, calculated) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_CHECKSUM, "Uploaded/downloaded object '" +
        path +
        "' has checksum '" +
        checksum +
        "' which does not match '" +
        calculated +
        "'. Please retry the upload/download.");
***REMOVED***
export function canceled() ***REMOVED***
    return new FirebaseStorageError(Code.CANCELED, 'User canceled the upload/download.');
***REMOVED***
export function invalidEventName(name) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_EVENT_NAME, "Invalid event name '" + name + "'.");
***REMOVED***
export function invalidUrl(url) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_URL, "Invalid URL '" + url + "'.");
***REMOVED***
export function invalidDefaultBucket(bucket) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_DEFAULT_BUCKET, "Invalid default bucket '" + bucket + "'.");
***REMOVED***
export function noDefaultBucket() ***REMOVED***
    return new FirebaseStorageError(Code.NO_DEFAULT_BUCKET, 'No default bucket ' +
        "found. Did you set the '" +
        configOption +
        "' property when initializing the app?");
***REMOVED***
export function cannotSliceBlob() ***REMOVED***
    return new FirebaseStorageError(Code.CANNOT_SLICE_BLOB, 'Cannot slice blob for upload. Please retry the upload.');
***REMOVED***
export function serverFileWrongSize() ***REMOVED***
    return new FirebaseStorageError(Code.SERVER_FILE_WRONG_SIZE, 'Server recorded incorrect upload file size, please retry the upload.');
***REMOVED***
export function noDownloadURL() ***REMOVED***
    return new FirebaseStorageError(Code.NO_DOWNLOAD_URL, 'The given file does not have any download URLs.');
***REMOVED***
export function invalidArgument(index, fnName, message) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_ARGUMENT, 'Invalid argument in `' + fnName + '` at index ' + index + ': ' + message);
***REMOVED***
export function invalidArgumentCount(argMin, argMax, fnName, real) ***REMOVED***
    var countPart;
    var plural;
    if (argMin === argMax) ***REMOVED***
        countPart = argMin;
        plural = argMin === 1 ? 'argument' : 'arguments';
    ***REMOVED***
    else ***REMOVED***
        countPart = 'between ' + argMin + ' and ' + argMax;
        plural = 'arguments';
    ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_ARGUMENT_COUNT, 'Invalid argument count in `' +
        fnName +
        '`: Expected ' +
        countPart +
        ' ' +
        plural +
        ', received ' +
        real +
        '.');
***REMOVED***
export function appDeleted() ***REMOVED***
    return new FirebaseStorageError(Code.APP_DELETED, 'The Firebase app was deleted.');
***REMOVED***
/**
 * @param name The name of the operation that was invalid.
 */
export function invalidRootOperation(name) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_ROOT_OPERATION, "The operation '" +
        name +
        "' cannot be performed on a root reference, create a non-root " +
        "reference using child, such as .child('file.png').");
***REMOVED***
/**
 * @param format The format that was not valid.
 * @param message A message describing the format violation.
 */
export function invalidFormat(format, message) ***REMOVED***
    return new FirebaseStorageError(Code.INVALID_FORMAT, "String does not match format '" + format + "': " + message);
***REMOVED***
/**
 * @param message A message describing the internal error.
 */
export function internalError(message) ***REMOVED***
    throw new FirebaseStorageError(Code.INTERNAL_ERROR, 'Internal error: ' + message);
***REMOVED***
