// @firebase/storage/dist/esm/src/implementation/type.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @return False if the object is undefined or null, true otherwise.
 */
export function isDef(p) ***REMOVED***
    return p != null;
***REMOVED***
export function isJustDef(p) ***REMOVED***
    return p !== void 0;
***REMOVED***
export function isFunction(p) ***REMOVED***
    return typeof p === 'function';
***REMOVED***
export function isObject(p) ***REMOVED***
    return typeof p === 'object';
***REMOVED***
export function isNonNullObject(p) ***REMOVED***
    return isObject(p) && p !== null;
***REMOVED***
export function isNonArrayObject(p) ***REMOVED***
    return isObject(p) && !Array.isArray(p);
***REMOVED***
export function isString(p) ***REMOVED***
    return typeof p === 'string' || p instanceof String;
***REMOVED***
export function isNumber(p) ***REMOVED***
    return typeof p === 'number' || p instanceof Number;
***REMOVED***
export function isNativeBlob(p) ***REMOVED***
    return isNativeBlobDefined() && p instanceof Blob;
***REMOVED***
export function isNativeBlobDefined() ***REMOVED***
    return typeof Blob !== 'undefined';
***REMOVED***
