// pbkdf2/lib/sync-browser.js
var md5 = require('create-hash/md5')
var rmd160 = require('ripemd160')
var sha = require('sha.js')

var checkParameters = require('./precondition')
var defaultEncoding = require('./default-encoding')
var Buffer = require('safe-buffer').Buffer
var ZEROS = Buffer.alloc(128)
var sizes = ***REMOVED***
  md5: 16,
  sha1: 20,
  sha224: 28,
  sha256: 32,
  sha384: 48,
  sha512: 64,
  rmd160: 20,
  ripemd160: 20
***REMOVED***

function Hmac (alg, key, saltLen) ***REMOVED***
  var hash = getDigest(alg)
  var blocksize = (alg === 'sha512' || alg === 'sha384') ? 128 : 64

  if (key.length > blocksize) ***REMOVED***
    key = hash(key)
  ***REMOVED*** else if (key.length < blocksize) ***REMOVED***
    key = Buffer.concat([key, ZEROS], blocksize)
  ***REMOVED***

  var ipad = Buffer.allocUnsafe(blocksize + sizes[alg])
  var opad = Buffer.allocUnsafe(blocksize + sizes[alg])
  for (var i = 0; i < blocksize; i++) ***REMOVED***
    ipad[i] = key[i] ^ 0x36
    opad[i] = key[i] ^ 0x5C
  ***REMOVED***

  var ipad1 = Buffer.allocUnsafe(blocksize + saltLen + 4)
  ipad.copy(ipad1, 0, 0, blocksize)
  this.ipad1 = ipad1
  this.ipad2 = ipad
  this.opad = opad
  this.alg = alg
  this.blocksize = blocksize
  this.hash = hash
  this.size = sizes[alg]
***REMOVED***

Hmac.prototype.run = function (data, ipad) ***REMOVED***
  data.copy(ipad, this.blocksize)
  var h = this.hash(ipad)
  h.copy(this.opad, this.blocksize)
  return this.hash(this.opad)
***REMOVED***

function getDigest (alg) ***REMOVED***
  function shaFunc (data) ***REMOVED***
    return sha(alg).update(data).digest()
  ***REMOVED***

  if (alg === 'rmd160' || alg === 'ripemd160') return rmd160
  if (alg === 'md5') return md5
  return shaFunc
***REMOVED***

function pbkdf2 (password, salt, iterations, keylen, digest) ***REMOVED***
  if (!Buffer.isBuffer(password)) password = Buffer.from(password, defaultEncoding)
  if (!Buffer.isBuffer(salt)) salt = Buffer.from(salt, defaultEncoding)

  checkParameters(iterations, keylen)

  digest = digest || 'sha1'

  var hmac = new Hmac(digest, password, salt.length)

  var DK = Buffer.allocUnsafe(keylen)
  var block1 = Buffer.allocUnsafe(salt.length + 4)
  salt.copy(block1, 0, 0, salt.length)

  var destPos = 0
  var hLen = sizes[digest]
  var l = Math.ceil(keylen / hLen)

  for (var i = 1; i <= l; i++) ***REMOVED***
    block1.writeUInt32BE(i, salt.length)

    var T = hmac.run(block1, hmac.ipad1)
    var U = T

    for (var j = 1; j < iterations; j++) ***REMOVED***
      U = hmac.run(U, hmac.ipad2)
      for (var k = 0; k < hLen; k++) T[k] ^= U[k]
    ***REMOVED***

    T.copy(DK, destPos)
    destPos += hLen
  ***REMOVED***

  return DK
***REMOVED***

module.exports = pbkdf2
