// escope/lib/reference.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
***REMOVED***);

var _createClass = function () ***REMOVED*** function defineProperties(target, props) ***REMOVED*** for (var i = 0; i < props.length; i++) ***REMOVED*** var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); ***REMOVED*** ***REMOVED*** return function (Constructor, protoProps, staticProps) ***REMOVED*** if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; ***REMOVED***; ***REMOVED***();

function _classCallCheck(instance, Constructor) ***REMOVED*** if (!(instance instanceof Constructor)) ***REMOVED*** throw new TypeError("Cannot call a class as a function"); ***REMOVED*** ***REMOVED***

/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

var READ = 0x1;
var WRITE = 0x2;
var RW = READ | WRITE;

/**
 * A Reference represents a single occurrence of an identifier in code.
 * @class Reference
 */

var Reference = function () ***REMOVED***
  function Reference(ident, scope, flag, writeExpr, maybeImplicitGlobal, partial, init) ***REMOVED***
    _classCallCheck(this, Reference);

    /**
     * Identifier syntax node.
     * @member ***REMOVED***esprima#Identifier***REMOVED*** Reference#identifier
     */
    this.identifier = ident;
    /**
     * Reference to the enclosing Scope.
     * @member ***REMOVED***Scope***REMOVED*** Reference#from
     */
    this.from = scope;
    /**
     * Whether the reference comes from a dynamic scope (such as 'eval',
     * 'with', etc.), and may be trapped by dynamic scopes.
     * @member ***REMOVED***boolean***REMOVED*** Reference#tainted
     */
    this.tainted = false;
    /**
     * The variable this reference is resolved with.
     * @member ***REMOVED***Variable***REMOVED*** Reference#resolved
     */
    this.resolved = null;
    /**
     * The read-write mode of the reference. (Value is one of ***REMOVED***@link
     * Reference.READ***REMOVED***, ***REMOVED***@link Reference.RW***REMOVED***, ***REMOVED***@link Reference.WRITE***REMOVED***).
     * @member ***REMOVED***number***REMOVED*** Reference#flag
     * @private
     */
    this.flag = flag;
    if (this.isWrite()) ***REMOVED***
      /**
       * If reference is writeable, this is the tree being written to it.
       * @member ***REMOVED***esprima#Node***REMOVED*** Reference#writeExpr
       */
      this.writeExpr = writeExpr;
      /**
       * Whether the Reference might refer to a partial value of writeExpr.
       * @member ***REMOVED***boolean***REMOVED*** Reference#partial
       */
      this.partial = partial;
      /**
       * Whether the Reference is to write of initialization.
       * @member ***REMOVED***boolean***REMOVED*** Reference#init
       */
      this.init = init;
    ***REMOVED***
    this.__maybeImplicitGlobal = maybeImplicitGlobal;
  ***REMOVED***

  /**
   * Whether the reference is static.
   * @method Reference#isStatic
   * @return ***REMOVED***boolean***REMOVED***
   */


  _createClass(Reference, [***REMOVED***
    key: "isStatic",
    value: function isStatic() ***REMOVED***
      return !this.tainted && this.resolved && this.resolved.scope.isStatic();
    ***REMOVED***

    /**
     * Whether the reference is writeable.
     * @method Reference#isWrite
     * @return ***REMOVED***boolean***REMOVED***
     */

  ***REMOVED***, ***REMOVED***
    key: "isWrite",
    value: function isWrite() ***REMOVED***
      return !!(this.flag & Reference.WRITE);
    ***REMOVED***

    /**
     * Whether the reference is readable.
     * @method Reference#isRead
     * @return ***REMOVED***boolean***REMOVED***
     */

  ***REMOVED***, ***REMOVED***
    key: "isRead",
    value: function isRead() ***REMOVED***
      return !!(this.flag & Reference.READ);
    ***REMOVED***

    /**
     * Whether the reference is read-only.
     * @method Reference#isReadOnly
     * @return ***REMOVED***boolean***REMOVED***
     */

  ***REMOVED***, ***REMOVED***
    key: "isReadOnly",
    value: function isReadOnly() ***REMOVED***
      return this.flag === Reference.READ;
    ***REMOVED***

    /**
     * Whether the reference is write-only.
     * @method Reference#isWriteOnly
     * @return ***REMOVED***boolean***REMOVED***
     */

  ***REMOVED***, ***REMOVED***
    key: "isWriteOnly",
    value: function isWriteOnly() ***REMOVED***
      return this.flag === Reference.WRITE;
    ***REMOVED***

    /**
     * Whether the reference is read-write.
     * @method Reference#isReadWrite
     * @return ***REMOVED***boolean***REMOVED***
     */

  ***REMOVED***, ***REMOVED***
    key: "isReadWrite",
    value: function isReadWrite() ***REMOVED***
      return this.flag === Reference.RW;
    ***REMOVED***
  ***REMOVED***]);

  return Reference;
***REMOVED***();

/**
 * @constant Reference.READ
 * @private
 */


exports.default = Reference;
Reference.READ = READ;
/**
 * @constant Reference.WRITE
 * @private
 */
Reference.WRITE = WRITE;
/**
 * @constant Reference.RW
 * @private
 */
Reference.RW = RW;

/* vim: set sw=4 ts=4 et tw=80 : */
