// webpack/lib/ContextReplacementPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const path = require("path");
const ContextElementDependency = require("./dependencies/ContextElementDependency");

class ContextReplacementPlugin ***REMOVED***
	constructor(resourceRegExp, newContentResource, newContentRecursive, newContentRegExp) ***REMOVED***
		this.resourceRegExp = resourceRegExp;

		if(typeof newContentResource === "function") ***REMOVED***
			this.newContentCallback = newContentResource;
		***REMOVED*** else if(typeof newContentResource === "string" && typeof newContentRecursive === "object") ***REMOVED***
			this.newContentResource = newContentResource;
			this.newContentCreateContextMap = (fs, callback) => ***REMOVED***
				callback(null, newContentRecursive);
			***REMOVED***;
		***REMOVED*** else if(typeof newContentResource === "string" && typeof newContentRecursive === "function") ***REMOVED***
			this.newContentResource = newContentResource;
			this.newContentCreateContextMap = newContentRecursive;
		***REMOVED*** else ***REMOVED***
			if(typeof newContentResource !== "string") ***REMOVED***
				newContentRegExp = newContentRecursive;
				newContentRecursive = newContentResource;
				newContentResource = undefined;
			***REMOVED***
			if(typeof newContentRecursive !== "boolean") ***REMOVED***
				newContentRegExp = newContentRecursive;
				newContentRecursive = undefined;
			***REMOVED***
			this.newContentResource = newContentResource;
			this.newContentRecursive = newContentRecursive;
			this.newContentRegExp = newContentRegExp;
		***REMOVED***
	***REMOVED***

	apply(compiler) ***REMOVED***
		const resourceRegExp = this.resourceRegExp;
		const newContentCallback = this.newContentCallback;
		const newContentResource = this.newContentResource;
		const newContentRecursive = this.newContentRecursive;
		const newContentRegExp = this.newContentRegExp;
		const newContentCreateContextMap = this.newContentCreateContextMap;

		compiler.plugin("context-module-factory", (cmf) => ***REMOVED***
			cmf.plugin("before-resolve", (result, callback) => ***REMOVED***
				if(!result) return callback();
				if(resourceRegExp.test(result.request)) ***REMOVED***
					if(typeof newContentResource !== "undefined")
						result.request = newContentResource;
					if(typeof newContentRecursive !== "undefined")
						result.recursive = newContentRecursive;
					if(typeof newContentRegExp !== "undefined")
						result.regExp = newContentRegExp;
					if(typeof newContentCallback === "function") ***REMOVED***
						newContentCallback(result);
					***REMOVED*** else ***REMOVED***
						result.dependencies.forEach((d) => ***REMOVED***
							if(d.critical)
								d.critical = false;
						***REMOVED***);
					***REMOVED***
				***REMOVED***
				return callback(null, result);
			***REMOVED***);
			cmf.plugin("after-resolve", (result, callback) => ***REMOVED***
				if(!result) return callback();
				if(resourceRegExp.test(result.resource)) ***REMOVED***
					if(typeof newContentResource !== "undefined")
						result.resource = path.resolve(result.resource, newContentResource);
					if(typeof newContentRecursive !== "undefined")
						result.recursive = newContentRecursive;
					if(typeof newContentRegExp !== "undefined")
						result.regExp = newContentRegExp;
					if(typeof newContentCreateContextMap === "function")
						result.resolveDependencies = createResolveDependenciesFromContextMap(newContentCreateContextMap);
					if(typeof newContentCallback === "function") ***REMOVED***
						const origResource = result.resource;
						newContentCallback(result);
						if(result.resource !== origResource) ***REMOVED***
							result.resource = path.resolve(origResource, result.resource);
						***REMOVED***
					***REMOVED*** else ***REMOVED***
						result.dependencies.forEach((d) => ***REMOVED***
							if(d.critical)
								d.critical = false;
						***REMOVED***);
					***REMOVED***
				***REMOVED***
				return callback(null, result);
			***REMOVED***);
		***REMOVED***);
	***REMOVED***
***REMOVED***

const createResolveDependenciesFromContextMap = (createContextMap) => ***REMOVED***
	return function resolveDependenciesFromContextMap(fs, resource, recursive, regExp, callback) ***REMOVED***
		createContextMap(fs, (err, map) => ***REMOVED***
			if(err) return callback(err);
			const dependencies = Object.keys(map).map((key) => ***REMOVED***
				return new ContextElementDependency(map[key], key);
			***REMOVED***);
			callback(null, dependencies);
		***REMOVED***);
	***REMOVED***;
***REMOVED***;

module.exports = ContextReplacementPlugin;
