// webpack/lib/FlagDependencyUsagePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

class FlagDependencyUsagePlugin ***REMOVED***
	apply(compiler) ***REMOVED***
		compiler.plugin("compilation", compilation => ***REMOVED***
			compilation.plugin("optimize-modules-advanced", modules => ***REMOVED***

				modules.forEach(module => module.used = false);

				const queue = [];
				compilation.chunks.forEach(chunk => ***REMOVED***
					if(chunk.entryModule) ***REMOVED***
						processModule(chunk.entryModule, true);
					***REMOVED***
				***REMOVED***);

				while(queue.length) ***REMOVED***
					const queueItem = queue.pop();
					processDependenciesBlock(queueItem[0], queueItem[1]);
				***REMOVED***

				function processModule(module, usedExports) ***REMOVED***
					module.used = true;
					if(module.usedExports === true)
						return;
					else if(usedExports === true)
						module.usedExports = true;
					else if(Array.isArray(usedExports)) ***REMOVED***
						const old = module.usedExports ? module.usedExports.length : -1;
						module.usedExports = addToSet(module.usedExports || [], usedExports);
						if(module.usedExports.length === old)
							return;
					***REMOVED*** else if(Array.isArray(module.usedExports))
						return;
					else
						module.usedExports = false;

					queue.push([module, module.usedExports]);
				***REMOVED***

				function processDependenciesBlock(depBlock, usedExports) ***REMOVED***
					depBlock.dependencies.forEach(dep => processDependency(dep));
					depBlock.variables.forEach(variable => variable.dependencies.forEach(dep => processDependency(dep)));
					depBlock.blocks.forEach(block => queue.push([block, usedExports]));
				***REMOVED***

				function processDependency(dep) ***REMOVED***
					const reference = dep.getReference && dep.getReference();
					if(!reference) return;
					const module = reference.module;
					const importedNames = reference.importedNames;
					const oldUsed = module.used;
					const oldUsedExports = module.usedExports;
					if(!oldUsed || (importedNames && (!oldUsedExports || !isSubset(oldUsedExports, importedNames)))) ***REMOVED***
						processModule(module, importedNames);
					***REMOVED***
				***REMOVED***

			***REMOVED***);

			function addToSet(a, b) ***REMOVED***
				b.forEach(item => ***REMOVED***
					if(a.indexOf(item) < 0)
						a.push(item);
				***REMOVED***);
				return a;
			***REMOVED***

			function isSubset(biggerSet, subset) ***REMOVED***
				if(biggerSet === true) return true;
				if(subset === true) return false;
				return subset.every(item => biggerSet.indexOf(item) >= 0);
			***REMOVED***
		***REMOVED***);
	***REMOVED***
***REMOVED***
module.exports = FlagDependencyUsagePlugin;
