// webpack/lib/HashedModuleIdsPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";
const createHash = require("crypto").createHash;

class HashedModuleIdsPlugin ***REMOVED***
	constructor(options) ***REMOVED***
		this.options = Object.assign(***REMOVED***
			hashFunction: "md5",
			hashDigest: "base64",
			hashDigestLength: 4
		***REMOVED***, options);
	***REMOVED***

	apply(compiler) ***REMOVED***
		const options = this.options;
		compiler.plugin("compilation", (compilation) => ***REMOVED***
			const usedIds = new Set();
			compilation.plugin("before-module-ids", (modules) => ***REMOVED***
				modules.forEach((module) => ***REMOVED***
					if(module.id === null && module.libIdent) ***REMOVED***
						const id = module.libIdent(***REMOVED***
							context: this.options.context || compiler.options.context
						***REMOVED***);
						const hash = createHash(options.hashFunction);
						hash.update(id);
						const hashId = hash.digest(options.hashDigest);
						let len = options.hashDigestLength;
						while(usedIds.has(hashId.substr(0, len)))
							len++;
						module.id = hashId.substr(0, len);
						usedIds.add(module.id);
					***REMOVED***
				***REMOVED***);
			***REMOVED***);
		***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports = HashedModuleIdsPlugin;
