// webpack/lib/ModuleBuildError.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const WebpackError = require("./WebpackError");
const cutOffLoaderExecution = require("./ErrorHelpers").cutOffLoaderExecution;

class ModuleBuildError extends WebpackError ***REMOVED***
	constructor(module, err) ***REMOVED***
		super();

		this.name = "ModuleBuildError";
		this.message = "Module build failed: ";
		if(err !== null && typeof err === "object") ***REMOVED***
			if(typeof err.stack === "string" && err.stack) ***REMOVED***
				var stack = cutOffLoaderExecution(err.stack);
				if(!err.hideStack) ***REMOVED***
					this.message += stack;
				***REMOVED*** else ***REMOVED***
					this.details = stack;
					if(typeof err.message === "string" && err.message) ***REMOVED***
						this.message += err.message;
					***REMOVED*** else ***REMOVED***
						this.message += err;
					***REMOVED***
				***REMOVED***
			***REMOVED*** else if(typeof err.message === "string" && err.message) ***REMOVED***
				this.message += err.message;
			***REMOVED*** else ***REMOVED***
				this.message += err;
			***REMOVED***
		***REMOVED***
		this.module = module;
		this.error = err;

		Error.captureStackTrace(this, this.constructor);
	***REMOVED***
***REMOVED***

module.exports = ModuleBuildError;
