// webpack/lib/ModuleFilenameHelpers.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const ModuleFilenameHelpers = exports;

ModuleFilenameHelpers.ALL_LOADERS_RESOURCE = "[all-loaders][resource]";
ModuleFilenameHelpers.REGEXP_ALL_LOADERS_RESOURCE = /\[all-?loaders\]\[resource\]/gi;
ModuleFilenameHelpers.LOADERS_RESOURCE = "[loaders][resource]";
ModuleFilenameHelpers.REGEXP_LOADERS_RESOURCE = /\[loaders\]\[resource\]/gi;
ModuleFilenameHelpers.RESOURCE = "[resource]";
ModuleFilenameHelpers.REGEXP_RESOURCE = /\[resource\]/gi;
ModuleFilenameHelpers.ABSOLUTE_RESOURCE_PATH = "[absolute-resource-path]";
ModuleFilenameHelpers.REGEXP_ABSOLUTE_RESOURCE_PATH = /\[abs(olute)?-?resource-?path\]/gi;
ModuleFilenameHelpers.RESOURCE_PATH = "[resource-path]";
ModuleFilenameHelpers.REGEXP_RESOURCE_PATH = /\[resource-?path\]/gi;
ModuleFilenameHelpers.ALL_LOADERS = "[all-loaders]";
ModuleFilenameHelpers.REGEXP_ALL_LOADERS = /\[all-?loaders\]/gi;
ModuleFilenameHelpers.LOADERS = "[loaders]";
ModuleFilenameHelpers.REGEXP_LOADERS = /\[loaders\]/gi;
ModuleFilenameHelpers.QUERY = "[query]";
ModuleFilenameHelpers.REGEXP_QUERY = /\[query\]/gi;
ModuleFilenameHelpers.ID = "[id]";
ModuleFilenameHelpers.REGEXP_ID = /\[id\]/gi;
ModuleFilenameHelpers.HASH = "[hash]";
ModuleFilenameHelpers.REGEXP_HASH = /\[hash\]/gi;

function getAfter(str, token) ***REMOVED***
	const idx = str.indexOf(token);
	return idx < 0 ? "" : str.substr(idx);
***REMOVED***

function getBefore(str, token) ***REMOVED***
	const idx = str.lastIndexOf(token);
	return idx < 0 ? "" : str.substr(0, idx);
***REMOVED***

function getHash(str) ***REMOVED***
	const hash = require("crypto").createHash("md5");
	hash.update(str);
	return hash.digest("hex").substr(0, 4);
***REMOVED***

function asRegExp(test) ***REMOVED***
	if(typeof test === "string") test = new RegExp("^" + test.replace(/[-[\]***REMOVED******REMOVED***()*+?.,\\^$|#\s]/g, "\\$&"));
	return test;
***REMOVED***

ModuleFilenameHelpers.createFilename = function createFilename(module, moduleFilenameTemplate, requestShortener) ***REMOVED***
	let absoluteResourcePath;
	let hash;
	let identifier;
	let moduleId;
	let shortIdentifier;
	if(module === undefined) module = "";
	if(typeof module === "string") ***REMOVED***
		shortIdentifier = requestShortener.shorten(module);
		identifier = shortIdentifier;
		moduleId = "";
		absoluteResourcePath = module.split("!").pop();
		hash = getHash(identifier);
	***REMOVED*** else ***REMOVED***
		shortIdentifier = module.readableIdentifier(requestShortener);
		identifier = requestShortener.shorten(module.identifier());
		moduleId = module.id;
		absoluteResourcePath = module.identifier().split("!").pop();
		hash = getHash(identifier);
	***REMOVED***
	const resource = shortIdentifier.split("!").pop();
	const loaders = getBefore(shortIdentifier, "!");
	const allLoaders = getBefore(identifier, "!");
	const query = getAfter(resource, "?");
	const resourcePath = resource.substr(0, resource.length - query.length);
	if(typeof moduleFilenameTemplate === "function") ***REMOVED***
		return moduleFilenameTemplate(***REMOVED***
			identifier: identifier,
			shortIdentifier: shortIdentifier,
			resource: resource,
			resourcePath: resourcePath,
			absoluteResourcePath: absoluteResourcePath,
			allLoaders: allLoaders,
			query: query,
			moduleId: moduleId,
			hash: hash
		***REMOVED***);
	***REMOVED***
	return moduleFilenameTemplate
		.replace(ModuleFilenameHelpers.REGEXP_ALL_LOADERS_RESOURCE, identifier)
		.replace(ModuleFilenameHelpers.REGEXP_LOADERS_RESOURCE, shortIdentifier)
		.replace(ModuleFilenameHelpers.REGEXP_RESOURCE, resource)
		.replace(ModuleFilenameHelpers.REGEXP_RESOURCE_PATH, resourcePath)
		.replace(ModuleFilenameHelpers.REGEXP_ABSOLUTE_RESOURCE_PATH, absoluteResourcePath)
		.replace(ModuleFilenameHelpers.REGEXP_ALL_LOADERS, allLoaders)
		.replace(ModuleFilenameHelpers.REGEXP_LOADERS, loaders)
		.replace(ModuleFilenameHelpers.REGEXP_QUERY, query)
		.replace(ModuleFilenameHelpers.REGEXP_ID, moduleId)
		.replace(ModuleFilenameHelpers.REGEXP_HASH, hash);
***REMOVED***;

ModuleFilenameHelpers.createFooter = function createFooter(module, requestShortener) ***REMOVED***
	if(!module) module = "";
	if(typeof module === "string") ***REMOVED***
		return [
			"// WEBPACK FOOTER //",
			`// $***REMOVED***requestShortener.shorten(module)***REMOVED***`
		].join("\n");
	***REMOVED*** else ***REMOVED***
		return [
			"//////////////////",
			"// WEBPACK FOOTER",
			`// $***REMOVED***module.readableIdentifier(requestShortener)***REMOVED***`,
			`// module id = $***REMOVED***module.id***REMOVED***`,
			`// module chunks = $***REMOVED***module.mapChunks(c => c.id).join(" ")***REMOVED***`
		].join("\n");
	***REMOVED***
***REMOVED***;

ModuleFilenameHelpers.replaceDuplicates = function replaceDuplicates(array, fn, comparator) ***REMOVED***
	const countMap = Object.create(null);
	const posMap = Object.create(null);
	array.forEach((item, idx) => ***REMOVED***
		countMap[item] = (countMap[item] || []);
		countMap[item].push(idx);
		posMap[item] = 0;
	***REMOVED***);
	if(comparator) ***REMOVED***
		Object.keys(countMap).forEach(item => ***REMOVED***
			countMap[item].sort(comparator);
		***REMOVED***);
	***REMOVED***
	return array.map((item, i) => ***REMOVED***
		if(countMap[item].length > 1) ***REMOVED***
			if(comparator && countMap[item][0] === i)
				return item;
			return fn(item, i, posMap[item]++);
		***REMOVED*** else return item;
	***REMOVED***);
***REMOVED***;

ModuleFilenameHelpers.matchPart = function matchPart(str, test) ***REMOVED***
	if(!test) return true;
	test = asRegExp(test);
	if(Array.isArray(test)) ***REMOVED***
		return test.map(asRegExp).filter(function(regExp) ***REMOVED***
			return regExp.test(str);
		***REMOVED***).length > 0;
	***REMOVED*** else ***REMOVED***
		return test.test(str);
	***REMOVED***
***REMOVED***;

ModuleFilenameHelpers.matchObject = function matchObject(obj, str) ***REMOVED***
	if(obj.test)
		if(!ModuleFilenameHelpers.matchPart(str, obj.test)) return false;
	if(obj.include)
		if(!ModuleFilenameHelpers.matchPart(str, obj.include)) return false;
	if(obj.exclude)
		if(ModuleFilenameHelpers.matchPart(str, obj.exclude)) return false;
	return true;
***REMOVED***;
