// webpack/lib/ParserHelpers.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";
const path = require("path");

const BasicEvaluatedExpression = require("./BasicEvaluatedExpression");
const ConstDependency = require("./dependencies/ConstDependency");
const UnsupportedFeatureWarning = require("./UnsupportedFeatureWarning");

const ParserHelpers = exports;

ParserHelpers.addParsedVariableToModule = function(parser, name, expression) ***REMOVED***
	if(!parser.state.current.addVariable) return false;
	var deps = [];
	parser.parse(expression, ***REMOVED***
		current: ***REMOVED***
			addDependency: function(dep) ***REMOVED***
				dep.userRequest = name;
				deps.push(dep);
			***REMOVED***
		***REMOVED***,
		module: parser.state.module
	***REMOVED***);
	parser.state.current.addVariable(name, expression, deps);
	return true;
***REMOVED***;

ParserHelpers.requireFileAsExpression = function(context, pathToModule) ***REMOVED***
	var moduleJsPath = path.relative(context, pathToModule);
	if(!/^[A-Z]:/i.test(moduleJsPath)) ***REMOVED***
		moduleJsPath = "./" + moduleJsPath.replace(/\\/g, "/");
	***REMOVED***
	return "require(" + JSON.stringify(moduleJsPath) + ")";
***REMOVED***;

ParserHelpers.toConstantDependency = function(value) ***REMOVED***
	return function constDependency(expr) ***REMOVED***
		var dep = new ConstDependency(value, expr.range);
		dep.loc = expr.loc;
		this.state.current.addDependency(dep);
		return true;
	***REMOVED***;
***REMOVED***;

ParserHelpers.evaluateToString = function(value) ***REMOVED***
	return function stringExpression(expr) ***REMOVED***
		return new BasicEvaluatedExpression().setString(value).setRange(expr.range);
	***REMOVED***;
***REMOVED***;

ParserHelpers.evaluateToBoolean = function(value) ***REMOVED***
	return function booleanExpression(expr) ***REMOVED***
		return new BasicEvaluatedExpression().setBoolean(value).setRange(expr.range);
	***REMOVED***;
***REMOVED***;

ParserHelpers.evaluateToIdentifier = function(identifier, truthy) ***REMOVED***
	return function identifierExpression(expr) ***REMOVED***
		let evex = new BasicEvaluatedExpression().setIdentifier(identifier).setRange(expr.range);
		if(truthy === true) evex = evex.setTruthy();
		else if(truthy === false) evex = evex.setFalsy();
		return evex;
	***REMOVED***;
***REMOVED***;

ParserHelpers.expressionIsUnsupported = function(message) ***REMOVED***
	return function unsupportedExpression(expr) ***REMOVED***
		var dep = new ConstDependency("(void 0)", expr.range);
		dep.loc = expr.loc;
		this.state.current.addDependency(dep);
		if(!this.state.module) return;
		this.state.module.warnings.push(new UnsupportedFeatureWarning(this.state.module, message));
		return true;
	***REMOVED***;
***REMOVED***;

ParserHelpers.skipTraversal = function skipTraversal() ***REMOVED***
	return true;
***REMOVED***;

ParserHelpers.approve = function approve() ***REMOVED***
	return true;
***REMOVED***;
