// webpack/lib/ProgressPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

class ProgressPlugin ***REMOVED***

	constructor(options) ***REMOVED***
		if(typeof options === "function") ***REMOVED***
			options = ***REMOVED***
				handler: options
			***REMOVED***;
		***REMOVED***
		options = options || ***REMOVED******REMOVED***;
		this.profile = options.profile;
		this.handler = options.handler;
	***REMOVED***

	apply(compiler) ***REMOVED***
		const handler = this.handler || defaultHandler;
		const profile = this.profile;
		if(compiler.compilers) ***REMOVED***
			const states = new Array(compiler.compilers.length);
			compiler.compilers.forEach(function(compiler, idx) ***REMOVED***
				compiler.apply(new ProgressPlugin(function(p, msg) ***REMOVED***
					states[idx] = Array.prototype.slice.apply(arguments);
					handler.apply(null, [
						states.map(state => state && state[0] || 0).reduce((a, b) => a + b) / states.length,
						`[$***REMOVED***idx***REMOVED***] $***REMOVED***msg***REMOVED***`
					].concat(Array.prototype.slice.call(arguments, 2)));
				***REMOVED***));
			***REMOVED***);
		***REMOVED*** else ***REMOVED***
			let lastModulesCount = 0;
			let moduleCount = 500;
			let doneModules = 0;
			const activeModules = [];

			const update = function update(module) ***REMOVED***
				handler(
					0.1 + (doneModules / Math.max(lastModulesCount, moduleCount)) * 0.6,
					"building modules",
					`$***REMOVED***doneModules***REMOVED***/$***REMOVED***moduleCount***REMOVED*** modules`,
					`$***REMOVED***activeModules.length***REMOVED*** active`,
					activeModules[activeModules.length - 1]
				);
			***REMOVED***;

			const moduleDone = function moduleDone(module) ***REMOVED***
				doneModules++;
				const ident = module.identifier();
				if(ident) ***REMOVED***
					const idx = activeModules.indexOf(ident);
					if(idx >= 0) activeModules.splice(idx, 1);
				***REMOVED***
				update();
			***REMOVED***;
			compiler.plugin("compilation", function(compilation) ***REMOVED***
				if(compilation.compiler.isChild()) return;
				lastModulesCount = moduleCount;
				moduleCount = 0;
				doneModules = 0;
				handler(0, "compiling");
				compilation.plugin("build-module", function(module) ***REMOVED***
					moduleCount++;
					const ident = module.identifier();
					if(ident) ***REMOVED***
						activeModules.push(ident);
					***REMOVED***
					update();
				***REMOVED***);
				compilation.plugin("failed-module", moduleDone);
				compilation.plugin("succeed-module", moduleDone);
				const syncHooks = ***REMOVED***
					"seal": [0.71, "sealing"],
					"optimize": [0.72, "optimizing"],
					"optimize-modules-basic": [0.73, "basic module optimization"],
					"optimize-modules": [0.74, "module optimization"],
					"optimize-modules-advanced": [0.75, "advanced module optimization"],
					"optimize-chunks-basic": [0.76, "basic chunk optimization"],
					"optimize-chunks": [0.77, "chunk optimization"],
					"optimize-chunks-advanced": [0.78, "advanced chunk optimization"],
					// optimize-tree
					"optimize-chunk-modules": [0.80, "chunk modules optimization"],
					"optimize-chunk-modules-advanced": [0.81, "advanced chunk modules optimization"],
					"revive-modules": [0.82, "module reviving"],
					"optimize-module-order": [0.83, "module order optimization"],
					"optimize-module-ids": [0.84, "module id optimization"],
					"revive-chunks": [0.85, "chunk reviving"],
					"optimize-chunk-order": [0.86, "chunk order optimization"],
					"optimize-chunk-ids": [0.87, "chunk id optimization"],
					"before-hash": [0.88, "hashing"],
					"before-module-assets": [0.89, "module assets processing"],
					"before-chunk-assets": [0.90, "chunk assets processing"],
					"additional-chunk-assets": [0.91, "additional chunk assets processing"],
					"record": [0.92, "recording"]
				***REMOVED***;
				Object.keys(syncHooks).forEach(name => ***REMOVED***
					let pass = 0;
					const settings = syncHooks[name];
					compilation.plugin(name, () => ***REMOVED***
						if(pass++ > 0)
							handler(settings[0], settings[1], `pass $***REMOVED***pass***REMOVED***`);
						else
							handler(settings[0], settings[1]);
					***REMOVED***);
				***REMOVED***);
				compilation.plugin("optimize-tree", (chunks, modules, callback) => ***REMOVED***
					handler(0.79, "module and chunk tree optimization");
					callback();
				***REMOVED***);
				compilation.plugin("additional-assets", callback => ***REMOVED***
					handler(0.91, "additional asset processing");
					callback();
				***REMOVED***);
				compilation.plugin("optimize-chunk-assets", (chunks, callback) => ***REMOVED***
					handler(0.92, "chunk asset optimization");
					callback();
				***REMOVED***);
				compilation.plugin("optimize-assets", (assets, callback) => ***REMOVED***
					handler(0.94, "asset optimization");
					callback();
				***REMOVED***);
			***REMOVED***);
			compiler.plugin("emit", (compilation, callback) => ***REMOVED***
				handler(0.95, "emitting");
				callback();
			***REMOVED***);
			compiler.plugin("done", () => ***REMOVED***
				handler(1, "");
			***REMOVED***);
		***REMOVED***

		let lineCaretPosition = 0,
			lastState, lastStateTime;

		function defaultHandler(percentage, msg) ***REMOVED***
			let state = msg;
			const details = Array.prototype.slice.call(arguments, 2);
			if(percentage < 1) ***REMOVED***
				percentage = Math.floor(percentage * 100);
				msg = `$***REMOVED***percentage***REMOVED***% $***REMOVED***msg***REMOVED***`;
				if(percentage < 100) ***REMOVED***
					msg = ` $***REMOVED***msg***REMOVED***`;
				***REMOVED***
				if(percentage < 10) ***REMOVED***
					msg = ` $***REMOVED***msg***REMOVED***`;
				***REMOVED***
				details.forEach(detail => ***REMOVED***
					if(!detail) return;
					if(detail.length > 40) ***REMOVED***
						detail = `...$***REMOVED***detail.substr(detail.length - 37)***REMOVED***`;
					***REMOVED***
					msg += ` $***REMOVED***detail***REMOVED***`;
				***REMOVED***);
			***REMOVED***
			if(profile) ***REMOVED***
				state = state.replace(/^\d+\/\d+\s+/, "");
				if(percentage === 0) ***REMOVED***
					lastState = null;
					lastStateTime = Date.now();
				***REMOVED*** else if(state !== lastState || percentage === 1) ***REMOVED***
					const now = Date.now();
					if(lastState) ***REMOVED***
						const stateMsg = `$***REMOVED***now - lastStateTime***REMOVED***ms $***REMOVED***lastState***REMOVED***`;
						goToLineStart(stateMsg);
						process.stderr.write(stateMsg + "\n");
						lineCaretPosition = 0;
					***REMOVED***
					lastState = state;
					lastStateTime = now;
				***REMOVED***
			***REMOVED***
			goToLineStart(msg);
			process.stderr.write(msg);
		***REMOVED***

		function goToLineStart(nextMessage) ***REMOVED***
			let str = "";
			for(; lineCaretPosition > nextMessage.length; lineCaretPosition--) ***REMOVED***
				str += "\b \b";
			***REMOVED***
			for(var i = 0; i < lineCaretPosition; i++) ***REMOVED***
				str += "\b";
			***REMOVED***
			lineCaretPosition = nextMessage.length;
			if(str) process.stderr.write(str);
		***REMOVED***
	***REMOVED***
***REMOVED***
module.exports = ProgressPlugin;
