// webpack/lib/RawModule.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Module = require("./Module");
const OriginalSource = require("webpack-sources").OriginalSource;
const RawSource = require("webpack-sources").RawSource;

module.exports = class RawModule extends Module ***REMOVED***

	constructor(source, identifier, readableIdentifier) ***REMOVED***
		super();
		this.sourceStr = source;
		this.identifierStr = identifier || this.sourceStr;
		this.readableIdentifierStr = readableIdentifier || this.identifierStr;
		this.cacheable = true;
		this.built = false;
	***REMOVED***

	identifier() ***REMOVED***
		return this.identifierStr;
	***REMOVED***

	size() ***REMOVED***
		return this.sourceStr.length;
	***REMOVED***

	readableIdentifier(requestShortener) ***REMOVED***
		return requestShortener.shorten(this.readableIdentifierStr);
	***REMOVED***

	needRebuild() ***REMOVED***
		return false;
	***REMOVED***

	build(options, compilations, resolver, fs, callback) ***REMOVED***
		this.builtTime = Date.now();
		callback();
	***REMOVED***

	source() ***REMOVED***
		if(this.useSourceMap)
			return new OriginalSource(this.sourceStr, this.identifier());
		else
			return new RawSource(this.sourceStr);
	***REMOVED***

	updateHash(hash) ***REMOVED***
		hash.update(this.sourceStr);
		super.updateHash(hash);
	***REMOVED***
***REMOVED***;
