// webpack/lib/SourceMapDevToolPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const path = require("path");
const crypto = require("crypto");
const RequestShortener = require("./RequestShortener");
const ConcatSource = require("webpack-sources").ConcatSource;
const RawSource = require("webpack-sources").RawSource;
const ModuleFilenameHelpers = require("./ModuleFilenameHelpers");
const SourceMapDevToolModuleOptionsPlugin = require("./SourceMapDevToolModuleOptionsPlugin");

const basename = (name) => ***REMOVED***
	if(name.indexOf("/") < 0) return name;
	return name.substr(name.lastIndexOf("/") + 1);
***REMOVED***;

function getTaskForFile(file, chunk, options, compilation) ***REMOVED***
	const asset = compilation.assets[file];
	if(asset.__SourceMapDevToolFile === file && asset.__SourceMapDevToolData) ***REMOVED***
		const data = asset.__SourceMapDevToolData;
		for(const cachedFile in data) ***REMOVED***
			compilation.assets[cachedFile] = data[cachedFile];
			if(cachedFile !== file)
				chunk.files.push(cachedFile);
		***REMOVED***
		return;
	***REMOVED***
	let source, sourceMap;
	if(asset.sourceAndMap) ***REMOVED***
		const sourceAndMap = asset.sourceAndMap(options);
		sourceMap = sourceAndMap.map;
		source = sourceAndMap.source;
	***REMOVED*** else ***REMOVED***
		sourceMap = asset.map(options);
		source = asset.source();
	***REMOVED***
	if(sourceMap) ***REMOVED***
		return ***REMOVED***
			chunk,
			file,
			asset,
			source,
			sourceMap,
			modules: undefined
		***REMOVED***;
	***REMOVED***
***REMOVED***

class SourceMapDevToolPlugin ***REMOVED***
	constructor(options) ***REMOVED***
		if(arguments.length > 1)
			throw new Error("SourceMapDevToolPlugin only takes one argument (pass an options object)");
		// TODO: remove in webpack 3
		if(typeof options === "string") ***REMOVED***
			options = ***REMOVED***
				sourceMapFilename: options
			***REMOVED***;
		***REMOVED***
		if(!options) options = ***REMOVED******REMOVED***;
		this.sourceMapFilename = options.filename;
		this.sourceMappingURLComment = options.append === false ? false : options.append || "\n//# sourceMappingURL=[url]";
		this.moduleFilenameTemplate = options.moduleFilenameTemplate || "webpack:///[resourcePath]";
		this.fallbackModuleFilenameTemplate = options.fallbackModuleFilenameTemplate || "webpack:///[resourcePath]?[hash]";
		this.options = options;
	***REMOVED***

	apply(compiler) ***REMOVED***
		const sourceMapFilename = this.sourceMapFilename;
		const sourceMappingURLComment = this.sourceMappingURLComment;
		const moduleFilenameTemplate = this.moduleFilenameTemplate;
		const fallbackModuleFilenameTemplate = this.fallbackModuleFilenameTemplate;
		const requestShortener = new RequestShortener(compiler.context);
		const options = this.options;
		options.test = options.test || /\.(js|css)($|\?)/i;

		const matchObject = ModuleFilenameHelpers.matchObject.bind(undefined, options);

		compiler.plugin("compilation", compilation => ***REMOVED***
			new SourceMapDevToolModuleOptionsPlugin(options).apply(compilation);

			compilation.plugin("after-optimize-chunk-assets", function(chunks) ***REMOVED***
				const moduleToSourceNameMapping = new Map();
				const tasks = [];

				chunks.forEach(function(chunk) ***REMOVED***
					chunk.files.forEach(file => ***REMOVED***
						if(matchObject(file)) ***REMOVED***
							const task = getTaskForFile(file, chunk, options, compilation);

							if(task) ***REMOVED***
								const modules = task.sourceMap.sources.map(source => ***REMOVED***
									const module = compilation.findModule(source);
									return module || source;
								***REMOVED***);

								for(let idx = 0; idx < modules.length; idx++) ***REMOVED***
									const module = modules[idx];
									if(!moduleToSourceNameMapping.get(module)) ***REMOVED***
										moduleToSourceNameMapping.set(module, ModuleFilenameHelpers.createFilename(module, moduleFilenameTemplate, requestShortener));
									***REMOVED***
								***REMOVED***

								task.modules = modules;

								tasks.push(task);
							***REMOVED***
						***REMOVED***
					***REMOVED***);
				***REMOVED***);

				const usedNamesSet = new Set(moduleToSourceNameMapping.values());
				const conflictDetectionSet = new Set();

				// all modules in defined order (longest identifier first)
				const allModules = Array.from(moduleToSourceNameMapping.keys()).sort((a, b) => ***REMOVED***
					const ai = typeof a === "string" ? a : a.identifier();
					const bi = typeof b === "string" ? b : b.identifier();
					return ai.length - bi.length;
				***REMOVED***);

				// find modules with conflicting source names
				for(let idx = 0; idx < allModules.length; idx++) ***REMOVED***
					const module = allModules[idx];
					let sourceName = moduleToSourceNameMapping.get(module);
					let hasName = conflictDetectionSet.has(sourceName);
					if(!hasName) ***REMOVED***
						conflictDetectionSet.add(sourceName);
						continue;
					***REMOVED***

					// try the fallback name first
					sourceName = ModuleFilenameHelpers.createFilename(module, fallbackModuleFilenameTemplate, requestShortener);
					hasName = usedNamesSet.has(sourceName);
					if(!hasName) ***REMOVED***
						moduleToSourceNameMapping.set(module, sourceName);
						usedNamesSet.add(sourceName);
						continue;
					***REMOVED***

					// elsewise just append stars until we have a valid name
					while(hasName) ***REMOVED***
						sourceName += "*";
						hasName = usedNamesSet.has(sourceName);
					***REMOVED***
					moduleToSourceNameMapping.set(module, sourceName);
					usedNamesSet.add(sourceName);
				***REMOVED***
				tasks.forEach(function(task) ***REMOVED***
					const chunk = task.chunk;
					const file = task.file;
					const asset = task.asset;
					const sourceMap = task.sourceMap;
					const source = task.source;
					const modules = task.modules;
					const moduleFilenames = modules.map(m => moduleToSourceNameMapping.get(m));
					sourceMap.sources = moduleFilenames;
					if(sourceMap.sourcesContent && !options.noSources) ***REMOVED***
						sourceMap.sourcesContent = sourceMap.sourcesContent.map((content, i) => typeof content === "string" ? `$***REMOVED***content***REMOVED***\n\n\n$***REMOVED***ModuleFilenameHelpers.createFooter(modules[i], requestShortener)***REMOVED***` : null);
					***REMOVED*** else ***REMOVED***
						sourceMap.sourcesContent = undefined;
					***REMOVED***
					sourceMap.sourceRoot = options.sourceRoot || "";
					sourceMap.file = file;
					asset.__SourceMapDevToolFile = file;
					asset.__SourceMapDevToolData = ***REMOVED******REMOVED***;
					let currentSourceMappingURLComment = sourceMappingURLComment;
					if(currentSourceMappingURLComment !== false && /\.css($|\?)/i.test(file)) ***REMOVED***
						currentSourceMappingURLComment = currentSourceMappingURLComment.replace(/^\n\/\/(.*)$/, "\n/*$1*/");
					***REMOVED***
					const sourceMapString = JSON.stringify(sourceMap);
					if(sourceMapFilename) ***REMOVED***
						let filename = file;
						let query = "";
						const idx = filename.indexOf("?");
						if(idx >= 0) ***REMOVED***
							query = filename.substr(idx);
							filename = filename.substr(0, idx);
						***REMOVED***
						let sourceMapFile = compilation.getPath(sourceMapFilename, ***REMOVED***
							chunk,
							filename: options.fileContext ? path.relative(options.fileContext, filename) : filename,
							query,
							basename: basename(filename)
						***REMOVED***);
						if(sourceMapFile.indexOf("[contenthash]") !== -1) ***REMOVED***
							sourceMapFile = sourceMapFile.replace(/\[contenthash\]/g, crypto.createHash("md5").update(sourceMapString).digest("hex"));
						***REMOVED***
						const sourceMapUrl = options.publicPath ? options.publicPath + sourceMapFile.replace(/\\/g, "/") : path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
						if(currentSourceMappingURLComment !== false) ***REMOVED***
							asset.__SourceMapDevToolData[file] = compilation.assets[file] = new ConcatSource(new RawSource(source), currentSourceMappingURLComment.replace(/\[url\]/g, sourceMapUrl));
						***REMOVED***
						asset.__SourceMapDevToolData[sourceMapFile] = compilation.assets[sourceMapFile] = new RawSource(sourceMapString);
						chunk.files.push(sourceMapFile);
					***REMOVED*** else ***REMOVED***
						asset.__SourceMapDevToolData[file] = compilation.assets[file] = new ConcatSource(new RawSource(source), currentSourceMappingURLComment
							.replace(/\[map\]/g, () => sourceMapString)
							.replace(/\[url\]/g, () => `data:application/json;charset=utf-8;base64,$***REMOVED***new Buffer(sourceMapString, "utf-8").toString("base64")***REMOVED***`) // eslint-disable-line
						);
					***REMOVED***
				***REMOVED***);
			***REMOVED***);
		***REMOVED***);
	***REMOVED***
***REMOVED***

module.exports = SourceMapDevToolPlugin;
