// google-auto-auth/index.js
'use strict';

var async = require('async');
var fs = require('fs');
var GoogleAuth = require('google-auth-library');
var gcpMetadata = require('gcp-metadata');
var path = require('path');
var request = require('request');

class Auth ***REMOVED***
  constructor(config) ***REMOVED***
    this.authClientPromise = null;
    this.authClient = null;
    this.config = config || ***REMOVED******REMOVED***;
    this.environment = ***REMOVED******REMOVED***;
  ***REMOVED***

  authorizeRequest (reqOpts, callback) ***REMOVED***
    this.getToken((err, token) => ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      var authorizedReqOpts = Object.assign(***REMOVED******REMOVED***, reqOpts, ***REMOVED***
        headers: Object.assign(***REMOVED******REMOVED***, reqOpts.headers, ***REMOVED***
          Authorization: `Bearer $***REMOVED***token***REMOVED***`
        ***REMOVED***)
      ***REMOVED***);

      callback(null, authorizedReqOpts);
    ***REMOVED***);
  ***REMOVED***

  getAuthClient (callback) ***REMOVED***
    if (this.authClient) ***REMOVED***
      // This code works around an issue with context loss with async-listener.
      // Strictly speaking, this should not be necessary as the call to
      // authClientPromise.then(..) below would resolve to the same value.
      // However, async-listener defaults to resuming the `then` callbacks with
      // the context at the point of resolution rather than the context from the
      // point where the `then` callback was added. In this case, the promise
      // will be resolved on the very first incoming http request, and that
      // context will become sticky (will be restored by async-listener) around
      // the `then` callbacks for all subsequent requests.
      //
      // This breaks APM tools like Stackdriver Trace & others and tools like 
      // long stack traces (they will provide an incorrect stack trace).
      //
      // NOTE: this doesn't solve the problem generally. Any request concurrent
      // to the first call to this function, before the promise resolves, will
      // still lose context. We don't have a better solution at the moment :(.
      return setImmediate(callback.bind(null, null, this.authClient));
    ***REMOVED***

    var createAuthClientPromise = (resolve, reject) => ***REMOVED***
      var googleAuth = new GoogleAuth();

      var config = this.config;
      var keyFile = config.keyFilename || config.keyFile;

      var addScope = (err, authClient, projectId) => ***REMOVED***
        if (err) ***REMOVED***
          reject(err);
          return;
        ***REMOVED***

        if (authClient.createScopedRequired && authClient.createScopedRequired()) ***REMOVED***
          if (!config.scopes || config.scopes.length === 0) ***REMOVED***
            var scopeError = new Error('Scopes are required for this request.');
            scopeError.code = 'MISSING_SCOPE';
            reject(scopeError);
            return;
          ***REMOVED***
        ***REMOVED***

        authClient.scopes = config.scopes;
        this.authClient = authClient;
        this.projectId = projectId || authClient.projectId;

        resolve(authClient);
      ***REMOVED***;

      if (config.credentials) ***REMOVED***
        googleAuth.fromJSON(config.credentials, addScope);
      ***REMOVED*** else if (keyFile) ***REMOVED***
        keyFile = path.resolve(process.cwd(), keyFile);

        fs.readFile(keyFile, (err, contents) => ***REMOVED***
          if (err) ***REMOVED***
            reject(err);
            return;
          ***REMOVED***

          try ***REMOVED***
            googleAuth.fromJSON(JSON.parse(contents), addScope);
          ***REMOVED*** catch(e) ***REMOVED***
            var authClient = new googleAuth.JWT();
            authClient.keyFile = keyFile;
            authClient.email = config.email;
            addScope(null, authClient);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED*** else ***REMOVED***
        googleAuth.getApplicationDefault(addScope);
      ***REMOVED***
    ***REMOVED***;

    if (!this.authClientPromise) ***REMOVED***
      this.authClientPromise = new Promise(createAuthClientPromise);
    ***REMOVED***

    this.authClientPromise.then(callback.bind(null, null)).catch(callback);
  ***REMOVED***

  getCredentials (callback) ***REMOVED***
    this.getAuthClient((err, client) => ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      if (client.email && client.key) ***REMOVED***
        callback(null, ***REMOVED***
          client_email: client.email,
          private_key: client.key
        ***REMOVED***);
        return;
      ***REMOVED***

      if (!client.authorize) ***REMOVED***
        callback(new Error('Could not get credentials without a JSON, pem, or p12 keyfile.'));
        return;
      ***REMOVED***

      client.authorize(err => ***REMOVED***
        if (err) ***REMOVED***
          callback(err);
          return;
        ***REMOVED***

        this.getCredentials(callback);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***

  getEnvironment (callback) ***REMOVED***
    async.parallel([
      cb => this.isAppEngine(cb),
      cb => this.isCloudFunction(cb),
      cb => this.isComputeEngine(cb),
      cb => this.isContainerEngine(cb)
    ], () => ***REMOVED***
      callback(null, this.environment);
    ***REMOVED***);
  ***REMOVED***

  getProjectId (callback) ***REMOVED***
    if (this.projectId) ***REMOVED***
      setImmediate(() => ***REMOVED***
        callback(null, this.projectId);
      ***REMOVED***);
      return;
    ***REMOVED***

    this.getAuthClient(err => ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      callback(null, this.projectId);
    ***REMOVED***);
  ***REMOVED***

  getToken (callback) ***REMOVED***
    this.getAuthClient((err, client) => ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      client.getAccessToken(callback);
    ***REMOVED***);
  ***REMOVED***

  isAppEngine (callback) ***REMOVED***
    setImmediate(() => ***REMOVED***
      var env = this.environment;

      if (typeof env.IS_APP_ENGINE === 'undefined') ***REMOVED***
        env.IS_APP_ENGINE = !!(process.env.GAE_SERVICE || process.env.GAE_MODULE_NAME);
      ***REMOVED***

      callback(null, env.IS_APP_ENGINE);
    ***REMOVED***);
  ***REMOVED***

  isCloudFunction (callback) ***REMOVED***
    setImmediate(() => ***REMOVED***
      var env = this.environment;

      if (typeof env.IS_CLOUD_FUNCTION === 'undefined') ***REMOVED***
        env.IS_CLOUD_FUNCTION = !!process.env.FUNCTION_NAME;
      ***REMOVED***

      callback(null, env.IS_CLOUD_FUNCTION);
    ***REMOVED***);
  ***REMOVED***

  isComputeEngine (callback) ***REMOVED***
    var env = this.environment;

    if (typeof env.IS_COMPUTE_ENGINE !== 'undefined') ***REMOVED***
      setImmediate(() => ***REMOVED***
        callback(null, env.IS_COMPUTE_ENGINE);
      ***REMOVED***);
      return;
    ***REMOVED***

    request('http://metadata.google.internal', (err, res) => ***REMOVED***
      env.IS_COMPUTE_ENGINE = !err && res.headers['metadata-flavor'] === 'Google';

      callback(null, env.IS_COMPUTE_ENGINE);
    ***REMOVED***);
  ***REMOVED***

  isContainerEngine (callback) ***REMOVED***
    var env = this.environment;

    if (typeof env.IS_CONTAINER_ENGINE !== 'undefined') ***REMOVED***
      setImmediate(() => ***REMOVED***
        callback(null, env.IS_CONTAINER_ENGINE);
      ***REMOVED***);
      return;
    ***REMOVED***

    gcpMetadata.instance('/attributes/cluster-name', err => ***REMOVED***
      env.IS_CONTAINER_ENGINE = !err;

      callback(null, env.IS_CONTAINER_ENGINE);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

module.exports = config => ***REMOVED***
  return new Auth(config);
***REMOVED***;
