// @firebase/firestore/dist/cjs/src/platform_browser/browser_platform.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var serializer_1 = require("../remote/serializer");
var webchannel_connection_1 = require("./webchannel_connection");
var BrowserPlatform = /** @class */ (function () ***REMOVED***
    function BrowserPlatform() ***REMOVED***
        this.emptyByteString = '';
        this.base64Available = typeof atob !== 'undefined';
    ***REMOVED***
    BrowserPlatform.prototype.loadConnection = function (databaseInfo) ***REMOVED***
        return Promise.resolve(new webchannel_connection_1.WebChannelConnection(databaseInfo));
    ***REMOVED***;
    BrowserPlatform.prototype.newSerializer = function (databaseId) ***REMOVED***
        return new serializer_1.JsonProtoSerializer(databaseId, ***REMOVED*** useProto3Json: true ***REMOVED***);
    ***REMOVED***;
    BrowserPlatform.prototype.formatJSON = function (value) ***REMOVED***
        return JSON.stringify(value);
    ***REMOVED***;
    BrowserPlatform.prototype.atob = function (encoded) ***REMOVED***
        return atob(encoded);
    ***REMOVED***;
    BrowserPlatform.prototype.btoa = function (raw) ***REMOVED***
        return btoa(raw);
    ***REMOVED***;
    return BrowserPlatform;
***REMOVED***());
exports.BrowserPlatform = BrowserPlatform;
