// object-copy/index.js
'use strict';

var typeOf = require('kind-of');
var copyDescriptor = require('copy-descriptor');
var define = require('define-property');

/**
 * Copy static properties, prototype properties, and descriptors from one object to another.
 *
 * ```js
 * function App() ***REMOVED******REMOVED***
 * var proto = App.prototype;
 * App.prototype.set = function() ***REMOVED******REMOVED***;
 * App.prototype.get = function() ***REMOVED******REMOVED***;
 *
 * var obj = ***REMOVED******REMOVED***;
 * copy(obj, proto);
 * ```
 * @param ***REMOVED***Object***REMOVED*** `receiver`
 * @param ***REMOVED***Object***REMOVED*** `provider`
 * @param ***REMOVED***String|Array***REMOVED*** `omit` One or more properties to omit
 * @return ***REMOVED***Object***REMOVED***
 * @api public
 */

function copy(receiver, provider, omit) ***REMOVED***
  if (!isObject(receiver)) ***REMOVED***
    throw new TypeError('expected receiving object to be an object.');
  ***REMOVED***
  if (!isObject(provider)) ***REMOVED***
    throw new TypeError('expected providing object to be an object.');
  ***REMOVED***

  var props = nativeKeys(provider);
  var keys = Object.keys(provider);
  var len = props.length;
  omit = arrayify(omit);

  while (len--) ***REMOVED***
    var key = props[len];

    if (has(keys, key)) ***REMOVED***
      define(receiver, key, provider[key]);
    ***REMOVED*** else if (!(key in receiver) && !has(omit, key)) ***REMOVED***
      copyDescriptor(receiver, provider, key);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Return true if the given value is an object or function
 */

function isObject(val) ***REMOVED***
  return typeOf(val) === 'object' || typeof val === 'function';
***REMOVED***

/**
 * Returns true if an array has any of the given elements, or an
 * object has any of the give keys.
 *
 * ```js
 * has(['a', 'b', 'c'], 'c');
 * //=> true
 *
 * has(['a', 'b', 'c'], ['c', 'z']);
 * //=> true
 *
 * has(***REMOVED***a: 'b', c: 'd'***REMOVED***, ['c', 'z']);
 * //=> true
 * ```
 * @param ***REMOVED***Object***REMOVED*** `obj`
 * @param ***REMOVED***String|Array***REMOVED*** `val`
 * @return ***REMOVED***Boolean***REMOVED***
 */

function has(obj, val) ***REMOVED***
  val = arrayify(val);
  var len = val.length;

  if (isObject(obj)) ***REMOVED***
    for (var key in obj) ***REMOVED***
      if (val.indexOf(key) > -1) ***REMOVED***
        return true;
      ***REMOVED***
    ***REMOVED***

    var keys = nativeKeys(obj);
    return has(keys, val);
  ***REMOVED***

  if (Array.isArray(obj)) ***REMOVED***
    var arr = obj;
    while (len--) ***REMOVED***
      if (arr.indexOf(val[len]) > -1) ***REMOVED***
        return true;
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***

  throw new TypeError('expected an array or object.');
***REMOVED***

/**
 * Cast the given value to an array.
 *
 * ```js
 * arrayify('foo');
 * //=> ['foo']
 *
 * arrayify(['foo']);
 * //=> ['foo']
 * ```
 *
 * @param ***REMOVED***String|Array***REMOVED*** `val`
 * @return ***REMOVED***Array***REMOVED***
 */

function arrayify(val) ***REMOVED***
  return val ? (Array.isArray(val) ? val : [val]) : [];
***REMOVED***

/**
 * Returns true if a value has a `contructor`
 *
 * ```js
 * hasConstructor(***REMOVED******REMOVED***);
 * //=> true
 *
 * hasConstructor(Object.create(null));
 * //=> false
 * ```
 * @param  ***REMOVED***Object***REMOVED*** `value`
 * @return ***REMOVED***Boolean***REMOVED***
 */

function hasConstructor(val) ***REMOVED***
  return isObject(val) && typeof val.constructor !== 'undefined';
***REMOVED***

/**
 * Get the native `ownPropertyNames` from the constructor of the
 * given `object`. An empty array is returned if the object does
 * not have a constructor.
 *
 * ```js
 * nativeKeys(***REMOVED***a: 'b', b: 'c', c: 'd'***REMOVED***)
 * //=> ['a', 'b', 'c']
 *
 * nativeKeys(function()***REMOVED******REMOVED***)
 * //=> ['length', 'caller']
 * ```
 *
 * @param  ***REMOVED***Object***REMOVED*** `obj` Object that has a `constructor`.
 * @return ***REMOVED***Array***REMOVED*** Array of keys.
 */

function nativeKeys(val) ***REMOVED***
  if (!hasConstructor(val)) return [];
  return Object.getOwnPropertyNames(val);
***REMOVED***

/**
 * Expose `copy`
 */

module.exports = copy;

/**
 * Expose `copy.has` for tests
 */

module.exports.has = has;
