// node-pre-gyp/lib/node-pre-gyp.js
"use strict";

/**
 * Module exports.
 */

module.exports = exports;

/**
 * Module dependencies.
 */

var path = require('path');
var nopt = require('nopt');
var log = require('npmlog');
log.disableProgress();

var EE = require('events').EventEmitter;
var inherits = require('util').inherits;
var commands = [
      'clean',
      'install',
      'reinstall',
      'build',
      'rebuild',
      'package',
      'testpackage',
      'publish',
      'unpublish',
      'info',
      'testbinary',
      'reveal',
      'configure'
    ];
var aliases = ***REMOVED******REMOVED***;

// differentiate node-pre-gyp's logs from npm's
log.heading = 'node-pre-gyp';

exports.find = require('./pre-binding').find;

function Run() ***REMOVED***
  var self = this;

  this.commands = ***REMOVED******REMOVED***;

  commands.forEach(function (command) ***REMOVED***
    self.commands[command] = function (argv, callback) ***REMOVED***
      log.verbose('command', command, argv);
      return require('./' + command)(self, argv, callback);
    ***REMOVED***;
  ***REMOVED***);
***REMOVED***
inherits(Run, EE);
exports.Run = Run;
var proto = Run.prototype;

/**
 * Export the contents of the package.json.
 */

proto.package = require('../package.json');

/**
 * nopt configuration definitions
 */

proto.configDefs = ***REMOVED***
    help: Boolean,     // everywhere
    arch: String,      // 'configure'
    debug: Boolean,    // 'build'
    directory: String, // bin
    proxy: String,     // 'install'
    loglevel: String,  // everywhere
***REMOVED***;

/**
 * nopt shorthands
 */

proto.shorthands = ***REMOVED***
    release: '--no-debug',
    C: '--directory',
    debug: '--debug',
    j: '--jobs',
    silent: '--loglevel=silent',
    silly: '--loglevel=silly',
    verbose: '--loglevel=verbose',
***REMOVED***;

/**
 * expose the command aliases for the bin file to use.
 */

proto.aliases = aliases;

/**
 * Parses the given argv array and sets the 'opts',
 * 'argv' and 'command' properties.
 */

proto.parseArgv = function parseOpts (argv) ***REMOVED***
  this.opts = nopt(this.configDefs, this.shorthands, argv);
  this.argv = this.opts.argv.remain.slice();
  var commands = this.todo = [];

  // create a copy of the argv array with aliases mapped
  argv = this.argv.map(function (arg) ***REMOVED***
    // is this an alias?
    if (arg in this.aliases) ***REMOVED***
      arg = this.aliases[arg];
    ***REMOVED***
    return arg;
  ***REMOVED***, this);

  // process the mapped args into "command" objects ("name" and "args" props)
  argv.slice().forEach(function (arg) ***REMOVED***
    if (arg in this.commands) ***REMOVED***
      var args = argv.splice(0, argv.indexOf(arg));
      argv.shift();
      if (commands.length > 0) ***REMOVED***
        commands[commands.length - 1].args = args;
      ***REMOVED***
      commands.push(***REMOVED*** name: arg, args: [] ***REMOVED***);
    ***REMOVED***
  ***REMOVED***, this);
  if (commands.length > 0) ***REMOVED***
    commands[commands.length - 1].args = argv.splice(0);
  ***REMOVED***

  // support for inheriting config env variables from npm
  var npm_config_prefix = 'npm_config_';
  Object.keys(process.env).forEach(function (name) ***REMOVED***
    if (name.indexOf(npm_config_prefix) !== 0) return;
    var val = process.env[name];
    if (name === npm_config_prefix + 'loglevel') ***REMOVED***
      log.level = val;
    ***REMOVED*** else ***REMOVED***
      // add the user-defined options to the config
      name = name.substring(npm_config_prefix.length);
      // avoid npm argv clobber already present args
      // which avoids problem of 'npm test' calling
      // script that runs unique npm install commands
      if (name === 'argv') ***REMOVED***
         if (this.opts.argv &&
             this.opts.argv.remain &&
             this.opts.argv.remain.length) ***REMOVED***
            // do nothing
         ***REMOVED*** else ***REMOVED***
            this.opts[name] = val;
         ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        this.opts[name] = val;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***, this);

  if (this.opts.loglevel) ***REMOVED***
    log.level = this.opts.loglevel;
  ***REMOVED***
  log.resume();
***REMOVED***;

/**
 * Returns the usage instructions for node-pre-gyp.
 */

proto.usage = function usage () ***REMOVED***
  var str = [
      '',
      '  Usage: node-pre-gyp <command> [options]',
      '',
      '  where <command> is one of:',
      commands.map(function (c) ***REMOVED***
        return '    - ' + c + ' - ' + require('./' + c).usage;
      ***REMOVED***).join('\n'),
      '',
      'node-pre-gyp@' + this.version + '  ' + path.resolve(__dirname, '..'),
      'node@' + process.versions.node
  ].join('\n');
  return str;
***REMOVED***;

/**
 * Version number getter.
 */

Object.defineProperty(proto, 'version', ***REMOVED***
    get: function () ***REMOVED***
      return this.package.version;
    ***REMOVED***,
    enumerable: true
***REMOVED***);
