// escope/src/index.js
/*
  Copyright (C) 2012-2014 Yusuke Suzuki <utatane.tea@gmail.com>
  Copyright (C) 2013 Alex Seville <hi@alexanderseville.com>
  Copyright (C) 2014 Thiago de Arruda <tpadilha84@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/**
 * Escope (<a href="http://github.com/estools/escope">escope</a>) is an <a
 * href="http://www.ecma-international.org/publications/standards/Ecma-262.htm">ECMAScript</a>
 * scope analyzer extracted from the <a
 * href="http://github.com/estools/esmangle">esmangle project</a/>.
 * <p>
 * <em>escope</em> finds lexical scopes in a source program, i.e. areas of that
 * program where different occurrences of the same identifier refer to the same
 * variable. With each scope the contained variables are collected, and each
 * identifier reference in code is linked to its corresponding variable (if
 * possible).
 * <p>
 * <em>escope</em> works on a syntax tree of the parsed source code which has
 * to adhere to the <a
 * href="https://developer.mozilla.org/en-US/docs/SpiderMonkey/Parser_API">
 * Mozilla Parser API</a>. E.g. <a href="http://esprima.org">esprima</a> is a parser
 * that produces such syntax trees.
 * <p>
 * The main interface is the ***REMOVED***@link analyze***REMOVED*** function.
 * @module escope
 */

/*jslint bitwise:true */

import assert from 'assert';

import ScopeManager from './scope-manager';
import Referencer from './referencer';
import Reference from './reference';
import Variable from './variable';
import Scope from './scope';
import ***REMOVED*** version ***REMOVED*** from '../package.json';

function defaultOptions() ***REMOVED***
    return ***REMOVED***
        optimistic: false,
        directive: false,
        nodejsScope: false,
        impliedStrict: false,
        sourceType: 'script',  // one of ['script', 'module']
        ecmaVersion: 5,
        childVisitorKeys: null,
        fallback: 'iteration'
    ***REMOVED***;
***REMOVED***

function updateDeeply(target, override) ***REMOVED***
    var key, val;

    function isHashObject(target) ***REMOVED***
        return typeof target === 'object' && target instanceof Object && !(target instanceof Array) && !(target instanceof RegExp);
    ***REMOVED***

    for (key in override) ***REMOVED***
        if (override.hasOwnProperty(key)) ***REMOVED***
            val = override[key];
            if (isHashObject(val)) ***REMOVED***
                if (isHashObject(target[key])) ***REMOVED***
                    updateDeeply(target[key], val);
                ***REMOVED*** else ***REMOVED***
                    target[key] = updateDeeply(***REMOVED******REMOVED***, val);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                target[key] = val;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return target;
***REMOVED***

/**
 * Main interface function. Takes an Esprima syntax tree and returns the
 * analyzed scopes.
 * @function analyze
 * @param ***REMOVED***esprima.Tree***REMOVED*** tree
 * @param ***REMOVED***Object***REMOVED*** providedOptions - Options that tailor the scope analysis
 * @param ***REMOVED***boolean***REMOVED*** [providedOptions.optimistic=false] - the optimistic flag
 * @param ***REMOVED***boolean***REMOVED*** [providedOptions.directive=false]- the directive flag
 * @param ***REMOVED***boolean***REMOVED*** [providedOptions.ignoreEval=false]- whether to check 'eval()' calls
 * @param ***REMOVED***boolean***REMOVED*** [providedOptions.nodejsScope=false]- whether the whole
 * script is executed under node.js environment. When enabled, escope adds
 * a function scope immediately following the global scope.
 * @param ***REMOVED***boolean***REMOVED*** [providedOptions.impliedStrict=false]- implied strict mode
 * (if ecmaVersion >= 5).
 * @param ***REMOVED***string***REMOVED*** [providedOptions.sourceType='script']- the source type of the script. one of 'script' and 'module'
 * @param ***REMOVED***number***REMOVED*** [providedOptions.ecmaVersion=5]- which ECMAScript version is considered
 * @param ***REMOVED***Object***REMOVED*** [providedOptions.childVisitorKeys=null] - Additional known visitor keys. See [esrecurse](https://github.com/estools/esrecurse)'s the `childVisitorKeys` option.
 * @param ***REMOVED***string***REMOVED*** [providedOptions.fallback='iteration'] - A kind of the fallback in order to encounter with unknown node. See [esrecurse](https://github.com/estools/esrecurse)'s the `fallback` option.
 * @return ***REMOVED***ScopeManager***REMOVED***
 */
export function analyze(tree, providedOptions) ***REMOVED***
    var scopeManager, referencer, options;

    options = updateDeeply(defaultOptions(), providedOptions);

    scopeManager = new ScopeManager(options);

    referencer = new Referencer(options, scopeManager);
    referencer.visit(tree);

    assert(scopeManager.__currentScope === null, 'currentScope should be null.');

    return scopeManager;
***REMOVED***

export ***REMOVED***
    /** @name module:escope.version */
    version,
    /** @name module:escope.Reference */
    Reference,
    /** @name module:escope.Variable */
    Variable,
    /** @name module:escope.Scope */
    Scope,
    /** @name module:escope.ScopeManager */
    ScopeManager
***REMOVED***;


/* vim: set sw=4 ts=4 et tw=80 : */
