// escope/src/referencer.js
/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
import ***REMOVED*** Syntax ***REMOVED*** from 'estraverse';
import esrecurse from 'esrecurse';
import Reference from './reference';
import Variable from './variable';
import PatternVisitor from './pattern-visitor';
import ***REMOVED*** ParameterDefinition, Definition ***REMOVED*** from './definition';
import assert from 'assert';

function traverseIdentifierInPattern(options, rootPattern, referencer, callback) ***REMOVED***
    // Call the callback at left hand identifier nodes, and Collect right hand nodes.
    var visitor = new PatternVisitor(options, rootPattern, callback);
    visitor.visit(rootPattern);

    // Process the right hand nodes recursively.
    if (referencer != null) ***REMOVED***
        visitor.rightHandNodes.forEach(referencer.visit, referencer);
    ***REMOVED***
***REMOVED***

// Importing ImportDeclaration.
// http://people.mozilla.org/~jorendorff/es6-draft.html#sec-moduledeclarationinstantiation
// https://github.com/estree/estree/blob/master/es6.md#importdeclaration
// FIXME: Now, we don't create module environment, because the context is
// implementation dependent.

class Importer extends esrecurse.Visitor ***REMOVED***
    constructor(declaration, referencer) ***REMOVED***
        super(null, referencer.options);
        this.declaration = declaration;
        this.referencer = referencer;
    ***REMOVED***

    visitImport(id, specifier) ***REMOVED***
        this.referencer.visitPattern(id, (pattern) => ***REMOVED***
            this.referencer.currentScope().__define(pattern,
                new Definition(
                    Variable.ImportBinding,
                    pattern,
                    specifier,
                    this.declaration,
                    null,
                    null
                    ));
        ***REMOVED***);
    ***REMOVED***

    ImportNamespaceSpecifier(node) ***REMOVED***
        let local = (node.local || node.id);
        if (local) ***REMOVED***
            this.visitImport(local, node);
        ***REMOVED***
    ***REMOVED***

    ImportDefaultSpecifier(node) ***REMOVED***
        let local = (node.local || node.id);
        this.visitImport(local, node);
    ***REMOVED***

    ImportSpecifier(node) ***REMOVED***
        let local = (node.local || node.id);
        if (node.name) ***REMOVED***
            this.visitImport(node.name, node);
        ***REMOVED*** else ***REMOVED***
            this.visitImport(local, node);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

// Referencing variables and creating bindings.
export default class Referencer extends esrecurse.Visitor ***REMOVED***
    constructor(options, scopeManager) ***REMOVED***
        super(null, options);
        this.options = options;
        this.scopeManager = scopeManager;
        this.parent = null;
        this.isInnerMethodDefinition = false;
    ***REMOVED***

    currentScope() ***REMOVED***
        return this.scopeManager.__currentScope;
    ***REMOVED***

    close(node) ***REMOVED***
        while (this.currentScope() && node === this.currentScope().block) ***REMOVED***
            this.scopeManager.__currentScope = this.currentScope().__close(this.scopeManager);
        ***REMOVED***
    ***REMOVED***

    pushInnerMethodDefinition(isInnerMethodDefinition) ***REMOVED***
        var previous = this.isInnerMethodDefinition;
        this.isInnerMethodDefinition = isInnerMethodDefinition;
        return previous;
    ***REMOVED***

    popInnerMethodDefinition(isInnerMethodDefinition) ***REMOVED***
        this.isInnerMethodDefinition = isInnerMethodDefinition;
    ***REMOVED***

    materializeTDZScope(node, iterationNode) ***REMOVED***
        // http://people.mozilla.org/~jorendorff/es6-draft.html#sec-runtime-semantics-forin-div-ofexpressionevaluation-abstract-operation
        // TDZ scope hides the declaration's names.
        this.scopeManager.__nestTDZScope(node, iterationNode);
        this.visitVariableDeclaration(this.currentScope(), Variable.TDZ, iterationNode.left, 0, true);
    ***REMOVED***

    materializeIterationScope(node) ***REMOVED***
        // Generate iteration scope for upper ForIn/ForOf Statements.
        var letOrConstDecl;
        this.scopeManager.__nestForScope(node);
        letOrConstDecl = node.left;
        this.visitVariableDeclaration(this.currentScope(), Variable.Variable, letOrConstDecl, 0);
        this.visitPattern(letOrConstDecl.declarations[0].id, (pattern) => ***REMOVED***
            this.currentScope().__referencing(pattern, Reference.WRITE, node.right, null, true, true);
        ***REMOVED***);
    ***REMOVED***

    referencingDefaultValue(pattern, assignments, maybeImplicitGlobal, init) ***REMOVED***
        const scope = this.currentScope();
        assignments.forEach(assignment => ***REMOVED***
            scope.__referencing(
                pattern,
                Reference.WRITE,
                assignment.right,
                maybeImplicitGlobal,
                pattern !== assignment.left,
                init);
        ***REMOVED***);
    ***REMOVED***

    visitPattern(node, options, callback) ***REMOVED***
        if (typeof options === 'function') ***REMOVED***
            callback = options;
            options = ***REMOVED***processRightHandNodes: false***REMOVED***
        ***REMOVED***
        traverseIdentifierInPattern(
            this.options,
            node,
            options.processRightHandNodes ? this : null,
            callback);
    ***REMOVED***

    visitFunction(node) ***REMOVED***
        var i, iz;
        // FunctionDeclaration name is defined in upper scope
        // NOTE: Not referring variableScope. It is intended.
        // Since
        //  in ES5, FunctionDeclaration should be in FunctionBody.
        //  in ES6, FunctionDeclaration should be block scoped.
        if (node.type === Syntax.FunctionDeclaration) ***REMOVED***
            // id is defined in upper scope
            this.currentScope().__define(node.id,
                    new Definition(
                        Variable.FunctionName,
                        node.id,
                        node,
                        null,
                        null,
                        null
                    ));
        ***REMOVED***

        // FunctionExpression with name creates its special scope;
        // FunctionExpressionNameScope.
        if (node.type === Syntax.FunctionExpression && node.id) ***REMOVED***
            this.scopeManager.__nestFunctionExpressionNameScope(node);
        ***REMOVED***

        // Consider this function is in the MethodDefinition.
        this.scopeManager.__nestFunctionScope(node, this.isInnerMethodDefinition);

        // Process parameter declarations.
        for (i = 0, iz = node.params.length; i < iz; ++i) ***REMOVED***
            this.visitPattern(node.params[i], ***REMOVED***processRightHandNodes: true***REMOVED***, (pattern, info) => ***REMOVED***
                this.currentScope().__define(pattern,
                    new ParameterDefinition(
                        pattern,
                        node,
                        i,
                        info.rest
                    ));

                this.referencingDefaultValue(pattern, info.assignments, null, true);
            ***REMOVED***);
        ***REMOVED***

        // if there's a rest argument, add that
        if (node.rest) ***REMOVED***
            this.visitPattern(***REMOVED***
                type: 'RestElement',
                argument: node.rest
            ***REMOVED***, (pattern) => ***REMOVED***
                this.currentScope().__define(pattern,
                    new ParameterDefinition(
                        pattern,
                        node,
                        node.params.length,
                        true
                    ));
            ***REMOVED***);
        ***REMOVED***

        // Skip BlockStatement to prevent creating BlockStatement scope.
        if (node.body.type === Syntax.BlockStatement) ***REMOVED***
            this.visitChildren(node.body);
        ***REMOVED*** else ***REMOVED***
            this.visit(node.body);
        ***REMOVED***

        this.close(node);
    ***REMOVED***

    visitClass(node) ***REMOVED***
        if (node.type === Syntax.ClassDeclaration) ***REMOVED***
            this.currentScope().__define(node.id,
                    new Definition(
                        Variable.ClassName,
                        node.id,
                        node,
                        null,
                        null,
                        null
                    ));
        ***REMOVED***

        // FIXME: Maybe consider TDZ.
        this.visit(node.superClass);

        this.scopeManager.__nestClassScope(node);

        if (node.id) ***REMOVED***
            this.currentScope().__define(node.id,
                    new Definition(
                        Variable.ClassName,
                        node.id,
                        node
                    ));
        ***REMOVED***
        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    visitProperty(node) ***REMOVED***
        var previous, isMethodDefinition;
        if (node.computed) ***REMOVED***
            this.visit(node.key);
        ***REMOVED***

        isMethodDefinition = node.type === Syntax.MethodDefinition;
        if (isMethodDefinition) ***REMOVED***
            previous = this.pushInnerMethodDefinition(true);
        ***REMOVED***
        this.visit(node.value);
        if (isMethodDefinition) ***REMOVED***
            this.popInnerMethodDefinition(previous);
        ***REMOVED***
    ***REMOVED***

    visitForIn(node) ***REMOVED***
        if (node.left.type === Syntax.VariableDeclaration && node.left.kind !== 'var') ***REMOVED***
            this.materializeTDZScope(node.right, node);
            this.visit(node.right);
            this.close(node.right);

            this.materializeIterationScope(node);
            this.visit(node.body);
            this.close(node);
        ***REMOVED*** else ***REMOVED***
            if (node.left.type === Syntax.VariableDeclaration) ***REMOVED***
                this.visit(node.left);
                this.visitPattern(node.left.declarations[0].id, (pattern) => ***REMOVED***
                    this.currentScope().__referencing(pattern, Reference.WRITE, node.right, null, true, true);
                ***REMOVED***);
            ***REMOVED*** else ***REMOVED***
                this.visitPattern(node.left, ***REMOVED***processRightHandNodes: true***REMOVED***, (pattern, info) => ***REMOVED***
                    var maybeImplicitGlobal = null;
                    if (!this.currentScope().isStrict) ***REMOVED***
                        maybeImplicitGlobal = ***REMOVED***
                            pattern: pattern,
                            node: node
                        ***REMOVED***;
                    ***REMOVED***
                    this.referencingDefaultValue(pattern, info.assignments, maybeImplicitGlobal, false);
                    this.currentScope().__referencing(pattern, Reference.WRITE, node.right, maybeImplicitGlobal, true, false);
                ***REMOVED***);
            ***REMOVED***
            this.visit(node.right);
            this.visit(node.body);
        ***REMOVED***
    ***REMOVED***

    visitVariableDeclaration(variableTargetScope, type, node, index, fromTDZ) ***REMOVED***
        // If this was called to initialize a TDZ scope, this needs to make definitions, but doesn't make references.
        var decl, init;

        decl = node.declarations[index];
        init = decl.init;
        this.visitPattern(decl.id, ***REMOVED***processRightHandNodes: !fromTDZ***REMOVED***, (pattern, info) => ***REMOVED***
            variableTargetScope.__define(pattern,
                new Definition(
                    type,
                    pattern,
                    decl,
                    node,
                    index,
                    node.kind
                ));

            if (!fromTDZ) ***REMOVED***
                this.referencingDefaultValue(pattern, info.assignments, null, true);
            ***REMOVED***
            if (init) ***REMOVED***
                this.currentScope().__referencing(pattern, Reference.WRITE, init, null, !info.topLevel, true);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    AssignmentExpression(node) ***REMOVED***
        if (PatternVisitor.isPattern(node.left)) ***REMOVED***
            if (node.operator === '=') ***REMOVED***
                this.visitPattern(node.left, ***REMOVED***processRightHandNodes: true***REMOVED***, (pattern, info) => ***REMOVED***
                    var maybeImplicitGlobal = null;
                    if (!this.currentScope().isStrict) ***REMOVED***
                        maybeImplicitGlobal = ***REMOVED***
                            pattern: pattern,
                            node: node
                        ***REMOVED***;
                    ***REMOVED***
                    this.referencingDefaultValue(pattern, info.assignments, maybeImplicitGlobal, false);
                    this.currentScope().__referencing(pattern, Reference.WRITE, node.right, maybeImplicitGlobal, !info.topLevel, false);
                ***REMOVED***);
            ***REMOVED*** else ***REMOVED***
                this.currentScope().__referencing(node.left, Reference.RW, node.right);
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            this.visit(node.left);
        ***REMOVED***
        this.visit(node.right);
    ***REMOVED***

    CatchClause(node) ***REMOVED***
        this.scopeManager.__nestCatchScope(node);

        this.visitPattern(node.param, ***REMOVED***processRightHandNodes: true***REMOVED***, (pattern, info) => ***REMOVED***
            this.currentScope().__define(pattern,
                new Definition(
                    Variable.CatchClause,
                    node.param,
                    node,
                    null,
                    null,
                    null
                ));
            this.referencingDefaultValue(pattern, info.assignments, null, true);
        ***REMOVED***);
        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    Program(node) ***REMOVED***
        this.scopeManager.__nestGlobalScope(node);

        if (this.scopeManager.__isNodejsScope()) ***REMOVED***
            // Force strictness of GlobalScope to false when using node.js scope.
            this.currentScope().isStrict = false;
            this.scopeManager.__nestFunctionScope(node, false);
        ***REMOVED***

        if (this.scopeManager.__isES6() && this.scopeManager.isModule()) ***REMOVED***
            this.scopeManager.__nestModuleScope(node);
        ***REMOVED***

        if (this.scopeManager.isStrictModeSupported() && this.scopeManager.isImpliedStrict()) ***REMOVED***
            this.currentScope().isStrict = true;
        ***REMOVED***

        this.visitChildren(node);
        this.close(node);
    ***REMOVED***

    Identifier(node) ***REMOVED***
        this.currentScope().__referencing(node);
    ***REMOVED***

    UpdateExpression(node) ***REMOVED***
        if (PatternVisitor.isPattern(node.argument)) ***REMOVED***
            this.currentScope().__referencing(node.argument, Reference.RW, null);
        ***REMOVED*** else ***REMOVED***
            this.visitChildren(node);
        ***REMOVED***
    ***REMOVED***

    MemberExpression(node) ***REMOVED***
        this.visit(node.object);
        if (node.computed) ***REMOVED***
            this.visit(node.property);
        ***REMOVED***
    ***REMOVED***

    Property(node) ***REMOVED***
        this.visitProperty(node);
    ***REMOVED***

    MethodDefinition(node) ***REMOVED***
        this.visitProperty(node);
    ***REMOVED***

    BreakStatement() ***REMOVED******REMOVED***

    ContinueStatement() ***REMOVED******REMOVED***

    LabeledStatement(node) ***REMOVED***
        this.visit(node.body);
    ***REMOVED***

    ForStatement(node) ***REMOVED***
        // Create ForStatement declaration.
        // NOTE: In ES6, ForStatement dynamically generates
        // per iteration environment. However, escope is
        // a static analyzer, we only generate one scope for ForStatement.
        if (node.init && node.init.type === Syntax.VariableDeclaration && node.init.kind !== 'var') ***REMOVED***
            this.scopeManager.__nestForScope(node);
        ***REMOVED***

        this.visitChildren(node);

        this.close(node);
    ***REMOVED***

    ClassExpression(node) ***REMOVED***
        this.visitClass(node);
    ***REMOVED***

    ClassDeclaration(node) ***REMOVED***
        this.visitClass(node);
    ***REMOVED***

    CallExpression(node) ***REMOVED***
        // Check this is direct call to eval
        if (!this.scopeManager.__ignoreEval() && node.callee.type === Syntax.Identifier && node.callee.name === 'eval') ***REMOVED***
            // NOTE: This should be `variableScope`. Since direct eval call always creates Lexical environment and
            // let / const should be enclosed into it. Only VariableDeclaration affects on the caller's environment.
            this.currentScope().variableScope.__detectEval();
        ***REMOVED***
        this.visitChildren(node);
    ***REMOVED***

    BlockStatement(node) ***REMOVED***
        if (this.scopeManager.__isES6()) ***REMOVED***
            this.scopeManager.__nestBlockScope(node);
        ***REMOVED***

        this.visitChildren(node);

        this.close(node);
    ***REMOVED***

    ThisExpression() ***REMOVED***
        this.currentScope().variableScope.__detectThis();
    ***REMOVED***

    WithStatement(node) ***REMOVED***
        this.visit(node.object);
        // Then nest scope for WithStatement.
        this.scopeManager.__nestWithScope(node);

        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    VariableDeclaration(node) ***REMOVED***
        var variableTargetScope, i, iz, decl;
        variableTargetScope = (node.kind === 'var') ? this.currentScope().variableScope : this.currentScope();
        for (i = 0, iz = node.declarations.length; i < iz; ++i) ***REMOVED***
            decl = node.declarations[i];
            this.visitVariableDeclaration(variableTargetScope, Variable.Variable, node, i);
            if (decl.init) ***REMOVED***
                this.visit(decl.init);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // sec 13.11.8
    SwitchStatement(node) ***REMOVED***
        var i, iz;

        this.visit(node.discriminant);

        if (this.scopeManager.__isES6()) ***REMOVED***
            this.scopeManager.__nestSwitchScope(node);
        ***REMOVED***

        for (i = 0, iz = node.cases.length; i < iz; ++i) ***REMOVED***
            this.visit(node.cases[i]);
        ***REMOVED***

        this.close(node);
    ***REMOVED***

    FunctionDeclaration(node) ***REMOVED***
        this.visitFunction(node);
    ***REMOVED***

    FunctionExpression(node) ***REMOVED***
        this.visitFunction(node);
    ***REMOVED***

    ForOfStatement(node) ***REMOVED***
        this.visitForIn(node);
    ***REMOVED***

    ForInStatement(node) ***REMOVED***
        this.visitForIn(node);
    ***REMOVED***

    ArrowFunctionExpression(node) ***REMOVED***
        this.visitFunction(node);
    ***REMOVED***

    ImportDeclaration(node) ***REMOVED***
        var importer;

        assert(this.scopeManager.__isES6() && this.scopeManager.isModule(), 'ImportDeclaration should appear when the mode is ES6 and in the module context.');

        importer = new Importer(node, this);
        importer.visit(node);
    ***REMOVED***

    visitExportDeclaration(node) ***REMOVED***
        if (node.source) ***REMOVED***
            return;
        ***REMOVED***
        if (node.declaration) ***REMOVED***
            this.visit(node.declaration);
            return;
        ***REMOVED***

        this.visitChildren(node);
    ***REMOVED***

    ExportDeclaration(node) ***REMOVED***
        this.visitExportDeclaration(node);
    ***REMOVED***

    ExportNamedDeclaration(node) ***REMOVED***
        this.visitExportDeclaration(node);
    ***REMOVED***

    ExportSpecifier(node) ***REMOVED***
        let local = (node.id || node.local);
        this.visit(local);
    ***REMOVED***

    MetaProperty() ***REMOVED***
        // do nothing.
    ***REMOVED***
***REMOVED***

/* vim: set sw=4 ts=4 et tw=80 : */
