// google-gax/lib/gax.js
/**
 * Copyright 2016, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * Google API Extensions
 */

'use strict';

/**
 * Encapsulates the overridable settings for a particular API call.
 *
 * ``CallOptions`` is an optional arg for all GAX API calls.  It is used to
 * configure the settings of a specific API call.
 *
 * When provided, its values override the GAX service defaults for that
 * particular call.
 *
 * Typically the API clients will accept this as the second to the last
 * argument. See the examples below.
 * @typedef ***REMOVED***Object***REMOVED*** CallOptions
 * @property ***REMOVED***number=***REMOVED*** timeout - The client-side timeout for API calls.
 * @property ***REMOVED***RetryOptions=***REMOVED*** retry - determines whether and how to retry
 *   on transient errors. When set to null, the call will not retry.
 * @property ***REMOVED***boolean=***REMOVED*** autoPaginate - If set to false and the call is
 *   configured for paged iteration, page unrolling is not performed, instead
 *   the callback will be called with the response object.
 * @property ***REMOVED***Object=***REMOVED*** pageToken - If set and the call is configured for
 *   paged iteration, paged iteration is not performed and requested with this
 *   pageToken.
 * @property ***REMOVED***number***REMOVED*** maxResults - If set and the call is configured for
 *   paged iteration, the call will stop when the number of response elements
 *   reaches to the specified size. By default, it will unroll the page to
 *   the end of the list.
 * @property ***REMOVED***boolean=***REMOVED*** isBundling - If set to false and the call is configured
 *   for bundling, bundling is not performed.
 * @property ***REMOVED***BackoffSettings=***REMOVED*** longrunning - BackoffSettings used for polling.
 * @property ***REMOVED***Function=***REMOVED*** promise - A constructor for a promise that implements the ES6
 * specification of promise which will be used to create promises. If not
 * provided, native promises will be used.
 * @example
 * // suppress bundling for bundled method.
 * api.bundlingMethod(
 *     param, ***REMOVED***optParam: aValue, isBundling: false***REMOVED***, function(err, response) ***REMOVED***
 *   // handle response.
 * ***REMOVED***);
 * @example
 * // suppress streaming for page-streaming method.
 * api.pageStreamingMethod(
 *     param, ***REMOVED***optParam: aValue, autoPaginate: false***REMOVED***, function(err, page) ***REMOVED***
 *   // not returning a stream, but callback is called with the paged response.
 * ***REMOVED***);
 */

/**
 * Per-call configurable settings for retrying upon transient failure.
 * @typedef ***REMOVED***Object***REMOVED*** RetryOptions
 * @property ***REMOVED***String[]***REMOVED*** retryCodes
 * @property ***REMOVED***BackoffSettings***REMOVED*** backoffSettings
 */

/**
 * Parameters to the exponential backoff algorithm for retrying.
 * @typedef ***REMOVED***Object***REMOVED*** BackoffSettings
 * @property ***REMOVED***number***REMOVED*** initialRetryDelayMillis - the initial delay time,
 *   in milliseconds, between the completion of the first failed request and the
 *   initiation of the first retrying request.
 * @property ***REMOVED***number***REMOVED*** retryDelayMultiplier - the multiplier by which to
 *   increase the delay time between the completion of failed requests, and the
 *   initiation of the subsequent retrying request.
 * @property ***REMOVED***number***REMOVED*** maxRetryDelayMillis - the maximum delay time, in
 *   milliseconds, between requests. When this value is reached,
 *   ``retryDelayMultiplier`` will no longer be used to increase delay time.
 * @property ***REMOVED***number***REMOVED*** initialRpcTimeoutMillis - the initial timeout parameter
 *   to the request.
 * @propetry ***REMOVED***number***REMOVED*** rpcTimeoutMultiplier - the multiplier by which to
 *   increase the timeout parameter between failed requests.
 * @property ***REMOVED***number***REMOVED*** maxRpcTimeoutMillis - the maximum timeout parameter, in
 *   milliseconds, for a request. When this value is reached,
 *   ``rpcTimeoutMultiplier`` will no longer be used to increase the timeout.
 * @property ***REMOVED***number***REMOVED*** totalTimeoutMillis - the total time, in milliseconds,
 *   starting from when the initial request is sent, after which an error will
 *   be returned, regardless of the retrying attempts made meanwhile.
 */

/**
 * Parameter to configure bundling behavior.
 * @typedef ***REMOVED***Object***REMOVED*** BundleOptions
 * @property ***REMOVED***number***REMOVED*** elementCountThreshold -
 *   the bundled request will be sent once the count of outstanding elements
 *   in the repeated field reaches this value.
 * @property ***REMOVED***number***REMOVED*** elementCountLimit -
 *   represents a hard limit on the number of elements in the repeated field
 *   of the bundle; if adding a request to a bundle would exceed this value,
 *   the bundle is sent and the new request is added to a fresh bundle. It is
 *   invalid for a single request to exceed this limit.
 * @property ***REMOVED***number***REMOVED*** requestByteThreshold -
 *   the bundled request will be sent once the count of bytes in the request
 *   reaches this value. Note that this value is pessimistically approximated
 *   by summing the bytesizes of the elements in the repeated field, and
 *   therefore may be an under-approximation.
 * @property ***REMOVED***number***REMOVED*** requestByteLimit -
 *   represents a hard limit on the size of the bundled request; if adding
 *   a request to a bundle would exceed this value, the bundle is sent and
 *   the new request is added to a fresh bundle. It is invalid for a single
 *   request to exceed this limit. Note that this value is pessimistically
 *   approximated by summing the bytesizes of the elements in the repeated
 *   field, with a buffer applied to correspond to the resulting
 *   under-approximation.
 * @property ***REMOVED***number***REMOVED*** delayThreshold -
 *   the bundled request will be sent this amount of time after the first
 *   element in the bundle was added to it.
 */

/**
 * @param ***REMOVED***Object***REMOVED*** settings - An object containing parameters of this settings.
 * @param ***REMOVED***number***REMOVED*** settings.timeout - The client-side timeout for API calls.
 *   This parameter is ignored for retrying calls.
 * @param ***REMOVED***RetryOptions***REMOVED*** settings.retry - The configuration for retrying upon
 *   transient error. If set to null, this call will not retry.
 * @param ***REMOVED***boolean***REMOVED*** settings.autoPaginate - If there is no `pageDescriptor`,
 *   this attrbute has no meaning. Otherwise, determines whether a page streamed
 *   response should make the page structure transparent to the user by
 *   flattening the repeated field in the returned generator.
 * @param ***REMOVED***number***REMOVED*** settings.pageToken - If there is no `pageDescriptor`,
 *   this attribute has no meaning. Otherwise, determines the page token used in
 *   the page streaming request.
 * @param ***REMOVED***Object***REMOVED*** settings.otherArgs - Additional arguments to be passed to
 *   the API calls.
 * @param ***REMOVED***Function=***REMOVED*** settings.promise - A constructor for a promise that
 * implements the ES6 specification of promise. If not provided, native promises
 * will be used.
 *
 * @constructor
 */
function CallSettings(settings) ***REMOVED***
  settings = settings || ***REMOVED******REMOVED***;
  this.timeout = settings.timeout || 30 * 1000;
  this.retry = settings.retry;
  this.autoPaginate = 'autoPaginate' in settings ? settings.autoPaginate : true;
  this.pageToken = settings.pageToken;
  this.maxResults = settings.maxResults;
  this.otherArgs = settings.otherArgs || ***REMOVED******REMOVED***;
  this.bundleOptions = settings.bundleOptions;
  this.isBundling = 'isBundling' in settings ? settings.isBundling : true;
  this.longrunning = 'longrunning' in settings ? settings.longrunning : null;
  this.promise = 'promise' in settings ? settings.promise : Promise;
***REMOVED***
exports.CallSettings = CallSettings;

/**
 * Returns a new CallSettings merged from this and a CallOptions object.
 *
 * @param ***REMOVED***CallOptions***REMOVED*** options - an instance whose values override
 *   those in this object. If null, ``merge`` returns a copy of this
 *   object
 * @return ***REMOVED***CallSettings***REMOVED*** The merged CallSettings instance.
 */
CallSettings.prototype.merge = function merge(options) ***REMOVED***
  if (!options) ***REMOVED***
    return new CallSettings(this);
  ***REMOVED***
  var timeout = this.timeout;
  var retry = this.retry;
  var autoPaginate = this.autoPaginate;
  var pageToken = this.pageToken;
  var maxResults = this.maxResults;
  var otherArgs = this.otherArgs;
  var isBundling = this.isBundling;
  var longrunning = this.longrunning;
  var promise = this.promise;
  if ('timeout' in options) ***REMOVED***
    timeout = options.timeout;
  ***REMOVED***
  if ('retry' in options) ***REMOVED***
    retry = options.retry;
  ***REMOVED***

  if ('autoPaginate' in options && !options.autoPaginate) ***REMOVED***
    autoPaginate = false;
  ***REMOVED***

  if ('pageToken' in options) ***REMOVED***
    autoPaginate = false;
    pageToken = options.pageToken;
  ***REMOVED***

  if ('maxResults' in options) ***REMOVED***
    maxResults = options.maxResults;
  ***REMOVED***

  if ('otherArgs' in options) ***REMOVED***
    otherArgs = ***REMOVED******REMOVED***;
    for (var key in this.otherArgs) ***REMOVED***
      otherArgs[key] = this.otherArgs[key];
    ***REMOVED***
    for (var optionsKey in options.otherArgs) ***REMOVED***
      otherArgs[optionsKey] = options.otherArgs[optionsKey];
    ***REMOVED***
  ***REMOVED***

  if ('isBundling' in options) ***REMOVED***
    isBundling = options.isBundling;
  ***REMOVED***

  if ('maxRetries' in options) ***REMOVED***
    retry.backoffSettings.maxRetries = options.maxRetries;
    delete retry.backoffSettings.totalTimeoutMillis;
  ***REMOVED***

  if ('longrunning' in options) ***REMOVED***
    longrunning = options.longrunning;
  ***REMOVED***

  if ('promise' in options) ***REMOVED***
    promise = options.promise;
  ***REMOVED***

  return new CallSettings(***REMOVED***
    timeout: timeout,
    retry: retry,
    bundleOptions: this.bundleOptions,
    longrunning: longrunning,
    autoPaginate: autoPaginate,
    pageToken: pageToken,
    maxResults: maxResults,
    otherArgs: otherArgs,
    isBundling: isBundling,
    promise: promise,
  ***REMOVED***);
***REMOVED***;

/**
 * Per-call configurable settings for retrying upon transient failure.
 *
 * @param ***REMOVED***String[]***REMOVED*** retryCodes - a list of Google API canonical error codes
 *   upon which a retry should be attempted.
 * @param ***REMOVED***BackoffSettings***REMOVED*** backoffSettings - configures the retry
 *   exponential backoff algorithm.
 * @return ***REMOVED***RetryOptions***REMOVED*** A new RetryOptions object.
 *
 */
function createRetryOptions(retryCodes, backoffSettings) ***REMOVED***
  return ***REMOVED***
    retryCodes: retryCodes,
    backoffSettings: backoffSettings,
  ***REMOVED***;
***REMOVED***
exports.createRetryOptions = createRetryOptions;

/**
 * Parameters to the exponential backoff algorithm for retrying.
 *
 * @param ***REMOVED***number***REMOVED*** initialRetryDelayMillis - the initial delay time,
 *   in milliseconds, between the completion of the first failed request and the
 *   initiation of the first retrying request.
 * @param ***REMOVED***number***REMOVED*** retryDelayMultiplier - the multiplier by which to
 *   increase the delay time between the completion of failed requests, and the
 *   initiation of the subsequent retrying request.
 * @param ***REMOVED***number***REMOVED*** maxRetryDelayMillis - the maximum delay time, in
 *   milliseconds, between requests. When this value is reached,
 *   ``retryDelayMultiplier`` will no longer be used to increase delay time.
 * @param ***REMOVED***number***REMOVED*** initialRpcTimeoutMillis - the initial timeout parameter
 *   to the request.
 * @param ***REMOVED***number***REMOVED*** rpcTimeoutMultiplier - the multiplier by which to
 *   increase the timeout parameter between failed requests.
 * @param ***REMOVED***number***REMOVED*** maxRpcTimeoutMillis - the maximum timeout parameter, in
 *   milliseconds, for a request. When this value is reached,
 *   ``rpcTimeoutMultiplier`` will no longer be used to increase the timeout.
 * @param ***REMOVED***number***REMOVED*** totalTimeoutMillis - the total time, in milliseconds,
 *   starting from when the initial request is sent, after which an error will
 *   be returned, regardless of the retrying attempts made meanwhile.
 * @return ***REMOVED***BackoffSettings***REMOVED*** a new settings.
 *
 */
function createBackoffSettings(
  initialRetryDelayMillis,
  retryDelayMultiplier,
  maxRetryDelayMillis,
  initialRpcTimeoutMillis,
  rpcTimeoutMultiplier,
  maxRpcTimeoutMillis,
  totalTimeoutMillis
) ***REMOVED***
  return ***REMOVED***
    initialRetryDelayMillis: initialRetryDelayMillis,
    retryDelayMultiplier: retryDelayMultiplier,
    maxRetryDelayMillis: maxRetryDelayMillis,
    initialRpcTimeoutMillis: initialRpcTimeoutMillis,
    rpcTimeoutMultiplier: rpcTimeoutMultiplier,
    maxRpcTimeoutMillis: maxRpcTimeoutMillis,
    totalTimeoutMillis: totalTimeoutMillis,
  ***REMOVED***;
***REMOVED***
exports.createBackoffSettings = createBackoffSettings;

/**
 * Parameters to the exponential backoff algorithm for retrying.
 * This function is unsupported, and intended for internal use only.
 *
 * @param ***REMOVED***number***REMOVED*** initialRetryDelayMillis - the initial delay time,
 *   in milliseconds, between the completion of the first failed request and the
 *   initiation of the first retrying request.
 * @param ***REMOVED***number***REMOVED*** retryDelayMultiplier - the multiplier by which to
 *   increase the delay time between the completion of failed requests, and the
 *   initiation of the subsequent retrying request.
 * @param ***REMOVED***number***REMOVED*** maxRetryDelayMillis - the maximum delay time, in
 *   milliseconds, between requests. When this value is reached,
 *   ``retryDelayMultiplier`` will no longer be used to increase delay time.
 * @param ***REMOVED***number***REMOVED*** initialRpcTimeoutMillis - the initial timeout parameter
 *   to the request.
 * @param ***REMOVED***number***REMOVED*** rpcTimeoutMultiplier - the multiplier by which to
 *   increase the timeout parameter between failed requests.
 * @param ***REMOVED***number***REMOVED*** maxRpcTimeoutMillis - the maximum timeout parameter, in
 *   milliseconds, for a request. When this value is reached,
 *   ``rpcTimeoutMultiplier`` will no longer be used to increase the timeout.
 * @param ***REMOVED***number***REMOVED*** maxRetries - the maximum number of retrying attempts that
 *   will be made. If reached, an error will be returned.
 * @return ***REMOVED***BackoffSettings***REMOVED*** a new settings.
 *
 */
function createMaxRetriesBackoffSettings(
  initialRetryDelayMillis,
  retryDelayMultiplier,
  maxRetryDelayMillis,
  initialRpcTimeoutMillis,
  rpcTimeoutMultiplier,
  maxRpcTimeoutMillis,
  maxRetries
) ***REMOVED***
  return ***REMOVED***
    initialRetryDelayMillis: initialRetryDelayMillis,
    retryDelayMultiplier: retryDelayMultiplier,
    maxRetryDelayMillis: maxRetryDelayMillis,
    initialRpcTimeoutMillis: initialRpcTimeoutMillis,
    rpcTimeoutMultiplier: rpcTimeoutMultiplier,
    maxRpcTimeoutMillis: maxRpcTimeoutMillis,
    maxRetries: maxRetries,
  ***REMOVED***;
***REMOVED***
exports.createMaxRetriesBackoffSettings = createMaxRetriesBackoffSettings;

/**
 * Creates a new ***REMOVED***@link BundleOptions***REMOVED***.
 *
 * @private
 * @param ***REMOVED***Object***REMOVED*** options - An object to hold optional parameters. See
 *   properties for the content of options.
 * @return ***REMOVED***BundleOptions***REMOVED*** - A new options.
 */
function createBundleOptions(options) ***REMOVED***
  var params = [
    'element_count_threshold',
    'element_count_limit',
    'request_byte_threshold',
    'request_byte_limit',
    'delay_threshold_millis',
  ];
  params.forEach(function(param) ***REMOVED***
    if (param in options && typeof options[param] !== 'number') ***REMOVED***
      throw new Error(param + ' should be a number');
    ***REMOVED***
  ***REMOVED***);

  var elementCountThreshold = options.element_count_threshold || 0;
  var elementCountLimit = options.element_count_limit || 0;
  var requestByteThreshold = options.request_byte_threshold || 0;
  var requestByteLimit = options.request_byte_limit || 0;
  var delayThreshold = options.delay_threshold_millis || 0;

  if (
    elementCountThreshold === 0 &&
    requestByteThreshold === 0 &&
    delayThreshold === 0
  ) ***REMOVED***
    throw new Error('one threshold should be > 0');
  ***REMOVED***
  return ***REMOVED***
    elementCountThreshold: elementCountThreshold,
    elementCountLimit: elementCountLimit,
    requestByteThreshold: requestByteThreshold,
    requestByteLimit: requestByteLimit,
    delayThreshold: delayThreshold,
  ***REMOVED***;
***REMOVED***
exports.createBundleOptions = createBundleOptions;

/**
 * Helper for ***REMOVED***@link constructSettings***REMOVED***
 *
 * @private
 *
 * @param ***REMOVED***Object***REMOVED*** methodConfig - A dictionary representing a single
 *   `methods` entry of the standard API client config file. (See
 *   ***REMOVED***@link constructSettings***REMOVED*** for information on this yaml.)
 * @param ***REMOVED***?Object***REMOVED*** retryCodes - A dictionary parsed from the
 *   `retry_codes_def` entry of the standard API client config
 *   file. (See ***REMOVED***@link constructSettings***REMOVED*** for information on this yaml.)
 * @param ***REMOVED***Object***REMOVED*** retryParams - A dictionary parsed from the
 *   `retry_params` entry of the standard API client config
 *   file. (See ***REMOVED***@link constructSettings***REMOVED*** for information on this yaml.)
 * @param ***REMOVED***Object***REMOVED*** retryNames - A dictionary mapping the string names
 *   used in the standard API client config file to API response
 *   status codes.
 * @return ***REMOVED***?RetryOptions***REMOVED*** The new retry options.
 */
function constructRetry(methodConfig, retryCodes, retryParams, retryNames) ***REMOVED***
  if (!methodConfig) ***REMOVED***
    return null;
  ***REMOVED***

  var codes = null;
  if (retryCodes && 'retry_codes_name' in methodConfig) ***REMOVED***
    var retryCodesName = methodConfig.retry_codes_name;
    codes = (retryCodes[retryCodesName] || []).map(function(name) ***REMOVED***
      return retryNames[name];
    ***REMOVED***);
  ***REMOVED***

  var backoffSettings = null;
  if (retryParams && 'retry_params_name' in methodConfig) ***REMOVED***
    var params = retryParams[methodConfig.retry_params_name];
    backoffSettings = createBackoffSettings(
      params.initial_retry_delay_millis,
      params.retry_delay_multiplier,
      params.max_retry_delay_millis,
      params.initial_rpc_timeout_millis,
      params.rpc_timeout_multiplier,
      params.max_rpc_timeout_millis,
      params.total_timeout_millis
    );
  ***REMOVED***
  return createRetryOptions(codes, backoffSettings);
***REMOVED***

/**
 * Helper for ***REMOVED***@link constructSettings***REMOVED***
 *
 * Takes two retry options, and merges them into a single RetryOption instance.
 *
 * @private
 *
 * @param ***REMOVED***RetryOptions***REMOVED*** retry - The base RetryOptions.
 * @param ***REMOVED***RetryOptions***REMOVED*** overrides - The RetryOptions used for overriding
 *   `retry`. Use the values if it is not null. If entire `overrides` is null,
 *   ignore the base retry and return null.
 * @return ***REMOVED***?RetryOptions***REMOVED*** The merged RetryOptions.
 */
function mergeRetryOptions(retry, overrides) ***REMOVED***
  if (!overrides) ***REMOVED***
    return null;
  ***REMOVED***

  if (!overrides.retryCodes && !overrides.backoffSettings) ***REMOVED***
    return retry;
  ***REMOVED***

  var codes = retry.retryCodes;
  if (overrides.retryCodes) ***REMOVED***
    codes = overrides.retryCodes;
  ***REMOVED***
  var backoffSettings = retry.backoffSettings;
  if (overrides.backoffSettings) ***REMOVED***
    backoffSettings = overrides.backoffSettings;
  ***REMOVED***
  return createRetryOptions(codes, backoffSettings);
***REMOVED***

/**
 * Constructs a dictionary mapping method names to ***REMOVED***@link CallSettings***REMOVED***.
 *
 * The `clientConfig` parameter is parsed from a client configuration JSON
 * file of the form:
 *
 *     ***REMOVED***
 *       "interfaces": ***REMOVED***
 *         "google.fake.v1.ServiceName": ***REMOVED***
 *           "retry_codes": ***REMOVED***
 *             "idempotent": ["UNAVAILABLE", "DEADLINE_EXCEEDED"],
 *             "non_idempotent": []
 *           ***REMOVED***,
 *           "retry_params": ***REMOVED***
 *             "default": ***REMOVED***
 *               "initial_retry_delay_millis": 100,
 *               "retry_delay_multiplier": 1.2,
 *               "max_retry_delay_millis": 1000,
 *               "initial_rpc_timeout_millis": 2000,
 *               "rpc_timeout_multiplier": 1.5,
 *               "max_rpc_timeout_millis": 30000,
 *               "total_timeout_millis": 45000
 *             ***REMOVED***
 *           ***REMOVED***,
 *           "methods": ***REMOVED***
 *             "CreateFoo": ***REMOVED***
 *               "retry_codes_name": "idempotent",
 *               "retry_params_name": "default"
 *             ***REMOVED***,
 *             "Publish": ***REMOVED***
 *               "retry_codes_name": "non_idempotent",
 *               "retry_params_name": "default",
 *               "bundling": ***REMOVED***
 *                 "element_count_threshold": 40,
 *                 "element_count_limit": 200,
 *                 "request_byte_threshold": 90000,
 *                 "request_byte_limit": 100000,
 *                 "delay_threshold_millis": 100
 *               ***REMOVED***
 *             ***REMOVED***
 *           ***REMOVED***
 *         ***REMOVED***
 *       ***REMOVED***
 *     ***REMOVED***
 *
 * @param ***REMOVED***String***REMOVED*** serviceName - The fully-qualified name of this
 *   service, used as a key into the client config file (in the
 *   example above, this value should be 'google.fake.v1.ServiceName').
 * @param ***REMOVED***Object***REMOVED*** clientConfig - A dictionary parsed from the
 *   standard API client config file.
 * @param ***REMOVED***Object***REMOVED*** configOverrides - A dictionary in the same structure of
 *   client_config to override the settings.
 * @param ***REMOVED***Object.<string, string[]>***REMOVED*** retryNames - A dictionary mapping the strings
 *   referring to response status codes to objects representing
 *   those codes.
 * @param ***REMOVED***Object***REMOVED*** otherArgs - the non-request arguments to be passed to the API
 *   calls.
 * @param ***REMOVED***Function=***REMOVED*** promise - A constructor for a promise that implements the
 * ES6 specification of promise. If not provided, native promises will be used.
 * @return ***REMOVED***Object***REMOVED*** A mapping from method name to CallSettings, or null if the
 *   service is not found in the config.
 */
exports.constructSettings = function constructSettings(
  serviceName,
  clientConfig,
  configOverrides,
  retryNames,
  otherArgs,
  promise
) ***REMOVED***
  otherArgs = otherArgs || ***REMOVED******REMOVED***;
  var defaults = ***REMOVED******REMOVED***;

  var serviceConfig = (clientConfig.interfaces || ***REMOVED******REMOVED***)[serviceName];
  if (!serviceConfig) ***REMOVED***
    return null;
  ***REMOVED***

  var overrides = (configOverrides.interfaces || ***REMOVED******REMOVED***)[serviceName] || ***REMOVED******REMOVED***;

  var methods = serviceConfig.methods;
  var overridingMethods = overrides.methods || ***REMOVED******REMOVED***;
  for (var methodName in methods) ***REMOVED***
    var methodConfig = methods[methodName];
    var jsName = methodName[0].toLowerCase() + methodName.slice(1);

    var retry = constructRetry(
      methodConfig,
      serviceConfig.retry_codes,
      serviceConfig.retry_params,
      retryNames
    );
    var bundlingConfig = methodConfig.bundling;
    var timeout = methodConfig.timeout_millis;
    if (methodName in overridingMethods) ***REMOVED***
      var overridingMethod = overridingMethods[methodName];
      if (overridingMethod) ***REMOVED***
        if ('bundling' in overridingMethod) ***REMOVED***
          bundlingConfig = overridingMethod.bundling;
        ***REMOVED***
        if ('timeout_millis' in overridingMethod) ***REMOVED***
          timeout = overridingMethod.timeout_millis;
        ***REMOVED***
      ***REMOVED***
      retry = mergeRetryOptions(
        retry,
        constructRetry(
          overridingMethod,
          overrides.retry_codes,
          overrides.retry_params,
          retryNames
        )
      );
    ***REMOVED***

    defaults[jsName] = new CallSettings(***REMOVED***
      timeout: timeout,
      retry: retry,
      bundleOptions: bundlingConfig
        ? createBundleOptions(bundlingConfig)
        : null,
      otherArgs: otherArgs,
      promise: promise || Promise,
    ***REMOVED***);
  ***REMOVED***
  return defaults;
***REMOVED***;
