// google-gax/lib/grpc.js
/*
 * Copyright 2016, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */
'use strict';

var autoAuth = require('google-auto-auth');

var fs = require('fs');
var gax = require('./gax');
var globby = require('globby');
var googleProtoFilesDir = require('google-proto-files')('..');
var path = require('path');
var protobuf = require('protobufjs');
var util = require('util');

googleProtoFilesDir = path.normalize(googleProtoFilesDir);

var COMMON_PROTO_DIRS = [
  // This list of directories is defined here:
  // https://github.com/googleapis/googleapis/blob/master/gapic/packaging/common_protos.yaml
  'api',
  path.join('iam', 'v1'),
  path.join('logging', 'type'),
  'longrunning',
  'protobuf', // This is an additional path that the common protos depend on.
  'rpc',
  'type',
];

var COMMON_PROTO_GLOB_PATTERNS = COMMON_PROTO_DIRS.map(function(dir) ***REMOVED***
  return path.join(googleProtoFilesDir, 'google', dir, '**', '*.proto');
***REMOVED***);

var COMMON_PROTO_FILES = globby
  .sync(COMMON_PROTO_GLOB_PATTERNS)
  .map(function(filename) ***REMOVED***
    return path.normalize(filename);
  ***REMOVED***)
  .map(function(filename) ***REMOVED***
    return filename.substring(googleProtoFilesDir.length + 1);
  ***REMOVED***);

/**
 * A class which keeps the context of gRPC and auth for the gRPC.
 *
 * @param ***REMOVED***Object=***REMOVED*** options - The optional parameters. It will be directly
 *   passed to google-auto-auth library, so parameters like keyFile or
 *   credentials will be valid.
 * @param ***REMOVED***Object=***REMOVED*** options.auth - An instance of google-auto-auth.
 *   When specified, this auth instance will be used instead of creating
 *   a new one.
 * @param ***REMOVED***Object=***REMOVED*** options.grpc - When specified, this will be used
 *   for the 'grpc' module in this context. By default, it will load the grpc
 *   module in the standard way.
 * @param ***REMOVED***Function=***REMOVED*** options.promise - A constructor for a promise that
 * implements the ES6 specification of promise. If not provided, native promises
 * will be used.
 * @constructor
 */
function GrpcClient(options) ***REMOVED***
  if (!(this instanceof GrpcClient)) ***REMOVED***
    return new GrpcClient(options);
  ***REMOVED***
  options = options || ***REMOVED******REMOVED***;
  this.auth = options.auth || autoAuth(options);
  this.promise = options.promise || Promise;
  if ('grpc' in options) ***REMOVED***
    this.grpc = options.grpc;
    this.grpcVersion = '';
  ***REMOVED*** else ***REMOVED***
    this.grpc = require('grpc');
    this.grpcVersion = require('grpc/package.json').version;
  ***REMOVED***
***REMOVED***
module.exports = GrpcClient;

/**
 * Creates a gRPC credentials. It asks the auth data if necessary.
 * @private
 * @param ***REMOVED***Object***REMOVED*** opts - options values for configuring credentials.
 * @param ***REMOVED***Object=***REMOVED*** opts.sslCreds - when specified, this is used instead
 *   of default channel credentials.
 * @return ***REMOVED***Promise***REMOVED*** The promise which will be resolved to the gRPC credential.
 */
GrpcClient.prototype._getCredentials = function _getCredentials(opts) ***REMOVED***
  var PromiseCtor = this.promise;
  if (opts.sslCreds) ***REMOVED***
    return PromiseCtor.resolve(opts.sslCreds);
  ***REMOVED***
  var grpc = this.grpc;
  var getAuthClient = this.auth.getAuthClient.bind(this.auth);
  var sslCreds = grpc.credentials.createSsl();
  return new PromiseCtor(function(resolve, reject) ***REMOVED***
    getAuthClient(function(err, auth) ***REMOVED***
      if (err) ***REMOVED***
        reject(err);
      ***REMOVED*** else ***REMOVED***
        var credentials = grpc.credentials.combineChannelCredentials(
          sslCreds,
          grpc.credentials.createFromGoogleCredential(auth)
        );
        resolve(credentials);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * Load grpc proto services with the specific arguments.
 * @param ***REMOVED***Array=***REMOVED*** args - The argument list to be passed to grpc.load().
 * @return ***REMOVED***Object***REMOVED*** The gRPC loaded result (the toplevel namespace object).
 */
GrpcClient.prototype.load = function(args) ***REMOVED***
  if (!args) ***REMOVED***
    args = [];
  ***REMOVED*** else if (!Array.isArray(args)) ***REMOVED***
    args = [args];
  ***REMOVED***
  if (args.length === 1) ***REMOVED***
    args.push('proto', ***REMOVED***convertFieldsToCamelCase: true***REMOVED***);
  ***REMOVED***
  return this.grpc.load.apply(this.grpc, args);
***REMOVED***;

/**
 * Load grpc proto service from a filename hooking in googleapis common protos
 * when necessary.
 * @param ***REMOVED***String***REMOVED*** protoPath - The directory to search for the protofile.
 * @param ***REMOVED***String***REMOVED*** filename - The filename of the proto to be loaded.
 * @return ***REMOVED***Object<string, *>***REMOVED*** The gRPC loaded result (the toplevel namespace
 *   object).
 */
GrpcClient.prototype.loadProto = function(protoPath, filename) ***REMOVED***
  var resolvedPath = GrpcClient._resolveFile(protoPath, filename);
  return this.grpc.loadObject(
    protobuf.loadSync(resolvedPath, new GoogleProtoFilesRoot())
  );
***REMOVED***;

GrpcClient._resolveFile = function(protoPath, filename) ***REMOVED***
  if (fs.existsSync(path.join(protoPath, filename))) ***REMOVED***
    return path.join(protoPath, filename);
  ***REMOVED*** else if (COMMON_PROTO_FILES.indexOf(filename) > -1) ***REMOVED***
    return path.join(googleProtoFilesDir, filename);
  ***REMOVED***
  throw new Error(filename + ' could not be found in ' + protoPath);
***REMOVED***;

GrpcClient.prototype.metadataBuilder = function(headers) ***REMOVED***
  var Metadata = this.grpc.Metadata;
  var baseMetadata = new Metadata();
  for (var key in headers) ***REMOVED***
    baseMetadata.set(key, headers[key]);
  ***REMOVED***
  return function buildMetadata(abTests, moreHeaders) ***REMOVED***
    // TODO: bring the A/B testing info into the metadata.
    var copied = false;
    var metadata = baseMetadata;
    for (var key in moreHeaders) ***REMOVED***
      if (
        key.toLowerCase() !== 'x-goog-api-client' &&
        moreHeaders.hasOwnProperty(key)
      ) ***REMOVED***
        if (!copied) ***REMOVED***
          copied = true;
          metadata = metadata.clone();
        ***REMOVED***
        metadata.set(key, moreHeaders[key]);
      ***REMOVED***
    ***REMOVED***
    return metadata;
  ***REMOVED***;
***REMOVED***;

/**
 * A wrapper of ***REMOVED***@link constructSettings***REMOVED*** function with under the gRPC context.
 *
 * Most of parameters are common among constructSettings, please take a look.
 * @param ***REMOVED***string***REMOVED*** serviceName - The fullly-qualified name of the service.
 * @param ***REMOVED***Object***REMOVED*** clientConfig - A dictionary of the client config.
 * @param ***REMOVED***Object***REMOVED*** configOverrides - A dictionary of overriding configs.
 * @param ***REMOVED***Object***REMOVED*** headers - A dictionary of additional HTTP header name to
 *   its value.
 * @return ***REMOVED***Object***REMOVED*** A mapping of method names to CallSettings.
 */
GrpcClient.prototype.constructSettings = function constructSettings(
  serviceName,
  clientConfig,
  configOverrides,
  headers
) ***REMOVED***
  return gax.constructSettings(
    serviceName,
    clientConfig,
    configOverrides,
    this.grpc.status,
    ***REMOVED***metadataBuilder: this.metadataBuilder(headers)***REMOVED***,
    this.promise
  );
***REMOVED***;

/**
 * Creates a gRPC stub with current gRPC and auth.
 * @param ***REMOVED***function***REMOVED*** CreateStub - The constructor function of the stub.
 * @param ***REMOVED***Object***REMOVED*** options - The optional arguments to customize
 *   gRPC connection. This options will be passed to the constructor of
 *   gRPC client too.
 * @param ***REMOVED***string***REMOVED*** options.servicePath - The name of the server of the service.
 * @param ***REMOVED***number***REMOVED*** options.port - The port of the service.
 * @param ***REMOVED***grpc.ClientCredentials=***REMOVED*** options.sslCreds - The credentials to be used
 *   to set up gRPC connection.
 * @return ***REMOVED***Promise***REMOVED*** A promse which resolves to a gRPC stub instance.
 */
GrpcClient.prototype.createStub = function(CreateStub, options) ***REMOVED***
  var serviceAddress = options.servicePath + ':' + options.port;
  return this._getCredentials(options).then(function(credentials) ***REMOVED***
    var grpcOptions = ***REMOVED******REMOVED***;
    Object.keys(options).forEach(function(key) ***REMOVED***
      if (key.indexOf('grpc.') === 0) ***REMOVED***
        grpcOptions[key] = options[key];
      ***REMOVED***
    ***REMOVED***);
    return new CreateStub(serviceAddress, credentials, grpcOptions);
  ***REMOVED***);
***REMOVED***;

/**
 * Creates a 'bytelength' function for a given proto message class.
 *
 * See ***REMOVED***@link BundleDescriptor***REMOVED*** about the meaning of the return value.
 *
 * @param ***REMOVED***function***REMOVED*** message - a constructor function that is generated by
 *   protobuf.js. Assumes 'encoder' field in the message.
 * @return ***REMOVED***function(Object):number***REMOVED*** - a function to compute the byte length
 *   for an object.
 */
GrpcClient.createByteLengthFunction = function createByteLengthFunction(
  message
) ***REMOVED***
  return function getByteLength(obj) ***REMOVED***
    return message.encode(obj).finish().length;
  ***REMOVED***;
***REMOVED***;

function GoogleProtoFilesRoot() ***REMOVED***
  protobuf.Root.call(this, [].slice.apply(arguments));
***REMOVED***
util.inherits(GoogleProtoFilesRoot, protobuf.Root);

module.exports.GoogleProtoFilesRoot = GoogleProtoFilesRoot;

// Causes the loading of an included proto to check if it is a common
// proto. If it is a common proto, use the google-proto-files proto.
GoogleProtoFilesRoot.prototype.resolvePath = function(originPath, includePath) ***REMOVED***
  originPath = path.normalize(originPath);
  includePath = path.normalize(includePath);

  // Fully qualified paths don't need to be resolved.
  if (path.isAbsolute(includePath)) ***REMOVED***
    if (!fs.existsSync(includePath)) ***REMOVED***
      throw new Error('The include `' + includePath + '` was not found.');
    ***REMOVED***
    return includePath;
  ***REMOVED***

  if (COMMON_PROTO_FILES.indexOf(includePath) > -1) ***REMOVED***
    return path.join(googleProtoFilesDir, includePath);
  ***REMOVED***

  return GoogleProtoFilesRoot._findIncludePath(originPath, includePath);
***REMOVED***;

GoogleProtoFilesRoot._findIncludePath = function(originPath, includePath) ***REMOVED***
  originPath = path.normalize(originPath);
  includePath = path.normalize(includePath);

  var current = originPath;
  var found = fs.existsSync(path.join(current, includePath));
  while (!found && current.length > 0) ***REMOVED***
    current = current.substring(0, current.lastIndexOf(path.sep));
    found = fs.existsSync(path.join(current, includePath));
  ***REMOVED***
  if (!found) ***REMOVED***
    throw new Error('The include `' + includePath + '` was not found.');
  ***REMOVED***
  return path.join(current, includePath);
***REMOVED***;
