// google-gax/lib/operations_client.js
/*
 * Copyright 2016 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * EDITING INSTRUCTIONS
 * This file was generated from the file
 * https://github.com/googleapis/googleapis/blob/master/google/longrunning/operations.proto,
 * and updates to that file get reflected here through a refresh process.
 * For the short term, the refresh process will only be runnable by Google
 * engineers.
 *
 * The only allowed edits are to method and file documentation. A 3-way
 * merge preserves those additions if the generated source changes.
 */
/* TODO: introduce line-wrapping so that it never exceeds the limit. */
/* jscs: disable maximumLineLength */
'use strict';

var configData = require('./operations_client_config');
var extend = require('extend');
var gax = require('./gax');
extend(gax, require('./api_callable'));
extend(gax, require('./path_template'));
extend(gax, require('./paged_iteration'));

var SERVICE_ADDRESS = 'longrunning.googleapis.com';

var DEFAULT_SERVICE_PORT = 443;

var CODE_GEN_NAME_VERSION = 'gapic/0.7.1';

var PAGE_DESCRIPTORS = ***REMOVED***
  listOperations: new gax.PageDescriptor(
    'pageToken',
    'nextPageToken',
    'operations'
  ),
***REMOVED***;

/**
 * The scopes needed to make gRPC calls to all of the methods defined in
 * this service.
 */
var ALL_SCOPES = [];

/**
 * Manages long-running operations with an API service.
 *
 * When an API method normally takes long time to complete, it can be designed
 * to return ***REMOVED***@link Operation***REMOVED*** to the client, and the client can use this
 * interface to receive the real response asynchronously by polling the
 * operation resource, or pass the operation resource to another API (such as
 * Google Cloud Pub/Sub API) to receive the response.  Any API service that
 * returns long-running operations should implement the `Operations` interface
 * so developers can have a consistent client experience.
 *
 * This will be created through a builder function which can be obtained by the module.
 * See the following example of how to initialize the module and how to access to the builder.
 * @see ***REMOVED***@link operationsClient***REMOVED***
 *
 * @class
 */
function OperationsClient(gaxGrpc, grpcClients, opts) ***REMOVED***
  opts = extend(
    ***REMOVED***
      servicePath: SERVICE_ADDRESS,
      port: DEFAULT_SERVICE_PORT,
      clientConfig: ***REMOVED******REMOVED***,
    ***REMOVED***,
    opts
  );

  var googleApiClient = ['gl-node/' + process.versions.node];
  if (opts.libName && opts.libVersion) ***REMOVED***
    googleApiClient.push(opts.libName + '/' + opts.libVersion);
  ***REMOVED***
  googleApiClient.push(
    CODE_GEN_NAME_VERSION,
    'gax/' + gax.version,
    'grpc/' + gaxGrpc.grpcVersion
  );

  var defaults = gaxGrpc.constructSettings(
    'google.longrunning.Operations',
    configData,
    opts.clientConfig,
    ***REMOVED***'x-goog-api-client': googleApiClient.join(' ')***REMOVED***
  );

  var self = this;

  this.auth = gaxGrpc.auth;
  var operationsStub = gaxGrpc.createStub(
    grpcClients.google.longrunning.Operations,
    opts
  );
  var operationsStubMethods = [
    'getOperation',
    'listOperations',
    'cancelOperation',
    'deleteOperation',
  ];
  operationsStubMethods.forEach(function(methodName) ***REMOVED***
    self['_' + methodName] = gax.createApiCall(
      operationsStub.then(function(operationsStub) ***REMOVED***
        return function() ***REMOVED***
          var args = Array.prototype.slice.call(arguments, 0);
          return operationsStub[methodName].apply(operationsStub, args);
        ***REMOVED***;
      ***REMOVED***),
      defaults[methodName],
      PAGE_DESCRIPTORS[methodName]
    );
  ***REMOVED***);
***REMOVED***

/**
 * Get the project ID used by this class.
 * @aram ***REMOVED***function(Error, string)***REMOVED*** callback - the callback to be called with
 *   the current project Id.
 */
OperationsClient.prototype.getProjectId = function(callback) ***REMOVED***
  return this.auth.getProjectId(callback);
***REMOVED***;

// Service calls

/**
 * Gets the latest state of a long-running operation.  Clients can use this
 * method to poll the operation result at intervals as recommended by the API
 * service.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The name of the operation resource.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is an object representing [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"***REMOVED***.
 * @return ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is an object representing [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"***REMOVED***.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var client = longrunning.operationsClient();
 * var name = '';
 * client.getOperation(***REMOVED***name: name***REMOVED***).then(function(responses) ***REMOVED***
 *     var response = responses[0];
 *     // doThingsWith(response)
 * ***REMOVED***).catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
OperationsClient.prototype.getOperation = function(request, options, callback) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._getOperation(request, options, callback);
***REMOVED***;

/**
 * Lists operations that match the specified filter in the request. If the
 * server doesn't support this method, it returns `UNIMPLEMENTED`.
 *
 * NOTE: the `name` binding below allows API services to override the binding
 * to use different resource name schemes.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The name of the operation collection.
 * @param ***REMOVED***string***REMOVED*** request.filter
 *   The standard list filter.
 * @param ***REMOVED***number=***REMOVED*** request.pageSize
 *   The maximum number of resources contained in the underlying API
 *   response. If page streaming is performed per-resource, this
 *   parameter does not affect the return value. If page streaming is
 *   performed per-page, this determines the maximum number of
 *   resources in a page.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Array, ?Object, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is Array of [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"***REMOVED***.
 *
 *   When autoPaginate: false is specified through options, it contains the result
 *   in a single response. If the response indicates the next page exists, the third
 *   parameter is set to be used for the next request object. The fourth parameter keeps
 *   the raw response object of an object representing [google.longrunning.ListOperationsResponse]***REMOVED***@link external:"google.longrunning.ListOperationsResponse"***REMOVED***.
 * @return ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is Array of [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"***REMOVED***.
 *
 *   When autoPaginate: false is specified through options, the array has three elements.
 *   The first element is Array of [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"***REMOVED*** in a single response.
 *   The second element is the next request object if the response
 *   indicates the next page exists, or null. The third element is
 *   an object representing [google.longrunning.ListOperationsResponse]***REMOVED***@link external:"google.longrunning.ListOperationsResponse"***REMOVED***.
 *
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var client = longrunning.operationsClient();
 * var name = '';
 * var filter = '';
 * var request = ***REMOVED***
 *     name: name,
 *     filter: filter
 * ***REMOVED***;
 * // Iterate over all elements.
 * client.listOperations(request).then(function(responses) ***REMOVED***
 *     var resources = responses[0];
 *     for (var i = 0; i < resources.length; ++i) ***REMOVED***
 *         // doThingsWith(resources[i])
 *     ***REMOVED***
 * ***REMOVED***).catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 *
 * // Or obtain the paged response.
 * var options = ***REMOVED***autoPaginate: false***REMOVED***;
 * function callback(responses) ***REMOVED***
 *     // The actual resources in a response.
 *     var resources = responses[0];
 *     // The next request if the response shows there's more responses.
 *     var nextRequest = responses[1];
 *     // The actual response object, if necessary.
 *     // var rawResponse = responses[2];
 *     for (var i = 0; i < resources.length; ++i) ***REMOVED***
 *         // doThingsWith(resources[i]);
 *     ***REMOVED***
 *     if (nextRequest) ***REMOVED***
 *         // Fetch the next page.
 *         return client.listOperations(nextRequest, options).then(callback);
 *     ***REMOVED***
 * ***REMOVED***
 * client.listOperations(request, options)
 *     .then(callback)
 *     .catch(function(err) ***REMOVED***
 *         console.error(err);
 *     ***REMOVED***);
 */
OperationsClient.prototype.listOperations = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._listOperations(request, options, callback);
***REMOVED***;

/**
 * Equivalent to ***REMOVED***@link listOperations***REMOVED***, but returns a NodeJS Stream object.
 *
 * This fetches the paged responses for ***REMOVED***@link listOperations***REMOVED*** continuously
 * and invokes the callback registered for 'data' event for each element in the
 * responses.
 *
 * The returned object has 'end' method when no more elements are required.
 *
 * autoPaginate option will be ignored.
 *
 * @see ***REMOVED***@link https://nodejs.org/api/stream.html***REMOVED***
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The name of the operation collection.
 * @param ***REMOVED***string***REMOVED*** request.filter
 *   The standard list filter.
 * @param ***REMOVED***number=***REMOVED*** request.pageSize
 *   The maximum number of resources contained in the underlying API
 *   response. If page streaming is performed per-resource, this
 *   parameter does not affect the return value. If page streaming is
 *   performed per-page, this determines the maximum number of
 *   resources in a page.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @return ***REMOVED***Stream***REMOVED***
 *   An object stream which emits an object representing [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"***REMOVED*** on 'data' event.
 *
 * @example
 *
 * var client = longrunning.operationsClient();
 * var name = '';
 * var filter = '';
 * var request = ***REMOVED***
 *     name: name,
 *     filter: filter
 * ***REMOVED***;
 * client.listOperationsStream(request).on('data', function(element) ***REMOVED***
 *     // doThingsWith(element)
 * ***REMOVED***).on('error', function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
OperationsClient.prototype.listOperationsStream = function(request, options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return PAGE_DESCRIPTORS.listOperations.createStream(
    this._listOperations,
    request,
    options
  );
***REMOVED***;

/**
 * Starts asynchronous cancellation on a long-running operation.  The server
 * makes a best effort to cancel the operation, but success is not
 * guaranteed.  If the server doesn't support this method, it returns
 * `google.rpc.Code.UNIMPLEMENTED`.  Clients can use
 * ***REMOVED***@link Operations.GetOperation***REMOVED*** or
 * other methods to check whether the cancellation succeeded or whether the
 * operation completed despite cancellation. On successful cancellation,
 * the operation is not deleted; instead, it becomes an operation with
 * an ***REMOVED***@link Operation.error***REMOVED*** value with a ***REMOVED***@link google.rpc.Status.code***REMOVED*** of 1,
 * corresponding to `Code.CANCELLED`.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The name of the operation resource to be cancelled.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 * @return ***REMOVED***Promise***REMOVED*** - The promise which resolves when API call finishes.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var client = longrunning.operationsClient();
 * var name = '';
 * client.cancelOperation(***REMOVED***name: name***REMOVED***).catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
OperationsClient.prototype.cancelOperation = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._cancelOperation(request, options, callback);
***REMOVED***;

/**
 * Deletes a long-running operation. This method indicates that the client is
 * no longer interested in the operation result. It does not cancel the
 * operation. If the server doesn't support this method, it returns
 * `google.rpc.Code.UNIMPLEMENTED`.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The name of the operation resource to be deleted.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 * @return ***REMOVED***Promise***REMOVED*** - The promise which resolves when API call finishes.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var client = longrunning.operationsClient();
 * var name = '';
 * client.deleteOperation(***REMOVED***name: name***REMOVED***).catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
OperationsClient.prototype.deleteOperation = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._deleteOperation(request, options, callback);
***REMOVED***;

function OperationsClientBuilder(gaxGrpc) ***REMOVED***
  if (!(this instanceof OperationsClientBuilder)) ***REMOVED***
    return new OperationsClientBuilder(gaxGrpc);
  ***REMOVED***

  var operationsClient = gaxGrpc.load([
    ***REMOVED***
      root: require('google-proto-files')('..'),
      file: 'google/longrunning/operations.proto',
    ***REMOVED***,
  ]);
  extend(this, operationsClient.google.longrunning);

  /**
   * Build a new instance of ***REMOVED***@link OperationsClient***REMOVED***.
   *
   * @param ***REMOVED***Object=***REMOVED*** opts - The optional parameters.
   * @param ***REMOVED***String=***REMOVED*** opts.servicePath
   *   The domain name of the API remote host.
   * @param ***REMOVED***number=***REMOVED*** opts.port
   *   The port on which to connect to the remote host.
   * @param ***REMOVED***grpc.ClientCredentials=***REMOVED*** opts.sslCreds
   *   A ClientCredentials for use with an SSL-enabled channel.
   * @param ***REMOVED***Object=***REMOVED*** opts.clientConfig
   *   The customized config to build the call settings. See
   *   ***REMOVED***@link gax.constructSettings***REMOVED*** for the format.
   */
  this.operationsClient = function(opts) ***REMOVED***
    return new OperationsClient(gaxGrpc, operationsClient, opts);
  ***REMOVED***;
  extend(this.operationsClient, OperationsClient);
***REMOVED***
module.exports = OperationsClientBuilder;
module.exports.SERVICE_ADDRESS = SERVICE_ADDRESS;
module.exports.ALL_SCOPES = ALL_SCOPES;
