// google-gax/lib/paged_iteration.js
/**
 * Copyright 2016, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
'use strict';

var extend = require('extend');
var util = require('util');
var NormalApiCaller = require('./api_callable').NormalApiCaller;
var through2 = require('through2');
var ended = require('is-stream-ended');

/**
 * Creates an API caller that returns a stream to performs page-streaming.
 *
 * @private
 * @constructor
 * @param ***REMOVED***PageDescriptor***REMOVED*** pageDescriptor - indicates the structure
 *   of page streaming to be performed.
 */
function PagedIteration(pageDescriptor) ***REMOVED***
  NormalApiCaller.call(this);
  this.pageDescriptor = pageDescriptor;
***REMOVED***

util.inherits(PagedIteration, NormalApiCaller);

PagedIteration.prototype.createActualCallback = function(request, callback) ***REMOVED***
  var self = this;
  return function fetchNextPageToken(err, response) ***REMOVED***
    if (err) ***REMOVED***
      callback(err);
      return;
    ***REMOVED***
    var resources = response[self.pageDescriptor.resourceField];
    var pageToken = response[self.pageDescriptor.responsePageTokenField];
    if (pageToken) ***REMOVED***
      request[self.pageDescriptor.requestPageTokenField] = pageToken;
      callback(err, resources, request, response);
    ***REMOVED*** else ***REMOVED***
      callback(err, resources, null, response);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***;

PagedIteration.prototype.wrap = function(func) ***REMOVED***
  var self = this;
  return function wrappedCall(argument, metadata, options, callback) ***REMOVED***
    return func(
      argument,
      metadata,
      options,
      self.createActualCallback(argument, callback)
    );
  ***REMOVED***;
***REMOVED***;

PagedIteration.prototype.init = function(settings, callback) ***REMOVED***
  return NormalApiCaller.prototype.init.call(this, settings, callback);
***REMOVED***;

PagedIteration.prototype.call = function(
  apiCall,
  argument,
  settings,
  canceller
) ***REMOVED***
  argument = extend(***REMOVED******REMOVED***, argument);
  if (settings.pageToken) ***REMOVED***
    argument[this.pageDescriptor.requestPageTokenField] = settings.pageToken;
  ***REMOVED***
  if (settings.pageSize) ***REMOVED***
    argument[this.pageDescriptor.requestPageSizeField] = settings.pageSize;
  ***REMOVED***
  if (!settings.autoPaginate) ***REMOVED***
    NormalApiCaller.prototype.call.call(
      this,
      apiCall,
      argument,
      settings,
      canceller
    );
    return;
  ***REMOVED***

  var maxResults = settings.maxResults || -1;
  var allResources = [];
  function pushResources(err, resources, next) ***REMOVED***
    if (err) ***REMOVED***
      canceller.callback(err);
      return;
    ***REMOVED***

    for (var i = 0; i < resources.length; ++i) ***REMOVED***
      allResources.push(resources[i]);
      if (allResources.length === maxResults) ***REMOVED***
        next = null;
        break;
      ***REMOVED***
    ***REMOVED***
    if (!next) ***REMOVED***
      canceller.callback(null, allResources);
      return;
    ***REMOVED***
    setImmediate(apiCall, next, pushResources);
  ***REMOVED***

  setImmediate(apiCall, argument, pushResources);
***REMOVED***;

/**
 * Describes the structure of a page-streaming call.
 *
 * @property ***REMOVED***String***REMOVED*** requestPageTokenField
 * @property ***REMOVED***String***REMOVED*** responsePageTokenField
 * @property ***REMOVED***String***REMOVED*** resourceField
 *
 * @param ***REMOVED***String***REMOVED*** requestPageTokenField - The field name of the page token in
 *   the request.
 * @param ***REMOVED***String***REMOVED*** responsePageTokenField - The field name of the page token in
 *   the response.
 * @param ***REMOVED***String***REMOVED*** resourceField - The resource field name.
 *
 * @constructor
 */
function PageDescriptor(
  requestPageTokenField,
  responsePageTokenField,
  resourceField
) ***REMOVED***
  this.requestPageTokenField = requestPageTokenField;
  this.responsePageTokenField = responsePageTokenField;
  this.resourceField = resourceField;
***REMOVED***
exports.PageDescriptor = PageDescriptor;

/**
 * Creates a new object Stream which emits the resource on 'data' event.
 * @private
 * @param ***REMOVED***ApiCall***REMOVED*** apiCall - the callable object.
 * @param ***REMOVED***Object***REMOVED*** request - the request object.
 * @param ***REMOVED***CallOptions=***REMOVED*** options - the call options to customize the api call.
 * @return ***REMOVED***Stream***REMOVED*** - a new object Stream.
 */
PageDescriptor.prototype.createStream = function(apiCall, request, options) ***REMOVED***
  var stream = through2.obj();
  options = extend(***REMOVED******REMOVED***, options, ***REMOVED***autoPaginate: false***REMOVED***);
  var maxResults = 'maxResults' in options ? options.maxResults : -1;
  var pushCount = 0;
  var started = false;
  function callback(err, resources, next) ***REMOVED***
    if (err) ***REMOVED***
      stream.emit('error', err);
      return;
    ***REMOVED***
    for (var i = 0; i < resources.length; ++i) ***REMOVED***
      if (ended(stream)) ***REMOVED***
        return;
      ***REMOVED***
      if (resources[i] === null) ***REMOVED***
        continue;
      ***REMOVED***
      stream.push(resources[i]);
      pushCount++;
      if (pushCount === maxResults) ***REMOVED***
        stream.end();
      ***REMOVED***
    ***REMOVED***
    if (ended(stream)) ***REMOVED***
      return;
    ***REMOVED***
    if (!next) ***REMOVED***
      stream.end();
      return;
    ***REMOVED***
    // When pageToken is specified in the original options, it will overwrite
    // the page token field in the next request. Therefore it must be cleared.
    if ('pageToken' in options) ***REMOVED***
      delete options.pageToken;
    ***REMOVED***
    if (stream.isPaused()) ***REMOVED***
      request = next;
      started = false;
    ***REMOVED*** else ***REMOVED***
      setImmediate(apiCall, next, options, callback);
    ***REMOVED***
  ***REMOVED***
  stream.on('resume', function() ***REMOVED***
    if (!started) ***REMOVED***
      started = true;
      apiCall(request, options, callback);
    ***REMOVED***
  ***REMOVED***);
  return stream;
***REMOVED***;

/**
 * Returns a new API caller.
 * @private
 * @return ***REMOVED***PageStreamable***REMOVED*** - the page streaming caller.
 */
PageDescriptor.prototype.apiCaller = function() ***REMOVED***
  return new PagedIteration(this);
***REMOVED***;
