// estraverse/estraverse.js
/*
  Copyright (C) 2012-2013 Yusuke Suzuki <utatane.tea@gmail.com>
  Copyright (C) 2012 Ariya Hidayat <ariya.hidayat@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
/*jslint vars:false, bitwise:true*/
/*jshint indent:4*/
/*global exports:true*/
(function clone(exports) ***REMOVED***
    'use strict';

    var Syntax,
        isArray,
        VisitorOption,
        VisitorKeys,
        objectCreate,
        objectKeys,
        BREAK,
        SKIP,
        REMOVE;

    function ignoreJSHintError() ***REMOVED*** ***REMOVED***

    isArray = Array.isArray;
    if (!isArray) ***REMOVED***
        isArray = function isArray(array) ***REMOVED***
            return Object.prototype.toString.call(array) === '[object Array]';
        ***REMOVED***;
    ***REMOVED***

    function deepCopy(obj) ***REMOVED***
        var ret = ***REMOVED******REMOVED***, key, val;
        for (key in obj) ***REMOVED***
            if (obj.hasOwnProperty(key)) ***REMOVED***
                val = obj[key];
                if (typeof val === 'object' && val !== null) ***REMOVED***
                    ret[key] = deepCopy(val);
                ***REMOVED*** else ***REMOVED***
                    ret[key] = val;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***

    function shallowCopy(obj) ***REMOVED***
        var ret = ***REMOVED******REMOVED***, key;
        for (key in obj) ***REMOVED***
            if (obj.hasOwnProperty(key)) ***REMOVED***
                ret[key] = obj[key];
            ***REMOVED***
        ***REMOVED***
        return ret;
    ***REMOVED***
    ignoreJSHintError(shallowCopy);

    // based on LLVM libc++ upper_bound / lower_bound
    // MIT License

    function upperBound(array, func) ***REMOVED***
        var diff, len, i, current;

        len = array.length;
        i = 0;

        while (len) ***REMOVED***
            diff = len >>> 1;
            current = i + diff;
            if (func(array[current])) ***REMOVED***
                len = diff;
            ***REMOVED*** else ***REMOVED***
                i = current + 1;
                len -= diff + 1;
            ***REMOVED***
        ***REMOVED***
        return i;
    ***REMOVED***

    function lowerBound(array, func) ***REMOVED***
        var diff, len, i, current;

        len = array.length;
        i = 0;

        while (len) ***REMOVED***
            diff = len >>> 1;
            current = i + diff;
            if (func(array[current])) ***REMOVED***
                i = current + 1;
                len -= diff + 1;
            ***REMOVED*** else ***REMOVED***
                len = diff;
            ***REMOVED***
        ***REMOVED***
        return i;
    ***REMOVED***
    ignoreJSHintError(lowerBound);

    objectCreate = Object.create || (function () ***REMOVED***
        function F() ***REMOVED*** ***REMOVED***

        return function (o) ***REMOVED***
            F.prototype = o;
            return new F();
        ***REMOVED***;
    ***REMOVED***)();

    objectKeys = Object.keys || function (o) ***REMOVED***
        var keys = [], key;
        for (key in o) ***REMOVED***
            keys.push(key);
        ***REMOVED***
        return keys;
    ***REMOVED***;

    function extend(to, from) ***REMOVED***
        var keys = objectKeys(from), key, i, len;
        for (i = 0, len = keys.length; i < len; i += 1) ***REMOVED***
            key = keys[i];
            to[key] = from[key];
        ***REMOVED***
        return to;
    ***REMOVED***

    Syntax = ***REMOVED***
        AssignmentExpression: 'AssignmentExpression',
        AssignmentPattern: 'AssignmentPattern',
        ArrayExpression: 'ArrayExpression',
        ArrayPattern: 'ArrayPattern',
        ArrowFunctionExpression: 'ArrowFunctionExpression',
        AwaitExpression: 'AwaitExpression', // CAUTION: It's deferred to ES7.
        BlockStatement: 'BlockStatement',
        BinaryExpression: 'BinaryExpression',
        BreakStatement: 'BreakStatement',
        CallExpression: 'CallExpression',
        CatchClause: 'CatchClause',
        ClassBody: 'ClassBody',
        ClassDeclaration: 'ClassDeclaration',
        ClassExpression: 'ClassExpression',
        ComprehensionBlock: 'ComprehensionBlock',  // CAUTION: It's deferred to ES7.
        ComprehensionExpression: 'ComprehensionExpression',  // CAUTION: It's deferred to ES7.
        ConditionalExpression: 'ConditionalExpression',
        ContinueStatement: 'ContinueStatement',
        DebuggerStatement: 'DebuggerStatement',
        DirectiveStatement: 'DirectiveStatement',
        DoWhileStatement: 'DoWhileStatement',
        EmptyStatement: 'EmptyStatement',
        ExportAllDeclaration: 'ExportAllDeclaration',
        ExportDefaultDeclaration: 'ExportDefaultDeclaration',
        ExportNamedDeclaration: 'ExportNamedDeclaration',
        ExportSpecifier: 'ExportSpecifier',
        ExpressionStatement: 'ExpressionStatement',
        ForStatement: 'ForStatement',
        ForInStatement: 'ForInStatement',
        ForOfStatement: 'ForOfStatement',
        FunctionDeclaration: 'FunctionDeclaration',
        FunctionExpression: 'FunctionExpression',
        GeneratorExpression: 'GeneratorExpression',  // CAUTION: It's deferred to ES7.
        Identifier: 'Identifier',
        IfStatement: 'IfStatement',
        ImportDeclaration: 'ImportDeclaration',
        ImportDefaultSpecifier: 'ImportDefaultSpecifier',
        ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
        ImportSpecifier: 'ImportSpecifier',
        Literal: 'Literal',
        LabeledStatement: 'LabeledStatement',
        LogicalExpression: 'LogicalExpression',
        MemberExpression: 'MemberExpression',
        MetaProperty: 'MetaProperty',
        MethodDefinition: 'MethodDefinition',
        ModuleSpecifier: 'ModuleSpecifier',
        NewExpression: 'NewExpression',
        ObjectExpression: 'ObjectExpression',
        ObjectPattern: 'ObjectPattern',
        Program: 'Program',
        Property: 'Property',
        RestElement: 'RestElement',
        ReturnStatement: 'ReturnStatement',
        SequenceExpression: 'SequenceExpression',
        SpreadElement: 'SpreadElement',
        Super: 'Super',
        SwitchStatement: 'SwitchStatement',
        SwitchCase: 'SwitchCase',
        TaggedTemplateExpression: 'TaggedTemplateExpression',
        TemplateElement: 'TemplateElement',
        TemplateLiteral: 'TemplateLiteral',
        ThisExpression: 'ThisExpression',
        ThrowStatement: 'ThrowStatement',
        TryStatement: 'TryStatement',
        UnaryExpression: 'UnaryExpression',
        UpdateExpression: 'UpdateExpression',
        VariableDeclaration: 'VariableDeclaration',
        VariableDeclarator: 'VariableDeclarator',
        WhileStatement: 'WhileStatement',
        WithStatement: 'WithStatement',
        YieldExpression: 'YieldExpression'
    ***REMOVED***;

    VisitorKeys = ***REMOVED***
        AssignmentExpression: ['left', 'right'],
        AssignmentPattern: ['left', 'right'],
        ArrayExpression: ['elements'],
        ArrayPattern: ['elements'],
        ArrowFunctionExpression: ['params', 'body'],
        AwaitExpression: ['argument'], // CAUTION: It's deferred to ES7.
        BlockStatement: ['body'],
        BinaryExpression: ['left', 'right'],
        BreakStatement: ['label'],
        CallExpression: ['callee', 'arguments'],
        CatchClause: ['param', 'body'],
        ClassBody: ['body'],
        ClassDeclaration: ['id', 'superClass', 'body'],
        ClassExpression: ['id', 'superClass', 'body'],
        ComprehensionBlock: ['left', 'right'],  // CAUTION: It's deferred to ES7.
        ComprehensionExpression: ['blocks', 'filter', 'body'],  // CAUTION: It's deferred to ES7.
        ConditionalExpression: ['test', 'consequent', 'alternate'],
        ContinueStatement: ['label'],
        DebuggerStatement: [],
        DirectiveStatement: [],
        DoWhileStatement: ['body', 'test'],
        EmptyStatement: [],
        ExportAllDeclaration: ['source'],
        ExportDefaultDeclaration: ['declaration'],
        ExportNamedDeclaration: ['declaration', 'specifiers', 'source'],
        ExportSpecifier: ['exported', 'local'],
        ExpressionStatement: ['expression'],
        ForStatement: ['init', 'test', 'update', 'body'],
        ForInStatement: ['left', 'right', 'body'],
        ForOfStatement: ['left', 'right', 'body'],
        FunctionDeclaration: ['id', 'params', 'body'],
        FunctionExpression: ['id', 'params', 'body'],
        GeneratorExpression: ['blocks', 'filter', 'body'],  // CAUTION: It's deferred to ES7.
        Identifier: [],
        IfStatement: ['test', 'consequent', 'alternate'],
        ImportDeclaration: ['specifiers', 'source'],
        ImportDefaultSpecifier: ['local'],
        ImportNamespaceSpecifier: ['local'],
        ImportSpecifier: ['imported', 'local'],
        Literal: [],
        LabeledStatement: ['label', 'body'],
        LogicalExpression: ['left', 'right'],
        MemberExpression: ['object', 'property'],
        MetaProperty: ['meta', 'property'],
        MethodDefinition: ['key', 'value'],
        ModuleSpecifier: [],
        NewExpression: ['callee', 'arguments'],
        ObjectExpression: ['properties'],
        ObjectPattern: ['properties'],
        Program: ['body'],
        Property: ['key', 'value'],
        RestElement: [ 'argument' ],
        ReturnStatement: ['argument'],
        SequenceExpression: ['expressions'],
        SpreadElement: ['argument'],
        Super: [],
        SwitchStatement: ['discriminant', 'cases'],
        SwitchCase: ['test', 'consequent'],
        TaggedTemplateExpression: ['tag', 'quasi'],
        TemplateElement: [],
        TemplateLiteral: ['quasis', 'expressions'],
        ThisExpression: [],
        ThrowStatement: ['argument'],
        TryStatement: ['block', 'handler', 'finalizer'],
        UnaryExpression: ['argument'],
        UpdateExpression: ['argument'],
        VariableDeclaration: ['declarations'],
        VariableDeclarator: ['id', 'init'],
        WhileStatement: ['test', 'body'],
        WithStatement: ['object', 'body'],
        YieldExpression: ['argument']
    ***REMOVED***;

    // unique id
    BREAK = ***REMOVED******REMOVED***;
    SKIP = ***REMOVED******REMOVED***;
    REMOVE = ***REMOVED******REMOVED***;

    VisitorOption = ***REMOVED***
        Break: BREAK,
        Skip: SKIP,
        Remove: REMOVE
    ***REMOVED***;

    function Reference(parent, key) ***REMOVED***
        this.parent = parent;
        this.key = key;
    ***REMOVED***

    Reference.prototype.replace = function replace(node) ***REMOVED***
        this.parent[this.key] = node;
    ***REMOVED***;

    Reference.prototype.remove = function remove() ***REMOVED***
        if (isArray(this.parent)) ***REMOVED***
            this.parent.splice(this.key, 1);
            return true;
        ***REMOVED*** else ***REMOVED***
            this.replace(null);
            return false;
        ***REMOVED***
    ***REMOVED***;

    function Element(node, path, wrap, ref) ***REMOVED***
        this.node = node;
        this.path = path;
        this.wrap = wrap;
        this.ref = ref;
    ***REMOVED***

    function Controller() ***REMOVED*** ***REMOVED***

    // API:
    // return property path array from root to current node
    Controller.prototype.path = function path() ***REMOVED***
        var i, iz, j, jz, result, element;

        function addToPath(result, path) ***REMOVED***
            if (isArray(path)) ***REMOVED***
                for (j = 0, jz = path.length; j < jz; ++j) ***REMOVED***
                    result.push(path[j]);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                result.push(path);
            ***REMOVED***
        ***REMOVED***

        // root node
        if (!this.__current.path) ***REMOVED***
            return null;
        ***REMOVED***

        // first node is sentinel, second node is root element
        result = [];
        for (i = 2, iz = this.__leavelist.length; i < iz; ++i) ***REMOVED***
            element = this.__leavelist[i];
            addToPath(result, element.path);
        ***REMOVED***
        addToPath(result, this.__current.path);
        return result;
    ***REMOVED***;

    // API:
    // return type of current node
    Controller.prototype.type = function () ***REMOVED***
        var node = this.current();
        return node.type || this.__current.wrap;
    ***REMOVED***;

    // API:
    // return array of parent elements
    Controller.prototype.parents = function parents() ***REMOVED***
        var i, iz, result;

        // first node is sentinel
        result = [];
        for (i = 1, iz = this.__leavelist.length; i < iz; ++i) ***REMOVED***
            result.push(this.__leavelist[i].node);
        ***REMOVED***

        return result;
    ***REMOVED***;

    // API:
    // return current node
    Controller.prototype.current = function current() ***REMOVED***
        return this.__current.node;
    ***REMOVED***;

    Controller.prototype.__execute = function __execute(callback, element) ***REMOVED***
        var previous, result;

        result = undefined;

        previous  = this.__current;
        this.__current = element;
        this.__state = null;
        if (callback) ***REMOVED***
            result = callback.call(this, element.node, this.__leavelist[this.__leavelist.length - 1].node);
        ***REMOVED***
        this.__current = previous;

        return result;
    ***REMOVED***;

    // API:
    // notify control skip / break
    Controller.prototype.notify = function notify(flag) ***REMOVED***
        this.__state = flag;
    ***REMOVED***;

    // API:
    // skip child nodes of current node
    Controller.prototype.skip = function () ***REMOVED***
        this.notify(SKIP);
    ***REMOVED***;

    // API:
    // break traversals
    Controller.prototype['break'] = function () ***REMOVED***
        this.notify(BREAK);
    ***REMOVED***;

    // API:
    // remove node
    Controller.prototype.remove = function () ***REMOVED***
        this.notify(REMOVE);
    ***REMOVED***;

    Controller.prototype.__initialize = function(root, visitor) ***REMOVED***
        this.visitor = visitor;
        this.root = root;
        this.__worklist = [];
        this.__leavelist = [];
        this.__current = null;
        this.__state = null;
        this.__fallback = null;
        if (visitor.fallback === 'iteration') ***REMOVED***
            this.__fallback = objectKeys;
        ***REMOVED*** else if (typeof visitor.fallback === 'function') ***REMOVED***
            this.__fallback = visitor.fallback;
        ***REMOVED***

        this.__keys = VisitorKeys;
        if (visitor.keys) ***REMOVED***
            this.__keys = extend(objectCreate(this.__keys), visitor.keys);
        ***REMOVED***
    ***REMOVED***;

    function isNode(node) ***REMOVED***
        if (node == null) ***REMOVED***
            return false;
        ***REMOVED***
        return typeof node === 'object' && typeof node.type === 'string';
    ***REMOVED***

    function isProperty(nodeType, key) ***REMOVED***
        return (nodeType === Syntax.ObjectExpression || nodeType === Syntax.ObjectPattern) && 'properties' === key;
    ***REMOVED***

    Controller.prototype.traverse = function traverse(root, visitor) ***REMOVED***
        var worklist,
            leavelist,
            element,
            node,
            nodeType,
            ret,
            key,
            current,
            current2,
            candidates,
            candidate,
            sentinel;

        this.__initialize(root, visitor);

        sentinel = ***REMOVED******REMOVED***;

        // reference
        worklist = this.__worklist;
        leavelist = this.__leavelist;

        // initialize
        worklist.push(new Element(root, null, null, null));
        leavelist.push(new Element(null, null, null, null));

        while (worklist.length) ***REMOVED***
            element = worklist.pop();

            if (element === sentinel) ***REMOVED***
                element = leavelist.pop();

                ret = this.__execute(visitor.leave, element);

                if (this.__state === BREAK || ret === BREAK) ***REMOVED***
                    return;
                ***REMOVED***
                continue;
            ***REMOVED***

            if (element.node) ***REMOVED***

                ret = this.__execute(visitor.enter, element);

                if (this.__state === BREAK || ret === BREAK) ***REMOVED***
                    return;
                ***REMOVED***

                worklist.push(sentinel);
                leavelist.push(element);

                if (this.__state === SKIP || ret === SKIP) ***REMOVED***
                    continue;
                ***REMOVED***

                node = element.node;
                nodeType = node.type || element.wrap;
                candidates = this.__keys[nodeType];
                if (!candidates) ***REMOVED***
                    if (this.__fallback) ***REMOVED***
                        candidates = this.__fallback(node);
                    ***REMOVED*** else ***REMOVED***
                        throw new Error('Unknown node type ' + nodeType + '.');
                    ***REMOVED***
                ***REMOVED***

                current = candidates.length;
                while ((current -= 1) >= 0) ***REMOVED***
                    key = candidates[current];
                    candidate = node[key];
                    if (!candidate) ***REMOVED***
                        continue;
                    ***REMOVED***

                    if (isArray(candidate)) ***REMOVED***
                        current2 = candidate.length;
                        while ((current2 -= 1) >= 0) ***REMOVED***
                            if (!candidate[current2]) ***REMOVED***
                                continue;
                            ***REMOVED***
                            if (isProperty(nodeType, candidates[current])) ***REMOVED***
                                element = new Element(candidate[current2], [key, current2], 'Property', null);
                            ***REMOVED*** else if (isNode(candidate[current2])) ***REMOVED***
                                element = new Element(candidate[current2], [key, current2], null, null);
                            ***REMOVED*** else ***REMOVED***
                                continue;
                            ***REMOVED***
                            worklist.push(element);
                        ***REMOVED***
                    ***REMOVED*** else if (isNode(candidate)) ***REMOVED***
                        worklist.push(new Element(candidate, key, null, null));
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;

    Controller.prototype.replace = function replace(root, visitor) ***REMOVED***
        var worklist,
            leavelist,
            node,
            nodeType,
            target,
            element,
            current,
            current2,
            candidates,
            candidate,
            sentinel,
            outer,
            key;

        function removeElem(element) ***REMOVED***
            var i,
                key,
                nextElem,
                parent;

            if (element.ref.remove()) ***REMOVED***
                // When the reference is an element of an array.
                key = element.ref.key;
                parent = element.ref.parent;

                // If removed from array, then decrease following items' keys.
                i = worklist.length;
                while (i--) ***REMOVED***
                    nextElem = worklist[i];
                    if (nextElem.ref && nextElem.ref.parent === parent) ***REMOVED***
                        if  (nextElem.ref.key < key) ***REMOVED***
                            break;
                        ***REMOVED***
                        --nextElem.ref.key;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        this.__initialize(root, visitor);

        sentinel = ***REMOVED******REMOVED***;

        // reference
        worklist = this.__worklist;
        leavelist = this.__leavelist;

        // initialize
        outer = ***REMOVED***
            root: root
        ***REMOVED***;
        element = new Element(root, null, null, new Reference(outer, 'root'));
        worklist.push(element);
        leavelist.push(element);

        while (worklist.length) ***REMOVED***
            element = worklist.pop();

            if (element === sentinel) ***REMOVED***
                element = leavelist.pop();

                target = this.__execute(visitor.leave, element);

                // node may be replaced with null,
                // so distinguish between undefined and null in this place
                if (target !== undefined && target !== BREAK && target !== SKIP && target !== REMOVE) ***REMOVED***
                    // replace
                    element.ref.replace(target);
                ***REMOVED***

                if (this.__state === REMOVE || target === REMOVE) ***REMOVED***
                    removeElem(element);
                ***REMOVED***

                if (this.__state === BREAK || target === BREAK) ***REMOVED***
                    return outer.root;
                ***REMOVED***
                continue;
            ***REMOVED***

            target = this.__execute(visitor.enter, element);

            // node may be replaced with null,
            // so distinguish between undefined and null in this place
            if (target !== undefined && target !== BREAK && target !== SKIP && target !== REMOVE) ***REMOVED***
                // replace
                element.ref.replace(target);
                element.node = target;
            ***REMOVED***

            if (this.__state === REMOVE || target === REMOVE) ***REMOVED***
                removeElem(element);
                element.node = null;
            ***REMOVED***

            if (this.__state === BREAK || target === BREAK) ***REMOVED***
                return outer.root;
            ***REMOVED***

            // node may be null
            node = element.node;
            if (!node) ***REMOVED***
                continue;
            ***REMOVED***

            worklist.push(sentinel);
            leavelist.push(element);

            if (this.__state === SKIP || target === SKIP) ***REMOVED***
                continue;
            ***REMOVED***

            nodeType = node.type || element.wrap;
            candidates = this.__keys[nodeType];
            if (!candidates) ***REMOVED***
                if (this.__fallback) ***REMOVED***
                    candidates = this.__fallback(node);
                ***REMOVED*** else ***REMOVED***
                    throw new Error('Unknown node type ' + nodeType + '.');
                ***REMOVED***
            ***REMOVED***

            current = candidates.length;
            while ((current -= 1) >= 0) ***REMOVED***
                key = candidates[current];
                candidate = node[key];
                if (!candidate) ***REMOVED***
                    continue;
                ***REMOVED***

                if (isArray(candidate)) ***REMOVED***
                    current2 = candidate.length;
                    while ((current2 -= 1) >= 0) ***REMOVED***
                        if (!candidate[current2]) ***REMOVED***
                            continue;
                        ***REMOVED***
                        if (isProperty(nodeType, candidates[current])) ***REMOVED***
                            element = new Element(candidate[current2], [key, current2], 'Property', new Reference(candidate, current2));
                        ***REMOVED*** else if (isNode(candidate[current2])) ***REMOVED***
                            element = new Element(candidate[current2], [key, current2], null, new Reference(candidate, current2));
                        ***REMOVED*** else ***REMOVED***
                            continue;
                        ***REMOVED***
                        worklist.push(element);
                    ***REMOVED***
                ***REMOVED*** else if (isNode(candidate)) ***REMOVED***
                    worklist.push(new Element(candidate, key, null, new Reference(node, key)));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return outer.root;
    ***REMOVED***;

    function traverse(root, visitor) ***REMOVED***
        var controller = new Controller();
        return controller.traverse(root, visitor);
    ***REMOVED***

    function replace(root, visitor) ***REMOVED***
        var controller = new Controller();
        return controller.replace(root, visitor);
    ***REMOVED***

    function extendCommentRange(comment, tokens) ***REMOVED***
        var target;

        target = upperBound(tokens, function search(token) ***REMOVED***
            return token.range[0] > comment.range[0];
        ***REMOVED***);

        comment.extendedRange = [comment.range[0], comment.range[1]];

        if (target !== tokens.length) ***REMOVED***
            comment.extendedRange[1] = tokens[target].range[0];
        ***REMOVED***

        target -= 1;
        if (target >= 0) ***REMOVED***
            comment.extendedRange[0] = tokens[target].range[1];
        ***REMOVED***

        return comment;
    ***REMOVED***

    function attachComments(tree, providedComments, tokens) ***REMOVED***
        // At first, we should calculate extended comment ranges.
        var comments = [], comment, len, i, cursor;

        if (!tree.range) ***REMOVED***
            throw new Error('attachComments needs range information');
        ***REMOVED***

        // tokens array is empty, we attach comments to tree as 'leadingComments'
        if (!tokens.length) ***REMOVED***
            if (providedComments.length) ***REMOVED***
                for (i = 0, len = providedComments.length; i < len; i += 1) ***REMOVED***
                    comment = deepCopy(providedComments[i]);
                    comment.extendedRange = [0, tree.range[0]];
                    comments.push(comment);
                ***REMOVED***
                tree.leadingComments = comments;
            ***REMOVED***
            return tree;
        ***REMOVED***

        for (i = 0, len = providedComments.length; i < len; i += 1) ***REMOVED***
            comments.push(extendCommentRange(deepCopy(providedComments[i]), tokens));
        ***REMOVED***

        // This is based on John Freeman's implementation.
        cursor = 0;
        traverse(tree, ***REMOVED***
            enter: function (node) ***REMOVED***
                var comment;

                while (cursor < comments.length) ***REMOVED***
                    comment = comments[cursor];
                    if (comment.extendedRange[1] > node.range[0]) ***REMOVED***
                        break;
                    ***REMOVED***

                    if (comment.extendedRange[1] === node.range[0]) ***REMOVED***
                        if (!node.leadingComments) ***REMOVED***
                            node.leadingComments = [];
                        ***REMOVED***
                        node.leadingComments.push(comment);
                        comments.splice(cursor, 1);
                    ***REMOVED*** else ***REMOVED***
                        cursor += 1;
                    ***REMOVED***
                ***REMOVED***

                // already out of owned node
                if (cursor === comments.length) ***REMOVED***
                    return VisitorOption.Break;
                ***REMOVED***

                if (comments[cursor].extendedRange[0] > node.range[1]) ***REMOVED***
                    return VisitorOption.Skip;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***);

        cursor = 0;
        traverse(tree, ***REMOVED***
            leave: function (node) ***REMOVED***
                var comment;

                while (cursor < comments.length) ***REMOVED***
                    comment = comments[cursor];
                    if (node.range[1] < comment.extendedRange[0]) ***REMOVED***
                        break;
                    ***REMOVED***

                    if (node.range[1] === comment.extendedRange[0]) ***REMOVED***
                        if (!node.trailingComments) ***REMOVED***
                            node.trailingComments = [];
                        ***REMOVED***
                        node.trailingComments.push(comment);
                        comments.splice(cursor, 1);
                    ***REMOVED*** else ***REMOVED***
                        cursor += 1;
                    ***REMOVED***
                ***REMOVED***

                // already out of owned node
                if (cursor === comments.length) ***REMOVED***
                    return VisitorOption.Break;
                ***REMOVED***

                if (comments[cursor].extendedRange[0] > node.range[1]) ***REMOVED***
                    return VisitorOption.Skip;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***);

        return tree;
    ***REMOVED***

    exports.version = require('./package.json').version;
    exports.Syntax = Syntax;
    exports.traverse = traverse;
    exports.replace = replace;
    exports.attachComments = attachComments;
    exports.VisitorKeys = VisitorKeys;
    exports.VisitorOption = VisitorOption;
    exports.Controller = Controller;
    exports.cloneEnvironment = function () ***REMOVED*** return clone(***REMOVED******REMOVED***); ***REMOVED***;

    return exports;
***REMOVED***(exports));
/* vim: set sw=4 ts=4 et tw=80 : */
