// grpc/deps/grpc/src/node/src/protobuf_js_6_common.js
/**
 * @license
 * Copyright 2017 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

/**
 * @module
 * @private
 */

'use strict';

var _ = require('lodash');
var client = require('./client');

/**
 * Get a function that deserializes a specific type of protobuf.
 * @param ***REMOVED***function()***REMOVED*** cls The constructor of the message type to deserialize
 * @param ***REMOVED***bool=***REMOVED*** binaryAsBase64 Deserialize bytes fields as base64 strings
 *     instead of Buffers. Defaults to false
 * @param ***REMOVED***bool=***REMOVED*** longsAsStrings Deserialize long values as strings instead of
 *     objects. Defaults to true
 * @return ***REMOVED***function(Buffer):cls***REMOVED*** The deserialization function
 */
exports.deserializeCls = function deserializeCls(cls, options) ***REMOVED***
  var conversion_options = ***REMOVED***
    defaults: true,
    bytes: options.binaryAsBase64 ? String : Buffer,
    longs: options.longsAsStrings ? String : null,
    enums: options.enumsAsStrings ? String : null,
    oneofs: true
  ***REMOVED***;
  /**
   * Deserialize a buffer to a message object
   * @param ***REMOVED***Buffer***REMOVED*** arg_buf The buffer to deserialize
   * @return ***REMOVED***cls***REMOVED*** The resulting object
   */
  return function deserialize(arg_buf) ***REMOVED***
    return cls.toObject(cls.decode(arg_buf), conversion_options);
  ***REMOVED***;
***REMOVED***;

var deserializeCls = exports.deserializeCls;

/**
 * Get a function that serializes objects to a buffer by protobuf class.
 * @param ***REMOVED***function()***REMOVED*** Cls The constructor of the message type to serialize
 * @return ***REMOVED***function(Cls):Buffer***REMOVED*** The serialization function
 */
exports.serializeCls = function serializeCls(cls) ***REMOVED***
  /**
   * Serialize an object to a Buffer
   * @param ***REMOVED***Object***REMOVED*** arg The object to serialize
   * @return ***REMOVED***Buffer***REMOVED*** The serialized object
   */
  return function serialize(arg) ***REMOVED***
    var message = cls.fromObject(arg);
    return cls.encode(message).finish();
  ***REMOVED***;
***REMOVED***;

var serializeCls = exports.serializeCls;

/**
 * Get the fully qualified (dotted) name of a ProtoBuf.Reflect value.
 * @param ***REMOVED***ProtoBuf.ReflectionObject***REMOVED*** value The value to get the name of
 * @return ***REMOVED***string***REMOVED*** The fully qualified name of the value
 */
exports.fullyQualifiedName = function fullyQualifiedName(value) ***REMOVED***
  if (value === null || value === undefined) ***REMOVED***
    return '';
  ***REMOVED***
  var name = value.name;
  var parent_fqn = fullyQualifiedName(value.parent);
  if (parent_fqn !== '') ***REMOVED***
    name = parent_fqn + '.' + name;
  ***REMOVED***
  return name;
***REMOVED***;

var fullyQualifiedName = exports.fullyQualifiedName;

/**
 * Return a map from method names to method attributes for the service.
 * @param ***REMOVED***ProtoBuf.Service***REMOVED*** service The service to get attributes for
 * @param ***REMOVED***Object=***REMOVED*** options Options to apply to these attributes
 * @return ***REMOVED***Object***REMOVED*** The attributes map
 */
exports.getProtobufServiceAttrs = function getProtobufServiceAttrs(service,
                                                                   options) ***REMOVED***
  var prefix = '/' + fullyQualifiedName(service) + '/';
  service.resolveAll();
  return _.zipObject(_.map(service.methods, function(method) ***REMOVED***
    return _.camelCase(method.name);
  ***REMOVED***), _.map(service.methods, function(method) ***REMOVED***
    return ***REMOVED***
      originalName: method.name,
      path: prefix + method.name,
      requestStream: !!method.requestStream,
      responseStream: !!method.responseStream,
      requestType: method.resolvedRequestType,
      responseType: method.resolvedResponseType,
      requestSerialize: serializeCls(method.resolvedRequestType),
      requestDeserialize: deserializeCls(method.resolvedRequestType, options),
      responseSerialize: serializeCls(method.resolvedResponseType),
      responseDeserialize: deserializeCls(method.resolvedResponseType, options)
    ***REMOVED***;
  ***REMOVED***));
***REMOVED***;

var getProtobufServiceAttrs = exports.getProtobufServiceAttrs;

exports.loadObject = function loadObject(value, options) ***REMOVED***
  var result = ***REMOVED******REMOVED***;
  if (!value) ***REMOVED***
    return value;
  ***REMOVED***
  if (value.hasOwnProperty('methods')) ***REMOVED***
    // It's a service object
    var service_attrs = getProtobufServiceAttrs(value, options);
    return client.makeClientConstructor(service_attrs);
  ***REMOVED***

  if (value.hasOwnProperty('nested')) ***REMOVED***
    // It's a namespace or root object
    _.each(value.nested, function(nested, name) ***REMOVED***
      result[name] = loadObject(nested, options);
    ***REMOVED***);
    return result;
  ***REMOVED***

  // Otherwise, it's not something we need to change
  return value;
***REMOVED***;

/**
 * The primary purpose of this method is to distinguish between reflection
 * objects from different versions of ProtoBuf.js. This is just a heuristic,
 * checking for properties that are (currently) specific to this version of
 * ProtoBuf.js
 * @param ***REMOVED***Object***REMOVED*** obj The object to check
 * @return ***REMOVED***boolean***REMOVED*** Whether the object appears to be a Protobuf.js 6
 *   ReflectionObject
 */
exports.isProbablyProtobufJs6 = function isProbablyProtobufJs6(obj) ***REMOVED***
  return (typeof obj.root === 'object') && (typeof obj.resolve === 'function');
***REMOVED***;
