// randomfill/browser.js
'use strict'

function oldBrowser () ***REMOVED***
  throw new Error('secure random number generation not supported by this browser\nuse chrome, FireFox or Internet Explorer 11')
***REMOVED***
var safeBuffer = require('safe-buffer')
var randombytes = require('randombytes')
var Buffer = safeBuffer.Buffer
var kBufferMaxLength = safeBuffer.kMaxLength
var crypto = global.crypto || global.msCrypto
var kMaxUint32 = Math.pow(2, 32) - 1
function assertOffset (offset, length) ***REMOVED***
  if (typeof offset !== 'number' || offset !== offset) ***REMOVED*** // eslint-disable-line no-self-compare
    throw new TypeError('offset must be a number')
  ***REMOVED***

  if (offset > kMaxUint32 || offset < 0) ***REMOVED***
    throw new TypeError('offset must be a uint32')
  ***REMOVED***

  if (offset > kBufferMaxLength || offset > length) ***REMOVED***
    throw new RangeError('offset out of range')
  ***REMOVED***
***REMOVED***

function assertSize (size, offset, length) ***REMOVED***
  if (typeof size !== 'number' || size !== size) ***REMOVED*** // eslint-disable-line no-self-compare
    throw new TypeError('size must be a number')
  ***REMOVED***

  if (size > kMaxUint32 || size < 0) ***REMOVED***
    throw new TypeError('size must be a uint32')
  ***REMOVED***

  if (size + offset > length || size > kBufferMaxLength) ***REMOVED***
    throw new RangeError('buffer too small')
  ***REMOVED***
***REMOVED***
if ((crypto && crypto.getRandomValues) || !process.browser) ***REMOVED***
  exports.randomFill = randomFill
  exports.randomFillSync = randomFillSync
***REMOVED*** else ***REMOVED***
  exports.randomFill = oldBrowser
  exports.randomFillSync = oldBrowser
***REMOVED***
function randomFill (buf, offset, size, cb) ***REMOVED***
  if (!Buffer.isBuffer(buf) && !(buf instanceof global.Uint8Array)) ***REMOVED***
    throw new TypeError('"buf" argument must be a Buffer or Uint8Array')
  ***REMOVED***

  if (typeof offset === 'function') ***REMOVED***
    cb = offset
    offset = 0
    size = buf.length
  ***REMOVED*** else if (typeof size === 'function') ***REMOVED***
    cb = size
    size = buf.length - offset
  ***REMOVED*** else if (typeof cb !== 'function') ***REMOVED***
    throw new TypeError('"cb" argument must be a function')
  ***REMOVED***
  assertOffset(offset, buf.length)
  assertSize(size, offset, buf.length)
  return actualFill(buf, offset, size, cb)
***REMOVED***

function actualFill (buf, offset, size, cb) ***REMOVED***
  if (process.browser) ***REMOVED***
    var ourBuf = buf.buffer
    var uint = new Uint8Array(ourBuf, offset, size)
    crypto.getRandomValues(uint)
    if (cb) ***REMOVED***
      process.nextTick(function () ***REMOVED***
        cb(null, buf)
      ***REMOVED***)
      return
    ***REMOVED***
    return buf
  ***REMOVED***
  if (cb) ***REMOVED***
    randombytes(size, function (err, bytes) ***REMOVED***
      if (err) ***REMOVED***
        return cb(err)
      ***REMOVED***
      bytes.copy(buf, offset)
      cb(null, buf)
    ***REMOVED***)
    return
  ***REMOVED***
  var bytes = randombytes(size)
  bytes.copy(buf, offset)
  return buf
***REMOVED***
function randomFillSync (buf, offset, size) ***REMOVED***
  if (typeof offset === 'undefined') ***REMOVED***
    offset = 0
  ***REMOVED***
  if (!Buffer.isBuffer(buf) && !(buf instanceof global.Uint8Array)) ***REMOVED***
    throw new TypeError('"buf" argument must be a Buffer or Uint8Array')
  ***REMOVED***

  assertOffset(offset, buf.length)

  if (size === undefined) size = buf.length - offset

  assertSize(size, offset, buf.length)

  return actualFill(buf, offset, size)
***REMOVED***
