// protobufjs/cli/util.js
"use strict";
var fs            = require("fs"),
    path          = require("path"),
    child_process = require("child_process");

var semver;

try ***REMOVED***
    // installed as a peer dependency
    require.resolve("protobufjs");
    exports.pathToProtobufJs = "protobufjs";
***REMOVED*** catch (e) ***REMOVED***
    // local development, i.e. forked from github
    exports.pathToProtobufJs = "..";
***REMOVED***

var protobuf = require(exports.pathToProtobufJs);

function basenameCompare(a, b) ***REMOVED***
    var aa = String(a).replace(/\.\w+$/, "").split(/(-?\d*\.?\d+)/g),
        bb = String(b).replace(/\.\w+$/, "").split(/(-?\d*\.?\d+)/g);
    for (var i = 0, k = Math.min(aa.length, bb.length); i < k; i++) ***REMOVED***
        var x = parseFloat(aa[i]) || aa[i].toLowerCase(),
            y = parseFloat(bb[i]) || bb[i].toLowerCase();
        if (x < y)
            return -1;
        if (x > y)
            return 1;
    ***REMOVED***
    return a.length < b.length ? -1 : 0;
***REMOVED***

exports.requireAll = function requireAll(dirname) ***REMOVED***
    dirname   = path.join(__dirname, dirname);
    var files = fs.readdirSync(dirname).sort(basenameCompare),
        all = ***REMOVED******REMOVED***;
    files.forEach(function(file) ***REMOVED***
        var basename = path.basename(file, ".js"),
            extname  = path.extname(file);
        if (extname === ".js")
            all[basename] = require(path.join(dirname, file));
    ***REMOVED***);
    return all;
***REMOVED***;

exports.traverse = function traverse(current, fn) ***REMOVED***
    fn(current);
    if (current.fieldsArray)
        current.fieldsArray.forEach(function(field) ***REMOVED***
            traverse(field, fn);
        ***REMOVED***);
    if (current.oneofsArray)
        current.oneofsArray.forEach(function(oneof) ***REMOVED***
            traverse(oneof, fn);
        ***REMOVED***);
    if (current.methodsArray)
        current.methodsArray.forEach(function(method) ***REMOVED***
            traverse(method, fn);
        ***REMOVED***);
    if (current.nestedArray)
        current.nestedArray.forEach(function(nested) ***REMOVED***
            traverse(nested, fn);
        ***REMOVED***);
***REMOVED***;

exports.traverseResolved = function traverseResolved(current, fn) ***REMOVED***
    fn(current);
    if (current.resolvedType)
        traverseResolved(current.resolvedType, fn);
    if (current.resolvedKeyType)
        traverseResolved(current.resolvedKeyType, fn);
    if (current.resolvedRequestType)
        traverseResolved(current.resolvedRequestType, fn);
    if (current.resolvedResponseType)
        traverseResolved(current.resolvedResponseType, fn);
***REMOVED***;

exports.inspect = function inspect(object, indent) ***REMOVED***
    if (!object)
        return "";
    var chalk = require("chalk");
    var sb = [];
    if (!indent)
        indent = "";
    var ind = indent ? indent.substring(0, indent.length - 2) + "└ " : "";
    sb.push(
        ind + chalk.bold(object.toString()) + (object.visible ? " (visible)" : ""),
        indent + chalk.gray("parent: ") + object.parent
    );
    if (object instanceof protobuf.Field) ***REMOVED***
        if (object.extend !== undefined)
            sb.push(indent + chalk.gray("extend: ") + object.extend);
        if (object.partOf)
            sb.push(indent + chalk.gray("oneof : ") + object.oneof);
    ***REMOVED***
    sb.push("");
    if (object.fieldsArray)
        object.fieldsArray.forEach(function(field) ***REMOVED***
            sb.push(inspect(field, indent + "  "));
        ***REMOVED***);
    if (object.oneofsArray)
        object.oneofsArray.forEach(function(oneof) ***REMOVED***
            sb.push(inspect(oneof, indent + "  "));
        ***REMOVED***);
    if (object.methodsArray)
        object.methodsArray.forEach(function(service) ***REMOVED***
            sb.push(inspect(service, indent + "  "));
        ***REMOVED***);
    if (object.nestedArray)
        object.nestedArray.forEach(function(nested) ***REMOVED***
            sb.push(inspect(nested, indent + "  "));
        ***REMOVED***);
    return sb.join("\n");
***REMOVED***;

function modExists(name, version) ***REMOVED***
    for (var i = 0; i < module.paths.length; ++i) ***REMOVED***
        try ***REMOVED***
            var pkg = JSON.parse(fs.readFileSync(path.join(module.paths[i], name, "package.json")));
            return semver
                ? semver.satisfies(pkg.version, version)
                : parseInt(pkg.version, 10) === parseInt(version.replace(/^[\^~]/, ""), 10); // used for semver only
        ***REMOVED*** catch (e) ***REMOVED***/**/***REMOVED***
    ***REMOVED***
    return false;
***REMOVED***

function modInstall(install) ***REMOVED***
    child_process.execSync("npm --silent install " + (typeof install === "string" ? install : install.join(" ")), ***REMOVED***
        cwd: __dirname,
        stdio: "ignore"
    ***REMOVED***);
***REMOVED***

exports.setup = function() ***REMOVED***
    var pkg = require(path.join(__dirname, "..", "package.json"));
    var version = pkg.dependencies["semver"] || pkg.devDependencies["semver"];
    if (!modExists("semver", version)) ***REMOVED***
        process.stderr.write("installing semver@" + version + "\n");
        modInstall("semver@" + version);
    ***REMOVED***
    semver = require("semver"); // used from now on for version comparison
    var install = [];
    pkg.cliDependencies.forEach(function(name) ***REMOVED***
        if (name === "semver")
            return;
        version = pkg.dependencies[name] || pkg.devDependencies[name];
        if (!modExists(name, version)) ***REMOVED***
            process.stderr.write("installing " + name + "@" + version + "\n");
            install.push(name + "@" + version);
        ***REMOVED***
    ***REMOVED***);
    require("../scripts/postinstall"); // emit postinstall warning, if any
    if (!install.length)
        return;
    modInstall(install);
***REMOVED***;

exports.wrap = function(OUTPUT, options) ***REMOVED***
    var name = options.wrap || "default";
    var wrap;
    try ***REMOVED***
        // try built-in wrappers first
        wrap = fs.readFileSync(path.join(__dirname, "wrappers", name + ".js")).toString("utf8");
    ***REMOVED*** catch (e) ***REMOVED***
        // otherwise fetch the custom one
        wrap = fs.readFileSync(path.resolve(process.cwd(), name)).toString("utf8");
    ***REMOVED***
    wrap = wrap.replace(/\$DEPENDENCY/g, JSON.stringify(options.dependency || "protobufjs"));
    wrap = wrap.replace(/( *)\$OUTPUT;/, function($0, $1) ***REMOVED***
        return $1.length ? OUTPUT.replace(/^/mg, $1) : OUTPUT;
    ***REMOVED***);
    if (options.lint !== "")
        wrap = "/*" + options.lint + "*/\n" + wrap;
    return wrap.replace(/\r?\n/g, "\n");
***REMOVED***;

exports.pad = function(str, len, l) ***REMOVED***
    while (str.length < len)
        str = l ? str + " " : " " + str;
    return str;
***REMOVED***;
