// boom/test/index.js
// Load modules

var Code = require('code');
var Boom = require('../lib');
var Lab = require('lab');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.describe;
var it = lab.it;
var expect = Code.expect;


it('returns the same object when already boom', function (done) ***REMOVED***

    var error = Boom.badRequest();
    var wrapped = Boom.wrap(error);
    expect(error).to.equal(wrapped);
    done();
***REMOVED***);

it('returns an error with info when constructed using another error', function (done) ***REMOVED***

    var error = new Error('ka-boom');
    error.xyz = 123;
    var err = Boom.wrap(error);
    expect(err.xyz).to.equal(123);
    expect(err.message).to.equal('ka-boom');
    expect(err.output).to.deep.equal(***REMOVED***
        statusCode: 500,
        payload: ***REMOVED***
            statusCode: 500,
            error: 'Internal Server Error',
            message: 'An internal server error occurred'
        ***REMOVED***,
        headers: ***REMOVED******REMOVED***
    ***REMOVED***);
    expect(err.data).to.equal(null);
    done();
***REMOVED***);

it('does not override data when constructed using another error', function (done) ***REMOVED***

    var error = new Error('ka-boom');
    error.data = ***REMOVED*** useful: 'data' ***REMOVED***;
    var err = Boom.wrap(error);
    expect(err.data).to.equal(error.data);
    done();
***REMOVED***);

it('sets new message when none exists', function (done) ***REMOVED***

    var error = new Error();
    var wrapped = Boom.wrap(error, 400, 'something bad');
    expect(wrapped.message).to.equal('something bad');
    done();
***REMOVED***);

it('throws when statusCode is not a number', function (done) ***REMOVED***

    expect(function () ***REMOVED***

        Boom.create('x');
    ***REMOVED***).to.throw('First argument must be a number (400+): x');
    done();
***REMOVED***);

it('will cast a number-string to an integer', function (done) ***REMOVED***

    var codes = [
        ***REMOVED*** input: '404', result: 404 ***REMOVED***,
        ***REMOVED*** input: '404.1', result: 404 ***REMOVED***,
        ***REMOVED*** input: 400, result: 400 ***REMOVED***,
        ***REMOVED*** input: 400.123, result: 400 ***REMOVED***];
    for (var i = 0, il = codes.length; i < il; ++i) ***REMOVED***
        var code = codes[i];
        var err = Boom.create(code.input);
        expect(err.output.statusCode).to.equal(code.result);
    ***REMOVED***

    done();
***REMOVED***);

it('throws when statusCode is not finite', function (done) ***REMOVED***

    expect(function () ***REMOVED***

        Boom.create(1 / 0);
    ***REMOVED***).to.throw('First argument must be a number (400+): null');
    done();
***REMOVED***);

it('sets error code to unknown', function (done) ***REMOVED***

    var err = Boom.create(999);
    expect(err.output.payload.error).to.equal('Unknown');
    done();
***REMOVED***);

describe('create()', function () ***REMOVED***

    it('does not sets null message', function (done) ***REMOVED***

        var error = Boom.unauthorized(null);
        expect(error.output.payload.message).to.not.exist();
        expect(error.isServer).to.be.false();
        done();
    ***REMOVED***);

    it('sets message and data', function (done) ***REMOVED***

        var error = Boom.badRequest('Missing data', ***REMOVED*** type: 'user' ***REMOVED***);
        expect(error.data.type).to.equal('user');
        expect(error.output.payload.message).to.equal('Missing data');
        done();
    ***REMOVED***);
***REMOVED***);

describe('isBoom()', function () ***REMOVED***

    it('returns true for Boom object', function (done) ***REMOVED***

        expect(Boom.badRequest().isBoom).to.equal(true);
        done();
    ***REMOVED***);

    it('returns false for Error object', function (done) ***REMOVED***

        expect((new Error()).isBoom).to.not.exist();
        done();
    ***REMOVED***);
***REMOVED***);

describe('badRequest()', function () ***REMOVED***

    it('returns a 400 error statusCode', function (done) ***REMOVED***

        var error = Boom.badRequest();

        expect(error.output.statusCode).to.equal(400);
        expect(error.isServer).to.be.false();
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.badRequest('my message').message).to.equal('my message');
        done();
    ***REMOVED***);

    it('sets the message to HTTP status if none provided', function (done) ***REMOVED***

        expect(Boom.badRequest().message).to.equal('Bad Request');
        done();
    ***REMOVED***);
***REMOVED***);

describe('unauthorized()', function () ***REMOVED***

    it('returns a 401 error statusCode', function (done) ***REMOVED***

        var err = Boom.unauthorized();
        expect(err.output.statusCode).to.equal(401);
        expect(err.output.headers).to.deep.equal(***REMOVED******REMOVED***);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.unauthorized('my message').message).to.equal('my message');
        done();
    ***REMOVED***);

    it('returns a WWW-Authenticate header when passed a scheme', function (done) ***REMOVED***

        var err = Boom.unauthorized('boom', 'Test');
        expect(err.output.statusCode).to.equal(401);
        expect(err.output.headers['WWW-Authenticate']).to.equal('Test error="boom"');
        done();
    ***REMOVED***);

    it('returns a WWW-Authenticate header set to the schema array value', function (done) ***REMOVED***

        var err = Boom.unauthorized(null, ['Test','one','two']);
        expect(err.output.statusCode).to.equal(401);
        expect(err.output.headers['WWW-Authenticate']).to.equal('Test, one, two');
        done();
    ***REMOVED***);

    it('returns a WWW-Authenticate header when passed a scheme and attributes', function (done) ***REMOVED***

        var err = Boom.unauthorized('boom', 'Test', ***REMOVED*** a: 1, b: 'something', c: null, d: 0 ***REMOVED***);
        expect(err.output.statusCode).to.equal(401);
        expect(err.output.headers['WWW-Authenticate']).to.equal('Test a="1", b="something", c="", d="0", error="boom"');
        expect(err.output.payload.attributes).to.deep.equal(***REMOVED*** a: 1, b: 'something', c: '', d: 0, error: 'boom' ***REMOVED***);
        done();
    ***REMOVED***);

    it('returns a WWW-Authenticate header when passed attributes, missing error', function (done) ***REMOVED***

        var err = Boom.unauthorized(null, 'Test', ***REMOVED*** a: 1, b: 'something', c: null, d: 0 ***REMOVED***);
        expect(err.output.statusCode).to.equal(401);
        expect(err.output.headers['WWW-Authenticate']).to.equal('Test a="1", b="something", c="", d="0"');
        expect(err.isMissing).to.equal(true);
        done();
    ***REMOVED***);

    it('sets the isMissing flag when error message is empty', function (done) ***REMOVED***

        var err = Boom.unauthorized('', 'Basic');
        expect(err.isMissing).to.equal(true);
        done();
    ***REMOVED***);

    it('does not set the isMissing flag when error message is not empty', function (done) ***REMOVED***

        var err = Boom.unauthorized('message', 'Basic');
        expect(err.isMissing).to.equal(undefined);
        done();
    ***REMOVED***);

    it('sets a WWW-Authenticate when passed as an array', function (done) ***REMOVED***

        var err = Boom.unauthorized('message', ['Basic', 'Example e="1"', 'Another x="3", y="4"']);
        expect(err.output.headers['WWW-Authenticate']).to.equal('Basic, Example e="1", Another x="3", y="4"');
        done();
    ***REMOVED***);
***REMOVED***);


describe('methodNotAllowed()', function () ***REMOVED***

    it('returns a 405 error statusCode', function (done) ***REMOVED***

        expect(Boom.methodNotAllowed().output.statusCode).to.equal(405);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.methodNotAllowed('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('notAcceptable()', function () ***REMOVED***

    it('returns a 406 error statusCode', function (done) ***REMOVED***

        expect(Boom.notAcceptable().output.statusCode).to.equal(406);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.notAcceptable('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('proxyAuthRequired()', function () ***REMOVED***

    it('returns a 407 error statusCode', function (done) ***REMOVED***

        expect(Boom.proxyAuthRequired().output.statusCode).to.equal(407);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.proxyAuthRequired('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('clientTimeout()', function () ***REMOVED***

    it('returns a 408 error statusCode', function (done) ***REMOVED***

        expect(Boom.clientTimeout().output.statusCode).to.equal(408);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.clientTimeout('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('conflict()', function () ***REMOVED***

    it('returns a 409 error statusCode', function (done) ***REMOVED***

        expect(Boom.conflict().output.statusCode).to.equal(409);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.conflict('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('resourceGone()', function () ***REMOVED***

    it('returns a 410 error statusCode', function (done) ***REMOVED***

        expect(Boom.resourceGone().output.statusCode).to.equal(410);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.resourceGone('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('lengthRequired()', function () ***REMOVED***

    it('returns a 411 error statusCode', function (done) ***REMOVED***

        expect(Boom.lengthRequired().output.statusCode).to.equal(411);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.lengthRequired('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('preconditionFailed()', function () ***REMOVED***

    it('returns a 412 error statusCode', function (done) ***REMOVED***

        expect(Boom.preconditionFailed().output.statusCode).to.equal(412);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.preconditionFailed('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('entityTooLarge()', function () ***REMOVED***

    it('returns a 413 error statusCode', function (done) ***REMOVED***

        expect(Boom.entityTooLarge().output.statusCode).to.equal(413);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.entityTooLarge('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('uriTooLong()', function () ***REMOVED***

    it('returns a 414 error statusCode', function (done) ***REMOVED***

        expect(Boom.uriTooLong().output.statusCode).to.equal(414);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.uriTooLong('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('unsupportedMediaType()', function () ***REMOVED***

    it('returns a 415 error statusCode', function (done) ***REMOVED***

        expect(Boom.unsupportedMediaType().output.statusCode).to.equal(415);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.unsupportedMediaType('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('rangeNotSatisfiable()', function () ***REMOVED***

    it('returns a 416 error statusCode', function (done) ***REMOVED***

        expect(Boom.rangeNotSatisfiable().output.statusCode).to.equal(416);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.rangeNotSatisfiable('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('expectationFailed()', function () ***REMOVED***

    it('returns a 417 error statusCode', function (done) ***REMOVED***

        expect(Boom.expectationFailed().output.statusCode).to.equal(417);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.expectationFailed('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('badData()', function () ***REMOVED***

    it('returns a 422 error statusCode', function (done) ***REMOVED***

        expect(Boom.badData().output.statusCode).to.equal(422);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.badData('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('preconditionRequired()', function () ***REMOVED***

    it('returns a 428 error statusCode', function (done) ***REMOVED***

        expect(Boom.preconditionRequired().output.statusCode).to.equal(428);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.preconditionRequired('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('tooManyRequests()', function () ***REMOVED***

    it('returns a 429 error statusCode', function (done) ***REMOVED***

        expect(Boom.tooManyRequests().output.statusCode).to.equal(429);
        done();
    ***REMOVED***);

    it('sets the message with the passed-in message', function (done) ***REMOVED***

        expect(Boom.tooManyRequests('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);

describe('serverTimeout()', function () ***REMOVED***

    it('returns a 503 error statusCode', function (done) ***REMOVED***

        expect(Boom.serverTimeout().output.statusCode).to.equal(503);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.serverTimeout('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);

describe('forbidden()', function () ***REMOVED***

    it('returns a 403 error statusCode', function (done) ***REMOVED***

        expect(Boom.forbidden().output.statusCode).to.equal(403);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.forbidden('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);

describe('notFound()', function () ***REMOVED***

    it('returns a 404 error statusCode', function (done) ***REMOVED***

        expect(Boom.notFound().output.statusCode).to.equal(404);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.notFound('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);

describe('internal()', function () ***REMOVED***

    it('returns a 500 error statusCode', function (done) ***REMOVED***

        expect(Boom.internal().output.statusCode).to.equal(500);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        var err = Boom.internal('my message');
        expect(err.message).to.equal('my message');
        expect(err.isServer).to.true();
        expect(err.output.payload.message).to.equal('An internal server error occurred');
        done();
    ***REMOVED***);

    it('passes data on the callback if its passed in', function (done) ***REMOVED***

        expect(Boom.internal('my message', ***REMOVED*** my: 'data' ***REMOVED***).data.my).to.equal('data');
        done();
    ***REMOVED***);

    it('returns an error with composite message', function (done) ***REMOVED***

        try ***REMOVED***
            JSON.parse('***REMOVED***');
        ***REMOVED***
        catch (err) ***REMOVED***
            var boom = Boom.internal('Someting bad', err);
            expect(boom.message).to.equal('Someting bad: Unexpected end of input');
            expect(boom.isServer).to.be.true();
            done();
        ***REMOVED***
    ***REMOVED***);
***REMOVED***);

describe('notImplemented()', function () ***REMOVED***

    it('returns a 501 error statusCode', function (done) ***REMOVED***

        expect(Boom.notImplemented().output.statusCode).to.equal(501);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.notImplemented('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);


describe('badGateway()', function () ***REMOVED***

    it('returns a 502 error statusCode', function (done) ***REMOVED***

        expect(Boom.badGateway().output.statusCode).to.equal(502);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.badGateway('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);

describe('gatewayTimeout()', function () ***REMOVED***

    it('returns a 504 error statusCode', function (done) ***REMOVED***

        expect(Boom.gatewayTimeout().output.statusCode).to.equal(504);
        done();
    ***REMOVED***);

    it('sets the message with the passed in message', function (done) ***REMOVED***

        expect(Boom.gatewayTimeout('my message').message).to.equal('my message');
        done();
    ***REMOVED***);
***REMOVED***);

describe('badImplementation()', function () ***REMOVED***

    it('returns a 500 error statusCode', function (done) ***REMOVED***

        var err = Boom.badImplementation();
        expect(err.output.statusCode).to.equal(500);
        expect(err.isDeveloperError).to.equal(true);
        expect(err.isServer).to.be.true();
        done();
    ***REMOVED***);
***REMOVED***);

describe('stack trace', function () ***REMOVED***

    it('should omit lib', function (done) ***REMOVED***

        ['badRequest', 'unauthorized', 'forbidden', 'notFound', 'methodNotAllowed',
            'notAcceptable', 'proxyAuthRequired', 'clientTimeout', 'conflict',
            'resourceGone', 'lengthRequired', 'preconditionFailed', 'entityTooLarge',
            'uriTooLong', 'unsupportedMediaType', 'rangeNotSatisfiable', 'expectationFailed',
            'badData', 'preconditionRequired', 'tooManyRequests',

            // 500s
            'internal', 'notImplemented', 'badGateway', 'serverTimeout', 'gatewayTimeout',
            'badImplementation'
        ].forEach(function (name) ***REMOVED***

            var err = Boom[name]();
            expect(err.stack).to.not.match(/\/lib\/index\.js/);
        ***REMOVED***);

        done();
    ***REMOVED***);
***REMOVED***);
