// protobufjs/cli/pbjs/targets/json.js
/*
 Copyright 2013 Daniel Wirtz <dcode@dcode.io>

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
var description = "Plain JSON descriptor";

var ProtoBuf = require(__dirname+"/../../../index.js"),
    util = require("../util.js");

/**
 * pbjs target: Plain JSON descriptor
 * @exports pbjs/targets/json
 * @function
 * @param ***REMOVED***!ProtoBuf.Builder***REMOVED*** builder Builder
 * @param ***REMOVED***!Object.<string,*>=***REMOVED*** options Options
 * @returns ***REMOVED***string***REMOVED***
 */
var json = module.exports = function(builder, options) ***REMOVED***
    options = options || ***REMOVED******REMOVED***;
    builder.resolveAll();

    // Set the pointer to the lowest common namespace (with options)
    var ptr = builder.ns;
    while (ptr.children.length === 1 && Object.keys(ptr.options).length === 0 && ptr.children[0].className === "Namespace")
        ptr = ptr.children[0];

    // Start by building the package namespace
    var pkg = ptr.fqn().substring(1),
        out = ***REMOVED***
            "package": pkg !== "" ? pkg : null
        ***REMOVED***;
    buildNamespace(ptr, out);
    return JSON.stringify(out, null, options.min ? 0 : 4);
***REMOVED***;

/**
 * Module description.
 * @type ***REMOVED***string***REMOVED***
 */
json.description = description;

/**
 * Builds all structures in a namespace.
 * @param ***REMOVED***!ProtoBuf.Reflect.Namespace***REMOVED*** ns Namespace to build
 * @param ***REMOVED***!Object.<string,*>***REMOVED*** out Extended output object
 */
function buildNamespace(ns, out) ***REMOVED***
    var messages, enums, services;
    util.extend(out, ***REMOVED***
        "syntax"   : ns.syntax   || 'proto2',
        "options"  : out.options || ***REMOVED******REMOVED***,
        "messages" : messages = [],
        "enums"    : enums    = [],
        "services" : services = []
    ***REMOVED***);
    if (!(ns instanceof ProtoBuf.Reflect.Message))
        out['isNamespace'] = true;
    util.extend(out["options"], buildOptions(ns.options));
    ns.getChildren(ProtoBuf.Reflect.Enum).forEach(function(enm) ***REMOVED***
        enums.push(buildEnum(enm));
    ***REMOVED***);
    if (enums.length === 0)
        delete out["enums"];
    ns.getChildren(ProtoBuf.Reflect.Message).forEach(function(msg) ***REMOVED***
        messages.push(buildMessage(msg));
    ***REMOVED***);
    ns.getChildren(ProtoBuf.Reflect.Service).forEach(function(svc) ***REMOVED***
        services.push(buildService(svc));
    ***REMOVED***);
    if (services.length === 0)
        delete out["services"];
    Array.prototype.push.apply(messages, buildExtensions(ns));
    ns.getChildren(ProtoBuf.Reflect.Namespace).forEach(function(innerNs) ***REMOVED***
        if (innerNs.className !== "Namespace")
            return;
        var emptyMessage = ***REMOVED***
            "name": innerNs.name,
            "fields": []
        ***REMOVED***;
        buildNamespace(innerNs, emptyMessage);
        messages.push(emptyMessage);
    ***REMOVED***);
    if (messages.length === 0)
        delete out["messages"];
    if (Object.keys(out["options"]).length === 0)
        delete out["options"];
***REMOVED***

/**
 * Builds extensions declared in the specified namespace.
 * @param ***REMOVED***!ProtoBuf.Reflect.Namespace***REMOVED*** ns Namespace
 * @returns ***REMOVED***!Array.<!*>***REMOVED***
 */
function buildExtensions(ns) ***REMOVED***
    var exts = util.groupExtensions(ns);
    if (exts === null)
        return [];
    var messages = [];
    Object.keys(exts).forEach(function(extFqn) ***REMOVED***
        var extMsg = ns.resolve(extFqn),
            extFields = exts[extFqn];
        var fields, ext = ***REMOVED***
            "ref"    : ns.qn(extMsg),
            "fields" : fields = []
        ***REMOVED***;
        extFields.forEach(function(extField) ***REMOVED***
            fields.push(buildMessageField(extField));
        ***REMOVED***);
        messages.push(ext);
    ***REMOVED***);
    return messages;
***REMOVED***

/**
 * Builds block-level options.
 * @param ***REMOVED***!Object.<string,*>***REMOVED*** options Options
 * @returns ***REMOVED***!Object.<string,*>***REMOVED***
 */
function buildOptions(options) ***REMOVED***
    Object.keys(options = options || ***REMOVED******REMOVED***).forEach(function(key) ***REMOVED***
        var val = options[key];
        switch (typeof val) ***REMOVED***
            case 'string':
            case 'number':
            case 'boolean':
            case 'object':
                break;
            default:
                throw Error("Illegal option type: "+typeof(val));
        ***REMOVED***
    ***REMOVED***);
    return options;
***REMOVED***

/**
 * Builds a message.
 * @param ***REMOVED***!ProtoBuf.Reflect.Message***REMOVED*** msg Message
 * @returns ***REMOVED***!****REMOVED***
 */
function buildMessage(msg) ***REMOVED***
    var fields, oneofs;
    var out = ***REMOVED***
        "name"     : msg.name,
        "syntax"   : msg.syntax || 'proto2',
        "options"  : ***REMOVED******REMOVED***,
        "fields"   : fields   = [],
        "oneofs"   : oneofs   = ***REMOVED******REMOVED***
    ***REMOVED***;
    msg.getChildren(ProtoBuf.Reflect.Message.Field).forEach(function(fld) ***REMOVED***
        if (fld instanceof ProtoBuf.Reflect.Message.ExtensionField)
            return;
        fields.push(buildMessageField(fld));
    ***REMOVED***);
    msg.getChildren(ProtoBuf.Reflect.Message.OneOf).forEach(function(oneof) ***REMOVED***
        oneofs[oneof.name] = buildMessageOneof(oneof);
    ***REMOVED***);
    if (msg.extensions)
        out["extensions"] = msg.extensions;
    if (Object.keys(oneofs).length === 0)
        delete out["oneofs"];
    buildNamespace(msg, out);
    return out;
***REMOVED***

/**
 * Builds a message field.
 * @param ***REMOVED***!ProtoBuf.Reflect.Message.Field***REMOVED*** fld Message field
 * @returns ***REMOVED***!****REMOVED***
 */
function buildMessageField(fld) ***REMOVED***
    return ***REMOVED***
        "rule"    : fld.map ? "map" : (fld.repeated ? "repeated" : (fld.required ? "required" : "optional")),
        "type"    : fld.resolvedType ? fld.parent.qn(fld.resolvedType) : fld.type['name'],
        "keytype" : (typeof(fld.keyType) === 'string') ? fld.keyType : (fld.keyType !== null ? fld.keyType.name : undefined),
        "name"    : fld instanceof ProtoBuf.Reflect.Message.ExtensionField ? fld.name.substring(fld.name.lastIndexOf(".")+1): fld.name,
        "id"      : fld.id,
        "options" : Object.keys(fld.options).length > 0 ? buildOptions(fld.options) : undefined,
        "oneof"   : fld.oneof ? fld.oneof.name : undefined
    ***REMOVED***;
***REMOVED***

/**
 * Builds a message oneof.
 * @param ***REMOVED***!ProtoBuf.Reflect.message.OneOf***REMOVED*** oneof Message oneof
 * @returns ***REMOVED***!Array.<!*>***REMOVED***
 */
function buildMessageOneof(oneof) ***REMOVED***
    var out = [];
    oneof.fields.forEach(function(fld) ***REMOVED***
        out.push(fld.id);
    ***REMOVED***);
    return out;
***REMOVED***

/**
 * Builds an enum.
 * @param ***REMOVED***!ProtoBuf.Reflect.Enum***REMOVED*** enm Enum
 * @returns ***REMOVED***!****REMOVED***
 */
function buildEnum(enm) ***REMOVED***
    var values;
    var out = ***REMOVED***
        "name"    : enm.name,
        "syntax"  : enm.syntax || 'proto2',
        "values"  : values = []
    ***REMOVED***;
    enm.getChildren(ProtoBuf.Reflect.Enum.Value).forEach(function(val) ***REMOVED***
        values.push(buildEnumValue(val));
    ***REMOVED***);
    if (Object.keys(enm.options).length > 0)
        out["options"] = buildOptions(enm.options);
    return out;
***REMOVED***

/**
 * Builds an enum value.
 * @param ***REMOVED***!ProtoBuf.Reflect.Enum.Value***REMOVED*** val Enum value
 * @returns ***REMOVED***!****REMOVED***
 */
function buildEnumValue(val) ***REMOVED***
    return ***REMOVED***
        "name"    : val.name,
        "id"      : val.id
    ***REMOVED***;
***REMOVED***

/**
 * Builds a service.
 * @param ***REMOVED***!ProtoBuf.Reflect.Service***REMOVED*** svc Service
 * @returns ***REMOVED***!****REMOVED***
 */
function buildService(svc) ***REMOVED***
    var rpc;
    var out = ***REMOVED***
        "name": svc.name,
        "options": buildOptions(svc.options),
        "rpc": rpc = ***REMOVED******REMOVED***
    ***REMOVED***;
    svc.getChildren(ProtoBuf.Reflect.Service.RPCMethod).forEach(function(mtd) ***REMOVED***
        rpc[mtd.name] = ***REMOVED***
            "request": svc.qn(mtd.resolvedRequestType),
            "request_stream": mtd.requestStream,
            "response": svc.qn(mtd.resolvedResponseType),
            "response_stream": mtd.responseStream,
            "options": buildOptions(mtd.options)
        ***REMOVED***;
    ***REMOVED***);
    return out;
***REMOVED***
