// @google-cloud/storage/src/acl.js
/*!
 * Copyright 2014 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var arrify = require('arrify');
var common = require('@google-cloud/common');
var extend = require('extend');
var is = require('is');
var util = require('util');

/**
 * Cloud Storage uses access control lists (ACLs) to manage object and
 * bucket access. ACLs are the mechanism you use to share objects with other
 * users and allow other users to access your buckets and objects.
 *
 * An ACL consists of one or more entries, where each entry grants permissions
 * to an entity. Permissions define the actions that can be performed against an
 * object or bucket (for example, `READ` or `WRITE`); the entity defines who the
 * permission applies to (for example, a specific user or group of users).
 *
 * Where an `entity` value is accepted, we follow the format the Cloud Storage
 * API expects.
 *
 * Refer to
 * https://cloud.google.com/storage/docs/json_api/v1/defaultObjectAccessControls
 * for the most up-to-date values.
 *
 *   - `user-userId`
 *   - `user-email`
 *   - `group-groupId`
 *   - `group-email`
 *   - `domain-domain`
 *   - `project-team-projectId`
 *   - `allUsers`
 *   - `allAuthenticatedUsers`
 *
 * Examples:
 *
 *   - The user "liz@example.com" would be `user-liz@example.com`.
 *   - The group "example@googlegroups.com" would be
 *     `group-example@googlegroups.com`.
 *   - To refer to all members of the Google Apps for Business domain
 *     "example.com", the entity would be `domain-example.com`.
 *
 * For more detailed information, see
 * [About Access Control Lists](http://goo.gl/6qBBPO).
 *
 * @constructor Acl
 * @mixin
 * @param ***REMOVED***object***REMOVED*** options Configuration options.
 */
function Acl(options) ***REMOVED***
  AclRoleAccessorMethods.call(this);

  this.pathPrefix = options.pathPrefix;
  this.request_ = options.request;
***REMOVED***

/**
 * An object of convenience methods to add or delete owner ACL permissions for a
 * given entity.
 *
 * The supported methods include:
 *
 *   - `myFile.acl.owners.addAllAuthenticatedUsers`
 *   - `myFile.acl.owners.deleteAllAuthenticatedUsers`
 *   - `myFile.acl.owners.addAllUsers`
 *   - `myFile.acl.owners.deleteAllUsers`
 *   - `myFile.acl.owners.addDomain`
 *   - `myFile.acl.owners.deleteDomain`
 *   - `myFile.acl.owners.addGroup`
 *   - `myFile.acl.owners.deleteGroup`
 *   - `myFile.acl.owners.addProject`
 *   - `myFile.acl.owners.deleteProject`
 *   - `myFile.acl.owners.addUser`
 *   - `myFile.acl.owners.deleteUser`
 *
 * @return ***REMOVED***object***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * //-
 * // Add a user as an owner of a file.
 * //-
 * var myBucket = gcs.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 * myFile.acl.owners.addUser('email@example.com', function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // For reference, the above command is the same as running the following.
 * //-
 * myFile.acl.add(***REMOVED***
 *   entity: 'user-email@example.com',
 *   role: gcs.acl.OWNER_ROLE
 * ***REMOVED***, function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myFile.acl.owners.addUser('email@example.com').then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Acl.prototype.owners = ***REMOVED******REMOVED***;

/**
 * An object of convenience methods to add or delete reader ACL permissions for
 * a given entity.
 *
 * The supported methods include:
 *
 *   - `myFile.acl.readers.addAllAuthenticatedUsers`
 *   - `myFile.acl.readers.deleteAllAuthenticatedUsers`
 *   - `myFile.acl.readers.addAllUsers`
 *   - `myFile.acl.readers.deleteAllUsers`
 *   - `myFile.acl.readers.addDomain`
 *   - `myFile.acl.readers.deleteDomain`
 *   - `myFile.acl.readers.addGroup`
 *   - `myFile.acl.readers.deleteGroup`
 *   - `myFile.acl.readers.addProject`
 *   - `myFile.acl.readers.deleteProject`
 *   - `myFile.acl.readers.addUser`
 *   - `myFile.acl.readers.deleteUser`
 *
 * @return ***REMOVED***object***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * //-
 * // Add a user as a reader of a file.
 * //-
 * myFile.acl.readers.addUser('email@example.com', function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // For reference, the above command is the same as running the following.
 * //-
 * myFile.acl.add(***REMOVED***
 *   entity: 'user-email@example.com',
 *   role: gcs.acl.READER_ROLE
 * ***REMOVED***, function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myFile.acl.readers.addUser('email@example.com').then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Acl.prototype.readers = ***REMOVED******REMOVED***;

/**
 * An object of convenience methods to add or delete writer ACL permissions for
 * a given entity.
 *
 * The supported methods include:
 *
 *   - `myFile.acl.writers.addAllAuthenticatedUsers`
 *   - `myFile.acl.writers.deleteAllAuthenticatedUsers`
 *   - `myFile.acl.writers.addAllUsers`
 *   - `myFile.acl.writers.deleteAllUsers`
 *   - `myFile.acl.writers.addDomain`
 *   - `myFile.acl.writers.deleteDomain`
 *   - `myFile.acl.writers.addGroup`
 *   - `myFile.acl.writers.deleteGroup`
 *   - `myFile.acl.writers.addProject`
 *   - `myFile.acl.writers.deleteProject`
 *   - `myFile.acl.writers.addUser`
 *   - `myFile.acl.writers.deleteUser`
 *
 * @return ***REMOVED***object***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * //-
 * // Add a user as a writer of a file.
 * //-
 * myFile.acl.writers.addUser('email@example.com', function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // For reference, the above command is the same as running the following.
 * //-
 * myFile.acl.add(***REMOVED***
 *   entity: 'user-email@example.com',
 *   role: gcs.acl.WRITER_ROLE
 * ***REMOVED***, function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myFile.acl.writers.addUser('email@example.com').then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Acl.prototype.writers = ***REMOVED******REMOVED***;

util.inherits(Acl, AclRoleAccessorMethods);

/**
 * @typedef ***REMOVED***array***REMOVED*** AddAclResponse
 * @property ***REMOVED***object***REMOVED*** 0 The Acl Objects.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback AddAclCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** acl The Acl Objects.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Add access controls on a ***REMOVED***@link Bucket***REMOVED*** or ***REMOVED***@link File***REMOVED***.
 *
 * @see [BucketAccessControls: insert API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/insert***REMOVED***
 * @see [ObjectAccessControls: insert API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objectAccessControls/insert***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** options Configuration options.
 * @param ***REMOVED***string***REMOVED*** options.entity Whose permissions will be added.
 * @param ***REMOVED***string***REMOVED*** options.role Permissions allowed for the defined entity.
 *     See ***REMOVED***@link https://cloud.google.com/storage/docs/access-control Access Control***REMOVED***.
 * @param ***REMOVED***number***REMOVED*** [options.generation] **File Objects Only** Select a specific
 *     revision of this file (as opposed to the latest version, the default).
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***AddAclCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<AddAclResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * var options = ***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   role: gcs.acl.OWNER_ROLE
 * ***REMOVED***;
 *
 * myBucket.acl.add(options, function(err, aclObject, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // For file ACL operations, you can also specify a `generation` property.
 * // Here is how you would grant ownership permissions to a user on a specific
 * // revision of a file.
 * //-
 * myFile.acl.add(***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   role: gcs.acl.OWNER_ROLE,
 *   generation: 1
 * ***REMOVED***, function(err, aclObject, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myBucket.acl.add(options).then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_add_file_owner
 * Example of adding an owner to a file:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_add_bucket_owner
 * Example of adding an owner to a bucket:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_add_bucket_default_owner
 * Example of adding a default owner to a bucket:
 */
Acl.prototype.add = function(options, callback) ***REMOVED***
  var self = this;

  var query = ***REMOVED******REMOVED***;

  if (options.generation) ***REMOVED***
    query.generation = options.generation;
  ***REMOVED***

  if (options.userProject) ***REMOVED***
    query.userProject = options.userProject;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'POST',
      uri: '',
      qs: query,
      json: ***REMOVED***
        entity: options.entity,
        role: options.role.toUpperCase(),
      ***REMOVED***,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      callback(null, self.makeAclObject_(resp), resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** RemoveAclResponse
 * @property ***REMOVED***object***REMOVED*** 0 The full API response.
 */
/**
 * @callback RemoveAclCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Delete access controls on a ***REMOVED***@link Bucket***REMOVED*** or ***REMOVED***@link File***REMOVED***.
 *
 * @see [BucketAccessControls: delete API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/delete***REMOVED***
 * @see [ObjectAccessControls: delete API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objectAccessControls/delete***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** options Configuration object.
 * @param ***REMOVED***string***REMOVED*** options.entity Whose permissions will be revoked.
 * @param ***REMOVED***int***REMOVED*** [options.generation] **File Objects Only** Select a specific
 *     revision of this file (as opposed to the latest version, the default).
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***RemoveAclCallback***REMOVED*** callback The callback function.
 * @returns ***REMOVED***Promise<RemoveAclResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * myBucket.acl.delete(***REMOVED***
 *   entity: 'user-useremail@example.com'
 * ***REMOVED***, function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // For file ACL operations, you can also specify a `generation` property.
 * //-
 * myFile.acl.delete(***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   generation: 1
 * ***REMOVED***, function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myFile.acl.delete().then(function(data) ***REMOVED***
 *   var apiResponse = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_remove_bucket_owner
 * Example of removing an owner from a bucket:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_remove_bucket_default_owner
 * Example of removing a default owner from a bucket:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_remove_file_owner
 * Example of removing an owner from a bucket:
 */
Acl.prototype.delete = function(options, callback) ***REMOVED***
  var query = ***REMOVED******REMOVED***;

  if (options.generation) ***REMOVED***
    query.generation = options.generation;
  ***REMOVED***

  if (options.userProject) ***REMOVED***
    query.userProject = options.userProject;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'DELETE',
      uri: '/' + encodeURIComponent(options.entity),
      qs: query,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      callback(err, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** GetAclResponse
 * @property ***REMOVED***object|object[]***REMOVED*** 0 Single or array of Acl Objects.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback GetAclCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object|object[]***REMOVED*** acl Single or array of Acl Objects.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Get access controls on a ***REMOVED***@link Bucket***REMOVED*** or ***REMOVED***@link File***REMOVED***. If
 * an entity is omitted, you will receive an array of all applicable access
 * controls.
 *
 * @see [BucketAccessControls: get API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/get***REMOVED***
 * @see [ObjectAccessControls: get API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objectAccessControls/get***REMOVED***
 *
 * @param ***REMOVED***object|function***REMOVED*** [options] Configuration options. If you want to
 *     receive a list of all access controls, pass the callback function as the
 *     only argument.
 * @param ***REMOVED***string***REMOVED*** [options.entity] Whose permissions will be fetched.
 * @param ***REMOVED***number***REMOVED*** [options.generation] **File Objects Only** Select a specific
 *     revision of this file (as opposed to the latest version, the default).
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***GetAclCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetAclResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * myBucket.acl.get(***REMOVED***
 *   entity: 'user-useremail@example.com'
 * ***REMOVED***, function(err, aclObject, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // Get all access controls.
 * //-
 * myBucket.acl.get(function(err, aclObjects, apiResponse) ***REMOVED***
 *   // aclObjects = [
 *   //   ***REMOVED***
 *   //     entity: 'user-useremail@example.com',
 *   //     role: 'owner'
 *   //   ***REMOVED***
 *   // ]
 * ***REMOVED***);
 *
 * //-
 * // For file ACL operations, you can also specify a `generation` property.
 * //-
 * myFile.acl.get(***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   generation: 1
 * ***REMOVED***, function(err, aclObject, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myBucket.acl.get().then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_print_file_acl
 * Example of printing a file's ACL:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_print_file_acl_for_user
 * Example of printing a file's ACL for a specific user:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_print_bucket_acl
 * Example of printing a bucket's ACL:
 *
 * @example <caption>include:samples/acl.js</caption>
 * region_tag:storage_print_bucket_acl_for_user
 * Example of printing a bucket's ACL for a specific user:
 */
Acl.prototype.get = function(options, callback) ***REMOVED***
  var self = this;
  var path = '';
  var query = ***REMOVED******REMOVED***;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = null;
  ***REMOVED*** else ***REMOVED***
    path = '/' + encodeURIComponent(options.entity);

    if (options.generation) ***REMOVED***
      query.generation = options.generation;
    ***REMOVED***

    if (options.userProject) ***REMOVED***
      query.userProject = options.userProject;
    ***REMOVED***
  ***REMOVED***

  this.request(
    ***REMOVED***
      uri: path,
      qs: query,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      var results;

      if (resp.items) ***REMOVED***
        results = arrify(resp.items).map(self.makeAclObject_);
      ***REMOVED*** else ***REMOVED***
        results = self.makeAclObject_(resp);
      ***REMOVED***

      callback(null, results, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** UpdateAclResponse
 * @property ***REMOVED***object***REMOVED*** 0 The updated Acl Objects.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback UpdateAclCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** acl The updated Acl Objects.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Update access controls on a ***REMOVED***@link Bucket***REMOVED*** or ***REMOVED***@link File***REMOVED***.
 *
 * @see [BucketAccessControls: update API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/update***REMOVED***
 * @see [ObjectAccessControls: update API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objectAccessControls/update***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** options Configuration options.
 * @param ***REMOVED***string***REMOVED*** options.entity Whose permissions will be updated.
 * @param ***REMOVED***string***REMOVED*** options.role Permissions allowed for the defined entity.
 *     See ***REMOVED***@link Storage.acl***REMOVED***.
 * @param ***REMOVED***number***REMOVED*** [options.generation] **File Objects Only** Select a specific
 *     revision of this file (as opposed to the latest version, the default).
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***UpdateAclCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<UpdateAclResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 * var myFile = myBucket.file('my-file');
 *
 * var options = ***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   role: gcs.acl.WRITER_ROLE
 * ***REMOVED***;
 *
 * myBucket.acl.update(options, function(err, aclObject, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // For file ACL operations, you can also specify a `generation` property.
 * //-
 * myFile.acl.update(***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   role: gcs.acl.WRITER_ROLE,
 *   generation: 1
 * ***REMOVED***, function(err, aclObject, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myFile.acl.update(options).then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Acl.prototype.update = function(options, callback) ***REMOVED***
  var self = this;

  var query = ***REMOVED******REMOVED***;

  if (options.generation) ***REMOVED***
    query.generation = options.generation;
  ***REMOVED***

  if (options.userProject) ***REMOVED***
    query.userProject = options.userProject;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'PUT',
      uri: '/' + encodeURIComponent(options.entity),
      qs: query,
      json: ***REMOVED***
        role: options.role.toUpperCase(),
      ***REMOVED***,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      callback(null, self.makeAclObject_(resp), resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * Transform API responses to a consistent object format.
 *
 * @private
 */
Acl.prototype.makeAclObject_ = function(accessControlObject) ***REMOVED***
  var obj = ***REMOVED***
    entity: accessControlObject.entity,
    role: accessControlObject.role,
  ***REMOVED***;

  if (accessControlObject.projectTeam) ***REMOVED***
    obj.projectTeam = accessControlObject.projectTeam;
  ***REMOVED***

  return obj;
***REMOVED***;

/**
 * Patch requests up to the bucket's request object.
 *
 * @private
 *
 * @param ***REMOVED***string***REMOVED*** method Action.
 * @param ***REMOVED***string***REMOVED*** path Request path.
 * @param ***REMOVED*******REMOVED*** query Request query object.
 * @param ***REMOVED*******REMOVED*** body Request body contents.
 * @param ***REMOVED***function***REMOVED*** callback Callback function.
 */
Acl.prototype.request = function(reqOpts, callback) ***REMOVED***
  reqOpts.uri = this.pathPrefix + reqOpts.uri;
  this.request_(reqOpts, callback);
***REMOVED***;

/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
common.util.promisifyAll(Acl);

module.exports = Acl;

/**
 * Attach functionality to a ***REMOVED***@link Storage.acl***REMOVED*** instance. This will add an
 * object for each role group (owners, readers, and writers), with each object
 * containing methods to add or delete a type of entity.
 *
 * As an example, here are a few methods that are created.
 *
 *   myBucket.acl.readers.deleteGroup('groupId', function(err) ***REMOVED******REMOVED***);
 *
 *   myBucket.acl.owners.addUser('email@example.com', function(err, acl) ***REMOVED******REMOVED***);
 *
 *   myBucket.acl.writers.addDomain('example.com', function(err, acl) ***REMOVED******REMOVED***);
 *
 * @private
 */
function AclRoleAccessorMethods() ***REMOVED***
  AclRoleAccessorMethods.roles.forEach(this._assignAccessMethods.bind(this));
***REMOVED***

AclRoleAccessorMethods.accessMethods = ['add', 'delete'];

AclRoleAccessorMethods.entities = [
  // Special entity groups that do not require further specification.
  'allAuthenticatedUsers',
  'allUsers',

  // Entity groups that require specification, e.g. `user-email@example.com`.
  'domain-',
  'group-',
  'project-',
  'user-',
];

AclRoleAccessorMethods.roles = ['OWNER', 'READER', 'WRITER'];

AclRoleAccessorMethods.prototype._assignAccessMethods = function(role) ***REMOVED***
  var self = this;

  var accessMethods = AclRoleAccessorMethods.accessMethods;
  var entities = AclRoleAccessorMethods.entities;
  var roleGroup = role.toLowerCase() + 's';

  this[roleGroup] = entities.reduce(function(acc, entity) ***REMOVED***
    var isPrefix = entity.charAt(entity.length - 1) === '-';

    accessMethods.forEach(function(accessMethod) ***REMOVED***
      var method = accessMethod + entity[0].toUpperCase() + entity.substr(1);

      if (isPrefix) ***REMOVED***
        method = method.replace('-', '');
      ***REMOVED***

      // Wrap the parent accessor method (e.g. `add` or `delete`) to avoid the
      // more complex API of specifying an `entity` and `role`.
      acc[method] = function(entityId, options, callback) ***REMOVED***
        var apiEntity;

        if (is.fn(options)) ***REMOVED***
          callback = options;
          options = ***REMOVED******REMOVED***;
        ***REMOVED***

        if (isPrefix) ***REMOVED***
          apiEntity = entity + entityId;
        ***REMOVED*** else ***REMOVED***
          // If the entity is not a prefix, it is a special entity group that
          // does not require further details. The accessor methods only accept
          // a callback.
          apiEntity = entity;
          callback = entityId;
        ***REMOVED***

        options = extend(
          ***REMOVED***
            entity: apiEntity,
            role: role,
          ***REMOVED***,
          options
        );

        var args = [options];

        if (is.fn(callback)) ***REMOVED***
          args.push(callback);
        ***REMOVED***

        return self[accessMethod].apply(self, args);
      ***REMOVED***;
    ***REMOVED***);

    return acc;
  ***REMOVED***, ***REMOVED******REMOVED***);
***REMOVED***;

module.exports.AclRoleAccessorMethods = AclRoleAccessorMethods;
