// @google-cloud/storage/src/bucket.js
/**
 * Copyright 2014-2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var arrify = require('arrify');
var async = require('async');
var common = require('@google-cloud/common');
var extend = require('extend');
var fs = require('fs');
var is = require('is');
var mime = require('mime-types');
var path = require('path');
var snakeize = require('snakeize');
var util = require('util');
var request = require('request');

var Acl = require('./acl.js');
var File = require('./file.js');
var Iam = require('./iam.js');
var Notification = require('./notification.js');

/**
 * The size of a file (in bytes) must be greater than this number to
 * automatically trigger a resumable upload.
 *
 * @const ***REMOVED***number***REMOVED***
 * @private
 */
var RESUMABLE_THRESHOLD = 5000000;

/**
 * Create a Bucket object to interact with a Cloud Storage bucket.
 *
 * @class
 * @hideconstructor
 *
 * @param ***REMOVED***Storage***REMOVED*** storage A ***REMOVED***@link Storage***REMOVED*** instance.
 * @param ***REMOVED***string***REMOVED*** name The name of the bucket.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration object.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] User project.
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 */
function Bucket(storage, name, options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;

  var methods = ***REMOVED***
    /**
     * Create a bucket.
     *
     * @method Bucket#create
     * @param ***REMOVED***CreateBucketRequest***REMOVED*** [metadata] Metadata to set for the bucket.
     * @param ***REMOVED***CreateBucketCallback***REMOVED*** [callback] Callback function.
     * @returns ***REMOVED***Promise<CreateBucketResponse>***REMOVED***
     *
     * @example
     * var storage = require('@google-cloud/storage')();
     * var bucket = storage.bucket('albums');
     * bucket.create(function(err, bucket, apiResponse) ***REMOVED***
     *   if (!err) ***REMOVED***
     *     // The bucket was created successfully.
     *   ***REMOVED***
     * ***REMOVED***);
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * bucket.create().then(function(data) ***REMOVED***
     *   var bucket = data[0];
     *   var apiResponse = data[1];
     * ***REMOVED***);
     */
    create: true,
  ***REMOVED***;

  common.ServiceObject.call(this, ***REMOVED***
    parent: storage,
    baseUrl: '/b',
    id: name,
    createMethod: storage.createBucket.bind(storage),
    methods: methods,
  ***REMOVED***);

  /**
   * The bucket's name.
   * @name Bucket#name
   * @type ***REMOVED***string***REMOVED***
   */
  this.name = name;
  /**
   * A reference to the ***REMOVED***@link Storage***REMOVED*** associated with this ***REMOVED***@link Bucket***REMOVED***
   * instance.
   * @name Bucket#storage
   * @type ***REMOVED***string***REMOVED***
   */
  this.storage = storage;

  /**
   * A user project to apply to each request from this bucket.
   * @name Bucket#userProject
   * @type ***REMOVED***string***REMOVED***
   */
  this.userProject = options.userProject;

  /**
   * Cloud Storage uses access control lists (ACLs) to manage object and
   * bucket access. ACLs are the mechanism you use to share objects with other
   * users and allow other users to access your buckets and objects.
   *
   * An ACL consists of one or more entries, where each entry grants permissions
   * to an entity. Permissions define the actions that can be performed against
   * an object or bucket (for example, `READ` or `WRITE`); the entity defines
   * who the permission applies to (for example, a specific user or group of
   * users).
   *
   * The `acl` object on a Bucket instance provides methods to get you a list of
   * the ACLs defined on your bucket, as well as set, update, and delete them.
   *
   * Buckets also have
   * [default ACLs](https://cloud.google.com/storage/docs/access-control/lists#default)
   * for all created files. Default ACLs specify permissions that all new
   * objects added to the bucket will inherit by default. You can add, delete,
   * get, and update entities and permissions for these as well with
   * ***REMOVED***@link Bucket#acl.default***REMOVED***.
   *
   * @see [About Access Control Lists]***REMOVED***@link http://goo.gl/6qBBPO***REMOVED***
   * @see [Default ACLs]***REMOVED***@link https://cloud.google.com/storage/docs/access-control/lists#default***REMOVED***
   *
   * @name Bucket#acl
   * @mixes Acl
   * @property ***REMOVED***Acl***REMOVED*** default Cloud Storage Buckets have
   * [default ACLs](https://cloud.google.com/storage/docs/access-control/lists#default)
   * for all created files. You can add, delete, get, and update entities and
   * permissions for these as well. The method signatures and examples are all
   * the same, after only prefixing the method call with `default`.
   *
   * @example
   * var storage = require('@google-cloud/storage')();
   *
   * //-
   * // Make a bucket's contents publicly readable.
   * //-
   * var myBucket = storage.bucket('my-bucket');
   *
   * var options = ***REMOVED***
   *   entity: 'allUsers',
   *   role: storage.acl.READER_ROLE
   * ***REMOVED***;
   *
   * myBucket.acl.add(options, function(err, aclObject) ***REMOVED******REMOVED***);
   *
   * //-
   * // If the callback is omitted, we'll return a Promise.
   * //-
   * myBucket.acl.add(options).then(function(data) ***REMOVED***
   *   var aclObject = data[0];
   *   var apiResponse = data[1];
   * ***REMOVED***);
   *
   * @example <caption>include:samples/acl.js</caption>
   * region_tag:storage_print_bucket_acl
   * Example of printing a bucket's ACL:
   *
   * @example <caption>include:samples/acl.js</caption>
   * region_tag:storage_print_bucket_acl_for_user
   * Example of printing a bucket's ACL for a specific user:
   *
   * @example <caption>include:samples/acl.js</caption>
   * region_tag:storage_add_bucket_owner
   * Example of adding an owner to a bucket:
   *
   * @example <caption>include:samples/acl.js</caption>
   * region_tag:storage_remove_bucket_owner
   * Example of removing an owner from a bucket:
   *
   * @example <caption>include:samples/acl.js</caption>
   * region_tag:storage_add_bucket_default_owner
   * Example of adding a default owner to a bucket:
   *
   * @example <caption>include:samples/acl.js</caption>
   * region_tag:storage_remove_bucket_default_owner
   * Example of removing a default owner from a bucket:
   */
  this.acl = new Acl(***REMOVED***
    request: this.request.bind(this),
    pathPrefix: '/acl',
  ***REMOVED***);

  this.acl.default = new Acl(***REMOVED***
    request: this.request.bind(this),
    pathPrefix: '/defaultObjectAcl',
  ***REMOVED***);

  /**
   * Get and set IAM policies for your bucket.
   *
   * @name Bucket#iam
   * @mixes Iam
   *
   * @see [Cloud Storage IAM Management](https://cloud.google.com/storage/docs/access-control/iam#short_title_iam_management)
   * @see [Granting, Changing, and Revoking Access](https://cloud.google.com/iam/docs/granting-changing-revoking-access)
   * @see [IAM Roles](https://cloud.google.com/iam/docs/understanding-roles)
   *
   * @example
   * var storage = require('@google-cloud/storage')();
   * var bucket = storage.bucket('albums');
   *
   * //-
   * // Get the IAM policy for your bucket.
   * //-
   * bucket.iam.getPolicy(function(err, policy) ***REMOVED***
   *   console.log(policy);
   * ***REMOVED***);
   *
   * //-
   * // If the callback is omitted, we'll return a Promise.
   * //-
   * bucket.iam.getPolicy().then(function(data) ***REMOVED***
   *   var policy = data[0];
   *   var apiResponse = data[1];
   * ***REMOVED***);
   *
   * @example <caption>include:samples/iam.js</caption>
   * region_tag:storage_view_bucket_iam_members
   * Example of retrieving a bucket's IAM policy:
   *
   * @example <caption>include:samples/iam.js</caption>
   * region_tag:storage_add_bucket_iam_member
   * Example of adding to a bucket's IAM policy:
   *
   * @example <caption>include:samples/iam.js</caption>
   * region_tag:storage_remove_bucket_iam_member
   * Example of removing from a bucket's IAM policy:
   */
  this.iam = new Iam(this);
***REMOVED***

util.inherits(Bucket, common.ServiceObject);

/**
 * @typedef ***REMOVED***array***REMOVED*** CombineResponse
 * @property ***REMOVED***File***REMOVED*** 0 The new ***REMOVED***@link File***REMOVED***.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback CombineCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***File***REMOVED*** newFile The new ***REMOVED***@link File***REMOVED***.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Combine multiple files into one new file.
 *
 * @see [Objects: compose API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/compose***REMOVED***
 *
 * @throws ***REMOVED***Error***REMOVED*** if a non-array is provided as sources argument.
 * @throws ***REMOVED***Error***REMOVED*** if less than two sources are provided.
 * @throws ***REMOVED***Error***REMOVED*** if no destination is provided.
 * @throws ***REMOVED***Error***REMOVED*** if content type can't be determined for the destination file.
 *
 * @param ***REMOVED***string[]|File[]***REMOVED*** sources The source files that will be
 *     combined.
 * @param ***REMOVED***string|File***REMOVED*** destination The file you would like the
 *     source files combined into.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***CombineCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<CombineResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var logBucket = storage.bucket('log-bucket');
 *
 * var sources = [
 *   logBucket.file('2013-logs.txt'),
 *   logBucket.file('2014-logs.txt')
 * ];
 *
 * var allLogs = logBucket.file('all-logs.txt');
 *
 * logBucket.combine(sources, allLogs, function(err, newFile, apiResponse) ***REMOVED***
 *   // newFile === allLogs
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * logBucket.combine(sources, allLogs).then(function(data) ***REMOVED***
 *   var newFile = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Bucket.prototype.combine = function(sources, destination, options, callback) ***REMOVED***
  if (!is.array(sources) || sources.length < 2) ***REMOVED***
    throw new Error('You must provide at least two source files.');
  ***REMOVED***

  if (!destination) ***REMOVED***
    throw new Error('A destination file must be specified.');
  ***REMOVED***

  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  sources = sources.map(convertToFile);
  destination = convertToFile(destination);
  callback = callback || common.util.noop;

  if (!destination.metadata.contentType) ***REMOVED***
    var destinationContentType = mime.contentType(destination.name);

    if (destinationContentType) ***REMOVED***
      destination.metadata.contentType = destinationContentType;
    ***REMOVED*** else ***REMOVED***
      throw new Error(
        'A content type could not be detected for the destination file.'
      );
    ***REMOVED***
  ***REMOVED***

  // Make the request from the destination File object.
  destination.request(
    ***REMOVED***
      method: 'POST',
      uri: '/compose',
      json: ***REMOVED***
        destination: ***REMOVED***
          contentType: destination.metadata.contentType,
        ***REMOVED***,
        sourceObjects: sources.map(function(source) ***REMOVED***
          var sourceObject = ***REMOVED***
            name: source.name,
          ***REMOVED***;

          if (source.metadata && source.metadata.generation) ***REMOVED***
            sourceObject.generation = source.metadata.generation;
          ***REMOVED***

          return sourceObject;
        ***REMOVED***),
      ***REMOVED***,
      qs: options,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      callback(null, destination, resp);
    ***REMOVED***
  );

  function convertToFile(file) ***REMOVED***
    if (file instanceof File) ***REMOVED***
      return file;
    ***REMOVED***

    return self.file(file);
  ***REMOVED***
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** CreateChannelResponse
 * @property ***REMOVED***Channel***REMOVED*** 0 The new ***REMOVED***@link Channel***REMOVED***.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback CreateChannelCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***Channel***REMOVED*** channel The new ***REMOVED***@link Channel***REMOVED***.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Create a channel that will be notified when objects in this bucket changes.
 *
 * @throws ***REMOVED***Error***REMOVED*** If an ID is not provided.
 * @throws ***REMOVED***Error***REMOVED*** If an address is not provided.
 *
 * @see [Objects: watchAll API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/watchAll***REMOVED***
 *
 * @param ***REMOVED***string***REMOVED*** id The ID of the channel to create.
 * @param ***REMOVED***object***REMOVED*** config See a
 *     [Objects: watchAll request body](https://cloud.google.com/storage/docs/json_api/v1/objects/watchAll).
 * @param ***REMOVED***string***REMOVED*** config.address The address where notifications are
 *     delivered for this channel.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***CreateChannelCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<CreateChannelResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 * var id = 'new-channel-id';
 *
 * var config = ***REMOVED***
 *   address: 'https://...'
 * ***REMOVED***;
 *
 * bucket.createChannel(id, config, function(err, channel, apiResponse) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // Channel created successfully.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.createChannel(id, config).then(function(data) ***REMOVED***
 *   var channel = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Bucket.prototype.createChannel = function(id, config, options, callback) ***REMOVED***
  var self = this;

  if (!is.string(id)) ***REMOVED***
    throw new Error('An ID is required to create a channel.');
  ***REMOVED***

  if (!is.string(config.address)) ***REMOVED***
    throw new Error('An address is required to create a channel.');
  ***REMOVED***

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'POST',
      uri: '/o/watch',
      json: extend(
        ***REMOVED***
          id: id,
          type: 'web_hook',
        ***REMOVED***,
        config
      ),
      qs: options,
    ***REMOVED***,
    function(err, apiResponse) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, apiResponse);
        return;
      ***REMOVED***

      var resourceId = apiResponse.resourceId;
      var channel = self.storage.channel(id, resourceId);

      channel.metadata = apiResponse;

      callback(null, channel, apiResponse);
    ***REMOVED***
  );
***REMOVED***;

/**
 * Metadata to set for the Notification.
 *
 * @typedef ***REMOVED***object***REMOVED*** CreateNotificationRequest
 * @property ***REMOVED***object***REMOVED*** [customAttributes] An optional list of additional
 *     attributes to attach to each Cloud PubSub message published for this
 *     notification subscription.
 * @property ***REMOVED***string[]***REMOVED*** [eventTypes] If present, only send notifications about
 *     listed event types. If empty, sent notifications for all event types.
 * @property ***REMOVED***string***REMOVED*** [objectNamePrefix] If present, only apply this
 *     notification configuration to object names that begin with this prefix.
 * @property ***REMOVED***string***REMOVED*** [payloadFormat] The desired content of the Payload.
 *     Defaults to `JSON_API_V1`.
 *
 *     Acceptable values are:
 *     - `JSON_API_V1`
 *
 *     - `NONE`
 * @property ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 */
/**
 * @typedef ***REMOVED***array***REMOVED*** CreateNotificationResponse
 * @property ***REMOVED***Notification***REMOVED*** 0 The new ***REMOVED***@link Notification***REMOVED***.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback CreateNotificationCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***Notification***REMOVED*** notification The new ***REMOVED***@link Notification***REMOVED***.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Creates a notification subscription for the bucket.
 *
 * @see [Notifications: insert]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/notifications/insert***REMOVED***
 *
 * @param ***REMOVED***Topic|string***REMOVED*** topic The Cloud PubSub topic to which this
 *     subscription publishes. If the project ID is omitted, the current project
 *     ID will be used.
 *
 *     Acceptable formats are:
 *     - `projects/grape-spaceship-123/topics/my-topic`
 *
 *     - `my-topic`
 * @param ***REMOVED***CreateNotificationRequest***REMOVED*** [options] Metadata to set for the
 *     notification.
 * @param ***REMOVED***CreateNotificationCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<CreateNotificationResponse>***REMOVED***
 * @throws ***REMOVED***Error***REMOVED*** If a valid topic is not provided.
 * @see Notification#create
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var myBucket = storage.bucket('my-bucket');
 *
 * var callback = function(err, notification, apiResponse) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // The notification was created successfully.
 *   ***REMOVED***
 * ***REMOVED***;
 *
 * myBucket.createNotification('my-topic', callback);
 *
 * //-
 * // Configure the nofiication by providing Notification metadata.
 * //-
 * var metadata = ***REMOVED***
 *   objectNamePrefix: 'prefix-'
 * ***REMOVED***;
 *
 * myBucket.createNotification('my-topic', metadata, callback);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * myBucket.createNotification('my-topic').then(function(data) ***REMOVED***
 *   var notification = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/notifications.js</caption>
 * region_tag:storage_create_notification
 * Another example:
 */
Bucket.prototype.createNotification = function(topic, options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  if (is.object(topic) && common.util.isCustomType(topic, 'pubsub/topic')) ***REMOVED***
    topic = topic.name;
  ***REMOVED***

  if (!is.string(topic)) ***REMOVED***
    throw new Error('A valid topic name is required.');
  ***REMOVED***

  var body = extend(***REMOVED***topic: topic***REMOVED***, options);

  if (body.topic.indexOf('projects') !== 0) ***REMOVED***
    body.topic = 'projects/***REMOVED******REMOVED***projectId***REMOVED******REMOVED***/topics/' + body.topic;
  ***REMOVED***

  body.topic = '//pubsub.googleapis.com/' + body.topic;

  if (!body.payloadFormat) ***REMOVED***
    body.payloadFormat = 'JSON_API_V1';
  ***REMOVED***

  var query = ***REMOVED******REMOVED***;

  if (body.userProject) ***REMOVED***
    query.userProject = body.userProject;
    delete body.userProject;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'POST',
      uri: '/notificationConfigs',
      json: snakeize(body),
      qs: query,
    ***REMOVED***,
    function(err, apiResponse) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, apiResponse);
        return;
      ***REMOVED***

      var notification = self.notification(apiResponse.id);

      notification.metadata = apiResponse;

      callback(null, notification, apiResponse);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** DeleteBucketResponse
 * @property ***REMOVED***object***REMOVED*** 0 The full API response.
 */
/**
 * @callback DeleteBucketCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Delete the bucket.
 *
 * @see [Buckets: delete API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/delete***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***DeleteBucketCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<DeleteBucketResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 * bucket.delete(function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.delete().then(function(data) ***REMOVED***
 *   var apiResponse = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/buckets.js</caption>
 * region_tag:storage_delete_bucket
 * Another example:
 */
Bucket.prototype.delete = function(options, callback) ***REMOVED***
  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'DELETE',
      uri: '',
      qs: options,
    ***REMOVED***,
    callback || common.util.noop
  );
***REMOVED***;

/**
 * @callback DeleteFilesCallback
 * @param ***REMOVED***?Error|?Error[]***REMOVED*** err Request error, if any, or array of errors from
 *     files that were not able to be deleted.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Iterate over the bucket's files, calling `file.delete()` on each.
 *
 * <strong>This is not an atomic request.</strong> A delete attempt will be made
 * for each file individually. Any one can fail, in which case only a portion of
 * the files you intended to be deleted would have.
 *
 * Operations are performed in parallel, up to 10 at once. The first error
 * breaks the loop and will execute the provided callback with it. Specify
 * `***REMOVED*** force: true ***REMOVED***` to suppress the errors until all files have had a chance to
 * be processed.
 *
 * The `query` object passed as the first argument will also be passed to
 * ***REMOVED***@link Bucket#getFiles***REMOVED***.
 *
 * @see [Objects: delete API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/delete***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** [query] Query object. See ***REMOVED***@link Bucket#getFiles***REMOVED***
 *     for all of the supported properties.
 * @param ***REMOVED***boolean***REMOVED*** [query.force] Suppress errors until all files have been
 *     processed.
 * @param ***REMOVED***string***REMOVED*** [query.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***DeleteFilesCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * //-
 * // Delete all of the files in the bucket.
 * //-
 * bucket.deleteFiles(function(err) ***REMOVED******REMOVED***);
 *
 * //-
 * // By default, if a file cannot be deleted, this method will stop deleting
 * // files from your bucket. You can override this setting with `force: true`.
 * //-
 * bucket.deleteFiles(***REMOVED***
 *   force: true
 * ***REMOVED***, function(errors) ***REMOVED***
 *   // `errors`:
 *   //    Array of errors if any occurred, otherwise null.
 * ***REMOVED***);
 *
 * //-
 * // The first argument to this method acts as a query to
 * // ***REMOVED***@link Bucket#getFiles***REMOVED***. As an example, you can delete files
 * // which match a prefix.
 * //-
 * bucket.deleteFiles(***REMOVED***
 *   prefix: 'images/'
 * ***REMOVED***, function(err) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // All files in the `images` directory have been deleted.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.deleteFiles().then(function() ***REMOVED******REMOVED***);
 */
Bucket.prototype.deleteFiles = function(query, callback) ***REMOVED***
  if (is.fn(query)) ***REMOVED***
    callback = query;
    query = ***REMOVED******REMOVED***;
  ***REMOVED***

  query = query || ***REMOVED******REMOVED***;

  var MAX_PARALLEL_LIMIT = 10;
  var errors = [];

  this.getFiles(query, function(err, files) ***REMOVED***
    if (err) ***REMOVED***
      callback(err);
      return;
    ***REMOVED***

    function deleteFile(file, callback) ***REMOVED***
      file.delete(query, function(err) ***REMOVED***
        if (err) ***REMOVED***
          if (query.force) ***REMOVED***
            errors.push(err);
            callback();
            return;
          ***REMOVED***

          callback(err);
          return;
        ***REMOVED***

        callback();
      ***REMOVED***);
    ***REMOVED***

    // Iterate through each file and attempt to delete it.
    async.eachLimit(files, MAX_PARALLEL_LIMIT, deleteFile, function(err) ***REMOVED***
      if (err || errors.length > 0) ***REMOVED***
        callback(err || errors);
        return;
      ***REMOVED***

      callback();
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** DeleteLabelsResponse
 * @property ***REMOVED***object***REMOVED*** 0 The full API response.
 */
/**
 * @callback DeleteLabelsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Delete one or more labels from this bucket.
 *
 * @param ***REMOVED***string|string[]***REMOVED*** labels The labels to delete. If no labels are
 *     provided, all of the labels are removed.
 * @param ***REMOVED***DeleteLabelsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<DeleteLabelsResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * //-
 * // Delete all of the labels from this bucket.
 * //-
 * bucket.deleteLabels(function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // Delete a single label.
 * //-
 * bucket.deleteLabels('labelone', function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // Delete a specific set of labels.
 * //-
 * bucket.deleteLabels([
 *   'labelone',
 *   'labeltwo'
 * ], function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.deleteLabels().then(function(data) ***REMOVED***
 *   var apiResponse = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.deleteLabels = function(labels, callback) ***REMOVED***
  var self = this;

  if (is.fn(labels)) ***REMOVED***
    callback = labels;
    labels = [];
  ***REMOVED***

  labels = arrify(labels);

  if (labels.length === 0) ***REMOVED***
    this.getLabels(function(err, labels) ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      deleteLabels(Object.keys(labels));
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    deleteLabels(labels);
  ***REMOVED***

  function deleteLabels(labels) ***REMOVED***
    var nullLabelMap = labels.reduce(function(nullLabelMap, labelKey) ***REMOVED***
      nullLabelMap[labelKey] = null;
      return nullLabelMap;
    ***REMOVED***, ***REMOVED******REMOVED***);

    self.setLabels(nullLabelMap, callback);
  ***REMOVED***
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** DisableRequesterPaysResponse
 * @property ***REMOVED***object***REMOVED*** 0 The full API response.
 */
/**
 * @callback DisableRequesterPaysCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * <div class="notice">
 *   <strong>Early Access Testers Only</strong>
 *   <p>
 *     This feature is not yet widely-available.
 *   </p>
 * </div>
 *
 * Disable `requesterPays` functionality from this bucket.
 *
 * @param ***REMOVED***DisableRequesterPaysCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<DisableRequesterPaysCallback>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.disableRequesterPays(function(err, apiResponse) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // requesterPays functionality disabled successfully.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.disableRequesterPays().then(function(data) ***REMOVED***
 *   var apiResponse = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/requesterPays.js</caption>
 * region_tag:storage_disable_requester_pays
 * Example of disabling requester pays:
 */
Bucket.prototype.disableRequesterPays = function(callback) ***REMOVED***
  this.setMetadata(
    ***REMOVED***
      billing: ***REMOVED***
        requesterPays: false,
      ***REMOVED***,
    ***REMOVED***,
    callback || common.util.noop
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** EnableRequesterPaysResponse
 * @property ***REMOVED***object***REMOVED*** 0 The full API response.
 */
/**
 * @callback EnableRequesterPaysCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * <div class="notice">
 *   <strong>Early Access Testers Only</strong>
 *   <p>
 *     This feature is not yet widely-available.
 *   </p>
 * </div>
 *
 * Enable `requesterPays` functionality for this bucket. This enables you, the
 * bucket owner, to have the requesting user assume the charges for the access
 * to your bucket and its contents.
 *
 * @param ***REMOVED***EnableRequesterPaysCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<EnableRequesterPaysResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.enableRequesterPays(function(err, apiResponse) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // requesterPays functionality enabled successfully.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.enableRequesterPays().then(function(data) ***REMOVED***
 *   var apiResponse = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/requesterPays.js</caption>
 * region_tag:storage_enable_requester_pays
 * Example of enabling requester pays:
 */
Bucket.prototype.enableRequesterPays = function(callback) ***REMOVED***
  this.setMetadata(
    ***REMOVED***
      billing: ***REMOVED***
        requesterPays: true,
      ***REMOVED***,
    ***REMOVED***,
    callback || common.util.noop
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** BucketExistsResponse
 * @property ***REMOVED***boolean***REMOVED*** 0 Whether the ***REMOVED***@link Bucket***REMOVED*** exists.
 */
/**
 * @callback BucketExistsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***boolean***REMOVED*** exists Whether the ***REMOVED***@link Bucket***REMOVED*** exists.
 */
/**
 * Check if the bucket exists.
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***BucketExistsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<BucketExistsResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.exists(function(err, exists) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.exists().then(function(data) ***REMOVED***
 *   var exists = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.exists = function(options, callback) ***REMOVED***
  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  options = options || ***REMOVED******REMOVED***;

  this.get(options, function(err) ***REMOVED***
    if (err) ***REMOVED***
      if (err.code === 404) ***REMOVED***
        callback(null, false);
      ***REMOVED*** else ***REMOVED***
        callback(err);
      ***REMOVED***

      return;
    ***REMOVED***

    callback(null, true);
  ***REMOVED***);
***REMOVED***;

/**
 * Create a ***REMOVED***@link File***REMOVED*** object. See ***REMOVED***@link File***REMOVED*** to see how to handle
 * the different use cases you may have.
 *
 * @param ***REMOVED***string***REMOVED*** name The name of the file in this bucket.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string|number***REMOVED*** [options.generation] Only use a specific revision of
 *     this file.
 * @param ***REMOVED***string***REMOVED*** [options.key] A custom encryption key. See
 *     [Customer-supplied Encryption Keys](https://cloud.google.com/storage/docs/encryption#customer-supplied).
 * @returns ***REMOVED***File***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 * var file = bucket.file('my-existing-file.png');
 */
Bucket.prototype.file = function(name, options) ***REMOVED***
  if (!name) ***REMOVED***
    throw Error('A file name must be specified.');
  ***REMOVED***

  return new File(this, name, options);
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** GetBucketResponse
 * @property ***REMOVED***Bucket***REMOVED*** 0 The ***REMOVED***@link Bucket***REMOVED***.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback GetBucketCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***Bucket***REMOVED*** bucket The ***REMOVED***@link Bucket***REMOVED***.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Get a bucket if it exists.
 *
 * You may optionally use this to "get or create" an object by providing an
 * object with `autoCreate` set to `true`. Any extra configuration that is
 * normally required for the `create` method must be contained within this
 * object as well.
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***boolean***REMOVED*** [options.autoCreate] Automatically create the object if
 *     it does not exist. Default: `false`
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***GetBucketCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetBucketResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.get(function(err, bucket, apiResponse) ***REMOVED***
 *   // `bucket.metadata` has been populated.
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.get().then(function(data) ***REMOVED***
 *   var bucket = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Bucket.prototype.get = function(options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  options = options || ***REMOVED******REMOVED***;

  var autoCreate = options.autoCreate;
  delete options.autoCreate;

  function onCreate(err, bucket, apiResponse) ***REMOVED***
    if (err) ***REMOVED***
      if (err.code === 409) ***REMOVED***
        self.get(options, callback);
        return;
      ***REMOVED***

      callback(err, null, apiResponse);
      return;
    ***REMOVED***

    callback(null, bucket, apiResponse);
  ***REMOVED***

  this.getMetadata(options, function(err, metadata) ***REMOVED***
    if (err) ***REMOVED***
      if (err.code === 404 && autoCreate) ***REMOVED***
        var args = [];

        if (!is.empty(options)) ***REMOVED***
          args.push(options);
        ***REMOVED***

        args.push(onCreate);

        self.create.apply(self, args);
        return;
      ***REMOVED***

      callback(err, null, metadata);
      return;
    ***REMOVED***

    callback(null, self, metadata);
  ***REMOVED***);
***REMOVED***;

/**
 * Query object for listing files.
 *
 * @typedef ***REMOVED***object***REMOVED*** GetFilesRequest
 * @property ***REMOVED***boolean***REMOVED*** [autoPaginate=true] Have pagination handled
 *     automatically.
 * @property ***REMOVED***string***REMOVED*** [delimiter] Results will contain only objects whose
 *     names, aside from the prefix, do not contain delimiter. Objects whose
 *     names, aside from the prefix, contain delimiter will have their name
 *     truncated after the delimiter, returned in `apiResponse.prefixes`.
 *     Duplicate prefixes are omitted.
 * @property ***REMOVED***string***REMOVED*** [prefix] Filter results to objects whose names begin
 *     with this prefix.
 * @property ***REMOVED***number***REMOVED*** [maxApiCalls] Maximum number of API calls to make.
 * @property ***REMOVED***number***REMOVED*** [maxResults] Maximum number of items plus prefixes to
 *     return.
 * @property ***REMOVED***string***REMOVED*** [pageToken] A previously-returned page token
 *     representing part of the larger set of results to view.
 * @property ***REMOVED***string***REMOVED*** [userProject] The ID of the project which will be
 *     billed for the request.
 * @property ***REMOVED***boolean***REMOVED*** [versions] If true, returns File objects scoped to
 *     their versions.
 */
/**
 * @typedef ***REMOVED***array***REMOVED*** GetFilesResponse
 * @property ***REMOVED***File[]***REMOVED*** 0 Array of ***REMOVED***@link File***REMOVED*** instances.
 */
/**
 * @callback GetFilesCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***File[]***REMOVED*** files Array of ***REMOVED***@link File***REMOVED*** instances.
 */
/**
 * Get ***REMOVED***@link File***REMOVED*** objects for the files currently in the bucket.
 *
 * @see [Objects: list API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/list***REMOVED***
 *
 * @param ***REMOVED***GetFilesRequest***REMOVED*** [query] Query object for listing files.
 * @param ***REMOVED***GetFilesCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetFilesResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.getFiles(function(err, files) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // files is an array of File objects.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If your bucket has versioning enabled, you can get all of your files
 * // scoped to their generation.
 * //-
 * bucket.getFiles(***REMOVED***
 *   versions: true
 * ***REMOVED***, function(err, files) ***REMOVED***
 *   // Each file is scoped to its generation.
 * ***REMOVED***);
 *
 * //-
 * // To control how many API requests are made and page through the results
 * // manually, set `autoPaginate` to `false`.
 * //-
 * var callback = function(err, files, nextQuery, apiResponse) ***REMOVED***
 *   if (nextQuery) ***REMOVED***
 *     // More results exist.
 *     bucket.getFiles(nextQuery, callback);
 *   ***REMOVED***
 *
 *   // The `metadata` property is populated for you with the metadata at the
 *   // time of fetching.
 *   files[0].metadata;
 *
 *   // However, in cases where you are concerned the metadata could have
 *   // changed, use the `getMetadata` method.
 *   files[0].getMetadata(function(err, metadata) ***REMOVED******REMOVED***);
 * ***REMOVED***;
 *
 * bucket.getFiles(***REMOVED***
 *   autoPaginate: false
 * ***REMOVED***, callback);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.getFiles().then(function(data) ***REMOVED***
 *   var files = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/files.js</caption>
 * region_tag:storage_list_files
 * Another example:
 *
 * @example <caption>include:samples/files.js</caption>
 * region_tag:storage_list_files_with_prefix
 * Example of listing files, filtered by a prefix:
 */
Bucket.prototype.getFiles = function(query, callback) ***REMOVED***
  var self = this;

  if (!callback) ***REMOVED***
    callback = query;
    query = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request(
    ***REMOVED***
      uri: '/o',
      qs: query,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, null, resp);
        return;
      ***REMOVED***

      var files = arrify(resp.items).map(function(file) ***REMOVED***
        var options = ***REMOVED******REMOVED***;

        if (query.versions) ***REMOVED***
          options.generation = file.generation;
        ***REMOVED***

        var fileInstance = self.file(file.name, options);
        fileInstance.metadata = file;

        return fileInstance;
      ***REMOVED***);

      var nextQuery = null;
      if (resp.nextPageToken) ***REMOVED***
        nextQuery = extend(***REMOVED******REMOVED***, query, ***REMOVED***
          pageToken: resp.nextPageToken,
        ***REMOVED***);
      ***REMOVED***

      callback(null, files, nextQuery, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * Get ***REMOVED***@link File***REMOVED*** objects for the files currently in the bucket as a
 * readable object stream.
 *
 * @method Bucket#getFilesStream
 * @param ***REMOVED***GetFilesRequest***REMOVED*** [query] Query object for listing files.
 * @returns ***REMOVED***ReadableStream***REMOVED*** A readable stream that emits ***REMOVED***@link File***REMOVED*** instances.
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.getFilesStream()
 *   .on('error', console.error)
 *   .on('data', function(file) ***REMOVED***
 *     // file is a File object.
 *   ***REMOVED***)
 *   .on('end', function() ***REMOVED***
 *     // All files retrieved.
 *   ***REMOVED***);
 *
 * //-
 * // If you anticipate many results, you can end a stream early to prevent
 * // unnecessary processing and API requests.
 * //-
 * bucket.getFilesStream()
 *   .on('data', function(file) ***REMOVED***
 *     this.end();
 *   ***REMOVED***);
 */
Bucket.prototype.getFilesStream = common.paginator.streamify('getFiles');

/**
 * @typedef ***REMOVED***array***REMOVED*** GetLabelsResponse
 * @property ***REMOVED***object***REMOVED*** 0 Object of labels currently set on this bucket.
 */
/**
 * @callback GetLabelsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** labels Object of labels currently set on this bucket.
 */
/**
 * Get the labels currently set on this bucket.
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***GetLabelsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetLabelsCallback>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.getLabels(function(err, labels) ***REMOVED***
 *   if (err) ***REMOVED***
 *     // Error handling omitted.
 *   ***REMOVED***
 *
 *   // labels = ***REMOVED***
 *   //   label: 'labelValue',
 *   //   ...
 *   // ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.getLabels().then(function(data) ***REMOVED***
 *   var labels = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.getLabels = function(options, callback) ***REMOVED***
  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.getMetadata(options, function(err, metadata) ***REMOVED***
    if (err) ***REMOVED***
      callback(err);
      return;
    ***REMOVED***

    callback(null, metadata.labels || ***REMOVED******REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** GetBucketMetadataResponse
 * @property ***REMOVED***object***REMOVED*** 0 The bucket metadata.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback GetBucketMetadataCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** files The bucket metadata.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Get the bucket's metadata.
 *
 * To set metadata, see ***REMOVED***@link Bucket#setMetadata***REMOVED***.
 *
 * @see [Buckets: get API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/get***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***GetBucketMetadataCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetBucketMetadataResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.getMetadata(function(err, metadata, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.getMetadata().then(function(data) ***REMOVED***
 *   var metadata = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/requesterPays.js</caption>
 * region_tag:storage_get_requester_pays_status
 * Example of retrieving the requester pays status of a bucket:
 */
Bucket.prototype.getMetadata = function(options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request(
    ***REMOVED***
      uri: '',
      qs: options,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      self.metadata = resp;

      callback(null, self.metadata, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** GetNotificationsResponse
 * @property ***REMOVED***Notification[]***REMOVED*** 0 Array of ***REMOVED***@link Notification***REMOVED*** instances.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback GetNotificationsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***Notification[]***REMOVED*** notifications Array of ***REMOVED***@link Notification***REMOVED***
 *     instances.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Retrieves a list of notification subscriptions for a given bucket.
 *
 * @see [Notifications: list]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/notifications/list***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***GetNotificationsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetNotificationsResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('my-bucket');
 *
 * bucket.getNotifications(function(err, notifications, apiResponse) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // notifications is an array of Notification objects.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.getNotifications().then(function(data) ***REMOVED***
 *   var notifications = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/notifications.js</caption>
 * region_tag:storage_list_notifications
 * Another example:
 */
Bucket.prototype.getNotifications = function(options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request(
    ***REMOVED***
      uri: '/notificationConfigs',
      qs: options,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      var notifications = arrify(resp.items).map(function(notification) ***REMOVED***
        var notificationInstance = self.notification(notification.id);
        notificationInstance.metadata = notification;
        return notificationInstance;
      ***REMOVED***);

      callback(null, notifications, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** MakeBucketPrivateResponse
 * @property ***REMOVED***File[]***REMOVED*** 0 List of files made private.
 */
/**
 * @callback MakeBucketPrivateCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***File[]***REMOVED*** files List of files made private.
 */
/**
 * Make the bucket listing private.
 *
 * You may also choose to make the contents of the bucket private by specifying
 * `includeFiles: true`. This will automatically run
 * ***REMOVED***@link File#makePrivate***REMOVED*** for every file in the bucket.
 *
 * When specifying `includeFiles: true`, use `force: true` to delay execution of
 * your callback until all files have been processed. By default, the callback
 * is executed after the first error. Use `force` to queue such errors until all
 * files have been processed, after which they will be returned as an array as
 * the first argument to your callback.
 *
 * NOTE: This may cause the process to be long-running and use a high number of
 * requests. Use with caution.
 *
 * @see [Buckets: patch API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/patch***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***boolean***REMOVED*** [options.includeFiles=false] Make each file in the bucket
 *     private.
 * @param ***REMOVED***boolean***REMOVED*** [options.force] Queue errors occurred while making files
 *     private until all files have been processed.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***MakeBucketPrivateCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<MakeBucketPrivateResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * //-
 * // Make the bucket private.
 * //-
 * bucket.makePrivate(function(err) ***REMOVED******REMOVED***);
 *
 * //-
 * // Make the bucket and its contents private.
 * //-
 * var opts = ***REMOVED***
 *   includeFiles: true
 * ***REMOVED***;
 *
 * bucket.makePrivate(opts, function(err, files) ***REMOVED***
 *   // `err`:
 *   //    The first error to occur, otherwise null.
 *   //
 *   // `files`:
 *   //    Array of files successfully made private in the bucket.
 * ***REMOVED***);
 *
 * //-
 * // Make the bucket and its contents private, using force to suppress errors
 * // until all files have been processed.
 * //-
 * var opts = ***REMOVED***
 *   includeFiles: true,
 *   force: true
 * ***REMOVED***;
 *
 * bucket.makePrivate(opts, function(errors, files) ***REMOVED***
 *   // `errors`:
 *   //    Array of errors if any occurred, otherwise null.
 *   //
 *   // `files`:
 *   //    Array of files successfully made private in the bucket.
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.makePrivate(opts).then(function(data) ***REMOVED***
 *   var files = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.makePrivate = function(options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  options = options || ***REMOVED******REMOVED***;
  options.private = true;

  async.series([setPredefinedAcl, makeFilesPrivate], callback);

  function setPredefinedAcl(done) ***REMOVED***
    var query = ***REMOVED***
      predefinedAcl: 'projectPrivate',
    ***REMOVED***;

    if (options.userProject) ***REMOVED***
      query.userProject = options.userProject;
    ***REMOVED***

    self.setMetadata(
      ***REMOVED***
        // You aren't allowed to set both predefinedAcl & acl properties on a
        // bucket so acl must explicitly be nullified.
        acl: null,
      ***REMOVED***,
      query,
      done
    );
  ***REMOVED***

  function makeFilesPrivate(done) ***REMOVED***
    if (!options.includeFiles) ***REMOVED***
      done();
      return;
    ***REMOVED***

    self.makeAllFilesPublicPrivate_(options, done);
  ***REMOVED***
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** MakeBucketPublicResponse
 * @property ***REMOVED***File[]***REMOVED*** 0 List of files made public.
 */
/**
 * @callback MakeBucketPublicCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***File[]***REMOVED*** files List of files made public.
 */
/**
 * Make the bucket publicly readable.
 *
 * You may also choose to make the contents of the bucket publicly readable by
 * specifying `includeFiles: true`. This will automatically run
 * ***REMOVED***@link File#makePublic***REMOVED*** for every file in the bucket.
 *
 * When specifying `includeFiles: true`, use `force: true` to delay execution of
 * your callback until all files have been processed. By default, the callback
 * is executed after the first error. Use `force` to queue such errors until all
 * files have been processed, after which they will be returned as an array as
 * the first argument to your callback.
 *
 * NOTE: This may cause the process to be long-running and use a high number of
 * requests. Use with caution.
 *
 * @see [Buckets: patch API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/patch***REMOVED***
 *
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***boolean***REMOVED*** [options.includeFiles=false] Make each file in the bucket
 *     publicly readable.
 * @param ***REMOVED***boolean***REMOVED*** [options.force] Queue errors occurred while making files
 *     public until all files have been processed.
 * @param ***REMOVED***MakeBucketPublicCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<MakeBucketPublicResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * //-
 * // Make the bucket publicly readable.
 * //-
 * bucket.makePublic(function(err) ***REMOVED******REMOVED***);
 *
 * //-
 * // Make the bucket and its contents publicly readable.
 * //-
 * var opts = ***REMOVED***
 *   includeFiles: true
 * ***REMOVED***;
 *
 * bucket.makePublic(opts, function(err, files) ***REMOVED***
 *   // `err`:
 *   //    The first error to occur, otherwise null.
 *   //
 *   // `files`:
 *   //    Array of files successfully made public in the bucket.
 * ***REMOVED***);
 *
 * //-
 * // Make the bucket and its contents publicly readable, using force to
 * // suppress errors until all files have been processed.
 * //-
 * var opts = ***REMOVED***
 *   includeFiles: true,
 *   force: true
 * ***REMOVED***;
 *
 * bucket.makePublic(opts, function(errors, files) ***REMOVED***
 *   // `errors`:
 *   //    Array of errors if any occurred, otherwise null.
 *   //
 *   // `files`:
 *   //    Array of files successfully made public in the bucket.
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.makePublic(opts).then(function(data) ***REMOVED***
 *   var files = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.makePublic = function(options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  options = options || ***REMOVED******REMOVED***;
  options.public = true;

  async.series(
    [addAclPermissions, addDefaultAclPermissions, makeFilesPublic],
    callback
  );

  function addAclPermissions(done) ***REMOVED***
    // Allow reading bucket contents while preserving original permissions.
    self.acl.add(
      ***REMOVED***
        entity: 'allUsers',
        role: 'READER',
      ***REMOVED***,
      done
    );
  ***REMOVED***

  function addDefaultAclPermissions(done) ***REMOVED***
    self.acl.default.add(
      ***REMOVED***
        entity: 'allUsers',
        role: 'READER',
      ***REMOVED***,
      done
    );
  ***REMOVED***

  function makeFilesPublic(done) ***REMOVED***
    if (!options.includeFiles) ***REMOVED***
      done();
      return;
    ***REMOVED***

    self.makeAllFilesPublicPrivate_(options, done);
  ***REMOVED***
***REMOVED***;

/**
 * Get a reference to a Cloud Pub/Sub Notification.
 *
 * @param ***REMOVED***string***REMOVED*** id ID of notification.
 * @returns ***REMOVED***Notification***REMOVED***
 * @see Notification
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('my-bucket');
 * var notification = bucket.notification('1');
 */
Bucket.prototype.notification = function(id) ***REMOVED***
  if (!id) ***REMOVED***
    throw new Error('You must supply a notification ID.');
  ***REMOVED***

  return new Notification(this, id);
***REMOVED***;

/**
 * Makes request and applies userProject query parameter if necessary.
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** reqOpts - The request options.
 * @param ***REMOVED***function***REMOVED*** callback - The callback function.
 */
Bucket.prototype.request = function(reqOpts, callback) ***REMOVED***
  if (this.userProject && (!reqOpts.qs || !reqOpts.qs.userProject)) ***REMOVED***
    reqOpts.qs = extend(reqOpts.qs, ***REMOVED***userProject: this.userProject***REMOVED***);
  ***REMOVED***

  return common.ServiceObject.prototype.request.call(this, reqOpts, callback);
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** SetLabelsResponse
 * @property ***REMOVED***object***REMOVED*** 0 The bucket metadata.
 */
/**
 * @callback SetLabelsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** metadata The bucket metadata.
 */
/**
 * Set labels on the bucket.
 *
 * This makes an underlying call to ***REMOVED***@link Bucket#setMetadata***REMOVED***, which
 * is a PATCH request. This means an individual label can be overwritten, but
 * unmentioned labels will not be touched.
 *
 * @param ***REMOVED***object<string, string>***REMOVED*** labels Labels to set on the bucket.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***SetLabelsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<SetLabelsResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * var labels = ***REMOVED***
 *   labelone: 'labelonevalue',
 *   labeltwo: 'labeltwovalue'
 * ***REMOVED***;
 *
 * bucket.setLabels(labels, function(err, metadata) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // Labels set successfully.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.setLabels(labels).then(function(data) ***REMOVED***
 *   var metadata = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.setLabels = function(labels, options, callback) ***REMOVED***
  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  callback = callback || common.util.noop;

  this.setMetadata(***REMOVED***labels***REMOVED***, options, callback);
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** SetBucketMetadataResponse
 * @property ***REMOVED***object***REMOVED*** 0 The bucket metadata.
 */
/**
 * @callback SetBucketMetadataCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** metadata The bucket metadata.
 */
/**
 * Set the bucket's metadata.
 *
 * @see [Buckets: patch API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/patch***REMOVED***
 *
 * @param ***REMOVED***object<string, *>***REMOVED*** metadata The metadata you wish to set.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***SetBucketMetadataCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<SetBucketMetadataResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * //-
 * // Set website metadata field on the bucket.
 * //-
 * var metadata = ***REMOVED***
 *   website: ***REMOVED***
 *     mainPageSuffix: 'http://example.com',
 *     notFoundPage: 'http://example.com/404.html'
 *   ***REMOVED***
 * ***REMOVED***;
 *
 * bucket.setMetadata(metadata, function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // Enable versioning for your bucket.
 * //-
 * bucket.setMetadata(***REMOVED***
 *   versioning: ***REMOVED***
 *     enabled: true
 *   ***REMOVED***
 * ***REMOVED***, function(err, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.setMetadata(metadata).then(function(data) ***REMOVED***
 *   var apiResponse = data[0];
 * ***REMOVED***);
 */
Bucket.prototype.setMetadata = function(metadata, options, callback) ***REMOVED***
  var self = this;

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  callback = callback || common.util.noop;

  this.request(
    ***REMOVED***
      method: 'PATCH',
      uri: '',
      json: metadata,
      qs: options,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, resp);
        return;
      ***REMOVED***

      self.metadata = resp;

      callback(null, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * @callback SetStorageClassCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 */
/**
 * Set the default storage class for new files in this bucket.
 *
 * @see [Storage Classes]***REMOVED***@link https://cloud.google.com/storage/docs/storage-classes***REMOVED***
 *
 * @param ***REMOVED***string***REMOVED*** storageClass The new storage class. (`multi_regional`,
 *     `regional`, `standard`, `nearline`, `coldline`, or
 *     `durable_reduced_availability`)
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] - The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***SetStorageClassCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.setStorageClass('regional', function(err, apiResponse) ***REMOVED***
 *   if (err) ***REMOVED***
 *     // Error handling omitted.
 *   ***REMOVED***
 *
 *   // The storage class was updated successfully.
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.setStorageClass('regional').then(function() ***REMOVED******REMOVED***);
 */
Bucket.prototype.setStorageClass = function(storageClass, options, callback) ***REMOVED***
  // In case we get input like `storageClass`, convert to `storage_class`.
  storageClass = storageClass
    .replace(/-/g, '_')
    .replace(/([a-z])([A-Z])/g, function(_, low, up) ***REMOVED***
      return low + '_' + up;
    ***REMOVED***)
    .toUpperCase();

  this.setMetadata(***REMOVED***storageClass***REMOVED***, options, callback);
***REMOVED***;

/**
 * Set a user project to be billed for all requests made from this Bucket
 * object and any files referenced from this Bucket object.
 *
 * @param ***REMOVED***string***REMOVED*** userProject The user project.
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * bucket.setUserProject('grape-spaceship-123');
 */
Bucket.prototype.setUserProject = function(userProject) ***REMOVED***
  this.userProject = userProject;
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** UploadResponse
 * @property ***REMOVED***object***REMOVED*** 0 The uploaded ***REMOVED***@link File***REMOVED***.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback UploadCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** metadata The uploaded ***REMOVED***@link File***REMOVED***.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Upload a file to the bucket. This is a convenience method that wraps
 * ***REMOVED***@link File#createWriteStream***REMOVED***.
 *
 * You can specify whether or not an upload is resumable by setting
 * `options.resumable`. *Resumable uploads are enabled by default if your input
 * file is larger than 5 MB.*
 *
 * For faster crc32c computation, you must manually install
 * [`fast-crc32c`](http://www.gitnpm.com/fast-crc32c):
 *
 *     $ npm install --save fast-crc32c
 *
 * @see [Upload Options (Simple or Resumable)]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/how-tos/upload#uploads***REMOVED***
 * @see [Objects: insert API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/insert***REMOVED***
 *
 * @param ***REMOVED***string***REMOVED*** pathString The fully qualified path or url to the file you
 *     wish to upload to your bucket.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration options.
 * @param ***REMOVED***string|File***REMOVED*** [options.destination] The place to save
 *     your file. If given a string, the file will be uploaded to the bucket
 *     using the string as a filename. When given a File object, your local file
 *     will be uploaded to the File object's bucket and under the File object's
 *     name. Lastly, when this argument is omitted, the file is uploaded to your
 *     bucket using the name of the local file or the path of the url relative to it's domain.
 * @param ***REMOVED***string***REMOVED*** [options.encryptionKey] A custom encryption key. See
 *     [Customer-supplied Encryption Keys](https://cloud.google.com/storage/docs/encryption#customer-supplied).
 * @param ***REMOVED***boolean***REMOVED*** [options.gzip] Automatically gzip the file. This will set
 *     `options.metadata.contentEncoding` to `gzip`.
 * @param ***REMOVED***object***REMOVED*** [options.metadata] See an
 *     [Objects: insert request body](https://cloud.google.com/storage/docs/json_api/v1/objects/insert#request_properties_JSON).
 * @param ***REMOVED***string***REMOVED*** [options.offset] The starting byte of the upload stream, for
 *     resuming an interrupted upload. Defaults to 0.
 * @param ***REMOVED***string***REMOVED*** [options.predefinedAcl] Apply a predefined set of access
 *     controls to this object.
 *
 *     Acceptable values are:
 *     - **`authenticatedRead`** - Object owner gets `OWNER` access, and
 *       `allAuthenticatedUsers` get `READER` access.
 *
 *     - **`bucketOwnerFullControl`** - Object owner gets `OWNER` access, and
 *       project team owners get `OWNER` access.
 *
 *     - **`bucketOwnerRead`** - Object owner gets `OWNER` access, and project
 *       team owners get `READER` access.
 *
 *     - **`private`** - Object owner gets `OWNER` access.
 *
 *     - **`projectPrivate`** - Object owner gets `OWNER` access, and project
 *       team members get access according to their roles.
 *
 *     - **`publicRead`** - Object owner gets `OWNER` access, and `allUsers` get
 *       `READER` access.
 * @param ***REMOVED***boolean***REMOVED*** [options.private] Make the uploaded file private. (Alias for
 *     `options.predefinedAcl = 'private'`)
 * @param ***REMOVED***boolean***REMOVED*** [options.public] Make the uploaded file public. (Alias for
 *     `options.predefinedAcl = 'publicRead'`)
 * @param ***REMOVED***boolean***REMOVED*** [options.resumable] Force a resumable upload. (default:
 *     true for files larger than 5 MB).
 * @param ***REMOVED***string***REMOVED*** [options.uri] The URI for an already-created resumable
 *     upload. See ***REMOVED***@link File#createResumableUpload***REMOVED***.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***string|boolean***REMOVED*** [options.validation] Possible values: `"md5"`,
 *     `"crc32c"`, or `false`. By default, data integrity is validated with an
 *     MD5 checksum for maximum reliability. CRC32c will provide better
 *     performance with less reliability. You may also choose to skip validation
 *     completely, however this is **not recommended**.
 * @param ***REMOVED***UploadCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<UploadResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('albums');
 *
 * //-
 * // Upload a file from a local path.
 * //-
 * bucket.upload('/local/path/image.png', function(err, file, apiResponse) ***REMOVED***
 *   // Your bucket now contains:
 *   // - "image.png" (with the contents of `/local/path/image.png')
 *
 *   // `file` is an instance of a File object that refers to your new file.
 * ***REMOVED***);
 *
 * //-
 * // You can also upload a file from a URL.
 * //-
 *
 * bucket.upload('https://example.com/images/image.png', function(err, file, apiResponse) ***REMOVED***
 *   // Your bucket now contains:
 *   // - "image.png"
 *
 *   // `file` is an instance of a File object that refers to your new file.
 * ***REMOVED***);
 *
 * //-
 * // It's not always that easy. You will likely want to specify the filename
 * // used when your new file lands in your bucket.
 * //
 * // You may also want to set metadata or customize other options.
 * //-
 * var options = ***REMOVED***
 *   destination: 'new-image.png',
 *   resumable: true,
 *   validation: 'crc32c',
 *   metadata: ***REMOVED***
 *     metadata: ***REMOVED***
 *       event: 'Fall trip to the zoo'
 *     ***REMOVED***
 *   ***REMOVED***
 * ***REMOVED***;
 *
 * bucket.upload('local-image.png', options, function(err, file) ***REMOVED***
 *   // Your bucket now contains:
 *   // - "new-image.png" (with the contents of `local-image.png')
 *
 *   // `file` is an instance of a File object that refers to your new file.
 * ***REMOVED***);
 *
 * //-
 * // You can also have a file gzip'd on the fly.
 * //-
 * bucket.upload('index.html', ***REMOVED*** gzip: true ***REMOVED***, function(err, file) ***REMOVED***
 *   // Your bucket now contains:
 *   // - "index.html" (automatically compressed with gzip)
 *
 *   // Downloading the file with `file.download` will automatically decode the
 *   // file.
 * ***REMOVED***);
 *
 * //-
 * // You may also re-use a File object, ***REMOVED***File***REMOVED***, that references
 * // the file you wish to create or overwrite.
 * //-
 * var options = ***REMOVED***
 *   destination: bucket.file('existing-file.png'),
 *   resumable: false
 * ***REMOVED***;
 *
 * bucket.upload('local-img.png', options, function(err, newFile) ***REMOVED***
 *   // Your bucket now contains:
 *   // - "existing-file.png" (with the contents of `local-img.png')
 *
 *   // Note:
 *   // The `newFile` parameter is equal to `file`.
 * ***REMOVED***);
 *
 * //-
 * // To use
 * // <a href="https://cloud.google.com/storage/docs/encryption#customer-supplied">
 * // Customer-supplied Encryption Keys</a>, provide the `encryptionKey` option.
 * //-
 * var crypto = require('crypto');
 * var encryptionKey = crypto.randomBytes(32);
 *
 * bucket.upload('img.png', ***REMOVED***
 *   encryptionKey: encryptionKey
 * ***REMOVED***, function(err, newFile) ***REMOVED***
 *   // `img.png` was uploaded with your custom encryption key.
 *
 *   // `newFile` is already configured to use the encryption key when making
 *   // operations on the remote object.
 *
 *   // However, to use your encryption key later, you must create a `File`
 *   // instance with the `key` supplied:
 *   var file = bucket.file('img.png', ***REMOVED***
 *     encryptionKey: encryptionKey
 *   ***REMOVED***);
 *
 *   // Or with `file#setEncryptionKey`:
 *   var file = bucket.file('img.png');
 *   file.setEncryptionKey(encryptionKey);
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.upload('local-image.png').then(function(data) ***REMOVED***
 *   var file = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/files.js</caption>
 * region_tag:storage_upload_file
 * Another example:
 *
 * @example <caption>include:samples/encryption.js</caption>
 * region_tag:storage_upload_encrypted_file
 * Example of uploading an encrypted file:
 */
Bucket.prototype.upload = function(pathString, options, callback) ***REMOVED***
  if (global.GCLOUD_SANDBOX_ENV) ***REMOVED***
    return;
  ***REMOVED***

  var isURL = /^(http|https):/.test(pathString);

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  options = extend(
    ***REMOVED***
      metadata: ***REMOVED******REMOVED***,
    ***REMOVED***,
    options
  );

  var newFile;
  if (options.destination instanceof File) ***REMOVED***
    newFile = options.destination;
  ***REMOVED*** else if (is.string(options.destination)) ***REMOVED***
    // Use the string as the name of the file.
    newFile = this.file(options.destination, ***REMOVED***
      encryptionKey: options.encryptionKey,
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    // Resort to using the name of the incoming file.
    var destination = path.basename(pathString);
    newFile = this.file(destination, ***REMOVED***
      encryptionKey: options.encryptionKey,
    ***REMOVED***);
  ***REMOVED***

  var contentType = mime.contentType(path.basename(pathString));

  if (contentType && !options.metadata.contentType) ***REMOVED***
    options.metadata.contentType = contentType;
  ***REMOVED***

  if (is.boolean(options.resumable)) ***REMOVED***
    upload();
  ***REMOVED*** else if (isURL) ***REMOVED***
    request.head(pathString, function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      var contentLength = resp.headers['content-length'];

      if (is.number(contentLength)) ***REMOVED***
        options.resumable = contentLength > RESUMABLE_THRESHOLD;
      ***REMOVED***

      upload();
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    // Determine if the upload should be resumable if it's over the threshold.
    fs.stat(pathString, function(err, fd) ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***

      options.resumable = fd.size > RESUMABLE_THRESHOLD;

      upload();
    ***REMOVED***);
  ***REMOVED***

  function upload() ***REMOVED***
    var sourceStream;

    if (isURL) ***REMOVED***
      sourceStream = request.get(pathString);
    ***REMOVED*** else ***REMOVED***
      sourceStream = fs.createReadStream(pathString);
    ***REMOVED***

    sourceStream
      .on('error', callback)
      .pipe(newFile.createWriteStream(options))
      .on('error', callback)
      .on('finish', function() ***REMOVED***
        callback(null, newFile);
      ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Iterate over all of a bucket's files, calling `file.makePublic()` (public)
 * or `file.makePrivate()` (private) on each.
 *
 * Operations are performed in parallel, up to 10 at once. The first error
 * breaks the loop, and will execute the provided callback with it. Specify
 * `***REMOVED*** force: true ***REMOVED***` to suppress the errors.
 *
 * @private
 *
 * @param ***REMOVED***object***REMOVED*** options] Configuration options.
 * @param ***REMOVED***boolean***REMOVED*** [options.force] Suppress errors until all files have been
 *     processed.
 * @param ***REMOVED***boolean***REMOVED*** [options.private] Make files private.
 * @param ***REMOVED***boolean***REMOVED*** [options.public] Make files public.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***function***REMOVED*** callback Callback function.
 */
Bucket.prototype.makeAllFilesPublicPrivate_ = function(options, callback) ***REMOVED***
  var MAX_PARALLEL_LIMIT = 10;
  var errors = [];
  var updatedFiles = [];

  this.getFiles(options, function(err, files) ***REMOVED***
    if (err) ***REMOVED***
      callback(err);
      return;
    ***REMOVED***

    function processFile(file, callback) ***REMOVED***
      if (options.public) ***REMOVED***
        file.makePublic(processedCallback);
      ***REMOVED*** else if (options.private) ***REMOVED***
        file.makePrivate(options, processedCallback);
      ***REMOVED***

      function processedCallback(err) ***REMOVED***
        if (err) ***REMOVED***
          if (options.force) ***REMOVED***
            errors.push(err);
            callback();
            return;
          ***REMOVED***

          callback(err);
          return;
        ***REMOVED***

        updatedFiles.push(file);
        callback();
      ***REMOVED***
    ***REMOVED***

    // Iterate through each file and make it public or private.
    async.eachLimit(files, MAX_PARALLEL_LIMIT, processFile, function(err) ***REMOVED***
      if (err || errors.length > 0) ***REMOVED***
        callback(err || errors, updatedFiles);
        return;
      ***REMOVED***

      callback(null, updatedFiles);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/*! Developer Documentation
 *
 * These methods can be auto-paginated.
 */
common.paginator.extend(Bucket, 'getFiles');

/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
common.util.promisifyAll(Bucket, ***REMOVED***
  exclude: ['file', 'notification'],
***REMOVED***);

/**
 * Reference to the ***REMOVED***@link Bucket***REMOVED*** class.
 * @name module:@google-cloud/storage.Bucket
 * @see Bucket
 */
module.exports = Bucket;
