// @google-cloud/storage/src/iam.js
/*!
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var arrify = require('arrify');
var common = require('@google-cloud/common');
var extend = require('extend');
var is = require('is');

/**
 * Get and set IAM policies for your Cloud Storage bucket.
 *
 * @see [Cloud Storage IAM Management](https://cloud.google.com/storage/docs/access-control/iam#short_title_iam_management)
 * @see [Granting, Changing, and Revoking Access](https://cloud.google.com/iam/docs/granting-changing-revoking-access)
 * @see [IAM Roles](https://cloud.google.com/iam/docs/understanding-roles)
 *
 * @constructor Iam
 * @mixin
 *
 * @param ***REMOVED***Bucket***REMOVED*** bucket The parent instance.
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('my-bucket');
 * // bucket.iam
 */
function Iam(bucket) ***REMOVED***
  this.request_ = bucket.request.bind(bucket);
  this.resourceId_ = 'buckets/' + bucket.id;
***REMOVED***

/**
 * @typedef ***REMOVED***object***REMOVED*** GetPolicyRequest
 * @property ***REMOVED***string***REMOVED*** userProject The ID of the project which will be billed for
 *     the request.
 */
/**
 * @typedef ***REMOVED***array***REMOVED*** GetPolicyResponse
 * @property ***REMOVED***object***REMOVED*** 0 The policy.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback GetPolicyCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** acl The policy.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Get the IAM policy.
 *
 * @param ***REMOVED***GetPolicyRequest***REMOVED*** [options] Request options.
 * @param ***REMOVED***GetPolicyCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetPolicyResponse>***REMOVED***
 *
 * @see [Buckets: setIamPolicy API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/getIamPolicy***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('my-bucket');
 * bucket.iam.getPolicy(function(err, policy, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.iam.getPolicy().then(function(data) ***REMOVED***
 *   var policy = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/iam.js</caption>
 * region_tag:storage_view_bucket_iam_members
 * Example of retrieving a bucket's IAM policy:
 */
Iam.prototype.getPolicy = function(options, callback) ***REMOVED***
  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request_(
    ***REMOVED***
      uri: '/iam',
      qs: options,
    ***REMOVED***,
    callback
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** SetPolicyResponse
 * @property ***REMOVED***object***REMOVED*** 0 The policy.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback SetPolicyCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object***REMOVED*** acl The policy.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Set the IAM policy.
 *
 * @throws ***REMOVED***Error***REMOVED*** If no policy is provided.
 *
 * @param ***REMOVED***object***REMOVED*** policy The policy.
 * @param ***REMOVED***array***REMOVED*** policy.bindings Bindings associate members with roles.
 * @param ***REMOVED***string***REMOVED*** [policy.etag] Etags are used to perform a read-modify-write.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration opbject.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***SetPolicyCallback***REMOVED*** callback Callback function.
 * @returns ***REMOVED***Promise<SetPolicyResponse>***REMOVED***
 *
 * @see [Buckets: setIamPolicy API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/setIamPolicy***REMOVED***
 * @see [IAM Roles](https://cloud.google.com/iam/docs/understanding-roles)
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('my-bucket');
 *
 * var myPolicy = ***REMOVED***
 *   bindings: [
 *     ***REMOVED***
 *       role: 'roles/storage.admin',
 *       members: ['serviceAccount:myotherproject@appspot.gserviceaccount.com']
 *     ***REMOVED***
 *   ]
 * ***REMOVED***;
 *
 * bucket.iam.setPolicy(myPolicy, function(err, policy, apiResponse) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.iam.setPolicy(myPolicy).then(function(data) ***REMOVED***
 *   var policy = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/iam.js</caption>
 * region_tag:storage_add_bucket_iam_member
 * Example of adding to a bucket's IAM policy:
 *
 * @example <caption>include:samples/iam.js</caption>
 * region_tag:storage_remove_bucket_iam_member
 * Example of removing from a bucket's IAM policy:
 */
Iam.prototype.setPolicy = function(policy, options, callback) ***REMOVED***
  if (!is.object(policy)) ***REMOVED***
    throw new Error('A policy object is required.');
  ***REMOVED***

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  this.request_(
    ***REMOVED***
      method: 'PUT',
      uri: '/iam',
      json: extend(
        ***REMOVED***
          resourceId: this.resourceId_,
        ***REMOVED***,
        policy
      ),
      qs: options,
    ***REMOVED***,
    callback
  );
***REMOVED***;

/**
 * @typedef ***REMOVED***array***REMOVED*** TestIamPermissionsResponse
 * @property ***REMOVED***object[]***REMOVED*** 0 A subset of permissions that the caller is allowed.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback TestIamPermissionsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***object[]***REMOVED*** acl A subset of permissions that the caller is allowed.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Test a set of permissions for a resource.
 *
 * @throws ***REMOVED***Error***REMOVED*** If permissions are not provided.
 *
 * @param ***REMOVED***string|string[]***REMOVED*** permissions The permission(s) to test for.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration object.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] The ID of the project which will be
 *     billed for the request.
 * @param ***REMOVED***TestIamPermissionsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<TestIamPermissionsResponse>***REMOVED***
 *
 * @see [Buckets: testIamPermissions API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/testIamPermissions***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var bucket = storage.bucket('my-bucket');
 *
 * //-
 * // Test a single permission.
 * //-
 * var test = 'storage.buckets.delete';
 *
 * bucket.iam.testPermissions(test, function(err, permissions, apiResponse) ***REMOVED***
 *   console.log(permissions);
 *   // ***REMOVED***
 *   //   "storage.buckets.delete": true
 *   // ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // Test several permissions at once.
 * //-
 * var tests = [
 *   'storage.buckets.delete',
 *   'storage.buckets.get'
 * ];
 *
 * bucket.iam.testPermissions(tests, function(err, permissions) ***REMOVED***
 *   console.log(permissions);
 *   // ***REMOVED***
 *   //   "storage.buckets.delete": false,
 *   //   "storage.buckets.get": true
 *   // ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * bucket.iam.testPermissions(test).then(function(data) ***REMOVED***
 *   var permissions = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Iam.prototype.testPermissions = function(permissions, options, callback) ***REMOVED***
  if (!is.array(permissions) && !is.string(permissions)) ***REMOVED***
    throw new Error('Permissions are required.');
  ***REMOVED***

  if (is.fn(options)) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  options = extend(
    ***REMOVED***
      permissions: arrify(permissions),
    ***REMOVED***,
    options
  );

  this.request_(
    ***REMOVED***
      uri: '/iam/testPermissions',
      qs: options,
      useQuerystring: true,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      var availablePermissions = arrify(resp.permissions);

      var permissionsHash = permissions.reduce(function(acc, permission) ***REMOVED***
        acc[permission] = availablePermissions.indexOf(permission) > -1;
        return acc;
      ***REMOVED***, ***REMOVED******REMOVED***);

      callback(null, permissionsHash, resp);
    ***REMOVED***
  );
***REMOVED***;

/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
common.util.promisifyAll(Iam);

module.exports = Iam;
