// @google-cloud/storage/src/index.js
/**
 * Copyright 2014-2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var arrify = require('arrify');
var common = require('@google-cloud/common');
var extend = require('extend');
var util = require('util');

var Bucket = require('./bucket.js');
var Channel = require('./channel.js');
var File = require('./file.js');

/**
 * @typedef ***REMOVED***object***REMOVED*** ClientConfig
 * @property ***REMOVED***string***REMOVED*** [projectId] The project ID from the Google Developer's
 *     Console, e.g. 'grape-spaceship-123'. We will also check the environment
 *     variable `GCLOUD_PROJECT` for your project ID. If your app is running in
 *     an environment which supports ***REMOVED***@link https://cloud.google.com/docs/authentication/production#providing_credentials_to_your_application Application Default Credentials***REMOVED***,
 *     your project ID will be detected automatically.
 * @property ***REMOVED***string***REMOVED*** [keyFilename] Full path to the a .json, .pem, or .p12 key
 *     downloaded from the Google Developers Console. If you provide a path to a
 *     JSON file, the `projectId` option above is not necessary. NOTE: .pem and
 *     .p12 require you to specify the `email` option as well.
 * @property ***REMOVED***string***REMOVED*** [email] Account email address. Required when using a .pem
 *     or .p12 keyFilename.
 * @property ***REMOVED***object***REMOVED*** [credentials] Credentials object.
 * @property ***REMOVED***string***REMOVED*** [credentials.client_email]
 * @property ***REMOVED***string***REMOVED*** [credentials.private_key]
 * @property ***REMOVED***boolean***REMOVED*** [autoRetry=true] Automatically retry requests if the
 *     response is related to rate limits or certain intermittent server errors.
 *     We will exponentially backoff subsequent requests by default.
 * @property ***REMOVED***number***REMOVED*** [maxRetries=3] Maximum number of automatic retries
 *     attempted before returning the error.
 * @property ***REMOVED***Constructor***REMOVED*** [promise] Custom promise module to use instead of
 *     native Promises.
 */

/*! Developer Documentation
 *
 * Invoke this method to create a new Storage object bound with pre-determined
 * configuration options. For each object that can be created (e.g., a bucket),
 * there is an equivalent static and instance method. While they are classes,
 * they can be instantiated without use of the `new` keyword.
 */
/**
 * <h4>ACLs</h4>
 * Cloud Storage uses access control lists (ACLs) to manage object and
 * bucket access. ACLs are the mechanism you use to share files with other users
 * and allow other users to access your buckets and files.
 *
 * To learn more about ACLs, read this overview on
 * [Access Control](https://cloud.google.com/storage/docs/access-control).
 *
 * @see [Cloud Storage overview]***REMOVED***@link https://cloud.google.com/storage/docs/overview***REMOVED***
 * @see [Access Control]***REMOVED***@link https://cloud.google.com/storage/docs/access-control***REMOVED***
 *
 * @class
 * @hideconstructor
 *
 * @example <caption>Create a client that uses Application Default Credentials (ADC)</caption>
 * var storage = require('@google-cloud/storage')();
 *
 * @example <caption>Create a client with explicit credentials</caption>
 * var storage = require('@google-cloud/storage')(***REMOVED***
 *   projectId: 'your-project-id',
 *   keyFilename: '/path/to/keyfile.json'
 * ***REMOVED***);
 *
 * @param ***REMOVED***ClientConfig***REMOVED*** [options] Configuration options.
 */
function Storage(options) ***REMOVED***
  if (!(this instanceof Storage)) ***REMOVED***
    options = common.util.normalizeArguments(this, options);
    return new Storage(options);
  ***REMOVED***

  var config = ***REMOVED***
    baseUrl: 'https://www.googleapis.com/storage/v1',
    projectIdRequired: false,
    scopes: [
      'https://www.googleapis.com/auth/iam',
      'https://www.googleapis.com/auth/cloud-platform',
      'https://www.googleapis.com/auth/devstorage.full_control',
    ],
    packageJson: require('../package.json'),
  ***REMOVED***;

  common.Service.call(this, config, options);
***REMOVED***

util.inherits(Storage, common.Service);

/**
 * Cloud Storage uses access control lists (ACLs) to manage object and
 * bucket access. ACLs are the mechanism you use to share objects with other
 * users and allow other users to access your buckets and objects.
 *
 * This object provides constants to refer to the three permission levels that
 * can be granted to an entity:
 *
 *   - `gcs.acl.OWNER_ROLE` - ("OWNER")
 *   - `gcs.acl.READER_ROLE` - ("READER")
 *   - `gcs.acl.WRITER_ROLE` - ("WRITER")
 *
 * @see [About Access Control Lists]***REMOVED***@link https://cloud.google.com/storage/docs/access-control/lists***REMOVED***
 *
 * @name Storage.acl
 * @type ***REMOVED***object***REMOVED***
 * @property ***REMOVED***string***REMOVED*** OWNER_ROLE
 * @property ***REMOVED***string***REMOVED*** READER_ROLE
 * @property ***REMOVED***string***REMOVED*** WRITER_ROLE
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var albums = storage.bucket('albums');
 *
 * //-
 * // Make all of the files currently in a bucket publicly readable.
 * //-
 * var options = ***REMOVED***
 *   entity: 'allUsers',
 *   role: storage.acl.READER_ROLE
 * ***REMOVED***;
 *
 * albums.acl.add(options, function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // Make any new objects added to a bucket publicly readable.
 * //-
 * albums.acl.default.add(options, function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // Grant a user ownership permissions to a bucket.
 * //-
 * albums.acl.add(***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   role: storage.acl.OWNER_ROLE
 * ***REMOVED***, function(err, aclObject) ***REMOVED******REMOVED***);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * albums.acl.add(options).then(function(data) ***REMOVED***
 *   var aclObject = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 */
Storage.acl = ***REMOVED***
  OWNER_ROLE: 'OWNER',
  READER_ROLE: 'READER',
  WRITER_ROLE: 'WRITER',
***REMOVED***;

/**
 * Reference to ***REMOVED***@link Storage.acl***REMOVED***.
 *
 * @name Storage#acl
 * @see Storage.acl
 */
Storage.prototype.acl = Storage.acl;

/**
 * Get a reference to a Cloud Storage bucket.
 *
 * @param ***REMOVED***string***REMOVED*** name Name of the bucket.
 * @param ***REMOVED***object***REMOVED*** [options] Configuration object.
 * @param ***REMOVED***string***REMOVED*** [options.userProject] User project to be billed for all
 *     requests made from this Bucket object.
 * @returns ***REMOVED***Bucket***REMOVED***
 * @see Bucket
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var albums = storage.bucket('albums');
 * var photos = storage.bucket('photos');
 */
Storage.prototype.bucket = function(name, options) ***REMOVED***
  if (!name) ***REMOVED***
    throw new Error('A bucket name is needed to use Cloud Storage.');
  ***REMOVED***

  return new Bucket(this, name, options);
***REMOVED***;

/**
 * Reference a channel to receive notifications about changes to your bucket.
 *
 * @param ***REMOVED***string***REMOVED*** id The ID of the channel.
 * @param ***REMOVED***string***REMOVED*** resourceId The resource ID of the channel.
 * @returns ***REMOVED***Channel***REMOVED***
 * @see Channel
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var channel = storage.channel('id', 'resource-id');
 */
Storage.prototype.channel = function(id, resourceId) ***REMOVED***
  return new Channel(this, id, resourceId);
***REMOVED***;

/**
 * Metadata to set for the bucket.
 *
 * @typedef ***REMOVED***object***REMOVED*** CreateBucketRequest
 * @property ***REMOVED***boolean***REMOVED*** [coldline=false] Specify the storage class as Coldline.
 * @property ***REMOVED***boolean***REMOVED*** [dra=false] Specify the storage class as Durable Reduced
 *     Availability.
 * @property ***REMOVED***boolean***REMOVED*** [multiRegional=false] Specify the storage class as
 *     Multi-Regional.
 * @property ***REMOVED***boolean***REMOVED*** [nearline=false] Specify the storage class as Nearline.
 * @property ***REMOVED***boolean***REMOVED*** [regional=false] Specify the storage class as Regional.
 * @property ***REMOVED***boolean***REMOVED*** [requesterPays=false] **Early Access Testers Only**
 *     Force the use of the User Project metadata field to assign operational
 *     costs when an operation is made on a Bucket and its objects.
 * @property ***REMOVED***string***REMOVED*** [userProject] The ID of the project which will be billed
 *     for the request.
 */
/**
 * @typedef ***REMOVED***array***REMOVED*** CreateBucketResponse
 * @property ***REMOVED***Bucket***REMOVED*** 0 The new ***REMOVED***@link Bucket***REMOVED***.
 * @property ***REMOVED***object***REMOVED*** 1 The full API response.
 */
/**
 * @callback CreateBucketCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***Bucket***REMOVED*** bucket The new ***REMOVED***@link Bucket***REMOVED***.
 * @param ***REMOVED***object***REMOVED*** apiResponse The full API response.
 */
/**
 * Create a bucket.
 *
 * Cloud Storage uses a flat namespace, so you can't create a bucket with
 * a name that is already in use. For more information, see
 * [Bucket Naming Guidelines](https://cloud.google.com/storage/docs/bucketnaming.html#requirements).
 *
 * @see [Buckets: insert API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/insert***REMOVED***
 * @see [Storage Classes]***REMOVED***@link https://cloud.google.com/storage/docs/storage-classes***REMOVED***
 *
 * @param ***REMOVED***string***REMOVED*** name Name of the bucket to create.
 * @param ***REMOVED***CreateBucketRequest***REMOVED*** [metadata] Metadata to set for the bucket.
 * @param ***REMOVED***CreateBucketCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<CreateBucketResponse>***REMOVED***
 * @throws ***REMOVED***Error***REMOVED*** If a name is not provided.
 * @see Bucket#create
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * var callback = function(err, bucket, apiResponse) ***REMOVED***
 *   // `bucket` is a Bucket object.
 * ***REMOVED***;
 *
 * storage.createBucket('new-bucket', callback);
 *
 * //-
 * // Create a bucket in a specific location and region. <em>See the <a
 * // href="https://cloud.google.com/storage/docs/json_api/v1/buckets/insert">
 * // Official JSON API docs</a> for complete details on the `location` option.
 * // </em>
 * //-
 * var metadata = ***REMOVED***
 *   location: 'US-CENTRAL1',
 *   regional: true
 * ***REMOVED***;
 *
 * storage.createBucket('new-bucket', metadata, callback);
 *
 * //-
 * // Enable versioning on a new bucket.
 * //-
 * var metadata = ***REMOVED***
 *   versioning: ***REMOVED***
 *     enabled: true
 *   ***REMOVED***
 * ***REMOVED***;
 *
 * storage.createBucket('new-bucket', metadata, callback);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * storage.createBucket('new-bucket').then(function(data) ***REMOVED***
 *   var bucket = data[0];
 *   var apiResponse = data[1];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/buckets.js</caption>
 * region_tag:storage_create_bucket
 * Another example:
 */
Storage.prototype.createBucket = function(name, metadata, callback) ***REMOVED***
  var self = this;

  if (!name) ***REMOVED***
    throw new Error('A name is required to create a bucket.');
  ***REMOVED***

  if (!callback) ***REMOVED***
    callback = metadata;
    metadata = ***REMOVED******REMOVED***;
  ***REMOVED***

  var body = extend(***REMOVED******REMOVED***, metadata, ***REMOVED***
    name: name,
  ***REMOVED***);

  var storageClasses = ***REMOVED***
    coldline: 'COLDLINE',
    dra: 'DURABLE_REDUCED_AVAILABILITY',
    multiRegional: 'MULTI_REGIONAL',
    nearline: 'NEARLINE',
    regional: 'REGIONAL',
  ***REMOVED***;

  Object.keys(storageClasses).forEach(function(storageClass) ***REMOVED***
    if (body[storageClass]) ***REMOVED***
      body.storageClass = storageClasses[storageClass];
      delete body[storageClass];
    ***REMOVED***
  ***REMOVED***);

  if (body.requesterPays) ***REMOVED***
    body.billing = ***REMOVED***
      requesterPays: body.requesterPays,
    ***REMOVED***;
    delete body.requesterPays;
  ***REMOVED***

  var query = ***REMOVED***
    project: this.projectId,
  ***REMOVED***;

  if (body.userProject) ***REMOVED***
    query.userProject = body.userProject;
    delete body.userProject;
  ***REMOVED***

  this.request(
    ***REMOVED***
      method: 'POST',
      uri: '/b',
      qs: query,
      json: body,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, resp);
        return;
      ***REMOVED***

      var bucket = self.bucket(name);
      bucket.metadata = resp;

      callback(null, bucket, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * Query object for listing buckets.
 *
 * @typedef ***REMOVED***object***REMOVED*** GetBucketsRequest
 * @property ***REMOVED***boolean***REMOVED*** [autoPaginate=true] Have pagination handled
 *     automatically.
 * @property ***REMOVED***number***REMOVED*** [maxApiCalls] Maximum number of API calls to make.
 * @property ***REMOVED***number***REMOVED*** [maxResults] Maximum number of items plus prefixes to
 *     return.
 * @property ***REMOVED***string***REMOVED*** [pageToken] A previously-returned page token
 *     representing part of the larger set of results to view.
 * @property ***REMOVED***string***REMOVED*** [userProject] The ID of the project which will be billed
 *     for the request.
 */
/**
 * @typedef ***REMOVED***array***REMOVED*** GetBucketsResponse
 * @property ***REMOVED***Bucket[]***REMOVED*** 0 Array of ***REMOVED***@link Bucket***REMOVED*** instances.
 */
/**
 * @callback GetBucketsCallback
 * @param ***REMOVED***?Error***REMOVED*** err Request error, if any.
 * @param ***REMOVED***Bucket[]***REMOVED*** buckets Array of ***REMOVED***@link Bucket***REMOVED*** instances.
 */
/**
 * Get Bucket objects for all of the buckets in your project.
 *
 * @see [Buckets: list API Documentation]***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/list***REMOVED***
 *
 * @param ***REMOVED***GetBucketsRequest***REMOVED*** [query] Query object for listing buckets.
 * @param ***REMOVED***GetBucketsCallback***REMOVED*** [callback] Callback function.
 * @returns ***REMOVED***Promise<GetBucketsResponse>***REMOVED***
 *
 * @example
 * var storage = require('@google-cloud/storage')();
 * storage.getBuckets(function(err, buckets) ***REMOVED***
 *   if (!err) ***REMOVED***
 *     // buckets is an array of Bucket objects.
 *   ***REMOVED***
 * ***REMOVED***);
 *
 * //-
 * // To control how many API requests are made and page through the results
 * // manually, set `autoPaginate` to `false`.
 * //-
 * var callback = function(err, buckets, nextQuery, apiResponse) ***REMOVED***
 *   if (nextQuery) ***REMOVED***
 *     // More results exist.
 *     storage.getBuckets(nextQuery, callback);
 *   ***REMOVED***
 *
 *   // The `metadata` property is populated for you with the metadata at the
 *   // time of fetching.
 *   buckets[0].metadata;
 *
 *   // However, in cases where you are concerned the metadata could have
 *   // changed, use the `getMetadata` method.
 *   buckets[0].getMetadata(function(err, metadata, apiResponse) ***REMOVED******REMOVED***);
 * ***REMOVED***;
 *
 * storage.getBuckets(***REMOVED***
 *   autoPaginate: false
 * ***REMOVED***, callback);
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * storage.getBuckets().then(function(data) ***REMOVED***
 *   var buckets = data[0];
 * ***REMOVED***);
 *
 * @example <caption>include:samples/buckets.js</caption>
 * region_tag:storage_list_buckets
 * Another example:
 */
Storage.prototype.getBuckets = function(query, callback) ***REMOVED***
  var self = this;

  if (!callback) ***REMOVED***
    callback = query;
    query = ***REMOVED******REMOVED***;
  ***REMOVED***

  query.project = query.project || this.projectId;

  this.request(
    ***REMOVED***
      uri: '/b',
      qs: query,
    ***REMOVED***,
    function(err, resp) ***REMOVED***
      if (err) ***REMOVED***
        callback(err, null, null, resp);
        return;
      ***REMOVED***

      var buckets = arrify(resp.items).map(function(bucket) ***REMOVED***
        var bucketInstance = self.bucket(bucket.id);
        bucketInstance.metadata = bucket;
        return bucketInstance;
      ***REMOVED***);

      var nextQuery = null;
      if (resp.nextPageToken) ***REMOVED***
        nextQuery = extend(***REMOVED******REMOVED***, query, ***REMOVED***pageToken: resp.nextPageToken***REMOVED***);
      ***REMOVED***

      callback(null, buckets, nextQuery, resp);
    ***REMOVED***
  );
***REMOVED***;

/**
 * Get ***REMOVED***@link Bucket***REMOVED*** objects for all of the buckets in your project as
 * a readable object stream.
 *
 * @method Storage#getBucketsStream
 * @param ***REMOVED***GetBucketsRequest***REMOVED*** [query] Query object for listing buckets.
 * @returns ***REMOVED***ReadableStream***REMOVED*** A readable stream that emits ***REMOVED***@link Bucket***REMOVED*** instances.
 *
 * @example
 * storage.getBucketsStream()
 *   .on('error', console.error)
 *   .on('data', function(bucket) ***REMOVED***
 *     // bucket is a Bucket object.
 *   ***REMOVED***)
 *   .on('end', function() ***REMOVED***
 *     // All buckets retrieved.
 *   ***REMOVED***);
 *
 * //-
 * // If you anticipate many results, you can end a stream early to prevent
 * // unnecessary processing and API requests.
 * //-
 * storage.getBucketsStream()
 *   .on('data', function(bucket) ***REMOVED***
 *     this.end();
 *   ***REMOVED***);
 */
Storage.prototype.getBucketsStream = common.paginator.streamify('getBuckets');

/*! Developer Documentation
 *
 * These methods can be auto-paginated.
 */
common.paginator.extend(Storage, 'getBuckets');

/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
common.util.promisifyAll(Storage, ***REMOVED***
  exclude: ['bucket', 'channel'],
***REMOVED***);

/**
 * ***REMOVED***@link Bucket***REMOVED*** class.
 *
 * @name Storage.Bucket
 * @see Bucket
 * @type ***REMOVED***Constructor***REMOVED***
 */
Storage.Bucket = Bucket;

/**
 * ***REMOVED***@link Channel***REMOVED*** class.
 *
 * @name Storage.Channel
 * @see Channel
 * @type ***REMOVED***Constructor***REMOVED***
 */
Storage.Channel = Channel;

/**
 * ***REMOVED***@link File***REMOVED*** class.
 *
 * @name Storage.File
 * @see File
 * @type ***REMOVED***Constructor***REMOVED***
 */
Storage.File = File;

/**
 * The default export of the `@google-cloud/storage` package is the
 * ***REMOVED***@link Storage***REMOVED*** class, which also serves as a factory function which produces
 * ***REMOVED***@link Storage***REMOVED*** instances.
 *
 * See ***REMOVED***@link Storage***REMOVED*** and ***REMOVED***@link ClientConfig***REMOVED*** for client methods and
 * configuration options.
 *
 * @module ***REMOVED***Storage***REMOVED*** @google-cloud/storage
 * @alias nodejs-storage
 *
 * @example <caption>Install the client library with <a href="https://www.npmjs.com/">npm</a>:</caption>
 * npm install --save @google-cloud/storage
 *
 * @example <caption>Import the client library</caption>
 * var Storage = require('@google-cloud/storage');
 *
 * @example <caption>Create a client that uses <a href="https://cloud.google.com/docs/authentication/production#providing_credentials_to_your_application">Application Default Credentials (ADC)</a>:</caption>
 * var storage = Storage();
 *
 * @example <caption>Create a client with <a href="https://cloud.google.com/docs/authentication/production#obtaining_and_providing_service_account_credentials_manually">explicit credentials</a>:</caption>
 * var storage = Storage(***REMOVED***
 *   projectId: 'your-project-id',
 *   keyFilename: '/path/to/keyfile.json'
 * ***REMOVED***);
 *
 * @example <caption>include:samples/quickstart.js</caption>
 * region_tag:storage_quickstart
 * Full quickstart example:
 */
module.exports = Storage;
