// es6-set/polyfill.js
'use strict';

var clear          = require('es5-ext/array/#/clear')
  , eIndexOf       = require('es5-ext/array/#/e-index-of')
  , setPrototypeOf = require('es5-ext/object/set-prototype-of')
  , callable       = require('es5-ext/object/valid-callable')
  , d              = require('d')
  , ee             = require('event-emitter')
  , Symbol         = require('es6-symbol')
  , iterator       = require('es6-iterator/valid-iterable')
  , forOf          = require('es6-iterator/for-of')
  , Iterator       = require('./lib/iterator')
  , isNative       = require('./is-native-implemented')

  , call = Function.prototype.call
  , defineProperty = Object.defineProperty, getPrototypeOf = Object.getPrototypeOf
  , SetPoly, getValues, NativeSet;

if (isNative) NativeSet = Set;

module.exports = SetPoly = function Set(/*iterable*/) ***REMOVED***
	var iterable = arguments[0], self;
	if (!(this instanceof SetPoly)) throw new TypeError('Constructor requires \'new\'');
	if (isNative && setPrototypeOf) self = setPrototypeOf(new NativeSet(), getPrototypeOf(this));
	else self = this;
	if (iterable != null) iterator(iterable);
	defineProperty(self, '__setData__', d('c', []));
	if (!iterable) return self;
	forOf(iterable, function (value) ***REMOVED***
		if (eIndexOf.call(this, value) !== -1) return;
		this.push(value);
	***REMOVED***, self.__setData__);
	return self;
***REMOVED***;

if (isNative) ***REMOVED***
	if (setPrototypeOf) setPrototypeOf(SetPoly, NativeSet);
	SetPoly.prototype = Object.create(NativeSet.prototype, ***REMOVED*** constructor: d(SetPoly) ***REMOVED***);
***REMOVED***

ee(Object.defineProperties(SetPoly.prototype, ***REMOVED***
	add: d(function (value) ***REMOVED***
		if (this.has(value)) return this;
		this.emit('_add', this.__setData__.push(value) - 1, value);
		return this;
	***REMOVED***),
	clear: d(function () ***REMOVED***
		if (!this.__setData__.length) return;
		clear.call(this.__setData__);
		this.emit('_clear');
	***REMOVED***),
	delete: d(function (value) ***REMOVED***
		var index = eIndexOf.call(this.__setData__, value);
		if (index === -1) return false;
		this.__setData__.splice(index, 1);
		this.emit('_delete', index, value);
		return true;
	***REMOVED***),
	entries: d(function () ***REMOVED*** return new Iterator(this, 'key+value'); ***REMOVED***),
	forEach: d(function (cb/*, thisArg*/) ***REMOVED***
		var thisArg = arguments[1], iterator, result, value;
		callable(cb);
		iterator = this.values();
		result = iterator._next();
		while (result !== undefined) ***REMOVED***
			value = iterator._resolve(result);
			call.call(cb, thisArg, value, value, this);
			result = iterator._next();
		***REMOVED***
	***REMOVED***),
	has: d(function (value) ***REMOVED***
		return (eIndexOf.call(this.__setData__, value) !== -1);
	***REMOVED***),
	keys: d(getValues = function () ***REMOVED*** return this.values(); ***REMOVED***),
	size: d.gs(function () ***REMOVED*** return this.__setData__.length; ***REMOVED***),
	values: d(function () ***REMOVED*** return new Iterator(this); ***REMOVED***),
	toString: d(function () ***REMOVED*** return '[object Set]'; ***REMOVED***)
***REMOVED***));
defineProperty(SetPoly.prototype, Symbol.iterator, d(getValues));
defineProperty(SetPoly.prototype, Symbol.toStringTag, d('c', 'Set'));
