// fsevents/fsevents.js
/*
 ** © 2014 by Philipp Dunkel <pip@pipobscure.com>
 ** Licensed under MIT License.
 */

/* jshint node:true */
'use strict';

if (process.platform !== 'darwin')
  throw new Error('Module \'fsevents\' is not compatible with platform \'' + process.platform + '\'');

var path = require('path');
var binary = require('node-pre-gyp');
var Native = require(binary.find(path.join(__dirname, 'package.json')));

var EventEmitter = require('events').EventEmitter;
var fs = require('fs');
var inherits = require('util').inherits;

function FSEvents(path, handler) ***REMOVED***
  EventEmitter.call(this);

  Object.defineProperty(this, '_impl', ***REMOVED***
    value: new Native.FSEvents(String(path || ''), handler),
    enumerable: false,
    writable: false
  ***REMOVED***);
***REMOVED***

inherits(FSEvents, EventEmitter);
proxies(FSEvents, Native.FSEvents);

module.exports = watch;
module.exports.getInfo = getInfo;
module.exports.FSEvents = Native.FSEvents;
module.exports.Constants = Native.Constants;

var defer = global.setImmediate || process.nextTick;

function watch(path) ***REMOVED***
  var fse = new FSEvents(String(path || ''), handler);
  EventEmitter.call(fse);
  return fse;

  function handler(path, flags, id) ***REMOVED***
    defer(function() ***REMOVED***
      fse.emit('fsevent', path, flags, id);
      var info = getInfo(path, flags);
      info.id = id;
      if (info.event === 'moved') ***REMOVED***
        fs.stat(info.path, function(err, stat) ***REMOVED***
          info.event = (err || !stat) ? 'moved-out' : 'moved-in';
          fse.emit('change', path, info);
          fse.emit(info.event, path, info);
        ***REMOVED***);
      ***REMOVED*** else ***REMOVED***
        fse.emit('change', path, info);
        fse.emit(info.event, path, info);
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
***REMOVED***

function proxies(ctor, target) ***REMOVED***
  Object.keys(target.prototype).filter(function(key) ***REMOVED***
    return typeof target.prototype[key] === 'function';
  ***REMOVED***).forEach(function(key) ***REMOVED***
    ctor.prototype[key] = function() ***REMOVED***
      this._impl[key].apply(this._impl, arguments);
      return this;
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

function getFileType(flags) ***REMOVED***
  if (Native.Constants.kFSEventStreamEventFlagItemIsFile & flags) return 'file';
  if (Native.Constants.kFSEventStreamEventFlagItemIsDir & flags) return 'directory';
  if (Native.Constants.kFSEventStreamEventFlagItemIsSymlink & flags) return 'symlink';
***REMOVED***

function getEventType(flags) ***REMOVED***
  if (Native.Constants.kFSEventStreamEventFlagItemRemoved & flags) return 'deleted';
  if (Native.Constants.kFSEventStreamEventFlagItemRenamed & flags) return 'moved';
  if (Native.Constants.kFSEventStreamEventFlagItemCreated & flags) return 'created';
  if (Native.Constants.kFSEventStreamEventFlagItemModified & flags) return 'modified';
  if (Native.Constants.kFSEventStreamEventFlagRootChanged & flags) return 'root-changed';

  return 'unknown';
***REMOVED***

function getFileChanges(flags) ***REMOVED***
  return ***REMOVED***
    inode: !! (Native.Constants.kFSEventStreamEventFlagItemInodeMetaMod & flags),
    finder: !! (Native.Constants.kFSEventStreamEventFlagItemFinderInfoMod & flags),
    access: !! (Native.Constants.kFSEventStreamEventFlagItemChangeOwner & flags),
    xattrs: !! (Native.Constants.kFSEventStreamEventFlagItemXattrMod & flags)
  ***REMOVED***;
***REMOVED***

function getInfo(path, flags) ***REMOVED***
  return ***REMOVED***
    path: path,
    event: getEventType(flags),
    type: getFileType(flags),
    changes: getFileChanges(flags),
    flags: flags
  ***REMOVED***;
***REMOVED***
