// which/which.js
module.exports = which
which.sync = whichSync

var isWindows = process.platform === 'win32' ||
    process.env.OSTYPE === 'cygwin' ||
    process.env.OSTYPE === 'msys'

var path = require('path')
var COLON = isWindows ? ';' : ':'
var isexe = require('isexe')

function getNotFoundError (cmd) ***REMOVED***
  var er = new Error('not found: ' + cmd)
  er.code = 'ENOENT'

  return er
***REMOVED***

function getPathInfo (cmd, opt) ***REMOVED***
  var colon = opt.colon || COLON
  var pathEnv = opt.path || process.env.PATH || ''
  var pathExt = ['']

  pathEnv = pathEnv.split(colon)

  var pathExtExe = ''
  if (isWindows) ***REMOVED***
    pathEnv.unshift(process.cwd())
    pathExtExe = (opt.pathExt || process.env.PATHEXT || '.EXE;.CMD;.BAT;.COM')
    pathExt = pathExtExe.split(colon)


    // Always test the cmd itself first.  isexe will check to make sure
    // it's found in the pathExt set.
    if (cmd.indexOf('.') !== -1 && pathExt[0] !== '')
      pathExt.unshift('')
  ***REMOVED***

  // If it has a slash, then we don't bother searching the pathenv.
  // just check the file itself, and that's it.
  if (cmd.match(/\//) || isWindows && cmd.match(/\\/))
    pathEnv = ['']

  return ***REMOVED***
    env: pathEnv,
    ext: pathExt,
    extExe: pathExtExe
  ***REMOVED***
***REMOVED***

function which (cmd, opt, cb) ***REMOVED***
  if (typeof opt === 'function') ***REMOVED***
    cb = opt
    opt = ***REMOVED******REMOVED***
  ***REMOVED***

  var info = getPathInfo(cmd, opt)
  var pathEnv = info.env
  var pathExt = info.ext
  var pathExtExe = info.extExe
  var found = []

  ;(function F (i, l) ***REMOVED***
    if (i === l) ***REMOVED***
      if (opt.all && found.length)
        return cb(null, found)
      else
        return cb(getNotFoundError(cmd))
    ***REMOVED***

    var pathPart = pathEnv[i]
    if (pathPart.charAt(0) === '"' && pathPart.slice(-1) === '"')
      pathPart = pathPart.slice(1, -1)

    var p = path.join(pathPart, cmd)
    if (!pathPart && (/^\.[\\\/]/).test(cmd)) ***REMOVED***
      p = cmd.slice(0, 2) + p
    ***REMOVED***
    ;(function E (ii, ll) ***REMOVED***
      if (ii === ll) return F(i + 1, l)
      var ext = pathExt[ii]
      isexe(p + ext, ***REMOVED*** pathExt: pathExtExe ***REMOVED***, function (er, is) ***REMOVED***
        if (!er && is) ***REMOVED***
          if (opt.all)
            found.push(p + ext)
          else
            return cb(null, p + ext)
        ***REMOVED***
        return E(ii + 1, ll)
      ***REMOVED***)
    ***REMOVED***)(0, pathExt.length)
  ***REMOVED***)(0, pathEnv.length)
***REMOVED***

function whichSync (cmd, opt) ***REMOVED***
  opt = opt || ***REMOVED******REMOVED***

  var info = getPathInfo(cmd, opt)
  var pathEnv = info.env
  var pathExt = info.ext
  var pathExtExe = info.extExe
  var found = []

  for (var i = 0, l = pathEnv.length; i < l; i ++) ***REMOVED***
    var pathPart = pathEnv[i]
    if (pathPart.charAt(0) === '"' && pathPart.slice(-1) === '"')
      pathPart = pathPart.slice(1, -1)

    var p = path.join(pathPart, cmd)
    if (!pathPart && /^\.[\\\/]/.test(cmd)) ***REMOVED***
      p = cmd.slice(0, 2) + p
    ***REMOVED***
    for (var j = 0, ll = pathExt.length; j < ll; j ++) ***REMOVED***
      var cur = p + pathExt[j]
      var is
      try ***REMOVED***
        is = isexe.sync(cur, ***REMOVED*** pathExt: pathExtExe ***REMOVED***)
        if (is) ***REMOVED***
          if (opt.all)
            found.push(cur)
          else
            return cur
        ***REMOVED***
      ***REMOVED*** catch (ex) ***REMOVED******REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (opt.all && found.length)
    return found

  if (opt.nothrow)
    return null

  throw getNotFoundError(cmd)
***REMOVED***
