// vm-browserify/index.js
var indexOf = require('indexof');

var Object_keys = function (obj) ***REMOVED***
    if (Object.keys) return Object.keys(obj)
    else ***REMOVED***
        var res = [];
        for (var key in obj) res.push(key)
        return res;
    ***REMOVED***
***REMOVED***;

var forEach = function (xs, fn) ***REMOVED***
    if (xs.forEach) return xs.forEach(fn)
    else for (var i = 0; i < xs.length; i++) ***REMOVED***
        fn(xs[i], i, xs);
    ***REMOVED***
***REMOVED***;

var defineProp = (function() ***REMOVED***
    try ***REMOVED***
        Object.defineProperty(***REMOVED******REMOVED***, '_', ***REMOVED******REMOVED***);
        return function(obj, name, value) ***REMOVED***
            Object.defineProperty(obj, name, ***REMOVED***
                writable: true,
                enumerable: false,
                configurable: true,
                value: value
            ***REMOVED***)
        ***REMOVED***;
    ***REMOVED*** catch(e) ***REMOVED***
        return function(obj, name, value) ***REMOVED***
            obj[name] = value;
        ***REMOVED***;
    ***REMOVED***
***REMOVED***());

var globals = ['Array', 'Boolean', 'Date', 'Error', 'EvalError', 'Function',
'Infinity', 'JSON', 'Math', 'NaN', 'Number', 'Object', 'RangeError',
'ReferenceError', 'RegExp', 'String', 'SyntaxError', 'TypeError', 'URIError',
'decodeURI', 'decodeURIComponent', 'encodeURI', 'encodeURIComponent', 'escape',
'eval', 'isFinite', 'isNaN', 'parseFloat', 'parseInt', 'undefined', 'unescape'];

function Context() ***REMOVED******REMOVED***
Context.prototype = ***REMOVED******REMOVED***;

var Script = exports.Script = function NodeScript (code) ***REMOVED***
    if (!(this instanceof Script)) return new Script(code);
    this.code = code;
***REMOVED***;

Script.prototype.runInContext = function (context) ***REMOVED***
    if (!(context instanceof Context)) ***REMOVED***
        throw new TypeError("needs a 'context' argument.");
    ***REMOVED***
    
    var iframe = document.createElement('iframe');
    if (!iframe.style) iframe.style = ***REMOVED******REMOVED***;
    iframe.style.display = 'none';
    
    document.body.appendChild(iframe);
    
    var win = iframe.contentWindow;
    var wEval = win.eval, wExecScript = win.execScript;

    if (!wEval && wExecScript) ***REMOVED***
        // win.eval() magically appears when this is called in IE:
        wExecScript.call(win, 'null');
        wEval = win.eval;
    ***REMOVED***
    
    forEach(Object_keys(context), function (key) ***REMOVED***
        win[key] = context[key];
    ***REMOVED***);
    forEach(globals, function (key) ***REMOVED***
        if (context[key]) ***REMOVED***
            win[key] = context[key];
        ***REMOVED***
    ***REMOVED***);
    
    var winKeys = Object_keys(win);

    var res = wEval.call(win, this.code);
    
    forEach(Object_keys(win), function (key) ***REMOVED***
        // Avoid copying circular objects like `top` and `window` by only
        // updating existing context properties or new properties in the `win`
        // that was only introduced after the eval.
        if (key in context || indexOf(winKeys, key) === -1) ***REMOVED***
            context[key] = win[key];
        ***REMOVED***
    ***REMOVED***);

    forEach(globals, function (key) ***REMOVED***
        if (!(key in context)) ***REMOVED***
            defineProp(context, key, win[key]);
        ***REMOVED***
    ***REMOVED***);
    
    document.body.removeChild(iframe);
    
    return res;
***REMOVED***;

Script.prototype.runInThisContext = function () ***REMOVED***
    return eval(this.code); // maybe...
***REMOVED***;

Script.prototype.runInNewContext = function (context) ***REMOVED***
    var ctx = Script.createContext(context);
    var res = this.runInContext(ctx);

    forEach(Object_keys(ctx), function (key) ***REMOVED***
        context[key] = ctx[key];
    ***REMOVED***);

    return res;
***REMOVED***;

forEach(Object_keys(Script.prototype), function (name) ***REMOVED***
    exports[name] = Script[name] = function (code) ***REMOVED***
        var s = Script(code);
        return s[name].apply(s, [].slice.call(arguments, 1));
    ***REMOVED***;
***REMOVED***);

exports.createScript = function (code) ***REMOVED***
    return exports.Script(code);
***REMOVED***;

exports.createContext = Script.createContext = function (context) ***REMOVED***
    var copy = new Context();
    if(typeof context === 'object') ***REMOVED***
        forEach(Object_keys(context), function (key) ***REMOVED***
            copy[key] = context[key];
        ***REMOVED***);
    ***REMOVED***
    return copy;
***REMOVED***;
