// power-assert-renderer-comparison/index.js
'use strict';

var BaseRenderer = require('power-assert-renderer-base');
var inherits = require('util').inherits;
var typeName = require('type-name');
var keys = require('core-js/library/fn/object/keys');
var forEach = require('core-js/library/fn/array/for-each');
var udiff = require('./lib/udiff');
var stringifier = require('stringifier');
var assign = require('core-js/library/fn/object/assign');
var defaultOptions = require('./lib/default-options');
var literalPattern = /^(?:String|Numeric|Null|Boolean|RegExp)?Literal$/;

function isLiteral (node) ***REMOVED***
    return literalPattern.test(node.type);
***REMOVED***

/**
 * options.stringify [function]
 * options.maxDepth [number]
 * options.lineSeparator [string]
 * options.anonymous [string]
 * options.circular [string]
 * 
 * options.diff [function]
 * options.lineDiffThreshold [number]
 */
function ComparisonRenderer (config) ***REMOVED***
    BaseRenderer.call(this);
    this.config = assign(***REMOVED******REMOVED***, defaultOptions(), config);
    if (typeof this.config.stringify === 'function') ***REMOVED***
        this.stringify = this.config.stringify;
    ***REMOVED*** else ***REMOVED***
        this.stringify = stringifier(this.config);
    ***REMOVED***
    if (typeof this.config.diff === 'function') ***REMOVED***
        this.diff = this.config.diff;
    ***REMOVED*** else ***REMOVED***
        this.diff = udiff(this.config);
    ***REMOVED***
    this.espathToPair = ***REMOVED******REMOVED***;
***REMOVED***
inherits(ComparisonRenderer, BaseRenderer);

ComparisonRenderer.prototype.onData = function (esNode) ***REMOVED***
    var pair;
    if (!esNode.isCaptured) ***REMOVED***
        if (isTargetBinaryExpression(esNode.parent) && isLiteral(esNode.node)) ***REMOVED***
            this.espathToPair[esNode.parent.espath][esNode.key] = ***REMOVED***code: esNode.code, value: esNode.value***REMOVED***;
        ***REMOVED***
        return;
    ***REMOVED***
    if (isTargetBinaryExpression(esNode.parent)) ***REMOVED***
        this.espathToPair[esNode.parent.espath][esNode.key] = ***REMOVED***code: esNode.code, value: esNode.value***REMOVED***;
    ***REMOVED***
    if (isTargetBinaryExpression(esNode)) ***REMOVED***
        pair = ***REMOVED***
            operator: esNode.node.operator,
            value: esNode.value
        ***REMOVED***;
        this.espathToPair[esNode.espath] = pair;
    ***REMOVED***
***REMOVED***;

ComparisonRenderer.prototype.onEnd = function () ***REMOVED***
    var _this = this;
    var pairs = [];
    forEach(keys(this.espathToPair), function (espath) ***REMOVED***
        var pair = _this.espathToPair[espath];
        if (pair.left && pair.right) ***REMOVED***
            pairs.push(pair);
        ***REMOVED***
    ***REMOVED***);
    forEach(pairs, function (pair) ***REMOVED***
        _this.compare(pair);
    ***REMOVED***);
***REMOVED***;

ComparisonRenderer.prototype.compare = function (pair) ***REMOVED***
    if (isStringDiffTarget(pair)) ***REMOVED***
        this.showStringDiff(pair);
    ***REMOVED*** else ***REMOVED***
        this.showExpectedAndActual(pair);
    ***REMOVED***
***REMOVED***;

ComparisonRenderer.prototype.showExpectedAndActual = function (pair) ***REMOVED***
    this.write('');
    this.write('[' + typeName(pair.right.value) + '] ' + pair.right.code);
    this.write('=> ' + this.stringify(pair.right.value));
    this.write('[' + typeName(pair.left.value)  + '] ' + pair.left.code);
    this.write('=> ' + this.stringify(pair.left.value));
***REMOVED***;

ComparisonRenderer.prototype.showStringDiff = function (pair) ***REMOVED***
    this.write('');
    this.write('--- [string] ' + pair.right.code);
    this.write('+++ [string] ' + pair.left.code);
    this.write(this.diff(pair.right.value, pair.left.value, this.config));
***REMOVED***;

function isTargetBinaryExpression (esNode) ***REMOVED***
    return esNode &&
        esNode.node.type === 'BinaryExpression' &&
        (esNode.node.operator === '===' || esNode.node.operator === '==') &&
        esNode.isCaptured &&
        !(esNode.value);
***REMOVED***

function isStringDiffTarget(pair) ***REMOVED***
    return typeof pair.left.value === 'string' && typeof pair.right.value === 'string';
***REMOVED***

ComparisonRenderer.udiff = udiff;
module.exports = ComparisonRenderer;
