// assert/assert.js
'use strict';

// compare and isBuffer taken from https://github.com/feross/buffer/blob/680e9e5e488f22aac27599a57dc844a6315928dd/index.js
// original notice:

/*!
 * The buffer module from node.js, for the browser.
 *
 * @author   Feross Aboukhadijeh <feross@feross.org> <http://feross.org>
 * @license  MIT
 */
function compare(a, b) ***REMOVED***
  if (a === b) ***REMOVED***
    return 0;
  ***REMOVED***

  var x = a.length;
  var y = b.length;

  for (var i = 0, len = Math.min(x, y); i < len; ++i) ***REMOVED***
    if (a[i] !== b[i]) ***REMOVED***
      x = a[i];
      y = b[i];
      break;
    ***REMOVED***
  ***REMOVED***

  if (x < y) ***REMOVED***
    return -1;
  ***REMOVED***
  if (y < x) ***REMOVED***
    return 1;
  ***REMOVED***
  return 0;
***REMOVED***
function isBuffer(b) ***REMOVED***
  if (global.Buffer && typeof global.Buffer.isBuffer === 'function') ***REMOVED***
    return global.Buffer.isBuffer(b);
  ***REMOVED***
  return !!(b != null && b._isBuffer);
***REMOVED***

// based on node assert, original notice:

// http://wiki.commonjs.org/wiki/Unit_Testing/1.0
//
// THIS IS NOT TESTED NOR LIKELY TO WORK OUTSIDE V8!
//
// Originally from narwhal.js (http://narwhaljs.org)
// Copyright (c) 2009 Thomas Robinson <280north.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the 'Software'), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

var util = require('util/');
var hasOwn = Object.prototype.hasOwnProperty;
var pSlice = Array.prototype.slice;
var functionsHaveNames = (function () ***REMOVED***
  return function foo() ***REMOVED******REMOVED***.name === 'foo';
***REMOVED***());
function pToString (obj) ***REMOVED***
  return Object.prototype.toString.call(obj);
***REMOVED***
function isView(arrbuf) ***REMOVED***
  if (isBuffer(arrbuf)) ***REMOVED***
    return false;
  ***REMOVED***
  if (typeof global.ArrayBuffer !== 'function') ***REMOVED***
    return false;
  ***REMOVED***
  if (typeof ArrayBuffer.isView === 'function') ***REMOVED***
    return ArrayBuffer.isView(arrbuf);
  ***REMOVED***
  if (!arrbuf) ***REMOVED***
    return false;
  ***REMOVED***
  if (arrbuf instanceof DataView) ***REMOVED***
    return true;
  ***REMOVED***
  if (arrbuf.buffer && arrbuf.buffer instanceof ArrayBuffer) ***REMOVED***
    return true;
  ***REMOVED***
  return false;
***REMOVED***
// 1. The assert module provides functions that throw
// AssertionError's when particular conditions are not met. The
// assert module must conform to the following interface.

var assert = module.exports = ok;

// 2. The AssertionError is defined in assert.
// new assert.AssertionError(***REMOVED*** message: message,
//                             actual: actual,
//                             expected: expected ***REMOVED***)

var regex = /\s*function\s+([^\(\s]*)\s*/;
// based on https://github.com/ljharb/function.prototype.name/blob/adeeeec8bfcc6068b187d7d9fb3d5bb1d3a30899/implementation.js
function getName(func) ***REMOVED***
  if (!util.isFunction(func)) ***REMOVED***
    return;
  ***REMOVED***
  if (functionsHaveNames) ***REMOVED***
    return func.name;
  ***REMOVED***
  var str = func.toString();
  var match = str.match(regex);
  return match && match[1];
***REMOVED***
assert.AssertionError = function AssertionError(options) ***REMOVED***
  this.name = 'AssertionError';
  this.actual = options.actual;
  this.expected = options.expected;
  this.operator = options.operator;
  if (options.message) ***REMOVED***
    this.message = options.message;
    this.generatedMessage = false;
  ***REMOVED*** else ***REMOVED***
    this.message = getMessage(this);
    this.generatedMessage = true;
  ***REMOVED***
  var stackStartFunction = options.stackStartFunction || fail;
  if (Error.captureStackTrace) ***REMOVED***
    Error.captureStackTrace(this, stackStartFunction);
  ***REMOVED*** else ***REMOVED***
    // non v8 browsers so we can have a stacktrace
    var err = new Error();
    if (err.stack) ***REMOVED***
      var out = err.stack;

      // try to strip useless frames
      var fn_name = getName(stackStartFunction);
      var idx = out.indexOf('\n' + fn_name);
      if (idx >= 0) ***REMOVED***
        // once we have located the function frame
        // we need to strip out everything before it (and its line)
        var next_line = out.indexOf('\n', idx + 1);
        out = out.substring(next_line + 1);
      ***REMOVED***

      this.stack = out;
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

// assert.AssertionError instanceof Error
util.inherits(assert.AssertionError, Error);

function truncate(s, n) ***REMOVED***
  if (typeof s === 'string') ***REMOVED***
    return s.length < n ? s : s.slice(0, n);
  ***REMOVED*** else ***REMOVED***
    return s;
  ***REMOVED***
***REMOVED***
function inspect(something) ***REMOVED***
  if (functionsHaveNames || !util.isFunction(something)) ***REMOVED***
    return util.inspect(something);
  ***REMOVED***
  var rawname = getName(something);
  var name = rawname ? ': ' + rawname : '';
  return '[Function' +  name + ']';
***REMOVED***
function getMessage(self) ***REMOVED***
  return truncate(inspect(self.actual), 128) + ' ' +
         self.operator + ' ' +
         truncate(inspect(self.expected), 128);
***REMOVED***

// At present only the three keys mentioned above are used and
// understood by the spec. Implementations or sub modules can pass
// other keys to the AssertionError's constructor - they will be
// ignored.

// 3. All of the following functions must throw an AssertionError
// when a corresponding condition is not met, with a message that
// may be undefined if not provided.  All assertion methods provide
// both the actual and expected values to the assertion error for
// display purposes.

function fail(actual, expected, message, operator, stackStartFunction) ***REMOVED***
  throw new assert.AssertionError(***REMOVED***
    message: message,
    actual: actual,
    expected: expected,
    operator: operator,
    stackStartFunction: stackStartFunction
  ***REMOVED***);
***REMOVED***

// EXTENSION! allows for well behaved errors defined elsewhere.
assert.fail = fail;

// 4. Pure assertion tests whether a value is truthy, as determined
// by !!guard.
// assert.ok(guard, message_opt);
// This statement is equivalent to assert.equal(true, !!guard,
// message_opt);. To test strictly for the value true, use
// assert.strictEqual(true, guard, message_opt);.

function ok(value, message) ***REMOVED***
  if (!value) fail(value, true, message, '==', assert.ok);
***REMOVED***
assert.ok = ok;

// 5. The equality assertion tests shallow, coercive equality with
// ==.
// assert.equal(actual, expected, message_opt);

assert.equal = function equal(actual, expected, message) ***REMOVED***
  if (actual != expected) fail(actual, expected, message, '==', assert.equal);
***REMOVED***;

// 6. The non-equality assertion tests for whether two objects are not equal
// with != assert.notEqual(actual, expected, message_opt);

assert.notEqual = function notEqual(actual, expected, message) ***REMOVED***
  if (actual == expected) ***REMOVED***
    fail(actual, expected, message, '!=', assert.notEqual);
  ***REMOVED***
***REMOVED***;

// 7. The equivalence assertion tests a deep equality relation.
// assert.deepEqual(actual, expected, message_opt);

assert.deepEqual = function deepEqual(actual, expected, message) ***REMOVED***
  if (!_deepEqual(actual, expected, false)) ***REMOVED***
    fail(actual, expected, message, 'deepEqual', assert.deepEqual);
  ***REMOVED***
***REMOVED***;

assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) ***REMOVED***
  if (!_deepEqual(actual, expected, true)) ***REMOVED***
    fail(actual, expected, message, 'deepStrictEqual', assert.deepStrictEqual);
  ***REMOVED***
***REMOVED***;

function _deepEqual(actual, expected, strict, memos) ***REMOVED***
  // 7.1. All identical values are equivalent, as determined by ===.
  if (actual === expected) ***REMOVED***
    return true;
  ***REMOVED*** else if (isBuffer(actual) && isBuffer(expected)) ***REMOVED***
    return compare(actual, expected) === 0;

  // 7.2. If the expected value is a Date object, the actual value is
  // equivalent if it is also a Date object that refers to the same time.
  ***REMOVED*** else if (util.isDate(actual) && util.isDate(expected)) ***REMOVED***
    return actual.getTime() === expected.getTime();

  // 7.3 If the expected value is a RegExp object, the actual value is
  // equivalent if it is also a RegExp object with the same source and
  // properties (`global`, `multiline`, `lastIndex`, `ignoreCase`).
  ***REMOVED*** else if (util.isRegExp(actual) && util.isRegExp(expected)) ***REMOVED***
    return actual.source === expected.source &&
           actual.global === expected.global &&
           actual.multiline === expected.multiline &&
           actual.lastIndex === expected.lastIndex &&
           actual.ignoreCase === expected.ignoreCase;

  // 7.4. Other pairs that do not both pass typeof value == 'object',
  // equivalence is determined by ==.
  ***REMOVED*** else if ((actual === null || typeof actual !== 'object') &&
             (expected === null || typeof expected !== 'object')) ***REMOVED***
    return strict ? actual === expected : actual == expected;

  // If both values are instances of typed arrays, wrap their underlying
  // ArrayBuffers in a Buffer each to increase performance
  // This optimization requires the arrays to have the same type as checked by
  // Object.prototype.toString (aka pToString). Never perform binary
  // comparisons for Float*Arrays, though, since e.g. +0 === -0 but their
  // bit patterns are not identical.
  ***REMOVED*** else if (isView(actual) && isView(expected) &&
             pToString(actual) === pToString(expected) &&
             !(actual instanceof Float32Array ||
               actual instanceof Float64Array)) ***REMOVED***
    return compare(new Uint8Array(actual.buffer),
                   new Uint8Array(expected.buffer)) === 0;

  // 7.5 For all other Object pairs, including Array objects, equivalence is
  // determined by having the same number of owned properties (as verified
  // with Object.prototype.hasOwnProperty.call), the same set of keys
  // (although not necessarily the same order), equivalent values for every
  // corresponding key, and an identical 'prototype' property. Note: this
  // accounts for both named and indexed properties on Arrays.
  ***REMOVED*** else if (isBuffer(actual) !== isBuffer(expected)) ***REMOVED***
    return false;
  ***REMOVED*** else ***REMOVED***
    memos = memos || ***REMOVED***actual: [], expected: []***REMOVED***;

    var actualIndex = memos.actual.indexOf(actual);
    if (actualIndex !== -1) ***REMOVED***
      if (actualIndex === memos.expected.indexOf(expected)) ***REMOVED***
        return true;
      ***REMOVED***
    ***REMOVED***

    memos.actual.push(actual);
    memos.expected.push(expected);

    return objEquiv(actual, expected, strict, memos);
  ***REMOVED***
***REMOVED***

function isArguments(object) ***REMOVED***
  return Object.prototype.toString.call(object) == '[object Arguments]';
***REMOVED***

function objEquiv(a, b, strict, actualVisitedObjects) ***REMOVED***
  if (a === null || a === undefined || b === null || b === undefined)
    return false;
  // if one is a primitive, the other must be same
  if (util.isPrimitive(a) || util.isPrimitive(b))
    return a === b;
  if (strict && Object.getPrototypeOf(a) !== Object.getPrototypeOf(b))
    return false;
  var aIsArgs = isArguments(a);
  var bIsArgs = isArguments(b);
  if ((aIsArgs && !bIsArgs) || (!aIsArgs && bIsArgs))
    return false;
  if (aIsArgs) ***REMOVED***
    a = pSlice.call(a);
    b = pSlice.call(b);
    return _deepEqual(a, b, strict);
  ***REMOVED***
  var ka = objectKeys(a);
  var kb = objectKeys(b);
  var key, i;
  // having the same number of owned properties (keys incorporates
  // hasOwnProperty)
  if (ka.length !== kb.length)
    return false;
  //the same set of keys (although not necessarily the same order),
  ka.sort();
  kb.sort();
  //~~~cheap key test
  for (i = ka.length - 1; i >= 0; i--) ***REMOVED***
    if (ka[i] !== kb[i])
      return false;
  ***REMOVED***
  //equivalent values for every corresponding key, and
  //~~~possibly expensive deep test
  for (i = ka.length - 1; i >= 0; i--) ***REMOVED***
    key = ka[i];
    if (!_deepEqual(a[key], b[key], strict, actualVisitedObjects))
      return false;
  ***REMOVED***
  return true;
***REMOVED***

// 8. The non-equivalence assertion tests for any deep inequality.
// assert.notDeepEqual(actual, expected, message_opt);

assert.notDeepEqual = function notDeepEqual(actual, expected, message) ***REMOVED***
  if (_deepEqual(actual, expected, false)) ***REMOVED***
    fail(actual, expected, message, 'notDeepEqual', assert.notDeepEqual);
  ***REMOVED***
***REMOVED***;

assert.notDeepStrictEqual = notDeepStrictEqual;
function notDeepStrictEqual(actual, expected, message) ***REMOVED***
  if (_deepEqual(actual, expected, true)) ***REMOVED***
    fail(actual, expected, message, 'notDeepStrictEqual', notDeepStrictEqual);
  ***REMOVED***
***REMOVED***


// 9. The strict equality assertion tests strict equality, as determined by ===.
// assert.strictEqual(actual, expected, message_opt);

assert.strictEqual = function strictEqual(actual, expected, message) ***REMOVED***
  if (actual !== expected) ***REMOVED***
    fail(actual, expected, message, '===', assert.strictEqual);
  ***REMOVED***
***REMOVED***;

// 10. The strict non-equality assertion tests for strict inequality, as
// determined by !==.  assert.notStrictEqual(actual, expected, message_opt);

assert.notStrictEqual = function notStrictEqual(actual, expected, message) ***REMOVED***
  if (actual === expected) ***REMOVED***
    fail(actual, expected, message, '!==', assert.notStrictEqual);
  ***REMOVED***
***REMOVED***;

function expectedException(actual, expected) ***REMOVED***
  if (!actual || !expected) ***REMOVED***
    return false;
  ***REMOVED***

  if (Object.prototype.toString.call(expected) == '[object RegExp]') ***REMOVED***
    return expected.test(actual);
  ***REMOVED***

  try ***REMOVED***
    if (actual instanceof expected) ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED*** catch (e) ***REMOVED***
    // Ignore.  The instanceof check doesn't work for arrow functions.
  ***REMOVED***

  if (Error.isPrototypeOf(expected)) ***REMOVED***
    return false;
  ***REMOVED***

  return expected.call(***REMOVED******REMOVED***, actual) === true;
***REMOVED***

function _tryBlock(block) ***REMOVED***
  var error;
  try ***REMOVED***
    block();
  ***REMOVED*** catch (e) ***REMOVED***
    error = e;
  ***REMOVED***
  return error;
***REMOVED***

function _throws(shouldThrow, block, expected, message) ***REMOVED***
  var actual;

  if (typeof block !== 'function') ***REMOVED***
    throw new TypeError('"block" argument must be a function');
  ***REMOVED***

  if (typeof expected === 'string') ***REMOVED***
    message = expected;
    expected = null;
  ***REMOVED***

  actual = _tryBlock(block);

  message = (expected && expected.name ? ' (' + expected.name + ').' : '.') +
            (message ? ' ' + message : '.');

  if (shouldThrow && !actual) ***REMOVED***
    fail(actual, expected, 'Missing expected exception' + message);
  ***REMOVED***

  var userProvidedMessage = typeof message === 'string';
  var isUnwantedException = !shouldThrow && util.isError(actual);
  var isUnexpectedException = !shouldThrow && actual && !expected;

  if ((isUnwantedException &&
      userProvidedMessage &&
      expectedException(actual, expected)) ||
      isUnexpectedException) ***REMOVED***
    fail(actual, expected, 'Got unwanted exception' + message);
  ***REMOVED***

  if ((shouldThrow && actual && expected &&
      !expectedException(actual, expected)) || (!shouldThrow && actual)) ***REMOVED***
    throw actual;
  ***REMOVED***
***REMOVED***

// 11. Expected to throw an error:
// assert.throws(block, Error_opt, message_opt);

assert.throws = function(block, /*optional*/error, /*optional*/message) ***REMOVED***
  _throws(true, block, error, message);
***REMOVED***;

// EXTENSION! This is annoying to write outside this module.
assert.doesNotThrow = function(block, /*optional*/error, /*optional*/message) ***REMOVED***
  _throws(false, block, error, message);
***REMOVED***;

assert.ifError = function(err) ***REMOVED*** if (err) throw err; ***REMOVED***;

var objectKeys = Object.keys || function (obj) ***REMOVED***
  var keys = [];
  for (var key in obj) ***REMOVED***
    if (hasOwn.call(obj, key)) keys.push(key);
  ***REMOVED***
  return keys;
***REMOVED***;
