// ajv-keywords/keywords/_formatLimit.js
'use strict';

var TIME = /^(\d\d):(\d\d):(\d\d)(\.\d+)?(z|[+-]\d\d:\d\d)?$/i;
var DATE_TIME_SEPARATOR = /t|\s/i;

var COMPARE_FORMATS = ***REMOVED***
  date: compareDate,
  time: compareTime,
  'date-time': compareDateTime
***REMOVED***;

module.exports = function (minMax) ***REMOVED***
  var keyword = 'format' + minMax;
  return function defFunc(ajv) ***REMOVED***
    defFunc.definition = ***REMOVED***
      type: 'string',
      inline: require('./dotjs/_formatLimit'),
      statements: true,
      errors: 'full',
      metaSchema: ***REMOVED***
        anyOf: [
          ***REMOVED*** type: 'string' ***REMOVED***,
          ***REMOVED***
            type: 'object',
            required: [ '$data' ],
            properties: ***REMOVED***
              $data: ***REMOVED***
                type: 'string',
                anyOf: [
                  ***REMOVED*** format: 'relative-json-pointer' ***REMOVED***,
                  ***REMOVED*** format: 'json-pointer' ***REMOVED***
                ]
              ***REMOVED***
            ***REMOVED***,
            additionalProperties: false
          ***REMOVED***
        ]
      ***REMOVED***
    ***REMOVED***;

    ajv.addKeyword(keyword, defFunc.definition);
    ajv.addKeyword('formatExclusive' + minMax);
    extendFormats(ajv);
    return ajv;
  ***REMOVED***;
***REMOVED***;


function extendFormats(ajv) ***REMOVED***
  var formats = ajv._formats;
  for (var name in COMPARE_FORMATS) ***REMOVED***
    var format = formats[name];
    // the last condition is needed if it's RegExp from another window
    if (typeof format != 'object' || format instanceof RegExp || !format.validate)
      format = formats[name] = ***REMOVED*** validate: format ***REMOVED***;
    if (!format.compare)
      format.compare = COMPARE_FORMATS[name];
  ***REMOVED***
***REMOVED***


function compareDate(d1, d2) ***REMOVED***
  if (!(d1 && d2)) return;
  if (d1 > d2) return 1;
  if (d1 < d2) return -1;
  if (d1 === d2) return 0;
***REMOVED***


function compareTime(t1, t2) ***REMOVED***
  if (!(t1 && t2)) return;
  t1 = t1.match(TIME);
  t2 = t2.match(TIME);
  if (!(t1 && t2)) return;
  t1 = t1[1] + t1[2] + t1[3] + (t1[4]||'');
  t2 = t2[1] + t2[2] + t2[3] + (t2[4]||'');
  if (t1 > t2) return 1;
  if (t1 < t2) return -1;
  if (t1 === t2) return 0;
***REMOVED***


function compareDateTime(dt1, dt2) ***REMOVED***
  if (!(dt1 && dt2)) return;
  dt1 = dt1.split(DATE_TIME_SEPARATOR);
  dt2 = dt2.split(DATE_TIME_SEPARATOR);
  var res = compareDate(dt1[0], dt2[0]);
  if (res === undefined) return;
  return res || compareTime(dt1[1], dt2[1]);
***REMOVED***
