// ajv-keywords/keywords/deepRequired.js
'use strict';

module.exports = function defFunc(ajv) ***REMOVED***
  defFunc.definition = ***REMOVED***
    type: 'object',
    inline: function (it, keyword, schema) ***REMOVED***
      var expr = '';
      for (var i=0; i<schema.length; i++) ***REMOVED***
        if (i) expr += ' && ';
        expr += '(' + getData(schema[i], it.dataLevel) + ' !== undefined)';
      ***REMOVED***
      return expr;
    ***REMOVED***,
    metaSchema: ***REMOVED***
      type: 'array',
      items: ***REMOVED***
        type: 'string',
        format: 'json-pointer'
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;

  ajv.addKeyword('deepRequired', defFunc.definition);
  return ajv;
***REMOVED***;


function getData(jsonPointer, lvl) ***REMOVED***
  var data = 'data' + (lvl || '');
  if (!jsonPointer) return data;

  var expr = data;
  var segments = jsonPointer.split('/');
  for (var i=1; i<segments.length; i++) ***REMOVED***
    var segment = segments[i];
    data += getProperty(unescapeJsonPointer(segment));
    expr += ' && ' + data;
  ***REMOVED***
  return expr;
***REMOVED***


var IDENTIFIER = /^[a-z$_][a-z$_0-9]*$/i;
var INTEGER = /^[0-9]+$/;
var SINGLE_QUOTE = /'|\\/g;
function getProperty(key) ***REMOVED***
  return INTEGER.test(key)
          ? '[' + key + ']'
          : IDENTIFIER.test(key)
            ? '.' + key
            : "['" + key.replace(SINGLE_QUOTE, '\\$&') + "']";
***REMOVED***


function unescapeJsonPointer(str) ***REMOVED***
  return str.replace(/~1/g, '/').replace(/~0/g, '~');
***REMOVED***
