// ajv-keywords/keywords/select.js
'use strict';

var util = require('./_util');

module.exports = function defFunc(ajv) ***REMOVED***
  if (!ajv._opts.$data) ***REMOVED***
    console.warn('keyword select requires $data option');
    return ajv;
  ***REMOVED***
  var metaSchemaRef = util.metaSchemaRef(ajv);
  var compiledCaseSchemas = [];

  defFunc.definition = ***REMOVED***
    validate: function v(schema, data, parentSchema) ***REMOVED***
      if (parentSchema.selectCases === undefined)
        throw new Error('keyword "selectCases" is absent');
      var compiled = getCompiledSchemas(parentSchema, false);
      var validate = compiled.cases[schema];
      if (validate === undefined) validate = compiled.default;
      if (typeof validate == 'boolean') return validate;
      var valid = validate(data);
      if (!valid) v.errors = validate.errors;
      return valid;
    ***REMOVED***,
    $data: true,
    metaSchema: ***REMOVED*** type: ['string', 'number', 'boolean', 'null'] ***REMOVED***
  ***REMOVED***;

  ajv.addKeyword('select', defFunc.definition);
  ajv.addKeyword('selectCases', ***REMOVED***
    compile: function (schemas, parentSchema) ***REMOVED***
      var compiled = getCompiledSchemas(parentSchema);
      for (var value in schemas)
        compiled.cases[value] = compileOrBoolean(schemas[value]);
      return function() ***REMOVED*** return true; ***REMOVED***;
    ***REMOVED***,
    valid: true,
    metaSchema: ***REMOVED***
      type: 'object',
      additionalProperties: metaSchemaRef
    ***REMOVED***
  ***REMOVED***);
  ajv.addKeyword('selectDefault', ***REMOVED***
    compile: function (schema, parentSchema) ***REMOVED***
      var compiled = getCompiledSchemas(parentSchema);
      compiled.default = compileOrBoolean(schema);
      return function() ***REMOVED*** return true; ***REMOVED***;
    ***REMOVED***,
    valid: true,
    metaSchema: metaSchemaRef
  ***REMOVED***);
  return ajv;


  function getCompiledSchemas(parentSchema, create) ***REMOVED***
    var compiled;
    compiledCaseSchemas.some(function (c) ***REMOVED***
      if (c.parentSchema === parentSchema) ***REMOVED***
        compiled = c;
        return true;
      ***REMOVED***
    ***REMOVED***);
    if (!compiled && create !== false) ***REMOVED***
      compiled = ***REMOVED***
        parentSchema: parentSchema,
        cases: ***REMOVED******REMOVED***,
        default: true
      ***REMOVED***;
      compiledCaseSchemas.push(compiled);
    ***REMOVED***
    return compiled;
  ***REMOVED***

  function compileOrBoolean(schema) ***REMOVED***
    return typeof schema == 'boolean'
            ? schema
            : ajv.compile(schema);
  ***REMOVED***
***REMOVED***;
