// ajv-keywords/keywords/switch.js
'use strict';

var util = require('./_util');

module.exports = function defFunc(ajv) ***REMOVED***
  if (ajv.RULES.keywords.switch && ajv.RULES.keywords.if) return;

  var metaSchemaRef = util.metaSchemaRef(ajv);

  defFunc.definition = ***REMOVED***
    inline: require('./dotjs/switch'),
    statements: true,
    errors: 'full',
    metaSchema: ***REMOVED***
      type: 'array',
      items: ***REMOVED***
        required: [ 'then' ],
        properties: ***REMOVED***
          'if': metaSchemaRef,
          'then': ***REMOVED***
            anyOf: [
              ***REMOVED*** type: 'boolean' ***REMOVED***,
              metaSchemaRef
            ]
          ***REMOVED***,
          'continue': ***REMOVED*** type: 'boolean' ***REMOVED***
        ***REMOVED***,
        additionalProperties: false,
        dependencies: ***REMOVED***
          'continue': [ 'if' ]
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;

  ajv.addKeyword('switch', defFunc.definition);
  return ajv;
***REMOVED***;
