// gtoken/lib/index.js
var gp12pem = require('google-p12-pem');
var request = require('request');
var mime = require('mime');
var jws = require('jws');
var fs = require('fs');

var GOOGLE_TOKEN_URL = 'https://accounts.google.com/o/oauth2/token';
var GOOGLE_REVOKE_TOKEN_URL = 'https://accounts.google.com/o/oauth2/revoke?token=';

/**
 * Create a GoogleToken.
 *
 * @param ***REMOVED***object***REMOVED***   options  Configuration object.
 */
function GoogleToken(options) ***REMOVED***
  if (!(this instanceof GoogleToken)) ***REMOVED***
    return new GoogleToken(options);
  ***REMOVED***
  this._configure(options);
***REMOVED***

GoogleToken.prototype._mime = mime;

/**
 * Returns whether the token has expired.
 *
 * @return ***REMOVED***Boolean***REMOVED*** true if the token has expired, false otherwise.
 */
GoogleToken.prototype.hasExpired = function() ***REMOVED***
  var now = (new Date()).getTime();
  if (this.token && this.expires_at) ***REMOVED***
    return now >= this.expires_at;
  ***REMOVED*** else ***REMOVED***
    return true;
  ***REMOVED***
***REMOVED***;

GoogleToken.prototype._gp12pem = gp12pem;

/**
 * Returns a cached token or retrieves a new one from Google.
 *
 * @param  ***REMOVED***Function***REMOVED*** callback The callback function.
 */
GoogleToken.prototype.getToken = function(callback) ***REMOVED***
  var self = this;

  if (!this.hasExpired()) ***REMOVED***
    return callback(null, this.token);
  ***REMOVED*** else ***REMOVED***
    if (!this.key && !this.keyFile) ***REMOVED***
      callback(new Error('No key or keyFile set.'));
      return;
    ***REMOVED*** else if (!this.key && this.keyFile) ***REMOVED***
      var mimeType = this._mime.lookup(this.keyFile);
      if (mimeType === 'application/json') ***REMOVED***
        // json file
        fs.readFile(this.keyFile, handleJSONKey);
      ***REMOVED*** else ***REMOVED***
        // Must be a .p12 file or .pem key
        if (!self.iss) ***REMOVED***
          var error = new Error('email is required.');
          error.code = 'MISSING_CREDENTIALS';
          callback(error);
          return;
        ***REMOVED***

        if (mimeType === 'application/x-pkcs12') ***REMOVED***
          // convert to .pem on the fly
          self._gp12pem(this.keyFile, handleKey);
        ***REMOVED*** else ***REMOVED***
          // assume .pem key otherwise
          fs.readFile(this.keyFile, handleKey);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      return this._requestToken(callback);
    ***REMOVED***
  ***REMOVED***

  function handleJSONKey(err, key) ***REMOVED***
    if (err) ***REMOVED***
      callback(err);
      return;
    ***REMOVED***
    try ***REMOVED***
      var body = JSON.parse(key);
      self.key = body.private_key;
      self.iss = body.client_email;
    ***REMOVED*** catch (e) ***REMOVED***
      callback(e);
      return;
    ***REMOVED***

    if (!self.key || !self.iss) ***REMOVED***
      var error = new Error('private_key and client_email are required.');
      error.code = 'MISSING_CREDENTIALS';
      callback(error);
      return;
    ***REMOVED***

    self._requestToken(callback);
  ***REMOVED***

  function handleKey(err, key) ***REMOVED***
    if (err) ***REMOVED***
      callback(err);
      return;
    ***REMOVED***
    self.key = key;
    self._requestToken(callback);
  ***REMOVED***
***REMOVED***;

/**
 * Revoke the token if one is set.
 *
 * @param  ***REMOVED***Function***REMOVED*** callback The callback function.
 */
GoogleToken.prototype.revokeToken = function(callback) ***REMOVED***
  var self = this;
  if (this.token) ***REMOVED***
    this._request(GOOGLE_REVOKE_TOKEN_URL + this.token, function(err, res) ***REMOVED***
      if (err) ***REMOVED***
        callback(err);
        return;
      ***REMOVED***
      self._configure(***REMOVED***
        email: self.iss,
        sub: self.sub,
        key: self.key,
        keyFile: self.keyFile,
        scope: self.scope
      ***REMOVED***);
      callback();
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    callback(new Error('No token to revoke.'));
  ***REMOVED***
***REMOVED***;

/**
 * Configure the GoogleToken for re-use.
 * @param  ***REMOVED***object***REMOVED*** options Configuration object.
 */
GoogleToken.prototype._configure = function(options) ***REMOVED***
  var self = this;
  options = options || ***REMOVED******REMOVED***;
  this.keyFile = options.keyFile;
  this.key = options.key;
  this._request = request;
  this.token = this.expires_at = this.raw_token = null;
  this.iss = options.email || options.iss;

  if (options.sub) ***REMOVED***
    this.sub = options.sub;
  ***REMOVED***

  if (typeof options.scope === 'object') ***REMOVED***
    this.scope = options.scope.join(' ');
  ***REMOVED*** else ***REMOVED***
    this.scope = options.scope;
  ***REMOVED***
***REMOVED***;

/**
 * Request the token from Google.
 *
 * @param  ***REMOVED***Function***REMOVED*** callback The callback function.
 */
GoogleToken.prototype._requestToken = function(callback) ***REMOVED***
  var self = this;
  var iat = Math.floor(new Date().getTime() / 1000);
  var payload = ***REMOVED***
    iss: this.iss,
    scope: this.scope,
    aud: GOOGLE_TOKEN_URL,
    exp: iat + 3600, // 3600 seconds = 1 hour
    iat: iat
  ***REMOVED***;

  if (this.sub) ***REMOVED***
    payload.sub = this.sub;
  ***REMOVED***

  var toSign = ***REMOVED***
    header: ***REMOVED***
      alg: 'RS256',
      typ: 'JWT'
    ***REMOVED***,
    payload: payload,
    secret: this.key
  ***REMOVED***;

  return this._signJWT(toSign, function(err, signedJWT) ***REMOVED***
    if (err) ***REMOVED***
      callback(err, null);
      return;
    ***REMOVED***

    return self._request(***REMOVED***
      method: 'post',
      url: GOOGLE_TOKEN_URL,
      form: ***REMOVED***
        grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
        assertion: signedJWT
      ***REMOVED***
    ***REMOVED***, function(err, res, body) ***REMOVED***
      try ***REMOVED***
        body = JSON.parse(body);
      ***REMOVED*** catch (e) ***REMOVED***
        body = ***REMOVED******REMOVED***;
      ***REMOVED***

      err = err || body.error && new Error(body.error +
          (body.error_description ? ': ' + body.error_description : ''));

      if (err) ***REMOVED***
        self.token = null;
        self.token_expires = null;
        callback(err, null);
        return;
      ***REMOVED***

      self.raw_token = body;
      self.token = body.access_token;
      self.expires_at = (iat + body.expires_in) * 1000;
      return callback(null, self.token);
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

/**
 * Sign the JWT object, returning any errors in the callback.
 *
 * @param  ***REMOVED***object***REMOVED***   opts     The configuration object.
 * @param  ***REMOVED***Function***REMOVED*** callback The callback function.
 */
GoogleToken.prototype._signJWT = function(opts, callback) ***REMOVED***
  try ***REMOVED***
    var signedJWT = jws.sign(opts);
    return callback(null, signedJWT);
  ***REMOVED*** catch (err) ***REMOVED***
    callback(err, null);
  ***REMOVED***
***REMOVED***;

module.exports = GoogleToken;
