// update-notifier/index.js
'use strict';
const spawn = require('child_process').spawn;
const path = require('path');
const format = require('util').format;
const importLazy = require('import-lazy')(require);

const configstore = importLazy('configstore');
const chalk = importLazy('chalk');
const semverDiff = importLazy('semver-diff');
const latestVersion = importLazy('latest-version');
const isNpm = importLazy('is-npm');
const isInstalledGlobally = importLazy('is-installed-globally');
const boxen = importLazy('boxen');
const xdgBasedir = importLazy('xdg-basedir');
const ONE_DAY = 1000 * 60 * 60 * 24;

class UpdateNotifier ***REMOVED***
	constructor(options) ***REMOVED***
		options = options || ***REMOVED******REMOVED***;
		this.options = options;
		options.pkg = options.pkg || ***REMOVED******REMOVED***;

		// Reduce pkg to the essential keys. with fallback to deprecated options
		// TODO: Remove deprecated options at some point far into the future
		options.pkg = ***REMOVED***
			name: options.pkg.name || options.packageName,
			version: options.pkg.version || options.packageVersion
		***REMOVED***;

		if (!options.pkg.name || !options.pkg.version) ***REMOVED***
			throw new Error('pkg.name and pkg.version required');
		***REMOVED***

		this.packageName = options.pkg.name;
		this.packageVersion = options.pkg.version;
		this.updateCheckInterval = typeof options.updateCheckInterval === 'number' ? options.updateCheckInterval : ONE_DAY;
		this.hasCallback = typeof options.callback === 'function';
		this.callback = options.callback || (() => ***REMOVED******REMOVED***);
		this.disabled = 'NO_UPDATE_NOTIFIER' in process.env ||
			process.argv.indexOf('--no-update-notifier') !== -1;

		if (!this.disabled && !this.hasCallback) ***REMOVED***
			try ***REMOVED***
				const ConfigStore = configstore();
				this.config = new ConfigStore(`update-notifier-$***REMOVED***this.packageName***REMOVED***`, ***REMOVED***
					optOut: false,
					// Init with the current time so the first check is only
					// after the set interval, so not to bother users right away
					lastUpdateCheck: Date.now()
				***REMOVED***);
			***REMOVED*** catch (err) ***REMOVED***
				// Expecting error code EACCES or EPERM
				const msg =
					chalk().yellow(format(' %s update check failed ', options.pkg.name)) +
					format('\n Try running with %s or get access ', chalk().cyan('sudo')) +
					'\n to the local update config store via \n' +
					chalk().cyan(format(' sudo chown -R $USER:$(id -gn $USER) %s ', xdgBasedir().config));

				process.on('exit', () => ***REMOVED***
					console.error('\n' + boxen()(msg, ***REMOVED***align: 'center'***REMOVED***));
				***REMOVED***);
			***REMOVED***
		***REMOVED***
	***REMOVED***
	check() ***REMOVED***
		if (this.hasCallback) ***REMOVED***
			this.checkNpm()
				.then(update => this.callback(null, update))
				.catch(err => this.callback(err));
			return;
		***REMOVED***

		if (
			!this.config ||
			this.config.get('optOut') ||
			this.disabled
		) ***REMOVED***
			return;
		***REMOVED***

		this.update = this.config.get('update');

		if (this.update) ***REMOVED***
			this.config.delete('update');
		***REMOVED***

		// Only check for updates on a set interval
		if (Date.now() - this.config.get('lastUpdateCheck') < this.updateCheckInterval) ***REMOVED***
			return;
		***REMOVED***

		// Spawn a detached process, passing the options as an environment property
		spawn(process.execPath, [path.join(__dirname, 'check.js'), JSON.stringify(this.options)], ***REMOVED***
			detached: true,
			stdio: 'ignore'
		***REMOVED***).unref();
	***REMOVED***
	checkNpm() ***REMOVED***
		return latestVersion()(this.packageName).then(latestVersion => ***REMOVED***
			return ***REMOVED***
				latest: latestVersion,
				current: this.packageVersion,
				type: semverDiff()(this.packageVersion, latestVersion) || 'latest',
				name: this.packageName
			***REMOVED***;
		***REMOVED***);
	***REMOVED***
	notify(opts) ***REMOVED***
		if (!process.stdout.isTTY || isNpm() || !this.update) ***REMOVED***
			return this;
		***REMOVED***

		opts = Object.assign(***REMOVED***isGlobal: isInstalledGlobally()***REMOVED***, opts);

		opts.message = opts.message || 'Update available ' + chalk().dim(this.update.current) + chalk().reset(' → ') +
			chalk().green(this.update.latest) + ' \nRun ' + chalk().cyan('npm i ' + (opts.isGlobal ? '-g ' : '') + this.packageName) + ' to update';

		opts.boxenOpts = opts.boxenOpts || ***REMOVED***
			padding: 1,
			margin: 1,
			align: 'center',
			borderColor: 'yellow',
			borderStyle: 'round'
		***REMOVED***;

		const message = '\n' + boxen()(opts.message, opts.boxenOpts);

		if (opts.defer === false) ***REMOVED***
			console.error(message);
		***REMOVED*** else ***REMOVED***
			process.on('exit', () => ***REMOVED***
				console.error(message);
			***REMOVED***);

			process.on('SIGINT', () => ***REMOVED***
				console.error('');
				process.exit();
			***REMOVED***);
		***REMOVED***

		return this;
	***REMOVED***
***REMOVED***

module.exports = options => ***REMOVED***
	const updateNotifier = new UpdateNotifier(options);
	updateNotifier.check();
	return updateNotifier;
***REMOVED***;

module.exports.UpdateNotifier = UpdateNotifier;
