// timers-browserify/main.js
var apply = Function.prototype.apply;

// DOM APIs, for completeness

exports.setTimeout = function() ***REMOVED***
  return new Timeout(apply.call(setTimeout, window, arguments), clearTimeout);
***REMOVED***;
exports.setInterval = function() ***REMOVED***
  return new Timeout(apply.call(setInterval, window, arguments), clearInterval);
***REMOVED***;
exports.clearTimeout =
exports.clearInterval = function(timeout) ***REMOVED***
  if (timeout) ***REMOVED***
    timeout.close();
  ***REMOVED***
***REMOVED***;

function Timeout(id, clearFn) ***REMOVED***
  this._id = id;
  this._clearFn = clearFn;
***REMOVED***
Timeout.prototype.unref = Timeout.prototype.ref = function() ***REMOVED******REMOVED***;
Timeout.prototype.close = function() ***REMOVED***
  this._clearFn.call(window, this._id);
***REMOVED***;

// Does not start the time, just sets up the members needed.
exports.enroll = function(item, msecs) ***REMOVED***
  clearTimeout(item._idleTimeoutId);
  item._idleTimeout = msecs;
***REMOVED***;

exports.unenroll = function(item) ***REMOVED***
  clearTimeout(item._idleTimeoutId);
  item._idleTimeout = -1;
***REMOVED***;

exports._unrefActive = exports.active = function(item) ***REMOVED***
  clearTimeout(item._idleTimeoutId);

  var msecs = item._idleTimeout;
  if (msecs >= 0) ***REMOVED***
    item._idleTimeoutId = setTimeout(function onTimeout() ***REMOVED***
      if (item._onTimeout)
        item._onTimeout();
    ***REMOVED***, msecs);
  ***REMOVED***
***REMOVED***;

// setimmediate attaches itself to the global object
require("setimmediate");
// On some exotic environments, it's not clear which object `setimmeidate` was
// able to install onto.  Search each possibility in the same order as the
// `setimmediate` library.
exports.setImmediate = (typeof self !== "undefined" && self.setImmediate) ||
                       (typeof global !== "undefined" && global.setImmediate) ||
                       (this && this.setImmediate);
exports.clearImmediate = (typeof self !== "undefined" && self.clearImmediate) ||
                         (typeof global !== "undefined" && global.clearImmediate) ||
                         (this && this.clearImmediate);
