// extglob/index.js
'use strict';

/**
 * Module dependencies
 */

var extend = require('extend-shallow');
var unique = require('array-unique');
var toRegex = require('to-regex');

/**
 * Local dependencies
 */

var compilers = require('./lib/compilers');
var parsers = require('./lib/parsers');
var Extglob = require('./lib/extglob');
var utils = require('./lib/utils');
var MAX_LENGTH = 1024 * 64;

/**
 * Convert the given `extglob` pattern into a regex-compatible string. Returns
 * an object with the compiled result and the parsed AST.
 *
 * ```js
 * var extglob = require('extglob');
 * console.log(extglob('*.!(*a)'));
 * //=> '(?!\\.)[^/]*?\\.(?!(?!\\.)[^/]*?a\\b).*?'
 * ```
 * @param ***REMOVED***String***REMOVED*** `pattern`
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @return ***REMOVED***String***REMOVED***
 * @api public
 */

function extglob(pattern, options) ***REMOVED***
  return extglob.create(pattern, options).output;
***REMOVED***

/**
 * Takes an array of strings and an extglob pattern and returns a new
 * array that contains only the strings that match the pattern.
 *
 * ```js
 * var extglob = require('extglob');
 * console.log(extglob.match(['a.a', 'a.b', 'a.c'], '*.!(*a)'));
 * //=> ['a.b', 'a.c']
 * ```
 * @param ***REMOVED***Array***REMOVED*** `list` Array of strings to match
 * @param ***REMOVED***String***REMOVED*** `pattern` Extglob pattern
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @return ***REMOVED***Array***REMOVED*** Returns an array of matches
 * @api public
 */

extglob.match = function(list, pattern, options) ***REMOVED***
  if (typeof pattern !== 'string') ***REMOVED***
    throw new TypeError('expected pattern to be a string');
  ***REMOVED***

  list = utils.arrayify(list);
  var isMatch = extglob.matcher(pattern, options);
  var len = list.length;
  var idx = -1;
  var matches = [];

  while (++idx < len) ***REMOVED***
    var ele = list[idx];

    if (isMatch(ele)) ***REMOVED***
      matches.push(ele);
    ***REMOVED***
  ***REMOVED***

  // if no options were passed, uniquify results and return
  if (typeof options === 'undefined') ***REMOVED***
    return unique(matches);
  ***REMOVED***

  if (matches.length === 0) ***REMOVED***
    if (options.failglob === true) ***REMOVED***
      throw new Error('no matches found for "' + pattern + '"');
    ***REMOVED***
    if (options.nonull === true || options.nullglob === true) ***REMOVED***
      return [pattern.split('\\').join('')];
    ***REMOVED***
  ***REMOVED***

  return options.nodupes !== false ? unique(matches) : matches;
***REMOVED***;

/**
 * Returns true if the specified `string` matches the given
 * extglob `pattern`.
 *
 * ```js
 * var extglob = require('extglob');
 *
 * console.log(extglob.isMatch('a.a', '*.!(*a)'));
 * //=> false
 * console.log(extglob.isMatch('a.b', '*.!(*a)'));
 * //=> true
 * ```
 * @param ***REMOVED***String***REMOVED*** `string` String to match
 * @param ***REMOVED***String***REMOVED*** `pattern` Extglob pattern
 * @param ***REMOVED***String***REMOVED*** `options`
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

extglob.isMatch = function(str, pattern, options) ***REMOVED***
  if (typeof pattern !== 'string') ***REMOVED***
    throw new TypeError('expected pattern to be a string');
  ***REMOVED***

  if (typeof str !== 'string') ***REMOVED***
    throw new TypeError('expected a string');
  ***REMOVED***

  if (pattern === str) ***REMOVED***
    return true;
  ***REMOVED***

  if (pattern === '' || pattern === ' ' || pattern === '.') ***REMOVED***
    return pattern === str;
  ***REMOVED***

  var isMatch = utils.memoize('isMatch', pattern, options, extglob.matcher);
  return isMatch(str);
***REMOVED***;

/**
 * Returns true if the given `string` contains the given pattern. Similar to `.isMatch` but
 * the pattern can match any part of the string.
 *
 * ```js
 * var extglob = require('extglob');
 * console.log(extglob.contains('aa/bb/cc', '*b'));
 * //=> true
 * console.log(extglob.contains('aa/bb/cc', '*d'));
 * //=> false
 * ```
 * @param ***REMOVED***String***REMOVED*** `str` The string to match.
 * @param ***REMOVED***String***REMOVED*** `pattern` Glob pattern to use for matching.
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @return ***REMOVED***Boolean***REMOVED*** Returns true if the patter matches any part of `str`.
 * @api public
 */

extglob.contains = function(str, pattern, options) ***REMOVED***
  if (typeof str !== 'string') ***REMOVED***
    throw new TypeError('expected a string');
  ***REMOVED***

  if (pattern === '' || pattern === ' ' || pattern === '.') ***REMOVED***
    return pattern === str;
  ***REMOVED***

  var opts = extend(***REMOVED******REMOVED***, options, ***REMOVED***contains: true***REMOVED***);
  opts.strictClose = false;
  opts.strictOpen = false;
  return extglob.isMatch(str, pattern, opts);
***REMOVED***;

/**
 * Takes an extglob pattern and returns a matcher function. The returned
 * function takes the string to match as its only argument.
 *
 * ```js
 * var extglob = require('extglob');
 * var isMatch = extglob.matcher('*.!(*a)');
 *
 * console.log(isMatch('a.a'));
 * //=> false
 * console.log(isMatch('a.b'));
 * //=> true
 * ```
 * @param ***REMOVED***String***REMOVED*** `pattern` Extglob pattern
 * @param ***REMOVED***String***REMOVED*** `options`
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

extglob.matcher = function(pattern, options) ***REMOVED***
  if (typeof pattern !== 'string') ***REMOVED***
    throw new TypeError('expected pattern to be a string');
  ***REMOVED***

  function matcher() ***REMOVED***
    var re = extglob.makeRe(pattern, options);
    return function(str) ***REMOVED***
      return re.test(str);
    ***REMOVED***;
  ***REMOVED***

  return utils.memoize('matcher', pattern, options, matcher);
***REMOVED***;

/**
 * Convert the given `extglob` pattern into a regex-compatible string. Returns
 * an object with the compiled result and the parsed AST.
 *
 * ```js
 * var extglob = require('extglob');
 * console.log(extglob.create('*.!(*a)').output);
 * //=> '(?!\\.)[^/]*?\\.(?!(?!\\.)[^/]*?a\\b).*?'
 * ```
 * @param ***REMOVED***String***REMOVED*** `str`
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @return ***REMOVED***String***REMOVED***
 * @api public
 */

extglob.create = function(pattern, options) ***REMOVED***
  if (typeof pattern !== 'string') ***REMOVED***
    throw new TypeError('expected pattern to be a string');
  ***REMOVED***

  function create() ***REMOVED***
    var ext = new Extglob(options);
    var ast = ext.parse(pattern, options);
    return ext.compile(ast, options);
  ***REMOVED***

  return utils.memoize('create', pattern, options, create);
***REMOVED***;

/**
 * Returns an array of matches captured by `pattern` in `string`, or `null`
 * if the pattern did not match.
 *
 * ```js
 * var extglob = require('extglob');
 * extglob.capture(pattern, string[, options]);
 *
 * console.log(extglob.capture('test/*.js', 'test/foo.js'));
 * //=> ['foo']
 * console.log(extglob.capture('test/*.js', 'foo/bar.css'));
 * //=> null
 * ```
 * @param ***REMOVED***String***REMOVED*** `pattern` Glob pattern to use for matching.
 * @param ***REMOVED***String***REMOVED*** `string` String to match
 * @param ***REMOVED***Object***REMOVED*** `options` See available [options](#options) for changing how matches are performed
 * @return ***REMOVED***Boolean***REMOVED*** Returns an array of captures if the string matches the glob pattern, otherwise `null`.
 * @api public
 */

extglob.capture = function(pattern, str, options) ***REMOVED***
  var re = extglob.makeRe(pattern, extend(***REMOVED***capture: true***REMOVED***, options));

  function match() ***REMOVED***
    return function(string) ***REMOVED***
      var match = re.exec(string);
      if (!match) ***REMOVED***
        return null;
      ***REMOVED***

      return match.slice(1);
    ***REMOVED***;
  ***REMOVED***

  var capture = utils.memoize('capture', pattern, options, match);
  return capture(str);
***REMOVED***;

/**
 * Create a regular expression from the given `pattern` and `options`.
 *
 * ```js
 * var extglob = require('extglob');
 * var re = extglob.makeRe('*.!(*a)');
 * console.log(re);
 * //=> /^[^\/]*?\.(?![^\/]*?a)[^\/]*?$/
 * ```
 * @param ***REMOVED***String***REMOVED*** `pattern` The pattern to convert to regex.
 * @param ***REMOVED***Object***REMOVED*** `options`
 * @return ***REMOVED***RegExp***REMOVED***
 * @api public
 */

extglob.makeRe = function(pattern, options) ***REMOVED***
  if (pattern instanceof RegExp) ***REMOVED***
    return pattern;
  ***REMOVED***

  if (typeof pattern !== 'string') ***REMOVED***
    throw new TypeError('expected pattern to be a string');
  ***REMOVED***

  if (pattern.length > MAX_LENGTH) ***REMOVED***
    throw new Error('expected pattern to be less than ' + MAX_LENGTH + ' characters');
  ***REMOVED***

  function makeRe() ***REMOVED***
    var opts = extend(***REMOVED***strictErrors: false***REMOVED***, options);
    if (opts.strictErrors === true) opts.strict = true;
    var res = extglob.create(pattern, opts);
    return toRegex(res.output, opts);
  ***REMOVED***

  var regex = utils.memoize('makeRe', pattern, options, makeRe);
  if (regex.source.length > MAX_LENGTH) ***REMOVED***
    throw new SyntaxError('potentially malicious regex detected');
  ***REMOVED***

  return regex;
***REMOVED***;

/**
 * Cache
 */

extglob.cache = utils.cache;
extglob.clearCache = function() ***REMOVED***
  extglob.cache.__data__ = ***REMOVED******REMOVED***;
***REMOVED***;

/**
 * Expose `Extglob` constructor, parsers and compilers
 */

extglob.Extglob = Extglob;
extglob.compilers = compilers;
extglob.parsers = parsers;

/**
 * Expose `extglob`
 * @type ***REMOVED***Function***REMOVED***
 */

module.exports = extglob;
