// @firebase/firestore/dist/cjs/src/util/sorted_set.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var sorted_map_1 = require("./sorted_map");
/**
 * SortedSet is an immutable (copy-on-write) collection that holds elements
 * in order specified by the provided comparator.
 *
 * NOTE: if provided comparator returns 0 for two elements, we consider them to
 * be equal!
 */
var SortedSet = /** @class */ (function () ***REMOVED***
    function SortedSet(comparator) ***REMOVED***
        this.comparator = comparator;
        this.data = new sorted_map_1.SortedMap(this.comparator);
    ***REMOVED***
    /**
     * Creates a SortedSet from the keys of the map.
     * This is currently implemented as an O(n) copy.
     */
    SortedSet.fromMapKeys = function (map) ***REMOVED***
        var keys = new SortedSet(map.comparator);
        map.forEach(function (key) ***REMOVED***
            keys = keys.add(key);
        ***REMOVED***);
        return keys;
    ***REMOVED***;
    SortedSet.prototype.has = function (elem) ***REMOVED***
        return this.data.get(elem) !== null;
    ***REMOVED***;
    SortedSet.prototype.first = function () ***REMOVED***
        return this.data.minKey();
    ***REMOVED***;
    SortedSet.prototype.last = function () ***REMOVED***
        return this.data.maxKey();
    ***REMOVED***;
    Object.defineProperty(SortedSet.prototype, "size", ***REMOVED***
        get: function () ***REMOVED***
            return this.data.size;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    SortedSet.prototype.indexOf = function (elem) ***REMOVED***
        return this.data.indexOf(elem);
    ***REMOVED***;
    /** Iterates elements in order defined by "comparator" */
    SortedSet.prototype.forEach = function (cb) ***REMOVED***
        this.data.inorderTraversal(function (k, v) ***REMOVED***
            cb(k);
            return false;
        ***REMOVED***);
    ***REMOVED***;
    /** Iterates over `elem`s such that: range[0] <= elem < range[1]. */
    SortedSet.prototype.forEachInRange = function (range, cb) ***REMOVED***
        var iter = this.data.getIteratorFrom(range[0]);
        while (iter.hasNext()) ***REMOVED***
            var elem = iter.getNext();
            if (this.comparator(elem.key, range[1]) >= 0)
                return;
            cb(elem.key);
        ***REMOVED***
    ***REMOVED***;
    /**
     * Iterates over `elem`s such that: start <= elem until false is returned.
     */
    SortedSet.prototype.forEachWhile = function (cb, start) ***REMOVED***
        var iter;
        if (start !== undefined) ***REMOVED***
            iter = this.data.getIteratorFrom(start);
        ***REMOVED***
        else ***REMOVED***
            iter = this.data.getIterator();
        ***REMOVED***
        while (iter.hasNext()) ***REMOVED***
            var elem = iter.getNext();
            var result = cb(elem.key);
            if (!result)
                return;
        ***REMOVED***
    ***REMOVED***;
    /** Finds the least element greater than or equal to `elem`. */
    SortedSet.prototype.firstAfterOrEqual = function (elem) ***REMOVED***
        var iter = this.data.getIteratorFrom(elem);
        return iter.hasNext() ? iter.getNext().key : null;
    ***REMOVED***;
    /** Inserts or updates an element */
    SortedSet.prototype.add = function (elem) ***REMOVED***
        return this.copy(this.data.remove(elem).insert(elem, true));
    ***REMOVED***;
    /** Deletes an element */
    SortedSet.prototype.delete = function (elem) ***REMOVED***
        if (!this.has(elem))
            return this;
        return this.copy(this.data.remove(elem));
    ***REMOVED***;
    SortedSet.prototype.isEmpty = function () ***REMOVED***
        return this.data.isEmpty();
    ***REMOVED***;
    SortedSet.prototype.unionWith = function (other) ***REMOVED***
        var result = this;
        other.forEach(function (elem) ***REMOVED***
            result = result.add(elem);
        ***REMOVED***);
        return result;
    ***REMOVED***;
    SortedSet.prototype.isEqual = function (other) ***REMOVED***
        if (!(other instanceof SortedSet))
            return false;
        if (this.size !== other.size)
            return false;
        var thisIt = this.data.getIterator();
        var otherIt = other.data.getIterator();
        while (thisIt.hasNext()) ***REMOVED***
            var thisElem = thisIt.getNext().key;
            var otherElem = otherIt.getNext().key;
            if (this.comparator(thisElem, otherElem) !== 0)
                return false;
        ***REMOVED***
        return true;
    ***REMOVED***;
    SortedSet.prototype.toString = function () ***REMOVED***
        var result = [];
        this.forEach(function (elem) ***REMOVED*** return result.push(elem); ***REMOVED***);
        return 'SortedSet(' + result.toString() + ')';
    ***REMOVED***;
    SortedSet.prototype.copy = function (data) ***REMOVED***
        var result = new SortedSet(this.comparator);
        result.data = data;
        return result;
    ***REMOVED***;
    return SortedSet;
***REMOVED***());
exports.SortedSet = SortedSet;
