// colour/colour.js
/*
 Copyright (c) 2010 Marak Squires, Alexis Sellier (cloudhead)
 Copyright (c) 2013 Daniel Wirtz <dcode@dcode.io>

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */
/**
 * @license colour.js (c) 2013 Daniel Wirtz <dcode@dcode.io>
 * Released under the MIT-License
 * see: https://github.com/dcodeIO/colour.js for details
 */
(function (global) ***REMOVED*** // #22
    'use strict';
    
    /**
     * colour namespace.
     * @namespace
     */
    var colour = ***REMOVED******REMOVED***;

    /**
     * Colour mode. To generate html in node, use colour.setTheme("html") instead.
     * @type ***REMOVED***string***REMOVED*** May be "console", "browser" or "none".
     * @expose
     */
    colour.mode = "console";

    /**
     * Whether running in a headless environment like node.js or not.
     * @type ***REMOVED***boolean***REMOVED***
     * @expose
     */
    colour.headless = typeof global['window'] === 'undefined';

    /**
     * All themes we know about. Contains only the "default" theme by default.
     * @type ***REMOVED***!Object.<string,!Object>***REMOVED***
     * @expose
     */
    colour.themes = ***REMOVED******REMOVED***;

    /**
     * Terminal colours.
     * @type ***REMOVED***!Object***REMOVED*** Hello Closure
     */
    var consoleStyles = ***REMOVED***
        'bold': ['\x1B[1m', '\x1B[22m'],
        'italic': ['\x1B[3m', '\x1B[23m'],
        'underline': ['\x1B[4m', '\x1B[24m'],
        'inverse': ['\x1B[7m', '\x1B[27m'],
        'strikethrough': ['\x1B[9m', '\x1B[29m'],
        'white': ['\x1B[37m', '\x1B[39m'],
        'gray': ['\x1B[90m', '\x1B[39m'],
        'grey': ['\x1B[90m', '\x1B[39m'],
        'black': ['\x1B[30m', '\x1B[39m'],
        'blue': ['\x1B[34m', '\x1B[39m'],
        'cyan': ['\x1B[36m', '\x1B[39m'],
        'green': ['\x1B[32m', '\x1B[39m'],
        'magenta': ['\x1B[35m', '\x1B[39m'],
        'red': ['\x1B[31m', '\x1B[39m'],
        'yellow': ['\x1B[33m', '\x1B[39m']
    ***REMOVED***;

    /**
     * HTML.
     * @type ***REMOVED***!Object.<string,Array.<string>>***REMOVED***
     */
    var browserStyles = ***REMOVED***
        'bold' : ['<b>', '</b>'],
        'italic' : ['<i>', '</i>'],
        'underline' : ['<u>', '</u>'],
        'inverse' : ['<span style="background-color:black;color:white;">', '</span>'],
        'strikethrough' : ['<del>', '</del>'],
        'white' : ['<span style="color:white;">', '</span>'],
        'gray' : ['<span style="color:gray;">', '</span>'], // #46
        'grey' : ['<span style="color:grey;">', '</span>'], // Let the user decide
        'black' : ['<span style="color:black;">', '</span>'],
        'blue' : ['<span style="color:blue;">', '</span>'],
        'cyan' : ['<span style="color:cyan;">', '</span>'],
        'green' : ['<span style="color:green;">', '</span>'],
        'magenta' : ['<span style="color:magenta;">', '</span>'],
        'red' : ['<span style="color:red;">', '</span>'],
        'yellow' : ['<span style="color:yellow;">', '</span>']
    ***REMOVED***;

    /**
     * CSS.
     * @type ***REMOVED***!Object.<string,Array.<string>>***REMOVED***
     */
    var cssStyles = ***REMOVED*** // #39
        'bold' : ['<span class="ansi-escape ansi-escape-bold">', '</span>'],
        'italic' : ['<span class="ansi-escape ansi-escape-italic">', '</span>'],
        'underline' : ['<span class="ansi-escape ansi-escape-underline">', '</span>'],
        'inverse' : ['<span class="ansi-escape ansi-escape-inverse">', '</span>'],
        'strikethrough' : ['<span class="ansi-escape ansi-escape-strikethrough">', '</span>'],
        'white' : ['<span class="ansi-escape ansi-escape-white">', '</span>'],
        'gray' : ['<span class="ansi-escape ansi-escape-gray">', '</span>'],
        'grey' : ['<span class="ansi-escape ansi-escape-grey">', '</span>'],
        'black' : ['<span class="ansi-escape ansi-escape-black">', '</span>'],
        'blue' : ['<span class="ansi-escape ansi-escape-blue">', '</span>'],
        'cyan' : ['<span class="ansi-escape ansi-escape-cyan">', '</span>'],
        'green' : ['<span class="ansi-escape ansi-escape-green">', '</span>'],
        'magenta' : ['<span class="ansi-escape ansi-escape-magenta">', '</span>'],
        'red' : ['<span class="ansi-escape ansi-escape-red">', '</span>'],
        'yellow' : ['<span class="ansi-escape ansi-escape-yellow">', '</span>']
    ***REMOVED***;

    /**
     * Remember all getters that we defined.
     * @type ***REMOVED***!Object***REMOVED***
     */
    var definedGetters = ***REMOVED******REMOVED***;

    /**
     * Prototypes the string object to have additional properties that wraps the current string in colours when accessed.
     * @param ***REMOVED***string***REMOVED*** col Colour / property name
     * @param ***REMOVED***function(string):string***REMOVED*** func Wrapper function
     * @private
     */
    function addProperty(col, func) ***REMOVED***
        // Exposed on top of the namespace
        colour[col] = function(str) ***REMOVED***
            return func.apply(str);
        ***REMOVED***;
        // And on top of all strings
        try ***REMOVED***
            String.prototype.__defineGetter__(col, func);
            definedGetters[col] = func;
        ***REMOVED*** catch (e) ***REMOVED******REMOVED*** // #25
    ***REMOVED***
    
    /**
     * Whether colour are currently installed on the global scope.
     * @type ***REMOVED***boolean***REMOVED***
     * @private
     **/
    var installed = true;

    /**
     * Uninstalls colour from the global scope.
     * @returns ***REMOVED***boolean***REMOVED*** true if successfully uninstalled, false if already uninstalled
     * @expose
     */
    colour.uninstall = function() ***REMOVED*** // #41
        if (installed) ***REMOVED***
            Object.keys(definedGetters).forEach(function(color) ***REMOVED***
                try ***REMOVED***
                    String.prototype.__defineGetter__(color, null);
                ***REMOVED*** catch (e) ***REMOVED***
                    delete String.prototype[color];
                ***REMOVED***
            ***REMOVED***);
            installed = false;
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***;

    /**
     * Reinstalls colour on the global scope.
     * @returns ***REMOVED***boolean***REMOVED*** true if successfully reinstalled, false if already installed
     * @expose
     */
    colour.install = function() ***REMOVED***
        if (!installed) ***REMOVED***
            Object.keys(definedGetters).forEach(function(color) ***REMOVED***
                String.prototype.__defineGetter__(color, definedGetters[color]);
            ***REMOVED***);
            installed = true;
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***;

    /**
     * Applies a style to a string.
     * @param ***REMOVED***string***REMOVED*** str String to stylize
     * @param ***REMOVED***string***REMOVED*** style Style to apply
     * @returns ***REMOVED***string***REMOVED***
     * @private
     */
    function stylize(str, style) ***REMOVED***
        if (colour.mode == 'console') ***REMOVED***
            return consoleStyles[style][0] + str + consoleStyles[style][1];
        ***REMOVED*** else if (colour.mode == 'browser') ***REMOVED***
            return browserStyles[style][0] + str + browserStyles[style][1];
        ***REMOVED*** else if (colour.mode == 'browser-css') ***REMOVED***
            return cssStyles[style][0] + str + browserStyles[style][1];
        ***REMOVED***
        return str+'';
    ***REMOVED***

    /**
     * Rainbow colours.
     * @type ***REMOVED***!Array.<string>***REMOVED***
     * @const
     * @private
     */
    var rainbowColours = ['red', 'yellow', 'green', 'blue', 'magenta'];

    /**
     * String properties that should never be overwritten.
     * @type ***REMOVED***!Array.<string>***REMOVED***
     * @const
     */
    var prototypeBlacklist = [
        '__defineGetter__', '__defineSetter__', '__lookupGetter__', '__lookupSetter__', 'charAt', 'constructor',
        'hasOwnProperty', 'isPrototypeOf', 'propertyIsEnumerable', 'toLocaleString', 'toString', 'valueOf', 'charCodeAt',
        'indexOf', 'lastIndexof', 'length', 'localeCompare', 'match', 'replace', 'search', 'slice', 'split', 'substring',
        'toLocaleLowerCase', 'toLocaleUpperCase', 'toLowerCase', 'toUpperCase', 'trim', 'trimLeft', 'trimRight'
    ];

    /**
     * Applies a theme.
     * @param ***REMOVED***!Object***REMOVED*** theme Theme to apply
     */
    function applyTheme(theme) ***REMOVED***
        Object.keys(theme).forEach(function(prop) ***REMOVED***
            if (prototypeBlacklist.indexOf(prop) >= 0) ***REMOVED***
                return;
            ***REMOVED***
            if (typeof theme[prop] == 'string') ***REMOVED***
                // Multiple colours white-space seperated #45, e.g. "red bold", #18
                theme[prop] = theme[prop].split(' ');
            ***REMOVED***
            addProperty(prop, function () ***REMOVED***
                var ret = this;
                for (var t=0; t<theme[prop].length; t++) ***REMOVED***
                    ret = colour[theme[prop][t]](ret);
                ***REMOVED***
                return ret;
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***

    /**
     * Sets another theme.
     * @param ***REMOVED***string|!Object***REMOVED*** theme Theme name or 
     * @returns ***REMOVED***!Object|!Error|undefined***REMOVED***
     * @expose
     */
    colour.setTheme = function(theme) ***REMOVED***
        if (typeof theme === 'string') ***REMOVED***
            if (typeof colour.themes[theme] != 'undefined') ***REMOVED***
                applyTheme(colour.themes[theme]);
                return colour.themes[theme];
            ***REMOVED***
            /* else */ try /* to load it */ ***REMOVED***
                colour.themes[theme] = require(theme);
                applyTheme(colour.themes[theme]);
                return colour.themes[theme];
            ***REMOVED*** catch (err) ***REMOVED***
                return err;
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            applyTheme(theme);
        ***REMOVED***
    ***REMOVED***;

    /**
     * Extends a mapper with the current index inside the string as a second argument.
     * @param ***REMOVED***function(string, number):string***REMOVED*** map Sequencing function
     * @returns ***REMOVED***function(string):string***REMOVED*** Wrapped sequencer
     * @private
     */
    function sequencer(map) ***REMOVED***
        return function () ***REMOVED***
            if (this == undefined) return "";
            var i=0;
            return String.prototype.split.apply(this, [""]).map(map).join("");
        ***REMOVED***;
    ***REMOVED***

    /**
     * Adds a sequencer that generates output depending on the index inside a string.
     * @param ***REMOVED***string***REMOVED*** name Sequencer name
     * @param ***REMOVED***function(string, number):string***REMOVED*** map Mapping function called for every character as the first and the
     *  current index of that character as the second argument.
     * @expose
     */
    colour.addSequencer = function (name, map) ***REMOVED***
        addProperty(name, sequencer(map));
    ***REMOVED***;
    
    // Apply defaults
    Object.keys(consoleStyles).forEach(
        function (style) ***REMOVED***
            addProperty(style, function () ***REMOVED***
                return stylize(this, style);
            ***REMOVED***);
        ***REMOVED***
    );

    colour.addSequencer('rainbow', function(letter, i) ***REMOVED***
        return letter === " " ? letter : stylize(letter, rainbowColours[i++ % rainbowColours.length]);
    ***REMOVED***);
    colour.addSequencer('zebra', sequencer(function (letter, i) ***REMOVED***
        return i % 2 === 0 ? letter : letter.inverse;
    ***REMOVED***));

    function strip() ***REMOVED***
        return this.replace(/\x1B\[\d+m/g, '')
                   .replace(/<\/?(?:span|u|i|u|del)\b[^>]*>/g, '');
    ***REMOVED***

    addProperty('strip', strip);
    addProperty('stripColors', strip); // Compatibility
    
    if (typeof module !== 'undefined' && module['exports']) ***REMOVED***
        module.exports = colour;
    ***REMOVED*** else if (typeof define !== 'undefined' && define.amd) ***REMOVED***
        define("colour", function() ***REMOVED*** return colour; ***REMOVED***);
        define("colors", function() ***REMOVED*** return colour; ***REMOVED***);
    ***REMOVED*** else ***REMOVED***
        colour.mode = 'browser';
        global['colour'] = global['colors'] = colour;
    ***REMOVED***

***REMOVED***)(this);
