// parse-glob/index.js
/*!
 * parse-glob <https://github.com/jonschlinkert/parse-glob>
 *
 * Copyright (c) 2015, Jon Schlinkert.
 * Licensed under the MIT License.
 */

'use strict';

var isGlob = require('is-glob');
var findBase = require('glob-base');
var extglob = require('is-extglob');
var dotfile = require('is-dotfile');

/**
 * Expose `cache`
 */

var cache = module.exports.cache = ***REMOVED******REMOVED***;

/**
 * Parse a glob pattern into tokens.
 *
 * When no paths or '**' are in the glob, we use a
 * different strategy for parsing the filename, since
 * file names can contain braces and other difficult
 * patterns. such as:
 *
 *  - `*.***REMOVED***a,b***REMOVED***`
 *  - `(**|*.js)`
 */

module.exports = function parseGlob(glob) ***REMOVED***
  if (cache.hasOwnProperty(glob)) ***REMOVED***
    return cache[glob];
  ***REMOVED***

  var tok = ***REMOVED******REMOVED***;
  tok.orig = glob;
  tok.is = ***REMOVED******REMOVED***;

  // unescape dots and slashes in braces/brackets
  glob = escape(glob);

  var parsed = findBase(glob);
  tok.is.glob = parsed.isGlob;

  tok.glob = parsed.glob;
  tok.base = parsed.base;
  var segs = /([^\/]*)$/.exec(glob);

  tok.path = ***REMOVED******REMOVED***;
  tok.path.dirname = '';
  tok.path.basename = segs[1] || '';
  tok.path.dirname = glob.split(tok.path.basename).join('') || '';
  var basename = (tok.path.basename || '').split('.') || '';
  tok.path.filename = basename[0] || '';
  tok.path.extname = basename.slice(1).join('.') || '';
  tok.path.ext = '';

  if (isGlob(tok.path.dirname) && !tok.path.basename) ***REMOVED***
    if (!/\/$/.test(tok.glob)) ***REMOVED***
      tok.path.basename = tok.glob;
    ***REMOVED***
    tok.path.dirname = tok.base;
  ***REMOVED***

  if (glob.indexOf('/') === -1 && !tok.is.globstar) ***REMOVED***
    tok.path.dirname = '';
    tok.path.basename = tok.orig;
  ***REMOVED***

  var dot = tok.path.basename.indexOf('.');
  if (dot !== -1) ***REMOVED***
    tok.path.filename = tok.path.basename.slice(0, dot);
    tok.path.extname = tok.path.basename.slice(dot);
  ***REMOVED***

  if (tok.path.extname.charAt(0) === '.') ***REMOVED***
    var exts = tok.path.extname.split('.');
    tok.path.ext = exts[exts.length - 1];
  ***REMOVED***

  // unescape dots and slashes in braces/brackets
  tok.glob = unescape(tok.glob);
  tok.path.dirname = unescape(tok.path.dirname);
  tok.path.basename = unescape(tok.path.basename);
  tok.path.filename = unescape(tok.path.filename);
  tok.path.extname = unescape(tok.path.extname);

  // Booleans
  var is = (glob && tok.is.glob);
  tok.is.negated  = glob && glob.charAt(0) === '!';
  tok.is.extglob  = glob && extglob(glob);
  tok.is.braces   = has(is, glob, '***REMOVED***');
  tok.is.brackets = has(is, glob, '[:');
  tok.is.globstar = has(is, glob, '**');
  tok.is.dotfile  = dotfile(tok.path.basename) || dotfile(tok.path.filename);
  tok.is.dotdir   = dotdir(tok.path.dirname);
  return (cache[glob] = tok);
***REMOVED***

/**
 * Returns true if the glob matches dot-directories.
 *
 * @param  ***REMOVED***Object***REMOVED*** `tok` The tokens object
 * @param  ***REMOVED***Object***REMOVED*** `path` The path object
 * @return ***REMOVED***Object***REMOVED***
 */

function dotdir(base) ***REMOVED***
  if (base.indexOf('/.') !== -1) ***REMOVED***
    return true;
  ***REMOVED***
  if (base.charAt(0) === '.' && base.charAt(1) !== '/') ***REMOVED***
    return true;
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Returns true if the pattern has the given `ch`aracter(s)
 *
 * @param  ***REMOVED***Object***REMOVED*** `glob` The glob pattern.
 * @param  ***REMOVED***Object***REMOVED*** `ch` The character to test for
 * @return ***REMOVED***Object***REMOVED***
 */

function has(is, glob, ch) ***REMOVED***
  return is && glob.indexOf(ch) !== -1;
***REMOVED***

/**
 * Escape/unescape utils
 */

function escape(str) ***REMOVED***
  var re = /\***REMOVED***([^***REMOVED******REMOVED***]*?)***REMOVED***|\(([^()]*?)\)|\[([^\[\]]*?)\]/g;
  return str.replace(re, function (outter, braces, parens, brackets) ***REMOVED***
    var inner = braces || parens || brackets;
    if (!inner) ***REMOVED*** return outter; ***REMOVED***
    return outter.split(inner).join(esc(inner));
  ***REMOVED***);
***REMOVED***

function esc(str) ***REMOVED***
  str = str.split('/').join('__SLASH__');
  str = str.split('.').join('__DOT__');
  return str;
***REMOVED***

function unescape(str) ***REMOVED***
  str = str.split('__SLASH__').join('/');
  str = str.split('__DOT__').join('.');
  return str;
***REMOVED***
