// @firebase/database/dist/esm/src/core/RepoInfo.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** assert ***REMOVED*** from '@firebase/util';
import ***REMOVED*** forEach ***REMOVED*** from '@firebase/util';
import ***REMOVED*** PersistentStorage ***REMOVED*** from './storage/storage';
import ***REMOVED*** LONG_POLLING, WEBSOCKET ***REMOVED*** from '../realtime/Constants';
/**
 * A class that holds metadata about a Repo object
 *
 * @constructor
 */
var RepoInfo = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***string***REMOVED*** host Hostname portion of the url for the repo
     * @param ***REMOVED***boolean***REMOVED*** secure Whether or not this repo is accessed over ssl
     * @param ***REMOVED***string***REMOVED*** namespace The namespace represented by the repo
     * @param ***REMOVED***boolean***REMOVED*** webSocketOnly Whether to prefer websockets over all other transports (used by Nest).
     * @param ***REMOVED***string=***REMOVED*** persistenceKey Override the default session persistence storage key
     */
    function RepoInfo(host, secure, namespace, webSocketOnly, persistenceKey) ***REMOVED***
        if (persistenceKey === void 0) ***REMOVED*** persistenceKey = ''; ***REMOVED***
        this.secure = secure;
        this.namespace = namespace;
        this.webSocketOnly = webSocketOnly;
        this.persistenceKey = persistenceKey;
        this.host = host.toLowerCase();
        this.domain = this.host.substr(this.host.indexOf('.') + 1);
        this.internalHost = PersistentStorage.get('host:' + host) || this.host;
    ***REMOVED***
    RepoInfo.prototype.needsQueryParam = function () ***REMOVED***
        return this.host !== this.internalHost;
    ***REMOVED***;
    RepoInfo.prototype.isCacheableHost = function () ***REMOVED***
        return this.internalHost.substr(0, 2) === 's-';
    ***REMOVED***;
    RepoInfo.prototype.isDemoHost = function () ***REMOVED***
        return this.domain === 'firebaseio-demo.com';
    ***REMOVED***;
    RepoInfo.prototype.isCustomHost = function () ***REMOVED***
        return (this.domain !== 'firebaseio.com' && this.domain !== 'firebaseio-demo.com');
    ***REMOVED***;
    RepoInfo.prototype.updateHost = function (newHost) ***REMOVED***
        if (newHost !== this.internalHost) ***REMOVED***
            this.internalHost = newHost;
            if (this.isCacheableHost()) ***REMOVED***
                PersistentStorage.set('host:' + this.host, this.internalHost);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    /**
     * Returns the websocket URL for this repo
     * @param ***REMOVED***string***REMOVED*** type of connection
     * @param ***REMOVED***Object***REMOVED*** params list
     * @return ***REMOVED***string***REMOVED*** The URL for this repo
     */
    RepoInfo.prototype.connectionURL = function (type, params) ***REMOVED***
        assert(typeof type === 'string', 'typeof type must == string');
        assert(typeof params === 'object', 'typeof params must == object');
        var connURL;
        if (type === WEBSOCKET) ***REMOVED***
            connURL =
                (this.secure ? 'wss://' : 'ws://') + this.internalHost + '/.ws?';
        ***REMOVED***
        else if (type === LONG_POLLING) ***REMOVED***
            connURL =
                (this.secure ? 'https://' : 'http://') + this.internalHost + '/.lp?';
        ***REMOVED***
        else ***REMOVED***
            throw new Error('Unknown connection type: ' + type);
        ***REMOVED***
        if (this.needsQueryParam()) ***REMOVED***
            params['ns'] = this.namespace;
        ***REMOVED***
        var pairs = [];
        forEach(params, function (key, value) ***REMOVED***
            pairs.push(key + '=' + value);
        ***REMOVED***);
        return connURL + pairs.join('&');
    ***REMOVED***;
    /** @return ***REMOVED***string***REMOVED*** */
    RepoInfo.prototype.toString = function () ***REMOVED***
        var str = this.toURLString();
        if (this.persistenceKey) ***REMOVED***
            str += '<' + this.persistenceKey + '>';
        ***REMOVED***
        return str;
    ***REMOVED***;
    /** @return ***REMOVED***string***REMOVED*** */
    RepoInfo.prototype.toURLString = function () ***REMOVED***
        return (this.secure ? 'https://' : 'http://') + this.host;
    ***REMOVED***;
    return RepoInfo;
***REMOVED***());
export ***REMOVED*** RepoInfo ***REMOVED***;
