// @firebase/database/dist/esm/src/core/RepoManager.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** safeGet ***REMOVED*** from '@firebase/util';
import ***REMOVED*** Repo ***REMOVED*** from './Repo';
import ***REMOVED*** fatal ***REMOVED*** from './util/util';
import ***REMOVED*** parseRepoInfo ***REMOVED*** from './util/libs/parser';
import ***REMOVED*** validateUrl ***REMOVED*** from './util/validation';
import './Repo_transaction';
/** @const ***REMOVED***string***REMOVED*** */
var DATABASE_URL_OPTION = 'databaseURL';
var _staticInstance;
/**
 * Creates and caches Repo instances.
 */
var RepoManager = /** @class */ (function () ***REMOVED***
    function RepoManager() ***REMOVED***
        /**
         * @private ***REMOVED***!Object.<string, Object<string, !fb.core.Repo>>***REMOVED***
         */
        this.repos_ = ***REMOVED******REMOVED***;
        /**
         * If true, new Repos will be created to use ReadonlyRestClient (for testing purposes).
         * @private ***REMOVED***boolean***REMOVED***
         */
        this.useRestClient_ = false;
    ***REMOVED***
    RepoManager.getInstance = function () ***REMOVED***
        if (!_staticInstance) ***REMOVED***
            _staticInstance = new RepoManager();
        ***REMOVED***
        return _staticInstance;
    ***REMOVED***;
    // TODO(koss): Remove these functions unless used in tests?
    RepoManager.prototype.interrupt = function () ***REMOVED***
        for (var appName in this.repos_) ***REMOVED***
            for (var dbUrl in this.repos_[appName]) ***REMOVED***
                this.repos_[appName][dbUrl].interrupt();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    RepoManager.prototype.resume = function () ***REMOVED***
        for (var appName in this.repos_) ***REMOVED***
            for (var dbUrl in this.repos_[appName]) ***REMOVED***
                this.repos_[appName][dbUrl].resume();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    /**
     * This function should only ever be called to CREATE a new database instance.
     *
     * @param ***REMOVED***!FirebaseApp***REMOVED*** app
     * @return ***REMOVED***!Database***REMOVED***
     */
    RepoManager.prototype.databaseFromApp = function (app, url) ***REMOVED***
        var dbUrl = url || app.options[DATABASE_URL_OPTION];
        if (dbUrl === undefined) ***REMOVED***
            fatal("Can't determine Firebase Database URL.  Be sure to include " +
                DATABASE_URL_OPTION +
                ' option when calling firebase.initializeApp().');
        ***REMOVED***
        var parsedUrl = parseRepoInfo(dbUrl);
        var repoInfo = parsedUrl.repoInfo;
        validateUrl('Invalid Firebase Database URL', 1, parsedUrl);
        if (!parsedUrl.path.isEmpty()) ***REMOVED***
            fatal('Database URL must point to the root of a Firebase Database ' +
                '(not including a child path).');
        ***REMOVED***
        var repo = this.createRepo(repoInfo, app);
        return repo.database;
    ***REMOVED***;
    /**
     * Remove the repo and make sure it is disconnected.
     *
     * @param ***REMOVED***!Repo***REMOVED*** repo
     */
    RepoManager.prototype.deleteRepo = function (repo) ***REMOVED***
        var appRepos = safeGet(this.repos_, repo.app.name);
        // This should never happen...
        if (!appRepos || safeGet(appRepos, repo.repoInfo_.toURLString()) !== repo) ***REMOVED***
            fatal("Database " + repo.app.name + "(" + repo.repoInfo_ + ") has already been deleted.");
        ***REMOVED***
        repo.interrupt();
        delete appRepos[repo.repoInfo_.toURLString()];
    ***REMOVED***;
    /**
     * Ensures a repo doesn't already exist and then creates one using the
     * provided app.
     *
     * @param ***REMOVED***!RepoInfo***REMOVED*** repoInfo The metadata about the Repo
     * @param ***REMOVED***!FirebaseApp***REMOVED*** app
     * @return ***REMOVED***!Repo***REMOVED*** The Repo object for the specified server / repoName.
     */
    RepoManager.prototype.createRepo = function (repoInfo, app) ***REMOVED***
        var appRepos = safeGet(this.repos_, app.name);
        if (!appRepos) ***REMOVED***
            appRepos = ***REMOVED******REMOVED***;
            this.repos_[app.name] = appRepos;
        ***REMOVED***
        var repo = safeGet(appRepos, repoInfo.toURLString());
        if (repo) ***REMOVED***
            fatal('Database initialized multiple times. Please make sure the format of the database URL matches with each database() call.');
        ***REMOVED***
        repo = new Repo(repoInfo, this.useRestClient_, app);
        appRepos[repoInfo.toURLString()] = repo;
        return repo;
    ***REMOVED***;
    /**
     * Forces us to use ReadonlyRestClient instead of PersistentConnection for new Repos.
     * @param ***REMOVED***boolean***REMOVED*** forceRestClient
     */
    RepoManager.prototype.forceRestClient = function (forceRestClient) ***REMOVED***
        this.useRestClient_ = forceRestClient;
    ***REMOVED***;
    return RepoManager;
***REMOVED***());
export ***REMOVED*** RepoManager ***REMOVED***;
