// enhanced-resolve/lib/concord.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var globToRegExp = require("./globToRegExp").globToRegExp;

function parseType(type) ***REMOVED***
	var items = type.split("+");
	var t = items.shift();
	return ***REMOVED***
		type: t === "*" ? null : t,
		features: items
	***REMOVED***;
***REMOVED***

function isTypeMatched(baseType, testedType) ***REMOVED***
	if(typeof baseType === "string") baseType = parseType(baseType);
	if(typeof testedType === "string") testedType = parseType(testedType);
	if(testedType.type && testedType.type !== baseType.type) return false;
	return testedType.features.every(function(requiredFeature) ***REMOVED***
		return baseType.features.indexOf(requiredFeature) >= 0;
	***REMOVED***);
***REMOVED***

function isResourceTypeMatched(baseType, testedType) ***REMOVED***
	baseType = baseType.split("/");
	testedType = testedType.split("/");
	if(baseType.length !== testedType.length) return false;
	for(var i = 0; i < baseType.length; i++) ***REMOVED***
		if(!isTypeMatched(baseType[i], testedType[i]))
			return false;
	***REMOVED***
	return true;
***REMOVED***

function isResourceTypeSupported(context, type) ***REMOVED***
	return context.supportedResourceTypes && context.supportedResourceTypes.some(function(supportedType) ***REMOVED***
		return isResourceTypeMatched(supportedType, type);
	***REMOVED***);
***REMOVED***

function isEnvironment(context, env) ***REMOVED***
	return context.environments && context.environments.every(function(environment) ***REMOVED***
		return isTypeMatched(environment, env);
	***REMOVED***);
***REMOVED***

var globCache = ***REMOVED******REMOVED***;

function getGlobRegExp(glob) ***REMOVED***
	var regExp = globCache[glob] || (globCache[glob] = globToRegExp(glob));
	return regExp;
***REMOVED***

function matchGlob(glob, relativePath) ***REMOVED***
	var regExp = getGlobRegExp(glob);
	return regExp.exec(relativePath);
***REMOVED***

function isGlobMatched(glob, relativePath) ***REMOVED***
	return !!matchGlob(glob, relativePath);
***REMOVED***

function isConditionMatched(context, condition) ***REMOVED***
	var items = condition.split("|");
	return items.some(function testFn(item) ***REMOVED***
		item = item.trim();
		var inverted = /^!/.test(item);
		if(inverted) return !testFn(item.substr(1));
		if(/^[a-z]+:/.test(item)) ***REMOVED***
			// match named condition
			var match = /^([a-z]+):\s*/.exec(item);
			var value = item.substr(match[0].length);
			var name = match[1];
			switch(name) ***REMOVED***
				case "referrer":
					return isGlobMatched(value, context.referrer);
				default:
					return false;
			***REMOVED***
		***REMOVED*** else if(item.indexOf("/") >= 0) ***REMOVED***
			// match supported type
			return isResourceTypeSupported(context, item);
		***REMOVED*** else ***REMOVED***
			// match environment
			return isEnvironment(context, item);
		***REMOVED***
	***REMOVED***);
***REMOVED***

function isKeyMatched(context, key) ***REMOVED***
	while(true) ***REMOVED*** //eslint-disable-line
		var match = /^\[([^\]]+)\]\s*/.exec(key);
		if(!match) return key;
		key = key.substr(match[0].length);
		var condition = match[1];
		if(!isConditionMatched(context, condition)) ***REMOVED***
			return false;
		***REMOVED***
	***REMOVED***
***REMOVED***

function getField(context, configuration, field) ***REMOVED***
	var value;
	Object.keys(configuration).forEach(function(key) ***REMOVED***
		var pureKey = isKeyMatched(context, key);
		if(pureKey === field) ***REMOVED***
			value = configuration[key];
		***REMOVED***
	***REMOVED***);
	return value;
***REMOVED***

function getMain(context, configuration) ***REMOVED***
	return getField(context, configuration, "main");
***REMOVED***

function getExtensions(context, configuration) ***REMOVED***
	return getField(context, configuration, "extensions");
***REMOVED***

function matchModule(context, configuration, request) ***REMOVED***
	var modulesField = getField(context, configuration, "modules");
	if(!modulesField) return request;
	var newRequest = request;
	var keys = Object.keys(modulesField);
	var iteration = 0;
	for(var i = 0; i < keys.length; i++) ***REMOVED***
		var key = keys[i];
		var pureKey = isKeyMatched(context, key);
		var match = matchGlob(pureKey, newRequest);
		if(match) ***REMOVED***
			var value = modulesField[key];
			if(typeof value !== "string") ***REMOVED***
				return value;
			***REMOVED*** else if(/^\(.+\)$/.test(pureKey)) ***REMOVED***
				newRequest = newRequest.replace(getGlobRegExp(pureKey), value);
			***REMOVED*** else ***REMOVED***
				var index = 1;
				newRequest = value.replace(/(\/?\*)?\*/g, replaceMatcher);
			***REMOVED***
			i = -1;
			if(iteration++ > keys.length) ***REMOVED***
				throw new Error("Request '" + request + "' matches recursively");
			***REMOVED***
		***REMOVED***
	***REMOVED***
	return newRequest;

	function replaceMatcher(find) ***REMOVED***
		switch(find) ***REMOVED***
			case "/**":
				var m = match[index++];
				return m ? "/" + m : "";
			case "**":
			case "*":
				return match[index++];
		***REMOVED***
	***REMOVED***
***REMOVED***

function matchType(context, configuration, relativePath) ***REMOVED***
	var typesField = getField(context, configuration, "types");
	if(!typesField) return undefined;
	var type;
	Object.keys(typesField).forEach(function(key) ***REMOVED***
		var pureKey = isKeyMatched(context, key);
		if(isGlobMatched(pureKey, relativePath)) ***REMOVED***
			var value = typesField[key];
			if(!type && /\/\*$/.test(value))
				throw new Error("value ('" + value + "') of key '" + key + "' contains '*', but there is no previous value defined");
			type = value.replace(/\/\*$/, "/" + type);
		***REMOVED***
	***REMOVED***);
	return type;
***REMOVED***

exports.parseType = parseType;
exports.isTypeMatched = isTypeMatched;
exports.isResourceTypeSupported = isResourceTypeSupported;
exports.isEnvironment = isEnvironment;
exports.isGlobMatched = isGlobMatched;
exports.isConditionMatched = isConditionMatched;
exports.isKeyMatched = isKeyMatched;
exports.getField = getField;
exports.getMain = getMain;
exports.getExtensions = getExtensions;
exports.matchModule = matchModule;
exports.matchType = matchType;
