// enhanced-resolve/lib/ConcordModulesPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var concord = require("./concord");
var DescriptionFileUtils = require("./DescriptionFileUtils");
var createInnerCallback = require("./createInnerCallback");
var getInnerRequest = require("./getInnerRequest");

function ConcordModulesPlugin(source, options, target) ***REMOVED***
	this.source = source;
	this.options = options;
	this.target = target;
***REMOVED***
module.exports = ConcordModulesPlugin;

ConcordModulesPlugin.prototype.apply = function(resolver) ***REMOVED***
	var target = this.target;
	resolver.plugin(this.source, function(request, callback) ***REMOVED***
		var innerRequest = getInnerRequest(resolver, request);
		if(!innerRequest) return callback();
		var concordField = DescriptionFileUtils.getField(request.descriptionFileData, "concord");
		if(!concordField) return callback();
		var data = concord.matchModule(request.context, concordField, innerRequest);
		if(data === innerRequest) return callback();
		if(data === undefined) return callback();
		if(data === false) ***REMOVED***
			var ignoreObj = Object.assign(***REMOVED******REMOVED***, request, ***REMOVED***
				path: false
			***REMOVED***);
			return callback(null, ignoreObj);
		***REMOVED***
		var obj = Object.assign(***REMOVED******REMOVED***, request, ***REMOVED***
			path: request.descriptionFileRoot,
			request: data
		***REMOVED***);
		resolver.doResolve(target, obj, "aliased from description file " + request.descriptionFilePath + " with mapping '" + innerRequest + "' to '" + data + "'", createInnerCallback(function(err, result) ***REMOVED***
			if(arguments.length > 0) return callback(err, result);

			// Don't allow other aliasing or raw request
			callback(null, null);
		***REMOVED***, callback));
	***REMOVED***);
***REMOVED***;
