// enhanced-resolve/lib/ConcordMainPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var path = require("path");
var concord = require("./concord");
var DescriptionFileUtils = require("./DescriptionFileUtils");

function ConcordMainPlugin(source, options, target) ***REMOVED***
	this.source = source;
	this.options = options;
	this.target = target;
***REMOVED***
module.exports = ConcordMainPlugin;

ConcordMainPlugin.prototype.apply = function(resolver) ***REMOVED***
	var target = this.target;
	resolver.plugin(this.source, function(request, callback) ***REMOVED***
		if(request.path !== request.descriptionFileRoot) return callback();
		var concordField = DescriptionFileUtils.getField(request.descriptionFileData, "concord");
		if(!concordField) return callback();
		var mainModule = concord.getMain(request.context, concordField);
		if(!mainModule) return callback();
		var obj = Object.assign(***REMOVED******REMOVED***, request, ***REMOVED***
			request: mainModule
		***REMOVED***);
		var filename = path.basename(request.descriptionFilePath);
		return resolver.doResolve(target, obj, "use " + mainModule + " from " + filename, callback);
	***REMOVED***);
***REMOVED***;
