// enhanced-resolve/lib/MainFieldPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var path = require("path");

function MainFieldPlugin(source, options, target) ***REMOVED***
	this.source = source;
	this.options = options;
	this.target = target;
***REMOVED***
module.exports = MainFieldPlugin;

MainFieldPlugin.prototype.apply = function(resolver) ***REMOVED***
	var target = this.target;
	var options = this.options;
	resolver.plugin(this.source, function mainField(request, callback) ***REMOVED***
		if(request.path !== request.descriptionFileRoot) return callback();
		var content = request.descriptionFileData;
		var filename = path.basename(request.descriptionFilePath);
		var mainModule;
		var field = options.name;
		if(Array.isArray(field)) ***REMOVED***
			var current = content;
			for(var j = 0; j < field.length; j++) ***REMOVED***
				if(current === null || typeof current !== "object") ***REMOVED***
					current = null;
					break;
				***REMOVED***
				current = current[field[j]];
			***REMOVED***
			if(typeof current === "string") ***REMOVED***
				mainModule = current;
			***REMOVED***
		***REMOVED*** else ***REMOVED***
			if(typeof content[field] === "string") ***REMOVED***
				mainModule = content[field];
			***REMOVED***
		***REMOVED***
		if(!mainModule) return callback();
		if(options.forceRelative && !/^\.\.?\//.test(mainModule))
			mainModule = "./" + mainModule;
		var obj = Object.assign(***REMOVED******REMOVED***, request, ***REMOVED***
			request: mainModule
		***REMOVED***);
		return resolver.doResolve(target, obj, "use " + mainModule + " from " + options.name + " in " + filename, callback);
	***REMOVED***);
***REMOVED***;
