// enhanced-resolve/lib/ModuleAppendPlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
function ModuleAppendPlugin(source, appending, target) ***REMOVED***
	this.source = source;
	this.appending = appending;
	this.target = target;
***REMOVED***
module.exports = ModuleAppendPlugin;

ModuleAppendPlugin.prototype.apply = function(resolver) ***REMOVED***
	var appending = this.appending;
	var target = this.target;
	resolver.plugin(this.source, function(request, callback) ***REMOVED***
		var i = request.request.indexOf("/"),
			j = request.request.indexOf("\\");
		var p = i < 0 ? j : j < 0 ? i : i < j ? i : j;
		var moduleName, remainingRequest;
		if(p < 0) ***REMOVED***
			moduleName = request.request;
			remainingRequest = "";
		***REMOVED*** else ***REMOVED***
			moduleName = request.request.substr(0, p);
			remainingRequest = request.request.substr(p);
		***REMOVED***
		if(moduleName === "." || moduleName === "..")
			return callback();
		var moduleFinalName = moduleName + appending;
		var obj = Object.assign(***REMOVED******REMOVED***, request, ***REMOVED***
			request: moduleFinalName + remainingRequest
		***REMOVED***);
		resolver.doResolve(target, obj, "module variation " + moduleFinalName, callback);
	***REMOVED***);
***REMOVED***;
