// enhanced-resolve/lib/SyncAsyncFileSystemDecorator.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
function SyncAsyncFileSystemDecorator(fs) ***REMOVED***
	this.fs = fs;
	if(fs.statSync) ***REMOVED***
		this.stat = function(arg, callback) ***REMOVED***
			try ***REMOVED***
				var result = fs.statSync(arg);
			***REMOVED*** catch(e) ***REMOVED***
				return callback(e);
			***REMOVED***
			callback(null, result);
		***REMOVED***;
	***REMOVED***
	if(fs.readdirSync) ***REMOVED***
		this.readdir = function(arg, callback) ***REMOVED***
			try ***REMOVED***
				var result = fs.readdirSync(arg);
			***REMOVED*** catch(e) ***REMOVED***
				return callback(e);
			***REMOVED***
			callback(null, result);
		***REMOVED***;
	***REMOVED***
	if(fs.readFileSync) ***REMOVED***
		this.readFile = function(arg, callback) ***REMOVED***
			try ***REMOVED***
				var result = fs.readFileSync(arg);
			***REMOVED*** catch(e) ***REMOVED***
				return callback(e);
			***REMOVED***
			callback(null, result);
		***REMOVED***;
	***REMOVED***
	if(fs.readlinkSync) ***REMOVED***
		this.readlink = function(arg, callback) ***REMOVED***
			try ***REMOVED***
				var result = fs.readlinkSync(arg);
			***REMOVED*** catch(e) ***REMOVED***
				return callback(e);
			***REMOVED***
			callback(null, result);
		***REMOVED***;
	***REMOVED***
	if(fs.readJsonSync) ***REMOVED***
		this.readJson = function(arg, callback) ***REMOVED***
			try ***REMOVED***
				var result = fs.readJsonSync(arg);
			***REMOVED*** catch(e) ***REMOVED***
				return callback(e);
			***REMOVED***
			callback(null, result);
		***REMOVED***;
	***REMOVED***
***REMOVED***
module.exports = SyncAsyncFileSystemDecorator;
