// source-list-map/lib/MappingsContext.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

class MappingsContext ***REMOVED***
	constructor() ***REMOVED***
		this.sourcesIndices = new Map();
		this.sourcesContent = new Map();
		this.hasSourceContent = false;
		this.currentOriginalLine = 1;
		this.currentSource = 0;
		this.unfinishedGeneratedLine = false;
	***REMOVED***

	ensureSource(source, originalSource) ***REMOVED***
		let idx = this.sourcesIndices.get(source);
		if(typeof idx === "number") ***REMOVED***
			return idx;
		***REMOVED***
		idx = this.sourcesIndices.size;
		this.sourcesIndices.set(source, idx);
		this.sourcesContent.set(source, originalSource)
		if(typeof originalSource === "string")
			this.hasSourceContent = true;
		return idx;
	***REMOVED***

	getArrays() ***REMOVED***
		const sources = [];
		const sourcesContent = [];

		for(const pair of this.sourcesContent) ***REMOVED***
			sources.push(pair[0]);
			sourcesContent.push(pair[1]);
		***REMOVED***

		return ***REMOVED***
			sources,
			sourcesContent
		***REMOVED***;
	***REMOVED***
***REMOVED***
module.exports = MappingsContext;
