// firebase-admin/lib/firebase-namespace.js
/*! firebase-admin v5.8.2 */
"use strict";
/*!
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var fs = require("fs");
var deep_copy_1 = require("./utils/deep-copy");
var error_1 = require("./utils/error");
var firebase_app_1 = require("./firebase-app");
var credential_1 = require("./auth/credential");
var auth_1 = require("./auth/auth");
var messaging_1 = require("./messaging/messaging");
var storage_1 = require("./storage/storage");
var instance_id_1 = require("./instance-id/instance-id");
var validator = require("./utils/validator");
var DEFAULT_APP_NAME = '[DEFAULT]';
/**
 * Constant holding the environment variable name with the default config.
 * If the environmet variable contains a string that starts with '***REMOVED***' it will be parsed as JSON,
 * otherwise it will be assumed to be pointing to a file.
 */
exports.FIREBASE_CONFIG_VAR = 'FIREBASE_CONFIG';
var globalAppDefaultCred;
var globalCertCreds = ***REMOVED******REMOVED***;
var globalRefreshTokenCreds = ***REMOVED******REMOVED***;
/**
 * Internals of a FirebaseNamespace instance.
 */
var FirebaseNamespaceInternals = /** @class */ (function () ***REMOVED***
    function FirebaseNamespaceInternals(firebase_) ***REMOVED***
        this.firebase_ = firebase_;
        this.serviceFactories = ***REMOVED******REMOVED***;
        this.apps_ = ***REMOVED******REMOVED***;
        this.appHooks_ = ***REMOVED******REMOVED***;
    ***REMOVED***
    /**
     * Initializes the FirebaseApp instance.
     *
     * @param ***REMOVED***FirebaseAppOptions***REMOVED*** options Optional options for the FirebaseApp instance. If none present
     *                             will try to initialize from the FIREBASE_CONFIG environment variable.
     *                             If the environmet variable contains a string that starts with '***REMOVED***'
     *                             it will be parsed as JSON,
     *                             otherwise it will be assumed to be pointing to a file.
     * @param ***REMOVED***string***REMOVED*** [appName] Optional name of the FirebaseApp instance.
     *
     * @return ***REMOVED***FirebaseApp***REMOVED*** A new FirebaseApp instance.
     */
    FirebaseNamespaceInternals.prototype.initializeApp = function (options, appName) ***REMOVED***
        if (appName === void 0) ***REMOVED*** appName = DEFAULT_APP_NAME; ***REMOVED***
        if (typeof options === 'undefined') ***REMOVED***
            options = this.loadOptionsFromEnvVar();
            options.credential = new credential_1.ApplicationDefaultCredential();
        ***REMOVED***
        if (typeof appName !== 'string' || appName === '') ***REMOVED***
            throw new error_1.FirebaseAppError(error_1.AppErrorCodes.INVALID_APP_NAME, "Invalid Firebase app name \"" + appName + "\" provided. App name must be a non-empty string.");
        ***REMOVED***
        else if (appName in this.apps_) ***REMOVED***
            if (appName === DEFAULT_APP_NAME) ***REMOVED***
                throw new error_1.FirebaseAppError(error_1.AppErrorCodes.DUPLICATE_APP, 'The default Firebase app already exists. This means you called initializeApp() ' +
                    'more than once without providing an app name as the second argument. In most cases ' +
                    'you only need to call initializeApp() once. But if you do want to initialize ' +
                    'multiple apps, pass a second argument to initializeApp() to give each app a unique ' +
                    'name.');
            ***REMOVED***
            else ***REMOVED***
                throw new error_1.FirebaseAppError(error_1.AppErrorCodes.DUPLICATE_APP, "Firebase app named \"" + appName + "\" already exists. This means you called initializeApp() " +
                    'more than once with the same app name as the second argument. Make sure you provide a ' +
                    'unique name every time you call initializeApp().');
            ***REMOVED***
        ***REMOVED***
        var app = new firebase_app_1.FirebaseApp(options, appName, this);
        this.apps_[appName] = app;
        this.callAppHooks_(app, 'create');
        return app;
    ***REMOVED***;
    /**
     * Returns the FirebaseApp instance with the provided name (or the default FirebaseApp instance
     * if no name is provided).
     *
     * @param ***REMOVED***string***REMOVED*** [appName=DEFAULT_APP_NAME] Optional name of the FirebaseApp instance to return.
     * @return ***REMOVED***FirebaseApp***REMOVED*** The FirebaseApp instance which has the provided name.
     */
    FirebaseNamespaceInternals.prototype.app = function (appName) ***REMOVED***
        if (appName === void 0) ***REMOVED*** appName = DEFAULT_APP_NAME; ***REMOVED***
        if (typeof appName !== 'string' || appName === '') ***REMOVED***
            throw new error_1.FirebaseAppError(error_1.AppErrorCodes.INVALID_APP_NAME, "Invalid Firebase app name \"" + appName + "\" provided. App name must be a non-empty string.");
        ***REMOVED***
        else if (!(appName in this.apps_)) ***REMOVED***
            var errorMessage = void 0;
            if (appName === DEFAULT_APP_NAME) ***REMOVED***
                errorMessage = 'The default Firebase app does not exist. ';
            ***REMOVED***
            else ***REMOVED***
                errorMessage = "Firebase app named \"" + appName + "\" does not exist. ";
            ***REMOVED***
            errorMessage += 'Make sure you call initializeApp() before using any of the Firebase services.';
            throw new error_1.FirebaseAppError(error_1.AppErrorCodes.NO_APP, errorMessage);
        ***REMOVED***
        return this.apps_[appName];
    ***REMOVED***;
    Object.defineProperty(FirebaseNamespaceInternals.prototype, "apps", ***REMOVED***
        /*
         * Returns an array of all the non-deleted FirebaseApp instances.
         *
         * @return ***REMOVED***Array<FirebaseApp>***REMOVED*** An array of all the non-deleted FirebaseApp instances
         */
        get: function () ***REMOVED***
            var _this = this;
            // Return a copy so the caller cannot mutate the array
            return Object.keys(this.apps_).map(function (appName) ***REMOVED*** return _this.apps_[appName]; ***REMOVED***);
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    /*
     * Removes the specified FirebaseApp instance.
     *
     * @param ***REMOVED***string***REMOVED*** appName The name of the FirebaseApp instance to remove.
     */
    FirebaseNamespaceInternals.prototype.removeApp = function (appName) ***REMOVED***
        if (typeof appName === 'undefined') ***REMOVED***
            throw new error_1.FirebaseAppError(error_1.AppErrorCodes.INVALID_APP_NAME, "No Firebase app name provided. App name must be a non-empty string.");
        ***REMOVED***
        var appToRemove = this.app(appName);
        this.callAppHooks_(appToRemove, 'delete');
        delete this.apps_[appName];
    ***REMOVED***;
    /*
     * Registers a new service on this Firebase namespace.
     *
     * @param ***REMOVED***string***REMOVED*** serviceName The name of the Firebase service to register.
     * @param ***REMOVED***FirebaseServiceFactory***REMOVED*** createService A factory method to generate an instance of the Firebase service.
     * @param ***REMOVED***Object***REMOVED*** [serviceProperties] Optional properties to extend this Firebase namespace with.
     * @param ***REMOVED***AppHook***REMOVED*** [appHook] Optional callback that handles app-related events like app creation and deletion.
     * @return ***REMOVED***FirebaseServiceNamespace<FirebaseServiceInterface>***REMOVED*** The Firebase service's namespace.
     */
    FirebaseNamespaceInternals.prototype.registerService = function (serviceName, createService, serviceProperties, appHook) ***REMOVED***
        var _this = this;
        var errorMessage;
        if (typeof serviceName === 'undefined') ***REMOVED***
            errorMessage = "No service name provided. Service name must be a non-empty string.";
        ***REMOVED***
        else if (typeof serviceName !== 'string' || serviceName === '') ***REMOVED***
            errorMessage = "Invalid service name \"" + serviceName + "\" provided. Service name must be a non-empty string.";
        ***REMOVED***
        else if (serviceName in this.serviceFactories) ***REMOVED***
            errorMessage = "Firebase service named \"" + serviceName + "\" has already been registered.";
        ***REMOVED***
        if (typeof errorMessage !== 'undefined') ***REMOVED***
            throw new error_1.FirebaseAppError(error_1.AppErrorCodes.INTERNAL_ERROR, "INTERNAL ASSERT FAILED: " + errorMessage);
        ***REMOVED***
        this.serviceFactories[serviceName] = createService;
        if (appHook) ***REMOVED***
            this.appHooks_[serviceName] = appHook;
        ***REMOVED***
        var serviceNamespace;
        // The service namespace is an accessor function which takes a FirebaseApp instance
        // or uses the default app if no FirebaseApp instance is provided
        serviceNamespace = function (appArg) ***REMOVED***
            if (typeof appArg === 'undefined') ***REMOVED***
                appArg = _this.app();
            ***REMOVED***
            // Forward service instance lookup to the FirebaseApp
            return appArg[serviceName]();
        ***REMOVED***;
        // ... and a container for service-level properties.
        if (serviceProperties !== undefined) ***REMOVED***
            deep_copy_1.deepExtend(serviceNamespace, serviceProperties);
        ***REMOVED***
        // Monkey-patch the service namespace onto the Firebase namespace
        this.firebase_[serviceName] = serviceNamespace;
        return serviceNamespace;
    ***REMOVED***;
    /**
     * Calls the app hooks corresponding to the provided event name for each service within the
     * provided FirebaseApp instance.
     *
     * @param ***REMOVED***FirebaseApp***REMOVED*** app The FirebaseApp instance whose app hooks to call.
     * @param ***REMOVED***string***REMOVED*** eventName The event name representing which app hooks to call.
     */
    FirebaseNamespaceInternals.prototype.callAppHooks_ = function (app, eventName) ***REMOVED***
        var _this = this;
        Object.keys(this.serviceFactories).forEach(function (serviceName) ***REMOVED***
            if (_this.appHooks_[serviceName]) ***REMOVED***
                _this.appHooks_[serviceName](eventName, app);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Parse the file pointed to by the FIREBASE_CONFIG_VAR, if it exists.
     * Or if the FIREBASE_CONFIG_ENV contains a valid JSON object, parse it directly.
     * If the environmet variable contains a string that starts with '***REMOVED***' it will be parsed as JSON,
     * otherwise it will be assumed to be pointing to a file.
     */
    FirebaseNamespaceInternals.prototype.loadOptionsFromEnvVar = function () ***REMOVED***
        var config = process.env[exports.FIREBASE_CONFIG_VAR];
        if (!validator.isNonEmptyString(config)) ***REMOVED***
            return ***REMOVED******REMOVED***;
        ***REMOVED***
        try ***REMOVED***
            var contents = void 0;
            if (config.startsWith('***REMOVED***')) ***REMOVED***
                // Assume json object.
                contents = config;
            ***REMOVED***
            else ***REMOVED***
                // Assume filename.
                contents = fs.readFileSync(config, 'utf8');
            ***REMOVED***
            return JSON.parse(contents);
        ***REMOVED***
        catch (error) ***REMOVED***
            // Throw a nicely formed error message if the file contents cannot be parsed
            throw new error_1.FirebaseAppError(error_1.AppErrorCodes.INVALID_APP_OPTIONS, 'Failed to parse app options file: ' + error);
        ***REMOVED***
    ***REMOVED***;
    return FirebaseNamespaceInternals;
***REMOVED***());
exports.FirebaseNamespaceInternals = FirebaseNamespaceInternals;
var firebaseCredential = ***REMOVED***
    cert: function (serviceAccountPathOrObject) ***REMOVED***
        var stringifiedServiceAccount = JSON.stringify(serviceAccountPathOrObject);
        if (!(stringifiedServiceAccount in globalCertCreds)) ***REMOVED***
            globalCertCreds[stringifiedServiceAccount] = new credential_1.CertCredential(serviceAccountPathOrObject);
        ***REMOVED***
        return globalCertCreds[stringifiedServiceAccount];
    ***REMOVED***,
    refreshToken: function (refreshTokenPathOrObject) ***REMOVED***
        var stringifiedRefreshToken = JSON.stringify(refreshTokenPathOrObject);
        if (!(stringifiedRefreshToken in globalRefreshTokenCreds)) ***REMOVED***
            globalRefreshTokenCreds[stringifiedRefreshToken] = new credential_1.RefreshTokenCredential(refreshTokenPathOrObject);
        ***REMOVED***
        return globalRefreshTokenCreds[stringifiedRefreshToken];
    ***REMOVED***,
    applicationDefault: function () ***REMOVED***
        if (typeof globalAppDefaultCred === 'undefined') ***REMOVED***
            globalAppDefaultCred = new credential_1.ApplicationDefaultCredential();
        ***REMOVED***
        return globalAppDefaultCred;
    ***REMOVED***,
***REMOVED***;
/**
 * Global Firebase context object.
 */
var FirebaseNamespace = /** @class */ (function () ***REMOVED***
    /* tslint:enable */
    function FirebaseNamespace() ***REMOVED***
        // Hack to prevent Babel from modifying the object returned as the default admin namespace.
        /* tslint:disable:variable-name */
        this.__esModule = true;
        /* tslint:enable:variable-name */
        this.credential = firebaseCredential;
        this.SDK_VERSION = '5.8.2';
        /* tslint:disable */
        // TODO(jwenger): Database is the only consumer of firebase.Promise. We should update it to use
        // use the native Promise and then remove this.
        this.Promise = Promise;
        this.INTERNAL = new FirebaseNamespaceInternals(this);
    ***REMOVED***
    Object.defineProperty(FirebaseNamespace.prototype, "auth", ***REMOVED***
        /**
         * Gets the `Auth` service namespace. The returned namespace can be used to get the
         * `Auth` service for the default app or an explicitly specified app.
         */
        get: function () ***REMOVED***
            var ns = this;
            var fn = function (app) ***REMOVED***
                return ns.ensureApp(app).auth();
            ***REMOVED***;
            return Object.assign(fn, ***REMOVED*** Auth: auth_1.Auth ***REMOVED***);
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseNamespace.prototype, "database", ***REMOVED***
        /**
         * Gets the `Database` service namespace. The returned namespace can be used to get the
         * `Database` service for the default app or an explicitly specified app.
         */
        get: function () ***REMOVED***
            var ns = this;
            var fn = function (app) ***REMOVED***
                return ns.ensureApp(app).database();
            ***REMOVED***;
            return Object.assign(fn, require('@firebase/database'));
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseNamespace.prototype, "messaging", ***REMOVED***
        /**
         * Gets the `Messaging` service namespace. The returned namespace can be used to get the
         * `Messaging` service for the default app or an explicitly specified app.
         */
        get: function () ***REMOVED***
            var ns = this;
            var fn = function (app) ***REMOVED***
                return ns.ensureApp(app).messaging();
            ***REMOVED***;
            return Object.assign(fn, ***REMOVED*** Messaging: messaging_1.Messaging ***REMOVED***);
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseNamespace.prototype, "storage", ***REMOVED***
        /**
         * Gets the `Storage` service namespace. The returned namespace can be used to get the
         * `Storage` service for the default app or an explicitly specified app.
         */
        get: function () ***REMOVED***
            var ns = this;
            var fn = function (app) ***REMOVED***
                return ns.ensureApp(app).storage();
            ***REMOVED***;
            return Object.assign(fn, ***REMOVED*** Storage: storage_1.Storage ***REMOVED***);
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseNamespace.prototype, "firestore", ***REMOVED***
        /**
         * Gets the `Firestore` service namespace. The returned namespace can be used to get the
         * `Firestore` service for the default app or an explicitly specified app.
         */
        get: function () ***REMOVED***
            var ns = this;
            var fn = function (app) ***REMOVED***
                return ns.ensureApp(app).firestore();
            ***REMOVED***;
            return Object.assign(fn, require('@google-cloud/firestore'));
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(FirebaseNamespace.prototype, "instanceId", ***REMOVED***
        /**
         * Gets the `InstanceId` service namespace. The returned namespace can be used to get the
         * `Instance` service for the default app or an explicitly specified app.
         */
        get: function () ***REMOVED***
            var ns = this;
            var fn = function (app) ***REMOVED***
                return ns.ensureApp(app).instanceId();
            ***REMOVED***;
            return Object.assign(fn, ***REMOVED*** InstanceId: instance_id_1.InstanceId ***REMOVED***);
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    /**
     * Initializes the FirebaseApp instance.
     *
     * @param ***REMOVED***FirebaseAppOptions***REMOVED*** [options] Optional options for the FirebaseApp instance.
     *   If none present will try to initialize from the FIREBASE_CONFIG environment variable.
     *   If the environmet variable contains a string that starts with '***REMOVED***' it will be parsed as JSON,
     *   otherwise it will be assumed to be pointing to a file.
     * @param ***REMOVED***string***REMOVED*** [appName] Optional name of the FirebaseApp instance.
     *
     * @return ***REMOVED***FirebaseApp***REMOVED*** A new FirebaseApp instance.
     */
    FirebaseNamespace.prototype.initializeApp = function (options, appName) ***REMOVED***
        return this.INTERNAL.initializeApp(options, appName);
    ***REMOVED***;
    /**
     * Returns the FirebaseApp instance with the provided name (or the default FirebaseApp instance
     * if no name is provided).
     *
     * @param ***REMOVED***string***REMOVED*** [appName] Optional name of the FirebaseApp instance to return.
     * @return ***REMOVED***FirebaseApp***REMOVED*** The FirebaseApp instance which has the provided name.
     */
    FirebaseNamespace.prototype.app = function (appName) ***REMOVED***
        return this.INTERNAL.app(appName);
    ***REMOVED***;
    Object.defineProperty(FirebaseNamespace.prototype, "apps", ***REMOVED***
        /*
         * Returns an array of all the non-deleted FirebaseApp instances.
         *
         * @return ***REMOVED***Array<FirebaseApp>***REMOVED*** An array of all the non-deleted FirebaseApp instances
         */
        get: function () ***REMOVED***
            return this.INTERNAL.apps;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    FirebaseNamespace.prototype.ensureApp = function (app) ***REMOVED***
        if (typeof app === 'undefined') ***REMOVED***
            app = this.app();
        ***REMOVED***
        return app;
    ***REMOVED***;
    return FirebaseNamespace;
***REMOVED***());
exports.FirebaseNamespace = FirebaseNamespace;
