// webpack/lib/dependencies/AMDRequireDependenciesBlock.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";
const AsyncDependenciesBlock = require("../AsyncDependenciesBlock");
const AMDRequireDependency = require("./AMDRequireDependency");

module.exports = class AMDRequireDependenciesBlock extends AsyncDependenciesBlock ***REMOVED***
	constructor(expr, arrayRange, functionRange, errorCallbackRange, module, loc) ***REMOVED***
		super(null, module, loc);
		this.expr = expr;
		this.outerRange = expr.range;
		this.arrayRange = arrayRange;
		this.functionRange = functionRange;
		this.errorCallbackRange = errorCallbackRange;
		this.bindThis = true;
		if(arrayRange && functionRange && errorCallbackRange) ***REMOVED***
			this.range = [arrayRange[0], errorCallbackRange[1]];
		***REMOVED*** else if(arrayRange && functionRange) ***REMOVED***
			this.range = [arrayRange[0], functionRange[1]];
		***REMOVED*** else if(arrayRange) ***REMOVED***
			this.range = arrayRange;
		***REMOVED*** else if(functionRange) ***REMOVED***
			this.range = functionRange;
		***REMOVED*** else ***REMOVED***
			this.range = expr.range;
		***REMOVED***
		const dep = new AMDRequireDependency(this);
		dep.loc = loc;
		this.addDependency(dep);
	***REMOVED***
***REMOVED***;
