// webpack/lib/dependencies/ContextDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";
const Dependency = require("../Dependency");
const CriticalDependencyWarning = require("./CriticalDependencyWarning");

class ContextDependency extends Dependency ***REMOVED***
	constructor(request, recursive, regExp) ***REMOVED***
		super();
		this.request = request;
		this.userRequest = request;
		this.recursive = recursive;
		this.regExp = regExp;
		this.async = false;

		this.hadGlobalOrStickyRegExp = false;
		if(this.regExp.global || this.regExp.sticky) ***REMOVED***
			this.regExp = null;
			this.hadGlobalOrStickyRegExp = true;
		***REMOVED***

	***REMOVED***

	isEqualResource(other) ***REMOVED***
		if(!(other instanceof ContextDependency))
			return false;

		return this.request === other.request &&
			this.recursive === other.recursive &&
			this.regExp === other.regExp &&
			this.async === other.async;
	***REMOVED***

	getWarnings() ***REMOVED***
		let warnings = super.getWarnings() || [];
		if(this.critical) ***REMOVED***
			warnings.push(new CriticalDependencyWarning(this.critical));
		***REMOVED***
		if(this.hadGlobalOrStickyRegExp) ***REMOVED***
			warnings.push(new CriticalDependencyWarning("Contexts can't use RegExps with the 'g' or 'y' flags."));
		***REMOVED***
		return warnings;
	***REMOVED***

***REMOVED***

module.exports = ContextDependency;
