// webpack/lib/dependencies/HarmonyExportExpressionDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";
const NullDependency = require("./NullDependency");

class HarmonyExportExpressionDependency extends NullDependency ***REMOVED***
	constructor(originModule, range, rangeStatement) ***REMOVED***
		super();
		this.originModule = originModule;
		this.range = range;
		this.rangeStatement = rangeStatement;
	***REMOVED***

	get type() ***REMOVED***
		return "harmony export expression";
	***REMOVED***

	getExports() ***REMOVED***
		return ***REMOVED***
			exports: ["default"]
		***REMOVED***;
	***REMOVED***
***REMOVED***

HarmonyExportExpressionDependency.Template = class HarmonyExportDependencyTemplate ***REMOVED***
	apply(dep, source) ***REMOVED***
		const used = dep.originModule.isUsed("default");
		const content = this.getContent(dep.originModule, used);

		if(dep.range) ***REMOVED***
			source.replace(dep.rangeStatement[0], dep.range[0] - 1, content + "(");
			source.replace(dep.range[1], dep.rangeStatement[1] - 1, ");");
			return;
		***REMOVED***

		source.replace(dep.rangeStatement[0], dep.rangeStatement[1] - 1, content);
	***REMOVED***

	getContent(module, used) ***REMOVED***
		const exportsName = module.exportsArgument || "exports";
		if(used) ***REMOVED***
			return `/* harmony default export */ $***REMOVED***exportsName***REMOVED***[$***REMOVED***JSON.stringify(used)***REMOVED***] = `;
		***REMOVED***
		return "/* unused harmony default export */ var _unused_webpack_default_export = ";
	***REMOVED***
***REMOVED***;

module.exports = HarmonyExportExpressionDependency;
