// webpack/lib/dependencies/HarmonyImportSpecifierDependency.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";
const NullDependency = require("./NullDependency");

class HarmonyImportSpecifierDependency extends NullDependency ***REMOVED***
	constructor(importDependency, importedVar, id, name, range, strictExportPresence) ***REMOVED***
		super();
		this.importDependency = importDependency;
		this.importedVar = importedVar;
		this.id = id;
		this.name = name;
		this.range = range;
		this.strictExportPresence = strictExportPresence;
		this.namespaceObjectAsContext = false;
		this.callArgs = undefined;
		this.call = undefined;
		this.directImport = undefined;
	***REMOVED***

	get type() ***REMOVED***
		return "harmony import specifier";
	***REMOVED***

	getReference() ***REMOVED***
		if(!this.importDependency.module) return null;
		return ***REMOVED***
			module: this.importDependency.module,
			importedNames: this.id && !this.namespaceObjectAsContext ? [this.id] : true
		***REMOVED***;
	***REMOVED***

	getWarnings() ***REMOVED***
		if(this.strictExportPresence) ***REMOVED***
			return [];
		***REMOVED***
		return this._getErrors();
	***REMOVED***

	getErrors() ***REMOVED***
		if(this.strictExportPresence) ***REMOVED***
			return this._getErrors();
		***REMOVED***
		return [];
	***REMOVED***

	_getErrors() ***REMOVED***
		const importedModule = this.importDependency.module;
		if(!importedModule || !importedModule.meta || !importedModule.meta.harmonyModule) ***REMOVED***
			return;
		***REMOVED***

		if(!this.id) ***REMOVED***
			return;
		***REMOVED***

		if(importedModule.isProvided(this.id) !== false) ***REMOVED***
			return;
		***REMOVED***

		const idIsNotNameMessage = this.id !== this.name ? ` (imported as '$***REMOVED***this.name***REMOVED***')` : "";
		const errorMessage = `"export '$***REMOVED***this.id***REMOVED***'$***REMOVED***idIsNotNameMessage***REMOVED*** was not found in '$***REMOVED***this.importDependency.userRequest***REMOVED***'`;
		const err = new Error(errorMessage);
		err.hideStack = true;
		return [err];
	***REMOVED***

	updateHash(hash) ***REMOVED***
		super.updateHash(hash);
		const importedModule = this.importDependency.module;
		hash.update((importedModule && importedModule.id) + "");
		hash.update((importedModule && this.id) + "");
		hash.update((importedModule && this.importedVar) + "");
		hash.update((importedModule && this.id && importedModule.isUsed(this.id)) + "");
		hash.update((importedModule && (!importedModule.meta || importedModule.meta.harmonyModule)) + "");
		hash.update((importedModule && (importedModule.used + JSON.stringify(importedModule.usedExports))) + "");
	***REMOVED***
***REMOVED***

HarmonyImportSpecifierDependency.Template = class HarmonyImportSpecifierDependencyTemplate ***REMOVED***
	apply(dep, source) ***REMOVED***
		const content = this.getContent(dep);
		source.replace(dep.range[0], dep.range[1] - 1, content);
	***REMOVED***

	getContent(dep) ***REMOVED***
		const importedModule = dep.importDependency.module;
		const defaultImport = dep.directImport && dep.id === "default" && !(importedModule && (!importedModule.meta || importedModule.meta.harmonyModule));
		const shortHandPrefix = this.getShortHandPrefix(dep);
		const importedVar = dep.importedVar;
		const importedVarSuffix = this.getImportVarSuffix(dep, defaultImport, importedModule);

		if(dep.call && defaultImport) ***REMOVED***
			return `$***REMOVED***shortHandPrefix***REMOVED***$***REMOVED***importedVar***REMOVED***_default()`;
		***REMOVED***

		if(dep.call && dep.id) ***REMOVED***
			return `$***REMOVED***shortHandPrefix***REMOVED***Object($***REMOVED***importedVar***REMOVED***$***REMOVED***importedVarSuffix***REMOVED***)`;
		***REMOVED***

		return `$***REMOVED***shortHandPrefix***REMOVED***$***REMOVED***importedVar***REMOVED***$***REMOVED***importedVarSuffix***REMOVED***`;
	***REMOVED***

	getImportVarSuffix(dep, defaultImport, importedModule) ***REMOVED***
		if(defaultImport) ***REMOVED***
			return "_default.a";
		***REMOVED***

		if(dep.id) ***REMOVED***
			const used = importedModule ? importedModule.isUsed(dep.id) : dep.id;
			const optionalComment = dep.id !== used ? " /* " + dep.id + " */" : "";
			return `[$***REMOVED***JSON.stringify(used)***REMOVED***$***REMOVED***optionalComment***REMOVED***]`;
		***REMOVED***

		return "";
	***REMOVED***

	getShortHandPrefix(dep) ***REMOVED***
		if(!dep.shorthand) ***REMOVED***
			return "";
		***REMOVED***

		return dep.name + ": ";
	***REMOVED***
***REMOVED***;

module.exports = HarmonyImportSpecifierDependency;
