// faye-websocket/lib/faye/websocket/api.js
var Stream      = require('stream').Stream,
    util        = require('util'),
    driver      = require('websocket-driver'),
    EventTarget = require('./api/event_target'),
    Event       = require('./api/event');

var API = function(options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;
  driver.validateOptions(options, ['headers', 'extensions', 'maxLength', 'ping', 'proxy', 'tls', 'ca']);

  this.readable = this.writable = true;

  var headers = options.headers;
  if (headers) ***REMOVED***
    for (var name in headers) this._driver.setHeader(name, headers[name]);
  ***REMOVED***

  var extensions = options.extensions;
  if (extensions) ***REMOVED***
    [].concat(extensions).forEach(this._driver.addExtension, this._driver);
  ***REMOVED***

  this._ping          = options.ping;
  this._pingId        = 0;
  this.readyState     = API.CONNECTING;
  this.bufferedAmount = 0;
  this.protocol       = '';
  this.url            = this._driver.url;
  this.version        = this._driver.version;

  var self = this;

  this._driver.on('open',    function(e) ***REMOVED*** self._open() ***REMOVED***);
  this._driver.on('message', function(e) ***REMOVED*** self._receiveMessage(e.data) ***REMOVED***);
  this._driver.on('close',   function(e) ***REMOVED*** self._beginClose(e.reason, e.code) ***REMOVED***);

  this._driver.on('error', function(error) ***REMOVED***
    self._emitError(error.message);
  ***REMOVED***);
  this.on('error', function() ***REMOVED******REMOVED***);

  this._driver.messages.on('drain', function() ***REMOVED***
    self.emit('drain');
  ***REMOVED***);

  if (this._ping)
    this._pingTimer = setInterval(function() ***REMOVED***
      self._pingId += 1;
      self.ping(self._pingId.toString());
    ***REMOVED***, this._ping * 1000);

  this._configureStream();

  if (!this._proxy) ***REMOVED***
    this._stream.pipe(this._driver.io);
    this._driver.io.pipe(this._stream);
  ***REMOVED***
***REMOVED***;
util.inherits(API, Stream);

API.CONNECTING = 0;
API.OPEN       = 1;
API.CLOSING    = 2;
API.CLOSED     = 3;

API.CLOSE_TIMEOUT = 30000;

var instance = ***REMOVED***
  write: function(data) ***REMOVED***
    return this.send(data);
  ***REMOVED***,

  end: function(data) ***REMOVED***
    if (data !== undefined) this.send(data);
    this.close();
  ***REMOVED***,

  pause: function() ***REMOVED***
    return this._driver.messages.pause();
  ***REMOVED***,

  resume: function() ***REMOVED***
    return this._driver.messages.resume();
  ***REMOVED***,

  send: function(data) ***REMOVED***
    if (this.readyState > API.OPEN) return false;
    if (!(data instanceof Buffer)) data = String(data);
    return this._driver.messages.write(data);
  ***REMOVED***,

  ping: function(message, callback) ***REMOVED***
    if (this.readyState > API.OPEN) return false;
    return this._driver.ping(message, callback);
  ***REMOVED***,

  close: function(code, reason) ***REMOVED***
    if (code === undefined) code = 1000;
    if (reason === undefined) reason = '';

    if (code !== 1000 && (code < 3000 || code > 4999))
      throw new Error("Failed to execute 'close' on WebSocket: " +
                      "The code must be either 1000, or between 3000 and 4999. " +
                      code + " is neither.");

    if (this.readyState !== API.CLOSED) this.readyState = API.CLOSING;
    this._driver.close(reason, code);

    var self = this;

    this._closeTimer = setTimeout(function() ***REMOVED***
      self._beginClose('', 1006);
    ***REMOVED***, API.CLOSE_TIMEOUT);
  ***REMOVED***,

  _configureStream: function() ***REMOVED***
    var self = this;

    this._stream.setTimeout(0);
    this._stream.setNoDelay(true);

    ['close', 'end'].forEach(function(event) ***REMOVED***
      this._stream.on(event, function() ***REMOVED*** self._finalizeClose() ***REMOVED***);
    ***REMOVED***, this);

    this._stream.on('error', function(error) ***REMOVED***
      self._emitError('Network error: ' + self.url + ': ' + error.message);
      self._finalizeClose();
    ***REMOVED***);
  ***REMOVED***,

 _open: function() ***REMOVED***
    if (this.readyState !== API.CONNECTING) return;

    this.readyState = API.OPEN;
    this.protocol = this._driver.protocol || '';

    var event = new Event('open');
    event.initEvent('open', false, false);
    this.dispatchEvent(event);
  ***REMOVED***,

  _receiveMessage: function(data) ***REMOVED***
    if (this.readyState > API.OPEN) return false;

    if (this.readable) this.emit('data', data);

    var event = new Event('message', ***REMOVED***data: data***REMOVED***);
    event.initEvent('message', false, false);
    this.dispatchEvent(event);
  ***REMOVED***,

  _emitError: function(message) ***REMOVED***
    if (this.readyState >= API.CLOSING) return;

    var event = new Event('error', ***REMOVED***message: message***REMOVED***);
    event.initEvent('error', false, false);
    this.dispatchEvent(event);
  ***REMOVED***,

  _beginClose: function(reason, code) ***REMOVED***
    if (this.readyState === API.CLOSED) return;
    this.readyState = API.CLOSING;
    this._closeParams = [reason, code];

    if (this._stream) ***REMOVED***
      this._stream.destroy();
      if (!this._stream.readable) this._finalizeClose();
    ***REMOVED***
  ***REMOVED***,

  _finalizeClose: function() ***REMOVED***
    if (this.readyState === API.CLOSED) return;
    this.readyState = API.CLOSED;

    if (this._closeTimer) clearTimeout(this._closeTimer);
    if (this._pingTimer) clearInterval(this._pingTimer);
    if (this._stream) this._stream.end();

    if (this.readable) this.emit('end');
    this.readable = this.writable = false;

    var reason = this._closeParams ? this._closeParams[0] : '',
        code   = this._closeParams ? this._closeParams[1] : 1006;

    var event = new Event('close', ***REMOVED***code: code, reason: reason***REMOVED***);
    event.initEvent('close', false, false);
    this.dispatchEvent(event);
  ***REMOVED***
***REMOVED***;

for (var method in instance) API.prototype[method] = instance[method];
for (var key in EventTarget) API.prototype[key] = EventTarget[key];

module.exports = API;
