// @google-cloud/firestore/src/v1beta1/firestore_client.js
/*
 * Copyright 2017, Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * EDITING INSTRUCTIONS
 * This file was generated from the file
 * https://github.com/googleapis/googleapis/blob/master/google/firestore/v1beta1/firestore.proto,
 * and updates to that file get reflected here through a refresh process.
 * For the short term, the refresh process will only be runnable by Google
 * engineers.
 *
 * The only allowed edits are to method and file documentation. A 3-way
 * merge preserves those additions if the generated source changes.
 */
'use strict';

var configData = require('./firestore_client_config');
var extend = require('extend');
var gax = require('google-gax');
var path = require('path');

var SERVICE_ADDRESS = 'firestore.googleapis.com';

var DEFAULT_SERVICE_PORT = 443;

var CODE_GEN_NAME_VERSION = 'gapic/0.0.5';

var PAGE_DESCRIPTORS = ***REMOVED***
  listDocuments: new gax.PageDescriptor(
    'pageToken',
    'nextPageToken',
    'documents'
  ),
  listCollectionIds: new gax.PageDescriptor(
    'pageToken',
    'nextPageToken',
    'collectionIds'
  ),
***REMOVED***;

var STREAM_DESCRIPTORS = ***REMOVED***
  batchGetDocuments: new gax.StreamDescriptor(gax.StreamType.SERVER_STREAMING),
  runQuery: new gax.StreamDescriptor(gax.StreamType.SERVER_STREAMING),
  write: new gax.StreamDescriptor(gax.StreamType.BIDI_STREAMING),
  listen: new gax.StreamDescriptor(gax.StreamType.BIDI_STREAMING),
***REMOVED***;

/*!
 * The scopes needed to make gRPC calls to all of the methods defined in
 * this service.
 */
var ALL_SCOPES = [
  'https://www.googleapis.com/auth/cloud-platform',
  'https://www.googleapis.com/auth/datastore',
];

/**
 * The Cloud Firestore service.
 *
 * This service exposes several types of comparable timestamps:
 *
 * *    `create_time` - The time at which a document was created. Changes only
 *      when a document is deleted, then re-created. Increases in a strict
 *       monotonic fashion.
 * *    `update_time` - The time at which a document was last updated. Changes
 *      every time a document is modified. Does not change when a write results
 *      in no modifications. Increases in a strict monotonic fashion.
 * *    `read_time` - The time at which a particular state was observed. Used
 *      to denote a consistent snapshot of the database or the time at which a
 *      Document was observed to not exist.
 * *    `commit_time` - The time at which the writes in a transaction were
 *      committed. Any read with an equal or greater `read_time` is guaranteed
 *      to see the effects of the transaction.
 *
 *
 * @class
 */
function FirestoreClient(gaxGrpc, loadedProtos, opts) ***REMOVED***
  opts = extend(
    ***REMOVED***
      servicePath: SERVICE_ADDRESS,
      port: DEFAULT_SERVICE_PORT,
      clientConfig: ***REMOVED******REMOVED***,
    ***REMOVED***,
    opts
  );

  var googleApiClient = ['gl-node/' + process.versions.node];
  if (opts.libName && opts.libVersion) ***REMOVED***
    googleApiClient.push(opts.libName + '/' + opts.libVersion);
  ***REMOVED***
  googleApiClient.push(
    CODE_GEN_NAME_VERSION,
    'gax/' + gax.version,
    'grpc/' + gaxGrpc.grpcVersion
  );

  var defaults = gaxGrpc.constructSettings(
    'google.firestore.v1beta1.Firestore',
    configData,
    opts.clientConfig,
    ***REMOVED***'x-goog-api-client': googleApiClient.join(' ')***REMOVED***
  );

  var self = this;

  this.auth = gaxGrpc.auth;
  var firestoreStub = gaxGrpc.createStub(
    loadedProtos.google.firestore.v1beta1.Firestore,
    opts
  );
  var firestoreStubMethods = [
    'getDocument',
    'listDocuments',
    'createDocument',
    'updateDocument',
    'deleteDocument',
    'batchGetDocuments',
    'beginTransaction',
    'commit',
    'rollback',
    'runQuery',
    'write',
    'listen',
    'listCollectionIds',
  ];
  firestoreStubMethods.forEach(function(methodName) ***REMOVED***
    self['_' + methodName] = gax.createApiCall(
      firestoreStub.then(function(firestoreStub) ***REMOVED***
        return function() ***REMOVED***
          var args = Array.prototype.slice.call(arguments, 0);
          return firestoreStub[methodName].apply(firestoreStub, args);
        ***REMOVED***;
      ***REMOVED***),
      defaults[methodName],
      PAGE_DESCRIPTORS[methodName] || STREAM_DESCRIPTORS[methodName]
    );
  ***REMOVED***);
***REMOVED***

// Path templates

var DATABASE_ROOT_PATH_TEMPLATE = new gax.PathTemplate(
  'projects/***REMOVED***project***REMOVED***/databases/***REMOVED***database***REMOVED***'
);

var DOCUMENT_ROOT_PATH_TEMPLATE = new gax.PathTemplate(
  'projects/***REMOVED***project***REMOVED***/databases/***REMOVED***database***REMOVED***/documents'
);

var DOCUMENT_PATH_PATH_TEMPLATE = new gax.PathTemplate(
  'projects/***REMOVED***project***REMOVED***/databases/***REMOVED***database***REMOVED***/documents/***REMOVED***document_path=*****REMOVED***'
);

var ANY_PATH_PATH_TEMPLATE = new gax.PathTemplate(
  'projects/***REMOVED***project***REMOVED***/databases/***REMOVED***database***REMOVED***/documents/***REMOVED***document***REMOVED***/***REMOVED***any_path=*****REMOVED***'
);

/**
 * Returns a fully-qualified database_root resource name string.
 * @param ***REMOVED***String***REMOVED*** project
 * @param ***REMOVED***String***REMOVED*** database
 * @returns ***REMOVED***String***REMOVED***
 */
FirestoreClient.prototype.databaseRootPath = function(project, database) ***REMOVED***
  return DATABASE_ROOT_PATH_TEMPLATE.render(***REMOVED***
    project: project,
    database: database,
  ***REMOVED***);
***REMOVED***;

/**
 * Returns a fully-qualified document_root resource name string.
 * @param ***REMOVED***String***REMOVED*** project
 * @param ***REMOVED***String***REMOVED*** database
 * @returns ***REMOVED***String***REMOVED***
 */
FirestoreClient.prototype.documentRootPath = function(project, database) ***REMOVED***
  return DOCUMENT_ROOT_PATH_TEMPLATE.render(***REMOVED***
    project: project,
    database: database,
  ***REMOVED***);
***REMOVED***;

/**
 * Returns a fully-qualified document_path resource name string.
 * @param ***REMOVED***String***REMOVED*** project
 * @param ***REMOVED***String***REMOVED*** database
 * @param ***REMOVED***String***REMOVED*** documentPath
 * @returns ***REMOVED***String***REMOVED***
 */
FirestoreClient.prototype.documentPathPath = function(
  project,
  database,
  documentPath
) ***REMOVED***
  return DOCUMENT_PATH_PATH_TEMPLATE.render(***REMOVED***
    project: project,
    database: database,
    document_path: documentPath,
  ***REMOVED***);
***REMOVED***;

/**
 * Returns a fully-qualified any_path resource name string.
 * @param ***REMOVED***String***REMOVED*** project
 * @param ***REMOVED***String***REMOVED*** database
 * @param ***REMOVED***String***REMOVED*** document
 * @param ***REMOVED***String***REMOVED*** anyPath
 * @returns ***REMOVED***String***REMOVED***
 */
FirestoreClient.prototype.anyPathPath = function(
  project,
  database,
  document,
  anyPath
) ***REMOVED***
  return ANY_PATH_PATH_TEMPLATE.render(***REMOVED***
    project: project,
    database: database,
    document: document,
    any_path: anyPath,
  ***REMOVED***);
***REMOVED***;

/**
 * Parses the databaseRootName from a database_root resource.
 * @param ***REMOVED***String***REMOVED*** databaseRootName
 *   A fully-qualified path representing a database_root resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the project.
 */
FirestoreClient.prototype.matchProjectFromDatabaseRootName = function(
  databaseRootName
) ***REMOVED***
  return DATABASE_ROOT_PATH_TEMPLATE.match(databaseRootName).project;
***REMOVED***;

/**
 * Parses the databaseRootName from a database_root resource.
 * @param ***REMOVED***String***REMOVED*** databaseRootName
 *   A fully-qualified path representing a database_root resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the database.
 */
FirestoreClient.prototype.matchDatabaseFromDatabaseRootName = function(
  databaseRootName
) ***REMOVED***
  return DATABASE_ROOT_PATH_TEMPLATE.match(databaseRootName).database;
***REMOVED***;

/**
 * Parses the documentRootName from a document_root resource.
 * @param ***REMOVED***String***REMOVED*** documentRootName
 *   A fully-qualified path representing a document_root resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the project.
 */
FirestoreClient.prototype.matchProjectFromDocumentRootName = function(
  documentRootName
) ***REMOVED***
  return DOCUMENT_ROOT_PATH_TEMPLATE.match(documentRootName).project;
***REMOVED***;

/**
 * Parses the documentRootName from a document_root resource.
 * @param ***REMOVED***String***REMOVED*** documentRootName
 *   A fully-qualified path representing a document_root resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the database.
 */
FirestoreClient.prototype.matchDatabaseFromDocumentRootName = function(
  documentRootName
) ***REMOVED***
  return DOCUMENT_ROOT_PATH_TEMPLATE.match(documentRootName).database;
***REMOVED***;

/**
 * Parses the documentPathName from a document_path resource.
 * @param ***REMOVED***String***REMOVED*** documentPathName
 *   A fully-qualified path representing a document_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the project.
 */
FirestoreClient.prototype.matchProjectFromDocumentPathName = function(
  documentPathName
) ***REMOVED***
  return DOCUMENT_PATH_PATH_TEMPLATE.match(documentPathName).project;
***REMOVED***;

/**
 * Parses the documentPathName from a document_path resource.
 * @param ***REMOVED***String***REMOVED*** documentPathName
 *   A fully-qualified path representing a document_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the database.
 */
FirestoreClient.prototype.matchDatabaseFromDocumentPathName = function(
  documentPathName
) ***REMOVED***
  return DOCUMENT_PATH_PATH_TEMPLATE.match(documentPathName).database;
***REMOVED***;

/**
 * Parses the documentPathName from a document_path resource.
 * @param ***REMOVED***String***REMOVED*** documentPathName
 *   A fully-qualified path representing a document_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the document_path.
 */
FirestoreClient.prototype.matchDocumentPathFromDocumentPathName = function(
  documentPathName
) ***REMOVED***
  return DOCUMENT_PATH_PATH_TEMPLATE.match(documentPathName).document_path;
***REMOVED***;

/**
 * Parses the anyPathName from a any_path resource.
 * @param ***REMOVED***String***REMOVED*** anyPathName
 *   A fully-qualified path representing a any_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the project.
 */
FirestoreClient.prototype.matchProjectFromAnyPathName = function(anyPathName) ***REMOVED***
  return ANY_PATH_PATH_TEMPLATE.match(anyPathName).project;
***REMOVED***;

/**
 * Parses the anyPathName from a any_path resource.
 * @param ***REMOVED***String***REMOVED*** anyPathName
 *   A fully-qualified path representing a any_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the database.
 */
FirestoreClient.prototype.matchDatabaseFromAnyPathName = function(anyPathName) ***REMOVED***
  return ANY_PATH_PATH_TEMPLATE.match(anyPathName).database;
***REMOVED***;

/**
 * Parses the anyPathName from a any_path resource.
 * @param ***REMOVED***String***REMOVED*** anyPathName
 *   A fully-qualified path representing a any_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the document.
 */
FirestoreClient.prototype.matchDocumentFromAnyPathName = function(anyPathName) ***REMOVED***
  return ANY_PATH_PATH_TEMPLATE.match(anyPathName).document;
***REMOVED***;

/**
 * Parses the anyPathName from a any_path resource.
 * @param ***REMOVED***String***REMOVED*** anyPathName
 *   A fully-qualified path representing a any_path resources.
 * @returns ***REMOVED***String***REMOVED*** - A string representing the any_path.
 */
FirestoreClient.prototype.matchAnyPathFromAnyPathName = function(anyPathName) ***REMOVED***
  return ANY_PATH_PATH_TEMPLATE.match(anyPathName).any_path;
***REMOVED***;

/**
 * Get the project ID used by this class.
 * @param ***REMOVED***function(Error, string)***REMOVED*** callback - the callback to be called with
 *   the current project Id.
 */
FirestoreClient.prototype.getProjectId = function(callback) ***REMOVED***
  return this.auth.getProjectId(callback);
***REMOVED***;

// Service calls

/**
 * Gets a single document.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The resource name of the Document to get. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 * @param ***REMOVED***Object=***REMOVED*** request.mask
 *   The fields to return. If not set, returns all fields.
 *
 *   If the document has a field that is not present in this mask, that field
 *   will not be returned in the response.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***string=***REMOVED*** request.transaction
 *   Reads the document in a transaction.
 * @param ***REMOVED***Object=***REMOVED*** request.readTime
 *   Reads the version of the document at the given time.
 *   This may not be older than 60 seconds.
 *
 *   This object should have the same structure as [Timestamp]***REMOVED***@link google.protobuf.Timestamp***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedName = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * client.getDocument(***REMOVED***name: formattedName***REMOVED***).then(function(responses) ***REMOVED***
 *     var response = responses[0];
 *     // doThingsWith(response)
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.getDocument = function(request, options, callback) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._getDocument(request, options, callback);
***REMOVED***;

/**
 * Lists documents.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.parent
 *   The parent resource name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents` or
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 *   For example:
 *   `projects/my-project/databases/my-database/documents` or
 *   `projects/my-project/databases/my-database/documents/chatrooms/my-chatroom`
 * @param ***REMOVED***string***REMOVED*** request.collectionId
 *   The collection ID, relative to `parent`, to list. For example: `chatrooms`
 *   or `messages`.
 * @param ***REMOVED***number=***REMOVED*** request.pageSize
 *   The maximum number of resources contained in the underlying API
 *   response. If page streaming is performed per-resource, this
 *   parameter does not affect the return value. If page streaming is
 *   performed per-page, this determines the maximum number of
 *   resources in a page.
 * @param ***REMOVED***string=***REMOVED*** request.orderBy
 *   The order to sort results by. For example: `priority desc, name`.
 * @param ***REMOVED***Object=***REMOVED*** request.mask
 *   The fields to return. If not set, returns all fields.
 *
 *   If a document has a field that is not present in this mask, that field
 *   will not be returned in the response.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***string=***REMOVED*** request.transaction
 *   Reads documents in a transaction.
 * @param ***REMOVED***Object=***REMOVED*** request.readTime
 *   Reads documents as they were at the given time.
 *   This may not be older than 60 seconds.
 *
 *   This object should have the same structure as [Timestamp]***REMOVED***@link google.protobuf.Timestamp***REMOVED***
 * @param ***REMOVED***boolean=***REMOVED*** request.showMissing
 *   If the list should show missing documents. A missing document is a
 *   document that does not exist but has sub-documents. These documents will
 *   be returned with a key but will not have fields, ***REMOVED***@link Document.create_time***REMOVED***,
 *   or ***REMOVED***@link Document.update_time***REMOVED*** set.
 *
 *   Requests with `show_missing` may not specify `where` or
 *   `order_by`.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Array, ?Object, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is Array of [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 *
 *   When autoPaginate: false is specified through options, it contains the result
 *   in a single response. If the response indicates the next page exists, the third
 *   parameter is set to be used for the next request object. The fourth parameter keeps
 *   the raw response object of an object representing [ListDocumentsResponse]***REMOVED***@link google.firestore.v1beta1.ListDocumentsResponse***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is Array of [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 *
 *   When autoPaginate: false is specified through options, the array has three elements.
 *   The first element is Array of [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED*** in a single response.
 *   The second element is the next request object if the response
 *   indicates the next page exists, or null. The third element is
 *   an object representing [ListDocumentsResponse]***REMOVED***@link google.firestore.v1beta1.ListDocumentsResponse***REMOVED***.
 *
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * // Iterate over all elements.
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * var collectionId = '';
 * var request = ***REMOVED***
 *     parent: formattedParent,
 *     collectionId: collectionId
 * ***REMOVED***;
 *
 * client.listDocuments(request).then(function(responses) ***REMOVED***
 *     var resources = responses[0];
 *     for (var i = 0; i < resources.length; ++i) ***REMOVED***
 *         // doThingsWith(resources[i])
 *     ***REMOVED***
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 *
 * // Or obtain the paged response.
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * var collectionId = '';
 * var request = ***REMOVED***
 *     parent: formattedParent,
 *     collectionId: collectionId
 * ***REMOVED***;
 *
 *
 * var options = ***REMOVED***autoPaginate: false***REMOVED***;
 * function callback(responses) ***REMOVED***
 *     // The actual resources in a response.
 *     var resources = responses[0];
 *     // The next request if the response shows there's more responses.
 *     var nextRequest = responses[1];
 *     // The actual response object, if necessary.
 *     // var rawResponse = responses[2];
 *     for (var i = 0; i < resources.length; ++i) ***REMOVED***
 *         // doThingsWith(resources[i]);
 *     ***REMOVED***
 *     if (nextRequest) ***REMOVED***
 *         // Fetch the next page.
 *         return client.listDocuments(nextRequest, options).then(callback);
 *     ***REMOVED***
 * ***REMOVED***
 * client.listDocuments(request, options)
 *     .then(callback)
 *     .catch(function(err) ***REMOVED***
 *         console.error(err);
 *     ***REMOVED***);
 */
FirestoreClient.prototype.listDocuments = function(request, options, callback) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._listDocuments(request, options, callback);
***REMOVED***;

/**
 * Equivalent to ***REMOVED***@link listDocuments***REMOVED***, but returns a NodeJS Stream object.
 *
 * This fetches the paged responses for ***REMOVED***@link listDocuments***REMOVED*** continuously
 * and invokes the callback registered for 'data' event for each element in the
 * responses.
 *
 * The returned object has 'end' method when no more elements are required.
 *
 * autoPaginate option will be ignored.
 *
 * @see ***REMOVED***@link https://nodejs.org/api/stream.html***REMOVED***
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.parent
 *   The parent resource name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents` or
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 *   For example:
 *   `projects/my-project/databases/my-database/documents` or
 *   `projects/my-project/databases/my-database/documents/chatrooms/my-chatroom`
 * @param ***REMOVED***string***REMOVED*** request.collectionId
 *   The collection ID, relative to `parent`, to list. For example: `chatrooms`
 *   or `messages`.
 * @param ***REMOVED***number=***REMOVED*** request.pageSize
 *   The maximum number of resources contained in the underlying API
 *   response. If page streaming is performed per-resource, this
 *   parameter does not affect the return value. If page streaming is
 *   performed per-page, this determines the maximum number of
 *   resources in a page.
 * @param ***REMOVED***string=***REMOVED*** request.orderBy
 *   The order to sort results by. For example: `priority desc, name`.
 * @param ***REMOVED***Object=***REMOVED*** request.mask
 *   The fields to return. If not set, returns all fields.
 *
 *   If a document has a field that is not present in this mask, that field
 *   will not be returned in the response.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***string=***REMOVED*** request.transaction
 *   Reads documents in a transaction.
 * @param ***REMOVED***Object=***REMOVED*** request.readTime
 *   Reads documents as they were at the given time.
 *   This may not be older than 60 seconds.
 *
 *   This object should have the same structure as [Timestamp]***REMOVED***@link google.protobuf.Timestamp***REMOVED***
 * @param ***REMOVED***boolean=***REMOVED*** request.showMissing
 *   If the list should show missing documents. A missing document is a
 *   document that does not exist but has sub-documents. These documents will
 *   be returned with a key but will not have fields, ***REMOVED***@link Document.create_time***REMOVED***,
 *   or ***REMOVED***@link Document.update_time***REMOVED*** set.
 *
 *   Requests with `show_missing` may not specify `where` or
 *   `order_by`.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @returns ***REMOVED***Stream***REMOVED***
 *   An object stream which emits an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED*** on 'data' event.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * var collectionId = '';
 * var request = ***REMOVED***
 *     parent: formattedParent,
 *     collectionId: collectionId
 * ***REMOVED***;
 * client.listDocumentsStream(request)
 * .on('data', function(element) ***REMOVED***
 *     // doThingsWith(element)
 * ***REMOVED***).on('error', function(err) ***REMOVED***
 *     console.log(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.listDocumentsStream = function(request, options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return PAGE_DESCRIPTORS.listDocuments.createStream(
    this._listDocuments,
    request,
    options
  );
***REMOVED***;

/**
 * Creates a new document.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.parent
 *   The parent resource. For example:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents` or
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/chatrooms/***REMOVED***chatroom_id***REMOVED***`
 * @param ***REMOVED***string***REMOVED*** request.collectionId
 *   The collection ID, relative to `parent`, to list. For example: `chatrooms`.
 * @param ***REMOVED***string***REMOVED*** request.documentId
 *   The client-assigned document ID to use for this document.
 *
 *   Optional. If not specified, an ID will be assigned by the service.
 * @param ***REMOVED***Object***REMOVED*** request.document
 *   The document to create. `name` must not be set.
 *
 *   This object should have the same structure as [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** request.mask
 *   The fields to return. If not set, returns all fields.
 *
 *   If the document has a field that is not present in this mask, that field
 *   will not be returned in the response.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * var collectionId = '';
 * var documentId = '';
 * var document = ***REMOVED******REMOVED***;
 * var request = ***REMOVED***
 *     parent: formattedParent,
 *     collectionId: collectionId,
 *     documentId: documentId,
 *     document: document
 * ***REMOVED***;
 * client.createDocument(request).then(function(responses) ***REMOVED***
 *     var response = responses[0];
 *     // doThingsWith(response)
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.createDocument = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._createDocument(request, options, callback);
***REMOVED***;

/**
 * Updates or inserts a document.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***Object***REMOVED*** request.document
 *   The updated document.
 *   Creates the document if it does not already exist.
 *
 *   This object should have the same structure as [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***
 * @param ***REMOVED***Object***REMOVED*** request.updateMask
 *   The fields to update.
 *   None of the field paths in the mask may contain a reserved name.
 *
 *   If the document exists on the server and has fields not referenced in the
 *   mask, they are left unchanged.
 *   Fields referenced in the mask, but not present in the input document, are
 *   deleted from the document on the server.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** request.mask
 *   The fields to return. If not set, returns all fields.
 *
 *   If the document has a field that is not present in this mask, that field
 *   will not be returned in the response.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** request.currentDocument
 *   An optional precondition on the document.
 *   The request will fail if this is set and not met by the target document.
 *
 *   This object should have the same structure as [Precondition]***REMOVED***@link google.firestore.v1beta1.Precondition***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is an object representing [Document]***REMOVED***@link google.firestore.v1beta1.Document***REMOVED***.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var document = ***REMOVED******REMOVED***;
 * var updateMask = ***REMOVED******REMOVED***;
 * var request = ***REMOVED***
 *     document: document,
 *     updateMask: updateMask
 * ***REMOVED***;
 * client.updateDocument(request).then(function(responses) ***REMOVED***
 *     var response = responses[0];
 *     // doThingsWith(response)
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.updateDocument = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._updateDocument(request, options, callback);
***REMOVED***;

/**
 * Deletes a document.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.name
 *   The resource name of the Document to delete. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 * @param ***REMOVED***Object=***REMOVED*** request.currentDocument
 *   An optional precondition on the document.
 *   The request will fail if this is set and not met by the target document.
 *
 *   This object should have the same structure as [Precondition]***REMOVED***@link google.firestore.v1beta1.Precondition***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves when API call finishes.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedName = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * client.deleteDocument(***REMOVED***name: formattedName***REMOVED***).catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.deleteDocument = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._deleteDocument(request, options, callback);
***REMOVED***;

/**
 * Gets multiple documents.
 *
 * Documents returned by this method are not guaranteed to be returned in the
 * same order that they were requested.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.database
 *   The database name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***`.
 * @param ***REMOVED***string[]***REMOVED*** request.documents
 *   The names of the documents to retrieve. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 *   The request will fail if any of the document is not a child resource of the
 *   given `database`. Duplicate names will be elided.
 * @param ***REMOVED***Object=***REMOVED*** request.mask
 *   The fields to return. If not set, returns all fields.
 *
 *   If a document has a field that is not present in this mask, that field will
 *   not be returned in the response.
 *
 *   This object should have the same structure as [DocumentMask]***REMOVED***@link google.firestore.v1beta1.DocumentMask***REMOVED***
 * @param ***REMOVED***string=***REMOVED*** request.transaction
 *   Reads documents in a transaction.
 * @param ***REMOVED***Object=***REMOVED*** request.newTransaction
 *   Starts a new transaction and reads the documents.
 *   Defaults to a read-only transaction.
 *   The new transaction ID will be returned as the first response in the
 *   stream.
 *
 *   This object should have the same structure as [TransactionOptions]***REMOVED***@link google.firestore.v1beta1.TransactionOptions***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** request.readTime
 *   Reads documents as they were at the given time.
 *   This may not be older than 60 seconds.
 *
 *   This object should have the same structure as [Timestamp]***REMOVED***@link google.protobuf.Timestamp***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @returns ***REMOVED***Stream***REMOVED***
 *   An object stream which emits [BatchGetDocumentsResponse]***REMOVED***@link google.firestore.v1beta1.BatchGetDocumentsResponse***REMOVED*** on 'data' event.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedDatabase = client.databaseRootPath("[PROJECT]", "[DATABASE]");
 * var documents = [];
 * var request = ***REMOVED***
 *     database: formattedDatabase,
 *     documents: documents
 * ***REMOVED***;
 * client.batchGetDocuments(request).on('data', function(response) ***REMOVED***
 *   // doThingsWith(response)
 * ***REMOVED***);
 */
FirestoreClient.prototype.batchGetDocuments = function(request, options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._batchGetDocuments(request, options);
***REMOVED***;

/**
 * Starts a new transaction.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.database
 *   The database name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***`.
 * @param ***REMOVED***Object=***REMOVED*** request.options
 *   The options for the transaction.
 *   Defaults to a read-write transaction.
 *
 *   This object should have the same structure as [TransactionOptions]***REMOVED***@link google.firestore.v1beta1.TransactionOptions***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is an object representing [BeginTransactionResponse]***REMOVED***@link google.firestore.v1beta1.BeginTransactionResponse***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is an object representing [BeginTransactionResponse]***REMOVED***@link google.firestore.v1beta1.BeginTransactionResponse***REMOVED***.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedDatabase = client.databaseRootPath("[PROJECT]", "[DATABASE]");
 * client.beginTransaction(***REMOVED***database: formattedDatabase***REMOVED***).then(function(responses) ***REMOVED***
 *     var response = responses[0];
 *     // doThingsWith(response)
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.beginTransaction = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._beginTransaction(request, options, callback);
***REMOVED***;

/**
 * Commits a transaction, while optionally updating documents.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.database
 *   The database name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***`.
 * @param ***REMOVED***Object[]***REMOVED*** request.writes
 *   The writes to apply.
 *
 *   Always executed atomically and in order.
 *
 *   This object should have the same structure as [Write]***REMOVED***@link google.firestore.v1beta1.Write***REMOVED***
 * @param ***REMOVED***string=***REMOVED*** request.transaction
 *   If set, applies all writes in this transaction, and commits it.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is an object representing [CommitResponse]***REMOVED***@link google.firestore.v1beta1.CommitResponse***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is an object representing [CommitResponse]***REMOVED***@link google.firestore.v1beta1.CommitResponse***REMOVED***.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedDatabase = client.databaseRootPath("[PROJECT]", "[DATABASE]");
 * var writes = [];
 * var request = ***REMOVED***
 *     database: formattedDatabase,
 *     writes: writes
 * ***REMOVED***;
 * client.commit(request).then(function(responses) ***REMOVED***
 *     var response = responses[0];
 *     // doThingsWith(response)
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.commit = function(request, options, callback) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._commit(request, options, callback);
***REMOVED***;

/**
 * Rolls back a transaction.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.database
 *   The database name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***`.
 * @param ***REMOVED***string***REMOVED*** request.transaction
 *   The transaction to roll back.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves when API call finishes.
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedDatabase = client.databaseRootPath("[PROJECT]", "[DATABASE]");
 * var transaction = '';
 * var request = ***REMOVED***
 *     database: formattedDatabase,
 *     transaction: transaction
 * ***REMOVED***;
 * client.rollback(request).catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.rollback = function(request, options, callback) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._rollback(request, options, callback);
***REMOVED***;

/**
 * Runs a query.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.parent
 *   The parent resource name. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents` or
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 *   For example:
 *   `projects/my-project/databases/my-database/documents` or
 *   `projects/my-project/databases/my-database/documents/chatrooms/my-chatroom`
 * @param ***REMOVED***Object=***REMOVED*** request.structuredQuery
 *   A structured query.
 *
 *   This object should have the same structure as [StructuredQuery]***REMOVED***@link google.firestore.v1beta1.StructuredQuery***REMOVED***
 * @param ***REMOVED***string=***REMOVED*** request.transaction
 *   Reads documents in a transaction.
 * @param ***REMOVED***Object=***REMOVED*** request.newTransaction
 *   Starts a new transaction and reads the documents.
 *   Defaults to a read-only transaction.
 *   The new transaction ID will be returned as the first response in the
 *   stream.
 *
 *   This object should have the same structure as [TransactionOptions]***REMOVED***@link google.firestore.v1beta1.TransactionOptions***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** request.readTime
 *   Reads documents as they were at the given time.
 *   This may not be older than 60 seconds.
 *
 *   This object should have the same structure as [Timestamp]***REMOVED***@link google.protobuf.Timestamp***REMOVED***
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @returns ***REMOVED***Stream***REMOVED***
 *   An object stream which emits [RunQueryResponse]***REMOVED***@link google.firestore.v1beta1.RunQueryResponse***REMOVED*** on 'data' event.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * client.runQuery(***REMOVED***parent: formattedParent***REMOVED***).on('data', function(response) ***REMOVED***
 *   // doThingsWith(response)
 * ***REMOVED***);
 */
FirestoreClient.prototype.runQuery = function(request, options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._runQuery(request, options);
***REMOVED***;

/**
 * Streams batches of document updates and deletes, in order.
 *
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @returns ***REMOVED***Stream***REMOVED***
 *   An object stream which is both readable and writable. It accepts objects
 *   representing [WriteRequest]***REMOVED***@link google.firestore.v1beta1.WriteRequest***REMOVED*** for write() method, and
 *   will emit objects representing [WriteResponse]***REMOVED***@link google.firestore.v1beta1.WriteResponse***REMOVED*** on 'data' event asynchronously.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var stream = client.write().on('data', function(response) ***REMOVED***
 *     // doThingsWith(response)
 * ***REMOVED***);
 * var formattedDatabase = client.databaseRootPath("[PROJECT]", "[DATABASE]");
 * var request = ***REMOVED***
 *     database : formattedDatabase
 * ***REMOVED***;
 * // Write request objects.
 * stream.write(request);
 */
FirestoreClient.prototype.write = function(options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._write(options);
***REMOVED***;

/**
 * Listens to changes.
 *
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @returns ***REMOVED***Stream***REMOVED***
 *   An object stream which is both readable and writable. It accepts objects
 *   representing [ListenRequest]***REMOVED***@link google.firestore.v1beta1.ListenRequest***REMOVED*** for write() method, and
 *   will emit objects representing [ListenResponse]***REMOVED***@link google.firestore.v1beta1.ListenResponse***REMOVED*** on 'data' event asynchronously.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var stream = client.listen().on('data', function(response) ***REMOVED***
 *     // doThingsWith(response)
 * ***REMOVED***);
 * var formattedDatabase = client.databaseRootPath("[PROJECT]", "[DATABASE]");
 * var request = ***REMOVED***
 *     database : formattedDatabase
 * ***REMOVED***;
 * // Write request objects.
 * stream.write(request);
 */
FirestoreClient.prototype.listen = function(options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._listen(options);
***REMOVED***;

/**
 * Lists all the collection IDs underneath a document.
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.parent
 *   The parent document. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 *   For example:
 *   `projects/my-project/databases/my-database/documents/chatrooms/my-chatroom`
 * @param ***REMOVED***number=***REMOVED*** request.pageSize
 *   The maximum number of resources contained in the underlying API
 *   response. If page streaming is performed per-resource, this
 *   parameter does not affect the return value. If page streaming is
 *   performed per-page, this determines the maximum number of
 *   resources in a page.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @param ***REMOVED***function(?Error, ?Array, ?Object, ?Object)=***REMOVED*** callback
 *   The function which will be called with the result of the API call.
 *
 *   The second parameter to the callback is Array of string.
 *
 *   When autoPaginate: false is specified through options, it contains the result
 *   in a single response. If the response indicates the next page exists, the third
 *   parameter is set to be used for the next request object. The fourth parameter keeps
 *   the raw response object of an object representing [ListCollectionIdsResponse]***REMOVED***@link google.firestore.v1beta1.ListCollectionIdsResponse***REMOVED***.
 * @returns ***REMOVED***Promise***REMOVED*** - The promise which resolves to an array.
 *   The first element of the array is Array of string.
 *
 *   When autoPaginate: false is specified through options, the array has three elements.
 *   The first element is Array of string in a single response.
 *   The second element is the next request object if the response
 *   indicates the next page exists, or null. The third element is
 *   an object representing [ListCollectionIdsResponse]***REMOVED***@link google.firestore.v1beta1.ListCollectionIdsResponse***REMOVED***.
 *
 *   The promise has a method named "cancel" which cancels the ongoing API call.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * // Iterate over all elements.
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 *
 * client.listCollectionIds(***REMOVED***parent: formattedParent***REMOVED***).then(function(responses) ***REMOVED***
 *     var resources = responses[0];
 *     for (var i = 0; i < resources.length; ++i) ***REMOVED***
 *         // doThingsWith(resources[i])
 *     ***REMOVED***
 * ***REMOVED***)
 * .catch(function(err) ***REMOVED***
 *     console.error(err);
 * ***REMOVED***);
 *
 * // Or obtain the paged response.
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 *
 *
 * var options = ***REMOVED***autoPaginate: false***REMOVED***;
 * function callback(responses) ***REMOVED***
 *     // The actual resources in a response.
 *     var resources = responses[0];
 *     // The next request if the response shows there's more responses.
 *     var nextRequest = responses[1];
 *     // The actual response object, if necessary.
 *     // var rawResponse = responses[2];
 *     for (var i = 0; i < resources.length; ++i) ***REMOVED***
 *         // doThingsWith(resources[i]);
 *     ***REMOVED***
 *     if (nextRequest) ***REMOVED***
 *         // Fetch the next page.
 *         return client.listCollectionIds(nextRequest, options).then(callback);
 *     ***REMOVED***
 * ***REMOVED***
 * client.listCollectionIds(***REMOVED***parent: formattedParent***REMOVED***, options)
 *     .then(callback)
 *     .catch(function(err) ***REMOVED***
 *         console.error(err);
 *     ***REMOVED***);
 */
FirestoreClient.prototype.listCollectionIds = function(
  request,
  options,
  callback
) ***REMOVED***
  if (options instanceof Function && callback === undefined) ***REMOVED***
    callback = options;
    options = ***REMOVED******REMOVED***;
  ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return this._listCollectionIds(request, options, callback);
***REMOVED***;

/**
 * Equivalent to ***REMOVED***@link listCollectionIds***REMOVED***, but returns a NodeJS Stream object.
 *
 * This fetches the paged responses for ***REMOVED***@link listCollectionIds***REMOVED*** continuously
 * and invokes the callback registered for 'data' event for each element in the
 * responses.
 *
 * The returned object has 'end' method when no more elements are required.
 *
 * autoPaginate option will be ignored.
 *
 * @see ***REMOVED***@link https://nodejs.org/api/stream.html***REMOVED***
 *
 * @param ***REMOVED***Object***REMOVED*** request
 *   The request object that will be sent.
 * @param ***REMOVED***string***REMOVED*** request.parent
 *   The parent document. In the format:
 *   `projects/***REMOVED***project_id***REMOVED***/databases/***REMOVED***database_id***REMOVED***/documents/***REMOVED***document_path***REMOVED***`.
 *   For example:
 *   `projects/my-project/databases/my-database/documents/chatrooms/my-chatroom`
 * @param ***REMOVED***number=***REMOVED*** request.pageSize
 *   The maximum number of resources contained in the underlying API
 *   response. If page streaming is performed per-resource, this
 *   parameter does not affect the return value. If page streaming is
 *   performed per-page, this determines the maximum number of
 *   resources in a page.
 * @param ***REMOVED***Object=***REMOVED*** options
 *   Optional parameters. You can override the default settings for this call, e.g, timeout,
 *   retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link https://googleapis.github.io/gax-nodejs/global.html#CallOptions***REMOVED*** for the details.
 * @returns ***REMOVED***Stream***REMOVED***
 *   An object stream which emits a string on 'data' event.
 *
 * @example
 *
 * var firestore = require('firestore.v1beta1');
 *
 * var client = firestore.v1beta1(***REMOVED***
 *   // optional auth parameters.
 * ***REMOVED***);
 *
 * var formattedParent = client.anyPathPath("[PROJECT]", "[DATABASE]", "[DOCUMENT]", "[ANY_PATH]");
 * client.listCollectionIdsStream(***REMOVED***parent: formattedParent***REMOVED***)
 * .on('data', function(element) ***REMOVED***
 *     // doThingsWith(element)
 * ***REMOVED***).on('error', function(err) ***REMOVED***
 *     console.log(err);
 * ***REMOVED***);
 */
FirestoreClient.prototype.listCollectionIdsStream = function(request, options) ***REMOVED***
  if (options === undefined) ***REMOVED***
    options = ***REMOVED******REMOVED***;
  ***REMOVED***

  return PAGE_DESCRIPTORS.listCollectionIds.createStream(
    this._listCollectionIds,
    request,
    options
  );
***REMOVED***;

/**
 * @class
 * @param ***REMOVED*******REMOVED*** gaxGrpc
 */
function FirestoreClientBuilder(gaxGrpc) ***REMOVED***
  if (!(this instanceof FirestoreClientBuilder)) ***REMOVED***
    return new FirestoreClientBuilder(gaxGrpc);
  ***REMOVED***

  var firestoreStubProtos = gaxGrpc.loadProto(
    path.join(__dirname, '..', '..', 'protos'),
    'google/firestore/v1beta1/firestore.proto'
  );
  extend(this, firestoreStubProtos.google.firestore.v1beta1);

  /**
   * Build a new instance of ***REMOVED***@link FirestoreClient***REMOVED***.
   *
   * @method FirestoreClientBuilder#firestoreClient
   * @param ***REMOVED***Object=***REMOVED*** opts - The optional parameters.
   * @param ***REMOVED***String=***REMOVED*** opts.servicePath
   *   The domain name of the API remote host.
   * @param ***REMOVED***number=***REMOVED*** opts.port
   *   The port on which to connect to the remote host.
   * @param ***REMOVED***grpc.ClientCredentials=***REMOVED*** opts.sslCreds
   *   A ClientCredentials for use with an SSL-enabled channel.
   * @param ***REMOVED***Object=***REMOVED*** opts.clientConfig
   *   The customized config to build the call settings. See
   *   ***REMOVED***@link gax.constructSettings***REMOVED*** for the format.
   */
  this.firestoreClient = function(opts) ***REMOVED***
    return new FirestoreClient(gaxGrpc, firestoreStubProtos, opts);
  ***REMOVED***;
  extend(this.firestoreClient, FirestoreClient);
***REMOVED***
module.exports = FirestoreClientBuilder;
module.exports.SERVICE_ADDRESS = SERVICE_ADDRESS;
module.exports.ALL_SCOPES = ALL_SCOPES;
