// nopt/lib/nopt.js
// info about each config option.

var debug = process.env.DEBUG_NOPT || process.env.NOPT_DEBUG
  ? function () ***REMOVED*** console.error.apply(console, arguments) ***REMOVED***
  : function () ***REMOVED******REMOVED***

var url = require("url")
  , path = require("path")
  , Stream = require("stream").Stream
  , abbrev = require("abbrev")

module.exports = exports = nopt
exports.clean = clean

exports.typeDefs =
  ***REMOVED*** String  : ***REMOVED*** type: String,  validate: validateString  ***REMOVED***
  , Boolean : ***REMOVED*** type: Boolean, validate: validateBoolean ***REMOVED***
  , url     : ***REMOVED*** type: url,     validate: validateUrl     ***REMOVED***
  , Number  : ***REMOVED*** type: Number,  validate: validateNumber  ***REMOVED***
  , path    : ***REMOVED*** type: path,    validate: validatePath    ***REMOVED***
  , Stream  : ***REMOVED*** type: Stream,  validate: validateStream  ***REMOVED***
  , Date    : ***REMOVED*** type: Date,    validate: validateDate    ***REMOVED***
  ***REMOVED***

function nopt (types, shorthands, args, slice) ***REMOVED***
  args = args || process.argv
  types = types || ***REMOVED******REMOVED***
  shorthands = shorthands || ***REMOVED******REMOVED***
  if (typeof slice !== "number") slice = 2

  debug(types, shorthands, args, slice)

  args = args.slice(slice)
  var data = ***REMOVED******REMOVED***
    , key
    , remain = []
    , cooked = args
    , original = args.slice(0)

  parse(args, data, remain, types, shorthands)
  // now data is full
  clean(data, types, exports.typeDefs)
  data.argv = ***REMOVED***remain:remain,cooked:cooked,original:original***REMOVED***
  data.argv.toString = function () ***REMOVED***
    return this.original.map(JSON.stringify).join(" ")
  ***REMOVED***
  return data
***REMOVED***

function clean (data, types, typeDefs) ***REMOVED***
  typeDefs = typeDefs || exports.typeDefs
  var remove = ***REMOVED******REMOVED***
    , typeDefault = [false, true, null, String, Number]

  Object.keys(data).forEach(function (k) ***REMOVED***
    if (k === "argv") return
    var val = data[k]
      , isArray = Array.isArray(val)
      , type = types[k]
    if (!isArray) val = [val]
    if (!type) type = typeDefault
    if (type === Array) type = typeDefault.concat(Array)
    if (!Array.isArray(type)) type = [type]

    debug("val=%j", val)
    debug("types=", type)
    val = val.map(function (val) ***REMOVED***
      // if it's an unknown value, then parse false/true/null/numbers/dates
      if (typeof val === "string") ***REMOVED***
        debug("string %j", val)
        val = val.trim()
        if ((val === "null" && ~type.indexOf(null))
            || (val === "true" &&
               (~type.indexOf(true) || ~type.indexOf(Boolean)))
            || (val === "false" &&
               (~type.indexOf(false) || ~type.indexOf(Boolean)))) ***REMOVED***
          val = JSON.parse(val)
          debug("jsonable %j", val)
        ***REMOVED*** else if (~type.indexOf(Number) && !isNaN(val)) ***REMOVED***
          debug("convert to number", val)
          val = +val
        ***REMOVED*** else if (~type.indexOf(Date) && !isNaN(Date.parse(val))) ***REMOVED***
          debug("convert to date", val)
          val = new Date(val)
        ***REMOVED***
      ***REMOVED***

      if (!types.hasOwnProperty(k)) ***REMOVED***
        return val
      ***REMOVED***

      // allow `--no-blah` to set 'blah' to null if null is allowed
      if (val === false && ~type.indexOf(null) &&
          !(~type.indexOf(false) || ~type.indexOf(Boolean))) ***REMOVED***
        val = null
      ***REMOVED***

      var d = ***REMOVED******REMOVED***
      d[k] = val
      debug("prevalidated val", d, val, types[k])
      if (!validate(d, k, val, types[k], typeDefs)) ***REMOVED***
        if (exports.invalidHandler) ***REMOVED***
          exports.invalidHandler(k, val, types[k], data)
        ***REMOVED*** else if (exports.invalidHandler !== false) ***REMOVED***
          debug("invalid: "+k+"="+val, types[k])
        ***REMOVED***
        return remove
      ***REMOVED***
      debug("validated val", d, val, types[k])
      return d[k]
    ***REMOVED***).filter(function (val) ***REMOVED*** return val !== remove ***REMOVED***)

    if (!val.length) delete data[k]
    else if (isArray) ***REMOVED***
      debug(isArray, data[k], val)
      data[k] = val
    ***REMOVED*** else data[k] = val[0]

    debug("k=%s val=%j", k, val, data[k])
  ***REMOVED***)
***REMOVED***

function validateString (data, k, val) ***REMOVED***
  data[k] = String(val)
***REMOVED***

function validatePath (data, k, val) ***REMOVED***
  data[k] = path.resolve(String(val))
  return true
***REMOVED***

function validateNumber (data, k, val) ***REMOVED***
  debug("validate Number %j %j %j", k, val, isNaN(val))
  if (isNaN(val)) return false
  data[k] = +val
***REMOVED***

function validateDate (data, k, val) ***REMOVED***
  debug("validate Date %j %j %j", k, val, Date.parse(val))
  var s = Date.parse(val)
  if (isNaN(s)) return false
  data[k] = new Date(val)
***REMOVED***

function validateBoolean (data, k, val) ***REMOVED***
  if (val instanceof Boolean) val = val.valueOf()
  else if (typeof val === "string") ***REMOVED***
    if (!isNaN(val)) val = !!(+val)
    else if (val === "null" || val === "false") val = false
    else val = true
  ***REMOVED*** else val = !!val
  data[k] = val
***REMOVED***

function validateUrl (data, k, val) ***REMOVED***
  val = url.parse(String(val))
  if (!val.host) return false
  data[k] = val.href
***REMOVED***

function validateStream (data, k, val) ***REMOVED***
  if (!(val instanceof Stream)) return false
  data[k] = val
***REMOVED***

function validate (data, k, val, type, typeDefs) ***REMOVED***
  // arrays are lists of types.
  if (Array.isArray(type)) ***REMOVED***
    for (var i = 0, l = type.length; i < l; i ++) ***REMOVED***
      if (type[i] === Array) continue
      if (validate(data, k, val, type[i], typeDefs)) return true
    ***REMOVED***
    delete data[k]
    return false
  ***REMOVED***

  // an array of anything?
  if (type === Array) return true

  // NaN is poisonous.  Means that something is not allowed.
  if (type !== type) ***REMOVED***
    debug("Poison NaN", k, val, type)
    delete data[k]
    return false
  ***REMOVED***

  // explicit list of values
  if (val === type) ***REMOVED***
    debug("Explicitly allowed %j", val)
    // if (isArray) (data[k] = data[k] || []).push(val)
    // else data[k] = val
    data[k] = val
    return true
  ***REMOVED***

  // now go through the list of typeDefs, validate against each one.
  var ok = false
    , types = Object.keys(typeDefs)
  for (var i = 0, l = types.length; i < l; i ++) ***REMOVED***
    debug("test type %j %j %j", k, val, types[i])
    var t = typeDefs[types[i]]
    if (t && type === t.type) ***REMOVED***
      var d = ***REMOVED******REMOVED***
      ok = false !== t.validate(d, k, val)
      val = d[k]
      if (ok) ***REMOVED***
        // if (isArray) (data[k] = data[k] || []).push(val)
        // else data[k] = val
        data[k] = val
        break
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  debug("OK? %j (%j %j %j)", ok, k, val, types[i])

  if (!ok) delete data[k]
  return ok
***REMOVED***

function parse (args, data, remain, types, shorthands) ***REMOVED***
  debug("parse", args, data, remain)

  var key = null
    , abbrevs = abbrev(Object.keys(types))
    , shortAbbr = abbrev(Object.keys(shorthands))

  for (var i = 0; i < args.length; i ++) ***REMOVED***
    var arg = args[i]
    debug("arg", arg)

    if (arg.match(/^-***REMOVED***2,***REMOVED***$/)) ***REMOVED***
      // done with keys.
      // the rest are args.
      remain.push.apply(remain, args.slice(i + 1))
      args[i] = "--"
      break
    ***REMOVED***
    if (arg.charAt(0) === "-") ***REMOVED***
      if (arg.indexOf("=") !== -1) ***REMOVED***
        var v = arg.split("=")
        arg = v.shift()
        v = v.join("=")
        args.splice.apply(args, [i, 1].concat([arg, v]))
      ***REMOVED***
      // see if it's a shorthand
      // if so, splice and back up to re-parse it.
      var shRes = resolveShort(arg, shorthands, shortAbbr, abbrevs)
      debug("arg=%j shRes=%j", arg, shRes)
      if (shRes) ***REMOVED***
        debug(arg, shRes)
        args.splice.apply(args, [i, 1].concat(shRes))
        if (arg !== shRes[0]) ***REMOVED***
          i --
          continue
        ***REMOVED***
      ***REMOVED***
      arg = arg.replace(/^-+/, "")
      var no = false
      while (arg.toLowerCase().indexOf("no-") === 0) ***REMOVED***
        no = !no
        arg = arg.substr(3)
      ***REMOVED***

      if (abbrevs[arg]) arg = abbrevs[arg]

      var isArray = types[arg] === Array ||
        Array.isArray(types[arg]) && types[arg].indexOf(Array) !== -1

      var val
        , la = args[i + 1]

      var isBool = no ||
        types[arg] === Boolean ||
        Array.isArray(types[arg]) && types[arg].indexOf(Boolean) !== -1 ||
        (la === "false" &&
         (types[arg] === null ||
          Array.isArray(types[arg]) && ~types[arg].indexOf(null)))

      if (isBool) ***REMOVED***
        // just set and move along
        val = !no
        // however, also support --bool true or --bool false
        if (la === "true" || la === "false") ***REMOVED***
          val = JSON.parse(la)
          la = null
          if (no) val = !val
          i ++
        ***REMOVED***

        // also support "foo":[Boolean, "bar"] and "--foo bar"
        if (Array.isArray(types[arg]) && la) ***REMOVED***
          if (~types[arg].indexOf(la)) ***REMOVED***
            // an explicit type
            val = la
            i ++
          ***REMOVED*** else if ( la === "null" && ~types[arg].indexOf(null) ) ***REMOVED***
            // null allowed
            val = null
            i ++
          ***REMOVED*** else if ( !la.match(/^-***REMOVED***2,***REMOVED***[^-]/) &&
                      !isNaN(la) &&
                      ~types[arg].indexOf(Number) ) ***REMOVED***
            // number
            val = +la
            i ++
          ***REMOVED*** else if ( !la.match(/^-[^-]/) && ~types[arg].indexOf(String) ) ***REMOVED***
            // string
            val = la
            i ++
          ***REMOVED***
        ***REMOVED***

        if (isArray) (data[arg] = data[arg] || []).push(val)
        else data[arg] = val

        continue
      ***REMOVED***

      if (la && la.match(/^-***REMOVED***2,***REMOVED***$/)) ***REMOVED***
        la = undefined
        i --
      ***REMOVED***

      val = la === undefined ? true : la
      if (isArray) (data[arg] = data[arg] || []).push(val)
      else data[arg] = val

      i ++
      continue
    ***REMOVED***
    remain.push(arg)
  ***REMOVED***
***REMOVED***

function resolveShort (arg, shorthands, shortAbbr, abbrevs) ***REMOVED***
  // handle single-char shorthands glommed together, like
  // npm ls -glp, but only if there is one dash, and only if
  // all of the chars are single-char shorthands, and it's
  // not a match to some other abbrev.
  arg = arg.replace(/^-+/, '')
  if (abbrevs[arg] && !shorthands[arg]) ***REMOVED***
    return null
  ***REMOVED***
  if (shortAbbr[arg]) ***REMOVED***
    arg = shortAbbr[arg]
  ***REMOVED*** else ***REMOVED***
    var singles = shorthands.___singles
    if (!singles) ***REMOVED***
      singles = Object.keys(shorthands).filter(function (s) ***REMOVED***
        return s.length === 1
      ***REMOVED***).reduce(function (l,r) ***REMOVED*** l[r] = true ; return l ***REMOVED***, ***REMOVED******REMOVED***)
      shorthands.___singles = singles
    ***REMOVED***
    var chrs = arg.split("").filter(function (c) ***REMOVED***
      return singles[c]
    ***REMOVED***)
    if (chrs.join("") === arg) return chrs.map(function (c) ***REMOVED***
      return shorthands[c]
    ***REMOVED***).reduce(function (l, r) ***REMOVED***
      return l.concat(r)
    ***REMOVED***, [])
  ***REMOVED***

  if (shorthands[arg] && !Array.isArray(shorthands[arg])) ***REMOVED***
    shorthands[arg] = shorthands[arg].split(/\s+/)
  ***REMOVED***
  return shorthands[arg]
***REMOVED***

if (module === require.main) ***REMOVED***
var assert = require("assert")
  , util = require("util")

  , shorthands =
    ***REMOVED*** s : ["--loglevel", "silent"]
    , d : ["--loglevel", "info"]
    , dd : ["--loglevel", "verbose"]
    , ddd : ["--loglevel", "silly"]
    , noreg : ["--no-registry"]
    , reg : ["--registry"]
    , "no-reg" : ["--no-registry"]
    , silent : ["--loglevel", "silent"]
    , verbose : ["--loglevel", "verbose"]
    , h : ["--usage"]
    , H : ["--usage"]
    , "?" : ["--usage"]
    , help : ["--usage"]
    , v : ["--version"]
    , f : ["--force"]
    , desc : ["--description"]
    , "no-desc" : ["--no-description"]
    , "local" : ["--no-global"]
    , l : ["--long"]
    , p : ["--parseable"]
    , porcelain : ["--parseable"]
    , g : ["--global"]
    ***REMOVED***

  , types =
    ***REMOVED*** aoa: Array
    , nullstream: [null, Stream]
    , date: Date
    , str: String
    , browser : String
    , cache : path
    , color : ["always", Boolean]
    , depth : Number
    , description : Boolean
    , dev : Boolean
    , editor : path
    , force : Boolean
    , global : Boolean
    , globalconfig : path
    , group : [String, Number]
    , gzipbin : String
    , logfd : [Number, Stream]
    , loglevel : ["silent","win","error","warn","info","verbose","silly"]
    , long : Boolean
    , "node-version" : [false, String]
    , npaturl : url
    , npat : Boolean
    , "onload-script" : [false, String]
    , outfd : [Number, Stream]
    , parseable : Boolean
    , pre: Boolean
    , prefix: path
    , proxy : url
    , "rebuild-bundle" : Boolean
    , registry : url
    , searchopts : String
    , searchexclude: [null, String]
    , shell : path
    , t: [Array, String]
    , tag : String
    , tar : String
    , tmp : path
    , "unsafe-perm" : Boolean
    , usage : Boolean
    , user : String
    , username : String
    , userconfig : path
    , version : Boolean
    , viewer: path
    , _exit : Boolean
    ***REMOVED***

; [["-v", ***REMOVED***version:true***REMOVED***, []]
  ,["---v", ***REMOVED***version:true***REMOVED***, []]
  ,["ls -s --no-reg connect -d",
    ***REMOVED***loglevel:"info",registry:null***REMOVED***,["ls","connect"]]
  ,["ls ---s foo",***REMOVED***loglevel:"silent"***REMOVED***,["ls","foo"]]
  ,["ls --registry blargle", ***REMOVED******REMOVED***, ["ls"]]
  ,["--no-registry", ***REMOVED***registry:null***REMOVED***, []]
  ,["--no-color true", ***REMOVED***color:false***REMOVED***, []]
  ,["--no-color false", ***REMOVED***color:true***REMOVED***, []]
  ,["--no-color", ***REMOVED***color:false***REMOVED***, []]
  ,["--color false", ***REMOVED***color:false***REMOVED***, []]
  ,["--color --logfd 7", ***REMOVED***logfd:7,color:true***REMOVED***, []]
  ,["--color=true", ***REMOVED***color:true***REMOVED***, []]
  ,["--logfd=10", ***REMOVED***logfd:10***REMOVED***, []]
  ,["--tmp=/tmp -tar=gtar",***REMOVED***tmp:"/tmp",tar:"gtar"***REMOVED***,[]]
  ,["--tmp=tmp -tar=gtar",
    ***REMOVED***tmp:path.resolve(process.cwd(), "tmp"),tar:"gtar"***REMOVED***,[]]
  ,["--logfd x", ***REMOVED******REMOVED***, []]
  ,["a -true -- -no-false", ***REMOVED***true:true***REMOVED***,["a","-no-false"]]
  ,["a -no-false", ***REMOVED***false:false***REMOVED***,["a"]]
  ,["a -no-no-true", ***REMOVED***true:true***REMOVED***, ["a"]]
  ,["a -no-no-no-false", ***REMOVED***false:false***REMOVED***, ["a"]]
  ,["---NO-no-No-no-no-no-nO-no-no"+
    "-No-no-no-no-no-no-no-no-no"+
    "-no-no-no-no-NO-NO-no-no-no-no-no-no"+
    "-no-body-can-do-the-boogaloo-like-I-do"
   ,***REMOVED***"body-can-do-the-boogaloo-like-I-do":false***REMOVED***, []]
  ,["we are -no-strangers-to-love "+
    "--you-know the-rules --and so-do-i "+
    "---im-thinking-of=a-full-commitment "+
    "--no-you-would-get-this-from-any-other-guy "+
    "--no-gonna-give-you-up "+
    "-no-gonna-let-you-down=true "+
    "--no-no-gonna-run-around false "+
    "--desert-you=false "+
    "--make-you-cry false "+
    "--no-tell-a-lie "+
    "--no-no-and-hurt-you false"
   ,***REMOVED***"strangers-to-love":false
    ,"you-know":"the-rules"
    ,"and":"so-do-i"
    ,"you-would-get-this-from-any-other-guy":false
    ,"gonna-give-you-up":false
    ,"gonna-let-you-down":false
    ,"gonna-run-around":false
    ,"desert-you":false
    ,"make-you-cry":false
    ,"tell-a-lie":false
    ,"and-hurt-you":false
    ***REMOVED***,["we", "are"]]
  ,["-t one -t two -t three"
   ,***REMOVED***t: ["one", "two", "three"]***REMOVED***
   ,[]]
  ,["-t one -t null -t three four five null"
   ,***REMOVED***t: ["one", "null", "three"]***REMOVED***
   ,["four", "five", "null"]]
  ,["-t foo"
   ,***REMOVED***t:["foo"]***REMOVED***
   ,[]]
  ,["--no-t"
   ,***REMOVED***t:["false"]***REMOVED***
   ,[]]
  ,["-no-no-t"
   ,***REMOVED***t:["true"]***REMOVED***
   ,[]]
  ,["-aoa one -aoa null -aoa 100"
   ,***REMOVED***aoa:["one", null, 100]***REMOVED***
   ,[]]
  ,["-str 100"
   ,***REMOVED***str:"100"***REMOVED***
   ,[]]
  ,["--color always"
   ,***REMOVED***color:"always"***REMOVED***
   ,[]]
  ,["--no-nullstream"
   ,***REMOVED***nullstream:null***REMOVED***
   ,[]]
  ,["--nullstream false"
   ,***REMOVED***nullstream:null***REMOVED***
   ,[]]
  ,["--notadate 2011-01-25"
   ,***REMOVED***notadate: "2011-01-25"***REMOVED***
   ,[]]
  ,["--date 2011-01-25"
   ,***REMOVED***date: new Date("2011-01-25")***REMOVED***
   ,[]]
  ].forEach(function (test) ***REMOVED***
    var argv = test[0].split(/\s+/)
      , opts = test[1]
      , rem = test[2]
      , actual = nopt(types, shorthands, argv, 0)
      , parsed = actual.argv
    delete actual.argv
    console.log(util.inspect(actual, false, 2, true), parsed.remain)
    for (var i in opts) ***REMOVED***
      var e = JSON.stringify(opts[i])
        , a = JSON.stringify(actual[i] === undefined ? null : actual[i])
      if (e && typeof e === "object") ***REMOVED***
        assert.deepEqual(e, a)
      ***REMOVED*** else ***REMOVED***
        assert.equal(e, a)
      ***REMOVED***
    ***REMOVED***
    assert.deepEqual(rem, parsed.remain)
  ***REMOVED***)
***REMOVED***
