// @firebase/storage/dist/cjs/src/implementation/string.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var errorsExports = require("./error");
exports.StringFormat = ***REMOVED***
    RAW: 'raw',
    BASE64: 'base64',
    BASE64URL: 'base64url',
    DATA_URL: 'data_url'
***REMOVED***;
function formatValidator(stringFormat) ***REMOVED***
    switch (stringFormat) ***REMOVED***
        case exports.StringFormat.RAW:
        case exports.StringFormat.BASE64:
        case exports.StringFormat.BASE64URL:
        case exports.StringFormat.DATA_URL:
            return;
        default:
            throw 'Expected one of the event types: [' +
                exports.StringFormat.RAW +
                ', ' +
                exports.StringFormat.BASE64 +
                ', ' +
                exports.StringFormat.BASE64URL +
                ', ' +
                exports.StringFormat.DATA_URL +
                '].';
    ***REMOVED***
***REMOVED***
exports.formatValidator = formatValidator;
/**
 * @struct
 */
var StringData = /** @class */ (function () ***REMOVED***
    function StringData(data, opt_contentType) ***REMOVED***
        this.data = data;
        this.contentType = opt_contentType || null;
    ***REMOVED***
    return StringData;
***REMOVED***());
exports.StringData = StringData;
function dataFromString(format, string) ***REMOVED***
    switch (format) ***REMOVED***
        case exports.StringFormat.RAW:
            return new StringData(utf8Bytes_(string));
        case exports.StringFormat.BASE64:
        case exports.StringFormat.BASE64URL:
            return new StringData(base64Bytes_(format, string));
        case exports.StringFormat.DATA_URL:
            return new StringData(dataURLBytes_(string), dataURLContentType_(string));
    ***REMOVED***
    // assert(false);
    throw errorsExports.unknown();
***REMOVED***
exports.dataFromString = dataFromString;
function utf8Bytes_(string) ***REMOVED***
    var b = [];
    for (var i = 0; i < string.length; i++) ***REMOVED***
        var c = string.charCodeAt(i);
        if (c <= 127) ***REMOVED***
            b.push(c);
        ***REMOVED***
        else ***REMOVED***
            if (c <= 2047) ***REMOVED***
                b.push(192 | (c >> 6), 128 | (c & 63));
            ***REMOVED***
            else ***REMOVED***
                if ((c & 64512) == 55296) ***REMOVED***
                    // The start of a surrogate pair.
                    var valid = i < string.length - 1 &&
                        (string.charCodeAt(i + 1) & 64512) == 56320;
                    if (!valid) ***REMOVED***
                        // The second surrogate wasn't there.
                        b.push(239, 191, 189);
                    ***REMOVED***
                    else ***REMOVED***
                        var hi = c;
                        var lo = string.charCodeAt(++i);
                        c = 65536 | ((hi & 1023) << 10) | (lo & 1023);
                        b.push(240 | (c >> 18), 128 | ((c >> 12) & 63), 128 | ((c >> 6) & 63), 128 | (c & 63));
                    ***REMOVED***
                ***REMOVED***
                else ***REMOVED***
                    if ((c & 64512) == 56320) ***REMOVED***
                        // Invalid low surrogate.
                        b.push(239, 191, 189);
                    ***REMOVED***
                    else ***REMOVED***
                        b.push(224 | (c >> 12), 128 | ((c >> 6) & 63), 128 | (c & 63));
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return new Uint8Array(b);
***REMOVED***
exports.utf8Bytes_ = utf8Bytes_;
function percentEncodedBytes_(string) ***REMOVED***
    var decoded;
    try ***REMOVED***
        decoded = decodeURIComponent(string);
    ***REMOVED***
    catch (e) ***REMOVED***
        throw errorsExports.invalidFormat(exports.StringFormat.DATA_URL, 'Malformed data URL.');
    ***REMOVED***
    return utf8Bytes_(decoded);
***REMOVED***
exports.percentEncodedBytes_ = percentEncodedBytes_;
function base64Bytes_(format, string) ***REMOVED***
    switch (format) ***REMOVED***
        case exports.StringFormat.BASE64: ***REMOVED***
            var hasMinus = string.indexOf('-') !== -1;
            var hasUnder = string.indexOf('_') !== -1;
            if (hasMinus || hasUnder) ***REMOVED***
                var invalidChar = hasMinus ? '-' : '_';
                throw errorsExports.invalidFormat(format, "Invalid character '" +
                    invalidChar +
                    "' found: is it base64url encoded?");
            ***REMOVED***
            break;
        ***REMOVED***
        case exports.StringFormat.BASE64URL: ***REMOVED***
            var hasPlus = string.indexOf('+') !== -1;
            var hasSlash = string.indexOf('/') !== -1;
            if (hasPlus || hasSlash) ***REMOVED***
                var invalidChar = hasPlus ? '+' : '/';
                throw errorsExports.invalidFormat(format, "Invalid character '" + invalidChar + "' found: is it base64 encoded?");
            ***REMOVED***
            string = string.replace(/-/g, '+').replace(/_/g, '/');
            break;
        ***REMOVED***
    ***REMOVED***
    var bytes;
    try ***REMOVED***
        bytes = atob(string);
    ***REMOVED***
    catch (e) ***REMOVED***
        throw errorsExports.invalidFormat(format, 'Invalid character found');
    ***REMOVED***
    var array = new Uint8Array(bytes.length);
    for (var i = 0; i < bytes.length; i++) ***REMOVED***
        array[i] = bytes.charCodeAt(i);
    ***REMOVED***
    return array;
***REMOVED***
exports.base64Bytes_ = base64Bytes_;
/**
 * @struct
 */
var DataURLParts = /** @class */ (function () ***REMOVED***
    function DataURLParts(dataURL) ***REMOVED***
        this.base64 = false;
        this.contentType = null;
        var matches = dataURL.match(/^data:([^,]+)?,/);
        if (matches === null) ***REMOVED***
            throw errorsExports.invalidFormat(exports.StringFormat.DATA_URL, "Must be formatted 'data:[<mediatype>][;base64],<data>");
        ***REMOVED***
        var middle = matches[1] || null;
        if (middle != null) ***REMOVED***
            this.base64 = endsWith(middle, ';base64');
            this.contentType = this.base64
                ? middle.substring(0, middle.length - ';base64'.length)
                : middle;
        ***REMOVED***
        this.rest = dataURL.substring(dataURL.indexOf(',') + 1);
    ***REMOVED***
    return DataURLParts;
***REMOVED***());
function dataURLBytes_(string) ***REMOVED***
    var parts = new DataURLParts(string);
    if (parts.base64) ***REMOVED***
        return base64Bytes_(exports.StringFormat.BASE64, parts.rest);
    ***REMOVED***
    else ***REMOVED***
        return percentEncodedBytes_(parts.rest);
    ***REMOVED***
***REMOVED***
exports.dataURLBytes_ = dataURLBytes_;
function dataURLContentType_(string) ***REMOVED***
    var parts = new DataURLParts(string);
    return parts.contentType;
***REMOVED***
exports.dataURLContentType_ = dataURLContentType_;
function endsWith(s, end) ***REMOVED***
    var longEnough = s.length >= end.length;
    if (!longEnough) ***REMOVED***
        return false;
    ***REMOVED***
    return s.substring(s.length - end.length) === end;
***REMOVED***
