// @firebase/storage/dist/cjs/src/implementation/type.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
/**
 * @return False if the object is undefined or null, true otherwise.
 */
function isDef(p) ***REMOVED***
    return p != null;
***REMOVED***
exports.isDef = isDef;
function isJustDef(p) ***REMOVED***
    return p !== void 0;
***REMOVED***
exports.isJustDef = isJustDef;
function isFunction(p) ***REMOVED***
    return typeof p === 'function';
***REMOVED***
exports.isFunction = isFunction;
function isObject(p) ***REMOVED***
    return typeof p === 'object';
***REMOVED***
exports.isObject = isObject;
function isNonNullObject(p) ***REMOVED***
    return isObject(p) && p !== null;
***REMOVED***
exports.isNonNullObject = isNonNullObject;
function isNonArrayObject(p) ***REMOVED***
    return isObject(p) && !Array.isArray(p);
***REMOVED***
exports.isNonArrayObject = isNonArrayObject;
function isString(p) ***REMOVED***
    return typeof p === 'string' || p instanceof String;
***REMOVED***
exports.isString = isString;
function isNumber(p) ***REMOVED***
    return typeof p === 'number' || p instanceof Number;
***REMOVED***
exports.isNumber = isNumber;
function isNativeBlob(p) ***REMOVED***
    return isNativeBlobDefined() && p instanceof Blob;
***REMOVED***
exports.isNativeBlob = isNativeBlob;
function isNativeBlobDefined() ***REMOVED***
    return typeof Blob !== 'undefined';
***REMOVED***
exports.isNativeBlobDefined = isNativeBlobDefined;
