// loader-runner/lib/LoaderRunner.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var fs = require("fs");
var readFile = fs.readFile.bind(fs);
var loadLoader = require("./loadLoader");

function utf8BufferToString(buf) ***REMOVED***
	var str = buf.toString("utf-8");
	if(str.charCodeAt(0) === 0xFEFF) ***REMOVED***
		return str.substr(1);
	***REMOVED*** else ***REMOVED***
		return str;
	***REMOVED***
***REMOVED***

function splitQuery(req) ***REMOVED***
	var i = req.indexOf("?");
	if(i < 0) return [req, ""];
	return [req.substr(0, i), req.substr(i)];
***REMOVED***

function dirname(path) ***REMOVED***
	if(path === "/") return "/";
	var i = path.lastIndexOf("/");
	var j = path.lastIndexOf("\\");
	var i2 = path.indexOf("/");
	var j2 = path.indexOf("\\");
	var idx = i > j ? i : j;
	var idx2 = i > j ? i2 : j2;
	if(idx < 0) return path;
	if(idx === idx2) return path.substr(0, idx + 1);
	return path.substr(0, idx);
***REMOVED***

function createLoaderObject(loader) ***REMOVED***
	var obj = ***REMOVED***
		path: null,
		query: null,
		options: null,
		ident: null,
		normal: null,
		pitch: null,
		raw: null,
		data: null,
		pitchExecuted: false,
		normalExecuted: false
	***REMOVED***;
	Object.defineProperty(obj, "request", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			return obj.path + obj.query;
		***REMOVED***,
		set: function(value) ***REMOVED***
			if(typeof value === "string") ***REMOVED***
				var splittedRequest = splitQuery(value);
				obj.path = splittedRequest[0];
				obj.query = splittedRequest[1];
				obj.options = undefined;
				obj.ident = undefined;
			***REMOVED*** else ***REMOVED***
				if(!value.loader)
					throw new Error("request should be a string or object with loader and object (" + JSON.stringify(value) + ")");
				obj.path = value.loader;
				obj.options = value.options;
				obj.ident = value.ident;
				if(obj.options === null)
					obj.query = "";
				else if(obj.options === undefined)
					obj.query = "";
				else if(typeof obj.options === "string")
					obj.query = "?" + obj.options;
				else if(obj.ident)
					obj.query = "??" + obj.ident;
				else if(typeof obj.options === "object" && obj.options.ident)
					obj.query = "??" + obj.options.ident;
				else
					obj.query = "?" + JSON.stringify(obj.options);
			***REMOVED***
		***REMOVED***
	***REMOVED***);
	obj.request = loader;
	if(Object.preventExtensions) ***REMOVED***
		Object.preventExtensions(obj);
	***REMOVED***
	return obj;
***REMOVED***

function runSyncOrAsync(fn, context, args, callback) ***REMOVED***
	var isSync = true;
	var isDone = false;
	var isError = false; // internal error
	var reportedError = false;
	context.async = function async() ***REMOVED***
		if(isDone) ***REMOVED***
			if(reportedError) return; // ignore
			throw new Error("async(): The callback was already called.");
		***REMOVED***
		isSync = false;
		return innerCallback;
	***REMOVED***;
	var innerCallback = context.callback = function() ***REMOVED***
		if(isDone) ***REMOVED***
			if(reportedError) return; // ignore
			throw new Error("callback(): The callback was already called.");
		***REMOVED***
		isDone = true;
		isSync = false;
		try ***REMOVED***
			callback.apply(null, arguments);
		***REMOVED*** catch(e) ***REMOVED***
			isError = true;
			throw e;
		***REMOVED***
	***REMOVED***;
	try ***REMOVED***
		var result = (function LOADER_EXECUTION() ***REMOVED***
			return fn.apply(context, args);
		***REMOVED***());
		if(isSync) ***REMOVED***
			isDone = true;
			if(result === undefined)
				return callback();
			if(result && typeof result === "object" && typeof result.then === "function") ***REMOVED***
				return result.catch(callback).then(function(r) ***REMOVED***
					callback(null, r);
				***REMOVED***);
			***REMOVED***
			return callback(null, result);
		***REMOVED***
	***REMOVED*** catch(e) ***REMOVED***
		if(isError) throw e;
		if(isDone) ***REMOVED***
			// loader is already "done", so we cannot use the callback function
			// for better debugging we print the error on the console
			if(typeof e === "object" && e.stack) console.error(e.stack);
			else console.error(e);
			return;
		***REMOVED***
		isDone = true;
		reportedError = true;
		callback(e);
	***REMOVED***

***REMOVED***

function convertArgs(args, raw) ***REMOVED***
	if(!raw && Buffer.isBuffer(args[0]))
		args[0] = utf8BufferToString(args[0]);
	else if(raw && typeof args[0] === "string")
		args[0] = new Buffer(args[0], "utf-8"); // eslint-disable-line
***REMOVED***

function iteratePitchingLoaders(options, loaderContext, callback) ***REMOVED***
	// abort after last loader
	if(loaderContext.loaderIndex >= loaderContext.loaders.length)
		return processResource(options, loaderContext, callback);

	var currentLoaderObject = loaderContext.loaders[loaderContext.loaderIndex];

	// iterate
	if(currentLoaderObject.pitchExecuted) ***REMOVED***
		loaderContext.loaderIndex++;
		return iteratePitchingLoaders(options, loaderContext, callback);
	***REMOVED***

	// load loader module
	loadLoader(currentLoaderObject, function(err) ***REMOVED***
		if(err) return callback(err);
		var fn = currentLoaderObject.pitch;
		currentLoaderObject.pitchExecuted = true;
		if(!fn) return iteratePitchingLoaders(options, loaderContext, callback);

		runSyncOrAsync(
			fn,
			loaderContext, [loaderContext.remainingRequest, loaderContext.previousRequest, currentLoaderObject.data = ***REMOVED******REMOVED***],
			function(err) ***REMOVED***
				if(err) return callback(err);
				var args = Array.prototype.slice.call(arguments, 1);
				if(args.length > 0) ***REMOVED***
					loaderContext.loaderIndex--;
					iterateNormalLoaders(options, loaderContext, args, callback);
				***REMOVED*** else ***REMOVED***
					iteratePitchingLoaders(options, loaderContext, callback);
				***REMOVED***
			***REMOVED***
		);
	***REMOVED***);
***REMOVED***

function processResource(options, loaderContext, callback) ***REMOVED***
	// set loader index to last loader
	loaderContext.loaderIndex = loaderContext.loaders.length - 1;

	var resourcePath = loaderContext.resourcePath;
	if(resourcePath) ***REMOVED***
		loaderContext.addDependency(resourcePath);
		options.readResource(resourcePath, function(err, buffer) ***REMOVED***
			if(err) return callback(err);
			options.resourceBuffer = buffer;
			iterateNormalLoaders(options, loaderContext, [buffer], callback);
		***REMOVED***);
	***REMOVED*** else ***REMOVED***
		iterateNormalLoaders(options, loaderContext, [null], callback);
	***REMOVED***
***REMOVED***

function iterateNormalLoaders(options, loaderContext, args, callback) ***REMOVED***
	if(loaderContext.loaderIndex < 0)
		return callback(null, args);

	var currentLoaderObject = loaderContext.loaders[loaderContext.loaderIndex];

	// iterate
	if(currentLoaderObject.normalExecuted) ***REMOVED***
		loaderContext.loaderIndex--;
		return iterateNormalLoaders(options, loaderContext, args, callback);
	***REMOVED***

	var fn = currentLoaderObject.normal;
	currentLoaderObject.normalExecuted = true;
	if(!fn) ***REMOVED***
		return iterateNormalLoaders(options, loaderContext, args, callback);
	***REMOVED***

	convertArgs(args, currentLoaderObject.raw);

	runSyncOrAsync(fn, loaderContext, args, function(err) ***REMOVED***
		if(err) return callback(err);

		var args = Array.prototype.slice.call(arguments, 1);
		iterateNormalLoaders(options, loaderContext, args, callback);
	***REMOVED***);
***REMOVED***

exports.getContext = function getContext(resource) ***REMOVED***
	var splitted = splitQuery(resource);
	return dirname(splitted[0]);
***REMOVED***;

exports.runLoaders = function runLoaders(options, callback) ***REMOVED***
	// read options
	var resource = options.resource || "";
	var loaders = options.loaders || [];
	var loaderContext = options.context || ***REMOVED******REMOVED***;
	var readResource = options.readResource || readFile;

	//
	var splittedResource = resource && splitQuery(resource);
	var resourcePath = splittedResource ? splittedResource[0] : undefined;
	var resourceQuery = splittedResource ? splittedResource[1] : undefined;
	var contextDirectory = resourcePath ? dirname(resourcePath) : null;

	// execution state
	var requestCacheable = true;
	var fileDependencies = [];
	var contextDependencies = [];

	// prepare loader objects
	loaders = loaders.map(createLoaderObject);

	loaderContext.context = contextDirectory;
	loaderContext.loaderIndex = 0;
	loaderContext.loaders = loaders;
	loaderContext.resourcePath = resourcePath;
	loaderContext.resourceQuery = resourceQuery;
	loaderContext.async = null;
	loaderContext.callback = null;
	loaderContext.cacheable = function cacheable(flag) ***REMOVED***
		if(flag === false) ***REMOVED***
			requestCacheable = false;
		***REMOVED***
	***REMOVED***;
	loaderContext.dependency = loaderContext.addDependency = function addDependency(file) ***REMOVED***
		fileDependencies.push(file);
	***REMOVED***;
	loaderContext.addContextDependency = function addContextDependency(context) ***REMOVED***
		contextDependencies.push(context);
	***REMOVED***;
	loaderContext.getDependencies = function getDependencies() ***REMOVED***
		return fileDependencies.slice();
	***REMOVED***;
	loaderContext.getContextDependencies = function getContextDependencies() ***REMOVED***
		return contextDependencies.slice();
	***REMOVED***;
	loaderContext.clearDependencies = function clearDependencies() ***REMOVED***
		fileDependencies.length = 0;
		contextDependencies.length = 0;
		requestCacheable = true;
	***REMOVED***;
	Object.defineProperty(loaderContext, "resource", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			if(loaderContext.resourcePath === undefined)
				return undefined;
			return loaderContext.resourcePath + loaderContext.resourceQuery;
		***REMOVED***,
		set: function(value) ***REMOVED***
			var splittedResource = value && splitQuery(value);
			loaderContext.resourcePath = splittedResource ? splittedResource[0] : undefined;
			loaderContext.resourceQuery = splittedResource ? splittedResource[1] : undefined;
		***REMOVED***
	***REMOVED***);
	Object.defineProperty(loaderContext, "request", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			return loaderContext.loaders.map(function(o) ***REMOVED***
				return o.request;
			***REMOVED***).concat(loaderContext.resource || "").join("!");
		***REMOVED***
	***REMOVED***);
	Object.defineProperty(loaderContext, "remainingRequest", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			if(loaderContext.loaderIndex >= loaderContext.loaders.length - 1 && !loaderContext.resource)
				return "";
			return loaderContext.loaders.slice(loaderContext.loaderIndex + 1).map(function(o) ***REMOVED***
				return o.request;
			***REMOVED***).concat(loaderContext.resource || "").join("!");
		***REMOVED***
	***REMOVED***);
	Object.defineProperty(loaderContext, "currentRequest", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			return loaderContext.loaders.slice(loaderContext.loaderIndex).map(function(o) ***REMOVED***
				return o.request;
			***REMOVED***).concat(loaderContext.resource || "").join("!");
		***REMOVED***
	***REMOVED***);
	Object.defineProperty(loaderContext, "previousRequest", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			return loaderContext.loaders.slice(0, loaderContext.loaderIndex).map(function(o) ***REMOVED***
				return o.request;
			***REMOVED***).join("!");
		***REMOVED***
	***REMOVED***);
	Object.defineProperty(loaderContext, "query", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			var entry = loaderContext.loaders[loaderContext.loaderIndex];
			return entry.options && typeof entry.options === "object" ? entry.options : entry.query;
		***REMOVED***
	***REMOVED***);
	Object.defineProperty(loaderContext, "data", ***REMOVED***
		enumerable: true,
		get: function() ***REMOVED***
			return loaderContext.loaders[loaderContext.loaderIndex].data;
		***REMOVED***
	***REMOVED***);

	// finish loader context
	if(Object.preventExtensions) ***REMOVED***
		Object.preventExtensions(loaderContext);
	***REMOVED***

	var processOptions = ***REMOVED***
		resourceBuffer: null,
		readResource: readResource
	***REMOVED***;
	iteratePitchingLoaders(processOptions, loaderContext, function(err, result) ***REMOVED***
		if(err) ***REMOVED***
			return callback(err, ***REMOVED***
				cacheable: requestCacheable,
				fileDependencies: fileDependencies,
				contextDependencies: contextDependencies
			***REMOVED***);
		***REMOVED***
		callback(null, ***REMOVED***
			result: result,
			resourceBuffer: processOptions.resourceBuffer,
			cacheable: requestCacheable,
			fileDependencies: fileDependencies,
			contextDependencies: contextDependencies
		***REMOVED***);
	***REMOVED***);
***REMOVED***;
