// esrecurse/esrecurse.js
/*
  Copyright (C) 2014 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
(function () ***REMOVED***
    'use strict';

    var assign,
        estraverse,
        isArray,
        objectKeys;

    assign = require('object-assign');
    estraverse = require('estraverse');

    isArray = Array.isArray || function isArray(array) ***REMOVED***
        return Object.prototype.toString.call(array) === '[object Array]';
    ***REMOVED***;

    objectKeys = Object.keys || function (o) ***REMOVED***
        var keys = [], key;
        for (key in o) ***REMOVED***
            keys.push(key);
        ***REMOVED***
        return keys;
    ***REMOVED***;

    function isNode(node) ***REMOVED***
        if (node == null) ***REMOVED***
            return false;
        ***REMOVED***
        return typeof node === 'object' && typeof node.type === 'string';
    ***REMOVED***

    function isProperty(nodeType, key) ***REMOVED***
        return (nodeType === estraverse.Syntax.ObjectExpression || nodeType === estraverse.Syntax.ObjectPattern) && key === 'properties';
    ***REMOVED***

    function Visitor(visitor, options) ***REMOVED***
        options = options || ***REMOVED******REMOVED***;

        this.__visitor = visitor ||  this;
        this.__childVisitorKeys = options.childVisitorKeys
            ? assign(***REMOVED******REMOVED***, estraverse.VisitorKeys, options.childVisitorKeys)
            : estraverse.VisitorKeys;
        if (options.fallback === 'iteration') ***REMOVED***
            this.__fallback = objectKeys;
        ***REMOVED*** else if (typeof options.fallback === 'function') ***REMOVED***
            this.__fallback = options.fallback;
        ***REMOVED***
    ***REMOVED***

    /* Default method for visiting children.
     * When you need to call default visiting operation inside custom visiting
     * operation, you can use it with `this.visitChildren(node)`.
     */
    Visitor.prototype.visitChildren = function (node) ***REMOVED***
        var type, children, i, iz, j, jz, child;

        if (node == null) ***REMOVED***
            return;
        ***REMOVED***

        type = node.type || estraverse.Syntax.Property;

        children = this.__childVisitorKeys[type];
        if (!children) ***REMOVED***
            if (this.__fallback) ***REMOVED***
                children = this.__fallback(node);
            ***REMOVED*** else ***REMOVED***
                throw new Error('Unknown node type ' + type + '.');
            ***REMOVED***
        ***REMOVED***

        for (i = 0, iz = children.length; i < iz; ++i) ***REMOVED***
            child = node[children[i]];
            if (child) ***REMOVED***
                if (isArray(child)) ***REMOVED***
                    for (j = 0, jz = child.length; j < jz; ++j) ***REMOVED***
                        if (child[j]) ***REMOVED***
                            if (isNode(child[j]) || isProperty(type, children[i])) ***REMOVED***
                                this.visit(child[j]);
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED*** else if (isNode(child)) ***REMOVED***
                    this.visit(child);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;

    /* Dispatching node. */
    Visitor.prototype.visit = function (node) ***REMOVED***
        var type;

        if (node == null) ***REMOVED***
            return;
        ***REMOVED***

        type = node.type || estraverse.Syntax.Property;
        if (this.__visitor[type]) ***REMOVED***
            this.__visitor[type].call(this, node);
            return;
        ***REMOVED***
        this.visitChildren(node);
    ***REMOVED***;

    exports.version = require('./package.json').version;
    exports.Visitor = Visitor;
    exports.visit = function (node, visitor, options) ***REMOVED***
        var v = new Visitor(visitor, options);
        v.visit(node);
    ***REMOVED***;
***REMOVED***());
/* vim: set sw=4 ts=4 et tw=80 : */
