// hawk/test/index.js
// Load modules

var Url = require('url');
var Code = require('code');
var Hawk = require('../lib');
var Lab = require('lab');


// Declare internals

var internals = ***REMOVED******REMOVED***;


// Test shortcuts

var lab = exports.lab = Lab.script();
var describe = lab.experiment;
var it = lab.test;
var expect = Code.expect;


describe('Hawk', function () ***REMOVED***

    var credentialsFunc = function (id, callback) ***REMOVED***

        var credentials = ***REMOVED***
            id: id,
            key: 'werxhqb98rpaxn39848xrunpaw3489ruxnpa98w4rxn',
            algorithm: (id === '1' ? 'sha1' : 'sha256'),
            user: 'steve'
        ***REMOVED***;

        return callback(null, credentials);
    ***REMOVED***;

    it('generates a header then successfully parse it (configuration)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header(Url.parse('http://example.com:8080/resource/4?filter=a'), req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data' ***REMOVED***).field;
            expect(req.authorization).to.exist();

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (node request)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
                expect(res.headers['server-authorization']).to.exist();

                expect(Hawk.client.authenticate(res, credentials2, artifacts, ***REMOVED*** payload: 'some reply' ***REMOVED***)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (absolute request uri)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: 'http://example.com:8080/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts, ***REMOVED*** payload: 'some reply', contentType: 'text/plain', ext: 'response-specific' ***REMOVED***);
                expect(res.headers['server-authorization']).to.exist();

                expect(Hawk.client.authenticate(res, credentials2, artifacts, ***REMOVED*** payload: 'some reply' ***REMOVED***)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (no server header options)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts);
                expect(res.headers['server-authorization']).to.exist();

                expect(Hawk.client.authenticate(res, credentials2, artifacts)).to.equal(true);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then fails to parse it (missing server header hash)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'POST',
            url: '/resource/4?filter=a',
            headers: ***REMOVED***
                host: 'example.com:8080',
                'content-type': 'text/plain;x=y'
            ***REMOVED***
        ***REMOVED***;

        var payload = 'some not so random text';

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            var reqHeader = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', payload: payload, contentType: req.headers['content-type'] ***REMOVED***);
            req.headers.authorization = reqHeader.field;

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload(payload, credentials2, artifacts, req.headers['content-type'])).to.equal(true);

                var res = ***REMOVED***
                    headers: ***REMOVED***
                        'content-type': 'text/plain'
                    ***REMOVED***
                ***REMOVED***;

                res.headers['server-authorization'] = Hawk.server.header(credentials2, artifacts);
                expect(res.headers['server-authorization']).to.exist();

                expect(Hawk.client.authenticate(res, credentials2, artifacts, ***REMOVED*** payload: 'some reply' ***REMOVED***)).to.equal(false);
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (with hash)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it then validate payload', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(Hawk.server.authenticatePayload('hola!', credentials2, artifacts)).to.be.true();
                expect(Hawk.server.authenticatePayload('hello!', credentials2, artifacts)).to.be.false();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parses and validates payload', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** payload: 'hola!' ***REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (app)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', app: 'asd23ased' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(artifacts.app).to.equal('asd23ased');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then successfully parse it (app, dlg)', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data', app: 'asd23ased', dlg: '23434szr3q4d' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.not.exist();
                expect(credentials2.user).to.equal('steve');
                expect(artifacts.ext).to.equal('some-app-data');
                expect(artifacts.app).to.equal('asd23ased');
                expect(artifacts.dlg).to.equal('23434szr3q4d');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header then fail authentication due to bad hash', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, payload: 'hola!', ext: 'some-app-data' ***REMOVED***).field;
            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED*** payload: 'byebye!' ***REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(err.output.payload.message).to.equal('Bad payload hash');
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);

    it('generates a header for one resource then fail to authenticate another', function (done) ***REMOVED***

        var req = ***REMOVED***
            method: 'GET',
            url: '/resource/4?filter=a',
            host: 'example.com',
            port: 8080
        ***REMOVED***;

        credentialsFunc('123456', function (err, credentials1) ***REMOVED***

            req.authorization = Hawk.client.header('http://example.com:8080/resource/4?filter=a', req.method, ***REMOVED*** credentials: credentials1, ext: 'some-app-data' ***REMOVED***).field;
            req.url = '/something/else';

            Hawk.server.authenticate(req, credentialsFunc, ***REMOVED******REMOVED***, function (err, credentials2, artifacts) ***REMOVED***

                expect(err).to.exist();
                expect(credentials2).to.exist();
                done();
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***);
