// @firebase/firestore/dist/esm/src/platform_browser/browser_platform.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** JsonProtoSerializer ***REMOVED*** from '../remote/serializer';
import ***REMOVED*** WebChannelConnection ***REMOVED*** from './webchannel_connection';
var BrowserPlatform = /** @class */ (function () ***REMOVED***
    function BrowserPlatform() ***REMOVED***
        this.emptyByteString = '';
        this.base64Available = typeof atob !== 'undefined';
    ***REMOVED***
    BrowserPlatform.prototype.loadConnection = function (databaseInfo) ***REMOVED***
        return Promise.resolve(new WebChannelConnection(databaseInfo));
    ***REMOVED***;
    BrowserPlatform.prototype.newSerializer = function (databaseId) ***REMOVED***
        return new JsonProtoSerializer(databaseId, ***REMOVED*** useProto3Json: true ***REMOVED***);
    ***REMOVED***;
    BrowserPlatform.prototype.formatJSON = function (value) ***REMOVED***
        return JSON.stringify(value);
    ***REMOVED***;
    BrowserPlatform.prototype.atob = function (encoded) ***REMOVED***
        return atob(encoded);
    ***REMOVED***;
    BrowserPlatform.prototype.btoa = function (raw) ***REMOVED***
        return btoa(raw);
    ***REMOVED***;
    return BrowserPlatform;
***REMOVED***());
export ***REMOVED*** BrowserPlatform ***REMOVED***;
