// google-proto-files/index.js
'use strict';

var path = require('path');
var loader = require('./load');

var googleProtoFiles = (module.exports = function() ***REMOVED***
  var args = [].slice.call(arguments);
  return path.join.apply(null, [__dirname, 'google'].concat(args));
***REMOVED***);

module.exports.load = loader.load;

module.exports.loadSync = loader.loadSync;

module.exports.appengine = ***REMOVED***
  v1: googleProtoFiles('appengine/v1/appengine.proto'),
***REMOVED***;

module.exports.bigtable = ***REMOVED***
  v1: googleProtoFiles('bigtable/v1/bigtable_service.proto'),
  v2: googleProtoFiles('bigtable/v2/bigtable.proto'),
***REMOVED***;

module.exports.bigtable.admin = ***REMOVED***
  v1: googleProtoFiles('bigtable/admin/table/v1/bigtable_table_service.proto'),
  v2: ***REMOVED***
    instance: googleProtoFiles(
      'bigtable/admin/v2/bigtable_instance_admin.proto'
    ),
    table: googleProtoFiles('bigtable/admin/v2/bigtable_table_admin.proto'),
  ***REMOVED***,
***REMOVED***;

module.exports.cloudtrace = ***REMOVED***
  v1: googleProtoFiles('devtools/cloudtrace/v1/trace.proto'),
***REMOVED***;

module.exports.container = ***REMOVED***
  v1: googleProtoFiles('container/v1/cluster_service.proto'),
***REMOVED***;

module.exports.datastore = ***REMOVED***
  v1: googleProtoFiles('datastore/v1/datastore.proto'),
  v1beta3: googleProtoFiles('datastore/v1beta3/datastore.proto'),
***REMOVED***;

module.exports.dlp = ***REMOVED***
  v2beta1: googleProtoFiles('privacy/dlp/v2beta1/dlp.proto'),
***REMOVED***;

module.exports.firestore = ***REMOVED***
  v1beta1: googleProtoFiles('firestore/v1beta1/firestore.proto'),
***REMOVED***;

module.exports.firestore.admin = ***REMOVED***
  v1beta1: googleProtoFiles('firestore/admin/v1beta1/firestore_admin.proto'),
***REMOVED***;

module.exports.functions = ***REMOVED***
  v1beta2: googleProtoFiles('cloud/functions/v1beta2/functions.proto'),
***REMOVED***;

module.exports.iam = ***REMOVED***
  v1: googleProtoFiles('iam/v1/iam_policy.proto'),
***REMOVED***;

module.exports.iam.admin = ***REMOVED***
  v1: googleProtoFiles('iam/admin/v1/iam.proto'),
***REMOVED***;

module.exports.language = ***REMOVED***
  v1beta1: googleProtoFiles('cloud/language/v1beta1/language_service.proto'),
  v1: googleProtoFiles('cloud/language/v1/language_service.proto'),
***REMOVED***;

module.exports.logging = ***REMOVED***
  v2: googleProtoFiles('logging/v2/logging.proto'),
***REMOVED***;

module.exports.monitoring = ***REMOVED***
  v3: ***REMOVED***
    group: googleProtoFiles('monitoring/v3/group_service.proto'),
    metric: googleProtoFiles('monitoring/v3/metric_service.proto'),
  ***REMOVED***,
***REMOVED***;

module.exports.pubsub = ***REMOVED***
  v1: googleProtoFiles('pubsub/v1/pubsub.proto'),
  v1beta2: googleProtoFiles('pubsub/v1beta2/pubsub.proto'),
***REMOVED***;

module.exports.spanner = ***REMOVED***
  v1: googleProtoFiles('spanner/v1/spanner.proto'),
***REMOVED***;

module.exports.spanner.admin = ***REMOVED***
  v1: ***REMOVED***
    database: googleProtoFiles(
      'spanner/admin/database/v1/spanner_database_admin.proto'
    ),
    instance: googleProtoFiles(
      'spanner/admin/instance/v1/spanner_instance_admin.proto'
    ),
  ***REMOVED***,
***REMOVED***;

module.exports.speech = ***REMOVED***
  v1beta1: googleProtoFiles('cloud/speech/v1beta1/cloud_speech.proto'),
  v1: googleProtoFiles('cloud/speech/v1/cloud_speech.proto'),
  v2: googleProtoFiles('cloud/speech/v2/cloud_speech.proto'),
***REMOVED***;

module.exports.storagetransfer = ***REMOVED***
  v1: googleProtoFiles('storagetransfer/v1/transfer.proto'),
***REMOVED***;

module.exports.videointelligence = ***REMOVED***
  v1beta1: googleProtoFiles(
    'cloud/videointelligence/v1beta1/video_intelligence.proto'
  ),
***REMOVED***;

module.exports.vision = ***REMOVED***
  v1: googleProtoFiles('cloud/vision/v1/image_annotator.proto'),
***REMOVED***;
