// fast-deep-equal/spec/tests.js
'use strict';

module.exports = [
  ***REMOVED***
    description: 'scalars',
    tests: [
      ***REMOVED***
        description: 'equal numbers',
        value1: 1,
        value2: 1,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal numbers',
        value1: 1,
        value2: 2,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'number and array are not equal',
        value1: 1,
        value2: [],
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: '0 and null are not equal',
        value1: 0,
        value2: null,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'equal strings',
        value1: 'a',
        value2: 'a',
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal strings',
        value1: 'a',
        value2: 'b',
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'empty string and null are not equal',
        value1: '',
        value2: null,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'null is equal to null',
        value1: null,
        value2: null,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'equal booleans (true)',
        value1: true,
        value2: true,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'equal booleans (false)',
        value1: false,
        value2: false,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal booleans',
        value1: true,
        value2: false,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: '1 and true are not equal',
        value1: 1,
        value2: true,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: '0 and false are not equal',
        value1: 0,
        value2: false,
        equal: false
      ***REMOVED***
    ]
  ***REMOVED***,

  ***REMOVED***
    description: 'objects',
    tests: [
      ***REMOVED***
        description: 'empty objects are equal',
        value1: ***REMOVED******REMOVED***,
        value2: ***REMOVED******REMOVED***,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'equal objects (same properties "order")',
        value1: ***REMOVED***a: 1, b: '2'***REMOVED***,
        value2: ***REMOVED***a: 1, b: '2'***REMOVED***,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'equal objects (different properties "order")',
        value1: ***REMOVED***a: 1, b: '2'***REMOVED***,
        value2: ***REMOVED***b: '2', a: 1***REMOVED***,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal objects (extra property)',
        value1: ***REMOVED***a: 1, b: '2'***REMOVED***,
        value2: ***REMOVED***a: 1, b: '2', c: []***REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal objects (different properties)',
        value1: ***REMOVED***a: 1, b: '2', c: 3***REMOVED***,
        value2: ***REMOVED***a: 1, b: '2', d: 3***REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal objects (different properties)',
        value1: ***REMOVED***a: 1, b: '2', c: 3***REMOVED***,
        value2: ***REMOVED***a: 1, b: '2', d: 3***REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'equal objects (same sub-properties)',
        value1: ***REMOVED*** a: [ ***REMOVED*** b: 'c' ***REMOVED*** ] ***REMOVED***,
        value2: ***REMOVED*** a: [ ***REMOVED*** b: 'c' ***REMOVED*** ] ***REMOVED***,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal objects (different sub-property value)',
        value1: ***REMOVED*** a: [ ***REMOVED*** b: 'c' ***REMOVED*** ] ***REMOVED***,
        value2: ***REMOVED*** a: [ ***REMOVED*** b: 'd' ***REMOVED*** ] ***REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal objects (different sub-property)',
        value1: ***REMOVED*** a: [ ***REMOVED*** b: 'c' ***REMOVED*** ] ***REMOVED***,
        value2: ***REMOVED*** a: [ ***REMOVED*** c: 'c' ***REMOVED*** ] ***REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'empty array and empty object are not equal',
        value1: ***REMOVED******REMOVED***,
        value2: [],
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'object with extra undefined properties are not equal #1',
        value1: ***REMOVED******REMOVED***,
        value2: ***REMOVED***foo: undefined***REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'object with extra undefined properties are not equal #2',
        value1: ***REMOVED***foo: undefined***REMOVED***,
        value2: ***REMOVED******REMOVED***,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'object with extra undefined properties are not equal #3',
        value1: ***REMOVED***foo: undefined***REMOVED***,
        value2: ***REMOVED***bar: undefined***REMOVED***,
        equal: false
      ***REMOVED***
    ]
  ***REMOVED***,

  ***REMOVED***
    description: 'arrays',
    tests: [
      ***REMOVED***
        description: 'two empty arrays are equal',
        value1: [],
        value2: [],
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'equal arrays',
        value1: [1, 2, 3],
        value2: [1, 2, 3],
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal arrays (different item)',
        value1: [1, 2, 3],
        value2: [1, 2, 4],
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal arrays (different length)',
        value1: [1, 2, 3],
        value2: [1, 2],
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'equal arrays of objects',
        value1: [***REMOVED***a: 'a'***REMOVED***, ***REMOVED***b: 'b'***REMOVED***],
        value2: [***REMOVED***a: 'a'***REMOVED***, ***REMOVED***b: 'b'***REMOVED***],
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal arrays of objects',
        value1: [***REMOVED***a: 'a'***REMOVED***, ***REMOVED***b: 'b'***REMOVED***],
        value2: [***REMOVED***a: 'a'***REMOVED***, ***REMOVED***b: 'c'***REMOVED***],
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'pseudo array and equivalent array are not equal',
        value1: ***REMOVED***'0': 0, '1': 1, length: 2***REMOVED***,
        value2: [0, 1],
        equal: false
      ***REMOVED***
    ]
  ***REMOVED***,
  ***REMOVED***
    description: 'Date objects',
    tests: [
      ***REMOVED***
        description: 'equal date objects',
        value1: new Date('2017-06-16T21:36:48.362Z'),
        value2: new Date('2017-06-16T21:36:48.362Z'),
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal date objects',
        value1: new Date('2017-06-16T21:36:48.362Z'),
        value2: new Date('2017-01-01T00:00:00.000Z'),
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'date and string are not equal',
        value1: new Date('2017-06-16T21:36:48.362Z'),
        value2: '2017-06-16T21:36:48.362Z',
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'date and object are not equal',
        value1: new Date('2017-06-16T21:36:48.362Z'),
        value2: ***REMOVED******REMOVED***,
        equal: false
      ***REMOVED***
    ]
  ***REMOVED***,
  ***REMOVED***
    description: 'RegExp objects',
    tests: [
      ***REMOVED***
        description: 'equal RegExp objects',
        value1: /foo/,
        value2: /foo/,
        equal: true
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal RegExp objects (different pattern)',
        value1: /foo/,
        value2: /bar/,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'not equal RegExp objects (different flags)',
        value1: /foo/,
        value2: /foo/i,
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'RegExp and string are not equal',
        value1: /foo/,
        value2: 'foo',
        equal: false
      ***REMOVED***,
      ***REMOVED***
        description: 'RegExp and object are not equal',
        value1: /foo/,
        value2: ***REMOVED******REMOVED***,
        equal: false
      ***REMOVED***
    ]
  ***REMOVED***,
  ***REMOVED***
    description: 'sample objects',
    tests: [
      ***REMOVED***
        description: 'big object',
        value1: ***REMOVED***
          prop1: 'value1',
          prop2: 'value2',
          prop3: 'value3',
          prop4: ***REMOVED***
            subProp1: 'sub value1',
            subProp2: ***REMOVED***
              subSubProp1: 'sub sub value1',
              subSubProp2: [1, 2, ***REMOVED***prop2: 1, prop: 2***REMOVED***, 4, 5]
            ***REMOVED***
          ***REMOVED***,
          prop5: 1000,
          prop6: new Date(2016, 2, 10)
        ***REMOVED***,
        value2: ***REMOVED***
          prop5: 1000,
          prop3: 'value3',
          prop1: 'value1',
          prop2: 'value2',
          prop6: new Date('2016/03/10'),
          prop4: ***REMOVED***
            subProp2: ***REMOVED***
              subSubProp1: 'sub sub value1',
              subSubProp2: [1, 2, ***REMOVED***prop2: 1, prop: 2***REMOVED***, 4, 5]
            ***REMOVED***,
            subProp1: 'sub value1'
          ***REMOVED***
        ***REMOVED***,
        equal: true
      ***REMOVED***
    ]
  ***REMOVED***
];
