// webpack/bin/config-yargs.js
var CONFIG_GROUP = "Config options:";
var BASIC_GROUP = "Basic options:";
var MODULE_GROUP = "Module options:";
var OUTPUT_GROUP = "Output options:";
var ADVANCED_GROUP = "Advanced options:";
var RESOLVE_GROUP = "Resolving options:";
var OPTIMIZE_GROUP = "Optimizing options:";

module.exports = function(yargs) ***REMOVED***
	yargs
		.help("help")
		.alias("help", "h")
		.version()
		.alias("version", "v")
		.options(***REMOVED***
			"config": ***REMOVED***
				type: "string",
				describe: "Path to the config file",
				group: CONFIG_GROUP,
				defaultDescription: "webpack.config.js or webpackfile.js",
				requiresArg: true
			***REMOVED***,
			"config-name": ***REMOVED***
				type: "string",
				describe: "Name of the config to use",
				group: CONFIG_GROUP,
				requiresArg: true
			***REMOVED***,
			"env": ***REMOVED***
				describe: "Environment passed to the config, when it is a function",
				group: CONFIG_GROUP
			***REMOVED***,
			"context": ***REMOVED***
				type: "string",
				describe: "The root directory for resolving entry point and stats",
				group: BASIC_GROUP,
				defaultDescription: "The current directory",
				requiresArg: true
			***REMOVED***,
			"entry": ***REMOVED***
				type: "string",
				describe: "The entry point",
				group: BASIC_GROUP,
				requiresArg: true
			***REMOVED***,
			"module-bind": ***REMOVED***
				type: "string",
				describe: "Bind an extension to a loader",
				group: MODULE_GROUP,
				requiresArg: true
			***REMOVED***,
			"module-bind-post": ***REMOVED***
				type: "string",
				describe: "",
				group: MODULE_GROUP,
				requiresArg: true
			***REMOVED***,
			"module-bind-pre": ***REMOVED***
				type: "string",
				describe: "",
				group: MODULE_GROUP,
				requiresArg: true
			***REMOVED***,
			"output-path": ***REMOVED***
				type: "string",
				describe: "The output path for compilation assets",
				group: OUTPUT_GROUP,
				defaultDescription: "The current directory",
				requiresArg: true
			***REMOVED***,
			"output-filename": ***REMOVED***
				type: "string",
				describe: "The output filename of the bundle",
				group: OUTPUT_GROUP,
				defaultDescription: "[name].js",
				requiresArg: true
			***REMOVED***,
			"output-chunk-filename": ***REMOVED***
				type: "string",
				describe: "The output filename for additional chunks",
				group: OUTPUT_GROUP,
				defaultDescription: "filename with [id] instead of [name] or [id] prefixed",
				requiresArg: true
			***REMOVED***,
			"output-source-map-filename": ***REMOVED***
				type: "string",
				describe: "The output filename for the SourceMap",
				group: OUTPUT_GROUP,
				requiresArg: true
			***REMOVED***,
			"output-public-path": ***REMOVED***
				type: "string",
				describe: "The public path for the assets",
				group: OUTPUT_GROUP,
				requiresArg: true
			***REMOVED***,
			"output-jsonp-function": ***REMOVED***
				type: "string",
				describe: "The name of the jsonp function used for chunk loading",
				group: OUTPUT_GROUP,
				requiresArg: true
			***REMOVED***,
			"output-pathinfo": ***REMOVED***
				type: "boolean",
				describe: "Include a comment with the request for every dependency (require, import, etc.)",
				group: OUTPUT_GROUP
			***REMOVED***,
			"output-library": ***REMOVED***
				type: "string",
				describe: "Expose the exports of the entry point as library",
				group: OUTPUT_GROUP,
				requiresArg: true
			***REMOVED***,
			"output-library-target": ***REMOVED***
				type: "string",
				describe: "The type for exposing the exports of the entry point as library",
				group: OUTPUT_GROUP,
				requiresArg: true
			***REMOVED***,
			"records-input-path": ***REMOVED***
				type: "string",
				describe: "Path to the records file (reading)",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"records-output-path": ***REMOVED***
				type: "string",
				describe: "Path to the records file (writing)",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"records-path": ***REMOVED***
				type: "string",
				describe: "Path to the records file",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"define": ***REMOVED***
				type: "string",
				describe: "Define any free var in the bundle",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"target": ***REMOVED***
				type: "string",
				describe: "The targeted execution environment",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"cache": ***REMOVED***
				type: "boolean",
				describe: "Enable in memory caching",
				default: null,
				group: ADVANCED_GROUP,
				defaultDescription: "It's enabled by default when watching"
			***REMOVED***,
			"watch": ***REMOVED***
				type: "boolean",
				alias: "w",
				describe: "Watch the filesystem for changes",
				group: BASIC_GROUP
			***REMOVED***,
			"watch-stdin": ***REMOVED***
				type: "boolean",
				alias: "stdin",
				describe: "Exit the process when stdin is closed",
				group: ADVANCED_GROUP
			***REMOVED***,
			"watch-aggregate-timeout": ***REMOVED***
				describe: "Timeout for gathering changes while watching",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"watch-poll": ***REMOVED***
				type: "string",
				describe: "The polling interval for watching (also enable polling)",
				group: ADVANCED_GROUP
			***REMOVED***,
			"hot": ***REMOVED***
				type: "boolean",
				describe: "Enables Hot Module Replacement",
				group: ADVANCED_GROUP
			***REMOVED***,
			"debug": ***REMOVED***
				type: "boolean",
				describe: "Switch loaders to debug mode",
				group: BASIC_GROUP
			***REMOVED***,
			"devtool": ***REMOVED***
				type: "string",
				describe: "Enable devtool for better debugging experience (Example: --devtool eval-cheap-module-source-map)",
				group: BASIC_GROUP,
				requiresArg: true
			***REMOVED***,
			"resolve-alias": ***REMOVED***
				type: "string",
				describe: "Setup a module alias for resolving (Example: jquery-plugin=jquery.plugin)",
				group: RESOLVE_GROUP,
				requiresArg: true
			***REMOVED***,
			"resolve-extensions": ***REMOVED***
				"type": "array",
				describe: "Setup extensions that should be used to resolve modules (Example: --resolve-extensions .es6,.js)",
				group: RESOLVE_GROUP,
				requiresArg: true
			***REMOVED***,
			"resolve-loader-alias": ***REMOVED***
				type: "string",
				describe: "Setup a loader alias for resolving",
				group: RESOLVE_GROUP,
				requiresArg: true
			***REMOVED***,
			"optimize-max-chunks": ***REMOVED***
				describe: "Try to keep the chunk count below a limit",
				group: OPTIMIZE_GROUP,
				requiresArg: true
			***REMOVED***,
			"optimize-min-chunk-size": ***REMOVED***
				describe: "Try to keep the chunk size above a limit",
				group: OPTIMIZE_GROUP,
				requiresArg: true
			***REMOVED***,
			"optimize-minimize": ***REMOVED***
				type: "boolean",
				describe: "Minimize javascript and switches loaders to minimizing",
				group: OPTIMIZE_GROUP
			***REMOVED***,
			"prefetch": ***REMOVED***
				type: "string",
				describe: "Prefetch this request (Example: --prefetch ./file.js)",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"provide": ***REMOVED***
				type: "string",
				describe: "Provide these modules as free vars in all modules (Example: --provide jQuery=jquery)",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"labeled-modules": ***REMOVED***
				type: "boolean",
				describe: "Enables labeled modules",
				group: ADVANCED_GROUP
			***REMOVED***,
			"plugin": ***REMOVED***
				type: "string",
				describe: "Load this plugin",
				group: ADVANCED_GROUP,
				requiresArg: true
			***REMOVED***,
			"bail": ***REMOVED***
				type: "boolean",
				describe: "Abort the compilation on first error",
				group: ADVANCED_GROUP,
				default: null
			***REMOVED***,
			"profile": ***REMOVED***
				type: "boolean",
				describe: "Profile the compilation and include information in stats",
				group: ADVANCED_GROUP,
				default: null
			***REMOVED***,
			"d": ***REMOVED***
				type: "boolean",
				describe: "shortcut for --debug --devtool eval-cheap-module-source-map --output-pathinfo",
				group: BASIC_GROUP
			***REMOVED***,
			"p": ***REMOVED***
				type: "boolean",
				describe: "shortcut for --optimize-minimize --define process.env.NODE_ENV=\"production\"",
				group: BASIC_GROUP
			***REMOVED***
		***REMOVED***).strict();
***REMOVED***;
