// webpack/bin/webpack.js
#!/usr/bin/env node

/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var path = require("path");

// Local version replace global one
try ***REMOVED***
	var localWebpack = require.resolve(path.join(process.cwd(), "node_modules", "webpack", "bin", "webpack.js"));
	if(__filename !== localWebpack) ***REMOVED***
		return require(localWebpack);
	***REMOVED***
***REMOVED*** catch(e) ***REMOVED******REMOVED***
var yargs = require("yargs")
	.usage("webpack " + require("../package.json").version + "\n" +
		"Usage: https://webpack.js.org/api/cli/\n" +
		"Usage without config file: webpack <entry> [<entry>] <output>\n" +
		"Usage with config file: webpack");

require("./config-yargs")(yargs);

var DISPLAY_GROUP = "Stats options:";
var BASIC_GROUP = "Basic options:";

yargs.options(***REMOVED***
	"json": ***REMOVED***
		type: "boolean",
		alias: "j",
		describe: "Prints the result as JSON."
	***REMOVED***,
	"progress": ***REMOVED***
		type: "boolean",
		describe: "Print compilation progress in percentage",
		group: BASIC_GROUP
	***REMOVED***,
	"color": ***REMOVED***
		type: "boolean",
		alias: "colors",
		default: function supportsColor() ***REMOVED***
			return require("supports-color");
		***REMOVED***,
		group: DISPLAY_GROUP,
		describe: "Enables/Disables colors on the console"
	***REMOVED***,
	"sort-modules-by": ***REMOVED***
		type: "string",
		group: DISPLAY_GROUP,
		describe: "Sorts the modules list by property in module"
	***REMOVED***,
	"sort-chunks-by": ***REMOVED***
		type: "string",
		group: DISPLAY_GROUP,
		describe: "Sorts the chunks list by property in chunk"
	***REMOVED***,
	"sort-assets-by": ***REMOVED***
		type: "string",
		group: DISPLAY_GROUP,
		describe: "Sorts the assets list by property in asset"
	***REMOVED***,
	"hide-modules": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Hides info about modules"
	***REMOVED***,
	"display-exclude": ***REMOVED***
		type: "string",
		group: DISPLAY_GROUP,
		describe: "Exclude modules in the output"
	***REMOVED***,
	"display-modules": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display even excluded modules in the output"
	***REMOVED***,
	"display-max-modules": ***REMOVED***
		type: "number",
		group: DISPLAY_GROUP,
		describe: "Sets the maximum number of visible modules in output"
	***REMOVED***,
	"display-chunks": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display chunks in the output"
	***REMOVED***,
	"display-entrypoints": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display entry points in the output"
	***REMOVED***,
	"display-origins": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display origins of chunks in the output"
	***REMOVED***,
	"display-cached": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display also cached modules in the output"
	***REMOVED***,
	"display-cached-assets": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display also cached assets in the output"
	***REMOVED***,
	"display-reasons": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display reasons about module inclusion in the output"
	***REMOVED***,
	"display-depth": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display distance from entry point for each module"
	***REMOVED***,
	"display-used-exports": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display information about used exports in modules (Tree Shaking)"
	***REMOVED***,
	"display-provided-exports": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display information about exports provided from modules"
	***REMOVED***,
	"display-optimization-bailout": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display information about why optimization bailed out for modules"
	***REMOVED***,
	"display-error-details": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Display details about errors"
	***REMOVED***,
	"display": ***REMOVED***
		type: "string",
		group: DISPLAY_GROUP,
		describe: "Select display preset (verbose, detailed, normal, minimal, errors-only, none)"
	***REMOVED***,
	"verbose": ***REMOVED***
		type: "boolean",
		group: DISPLAY_GROUP,
		describe: "Show more details"
	***REMOVED***
***REMOVED***);

// yargs will terminate the process early when the user uses help or version.
// This causes large help outputs to be cut short (https://github.com/nodejs/node/wiki/API-changes-between-v0.10-and-v4#process).
// To prevent this we use the yargs.parse API and exit the process normally
yargs.parse(process.argv.slice(2), (err, argv, output) => ***REMOVED***

	// arguments validation failed
	if(err && output) ***REMOVED***
		console.error(output);
		process.exitCode = 1;
		return;
	***REMOVED***

	// help or version info
	if(output) ***REMOVED***
		console.log(output);
		return;
	***REMOVED***

	if(argv.verbose) ***REMOVED***
		argv["display"] = "verbose";
	***REMOVED***

	var options = require("./convert-argv")(yargs, argv);

	function ifArg(name, fn, init) ***REMOVED***
		if(Array.isArray(argv[name])) ***REMOVED***
			if(init) init();
			argv[name].forEach(fn);
		***REMOVED*** else if(typeof argv[name] !== "undefined") ***REMOVED***
			if(init) init();
			fn(argv[name], -1);
		***REMOVED***
	***REMOVED***

	function processOptions(options) ***REMOVED***
		// process Promise
		if(typeof options.then === "function") ***REMOVED***
			options.then(processOptions).catch(function(err) ***REMOVED***
				console.error(err.stack || err);
				process.exit(1); // eslint-disable-line
			***REMOVED***);
			return;
		***REMOVED***

		var firstOptions = [].concat(options)[0];
		var statsPresetToOptions = require("../lib/Stats.js").presetToOptions;

		var outputOptions = options.stats;
		if(typeof outputOptions === "boolean" || typeof outputOptions === "string") ***REMOVED***
			outputOptions = statsPresetToOptions(outputOptions);
		***REMOVED*** else if(!outputOptions) ***REMOVED***
			outputOptions = ***REMOVED******REMOVED***;
		***REMOVED***

		ifArg("display", function(preset) ***REMOVED***
			outputOptions = statsPresetToOptions(preset);
		***REMOVED***);

		outputOptions = Object.create(outputOptions);
		if(Array.isArray(options) && !outputOptions.children) ***REMOVED***
			outputOptions.children = options.map(o => o.stats);
		***REMOVED***
		if(typeof outputOptions.context === "undefined")
			outputOptions.context = firstOptions.context;

		ifArg("env", function(value) ***REMOVED***
			if(outputOptions.env) ***REMOVED***
				outputOptions._env = value;
			***REMOVED***
		***REMOVED***);

		ifArg("json", function(bool) ***REMOVED***
			if(bool)
				outputOptions.json = bool;
		***REMOVED***);

		if(typeof outputOptions.colors === "undefined")
			outputOptions.colors = require("supports-color");

		ifArg("sort-modules-by", function(value) ***REMOVED***
			outputOptions.modulesSort = value;
		***REMOVED***);

		ifArg("sort-chunks-by", function(value) ***REMOVED***
			outputOptions.chunksSort = value;
		***REMOVED***);

		ifArg("sort-assets-by", function(value) ***REMOVED***
			outputOptions.assetsSort = value;
		***REMOVED***);

		ifArg("display-exclude", function(value) ***REMOVED***
			outputOptions.exclude = value;
		***REMOVED***);

		if(!outputOptions.json) ***REMOVED***
			if(typeof outputOptions.cached === "undefined")
				outputOptions.cached = false;
			if(typeof outputOptions.cachedAssets === "undefined")
				outputOptions.cachedAssets = false;

			ifArg("display-chunks", function(bool) ***REMOVED***
				if(bool) ***REMOVED***
					outputOptions.modules = false;
					outputOptions.chunks = true;
					outputOptions.chunkModules = true;
				***REMOVED***
			***REMOVED***);

			ifArg("display-entrypoints", function(bool) ***REMOVED***
				if(bool)
					outputOptions.entrypoints = true;
			***REMOVED***);

			ifArg("display-reasons", function(bool) ***REMOVED***
				if(bool)
					outputOptions.reasons = true;
			***REMOVED***);

			ifArg("display-depth", function(bool) ***REMOVED***
				if(bool)
					outputOptions.depth = true;
			***REMOVED***);

			ifArg("display-used-exports", function(bool) ***REMOVED***
				if(bool)
					outputOptions.usedExports = true;
			***REMOVED***);

			ifArg("display-provided-exports", function(bool) ***REMOVED***
				if(bool)
					outputOptions.providedExports = true;
			***REMOVED***);

			ifArg("display-optimization-bailout", function(bool) ***REMOVED***
				if(bool)
					outputOptions.optimizationBailout = bool;
			***REMOVED***);

			ifArg("display-error-details", function(bool) ***REMOVED***
				if(bool)
					outputOptions.errorDetails = true;
			***REMOVED***);

			ifArg("display-origins", function(bool) ***REMOVED***
				if(bool)
					outputOptions.chunkOrigins = true;
			***REMOVED***);

			ifArg("display-max-modules", function(value) ***REMOVED***
				outputOptions.maxModules = +value;
			***REMOVED***);

			ifArg("display-cached", function(bool) ***REMOVED***
				if(bool)
					outputOptions.cached = true;
			***REMOVED***);

			ifArg("display-cached-assets", function(bool) ***REMOVED***
				if(bool)
					outputOptions.cachedAssets = true;
			***REMOVED***);

			if(!outputOptions.exclude)
				outputOptions.exclude = ["node_modules", "bower_components", "components"];

			if(argv["display-modules"]) ***REMOVED***
				outputOptions.maxModules = Infinity;
				outputOptions.exclude = undefined;
				outputOptions.modules = true;
			***REMOVED***
		***REMOVED***

		ifArg("hide-modules", function(bool) ***REMOVED***
			if(bool) ***REMOVED***
				outputOptions.modules = false;
				outputOptions.chunkModules = false;
			***REMOVED***
		***REMOVED***);

		var webpack = require("../lib/webpack.js");

		Error.stackTraceLimit = 30;
		var lastHash = null;
		var compiler;
		try ***REMOVED***
			compiler = webpack(options);
		***REMOVED*** catch(err) ***REMOVED***
			if(err.name === "WebpackOptionsValidationError") ***REMOVED***
				if(argv.color)
					console.error(
						`\u001b[1m\u001b[31m$***REMOVED***err.message***REMOVED***\u001b[39m\u001b[22m`
					);
				else
					console.error(err.message);
				// eslint-disable-next-line no-process-exit
				process.exit(1);
			***REMOVED***

			throw err;
		***REMOVED***

		if(argv.progress) ***REMOVED***
			var ProgressPlugin = require("../lib/ProgressPlugin");
			compiler.apply(new ProgressPlugin(***REMOVED***
				profile: argv.profile
			***REMOVED***));
		***REMOVED***

		function compilerCallback(err, stats) ***REMOVED***
			if(!options.watch || err) ***REMOVED***
				// Do not keep cache anymore
				compiler.purgeInputFileSystem();
			***REMOVED***
			if(err) ***REMOVED***
				lastHash = null;
				console.error(err.stack || err);
				if(err.details) console.error(err.details);
				process.exit(1); // eslint-disable-line
			***REMOVED***
			if(outputOptions.json) ***REMOVED***
				process.stdout.write(JSON.stringify(stats.toJson(outputOptions), null, 2) + "\n");
			***REMOVED*** else if(stats.hash !== lastHash) ***REMOVED***
				lastHash = stats.hash;
				var statsString = stats.toString(outputOptions);
				if(statsString)
					process.stdout.write(statsString + "\n");
			***REMOVED***
			if(!options.watch && stats.hasErrors()) ***REMOVED***
				process.exitCode = 2;
			***REMOVED***
		***REMOVED***
		if(firstOptions.watch || options.watch) ***REMOVED***
			var watchOptions = firstOptions.watchOptions || firstOptions.watch || options.watch || ***REMOVED******REMOVED***;
			if(watchOptions.stdin) ***REMOVED***
				process.stdin.on("end", function() ***REMOVED***
					process.exit(); // eslint-disable-line
				***REMOVED***);
				process.stdin.resume();
			***REMOVED***
			compiler.watch(watchOptions, compilerCallback);
			console.log("\nWebpack is watching the files…\n");
		***REMOVED*** else
			compiler.run(compilerCallback);

	***REMOVED***

	processOptions(options);

***REMOVED***);
