// fs.realpath/old.js
// Copyright Joyent, Inc. and other Node contributors.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.

var pathModule = require('path');
var isWindows = process.platform === 'win32';
var fs = require('fs');

// JavaScript implementation of realpath, ported from node pre-v6

var DEBUG = process.env.NODE_DEBUG && /fs/.test(process.env.NODE_DEBUG);

function rethrow() ***REMOVED***
  // Only enable in debug mode. A backtrace uses ~1000 bytes of heap space and
  // is fairly slow to generate.
  var callback;
  if (DEBUG) ***REMOVED***
    var backtrace = new Error;
    callback = debugCallback;
  ***REMOVED*** else
    callback = missingCallback;

  return callback;

  function debugCallback(err) ***REMOVED***
    if (err) ***REMOVED***
      backtrace.message = err.message;
      err = backtrace;
      missingCallback(err);
    ***REMOVED***
  ***REMOVED***

  function missingCallback(err) ***REMOVED***
    if (err) ***REMOVED***
      if (process.throwDeprecation)
        throw err;  // Forgot a callback but don't know where? Use NODE_DEBUG=fs
      else if (!process.noDeprecation) ***REMOVED***
        var msg = 'fs: missing callback ' + (err.stack || err.message);
        if (process.traceDeprecation)
          console.trace(msg);
        else
          console.error(msg);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function maybeCallback(cb) ***REMOVED***
  return typeof cb === 'function' ? cb : rethrow();
***REMOVED***

var normalize = pathModule.normalize;

// Regexp that finds the next partion of a (partial) path
// result is [base_with_slash, base], e.g. ['somedir/', 'somedir']
if (isWindows) ***REMOVED***
  var nextPartRe = /(.*?)(?:[\/\\]+|$)/g;
***REMOVED*** else ***REMOVED***
  var nextPartRe = /(.*?)(?:[\/]+|$)/g;
***REMOVED***

// Regex to find the device root, including trailing slash. E.g. 'c:\\'.
if (isWindows) ***REMOVED***
  var splitRootRe = /^(?:[a-zA-Z]:|[\\\/]***REMOVED***2***REMOVED***[^\\\/]+[\\\/][^\\\/]+)?[\\\/]*/;
***REMOVED*** else ***REMOVED***
  var splitRootRe = /^[\/]*/;
***REMOVED***

exports.realpathSync = function realpathSync(p, cache) ***REMOVED***
  // make p is absolute
  p = pathModule.resolve(p);

  if (cache && Object.prototype.hasOwnProperty.call(cache, p)) ***REMOVED***
    return cache[p];
  ***REMOVED***

  var original = p,
      seenLinks = ***REMOVED******REMOVED***,
      knownHard = ***REMOVED******REMOVED***;

  // current character position in p
  var pos;
  // the partial path so far, including a trailing slash if any
  var current;
  // the partial path without a trailing slash (except when pointing at a root)
  var base;
  // the partial path scanned in the previous round, with slash
  var previous;

  start();

  function start() ***REMOVED***
    // Skip over roots
    var m = splitRootRe.exec(p);
    pos = m[0].length;
    current = m[0];
    base = m[0];
    previous = '';

    // On windows, check that the root exists. On unix there is no need.
    if (isWindows && !knownHard[base]) ***REMOVED***
      fs.lstatSync(base);
      knownHard[base] = true;
    ***REMOVED***
  ***REMOVED***

  // walk down the path, swapping out linked pathparts for their real
  // values
  // NB: p.length changes.
  while (pos < p.length) ***REMOVED***
    // find the next part
    nextPartRe.lastIndex = pos;
    var result = nextPartRe.exec(p);
    previous = current;
    current += result[0];
    base = previous + result[1];
    pos = nextPartRe.lastIndex;

    // continue if not a symlink
    if (knownHard[base] || (cache && cache[base] === base)) ***REMOVED***
      continue;
    ***REMOVED***

    var resolvedLink;
    if (cache && Object.prototype.hasOwnProperty.call(cache, base)) ***REMOVED***
      // some known symbolic link.  no need to stat again.
      resolvedLink = cache[base];
    ***REMOVED*** else ***REMOVED***
      var stat = fs.lstatSync(base);
      if (!stat.isSymbolicLink()) ***REMOVED***
        knownHard[base] = true;
        if (cache) cache[base] = base;
        continue;
      ***REMOVED***

      // read the link if it wasn't read before
      // dev/ino always return 0 on windows, so skip the check.
      var linkTarget = null;
      if (!isWindows) ***REMOVED***
        var id = stat.dev.toString(32) + ':' + stat.ino.toString(32);
        if (seenLinks.hasOwnProperty(id)) ***REMOVED***
          linkTarget = seenLinks[id];
        ***REMOVED***
      ***REMOVED***
      if (linkTarget === null) ***REMOVED***
        fs.statSync(base);
        linkTarget = fs.readlinkSync(base);
      ***REMOVED***
      resolvedLink = pathModule.resolve(previous, linkTarget);
      // track this, if given a cache.
      if (cache) cache[base] = resolvedLink;
      if (!isWindows) seenLinks[id] = linkTarget;
    ***REMOVED***

    // resolve the link, then start over
    p = pathModule.resolve(resolvedLink, p.slice(pos));
    start();
  ***REMOVED***

  if (cache) cache[original] = p;

  return p;
***REMOVED***;


exports.realpath = function realpath(p, cache, cb) ***REMOVED***
  if (typeof cb !== 'function') ***REMOVED***
    cb = maybeCallback(cache);
    cache = null;
  ***REMOVED***

  // make p is absolute
  p = pathModule.resolve(p);

  if (cache && Object.prototype.hasOwnProperty.call(cache, p)) ***REMOVED***
    return process.nextTick(cb.bind(null, null, cache[p]));
  ***REMOVED***

  var original = p,
      seenLinks = ***REMOVED******REMOVED***,
      knownHard = ***REMOVED******REMOVED***;

  // current character position in p
  var pos;
  // the partial path so far, including a trailing slash if any
  var current;
  // the partial path without a trailing slash (except when pointing at a root)
  var base;
  // the partial path scanned in the previous round, with slash
  var previous;

  start();

  function start() ***REMOVED***
    // Skip over roots
    var m = splitRootRe.exec(p);
    pos = m[0].length;
    current = m[0];
    base = m[0];
    previous = '';

    // On windows, check that the root exists. On unix there is no need.
    if (isWindows && !knownHard[base]) ***REMOVED***
      fs.lstat(base, function(err) ***REMOVED***
        if (err) return cb(err);
        knownHard[base] = true;
        LOOP();
      ***REMOVED***);
    ***REMOVED*** else ***REMOVED***
      process.nextTick(LOOP);
    ***REMOVED***
  ***REMOVED***

  // walk down the path, swapping out linked pathparts for their real
  // values
  function LOOP() ***REMOVED***
    // stop if scanned past end of path
    if (pos >= p.length) ***REMOVED***
      if (cache) cache[original] = p;
      return cb(null, p);
    ***REMOVED***

    // find the next part
    nextPartRe.lastIndex = pos;
    var result = nextPartRe.exec(p);
    previous = current;
    current += result[0];
    base = previous + result[1];
    pos = nextPartRe.lastIndex;

    // continue if not a symlink
    if (knownHard[base] || (cache && cache[base] === base)) ***REMOVED***
      return process.nextTick(LOOP);
    ***REMOVED***

    if (cache && Object.prototype.hasOwnProperty.call(cache, base)) ***REMOVED***
      // known symbolic link.  no need to stat again.
      return gotResolvedLink(cache[base]);
    ***REMOVED***

    return fs.lstat(base, gotStat);
  ***REMOVED***

  function gotStat(err, stat) ***REMOVED***
    if (err) return cb(err);

    // if not a symlink, skip to the next path part
    if (!stat.isSymbolicLink()) ***REMOVED***
      knownHard[base] = true;
      if (cache) cache[base] = base;
      return process.nextTick(LOOP);
    ***REMOVED***

    // stat & read the link if not read before
    // call gotTarget as soon as the link target is known
    // dev/ino always return 0 on windows, so skip the check.
    if (!isWindows) ***REMOVED***
      var id = stat.dev.toString(32) + ':' + stat.ino.toString(32);
      if (seenLinks.hasOwnProperty(id)) ***REMOVED***
        return gotTarget(null, seenLinks[id], base);
      ***REMOVED***
    ***REMOVED***
    fs.stat(base, function(err) ***REMOVED***
      if (err) return cb(err);

      fs.readlink(base, function(err, target) ***REMOVED***
        if (!isWindows) seenLinks[id] = target;
        gotTarget(err, target);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***

  function gotTarget(err, target, base) ***REMOVED***
    if (err) return cb(err);

    var resolvedLink = pathModule.resolve(previous, target);
    if (cache) cache[base] = resolvedLink;
    gotResolvedLink(resolvedLink);
  ***REMOVED***

  function gotResolvedLink(resolvedLink) ***REMOVED***
    // resolve the link, then start over
    p = pathModule.resolve(resolvedLink, p.slice(pos));
    start();
  ***REMOVED***
***REMOVED***;
