// fill-range/index.js
/*!
 * fill-range <https://github.com/jonschlinkert/fill-range>
 *
 * Copyright (c) 2014-2015, 2017, Jon Schlinkert.
 * Released under the MIT License.
 */

'use strict';

var util = require('util');
var isNumber = require('is-number');
var extend = require('extend-shallow');
var repeat = require('repeat-string');
var toRegex = require('to-regex-range');

/**
 * Return a range of numbers or letters.
 *
 * @param  ***REMOVED***String***REMOVED*** `start` Start of the range
 * @param  ***REMOVED***String***REMOVED*** `stop` End of the range
 * @param  ***REMOVED***String***REMOVED*** `step` Increment or decrement to use.
 * @param  ***REMOVED***Function***REMOVED*** `fn` Custom function to modify each element in the range.
 * @return ***REMOVED***Array***REMOVED***
 */

function fillRange(start, stop, step, options) ***REMOVED***
  if (typeof start === 'undefined') ***REMOVED***
    return [];
  ***REMOVED***

  if (typeof stop === 'undefined' || start === stop) ***REMOVED***
    // special case, for handling negative zero
    var isString = typeof start === 'string';
    if (isNumber(start) && !toNumber(start)) ***REMOVED***
      return [isString ? '0' : 0];
    ***REMOVED***
    return [start];
  ***REMOVED***

  if (typeof step !== 'number' && typeof step !== 'string') ***REMOVED***
    options = step;
    step = undefined;
  ***REMOVED***

  if (typeof options === 'function') ***REMOVED***
    options = ***REMOVED*** transform: options ***REMOVED***;
  ***REMOVED***

  var opts = extend(***REMOVED***step: step***REMOVED***, options);
  if (opts.step && !isValidNumber(opts.step)) ***REMOVED***
    if (opts.strictRanges === true) ***REMOVED***
      throw new TypeError('expected options.step to be a number');
    ***REMOVED***
    return [];
  ***REMOVED***

  opts.isNumber = isValidNumber(start) && isValidNumber(stop);
  if (!opts.isNumber && !isValid(start, stop)) ***REMOVED***
    if (opts.strictRanges === true) ***REMOVED***
      throw new RangeError('invalid range arguments: ' + util.inspect([start, stop]));
    ***REMOVED***
    return [];
  ***REMOVED***

  opts.isPadded = isPadded(start) || isPadded(stop);
  opts.toString = opts.stringify
    || typeof opts.step === 'string'
    || typeof start === 'string'
    || typeof stop === 'string'
    || !opts.isNumber;

  if (opts.isPadded) ***REMOVED***
    opts.maxLength = Math.max(String(start).length, String(stop).length);
  ***REMOVED***

  // support legacy minimatch/fill-range options
  if (typeof opts.optimize === 'boolean') opts.toRegex = opts.optimize;
  if (typeof opts.makeRe === 'boolean') opts.toRegex = opts.makeRe;
  return expand(start, stop, opts);
***REMOVED***

function expand(start, stop, options) ***REMOVED***
  var a = options.isNumber ? toNumber(start) : start.charCodeAt(0);
  var b = options.isNumber ? toNumber(stop) : stop.charCodeAt(0);

  var step = Math.abs(toNumber(options.step)) || 1;
  if (options.toRegex && step === 1) ***REMOVED***
    return toRange(a, b, start, stop, options);
  ***REMOVED***

  var zero = ***REMOVED***greater: [], lesser: []***REMOVED***;
  var asc = a < b;
  var arr = new Array(Math.round((asc ? b - a : a - b) / step));
  var idx = 0;

  while (asc ? a <= b : a >= b) ***REMOVED***
    var val = options.isNumber ? a : String.fromCharCode(a);
    if (options.toRegex && (val >= 0 || !options.isNumber)) ***REMOVED***
      zero.greater.push(val);
    ***REMOVED*** else ***REMOVED***
      zero.lesser.push(Math.abs(val));
    ***REMOVED***

    if (options.isPadded) ***REMOVED***
      val = zeros(val, options);
    ***REMOVED***

    if (options.toString) ***REMOVED***
      val = String(val);
    ***REMOVED***

    if (typeof options.transform === 'function') ***REMOVED***
      arr[idx++] = options.transform(val, a, b, step, idx, arr, options);
    ***REMOVED*** else ***REMOVED***
      arr[idx++] = val;
    ***REMOVED***

    if (asc) ***REMOVED***
      a += step;
    ***REMOVED*** else ***REMOVED***
      a -= step;
    ***REMOVED***
  ***REMOVED***

  if (options.toRegex === true) ***REMOVED***
    return toSequence(arr, zero, options);
  ***REMOVED***
  return arr;
***REMOVED***

function toRange(a, b, start, stop, options) ***REMOVED***
  if (options.isPadded) ***REMOVED***
    return toRegex(start, stop, options);
  ***REMOVED***

  if (options.isNumber) ***REMOVED***
    return toRegex(Math.min(a, b), Math.max(a, b), options);
  ***REMOVED***

  var start = String.fromCharCode(Math.min(a, b));
  var stop = String.fromCharCode(Math.max(a, b));
  return '[' + start + '-' + stop + ']';
***REMOVED***

function toSequence(arr, zeros, options) ***REMOVED***
  var greater = '', lesser = '';
  if (zeros.greater.length) ***REMOVED***
    greater = zeros.greater.join('|');
  ***REMOVED***
  if (zeros.lesser.length) ***REMOVED***
    lesser = '-(' + zeros.lesser.join('|') + ')';
  ***REMOVED***
  var res = greater && lesser
    ? greater + '|' + lesser
    : greater || lesser;

  if (options.capture) ***REMOVED***
    return '(' + res + ')';
  ***REMOVED***
  return res;
***REMOVED***

function zeros(val, options) ***REMOVED***
  if (options.isPadded) ***REMOVED***
    var str = String(val);
    var len = str.length;
    var dash = '';
    if (str.charAt(0) === '-') ***REMOVED***
      dash = '-';
      str = str.slice(1);
    ***REMOVED***
    var diff = options.maxLength - len;
    var pad = repeat('0', diff);
    val = (dash + pad + str);
  ***REMOVED***
  if (options.stringify) ***REMOVED***
    return String(val);
  ***REMOVED***
  return val;
***REMOVED***

function toNumber(val) ***REMOVED***
  return Number(val) || 0;
***REMOVED***

function isPadded(str) ***REMOVED***
  return /^-?0\d/.test(str);
***REMOVED***

function isValid(min, max) ***REMOVED***
  return (isValidNumber(min) || isValidLetter(min))
      && (isValidNumber(max) || isValidLetter(max));
***REMOVED***

function isValidLetter(ch) ***REMOVED***
  return typeof ch === 'string' && ch.length === 1 && /^\w+$/.test(ch);
***REMOVED***

function isValidNumber(n) ***REMOVED***
  return isNumber(n) && !/\./.test(n);
***REMOVED***

/**
 * Expose `fillRange`
 * @type ***REMOVED***Function***REMOVED***
 */

module.exports = fillRange;
