// power-assert-renderer-diagram/index.js
'use strict';

var BaseRenderer = require('power-assert-renderer-base');
var inherits = require('util').inherits;
var forEach = require('core-js/library/fn/array/for-each');
var stringifier = require('stringifier');
var stringWidth = require('power-assert-util-string-width');
var assign = require('core-js/library/fn/object/assign');
var defaultOptions = require('./lib/default-options');

/**
 * options.stringify [function]
 * options.maxDepth [number]
 * options.lineSeparator [string]
 * options.anonymous [string]
 * options.circular [string]
 * 
 * options.widthOf [function]
 * options.ambiguousEastAsianCharWidth [number]
 */
function DiagramRenderer (config) ***REMOVED***
    BaseRenderer.call(this);
    this.config = assign(***REMOVED******REMOVED***, defaultOptions(), config);
    this.events = [];
    if (typeof this.config.stringify === 'function') ***REMOVED***
        this.stringify = this.config.stringify;
    ***REMOVED*** else ***REMOVED***
        this.stringify = stringifier(this.config);
    ***REMOVED***
    if (typeof this.config.widthOf === 'function') ***REMOVED***
        this.widthOf = this.config.widthOf;
    ***REMOVED*** else ***REMOVED***
        this.widthOf = (this.config.ambiguousEastAsianCharWidth === 1) ? stringWidth.narrow : stringWidth;
    ***REMOVED***
    this.initialVertivalBarLength = 1;
***REMOVED***
inherits(DiagramRenderer, BaseRenderer);

DiagramRenderer.prototype.onStart = function (context) ***REMOVED***
    this.assertionLine = context.source.content;
    this.initializeRows();
***REMOVED***;

DiagramRenderer.prototype.onData = function (esNode) ***REMOVED***
    if (!esNode.isCaptured) ***REMOVED***
        return;
    ***REMOVED***
    this.events.push(***REMOVED***value: esNode.value, leftIndex: esNode.range[0]***REMOVED***);
***REMOVED***;

DiagramRenderer.prototype.onEnd = function () ***REMOVED***
    this.events.sort(rightToLeft);
    this.constructRows(this.events);
    var _this = this;
    forEach(this.rows, function (columns) ***REMOVED***
        _this.write(columns.join(''));
    ***REMOVED***);
***REMOVED***;

DiagramRenderer.prototype.initializeRows = function () ***REMOVED***
    this.rows = [];
    for (var i = 0; i <= this.initialVertivalBarLength; i += 1) ***REMOVED***
        this.addOneMoreRow();
    ***REMOVED***
***REMOVED***;

DiagramRenderer.prototype.newRowFor = function (assertionLine) ***REMOVED***
    return createRow(this.widthOf(assertionLine), ' ');
***REMOVED***;

DiagramRenderer.prototype.addOneMoreRow = function () ***REMOVED***
    this.rows.push(this.newRowFor(this.assertionLine));
***REMOVED***;

DiagramRenderer.prototype.lastRow = function () ***REMOVED***
    return this.rows[this.rows.length - 1];
***REMOVED***;

DiagramRenderer.prototype.renderVerticalBarAt = function (columnIndex) ***REMOVED***
    var i, lastRowIndex = this.rows.length - 1;
    for (i = 0; i < lastRowIndex; i += 1) ***REMOVED***
        this.rows[i].splice(columnIndex, 1, '|');
    ***REMOVED***
***REMOVED***;

DiagramRenderer.prototype.renderValueAt = function (columnIndex, dumpedValue) ***REMOVED***
    var i, width = this.widthOf(dumpedValue);
    for (i = 0; i < width; i += 1) ***REMOVED***
        this.lastRow().splice(columnIndex + i, 1, dumpedValue.charAt(i));
    ***REMOVED***
***REMOVED***;

DiagramRenderer.prototype.isOverlapped = function (prevCapturing, nextCaputuring, dumpedValue) ***REMOVED***
    return (typeof prevCapturing !== 'undefined') && this.startColumnFor(prevCapturing) <= (this.startColumnFor(nextCaputuring) + this.widthOf(dumpedValue));
***REMOVED***;

DiagramRenderer.prototype.constructRows = function (capturedEvents) ***REMOVED***
    var that = this;
    var prevCaptured;
    forEach(capturedEvents, function (captured) ***REMOVED***
        var dumpedValue = that.stringify(captured.value);
        if (that.isOverlapped(prevCaptured, captured, dumpedValue)) ***REMOVED***
            that.addOneMoreRow();
        ***REMOVED***
        that.renderVerticalBarAt(that.startColumnFor(captured));
        that.renderValueAt(that.startColumnFor(captured), dumpedValue);
        prevCaptured = captured;
    ***REMOVED***);
***REMOVED***;

DiagramRenderer.prototype.startColumnFor = function (captured) ***REMOVED***
    return this.widthOf(this.assertionLine.slice(0, captured.leftIndex));
***REMOVED***;

function createRow (numCols, initial) ***REMOVED***
    var row = [], i;
    for(i = 0; i < numCols; i += 1) ***REMOVED***
        row[i] = initial;
    ***REMOVED***
    return row;
***REMOVED***

function rightToLeft (a, b) ***REMOVED***
    return b.leftIndex - a.leftIndex;
***REMOVED***

module.exports = DiagramRenderer;
