// node-forge/lib/http.js
/**
 * HTTP client-side implementation that uses forge.net sockets.
 *
 * @author Dave Longley
 *
 * Copyright (c) 2010-2014 Digital Bazaar, Inc. All rights reserved.
 */
var forge = require('./forge');
require('./debug');
require('./tls');
require('./util');

// define http namespace
var http = module.exports = forge.http = forge.http || ***REMOVED******REMOVED***;

// logging category
var cat = 'forge.http';

// add array of clients to debug storage
if(forge.debug) ***REMOVED***
  forge.debug.set('forge.http', 'clients', []);
***REMOVED***

// normalizes an http header field name
var _normalize = function(name) ***REMOVED***
  return name.toLowerCase().replace(/(^.)|(-.)/g,
    function(a) ***REMOVED***return a.toUpperCase();***REMOVED***);
***REMOVED***;

/**
 * Gets the local storage ID for the given client.
 *
 * @param client the client to get the local storage ID for.
 *
 * @return the local storage ID to use.
 */
var _getStorageId = function(client) ***REMOVED***
  // TODO: include browser in ID to avoid sharing cookies between
  // browsers (if this is undesirable)
  // navigator.userAgent
  return 'forge.http.' +
    client.url.scheme + '.' +
    client.url.host + '.' +
    client.url.port;
***REMOVED***;

/**
 * Loads persistent cookies from disk for the given client.
 *
 * @param client the client.
 */
var _loadCookies = function(client) ***REMOVED***
  if(client.persistCookies) ***REMOVED***
    try ***REMOVED***
      var cookies = forge.util.getItem(
        client.socketPool.flashApi,
        _getStorageId(client), 'cookies');
      client.cookies = cookies || ***REMOVED******REMOVED***;
    ***REMOVED*** catch(ex) ***REMOVED***
      // no flash storage available, just silently fail
      // TODO: i assume we want this logged somewhere or
      // should it actually generate an error
      //forge.log.error(cat, ex);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Saves persistent cookies on disk for the given client.
 *
 * @param client the client.
 */
var _saveCookies = function(client) ***REMOVED***
  if(client.persistCookies) ***REMOVED***
    try ***REMOVED***
      forge.util.setItem(
        client.socketPool.flashApi,
        _getStorageId(client), 'cookies', client.cookies);
    ***REMOVED*** catch(ex) ***REMOVED***
      // no flash storage available, just silently fail
      // TODO: i assume we want this logged somewhere or
      // should it actually generate an error
      //forge.log.error(cat, ex);
    ***REMOVED***
  ***REMOVED***

  // FIXME: remove me
  _loadCookies(client);
***REMOVED***;

/**
 * Clears persistent cookies on disk for the given client.
 *
 * @param client the client.
 */
var _clearCookies = function(client) ***REMOVED***
  if(client.persistCookies) ***REMOVED***
    try ***REMOVED***
      // only thing stored is 'cookies', so clear whole storage
      forge.util.clearItems(
        client.socketPool.flashApi,
        _getStorageId(client));
    ***REMOVED*** catch(ex) ***REMOVED***
      // no flash storage available, just silently fail
      // TODO: i assume we want this logged somewhere or
      // should it actually generate an error
      //forge.log.error(cat, ex);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Connects and sends a request.
 *
 * @param client the http client.
 * @param socket the socket to use.
 */
var _doRequest = function(client, socket) ***REMOVED***
  if(socket.isConnected()) ***REMOVED***
    // already connected
    socket.options.request.connectTime = +new Date();
    socket.connected(***REMOVED***
      type: 'connect',
      id: socket.id
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    // connect
    socket.options.request.connectTime = +new Date();
    socket.connect(***REMOVED***
      host: client.url.host,
      port: client.url.port,
      policyPort: client.policyPort,
      policyUrl: client.policyUrl
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Handles the next request or marks a socket as idle.
 *
 * @param client the http client.
 * @param socket the socket.
 */
var _handleNextRequest = function(client, socket) ***REMOVED***
  // clear buffer
  socket.buffer.clear();

  // get pending request
  var pending = null;
  while(pending === null && client.requests.length > 0) ***REMOVED***
    pending = client.requests.shift();
    if(pending.request.aborted) ***REMOVED***
      pending = null;
    ***REMOVED***
  ***REMOVED***

  // mark socket idle if no pending requests
  if(pending === null) ***REMOVED***
    if(socket.options !== null) ***REMOVED***
      socket.options = null;
    ***REMOVED***
    client.idle.push(socket);
  ***REMOVED*** else ***REMOVED***
    // handle pending request, allow 1 retry
    socket.retries = 1;
    socket.options = pending;
    _doRequest(client, socket);
  ***REMOVED***
***REMOVED***;

/**
 * Sets up a socket for use with an http client.
 *
 * @param client the parent http client.
 * @param socket the socket to set up.
 * @param tlsOptions if the socket must use TLS, the TLS options.
 */
var _initSocket = function(client, socket, tlsOptions) ***REMOVED***
  // no socket options yet
  socket.options = null;

  // set up handlers
  socket.connected = function(e) ***REMOVED***
    // socket primed by caching TLS session, handle next request
    if(socket.options === null) ***REMOVED***
      _handleNextRequest(client, socket);
    ***REMOVED*** else ***REMOVED***
      // socket in use
      var request = socket.options.request;
      request.connectTime = +new Date() - request.connectTime;
      e.socket = socket;
      socket.options.connected(e);
      if(request.aborted) ***REMOVED***
        socket.close();
      ***REMOVED*** else ***REMOVED***
        var out = request.toString();
        if(request.body) ***REMOVED***
          out += request.body;
        ***REMOVED***
        request.time = +new Date();
        socket.send(out);
        request.time = +new Date() - request.time;
        socket.options.response.time = +new Date();
        socket.sending = true;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
  socket.closed = function(e) ***REMOVED***
    if(socket.sending) ***REMOVED***
      socket.sending = false;
      if(socket.retries > 0) ***REMOVED***
        --socket.retries;
        _doRequest(client, socket);
      ***REMOVED*** else ***REMOVED***
        // error, closed during send
        socket.error(***REMOVED***
          id: socket.id,
          type: 'ioError',
          message: 'Connection closed during send. Broken pipe.',
          bytesAvailable: 0
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      // handle unspecified content-length transfer
      var response = socket.options.response;
      if(response.readBodyUntilClose) ***REMOVED***
        response.time = +new Date() - response.time;
        response.bodyReceived = true;
        socket.options.bodyReady(***REMOVED***
          request: socket.options.request,
          response: response,
          socket: socket
        ***REMOVED***);
      ***REMOVED***
      socket.options.closed(e);
      _handleNextRequest(client, socket);
    ***REMOVED***
  ***REMOVED***;
  socket.data = function(e) ***REMOVED***
    socket.sending = false;
    var request = socket.options.request;
    if(request.aborted) ***REMOVED***
      socket.close();
    ***REMOVED*** else ***REMOVED***
      // receive all bytes available
      var response = socket.options.response;
      var bytes = socket.receive(e.bytesAvailable);
      if(bytes !== null) ***REMOVED***
        // receive header and then body
        socket.buffer.putBytes(bytes);
        if(!response.headerReceived) ***REMOVED***
          response.readHeader(socket.buffer);
          if(response.headerReceived) ***REMOVED***
            socket.options.headerReady(***REMOVED***
              request: socket.options.request,
              response: response,
              socket: socket
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
        if(response.headerReceived && !response.bodyReceived) ***REMOVED***
          response.readBody(socket.buffer);
        ***REMOVED***
        if(response.bodyReceived) ***REMOVED***
          socket.options.bodyReady(***REMOVED***
            request: socket.options.request,
            response: response,
            socket: socket
          ***REMOVED***);
          // close connection if requested or by default on http/1.0
          var value = response.getField('Connection') || '';
          if(value.indexOf('close') != -1 ||
            (response.version === 'HTTP/1.0' &&
            response.getField('Keep-Alive') === null)) ***REMOVED***
            socket.close();
          ***REMOVED*** else ***REMOVED***
            _handleNextRequest(client, socket);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
  socket.error = function(e) ***REMOVED***
    // do error callback, include request
    socket.options.error(***REMOVED***
      type: e.type,
      message: e.message,
      request: socket.options.request,
      response: socket.options.response,
      socket: socket
    ***REMOVED***);
    socket.close();
  ***REMOVED***;

  // wrap socket for TLS
  if(tlsOptions) ***REMOVED***
    socket = forge.tls.wrapSocket(***REMOVED***
      sessionId: null,
      sessionCache: ***REMOVED******REMOVED***,
      caStore: tlsOptions.caStore,
      cipherSuites: tlsOptions.cipherSuites,
      socket: socket,
      virtualHost: tlsOptions.virtualHost,
      verify: tlsOptions.verify,
      getCertificate: tlsOptions.getCertificate,
      getPrivateKey: tlsOptions.getPrivateKey,
      getSignature: tlsOptions.getSignature,
      deflate: tlsOptions.deflate || null,
      inflate: tlsOptions.inflate || null
    ***REMOVED***);

    socket.options = null;
    socket.buffer = forge.util.createBuffer();
    client.sockets.push(socket);
    if(tlsOptions.prime) ***REMOVED***
      // prime socket by connecting and caching TLS session, will do
      // next request from there
      socket.connect(***REMOVED***
        host: client.url.host,
        port: client.url.port,
        policyPort: client.policyPort,
        policyUrl: client.policyUrl
      ***REMOVED***);
    ***REMOVED*** else ***REMOVED***
      // do not prime socket, just add as idle
      client.idle.push(socket);
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    // no need to prime non-TLS sockets
    socket.buffer = forge.util.createBuffer();
    client.sockets.push(socket);
    client.idle.push(socket);
  ***REMOVED***
***REMOVED***;

/**
 * Checks to see if the given cookie has expired. If the cookie's max-age
 * plus its created time is less than the time now, it has expired, unless
 * its max-age is set to -1 which indicates it will never expire.
 *
 * @param cookie the cookie to check.
 *
 * @return true if it has expired, false if not.
 */
var _hasCookieExpired = function(cookie) ***REMOVED***
  var rval = false;

  if(cookie.maxAge !== -1) ***REMOVED***
    var now = _getUtcTime(new Date());
    var expires = cookie.created + cookie.maxAge;
    if(expires <= now) ***REMOVED***
      rval = true;
    ***REMOVED***
  ***REMOVED***

  return rval;
***REMOVED***;

/**
 * Adds cookies in the given client to the given request.
 *
 * @param client the client.
 * @param request the request.
 */
var _writeCookies = function(client, request) ***REMOVED***
  var expired = [];
  var url = client.url;
  var cookies = client.cookies;
  for(var name in cookies) ***REMOVED***
    // get cookie paths
    var paths = cookies[name];
    for(var p in paths) ***REMOVED***
      var cookie = paths[p];
      if(_hasCookieExpired(cookie)) ***REMOVED***
        // store for clean up
        expired.push(cookie);
      ***REMOVED*** else if(request.path.indexOf(cookie.path) === 0) ***REMOVED***
        // path or path's ancestor must match cookie.path
        request.addCookie(cookie);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // clean up expired cookies
  for(var i = 0; i < expired.length; ++i) ***REMOVED***
    var cookie = expired[i];
    client.removeCookie(cookie.name, cookie.path);
  ***REMOVED***
***REMOVED***;

/**
 * Gets cookies from the given response and adds the to the given client.
 *
 * @param client the client.
 * @param response the response.
 */
var _readCookies = function(client, response) ***REMOVED***
  var cookies = response.getCookies();
  for(var i = 0; i < cookies.length; ++i) ***REMOVED***
    try ***REMOVED***
      client.setCookie(cookies[i]);
    ***REMOVED*** catch(ex) ***REMOVED***
      // ignore failure to add other-domain, etc. cookies
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Creates an http client that uses forge.net sockets as a backend and
 * forge.tls for security.
 *
 * @param options:
 *   url: the url to connect to (scheme://host:port).
 *     socketPool: the flash socket pool to use.
 *   policyPort: the flash policy port to use (if other than the
 *     socket pool default), use 0 for flash default.
 *   policyUrl: the flash policy file URL to use (if provided will
 *     be used instead of a policy port).
 *   connections: number of connections to use to handle requests.
 *   caCerts: an array of certificates to trust for TLS, certs may
 *     be PEM-formatted or cert objects produced via forge.pki.
 *   cipherSuites: an optional array of cipher suites to use,
 *     see forge.tls.CipherSuites.
 *   virtualHost: the virtual server name to use in a TLS SNI
 *     extension, if not provided the url host will be used.
 *   verify: a custom TLS certificate verify callback to use.
 *   getCertificate: an optional callback used to get a client-side
 *     certificate (see forge.tls for details).
 *   getPrivateKey: an optional callback used to get a client-side
 *     private key (see forge.tls for details).
 *   getSignature: an optional callback used to get a client-side
 *     signature (see forge.tls for details).
 *   persistCookies: true to use persistent cookies via flash local
 *     storage, false to only keep cookies in javascript.
 *   primeTlsSockets: true to immediately connect TLS sockets on
 *     their creation so that they will cache TLS sessions for reuse.
 *
 * @return the client.
 */
http.createClient = function(options) ***REMOVED***
  // create CA store to share with all TLS connections
  var caStore = null;
  if(options.caCerts) ***REMOVED***
    caStore = forge.pki.createCaStore(options.caCerts);
  ***REMOVED***

  // get scheme, host, and port from url
  options.url = (options.url ||
    window.location.protocol + '//' + window.location.host);
  var url = http.parseUrl(options.url);
  if(!url) ***REMOVED***
    var error = new Error('Invalid url.');
    error.details = ***REMOVED***url: options.url***REMOVED***;
    throw error;
  ***REMOVED***

  // default to 1 connection
  options.connections = options.connections || 1;

  // create client
  var sp = options.socketPool;
  var client = ***REMOVED***
    // url
    url: url,
    // socket pool
    socketPool: sp,
    // the policy port to use
    policyPort: options.policyPort,
    // policy url to use
    policyUrl: options.policyUrl,
    // queue of requests to service
    requests: [],
    // all sockets
    sockets: [],
    // idle sockets
    idle: [],
    // whether or not the connections are secure
    secure: (url.scheme === 'https'),
    // cookie jar (key'd off of name and then path, there is only 1 domain
    // and one setting for secure per client so name+path is unique)
    cookies: ***REMOVED******REMOVED***,
    // default to flash storage of cookies
    persistCookies: (typeof(options.persistCookies) === 'undefined') ?
      true : options.persistCookies
  ***REMOVED***;

  // add client to debug storage
  if(forge.debug) ***REMOVED***
    forge.debug.get('forge.http', 'clients').push(client);
  ***REMOVED***

  // load cookies from disk
  _loadCookies(client);

  /**
   * A default certificate verify function that checks a certificate common
   * name against the client's URL host.
   *
   * @param c the TLS connection.
   * @param verified true if cert is verified, otherwise alert number.
   * @param depth the chain depth.
   * @param certs the cert chain.
   *
   * @return true if verified and the common name matches the host, error
   *         otherwise.
   */
  var _defaultCertificateVerify = function(c, verified, depth, certs) ***REMOVED***
    if(depth === 0 && verified === true) ***REMOVED***
      // compare common name to url host
      var cn = certs[depth].subject.getField('CN');
      if(cn === null || client.url.host !== cn.value) ***REMOVED***
        verified = ***REMOVED***
          message: 'Certificate common name does not match url host.'
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***
    return verified;
  ***REMOVED***;

  // determine if TLS is used
  var tlsOptions = null;
  if(client.secure) ***REMOVED***
    tlsOptions = ***REMOVED***
      caStore: caStore,
      cipherSuites: options.cipherSuites || null,
      virtualHost: options.virtualHost || url.host,
      verify: options.verify || _defaultCertificateVerify,
      getCertificate: options.getCertificate || null,
      getPrivateKey: options.getPrivateKey || null,
      getSignature: options.getSignature || null,
      prime: options.primeTlsSockets || false
    ***REMOVED***;

    // if socket pool uses a flash api, then add deflate support to TLS
    if(sp.flashApi !== null) ***REMOVED***
      tlsOptions.deflate = function(bytes) ***REMOVED***
        // strip 2 byte zlib header and 4 byte trailer
        return forge.util.deflate(sp.flashApi, bytes, true);
      ***REMOVED***;
      tlsOptions.inflate = function(bytes) ***REMOVED***
        return forge.util.inflate(sp.flashApi, bytes, true);
      ***REMOVED***;
    ***REMOVED***
  ***REMOVED***

  // create and initialize sockets
  for(var i = 0; i < options.connections; ++i) ***REMOVED***
    _initSocket(client, sp.createSocket(), tlsOptions);
  ***REMOVED***

  /**
   * Sends a request. A method 'abort' will be set on the request that
   * can be called to attempt to abort the request.
   *
   * @param options:
   *          request: the request to send.
   *          connected: a callback for when the connection is open.
   *          closed: a callback for when the connection is closed.
   *          headerReady: a callback for when the response header arrives.
   *          bodyReady: a callback for when the response body arrives.
   *          error: a callback for if an error occurs.
   */
  client.send = function(options) ***REMOVED***
    // add host header if not set
    if(options.request.getField('Host') === null) ***REMOVED***
      options.request.setField('Host', client.url.fullHost);
    ***REMOVED***

    // set default dummy handlers
    var opts = ***REMOVED******REMOVED***;
    opts.request = options.request;
    opts.connected = options.connected || function() ***REMOVED******REMOVED***;
    opts.closed = options.close || function() ***REMOVED******REMOVED***;
    opts.headerReady = function(e) ***REMOVED***
      // read cookies
      _readCookies(client, e.response);
      if(options.headerReady) ***REMOVED***
        options.headerReady(e);
      ***REMOVED***
    ***REMOVED***;
    opts.bodyReady = options.bodyReady || function() ***REMOVED******REMOVED***;
    opts.error = options.error || function() ***REMOVED******REMOVED***;

    // create response
    opts.response = http.createResponse();
    opts.response.time = 0;
    opts.response.flashApi = client.socketPool.flashApi;
    opts.request.flashApi = client.socketPool.flashApi;

    // create abort function
    opts.request.abort = function() ***REMOVED***
      // set aborted, clear handlers
      opts.request.aborted = true;
      opts.connected = function() ***REMOVED******REMOVED***;
      opts.closed = function() ***REMOVED******REMOVED***;
      opts.headerReady = function() ***REMOVED******REMOVED***;
      opts.bodyReady = function() ***REMOVED******REMOVED***;
      opts.error = function() ***REMOVED******REMOVED***;
    ***REMOVED***;

    // add cookies to request
    _writeCookies(client, opts.request);

    // queue request options if there are no idle sockets
    if(client.idle.length === 0) ***REMOVED***
      client.requests.push(opts);
    ***REMOVED*** else ***REMOVED***
      // use an idle socket, prefer an idle *connected* socket first
      var socket = null;
      var len = client.idle.length;
      for(var i = 0; socket === null && i < len; ++i) ***REMOVED***
        socket = client.idle[i];
        if(socket.isConnected()) ***REMOVED***
          client.idle.splice(i, 1);
        ***REMOVED*** else ***REMOVED***
          socket = null;
        ***REMOVED***
      ***REMOVED***
      // no connected socket available, get unconnected socket
      if(socket === null) ***REMOVED***
        socket = client.idle.pop();
      ***REMOVED***
      socket.options = opts;
      _doRequest(client, socket);
    ***REMOVED***
  ***REMOVED***;

  /**
   * Destroys this client.
   */
  client.destroy = function() ***REMOVED***
    // clear pending requests, close and destroy sockets
    client.requests = [];
    for(var i = 0; i < client.sockets.length; ++i) ***REMOVED***
      client.sockets[i].close();
      client.sockets[i].destroy();
    ***REMOVED***
    client.socketPool = null;
    client.sockets = [];
    client.idle = [];
  ***REMOVED***;

  /**
   * Sets a cookie for use with all connections made by this client. Any
   * cookie with the same name will be replaced. If the cookie's value
   * is undefined, null, or the blank string, the cookie will be removed.
   *
   * If the cookie's domain doesn't match this client's url host or the
   * cookie's secure flag doesn't match this client's url scheme, then
   * setting the cookie will fail with an exception.
   *
   * @param cookie the cookie with parameters:
   *   name: the name of the cookie.
   *   value: the value of the cookie.
   *   comment: an optional comment string.
   *   maxAge: the age of the cookie in seconds relative to created time.
   *   secure: true if the cookie must be sent over a secure protocol.
   *   httpOnly: true to restrict access to the cookie from javascript
   *     (inaffective since the cookies are stored in javascript).
   *   path: the path for the cookie.
   *   domain: optional domain the cookie belongs to (must start with dot).
   *   version: optional version of the cookie.
   *   created: creation time, in UTC seconds, of the cookie.
   */
  client.setCookie = function(cookie) ***REMOVED***
    var rval;
    if(typeof(cookie.name) !== 'undefined') ***REMOVED***
      if(cookie.value === null || typeof(cookie.value) === 'undefined' ||
        cookie.value === '') ***REMOVED***
        // remove cookie
        rval = client.removeCookie(cookie.name, cookie.path);
      ***REMOVED*** else ***REMOVED***
        // set cookie defaults
        cookie.comment = cookie.comment || '';
        cookie.maxAge = cookie.maxAge || 0;
        cookie.secure = (typeof(cookie.secure) === 'undefined') ?
          true : cookie.secure;
        cookie.httpOnly = cookie.httpOnly || true;
        cookie.path = cookie.path || '/';
        cookie.domain = cookie.domain || null;
        cookie.version = cookie.version || null;
        cookie.created = _getUtcTime(new Date());

        // do secure check
        if(cookie.secure !== client.secure) ***REMOVED***
          var error = new Error('Http client url scheme is incompatible ' +
            'with cookie secure flag.');
          error.url = client.url;
          error.cookie = cookie;
          throw error;
        ***REMOVED***
        // make sure url host is within cookie.domain
        if(!http.withinCookieDomain(client.url, cookie)) ***REMOVED***
          var error = new Error('Http client url scheme is incompatible ' +
            'with cookie secure flag.');
          error.url = client.url;
          error.cookie = cookie;
          throw error;
        ***REMOVED***

        // add new cookie
        if(!(cookie.name in client.cookies)) ***REMOVED***
          client.cookies[cookie.name] = ***REMOVED******REMOVED***;
        ***REMOVED***
        client.cookies[cookie.name][cookie.path] = cookie;
        rval = true;

        // save cookies
        _saveCookies(client);
      ***REMOVED***
    ***REMOVED***

    return rval;
  ***REMOVED***;

  /**
   * Gets a cookie by its name.
   *
   * @param name the name of the cookie to retrieve.
   * @param path an optional path for the cookie (if there are multiple
   *          cookies with the same name but different paths).
   *
   * @return the cookie or null if not found.
   */
  client.getCookie = function(name, path) ***REMOVED***
    var rval = null;
    if(name in client.cookies) ***REMOVED***
      var paths = client.cookies[name];

      // get path-specific cookie
      if(path) ***REMOVED***
        if(path in paths) ***REMOVED***
          rval = paths[path];
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        // get first cookie
        for(var p in paths) ***REMOVED***
          rval = paths[p];
          break;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return rval;
  ***REMOVED***;

  /**
   * Removes a cookie.
   *
   * @param name the name of the cookie to remove.
   * @param path an optional path for the cookie (if there are multiple
   *          cookies with the same name but different paths).
   *
   * @return true if a cookie was removed, false if not.
   */
  client.removeCookie = function(name, path) ***REMOVED***
    var rval = false;
    if(name in client.cookies) ***REMOVED***
      // delete the specific path
      if(path) ***REMOVED***
        var paths = client.cookies[name];
        if(path in paths) ***REMOVED***
          rval = true;
          delete client.cookies[name][path];
          // clean up entry if empty
          var empty = true;
          for(var i in client.cookies[name]) ***REMOVED***
            empty = false;
            break;
          ***REMOVED***
          if(empty) ***REMOVED***
            delete client.cookies[name];
          ***REMOVED***
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        // delete all cookies with the given name
        rval = true;
        delete client.cookies[name];
      ***REMOVED***
    ***REMOVED***
    if(rval) ***REMOVED***
      // save cookies
      _saveCookies(client);
    ***REMOVED***
    return rval;
  ***REMOVED***;

  /**
   * Clears all cookies stored in this client.
   */
  client.clearCookies = function() ***REMOVED***
    client.cookies = ***REMOVED******REMOVED***;
    _clearCookies(client);
  ***REMOVED***;

  if(forge.log) ***REMOVED***
    forge.log.debug('forge.http', 'created client', options);
  ***REMOVED***

  return client;
***REMOVED***;

/**
 * Trims the whitespace off of the beginning and end of a string.
 *
 * @param str the string to trim.
 *
 * @return the trimmed string.
 */
var _trimString = function(str) ***REMOVED***
  return str.replace(/^\s*/, '').replace(/\s*$/, '');
***REMOVED***;

/**
 * Creates an http header object.
 *
 * @return the http header object.
 */
var _createHeader = function() ***REMOVED***
  var header = ***REMOVED***
    fields: ***REMOVED******REMOVED***,
    setField: function(name, value) ***REMOVED***
      // normalize field name, trim value
      header.fields[_normalize(name)] = [_trimString('' + value)];
    ***REMOVED***,
    appendField: function(name, value) ***REMOVED***
      name = _normalize(name);
      if(!(name in header.fields)) ***REMOVED***
        header.fields[name] = [];
      ***REMOVED***
      header.fields[name].push(_trimString('' + value));
    ***REMOVED***,
    getField: function(name, index) ***REMOVED***
      var rval = null;
      name = _normalize(name);
      if(name in header.fields) ***REMOVED***
        index = index || 0;
        rval = header.fields[name][index];
      ***REMOVED***
      return rval;
    ***REMOVED***
  ***REMOVED***;
  return header;
***REMOVED***;

/**
 * Gets the time in utc seconds given a date.
 *
 * @param d the date to use.
 *
 * @return the time in utc seconds.
 */
var _getUtcTime = function(d) ***REMOVED***
  var utc = +d + d.getTimezoneOffset() * 60000;
  return Math.floor(+new Date() / 1000);
***REMOVED***;

/**
 * Creates an http request.
 *
 * @param options:
 *          version: the version.
 *          method: the method.
 *          path: the path.
 *          body: the body.
 *          headers: custom header fields to add,
 *            eg: [***REMOVED***'Content-Length': 0***REMOVED***].
 *
 * @return the http request.
 */
http.createRequest = function(options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;
  var request = _createHeader();
  request.version = options.version || 'HTTP/1.1';
  request.method = options.method || null;
  request.path = options.path || null;
  request.body = options.body || null;
  request.bodyDeflated = false;
  request.flashApi = null;

  // add custom headers
  var headers = options.headers || [];
  if(!forge.util.isArray(headers)) ***REMOVED***
    headers = [headers];
  ***REMOVED***
  for(var i = 0; i < headers.length; ++i) ***REMOVED***
    for(var name in headers[i]) ***REMOVED***
      request.appendField(name, headers[i][name]);
    ***REMOVED***
  ***REMOVED***

  /**
   * Adds a cookie to the request 'Cookie' header.
   *
   * @param cookie a cookie to add.
   */
  request.addCookie = function(cookie) ***REMOVED***
    var value = '';
    var field = request.getField('Cookie');
    if(field !== null) ***REMOVED***
      // separate cookies by semi-colons
      value = field + '; ';
    ***REMOVED***

    // get current time in utc seconds
    var now = _getUtcTime(new Date());

    // output cookie name and value
    value += cookie.name + '=' + cookie.value;
    request.setField('Cookie', value);
  ***REMOVED***;

  /**
   * Converts an http request into a string that can be sent as an
   * HTTP request. Does not include any data.
   *
   * @return the string representation of the request.
   */
  request.toString = function() ***REMOVED***
    /* Sample request header:
      GET /some/path/?query HTTP/1.1
      Host: www.someurl.com
      Connection: close
      Accept-Encoding: deflate
      Accept: image/gif, text/html
      User-Agent: Mozilla 4.0
     */

    // set default headers
    if(request.getField('User-Agent') === null) ***REMOVED***
      request.setField('User-Agent', 'forge.http 1.0');
    ***REMOVED***
    if(request.getField('Accept') === null) ***REMOVED***
      request.setField('Accept', '*/*');
    ***REMOVED***
    if(request.getField('Connection') === null) ***REMOVED***
      request.setField('Connection', 'keep-alive');
      request.setField('Keep-Alive', '115');
    ***REMOVED***

    // add Accept-Encoding if not specified
    if(request.flashApi !== null &&
      request.getField('Accept-Encoding') === null) ***REMOVED***
      request.setField('Accept-Encoding', 'deflate');
    ***REMOVED***

    // if the body isn't null, deflate it if its larger than 100 bytes
    if(request.flashApi !== null && request.body !== null &&
      request.getField('Content-Encoding') === null &&
      !request.bodyDeflated && request.body.length > 100) ***REMOVED***
      // use flash to compress data
      request.body = forge.util.deflate(request.flashApi, request.body);
      request.bodyDeflated = true;
      request.setField('Content-Encoding', 'deflate');
      request.setField('Content-Length', request.body.length);
    ***REMOVED*** else if(request.body !== null) ***REMOVED***
      // set content length for body
      request.setField('Content-Length', request.body.length);
    ***REMOVED***

    // build start line
    var rval =
      request.method.toUpperCase() + ' ' + request.path + ' ' +
      request.version + '\r\n';

    // add each header
    for(var name in request.fields) ***REMOVED***
      var fields = request.fields[name];
      for(var i = 0; i < fields.length; ++i) ***REMOVED***
        rval += name + ': ' + fields[i] + '\r\n';
      ***REMOVED***
    ***REMOVED***
    // final terminating CRLF
    rval += '\r\n';

    return rval;
  ***REMOVED***;

  return request;
***REMOVED***;

/**
 * Creates an empty http response header.
 *
 * @return the empty http response header.
 */
http.createResponse = function() ***REMOVED***
  // private vars
  var _first = true;
  var _chunkSize = 0;
  var _chunksFinished = false;

  // create response
  var response = _createHeader();
  response.version = null;
  response.code = 0;
  response.message = null;
  response.body = null;
  response.headerReceived = false;
  response.bodyReceived = false;
  response.flashApi = null;

  /**
   * Reads a line that ends in CRLF from a byte buffer.
   *
   * @param b the byte buffer.
   *
   * @return the line or null if none was found.
   */
  var _readCrlf = function(b) ***REMOVED***
    var line = null;
    var i = b.data.indexOf('\r\n', b.read);
    if(i != -1) ***REMOVED***
      // read line, skip CRLF
      line = b.getBytes(i - b.read);
      b.getBytes(2);
    ***REMOVED***
    return line;
  ***REMOVED***;

  /**
   * Parses a header field and appends it to the response.
   *
   * @param line the header field line.
   */
  var _parseHeader = function(line) ***REMOVED***
    var tmp = line.indexOf(':');
    var name = line.substring(0, tmp++);
    response.appendField(
      name, (tmp < line.length) ? line.substring(tmp) : '');
  ***REMOVED***;

  /**
   * Reads an http response header from a buffer of bytes.
   *
   * @param b the byte buffer to parse the header from.
   *
   * @return true if the whole header was read, false if not.
   */
  response.readHeader = function(b) ***REMOVED***
    // read header lines (each ends in CRLF)
    var line = '';
    while(!response.headerReceived && line !== null) ***REMOVED***
      line = _readCrlf(b);
      if(line !== null) ***REMOVED***
        // parse first line
        if(_first) ***REMOVED***
          _first = false;
          var tmp = line.split(' ');
          if(tmp.length >= 3) ***REMOVED***
            response.version = tmp[0];
            response.code = parseInt(tmp[1], 10);
            response.message = tmp.slice(2).join(' ');
          ***REMOVED*** else ***REMOVED***
            // invalid header
            var error = new Error('Invalid http response header.');
            error.details = ***REMOVED***'line': line***REMOVED***;
            throw error;
          ***REMOVED***
        ***REMOVED*** else if(line.length === 0) ***REMOVED***
          // handle final line, end of header
          response.headerReceived = true;
        ***REMOVED*** else ***REMOVED***
          _parseHeader(line);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return response.headerReceived;
  ***REMOVED***;

  /**
   * Reads some chunked http response entity-body from the given buffer of
   * bytes.
   *
   * @param b the byte buffer to read from.
   *
   * @return true if the whole body was read, false if not.
   */
  var _readChunkedBody = function(b) ***REMOVED***
    /* Chunked transfer-encoding sends data in a series of chunks,
      followed by a set of 0-N http trailers.
      The format is as follows:

      chunk-size (in hex) CRLF
      chunk data (with "chunk-size" many bytes) CRLF
      ... (N many chunks)
      chunk-size (of 0 indicating the last chunk) CRLF
      N many http trailers followed by CRLF
      blank line + CRLF (terminates the trailers)

      If there are no http trailers, then after the chunk-size of 0,
      there is still a single CRLF (indicating the blank line + CRLF
      that terminates the trailers). In other words, you always terminate
      the trailers with blank line + CRLF, regardless of 0-N trailers. */

      /* From RFC-2616, section 3.6.1, here is the pseudo-code for
      implementing chunked transfer-encoding:

      length := 0
      read chunk-size, chunk-extension (if any) and CRLF
      while (chunk-size > 0) ***REMOVED***
        read chunk-data and CRLF
        append chunk-data to entity-body
        length := length + chunk-size
        read chunk-size and CRLF
      ***REMOVED***
      read entity-header
      while (entity-header not empty) ***REMOVED***
        append entity-header to existing header fields
        read entity-header
      ***REMOVED***
      Content-Length := length
      Remove "chunked" from Transfer-Encoding
    */

    var line = '';
    while(line !== null && b.length() > 0) ***REMOVED***
      // if in the process of reading a chunk
      if(_chunkSize > 0) ***REMOVED***
        // if there are not enough bytes to read chunk and its
        // trailing CRLF,  we must wait for more data to be received
        if(_chunkSize + 2 > b.length()) ***REMOVED***
          break;
        ***REMOVED***

        // read chunk data, skip CRLF
        response.body += b.getBytes(_chunkSize);
        b.getBytes(2);
        _chunkSize = 0;
      ***REMOVED*** else if(!_chunksFinished) ***REMOVED***
        // more chunks, read next chunk-size line
        line = _readCrlf(b);
        if(line !== null) ***REMOVED***
          // parse chunk-size (ignore any chunk extension)
          _chunkSize = parseInt(line.split(';', 1)[0], 16);
          _chunksFinished = (_chunkSize === 0);
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        // chunks finished, read next trailer
        line = _readCrlf(b);
        while(line !== null) ***REMOVED***
          if(line.length > 0) ***REMOVED***
            // parse trailer
            _parseHeader(line);
            // read next trailer
            line = _readCrlf(b);
          ***REMOVED*** else ***REMOVED***
            // body received
            response.bodyReceived = true;
            line = null;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    return response.bodyReceived;
  ***REMOVED***;

  /**
   * Reads an http response body from a buffer of bytes.
   *
   * @param b the byte buffer to read from.
   *
   * @return true if the whole body was read, false if not.
   */
  response.readBody = function(b) ***REMOVED***
    var contentLength = response.getField('Content-Length');
    var transferEncoding = response.getField('Transfer-Encoding');
    if(contentLength !== null) ***REMOVED***
      contentLength = parseInt(contentLength);
    ***REMOVED***

    // read specified length
    if(contentLength !== null && contentLength >= 0) ***REMOVED***
      response.body = response.body || '';
      response.body += b.getBytes(contentLength);
      response.bodyReceived = (response.body.length === contentLength);
    ***REMOVED*** else if(transferEncoding !== null) ***REMOVED***
      // read chunked encoding
      if(transferEncoding.indexOf('chunked') != -1) ***REMOVED***
        response.body = response.body || '';
        _readChunkedBody(b);
      ***REMOVED*** else ***REMOVED***
        var error = new Error('Unknown Transfer-Encoding.');
        error.details = ***REMOVED***'transferEncoding': transferEncoding***REMOVED***;
        throw error;
      ***REMOVED***
    ***REMOVED*** else if((contentLength !== null && contentLength < 0) ||
      (contentLength === null &&
      response.getField('Content-Type') !== null)) ***REMOVED***
      // read all data in the buffer
      response.body = response.body || '';
      response.body += b.getBytes();
      response.readBodyUntilClose = true;
    ***REMOVED*** else ***REMOVED***
      // no body
      response.body = null;
      response.bodyReceived = true;
    ***REMOVED***

    if(response.bodyReceived) ***REMOVED***
      response.time = +new Date() - response.time;
    ***REMOVED***

    if(response.flashApi !== null &&
      response.bodyReceived && response.body !== null &&
      response.getField('Content-Encoding') === 'deflate') ***REMOVED***
      // inflate using flash api
      response.body = forge.util.inflate(
        response.flashApi, response.body);
    ***REMOVED***

    return response.bodyReceived;
  ***REMOVED***;

   /**
    * Parses an array of cookies from the 'Set-Cookie' field, if present.
    *
    * @return the array of cookies.
    */
   response.getCookies = function() ***REMOVED***
     var rval = [];

     // get Set-Cookie field
     if('Set-Cookie' in response.fields) ***REMOVED***
       var field = response.fields['Set-Cookie'];

       // get current local time in seconds
       var now = +new Date() / 1000;

       // regex for parsing 'name1=value1; name2=value2; name3'
       var regex = /\s*([^=]*)=?([^;]*)(;|$)/g;

       // examples:
       // Set-Cookie: cookie1_name=cookie1_value; max-age=0; path=/
       // Set-Cookie: c2=v2; expires=Thu, 21-Aug-2008 23:47:25 GMT; path=/
       for(var i = 0; i < field.length; ++i) ***REMOVED***
         var fv = field[i];
         var m;
         regex.lastIndex = 0;
         var first = true;
         var cookie = ***REMOVED******REMOVED***;
         do ***REMOVED***
           m = regex.exec(fv);
           if(m !== null) ***REMOVED***
             var name = _trimString(m[1]);
             var value = _trimString(m[2]);

             // cookie_name=value
             if(first) ***REMOVED***
               cookie.name = name;
               cookie.value = value;
               first = false;
             ***REMOVED*** else ***REMOVED***
               // property_name=value
               name = name.toLowerCase();
               switch(name) ***REMOVED***
               case 'expires':
                 // replace hyphens w/spaces so date will parse
                 value = value.replace(/-/g, ' ');
                 var secs = Date.parse(value) / 1000;
                 cookie.maxAge = Math.max(0, secs - now);
                 break;
               case 'max-age':
                 cookie.maxAge = parseInt(value, 10);
                 break;
               case 'secure':
                 cookie.secure = true;
                 break;
               case 'httponly':
                 cookie.httpOnly = true;
                 break;
               default:
                 if(name !== '') ***REMOVED***
                   cookie[name] = value;
                 ***REMOVED***
               ***REMOVED***
             ***REMOVED***
           ***REMOVED***
         ***REMOVED*** while(m !== null && m[0] !== '');
         rval.push(cookie);
       ***REMOVED***
     ***REMOVED***

     return rval;
  ***REMOVED***;

  /**
   * Converts an http response into a string that can be sent as an
   * HTTP response. Does not include any data.
   *
   * @return the string representation of the response.
   */
  response.toString = function() ***REMOVED***
    /* Sample response header:
      HTTP/1.0 200 OK
      Host: www.someurl.com
      Connection: close
     */

    // build start line
    var rval =
      response.version + ' ' + response.code + ' ' + response.message + '\r\n';

    // add each header
    for(var name in response.fields) ***REMOVED***
      var fields = response.fields[name];
      for(var i = 0; i < fields.length; ++i) ***REMOVED***
        rval += name + ': ' + fields[i] + '\r\n';
      ***REMOVED***
    ***REMOVED***
    // final terminating CRLF
    rval += '\r\n';

    return rval;
  ***REMOVED***;

  return response;
***REMOVED***;

/**
 * Parses the scheme, host, and port from an http(s) url.
 *
 * @param str the url string.
 *
 * @return the parsed url object or null if the url is invalid.
 */
http.parseUrl = forge.util.parseUrl;

/**
 * Returns true if the given url is within the given cookie's domain.
 *
 * @param url the url to check.
 * @param cookie the cookie or cookie domain to check.
 */
http.withinCookieDomain = function(url, cookie) ***REMOVED***
  var rval = false;

  // cookie may be null, a cookie object, or a domain string
  var domain = (cookie === null || typeof cookie === 'string') ?
    cookie : cookie.domain;

  // any domain will do
  if(domain === null) ***REMOVED***
    rval = true;
  ***REMOVED*** else if(domain.charAt(0) === '.') ***REMOVED***
    // ensure domain starts with a '.'
    // parse URL as necessary
    if(typeof url === 'string') ***REMOVED***
      url = http.parseUrl(url);
    ***REMOVED***

    // add '.' to front of URL host to match against domain
    var host = '.' + url.host;

    // if the host ends with domain then it falls within it
    var idx = host.lastIndexOf(domain);
    if(idx !== -1 && (idx + domain.length === host.length)) ***REMOVED***
      rval = true;
    ***REMOVED***
  ***REMOVED***

  return rval;
***REMOVED***;
