// node-forge/lib/log.js
/**
 * Cross-browser support for logging in a web application.
 *
 * @author David I. Lehn <dlehn@digitalbazaar.com>
 *
 * Copyright (c) 2008-2013 Digital Bazaar, Inc.
 */
var forge = require('./forge');
require('./util');

/* LOG API */
module.exports = forge.log = forge.log || ***REMOVED******REMOVED***;

/**
 * Application logging system.
 *
 * Each logger level available as it's own function of the form:
 *   forge.log.level(category, args...)
 * The category is an arbitrary string, and the args are the same as
 * Firebug's console.log API. By default the call will be output as:
 *   'LEVEL [category] <args[0]>, args[1], ...'
 * This enables proper % formatting via the first argument.
 * Each category is enabled by default but can be enabled or disabled with
 * the setCategoryEnabled() function.
 */
// list of known levels
forge.log.levels = [
  'none', 'error', 'warning', 'info', 'debug', 'verbose', 'max'];
// info on the levels indexed by name:
//   index: level index
//   name: uppercased display name
var sLevelInfo = ***REMOVED******REMOVED***;
// list of loggers
var sLoggers = [];
/**
 * Standard console logger. If no console support is enabled this will
 * remain null. Check before using.
 */
var sConsoleLogger = null;

// logger flags
/**
 * Lock the level at the current value. Used in cases where user config may
 * set the level such that only critical messages are seen but more verbose
 * messages are needed for debugging or other purposes.
 */
forge.log.LEVEL_LOCKED = (1 << 1);
/**
 * Always call log function. By default, the logging system will check the
 * message level against logger.level before calling the log function. This
 * flag allows the function to do its own check.
 */
forge.log.NO_LEVEL_CHECK = (1 << 2);
/**
 * Perform message interpolation with the passed arguments. "%" style
 * fields in log messages will be replaced by arguments as needed. Some
 * loggers, such as Firebug, may do this automatically. The original log
 * message will be available as 'message' and the interpolated version will
 * be available as 'fullMessage'.
 */
forge.log.INTERPOLATE = (1 << 3);

// setup each log level
for(var i = 0; i < forge.log.levels.length; ++i) ***REMOVED***
  var level = forge.log.levels[i];
  sLevelInfo[level] = ***REMOVED***
    index: i,
    name: level.toUpperCase()
  ***REMOVED***;
***REMOVED***

/**
 * Message logger. Will dispatch a message to registered loggers as needed.
 *
 * @param message message object
 */
forge.log.logMessage = function(message) ***REMOVED***
  var messageLevelIndex = sLevelInfo[message.level].index;
  for(var i = 0; i < sLoggers.length; ++i) ***REMOVED***
    var logger = sLoggers[i];
    if(logger.flags & forge.log.NO_LEVEL_CHECK) ***REMOVED***
      logger.f(message);
    ***REMOVED*** else ***REMOVED***
      // get logger level
      var loggerLevelIndex = sLevelInfo[logger.level].index;
      // check level
      if(messageLevelIndex <= loggerLevelIndex) ***REMOVED***
        // message critical enough, call logger
        logger.f(logger, message);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Sets the 'standard' key on a message object to:
 * "LEVEL [category] " + message
 *
 * @param message a message log object
 */
forge.log.prepareStandard = function(message) ***REMOVED***
  if(!('standard' in message)) ***REMOVED***
    message.standard =
      sLevelInfo[message.level].name +
      //' ' + +message.timestamp +
      ' [' + message.category + '] ' +
      message.message;
  ***REMOVED***
***REMOVED***;

/**
 * Sets the 'full' key on a message object to the original message
 * interpolated via % formatting with the message arguments.
 *
 * @param message a message log object.
 */
forge.log.prepareFull = function(message) ***REMOVED***
  if(!('full' in message)) ***REMOVED***
    // copy args and insert message at the front
    var args = [message.message];
    args = args.concat([] || message['arguments']);
    // format the message
    message.full = forge.util.format.apply(this, args);
  ***REMOVED***
***REMOVED***;

/**
 * Applies both preparseStandard() and prepareFull() to a message object and
 * store result in 'standardFull'.
 *
 * @param message a message log object.
 */
forge.log.prepareStandardFull = function(message) ***REMOVED***
  if(!('standardFull' in message)) ***REMOVED***
    // FIXME implement 'standardFull' logging
    forge.log.prepareStandard(message);
    message.standardFull = message.standard;
  ***REMOVED***
***REMOVED***;

// create log level functions
if(true) ***REMOVED***
  // levels for which we want functions
  var levels = ['error', 'warning', 'info', 'debug', 'verbose'];
  for(var i = 0; i < levels.length; ++i) ***REMOVED***
    // wrap in a function to ensure proper level var is passed
    (function(level) ***REMOVED***
      // create function for this level
      forge.log[level] = function(category, message/*, args...*/) ***REMOVED***
        // convert arguments to real array, remove category and message
        var args = Array.prototype.slice.call(arguments).slice(2);
        // create message object
        // Note: interpolation and standard formatting is done lazily
        var msg = ***REMOVED***
          timestamp: new Date(),
          level: level,
          category: category,
          message: message,
          'arguments': args
          /*standard*/
          /*full*/
          /*fullMessage*/
        ***REMOVED***;
        // process this message
        forge.log.logMessage(msg);
      ***REMOVED***;
    ***REMOVED***)(levels[i]);
  ***REMOVED***
***REMOVED***

/**
 * Creates a new logger with specified custom logging function.
 *
 * The logging function has a signature of:
 *   function(logger, message)
 * logger: current logger
 * message: object:
 *   level: level id
 *   category: category
 *   message: string message
 *   arguments: Array of extra arguments
 *   fullMessage: interpolated message and arguments if INTERPOLATE flag set
 *
 * @param logFunction a logging function which takes a log message object
 *          as a parameter.
 *
 * @return a logger object.
 */
forge.log.makeLogger = function(logFunction) ***REMOVED***
  var logger = ***REMOVED***
    flags: 0,
    f: logFunction
  ***REMOVED***;
  forge.log.setLevel(logger, 'none');
  return logger;
***REMOVED***;

/**
 * Sets the current log level on a logger.
 *
 * @param logger the target logger.
 * @param level the new maximum log level as a string.
 *
 * @return true if set, false if not.
 */
forge.log.setLevel = function(logger, level) ***REMOVED***
  var rval = false;
  if(logger && !(logger.flags & forge.log.LEVEL_LOCKED)) ***REMOVED***
    for(var i = 0; i < forge.log.levels.length; ++i) ***REMOVED***
      var aValidLevel = forge.log.levels[i];
      if(level == aValidLevel) ***REMOVED***
        // set level
        logger.level = level;
        rval = true;
        break;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  return rval;
***REMOVED***;

/**
 * Locks the log level at its current value.
 *
 * @param logger the target logger.
 * @param lock boolean lock value, default to true.
 */
forge.log.lock = function(logger, lock) ***REMOVED***
  if(typeof lock === 'undefined' || lock) ***REMOVED***
    logger.flags |= forge.log.LEVEL_LOCKED;
  ***REMOVED*** else ***REMOVED***
    logger.flags &= ~forge.log.LEVEL_LOCKED;
  ***REMOVED***
***REMOVED***;

/**
 * Adds a logger.
 *
 * @param logger the logger object.
 */
forge.log.addLogger = function(logger) ***REMOVED***
  sLoggers.push(logger);
***REMOVED***;

// setup the console logger if possible, else create fake console.log
if(typeof(console) !== 'undefined' && 'log' in console) ***REMOVED***
  var logger;
  if(console.error && console.warn && console.info && console.debug) ***REMOVED***
    // looks like Firebug-style logging is available
    // level handlers map
    var levelHandlers = ***REMOVED***
      error: console.error,
      warning: console.warn,
      info: console.info,
      debug: console.debug,
      verbose: console.debug
    ***REMOVED***;
    var f = function(logger, message) ***REMOVED***
      forge.log.prepareStandard(message);
      var handler = levelHandlers[message.level];
      // prepend standard message and concat args
      var args = [message.standard];
      args = args.concat(message['arguments'].slice());
      // apply to low-level console function
      handler.apply(console, args);
    ***REMOVED***;
    logger = forge.log.makeLogger(f);
  ***REMOVED*** else ***REMOVED***
    // only appear to have basic console.log
    var f = function(logger, message) ***REMOVED***
      forge.log.prepareStandardFull(message);
      console.log(message.standardFull);
    ***REMOVED***;
    logger = forge.log.makeLogger(f);
  ***REMOVED***
  forge.log.setLevel(logger, 'debug');
  forge.log.addLogger(logger);
  sConsoleLogger = logger;
***REMOVED*** else ***REMOVED***
  // define fake console.log to avoid potential script errors on
  // browsers that do not have console logging
  console = ***REMOVED***
    log: function() ***REMOVED******REMOVED***
  ***REMOVED***;
***REMOVED***

/*
 * Check for logging control query vars.
 *
 * console.level=<level-name>
 * Set's the console log level by name.  Useful to override defaults and
 * allow more verbose logging before a user config is loaded.
 *
 * console.lock=<true|false>
 * Lock the console log level at whatever level it is set at.  This is run
 * after console.level is processed.  Useful to force a level of verbosity
 * that could otherwise be limited by a user config.
 */
if(sConsoleLogger !== null) ***REMOVED***
  var query = forge.util.getQueryVariables();
  if('console.level' in query) ***REMOVED***
    // set with last value
    forge.log.setLevel(
      sConsoleLogger, query['console.level'].slice(-1)[0]);
  ***REMOVED***
  if('console.lock' in query) ***REMOVED***
    // set with last value
    var lock = query['console.lock'].slice(-1)[0];
    if(lock == 'true') ***REMOVED***
      forge.log.lock(sConsoleLogger);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// provide public access to console logger
forge.log.consoleLogger = sConsoleLogger;
