// node-forge/lib/pkcs7.js
/**
 * Javascript implementation of PKCS#7 v1.5.
 *
 * @author Stefan Siegl
 * @author Dave Longley
 *
 * Copyright (c) 2012 Stefan Siegl <stesie@brokenpipe.de>
 * Copyright (c) 2012-2015 Digital Bazaar, Inc.
 *
 * Currently this implementation only supports ContentType of EnvelopedData,
 * EncryptedData, or SignedData at the root level. The top level elements may
 * contain only a ContentInfo of ContentType Data, i.e. plain data. Further
 * nesting is not (yet) supported.
 *
 * The Forge validators for PKCS #7's ASN.1 structures are available from
 * a separate file pkcs7asn1.js, since those are referenced from other
 * PKCS standards like PKCS #12.
 */
var forge = require('./forge');
require('./aes');
require('./asn1');
require('./des');
require('./oids');
require('./pem');
require('./pkcs7asn1');
require('./random');
require('./util');
require('./x509');

// shortcut for ASN.1 API
var asn1 = forge.asn1;

// shortcut for PKCS#7 API
var p7 = module.exports = forge.pkcs7 = forge.pkcs7 || ***REMOVED******REMOVED***;

/**
 * Converts a PKCS#7 message from PEM format.
 *
 * @param pem the PEM-formatted PKCS#7 message.
 *
 * @return the PKCS#7 message.
 */
p7.messageFromPem = function(pem) ***REMOVED***
  var msg = forge.pem.decode(pem)[0];

  if(msg.type !== 'PKCS7') ***REMOVED***
    var error = new Error('Could not convert PKCS#7 message from PEM; PEM ' +
      'header type is not "PKCS#7".');
    error.headerType = msg.type;
    throw error;
  ***REMOVED***
  if(msg.procType && msg.procType.type === 'ENCRYPTED') ***REMOVED***
    throw new Error('Could not convert PKCS#7 message from PEM; PEM is encrypted.');
  ***REMOVED***

  // convert DER to ASN.1 object
  var obj = asn1.fromDer(msg.body);

  return p7.messageFromAsn1(obj);
***REMOVED***;

/**
 * Converts a PKCS#7 message to PEM format.
 *
 * @param msg The PKCS#7 message object
 * @param maxline The maximum characters per line, defaults to 64.
 *
 * @return The PEM-formatted PKCS#7 message.
 */
p7.messageToPem = function(msg, maxline) ***REMOVED***
  // convert to ASN.1, then DER, then PEM-encode
  var pemObj = ***REMOVED***
    type: 'PKCS7',
    body: asn1.toDer(msg.toAsn1()).getBytes()
  ***REMOVED***;
  return forge.pem.encode(pemObj, ***REMOVED***maxline: maxline***REMOVED***);
***REMOVED***;

/**
 * Converts a PKCS#7 message from an ASN.1 object.
 *
 * @param obj the ASN.1 representation of a ContentInfo.
 *
 * @return the PKCS#7 message.
 */
p7.messageFromAsn1 = function(obj) ***REMOVED***
  // validate root level ContentInfo and capture data
  var capture = ***REMOVED******REMOVED***;
  var errors = [];
  if(!asn1.validate(obj, p7.asn1.contentInfoValidator, capture, errors)) ***REMOVED***
    var error = new Error('Cannot read PKCS#7 message. ' +
      'ASN.1 object is not an PKCS#7 ContentInfo.');
    error.errors = errors;
    throw error;
  ***REMOVED***

  var contentType = asn1.derToOid(capture.contentType);
  var msg;

  switch(contentType) ***REMOVED***
    case forge.pki.oids.envelopedData:
      msg = p7.createEnvelopedData();
      break;

    case forge.pki.oids.encryptedData:
      msg = p7.createEncryptedData();
      break;

    case forge.pki.oids.signedData:
      msg = p7.createSignedData();
      break;

    default:
      throw new Error('Cannot read PKCS#7 message. ContentType with OID ' +
        contentType + ' is not (yet) supported.');
  ***REMOVED***

  msg.fromAsn1(capture.content.value[0]);
  return msg;
***REMOVED***;

p7.createSignedData = function() ***REMOVED***
  var msg = null;
  msg = ***REMOVED***
    type: forge.pki.oids.signedData,
    version: 1,
    certificates: [],
    crls: [],
    // TODO: add json-formatted signer stuff here?
    signers: [],
    // populated during sign()
    digestAlgorithmIdentifiers: [],
    contentInfo: null,
    signerInfos: [],

    fromAsn1: function(obj) ***REMOVED***
      // validate SignedData content block and capture data.
      _fromAsn1(msg, obj, p7.asn1.signedDataValidator);
      msg.certificates = [];
      msg.crls = [];
      msg.digestAlgorithmIdentifiers = [];
      msg.contentInfo = null;
      msg.signerInfos = [];

      var certs = msg.rawCapture.certificates.value;
      for(var i = 0; i < certs.length; ++i) ***REMOVED***
        msg.certificates.push(forge.pki.certificateFromAsn1(certs[i]));
      ***REMOVED***

      // TODO: parse crls
    ***REMOVED***,

    toAsn1: function() ***REMOVED***
      // degenerate case with no content
      if(!msg.contentInfo) ***REMOVED***
        msg.sign();
      ***REMOVED***

      var certs = [];
      for(var i = 0; i < msg.certificates.length; ++i) ***REMOVED***
        certs.push(forge.pki.certificateToAsn1(msg.certificates[i]));
      ***REMOVED***

      var crls = [];
      // TODO: implement CRLs

      // [0] SignedData
      var signedData = asn1.create(asn1.Class.CONTEXT_SPECIFIC, 0, true, [
        asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
          // Version
          asn1.create(asn1.Class.UNIVERSAL, asn1.Type.INTEGER, false,
            asn1.integerToDer(msg.version).getBytes()),
          // DigestAlgorithmIdentifiers
          asn1.create(
            asn1.Class.UNIVERSAL, asn1.Type.SET, true,
            msg.digestAlgorithmIdentifiers),
          // ContentInfo
          msg.contentInfo
        ])
      ]);
      if(certs.length > 0) ***REMOVED***
        // [0] IMPLICIT ExtendedCertificatesAndCertificates OPTIONAL
        signedData.value[0].value.push(
          asn1.create(asn1.Class.CONTEXT_SPECIFIC, 0, true, certs));
      ***REMOVED***
      if(crls.length > 0) ***REMOVED***
        // [1] IMPLICIT CertificateRevocationLists OPTIONAL
        signedData.value[0].value.push(
          asn1.create(asn1.Class.CONTEXT_SPECIFIC, 1, true, crls));
      ***REMOVED***
      // SignerInfos
      signedData.value[0].value.push(
        asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SET, true,
          msg.signerInfos));

      // ContentInfo
      return asn1.create(
        asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
          // ContentType
          asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
            asn1.oidToDer(msg.type).getBytes()),
          // [0] SignedData
          signedData
        ]);
    ***REMOVED***,

    /**
     * Add (another) entity to list of signers.
     *
     * Note: If authenticatedAttributes are provided, then, per RFC 2315,
     * they must include at least two attributes: content type and
     * message digest. The message digest attribute value will be
     * auto-calculated during signing and will be ignored if provided.
     *
     * Here's an example of providing these two attributes:
     *
     * forge.pkcs7.createSignedData();
     * p7.addSigner(***REMOVED***
     *   issuer: cert.issuer.attributes,
     *   serialNumber: cert.serialNumber,
     *   key: privateKey,
     *   digestAlgorithm: forge.pki.oids.sha1,
     *   authenticatedAttributes: [***REMOVED***
     *     type: forge.pki.oids.contentType,
     *     value: forge.pki.oids.data
     *   ***REMOVED***, ***REMOVED***
     *     type: forge.pki.oids.messageDigest
     *   ***REMOVED***]
     * ***REMOVED***);
     *
     * TODO: Support [subjectKeyIdentifier] as signer's ID.
     *
     * @param signer the signer information:
     *          key the signer's private key.
     *          [certificate] a certificate containing the public key
     *            associated with the signer's private key; use this option as
     *            an alternative to specifying signer.issuer and
     *            signer.serialNumber.
     *          [issuer] the issuer attributes (eg: cert.issuer.attributes).
     *          [serialNumber] the signer's certificate's serial number in
     *           hexadecimal (eg: cert.serialNumber).
     *          [digestAlgorithm] the message digest OID, as a string, to use
     *            (eg: forge.pki.oids.sha1).
     *          [authenticatedAttributes] an optional array of attributes
     *            to also sign along with the content.
     */
    addSigner: function(signer) ***REMOVED***
      var issuer = signer.issuer;
      var serialNumber = signer.serialNumber;
      if(signer.certificate) ***REMOVED***
        var cert = signer.certificate;
        if(typeof cert === 'string') ***REMOVED***
          cert = forge.pki.certificateFromPem(cert);
        ***REMOVED***
        issuer = cert.issuer.attributes;
        serialNumber = cert.serialNumber;
      ***REMOVED***
      var key = signer.key;
      if(!key) ***REMOVED***
        throw new Error(
          'Could not add PKCS#7 signer; no private key specified.');
      ***REMOVED***
      if(typeof key === 'string') ***REMOVED***
        key = forge.pki.privateKeyFromPem(key);
      ***REMOVED***

      // ensure OID known for digest algorithm
      var digestAlgorithm = signer.digestAlgorithm || forge.pki.oids.sha1;
      switch(digestAlgorithm) ***REMOVED***
      case forge.pki.oids.sha1:
      case forge.pki.oids.sha256:
      case forge.pki.oids.sha384:
      case forge.pki.oids.sha512:
      case forge.pki.oids.md5:
        break;
      default:
        throw new Error(
          'Could not add PKCS#7 signer; unknown message digest algorithm: ' +
          digestAlgorithm);
      ***REMOVED***

      // if authenticatedAttributes is present, then the attributes
      // must contain at least PKCS #9 content-type and message-digest
      var authenticatedAttributes = signer.authenticatedAttributes || [];
      if(authenticatedAttributes.length > 0) ***REMOVED***
        var contentType = false;
        var messageDigest = false;
        for(var i = 0; i < authenticatedAttributes.length; ++i) ***REMOVED***
          var attr = authenticatedAttributes[i];
          if(!contentType && attr.type === forge.pki.oids.contentType) ***REMOVED***
            contentType = true;
            if(messageDigest) ***REMOVED***
              break;
            ***REMOVED***
            continue;
          ***REMOVED***
          if(!messageDigest && attr.type === forge.pki.oids.messageDigest) ***REMOVED***
            messageDigest = true;
            if(contentType) ***REMOVED***
              break;
            ***REMOVED***
            continue;
          ***REMOVED***
        ***REMOVED***

        if(!contentType || !messageDigest) ***REMOVED***
          throw new Error('Invalid signer.authenticatedAttributes. If ' +
            'signer.authenticatedAttributes is specified, then it must ' +
            'contain at least two attributes, PKCS #9 content-type and ' +
            'PKCS #9 message-digest.');
        ***REMOVED***
      ***REMOVED***

      msg.signers.push(***REMOVED***
        key: key,
        version: 1,
        issuer: issuer,
        serialNumber: serialNumber,
        digestAlgorithm: digestAlgorithm,
        signatureAlgorithm: forge.pki.oids.rsaEncryption,
        signature: null,
        authenticatedAttributes: authenticatedAttributes,
        unauthenticatedAttributes: []
      ***REMOVED***);
    ***REMOVED***,

    /**
     * Signs the content.
     */
    sign: function() ***REMOVED***
      // auto-generate content info
      if(typeof msg.content !== 'object' || msg.contentInfo === null) ***REMOVED***
        // use Data ContentInfo
        msg.contentInfo = asn1.create(
          asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
            // ContentType
            asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
              asn1.oidToDer(forge.pki.oids.data).getBytes())
          ]);

        // add actual content, if present
        if('content' in msg) ***REMOVED***
          var content;
          if(msg.content instanceof forge.util.ByteBuffer) ***REMOVED***
            content = msg.content.bytes();
          ***REMOVED*** else if(typeof msg.content === 'string') ***REMOVED***
            content = forge.util.encodeUtf8(msg.content);
          ***REMOVED***

          msg.contentInfo.value.push(
            // [0] EXPLICIT content
            asn1.create(asn1.Class.CONTEXT_SPECIFIC, 0, true, [
              asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
                content)
            ]));
        ***REMOVED***
      ***REMOVED***

      // no signers, return early (degenerate case for certificate container)
      if(msg.signers.length === 0) ***REMOVED***
        return;
      ***REMOVED***

      // generate digest algorithm identifiers
      var mds = addDigestAlgorithmIds();

      // generate signerInfos
      addSignerInfos(mds);
    ***REMOVED***,

    verify: function() ***REMOVED***
      throw new Error('PKCS#7 signature verification not yet implemented.');
    ***REMOVED***,

    /**
     * Add a certificate.
     *
     * @param cert the certificate to add.
     */
    addCertificate: function(cert) ***REMOVED***
      // convert from PEM
      if(typeof cert === 'string') ***REMOVED***
        cert = forge.pki.certificateFromPem(cert);
      ***REMOVED***
      msg.certificates.push(cert);
    ***REMOVED***,

    /**
     * Add a certificate revokation list.
     *
     * @param crl the certificate revokation list to add.
     */
    addCertificateRevokationList: function(crl) ***REMOVED***
      throw new Error('PKCS#7 CRL support not yet implemented.');
    ***REMOVED***
  ***REMOVED***;
  return msg;

  function addDigestAlgorithmIds() ***REMOVED***
    var mds = ***REMOVED******REMOVED***;

    for(var i = 0; i < msg.signers.length; ++i) ***REMOVED***
      var signer = msg.signers[i];
      var oid = signer.digestAlgorithm;
      if(!(oid in mds)) ***REMOVED***
        // content digest
        mds[oid] = forge.md[forge.pki.oids[oid]].create();
      ***REMOVED***
      if(signer.authenticatedAttributes.length === 0) ***REMOVED***
        // no custom attributes to digest; use content message digest
        signer.md = mds[oid];
      ***REMOVED*** else ***REMOVED***
        // custom attributes to be digested; use own message digest
        // TODO: optimize to just copy message digest state if that
        // feature is ever supported with message digests
        signer.md = forge.md[forge.pki.oids[oid]].create();
      ***REMOVED***
    ***REMOVED***

    // add unique digest algorithm identifiers
    msg.digestAlgorithmIdentifiers = [];
    for(var oid in mds) ***REMOVED***
      msg.digestAlgorithmIdentifiers.push(
        // AlgorithmIdentifier
        asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
          // algorithm
          asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
            asn1.oidToDer(oid).getBytes()),
          // parameters (null)
          asn1.create(asn1.Class.UNIVERSAL, asn1.Type.NULL, false, '')
        ]));
    ***REMOVED***

    return mds;
  ***REMOVED***

  function addSignerInfos(mds) ***REMOVED***
    // Note: ContentInfo is a SEQUENCE with 2 values, second value is
    // the content field and is optional for a ContentInfo but required here
    // since signers are present
    if(msg.contentInfo.value.length < 2) ***REMOVED***
      throw new Error(
        'Could not sign PKCS#7 message; there is no content to sign.');
    ***REMOVED***

    // get ContentInfo content type
    var contentType = asn1.derToOid(msg.contentInfo.value[0].value);

    // get ContentInfo content
    var content = msg.contentInfo.value[1];
    // skip [0] EXPLICIT content wrapper
    content = content.value[0];

    // serialize content
    var bytes = asn1.toDer(content);

    // skip identifier and length per RFC 2315 9.3
    // skip identifier (1 byte)
    bytes.getByte();
    // read and discard length bytes
    asn1.getBerValueLength(bytes);
    bytes = bytes.getBytes();

    // digest content DER value bytes
    for(var oid in mds) ***REMOVED***
      mds[oid].start().update(bytes);
    ***REMOVED***

    // sign content
    var signingTime = new Date();
    for(var i = 0; i < msg.signers.length; ++i) ***REMOVED***
      var signer = msg.signers[i];

      if(signer.authenticatedAttributes.length === 0) ***REMOVED***
        // if ContentInfo content type is not "Data", then
        // authenticatedAttributes must be present per RFC 2315
        if(contentType !== forge.pki.oids.data) ***REMOVED***
          throw new Error(
            'Invalid signer; authenticatedAttributes must be present ' +
            'when the ContentInfo content type is not PKCS#7 Data.');
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        // process authenticated attributes
        // [0] IMPLICIT
        signer.authenticatedAttributesAsn1 = asn1.create(
          asn1.Class.CONTEXT_SPECIFIC, 0, true, []);

        // per RFC 2315, attributes are to be digested using a SET container
        // not the above [0] IMPLICIT container
        var attrsAsn1 = asn1.create(
          asn1.Class.UNIVERSAL, asn1.Type.SET, true, []);

        for(var ai = 0; ai < signer.authenticatedAttributes.length; ++ai) ***REMOVED***
          var attr = signer.authenticatedAttributes[ai];
          if(attr.type === forge.pki.oids.messageDigest) ***REMOVED***
            // use content message digest as value
            attr.value = mds[signer.digestAlgorithm].digest();
          ***REMOVED*** else if(attr.type === forge.pki.oids.signingTime) ***REMOVED***
            // auto-populate signing time if not already set
            if(!attr.value) ***REMOVED***
              attr.value = signingTime;
            ***REMOVED***
          ***REMOVED***

          // convert to ASN.1 and push onto Attributes SET (for signing) and
          // onto authenticatedAttributesAsn1 to complete SignedData ASN.1
          // TODO: optimize away duplication
          attrsAsn1.value.push(_attributeToAsn1(attr));
          signer.authenticatedAttributesAsn1.value.push(_attributeToAsn1(attr));
        ***REMOVED***

        // DER-serialize and digest SET OF attributes only
        bytes = asn1.toDer(attrsAsn1).getBytes();
        signer.md.start().update(bytes);
      ***REMOVED***

      // sign digest
      signer.signature = signer.key.sign(signer.md, 'RSASSA-PKCS1-V1_5');
    ***REMOVED***

    // add signer info
    msg.signerInfos = _signersToAsn1(msg.signers);
  ***REMOVED***
***REMOVED***;

/**
 * Creates an empty PKCS#7 message of type EncryptedData.
 *
 * @return the message.
 */
p7.createEncryptedData = function() ***REMOVED***
  var msg = null;
  msg = ***REMOVED***
    type: forge.pki.oids.encryptedData,
    version: 0,
    encryptedContent: ***REMOVED***
      algorithm: forge.pki.oids['aes256-CBC']
    ***REMOVED***,

    /**
     * Reads an EncryptedData content block (in ASN.1 format)
     *
     * @param obj The ASN.1 representation of the EncryptedData content block
     */
    fromAsn1: function(obj) ***REMOVED***
      // Validate EncryptedData content block and capture data.
      _fromAsn1(msg, obj, p7.asn1.encryptedDataValidator);
    ***REMOVED***,

    /**
     * Decrypt encrypted content
     *
     * @param key The (symmetric) key as a byte buffer
     */
    decrypt: function(key) ***REMOVED***
      if(key !== undefined) ***REMOVED***
        msg.encryptedContent.key = key;
      ***REMOVED***
      _decryptContent(msg);
    ***REMOVED***
  ***REMOVED***;
  return msg;
***REMOVED***;

/**
 * Creates an empty PKCS#7 message of type EnvelopedData.
 *
 * @return the message.
 */
p7.createEnvelopedData = function() ***REMOVED***
  var msg = null;
  msg = ***REMOVED***
    type: forge.pki.oids.envelopedData,
    version: 0,
    recipients: [],
    encryptedContent: ***REMOVED***
      algorithm: forge.pki.oids['aes256-CBC']
    ***REMOVED***,

    /**
     * Reads an EnvelopedData content block (in ASN.1 format)
     *
     * @param obj the ASN.1 representation of the EnvelopedData content block.
     */
    fromAsn1: function(obj) ***REMOVED***
      // validate EnvelopedData content block and capture data
      var capture = _fromAsn1(msg, obj, p7.asn1.envelopedDataValidator);
      msg.recipients = _recipientsFromAsn1(capture.recipientInfos.value);
    ***REMOVED***,

    toAsn1: function() ***REMOVED***
      // ContentInfo
      return asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
        // ContentType
        asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
          asn1.oidToDer(msg.type).getBytes()),
        // [0] EnvelopedData
        asn1.create(asn1.Class.CONTEXT_SPECIFIC, 0, true, [
          asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
            // Version
            asn1.create(asn1.Class.UNIVERSAL, asn1.Type.INTEGER, false,
              asn1.integerToDer(msg.version).getBytes()),
            // RecipientInfos
            asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SET, true,
              _recipientsToAsn1(msg.recipients)),
            // EncryptedContentInfo
            asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true,
              _encryptedContentToAsn1(msg.encryptedContent))
          ])
        ])
      ]);
    ***REMOVED***,

    /**
     * Find recipient by X.509 certificate's issuer.
     *
     * @param cert the certificate with the issuer to look for.
     *
     * @return the recipient object.
     */
    findRecipient: function(cert) ***REMOVED***
      var sAttr = cert.issuer.attributes;

      for(var i = 0; i < msg.recipients.length; ++i) ***REMOVED***
        var r = msg.recipients[i];
        var rAttr = r.issuer;

        if(r.serialNumber !== cert.serialNumber) ***REMOVED***
          continue;
        ***REMOVED***

        if(rAttr.length !== sAttr.length) ***REMOVED***
          continue;
        ***REMOVED***

        var match = true;
        for(var j = 0; j < sAttr.length; ++j) ***REMOVED***
          if(rAttr[j].type !== sAttr[j].type ||
            rAttr[j].value !== sAttr[j].value) ***REMOVED***
            match = false;
            break;
          ***REMOVED***
        ***REMOVED***

        if(match) ***REMOVED***
          return r;
        ***REMOVED***
      ***REMOVED***

      return null;
    ***REMOVED***,

    /**
     * Decrypt enveloped content
     *
     * @param recipient The recipient object related to the private key
     * @param privKey The (RSA) private key object
     */
    decrypt: function(recipient, privKey) ***REMOVED***
      if(msg.encryptedContent.key === undefined && recipient !== undefined &&
        privKey !== undefined) ***REMOVED***
        switch(recipient.encryptedContent.algorithm) ***REMOVED***
          case forge.pki.oids.rsaEncryption:
          case forge.pki.oids.desCBC:
            var key = privKey.decrypt(recipient.encryptedContent.content);
            msg.encryptedContent.key = forge.util.createBuffer(key);
            break;

          default:
            throw new Error('Unsupported asymmetric cipher, ' +
              'OID ' + recipient.encryptedContent.algorithm);
        ***REMOVED***
      ***REMOVED***

      _decryptContent(msg);
    ***REMOVED***,

    /**
     * Add (another) entity to list of recipients.
     *
     * @param cert The certificate of the entity to add.
     */
    addRecipient: function(cert) ***REMOVED***
      msg.recipients.push(***REMOVED***
        version: 0,
        issuer: cert.issuer.attributes,
        serialNumber: cert.serialNumber,
        encryptedContent: ***REMOVED***
          // We simply assume rsaEncryption here, since forge.pki only
          // supports RSA so far.  If the PKI module supports other
          // ciphers one day, we need to modify this one as well.
          algorithm: forge.pki.oids.rsaEncryption,
          key: cert.publicKey
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***,

    /**
     * Encrypt enveloped content.
     *
     * This function supports two optional arguments, cipher and key, which
     * can be used to influence symmetric encryption.  Unless cipher is
     * provided, the cipher specified in encryptedContent.algorithm is used
     * (defaults to AES-256-CBC).  If no key is provided, encryptedContent.key
     * is (re-)used.  If that one's not set, a random key will be generated
     * automatically.
     *
     * @param [key] The key to be used for symmetric encryption.
     * @param [cipher] The OID of the symmetric cipher to use.
     */
    encrypt: function(key, cipher) ***REMOVED***
      // Part 1: Symmetric encryption
      if(msg.encryptedContent.content === undefined) ***REMOVED***
        cipher = cipher || msg.encryptedContent.algorithm;
        key = key || msg.encryptedContent.key;

        var keyLen, ivLen, ciphFn;
        switch(cipher) ***REMOVED***
          case forge.pki.oids['aes128-CBC']:
            keyLen = 16;
            ivLen = 16;
            ciphFn = forge.aes.createEncryptionCipher;
            break;

          case forge.pki.oids['aes192-CBC']:
            keyLen = 24;
            ivLen = 16;
            ciphFn = forge.aes.createEncryptionCipher;
            break;

          case forge.pki.oids['aes256-CBC']:
            keyLen = 32;
            ivLen = 16;
            ciphFn = forge.aes.createEncryptionCipher;
            break;

          case forge.pki.oids['des-EDE3-CBC']:
            keyLen = 24;
            ivLen = 8;
            ciphFn = forge.des.createEncryptionCipher;
            break;

          default:
            throw new Error('Unsupported symmetric cipher, OID ' + cipher);
        ***REMOVED***

        if(key === undefined) ***REMOVED***
          key = forge.util.createBuffer(forge.random.getBytes(keyLen));
        ***REMOVED*** else if(key.length() != keyLen) ***REMOVED***
          throw new Error('Symmetric key has wrong length; ' +
            'got ' + key.length() + ' bytes, expected ' + keyLen + '.');
        ***REMOVED***

        // Keep a copy of the key & IV in the object, so the caller can
        // use it for whatever reason.
        msg.encryptedContent.algorithm = cipher;
        msg.encryptedContent.key = key;
        msg.encryptedContent.parameter = forge.util.createBuffer(
          forge.random.getBytes(ivLen));

        var ciph = ciphFn(key);
        ciph.start(msg.encryptedContent.parameter.copy());
        ciph.update(msg.content);

        // The finish function does PKCS#7 padding by default, therefore
        // no action required by us.
        if(!ciph.finish()) ***REMOVED***
          throw new Error('Symmetric encryption failed.');
        ***REMOVED***

        msg.encryptedContent.content = ciph.output;
      ***REMOVED***

      // Part 2: asymmetric encryption for each recipient
      for(var i = 0; i < msg.recipients.length; ++i) ***REMOVED***
        var recipient = msg.recipients[i];

        // Nothing to do, encryption already done.
        if(recipient.encryptedContent.content !== undefined) ***REMOVED***
          continue;
        ***REMOVED***

        switch(recipient.encryptedContent.algorithm) ***REMOVED***
          case forge.pki.oids.rsaEncryption:
            recipient.encryptedContent.content =
              recipient.encryptedContent.key.encrypt(
                msg.encryptedContent.key.data);
            break;

          default:
            throw new Error('Unsupported asymmetric cipher, OID ' +
              recipient.encryptedContent.algorithm);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
  return msg;
***REMOVED***;

/**
 * Converts a single recipient from an ASN.1 object.
 *
 * @param obj the ASN.1 RecipientInfo.
 *
 * @return the recipient object.
 */
function _recipientFromAsn1(obj) ***REMOVED***
  // validate EnvelopedData content block and capture data
  var capture = ***REMOVED******REMOVED***;
  var errors = [];
  if(!asn1.validate(obj, p7.asn1.recipientInfoValidator, capture, errors)) ***REMOVED***
    var error = new Error('Cannot read PKCS#7 RecipientInfo. ' +
      'ASN.1 object is not an PKCS#7 RecipientInfo.');
    error.errors = errors;
    throw error;
  ***REMOVED***

  return ***REMOVED***
    version: capture.version.charCodeAt(0),
    issuer: forge.pki.RDNAttributesAsArray(capture.issuer),
    serialNumber: forge.util.createBuffer(capture.serial).toHex(),
    encryptedContent: ***REMOVED***
      algorithm: asn1.derToOid(capture.encAlgorithm),
      parameter: capture.encParameter.value,
      content: capture.encKey
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

/**
 * Converts a single recipient object to an ASN.1 object.
 *
 * @param obj the recipient object.
 *
 * @return the ASN.1 RecipientInfo.
 */
function _recipientToAsn1(obj) ***REMOVED***
  return asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
    // Version
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.INTEGER, false,
      asn1.integerToDer(obj.version).getBytes()),
    // IssuerAndSerialNumber
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
      // Name
      forge.pki.distinguishedNameToAsn1(***REMOVED***attributes: obj.issuer***REMOVED***),
      // Serial
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.INTEGER, false,
        forge.util.hexToBytes(obj.serialNumber))
    ]),
    // KeyEncryptionAlgorithmIdentifier
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
      // Algorithm
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
        asn1.oidToDer(obj.encryptedContent.algorithm).getBytes()),
      // Parameter, force NULL, only RSA supported for now.
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.NULL, false, '')
    ]),
    // EncryptedKey
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
      obj.encryptedContent.content)
  ]);
***REMOVED***

/**
 * Map a set of RecipientInfo ASN.1 objects to recipient objects.
 *
 * @param infos an array of ASN.1 representations RecipientInfo (i.e. SET OF).
 *
 * @return an array of recipient objects.
 */
function _recipientsFromAsn1(infos) ***REMOVED***
  var ret = [];
  for(var i = 0; i < infos.length; ++i) ***REMOVED***
    ret.push(_recipientFromAsn1(infos[i]));
  ***REMOVED***
  return ret;
***REMOVED***

/**
 * Map an array of recipient objects to ASN.1 RecipientInfo objects.
 *
 * @param recipients an array of recipientInfo objects.
 *
 * @return an array of ASN.1 RecipientInfos.
 */
function _recipientsToAsn1(recipients) ***REMOVED***
  var ret = [];
  for(var i = 0; i < recipients.length; ++i) ***REMOVED***
    ret.push(_recipientToAsn1(recipients[i]));
  ***REMOVED***
  return ret;
***REMOVED***

/**
 * Converts a single signer from an ASN.1 object.
 *
 * @param obj the ASN.1 representation of a SignerInfo.
 *
 * @return the signer object.
 */
function _signerFromAsn1(obj) ***REMOVED***
  // validate EnvelopedData content block and capture data
  var capture = ***REMOVED******REMOVED***;
  var errors = [];
  if(!asn1.validate(obj, p7.asn1.signerInfoValidator, capture, errors)) ***REMOVED***
    var error = new Error('Cannot read PKCS#7 SignerInfo. ' +
      'ASN.1 object is not an PKCS#7 SignerInfo.');
    error.errors = errors;
    throw error;
  ***REMOVED***

  var rval = ***REMOVED***
    version: capture.version.charCodeAt(0),
    issuer: forge.pki.RDNAttributesAsArray(capture.issuer),
    serialNumber: forge.util.createBuffer(capture.serial).toHex(),
    digestAlgorithm: asn1.derToOid(capture.digestAlgorithm),
    signatureAlgorithm: asn1.derToOid(capture.signatureAlgorithm),
    signature: capture.signature,
    authenticatedAttributes: [],
    unauthenticatedAttributes: []
  ***REMOVED***;

  // TODO: convert attributes
  var authenticatedAttributes = capture.authenticatedAttributes || [];
  var unauthenticatedAttributes = capture.unauthenticatedAttributes || [];

  return rval;
***REMOVED***

/**
 * Converts a single signerInfo object to an ASN.1 object.
 *
 * @param obj the signerInfo object.
 *
 * @return the ASN.1 representation of a SignerInfo.
 */
function _signerToAsn1(obj) ***REMOVED***
  // SignerInfo
  var rval = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
    // version
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.INTEGER, false,
      asn1.integerToDer(obj.version).getBytes()),
    // issuerAndSerialNumber
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
      // name
      forge.pki.distinguishedNameToAsn1(***REMOVED***attributes: obj.issuer***REMOVED***),
      // serial
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.INTEGER, false,
        forge.util.hexToBytes(obj.serialNumber))
    ]),
    // digestAlgorithm
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
      // algorithm
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
        asn1.oidToDer(obj.digestAlgorithm).getBytes()),
      // parameters (null)
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.NULL, false, '')
    ])
  ]);

  // authenticatedAttributes (OPTIONAL)
  if(obj.authenticatedAttributesAsn1) ***REMOVED***
    // add ASN.1 previously generated during signing
    rval.value.push(obj.authenticatedAttributesAsn1);
  ***REMOVED***

  // digestEncryptionAlgorithm
  rval.value.push(asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
    // algorithm
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
      asn1.oidToDer(obj.signatureAlgorithm).getBytes()),
    // parameters (null)
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.NULL, false, '')
  ]));

  // encryptedDigest
  rval.value.push(asn1.create(
    asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false, obj.signature));

  // unauthenticatedAttributes (OPTIONAL)
  if(obj.unauthenticatedAttributes.length > 0) ***REMOVED***
    // [1] IMPLICIT
    var attrsAsn1 = asn1.create(asn1.Class.CONTEXT_SPECIFIC, 1, true, []);
    for(var i = 0; i < obj.unauthenticatedAttributes.length; ++i) ***REMOVED***
      var attr = obj.unauthenticatedAttributes[i];
      attrsAsn1.values.push(_attributeToAsn1(attr));
    ***REMOVED***
    rval.value.push(attrsAsn1);
  ***REMOVED***

  return rval;
***REMOVED***

/**
 * Map a set of SignerInfo ASN.1 objects to an array of signer objects.
 *
 * @param signerInfoAsn1s an array of ASN.1 SignerInfos (i.e. SET OF).
 *
 * @return an array of signers objects.
 */
function _signersFromAsn1(signerInfoAsn1s) ***REMOVED***
  var ret = [];
  for(var i = 0; i < signerInfoAsn1s.length; ++i) ***REMOVED***
    ret.push(_signerFromAsn1(signerInfoAsn1s[i]));
  ***REMOVED***
  return ret;
***REMOVED***

/**
 * Map an array of signer objects to ASN.1 objects.
 *
 * @param signers an array of signer objects.
 *
 * @return an array of ASN.1 SignerInfos.
 */
function _signersToAsn1(signers) ***REMOVED***
  var ret = [];
  for(var i = 0; i < signers.length; ++i) ***REMOVED***
    ret.push(_signerToAsn1(signers[i]));
  ***REMOVED***
  return ret;
***REMOVED***

/**
 * Convert an attribute object to an ASN.1 Attribute.
 *
 * @param attr the attribute object.
 *
 * @return the ASN.1 Attribute.
 */
function _attributeToAsn1(attr) ***REMOVED***
  var value;

  // TODO: generalize to support more attributes
  if(attr.type === forge.pki.oids.contentType) ***REMOVED***
    value = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
      asn1.oidToDer(attr.value).getBytes());
  ***REMOVED*** else if(attr.type === forge.pki.oids.messageDigest) ***REMOVED***
    value = asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
      attr.value.bytes());
  ***REMOVED*** else if(attr.type === forge.pki.oids.signingTime) ***REMOVED***
    /* Note per RFC 2985: Dates between 1 January 1950 and 31 December 2049
      (inclusive) MUST be encoded as UTCTime. Any dates with year values
      before 1950 or after 2049 MUST be encoded as GeneralizedTime. [Further,]
      UTCTime values MUST be expressed in Greenwich Mean Time (Zulu) and MUST
      include seconds (i.e., times are YYMMDDHHMMSSZ), even where the
      number of seconds is zero.  Midnight (GMT) must be represented as
      "YYMMDD000000Z". */
    // TODO: make these module-level constants
    var jan_1_1950 = new Date('1950-01-01T00:00:00Z');
    var jan_1_2050 = new Date('2050-01-01T00:00:00Z');
    var date = attr.value;
    if(typeof date === 'string') ***REMOVED***
      // try to parse date
      var timestamp = Date.parse(date);
      if(!isNaN(timestamp)) ***REMOVED***
        date = new Date(timestamp);
      ***REMOVED*** else if(date.length === 13) ***REMOVED***
        // YYMMDDHHMMSSZ (13 chars for UTCTime)
        date = asn1.utcTimeToDate(date);
      ***REMOVED*** else ***REMOVED***
        // assume generalized time
        date = asn1.generalizedTimeToDate(date);
      ***REMOVED***
    ***REMOVED***

    if(date >= jan_1_1950 && date < jan_1_2050) ***REMOVED***
      value = asn1.create(
        asn1.Class.UNIVERSAL, asn1.Type.UTCTIME, false,
        asn1.dateToUtcTime(date));
    ***REMOVED*** else ***REMOVED***
      value = asn1.create(
        asn1.Class.UNIVERSAL, asn1.Type.GENERALIZEDTIME, false,
        asn1.dateToGeneralizedTime(date));
    ***REMOVED***
  ***REMOVED***

  // TODO: expose as common API call
  // create a RelativeDistinguishedName set
  // each value in the set is an AttributeTypeAndValue first
  // containing the type (an OID) and second the value
  return asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
    // AttributeType
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
      asn1.oidToDer(attr.type).getBytes()),
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SET, true, [
      // AttributeValue
      value
    ])
  ]);
***REMOVED***

/**
 * Map messages encrypted content to ASN.1 objects.
 *
 * @param ec The encryptedContent object of the message.
 *
 * @return ASN.1 representation of the encryptedContent object (SEQUENCE).
 */
function _encryptedContentToAsn1(ec) ***REMOVED***
  return [
    // ContentType, always Data for the moment
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
      asn1.oidToDer(forge.pki.oids.data).getBytes()),
    // ContentEncryptionAlgorithmIdentifier
    asn1.create(asn1.Class.UNIVERSAL, asn1.Type.SEQUENCE, true, [
      // Algorithm
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OID, false,
        asn1.oidToDer(ec.algorithm).getBytes()),
      // Parameters (IV)
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
        ec.parameter.getBytes())
    ]),
    // [0] EncryptedContent
    asn1.create(asn1.Class.CONTEXT_SPECIFIC, 0, true, [
      asn1.create(asn1.Class.UNIVERSAL, asn1.Type.OCTETSTRING, false,
        ec.content.getBytes())
    ])
  ];
***REMOVED***

/**
 * Reads the "common part" of an PKCS#7 content block (in ASN.1 format)
 *
 * This function reads the "common part" of the PKCS#7 content blocks
 * EncryptedData and EnvelopedData, i.e. version number and symmetrically
 * encrypted content block.
 *
 * The result of the ASN.1 validate and capture process is returned
 * to allow the caller to extract further data, e.g. the list of recipients
 * in case of a EnvelopedData object.
 *
 * @param msg the PKCS#7 object to read the data to.
 * @param obj the ASN.1 representation of the content block.
 * @param validator the ASN.1 structure validator object to use.
 *
 * @return the value map captured by validator object.
 */
function _fromAsn1(msg, obj, validator) ***REMOVED***
  var capture = ***REMOVED******REMOVED***;
  var errors = [];
  if(!asn1.validate(obj, validator, capture, errors)) ***REMOVED***
    var error = new Error('Cannot read PKCS#7 message. ' +
      'ASN.1 object is not a supported PKCS#7 message.');
    error.errors = error;
    throw error;
  ***REMOVED***

  // Check contentType, so far we only support (raw) Data.
  var contentType = asn1.derToOid(capture.contentType);
  if(contentType !== forge.pki.oids.data) ***REMOVED***
    throw new Error('Unsupported PKCS#7 message. ' +
      'Only wrapped ContentType Data supported.');
  ***REMOVED***

  if(capture.encryptedContent) ***REMOVED***
    var content = '';
    if(forge.util.isArray(capture.encryptedContent)) ***REMOVED***
      for(var i = 0; i < capture.encryptedContent.length; ++i) ***REMOVED***
        if(capture.encryptedContent[i].type !== asn1.Type.OCTETSTRING) ***REMOVED***
          throw new Error('Malformed PKCS#7 message, expecting encrypted ' +
            'content constructed of only OCTET STRING objects.');
        ***REMOVED***
        content += capture.encryptedContent[i].value;
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      content = capture.encryptedContent;
    ***REMOVED***
    msg.encryptedContent = ***REMOVED***
      algorithm: asn1.derToOid(capture.encAlgorithm),
      parameter: forge.util.createBuffer(capture.encParameter.value),
      content: forge.util.createBuffer(content)
    ***REMOVED***;
  ***REMOVED***

  if(capture.content) ***REMOVED***
    var content = '';
    if(forge.util.isArray(capture.content)) ***REMOVED***
      for(var i = 0; i < capture.content.length; ++i) ***REMOVED***
        if(capture.content[i].type !== asn1.Type.OCTETSTRING) ***REMOVED***
          throw new Error('Malformed PKCS#7 message, expecting ' +
            'content constructed of only OCTET STRING objects.');
        ***REMOVED***
        content += capture.content[i].value;
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      content = capture.content;
    ***REMOVED***
    msg.content = forge.util.createBuffer(content);
  ***REMOVED***

  msg.version = capture.version.charCodeAt(0);
  msg.rawCapture = capture;

  return capture;
***REMOVED***

/**
 * Decrypt the symmetrically encrypted content block of the PKCS#7 message.
 *
 * Decryption is skipped in case the PKCS#7 message object already has a
 * (decrypted) content attribute.  The algorithm, key and cipher parameters
 * (probably the iv) are taken from the encryptedContent attribute of the
 * message object.
 *
 * @param The PKCS#7 message object.
 */
function _decryptContent(msg) ***REMOVED***
  if(msg.encryptedContent.key === undefined) ***REMOVED***
    throw new Error('Symmetric key not available.');
  ***REMOVED***

  if(msg.content === undefined) ***REMOVED***
    var ciph;

    switch(msg.encryptedContent.algorithm) ***REMOVED***
      case forge.pki.oids['aes128-CBC']:
      case forge.pki.oids['aes192-CBC']:
      case forge.pki.oids['aes256-CBC']:
        ciph = forge.aes.createDecryptionCipher(msg.encryptedContent.key);
        break;

      case forge.pki.oids['desCBC']:
      case forge.pki.oids['des-EDE3-CBC']:
        ciph = forge.des.createDecryptionCipher(msg.encryptedContent.key);
        break;

      default:
        throw new Error('Unsupported symmetric cipher, OID ' +
          msg.encryptedContent.algorithm);
    ***REMOVED***
    ciph.start(msg.encryptedContent.parameter);
    ciph.update(msg.encryptedContent.content);

    if(!ciph.finish()) ***REMOVED***
      throw new Error('Symmetric decryption failed.');
    ***REMOVED***

    msg.content = ciph.output;
  ***REMOVED***
***REMOVED***
