// node-forge/lib/socket.js
/**
 * Socket implementation that uses flash SocketPool class as a backend.
 *
 * @author Dave Longley
 *
 * Copyright (c) 2010-2013 Digital Bazaar, Inc.
 */
var forge = require('./forge');
require('./util');

// define net namespace
var net = module.exports = forge.net = forge.net || ***REMOVED******REMOVED***;

// map of flash ID to socket pool
net.socketPools = ***REMOVED******REMOVED***;

/**
 * Creates a flash socket pool.
 *
 * @param options:
 *          flashId: the dom ID for the flash object element.
 *          policyPort: the default policy port for sockets, 0 to use the
 *            flash default.
 *          policyUrl: the default policy file URL for sockets (if provided
 *            used instead of a policy port).
 *          msie: true if the browser is msie, false if not.
 *
 * @return the created socket pool.
 */
net.createSocketPool = function(options) ***REMOVED***
  // set default
  options.msie = options.msie || false;

  // initialize the flash interface
  var spId = options.flashId;
  var api = document.getElementById(spId);
  api.init(***REMOVED***marshallExceptions: !options.msie***REMOVED***);

  // create socket pool entry
  var sp = ***REMOVED***
    // ID of the socket pool
    id: spId,
    // flash interface
    flashApi: api,
    // map of socket ID to sockets
    sockets: ***REMOVED******REMOVED***,
    // default policy port
    policyPort: options.policyPort || 0,
    // default policy URL
    policyUrl: options.policyUrl || null
  ***REMOVED***;
  net.socketPools[spId] = sp;

  // create event handler, subscribe to flash events
  if(options.msie === true) ***REMOVED***
    sp.handler = function(e) ***REMOVED***
      if(e.id in sp.sockets) ***REMOVED***
        // get handler function
        var f;
        switch(e.type) ***REMOVED***
        case 'connect':
          f = 'connected';
          break;
        case 'close':
          f = 'closed';
          break;
        case 'socketData':
          f = 'data';
          break;
        default:
          f = 'error';
          break;
        ***REMOVED***
        /* IE calls javascript on the thread of the external object
          that triggered the event (in this case flash) ... which will
          either run concurrently with other javascript or pre-empt any
          running javascript in the middle of its execution (BAD!) ...
          calling setTimeout() will schedule the javascript to run on
          the javascript thread and solve this EVIL problem. */
        setTimeout(function() ***REMOVED***sp.sockets[e.id][f](e);***REMOVED***, 0);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED*** else ***REMOVED***
    sp.handler = function(e) ***REMOVED***
      if(e.id in sp.sockets) ***REMOVED***
        // get handler function
        var f;
        switch(e.type) ***REMOVED***
        case 'connect':
          f = 'connected';
          break;
        case 'close':
          f = 'closed';
          break;
        case 'socketData':
          f = 'data';
          break;
        default:
          f = 'error';
          break;
        ***REMOVED***
        sp.sockets[e.id][f](e);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
  var handler = 'forge.net.socketPools[\'' + spId + '\'].handler';
  api.subscribe('connect', handler);
  api.subscribe('close', handler);
  api.subscribe('socketData', handler);
  api.subscribe('ioError', handler);
  api.subscribe('securityError', handler);

  /**
   * Destroys a socket pool. The socket pool still needs to be cleaned
   * up via net.cleanup().
   */
  sp.destroy = function() ***REMOVED***
    delete net.socketPools[options.flashId];
    for(var id in sp.sockets) ***REMOVED***
      sp.sockets[id].destroy();
    ***REMOVED***
    sp.sockets = ***REMOVED******REMOVED***;
    api.cleanup();
  ***REMOVED***;

  /**
   * Creates a new socket.
   *
   * @param options:
   *          connected: function(event) called when the socket connects.
   *          closed: function(event) called when the socket closes.
   *          data: function(event) called when socket data has arrived,
   *            it can be read from the socket using receive().
   *          error: function(event) called when a socket error occurs.
   */
   sp.createSocket = function(options) ***REMOVED***
     // default to empty options
     options = options || ***REMOVED******REMOVED***;

     // create flash socket
     var id = api.create();

     // create javascript socket wrapper
     var socket = ***REMOVED***
       id: id,
       // set handlers
       connected: options.connected || function(e) ***REMOVED******REMOVED***,
       closed: options.closed || function(e) ***REMOVED******REMOVED***,
       data: options.data || function(e) ***REMOVED******REMOVED***,
       error: options.error || function(e) ***REMOVED******REMOVED***
     ***REMOVED***;

     /**
      * Destroys this socket.
      */
     socket.destroy = function() ***REMOVED***
       api.destroy(id);
       delete sp.sockets[id];
     ***REMOVED***;

     /**
      * Connects this socket.
      *
      * @param options:
      *          host: the host to connect to.
      *          port: the port to connect to.
      *          policyPort: the policy port to use (if non-default), 0 to
      *            use the flash default.
      *          policyUrl: the policy file URL to use (instead of port).
      */
     socket.connect = function(options) ***REMOVED***
       // give precedence to policy URL over policy port
       // if no policy URL and passed port isn't 0, use default port,
       // otherwise use 0 for the port
       var policyUrl = options.policyUrl || null;
       var policyPort = 0;
       if(policyUrl === null && options.policyPort !== 0) ***REMOVED***
         policyPort = options.policyPort || sp.policyPort;
       ***REMOVED***
       api.connect(id, options.host, options.port, policyPort, policyUrl);
     ***REMOVED***;

     /**
      * Closes this socket.
      */
     socket.close = function() ***REMOVED***
       api.close(id);
       socket.closed(***REMOVED***
         id: socket.id,
         type: 'close',
         bytesAvailable: 0
       ***REMOVED***);
     ***REMOVED***;

     /**
      * Determines if the socket is connected or not.
      *
      * @return true if connected, false if not.
      */
     socket.isConnected = function() ***REMOVED***
       return api.isConnected(id);
     ***REMOVED***;

     /**
      * Writes bytes to this socket.
      *
      * @param bytes the bytes (as a string) to write.
      *
      * @return true on success, false on failure.
      */
     socket.send = function(bytes) ***REMOVED***
       return api.send(id, forge.util.encode64(bytes));
     ***REMOVED***;

     /**
      * Reads bytes from this socket (non-blocking). Fewer than the number
      * of bytes requested may be read if enough bytes are not available.
      *
      * This method should be called from the data handler if there are
      * enough bytes available. To see how many bytes are available, check
      * the 'bytesAvailable' property on the event in the data handler or
      * call the bytesAvailable() function on the socket. If the browser is
      * msie, then the bytesAvailable() function should be used to avoid
      * race conditions. Otherwise, using the property on the data handler's
      * event may be quicker.
      *
      * @param count the maximum number of bytes to read.
      *
      * @return the bytes read (as a string) or null on error.
      */
     socket.receive = function(count) ***REMOVED***
       var rval = api.receive(id, count).rval;
       return (rval === null) ? null : forge.util.decode64(rval);
     ***REMOVED***;

     /**
      * Gets the number of bytes available for receiving on the socket.
      *
      * @return the number of bytes available for receiving.
      */
     socket.bytesAvailable = function() ***REMOVED***
       return api.getBytesAvailable(id);
     ***REMOVED***;

     // store and return socket
     sp.sockets[id] = socket;
     return socket;
  ***REMOVED***;

  return sp;
***REMOVED***;

/**
 * Destroys a flash socket pool.
 *
 * @param options:
 *          flashId: the dom ID for the flash object element.
 */
net.destroySocketPool = function(options) ***REMOVED***
  if(options.flashId in net.socketPools) ***REMOVED***
    var sp = net.socketPools[options.flashId];
    sp.destroy();
  ***REMOVED***
***REMOVED***;

/**
 * Creates a new socket.
 *
 * @param options:
 *          flashId: the dom ID for the flash object element.
 *          connected: function(event) called when the socket connects.
 *          closed: function(event) called when the socket closes.
 *          data: function(event) called when socket data has arrived, it
 *            can be read from the socket using receive().
 *          error: function(event) called when a socket error occurs.
 *
 * @return the created socket.
 */
net.createSocket = function(options) ***REMOVED***
  var socket = null;
  if(options.flashId in net.socketPools) ***REMOVED***
    // get related socket pool
    var sp = net.socketPools[options.flashId];
    socket = sp.createSocket(options);
  ***REMOVED***
  return socket;
***REMOVED***;
