// class-utils/index.js
'use strict';

var util = require('util');
var union = require('arr-union');
var define = require('define-property');
var staticExtend = require('static-extend');
var isObj = require('isobject');

/**
 * Expose class utils
 */

var cu = module.exports;

/**
 * Expose class utils: `cu`
 */

cu.isObject = function isObject(val) ***REMOVED***
  return isObj(val) || typeof val === 'function';
***REMOVED***;

/**
 * Returns true if an array has any of the given elements, or an
 * object has any of the give keys.
 *
 * ```js
 * cu.has(['a', 'b', 'c'], 'c');
 * //=> true
 *
 * cu.has(['a', 'b', 'c'], ['c', 'z']);
 * //=> true
 *
 * cu.has(***REMOVED***a: 'b', c: 'd'***REMOVED***, ['c', 'z']);
 * //=> true
 * ```
 * @param ***REMOVED***Object***REMOVED*** `obj`
 * @param ***REMOVED***String|Array***REMOVED*** `val`
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

cu.has = function has(obj, val) ***REMOVED***
  val = cu.arrayify(val);
  var len = val.length;

  if (cu.isObject(obj)) ***REMOVED***
    for (var key in obj) ***REMOVED***
      if (val.indexOf(key) > -1) ***REMOVED***
        return true;
      ***REMOVED***
    ***REMOVED***

    var keys = cu.nativeKeys(obj);
    return cu.has(keys, val);
  ***REMOVED***

  if (Array.isArray(obj)) ***REMOVED***
    var arr = obj;
    while (len--) ***REMOVED***
      if (arr.indexOf(val[len]) > -1) ***REMOVED***
        return true;
      ***REMOVED***
    ***REMOVED***
    return false;
  ***REMOVED***

  throw new TypeError('expected an array or object.');
***REMOVED***;

/**
 * Returns true if an array or object has all of the given values.
 *
 * ```js
 * cu.hasAll(['a', 'b', 'c'], 'c');
 * //=> true
 *
 * cu.hasAll(['a', 'b', 'c'], ['c', 'z']);
 * //=> false
 *
 * cu.hasAll(***REMOVED***a: 'b', c: 'd'***REMOVED***, ['c', 'z']);
 * //=> false
 * ```
 * @param ***REMOVED***Object|Array***REMOVED*** `val`
 * @param ***REMOVED***String|Array***REMOVED*** `values`
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

cu.hasAll = function hasAll(val, values) ***REMOVED***
  values = cu.arrayify(values);
  var len = values.length;
  while (len--) ***REMOVED***
    if (!cu.has(val, values[len])) ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***
  return true;
***REMOVED***;

/**
 * Cast the given value to an array.
 *
 * ```js
 * cu.arrayify('foo');
 * //=> ['foo']
 *
 * cu.arrayify(['foo']);
 * //=> ['foo']
 * ```
 *
 * @param ***REMOVED***String|Array***REMOVED*** `val`
 * @return ***REMOVED***Array***REMOVED***
 * @api public
 */

cu.arrayify = function arrayify(val) ***REMOVED***
  return val ? (Array.isArray(val) ? val : [val]) : [];
***REMOVED***;

/**
 * Noop
 */

cu.noop = function noop() ***REMOVED***
  return;
***REMOVED***;

/**
 * Returns the first argument passed to the function.
 */

cu.identity = function identity(val) ***REMOVED***
  return val;
***REMOVED***;

/**
 * Returns true if a value has a `contructor`
 *
 * ```js
 * cu.hasConstructor(***REMOVED******REMOVED***);
 * //=> true
 *
 * cu.hasConstructor(Object.create(null));
 * //=> false
 * ```
 * @param  ***REMOVED***Object***REMOVED*** `value`
 * @return ***REMOVED***Boolean***REMOVED***
 * @api public
 */

cu.hasConstructor = function hasConstructor(val) ***REMOVED***
  return cu.isObject(val) && typeof val.constructor !== 'undefined';
***REMOVED***;

/**
 * Get the native `ownPropertyNames` from the constructor of the
 * given `object`. An empty array is returned if the object does
 * not have a constructor.
 *
 * ```js
 * cu.nativeKeys(***REMOVED***a: 'b', b: 'c', c: 'd'***REMOVED***)
 * //=> ['a', 'b', 'c']
 *
 * cu.nativeKeys(function()***REMOVED******REMOVED***)
 * //=> ['length', 'caller']
 * ```
 *
 * @param  ***REMOVED***Object***REMOVED*** `obj` Object that has a `constructor`.
 * @return ***REMOVED***Array***REMOVED*** Array of keys.
 * @api public
 */

cu.nativeKeys = function nativeKeys(val) ***REMOVED***
  if (!cu.hasConstructor(val)) return [];
  var keys = Object.getOwnPropertyNames(val);
  if ('caller' in val) keys.push('caller');
  return keys;
***REMOVED***;

/**
 * Returns property descriptor `key` if it's an "own" property
 * of the given object.
 *
 * ```js
 * function App() ***REMOVED******REMOVED***
 * Object.defineProperty(App.prototype, 'count', ***REMOVED***
 *   get: function() ***REMOVED***
 *     return Object.keys(this).length;
 *   ***REMOVED***
 * ***REMOVED***);
 * cu.getDescriptor(App.prototype, 'count');
 * // returns:
 * // ***REMOVED***
 * //   get: [Function],
 * //   set: undefined,
 * //   enumerable: false,
 * //   configurable: false
 * // ***REMOVED***
 * ```
 *
 * @param ***REMOVED***Object***REMOVED*** `obj`
 * @param ***REMOVED***String***REMOVED*** `key`
 * @return ***REMOVED***Object***REMOVED*** Returns descriptor `key`
 * @api public
 */

cu.getDescriptor = function getDescriptor(obj, key) ***REMOVED***
  if (!cu.isObject(obj)) ***REMOVED***
    throw new TypeError('expected an object.');
  ***REMOVED***
  if (typeof key !== 'string') ***REMOVED***
    throw new TypeError('expected key to be a string.');
  ***REMOVED***
  return Object.getOwnPropertyDescriptor(obj, key);
***REMOVED***;

/**
 * Copy a descriptor from one object to another.
 *
 * ```js
 * function App() ***REMOVED******REMOVED***
 * Object.defineProperty(App.prototype, 'count', ***REMOVED***
 *   get: function() ***REMOVED***
 *     return Object.keys(this).length;
 *   ***REMOVED***
 * ***REMOVED***);
 * var obj = ***REMOVED******REMOVED***;
 * cu.copyDescriptor(obj, App.prototype, 'count');
 * ```
 * @param ***REMOVED***Object***REMOVED*** `receiver`
 * @param ***REMOVED***Object***REMOVED*** `provider`
 * @param ***REMOVED***String***REMOVED*** `name`
 * @return ***REMOVED***Object***REMOVED***
 * @api public
 */

cu.copyDescriptor = function copyDescriptor(receiver, provider, name) ***REMOVED***
  if (!cu.isObject(receiver)) ***REMOVED***
    throw new TypeError('expected receiving object to be an object.');
  ***REMOVED***
  if (!cu.isObject(provider)) ***REMOVED***
    throw new TypeError('expected providing object to be an object.');
  ***REMOVED***
  if (typeof name !== 'string') ***REMOVED***
    throw new TypeError('expected name to be a string.');
  ***REMOVED***

  var val = cu.getDescriptor(provider, name);
  if (val) Object.defineProperty(receiver, name, val);
***REMOVED***;

/**
 * Copy static properties, prototype properties, and descriptors
 * from one object to another.
 *
 * @param ***REMOVED***Object***REMOVED*** `receiver`
 * @param ***REMOVED***Object***REMOVED*** `provider`
 * @param ***REMOVED***String|Array***REMOVED*** `omit` One or more properties to omit
 * @return ***REMOVED***Object***REMOVED***
 * @api public
 */

cu.copy = function copy(receiver, provider, omit) ***REMOVED***
  if (!cu.isObject(receiver)) ***REMOVED***
    throw new TypeError('expected receiving object to be an object.');
  ***REMOVED***
  if (!cu.isObject(provider)) ***REMOVED***
    throw new TypeError('expected providing object to be an object.');
  ***REMOVED***
  var props = Object.getOwnPropertyNames(provider);
  var keys = Object.keys(provider);
  var len = props.length,
    key;
  omit = cu.arrayify(omit);

  while (len--) ***REMOVED***
    key = props[len];

    if (cu.has(keys, key)) ***REMOVED***
      define(receiver, key, provider[key]);
    ***REMOVED*** else if (!(key in receiver) && !cu.has(omit, key)) ***REMOVED***
      cu.copyDescriptor(receiver, provider, key);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

/**
 * Inherit the static properties, prototype properties, and descriptors
 * from of an object.
 *
 * @param ***REMOVED***Object***REMOVED*** `receiver`
 * @param ***REMOVED***Object***REMOVED*** `provider`
 * @param ***REMOVED***String|Array***REMOVED*** `omit` One or more properties to omit
 * @return ***REMOVED***Object***REMOVED***
 * @api public
 */

cu.inherit = function inherit(receiver, provider, omit) ***REMOVED***
  if (!cu.isObject(receiver)) ***REMOVED***
    throw new TypeError('expected receiving object to be an object.');
  ***REMOVED***
  if (!cu.isObject(provider)) ***REMOVED***
    throw new TypeError('expected providing object to be an object.');
  ***REMOVED***

  var keys = [];
  for (var key in provider) ***REMOVED***
    keys.push(key);
    receiver[key] = provider[key];
  ***REMOVED***

  keys = keys.concat(cu.arrayify(omit));

  var a = provider.prototype || provider;
  var b = receiver.prototype || receiver;
  cu.copy(b, a, keys);
***REMOVED***;

/**
 * Returns a function for extending the static properties,
 * prototype properties, and descriptors from the `Parent`
 * constructor onto `Child` constructors.
 *
 * ```js
 * var extend = cu.extend(Parent);
 * Parent.extend(Child);
 *
 * // optional methods
 * Parent.extend(Child, ***REMOVED***
 *   foo: function() ***REMOVED******REMOVED***,
 *   bar: function() ***REMOVED******REMOVED***
 * ***REMOVED***);
 * ```
 * @param ***REMOVED***Function***REMOVED*** `Parent` Parent ctor
 * @param ***REMOVED***Function***REMOVED*** `extend` Optional extend function to handle custom extensions. Useful when updating methods that require a specific prototype.
 *   @param ***REMOVED***Function***REMOVED*** `Child` Child ctor
 *   @param ***REMOVED***Object***REMOVED*** `proto` Optionally pass additional prototype properties to inherit.
 *   @return ***REMOVED***Object***REMOVED***
 * @api public
 */

cu.extend = function() ***REMOVED***
  // keep it lazy, instead of assigning to `cu.extend`
  return staticExtend.apply(null, arguments);
***REMOVED***;

/**
 * Bubble up events emitted from static methods on the Parent ctor.
 *
 * @param ***REMOVED***Object***REMOVED*** `Parent`
 * @param ***REMOVED***Array***REMOVED*** `events` Event names to bubble up
 * @api public
 */

cu.bubble = function(Parent, events) ***REMOVED***
  events = events || [];
  Parent.bubble = function(Child, arr) ***REMOVED***
    if (Array.isArray(arr)) ***REMOVED***
      events = union([], events, arr);
    ***REMOVED***
    var len = events.length;
    var idx = -1;
    while (++idx < len) ***REMOVED***
      var name = events[idx];
      Parent.on(name, Child.emit.bind(Child, name));
    ***REMOVED***
    cu.bubble(Child, events);
  ***REMOVED***;
***REMOVED***;
