// webpack/lib/node/NodeWatchFileSystem.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Watchpack = require("watchpack");

class NodeWatchFileSystem ***REMOVED***
	constructor(inputFileSystem) ***REMOVED***
		this.inputFileSystem = inputFileSystem;
		this.watcherOptions = ***REMOVED***
			aggregateTimeout: 0
		***REMOVED***;
		this.watcher = new Watchpack(this.watcherOptions);
	***REMOVED***

	watch(files, dirs, missing, startTime, options, callback, callbackUndelayed) ***REMOVED***
		if(!Array.isArray(files))
			throw new Error("Invalid arguments: 'files'");
		if(!Array.isArray(dirs))
			throw new Error("Invalid arguments: 'dirs'");
		if(!Array.isArray(missing))
			throw new Error("Invalid arguments: 'missing'");
		if(typeof callback !== "function")
			throw new Error("Invalid arguments: 'callback'");
		if(typeof startTime !== "number" && startTime)
			throw new Error("Invalid arguments: 'startTime'");
		if(typeof options !== "object")
			throw new Error("Invalid arguments: 'options'");
		if(typeof callbackUndelayed !== "function" && callbackUndelayed)
			throw new Error("Invalid arguments: 'callbackUndelayed'");
		const oldWatcher = this.watcher;
		this.watcher = new Watchpack(options);

		if(callbackUndelayed)
			this.watcher.once("change", callbackUndelayed);

		this.watcher.once("aggregated", (changes, removals) => ***REMOVED***
			changes = changes.concat(removals);
			if(this.inputFileSystem && this.inputFileSystem.purge) ***REMOVED***
				this.inputFileSystem.purge(changes);
			***REMOVED***
			const times = this.watcher.getTimes();
			callback(null,
				changes.filter(file => files.indexOf(file) >= 0).sort(),
				changes.filter(file => dirs.indexOf(file) >= 0).sort(),
				changes.filter(file => missing.indexOf(file) >= 0).sort(), times, times);
		***REMOVED***);

		this.watcher.watch(files.concat(missing), dirs.concat(missing), startTime);

		if(oldWatcher) ***REMOVED***
			oldWatcher.close();
		***REMOVED***
		return ***REMOVED***
			close: () => ***REMOVED***
				if(this.watcher) ***REMOVED***
					this.watcher.close();
					this.watcher = null;
				***REMOVED***
			***REMOVED***,
			pause: () => ***REMOVED***
				if(this.watcher) ***REMOVED***
					this.watcher.pause();
				***REMOVED***
			***REMOVED***
		***REMOVED***;
	***REMOVED***
***REMOVED***

module.exports = NodeWatchFileSystem;
