// webpack/lib/node/NodeMainTemplatePlugin.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
"use strict";

const Template = require("../Template");

module.exports = class NodeMainTemplatePlugin ***REMOVED***
	constructor(asyncChunkLoading) ***REMOVED***
		this.asyncChunkLoading = asyncChunkLoading;
	***REMOVED***

	apply(mainTemplate) ***REMOVED***
		const asyncChunkLoading = this.asyncChunkLoading;
		mainTemplate.plugin("local-vars", function(source, chunk) ***REMOVED***
			if(chunk.chunks.length > 0) ***REMOVED***
				return this.asString([
					source,
					"",
					"// object to store loaded chunks",
					"// \"0\" means \"already loaded\"",
					"var installedChunks = ***REMOVED***",
					this.indent(chunk.ids.map((id) => `$***REMOVED***id***REMOVED***: 0`).join(",\n")),
					"***REMOVED***;"
				]);
			***REMOVED***
			return source;
		***REMOVED***);
		mainTemplate.plugin("require-extensions", function(source, chunk) ***REMOVED***
			if(chunk.chunks.length > 0) ***REMOVED***
				return this.asString([
					source,
					"",
					"// uncatched error handler for webpack runtime",
					`$***REMOVED***this.requireFn***REMOVED***.oe = function(err) ***REMOVED***`,
					this.indent([
						"process.nextTick(function() ***REMOVED***",
						this.indent("throw err; // catch this error by using System.import().catch()"),
						"***REMOVED***);"
					]),
					"***REMOVED***;"
				]);
			***REMOVED***
			return source;
		***REMOVED***);
		mainTemplate.plugin("require-ensure", function(_, chunk, hash) ***REMOVED***
			const chunkFilename = this.outputOptions.chunkFilename;
			const chunkMaps = chunk.getChunkMaps();
			const insertMoreModules = [
				"var moreModules = chunk.modules, chunkIds = chunk.ids;",
				"for(var moduleId in moreModules) ***REMOVED***",
				this.indent(this.renderAddModule(hash, chunk, "moduleId", "moreModules[moduleId]")),
				"***REMOVED***"
			];
			if(asyncChunkLoading) ***REMOVED***
				return this.asString([
					"// \"0\" is the signal for \"already loaded\"",
					"if(installedChunks[chunkId] === 0)",
					this.indent([
						"return Promise.resolve();"
					]),
					"// array of [resolve, reject, promise] means \"currently loading\"",
					"if(installedChunks[chunkId])",
					this.indent([
						"return installedChunks[chunkId][2];"
					]),
					"// load the chunk and return promise to it",
					"var promise = new Promise(function(resolve, reject) ***REMOVED***",
					this.indent([
						"installedChunks[chunkId] = [resolve, reject];",
						"var filename = __dirname + " + this.applyPluginsWaterfall("asset-path", JSON.stringify(`/$***REMOVED***chunkFilename***REMOVED***`), ***REMOVED***
							hash: `" + $***REMOVED***this.renderCurrentHashCode(hash)***REMOVED*** + "`,
							hashWithLength: (length) => `" + $***REMOVED***this.renderCurrentHashCode(hash, length)***REMOVED*** + "`,
							chunk: ***REMOVED***
								id: "\" + chunkId + \"",
								hash: `" + $***REMOVED***JSON.stringify(chunkMaps.hash)***REMOVED***[chunkId] + "`,
								hashWithLength: (length) => ***REMOVED***
									const shortChunkHashMap = ***REMOVED******REMOVED***;
									Object.keys(chunkMaps.hash).forEach((chunkId) => ***REMOVED***
										if(typeof chunkMaps.hash[chunkId] === "string")
											shortChunkHashMap[chunkId] = chunkMaps.hash[chunkId].substr(0, length);
									***REMOVED***);
									return `" + $***REMOVED***JSON.stringify(shortChunkHashMap)***REMOVED***[chunkId] + "`;
								***REMOVED***,
								name: `" + ($***REMOVED***JSON.stringify(chunkMaps.name)***REMOVED***[chunkId]||chunkId) + "`
							***REMOVED***
						***REMOVED***) + ";",
						"require('fs').readFile(filename, 'utf-8',  function(err, content) ***REMOVED***",
						this.indent([
							"if(err) return reject(err);",
							"var chunk = ***REMOVED******REMOVED***;",
							"require('vm').runInThisContext('(function(exports, require, __dirname, __filename) ***REMOVED***' + content + '\\n***REMOVED***)', filename)" +
							"(chunk, require, require('path').dirname(filename), filename);"
						].concat(insertMoreModules).concat([
							"var callbacks = [];",
							"for(var i = 0; i < chunkIds.length; i++) ***REMOVED***",
							this.indent([
								"if(installedChunks[chunkIds[i]])",
								this.indent([
									"callbacks = callbacks.concat(installedChunks[chunkIds[i]][0]);"
								]),
								"installedChunks[chunkIds[i]] = 0;"
							]),
							"***REMOVED***",
							"for(i = 0; i < callbacks.length; i++)",
							this.indent("callbacks[i]();")
						])),
						"***REMOVED***);"
					]),
					"***REMOVED***);",
					"return installedChunks[chunkId][2] = promise;"
				]);
			***REMOVED*** else ***REMOVED***
				const request = this.applyPluginsWaterfall("asset-path", JSON.stringify(`./$***REMOVED***chunkFilename***REMOVED***`), ***REMOVED***
					hash: `" + $***REMOVED***this.renderCurrentHashCode(hash)***REMOVED*** + "`,
					hashWithLength: (length) => `" + $***REMOVED***this.renderCurrentHashCode(hash, length)***REMOVED*** + "`,
					chunk: ***REMOVED***
						id: "\" + chunkId + \"",
						hash: `" + $***REMOVED***JSON.stringify(chunkMaps.hash)***REMOVED***[chunkId] + "`,
						hashWithLength: (length) => ***REMOVED***
							const shortChunkHashMap = ***REMOVED******REMOVED***;
							Object.keys(chunkMaps.hash).forEach((chunkId) => ***REMOVED***
								if(typeof chunkMaps.hash[chunkId] === "string")
									shortChunkHashMap[chunkId] = chunkMaps.hash[chunkId].substr(0, length);
							***REMOVED***);
							return `" + $***REMOVED***JSON.stringify(shortChunkHashMap)***REMOVED***[chunkId] + "`;
						***REMOVED***,
						name: `" + ($***REMOVED***JSON.stringify(chunkMaps.name)***REMOVED***[chunkId]||chunkId) + "`
					***REMOVED***
				***REMOVED***);
				return this.asString([
					"// \"0\" is the signal for \"already loaded\"",
					"if(installedChunks[chunkId] !== 0) ***REMOVED***",
					this.indent([
						`var chunk = require($***REMOVED***request***REMOVED***);`
					].concat(insertMoreModules).concat([
						"for(var i = 0; i < chunkIds.length; i++)",
						this.indent("installedChunks[chunkIds[i]] = 0;")
					])),
					"***REMOVED***",
					"return Promise.resolve();"
				]);
			***REMOVED***
		***REMOVED***);
		mainTemplate.plugin("hot-bootstrap", function(source, chunk, hash) ***REMOVED***
			const hotUpdateChunkFilename = this.outputOptions.hotUpdateChunkFilename;
			const hotUpdateMainFilename = this.outputOptions.hotUpdateMainFilename;
			const chunkMaps = chunk.getChunkMaps();
			const currentHotUpdateChunkFilename = this.applyPluginsWaterfall("asset-path", JSON.stringify(hotUpdateChunkFilename), ***REMOVED***
				hash: `" + $***REMOVED***this.renderCurrentHashCode(hash)***REMOVED*** + "`,
				hashWithLength: (length) => `" + $***REMOVED***this.renderCurrentHashCode(hash, length)***REMOVED*** + "`,
				chunk: ***REMOVED***
					id: "\" + chunkId + \"",
					hash: `" + $***REMOVED***JSON.stringify(chunkMaps.hash)***REMOVED***[chunkId] + "`,
					hashWithLength: (length) => ***REMOVED***
						const shortChunkHashMap = ***REMOVED******REMOVED***;
						Object.keys(chunkMaps.hash).forEach((chunkId) => ***REMOVED***
							if(typeof chunkMaps.hash[chunkId] === "string")
								shortChunkHashMap[chunkId] = chunkMaps.hash[chunkId].substr(0, length);
						***REMOVED***);
						return `" + $***REMOVED***JSON.stringify(shortChunkHashMap)***REMOVED***[chunkId] + "`;
					***REMOVED***,
					name: `" + ($***REMOVED***JSON.stringify(chunkMaps.name)***REMOVED***[chunkId]||chunkId) + "`
				***REMOVED***
			***REMOVED***);
			const currentHotUpdateMainFilename = this.applyPluginsWaterfall("asset-path", JSON.stringify(hotUpdateMainFilename), ***REMOVED***
				hash: `" + $***REMOVED***this.renderCurrentHashCode(hash)***REMOVED*** + "`,
				hashWithLength: (length) => `" + $***REMOVED***this.renderCurrentHashCode(hash, length)***REMOVED*** + "`
			***REMOVED***);
			return Template.getFunctionContent(asyncChunkLoading ? require("./NodeMainTemplateAsync.runtime.js") : require("./NodeMainTemplate.runtime.js"))
				.replace(/\$require\$/g, this.requireFn)
				.replace(/\$hotMainFilename\$/g, currentHotUpdateMainFilename)
				.replace(/\$hotChunkFilename\$/g, currentHotUpdateChunkFilename);
		***REMOVED***);
		mainTemplate.plugin("hash", function(hash) ***REMOVED***
			hash.update("node");
			hash.update("3");
			hash.update(this.outputOptions.filename + "");
			hash.update(this.outputOptions.chunkFilename + "");
		***REMOVED***);
	***REMOVED***
***REMOVED***;
