// @firebase/database/dist/cjs/src/core/PersistentConnection.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = Object.setPrototypeOf ||
        (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
        function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var app_1 = require("@firebase/app");
var util_1 = require("@firebase/util");
var util_2 = require("@firebase/util");
var util_3 = require("@firebase/util");
var util_4 = require("./util/util");
var Path_1 = require("./util/Path");
var VisibilityMonitor_1 = require("./util/VisibilityMonitor");
var OnlineMonitor_1 = require("./util/OnlineMonitor");
var util_5 = require("@firebase/util");
var Connection_1 = require("../realtime/Connection");
var util_6 = require("@firebase/util");
var util_7 = require("@firebase/util");
var ServerActions_1 = require("./ServerActions");
var RECONNECT_MIN_DELAY = 1000;
var RECONNECT_MAX_DELAY_DEFAULT = 60 * 5 * 1000; // 5 minutes in milliseconds (Case: 1858)
var RECONNECT_MAX_DELAY_FOR_ADMINS = 30 * 1000; // 30 seconds for admin clients (likely to be a backend server)
var RECONNECT_DELAY_MULTIPLIER = 1.3;
var RECONNECT_DELAY_RESET_TIMEOUT = 30000; // Reset delay back to MIN_DELAY after being connected for 30sec.
var SERVER_KILL_INTERRUPT_REASON = 'server_kill';
// If auth fails repeatedly, we'll assume something is wrong and log a warning / back off.
var INVALID_AUTH_TOKEN_THRESHOLD = 3;
/**
 * Firebase connection.  Abstracts wire protocol and handles reconnecting.
 *
 * NOTE: All JSON objects sent to the realtime connection must have property names enclosed
 * in quotes to make sure the closure compiler does not minify them.
 */
var PersistentConnection = /** @class */ (function (_super) ***REMOVED***
    __extends(PersistentConnection, _super);
    /**
     * @implements ***REMOVED***ServerActions***REMOVED***
     * @param ***REMOVED***!RepoInfo***REMOVED*** repoInfo_ Data about the namespace we are connecting to
     * @param ***REMOVED***function(string, *, boolean, ?number)***REMOVED*** onDataUpdate_ A callback for new data from the server
     * @param onConnectStatus_
     * @param onServerInfoUpdate_
     * @param authTokenProvider_
     * @param authOverride_
     */
    function PersistentConnection(repoInfo_, onDataUpdate_, onConnectStatus_, onServerInfoUpdate_, authTokenProvider_, authOverride_) ***REMOVED***
        var _this = _super.call(this) || this;
        _this.repoInfo_ = repoInfo_;
        _this.onDataUpdate_ = onDataUpdate_;
        _this.onConnectStatus_ = onConnectStatus_;
        _this.onServerInfoUpdate_ = onServerInfoUpdate_;
        _this.authTokenProvider_ = authTokenProvider_;
        _this.authOverride_ = authOverride_;
        // Used for diagnostic logging.
        _this.id = PersistentConnection.nextPersistentConnectionId_++;
        _this.log_ = util_4.logWrapper('p:' + _this.id + ':');
        /** @private ***REMOVED***Object***REMOVED*** */
        _this.interruptReasons_ = ***REMOVED******REMOVED***;
        _this.listens_ = ***REMOVED******REMOVED***;
        _this.outstandingPuts_ = [];
        _this.outstandingPutCount_ = 0;
        _this.onDisconnectRequestQueue_ = [];
        _this.connected_ = false;
        _this.reconnectDelay_ = RECONNECT_MIN_DELAY;
        _this.maxReconnectDelay_ = RECONNECT_MAX_DELAY_DEFAULT;
        _this.securityDebugCallback_ = null;
        _this.lastSessionId = null;
        /** @private ***REMOVED***number|null***REMOVED*** */
        _this.establishConnectionTimer_ = null;
        /** @private ***REMOVED***boolean***REMOVED*** */
        _this.visible_ = false;
        // Before we get connected, we keep a queue of pending messages to send.
        _this.requestCBHash_ = ***REMOVED******REMOVED***;
        _this.requestNumber_ = 0;
        /** @private ***REMOVED***?***REMOVED***
         *   sendRequest(Object),
         *   close()
         * ***REMOVED******REMOVED*** */
        _this.realtime_ = null;
        /** @private ***REMOVED***string|null***REMOVED*** */
        _this.authToken_ = null;
        _this.forceTokenRefresh_ = false;
        _this.invalidAuthTokenCount_ = 0;
        _this.firstConnection_ = true;
        _this.lastConnectionAttemptTime_ = null;
        _this.lastConnectionEstablishedTime_ = null;
        if (authOverride_ && !util_7.isNodeSdk()) ***REMOVED***
            throw new Error('Auth override specified in options, but not supported on non Node.js platforms');
        ***REMOVED***
        _this.scheduleConnect_(0);
        VisibilityMonitor_1.VisibilityMonitor.getInstance().on('visible', _this.onVisible_, _this);
        if (repoInfo_.host.indexOf('fblocal') === -1) ***REMOVED***
            OnlineMonitor_1.OnlineMonitor.getInstance().on('online', _this.onOnline_, _this);
        ***REMOVED***
        return _this;
    ***REMOVED***
    /**
     * @param ***REMOVED***!string***REMOVED*** action
     * @param ***REMOVED*******REMOVED*** body
     * @param ***REMOVED***function(*)=***REMOVED*** onResponse
     * @protected
     */
    PersistentConnection.prototype.sendRequest = function (action, body, onResponse) ***REMOVED***
        var curReqNum = ++this.requestNumber_;
        var msg = ***REMOVED*** r: curReqNum, a: action, b: body ***REMOVED***;
        this.log_(util_2.stringify(msg));
        util_3.assert(this.connected_, "sendRequest call when we're not connected not allowed.");
        this.realtime_.sendRequest(msg);
        if (onResponse) ***REMOVED***
            this.requestCBHash_[curReqNum] = onResponse;
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.listen = function (query, currentHashFn, tag, onComplete) ***REMOVED***
        var queryId = query.queryIdentifier();
        var pathString = query.path.toString();
        this.log_('Listen called for ' + pathString + ' ' + queryId);
        this.listens_[pathString] = this.listens_[pathString] || ***REMOVED******REMOVED***;
        util_3.assert(query.getQueryParams().isDefault() ||
            !query.getQueryParams().loadsAllData(), 'listen() called for non-default but complete query');
        util_3.assert(!this.listens_[pathString][queryId], 'listen() called twice for same path/queryId.');
        var listenSpec = ***REMOVED***
            onComplete: onComplete,
            hashFn: currentHashFn,
            query: query,
            tag: tag
        ***REMOVED***;
        this.listens_[pathString][queryId] = listenSpec;
        if (this.connected_) ***REMOVED***
            this.sendListen_(listenSpec);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @param ***REMOVED***!***REMOVED***onComplete(),
     *           hashFn():!string,
     *           query: !Query,
     *           tag: ?number***REMOVED******REMOVED*** listenSpec
     * @private
     */
    PersistentConnection.prototype.sendListen_ = function (listenSpec) ***REMOVED***
        var _this = this;
        var query = listenSpec.query;
        var pathString = query.path.toString();
        var queryId = query.queryIdentifier();
        this.log_('Listen on ' + pathString + ' for ' + queryId);
        var req = ***REMOVED*** /*path*/ p: pathString ***REMOVED***;
        var action = 'q';
        // Only bother to send query if it's non-default.
        if (listenSpec.tag) ***REMOVED***
            req['q'] = query.queryObject();
            req['t'] = listenSpec.tag;
        ***REMOVED***
        req['h'] = listenSpec.hashFn();
        this.sendRequest(action, req, function (message) ***REMOVED***
            var payload = message['d'];
            var status = message['s'];
            // print warnings in any case...
            PersistentConnection.warnOnListenWarnings_(payload, query);
            var currentListenSpec = _this.listens_[pathString] && _this.listens_[pathString][queryId];
            // only trigger actions if the listen hasn't been removed and readded
            if (currentListenSpec === listenSpec) ***REMOVED***
                _this.log_('listen response', message);
                if (status !== 'ok') ***REMOVED***
                    _this.removeListen_(pathString, queryId);
                ***REMOVED***
                if (listenSpec.onComplete) ***REMOVED***
                    listenSpec.onComplete(status, payload);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    /**
     * @param ***REMOVED*******REMOVED*** payload
     * @param ***REMOVED***!Query***REMOVED*** query
     * @private
     */
    PersistentConnection.warnOnListenWarnings_ = function (payload, query) ***REMOVED***
        if (payload && typeof payload === 'object' && util_1.contains(payload, 'w')) ***REMOVED***
            var warnings = util_1.safeGet(payload, 'w');
            if (Array.isArray(warnings) && ~warnings.indexOf('no_index')) ***REMOVED***
                var indexSpec = '".indexOn": "' +
                    query
                        .getQueryParams()
                        .getIndex()
                        .toString() +
                    '"';
                var indexPath = query.path.toString();
                util_4.warn("Using an unspecified index. Your data will be downloaded and " +
                    ("filtered on the client. Consider adding " + indexSpec + " at ") +
                    (indexPath + " to your security rules for better performance."));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.refreshAuthToken = function (token) ***REMOVED***
        this.authToken_ = token;
        this.log_('Auth token refreshed');
        if (this.authToken_) ***REMOVED***
            this.tryAuth();
        ***REMOVED***
        else ***REMOVED***
            //If we're connected we want to let the server know to unauthenticate us. If we're not connected, simply delete
            //the credential so we dont become authenticated next time we connect.
            if (this.connected_) ***REMOVED***
                this.sendRequest('unauth', ***REMOVED******REMOVED***, function () ***REMOVED*** ***REMOVED***);
            ***REMOVED***
        ***REMOVED***
        this.reduceReconnectDelayIfAdminCredential_(token);
    ***REMOVED***;
    /**
     * @param ***REMOVED***!string***REMOVED*** credential
     * @private
     */
    PersistentConnection.prototype.reduceReconnectDelayIfAdminCredential_ = function (credential) ***REMOVED***
        // NOTE: This isn't intended to be bulletproof (a malicious developer can always just modify the client).
        // Additionally, we don't bother resetting the max delay back to the default if auth fails / expires.
        var isFirebaseSecret = credential && credential.length === 40;
        if (isFirebaseSecret || util_5.isAdmin(credential)) ***REMOVED***
            this.log_('Admin auth credential detected.  Reducing max reconnect time.');
            this.maxReconnectDelay_ = RECONNECT_MAX_DELAY_FOR_ADMINS;
        ***REMOVED***
    ***REMOVED***;
    /**
     * Attempts to authenticate with the given credentials. If the authentication attempt fails, it's triggered like
     * a auth revoked (the connection is closed).
     */
    PersistentConnection.prototype.tryAuth = function () ***REMOVED***
        var _this = this;
        if (this.connected_ && this.authToken_) ***REMOVED***
            var token_1 = this.authToken_;
            var authMethod = util_5.isValidFormat(token_1) ? 'auth' : 'gauth';
            var requestData = ***REMOVED*** cred: token_1 ***REMOVED***;
            if (this.authOverride_ === null) ***REMOVED***
                requestData['noauth'] = true;
            ***REMOVED***
            else if (typeof this.authOverride_ === 'object') ***REMOVED***
                requestData['authvar'] = this.authOverride_;
            ***REMOVED***
            this.sendRequest(authMethod, requestData, function (res) ***REMOVED***
                var status = res['s'];
                var data = res['d'] || 'error';
                if (_this.authToken_ === token_1) ***REMOVED***
                    if (status === 'ok') ***REMOVED***
                        _this.invalidAuthTokenCount_ = 0;
                    ***REMOVED***
                    else ***REMOVED***
                        // Triggers reconnect and force refresh for auth token
                        _this.onAuthRevoked_(status, data);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.unlisten = function (query, tag) ***REMOVED***
        var pathString = query.path.toString();
        var queryId = query.queryIdentifier();
        this.log_('Unlisten called for ' + pathString + ' ' + queryId);
        util_3.assert(query.getQueryParams().isDefault() ||
            !query.getQueryParams().loadsAllData(), 'unlisten() called for non-default but complete query');
        var listen = this.removeListen_(pathString, queryId);
        if (listen && this.connected_) ***REMOVED***
            this.sendUnlisten_(pathString, queryId, query.queryObject(), tag);
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.sendUnlisten_ = function (pathString, queryId, queryObj, tag) ***REMOVED***
        this.log_('Unlisten on ' + pathString + ' for ' + queryId);
        var req = ***REMOVED*** /*path*/ p: pathString ***REMOVED***;
        var action = 'n';
        // Only bother sending queryId if it's non-default.
        if (tag) ***REMOVED***
            req['q'] = queryObj;
            req['t'] = tag;
        ***REMOVED***
        this.sendRequest(action, req);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.onDisconnectPut = function (pathString, data, onComplete) ***REMOVED***
        if (this.connected_) ***REMOVED***
            this.sendOnDisconnect_('o', pathString, data, onComplete);
        ***REMOVED***
        else ***REMOVED***
            this.onDisconnectRequestQueue_.push(***REMOVED***
                pathString: pathString,
                action: 'o',
                data: data,
                onComplete: onComplete
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.onDisconnectMerge = function (pathString, data, onComplete) ***REMOVED***
        if (this.connected_) ***REMOVED***
            this.sendOnDisconnect_('om', pathString, data, onComplete);
        ***REMOVED***
        else ***REMOVED***
            this.onDisconnectRequestQueue_.push(***REMOVED***
                pathString: pathString,
                action: 'om',
                data: data,
                onComplete: onComplete
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.onDisconnectCancel = function (pathString, onComplete) ***REMOVED***
        if (this.connected_) ***REMOVED***
            this.sendOnDisconnect_('oc', pathString, null, onComplete);
        ***REMOVED***
        else ***REMOVED***
            this.onDisconnectRequestQueue_.push(***REMOVED***
                pathString: pathString,
                action: 'oc',
                data: null,
                onComplete: onComplete
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.sendOnDisconnect_ = function (action, pathString, data, onComplete) ***REMOVED***
        var request = ***REMOVED*** /*path*/ p: pathString, /*data*/ d: data ***REMOVED***;
        this.log_('onDisconnect ' + action, request);
        this.sendRequest(action, request, function (response) ***REMOVED***
            if (onComplete) ***REMOVED***
                setTimeout(function () ***REMOVED***
                    onComplete(response['s'], response['d']);
                ***REMOVED***, Math.floor(0));
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.put = function (pathString, data, onComplete, hash) ***REMOVED***
        this.putInternal('p', pathString, data, onComplete, hash);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.merge = function (pathString, data, onComplete, hash) ***REMOVED***
        this.putInternal('m', pathString, data, onComplete, hash);
    ***REMOVED***;
    PersistentConnection.prototype.putInternal = function (action, pathString, data, onComplete, hash) ***REMOVED***
        var request = ***REMOVED***
            /*path*/ p: pathString,
            /*data*/ d: data
        ***REMOVED***;
        if (hash !== undefined)
            request['h'] = hash;
        // TODO: Only keep track of the most recent put for a given path?
        this.outstandingPuts_.push(***REMOVED***
            action: action,
            request: request,
            onComplete: onComplete
        ***REMOVED***);
        this.outstandingPutCount_++;
        var index = this.outstandingPuts_.length - 1;
        if (this.connected_) ***REMOVED***
            this.sendPut_(index);
        ***REMOVED***
        else ***REMOVED***
            this.log_('Buffering put: ' + pathString);
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.sendPut_ = function (index) ***REMOVED***
        var _this = this;
        var action = this.outstandingPuts_[index].action;
        var request = this.outstandingPuts_[index].request;
        var onComplete = this.outstandingPuts_[index].onComplete;
        this.outstandingPuts_[index].queued = this.connected_;
        this.sendRequest(action, request, function (message) ***REMOVED***
            _this.log_(action + ' response', message);
            delete _this.outstandingPuts_[index];
            _this.outstandingPutCount_--;
            // Clean up array occasionally.
            if (_this.outstandingPutCount_ === 0) ***REMOVED***
                _this.outstandingPuts_ = [];
            ***REMOVED***
            if (onComplete)
                onComplete(message['s'], message['d']);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PersistentConnection.prototype.reportStats = function (stats) ***REMOVED***
        var _this = this;
        // If we're not connected, we just drop the stats.
        if (this.connected_) ***REMOVED***
            var request = ***REMOVED*** /*counters*/ c: stats ***REMOVED***;
            this.log_('reportStats', request);
            this.sendRequest(/*stats*/ 's', request, function (result) ***REMOVED***
                var status = result['s'];
                if (status !== 'ok') ***REMOVED***
                    var errorReason = result['d'];
                    _this.log_('reportStats', 'Error sending stats: ' + errorReason);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @param ***REMOVED*******REMOVED*** message
     * @private
     */
    PersistentConnection.prototype.onDataMessage_ = function (message) ***REMOVED***
        if ('r' in message) ***REMOVED***
            // this is a response
            this.log_('from server: ' + util_2.stringify(message));
            var reqNum = message['r'];
            var onResponse = this.requestCBHash_[reqNum];
            if (onResponse) ***REMOVED***
                delete this.requestCBHash_[reqNum];
                onResponse(message['b']);
            ***REMOVED***
        ***REMOVED***
        else if ('error' in message) ***REMOVED***
            throw 'A server-side error has occurred: ' + message['error'];
        ***REMOVED***
        else if ('a' in message) ***REMOVED***
            // a and b are action and body, respectively
            this.onDataPush_(message['a'], message['b']);
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.onDataPush_ = function (action, body) ***REMOVED***
        this.log_('handleServerMessage', action, body);
        if (action === 'd')
            this.onDataUpdate_(body['p'], body['d'], 
            /*isMerge*/ false, body['t']);
        else if (action === 'm')
            this.onDataUpdate_(body['p'], body['d'], 
            /*isMerge=*/ true, body['t']);
        else if (action === 'c')
            this.onListenRevoked_(body['p'], body['q']);
        else if (action === 'ac')
            this.onAuthRevoked_(body['s'], body['d']);
        else if (action === 'sd')
            this.onSecurityDebugPacket_(body);
        else
            util_4.error('Unrecognized action received from server: ' +
                util_2.stringify(action) +
                '\nAre you using the latest client?');
    ***REMOVED***;
    PersistentConnection.prototype.onReady_ = function (timestamp, sessionId) ***REMOVED***
        this.log_('connection ready');
        this.connected_ = true;
        this.lastConnectionEstablishedTime_ = new Date().getTime();
        this.handleTimestamp_(timestamp);
        this.lastSessionId = sessionId;
        if (this.firstConnection_) ***REMOVED***
            this.sendConnectStats_();
        ***REMOVED***
        this.restoreState_();
        this.firstConnection_ = false;
        this.onConnectStatus_(true);
    ***REMOVED***;
    PersistentConnection.prototype.scheduleConnect_ = function (timeout) ***REMOVED***
        var _this = this;
        util_3.assert(!this.realtime_, "Scheduling a connect when we're already connected/ing?");
        if (this.establishConnectionTimer_) ***REMOVED***
            clearTimeout(this.establishConnectionTimer_);
        ***REMOVED***
        // NOTE: Even when timeout is 0, it's important to do a setTimeout to work around an infuriating "Security Error" in
        // Firefox when trying to write to our long-polling iframe in some scenarios (e.g. Forge or our unit tests).
        this.establishConnectionTimer_ = setTimeout(function () ***REMOVED***
            _this.establishConnectionTimer_ = null;
            _this.establishConnection_();
        ***REMOVED***, Math.floor(timeout));
    ***REMOVED***;
    /**
     * @param ***REMOVED***boolean***REMOVED*** visible
     * @private
     */
    PersistentConnection.prototype.onVisible_ = function (visible) ***REMOVED***
        // NOTE: Tabbing away and back to a window will defeat our reconnect backoff, but I think that's fine.
        if (visible &&
            !this.visible_ &&
            this.reconnectDelay_ === this.maxReconnectDelay_) ***REMOVED***
            this.log_('Window became visible.  Reducing delay.');
            this.reconnectDelay_ = RECONNECT_MIN_DELAY;
            if (!this.realtime_) ***REMOVED***
                this.scheduleConnect_(0);
            ***REMOVED***
        ***REMOVED***
        this.visible_ = visible;
    ***REMOVED***;
    PersistentConnection.prototype.onOnline_ = function (online) ***REMOVED***
        if (online) ***REMOVED***
            this.log_('Browser went online.');
            this.reconnectDelay_ = RECONNECT_MIN_DELAY;
            if (!this.realtime_) ***REMOVED***
                this.scheduleConnect_(0);
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            this.log_('Browser went offline.  Killing connection.');
            if (this.realtime_) ***REMOVED***
                this.realtime_.close();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.onRealtimeDisconnect_ = function () ***REMOVED***
        this.log_('data client disconnected');
        this.connected_ = false;
        this.realtime_ = null;
        // Since we don't know if our sent transactions succeeded or not, we need to cancel them.
        this.cancelSentTransactions_();
        // Clear out the pending requests.
        this.requestCBHash_ = ***REMOVED******REMOVED***;
        if (this.shouldReconnect_()) ***REMOVED***
            if (!this.visible_) ***REMOVED***
                this.log_("Window isn't visible.  Delaying reconnect.");
                this.reconnectDelay_ = this.maxReconnectDelay_;
                this.lastConnectionAttemptTime_ = new Date().getTime();
            ***REMOVED***
            else if (this.lastConnectionEstablishedTime_) ***REMOVED***
                // If we've been connected long enough, reset reconnect delay to minimum.
                var timeSinceLastConnectSucceeded = new Date().getTime() - this.lastConnectionEstablishedTime_;
                if (timeSinceLastConnectSucceeded > RECONNECT_DELAY_RESET_TIMEOUT)
                    this.reconnectDelay_ = RECONNECT_MIN_DELAY;
                this.lastConnectionEstablishedTime_ = null;
            ***REMOVED***
            var timeSinceLastConnectAttempt = new Date().getTime() - this.lastConnectionAttemptTime_;
            var reconnectDelay = Math.max(0, this.reconnectDelay_ - timeSinceLastConnectAttempt);
            reconnectDelay = Math.random() * reconnectDelay;
            this.log_('Trying to reconnect in ' + reconnectDelay + 'ms');
            this.scheduleConnect_(reconnectDelay);
            // Adjust reconnect delay for next time.
            this.reconnectDelay_ = Math.min(this.maxReconnectDelay_, this.reconnectDelay_ * RECONNECT_DELAY_MULTIPLIER);
        ***REMOVED***
        this.onConnectStatus_(false);
    ***REMOVED***;
    PersistentConnection.prototype.establishConnection_ = function () ***REMOVED***
        if (this.shouldReconnect_()) ***REMOVED***
            this.log_('Making a connection attempt');
            this.lastConnectionAttemptTime_ = new Date().getTime();
            this.lastConnectionEstablishedTime_ = null;
            var onDataMessage_1 = this.onDataMessage_.bind(this);
            var onReady_1 = this.onReady_.bind(this);
            var onDisconnect_1 = this.onRealtimeDisconnect_.bind(this);
            var connId_1 = this.id + ':' + PersistentConnection.nextConnectionId_++;
            var self_1 = this;
            var lastSessionId_1 = this.lastSessionId;
            var canceled_1 = false;
            var connection_1 = null;
            var closeFn_1 = function () ***REMOVED***
                if (connection_1) ***REMOVED***
                    connection_1.close();
                ***REMOVED***
                else ***REMOVED***
                    canceled_1 = true;
                    onDisconnect_1();
                ***REMOVED***
            ***REMOVED***;
            var sendRequestFn = function (msg) ***REMOVED***
                util_3.assert(connection_1, "sendRequest call when we're not connected not allowed.");
                connection_1.sendRequest(msg);
            ***REMOVED***;
            this.realtime_ = ***REMOVED***
                close: closeFn_1,
                sendRequest: sendRequestFn
            ***REMOVED***;
            var forceRefresh = this.forceTokenRefresh_;
            this.forceTokenRefresh_ = false;
            // First fetch auth token, and establish connection after fetching the token was successful
            this.authTokenProvider_
                .getToken(forceRefresh)
                .then(function (result) ***REMOVED***
                if (!canceled_1) ***REMOVED***
                    util_4.log('getToken() completed. Creating connection.');
                    self_1.authToken_ = result && result.accessToken;
                    connection_1 = new Connection_1.Connection(connId_1, self_1.repoInfo_, onDataMessage_1, onReady_1, onDisconnect_1, 
                    /* onKill= */ function (reason) ***REMOVED***
                        util_4.warn(reason + ' (' + self_1.repoInfo_.toString() + ')');
                        self_1.interrupt(SERVER_KILL_INTERRUPT_REASON);
                    ***REMOVED***, lastSessionId_1);
                ***REMOVED***
                else ***REMOVED***
                    util_4.log('getToken() completed but was canceled');
                ***REMOVED***
            ***REMOVED***)
                .then(null, function (error) ***REMOVED***
                self_1.log_('Failed to get token: ' + error);
                if (!canceled_1) ***REMOVED***
                    if (util_6.CONSTANTS.NODE_ADMIN) ***REMOVED***
                        // This may be a critical error for the Admin Node.js SDK, so log a warning.
                        // But getToken() may also just have temporarily failed, so we still want to
                        // continue retrying.
                        util_4.warn(error);
                    ***REMOVED***
                    closeFn_1();
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** reason
     */
    PersistentConnection.prototype.interrupt = function (reason) ***REMOVED***
        util_4.log('Interrupting connection for reason: ' + reason);
        this.interruptReasons_[reason] = true;
        if (this.realtime_) ***REMOVED***
            this.realtime_.close();
        ***REMOVED***
        else ***REMOVED***
            if (this.establishConnectionTimer_) ***REMOVED***
                clearTimeout(this.establishConnectionTimer_);
                this.establishConnectionTimer_ = null;
            ***REMOVED***
            if (this.connected_) ***REMOVED***
                this.onRealtimeDisconnect_();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** reason
     */
    PersistentConnection.prototype.resume = function (reason) ***REMOVED***
        util_4.log('Resuming connection for reason: ' + reason);
        delete this.interruptReasons_[reason];
        if (util_1.isEmpty(this.interruptReasons_)) ***REMOVED***
            this.reconnectDelay_ = RECONNECT_MIN_DELAY;
            if (!this.realtime_) ***REMOVED***
                this.scheduleConnect_(0);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.handleTimestamp_ = function (timestamp) ***REMOVED***
        var delta = timestamp - new Date().getTime();
        this.onServerInfoUpdate_(***REMOVED*** serverTimeOffset: delta ***REMOVED***);
    ***REMOVED***;
    PersistentConnection.prototype.cancelSentTransactions_ = function () ***REMOVED***
        for (var i = 0; i < this.outstandingPuts_.length; i++) ***REMOVED***
            var put = this.outstandingPuts_[i];
            if (put && /*hash*/ 'h' in put.request && put.queued) ***REMOVED***
                if (put.onComplete)
                    put.onComplete('disconnect');
                delete this.outstandingPuts_[i];
                this.outstandingPutCount_--;
            ***REMOVED***
        ***REMOVED***
        // Clean up array occasionally.
        if (this.outstandingPutCount_ === 0)
            this.outstandingPuts_ = [];
    ***REMOVED***;
    /**
     * @param ***REMOVED***!string***REMOVED*** pathString
     * @param ***REMOVED***Array.<*>=***REMOVED*** query
     * @private
     */
    PersistentConnection.prototype.onListenRevoked_ = function (pathString, query) ***REMOVED***
        // Remove the listen and manufacture a "permission_denied" error for the failed listen.
        var queryId;
        if (!query) ***REMOVED***
            queryId = 'default';
        ***REMOVED***
        else ***REMOVED***
            queryId = query.map(function (q) ***REMOVED*** return util_4.ObjectToUniqueKey(q); ***REMOVED***).join('$');
        ***REMOVED***
        var listen = this.removeListen_(pathString, queryId);
        if (listen && listen.onComplete)
            listen.onComplete('permission_denied');
    ***REMOVED***;
    /**
     * @param ***REMOVED***!string***REMOVED*** pathString
     * @param ***REMOVED***!string***REMOVED*** queryId
     * @return ***REMOVED******REMOVED***queries:Array.<Query>, onComplete:function(string)***REMOVED******REMOVED***
     * @private
     */
    PersistentConnection.prototype.removeListen_ = function (pathString, queryId) ***REMOVED***
        var normalizedPathString = new Path_1.Path(pathString).toString(); // normalize path.
        var listen;
        if (this.listens_[normalizedPathString] !== undefined) ***REMOVED***
            listen = this.listens_[normalizedPathString][queryId];
            delete this.listens_[normalizedPathString][queryId];
            if (util_1.getCount(this.listens_[normalizedPathString]) === 0) ***REMOVED***
                delete this.listens_[normalizedPathString];
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            // all listens for this path has already been removed
            listen = undefined;
        ***REMOVED***
        return listen;
    ***REMOVED***;
    PersistentConnection.prototype.onAuthRevoked_ = function (statusCode, explanation) ***REMOVED***
        util_4.log('Auth token revoked: ' + statusCode + '/' + explanation);
        this.authToken_ = null;
        this.forceTokenRefresh_ = true;
        this.realtime_.close();
        if (statusCode === 'invalid_token' || statusCode === 'permission_denied') ***REMOVED***
            // We'll wait a couple times before logging the warning / increasing the
            // retry period since oauth tokens will report as "invalid" if they're
            // just expired. Plus there may be transient issues that resolve themselves.
            this.invalidAuthTokenCount_++;
            if (this.invalidAuthTokenCount_ >= INVALID_AUTH_TOKEN_THRESHOLD) ***REMOVED***
                // Set a long reconnect delay because recovery is unlikely
                this.reconnectDelay_ = RECONNECT_MAX_DELAY_FOR_ADMINS;
                // Notify the auth token provider that the token is invalid, which will log
                // a warning
                this.authTokenProvider_.notifyForInvalidToken();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.onSecurityDebugPacket_ = function (body) ***REMOVED***
        if (this.securityDebugCallback_) ***REMOVED***
            this.securityDebugCallback_(body);
        ***REMOVED***
        else ***REMOVED***
            if ('msg' in body && typeof console !== 'undefined') ***REMOVED***
                console.log('FIREBASE: ' + body['msg'].replace('\n', '\nFIREBASE: '));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    PersistentConnection.prototype.restoreState_ = function () ***REMOVED***
        var _this = this;
        //Re-authenticate ourselves if we have a credential stored.
        this.tryAuth();
        // Puts depend on having received the corresponding data update from the server before they complete, so we must
        // make sure to send listens before puts.
        util_1.forEach(this.listens_, function (pathString, queries) ***REMOVED***
            util_1.forEach(queries, function (key, listenSpec) ***REMOVED***
                _this.sendListen_(listenSpec);
            ***REMOVED***);
        ***REMOVED***);
        for (var i = 0; i < this.outstandingPuts_.length; i++) ***REMOVED***
            if (this.outstandingPuts_[i])
                this.sendPut_(i);
        ***REMOVED***
        while (this.onDisconnectRequestQueue_.length) ***REMOVED***
            var request = this.onDisconnectRequestQueue_.shift();
            this.sendOnDisconnect_(request.action, request.pathString, request.data, request.onComplete);
        ***REMOVED***
    ***REMOVED***;
    /**
     * Sends client stats for first connection
     * @private
     */
    PersistentConnection.prototype.sendConnectStats_ = function () ***REMOVED***
        var stats = ***REMOVED******REMOVED***;
        var clientName = 'js';
        if (util_6.CONSTANTS.NODE_ADMIN) ***REMOVED***
            clientName = 'admin_node';
        ***REMOVED***
        else if (util_6.CONSTANTS.NODE_CLIENT) ***REMOVED***
            clientName = 'node';
        ***REMOVED***
        stats['sdk.' + clientName + '.' + app_1.default.SDK_VERSION.replace(/\./g, '-')] = 1;
        if (util_7.isMobileCordova()) ***REMOVED***
            stats['framework.cordova'] = 1;
        ***REMOVED***
        else if (util_7.isReactNative()) ***REMOVED***
            stats['framework.reactnative'] = 1;
        ***REMOVED***
        this.reportStats(stats);
    ***REMOVED***;
    /**
     * @return ***REMOVED***boolean***REMOVED***
     * @private
     */
    PersistentConnection.prototype.shouldReconnect_ = function () ***REMOVED***
        var online = OnlineMonitor_1.OnlineMonitor.getInstance().currentlyOnline();
        return util_1.isEmpty(this.interruptReasons_) && online;
    ***REMOVED***;
    /**
     * @private
     */
    PersistentConnection.nextPersistentConnectionId_ = 0;
    /**
     * Counter for number of connections created. Mainly used for tagging in the logs
     * @type ***REMOVED***number***REMOVED***
     * @private
     */
    PersistentConnection.nextConnectionId_ = 0;
    return PersistentConnection;
***REMOVED***(ServerActions_1.ServerActions));
exports.PersistentConnection = PersistentConnection;
