// har-validator/src/async.js
import * as schemas from 'har-schema'
import Ajv from 'ajv'
import HARError from './error'

let ajv

export function validate (name, data = ***REMOVED******REMOVED***, next) ***REMOVED***
  // validator config
  ajv = ajv || new Ajv(***REMOVED***
    allErrors: true,
    schemas: schemas
  ***REMOVED***)

  let validate = ajv.getSchema(name + '.json')

  let valid = validate(data)

  // callback?
  if (typeof next === 'function') ***REMOVED***
    return next(!valid ? new HARError(validate.errors) : null, valid)
  ***REMOVED***

  return valid
***REMOVED***

export function afterRequest (data, next) ***REMOVED***
  return validate('afterRequest', data, next)
***REMOVED***

export function beforeRequest (data, next) ***REMOVED***
  return validate('beforeRequest', data, next)
***REMOVED***

export function browser (data, next) ***REMOVED***
  return validate('browser', data, next)
***REMOVED***

export function cache (data, next) ***REMOVED***
  return validate('cache', data, next)
***REMOVED***

export function content (data, next) ***REMOVED***
  return validate('content', data, next)
***REMOVED***

export function cookie (data, next) ***REMOVED***
  return validate('cookie', data, next)
***REMOVED***

export function creator (data, next) ***REMOVED***
  return validate('creator', data, next)
***REMOVED***

export function entry (data, next) ***REMOVED***
  return validate('entry', data, next)
***REMOVED***

export function har (data, next) ***REMOVED***
  return validate('har', data, next)
***REMOVED***

export function header (data, next) ***REMOVED***
  return validate('header', data, next)
***REMOVED***

export function log (data, next) ***REMOVED***
  return validate('log', data, next)
***REMOVED***

export function page (data, next) ***REMOVED***
  return validate('page', data, next)
***REMOVED***

export function pageTimings (data, next) ***REMOVED***
  return validate('pageTimings', data, next)
***REMOVED***

export function postData (data, next) ***REMOVED***
  return validate('postData', data, next)
***REMOVED***

export function query (data, next) ***REMOVED***
  return validate('query', data, next)
***REMOVED***

export function request (data, next) ***REMOVED***
  return validate('request', data, next)
***REMOVED***

export function response (data, next) ***REMOVED***
  return validate('response', data, next)
***REMOVED***

export function timings (data, next) ***REMOVED***
  return validate('timings', data, next)
***REMOVED***
