// es6-map/polyfill.js
'use strict';

var clear          = require('es5-ext/array/#/clear')
  , eIndexOf       = require('es5-ext/array/#/e-index-of')
  , setPrototypeOf = require('es5-ext/object/set-prototype-of')
  , callable       = require('es5-ext/object/valid-callable')
  , validValue     = require('es5-ext/object/valid-value')
  , d              = require('d')
  , ee             = require('event-emitter')
  , Symbol         = require('es6-symbol')
  , iterator       = require('es6-iterator/valid-iterable')
  , forOf          = require('es6-iterator/for-of')
  , Iterator       = require('./lib/iterator')
  , isNative       = require('./is-native-implemented')

  , call = Function.prototype.call
  , defineProperties = Object.defineProperties, getPrototypeOf = Object.getPrototypeOf
  , MapPoly;

module.exports = MapPoly = function (/*iterable*/) ***REMOVED***
	var iterable = arguments[0], keys, values, self;
	if (!(this instanceof MapPoly)) throw new TypeError('Constructor requires \'new\'');
	if (isNative && setPrototypeOf && (Map !== MapPoly)) ***REMOVED***
		self = setPrototypeOf(new Map(), getPrototypeOf(this));
	***REMOVED*** else ***REMOVED***
		self = this;
	***REMOVED***
	if (iterable != null) iterator(iterable);
	defineProperties(self, ***REMOVED***
		__mapKeysData__: d('c', keys = []),
		__mapValuesData__: d('c', values = [])
	***REMOVED***);
	if (!iterable) return self;
	forOf(iterable, function (value) ***REMOVED***
		var key = validValue(value)[0];
		value = value[1];
		if (eIndexOf.call(keys, key) !== -1) return;
		keys.push(key);
		values.push(value);
	***REMOVED***, self);
	return self;
***REMOVED***;

if (isNative) ***REMOVED***
	if (setPrototypeOf) setPrototypeOf(MapPoly, Map);
	MapPoly.prototype = Object.create(Map.prototype, ***REMOVED***
		constructor: d(MapPoly)
	***REMOVED***);
***REMOVED***

ee(defineProperties(MapPoly.prototype, ***REMOVED***
	clear: d(function () ***REMOVED***
		if (!this.__mapKeysData__.length) return;
		clear.call(this.__mapKeysData__);
		clear.call(this.__mapValuesData__);
		this.emit('_clear');
	***REMOVED***),
	delete: d(function (key) ***REMOVED***
		var index = eIndexOf.call(this.__mapKeysData__, key);
		if (index === -1) return false;
		this.__mapKeysData__.splice(index, 1);
		this.__mapValuesData__.splice(index, 1);
		this.emit('_delete', index, key);
		return true;
	***REMOVED***),
	entries: d(function () ***REMOVED*** return new Iterator(this, 'key+value'); ***REMOVED***),
	forEach: d(function (cb/*, thisArg*/) ***REMOVED***
		var thisArg = arguments[1], iterator, result;
		callable(cb);
		iterator = this.entries();
		result = iterator._next();
		while (result !== undefined) ***REMOVED***
			call.call(cb, thisArg, this.__mapValuesData__[result],
				this.__mapKeysData__[result], this);
			result = iterator._next();
		***REMOVED***
	***REMOVED***),
	get: d(function (key) ***REMOVED***
		var index = eIndexOf.call(this.__mapKeysData__, key);
		if (index === -1) return;
		return this.__mapValuesData__[index];
	***REMOVED***),
	has: d(function (key) ***REMOVED***
		return (eIndexOf.call(this.__mapKeysData__, key) !== -1);
	***REMOVED***),
	keys: d(function () ***REMOVED*** return new Iterator(this, 'key'); ***REMOVED***),
	set: d(function (key, value) ***REMOVED***
		var index = eIndexOf.call(this.__mapKeysData__, key), emit;
		if (index === -1) ***REMOVED***
			index = this.__mapKeysData__.push(key) - 1;
			emit = true;
		***REMOVED***
		this.__mapValuesData__[index] = value;
		if (emit) this.emit('_add', index, key);
		return this;
	***REMOVED***),
	size: d.gs(function () ***REMOVED*** return this.__mapKeysData__.length; ***REMOVED***),
	values: d(function () ***REMOVED*** return new Iterator(this, 'value'); ***REMOVED***),
	toString: d(function () ***REMOVED*** return '[object Map]'; ***REMOVED***)
***REMOVED***));
Object.defineProperty(MapPoly.prototype, Symbol.iterator, d(function () ***REMOVED***
	return this.entries();
***REMOVED***));
Object.defineProperty(MapPoly.prototype, Symbol.toStringTag, d('c', 'Map'));
