// ajv/lib/v5.js
'use strict';

var META_SCHEMA_ID = 'https://raw.githubusercontent.com/epoberezkin/ajv/master/lib/refs/json-schema-v5.json';

module.exports = ***REMOVED***
  enable: enableV5,
  META_SCHEMA_ID: META_SCHEMA_ID
***REMOVED***;


function enableV5(ajv) ***REMOVED***
  var inlineFunctions = ***REMOVED***
    'switch': require('./dotjs/switch'),
    'constant': require('./dotjs/constant'),
    '_formatLimit': require('./dotjs/_formatLimit'),
    'patternRequired': require('./dotjs/patternRequired')
  ***REMOVED***;

  if (ajv._opts.meta !== false) ***REMOVED***
    var metaSchema = require('./refs/json-schema-v5.json');
    ajv.addMetaSchema(metaSchema, META_SCHEMA_ID);
  ***REMOVED***
  _addKeyword('constant');
  ajv.addKeyword('contains', ***REMOVED*** type: 'array', macro: containsMacro ***REMOVED***);

  _addKeyword('formatMaximum', 'string', inlineFunctions._formatLimit);
  _addKeyword('formatMinimum', 'string', inlineFunctions._formatLimit);
  ajv.addKeyword('formatExclusiveMaximum');
  ajv.addKeyword('formatExclusiveMinimum');

  ajv.addKeyword('patternGroups'); // implemented in properties.jst
  _addKeyword('patternRequired', 'object');
  _addKeyword('switch');


  function _addKeyword(keyword, types, inlineFunc) ***REMOVED***
    var definition = ***REMOVED***
      inline: inlineFunc || inlineFunctions[keyword],
      statements: true,
      errors: 'full'
    ***REMOVED***;
    if (types) definition.type = types;
    ajv.addKeyword(keyword, definition);
  ***REMOVED***
***REMOVED***


function containsMacro(schema) ***REMOVED***
  return ***REMOVED***
    not: ***REMOVED*** items: ***REMOVED*** not: schema ***REMOVED*** ***REMOVED***
  ***REMOVED***;
***REMOVED***
