// @firebase/firestore/dist/esm/src/platform_node/grpc_connection.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as grpc from 'grpc';
import firebase from '@firebase/app';
var SDK_VERSION = firebase.SDK_VERSION;
var grpcVersion = require('grpc/package.json').version;
import ***REMOVED*** StreamBridge ***REMOVED*** from '../remote/stream_bridge';
import ***REMOVED*** mapCodeFromRpcCode ***REMOVED*** from '../remote/rpc_error';
import ***REMOVED*** assert ***REMOVED*** from '../util/assert';
import ***REMOVED*** FirestoreError ***REMOVED*** from '../util/error';
import * as log from '../util/log';
import ***REMOVED*** nodePromise ***REMOVED*** from '../util/node_api';
import ***REMOVED*** Deferred ***REMOVED*** from '../util/promise';
var LOG_TAG = 'Connection';
// TODO(b/38203344): The SDK_VERSION is set independently from Firebase because
// we are doing out-of-band releases. Once we release as part of Firebase, we
// should use the Firebase version instead.
var X_GOOG_API_CLIENT_VALUE = "gl-node/" + process.versions.node + " fire/" + SDK_VERSION + " grpc/" + grpcVersion;
function createHeaders(databaseInfo, token) ***REMOVED***
    assert(token === null || token.type === 'OAuth', 'If provided, token must be OAuth');
    var channelCredentials = databaseInfo.ssl
        ? grpc.credentials.createSsl()
        : grpc.credentials.createInsecure();
    var callCredentials = grpc.credentials.createFromMetadataGenerator(function (context, cb) ***REMOVED***
        var metadata = new grpc.Metadata();
        if (token) ***REMOVED***
            for (var header in token.authHeaders) ***REMOVED***
                if (token.authHeaders.hasOwnProperty(header)) ***REMOVED***
                    metadata.set(header, token.authHeaders[header]);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        metadata.set('x-goog-api-client', X_GOOG_API_CLIENT_VALUE);
        // This header is used to improve routing and project isolation by the
        // backend.
        metadata.set('google-cloud-resource-prefix', "projects/" + databaseInfo.databaseId.projectId + "/" +
            ("databases/" + databaseInfo.databaseId.database));
        cb(null, metadata);
    ***REMOVED***);
    return grpc.credentials.combineChannelCredentials(channelCredentials, callCredentials);
***REMOVED***
/**
 * A Connection implemented by GRPC-Node.
 */
var GrpcConnection = /** @class */ (function () ***REMOVED***
    function GrpcConnection(protos, databaseInfo) ***REMOVED***
        this.databaseInfo = databaseInfo;
        // We cache stubs for the most-recently-used token.
        this.cachedStub = null;
        this.firestore = protos['google']['firestore']['v1beta1'];
    ***REMOVED***
    GrpcConnection.prototype.sameToken = function (tokenA, tokenB) ***REMOVED***
        var valueA = tokenA && tokenA.authHeaders['Authorization'];
        var valueB = tokenB && tokenB.authHeaders['Authorization'];
        return valueA === valueB;
    ***REMOVED***;
    GrpcConnection.prototype.ensureActiveStub = function (token) ***REMOVED***
        if (!this.cachedStub || !this.sameToken(this.cachedStub.token, token)) ***REMOVED***
            log.debug(LOG_TAG, 'Creating Firestore stub.');
            var credentials = createHeaders(this.databaseInfo, token);
            this.cachedStub = ***REMOVED***
                stub: new this.firestore.Firestore(this.databaseInfo.host, credentials),
                token: token
            ***REMOVED***;
        ***REMOVED***
        return this.cachedStub.stub;
    ***REMOVED***;
    GrpcConnection.prototype.getRpcCallable = function (rpcName, token) ***REMOVED***
        // RPC Methods have the first character lower-cased
        // (e.g. Listen => listen(), BatchGetDocuments => batchGetDocuments()).
        var rpcMethod = rpcName.charAt(0).toLowerCase() + rpcName.slice(1);
        var stub = this.ensureActiveStub(token);
        var rpc = stub[rpcMethod];
        assert(rpc != null, 'Unknown RPC: ' + rpcName);
        return rpc.bind(stub);
    ***REMOVED***;
    GrpcConnection.prototype.invokeRPC = function (rpcName, request, token) ***REMOVED***
        var rpc = this.getRpcCallable(rpcName, token);
        return nodePromise(function (callback) ***REMOVED***
            log.debug(LOG_TAG, "RPC '" + rpcName + "' invoked with request:", request);
            return rpc(request, function (grpcError, value) ***REMOVED***
                if (grpcError) ***REMOVED***
                    log.debug(LOG_TAG, "RPC '" + rpcName + "' failed with error:", grpcError);
                    callback(new FirestoreError(mapCodeFromRpcCode(grpcError.code), grpcError.message));
                ***REMOVED***
                else ***REMOVED***
                    log.debug(LOG_TAG, "RPC '" + rpcName + "' completed with response:", value);
                    callback(undefined, value);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    GrpcConnection.prototype.invokeStreamingRPC = function (rpcName, request, token) ***REMOVED***
        var rpc = this.getRpcCallable(rpcName, token);
        var results = [];
        var responseDeferred = new Deferred();
        log.debug(LOG_TAG, "RPC '" + rpcName + "' invoked (streaming) with request:", request);
        var stream = rpc(request);
        stream.on('data', function (response) ***REMOVED***
            log.debug(LOG_TAG, "RPC " + rpcName + " received result:", response);
            results.push(response);
        ***REMOVED***);
        stream.on('end', function () ***REMOVED***
            log.debug(LOG_TAG, "RPC '" + rpcName + "' completed.");
            responseDeferred.resolve(results);
        ***REMOVED***);
        stream.on('error', function (grpcError) ***REMOVED***
            log.debug(LOG_TAG, "RPC '" + rpcName + "' failed with error:", grpcError);
            var code = mapCodeFromRpcCode(grpcError.code);
            responseDeferred.reject(new FirestoreError(code, grpcError.message));
        ***REMOVED***);
        return responseDeferred.promise;
    ***REMOVED***;
    // TODO(mikelehen): This "method" is a monster. Should be refactored.
    GrpcConnection.prototype.openStream = function (rpcName, token) ***REMOVED***
        var rpc = this.getRpcCallable(rpcName, token);
        var grpcStream = rpc();
        var closed = false;
        var close;
        var stream = new StreamBridge(***REMOVED***
            sendFn: function (msg) ***REMOVED***
                if (!closed) ***REMOVED***
                    log.debug(LOG_TAG, 'GRPC stream sending:', msg);
                    try ***REMOVED***
                        grpcStream.write(msg);
                    ***REMOVED***
                    catch (e) ***REMOVED***
                        // This probably means we didn't conform to the proto.  Make sure to
                        // log the message we sent.
                        log.error('Failure sending:', msg);
                        log.error('Error:', e);
                        throw e;
                    ***REMOVED***
                ***REMOVED***
                else ***REMOVED***
                    log.debug(LOG_TAG, 'Not sending because gRPC stream is closed:', msg);
                ***REMOVED***
            ***REMOVED***,
            closeFn: function () ***REMOVED***
                log.debug(LOG_TAG, 'GRPC stream closed locally via close().');
                close();
            ***REMOVED***
        ***REMOVED***);
        close = function (err) ***REMOVED***
            if (!closed) ***REMOVED***
                closed = true;
                stream.callOnClose(err);
                grpcStream.end();
            ***REMOVED***
        ***REMOVED***;
        grpcStream.on('data', function (msg) ***REMOVED***
            if (!closed) ***REMOVED***
                log.debug(LOG_TAG, 'GRPC stream received:', msg);
                stream.callOnMessage(msg);
            ***REMOVED***
        ***REMOVED***);
        grpcStream.on('end', function () ***REMOVED***
            log.debug(LOG_TAG, 'GRPC stream ended.');
            close();
        ***REMOVED***);
        grpcStream.on('finish', function () ***REMOVED***
            // TODO(mikelehen): I *believe* this assert is safe and we can just remove
            // the 'finish' event if we don't see the assert getting hit for a while.
            assert(closed, 'Received "finish" event without close() being called.');
        ***REMOVED***);
        grpcStream.on('error', function (grpcError) ***REMOVED***
            log.debug(LOG_TAG, 'GRPC stream error. Code:', grpcError.code, 'Message:', grpcError.message);
            var code = mapCodeFromRpcCode(grpcError.code);
            close(new FirestoreError(code, grpcError.message));
        ***REMOVED***);
        grpcStream.on('status', function (status) ***REMOVED***
            // TODO(mikelehen): I *believe* this assert is safe and we can just remove
            // the 'status' event if we don't see the assert getting hit for a while.
            assert(closed, "status event received before \"end\" or \"error\". " +
                ("code: " + status.code + " details: " + status.details));
        ***REMOVED***);
        log.debug(LOG_TAG, 'Opening GRPC stream');
        // TODO(dimond): Since grpc has no explicit open status (or does it?) we
        // simulate an onOpen in the next loop after the stream had it's listeners
        // registered
        setTimeout(function () ***REMOVED***
            stream.callOnOpen();
        ***REMOVED***, 0);
        return stream;
    ***REMOVED***;
    return GrpcConnection;
***REMOVED***());
export ***REMOVED*** GrpcConnection ***REMOVED***;
