// universal-deep-strict-equal/index.js
// port of https://github.com/nodejs/node/blob/v6.3.0/lib/assert.js#L145-L248

// http://wiki.commonjs.org/wiki/Unit_Testing/1.0
//
// Originally from narwhal.js (http://narwhaljs.org)
// Copyright (c) 2009 Thomas Robinson <280north.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the 'Software'), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'use strict';

var Buffer = require('buffer').Buffer;
var compare = Buffer.compare;
var indexOf = require('indexof');
var filter = require('array-filter');
var getPrototypeOf = Object.getPrototypeOf || function(obj) ***REMOVED***
  return obj.__proto__ || (
    obj.constructor
      ? obj.constructor.prototype
      : Object.prototype
  );
***REMOVED***;
function isEnumerable(obj, key) ***REMOVED***
  return Object.prototype.propertyIsEnumerable.call(obj, key);
***REMOVED***;
function pToString(obj) ***REMOVED***
  return Object.prototype.toString.call(obj);
***REMOVED***;
function isPrimitive(arg) ***REMOVED***
  return arg === null ||
         typeof arg === 'boolean' ||
         typeof arg === 'number' ||
         typeof arg === 'string' ||
         typeof arg === 'symbol' ||  // ES6 symbol
         typeof arg === 'undefined';
***REMOVED***
function isObject(arg) ***REMOVED***
  return typeof arg === 'object' && arg !== null;
***REMOVED***
function isDate(d) ***REMOVED***
  return isObject(d) && pToString(d) === '[object Date]';
***REMOVED***
function isRegExp(re) ***REMOVED***
  return isObject(re) && pToString(re) === '[object RegExp]';
***REMOVED***
var isArguments = (function () ***REMOVED***
  function isArg(obj) ***REMOVED***
    return isObject(obj) && pToString(obj) == '[object Arguments]';
  ***REMOVED***
  // dealing with old IEs (There's no Arguments type)
  if (!isArg(arguments)) ***REMOVED***
    return function(obj) ***REMOVED***
      return isObject(obj) &&
        typeof obj.length === 'number' &&
        obj.length >= 0 &&
        pToString(obj) !== '[object Array]' &&
        pToString(obj.callee) === '[object Function]';
    ***REMOVED***;
  ***REMOVED*** else ***REMOVED***
    return isArg;
  ***REMOVED***
***REMOVED***)();
function fromBufferSupport() ***REMOVED***
  try ***REMOVED***
    return typeof Buffer.from === 'function' && !!Buffer.from([0x62,0x75,0x66,0x66,0x65,0x72]);
  ***REMOVED*** catch (e) ***REMOVED***
    // Buffer.from under Node 4.x causes `TypeError: this is not a typed array.`
    return false;
  ***REMOVED***
***REMOVED***
var toBuffer = (function () ***REMOVED***
  // check whether Buffer constructor accepts ArrayBuffer or not
  function isBufferConstructorAcceptsArrayBuffer() ***REMOVED***
    try ***REMOVED***
      return typeof Uint8Array === 'function' && (new Buffer(new Uint8Array([1]).buffer)[0] === 1);
    ***REMOVED*** catch (e) ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***
  if (isBufferConstructorAcceptsArrayBuffer()) ***REMOVED***
    // Node 4.x
    return function (ab) ***REMOVED***
      return new Buffer(ab);
    ***REMOVED***;
  ***REMOVED*** else ***REMOVED***
    // Node 0.10.x and 0.12.x
    return function (ab) ***REMOVED***
      var buffer = new Buffer(ab.byteLength);
      var view = new Uint8Array(ab);
      for (var i = 0; i < buffer.length; ++i) ***REMOVED***
        buffer[i] = view[i];
      ***REMOVED***
      return buffer;
    ***REMOVED***;
  ***REMOVED***
***REMOVED***)();
var bufferFrom = fromBufferSupport() ? Buffer.from : toBuffer;
var objectKeys = (function () ***REMOVED***
  var OLD_V8_ARRAY_BUFFER_ENUM = ['BYTES_PER_ELEMENT','get','set','slice','subarray','buffer','length','byteOffset','byteLength'];
  var keys = Object.keys || require('object-keys');
  return function objectKeys(obj) ***REMOVED***
    // avoid iterating enumerable properties of ArrayBuffer under old V8
    if (isEnumerable(obj, 'buffer') &&
        isEnumerable(obj, 'byteOffset') &&
        isEnumerable(obj, 'byteLength')) ***REMOVED***
      return filter(keys(obj), function (k) ***REMOVED***
        return indexOf(OLD_V8_ARRAY_BUFFER_ENUM, k) === -1;
      ***REMOVED***);
    ***REMOVED*** else ***REMOVED***
      return keys(obj);
    ***REMOVED***
  ***REMOVED***;
***REMOVED***)();

function _deepEqual(actual, expected, strict, memos) ***REMOVED***
  // 7.1. All identical values are equivalent, as determined by ===.
  if (actual === expected) ***REMOVED***
    return true;
  ***REMOVED*** else if (actual instanceof Buffer && expected instanceof Buffer) ***REMOVED***
    return compare(actual, expected) === 0;

  // 7.2. If the expected value is a Date object, the actual value is
  // equivalent if it is also a Date object that refers to the same time.
  ***REMOVED*** else if (isDate(actual) && isDate(expected)) ***REMOVED***
    return actual.getTime() === expected.getTime();

  // 7.3 If the expected value is a RegExp object, the actual value is
  // equivalent if it is also a RegExp object with the same source and
  // properties (`global`, `multiline`, `lastIndex`, `ignoreCase`).
  ***REMOVED*** else if (isRegExp(actual) && isRegExp(expected)) ***REMOVED***
    return actual.source === expected.source &&
           actual.global === expected.global &&
           actual.multiline === expected.multiline &&
           actual.lastIndex === expected.lastIndex &&
           actual.ignoreCase === expected.ignoreCase;

  // 7.4. Other pairs that do not both pass typeof value == 'object',
  // equivalence is determined by ==.
  ***REMOVED*** else if ((actual === null || typeof actual !== 'object') &&
             (expected === null || typeof expected !== 'object')) ***REMOVED***
    return strict ? actual === expected : actual == expected;

  // If both values are instances of typed arrays, wrap their underlying
  // ArrayBuffers in a Buffer each to increase performance
  // This optimization requires the arrays to have the same type as checked by
  // Object.prototype.toString (aka pToString). Never perform binary
  // comparisons for Float*Arrays, though, since e.g. +0 === -0 but their
  // bit patterns are not identical.
  ***REMOVED*** else if (typeof ArrayBuffer === 'function' && typeof ArrayBuffer.isView === 'function' &&
             ArrayBuffer.isView(actual) && ArrayBuffer.isView(expected) &&
             pToString(actual) === pToString(expected) &&
             !(actual instanceof Float32Array ||
               actual instanceof Float64Array)) ***REMOVED***
    return compare(bufferFrom(actual.buffer),
                   bufferFrom(expected.buffer)) === 0;

  // 7.5 For all other Object pairs, including Array objects, equivalence is
  // determined by having the same number of owned properties (as verified
  // with Object.prototype.hasOwnProperty.call), the same set of keys
  // (although not necessarily the same order), equivalent values for every
  // corresponding key, and an identical 'prototype' property. Note: this
  // accounts for both named and indexed properties on Arrays.
  ***REMOVED*** else ***REMOVED***
    memos = memos || ***REMOVED***actual: [], expected: []***REMOVED***;

    var actualIndex = indexOf(memos.actual, actual);
    if (actualIndex !== -1) ***REMOVED***
      if (actualIndex === indexOf(memos.expected, expected)) ***REMOVED***
        return true;
      ***REMOVED***
    ***REMOVED***

    memos.actual.push(actual);
    memos.expected.push(expected);

    return objEquiv(actual, expected, strict, memos);
  ***REMOVED***
***REMOVED***

function objEquiv(a, b, strict, actualVisitedObjects) ***REMOVED***
  if (a === null || a === undefined || b === null || b === undefined)
    return false;
  // if one is a primitive, the other must be same
  if (isPrimitive(a) || isPrimitive(b))
    return a === b;
  if (strict && getPrototypeOf(a) !== getPrototypeOf(b))
    return false;
  var aIsArgs = isArguments(a),
      bIsArgs = isArguments(b);
  if ((aIsArgs && !bIsArgs) || (!aIsArgs && bIsArgs))
    return false;
  var ka = objectKeys(a),
      kb = objectKeys(b),
      key, i;
  // having the same number of owned properties (keys incorporates
  // hasOwnProperty)
  if (ka.length != kb.length)
    return false;
  //the same set of keys (although not necessarily the same order),
  ka.sort();
  kb.sort();
  //~~~cheap key test
  for (i = ka.length - 1; i >= 0; i--) ***REMOVED***
    if (ka[i] != kb[i])
      return false;
  ***REMOVED***
  //equivalent values for every corresponding key, and
  //~~~possibly expensive deep test
  for (i = ka.length - 1; i >= 0; i--) ***REMOVED***
    key = ka[i];
    if (!_deepEqual(a[key], b[key], strict, actualVisitedObjects))
      return false;
  ***REMOVED***
  return true;
***REMOVED***

module.exports = _deepEqual;
