// @firebase/firestore/dist/cjs/src/remote/remote_store.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var snapshot_version_1 = require("../core/snapshot_version");
var transaction_1 = require("../core/transaction");
var types_1 = require("../core/types");
var query_data_1 = require("../local/query_data");
var document_1 = require("../model/document");
var document_key_1 = require("../model/document_key");
var mutation_batch_1 = require("../model/mutation_batch");
var platform_1 = require("../platform/platform");
var assert_1 = require("../util/assert");
var error_1 = require("../util/error");
var log = require("../util/log");
var objUtils = require("../util/obj");
var remote_event_1 = require("./remote_event");
var rpc_error_1 = require("./rpc_error");
var watch_change_1 = require("./watch_change");
var LOG_TAG = 'RemoteStore';
// TODO(b/35853402): Negotiate this with the stream.
var MAX_PENDING_WRITES = 10;
// The RemoteStore notifies an onlineStateHandler with OnlineState.Failed if we
// fail to connect to the backend. This subsequently triggers get() requests to
// fail or use cached data, etc. Unfortunately, our connections have
// historically been subject to various transient failures. So we wait for
// multiple failures before notifying the onlineStateHandler.
var ONLINE_ATTEMPTS_BEFORE_FAILURE = 2;
/**
 * RemoteStore - An interface to remotely stored data, basically providing a
 * wrapper around the Datastore that is more reliable for the rest of the
 * system.
 *
 * RemoteStore is responsible for maintaining the connection to the server.
 * - maintaining a list of active listens.
 * - reconnecting when the connection is dropped.
 * - resuming all the active listens on reconnect.
 *
 * RemoteStore handles all incoming events from the Datastore.
 * - listening to the watch stream and repackaging the events as RemoteEvents
 * - notifying SyncEngine of any changes to the active listens.
 *
 * RemoteStore takes writes from other components and handles them reliably.
 * - pulling pending mutations from LocalStore and sending them to Datastore.
 * - retrying mutations that failed because of network problems.
 * - acking mutations to the SyncEngine once they are accepted or rejected.
 */
var RemoteStore = /** @class */ (function () ***REMOVED***
    function RemoteStore(
        /**
         * The local store, used to fill the write pipeline with outbound
         * mutations and resolve existence filter mismatches.
         */
        localStore, 
        /** The client-side proxy for interacting with the backend. */
        datastore, onlineStateHandler) ***REMOVED***
        this.localStore = localStore;
        this.datastore = datastore;
        this.onlineStateHandler = onlineStateHandler;
        this.pendingWrites = [];
        this.lastBatchSeen = mutation_batch_1.BATCHID_UNKNOWN;
        /**
         * A mapping of watched targets that the client cares about tracking and the
         * user has explicitly called a 'listen' for this target.
         *
         * These targets may or may not have been sent to or acknowledged by the
         * server. On re-establishing the listen stream, these targets should be sent
         * to the server. The targets removed with unlistens are removed eagerly
         * without waiting for confirmation from the listen stream.
         */
        this.listenTargets = ***REMOVED******REMOVED***;
        /**
         * A mapping of targetId to pending acks needed.
         *
         * If a targetId is present in this map, then we're waiting for watch to
         * acknowledge a removal or addition of the target. If a target is not in this
         * mapping, and it's in the listenTargets map, then we consider the target to
         * be active.
         *
         * We increment the count here every time we issue a request over the stream
         * to watch or unwatch. We then decrement the count every time we get a target
         * added or target removed message from the server. Once the count is equal to
         * 0 we know that the client and server are in the same state (once this state
         * is reached the targetId is removed from the map to free the memory).
         */
        this.pendingTargetResponses = ***REMOVED******REMOVED***;
        this.accumulatedWatchChanges = [];
        this.watchStream = null;
        this.writeStream = null;
        /**
         * The online state of the watch stream. The state is set to healthy if and
         * only if there are messages received by the backend.
         */
        this.watchStreamOnlineState = types_1.OnlineState.Unknown;
        /** A count of consecutive failures to open the stream. */
        this.watchStreamFailures = 0;
        /** Whether the client should fire offline warning. */
        this.shouldWarnOffline = true;
    ***REMOVED***
    /**
     * Starts up the remote store, creating streams, restoring state from
     * LocalStore, etc.
     */
    RemoteStore.prototype.start = function () ***REMOVED***
        return this.enableNetwork();
    ***REMOVED***;
    /**
     * Updates our OnlineState to the new state, updating local state
     * and notifying the onlineStateHandler as appropriate. Idempotent.
     */
    RemoteStore.prototype.updateOnlineState = function (newState) ***REMOVED***
        if (newState !== this.watchStreamOnlineState) ***REMOVED***
            if (newState === types_1.OnlineState.Healthy) ***REMOVED***
                // We've connected to watch at least once. Don't warn the developer about
                // being offline going forward.
                this.shouldWarnOffline = false;
            ***REMOVED***
            else if (newState === types_1.OnlineState.Unknown) ***REMOVED***
                // The state is set to unknown when a healthy stream is closed (e.g. due to
                // a token timeout) or when we have no active listens and therefore there's
                // no need to start the stream. Assuming there is (possibly in the future)
                // an active listen, then we will eventually move to state Online or Failed,
                // but we always want to make at least ONLINE_ATTEMPTS_BEFORE_FAILURE
                // attempts before failing, so we reset the count here.
                this.watchStreamFailures = 0;
            ***REMOVED***
            this.watchStreamOnlineState = newState;
            this.onlineStateHandler(newState);
        ***REMOVED***
    ***REMOVED***;
    /**
     * Updates our OnlineState as appropriate after the watch stream reports a
     * failure. The first failure moves us to the 'Unknown' state. We then may
     * allow multiple failures (based on ONLINE_ATTEMPTS_BEFORE_FAILURE) before we
     * actually transition to OnlineState.Failed.
     */
    RemoteStore.prototype.updateOnlineStateAfterFailure = function () ***REMOVED***
        if (this.watchStreamOnlineState === types_1.OnlineState.Healthy) ***REMOVED***
            this.updateOnlineState(types_1.OnlineState.Unknown);
        ***REMOVED***
        else ***REMOVED***
            this.watchStreamFailures++;
            if (this.watchStreamFailures >= ONLINE_ATTEMPTS_BEFORE_FAILURE) ***REMOVED***
                if (this.shouldWarnOffline) ***REMOVED***
                    log.error('Could not reach Firestore backend.');
                    this.shouldWarnOffline = false;
                ***REMOVED***
                this.updateOnlineState(types_1.OnlineState.Failed);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    RemoteStore.prototype.isNetworkEnabled = function () ***REMOVED***
        assert_1.assert((this.watchStream == null) === (this.writeStream == null), 'WatchStream and WriteStream should both be null or non-null');
        return this.watchStream != null;
    ***REMOVED***;
    /** Re-enables the network. Idempotent. */
    RemoteStore.prototype.enableNetwork = function () ***REMOVED***
        var _this = this;
        if (this.isNetworkEnabled()) ***REMOVED***
            return Promise.resolve();
        ***REMOVED***
        // Create new streams (but note they're not started yet).
        this.watchStream = this.datastore.newPersistentWatchStream();
        this.writeStream = this.datastore.newPersistentWriteStream();
        // Load any saved stream token from persistent storage
        return this.localStore.getLastStreamToken().then(function (token) ***REMOVED***
            _this.writeStream.lastStreamToken = token;
            if (_this.shouldStartWatchStream()) ***REMOVED***
                _this.startWatchStream();
            ***REMOVED***
            _this.updateOnlineState(types_1.OnlineState.Unknown);
            return _this.fillWritePipeline(); // This may start the writeStream.
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Temporarily disables the network. The network can be re-enabled using
     * enableNetwork().
     */
    RemoteStore.prototype.disableNetwork = function () ***REMOVED***
        this.disableNetworkInternal();
        // Set the OnlineState to failed so get()'s return from cache, etc.
        this.updateOnlineState(types_1.OnlineState.Failed);
        return Promise.resolve();
    ***REMOVED***;
    /**
     * Disables the network, if it is currently enabled.
     */
    RemoteStore.prototype.disableNetworkInternal = function () ***REMOVED***
        if (this.isNetworkEnabled()) ***REMOVED***
            // NOTE: We're guaranteed not to get any further events from these streams (not even a close
            // event).
            this.watchStream.stop();
            this.writeStream.stop();
            this.cleanUpWatchStreamState();
            this.cleanUpWriteStreamState();
            this.writeStream = null;
            this.watchStream = null;
        ***REMOVED***
    ***REMOVED***;
    RemoteStore.prototype.shutdown = function () ***REMOVED***
        log.debug(LOG_TAG, 'RemoteStore shutting down.');
        this.disableNetworkInternal();
        // Set the OnlineState to Unknown (rather than Failed) to avoid potentially
        // triggering spurious listener events with cached data, etc.
        this.updateOnlineState(types_1.OnlineState.Unknown);
        return Promise.resolve();
    ***REMOVED***;
    /** Starts new listen for the given query. Uses resume token if provided */
    RemoteStore.prototype.listen = function (queryData) ***REMOVED***
        assert_1.assert(!objUtils.contains(this.listenTargets, queryData.targetId), 'listen called with duplicate targetId!');
        // Mark this as something the client is currently listening for.
        this.listenTargets[queryData.targetId] = queryData;
        if (this.shouldStartWatchStream()) ***REMOVED***
            // The listen will be sent in onWatchStreamOpen
            this.startWatchStream();
        ***REMOVED***
        else if (this.isNetworkEnabled() && this.watchStream.isOpen()) ***REMOVED***
            this.sendWatchRequest(queryData);
        ***REMOVED***
    ***REMOVED***;
    /** Removes the listen from server */
    RemoteStore.prototype.unlisten = function (targetId) ***REMOVED***
        assert_1.assert(objUtils.contains(this.listenTargets, targetId), 'unlisten called without assigned target ID!');
        var queryData = this.listenTargets[targetId];
        delete this.listenTargets[targetId];
        if (this.isNetworkEnabled() && this.watchStream.isOpen()) ***REMOVED***
            this.sendUnwatchRequest(targetId);
            if (objUtils.isEmpty(this.listenTargets)) ***REMOVED***
                this.watchStream.markIdle();
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    /**
     * We need to increment the the expected number of pending responses we're due
     * from watch so we wait for the ack to process any messages from this target.
     */
    RemoteStore.prototype.sendWatchRequest = function (queryData) ***REMOVED***
        this.recordPendingTargetRequest(queryData.targetId);
        this.watchStream.watch(queryData);
    ***REMOVED***;
    /**
     * We need to increment the expected number of pending responses we're due
     * from watch so we wait for the removal on the server before we process any
     * messages from this target.
     */
    RemoteStore.prototype.sendUnwatchRequest = function (targetId) ***REMOVED***
        this.recordPendingTargetRequest(targetId);
        this.watchStream.unwatch(targetId);
    ***REMOVED***;
    /**
     * Increment the mapping of how many acks are needed from watch before we can
     * consider the server to be 'in-sync' with the client's active targets.
     */
    RemoteStore.prototype.recordPendingTargetRequest = function (targetId) ***REMOVED***
        // For each request we get we need to record we need a response for it.
        this.pendingTargetResponses[targetId] =
            (this.pendingTargetResponses[targetId] || 0) + 1;
    ***REMOVED***;
    RemoteStore.prototype.startWatchStream = function () ***REMOVED***
        assert_1.assert(this.shouldStartWatchStream(), 'startWriteStream() called when shouldStartWatchStream() is false.');
        this.watchStream.start(***REMOVED***
            onOpen: this.onWatchStreamOpen.bind(this),
            onClose: this.onWatchStreamClose.bind(this),
            onWatchChange: this.onWatchStreamChange.bind(this)
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Returns whether the watch stream should be started because there are
     * active targets trying to be listened too
     */
    RemoteStore.prototype.shouldStartWatchStream = function () ***REMOVED***
        return (this.isNetworkEnabled() &&
            !this.watchStream.isStarted() &&
            !objUtils.isEmpty(this.listenTargets));
    ***REMOVED***;
    RemoteStore.prototype.cleanUpWatchStreamState = function () ***REMOVED***
        // If the connection is closed then we'll never get a snapshot version for
        // the accumulated changes and so we'll never be able to complete the batch.
        // When we start up again the server is going to resend these changes
        // anyway, so just toss the accumulated state.
        this.accumulatedWatchChanges = [];
        this.pendingTargetResponses = ***REMOVED******REMOVED***;
    ***REMOVED***;
    RemoteStore.prototype.onWatchStreamOpen = function () ***REMOVED***
        var _this = this;
        // TODO(b/35852690): close the stream again (with some timeout?) if no watch
        // targets are active
        objUtils.forEachNumber(this.listenTargets, function (targetId, queryData) ***REMOVED***
            _this.sendWatchRequest(queryData);
        ***REMOVED***);
        return Promise.resolve();
    ***REMOVED***;
    RemoteStore.prototype.onWatchStreamClose = function (error) ***REMOVED***
        assert_1.assert(this.isNetworkEnabled(), 'onWatchStreamClose() should only be called when the network is enabled');
        this.cleanUpWatchStreamState();
        // If there was an error, retry the connection.
        if (this.shouldStartWatchStream()) ***REMOVED***
            this.updateOnlineStateAfterFailure();
            this.startWatchStream();
        ***REMOVED***
        else ***REMOVED***
            // No need to restart watch stream because there are no active targets.
            // The online state is set to unknown because there is no active attempt
            // at establishing a connection
            this.updateOnlineState(types_1.OnlineState.Unknown);
        ***REMOVED***
        return Promise.resolve();
    ***REMOVED***;
    RemoteStore.prototype.onWatchStreamChange = function (watchChange, snapshotVersion) ***REMOVED***
        // Mark the connection as healthy because we got a message from the server
        this.updateOnlineState(types_1.OnlineState.Healthy);
        if (watchChange instanceof watch_change_1.WatchTargetChange &&
            watchChange.state === watch_change_1.WatchTargetChangeState.Removed &&
            watchChange.cause) ***REMOVED***
            // There was an error on a target, don't wait for a consistent snapshot
            // to raise events
            return this.handleTargetError(watchChange);
        ***REMOVED***
        // Accumulate watch changes but don't process them if there's no
        // snapshotVersion or it's older than a previous snapshot we've processed
        // (can happen after we resume a target using a resume token).
        this.accumulatedWatchChanges.push(watchChange);
        if (!snapshotVersion.isEqual(snapshot_version_1.SnapshotVersion.MIN) &&
            snapshotVersion.compareTo(this.localStore.getLastRemoteSnapshotVersion()) >= 0) ***REMOVED***
            var changes = this.accumulatedWatchChanges;
            this.accumulatedWatchChanges = [];
            return this.handleWatchChangeBatch(snapshotVersion, changes);
        ***REMOVED***
        else ***REMOVED***
            return Promise.resolve();
        ***REMOVED***
    ***REMOVED***;
    /**
     * Takes a batch of changes from the Datastore, repackages them as a
     * RemoteEvent, and passes that on to the listener, which is typically the
     * SyncEngine.
     */
    RemoteStore.prototype.handleWatchChangeBatch = function (snapshotVersion, changes) ***REMOVED***
        var _this = this;
        var aggregator = new watch_change_1.WatchChangeAggregator(snapshotVersion, this.listenTargets, this.pendingTargetResponses);
        aggregator.addChanges(changes);
        var remoteEvent = aggregator.createRemoteEvent();
        // Get the new response counts from the aggregator
        this.pendingTargetResponses = aggregator.pendingTargetResponses;
        var promises = [];
        // Handle existence filters and existence filter mismatches.
        objUtils.forEachNumber(aggregator.existenceFilters, function (targetId, filter) ***REMOVED***
            var queryData = _this.listenTargets[targetId];
            if (!queryData) ***REMOVED***
                // A watched target might have been removed already.
                return;
            ***REMOVED***
            var query = queryData.query;
            if (query.isDocumentQuery()) ***REMOVED***
                if (filter.count === 0) ***REMOVED***
                    // The existence filter told us the document does not exist.
                    // We need to deduce that this document does not exist and apply
                    // a deleted document to our updates. Without applying a deleted
                    // document there might be another query that will raise this
                    // document as part of a snapshot until it is resolved,
                    // essentially exposing inconsistency between queries.
                    var key = new document_key_1.DocumentKey(query.path);
                    var deletedDoc = new document_1.NoDocument(key, snapshotVersion);
                    remoteEvent.addDocumentUpdate(deletedDoc);
                ***REMOVED***
                else ***REMOVED***
                    assert_1.assert(filter.count === 1, 'Single document existence filter with count: ' + filter.count);
                ***REMOVED***
            ***REMOVED***
            else ***REMOVED***
                // Not a document query.
                var promise = _this.localStore
                    .remoteDocumentKeys(targetId)
                    .then(function (trackedRemote) ***REMOVED***
                    if (remoteEvent.targetChanges[targetId]) ***REMOVED***
                        var mapping = remoteEvent.targetChanges[targetId].mapping;
                        if (mapping !== null) ***REMOVED***
                            if (mapping instanceof remote_event_1.UpdateMapping) ***REMOVED***
                                trackedRemote = mapping.applyToKeySet(trackedRemote);
                            ***REMOVED***
                            else ***REMOVED***
                                assert_1.assert(mapping instanceof remote_event_1.ResetMapping, 'Expected either reset or update mapping but got something else: ' +
                                    mapping);
                                trackedRemote = mapping.documents;
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                    if (trackedRemote.size !== filter.count) ***REMOVED***
                        // Existence filter mismatch, resetting mapping.
                        // Make sure the mismatch is exposed in the remote event.
                        remoteEvent.handleExistenceFilterMismatch(targetId);
                        // Clear the resume token for the query, since we're in a
                        // known mismatch state.
                        var newQueryData = new query_data_1.QueryData(query, targetId, queryData.purpose);
                        _this.listenTargets[targetId] = newQueryData;
                        // Cause a hard reset by unwatching and rewatching
                        // immediately, but deliberately don't send a resume token
                        // so that we get a full update.
                        // Make sure we expect that this acks are going to happen.
                        _this.sendUnwatchRequest(targetId);
                        // Mark the query we send as being on behalf of an existence
                        // filter mismatch, but don't actually retain that in
                        // listenTargets. This ensures that we flag the first
                        // re-listen this way without impacting future listens of
                        // this target (that might happen e.g. on reconnect).
                        var requestQueryData = new query_data_1.QueryData(query, targetId, query_data_1.QueryPurpose.ExistenceFilterMismatch);
                        _this.sendWatchRequest(requestQueryData);
                    ***REMOVED***
                ***REMOVED***);
                promises.push(promise);
            ***REMOVED***
        ***REMOVED***);
        return Promise.all(promises).then(function () ***REMOVED***
            // Update in-memory resume tokens. LocalStore will update the
            // persistent view of these when applying the completed RemoteEvent.
            objUtils.forEachNumber(remoteEvent.targetChanges, function (targetId, change) ***REMOVED***
                if (change.resumeToken.length > 0) ***REMOVED***
                    var queryData = _this.listenTargets[targetId];
                    // A watched target might have been removed already.
                    if (queryData) ***REMOVED***
                        _this.listenTargets[targetId] = queryData.update(***REMOVED***
                            resumeToken: change.resumeToken,
                            snapshotVersion: change.snapshotVersion
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***);
            // Finally handle remote event
            return _this.syncEngine.applyRemoteEvent(remoteEvent);
        ***REMOVED***);
    ***REMOVED***;
    /** Handles an error on a target */
    RemoteStore.prototype.handleTargetError = function (watchChange) ***REMOVED***
        var _this = this;
        assert_1.assert(!!watchChange.cause, 'Handling target error without a cause');
        var error = watchChange.cause;
        var promiseChain = Promise.resolve();
        watchChange.targetIds.forEach(function (targetId) ***REMOVED***
            promiseChain = promiseChain.then(function () ***REMOVED***
                if (objUtils.contains(_this.listenTargets, targetId)) ***REMOVED***
                    delete _this.listenTargets[targetId];
                    return _this.syncEngine.rejectListen(targetId, error);
                ***REMOVED***
                else ***REMOVED***
                    // A watched target might have been removed already.
                    return Promise.resolve();
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
        return promiseChain;
    ***REMOVED***;
    RemoteStore.prototype.cleanUpWriteStreamState = function () ***REMOVED***
        this.lastBatchSeen = mutation_batch_1.BATCHID_UNKNOWN;
        log.debug(LOG_TAG, 'Stopping write stream with ' +
            this.pendingWrites.length +
            ' pending writes');
        this.pendingWrites = [];
    ***REMOVED***;
    /**
     * Notifies that there are new mutations to process in the queue. This is
     * typically called by SyncEngine after it has sent mutations to LocalStore.
     */
    RemoteStore.prototype.fillWritePipeline = function () ***REMOVED***
        var _this = this;
        if (!this.canWriteMutations()) ***REMOVED***
            return Promise.resolve();
        ***REMOVED***
        else ***REMOVED***
            return this.localStore
                .nextMutationBatch(this.lastBatchSeen)
                .then(function (batch) ***REMOVED***
                if (batch === null) ***REMOVED***
                    if (_this.pendingWrites.length === 0) ***REMOVED***
                        _this.writeStream.markIdle();
                    ***REMOVED***
                    return Promise.resolve();
                ***REMOVED***
                else ***REMOVED***
                    _this.commit(batch);
                    return _this.fillWritePipeline();
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * Returns true if the backend can accept additional write requests.
     *
     * When sending mutations to the write stream (e.g. in fillWritePipeline),
     * call this method first to check if more mutations can be sent.
     *
     * Currently the only thing that can prevent the backend from accepting
     * write requests is if there are too many requests already outstanding. As
     * writes complete the backend will be able to accept more.
     */
    RemoteStore.prototype.canWriteMutations = function () ***REMOVED***
        return (this.isNetworkEnabled() && this.pendingWrites.length < MAX_PENDING_WRITES);
    ***REMOVED***;
    // For testing
    RemoteStore.prototype.outstandingWrites = function () ***REMOVED***
        return this.pendingWrites.length;
    ***REMOVED***;
    /**
     * Given mutations to commit, actually commits them to the Datastore. Note
     * that this does *not* return a Promise specifically because the AsyncQueue
     * should not block operations for this.
     */
    RemoteStore.prototype.commit = function (batch) ***REMOVED***
        assert_1.assert(this.canWriteMutations(), "commit called when batches can't be written");
        this.lastBatchSeen = batch.batchId;
        this.pendingWrites.push(batch);
        if (this.shouldStartWriteStream()) ***REMOVED***
            this.startWriteStream();
        ***REMOVED***
        else if (this.isNetworkEnabled() && this.writeStream.handshakeComplete) ***REMOVED***
            this.writeStream.writeMutations(batch.mutations);
        ***REMOVED***
    ***REMOVED***;
    RemoteStore.prototype.shouldStartWriteStream = function () ***REMOVED***
        return (this.isNetworkEnabled() &&
            !this.writeStream.isStarted() &&
            this.pendingWrites.length > 0);
    ***REMOVED***;
    RemoteStore.prototype.startWriteStream = function () ***REMOVED***
        assert_1.assert(this.shouldStartWriteStream(), 'startWriteStream() called when shouldStartWriteStream() is false.');
        this.writeStream.start(***REMOVED***
            onOpen: this.onWriteStreamOpen.bind(this),
            onClose: this.onWriteStreamClose.bind(this),
            onHandshakeComplete: this.onWriteHandshakeComplete.bind(this),
            onMutationResult: this.onMutationResult.bind(this)
        ***REMOVED***);
    ***REMOVED***;
    RemoteStore.prototype.onWriteStreamOpen = function () ***REMOVED***
        this.writeStream.writeHandshake();
        return Promise.resolve();
    ***REMOVED***;
    RemoteStore.prototype.onWriteHandshakeComplete = function () ***REMOVED***
        var _this = this;
        // Record the stream token.
        return this.localStore
            .setLastStreamToken(this.writeStream.lastStreamToken)
            .then(function () ***REMOVED***
            // Drain any pending writes.
            //
            // Note that at this point pendingWrites contains mutations that
            // have already been accepted by fillWritePipeline/commitBatch. If
            // the pipeline is full, canWriteMutations will be false, despite
            // the fact that we actually need to send mutations over.
            //
            // This also means that this method indirectly respects the limits
            // imposed by canWriteMutations since writes can't be added to the
            // pendingWrites array when canWriteMutations is false. If the
            // limits imposed by canWriteMutations actually protect us from
            // DOSing ourselves then those limits won't be exceeded here and
            // we'll continue to make progress.
            for (var _i = 0, _a = _this.pendingWrites; _i < _a.length; _i++) ***REMOVED***
                var batch = _a[_i];
                _this.writeStream.writeMutations(batch.mutations);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    RemoteStore.prototype.onMutationResult = function (commitVersion, results) ***REMOVED***
        var _this = this;
        // This is a response to a write containing mutations and should be
        // correlated to the first pending write.
        assert_1.assert(this.pendingWrites.length > 0, 'Got result for empty pending writes');
        var batch = this.pendingWrites.shift();
        var success = mutation_batch_1.MutationBatchResult.from(batch, commitVersion, results, this.writeStream.lastStreamToken);
        return this.syncEngine.applySuccessfulWrite(success).then(function () ***REMOVED***
            // It's possible that with the completion of this mutation another
            // slot has freed up.
            return _this.fillWritePipeline();
        ***REMOVED***);
    ***REMOVED***;
    RemoteStore.prototype.onWriteStreamClose = function (error) ***REMOVED***
        var _this = this;
        assert_1.assert(this.isNetworkEnabled(), 'onWriteStreamClose() should only be called when the network is enabled');
        // If the write stream closed due to an error, invoke the error callbacks if
        // there are pending writes.
        if (error && this.pendingWrites.length > 0) ***REMOVED***
            assert_1.assert(!!error, 'We have pending writes, but the write stream closed without an error');
            // A promise that is resolved after we processed the error
            var errorHandling = void 0;
            if (this.writeStream.handshakeComplete) ***REMOVED***
                // This error affects the actual write.
                errorHandling = this.handleWriteError(error);
            ***REMOVED***
            else ***REMOVED***
                // If there was an error before the handshake has finished, it's
                // possible that the server is unable to process the stream token
                // we're sending. (Perhaps it's too old?)
                errorHandling = this.handleHandshakeError(error);
            ***REMOVED***
            return errorHandling.then(function () ***REMOVED***
                // The write stream might have been started by refilling the write
                // pipeline for failed writes
                if (_this.shouldStartWriteStream()) ***REMOVED***
                    _this.startWriteStream();
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
        else ***REMOVED***
            // No pending writes, nothing to do
            return Promise.resolve();
        ***REMOVED***
    ***REMOVED***;
    RemoteStore.prototype.handleHandshakeError = function (error) ***REMOVED***
        // Reset the token if it's a permanent error or the error code is
        // ABORTED, signaling the write stream is no longer valid.
        if (rpc_error_1.isPermanentError(error.code) || error.code === error_1.Code.ABORTED) ***REMOVED***
            log.debug(LOG_TAG, 'RemoteStore error before completed handshake; resetting stream token: ', this.writeStream.lastStreamToken);
            this.writeStream.lastStreamToken = platform_1.emptyByteString();
            return this.localStore.setLastStreamToken(platform_1.emptyByteString());
        ***REMOVED***
        else ***REMOVED***
            // Some other error, don't reset stream token. Our stream logic will
            // just retry with exponential backoff.
            return Promise.resolve();
        ***REMOVED***
    ***REMOVED***;
    RemoteStore.prototype.handleWriteError = function (error) ***REMOVED***
        var _this = this;
        if (rpc_error_1.isPermanentError(error.code)) ***REMOVED***
            // This was a permanent error, the request itself was the problem
            // so it's not going to succeed if we resend it.
            var batch = this.pendingWrites.shift();
            // In this case it's also unlikely that the server itself is melting
            // down -- this was just a bad request so inhibit backoff on the next
            // restart.
            this.writeStream.inhibitBackoff();
            return this.syncEngine
                .rejectFailedWrite(batch.batchId, error)
                .then(function () ***REMOVED***
                // It's possible that with the completion of this mutation
                // another slot has freed up.
                return _this.fillWritePipeline();
            ***REMOVED***);
        ***REMOVED***
        else ***REMOVED***
            // Transient error, just let the retry logic kick in.
            return Promise.resolve();
        ***REMOVED***
    ***REMOVED***;
    RemoteStore.prototype.createTransaction = function () ***REMOVED***
        return new transaction_1.Transaction(this.datastore);
    ***REMOVED***;
    RemoteStore.prototype.handleUserChange = function (user) ***REMOVED***
        log.debug(LOG_TAG, 'RemoteStore changing users: uid=', user.uid);
        // If the network has been explicitly disabled, make sure we don't
        // accidentally re-enable it.
        if (this.isNetworkEnabled()) ***REMOVED***
            // Tear down and re-create our network streams. This will ensure we get a fresh auth token
            // for the new user and re-fill the write pipeline with new mutations from the LocalStore
            // (since mutations are per-user).
            this.disableNetworkInternal();
            this.updateOnlineState(types_1.OnlineState.Unknown);
            return this.enableNetwork();
        ***REMOVED***
    ***REMOVED***;
    return RemoteStore;
***REMOVED***());
exports.RemoteStore = RemoteStore;
