// @firebase/firestore/dist/cjs/src/remote/stream_bridge.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var assert_1 = require("../util/assert");
/**
 * Provides a simple helper class that implements the Stream interface to
 * bridge to other implementations that are streams but do not implement the
 * interface. The stream callbacks are invoked with the callOn... methods.
 */
var StreamBridge = /** @class */ (function () ***REMOVED***
    function StreamBridge(args) ***REMOVED***
        this.sendFn = args.sendFn;
        this.closeFn = args.closeFn;
    ***REMOVED***
    StreamBridge.prototype.onOpen = function (callback) ***REMOVED***
        assert_1.assert(!this.wrappedOnOpen, 'Called onOpen on stream twice!');
        this.wrappedOnOpen = callback;
    ***REMOVED***;
    StreamBridge.prototype.onClose = function (callback) ***REMOVED***
        assert_1.assert(!this.wrappedOnClose, 'Called onClose on stream twice!');
        this.wrappedOnClose = callback;
    ***REMOVED***;
    StreamBridge.prototype.onMessage = function (callback) ***REMOVED***
        assert_1.assert(!this.wrappedOnMessage, 'Called onMessage on stream twice!');
        this.wrappedOnMessage = callback;
    ***REMOVED***;
    StreamBridge.prototype.close = function () ***REMOVED***
        this.closeFn();
    ***REMOVED***;
    StreamBridge.prototype.send = function (msg) ***REMOVED***
        this.sendFn(msg);
    ***REMOVED***;
    StreamBridge.prototype.callOnOpen = function () ***REMOVED***
        assert_1.assert(this.wrappedOnOpen !== undefined, 'Cannot call onOpen because no callback was set');
        this.wrappedOnOpen();
    ***REMOVED***;
    StreamBridge.prototype.callOnClose = function (err) ***REMOVED***
        assert_1.assert(this.wrappedOnClose !== undefined, 'Cannot call onClose because no callback was set');
        this.wrappedOnClose(err);
    ***REMOVED***;
    StreamBridge.prototype.callOnMessage = function (msg) ***REMOVED***
        assert_1.assert(this.wrappedOnMessage !== undefined, 'Cannot call onMessage because no callback was set');
        this.wrappedOnMessage(msg);
    ***REMOVED***;
    return StreamBridge;
***REMOVED***());
exports.StreamBridge = StreamBridge;
