// watchpack/lib/watchpack.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var watcherManager = require("./watcherManager");
var EventEmitter = require("events").EventEmitter;

function Watchpack(options) ***REMOVED***
	EventEmitter.call(this);
	if(!options) options = ***REMOVED******REMOVED***;
	if(!options.aggregateTimeout) options.aggregateTimeout = 200;
	this.options = options;
	this.watcherOptions = ***REMOVED***
		ignored: options.ignored,
		poll: options.poll
	***REMOVED***;
	this.fileWatchers = [];
	this.dirWatchers = [];
	this.mtimes = Object.create(null);
	this.paused = false;
	this.aggregatedChanges = [];
	this.aggregatedRemovals = [];
	this.aggregateTimeout = 0;
	this._onTimeout = this._onTimeout.bind(this);
***REMOVED***

module.exports = Watchpack;

Watchpack.prototype = Object.create(EventEmitter.prototype);

Watchpack.prototype.watch = function watch(files, directories, startTime) ***REMOVED***
	this.paused = false;
	var oldFileWatchers = this.fileWatchers;
	var oldDirWatchers = this.dirWatchers;
	this.fileWatchers = files.map(function(file) ***REMOVED***
		return this._fileWatcher(file, watcherManager.watchFile(file, this.watcherOptions, startTime));
	***REMOVED***, this);
	this.dirWatchers = directories.map(function(dir) ***REMOVED***
		return this._dirWatcher(dir, watcherManager.watchDirectory(dir, this.watcherOptions, startTime));
	***REMOVED***, this);
	oldFileWatchers.forEach(function(w) ***REMOVED***
		w.close();
	***REMOVED***, this);
	oldDirWatchers.forEach(function(w) ***REMOVED***
		w.close();
	***REMOVED***, this);
***REMOVED***;

Watchpack.prototype.close = function resume() ***REMOVED***
	this.paused = true;
	if(this.aggregateTimeout)
		clearTimeout(this.aggregateTimeout);
	this.fileWatchers.forEach(function(w) ***REMOVED***
		w.close();
	***REMOVED***, this);
	this.dirWatchers.forEach(function(w) ***REMOVED***
		w.close();
	***REMOVED***, this);
	this.fileWatchers.length = 0;
	this.dirWatchers.length = 0;
***REMOVED***;

Watchpack.prototype.pause = function pause() ***REMOVED***
	this.paused = true;
	if(this.aggregateTimeout)
		clearTimeout(this.aggregateTimeout);
***REMOVED***;

function addWatchersToArray(watchers, array) ***REMOVED***
	watchers.forEach(function(w) ***REMOVED***
		if(array.indexOf(w.directoryWatcher) < 0) ***REMOVED***
			array.push(w.directoryWatcher);
			addWatchersToArray(Object.keys(w.directoryWatcher.directories).reduce(function(a, dir) ***REMOVED***
				if(w.directoryWatcher.directories[dir] !== true)
					a.push(w.directoryWatcher.directories[dir]);
				return a;
			***REMOVED***, []), array);
		***REMOVED***
	***REMOVED***);
***REMOVED***

Watchpack.prototype.getTimes = function() ***REMOVED***
	var directoryWatchers = [];
	addWatchersToArray(this.fileWatchers.concat(this.dirWatchers), directoryWatchers);
	var obj = Object.create(null);
	directoryWatchers.forEach(function(w) ***REMOVED***
		var times = w.getTimes();
		Object.keys(times).forEach(function(file) ***REMOVED***
			obj[file] = times[file];
		***REMOVED***);
	***REMOVED***);
	return obj;
***REMOVED***;

Watchpack.prototype._fileWatcher = function _fileWatcher(file, watcher) ***REMOVED***
	watcher.on("change", function(mtime, type) ***REMOVED***
		this._onChange(file, mtime, file, type);
	***REMOVED***.bind(this));
	watcher.on("remove", function(type) ***REMOVED***
		this._onRemove(file, file, type);
	***REMOVED***.bind(this));
	return watcher;
***REMOVED***;

Watchpack.prototype._dirWatcher = function _dirWatcher(item, watcher) ***REMOVED***
	watcher.on("change", function(file, mtime, type) ***REMOVED***
		this._onChange(item, mtime, file, type);
	***REMOVED***.bind(this));
	return watcher;
***REMOVED***;

Watchpack.prototype._onChange = function _onChange(item, mtime, file) ***REMOVED***
	file = file || item;
	this.mtimes[file] = mtime;
	if(this.paused) return;
	this.emit("change", file, mtime);
	if(this.aggregateTimeout)
		clearTimeout(this.aggregateTimeout);
	if(this.aggregatedChanges.indexOf(item) < 0)
		this.aggregatedChanges.push(item);
	this.aggregateTimeout = setTimeout(this._onTimeout, this.options.aggregateTimeout);
***REMOVED***;

Watchpack.prototype._onRemove = function _onRemove(item, file) ***REMOVED***
	file = file || item;
	delete this.mtimes[item];
	if(this.paused) return;
	this.emit("remove", item);
	if(this.aggregateTimeout)
		clearTimeout(this.aggregateTimeout);
	if(this.aggregatedRemovals.indexOf(item) < 0)
		this.aggregatedRemovals.push(item);
	this.aggregateTimeout = setTimeout(this._onTimeout, this.options.aggregateTimeout);
***REMOVED***;

Watchpack.prototype._onTimeout = function _onTimeout() ***REMOVED***
	this.aggregateTimeout = 0;
	var changes = this.aggregatedChanges;
	var removals = this.aggregatedRemovals;
	this.aggregatedChanges = [];
	this.aggregatedRemovals = [];
	this.emit("aggregated", changes, removals);
***REMOVED***;
