// @firebase/firestore/dist/cjs/src/local/simple_db.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var assert_1 = require("../util/assert");
var log_1 = require("../util/log");
var persistence_promise_1 = require("./persistence_promise");
var LOG_TAG = 'SimpleDb';
/**
 * Provides a wrapper around IndexedDb with a simplified interface that uses
 * Promise-like return values to chain operations. Real promises cannot be used
 * since .then() continuations are executed asynchronously (e.g. via
 * .setImmediate), which would cause IndexedDB to end the transaction.
 * See PersistencePromise for more details.
 */
var SimpleDb = /** @class */ (function () ***REMOVED***
    function SimpleDb(db) ***REMOVED***
        this.db = db;
    ***REMOVED***
    /** Opens the specified database, creating or upgrading it if necessary. */
    SimpleDb.openOrCreate = function (name, version, runUpgrade) ***REMOVED***
        assert_1.assert(SimpleDb.isAvailable(), 'IndexedDB not supported in current environment.');
        log_1.debug(LOG_TAG, 'Opening database:', name);
        return new persistence_promise_1.PersistencePromise(function (resolve, reject) ***REMOVED***
            // TODO(mikelehen): Investigate browser compatibility.
            // https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB
            // suggests IE9 and older WebKit browsers handle upgrade
            // differently.
            var request = window.indexedDB.open(name, version);
            request.onsuccess = function (event) ***REMOVED***
                var db = event.target.result;
                resolve(new SimpleDb(db));
            ***REMOVED***;
            request.onerror = function (event) ***REMOVED***
                reject(event.target.error);
            ***REMOVED***;
            request.onupgradeneeded = function (event) ***REMOVED***
                log_1.debug(LOG_TAG, 'Database "' + name + '" requires upgrade from version:', event.oldVersion);
                // TODO(mikelehen): If/when we need to do an actual data
                // migration, we'll want to wrap db in a SimpleDb and have the
                // runUpgrade function return a PersistencePromise, since we'll
                // likely need to do async reads and writes. For now we're
                // cheating and just passing the raw IndexedDB in, since
                // createObjectStore(), etc. are synchronous.
                var db = event.target.result;
                runUpgrade(db, event.oldVersion);
            ***REMOVED***;
        ***REMOVED***).toPromise();
    ***REMOVED***;
    /** Deletes the specified database. */
    SimpleDb.delete = function (name) ***REMOVED***
        log_1.debug(LOG_TAG, 'Removing database:', name);
        return wrapRequest(window.indexedDB.deleteDatabase(name)).toPromise();
    ***REMOVED***;
    /** Returns true if IndexedDB is available in the current environment. */
    SimpleDb.isAvailable = function () ***REMOVED***
        if (typeof window === 'undefined' || window.indexedDB == null) ***REMOVED***
            return false;
        ***REMOVED***
        // We extensively use indexed array values and compound keys,
        // which IE and Edge do not support. However, they still have indexedDB
        // defined on the window, so we need to check for them here and make sure
        // to return that persistence is not enabled for those browsers.
        // For tracking support of this feature, see here:
        // https://developer.microsoft.com/en-us/microsoft-edge/platform/status/indexeddbarraysandmultientrysupport/
        // Check the UA string to find out the browser.
        var ua = window.navigator.userAgent;
        // IE 10
        // ua = 'Mozilla/5.0 (compatible; MSIE 10.0; Windows NT 6.2; Trident/6.0)';
        // IE 11
        // ua = 'Mozilla/5.0 (Windows NT 6.3; Trident/7.0; rv:11.0) like Gecko';
        // Edge
        // ua = 'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML,
        // like Gecko) Chrome/39.0.2171.71 Safari/537.36 Edge/12.0';
        if (ua.indexOf('MSIE ') > 0 ||
            ua.indexOf('Trident/') > 0 ||
            ua.indexOf('Edge/') > 0) ***REMOVED***
            return false;
        ***REMOVED***
        else ***REMOVED***
            return true;
        ***REMOVED***
    ***REMOVED***;
    SimpleDb.prototype.runTransaction = function (mode, objectStores, transactionFn) ***REMOVED***
        var transaction = new SimpleDbTransaction(this.db, mode, objectStores);
        var transactionFnResult = transactionFn(transaction)
            .catch(function (error) ***REMOVED***
            // Abort the transaction if there was an
            // error.
            transaction.abort();
            return persistence_promise_1.PersistencePromise.reject(error);
        ***REMOVED***)
            .toPromise();
        // Wait for the transaction to complete (i.e. IndexedDb's onsuccess event to
        // fire), but still return the original transactionFnResult back to the
        // caller.
        return transaction.completionPromise.then(function () ***REMOVED*** return transactionFnResult; ***REMOVED***);
    ***REMOVED***;
    SimpleDb.prototype.close = function () ***REMOVED***
        this.db.close();
    ***REMOVED***;
    return SimpleDb;
***REMOVED***());
exports.SimpleDb = SimpleDb;
/**
 * A controller for iterating over a key range or index. It allows an iterate
 * callback to delete the currently-referenced object, or jump to a new key
 * within the key range or index.
 */
var IterationController = /** @class */ (function () ***REMOVED***
    function IterationController(dbCursor) ***REMOVED***
        this.dbCursor = dbCursor;
        this.shouldStop = false;
        this.nextKey = null;
    ***REMOVED***
    Object.defineProperty(IterationController.prototype, "isDone", ***REMOVED***
        get: function () ***REMOVED***
            return this.shouldStop;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(IterationController.prototype, "skipToKey", ***REMOVED***
        get: function () ***REMOVED***
            return this.nextKey;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(IterationController.prototype, "cursor", ***REMOVED***
        set: function (value) ***REMOVED***
            this.dbCursor = value;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    /**
     * This function can be called to stop iteration at any point.
     */
    IterationController.prototype.done = function () ***REMOVED***
        this.shouldStop = true;
    ***REMOVED***;
    /**
     * This function can be called to skip to that next key, which could be
     * an index or a primary key.
     */
    IterationController.prototype.skip = function (key) ***REMOVED***
        this.nextKey = key;
    ***REMOVED***;
    /**
     * Delete the current cursor value from the object store.
     *
     * NOTE: You CANNOT do this with a keysOnly query.
     */
    IterationController.prototype.delete = function () ***REMOVED***
        return wrapRequest(this.dbCursor.delete());
    ***REMOVED***;
    return IterationController;
***REMOVED***());
exports.IterationController = IterationController;
/**
 * Wraps an IDBTransaction and exposes a store() method to get a handle to a
 * specific object store.
 */
var SimpleDbTransaction = /** @class */ (function () ***REMOVED***
    function SimpleDbTransaction(db, mode, objectStoresNames) ***REMOVED***
        var _this = this;
        this.aborted = false;
        this.transaction = db.transaction(objectStoresNames, mode);
        this.completionPromise = new Promise(function (resolve, reject) ***REMOVED***
            // We consider aborting to be "normal" and just resolve the promise.
            // May need to revisit if/when we actually need to abort transactions.
            _this.transaction.onabort = _this.transaction.oncomplete = function (event) ***REMOVED***
                resolve();
            ***REMOVED***;
            _this.transaction.onerror = function (event) ***REMOVED***
                reject(event.target.error);
            ***REMOVED***;
        ***REMOVED***);
    ***REMOVED***
    SimpleDbTransaction.prototype.abort = function () ***REMOVED***
        if (!this.aborted) ***REMOVED***
            log_1.debug(LOG_TAG, 'Aborting transaction.');
            this.aborted = true;
            this.transaction.abort();
        ***REMOVED***
    ***REMOVED***;
    /**
     * Returns a SimpleDbStore<KeyType, ValueType> for the specified store. All
     * operations performed on the SimpleDbStore happen within the context of this
     * transaction and it cannot be used anymore once the transaction is
     * completed.
     *
     * Note that we can't actually enforce that the KeyType and ValueType are
     * correct, but they allow type safety through the rest of the consuming code.
     */
    SimpleDbTransaction.prototype.store = function (storeName) ***REMOVED***
        var store = this.transaction.objectStore(storeName);
        assert_1.assert(!!store, 'Object store not part of transaction: ' + storeName);
        return new SimpleDbStore(store);
    ***REMOVED***;
    return SimpleDbTransaction;
***REMOVED***());
exports.SimpleDbTransaction = SimpleDbTransaction;
/**
 * A wrapper around an IDBObjectStore providing an API that:
 *
 * 1) Has generic KeyType / ValueType parameters to provide strongly-typed
 * methods for acting against the object store.
 * 2) Deals with IndexedDB's onsuccess / onerror event callbacks, making every
 * method return a PersistencePromise instead.
 * 3) Provides a higher-level API to avoid needing to do excessive wrapping of
 * intermediate IndexedDB types (IDBCursorWithValue, etc.)
 */
var SimpleDbStore = /** @class */ (function () ***REMOVED***
    function SimpleDbStore(store) ***REMOVED***
        this.store = store;
    ***REMOVED***
    SimpleDbStore.prototype.put = function (keyOrValue, value) ***REMOVED***
        var request;
        if (value !== undefined) ***REMOVED***
            log_1.debug(LOG_TAG, 'PUT', this.store.name, keyOrValue, value);
            request = this.store.put(value, keyOrValue);
        ***REMOVED***
        else ***REMOVED***
            log_1.debug(LOG_TAG, 'PUT', this.store.name, '<auto-key>', keyOrValue);
            request = this.store.put(keyOrValue);
        ***REMOVED***
        return wrapRequest(request);
    ***REMOVED***;
    /**
     * Gets the object with the specified key from the specified store, or null
     * if no object exists with the specified key.
     *
     * @key The key of the object to get.
     * @return The object with the specified key or null if no object exists.
     */
    SimpleDbStore.prototype.get = function (key) ***REMOVED***
        var _this = this;
        var request = this.store.get(key);
        // tslint:disable-next-line:no-any We're doing an unsafe cast to ValueType.
        return wrapRequest(request).next(function (result) ***REMOVED***
            // Normalize nonexistence to null.
            if (result === undefined) ***REMOVED***
                result = null;
            ***REMOVED***
            log_1.debug(LOG_TAG, 'GET', _this.store.name, key, result);
            return result;
        ***REMOVED***);
    ***REMOVED***;
    SimpleDbStore.prototype.delete = function (key) ***REMOVED***
        log_1.debug(LOG_TAG, 'DELETE', this.store.name, key);
        var request = this.store.delete(key);
        return wrapRequest(request);
    ***REMOVED***;
    SimpleDbStore.prototype.loadAll = function (indexOrRange, range) ***REMOVED***
        var cursor = this.cursor(this.options(indexOrRange, range));
        var results = [];
        return this.iterateCursor(cursor, function (key, value) ***REMOVED***
            results.push(value);
        ***REMOVED***).next(function () ***REMOVED***
            return results;
        ***REMOVED***);
    ***REMOVED***;
    SimpleDbStore.prototype.deleteAll = function (indexOrRange, range) ***REMOVED***
        log_1.debug(LOG_TAG, 'DELETE ALL', this.store.name);
        var options = this.options(indexOrRange, range);
        options.keysOnly = false;
        var cursor = this.cursor(options);
        return this.iterateCursor(cursor, function (key, value, control) ***REMOVED***
            // NOTE: Calling delete() on a cursor is documented as more efficient than
            // calling delete() on an object store with a single key
            // (https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/delete),
            // however, this requires us *not* to use a keysOnly cursor
            // (https://developer.mozilla.org/en-US/docs/Web/API/IDBCursor/delete). We
            // may want to compare the performance of each method.
            return control.delete();
        ***REMOVED***);
    ***REMOVED***;
    SimpleDbStore.prototype.iterate = function (optionsOrCallback, callback) ***REMOVED***
        var options;
        if (!callback) ***REMOVED***
            options = ***REMOVED******REMOVED***;
            callback = optionsOrCallback;
        ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
        ***REMOVED***
        var cursor = this.cursor(options);
        return this.iterateCursor(cursor, callback);
    ***REMOVED***;
    SimpleDbStore.prototype.iterateCursor = function (cursorRequest, fn) ***REMOVED***
        var results = [];
        return new persistence_promise_1.PersistencePromise(function (resolve, reject) ***REMOVED***
            cursorRequest.onerror = function (event) ***REMOVED***
                reject(event.target.error);
            ***REMOVED***;
            cursorRequest.onsuccess = function (event) ***REMOVED***
                var cursor = event.target.result;
                if (!cursor) ***REMOVED***
                    resolve();
                    return;
                ***REMOVED***
                var controller = new IterationController(cursor);
                var userResult = fn(cursor.primaryKey, cursor.value, controller);
                if (userResult instanceof persistence_promise_1.PersistencePromise) ***REMOVED***
                    results.push(userResult);
                ***REMOVED***
                if (controller.isDone) ***REMOVED***
                    resolve();
                ***REMOVED***
                else if (controller.skipToKey === null) ***REMOVED***
                    cursor.continue();
                ***REMOVED***
                else ***REMOVED***
                    cursor.continue(controller.skipToKey);
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***).next(function () ***REMOVED***
            return persistence_promise_1.PersistencePromise.waitFor(results);
        ***REMOVED***);
    ***REMOVED***;
    SimpleDbStore.prototype.options = function (indexOrRange, range) ***REMOVED***
        var indexName = undefined;
        if (indexOrRange !== undefined) ***REMOVED***
            if (typeof indexOrRange === 'string') ***REMOVED***
                indexName = indexOrRange;
            ***REMOVED***
            else ***REMOVED***
                assert_1.assert(range === undefined, '3rd argument must not be defined if 2nd is a range.');
                range = indexOrRange;
            ***REMOVED***
        ***REMOVED***
        return ***REMOVED*** index: indexName, range: range ***REMOVED***;
    ***REMOVED***;
    SimpleDbStore.prototype.cursor = function (options) ***REMOVED***
        var direction = 'next';
        if (options.reverse) ***REMOVED***
            direction = 'prev';
        ***REMOVED***
        if (options.index) ***REMOVED***
            var index = this.store.index(options.index);
            if (options.keysOnly) ***REMOVED***
                return index.openKeyCursor(options.range, direction);
            ***REMOVED***
            else ***REMOVED***
                return index.openCursor(options.range, direction);
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            return this.store.openCursor(options.range, direction);
        ***REMOVED***
    ***REMOVED***;
    return SimpleDbStore;
***REMOVED***());
exports.SimpleDbStore = SimpleDbStore;
/**
 * Wraps an IDBRequest in a PersistencePromise, using the onsuccess / onerror
 * handlers to resolve / reject the PersistencePromise as appropriate.
 */
function wrapRequest(request) ***REMOVED***
    return new persistence_promise_1.PersistencePromise(function (resolve, reject) ***REMOVED***
        request.onsuccess = function (event) ***REMOVED***
            var result = event.target.result;
            resolve(result);
        ***REMOVED***;
        request.onerror = function (event) ***REMOVED***
            reject(event.target.error);
        ***REMOVED***;
    ***REMOVED***);
***REMOVED***
