// empower/lib/capturable.js
'use strict';

module.exports = function capturable () ***REMOVED***
    var events = [];

    function _capt (value, espath) ***REMOVED***
        events.push(***REMOVED***value: value, espath: espath***REMOVED***);
        return value;
    ***REMOVED***

    function _expr (value, args) ***REMOVED***
        var captured = events;
        events = [];
        var source = ***REMOVED***
            content: args.content,
            filepath: args.filepath,
            line: args.line
        ***REMOVED***;
        if (args.generator) ***REMOVED***
            source.generator = true;
        ***REMOVED***
        if (args.async) ***REMOVED***
            source.async = true;
        ***REMOVED***
        return ***REMOVED***
            powerAssertContext: ***REMOVED***
                value: value,
                events: captured
            ***REMOVED***,
            source: source
        ***REMOVED***;
    ***REMOVED***

    return ***REMOVED***
        _capt: _capt,
        _expr: _expr
    ***REMOVED***;
***REMOVED***;
