// google-auth-library/lib/auth/googleauth.js
/**
 * Copyright 2014 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var JWTClient = require('./jwtclient.js');
var ComputeClient = require('./computeclient.js');
var exec = require('child_process').exec;
var fs = require('fs');
var os = require('os');
var path = require('path');
var util = require('util');
var DefaultTransporter = require('../transporters.js');

/**
 * GoogleAuth account manager.
 *
 * @constructor
 */
function GoogleAuth() ***REMOVED***
  this.JWTClient = JWTClient;
  this.ComputeClient = ComputeClient;
  this._cachedCredential = null;
***REMOVED***

// Executes the given callback if it is not null.
function callback(c) ***REMOVED***
  if (c) ***REMOVED***
    return c.apply(null, Array.prototype.slice.call(arguments, 1));
  ***REMOVED***
***REMOVED***

// Creates an Error containing the given message, and includes the message from the optional err
// passed in.
function createError(message, err) ***REMOVED***
  var s = message || '';
  if (err) ***REMOVED***
    var errorMessage = String(err);
    if (errorMessage && errorMessage.length > 0) ***REMOVED***
      if (s.length > 0) ***REMOVED***
        s += ' ';
      ***REMOVED***
      s += errorMessage;
    ***REMOVED***
  ***REMOVED***
  return Error(s);
***REMOVED***

/**
 * Convenience field mapping in the IAM credential type.
 */
GoogleAuth.prototype.IAMAuth = require('./iam.js');

/**
 * Convenience field mapping in the Compute credential type.
 */
GoogleAuth.prototype.Compute = require('./computeclient.js');

/**
 * Convenience field mapping in the JWT credential type.
 */
GoogleAuth.prototype.JWT = require('./jwtclient.js');

/**
 * Convenience field mapping in the JWT Access credential type.
 */
GoogleAuth.prototype.JWTAccess = require('./jwtaccess.js');

/**
 * Convenience field mapping in the OAuth2 credential type.
 */
GoogleAuth.prototype.OAuth2 = require('./oauth2client.js');

/**
 * Convenience field mapping to the UserRefreshClient credential type.
 */
GoogleAuth.prototype.UserRefreshClient = require('./refreshclient.js');

/**
 * Caches a value indicating whether the auth layer is running on Google Compute Engine.
 * @private
 */
GoogleAuth.prototype._isGCE = false;

/**
 * Caches a value indicating whether we've checked whether the auth layer is running on
 * Google Compute Engine.
 * @private
 */
GoogleAuth.prototype._checked_isGCE = false;

/**
 * Obtains the default project ID for the application..
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 */
GoogleAuth.prototype.getDefaultProjectId = function(opt_callback) ***REMOVED***
  var that = this;

  // In implicit case, supports three environments. In order of precedence, the
  // implicit environments are:
  //
  // * GCLOUD_PROJECT or GOOGLE_CLOUD_PROJECT environment variable
  // * GOOGLE_APPLICATION_CREDENTIALS JSON file
  // * Get default service project from
  //  ``$ gcloud beta auth application-default login``
  // * Google App Engine application ID (Not implemented yet)
  // * Google Compute Engine project ID (from metadata server) (Not implemented yet)

  if (that._cachedProjectId) ***REMOVED***
    process.nextTick(function() ***REMOVED***
      callback(opt_callback, null, that._cachedProjectId);
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    var my_callback = function(err, projectId) ***REMOVED***
      if (!err && projectId) ***REMOVED***
        that._cachedprojectId = projectId;
      ***REMOVED***
      process.nextTick(function() ***REMOVED***
        callback(opt_callback, err, projectId);
      ***REMOVED***);
    ***REMOVED***;

    // environment variable
    if (that._getProductionProjectId(my_callback)) ***REMOVED***
      return;
    ***REMOVED***

    // json file
    that._getFileProjectId(function(err, projectId) ***REMOVED***
      if (err || projectId) ***REMOVED***
        my_callback(err, projectId);
        return;
      ***REMOVED***

      // Google Cloud SDK default project id
      that._getDefaultServiceProjectId(function(err, projectId) ***REMOVED***
        if (err || projectId) ***REMOVED***
          my_callback(err, projectId);
          return;
        ***REMOVED***

        // Get project ID from Compute Engine metadata server
        that._getGCEProjectId(my_callback);
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Loads the project id from environment variables.
 * @param ***REMOVED***function***REMOVED*** _callback Callback.
 * @api private
 */
GoogleAuth.prototype._getProductionProjectId = function(_callback) ***REMOVED***
  var projectId = this._getEnv('GCLOUD_PROJECT') || this._getEnv('GOOGLE_CLOUD_PROJECT');
  if (projectId) ***REMOVED***
    process.nextTick(function() ***REMOVED***
      callback(_callback, null, projectId);
    ***REMOVED***);
  ***REMOVED***
  return projectId;
***REMOVED***;

/**
 * Loads the project id from the GOOGLE_APPLICATION_CREDENTIALS json file.
 * @param ***REMOVED***function***REMOVED*** _callback Callback.
 * @api private
 */
GoogleAuth.prototype._getFileProjectId = function(_callback) ***REMOVED***
  var that = this;
  if (that._cachedCredential) ***REMOVED***
    // Try to read the project ID from the cached credentials file
    process.nextTick(function() ***REMOVED***
      callback(_callback, null, that._cachedCredential.projectId);
    ***REMOVED***);
    return;
  ***REMOVED***

  // Try to load a credentials file and read its project ID
  var pathExists = that._tryGetApplicationCredentialsFromEnvironmentVariable(function(err, result) ***REMOVED***
    if (!err && result) ***REMOVED***
      callback(_callback, null, result.projectId);
      return;
    ***REMOVED***
    callback(_callback, err);
  ***REMOVED***);

  if (!pathExists) ***REMOVED***
    callback(_callback, null);
  ***REMOVED***
***REMOVED***;

/**
 * Loads the default project of the Google Cloud SDK.
 * @param ***REMOVED***function***REMOVED*** _callback Callback.
 * @api private
 */
GoogleAuth.prototype._getDefaultServiceProjectId = function(_callback) ***REMOVED***
  this._getSDKDefaultProjectId(function(err, stdout) ***REMOVED***
    var projectId;
    if (!err && stdout) ***REMOVED***
      try ***REMOVED***
        projectId = JSON.parse(stdout).core.project;
      ***REMOVED*** catch (err) ***REMOVED***
        projectId = null;
      ***REMOVED***
    ***REMOVED***
    // Ignore any errors
    callback(_callback, null, projectId);
  ***REMOVED***);
***REMOVED***;

/**
 * Run the Google Cloud SDK command that prints the default project ID
 * @param ***REMOVED***function***REMOVED*** _callback Callback.
 * @api private
 */
GoogleAuth.prototype._getSDKDefaultProjectId = function(_callback) ***REMOVED***
  exec('gcloud -q config list core/project --format=json', _callback);
***REMOVED***;

/**
 * Gets the Compute Engine project ID if it can be inferred.
 * Uses 169.254.169.254 for the metadata server to avoid request
 * latency from DNS lookup.
 * See https://cloud.google.com/compute/docs/metadata#metadataserver
 * for information about this IP address. (This IP is also used for
 * Amazon EC2 instances, so the metadata flavor is crucial.)
 * See https://github.com/google/oauth2client/issues/93 for context about
 * DNS latency.
 *
 * @param ***REMOVED***function***REMOVED*** _callback Callback.
 * @api private
 */
GoogleAuth.prototype._getGCEProjectId = function(_callback) ***REMOVED***
  if (!this.transporter) ***REMOVED***
    this.transporter = new DefaultTransporter();
  ***REMOVED***
  this.transporter.request(***REMOVED***
    method: 'GET',
    uri: 'http://169.254.169.254/computeMetadata/v1/project/project-id',
    headers: ***REMOVED***
      'Metadata-Flavor': 'Google'
    ***REMOVED***
  ***REMOVED***, function(err, body, res) ***REMOVED***
    if (err || !res || res.statusCode !== 200 || !body) ***REMOVED***
      callback(_callback, null);
      return;
    ***REMOVED***
    // Ignore any errors
    callback(_callback, null, body);
  ***REMOVED***);
***REMOVED***;

/**
 * Obtains the default service-level credentials for the application..
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 */
GoogleAuth.prototype.getApplicationDefault = function(opt_callback) ***REMOVED***
  var that = this;

  // If we've already got a cached credential, just return it.
  if (that._cachedCredential) ***REMOVED***
    process.nextTick(function() ***REMOVED***
      callback(opt_callback, null, that._cachedCredential, that._cachedProjectId);
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    // Inject our own callback routine, which will cache the credential once it's been created.
    // It also allows us to ensure that the ultimate callback is always async.
    var my_callback = function(err, result) ***REMOVED***
      if (!err && result) ***REMOVED***
        that._cachedCredential = result;
        that.getDefaultProjectId(function(err, projectId) ***REMOVED***
          process.nextTick(function() ***REMOVED***
            // Ignore default project error
            callback(opt_callback, null, result, projectId);
          ***REMOVED***);
        ***REMOVED***);
      ***REMOVED*** else ***REMOVED***
        process.nextTick(function() ***REMOVED***
          callback(opt_callback, err, result);
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***;
    // Check for the existence of a local environment variable pointing to the
    // location of the credential file. This is typically used in local developer scenarios.
    if (that._tryGetApplicationCredentialsFromEnvironmentVariable(my_callback)) ***REMOVED***
      return;
    ***REMOVED***

    // Look in the well-known credential file location.
    if (that._tryGetApplicationCredentialsFromWellKnownFile(my_callback)) ***REMOVED***
      return;
    ***REMOVED***

    // Determine if we're running on GCE.
    that._checkIsGCE(function(gce) ***REMOVED***
      if (gce) ***REMOVED***
        // For GCE, just return a default ComputeClient. It will take care of the rest.
        my_callback(null, new that.ComputeClient());
      ***REMOVED*** else ***REMOVED***
        // We failed to find the default credentials. Bail out with an error.
        my_callback(new Error('Could not load the default credentials. Browse to ' +
          'https://developers.google.com/accounts/docs/application-default-credentials for ' +
          'more information.'));
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Determines whether the auth layer is running on Google Compute Engine.
 * @param ***REMOVED***function=***REMOVED*** callback The callback.
 * @api private
 */
GoogleAuth.prototype._checkIsGCE = function(callback) ***REMOVED***
  var that = this;
  if (that._checked_isGCE) ***REMOVED***
    callback(that._isGCE);
  ***REMOVED*** else ***REMOVED***
    if (!that.transporter) ***REMOVED***
      that.transporter = new DefaultTransporter();
    ***REMOVED***
    that.transporter.request(***REMOVED***
      method: 'GET',
      uri: 'http://metadata.google.internal',
      json: true
    ***REMOVED***, function(err, body, res) ***REMOVED***
      if (!err && res && res.headers) ***REMOVED***
        that._isGCE = res.headers['metadata-flavor'] === 'Google';
      ***REMOVED***
      that._checked_isGCE = true;
      callback(that._isGCE);
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

/**
 * Attempts to load default credentials from the environment variable path..
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 * @return ***REMOVED***boolean***REMOVED*** Returns true if the callback has been executed; false otherwise.
 * @api private
 */
GoogleAuth.prototype._tryGetApplicationCredentialsFromEnvironmentVariable =
  function(opt_callback) ***REMOVED***

    var that = this;
    var credentialsPath = that._getEnv('GOOGLE_APPLICATION_CREDENTIALS');
    if (!credentialsPath || credentialsPath.length === 0) ***REMOVED***
      return false;
    ***REMOVED***
    that._getApplicationCredentialsFromFilePath(credentialsPath, function(err, result) ***REMOVED***
      var wrappedError = null;
      if (err) ***REMOVED***
        wrappedError = createError(
            'Unable to read the credential file specified by the GOOGLE_APPLICATION_CREDENTIALS ' +
            'environment variable.',
          err);
      ***REMOVED***
      callback(opt_callback, wrappedError, result);
    ***REMOVED***);
    return true;
  ***REMOVED***;

/**
 * Attempts to load default credentials from a well-known file location
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 * @return ***REMOVED***boolean***REMOVED*** Returns true if the callback has been executed; false otherwise.
 * @api private
 */
GoogleAuth.prototype._tryGetApplicationCredentialsFromWellKnownFile =
  function(opt_callback) ***REMOVED***

    var that = this;
    // First, figure out the location of the file, depending upon the OS type.
    var location = null;
    if (that._isWindows()) ***REMOVED***
      // Windows
      location = that._getEnv('APPDATA');
    ***REMOVED*** else ***REMOVED***
      // Linux or Mac
      var home = that._getEnv('HOME');
      if (home) ***REMOVED***
        location = that._pathJoin(home, '.config');
      ***REMOVED***
    ***REMOVED***
    // If we found the root path, expand it.
    if (location) ***REMOVED***
      location = that._pathJoin(location, 'gcloud');
      location = that._pathJoin(location, 'application_default_credentials.json');
      location = that._mockWellKnownFilePath(location);
      // Check whether the file exists.
      if (!that._fileExists(location)) ***REMOVED***
        location = null;
      ***REMOVED***
    ***REMOVED***
    // The file does not exist.
    if (!location) ***REMOVED***
      return false;
    ***REMOVED***
    // The file seems to exist. Try to use it.
    this._getApplicationCredentialsFromFilePath(location, opt_callback);
    return true;
  ***REMOVED***;

/**
 * Attempts to load default credentials from a file at the given path..
 * @param ***REMOVED***string=***REMOVED*** filePath The path to the file to read.
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 * @api private
 */
GoogleAuth.prototype._getApplicationCredentialsFromFilePath =
  function(filePath, opt_callback) ***REMOVED***

    var that = this;
    var error = null;
    // Make sure the path looks like a string.
    if (!filePath || filePath.length === 0) ***REMOVED***
      error = new Error('The file path is invalid.');
    ***REMOVED***

    // Make sure there is a file at the path. lstatSync will throw if there is nothing there.
    if (!error) ***REMOVED***
      try ***REMOVED***
        // Resolve path to actual file in case of symlink. Expect a thrown error if not resolvable.
        filePath = fs.realpathSync(filePath);

        if (!fs.lstatSync(filePath).isFile()) ***REMOVED***
          throw '';
        ***REMOVED***
      ***REMOVED*** catch (err) ***REMOVED***
        error = createError(util.format('The file at %s does not exist, or it is not a file.',
          filePath), err);
      ***REMOVED***
    ***REMOVED***
    // Now open a read stream on the file, and parse it.
    if (!error) ***REMOVED***
      try ***REMOVED***
        var stream = that._createReadStream(filePath);
        that.fromStream(stream, opt_callback);
      ***REMOVED*** catch (err) ***REMOVED***
        error = createError(util.format('Unable to read the file at %s.', filePath), err);
      ***REMOVED***
    ***REMOVED***
    if (error) ***REMOVED***
      callback(opt_callback, error);
    ***REMOVED***
  ***REMOVED***;

/**
 * Create a credentials instance using the given input options.
 * @param ***REMOVED***object=***REMOVED*** json The input object.
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 */
GoogleAuth.prototype.fromJSON = function(json, opt_callback) ***REMOVED***
  var that = this;
  var client;
  if (!json) ***REMOVED***
    callback(opt_callback, new Error(
      'Must pass in a JSON object containing the Google auth settings.'));
    return;
  ***REMOVED***
  if (json.type === 'authorized_user') ***REMOVED***
    client = new that.UserRefreshClient();
  ***REMOVED*** else ***REMOVED***
    client = new that.JWTClient();
  ***REMOVED***
  client.fromJSON(json, function(err) ***REMOVED***
    if (err) ***REMOVED***
      callback(opt_callback, err);
    ***REMOVED*** else ***REMOVED***
      callback(opt_callback, null, client);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;

/**
 * Create a credentials instance using the given input stream.
 * @param ***REMOVED***object=***REMOVED*** stream The input stream.
 * @param ***REMOVED***function=***REMOVED*** opt_callback Optional callback.
 */
GoogleAuth.prototype.fromStream = function(stream, opt_callback) ***REMOVED***
  var that = this;
  if (!stream) ***REMOVED***
    process.nextTick(function() ***REMOVED***
      callback(opt_callback, new Error(
          'Must pass in a stream containing the Google auth settings.'));
    ***REMOVED***);
    return;
  ***REMOVED***
  var s = '';
  stream.setEncoding('utf8');
  stream.on('data', function (chunk) ***REMOVED***
    s += chunk;
  ***REMOVED***);
  stream.on('end', function () ***REMOVED***
    try ***REMOVED***
      var data = JSON.parse(s);
      that.fromJSON(data, opt_callback);
    ***REMOVED*** catch (err) ***REMOVED***
      callback(opt_callback, err);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;

/**
 * Determines whether the current operating system is Windows.
 * @api private
 * */
GoogleAuth.prototype._isWindows = function() ***REMOVED***
  var sys = this._osPlatform();
  if (sys && sys.length >= 3) ***REMOVED***
    if (sys.substring(0, 3).toLowerCase() === 'win') ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***
  return false;
***REMOVED***;

/**
 * Creates a file stream. Allows mocking.
 * @api private
 * */
GoogleAuth.prototype._createReadStream = function(filePath) ***REMOVED***
  return fs.createReadStream(filePath);
***REMOVED***;

/**
 * Gets the value of the environment variable with the given name. Allows mocking.
 * @api private
 * */
GoogleAuth.prototype._getEnv = function(name) ***REMOVED***
  return process.env[name];
***REMOVED***;

/**
 * Gets the current operating system platform. Allows mocking.
 * @api private
 * */
GoogleAuth.prototype._osPlatform = function() ***REMOVED***
  return os.platform();
***REMOVED***;

/**
 * Determines whether a file exists. Allows mocking.
 * @api private
 * */
GoogleAuth.prototype._fileExists = function(filePath) ***REMOVED***
  return fs.existsSync(filePath);
***REMOVED***;

/**
 * Joins two parts of a path. Allows mocking.
 * @api private
 * */
GoogleAuth.prototype._pathJoin = function(item1, item2) ***REMOVED***
  return path.join(item1, item2);
***REMOVED***;

/**
 * Allows mocking of the path to a well-known file.
 * @api private
 * */
GoogleAuth.prototype._mockWellKnownFilePath = function(filePath) ***REMOVED***
  return filePath;
***REMOVED***;

/**
 * Export GoogleAuth.
 */
module.exports = GoogleAuth;
