// google-auth-library/lib/auth/loginticket.js
/**
 * Copyright 2014 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var USER_ATTR = 'sub';

/**
 * Create a simple class to extract user ID from an ID Token
 *
 * @param ***REMOVED***string***REMOVED*** env Envelope of the jwt
 * @param ***REMOVED***string***REMOVED*** pay Payload of the jwt
 * @constructor
 */
function LoginTicket(env, pay) ***REMOVED***
  var envelope = env;
  var payload = pay;

  this.getEnvelope = function() ***REMOVED***
    return envelope;
  ***REMOVED***;

  this.getPayload = function() ***REMOVED***
    return payload;
  ***REMOVED***;
***REMOVED***

/**
 * Create a simple class to extract user ID from an ID Token
 *
 * @return ***REMOVED***string***REMOVED*** The user ID
 */
LoginTicket.prototype.getUserId = function() ***REMOVED***
  var payload = this.getPayload();
  if (payload && payload[USER_ATTR]) ***REMOVED***
    return payload[USER_ATTR];
  ***REMOVED***

  return null;
***REMOVED***;

/**
 * Returns attributes from the login ticket.  This can contain
 * various information about the user session.
 *
 * @return ***REMOVED***Object***REMOVED*** The envelope and payload
 */
LoginTicket.prototype.getAttributes = function() ***REMOVED***
  return ***REMOVED***
    'envelope': this.getEnvelope(),
    'payload': this.getPayload()
  ***REMOVED***;
***REMOVED***;

/**
 * Export LoginTicket.
 */
module.exports = LoginTicket;
