// fstream/lib/dir-writer.js
// It is expected that, when .add() returns false, the consumer
// of the DirWriter will pause until a "drain" event occurs. Note
// that this is *almost always going to be the case*, unless the
// thing being written is some sort of unsupported type, and thus
// skipped over.

module.exports = DirWriter

var Writer = require('./writer.js')
var inherits = require('inherits')
var mkdir = require('mkdirp')
var path = require('path')
var collect = require('./collect.js')

inherits(DirWriter, Writer)

function DirWriter (props) ***REMOVED***
  var self = this
  if (!(self instanceof DirWriter)) ***REMOVED***
    self.error('DirWriter must be called as constructor.', null, true)
  ***REMOVED***

  // should already be established as a Directory type
  if (props.type !== 'Directory' || !props.Directory) ***REMOVED***
    self.error('Non-directory type ' + props.type + ' ' +
      JSON.stringify(props), null, true)
  ***REMOVED***

  Writer.call(this, props)
***REMOVED***

DirWriter.prototype._create = function () ***REMOVED***
  var self = this
  mkdir(self._path, Writer.dirmode, function (er) ***REMOVED***
    if (er) return self.error(er)
    // ready to start getting entries!
    self.ready = true
    self.emit('ready')
    self._process()
  ***REMOVED***)
***REMOVED***

// a DirWriter has an add(entry) method, but its .write() doesn't
// do anything.  Why a no-op rather than a throw?  Because this
// leaves open the door for writing directory metadata for
// gnu/solaris style dumpdirs.
DirWriter.prototype.write = function () ***REMOVED***
  return true
***REMOVED***

DirWriter.prototype.end = function () ***REMOVED***
  this._ended = true
  this._process()
***REMOVED***

DirWriter.prototype.add = function (entry) ***REMOVED***
  var self = this

  // console.error('\tadd', entry._path, '->', self._path)
  collect(entry)
  if (!self.ready || self._currentEntry) ***REMOVED***
    self._buffer.push(entry)
    return false
  ***REMOVED***

  // create a new writer, and pipe the incoming entry into it.
  if (self._ended) ***REMOVED***
    return self.error('add after end')
  ***REMOVED***

  self._buffer.push(entry)
  self._process()

  return this._buffer.length === 0
***REMOVED***

DirWriter.prototype._process = function () ***REMOVED***
  var self = this

  // console.error('DW Process p=%j', self._processing, self.basename)

  if (self._processing) return

  var entry = self._buffer.shift()
  if (!entry) ***REMOVED***
    // console.error("DW Drain")
    self.emit('drain')
    if (self._ended) self._finish()
    return
  ***REMOVED***

  self._processing = true
  // console.error("DW Entry", entry._path)

  self.emit('entry', entry)

  // ok, add this entry
  //
  // don't allow recursive copying
  var p = entry
  var pp
  do ***REMOVED***
    pp = p._path || p.path
    if (pp === self.root._path || pp === self._path ||
      (pp && pp.indexOf(self._path) === 0)) ***REMOVED***
      // console.error('DW Exit (recursive)', entry.basename, self._path)
      self._processing = false
      if (entry._collected) entry.pipe()
      return self._process()
    ***REMOVED***
    p = p.parent
  ***REMOVED*** while (p)

  // console.error("DW not recursive")

  // chop off the entry's root dir, replace with ours
  var props = ***REMOVED***
    parent: self,
    root: self.root || self,
    type: entry.type,
    depth: self.depth + 1
  ***REMOVED***

  pp = entry._path || entry.path || entry.props.path
  if (entry.parent) ***REMOVED***
    pp = pp.substr(entry.parent._path.length + 1)
  ***REMOVED***
  // get rid of any ../../ shenanigans
  props.path = path.join(self.path, path.join('/', pp))

  // if i have a filter, the child should inherit it.
  props.filter = self.filter

  // all the rest of the stuff, copy over from the source.
  Object.keys(entry.props).forEach(function (k) ***REMOVED***
    if (!props.hasOwnProperty(k)) ***REMOVED***
      props[k] = entry.props[k]
    ***REMOVED***
  ***REMOVED***)

  // not sure at this point what kind of writer this is.
  var child = self._currentChild = new Writer(props)
  child.on('ready', function () ***REMOVED***
    // console.error("DW Child Ready", child.type, child._path)
    // console.error("  resuming", entry._path)
    entry.pipe(child)
    entry.resume()
  ***REMOVED***)

  // XXX Make this work in node.
  // Long filenames should not break stuff.
  child.on('error', function (er) ***REMOVED***
    if (child._swallowErrors) ***REMOVED***
      self.warn(er)
      child.emit('end')
      child.emit('close')
    ***REMOVED*** else ***REMOVED***
      self.emit('error', er)
    ***REMOVED***
  ***REMOVED***)

  // we fire _end internally *after* end, so that we don't move on
  // until any "end" listeners have had their chance to do stuff.
  child.on('close', onend)
  var ended = false
  function onend () ***REMOVED***
    if (ended) return
    ended = true
    // console.error("* DW Child end", child.basename)
    self._currentChild = null
    self._processing = false
    self._process()
  ***REMOVED***
***REMOVED***
