// asn1.js/lib/asn1/base/buffer.js
var inherits = require('inherits');
var Reporter = require('../base').Reporter;
var Buffer = require('buffer').Buffer;

function DecoderBuffer(base, options) ***REMOVED***
  Reporter.call(this, options);
  if (!Buffer.isBuffer(base)) ***REMOVED***
    this.error('Input not Buffer');
    return;
  ***REMOVED***

  this.base = base;
  this.offset = 0;
  this.length = base.length;
***REMOVED***
inherits(DecoderBuffer, Reporter);
exports.DecoderBuffer = DecoderBuffer;

DecoderBuffer.prototype.save = function save() ***REMOVED***
  return ***REMOVED*** offset: this.offset, reporter: Reporter.prototype.save.call(this) ***REMOVED***;
***REMOVED***;

DecoderBuffer.prototype.restore = function restore(save) ***REMOVED***
  // Return skipped data
  var res = new DecoderBuffer(this.base);
  res.offset = save.offset;
  res.length = this.offset;

  this.offset = save.offset;
  Reporter.prototype.restore.call(this, save.reporter);

  return res;
***REMOVED***;

DecoderBuffer.prototype.isEmpty = function isEmpty() ***REMOVED***
  return this.offset === this.length;
***REMOVED***;

DecoderBuffer.prototype.readUInt8 = function readUInt8(fail) ***REMOVED***
  if (this.offset + 1 <= this.length)
    return this.base.readUInt8(this.offset++, true);
  else
    return this.error(fail || 'DecoderBuffer overrun');
***REMOVED***

DecoderBuffer.prototype.skip = function skip(bytes, fail) ***REMOVED***
  if (!(this.offset + bytes <= this.length))
    return this.error(fail || 'DecoderBuffer overrun');

  var res = new DecoderBuffer(this.base);

  // Share reporter state
  res._reporterState = this._reporterState;

  res.offset = this.offset;
  res.length = this.offset + bytes;
  this.offset += bytes;
  return res;
***REMOVED***

DecoderBuffer.prototype.raw = function raw(save) ***REMOVED***
  return this.base.slice(save ? save.offset : this.offset, this.length);
***REMOVED***

function EncoderBuffer(value, reporter) ***REMOVED***
  if (Array.isArray(value)) ***REMOVED***
    this.length = 0;
    this.value = value.map(function(item) ***REMOVED***
      if (!(item instanceof EncoderBuffer))
        item = new EncoderBuffer(item, reporter);
      this.length += item.length;
      return item;
    ***REMOVED***, this);
  ***REMOVED*** else if (typeof value === 'number') ***REMOVED***
    if (!(0 <= value && value <= 0xff))
      return reporter.error('non-byte EncoderBuffer value');
    this.value = value;
    this.length = 1;
  ***REMOVED*** else if (typeof value === 'string') ***REMOVED***
    this.value = value;
    this.length = Buffer.byteLength(value);
  ***REMOVED*** else if (Buffer.isBuffer(value)) ***REMOVED***
    this.value = value;
    this.length = value.length;
  ***REMOVED*** else ***REMOVED***
    return reporter.error('Unsupported type: ' + typeof value);
  ***REMOVED***
***REMOVED***
exports.EncoderBuffer = EncoderBuffer;

EncoderBuffer.prototype.join = function join(out, offset) ***REMOVED***
  if (!out)
    out = new Buffer(this.length);
  if (!offset)
    offset = 0;

  if (this.length === 0)
    return out;

  if (Array.isArray(this.value)) ***REMOVED***
    this.value.forEach(function(item) ***REMOVED***
      item.join(out, offset);
      offset += item.length;
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
    if (typeof this.value === 'number')
      out[offset] = this.value;
    else if (typeof this.value === 'string')
      out.write(this.value, offset);
    else if (Buffer.isBuffer(this.value))
      this.value.copy(out, offset);
    offset += this.length;
  ***REMOVED***

  return out;
***REMOVED***;
