// asn1.js/lib/asn1/base/node.js
var Reporter = require('../base').Reporter;
var EncoderBuffer = require('../base').EncoderBuffer;
var DecoderBuffer = require('../base').DecoderBuffer;
var assert = require('minimalistic-assert');

// Supported tags
var tags = [
  'seq', 'seqof', 'set', 'setof', 'objid', 'bool',
  'gentime', 'utctime', 'null_', 'enum', 'int', 'objDesc',
  'bitstr', 'bmpstr', 'charstr', 'genstr', 'graphstr', 'ia5str', 'iso646str',
  'numstr', 'octstr', 'printstr', 't61str', 'unistr', 'utf8str', 'videostr'
];

// Public methods list
var methods = [
  'key', 'obj', 'use', 'optional', 'explicit', 'implicit', 'def', 'choice',
  'any', 'contains'
].concat(tags);

// Overrided methods list
var overrided = [
  '_peekTag', '_decodeTag', '_use',
  '_decodeStr', '_decodeObjid', '_decodeTime',
  '_decodeNull', '_decodeInt', '_decodeBool', '_decodeList',

  '_encodeComposite', '_encodeStr', '_encodeObjid', '_encodeTime',
  '_encodeNull', '_encodeInt', '_encodeBool'
];

function Node(enc, parent) ***REMOVED***
  var state = ***REMOVED******REMOVED***;
  this._baseState = state;

  state.enc = enc;

  state.parent = parent || null;
  state.children = null;

  // State
  state.tag = null;
  state.args = null;
  state.reverseArgs = null;
  state.choice = null;
  state.optional = false;
  state.any = false;
  state.obj = false;
  state.use = null;
  state.useDecoder = null;
  state.key = null;
  state['default'] = null;
  state.explicit = null;
  state.implicit = null;
  state.contains = null;

  // Should create new instance on each method
  if (!state.parent) ***REMOVED***
    state.children = [];
    this._wrap();
  ***REMOVED***
***REMOVED***
module.exports = Node;

var stateProps = [
  'enc', 'parent', 'children', 'tag', 'args', 'reverseArgs', 'choice',
  'optional', 'any', 'obj', 'use', 'alteredUse', 'key', 'default', 'explicit',
  'implicit', 'contains'
];

Node.prototype.clone = function clone() ***REMOVED***
  var state = this._baseState;
  var cstate = ***REMOVED******REMOVED***;
  stateProps.forEach(function(prop) ***REMOVED***
    cstate[prop] = state[prop];
  ***REMOVED***);
  var res = new this.constructor(cstate.parent);
  res._baseState = cstate;
  return res;
***REMOVED***;

Node.prototype._wrap = function wrap() ***REMOVED***
  var state = this._baseState;
  methods.forEach(function(method) ***REMOVED***
    this[method] = function _wrappedMethod() ***REMOVED***
      var clone = new this.constructor(this);
      state.children.push(clone);
      return clone[method].apply(clone, arguments);
    ***REMOVED***;
  ***REMOVED***, this);
***REMOVED***;

Node.prototype._init = function init(body) ***REMOVED***
  var state = this._baseState;

  assert(state.parent === null);
  body.call(this);

  // Filter children
  state.children = state.children.filter(function(child) ***REMOVED***
    return child._baseState.parent === this;
  ***REMOVED***, this);
  assert.equal(state.children.length, 1, 'Root node can have only one child');
***REMOVED***;

Node.prototype._useArgs = function useArgs(args) ***REMOVED***
  var state = this._baseState;

  // Filter children and args
  var children = args.filter(function(arg) ***REMOVED***
    return arg instanceof this.constructor;
  ***REMOVED***, this);
  args = args.filter(function(arg) ***REMOVED***
    return !(arg instanceof this.constructor);
  ***REMOVED***, this);

  if (children.length !== 0) ***REMOVED***
    assert(state.children === null);
    state.children = children;

    // Replace parent to maintain backward link
    children.forEach(function(child) ***REMOVED***
      child._baseState.parent = this;
    ***REMOVED***, this);
  ***REMOVED***
  if (args.length !== 0) ***REMOVED***
    assert(state.args === null);
    state.args = args;
    state.reverseArgs = args.map(function(arg) ***REMOVED***
      if (typeof arg !== 'object' || arg.constructor !== Object)
        return arg;

      var res = ***REMOVED******REMOVED***;
      Object.keys(arg).forEach(function(key) ***REMOVED***
        if (key == (key | 0))
          key |= 0;
        var value = arg[key];
        res[value] = key;
      ***REMOVED***);
      return res;
    ***REMOVED***);
  ***REMOVED***
***REMOVED***;

//
// Overrided methods
//

overrided.forEach(function(method) ***REMOVED***
  Node.prototype[method] = function _overrided() ***REMOVED***
    var state = this._baseState;
    throw new Error(method + ' not implemented for encoding: ' + state.enc);
  ***REMOVED***;
***REMOVED***);

//
// Public methods
//

tags.forEach(function(tag) ***REMOVED***
  Node.prototype[tag] = function _tagMethod() ***REMOVED***
    var state = this._baseState;
    var args = Array.prototype.slice.call(arguments);

    assert(state.tag === null);
    state.tag = tag;

    this._useArgs(args);

    return this;
  ***REMOVED***;
***REMOVED***);

Node.prototype.use = function use(item) ***REMOVED***
  assert(item);
  var state = this._baseState;

  assert(state.use === null);
  state.use = item;

  return this;
***REMOVED***;

Node.prototype.optional = function optional() ***REMOVED***
  var state = this._baseState;

  state.optional = true;

  return this;
***REMOVED***;

Node.prototype.def = function def(val) ***REMOVED***
  var state = this._baseState;

  assert(state['default'] === null);
  state['default'] = val;
  state.optional = true;

  return this;
***REMOVED***;

Node.prototype.explicit = function explicit(num) ***REMOVED***
  var state = this._baseState;

  assert(state.explicit === null && state.implicit === null);
  state.explicit = num;

  return this;
***REMOVED***;

Node.prototype.implicit = function implicit(num) ***REMOVED***
  var state = this._baseState;

  assert(state.explicit === null && state.implicit === null);
  state.implicit = num;

  return this;
***REMOVED***;

Node.prototype.obj = function obj() ***REMOVED***
  var state = this._baseState;
  var args = Array.prototype.slice.call(arguments);

  state.obj = true;

  if (args.length !== 0)
    this._useArgs(args);

  return this;
***REMOVED***;

Node.prototype.key = function key(newKey) ***REMOVED***
  var state = this._baseState;

  assert(state.key === null);
  state.key = newKey;

  return this;
***REMOVED***;

Node.prototype.any = function any() ***REMOVED***
  var state = this._baseState;

  state.any = true;

  return this;
***REMOVED***;

Node.prototype.choice = function choice(obj) ***REMOVED***
  var state = this._baseState;

  assert(state.choice === null);
  state.choice = obj;
  this._useArgs(Object.keys(obj).map(function(key) ***REMOVED***
    return obj[key];
  ***REMOVED***));

  return this;
***REMOVED***;

Node.prototype.contains = function contains(item) ***REMOVED***
  var state = this._baseState;

  assert(state.use === null);
  state.contains = item;

  return this;
***REMOVED***;

//
// Decoding
//

Node.prototype._decode = function decode(input, options) ***REMOVED***
  var state = this._baseState;

  // Decode root node
  if (state.parent === null)
    return input.wrapResult(state.children[0]._decode(input, options));

  var result = state['default'];
  var present = true;

  var prevKey = null;
  if (state.key !== null)
    prevKey = input.enterKey(state.key);

  // Check if tag is there
  if (state.optional) ***REMOVED***
    var tag = null;
    if (state.explicit !== null)
      tag = state.explicit;
    else if (state.implicit !== null)
      tag = state.implicit;
    else if (state.tag !== null)
      tag = state.tag;

    if (tag === null && !state.any) ***REMOVED***
      // Trial and Error
      var save = input.save();
      try ***REMOVED***
        if (state.choice === null)
          this._decodeGeneric(state.tag, input, options);
        else
          this._decodeChoice(input, options);
        present = true;
      ***REMOVED*** catch (e) ***REMOVED***
        present = false;
      ***REMOVED***
      input.restore(save);
    ***REMOVED*** else ***REMOVED***
      present = this._peekTag(input, tag, state.any);

      if (input.isError(present))
        return present;
    ***REMOVED***
  ***REMOVED***

  // Push object on stack
  var prevObj;
  if (state.obj && present)
    prevObj = input.enterObject();

  if (present) ***REMOVED***
    // Unwrap explicit values
    if (state.explicit !== null) ***REMOVED***
      var explicit = this._decodeTag(input, state.explicit);
      if (input.isError(explicit))
        return explicit;
      input = explicit;
    ***REMOVED***

    var start = input.offset;

    // Unwrap implicit and normal values
    if (state.use === null && state.choice === null) ***REMOVED***
      if (state.any)
        var save = input.save();
      var body = this._decodeTag(
        input,
        state.implicit !== null ? state.implicit : state.tag,
        state.any
      );
      if (input.isError(body))
        return body;

      if (state.any)
        result = input.raw(save);
      else
        input = body;
    ***REMOVED***

    if (options && options.track && state.tag !== null)
      options.track(input.path(), start, input.length, 'tagged');

    if (options && options.track && state.tag !== null)
      options.track(input.path(), input.offset, input.length, 'content');

    // Select proper method for tag
    if (state.any)
      result = result;
    else if (state.choice === null)
      result = this._decodeGeneric(state.tag, input, options);
    else
      result = this._decodeChoice(input, options);

    if (input.isError(result))
      return result;

    // Decode children
    if (!state.any && state.choice === null && state.children !== null) ***REMOVED***
      state.children.forEach(function decodeChildren(child) ***REMOVED***
        // NOTE: We are ignoring errors here, to let parser continue with other
        // parts of encoded data
        child._decode(input, options);
      ***REMOVED***);
    ***REMOVED***

    // Decode contained/encoded by schema, only in bit or octet strings
    if (state.contains && (state.tag === 'octstr' || state.tag === 'bitstr')) ***REMOVED***
      var data = new DecoderBuffer(result);
      result = this._getUse(state.contains, input._reporterState.obj)
          ._decode(data, options);
    ***REMOVED***
  ***REMOVED***

  // Pop object
  if (state.obj && present)
    result = input.leaveObject(prevObj);

  // Set key
  if (state.key !== null && (result !== null || present === true))
    input.leaveKey(prevKey, state.key, result);
  else if (prevKey !== null)
    input.exitKey(prevKey);

  return result;
***REMOVED***;

Node.prototype._decodeGeneric = function decodeGeneric(tag, input, options) ***REMOVED***
  var state = this._baseState;

  if (tag === 'seq' || tag === 'set')
    return null;
  if (tag === 'seqof' || tag === 'setof')
    return this._decodeList(input, tag, state.args[0], options);
  else if (/str$/.test(tag))
    return this._decodeStr(input, tag, options);
  else if (tag === 'objid' && state.args)
    return this._decodeObjid(input, state.args[0], state.args[1], options);
  else if (tag === 'objid')
    return this._decodeObjid(input, null, null, options);
  else if (tag === 'gentime' || tag === 'utctime')
    return this._decodeTime(input, tag, options);
  else if (tag === 'null_')
    return this._decodeNull(input, options);
  else if (tag === 'bool')
    return this._decodeBool(input, options);
  else if (tag === 'objDesc')
    return this._decodeStr(input, tag, options);
  else if (tag === 'int' || tag === 'enum')
    return this._decodeInt(input, state.args && state.args[0], options);

  if (state.use !== null) ***REMOVED***
    return this._getUse(state.use, input._reporterState.obj)
        ._decode(input, options);
  ***REMOVED*** else ***REMOVED***
    return input.error('unknown tag: ' + tag);
  ***REMOVED***
***REMOVED***;

Node.prototype._getUse = function _getUse(entity, obj) ***REMOVED***

  var state = this._baseState;
  // Create altered use decoder if implicit is set
  state.useDecoder = this._use(entity, obj);
  assert(state.useDecoder._baseState.parent === null);
  state.useDecoder = state.useDecoder._baseState.children[0];
  if (state.implicit !== state.useDecoder._baseState.implicit) ***REMOVED***
    state.useDecoder = state.useDecoder.clone();
    state.useDecoder._baseState.implicit = state.implicit;
  ***REMOVED***
  return state.useDecoder;
***REMOVED***;

Node.prototype._decodeChoice = function decodeChoice(input, options) ***REMOVED***
  var state = this._baseState;
  var result = null;
  var match = false;

  Object.keys(state.choice).some(function(key) ***REMOVED***
    var save = input.save();
    var node = state.choice[key];
    try ***REMOVED***
      var value = node._decode(input, options);
      if (input.isError(value))
        return false;

      result = ***REMOVED*** type: key, value: value ***REMOVED***;
      match = true;
    ***REMOVED*** catch (e) ***REMOVED***
      input.restore(save);
      return false;
    ***REMOVED***
    return true;
  ***REMOVED***, this);

  if (!match)
    return input.error('Choice not matched');

  return result;
***REMOVED***;

//
// Encoding
//

Node.prototype._createEncoderBuffer = function createEncoderBuffer(data) ***REMOVED***
  return new EncoderBuffer(data, this.reporter);
***REMOVED***;

Node.prototype._encode = function encode(data, reporter, parent) ***REMOVED***
  var state = this._baseState;
  if (state['default'] !== null && state['default'] === data)
    return;

  var result = this._encodeValue(data, reporter, parent);
  if (result === undefined)
    return;

  if (this._skipDefault(result, reporter, parent))
    return;

  return result;
***REMOVED***;

Node.prototype._encodeValue = function encode(data, reporter, parent) ***REMOVED***
  var state = this._baseState;

  // Decode root node
  if (state.parent === null)
    return state.children[0]._encode(data, reporter || new Reporter());

  var result = null;

  // Set reporter to share it with a child class
  this.reporter = reporter;

  // Check if data is there
  if (state.optional && data === undefined) ***REMOVED***
    if (state['default'] !== null)
      data = state['default']
    else
      return;
  ***REMOVED***

  // Encode children first
  var content = null;
  var primitive = false;
  if (state.any) ***REMOVED***
    // Anything that was given is translated to buffer
    result = this._createEncoderBuffer(data);
  ***REMOVED*** else if (state.choice) ***REMOVED***
    result = this._encodeChoice(data, reporter);
  ***REMOVED*** else if (state.contains) ***REMOVED***
    content = this._getUse(state.contains, parent)._encode(data, reporter);
    primitive = true;
  ***REMOVED*** else if (state.children) ***REMOVED***
    content = state.children.map(function(child) ***REMOVED***
      if (child._baseState.tag === 'null_')
        return child._encode(null, reporter, data);

      if (child._baseState.key === null)
        return reporter.error('Child should have a key');
      var prevKey = reporter.enterKey(child._baseState.key);

      if (typeof data !== 'object')
        return reporter.error('Child expected, but input is not object');

      var res = child._encode(data[child._baseState.key], reporter, data);
      reporter.leaveKey(prevKey);

      return res;
    ***REMOVED***, this).filter(function(child) ***REMOVED***
      return child;
    ***REMOVED***);
    content = this._createEncoderBuffer(content);
  ***REMOVED*** else ***REMOVED***
    if (state.tag === 'seqof' || state.tag === 'setof') ***REMOVED***
      // TODO(indutny): this should be thrown on DSL level
      if (!(state.args && state.args.length === 1))
        return reporter.error('Too many args for : ' + state.tag);

      if (!Array.isArray(data))
        return reporter.error('seqof/setof, but data is not Array');

      var child = this.clone();
      child._baseState.implicit = null;
      content = this._createEncoderBuffer(data.map(function(item) ***REMOVED***
        var state = this._baseState;

        return this._getUse(state.args[0], data)._encode(item, reporter);
      ***REMOVED***, child));
    ***REMOVED*** else if (state.use !== null) ***REMOVED***
      result = this._getUse(state.use, parent)._encode(data, reporter);
    ***REMOVED*** else ***REMOVED***
      content = this._encodePrimitive(state.tag, data);
      primitive = true;
    ***REMOVED***
  ***REMOVED***

  // Encode data itself
  var result;
  if (!state.any && state.choice === null) ***REMOVED***
    var tag = state.implicit !== null ? state.implicit : state.tag;
    var cls = state.implicit === null ? 'universal' : 'context';

    if (tag === null) ***REMOVED***
      if (state.use === null)
        reporter.error('Tag could be omitted only for .use()');
    ***REMOVED*** else ***REMOVED***
      if (state.use === null)
        result = this._encodeComposite(tag, primitive, cls, content);
    ***REMOVED***
  ***REMOVED***

  // Wrap in explicit
  if (state.explicit !== null)
    result = this._encodeComposite(state.explicit, false, 'context', result);

  return result;
***REMOVED***;

Node.prototype._encodeChoice = function encodeChoice(data, reporter) ***REMOVED***
  var state = this._baseState;

  var node = state.choice[data.type];
  if (!node) ***REMOVED***
    assert(
        false,
        data.type + ' not found in ' +
            JSON.stringify(Object.keys(state.choice)));
  ***REMOVED***
  return node._encode(data.value, reporter);
***REMOVED***;

Node.prototype._encodePrimitive = function encodePrimitive(tag, data) ***REMOVED***
  var state = this._baseState;

  if (/str$/.test(tag))
    return this._encodeStr(data, tag);
  else if (tag === 'objid' && state.args)
    return this._encodeObjid(data, state.reverseArgs[0], state.args[1]);
  else if (tag === 'objid')
    return this._encodeObjid(data, null, null);
  else if (tag === 'gentime' || tag === 'utctime')
    return this._encodeTime(data, tag);
  else if (tag === 'null_')
    return this._encodeNull();
  else if (tag === 'int' || tag === 'enum')
    return this._encodeInt(data, state.args && state.reverseArgs[0]);
  else if (tag === 'bool')
    return this._encodeBool(data);
  else if (tag === 'objDesc')
    return this._encodeStr(data, tag);
  else
    throw new Error('Unsupported tag: ' + tag);
***REMOVED***;

Node.prototype._isNumstr = function isNumstr(str) ***REMOVED***
  return /^[0-9 ]*$/.test(str);
***REMOVED***;

Node.prototype._isPrintstr = function isPrintstr(str) ***REMOVED***
  return /^[A-Za-z0-9 '\(\)\+,\-\.\/:=\?]*$/.test(str);
***REMOVED***;
