// asn1.js/lib/asn1/base/reporter.js
var inherits = require('inherits');

function Reporter(options) ***REMOVED***
  this._reporterState = ***REMOVED***
    obj: null,
    path: [],
    options: options || ***REMOVED******REMOVED***,
    errors: []
  ***REMOVED***;
***REMOVED***
exports.Reporter = Reporter;

Reporter.prototype.isError = function isError(obj) ***REMOVED***
  return obj instanceof ReporterError;
***REMOVED***;

Reporter.prototype.save = function save() ***REMOVED***
  var state = this._reporterState;

  return ***REMOVED*** obj: state.obj, pathLen: state.path.length ***REMOVED***;
***REMOVED***;

Reporter.prototype.restore = function restore(data) ***REMOVED***
  var state = this._reporterState;

  state.obj = data.obj;
  state.path = state.path.slice(0, data.pathLen);
***REMOVED***;

Reporter.prototype.enterKey = function enterKey(key) ***REMOVED***
  return this._reporterState.path.push(key);
***REMOVED***;

Reporter.prototype.exitKey = function exitKey(index) ***REMOVED***
  var state = this._reporterState;

  state.path = state.path.slice(0, index - 1);
***REMOVED***;

Reporter.prototype.leaveKey = function leaveKey(index, key, value) ***REMOVED***
  var state = this._reporterState;

  this.exitKey(index);
  if (state.obj !== null)
    state.obj[key] = value;
***REMOVED***;

Reporter.prototype.path = function path() ***REMOVED***
  return this._reporterState.path.join('/');
***REMOVED***;

Reporter.prototype.enterObject = function enterObject() ***REMOVED***
  var state = this._reporterState;

  var prev = state.obj;
  state.obj = ***REMOVED******REMOVED***;
  return prev;
***REMOVED***;

Reporter.prototype.leaveObject = function leaveObject(prev) ***REMOVED***
  var state = this._reporterState;

  var now = state.obj;
  state.obj = prev;
  return now;
***REMOVED***;

Reporter.prototype.error = function error(msg) ***REMOVED***
  var err;
  var state = this._reporterState;

  var inherited = msg instanceof ReporterError;
  if (inherited) ***REMOVED***
    err = msg;
  ***REMOVED*** else ***REMOVED***
    err = new ReporterError(state.path.map(function(elem) ***REMOVED***
      return '[' + JSON.stringify(elem) + ']';
    ***REMOVED***).join(''), msg.message || msg, msg.stack);
  ***REMOVED***

  if (!state.options.partial)
    throw err;

  if (!inherited)
    state.errors.push(err);

  return err;
***REMOVED***;

Reporter.prototype.wrapResult = function wrapResult(result) ***REMOVED***
  var state = this._reporterState;
  if (!state.options.partial)
    return result;

  return ***REMOVED***
    result: this.isError(result) ? null : result,
    errors: state.errors
  ***REMOVED***;
***REMOVED***;

function ReporterError(path, msg) ***REMOVED***
  this.path = path;
  this.rethrow(msg);
***REMOVED***;
inherits(ReporterError, Error);

ReporterError.prototype.rethrow = function rethrow(msg) ***REMOVED***
  this.message = msg + ' at: ' + (this.path || '(shallow)');
  if (Error.captureStackTrace)
    Error.captureStackTrace(this, ReporterError);

  if (!this.stack) ***REMOVED***
    try ***REMOVED***
      // IE only adds stack when thrown
      throw new Error(this.message);
    ***REMOVED*** catch (e) ***REMOVED***
      this.stack = e.stack;
    ***REMOVED***
  ***REMOVED***
  return this;
***REMOVED***;
