// har-validator/lib/node4/promise.js
'use strict';

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
***REMOVED***);
exports.validate = validate;
exports.afterRequest = afterRequest;
exports.beforeRequest = beforeRequest;
exports.browser = browser;
exports.cache = cache;
exports.content = content;
exports.cookie = cookie;
exports.creator = creator;
exports.entry = entry;
exports.har = har;
exports.header = header;
exports.log = log;
exports.page = page;
exports.pageTimings = pageTimings;
exports.postData = postData;
exports.query = query;
exports.request = request;
exports.response = response;
exports.timings = timings;

var _harSchema = require('har-schema');

var schemas = _interopRequireWildcard(_harSchema);

var _ajv = require('ajv');

var _ajv2 = _interopRequireDefault(_ajv);

var _error = require('./error');

var _error2 = _interopRequireDefault(_error);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

function _interopRequireWildcard(obj) ***REMOVED*** if (obj && obj.__esModule) ***REMOVED*** return obj; ***REMOVED*** else ***REMOVED*** var newObj = ***REMOVED******REMOVED***; if (obj != null) ***REMOVED*** for (var key in obj) ***REMOVED*** if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; ***REMOVED*** ***REMOVED*** newObj.default = obj; return newObj; ***REMOVED*** ***REMOVED***

var ajv = void 0;

function validate(name) ***REMOVED***
  var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : ***REMOVED******REMOVED***;

  // validator config
  ajv = ajv || new _ajv2.default(***REMOVED***
    allErrors: true,
    schemas: schemas
  ***REMOVED***);

  var validate = ajv.getSchema(name + '.json');

  return new Promise(function (resolve, reject) ***REMOVED***
    var valid = validate(data);

    !valid ? reject(new _error2.default(validate.errors)) : resolve(data);
  ***REMOVED***);
***REMOVED***

function afterRequest(data) ***REMOVED***
  return validate('afterRequest', data);
***REMOVED***

function beforeRequest(data) ***REMOVED***
  return validate('beforeRequest', data);
***REMOVED***

function browser(data) ***REMOVED***
  return validate('browser', data);
***REMOVED***

function cache(data) ***REMOVED***
  return validate('cache', data);
***REMOVED***

function content(data) ***REMOVED***
  return validate('content', data);
***REMOVED***

function cookie(data) ***REMOVED***
  return validate('cookie', data);
***REMOVED***

function creator(data) ***REMOVED***
  return validate('creator', data);
***REMOVED***

function entry(data) ***REMOVED***
  return validate('entry', data);
***REMOVED***

function har(data) ***REMOVED***
  return validate('har', data);
***REMOVED***

function header(data) ***REMOVED***
  return validate('header', data);
***REMOVED***

function log(data) ***REMOVED***
  return validate('log', data);
***REMOVED***

function page(data) ***REMOVED***
  return validate('page', data);
***REMOVED***

function pageTimings(data) ***REMOVED***
  return validate('pageTimings', data);
***REMOVED***

function postData(data) ***REMOVED***
  return validate('postData', data);
***REMOVED***

function query(data) ***REMOVED***
  return validate('query', data);
***REMOVED***

function request(data) ***REMOVED***
  return validate('request', data);
***REMOVED***

function response(data) ***REMOVED***
  return validate('response', data);
***REMOVED***

function timings(data) ***REMOVED***
  return validate('timings', data);
***REMOVED***
