// diffie-hellman/lib/dh.js
var BN = require('bn.js');
var MillerRabin = require('miller-rabin');
var millerRabin = new MillerRabin();
var TWENTYFOUR = new BN(24);
var ELEVEN = new BN(11);
var TEN = new BN(10);
var THREE = new BN(3);
var SEVEN = new BN(7);
var primes = require('./generatePrime');
var randomBytes = require('randombytes');
module.exports = DH;

function setPublicKey(pub, enc) ***REMOVED***
  enc = enc || 'utf8';
  if (!Buffer.isBuffer(pub)) ***REMOVED***
    pub = new Buffer(pub, enc);
  ***REMOVED***
  this._pub = new BN(pub);
  return this;
***REMOVED***

function setPrivateKey(priv, enc) ***REMOVED***
  enc = enc || 'utf8';
  if (!Buffer.isBuffer(priv)) ***REMOVED***
    priv = new Buffer(priv, enc);
  ***REMOVED***
  this._priv = new BN(priv);
  return this;
***REMOVED***

var primeCache = ***REMOVED******REMOVED***;
function checkPrime(prime, generator) ***REMOVED***
  var gen = generator.toString('hex');
  var hex = [gen, prime.toString(16)].join('_');
  if (hex in primeCache) ***REMOVED***
    return primeCache[hex];
  ***REMOVED***
  var error = 0;

  if (prime.isEven() ||
    !primes.simpleSieve ||
    !primes.fermatTest(prime) ||
    !millerRabin.test(prime)) ***REMOVED***
    //not a prime so +1
    error += 1;

    if (gen === '02' || gen === '05') ***REMOVED***
      // we'd be able to check the generator
      // it would fail so +8
      error += 8;
    ***REMOVED*** else ***REMOVED***
      //we wouldn't be able to test the generator
      // so +4
      error += 4;
    ***REMOVED***
    primeCache[hex] = error;
    return error;
  ***REMOVED***
  if (!millerRabin.test(prime.shrn(1))) ***REMOVED***
    //not a safe prime
    error += 2;
  ***REMOVED***
  var rem;
  switch (gen) ***REMOVED***
    case '02':
      if (prime.mod(TWENTYFOUR).cmp(ELEVEN)) ***REMOVED***
        // unsuidable generator
        error += 8;
      ***REMOVED***
      break;
    case '05':
      rem = prime.mod(TEN);
      if (rem.cmp(THREE) && rem.cmp(SEVEN)) ***REMOVED***
        // prime mod 10 needs to equal 3 or 7
        error += 8;
      ***REMOVED***
      break;
    default:
      error += 4;
  ***REMOVED***
  primeCache[hex] = error;
  return error;
***REMOVED***

function DH(prime, generator, malleable) ***REMOVED***
  this.setGenerator(generator);
  this.__prime = new BN(prime);
  this._prime = BN.mont(this.__prime);
  this._primeLen = prime.length;
  this._pub = undefined;
  this._priv = undefined;
  this._primeCode = undefined;
  if (malleable) ***REMOVED***
    this.setPublicKey = setPublicKey;
    this.setPrivateKey = setPrivateKey;
  ***REMOVED*** else ***REMOVED***
    this._primeCode = 8;
  ***REMOVED***
***REMOVED***
Object.defineProperty(DH.prototype, 'verifyError', ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    if (typeof this._primeCode !== 'number') ***REMOVED***
      this._primeCode = checkPrime(this.__prime, this.__gen);
    ***REMOVED***
    return this._primeCode;
  ***REMOVED***
***REMOVED***);
DH.prototype.generateKeys = function () ***REMOVED***
  if (!this._priv) ***REMOVED***
    this._priv = new BN(randomBytes(this._primeLen));
  ***REMOVED***
  this._pub = this._gen.toRed(this._prime).redPow(this._priv).fromRed();
  return this.getPublicKey();
***REMOVED***;

DH.prototype.computeSecret = function (other) ***REMOVED***
  other = new BN(other);
  other = other.toRed(this._prime);
  var secret = other.redPow(this._priv).fromRed();
  var out = new Buffer(secret.toArray());
  var prime = this.getPrime();
  if (out.length < prime.length) ***REMOVED***
    var front = new Buffer(prime.length - out.length);
    front.fill(0);
    out = Buffer.concat([front, out]);
  ***REMOVED***
  return out;
***REMOVED***;

DH.prototype.getPublicKey = function getPublicKey(enc) ***REMOVED***
  return formatReturnValue(this._pub, enc);
***REMOVED***;

DH.prototype.getPrivateKey = function getPrivateKey(enc) ***REMOVED***
  return formatReturnValue(this._priv, enc);
***REMOVED***;

DH.prototype.getPrime = function (enc) ***REMOVED***
  return formatReturnValue(this.__prime, enc);
***REMOVED***;

DH.prototype.getGenerator = function (enc) ***REMOVED***
  return formatReturnValue(this._gen, enc);
***REMOVED***;

DH.prototype.setGenerator = function (gen, enc) ***REMOVED***
  enc = enc || 'utf8';
  if (!Buffer.isBuffer(gen)) ***REMOVED***
    gen = new Buffer(gen, enc);
  ***REMOVED***
  this.__gen = gen;
  this._gen = new BN(gen);
  return this;
***REMOVED***;

function formatReturnValue(bn, enc) ***REMOVED***
  var buf = new Buffer(bn.toArray());
  if (!enc) ***REMOVED***
    return buf;
  ***REMOVED*** else ***REMOVED***
    return buf.toString(enc);
  ***REMOVED***
***REMOVED***
