// core-js/build/index.js
// Generated by LiveScript 1.4.0
(function()***REMOVED***
  var Promise, ref$, list, experimental, libraryBlacklist, es5SpecialCase, banner, readFile, writeFile, unlink, basename, dirname, join, webpack, temp;
  Promise = require('../library/fn/promise');
  ref$ = require('./config'), list = ref$.list, experimental = ref$.experimental, libraryBlacklist = ref$.libraryBlacklist, es5SpecialCase = ref$.es5SpecialCase, banner = ref$.banner;
  ref$ = require('fs'), readFile = ref$.readFile, writeFile = ref$.writeFile, unlink = ref$.unlink;
  ref$ = require('path'), basename = ref$.basename, dirname = ref$.dirname, join = ref$.join;
  webpack = require('webpack');
  temp = require('temp');
  module.exports = function(arg$)***REMOVED***
    var modules, ref$, blacklist, library, umd, this$ = this;
    modules = (ref$ = arg$.modules) != null
      ? ref$
      : [], blacklist = (ref$ = arg$.blacklist) != null
      ? ref$
      : [], library = (ref$ = arg$.library) != null ? ref$ : false, umd = (ref$ = arg$.umd) != null ? ref$ : true;
    return new Promise(function(resolve, reject)***REMOVED***
      (function()***REMOVED***
        var i$, x$, ref$, len$, y$, ns, name, j$, len1$, TARGET, this$ = this;
        if (this.exp) ***REMOVED***
          for (i$ = 0, len$ = (ref$ = experimental).length; i$ < len$; ++i$) ***REMOVED***
            x$ = ref$[i$];
            this[x$] = true;
          ***REMOVED***
        ***REMOVED***
        if (this.es5) ***REMOVED***
          for (i$ = 0, len$ = (ref$ = es5SpecialCase).length; i$ < len$; ++i$) ***REMOVED***
            y$ = ref$[i$];
            this[y$] = true;
          ***REMOVED***
        ***REMOVED***
        for (ns in this) ***REMOVED***
          if (this[ns]) ***REMOVED***
            for (i$ = 0, len$ = (ref$ = list).length; i$ < len$; ++i$) ***REMOVED***
              name = ref$[i$];
              if (name.indexOf(ns + ".") === 0 && !in$(name, experimental)) ***REMOVED***
                this[name] = true;
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
        if (library) ***REMOVED***
          blacklist = blacklist.concat(libraryBlacklist);
        ***REMOVED***
        for (i$ = 0, len$ = blacklist.length; i$ < len$; ++i$) ***REMOVED***
          ns = blacklist[i$];
          for (j$ = 0, len1$ = (ref$ = list).length; j$ < len1$; ++j$) ***REMOVED***
            name = ref$[j$];
            if (name === ns || name.indexOf(ns + ".") === 0) ***REMOVED***
              this[name] = false;
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
        TARGET = temp.path(***REMOVED***
          suffix: '.js'
        ***REMOVED***);
        webpack(***REMOVED***
          entry: list.filter(function(it)***REMOVED***
            return this$[it];
          ***REMOVED***).map(function(it)***REMOVED***
            if (library) ***REMOVED***
              return join(__dirname, '..', 'library', 'modules', it);
            ***REMOVED*** else ***REMOVED***
              return join(__dirname, '..', 'modules', it);
            ***REMOVED***
          ***REMOVED***),
          output: ***REMOVED***
            path: dirname(TARGET),
            filename: basename("./" + TARGET)
          ***REMOVED***
        ***REMOVED***, function(err, info)***REMOVED***
          if (err) ***REMOVED***
            return reject(err);
          ***REMOVED***
          readFile(TARGET, function(err, script)***REMOVED***
            if (err) ***REMOVED***
              return reject(err);
            ***REMOVED***
            unlink(TARGET, function(err)***REMOVED***
              var exportScript;
              if (err) ***REMOVED***
                return reject(err);
              ***REMOVED***
              if (umd) ***REMOVED***
                exportScript = "// CommonJS export\nif (typeof module != 'undefined' && module.exports) module.exports = __e;\n// RequireJS export\nelse if (typeof define == 'function' && define.amd) define(function () ***REMOVED*** return __e; ***REMOVED***);\n// Export to global object\nelse __g.core = __e;";
              ***REMOVED*** else ***REMOVED***
                exportScript = "";
              ***REMOVED***
              resolve("" + banner + "\n!function(__e, __g, undefined)***REMOVED***\n'use strict';\n" + script + "\n" + exportScript + "\n***REMOVED***(1, 1);");
            ***REMOVED***);
          ***REMOVED***);
        ***REMOVED***);
      ***REMOVED***.call(modules.reduce(function(memo, it)***REMOVED***
        memo[it] = true;
        return memo;
      ***REMOVED***, ***REMOVED******REMOVED***)));
    ***REMOVED***);
  ***REMOVED***;
  function in$(x, xs)***REMOVED***
    var i = -1, l = xs.length >>> 0;
    while (++i < l) if (x === xs[i]) return true;
    return false;
  ***REMOVED***
***REMOVED***).call(this);
