// d/test/index.js
'use strict';

var getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;

module.exports = function (t, a) ***REMOVED***
	var o, c, cg, cs, ce, ceg, ces, cew, cw, e, eg, es, ew, v, vg, vs, w, df, dfg
	  , dfs;

	o = Object.create(Object.prototype, ***REMOVED***
		c: t('c', c = ***REMOVED******REMOVED***),
		cgs: t.gs('c', cg = function () ***REMOVED******REMOVED***, cs = function () ***REMOVED******REMOVED***),
		ce: t('ce', ce = ***REMOVED******REMOVED***),
		cegs: t.gs('ce', ceg = function () ***REMOVED******REMOVED***, ces = function () ***REMOVED******REMOVED***),
		cew: t('cew', cew = ***REMOVED******REMOVED***),
		cw: t('cw', cw = ***REMOVED******REMOVED***),
		e: t('e', e = ***REMOVED******REMOVED***),
		egs: t.gs('e', eg = function () ***REMOVED******REMOVED***, es = function () ***REMOVED******REMOVED***),
		ew: t('ew', ew = ***REMOVED******REMOVED***),
		v: t('', v = ***REMOVED******REMOVED***),
		vgs: t.gs('', vg = function () ***REMOVED******REMOVED***, vs = function () ***REMOVED******REMOVED***),
		w: t('w', w = ***REMOVED******REMOVED***),

		df: t(df = ***REMOVED******REMOVED***),
		dfgs: t.gs(dfg = function () ***REMOVED******REMOVED***, dfs = function () ***REMOVED******REMOVED***)
	***REMOVED***);

	return ***REMOVED***
		c: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'c');
			a(d.value, c, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, true, "Configurable");
			a(d.enumerable, false, "Enumerable");
			a(d.writable, false, "Writable");

			d = getOwnPropertyDescriptor(o, 'cgs');
			a(d.value, undefined, "GS Value");
			a(d.get, cg, "GS Get");
			a(d.set, cs, "GS Set");
			a(d.configurable, true, "GS Configurable");
			a(d.enumerable, false, "GS Enumerable");
			a(d.writable, undefined, "GS Writable");
		***REMOVED***,
		ce: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'ce');
			a(d.value, ce, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, true, "Configurable");
			a(d.enumerable, true, "Enumerable");
			a(d.writable, false, "Writable");

			d = getOwnPropertyDescriptor(o, 'cegs');
			a(d.value, undefined, "GS Value");
			a(d.get, ceg, "GS Get");
			a(d.set, ces, "GS Set");
			a(d.configurable, true, "GS Configurable");
			a(d.enumerable, true, "GS Enumerable");
			a(d.writable, undefined, "GS Writable");
		***REMOVED***,
		cew: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'cew');
			a(d.value, cew, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, true, "Configurable");
			a(d.enumerable, true, "Enumerable");
			a(d.writable, true, "Writable");
		***REMOVED***,
		cw: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'cw');
			a(d.value, cw, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, true, "Configurable");
			a(d.enumerable, false, "Enumerable");
			a(d.writable, true, "Writable");
		***REMOVED***,
		e: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'e');
			a(d.value, e, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, false, "Configurable");
			a(d.enumerable, true, "Enumerable");
			a(d.writable, false, "Writable");

			d = getOwnPropertyDescriptor(o, 'egs');
			a(d.value, undefined, "GS Value");
			a(d.get, eg, "GS Get");
			a(d.set, es, "GS Set");
			a(d.configurable, false, "GS Configurable");
			a(d.enumerable, true, "GS Enumerable");
			a(d.writable, undefined, "GS Writable");
		***REMOVED***,
		ew: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'ew');
			a(d.value, ew, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, false, "Configurable");
			a(d.enumerable, true, "Enumerable");
			a(d.writable, true, "Writable");
		***REMOVED***,
		v: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'v');
			a(d.value, v, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, false, "Configurable");
			a(d.enumerable, false, "Enumerable");
			a(d.writable, false, "Writable");

			d = getOwnPropertyDescriptor(o, 'vgs');
			a(d.value, undefined, "GS Value");
			a(d.get, vg, "GS Get");
			a(d.set, vs, "GS Set");
			a(d.configurable, false, "GS Configurable");
			a(d.enumerable, false, "GS Enumerable");
			a(d.writable, undefined, "GS Writable");
		***REMOVED***,
		w: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'w');
			a(d.value, w, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, false, "Configurable");
			a(d.enumerable, false, "Enumerable");
			a(d.writable, true, "Writable");
		***REMOVED***,
		d: function (a) ***REMOVED***
			var d = getOwnPropertyDescriptor(o, 'df');
			a(d.value, df, "Value");
			a(d.get, undefined, "Get");
			a(d.set, undefined, "Set");
			a(d.configurable, true, "Configurable");
			a(d.enumerable, false, "Enumerable");
			a(d.writable, true, "Writable");

			d = getOwnPropertyDescriptor(o, 'dfgs');
			a(d.value, undefined, "GS Value");
			a(d.get, dfg, "GS Get");
			a(d.set, dfs, "GS Set");
			a(d.configurable, true, "GS Configurable");
			a(d.enumerable, false, "GS Enumerable");
			a(d.writable, undefined, "GS Writable");
		***REMOVED***,
		Options: ***REMOVED***
			v: function (a) ***REMOVED***
				var x = ***REMOVED******REMOVED***, d = t(x, ***REMOVED*** foo: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: true, enumerable: false, writable: true,
					value: x, foo: true ***REMOVED***, "No descriptor");
				d = t('c', 'foo', ***REMOVED*** marko: 'elo' ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: true, enumerable: false, writable: false,
					value: 'foo', marko: 'elo' ***REMOVED***, "Descriptor");
			***REMOVED***,
			gs: function (a) ***REMOVED***
				var gFn = function () ***REMOVED******REMOVED***, sFn = function () ***REMOVED******REMOVED***, d;
				d = t.gs(gFn, sFn, ***REMOVED*** foo: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: true, enumerable: false, get: gFn, set: sFn,
					foo: true ***REMOVED***, "No descriptor");
				d = t.gs(null, sFn, ***REMOVED*** foo: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: true, enumerable: false, get: undefined,
					set: sFn, foo: true ***REMOVED***, "No descriptor: Just set");
				d = t.gs(gFn, ***REMOVED*** foo: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: true, enumerable: false, get: gFn,
					set: undefined, foo: true ***REMOVED***, "No descriptor: Just get");

				d = t.gs('e', gFn, sFn, ***REMOVED*** bar: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: false, enumerable: true, get: gFn, set: sFn,
					bar: true ***REMOVED***, "Descriptor");
				d = t.gs('e', null, sFn, ***REMOVED*** bar: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: false, enumerable: true, get: undefined,
					set: sFn, bar: true ***REMOVED***, "Descriptor: Just set");
				d = t.gs('e', gFn, ***REMOVED*** bar: true ***REMOVED***);
				a.deep(d, ***REMOVED*** configurable: false, enumerable: true, get: gFn,
					set: undefined, bar: true ***REMOVED***, "Descriptor: Just get");
			***REMOVED***
		***REMOVED***
	***REMOVED***;
***REMOVED***;
