// d/test/lazy.js
'use strict';

var d = require('../')

  , getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;

module.exports = function (t, a) ***REMOVED***
	var Foo = function () ***REMOVED******REMOVED***, i = 1, o, o2, desc;
	Object.defineProperties(Foo.prototype, t(***REMOVED***
		bar: d(function () ***REMOVED*** return ++i; ***REMOVED***),
		bar2: d(function () ***REMOVED*** return this.bar + 23; ***REMOVED***),
		bar3: d(function () ***REMOVED*** return this.bar2 + 34; ***REMOVED***, ***REMOVED*** desc: 'ew' ***REMOVED***),
		bar4: d(function () ***REMOVED*** return this.bar3 + 12; ***REMOVED***, ***REMOVED*** cacheName: '_bar4_' ***REMOVED***),
		bar5: d(function () ***REMOVED*** return this.bar4 + 3; ***REMOVED***,
			***REMOVED*** cacheName: '_bar5_', desc: 'e' ***REMOVED***)
	***REMOVED***));

	desc = getOwnPropertyDescriptor(Foo.prototype, 'bar');
	a(desc.configurable, true, "Configurable: default");
	a(desc.enumerable, false, "Enumerable: default");

	o = new Foo();
	a.deep([o.bar, o.bar2, o.bar3, o.bar4, o.bar5], [2, 25, 59, 71, 74],
		"Values");

	a.deep(getOwnPropertyDescriptor(o, 'bar3'), ***REMOVED*** configurable: false,
		enumerable: true, writable: true, value: 59 ***REMOVED***, "Desc");
	a(o.hasOwnProperty('bar4'), false, "Cache not exposed");
	desc = getOwnPropertyDescriptor(o, 'bar5');
	a.deep(desc, ***REMOVED*** configurable: false,
		enumerable: true, get: desc.get, set: desc.set ***REMOVED***, "Cache & Desc: desc");

	o2 = Object.create(o);
	o2.bar = 30;
	o2.bar3 = 100;

	a.deep([o2.bar, o2.bar2, o2.bar3, o2.bar4, o2.bar5], [30, 25, 100, 112, 115],
		"Extension Values");

	Foo = function () ***REMOVED******REMOVED***;
	Object.defineProperties(Foo.prototype, t(***REMOVED***
		test: d('w', function () ***REMOVED*** return 'raz'; ***REMOVED***),
		test2: d('', function () ***REMOVED*** return 'raz'; ***REMOVED***, ***REMOVED*** desc: 'w' ***REMOVED***),
		test3: d('', function () ***REMOVED*** return 'raz'; ***REMOVED***,
			***REMOVED*** cacheName: '__test3__', desc: 'w' ***REMOVED***),
		test4: d('w', 'bar')
	***REMOVED***));

	o = new Foo();
	o.test = 'marko';
	a.deep(getOwnPropertyDescriptor(o, 'test'),
		***REMOVED*** configurable: false, enumerable: false, writable: true, value: 'marko' ***REMOVED***,
		"Set before get");
	o.test2 = 'marko2';
	a.deep(getOwnPropertyDescriptor(o, 'test2'),
		***REMOVED*** configurable: false, enumerable: false, writable: true, value: 'marko2' ***REMOVED***,
		"Set before get: Custom desc");
	o.test3 = 'marko3';
	a.deep(getOwnPropertyDescriptor(o, '__test3__'),
		***REMOVED*** configurable: false, enumerable: false, writable: true, value: 'marko3' ***REMOVED***,
		"Set before get: Custom cache name");
	a(o.test4, 'bar', "Resolve by value");

	a.h1("Flat");
	Object.defineProperties(Foo.prototype, t(***REMOVED***
		flat: d(function () ***REMOVED*** return 'foo'; ***REMOVED***, ***REMOVED*** flat: true ***REMOVED***),
		flat2: d(function () ***REMOVED*** return 'bar'; ***REMOVED***, ***REMOVED*** flat: true ***REMOVED***)
	***REMOVED***));

	a.h2("Instance");
	a(o.flat, 'foo', "Value");
	a(o.hasOwnProperty('flat'), false, "Instance");
	a(Foo.prototype.flat, 'foo', "Prototype");

	a.h2("Direct");
	a(Foo.prototype.flat2, 'bar');

	a.h2("Reset direct");
	Object.defineProperties(Foo.prototype, t(***REMOVED*** testResetDirect: d(false) ***REMOVED***));

	a.throws(function () ***REMOVED*** Foo.prototype.testResetDirect = false; ***REMOVED***, TypeError);
***REMOVED***;
