// core-js/modules/core.dict.js
'use strict';
var ctx = require('./_ctx');
var $export = require('./_export');
var createDesc = require('./_property-desc');
var assign = require('./_object-assign');
var create = require('./_object-create');
var getPrototypeOf = require('./_object-gpo');
var getKeys = require('./_object-keys');
var dP = require('./_object-dp');
var keyOf = require('./_keyof');
var aFunction = require('./_a-function');
var forOf = require('./_for-of');
var isIterable = require('./core.is-iterable');
var $iterCreate = require('./_iter-create');
var step = require('./_iter-step');
var isObject = require('./_is-object');
var toIObject = require('./_to-iobject');
var DESCRIPTORS = require('./_descriptors');
var has = require('./_has');

// 0 -> Dict.forEach
// 1 -> Dict.map
// 2 -> Dict.filter
// 3 -> Dict.some
// 4 -> Dict.every
// 5 -> Dict.find
// 6 -> Dict.findKey
// 7 -> Dict.mapPairs
var createDictMethod = function (TYPE) ***REMOVED***
  var IS_MAP = TYPE == 1;
  var IS_EVERY = TYPE == 4;
  return function (object, callbackfn, that /* = undefined */) ***REMOVED***
    var f = ctx(callbackfn, that, 3);
    var O = toIObject(object);
    var result = IS_MAP || TYPE == 7 || TYPE == 2
          ? new (typeof this == 'function' ? this : Dict)() : undefined;
    var key, val, res;
    for (key in O) if (has(O, key)) ***REMOVED***
      val = O[key];
      res = f(val, key, object);
      if (TYPE) ***REMOVED***
        if (IS_MAP) result[key] = res;          // map
        else if (res) switch (TYPE) ***REMOVED***
          case 2: result[key] = val; break;     // filter
          case 3: return true;                  // some
          case 5: return val;                   // find
          case 6: return key;                   // findKey
          case 7: result[res[0]] = res[1];      // mapPairs
        ***REMOVED*** else if (IS_EVERY) return false;      // every
      ***REMOVED***
    ***REMOVED***
    return TYPE == 3 || IS_EVERY ? IS_EVERY : result;
  ***REMOVED***;
***REMOVED***;
var findKey = createDictMethod(6);

var createDictIter = function (kind) ***REMOVED***
  return function (it) ***REMOVED***
    return new DictIterator(it, kind);
  ***REMOVED***;
***REMOVED***;
var DictIterator = function (iterated, kind) ***REMOVED***
  this._t = toIObject(iterated); // target
  this._a = getKeys(iterated);   // keys
  this._i = 0;                   // next index
  this._k = kind;                // kind
***REMOVED***;
$iterCreate(DictIterator, 'Dict', function () ***REMOVED***
  var that = this;
  var O = that._t;
  var keys = that._a;
  var kind = that._k;
  var key;
  do ***REMOVED***
    if (that._i >= keys.length) ***REMOVED***
      that._t = undefined;
      return step(1);
    ***REMOVED***
  ***REMOVED*** while (!has(O, key = keys[that._i++]));
  if (kind == 'keys') return step(0, key);
  if (kind == 'values') return step(0, O[key]);
  return step(0, [key, O[key]]);
***REMOVED***);

function Dict(iterable) ***REMOVED***
  var dict = create(null);
  if (iterable != undefined) ***REMOVED***
    if (isIterable(iterable)) ***REMOVED***
      forOf(iterable, true, function (key, value) ***REMOVED***
        dict[key] = value;
      ***REMOVED***);
    ***REMOVED*** else assign(dict, iterable);
  ***REMOVED***
  return dict;
***REMOVED***
Dict.prototype = null;

function reduce(object, mapfn, init) ***REMOVED***
  aFunction(mapfn);
  var O = toIObject(object);
  var keys = getKeys(O);
  var length = keys.length;
  var i = 0;
  var memo, key;
  if (arguments.length < 3) ***REMOVED***
    if (!length) throw TypeError('Reduce of empty object with no initial value');
    memo = O[keys[i++]];
  ***REMOVED*** else memo = Object(init);
  while (length > i) if (has(O, key = keys[i++])) ***REMOVED***
    memo = mapfn(memo, O[key], key, object);
  ***REMOVED***
  return memo;
***REMOVED***

function includes(object, el) ***REMOVED***
  // eslint-disable-next-line no-self-compare
  return (el == el ? keyOf(object, el) : findKey(object, function (it) ***REMOVED***
    // eslint-disable-next-line no-self-compare
    return it != it;
  ***REMOVED***)) !== undefined;
***REMOVED***

function get(object, key) ***REMOVED***
  if (has(object, key)) return object[key];
***REMOVED***
function set(object, key, value) ***REMOVED***
  if (DESCRIPTORS && key in Object) dP.f(object, key, createDesc(0, value));
  else object[key] = value;
  return object;
***REMOVED***

function isDict(it) ***REMOVED***
  return isObject(it) && getPrototypeOf(it) === Dict.prototype;
***REMOVED***

$export($export.G + $export.F, ***REMOVED*** Dict: Dict ***REMOVED***);

$export($export.S, 'Dict', ***REMOVED***
  keys: createDictIter('keys'),
  values: createDictIter('values'),
  entries: createDictIter('entries'),
  forEach: createDictMethod(0),
  map: createDictMethod(1),
  filter: createDictMethod(2),
  some: createDictMethod(3),
  every: createDictMethod(4),
  find: createDictMethod(5),
  findKey: findKey,
  mapPairs: createDictMethod(7),
  reduce: reduce,
  keyOf: keyOf,
  includes: includes,
  has: has,
  get: get,
  set: set,
  isDict: isDict
***REMOVED***);
