// grpc/src/metadata.js
/**
 * @license
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

var _ = require('lodash');

var grpc = require('./grpc_extension');

/**
 * Class for storing metadata. Keys are normalized to lowercase ASCII.
 * @memberof grpc
 * @constructor
 * @example
 * var metadata = new metadata_module.Metadata();
 * metadata.set('key1', 'value1');
 * metadata.add('key1', 'value2');
 * metadata.get('key1') // returns ['value1', 'value2']
 */
function Metadata() ***REMOVED***
  this._internal_repr = ***REMOVED******REMOVED***;
***REMOVED***

function normalizeKey(key) ***REMOVED***
  key = key.toLowerCase();
  if (grpc.metadataKeyIsLegal(key)) ***REMOVED***
    return key;
  ***REMOVED*** else ***REMOVED***
    throw new Error('Metadata key"' + key + '" contains illegal characters');
  ***REMOVED***
***REMOVED***

function validate(key, value) ***REMOVED***
  if (grpc.metadataKeyIsBinary(key)) ***REMOVED***
    if (!(value instanceof Buffer)) ***REMOVED***
      throw new Error('keys that end with \'-bin\' must have Buffer values');
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    if (!_.isString(value)) ***REMOVED***
      throw new Error(
          'keys that don\'t end with \'-bin\' must have String values');
    ***REMOVED***
    if (!grpc.metadataNonbinValueIsLegal(value)) ***REMOVED***
      throw new Error('Metadata string value "' + value +
                      '" contains illegal characters');
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Sets the given value for the given key, replacing any other values associated
 * with that key. Normalizes the key.
 * @param ***REMOVED***String***REMOVED*** key The key to set
 * @param ***REMOVED***String|Buffer***REMOVED*** value The value to set. Must be a buffer if and only
 *     if the normalized key ends with '-bin'
 */
Metadata.prototype.set = function(key, value) ***REMOVED***
  key = normalizeKey(key);
  validate(key, value);
  this._internal_repr[key] = [value];
***REMOVED***;

/**
 * Adds the given value for the given key. Normalizes the key.
 * @param ***REMOVED***String***REMOVED*** key The key to add to.
 * @param ***REMOVED***String|Buffer***REMOVED*** value The value to add. Must be a buffer if and only
 *     if the normalized key ends with '-bin'
 */
Metadata.prototype.add = function(key, value) ***REMOVED***
  key = normalizeKey(key);
  validate(key, value);
  if (!this._internal_repr[key]) ***REMOVED***
    this._internal_repr[key] = [];
  ***REMOVED***
  this._internal_repr[key].push(value);
***REMOVED***;

/**
 * Remove the given key and any associated values. Normalizes the key.
 * @param ***REMOVED***String***REMOVED*** key The key to remove
 */
Metadata.prototype.remove = function(key) ***REMOVED***
  key = normalizeKey(key);
  if (Object.prototype.hasOwnProperty.call(this._internal_repr, key)) ***REMOVED***
    delete this._internal_repr[key];
  ***REMOVED***
***REMOVED***;

/**
 * Gets a list of all values associated with the key. Normalizes the key.
 * @param ***REMOVED***String***REMOVED*** key The key to get
 * @return ***REMOVED***Array.<String|Buffer>***REMOVED*** The values associated with that key
 */
Metadata.prototype.get = function(key) ***REMOVED***
  key = normalizeKey(key);
  if (Object.prototype.hasOwnProperty.call(this._internal_repr, key)) ***REMOVED***
    return this._internal_repr[key];
  ***REMOVED*** else ***REMOVED***
    return [];
  ***REMOVED***
***REMOVED***;

/**
 * Get a map of each key to a single associated value. This reflects the most
 * common way that people will want to see metadata.
 * @return ***REMOVED***Object.<String,String|Buffer>***REMOVED*** A key/value mapping of the metadata
 */
Metadata.prototype.getMap = function() ***REMOVED***
  var result = ***REMOVED******REMOVED***;
  _.forOwn(this._internal_repr, function(values, key) ***REMOVED***
    if(values.length > 0) ***REMOVED***
      result[key] = values[0];
    ***REMOVED***
  ***REMOVED***);
  return result;
***REMOVED***;

/**
 * Clone the metadata object.
 * @return ***REMOVED***grpc.Metadata***REMOVED*** The new cloned object
 */
Metadata.prototype.clone = function() ***REMOVED***
  var copy = new Metadata();
  _.forOwn(this._internal_repr, function(value, key) ***REMOVED***
    copy._internal_repr[key] = _.clone(value);
  ***REMOVED***);
  return copy;
***REMOVED***;

/**
 * Gets the metadata in the format used by interal code. Intended for internal
 * use only. API stability is not guaranteed.
 * @private
 * @return ***REMOVED***Object.<String, Array.<String|Buffer>>***REMOVED*** The metadata
 */
Metadata.prototype._getCoreRepresentation = function() ***REMOVED***
  return this._internal_repr;
***REMOVED***;

/**
 * Creates a Metadata object from a metadata map in the internal format.
 * Intended for internal use only. API stability is not guaranteed.
 * @private
 * @param ***REMOVED***Object.<String, Array.<String|Buffer>>***REMOVED*** The metadata
 * @return ***REMOVED***Metadata***REMOVED*** The new Metadata object
 */
Metadata._fromCoreRepresentation = function(metadata) ***REMOVED***
  var newMetadata = new Metadata();
  if (metadata) ***REMOVED***
    _.forOwn(metadata, function(value, key) ***REMOVED***
      newMetadata._internal_repr[key] = _.clone(value);
    ***REMOVED***);
  ***REMOVED***
  return newMetadata;
***REMOVED***;

module.exports = Metadata;
