// tar/lib/entry-writer.js
module.exports = EntryWriter

var tar = require("../tar.js")
  , TarHeader = require("./header.js")
  , Entry = require("./entry.js")
  , inherits = require("inherits")
  , BlockStream = require("block-stream")
  , ExtendedHeaderWriter
  , Stream = require("stream").Stream
  , EOF = ***REMOVED******REMOVED***

inherits(EntryWriter, Stream)

function EntryWriter (props) ***REMOVED***
  var me = this

  if (!(me instanceof EntryWriter)) ***REMOVED***
    return new EntryWriter(props)
  ***REMOVED***

  Stream.apply(this)

  me.writable = true
  me.readable = true

  me._stream = new BlockStream(512)

  me._stream.on("data", function (c) ***REMOVED***
    me.emit("data", c)
  ***REMOVED***)

  me._stream.on("drain", function () ***REMOVED***
    me.emit("drain")
  ***REMOVED***)

  me._stream.on("end", function () ***REMOVED***
    me.emit("end")
    me.emit("close")
  ***REMOVED***)

  me.props = props
  if (props.type === "Directory") ***REMOVED***
    props.size = 0
  ***REMOVED***
  props.ustar = "ustar\0"
  props.ustarver = "00"
  me.path = props.path

  me._buffer = []
  me._didHeader = false
  me._meta = false

  me.on("pipe", function () ***REMOVED***
    me._process()
  ***REMOVED***)
***REMOVED***

EntryWriter.prototype.write = function (c) ***REMOVED***
  // console.error(".. ew write")
  if (this._ended) return this.emit("error", new Error("write after end"))
  this._buffer.push(c)
  this._process()
  this._needDrain = this._buffer.length > 0
  return !this._needDrain
***REMOVED***

EntryWriter.prototype.end = function (c) ***REMOVED***
  // console.error(".. ew end")
  if (c) this._buffer.push(c)
  this._buffer.push(EOF)
  this._ended = true
  this._process()
  this._needDrain = this._buffer.length > 0
***REMOVED***

EntryWriter.prototype.pause = function () ***REMOVED***
  // console.error(".. ew pause")
  this._paused = true
  this.emit("pause")
***REMOVED***

EntryWriter.prototype.resume = function () ***REMOVED***
  // console.error(".. ew resume")
  this._paused = false
  this.emit("resume")
  this._process()
***REMOVED***

EntryWriter.prototype.add = function (entry) ***REMOVED***
  // console.error(".. ew add")
  if (!this.parent) return this.emit("error", new Error("no parent"))

  // make sure that the _header and such is emitted, and clear out
  // the _currentEntry link on the parent.
  if (!this._ended) this.end()

  return this.parent.add(entry)
***REMOVED***

EntryWriter.prototype._header = function () ***REMOVED***
  // console.error(".. ew header")
  if (this._didHeader) return
  this._didHeader = true

  var headerBlock = TarHeader.encode(this.props)

  if (this.props.needExtended && !this._meta) ***REMOVED***
    var me = this

    ExtendedHeaderWriter = ExtendedHeaderWriter ||
      require("./extended-header-writer.js")

    ExtendedHeaderWriter(this.props)
      .on("data", function (c) ***REMOVED***
        me.emit("data", c)
      ***REMOVED***)
      .on("error", function (er) ***REMOVED***
        me.emit("error", er)
      ***REMOVED***)
      .end()
  ***REMOVED***

  // console.error(".. .. ew headerBlock emitting")
  this.emit("data", headerBlock)
  this.emit("header")
***REMOVED***

EntryWriter.prototype._process = function () ***REMOVED***
  // console.error(".. .. ew process")
  if (!this._didHeader && !this._meta) ***REMOVED***
    this._header()
  ***REMOVED***

  if (this._paused || this._processing) ***REMOVED***
    // console.error(".. .. .. paused=%j, processing=%j", this._paused, this._processing)
    return
  ***REMOVED***

  this._processing = true

  var buf = this._buffer
  for (var i = 0; i < buf.length; i ++) ***REMOVED***
    // console.error(".. .. .. i=%d", i)

    var c = buf[i]

    if (c === EOF) this._stream.end()
    else this._stream.write(c)

    if (this._paused) ***REMOVED***
      // console.error(".. .. .. paused mid-emission")
      this._processing = false
      if (i < buf.length) ***REMOVED***
        this._needDrain = true
        this._buffer = buf.slice(i + 1)
      ***REMOVED***
      return
    ***REMOVED***
  ***REMOVED***

  // console.error(".. .. .. emitted")
  this._buffer.length = 0
  this._processing = false

  // console.error(".. .. .. emitting drain")
  this.emit("drain")
***REMOVED***

EntryWriter.prototype.destroy = function () ***REMOVED******REMOVED***
