// asn1.js/lib/asn1/decoders/der.js
var inherits = require('inherits');

var asn1 = require('../../asn1');
var base = asn1.base;
var bignum = asn1.bignum;

// Import DER constants
var der = asn1.constants.der;

function DERDecoder(entity) ***REMOVED***
  this.enc = 'der';
  this.name = entity.name;
  this.entity = entity;

  // Construct base tree
  this.tree = new DERNode();
  this.tree._init(entity.body);
***REMOVED***;
module.exports = DERDecoder;

DERDecoder.prototype.decode = function decode(data, options) ***REMOVED***
  if (!(data instanceof base.DecoderBuffer))
    data = new base.DecoderBuffer(data, options);

  return this.tree._decode(data, options);
***REMOVED***;

// Tree methods

function DERNode(parent) ***REMOVED***
  base.Node.call(this, 'der', parent);
***REMOVED***
inherits(DERNode, base.Node);

DERNode.prototype._peekTag = function peekTag(buffer, tag, any) ***REMOVED***
  if (buffer.isEmpty())
    return false;

  var state = buffer.save();
  var decodedTag = derDecodeTag(buffer, 'Failed to peek tag: "' + tag + '"');
  if (buffer.isError(decodedTag))
    return decodedTag;

  buffer.restore(state);

  return decodedTag.tag === tag || decodedTag.tagStr === tag ||
    (decodedTag.tagStr + 'of') === tag || any;
***REMOVED***;

DERNode.prototype._decodeTag = function decodeTag(buffer, tag, any) ***REMOVED***
  var decodedTag = derDecodeTag(buffer,
                                'Failed to decode tag of "' + tag + '"');
  if (buffer.isError(decodedTag))
    return decodedTag;

  var len = derDecodeLen(buffer,
                         decodedTag.primitive,
                         'Failed to get length of "' + tag + '"');

  // Failure
  if (buffer.isError(len))
    return len;

  if (!any &&
      decodedTag.tag !== tag &&
      decodedTag.tagStr !== tag &&
      decodedTag.tagStr + 'of' !== tag) ***REMOVED***
    return buffer.error('Failed to match tag: "' + tag + '"');
  ***REMOVED***

  if (decodedTag.primitive || len !== null)
    return buffer.skip(len, 'Failed to match body of: "' + tag + '"');

  // Indefinite length... find END tag
  var state = buffer.save();
  var res = this._skipUntilEnd(
      buffer,
      'Failed to skip indefinite length body: "' + this.tag + '"');
  if (buffer.isError(res))
    return res;

  len = buffer.offset - state.offset;
  buffer.restore(state);
  return buffer.skip(len, 'Failed to match body of: "' + tag + '"');
***REMOVED***;

DERNode.prototype._skipUntilEnd = function skipUntilEnd(buffer, fail) ***REMOVED***
  while (true) ***REMOVED***
    var tag = derDecodeTag(buffer, fail);
    if (buffer.isError(tag))
      return tag;
    var len = derDecodeLen(buffer, tag.primitive, fail);
    if (buffer.isError(len))
      return len;

    var res;
    if (tag.primitive || len !== null)
      res = buffer.skip(len)
    else
      res = this._skipUntilEnd(buffer, fail);

    // Failure
    if (buffer.isError(res))
      return res;

    if (tag.tagStr === 'end')
      break;
  ***REMOVED***
***REMOVED***;

DERNode.prototype._decodeList = function decodeList(buffer, tag, decoder,
                                                    options) ***REMOVED***
  var result = [];
  while (!buffer.isEmpty()) ***REMOVED***
    var possibleEnd = this._peekTag(buffer, 'end');
    if (buffer.isError(possibleEnd))
      return possibleEnd;

    var res = decoder.decode(buffer, 'der', options);
    if (buffer.isError(res) && possibleEnd)
      break;
    result.push(res);
  ***REMOVED***
  return result;
***REMOVED***;

DERNode.prototype._decodeStr = function decodeStr(buffer, tag) ***REMOVED***
  if (tag === 'bitstr') ***REMOVED***
    var unused = buffer.readUInt8();
    if (buffer.isError(unused))
      return unused;
    return ***REMOVED*** unused: unused, data: buffer.raw() ***REMOVED***;
  ***REMOVED*** else if (tag === 'bmpstr') ***REMOVED***
    var raw = buffer.raw();
    if (raw.length % 2 === 1)
      return buffer.error('Decoding of string type: bmpstr length mismatch');

    var str = '';
    for (var i = 0; i < raw.length / 2; i++) ***REMOVED***
      str += String.fromCharCode(raw.readUInt16BE(i * 2));
    ***REMOVED***
    return str;
  ***REMOVED*** else if (tag === 'numstr') ***REMOVED***
    var numstr = buffer.raw().toString('ascii');
    if (!this._isNumstr(numstr)) ***REMOVED***
      return buffer.error('Decoding of string type: ' +
                          'numstr unsupported characters');
    ***REMOVED***
    return numstr;
  ***REMOVED*** else if (tag === 'octstr') ***REMOVED***
    return buffer.raw();
  ***REMOVED*** else if (tag === 'objDesc') ***REMOVED***
    return buffer.raw();
  ***REMOVED*** else if (tag === 'printstr') ***REMOVED***
    var printstr = buffer.raw().toString('ascii');
    if (!this._isPrintstr(printstr)) ***REMOVED***
      return buffer.error('Decoding of string type: ' +
                          'printstr unsupported characters');
    ***REMOVED***
    return printstr;
  ***REMOVED*** else if (/str$/.test(tag)) ***REMOVED***
    return buffer.raw().toString();
  ***REMOVED*** else ***REMOVED***
    return buffer.error('Decoding of string type: ' + tag + ' unsupported');
  ***REMOVED***
***REMOVED***;

DERNode.prototype._decodeObjid = function decodeObjid(buffer, values, relative) ***REMOVED***
  var result;
  var identifiers = [];
  var ident = 0;
  while (!buffer.isEmpty()) ***REMOVED***
    var subident = buffer.readUInt8();
    ident <<= 7;
    ident |= subident & 0x7f;
    if ((subident & 0x80) === 0) ***REMOVED***
      identifiers.push(ident);
      ident = 0;
    ***REMOVED***
  ***REMOVED***
  if (subident & 0x80)
    identifiers.push(ident);

  var first = (identifiers[0] / 40) | 0;
  var second = identifiers[0] % 40;

  if (relative)
    result = identifiers;
  else
    result = [first, second].concat(identifiers.slice(1));

  if (values) ***REMOVED***
    var tmp = values[result.join(' ')];
    if (tmp === undefined)
      tmp = values[result.join('.')];
    if (tmp !== undefined)
      result = tmp;
  ***REMOVED***

  return result;
***REMOVED***;

DERNode.prototype._decodeTime = function decodeTime(buffer, tag) ***REMOVED***
  var str = buffer.raw().toString();
  if (tag === 'gentime') ***REMOVED***
    var year = str.slice(0, 4) | 0;
    var mon = str.slice(4, 6) | 0;
    var day = str.slice(6, 8) | 0;
    var hour = str.slice(8, 10) | 0;
    var min = str.slice(10, 12) | 0;
    var sec = str.slice(12, 14) | 0;
  ***REMOVED*** else if (tag === 'utctime') ***REMOVED***
    var year = str.slice(0, 2) | 0;
    var mon = str.slice(2, 4) | 0;
    var day = str.slice(4, 6) | 0;
    var hour = str.slice(6, 8) | 0;
    var min = str.slice(8, 10) | 0;
    var sec = str.slice(10, 12) | 0;
    if (year < 70)
      year = 2000 + year;
    else
      year = 1900 + year;
  ***REMOVED*** else ***REMOVED***
    return buffer.error('Decoding ' + tag + ' time is not supported yet');
  ***REMOVED***

  return Date.UTC(year, mon - 1, day, hour, min, sec, 0);
***REMOVED***;

DERNode.prototype._decodeNull = function decodeNull(buffer) ***REMOVED***
  return null;
***REMOVED***;

DERNode.prototype._decodeBool = function decodeBool(buffer) ***REMOVED***
  var res = buffer.readUInt8();
  if (buffer.isError(res))
    return res;
  else
    return res !== 0;
***REMOVED***;

DERNode.prototype._decodeInt = function decodeInt(buffer, values) ***REMOVED***
  // Bigint, return as it is (assume big endian)
  var raw = buffer.raw();
  var res = new bignum(raw);

  if (values)
    res = values[res.toString(10)] || res;

  return res;
***REMOVED***;

DERNode.prototype._use = function use(entity, obj) ***REMOVED***
  if (typeof entity === 'function')
    entity = entity(obj);
  return entity._getDecoder('der').tree;
***REMOVED***;

// Utility methods

function derDecodeTag(buf, fail) ***REMOVED***
  var tag = buf.readUInt8(fail);
  if (buf.isError(tag))
    return tag;

  var cls = der.tagClass[tag >> 6];
  var primitive = (tag & 0x20) === 0;

  // Multi-octet tag - load
  if ((tag & 0x1f) === 0x1f) ***REMOVED***
    var oct = tag;
    tag = 0;
    while ((oct & 0x80) === 0x80) ***REMOVED***
      oct = buf.readUInt8(fail);
      if (buf.isError(oct))
        return oct;

      tag <<= 7;
      tag |= oct & 0x7f;
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    tag &= 0x1f;
  ***REMOVED***
  var tagStr = der.tag[tag];

  return ***REMOVED***
    cls: cls,
    primitive: primitive,
    tag: tag,
    tagStr: tagStr
  ***REMOVED***;
***REMOVED***

function derDecodeLen(buf, primitive, fail) ***REMOVED***
  var len = buf.readUInt8(fail);
  if (buf.isError(len))
    return len;

  // Indefinite form
  if (!primitive && len === 0x80)
    return null;

  // Definite form
  if ((len & 0x80) === 0) ***REMOVED***
    // Short form
    return len;
  ***REMOVED***

  // Long form
  var num = len & 0x7f;
  if (num > 4)
    return buf.error('length octect is too long');

  len = 0;
  for (var i = 0; i < num; i++) ***REMOVED***
    len <<= 8;
    var j = buf.readUInt8(fail);
    if (buf.isError(j))
      return j;
    len |= j;
  ***REMOVED***

  return len;
***REMOVED***
