// expand-brackets/lib/parsers.js
'use strict';

var utils = require('./utils');
var define = require('define-property');

/**
 * Text regex
 */

var TEXT_REGEX = '(\\[(?=.*\\])|\\])+';
var not = utils.createRegex(TEXT_REGEX);

/**
 * Brackets parsers
 */

function parsers(brackets) ***REMOVED***
  brackets.state = brackets.state || ***REMOVED******REMOVED***;
  brackets.parser.sets.bracket = brackets.parser.sets.bracket || [];
  brackets.parser

    .capture('escape', function() ***REMOVED***
      if (this.isInside('bracket')) return;
      var pos = this.position();
      var m = this.match(/^\\(.)/);
      if (!m) return;

      return pos(***REMOVED***
        type: 'escape',
        val: m[0]
      ***REMOVED***);
    ***REMOVED***)

    /**
     * Text parser
     */

    .capture('text', function() ***REMOVED***
      if (this.isInside('bracket')) return;
      var pos = this.position();
      var m = this.match(not);
      if (!m || !m[0]) return;

      return pos(***REMOVED***
        type: 'text',
        val: m[0]
      ***REMOVED***);
    ***REMOVED***)

    /**
     * POSIX character classes: "[[:alpha:][:digits:]]"
     */

    .capture('posix', function() ***REMOVED***
      var pos = this.position();
      var m = this.match(/^\[:(.*?):\](?=.*\])/);
      if (!m) return;

      var inside = this.isInside('bracket');
      if (inside) ***REMOVED***
        brackets.posix++;
      ***REMOVED***

      return pos(***REMOVED***
        type: 'posix',
        insideBracket: inside,
        inner: m[1],
        val: m[0]
      ***REMOVED***);
    ***REMOVED***)

    /**
     * Bracket (noop)
     */

    .capture('bracket', function() ***REMOVED******REMOVED***)

    /**
     * Open: '['
     */

    .capture('bracket.open', function() ***REMOVED***
      var parsed = this.parsed;
      var pos = this.position();
      var m = this.match(/^\[(?=.*\])/);
      if (!m) return;

      var prev = this.prev();
      var last = utils.last(prev.nodes);

      if (parsed.slice(-1) === '\\' && !this.isInside('bracket')) ***REMOVED***
        last.val = last.val.slice(0, last.val.length - 1);
        return pos(***REMOVED***
          type: 'escape',
          val: m[0]
        ***REMOVED***);
      ***REMOVED***

      var open = pos(***REMOVED***
        type: 'bracket.open',
        val: m[0]
      ***REMOVED***);

      if (last.type === 'bracket.open' || this.isInside('bracket')) ***REMOVED***
        open.val = '\\' + open.val;
        open.type = 'bracket.inner';
        open.escaped = true;
        return open;
      ***REMOVED***

      var node = pos(***REMOVED***
        type: 'bracket',
        nodes: [open]
      ***REMOVED***);

      define(node, 'parent', prev);
      define(open, 'parent', node);
      this.push('bracket', node);
      prev.nodes.push(node);
    ***REMOVED***)

    /**
     * Bracket text
     */

    .capture('bracket.inner', function() ***REMOVED***
      if (!this.isInside('bracket')) return;
      var pos = this.position();
      var m = this.match(not);
      if (!m || !m[0]) return;

      var next = this.input.charAt(0);
      var val = m[0];

      var node = pos(***REMOVED***
        type: 'bracket.inner',
        val: val
      ***REMOVED***);

      if (val === '\\\\') ***REMOVED***
        return node;
      ***REMOVED***

      var first = val.charAt(0);
      var last = val.slice(-1);

      if (first === '!') ***REMOVED***
        val = '^' + val.slice(1);
      ***REMOVED***

      if (last === '\\' || (val === '^' && next === ']')) ***REMOVED***
        val += this.input[0];
        this.consume(1);
      ***REMOVED***

      node.val = val;
      return node;
    ***REMOVED***)

    /**
     * Close: ']'
     */

    .capture('bracket.close', function() ***REMOVED***
      var parsed = this.parsed;
      var pos = this.position();
      var m = this.match(/^\]/);
      if (!m) return;

      var prev = this.prev();
      var last = utils.last(prev.nodes);

      if (parsed.slice(-1) === '\\' && !this.isInside('bracket')) ***REMOVED***
        last.val = last.val.slice(0, last.val.length - 1);

        return pos(***REMOVED***
          type: 'escape',
          val: m[0]
        ***REMOVED***);
      ***REMOVED***

      var node = pos(***REMOVED***
        type: 'bracket.close',
        rest: this.input,
        val: m[0]
      ***REMOVED***);

      if (last.type === 'bracket.open') ***REMOVED***
        node.type = 'bracket.inner';
        node.escaped = true;
        return node;
      ***REMOVED***

      var bracket = this.pop('bracket');
      if (!this.isType(bracket, 'bracket')) ***REMOVED***
        if (this.options.strict) ***REMOVED***
          throw new Error('missing opening "["');
        ***REMOVED***
        node.type = 'bracket.inner';
        node.escaped = true;
        return node;
      ***REMOVED***

      bracket.nodes.push(node);
      define(node, 'parent', bracket);
    ***REMOVED***);
***REMOVED***

/**
 * Brackets parsers
 */

module.exports = parsers;

/**
 * Expose text regex
 */

module.exports.TEXT_REGEX = TEXT_REGEX;
