// modelo/modelo/modelo.js
/*
The MIT License (MIT)
Copyright (c) 2012 Kevin Conway

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

/*jslint node: true, indent: 2, passfail: true, nomen: true */
/*global define */
"use strict";

function arrayFromArguments() ***REMOVED***
  var length = arguments.length,
    args = [],
    x;

  for (x = 0; x < length; x = x + 1) ***REMOVED***
    args[x] = arguments[x];
  ***REMOVED***

  return args;
***REMOVED***

function copyPrototype(take, give) ***REMOVED***
  var keys = Object.keys(give.prototype),
    length = keys.length,
    x;

  for (x = 0; x < length; x = x + 1) ***REMOVED***
    take.prototype[keys[x]] = give.prototype[keys[x]];
  ***REMOVED***
***REMOVED***

function ModeloBase() ***REMOVED***
  return this;
***REMOVED***

function isInstance(self, bases, base) ***REMOVED***

  var length = bases.length,
    x;

  if (base === self) ***REMOVED***
    return true;
  ***REMOVED***

  for (x = 0; x < length; x = x + 1) ***REMOVED***
    if (base === bases[x]) ***REMOVED***
      return true;
    ***REMOVED***

    if (!!bases[x].prototype.isInstance &&
          bases[x].prototype.isInstance(base)) ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***

  return false;

***REMOVED***

function extend() ***REMOVED***
  return define.apply(undefined, arguments);
***REMOVED***

function define() ***REMOVED***

  var constructors = arrayFromArguments.apply(undefined, arguments),
    length = constructors.length,
    x;

  function ModeloWrapper() ***REMOVED***

    var y;

    for (y = 0; y < length; y = y + 1) ***REMOVED***
      if (arguments.length > 0) ***REMOVED***
        constructors[y].apply(this, arguments);
      ***REMOVED*** else ***REMOVED***
        constructors[y].call(this, ***REMOVED******REMOVED***);
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***

  for (x = 0; x < length; x = x + 1) ***REMOVED***
    copyPrototype(ModeloWrapper, constructors[x]);
  ***REMOVED***
  if (length > 0) ***REMOVED***
    ModeloWrapper.constructor = constructors[length - 1].constructor;
  ***REMOVED***

  ModeloWrapper.prototype.isInstance = isInstance.bind(
    undefined,
    ModeloWrapper,
    constructors
  );
  ModeloWrapper.extend = extend.bind(undefined, ModeloWrapper);

  return ModeloWrapper;

***REMOVED***

function inherits(child, parent) ***REMOVED***

  var constructors = [parent],
    length,
    x;

  // This method brought to you by Node.js util module.
  // https://github.com/joyent/node/blob/master/lib/util.js
  child.super_ = parent;
  child.prototype = Object.create(
    parent.prototype,
    ***REMOVED***
      constructor: ***REMOVED***
        value: child,
        enumerable: false,
        writable: true,
        configurable: true
      ***REMOVED***
    ***REMOVED***
  );

  if (arguments.length > 2) ***REMOVED***

    constructors = arrayFromArguments.apply(undefined, arguments);
    constructors.shift();
    length = constructors.length;

    for (x = 1; x < length; x = x + 1) ***REMOVED***
      copyPrototype(child, constructors[x]);
    ***REMOVED***

  ***REMOVED***

  child.prototype.isInstance = isInstance.bind(
    undefined,
    child,
    constructors
  );
  child.extend = extend.bind(undefined, child);

  return child;

***REMOVED***

define.define = define;
define.inherits = inherits;

module.exports = define;
