// empower-core/index.js
/**
 * empower-core - Power Assert feature enhancer for assert function/object.
 *
 * https://github.com/twada/empower-core
 *
 * Copyright (c) 2013-2017 Takuto Wada
 * Licensed under the MIT license.
 *   https://github.com/twada/empower-core/blob/master/MIT-LICENSE.txt
 */
var create = require('core-js/library/fn/object/create');
var assign = require('core-js/library/fn/object/assign');
var defaultOptions = require('./lib/default-options');
var Decorator = require('./lib/decorator');
var define = require('./lib/define-properties');
var slice = Array.prototype.slice;

/**
 * Enhance Power Assert feature to assert function/object.
 * @param assert target assert function or object to enhance
 * @param options enhancement options
 * @return enhanced assert function/object
 */
function empowerCore (assert, options) ***REMOVED***
    var typeOfAssert = (typeof assert);
    var enhancedAssert;
    if ((typeOfAssert !== 'object' && typeOfAssert !== 'function') || assert === null) ***REMOVED***
        throw new TypeError('empower-core argument should be a function or object.');
    ***REMOVED***
    if (isEmpowered(assert)) ***REMOVED***
        return assert;
    ***REMOVED***
    switch (typeOfAssert) ***REMOVED***
    case 'function':
        enhancedAssert = empowerAssertFunction(assert, options);
        break;
    case 'object':
        enhancedAssert = empowerAssertObject(assert, options);
        break;
    default:
        throw new Error('Cannot be here');
    ***REMOVED***
    define(enhancedAssert, ***REMOVED*** _empowered: true ***REMOVED***);
    return enhancedAssert;
***REMOVED***

function empowerAssertObject (assertObject, options) ***REMOVED***
    var config = assign(defaultOptions(), options);
    var target = config.destructive ? assertObject : create(assertObject);
    var decorator = new Decorator(target, config);
    return assign(target, decorator.enhancement());
***REMOVED***

function empowerAssertFunction (assertFunction, options) ***REMOVED***
    var config = assign(defaultOptions(), options);
    if (config.destructive) ***REMOVED***
        throw new Error('cannot use destructive:true to function.');
    ***REMOVED***
    var decorator = new Decorator(assertFunction, config);
    var enhancement = decorator.enhancement();
    var powerAssert;
    if (typeof enhancement === 'function') ***REMOVED***
        powerAssert = function powerAssert () ***REMOVED***
            return enhancement.apply(null, slice.apply(arguments));
        ***REMOVED***;
    ***REMOVED*** else ***REMOVED***
        powerAssert = function powerAssert () ***REMOVED***
            return assertFunction.apply(null, slice.apply(arguments));
        ***REMOVED***;
    ***REMOVED***
    assign(powerAssert, assertFunction);
    return assign(powerAssert, enhancement);
***REMOVED***

function isEmpowered (assertObjectOrFunction) ***REMOVED***
    return assertObjectOrFunction._empowered;
***REMOVED***

empowerCore.defaultOptions = defaultOptions;
module.exports = empowerCore;
