// bytebuffer/src/bytebuffer.js
//? include("macros.js");
/**
 * Constructs a new ByteBuffer.
 * @class The swiss army knife for binary data in JavaScript.
 * @exports ByteBuffer
 * @constructor
 * @param ***REMOVED***number=***REMOVED*** capacity Initial capacity. Defaults to ***REMOVED***@link ByteBuffer.DEFAULT_CAPACITY***REMOVED***.
 * @param ***REMOVED***boolean=***REMOVED*** littleEndian Whether to use little or big endian byte order. Defaults to
 *  ***REMOVED***@link ByteBuffer.DEFAULT_ENDIAN***REMOVED***.
 * @param ***REMOVED***boolean=***REMOVED*** noAssert Whether to skip assertions of offsets and values. Defaults to
 *  ***REMOVED***@link ByteBuffer.DEFAULT_NOASSERT***REMOVED***.
 * @expose
 */
var ByteBuffer = function(capacity, littleEndian, noAssert) ***REMOVED***
    if (typeof capacity === 'undefined')
        capacity = ByteBuffer.DEFAULT_CAPACITY;
    if (typeof littleEndian === 'undefined')
        littleEndian = ByteBuffer.DEFAULT_ENDIAN;
    if (typeof noAssert === 'undefined')
        noAssert = ByteBuffer.DEFAULT_NOASSERT;
    if (!noAssert) ***REMOVED***
        capacity = capacity | 0;
        if (capacity < 0)
            throw RangeError("Illegal capacity");
        littleEndian = !!littleEndian;
        noAssert = !!noAssert;
    ***REMOVED***
    //? if (NODE) ***REMOVED***
    
    /**
     * Backing node Buffer.
     * @type ***REMOVED***!Buffer***REMOVED***
     * @expose
     */
    this.buffer = capacity === 0 ? EMPTY_BUFFER : new Buffer(capacity);
    //? ***REMOVED*** else ***REMOVED***
    
    /**
     * Backing ArrayBuffer.
     * @type ***REMOVED***!ArrayBuffer***REMOVED***
     * @expose
     */
    this.buffer = capacity === 0 ? EMPTY_BUFFER : new ArrayBuffer(capacity);

    //? if (DATAVIEW) ***REMOVED***
    /**
     * DataView utilized to manipulate the backing buffer. Becomes `null` if the backing buffer has a capacity of `0`.
     * @type ***REMOVED***?DataView***REMOVED***
     * @expose
     */
    this.view = capacity === 0 ? null : new DataView(this.buffer);
    //? ***REMOVED*** else ***REMOVED***
    /**
     * Uint8Array utilized to manipulate the backing buffer. Becomes `null` if the backing buffer has a capacity of `0`.
     * @type ***REMOVED***?Uint8Array***REMOVED***
     * @expose
     */
    this.view = capacity === 0 ? null : new Uint8Array(this.buffer);
    //? ***REMOVED***
    //? ***REMOVED***
    
    /**
     * Absolute read/write offset.
     * @type ***REMOVED***number***REMOVED***
     * @expose
     * @see ByteBuffer#flip
     * @see ByteBuffer#clear
     */
    this.offset = 0;

    /**
     * Marked offset.
     * @type ***REMOVED***number***REMOVED***
     * @expose
     * @see ByteBuffer#mark
     * @see ByteBuffer#reset
     */
    this.markedOffset = -1;

    /**
     * Absolute limit of the contained data. Set to the backing buffer's capacity upon allocation.
     * @type ***REMOVED***number***REMOVED***
     * @expose
     * @see ByteBuffer#flip
     * @see ByteBuffer#clear
     */
    this.limit = capacity;

    /**
     * Whether to use little endian byte order, defaults to `false` for big endian.
     * @type ***REMOVED***boolean***REMOVED***
     * @expose
     */
    this.littleEndian = littleEndian;

    /**
     * Whether to skip assertions of offsets and values, defaults to `false`.
     * @type ***REMOVED***boolean***REMOVED***
     * @expose
     */
    this.noAssert = noAssert;
***REMOVED***;

/**
 * ByteBuffer version.
 * @type ***REMOVED***string***REMOVED***
 * @const
 * @expose
 */
ByteBuffer.VERSION = "/*?= VERSION */";

/**
 * Little endian constant that can be used instead of its boolean value. Evaluates to `true`.
 * @type ***REMOVED***boolean***REMOVED***
 * @const
 * @expose
 */
ByteBuffer.LITTLE_ENDIAN = true;

/**
 * Big endian constant that can be used instead of its boolean value. Evaluates to `false`.
 * @type ***REMOVED***boolean***REMOVED***
 * @const
 * @expose
 */
ByteBuffer.BIG_ENDIAN = false;

/**
 * Default initial capacity of `16`.
 * @type ***REMOVED***number***REMOVED***
 * @expose
 */
ByteBuffer.DEFAULT_CAPACITY = 16;

/**
 * Default endianess of `false` for big endian.
 * @type ***REMOVED***boolean***REMOVED***
 * @expose
 */
ByteBuffer.DEFAULT_ENDIAN = ByteBuffer.BIG_ENDIAN;

/**
 * Default no assertions flag of `false`.
 * @type ***REMOVED***boolean***REMOVED***
 * @expose
 */
ByteBuffer.DEFAULT_NOASSERT = false;
//? if (NODE) ***REMOVED***

/**
 * A `Long` class for representing a 64-bit two's-complement integer value.
 * @type ***REMOVED***!Long***REMOVED***
 * @const
 * @see https://npmjs.org/package/long
 * @expose
 */
ByteBuffer.Long = Long;
//? ***REMOVED*** else ***REMOVED***

/**
 * A `Long` class for representing a 64-bit two's-complement integer value. May be `null` if Long.js has not been loaded
 *  and int64 support is not available.
 * @type ***REMOVED***?Long***REMOVED***
 * @const
 * @see https://github.com/dcodeIO/long.js
 * @expose
 */
ByteBuffer.Long = Long || null;
//? ***REMOVED***

/**
 * @alias ByteBuffer.prototype
 * @inner
 */
var ByteBufferPrototype = ByteBuffer.prototype;

/**
 * An indicator used to reliably determine if an object is a ByteBuffer or not.
 * @type ***REMOVED***boolean***REMOVED***
 * @const
 * @expose
 * @private
 */
ByteBufferPrototype.__isByteBuffer__;

Object.defineProperty(ByteBufferPrototype, "__isByteBuffer__", ***REMOVED***
    value: true,
    enumerable: false,
    configurable: false
***REMOVED***);

//? include("helpers.js");

//? include("methods/static/*.js");
//? if (BYTES) ***REMOVED***

//? include("types/bytes/*.js");
//? ***REMOVED***
//? if (INTS) ***REMOVED***

//? include("types/ints/*.js");
//? ***REMOVED***
//? if (FLOATS) ***REMOVED***

//? include("types/floats/float*.js");
//? ***REMOVED***
//? if (VARINTS) ***REMOVED***

//? include("types/varints/*.js");
//? ***REMOVED***
//? if (UTF8 && STRINGS) ***REMOVED***

//? include("types/strings/*.js");
//? ***REMOVED***

//? include("methods/*.js");
//? if (ENCODINGS) ***REMOVED***

//? include("encodings/*.js");
//? ***REMOVED***
