// acorn/src/locutil.js
import ***REMOVED***lineBreakG***REMOVED*** from "./whitespace"

// These are used when `options.locations` is on, for the
// `startLoc` and `endLoc` properties.

export class Position ***REMOVED***
  constructor(line, col) ***REMOVED***
    this.line = line
    this.column = col
  ***REMOVED***

  offset(n) ***REMOVED***
    return new Position(this.line, this.column + n)
  ***REMOVED***
***REMOVED***

export class SourceLocation ***REMOVED***
  constructor(p, start, end) ***REMOVED***
    this.start = start
    this.end = end
    if (p.sourceFile !== null) this.source = p.sourceFile
  ***REMOVED***
***REMOVED***

// The `getLineInfo` function is mostly useful when the
// `locations` option is off (for performance reasons) and you
// want to find the line/column position for a given character
// offset. `input` should be the code string that the offset refers
// into.

export function getLineInfo(input, offset) ***REMOVED***
  for (let line = 1, cur = 0;;) ***REMOVED***
    lineBreakG.lastIndex = cur
    let match = lineBreakG.exec(input)
    if (match && match.index < offset) ***REMOVED***
      ++line
      cur = match.index + match[0].length
    ***REMOVED*** else ***REMOVED***
      return new Position(line, offset - cur)
    ***REMOVED***
  ***REMOVED***
***REMOVED***
