// acorn/src/node.js
import ***REMOVED***Parser***REMOVED*** from "./state"
import ***REMOVED***SourceLocation***REMOVED*** from "./locutil"

export class Node ***REMOVED***
  constructor(parser, pos, loc) ***REMOVED***
    this.type = ""
    this.start = pos
    this.end = 0
    if (parser.options.locations)
      this.loc = new SourceLocation(parser, loc)
    if (parser.options.directSourceFile)
      this.sourceFile = parser.options.directSourceFile
    if (parser.options.ranges)
      this.range = [pos, 0]
  ***REMOVED***
***REMOVED***

// Start an AST node, attaching a start offset.

const pp = Parser.prototype

pp.startNode = function() ***REMOVED***
  return new Node(this, this.start, this.startLoc)
***REMOVED***

pp.startNodeAt = function(pos, loc) ***REMOVED***
  return new Node(this, pos, loc)
***REMOVED***

// Finish an AST node, adding `type` and `end` properties.

function finishNodeAt(node, type, pos, loc) ***REMOVED***
  node.type = type
  node.end = pos
  if (this.options.locations)
    node.loc.end = loc
  if (this.options.ranges)
    node.range[1] = pos
  return node
***REMOVED***

pp.finishNode = function(node, type) ***REMOVED***
  return finishNodeAt.call(this, node, type, this.lastTokEnd, this.lastTokEndLoc)
***REMOVED***

// Finish node at given position

pp.finishNodeAt = function(node, type, pos, loc) ***REMOVED***
  return finishNodeAt.call(this, node, type, pos, loc)
***REMOVED***
