// @firebase/firestore/dist/esm/src/local/memory_query_cache.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** SnapshotVersion ***REMOVED*** from '../core/snapshot_version';
import ***REMOVED*** ObjectMap ***REMOVED*** from '../util/obj_map';
import ***REMOVED*** PersistencePromise ***REMOVED*** from './persistence_promise';
import ***REMOVED*** ReferenceSet ***REMOVED*** from './reference_set';
var MemoryQueryCache = /** @class */ (function () ***REMOVED***
    function MemoryQueryCache() ***REMOVED***
        /**
         * Maps a query to the data about that query
         */
        this.queries = new ObjectMap(function (q) ***REMOVED*** return q.canonicalId(); ***REMOVED***);
        /** The last received snapshot version. */
        this.lastRemoteSnapshotVersion = SnapshotVersion.MIN;
        /** The highest numbered target ID encountered. */
        this.highestTargetId = 0;
        /**
         * A ordered bidirectional mapping between documents and the remote target
         * IDs.
         */
        this.references = new ReferenceSet();
    ***REMOVED***
    MemoryQueryCache.prototype.start = function (transaction) ***REMOVED***
        // Nothing to do.
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.getLastRemoteSnapshotVersion = function () ***REMOVED***
        return this.lastRemoteSnapshotVersion;
    ***REMOVED***;
    MemoryQueryCache.prototype.getHighestTargetId = function () ***REMOVED***
        return this.highestTargetId;
    ***REMOVED***;
    MemoryQueryCache.prototype.setLastRemoteSnapshotVersion = function (transaction, snapshotVersion) ***REMOVED***
        this.lastRemoteSnapshotVersion = snapshotVersion;
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.addQueryData = function (transaction, queryData) ***REMOVED***
        this.queries.set(queryData.query, queryData);
        var targetId = queryData.targetId;
        if (targetId > this.highestTargetId) ***REMOVED***
            this.highestTargetId = targetId;
        ***REMOVED***
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.removeQueryData = function (transaction, queryData) ***REMOVED***
        this.queries.delete(queryData.query);
        this.references.removeReferencesForId(queryData.targetId);
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.getQueryData = function (transaction, query) ***REMOVED***
        var queryData = this.queries.get(query) || null;
        return PersistencePromise.resolve(queryData);
    ***REMOVED***;
    MemoryQueryCache.prototype.addMatchingKeys = function (txn, keys, targetId) ***REMOVED***
        this.references.addReferences(keys, targetId);
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.removeMatchingKeys = function (txn, keys, targetId) ***REMOVED***
        this.references.removeReferences(keys, targetId);
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.removeMatchingKeysForTargetId = function (txn, targetId) ***REMOVED***
        this.references.removeReferencesForId(targetId);
        return PersistencePromise.resolve();
    ***REMOVED***;
    MemoryQueryCache.prototype.getMatchingKeysForTargetId = function (txn, targetId) ***REMOVED***
        var matchingKeys = this.references.referencesForId(targetId);
        return PersistencePromise.resolve(matchingKeys);
    ***REMOVED***;
    MemoryQueryCache.prototype.setGarbageCollector = function (gc) ***REMOVED***
        this.references.setGarbageCollector(gc);
    ***REMOVED***;
    MemoryQueryCache.prototype.containsKey = function (txn, key) ***REMOVED***
        return this.references.containsKey(txn, key);
    ***REMOVED***;
    return MemoryQueryCache;
***REMOVED***());
export ***REMOVED*** MemoryQueryCache ***REMOVED***;
