// @firebase/firestore/dist/esm/src/local/memory_persistence.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** assert ***REMOVED*** from '../util/assert';
import ***REMOVED*** debug ***REMOVED*** from '../util/log';
import ***REMOVED*** MemoryMutationQueue ***REMOVED*** from './memory_mutation_queue';
import ***REMOVED*** MemoryQueryCache ***REMOVED*** from './memory_query_cache';
import ***REMOVED*** MemoryRemoteDocumentCache ***REMOVED*** from './memory_remote_document_cache';
var LOG_TAG = 'MemoryPersistence';
/**
 * A memory-backed instance of Persistence. Data is stored only in RAM and
 * not persisted across sessions.
 */
var MemoryPersistence = /** @class */ (function () ***REMOVED***
    function MemoryPersistence() ***REMOVED***
        /**
         * Note that these are retained here to make it easier to write tests
         * affecting both the in-memory and IndexedDB-backed persistence layers. Tests
         * can create a new LocalStore wrapping this Persistence instance and this
         * will make the in-memory persistence layer behave as if it were actually
         * persisting values.
         */
        this.mutationQueues = ***REMOVED******REMOVED***;
        this.remoteDocumentCache = new MemoryRemoteDocumentCache();
        this.queryCache = new MemoryQueryCache();
        this.started = false;
    ***REMOVED***
    MemoryPersistence.prototype.start = function () ***REMOVED***
        assert(!this.started, 'MemoryPersistence double-started!');
        this.started = true;
        // No durable state to read on startup.
        return Promise.resolve();
    ***REMOVED***;
    MemoryPersistence.prototype.shutdown = function () ***REMOVED***
        // No durable state to ensure is closed on shutdown.
        assert(this.started, 'MemoryPersistence shutdown without start!');
        this.started = false;
        return Promise.resolve();
    ***REMOVED***;
    MemoryPersistence.prototype.getMutationQueue = function (user) ***REMOVED***
        var queue = this.mutationQueues[user.toKey()];
        if (!queue) ***REMOVED***
            queue = new MemoryMutationQueue();
            this.mutationQueues[user.toKey()] = queue;
        ***REMOVED***
        return queue;
    ***REMOVED***;
    MemoryPersistence.prototype.getQueryCache = function () ***REMOVED***
        return this.queryCache;
    ***REMOVED***;
    MemoryPersistence.prototype.getRemoteDocumentCache = function () ***REMOVED***
        return this.remoteDocumentCache;
    ***REMOVED***;
    MemoryPersistence.prototype.runTransaction = function (action, operation) ***REMOVED***
        debug(LOG_TAG, 'Starting transaction:', action);
        return operation(new MemoryPersistenceTransaction()).toPromise();
    ***REMOVED***;
    return MemoryPersistence;
***REMOVED***());
export ***REMOVED*** MemoryPersistence ***REMOVED***;
/** Dummy class since memory persistence doesn't actually use transactions. */
var MemoryPersistenceTransaction = /** @class */ (function () ***REMOVED***
    function MemoryPersistenceTransaction() ***REMOVED***
    ***REMOVED***
    return MemoryPersistenceTransaction;
***REMOVED***());
