// @firebase/firestore/dist/esm/src/local/remote_document_change_buffer.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** maybeDocumentMap ***REMOVED*** from '../model/collections';
import ***REMOVED*** assert ***REMOVED*** from '../util/assert';
import ***REMOVED*** PersistencePromise ***REMOVED*** from './persistence_promise';
/**
 * An in-memory buffer of entries to be written to a RemoteDocumentCache.
 * It can be used to batch up a set of changes to be written to the cache, but
 * additionally supports reading entries back with the `getEntry()` method,
 * falling back to the underlying RemoteDocumentCache if no entry is
 * buffered.
 *
 * NOTE: This class was introduced in iOS to work around a limitation in
 * LevelDB. Given IndexedDb has full transaction support with
 * read-your-own-writes capability, this class is not technically needed, but
 * has been preserved as a convenience and to aid portability.
 */
var RemoteDocumentChangeBuffer = /** @class */ (function () ***REMOVED***
    function RemoteDocumentChangeBuffer(remoteDocumentCache) ***REMOVED***
        this.remoteDocumentCache = remoteDocumentCache;
        this.changes = maybeDocumentMap();
    ***REMOVED***
    /** Buffers a `RemoteDocumentCache.addEntry()` call. */
    RemoteDocumentChangeBuffer.prototype.addEntry = function (maybeDocument) ***REMOVED***
        var changes = this.assertChanges();
        this.changes = changes.insert(maybeDocument.key, maybeDocument);
    ***REMOVED***;
    // NOTE: removeEntry() is not presently necessary and so is omitted.
    /**
     * Looks up an entry in the cache. The buffered changes will first be checked,
     * and if no buffered change applies, this will forward to
     * `RemoteDocumentCache.getEntry()`.
     *
     * @param transaction The transaction in which to perform any persistence
     *     operations.
     * @param documentKey The key of the entry to look up.
     * @return The cached Document or NoDocument entry, or null if we have nothing
     * cached.
     */
    RemoteDocumentChangeBuffer.prototype.getEntry = function (transaction, documentKey) ***REMOVED***
        var changes = this.assertChanges();
        var bufferedEntry = changes.get(documentKey);
        if (bufferedEntry) ***REMOVED***
            return PersistencePromise.resolve(bufferedEntry);
        ***REMOVED***
        else ***REMOVED***
            return this.remoteDocumentCache.getEntry(transaction, documentKey);
        ***REMOVED***
    ***REMOVED***;
    /**
     * Applies buffered changes to the underlying RemoteDocumentCache, using
     * the provided transaction.
     */
    RemoteDocumentChangeBuffer.prototype.apply = function (transaction) ***REMOVED***
        var _this = this;
        var changes = this.assertChanges();
        var promises = [];
        changes.forEach(function (key, maybeDoc) ***REMOVED***
            promises.push(_this.remoteDocumentCache.addEntry(transaction, maybeDoc));
        ***REMOVED***);
        // We should not be used to buffer any more changes.
        this.changes = null;
        return PersistencePromise.waitFor(promises);
    ***REMOVED***;
    /** Helper to assert this.changes is not null and return it. */
    RemoteDocumentChangeBuffer.prototype.assertChanges = function () ***REMOVED***
        assert(this.changes !== null, 'Changes have already been applied.');
        return this.changes;
    ***REMOVED***;
    return RemoteDocumentChangeBuffer;
***REMOVED***());
export ***REMOVED*** RemoteDocumentChangeBuffer ***REMOVED***;
