// acorn/dist/acorn_loose.es.js
import ***REMOVED*** defaultOptions, addLooseExports, SourceLocation, tokTypes, tokenizer, Node, lineBreak, isNewLine, getLineInfo, Token, lineBreakG ***REMOVED*** from './acorn';

// Registered plugins
var pluginsLoose = ***REMOVED******REMOVED***

var LooseParser = function LooseParser(input, options) ***REMOVED***
  if ( options === void 0 ) options = ***REMOVED******REMOVED***;

  this.toks = tokenizer(input, options)
  this.options = this.toks.options
  this.input = this.toks.input
  this.tok = this.last = ***REMOVED***type: tokTypes.eof, start: 0, end: 0***REMOVED***
  if (this.options.locations) ***REMOVED***
    var here = this.toks.curPosition()
    this.tok.loc = new SourceLocation(this.toks, here, here)
  ***REMOVED***
  this.ahead = [] // Tokens ahead
  this.context = [] // Indentation contexted
  this.curIndent = 0
  this.curLineStart = 0
  this.nextLineStart = this.lineEnd(this.curLineStart) + 1
  this.inAsync = false
  // Load plugins
  this.options.pluginsLoose = options.pluginsLoose || ***REMOVED******REMOVED***
  this.loadPlugins(this.options.pluginsLoose)
***REMOVED***;

LooseParser.prototype.startNode = function startNode () ***REMOVED***
  return new Node(this.toks, this.tok.start, this.options.locations ? this.tok.loc.start : null)
***REMOVED***;

LooseParser.prototype.storeCurrentPos = function storeCurrentPos () ***REMOVED***
  return this.options.locations ? [this.tok.start, this.tok.loc.start] : this.tok.start
***REMOVED***;

LooseParser.prototype.startNodeAt = function startNodeAt (pos) ***REMOVED***
  if (this.options.locations) ***REMOVED***
    return new Node(this.toks, pos[0], pos[1])
  ***REMOVED*** else ***REMOVED***
    return new Node(this.toks, pos)
  ***REMOVED***
***REMOVED***;

LooseParser.prototype.finishNode = function finishNode (node, type) ***REMOVED***
  node.type = type
  node.end = this.last.end
  if (this.options.locations)
    node.loc.end = this.last.loc.end
  if (this.options.ranges)
    node.range[1] = this.last.end
  return node
***REMOVED***;

LooseParser.prototype.dummyNode = function dummyNode (type) ***REMOVED***
  var dummy = this.startNode()
  dummy.type = type
  dummy.end = dummy.start
  if (this.options.locations)
    dummy.loc.end = dummy.loc.start
  if (this.options.ranges)
    dummy.range[1] = dummy.start
  this.last = ***REMOVED***type: tokTypes.name, start: dummy.start, end: dummy.start, loc: dummy.loc***REMOVED***
  return dummy
***REMOVED***;

LooseParser.prototype.dummyIdent = function dummyIdent () ***REMOVED***
  var dummy = this.dummyNode("Identifier")
  dummy.name = "✖"
  return dummy
***REMOVED***;

LooseParser.prototype.dummyString = function dummyString () ***REMOVED***
  var dummy = this.dummyNode("Literal")
  dummy.value = dummy.raw = "✖"
  return dummy
***REMOVED***;

LooseParser.prototype.eat = function eat (type) ***REMOVED***
  if (this.tok.type === type) ***REMOVED***
    this.next()
    return true
  ***REMOVED*** else ***REMOVED***
    return false
  ***REMOVED***
***REMOVED***;

LooseParser.prototype.isContextual = function isContextual (name) ***REMOVED***
  return this.tok.type === tokTypes.name && this.tok.value === name
***REMOVED***;

LooseParser.prototype.eatContextual = function eatContextual (name) ***REMOVED***
  return this.tok.value === name && this.eat(tokTypes.name)
***REMOVED***;

LooseParser.prototype.canInsertSemicolon = function canInsertSemicolon () ***REMOVED***
  return this.tok.type === tokTypes.eof || this.tok.type === tokTypes.braceR ||
    lineBreak.test(this.input.slice(this.last.end, this.tok.start))
***REMOVED***;

LooseParser.prototype.semicolon = function semicolon () ***REMOVED***
  return this.eat(tokTypes.semi)
***REMOVED***;

LooseParser.prototype.expect = function expect (type) ***REMOVED***
    var this$1 = this;

  if (this.eat(type)) return true
  for (var i = 1; i <= 2; i++) ***REMOVED***
    if (this$1.lookAhead(i).type == type) ***REMOVED***
      for (var j = 0; j < i; j++) this$1.next()
      return true
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

LooseParser.prototype.pushCx = function pushCx () ***REMOVED***
  this.context.push(this.curIndent)
***REMOVED***;

LooseParser.prototype.popCx = function popCx () ***REMOVED***
  this.curIndent = this.context.pop()
***REMOVED***;

LooseParser.prototype.lineEnd = function lineEnd (pos) ***REMOVED***
  while (pos < this.input.length && !isNewLine(this.input.charCodeAt(pos))) ++pos
  return pos
***REMOVED***;

LooseParser.prototype.indentationAfter = function indentationAfter (pos) ***REMOVED***
    var this$1 = this;

  for (var count = 0;; ++pos) ***REMOVED***
    var ch = this$1.input.charCodeAt(pos)
    if (ch === 32) ++count
    else if (ch === 9) count += this$1.options.tabSize
    else return count
  ***REMOVED***
***REMOVED***;

LooseParser.prototype.closes = function closes (closeTok, indent, line, blockHeuristic) ***REMOVED***
  if (this.tok.type === closeTok || this.tok.type === tokTypes.eof) return true
  return line != this.curLineStart && this.curIndent < indent && this.tokenStartsLine() &&
    (!blockHeuristic || this.nextLineStart >= this.input.length ||
     this.indentationAfter(this.nextLineStart) < indent)
***REMOVED***;

LooseParser.prototype.tokenStartsLine = function tokenStartsLine () ***REMOVED***
    var this$1 = this;

  for (var p = this.tok.start - 1; p >= this.curLineStart; --p) ***REMOVED***
    var ch = this$1.input.charCodeAt(p)
    if (ch !== 9 && ch !== 32) return false
  ***REMOVED***
  return true
***REMOVED***;

LooseParser.prototype.extend = function extend (name, f) ***REMOVED***
  this[name] = f(this[name])
***REMOVED***;

LooseParser.prototype.loadPlugins = function loadPlugins (pluginConfigs) ***REMOVED***
    var this$1 = this;

  for (var name in pluginConfigs) ***REMOVED***
    var plugin = pluginsLoose[name]
    if (!plugin) throw new Error("Plugin '" + name + "' not found")
    plugin(this$1, pluginConfigs[name])
  ***REMOVED***
***REMOVED***;

var lp = LooseParser.prototype

function isSpace(ch) ***REMOVED***
  return (ch < 14 && ch > 8) || ch === 32 || ch === 160 || isNewLine(ch)
***REMOVED***

lp.next = function() ***REMOVED***
  var this$1 = this;

  this.last = this.tok
  if (this.ahead.length)
    this.tok = this.ahead.shift()
  else
    this.tok = this.readToken()

  if (this.tok.start >= this.nextLineStart) ***REMOVED***
    while (this.tok.start >= this.nextLineStart) ***REMOVED***
      this$1.curLineStart = this$1.nextLineStart
      this$1.nextLineStart = this$1.lineEnd(this$1.curLineStart) + 1
    ***REMOVED***
    this.curIndent = this.indentationAfter(this.curLineStart)
  ***REMOVED***
***REMOVED***

lp.readToken = function() ***REMOVED***
  var this$1 = this;

  for (;;) ***REMOVED***
    try ***REMOVED***
      this$1.toks.next()
      if (this$1.toks.type === tokTypes.dot &&
          this$1.input.substr(this$1.toks.end, 1) === "." &&
          this$1.options.ecmaVersion >= 6) ***REMOVED***
        this$1.toks.end++
        this$1.toks.type = tokTypes.ellipsis
      ***REMOVED***
      return new Token(this$1.toks)
    ***REMOVED*** catch(e) ***REMOVED***
      if (!(e instanceof SyntaxError)) throw e

      // Try to skip some text, based on the error message, and then continue
      var msg = e.message, pos = e.raisedAt, replace = true
      if (/unterminated/i.test(msg)) ***REMOVED***
        pos = this$1.lineEnd(e.pos + 1)
        if (/string/.test(msg)) ***REMOVED***
          replace = ***REMOVED***start: e.pos, end: pos, type: tokTypes.string, value: this$1.input.slice(e.pos + 1, pos)***REMOVED***
        ***REMOVED*** else if (/regular expr/i.test(msg)) ***REMOVED***
          var re = this$1.input.slice(e.pos, pos)
          try ***REMOVED*** re = new RegExp(re) ***REMOVED*** catch(e) ***REMOVED******REMOVED***
          replace = ***REMOVED***start: e.pos, end: pos, type: tokTypes.regexp, value: re***REMOVED***
        ***REMOVED*** else if (/template/.test(msg)) ***REMOVED***
          replace = ***REMOVED***start: e.pos, end: pos,
                     type: tokTypes.template,
                     value: this$1.input.slice(e.pos, pos)***REMOVED***
        ***REMOVED*** else ***REMOVED***
          replace = false
        ***REMOVED***
      ***REMOVED*** else if (/invalid (unicode|regexp|number)|expecting unicode|octal literal|is reserved|directly after number|expected number in radix/i.test(msg)) ***REMOVED***
        while (pos < this.input.length && !isSpace(this.input.charCodeAt(pos))) ++pos
      ***REMOVED*** else if (/character escape|expected hexadecimal/i.test(msg)) ***REMOVED***
        while (pos < this.input.length) ***REMOVED***
          var ch = this$1.input.charCodeAt(pos++)
          if (ch === 34 || ch === 39 || isNewLine(ch)) break
        ***REMOVED***
      ***REMOVED*** else if (/unexpected character/i.test(msg)) ***REMOVED***
        pos++
        replace = false
      ***REMOVED*** else if (/regular expression/i.test(msg)) ***REMOVED***
        replace = true
      ***REMOVED*** else ***REMOVED***
        throw e
      ***REMOVED***
      this$1.resetTo(pos)
      if (replace === true) replace = ***REMOVED***start: pos, end: pos, type: tokTypes.name, value: "✖"***REMOVED***
      if (replace) ***REMOVED***
        if (this$1.options.locations)
          replace.loc = new SourceLocation(
            this$1.toks,
            getLineInfo(this$1.input, replace.start),
            getLineInfo(this$1.input, replace.end))
        return replace
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

lp.resetTo = function(pos) ***REMOVED***
  var this$1 = this;

  this.toks.pos = pos
  var ch = this.input.charAt(pos - 1)
  this.toks.exprAllowed = !ch || /[\[\***REMOVED***\(,;:?\/*=+\-~!|&%^<>]/.test(ch) ||
    /[enwfd]/.test(ch) &&
    /\b(keywords|case|else|return|throw|new|in|(instance|type)of|delete|void)$/.test(this.input.slice(pos - 10, pos))

  if (this.options.locations) ***REMOVED***
    this.toks.curLine = 1
    this.toks.lineStart = lineBreakG.lastIndex = 0
    var match
    while ((match = lineBreakG.exec(this.input)) && match.index < pos) ***REMOVED***
      ++this$1.toks.curLine
      this$1.toks.lineStart = match.index + match[0].length
    ***REMOVED***
  ***REMOVED***
***REMOVED***

lp.lookAhead = function(n) ***REMOVED***
  var this$1 = this;

  while (n > this.ahead.length)
    this$1.ahead.push(this$1.readToken())
  return this.ahead[n - 1]
***REMOVED***

function isDummy(node) ***REMOVED*** return node.name == "✖" ***REMOVED***

var lp$1 = LooseParser.prototype

lp$1.parseTopLevel = function() ***REMOVED***
  var this$1 = this;

  var node = this.startNodeAt(this.options.locations ? [0, getLineInfo(this.input, 0)] : 0)
  node.body = []
  while (this.tok.type !== tokTypes.eof) node.body.push(this$1.parseStatement())
  this.last = this.tok
  if (this.options.ecmaVersion >= 6) ***REMOVED***
    node.sourceType = this.options.sourceType
  ***REMOVED***
  return this.finishNode(node, "Program")
***REMOVED***

lp$1.parseStatement = function() ***REMOVED***
  var this$1 = this;

  var starttype = this.tok.type, node = this.startNode(), kind

  if (this.toks.isLet()) ***REMOVED***
    starttype = tokTypes._var
    kind = "let"
  ***REMOVED***

  switch (starttype) ***REMOVED***
  case tokTypes._break: case tokTypes._continue:
    this.next()
    var isBreak = starttype === tokTypes._break
    if (this.semicolon() || this.canInsertSemicolon()) ***REMOVED***
      node.label = null
    ***REMOVED*** else ***REMOVED***
      node.label = this.tok.type === tokTypes.name ? this.parseIdent() : null
      this.semicolon()
    ***REMOVED***
    return this.finishNode(node, isBreak ? "BreakStatement" : "ContinueStatement")

  case tokTypes._debugger:
    this.next()
    this.semicolon()
    return this.finishNode(node, "DebuggerStatement")

  case tokTypes._do:
    this.next()
    node.body = this.parseStatement()
    node.test = this.eat(tokTypes._while) ? this.parseParenExpression() : this.dummyIdent()
    this.semicolon()
    return this.finishNode(node, "DoWhileStatement")

  case tokTypes._for:
    this.next()
    this.pushCx()
    this.expect(tokTypes.parenL)
    if (this.tok.type === tokTypes.semi) return this.parseFor(node, null)
    var isLet = this.toks.isLet()
    if (isLet || this.tok.type === tokTypes._var || this.tok.type === tokTypes._const) ***REMOVED***
      var init$1 = this.parseVar(true, isLet ? "let" : this.tok.value)
      if (init$1.declarations.length === 1 && (this.tok.type === tokTypes._in || this.isContextual("of"))) ***REMOVED***
        return this.parseForIn(node, init$1)
      ***REMOVED***
      return this.parseFor(node, init$1)
    ***REMOVED***
    var init = this.parseExpression(true)
    if (this.tok.type === tokTypes._in || this.isContextual("of"))
      return this.parseForIn(node, this.toAssignable(init))
    return this.parseFor(node, init)

  case tokTypes._function:
    this.next()
    return this.parseFunction(node, true)

  case tokTypes._if:
    this.next()
    node.test = this.parseParenExpression()
    node.consequent = this.parseStatement()
    node.alternate = this.eat(tokTypes._else) ? this.parseStatement() : null
    return this.finishNode(node, "IfStatement")

  case tokTypes._return:
    this.next()
    if (this.eat(tokTypes.semi) || this.canInsertSemicolon()) node.argument = null
    else ***REMOVED*** node.argument = this.parseExpression(); this.semicolon() ***REMOVED***
    return this.finishNode(node, "ReturnStatement")

  case tokTypes._switch:
    var blockIndent = this.curIndent, line = this.curLineStart
    this.next()
    node.discriminant = this.parseParenExpression()
    node.cases = []
    this.pushCx()
    this.expect(tokTypes.braceL)

    var cur
    while (!this.closes(tokTypes.braceR, blockIndent, line, true)) ***REMOVED***
      if (this$1.tok.type === tokTypes._case || this$1.tok.type === tokTypes._default) ***REMOVED***
        var isCase = this$1.tok.type === tokTypes._case
        if (cur) this$1.finishNode(cur, "SwitchCase")
        node.cases.push(cur = this$1.startNode())
        cur.consequent = []
        this$1.next()
        if (isCase) cur.test = this$1.parseExpression()
        else cur.test = null
        this$1.expect(tokTypes.colon)
      ***REMOVED*** else ***REMOVED***
        if (!cur) ***REMOVED***
          node.cases.push(cur = this$1.startNode())
          cur.consequent = []
          cur.test = null
        ***REMOVED***
        cur.consequent.push(this$1.parseStatement())
      ***REMOVED***
    ***REMOVED***
    if (cur) this.finishNode(cur, "SwitchCase")
    this.popCx()
    this.eat(tokTypes.braceR)
    return this.finishNode(node, "SwitchStatement")

  case tokTypes._throw:
    this.next()
    node.argument = this.parseExpression()
    this.semicolon()
    return this.finishNode(node, "ThrowStatement")

  case tokTypes._try:
    this.next()
    node.block = this.parseBlock()
    node.handler = null
    if (this.tok.type === tokTypes._catch) ***REMOVED***
      var clause = this.startNode()
      this.next()
      this.expect(tokTypes.parenL)
      clause.param = this.toAssignable(this.parseExprAtom(), true)
      this.expect(tokTypes.parenR)
      clause.body = this.parseBlock()
      node.handler = this.finishNode(clause, "CatchClause")
    ***REMOVED***
    node.finalizer = this.eat(tokTypes._finally) ? this.parseBlock() : null
    if (!node.handler && !node.finalizer) return node.block
    return this.finishNode(node, "TryStatement")

  case tokTypes._var:
  case tokTypes._const:
    return this.parseVar(false, kind || this.tok.value)

  case tokTypes._while:
    this.next()
    node.test = this.parseParenExpression()
    node.body = this.parseStatement()
    return this.finishNode(node, "WhileStatement")

  case tokTypes._with:
    this.next()
    node.object = this.parseParenExpression()
    node.body = this.parseStatement()
    return this.finishNode(node, "WithStatement")

  case tokTypes.braceL:
    return this.parseBlock()

  case tokTypes.semi:
    this.next()
    return this.finishNode(node, "EmptyStatement")

  case tokTypes._class:
    return this.parseClass(true)

  case tokTypes._import:
    return this.parseImport()

  case tokTypes._export:
    return this.parseExport()

  default:
    if (this.toks.isAsyncFunction()) ***REMOVED***
      this.next()
      this.next()
      return this.parseFunction(node, true, true)
    ***REMOVED***
    var expr = this.parseExpression()
    if (isDummy(expr)) ***REMOVED***
      this.next()
      if (this.tok.type === tokTypes.eof) return this.finishNode(node, "EmptyStatement")
      return this.parseStatement()
    ***REMOVED*** else if (starttype === tokTypes.name && expr.type === "Identifier" && this.eat(tokTypes.colon)) ***REMOVED***
      node.body = this.parseStatement()
      node.label = expr
      return this.finishNode(node, "LabeledStatement")
    ***REMOVED*** else ***REMOVED***
      node.expression = expr
      this.semicolon()
      return this.finishNode(node, "ExpressionStatement")
    ***REMOVED***
  ***REMOVED***
***REMOVED***

lp$1.parseBlock = function() ***REMOVED***
  var this$1 = this;

  var node = this.startNode()
  this.pushCx()
  this.expect(tokTypes.braceL)
  var blockIndent = this.curIndent, line = this.curLineStart
  node.body = []
  while (!this.closes(tokTypes.braceR, blockIndent, line, true))
    node.body.push(this$1.parseStatement())
  this.popCx()
  this.eat(tokTypes.braceR)
  return this.finishNode(node, "BlockStatement")
***REMOVED***

lp$1.parseFor = function(node, init) ***REMOVED***
  node.init = init
  node.test = node.update = null
  if (this.eat(tokTypes.semi) && this.tok.type !== tokTypes.semi) node.test = this.parseExpression()
  if (this.eat(tokTypes.semi) && this.tok.type !== tokTypes.parenR) node.update = this.parseExpression()
  this.popCx()
  this.expect(tokTypes.parenR)
  node.body = this.parseStatement()
  return this.finishNode(node, "ForStatement")
***REMOVED***

lp$1.parseForIn = function(node, init) ***REMOVED***
  var type = this.tok.type === tokTypes._in ? "ForInStatement" : "ForOfStatement"
  this.next()
  node.left = init
  node.right = this.parseExpression()
  this.popCx()
  this.expect(tokTypes.parenR)
  node.body = this.parseStatement()
  return this.finishNode(node, type)
***REMOVED***

lp$1.parseVar = function(noIn, kind) ***REMOVED***
  var this$1 = this;

  var node = this.startNode()
  node.kind = kind
  this.next()
  node.declarations = []
  do ***REMOVED***
    var decl = this$1.startNode()
    decl.id = this$1.options.ecmaVersion >= 6 ? this$1.toAssignable(this$1.parseExprAtom(), true) : this$1.parseIdent()
    decl.init = this$1.eat(tokTypes.eq) ? this$1.parseMaybeAssign(noIn) : null
    node.declarations.push(this$1.finishNode(decl, "VariableDeclarator"))
  ***REMOVED*** while (this.eat(tokTypes.comma))
  if (!node.declarations.length) ***REMOVED***
    var decl$1 = this.startNode()
    decl$1.id = this.dummyIdent()
    node.declarations.push(this.finishNode(decl$1, "VariableDeclarator"))
  ***REMOVED***
  if (!noIn) this.semicolon()
  return this.finishNode(node, "VariableDeclaration")
***REMOVED***

lp$1.parseClass = function(isStatement) ***REMOVED***
  var this$1 = this;

  var node = this.startNode()
  this.next()
  if (isStatement == null) isStatement = this.tok.type === tokTypes.name
  if (this.tok.type === tokTypes.name) node.id = this.parseIdent()
  else if (isStatement) node.id = this.dummyIdent()
  else node.id = null
  node.superClass = this.eat(tokTypes._extends) ? this.parseExpression() : null
  node.body = this.startNode()
  node.body.body = []
  this.pushCx()
  var indent = this.curIndent + 1, line = this.curLineStart
  this.eat(tokTypes.braceL)
  if (this.curIndent + 1 < indent) ***REMOVED*** indent = this.curIndent; line = this.curLineStart ***REMOVED***
  while (!this.closes(tokTypes.braceR, indent, line)) ***REMOVED***
    if (this$1.semicolon()) continue
    var method = this$1.startNode(), isGenerator, isAsync
    if (this$1.options.ecmaVersion >= 6) ***REMOVED***
      method.static = false
      isGenerator = this$1.eat(tokTypes.star)
    ***REMOVED***
    this$1.parsePropertyName(method)
    if (isDummy(method.key)) ***REMOVED*** if (isDummy(this$1.parseMaybeAssign())) this$1.next(); this$1.eat(tokTypes.comma); continue ***REMOVED***
    if (method.key.type === "Identifier" && !method.computed && method.key.name === "static" &&
        (this$1.tok.type != tokTypes.parenL && this$1.tok.type != tokTypes.braceL)) ***REMOVED***
      method.static = true
      isGenerator = this$1.eat(tokTypes.star)
      this$1.parsePropertyName(method)
    ***REMOVED*** else ***REMOVED***
      method.static = false
    ***REMOVED***
    if (!method.computed &&
        method.key.type === "Identifier" && method.key.name === "async" && this$1.tok.type !== tokTypes.parenL &&
        !this$1.canInsertSemicolon()) ***REMOVED***
      this$1.parsePropertyName(method)
      isAsync = true
    ***REMOVED*** else ***REMOVED***
      isAsync = false
    ***REMOVED***
    if (this$1.options.ecmaVersion >= 5 && method.key.type === "Identifier" &&
        !method.computed && (method.key.name === "get" || method.key.name === "set") &&
        this$1.tok.type !== tokTypes.parenL && this$1.tok.type !== tokTypes.braceL) ***REMOVED***
      method.kind = method.key.name
      this$1.parsePropertyName(method)
      method.value = this$1.parseMethod(false)
    ***REMOVED*** else ***REMOVED***
      if (!method.computed && !method.static && !isGenerator && !isAsync && (
        method.key.type === "Identifier" && method.key.name === "constructor" ||
          method.key.type === "Literal" && method.key.value === "constructor")) ***REMOVED***
        method.kind = "constructor"
      ***REMOVED*** else ***REMOVED***
        method.kind =  "method"
      ***REMOVED***
      method.value = this$1.parseMethod(isGenerator, isAsync)
    ***REMOVED***
    node.body.body.push(this$1.finishNode(method, "MethodDefinition"))
  ***REMOVED***
  this.popCx()
  if (!this.eat(tokTypes.braceR)) ***REMOVED***
    // If there is no closing brace, make the node span to the start
    // of the next token (this is useful for Tern)
    this.last.end = this.tok.start
    if (this.options.locations) this.last.loc.end = this.tok.loc.start
  ***REMOVED***
  this.semicolon()
  this.finishNode(node.body, "ClassBody")
  return this.finishNode(node, isStatement ? "ClassDeclaration" : "ClassExpression")
***REMOVED***

lp$1.parseFunction = function(node, isStatement, isAsync) ***REMOVED***
  var oldInAsync = this.inAsync
  this.initFunction(node)
  if (this.options.ecmaVersion >= 6) ***REMOVED***
    node.generator = this.eat(tokTypes.star)
  ***REMOVED***
  if (this.options.ecmaVersion >= 8) ***REMOVED***
    node.async = !!isAsync
  ***REMOVED***
  if (isStatement == null) isStatement = this.tok.type === tokTypes.name
  if (this.tok.type === tokTypes.name) node.id = this.parseIdent()
  else if (isStatement) node.id = this.dummyIdent()
  this.inAsync = node.async
  node.params = this.parseFunctionParams()
  node.body = this.parseBlock()
  this.inAsync = oldInAsync
  return this.finishNode(node, isStatement ? "FunctionDeclaration" : "FunctionExpression")
***REMOVED***

lp$1.parseExport = function() ***REMOVED***
  var node = this.startNode()
  this.next()
  if (this.eat(tokTypes.star)) ***REMOVED***
    node.source = this.eatContextual("from") ? this.parseExprAtom() : this.dummyString()
    return this.finishNode(node, "ExportAllDeclaration")
  ***REMOVED***
  if (this.eat(tokTypes._default)) ***REMOVED***
    // export default (function foo() ***REMOVED******REMOVED***) // This is FunctionExpression.
    var isAsync
    if (this.tok.type === tokTypes._function || (isAsync = this.toks.isAsyncFunction())) ***REMOVED***
      var fNode = this.startNode()
      this.next()
      if (isAsync) this.next()
      node.declaration = this.parseFunction(fNode, null, isAsync)
    ***REMOVED*** else if (this.tok.type === tokTypes._class) ***REMOVED***
      node.declaration = this.parseClass(null)
    ***REMOVED*** else ***REMOVED***
      node.declaration = this.parseMaybeAssign()
      this.semicolon()
    ***REMOVED***
    return this.finishNode(node, "ExportDefaultDeclaration")
  ***REMOVED***
  if (this.tok.type.keyword || this.toks.isLet() || this.toks.isAsyncFunction()) ***REMOVED***
    node.declaration = this.parseStatement()
    node.specifiers = []
    node.source = null
  ***REMOVED*** else ***REMOVED***
    node.declaration = null
    node.specifiers = this.parseExportSpecifierList()
    node.source = this.eatContextual("from") ? this.parseExprAtom() : null
    this.semicolon()
  ***REMOVED***
  return this.finishNode(node, "ExportNamedDeclaration")
***REMOVED***

lp$1.parseImport = function() ***REMOVED***
  var node = this.startNode()
  this.next()
  if (this.tok.type === tokTypes.string) ***REMOVED***
    node.specifiers = []
    node.source = this.parseExprAtom()
    node.kind = ''
  ***REMOVED*** else ***REMOVED***
    var elt
    if (this.tok.type === tokTypes.name && this.tok.value !== "from") ***REMOVED***
      elt = this.startNode()
      elt.local = this.parseIdent()
      this.finishNode(elt, "ImportDefaultSpecifier")
      this.eat(tokTypes.comma)
    ***REMOVED***
    node.specifiers = this.parseImportSpecifierList()
    node.source = this.eatContextual("from") && this.tok.type == tokTypes.string ? this.parseExprAtom() : this.dummyString()
    if (elt) node.specifiers.unshift(elt)
  ***REMOVED***
  this.semicolon()
  return this.finishNode(node, "ImportDeclaration")
***REMOVED***

lp$1.parseImportSpecifierList = function() ***REMOVED***
  var this$1 = this;

  var elts = []
  if (this.tok.type === tokTypes.star) ***REMOVED***
    var elt = this.startNode()
    this.next()
    elt.local = this.eatContextual("as") ? this.parseIdent() : this.dummyIdent()
    elts.push(this.finishNode(elt, "ImportNamespaceSpecifier"))
  ***REMOVED*** else ***REMOVED***
    var indent = this.curIndent, line = this.curLineStart, continuedLine = this.nextLineStart
    this.pushCx()
    this.eat(tokTypes.braceL)
    if (this.curLineStart > continuedLine) continuedLine = this.curLineStart
    while (!this.closes(tokTypes.braceR, indent + (this.curLineStart <= continuedLine ? 1 : 0), line)) ***REMOVED***
      var elt$1 = this$1.startNode()
      if (this$1.eat(tokTypes.star)) ***REMOVED***
        elt$1.local = this$1.eatContextual("as") ? this$1.parseIdent() : this$1.dummyIdent()
        this$1.finishNode(elt$1, "ImportNamespaceSpecifier")
      ***REMOVED*** else ***REMOVED***
        if (this$1.isContextual("from")) break
        elt$1.imported = this$1.parseIdent()
        if (isDummy(elt$1.imported)) break
        elt$1.local = this$1.eatContextual("as") ? this$1.parseIdent() : elt$1.imported
        this$1.finishNode(elt$1, "ImportSpecifier")
      ***REMOVED***
      elts.push(elt$1)
      this$1.eat(tokTypes.comma)
    ***REMOVED***
    this.eat(tokTypes.braceR)
    this.popCx()
  ***REMOVED***
  return elts
***REMOVED***

lp$1.parseExportSpecifierList = function() ***REMOVED***
  var this$1 = this;

  var elts = []
  var indent = this.curIndent, line = this.curLineStart, continuedLine = this.nextLineStart
  this.pushCx()
  this.eat(tokTypes.braceL)
  if (this.curLineStart > continuedLine) continuedLine = this.curLineStart
  while (!this.closes(tokTypes.braceR, indent + (this.curLineStart <= continuedLine ? 1 : 0), line)) ***REMOVED***
    if (this$1.isContextual("from")) break
    var elt = this$1.startNode()
    elt.local = this$1.parseIdent()
    if (isDummy(elt.local)) break
    elt.exported = this$1.eatContextual("as") ? this$1.parseIdent() : elt.local
    this$1.finishNode(elt, "ExportSpecifier")
    elts.push(elt)
    this$1.eat(tokTypes.comma)
  ***REMOVED***
  this.eat(tokTypes.braceR)
  this.popCx()
  return elts
***REMOVED***

var lp$2 = LooseParser.prototype

lp$2.checkLVal = function(expr) ***REMOVED***
  if (!expr) return expr
  switch (expr.type) ***REMOVED***
  case "Identifier":
  case "MemberExpression":
    return expr

  case "ParenthesizedExpression":
    expr.expression = this.checkLVal(expr.expression)
    return expr

  default:
    return this.dummyIdent()
  ***REMOVED***
***REMOVED***

lp$2.parseExpression = function(noIn) ***REMOVED***
  var this$1 = this;

  var start = this.storeCurrentPos()
  var expr = this.parseMaybeAssign(noIn)
  if (this.tok.type === tokTypes.comma) ***REMOVED***
    var node = this.startNodeAt(start)
    node.expressions = [expr]
    while (this.eat(tokTypes.comma)) node.expressions.push(this$1.parseMaybeAssign(noIn))
    return this.finishNode(node, "SequenceExpression")
  ***REMOVED***
  return expr
***REMOVED***

lp$2.parseParenExpression = function() ***REMOVED***
  this.pushCx()
  this.expect(tokTypes.parenL)
  var val = this.parseExpression()
  this.popCx()
  this.expect(tokTypes.parenR)
  return val
***REMOVED***

lp$2.parseMaybeAssign = function(noIn) ***REMOVED***
  if (this.toks.isContextual("yield")) ***REMOVED***
    var node = this.startNode()
    this.next()
    if (this.semicolon() || this.canInsertSemicolon() || (this.tok.type != tokTypes.star && !this.tok.type.startsExpr)) ***REMOVED***
      node.delegate = false
      node.argument = null
    ***REMOVED*** else ***REMOVED***
      node.delegate = this.eat(tokTypes.star)
      node.argument = this.parseMaybeAssign()
    ***REMOVED***
    return this.finishNode(node, "YieldExpression")
  ***REMOVED***

  var start = this.storeCurrentPos()
  var left = this.parseMaybeConditional(noIn)
  if (this.tok.type.isAssign) ***REMOVED***
    var node$1 = this.startNodeAt(start)
    node$1.operator = this.tok.value
    node$1.left = this.tok.type === tokTypes.eq ? this.toAssignable(left) : this.checkLVal(left)
    this.next()
    node$1.right = this.parseMaybeAssign(noIn)
    return this.finishNode(node$1, "AssignmentExpression")
  ***REMOVED***
  return left
***REMOVED***

lp$2.parseMaybeConditional = function(noIn) ***REMOVED***
  var start = this.storeCurrentPos()
  var expr = this.parseExprOps(noIn)
  if (this.eat(tokTypes.question)) ***REMOVED***
    var node = this.startNodeAt(start)
    node.test = expr
    node.consequent = this.parseMaybeAssign()
    node.alternate = this.expect(tokTypes.colon) ? this.parseMaybeAssign(noIn) : this.dummyIdent()
    return this.finishNode(node, "ConditionalExpression")
  ***REMOVED***
  return expr
***REMOVED***

lp$2.parseExprOps = function(noIn) ***REMOVED***
  var start = this.storeCurrentPos()
  var indent = this.curIndent, line = this.curLineStart
  return this.parseExprOp(this.parseMaybeUnary(false), start, -1, noIn, indent, line)
***REMOVED***

lp$2.parseExprOp = function(left, start, minPrec, noIn, indent, line) ***REMOVED***
  if (this.curLineStart != line && this.curIndent < indent && this.tokenStartsLine()) return left
  var prec = this.tok.type.binop
  if (prec != null && (!noIn || this.tok.type !== tokTypes._in)) ***REMOVED***
    if (prec > minPrec) ***REMOVED***
      var node = this.startNodeAt(start)
      node.left = left
      node.operator = this.tok.value
      this.next()
      if (this.curLineStart != line && this.curIndent < indent && this.tokenStartsLine()) ***REMOVED***
        node.right = this.dummyIdent()
      ***REMOVED*** else ***REMOVED***
        var rightStart = this.storeCurrentPos()
        node.right = this.parseExprOp(this.parseMaybeUnary(false), rightStart, prec, noIn, indent, line)
      ***REMOVED***
      this.finishNode(node, /&&|\|\|/.test(node.operator) ? "LogicalExpression" : "BinaryExpression")
      return this.parseExprOp(node, start, minPrec, noIn, indent, line)
    ***REMOVED***
  ***REMOVED***
  return left
***REMOVED***

lp$2.parseMaybeUnary = function(sawUnary) ***REMOVED***
  var this$1 = this;

  var start = this.storeCurrentPos(), expr
  if (this.options.ecmaVersion >= 8 && this.inAsync && this.toks.isContextual("await")) ***REMOVED***
    expr = this.parseAwait()
    sawUnary = true
  ***REMOVED*** else if (this.tok.type.prefix) ***REMOVED***
    var node = this.startNode(), update = this.tok.type === tokTypes.incDec
    if (!update) sawUnary = true
    node.operator = this.tok.value
    node.prefix = true
    this.next()
    node.argument = this.parseMaybeUnary(true)
    if (update) node.argument = this.checkLVal(node.argument)
    expr = this.finishNode(node, update ? "UpdateExpression" : "UnaryExpression")
  ***REMOVED*** else if (this.tok.type === tokTypes.ellipsis) ***REMOVED***
    var node$1 = this.startNode()
    this.next()
    node$1.argument = this.parseMaybeUnary(sawUnary)
    expr = this.finishNode(node$1, "SpreadElement")
  ***REMOVED*** else ***REMOVED***
    expr = this.parseExprSubscripts()
    while (this.tok.type.postfix && !this.canInsertSemicolon()) ***REMOVED***
      var node$2 = this$1.startNodeAt(start)
      node$2.operator = this$1.tok.value
      node$2.prefix = false
      node$2.argument = this$1.checkLVal(expr)
      this$1.next()
      expr = this$1.finishNode(node$2, "UpdateExpression")
    ***REMOVED***
  ***REMOVED***

  if (!sawUnary && this.eat(tokTypes.starstar)) ***REMOVED***
    var node$3 = this.startNodeAt(start)
    node$3.operator = "**"
    node$3.left = expr
    node$3.right = this.parseMaybeUnary(false)
    return this.finishNode(node$3, "BinaryExpression")
  ***REMOVED***

  return expr
***REMOVED***

lp$2.parseExprSubscripts = function() ***REMOVED***
  var start = this.storeCurrentPos()
  return this.parseSubscripts(this.parseExprAtom(), start, false, this.curIndent, this.curLineStart)
***REMOVED***

lp$2.parseSubscripts = function(base, start, noCalls, startIndent, line) ***REMOVED***
  var this$1 = this;

  for (;;) ***REMOVED***
    if (this$1.curLineStart != line && this$1.curIndent <= startIndent && this$1.tokenStartsLine()) ***REMOVED***
      if (this$1.tok.type == tokTypes.dot && this$1.curIndent == startIndent)
        --startIndent
      else
        return base
    ***REMOVED***

    var maybeAsyncArrow = base.type === "Identifier" && base.name === "async" && !this$1.canInsertSemicolon()

    if (this$1.eat(tokTypes.dot)) ***REMOVED***
      var node = this$1.startNodeAt(start)
      node.object = base
      if (this$1.curLineStart != line && this$1.curIndent <= startIndent && this$1.tokenStartsLine())
        node.property = this$1.dummyIdent()
      else
        node.property = this$1.parsePropertyAccessor() || this$1.dummyIdent()
      node.computed = false
      base = this$1.finishNode(node, "MemberExpression")
    ***REMOVED*** else if (this$1.tok.type == tokTypes.bracketL) ***REMOVED***
      this$1.pushCx()
      this$1.next()
      var node$1 = this$1.startNodeAt(start)
      node$1.object = base
      node$1.property = this$1.parseExpression()
      node$1.computed = true
      this$1.popCx()
      this$1.expect(tokTypes.bracketR)
      base = this$1.finishNode(node$1, "MemberExpression")
    ***REMOVED*** else if (!noCalls && this$1.tok.type == tokTypes.parenL) ***REMOVED***
      var exprList = this$1.parseExprList(tokTypes.parenR)
      if (maybeAsyncArrow && this$1.eat(tokTypes.arrow))
        return this$1.parseArrowExpression(this$1.startNodeAt(start), exprList, true)
      var node$2 = this$1.startNodeAt(start)
      node$2.callee = base
      node$2.arguments = exprList
      base = this$1.finishNode(node$2, "CallExpression")
    ***REMOVED*** else if (this$1.tok.type == tokTypes.backQuote) ***REMOVED***
      var node$3 = this$1.startNodeAt(start)
      node$3.tag = base
      node$3.quasi = this$1.parseTemplate()
      base = this$1.finishNode(node$3, "TaggedTemplateExpression")
    ***REMOVED*** else ***REMOVED***
      return base
    ***REMOVED***
  ***REMOVED***
***REMOVED***

lp$2.parseExprAtom = function() ***REMOVED***
  var node
  switch (this.tok.type) ***REMOVED***
  case tokTypes._this:
  case tokTypes._super:
    var type = this.tok.type === tokTypes._this ? "ThisExpression" : "Super"
    node = this.startNode()
    this.next()
    return this.finishNode(node, type)

  case tokTypes.name:
    var start = this.storeCurrentPos()
    var id = this.parseIdent()
    var isAsync = false
    if (id.name === "async" && !this.canInsertSemicolon()) ***REMOVED***
      if (this.eat(tokTypes._function))
        return this.parseFunction(this.startNodeAt(start), false, true)
      if (this.tok.type === tokTypes.name) ***REMOVED***
        id = this.parseIdent()
        isAsync = true
      ***REMOVED***
    ***REMOVED***
    return this.eat(tokTypes.arrow) ? this.parseArrowExpression(this.startNodeAt(start), [id], isAsync) : id

  case tokTypes.regexp:
    node = this.startNode()
    var val = this.tok.value
    node.regex = ***REMOVED***pattern: val.pattern, flags: val.flags***REMOVED***
    node.value = val.value
    node.raw = this.input.slice(this.tok.start, this.tok.end)
    this.next()
    return this.finishNode(node, "Literal")

  case tokTypes.num: case tokTypes.string:
    node = this.startNode()
    node.value = this.tok.value
    node.raw = this.input.slice(this.tok.start, this.tok.end)
    this.next()
    return this.finishNode(node, "Literal")

  case tokTypes._null: case tokTypes._true: case tokTypes._false:
    node = this.startNode()
    node.value = this.tok.type === tokTypes._null ? null : this.tok.type === tokTypes._true
    node.raw = this.tok.type.keyword
    this.next()
    return this.finishNode(node, "Literal")

  case tokTypes.parenL:
    var parenStart = this.storeCurrentPos()
    this.next()
    var inner = this.parseExpression()
    this.expect(tokTypes.parenR)
    if (this.eat(tokTypes.arrow)) ***REMOVED***
      // (a,)=>a // SequenceExpression makes dummy in the last hole. Drop the dummy.
      var params = inner.expressions || [inner]
      if (params.length && isDummy(params[params.length - 1]))
        params.pop()
      return this.parseArrowExpression(this.startNodeAt(parenStart), params)
    ***REMOVED***
    if (this.options.preserveParens) ***REMOVED***
      var par = this.startNodeAt(parenStart)
      par.expression = inner
      inner = this.finishNode(par, "ParenthesizedExpression")
    ***REMOVED***
    return inner

  case tokTypes.bracketL:
    node = this.startNode()
    node.elements = this.parseExprList(tokTypes.bracketR, true)
    return this.finishNode(node, "ArrayExpression")

  case tokTypes.braceL:
    return this.parseObj()

  case tokTypes._class:
    return this.parseClass(false)

  case tokTypes._function:
    node = this.startNode()
    this.next()
    return this.parseFunction(node, false)

  case tokTypes._new:
    return this.parseNew()

  case tokTypes.backQuote:
    return this.parseTemplate()

  default:
    return this.dummyIdent()
  ***REMOVED***
***REMOVED***

lp$2.parseNew = function() ***REMOVED***
  var node = this.startNode(), startIndent = this.curIndent, line = this.curLineStart
  var meta = this.parseIdent(true)
  if (this.options.ecmaVersion >= 6 && this.eat(tokTypes.dot)) ***REMOVED***
    node.meta = meta
    node.property = this.parseIdent(true)
    return this.finishNode(node, "MetaProperty")
  ***REMOVED***
  var start = this.storeCurrentPos()
  node.callee = this.parseSubscripts(this.parseExprAtom(), start, true, startIndent, line)
  if (this.tok.type == tokTypes.parenL) ***REMOVED***
    node.arguments = this.parseExprList(tokTypes.parenR)
  ***REMOVED*** else ***REMOVED***
    node.arguments = []
  ***REMOVED***
  return this.finishNode(node, "NewExpression")
***REMOVED***

lp$2.parseTemplateElement = function() ***REMOVED***
  var elem = this.startNode()
  elem.value = ***REMOVED***
    raw: this.input.slice(this.tok.start, this.tok.end).replace(/\r\n?/g, '\n'),
    cooked: this.tok.value
  ***REMOVED***
  this.next()
  elem.tail = this.tok.type === tokTypes.backQuote
  return this.finishNode(elem, "TemplateElement")
***REMOVED***

lp$2.parseTemplate = function() ***REMOVED***
  var this$1 = this;

  var node = this.startNode()
  this.next()
  node.expressions = []
  var curElt = this.parseTemplateElement()
  node.quasis = [curElt]
  while (!curElt.tail) ***REMOVED***
    this$1.next()
    node.expressions.push(this$1.parseExpression())
    if (this$1.expect(tokTypes.braceR)) ***REMOVED***
      curElt = this$1.parseTemplateElement()
    ***REMOVED*** else ***REMOVED***
      curElt = this$1.startNode()
      curElt.value = ***REMOVED***cooked: '', raw: ''***REMOVED***
      curElt.tail = true
      this$1.finishNode(curElt, "TemplateElement")
    ***REMOVED***
    node.quasis.push(curElt)
  ***REMOVED***
  this.expect(tokTypes.backQuote)
  return this.finishNode(node, "TemplateLiteral")
***REMOVED***

lp$2.parseObj = function() ***REMOVED***
  var this$1 = this;

  var node = this.startNode()
  node.properties = []
  this.pushCx()
  var indent = this.curIndent + 1, line = this.curLineStart
  this.eat(tokTypes.braceL)
  if (this.curIndent + 1 < indent) ***REMOVED*** indent = this.curIndent; line = this.curLineStart ***REMOVED***
  while (!this.closes(tokTypes.braceR, indent, line)) ***REMOVED***
    var prop = this$1.startNode(), isGenerator, isAsync, start
    if (this$1.options.ecmaVersion >= 6) ***REMOVED***
      start = this$1.storeCurrentPos()
      prop.method = false
      prop.shorthand = false
      isGenerator = this$1.eat(tokTypes.star)
    ***REMOVED***
    this$1.parsePropertyName(prop)
    if (!prop.computed &&
        prop.key.type === "Identifier" && prop.key.name === "async" && this$1.tok.type !== tokTypes.parenL &&
        this$1.tok.type !== tokTypes.colon && !this$1.canInsertSemicolon()) ***REMOVED***
      this$1.parsePropertyName(prop)
      isAsync = true
    ***REMOVED*** else ***REMOVED***
      isAsync = false
    ***REMOVED***
    if (isDummy(prop.key)) ***REMOVED*** if (isDummy(this$1.parseMaybeAssign())) this$1.next(); this$1.eat(tokTypes.comma); continue ***REMOVED***
    if (this$1.eat(tokTypes.colon)) ***REMOVED***
      prop.kind = "init"
      prop.value = this$1.parseMaybeAssign()
    ***REMOVED*** else if (this$1.options.ecmaVersion >= 6 && (this$1.tok.type === tokTypes.parenL || this$1.tok.type === tokTypes.braceL)) ***REMOVED***
      prop.kind = "init"
      prop.method = true
      prop.value = this$1.parseMethod(isGenerator, isAsync)
    ***REMOVED*** else if (this$1.options.ecmaVersion >= 5 && prop.key.type === "Identifier" &&
               !prop.computed && (prop.key.name === "get" || prop.key.name === "set") &&
               (this$1.tok.type != tokTypes.comma && this$1.tok.type != tokTypes.braceR)) ***REMOVED***
      prop.kind = prop.key.name
      this$1.parsePropertyName(prop)
      prop.value = this$1.parseMethod(false)
    ***REMOVED*** else ***REMOVED***
      prop.kind = "init"
      if (this$1.options.ecmaVersion >= 6) ***REMOVED***
        if (this$1.eat(tokTypes.eq)) ***REMOVED***
          var assign = this$1.startNodeAt(start)
          assign.operator = "="
          assign.left = prop.key
          assign.right = this$1.parseMaybeAssign()
          prop.value = this$1.finishNode(assign, "AssignmentExpression")
        ***REMOVED*** else ***REMOVED***
          prop.value = prop.key
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        prop.value = this$1.dummyIdent()
      ***REMOVED***
      prop.shorthand = true
    ***REMOVED***
    node.properties.push(this$1.finishNode(prop, "Property"))
    this$1.eat(tokTypes.comma)
  ***REMOVED***
  this.popCx()
  if (!this.eat(tokTypes.braceR)) ***REMOVED***
    // If there is no closing brace, make the node span to the start
    // of the next token (this is useful for Tern)
    this.last.end = this.tok.start
    if (this.options.locations) this.last.loc.end = this.tok.loc.start
  ***REMOVED***
  return this.finishNode(node, "ObjectExpression")
***REMOVED***

lp$2.parsePropertyName = function(prop) ***REMOVED***
  if (this.options.ecmaVersion >= 6) ***REMOVED***
    if (this.eat(tokTypes.bracketL)) ***REMOVED***
      prop.computed = true
      prop.key = this.parseExpression()
      this.expect(tokTypes.bracketR)
      return
    ***REMOVED*** else ***REMOVED***
      prop.computed = false
    ***REMOVED***
  ***REMOVED***
  var key = (this.tok.type === tokTypes.num || this.tok.type === tokTypes.string) ? this.parseExprAtom() : this.parseIdent()
  prop.key = key || this.dummyIdent()
***REMOVED***

lp$2.parsePropertyAccessor = function() ***REMOVED***
  if (this.tok.type === tokTypes.name || this.tok.type.keyword) return this.parseIdent()
***REMOVED***

lp$2.parseIdent = function() ***REMOVED***
  var name = this.tok.type === tokTypes.name ? this.tok.value : this.tok.type.keyword
  if (!name) return this.dummyIdent()
  var node = this.startNode()
  this.next()
  node.name = name
  return this.finishNode(node, "Identifier")
***REMOVED***

lp$2.initFunction = function(node) ***REMOVED***
  node.id = null
  node.params = []
  if (this.options.ecmaVersion >= 6) ***REMOVED***
    node.generator = false
    node.expression = false
  ***REMOVED***
  if (this.options.ecmaVersion >= 8)
    node.async = false
***REMOVED***

// Convert existing expression atom to assignable pattern
// if possible.

lp$2.toAssignable = function(node, binding) ***REMOVED***
  var this$1 = this;

  if (!node || node.type == "Identifier" || (node.type == "MemberExpression" && !binding)) ***REMOVED***
    // Okay
  ***REMOVED*** else if (node.type == "ParenthesizedExpression") ***REMOVED***
    node.expression = this.toAssignable(node.expression, binding)
  ***REMOVED*** else if (this.options.ecmaVersion < 6) ***REMOVED***
    return this.dummyIdent()
  ***REMOVED*** else if (node.type == "ObjectExpression") ***REMOVED***
    node.type = "ObjectPattern"
    var props = node.properties
    for (var i = 0; i < props.length; i++)
      props[i].value = this$1.toAssignable(props[i].value, binding)
  ***REMOVED*** else if (node.type == "ArrayExpression") ***REMOVED***
    node.type = "ArrayPattern"
    this.toAssignableList(node.elements, binding)
  ***REMOVED*** else if (node.type == "SpreadElement") ***REMOVED***
    node.type = "RestElement"
    node.argument = this.toAssignable(node.argument, binding)
  ***REMOVED*** else if (node.type == "AssignmentExpression") ***REMOVED***
    node.type = "AssignmentPattern"
    delete node.operator
  ***REMOVED*** else ***REMOVED***
    return this.dummyIdent()
  ***REMOVED***
  return node
***REMOVED***

lp$2.toAssignableList = function(exprList, binding) ***REMOVED***
  var this$1 = this;

  for (var i = 0; i < exprList.length; i++)
    exprList[i] = this$1.toAssignable(exprList[i], binding)
  return exprList
***REMOVED***

lp$2.parseFunctionParams = function(params) ***REMOVED***
  params = this.parseExprList(tokTypes.parenR)
  return this.toAssignableList(params, true)
***REMOVED***

lp$2.parseMethod = function(isGenerator, isAsync) ***REMOVED***
  var node = this.startNode(), oldInAsync = this.inAsync
  this.initFunction(node)
  if (this.options.ecmaVersion >= 6)
    node.generator = !!isGenerator
  if (this.options.ecmaVersion >= 8)
    node.async = !!isAsync
  this.inAsync = node.async
  node.params = this.parseFunctionParams()
  node.expression = this.options.ecmaVersion >= 6 && this.tok.type !== tokTypes.braceL
  node.body = node.expression ? this.parseMaybeAssign() : this.parseBlock()
  this.inAsync = oldInAsync
  return this.finishNode(node, "FunctionExpression")
***REMOVED***

lp$2.parseArrowExpression = function(node, params, isAsync) ***REMOVED***
  var oldInAsync = this.inAsync
  this.initFunction(node)
  if (this.options.ecmaVersion >= 8)
    node.async = !!isAsync
  this.inAsync = node.async
  node.params = this.toAssignableList(params, true)
  node.expression = this.tok.type !== tokTypes.braceL
  node.body = node.expression ? this.parseMaybeAssign() : this.parseBlock()
  this.inAsync = oldInAsync
  return this.finishNode(node, "ArrowFunctionExpression")
***REMOVED***

lp$2.parseExprList = function(close, allowEmpty) ***REMOVED***
  var this$1 = this;

  this.pushCx()
  var indent = this.curIndent, line = this.curLineStart, elts = []
  this.next() // Opening bracket
  while (!this.closes(close, indent + 1, line)) ***REMOVED***
    if (this$1.eat(tokTypes.comma)) ***REMOVED***
      elts.push(allowEmpty ? null : this$1.dummyIdent())
      continue
    ***REMOVED***
    var elt = this$1.parseMaybeAssign()
    if (isDummy(elt)) ***REMOVED***
      if (this$1.closes(close, indent, line)) break
      this$1.next()
    ***REMOVED*** else ***REMOVED***
      elts.push(elt)
    ***REMOVED***
    this$1.eat(tokTypes.comma)
  ***REMOVED***
  this.popCx()
  if (!this.eat(close)) ***REMOVED***
    // If there is no closing brace, make the node span to the start
    // of the next token (this is useful for Tern)
    this.last.end = this.tok.start
    if (this.options.locations) this.last.loc.end = this.tok.loc.start
  ***REMOVED***
  return elts
***REMOVED***

lp$2.parseAwait = function() ***REMOVED***
  var node = this.startNode()
  this.next()
  node.argument = this.parseMaybeUnary()
  return this.finishNode(node, "AwaitExpression")
***REMOVED***

// Acorn: Loose parser
//
// This module provides an alternative parser (`parse_dammit`) that
// exposes that same interface as `parse`, but will try to parse
// anything as JavaScript, repairing syntax error the best it can.
// There are circumstances in which it will raise an error and give
// up, but they are very rare. The resulting AST will be a mostly
// valid JavaScript AST (as per the [Mozilla parser API][api], except
// that:
//
// - Return outside functions is allowed
//
// - Label consistency (no conflicts, break only to existing labels)
//   is not enforced.
//
// - Bogus Identifier nodes with a name of `"✖"` are inserted whenever
//   the parser got too confused to return anything meaningful.
//
// [api]: https://developer.mozilla.org/en-US/docs/SpiderMonkey/Parser_API
//
// The expected use for this is to *first* try `acorn.parse`, and only
// if that fails switch to `parse_dammit`. The loose parser might
// parse badly indented code incorrectly, so **don't** use it as
// your default parser.
//
// Quite a lot of acorn.js is duplicated here. The alternative was to
// add a *lot* of extra cruft to that file, making it less readable
// and slower. Copying and editing the code allowed me to make
// invasive changes and simplifications without creating a complicated
// tangle.

defaultOptions.tabSize = 4

function parse_dammit(input, options) ***REMOVED***
  var p = new LooseParser(input, options)
  p.next()
  return p.parseTopLevel()
***REMOVED***

addLooseExports(parse_dammit, LooseParser, pluginsLoose)

export ***REMOVED*** parse_dammit, LooseParser, pluginsLoose ***REMOVED***;
