// acorn/dist/walk.js
(function (global, factory) ***REMOVED***
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (factory((global.acorn = global.acorn || ***REMOVED******REMOVED***, global.acorn.walk = global.acorn.walk || ***REMOVED******REMOVED***)));
***REMOVED***(this, (function (exports) ***REMOVED*** 'use strict';

// AST walker module for Mozilla Parser API compatible trees

// A simple walk is one where you simply specify callbacks to be
// called on specific nodes. The last two arguments are optional. A
// simple use would be
//
//     walk.simple(myTree, ***REMOVED***
//         Expression: function(node) ***REMOVED*** ... ***REMOVED***
//     ***REMOVED***);
//
// to do something with all expressions. All Parser API node types
// can be used to identify node types, as well as Expression,
// Statement, and ScopeBody, which denote categories of nodes.
//
// The base argument can be used to pass a custom (recursive)
// walker, and state can be used to give this walked an initial
// state.

function simple(node, visitors, base, state, override) ***REMOVED***
  if (!base) base = exports.base
  ;(function c(node, st, override) ***REMOVED***
    var type = override || node.type, found = visitors[type]
    base[type](node, st, c)
    if (found) found(node, st)
  ***REMOVED***)(node, state, override)
***REMOVED***

// An ancestor walk keeps an array of ancestor nodes (including the
// current node) and passes them to the callback as third parameter
// (and also as state parameter when no other state is present).
function ancestor(node, visitors, base, state) ***REMOVED***
  if (!base) base = exports.base
  var ancestors = []
  ;(function c(node, st, override) ***REMOVED***
    var type = override || node.type, found = visitors[type]
    var isNew = node != ancestors[ancestors.length - 1]
    if (isNew) ancestors.push(node)
    base[type](node, st, c)
    if (found) found(node, st || ancestors, ancestors)
    if (isNew) ancestors.pop()
  ***REMOVED***)(node, state)
***REMOVED***

// A recursive walk is one where your functions override the default
// walkers. They can modify and replace the state parameter that's
// threaded through the walk, and can opt how and whether to walk
// their child nodes (by calling their third argument on these
// nodes).
function recursive(node, state, funcs, base, override) ***REMOVED***
  var visitor = funcs ? exports.make(funcs, base) : base
  ;(function c(node, st, override) ***REMOVED***
    visitor[override || node.type](node, st, c)
  ***REMOVED***)(node, state, override)
***REMOVED***

function makeTest(test) ***REMOVED***
  if (typeof test == "string")
    return function (type) ***REMOVED*** return type == test; ***REMOVED***
  else if (!test)
    return function () ***REMOVED*** return true; ***REMOVED***
  else
    return test
***REMOVED***

var Found = function Found(node, state) ***REMOVED*** this.node = node; this.state = state ***REMOVED***;

// Find a node with a given start, end, and type (all are optional,
// null can be used as wildcard). Returns a ***REMOVED***node, state***REMOVED*** object, or
// undefined when it doesn't find a matching node.
function findNodeAt(node, start, end, test, base, state) ***REMOVED***
  test = makeTest(test)
  if (!base) base = exports.base
  try ***REMOVED***
    ;(function c(node, st, override) ***REMOVED***
      var type = override || node.type
      if ((start == null || node.start <= start) &&
          (end == null || node.end >= end))
        base[type](node, st, c)
      if ((start == null || node.start == start) &&
          (end == null || node.end == end) &&
          test(type, node))
        throw new Found(node, st)
    ***REMOVED***)(node, state)
  ***REMOVED*** catch (e) ***REMOVED***
    if (e instanceof Found) return e
    throw e
  ***REMOVED***
***REMOVED***

// Find the innermost node of a given type that contains the given
// position. Interface similar to findNodeAt.
function findNodeAround(node, pos, test, base, state) ***REMOVED***
  test = makeTest(test)
  if (!base) base = exports.base
  try ***REMOVED***
    ;(function c(node, st, override) ***REMOVED***
      var type = override || node.type
      if (node.start > pos || node.end < pos) return
      base[type](node, st, c)
      if (test(type, node)) throw new Found(node, st)
    ***REMOVED***)(node, state)
  ***REMOVED*** catch (e) ***REMOVED***
    if (e instanceof Found) return e
    throw e
  ***REMOVED***
***REMOVED***

// Find the outermost matching node after a given position.
function findNodeAfter(node, pos, test, base, state) ***REMOVED***
  test = makeTest(test)
  if (!base) base = exports.base
  try ***REMOVED***
    ;(function c(node, st, override) ***REMOVED***
      if (node.end < pos) return
      var type = override || node.type
      if (node.start >= pos && test(type, node)) throw new Found(node, st)
      base[type](node, st, c)
    ***REMOVED***)(node, state)
  ***REMOVED*** catch (e) ***REMOVED***
    if (e instanceof Found) return e
    throw e
  ***REMOVED***
***REMOVED***

// Find the outermost matching node before a given position.
function findNodeBefore(node, pos, test, base, state) ***REMOVED***
  test = makeTest(test)
  if (!base) base = exports.base
  var max
  ;(function c(node, st, override) ***REMOVED***
    if (node.start > pos) return
    var type = override || node.type
    if (node.end <= pos && (!max || max.node.end < node.end) && test(type, node))
      max = new Found(node, st)
    base[type](node, st, c)
  ***REMOVED***)(node, state)
  return max
***REMOVED***

// Fallback to an Object.create polyfill for older environments.
var create = Object.create || function(proto) ***REMOVED***
  function Ctor() ***REMOVED******REMOVED***
  Ctor.prototype = proto
  return new Ctor
***REMOVED***

// Used to create a custom walker. Will fill in all missing node
// type properties with the defaults.
function make(funcs, base) ***REMOVED***
  if (!base) base = exports.base
  var visitor = create(base)
  for (var type in funcs) visitor[type] = funcs[type]
  return visitor
***REMOVED***

function skipThrough(node, st, c) ***REMOVED*** c(node, st) ***REMOVED***
function ignore(_node, _st, _c) ***REMOVED******REMOVED***

// Node walkers.

var base = ***REMOVED******REMOVED***

base.Program = base.BlockStatement = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.body.length; ++i)
    c(node.body[i], st, "Statement")
***REMOVED***
base.Statement = skipThrough
base.EmptyStatement = ignore
base.ExpressionStatement = base.ParenthesizedExpression =
  function (node, st, c) ***REMOVED*** return c(node.expression, st, "Expression"); ***REMOVED***
base.IfStatement = function (node, st, c) ***REMOVED***
  c(node.test, st, "Expression")
  c(node.consequent, st, "Statement")
  if (node.alternate) c(node.alternate, st, "Statement")
***REMOVED***
base.LabeledStatement = function (node, st, c) ***REMOVED*** return c(node.body, st, "Statement"); ***REMOVED***
base.BreakStatement = base.ContinueStatement = ignore
base.WithStatement = function (node, st, c) ***REMOVED***
  c(node.object, st, "Expression")
  c(node.body, st, "Statement")
***REMOVED***
base.SwitchStatement = function (node, st, c) ***REMOVED***
  c(node.discriminant, st, "Expression")
  for (var i = 0; i < node.cases.length; ++i) ***REMOVED***
    var cs = node.cases[i]
    if (cs.test) c(cs.test, st, "Expression")
    for (var j = 0; j < cs.consequent.length; ++j)
      c(cs.consequent[j], st, "Statement")
  ***REMOVED***
***REMOVED***
base.ReturnStatement = base.YieldExpression = base.AwaitExpression = function (node, st, c) ***REMOVED***
  if (node.argument) c(node.argument, st, "Expression")
***REMOVED***
base.ThrowStatement = base.SpreadElement =
  function (node, st, c) ***REMOVED*** return c(node.argument, st, "Expression"); ***REMOVED***
base.TryStatement = function (node, st, c) ***REMOVED***
  c(node.block, st, "Statement")
  if (node.handler) c(node.handler, st)
  if (node.finalizer) c(node.finalizer, st, "Statement")
***REMOVED***
base.CatchClause = function (node, st, c) ***REMOVED***
  c(node.param, st, "Pattern")
  c(node.body, st, "ScopeBody")
***REMOVED***
base.WhileStatement = base.DoWhileStatement = function (node, st, c) ***REMOVED***
  c(node.test, st, "Expression")
  c(node.body, st, "Statement")
***REMOVED***
base.ForStatement = function (node, st, c) ***REMOVED***
  if (node.init) c(node.init, st, "ForInit")
  if (node.test) c(node.test, st, "Expression")
  if (node.update) c(node.update, st, "Expression")
  c(node.body, st, "Statement")
***REMOVED***
base.ForInStatement = base.ForOfStatement = function (node, st, c) ***REMOVED***
  c(node.left, st, "ForInit")
  c(node.right, st, "Expression")
  c(node.body, st, "Statement")
***REMOVED***
base.ForInit = function (node, st, c) ***REMOVED***
  if (node.type == "VariableDeclaration") c(node, st)
  else c(node, st, "Expression")
***REMOVED***
base.DebuggerStatement = ignore

base.FunctionDeclaration = function (node, st, c) ***REMOVED*** return c(node, st, "Function"); ***REMOVED***
base.VariableDeclaration = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.declarations.length; ++i)
    c(node.declarations[i], st)
***REMOVED***
base.VariableDeclarator = function (node, st, c) ***REMOVED***
  c(node.id, st, "Pattern")
  if (node.init) c(node.init, st, "Expression")
***REMOVED***

base.Function = function (node, st, c) ***REMOVED***
  if (node.id) c(node.id, st, "Pattern")
  for (var i = 0; i < node.params.length; i++)
    c(node.params[i], st, "Pattern")
  c(node.body, st, node.expression ? "ScopeExpression" : "ScopeBody")
***REMOVED***
// FIXME drop these node types in next major version
// (They are awkward, and in ES6 every block can be a scope.)
base.ScopeBody = function (node, st, c) ***REMOVED*** return c(node, st, "Statement"); ***REMOVED***
base.ScopeExpression = function (node, st, c) ***REMOVED*** return c(node, st, "Expression"); ***REMOVED***

base.Pattern = function (node, st, c) ***REMOVED***
  if (node.type == "Identifier")
    c(node, st, "VariablePattern")
  else if (node.type == "MemberExpression")
    c(node, st, "MemberPattern")
  else
    c(node, st)
***REMOVED***
base.VariablePattern = ignore
base.MemberPattern = skipThrough
base.RestElement = function (node, st, c) ***REMOVED*** return c(node.argument, st, "Pattern"); ***REMOVED***
base.ArrayPattern =  function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.elements.length; ++i) ***REMOVED***
    var elt = node.elements[i]
    if (elt) c(elt, st, "Pattern")
  ***REMOVED***
***REMOVED***
base.ObjectPattern = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.properties.length; ++i)
    c(node.properties[i].value, st, "Pattern")
***REMOVED***

base.Expression = skipThrough
base.ThisExpression = base.Super = base.MetaProperty = ignore
base.ArrayExpression = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.elements.length; ++i) ***REMOVED***
    var elt = node.elements[i]
    if (elt) c(elt, st, "Expression")
  ***REMOVED***
***REMOVED***
base.ObjectExpression = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.properties.length; ++i)
    c(node.properties[i], st)
***REMOVED***
base.FunctionExpression = base.ArrowFunctionExpression = base.FunctionDeclaration
base.SequenceExpression = base.TemplateLiteral = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.expressions.length; ++i)
    c(node.expressions[i], st, "Expression")
***REMOVED***
base.UnaryExpression = base.UpdateExpression = function (node, st, c) ***REMOVED***
  c(node.argument, st, "Expression")
***REMOVED***
base.BinaryExpression = base.LogicalExpression = function (node, st, c) ***REMOVED***
  c(node.left, st, "Expression")
  c(node.right, st, "Expression")
***REMOVED***
base.AssignmentExpression = base.AssignmentPattern = function (node, st, c) ***REMOVED***
  c(node.left, st, "Pattern")
  c(node.right, st, "Expression")
***REMOVED***
base.ConditionalExpression = function (node, st, c) ***REMOVED***
  c(node.test, st, "Expression")
  c(node.consequent, st, "Expression")
  c(node.alternate, st, "Expression")
***REMOVED***
base.NewExpression = base.CallExpression = function (node, st, c) ***REMOVED***
  c(node.callee, st, "Expression")
  if (node.arguments) for (var i = 0; i < node.arguments.length; ++i)
    c(node.arguments[i], st, "Expression")
***REMOVED***
base.MemberExpression = function (node, st, c) ***REMOVED***
  c(node.object, st, "Expression")
  if (node.computed) c(node.property, st, "Expression")
***REMOVED***
base.ExportNamedDeclaration = base.ExportDefaultDeclaration = function (node, st, c) ***REMOVED***
  if (node.declaration)
    c(node.declaration, st, node.type == "ExportNamedDeclaration" || node.declaration.id ? "Statement" : "Expression")
  if (node.source) c(node.source, st, "Expression")
***REMOVED***
base.ExportAllDeclaration = function (node, st, c) ***REMOVED***
  c(node.source, st, "Expression")
***REMOVED***
base.ImportDeclaration = function (node, st, c) ***REMOVED***
  for (var i = 0; i < node.specifiers.length; i++)
    c(node.specifiers[i], st)
  c(node.source, st, "Expression")
***REMOVED***
base.ImportSpecifier = base.ImportDefaultSpecifier = base.ImportNamespaceSpecifier = base.Identifier = base.Literal = ignore

base.TaggedTemplateExpression = function (node, st, c) ***REMOVED***
  c(node.tag, st, "Expression")
  c(node.quasi, st)
***REMOVED***
base.ClassDeclaration = base.ClassExpression = function (node, st, c) ***REMOVED*** return c(node, st, "Class"); ***REMOVED***
base.Class = function (node, st, c) ***REMOVED***
  if (node.id) c(node.id, st, "Pattern")
  if (node.superClass) c(node.superClass, st, "Expression")
  for (var i = 0; i < node.body.body.length; i++)
    c(node.body.body[i], st)
***REMOVED***
base.MethodDefinition = base.Property = function (node, st, c) ***REMOVED***
  if (node.computed) c(node.key, st, "Expression")
  c(node.value, st, "Expression")
***REMOVED***

exports.simple = simple;
exports.ancestor = ancestor;
exports.recursive = recursive;
exports.findNodeAt = findNodeAt;
exports.findNodeAround = findNodeAround;
exports.findNodeAfter = findNodeAfter;
exports.findNodeBefore = findNodeBefore;
exports.make = make;
exports.base = base;

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true ***REMOVED***);

***REMOVED***)));
