// grpc/index.js
/**
 * @license
 * Copyright 2015 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

var path = require('path');
var fs = require('fs');

var SSL_ROOTS_PATH = path.resolve(__dirname, 'deps', 'grpc', 'etc', 'roots.pem');

var _ = require('lodash');

var ProtoBuf = require('protobufjs');

var client = require('./src/client.js');

var server = require('./src/server.js');

var common = require('./src/common.js');

var Metadata = require('./src/metadata.js');

var grpc = require('./src/grpc_extension');

var protobuf_js_5_common = require('./src/protobuf_js_5_common');
var protobuf_js_6_common = require('./src/protobuf_js_6_common');

var constants = require('./src/constants.js');

grpc.setDefaultRootsPem(fs.readFileSync(SSL_ROOTS_PATH, 'ascii'));

/**
 * @namespace grpc
 */

/**
 * Load a ProtoBuf.js object as a gRPC object.
 * @memberof grpc
 * @alias grpc.loadObject
 * @param ***REMOVED***Object***REMOVED*** value The ProtoBuf.js reflection object to load
 * @param ***REMOVED***Object=***REMOVED*** options Options to apply to the loaded file
 * @param ***REMOVED***bool=***REMOVED*** [options.binaryAsBase64=false] deserialize bytes values as
 *     base64 strings instead of Buffers
 * @param ***REMOVED***bool=***REMOVED*** [options.longsAsStrings=true] deserialize long values as
 *     strings instead of objects
 * @param ***REMOVED***bool=***REMOVED*** [options.enumsAsStrings=true] deserialize enum values as
 *     strings instead of numbers. Only works with Protobuf.js 6 values.
 * @param ***REMOVED***bool=***REMOVED*** [options.deprecatedArgumentOrder=false] use the beta method
 *     argument order for client methods, with optional arguments after the
 *     callback. This option is only a temporary stopgap measure to smooth an
 *     API breakage. It is deprecated, and new code should not use it.
 * @param ***REMOVED***(number|string)=***REMOVED*** [options.protobufjsVersion='detect'] 5 and 6
 *     respectively indicate that an object from the corresponding version of
 *     Protobuf.js is provided in the value argument. If the option is 'detect',
 *     gRPC will guess what the version is based on the structure of the value.
 * @return ***REMOVED***Object<string, *>***REMOVED*** The resulting gRPC object.
 */
exports.loadObject = function loadObject(value, options) ***REMOVED***
  options = _.defaults(options, common.defaultGrpcOptions);
  options = _.defaults(options, ***REMOVED***'protobufjsVersion': 'detect'***REMOVED***);
  var protobufjsVersion;
  if (options.protobufjsVersion === 'detect') ***REMOVED***
    if (protobuf_js_6_common.isProbablyProtobufJs6(value)) ***REMOVED***
      protobufjsVersion = 6;
    ***REMOVED*** else if (protobuf_js_5_common.isProbablyProtobufJs5(value)) ***REMOVED***
      protobufjsVersion = 5;
    ***REMOVED*** else ***REMOVED***
      var error_message = 'Could not detect ProtoBuf.js version. Please ' +
          'specify the version number with the "protobufjs_version" option';
      throw new Error(error_message);
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    protobufjsVersion = options.protobufjsVersion;
  ***REMOVED***
  switch (protobufjsVersion) ***REMOVED***
    case 6: return protobuf_js_6_common.loadObject(value, options);
    case 5:
    return protobuf_js_5_common.loadObject(value, options);
    default:
    throw new Error('Unrecognized protobufjsVersion', protobufjsVersion);
  ***REMOVED***
***REMOVED***;

var loadObject = exports.loadObject;

/**
 * Load a gRPC object from a .proto file.
 * @memberof grpc
 * @alias grpc.load
 * @param ***REMOVED***string|***REMOVED***root: string, file: string***REMOVED******REMOVED*** filename The file to load
 * @param ***REMOVED***string=***REMOVED*** format The file format to expect. Must be either 'proto' or
 *     'json'. Defaults to 'proto'
 * @param ***REMOVED***Object=***REMOVED*** options Options to apply to the loaded file
 * @param ***REMOVED***bool=***REMOVED*** [options.convertFieldsToCamelCase=false] Load this file with
 *     field names in camel case instead of their original case
 * @param ***REMOVED***bool=***REMOVED*** [options.binaryAsBase64=false] deserialize bytes values as
 *     base64 strings instead of Buffers
 * @param ***REMOVED***bool=***REMOVED*** [options.longsAsStrings=true] deserialize long values as
 *     strings instead of objects
 * @param ***REMOVED***bool=***REMOVED*** [options.deprecatedArgumentOrder=false] use the beta method
 *     argument order for client methods, with optional arguments after the
 *     callback. This option is only a temporary stopgap measure to smooth an
 *     API breakage. It is deprecated, and new code should not use it.
 * @return ***REMOVED***Object<string, *>***REMOVED*** The resulting gRPC object
 */
exports.load = function load(filename, format, options) ***REMOVED***
  options = _.defaults(options, common.defaultGrpcOptions);
  options.protobufjsVersion = 5;
  if (!format) ***REMOVED***
    format = 'proto';
  ***REMOVED***
  var convertFieldsToCamelCaseOriginal = ProtoBuf.convertFieldsToCamelCase;
  if(options && options.hasOwnProperty('convertFieldsToCamelCase')) ***REMOVED***
    ProtoBuf.convertFieldsToCamelCase = options.convertFieldsToCamelCase;
  ***REMOVED***
  var builder;
  try ***REMOVED***
    switch(format) ***REMOVED***
      case 'proto':
      builder = ProtoBuf.loadProtoFile(filename);
      break;
      case 'json':
      builder = ProtoBuf.loadJsonFile(filename);
      break;
      default:
      throw new Error('Unrecognized format "' + format + '"');
    ***REMOVED***
  ***REMOVED*** finally ***REMOVED***
    ProtoBuf.convertFieldsToCamelCase = convertFieldsToCamelCaseOriginal;
  ***REMOVED***
  return loadObject(builder.ns, options);
***REMOVED***;

var log_template = _.template(
    '***REMOVED***severity***REMOVED*** ***REMOVED***timestamp***REMOVED***\t***REMOVED***file***REMOVED***:***REMOVED***line***REMOVED***]\t***REMOVED***message***REMOVED***',
    ***REMOVED***interpolate: /***REMOVED***([\s\S]+?)***REMOVED***/g***REMOVED***);

/**
 * Sets the logger function for the gRPC module. For debugging purposes, the C
 * core will log synchronously directly to stdout unless this function is
 * called. Note: the output format here is intended to be informational, and
 * is not guaranteed to stay the same in the future.
 * Logs will be directed to logger.error.
 * @memberof grpc
 * @alias grpc.setLogger
 * @param ***REMOVED***Console***REMOVED*** logger A Console-like object.
 */
exports.setLogger = function setLogger(logger) ***REMOVED***
  common.logger = logger;
  grpc.setDefaultLoggerCallback(function(file, line, severity,
                                         message, timestamp) ***REMOVED***
    logger.error(log_template(***REMOVED***
      file: path.basename(file),
      line: line,
      severity: severity,
      message: message,
      timestamp: timestamp.toISOString()
    ***REMOVED***));
  ***REMOVED***);
***REMOVED***;

/**
 * Sets the logger verbosity for gRPC module logging. The options are members
 * of the grpc.logVerbosity map.
 * @memberof grpc
 * @alias grpc.setLogVerbosity
 * @param ***REMOVED***Number***REMOVED*** verbosity The minimum severity to log
 */
exports.setLogVerbosity = function setLogVerbosity(verbosity) ***REMOVED***
  common.logVerbosity = verbosity;
  grpc.setLogVerbosity(verbosity);
***REMOVED***;

exports.Server = server.Server;

exports.Metadata = Metadata;

exports.status = constants.status;

exports.propagate = constants.propagate;

exports.callError = constants.callError;

exports.writeFlags = constants.writeFlags;

exports.logVerbosity = constants.logVerbosity;

exports.credentials = require('./src/credentials.js');

/**
 * ServerCredentials factories
 * @constructor ServerCredentials
 * @memberof grpc
 */
exports.ServerCredentials = grpc.ServerCredentials;

/**
 * Create insecure server credentials
 * @name grpc.ServerCredentials.createInsecure
 * @kind function
 * @return ***REMOVED***grpc.ServerCredentials***REMOVED***
 */

/**
 * A private key and certificate pair
 * @typedef ***REMOVED***Object***REMOVED*** grpc.ServerCredentials~keyCertPair
 * @property ***REMOVED***Buffer***REMOVED*** privateKey The server's private key
 * @property ***REMOVED***Buffer***REMOVED*** certChain The server's certificate chain
 */

/**
 * Create SSL server credentials
 * @name grpc.ServerCredentials.createSsl
 * @kind function
 * @param ***REMOVED***?Buffer***REMOVED*** rootCerts Root CA certificates for validating client
 *     certificates
 * @param ***REMOVED***Array<grpc.ServerCredentials~keyCertPair>***REMOVED*** keyCertPairs A list of
 *     private key and certificate chain pairs to be used for authenticating
 *     the server
 * @param ***REMOVED***boolean***REMOVED*** [checkClientCertificate=false] Indicates that the server
 *     should request and verify the client's certificates
 * @return ***REMOVED***grpc.ServerCredentials***REMOVED***
 */

exports.makeGenericClientConstructor = client.makeClientConstructor;

exports.getClientChannel = client.getClientChannel;

exports.waitForClientReady = client.waitForClientReady;

/**
 * @memberof grpc
 * @alias grpc.closeClient
 * @param ***REMOVED***grpc.Client***REMOVED*** client_obj The client to close
 */
exports.closeClient = function closeClient(client_obj) ***REMOVED***
  client.Client.prototype.close.apply(client_obj);
***REMOVED***;

exports.Client = client.Client;
