// @firebase/database/dist/cjs/src/core/view/filter/RangedFilter.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var IndexedFilter_1 = require("./IndexedFilter");
var PriorityIndex_1 = require("../../snap/indexes/PriorityIndex");
var Node_1 = require("../../../core/snap/Node");
var ChildrenNode_1 = require("../../snap/ChildrenNode");
/**
 * Filters nodes by range and uses an IndexFilter to track any changes after filtering the node
 *
 * @constructor
 * @implements ***REMOVED***NodeFilter***REMOVED***
 */
var RangedFilter = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***!QueryParams***REMOVED*** params
     */
    function RangedFilter(params) ***REMOVED***
        this.indexedFilter_ = new IndexedFilter_1.IndexedFilter(params.getIndex());
        this.index_ = params.getIndex();
        this.startPost_ = RangedFilter.getStartPost_(params);
        this.endPost_ = RangedFilter.getEndPost_(params);
    ***REMOVED***
    /**
     * @return ***REMOVED***!NamedNode***REMOVED***
     */
    RangedFilter.prototype.getStartPost = function () ***REMOVED***
        return this.startPost_;
    ***REMOVED***;
    /**
     * @return ***REMOVED***!NamedNode***REMOVED***
     */
    RangedFilter.prototype.getEndPost = function () ***REMOVED***
        return this.endPost_;
    ***REMOVED***;
    /**
     * @param ***REMOVED***!NamedNode***REMOVED*** node
     * @return ***REMOVED***boolean***REMOVED***
     */
    RangedFilter.prototype.matches = function (node) ***REMOVED***
        return (this.index_.compare(this.getStartPost(), node) <= 0 &&
            this.index_.compare(node, this.getEndPost()) <= 0);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    RangedFilter.prototype.updateChild = function (snap, key, newChild, affectedPath, source, optChangeAccumulator) ***REMOVED***
        if (!this.matches(new Node_1.NamedNode(key, newChild))) ***REMOVED***
            newChild = ChildrenNode_1.ChildrenNode.EMPTY_NODE;
        ***REMOVED***
        return this.indexedFilter_.updateChild(snap, key, newChild, affectedPath, source, optChangeAccumulator);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    RangedFilter.prototype.updateFullNode = function (oldSnap, newSnap, optChangeAccumulator) ***REMOVED***
        if (newSnap.isLeafNode()) ***REMOVED***
            // Make sure we have a children node with the correct index, not a leaf node;
            newSnap = ChildrenNode_1.ChildrenNode.EMPTY_NODE;
        ***REMOVED***
        var filtered = newSnap.withIndex(this.index_);
        // Don't support priorities on queries
        filtered = filtered.updatePriority(ChildrenNode_1.ChildrenNode.EMPTY_NODE);
        var self = this;
        newSnap.forEachChild(PriorityIndex_1.PRIORITY_INDEX, function (key, childNode) ***REMOVED***
            if (!self.matches(new Node_1.NamedNode(key, childNode))) ***REMOVED***
                filtered = filtered.updateImmediateChild(key, ChildrenNode_1.ChildrenNode.EMPTY_NODE);
            ***REMOVED***
        ***REMOVED***);
        return this.indexedFilter_.updateFullNode(oldSnap, filtered, optChangeAccumulator);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    RangedFilter.prototype.updatePriority = function (oldSnap, newPriority) ***REMOVED***
        // Don't support priorities on queries
        return oldSnap;
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    RangedFilter.prototype.filtersNodes = function () ***REMOVED***
        return true;
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    RangedFilter.prototype.getIndexedFilter = function () ***REMOVED***
        return this.indexedFilter_;
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    RangedFilter.prototype.getIndex = function () ***REMOVED***
        return this.index_;
    ***REMOVED***;
    /**
     * @param ***REMOVED***!QueryParams***REMOVED*** params
     * @return ***REMOVED***!NamedNode***REMOVED***
     * @private
     */
    RangedFilter.getStartPost_ = function (params) ***REMOVED***
        if (params.hasStart()) ***REMOVED***
            var startName = params.getIndexStartName();
            return params.getIndex().makePost(params.getIndexStartValue(), startName);
        ***REMOVED***
        else ***REMOVED***
            return params.getIndex().minPost();
        ***REMOVED***
    ***REMOVED***;
    /**
     * @param ***REMOVED***!QueryParams***REMOVED*** params
     * @return ***REMOVED***!NamedNode***REMOVED***
     * @private
     */
    RangedFilter.getEndPost_ = function (params) ***REMOVED***
        if (params.hasEnd()) ***REMOVED***
            var endName = params.getIndexEndName();
            return params.getIndex().makePost(params.getIndexEndValue(), endName);
        ***REMOVED***
        else ***REMOVED***
            return params.getIndex().maxPost();
        ***REMOVED***
    ***REMOVED***;
    return RangedFilter;
***REMOVED***());
exports.RangedFilter = RangedFilter;
