// protobufjs/src/ProtoBuf/Builder.js
/**
 * @alias ProtoBuf.Builder
 * @expose
 */
ProtoBuf.Builder = (function(ProtoBuf, Lang, Reflect) ***REMOVED***
    "use strict";

    /**
     * Constructs a new Builder.
     * @exports ProtoBuf.Builder
     * @class Provides the functionality to build protocol messages.
     * @param ***REMOVED***Object.<string,*>=***REMOVED*** options Options
     * @constructor
     */
    var Builder = function(options) ***REMOVED***

        /**
         * Namespace.
         * @type ***REMOVED***ProtoBuf.Reflect.Namespace***REMOVED***
         * @expose
         */
        this.ns = new Reflect.Namespace(this, null, ""); // Global namespace

        /**
         * Namespace pointer.
         * @type ***REMOVED***ProtoBuf.Reflect.T***REMOVED***
         * @expose
         */
        this.ptr = this.ns;

        /**
         * Resolved flag.
         * @type ***REMOVED***boolean***REMOVED***
         * @expose
         */
        this.resolved = false;

        /**
         * The current building result.
         * @type ***REMOVED***Object.<string,ProtoBuf.Builder.Message|Object>|null***REMOVED***
         * @expose
         */
        this.result = null;

        /**
         * Imported files.
         * @type ***REMOVED***Array.<string>***REMOVED***
         * @expose
         */
        this.files = ***REMOVED******REMOVED***;

        /**
         * Import root override.
         * @type ***REMOVED***?string***REMOVED***
         * @expose
         */
        this.importRoot = null;

        /**
         * Options.
         * @type ***REMOVED***!Object.<string, *>***REMOVED***
         * @expose
         */
        this.options = options || ***REMOVED******REMOVED***;
    ***REMOVED***;

    /**
     * @alias ProtoBuf.Builder.prototype
     * @inner
     */
    var BuilderPrototype = Builder.prototype;

    // ----- Definition tests -----

    /**
     * Tests if a definition most likely describes a message.
     * @param ***REMOVED***!Object***REMOVED*** def
     * @returns ***REMOVED***boolean***REMOVED***
     * @expose
     */
    Builder.isMessage = function(def) ***REMOVED***
        // Messages require a string name
        if (typeof def["name"] !== 'string')
            return false;
        // Messages do not contain values (enum) or rpc methods (service)
        if (typeof def["values"] !== 'undefined' || typeof def["rpc"] !== 'undefined')
            return false;
        return true;
    ***REMOVED***;

    /**
     * Tests if a definition most likely describes a message field.
     * @param ***REMOVED***!Object***REMOVED*** def
     * @returns ***REMOVED***boolean***REMOVED***
     * @expose
     */
    Builder.isMessageField = function(def) ***REMOVED***
        // Message fields require a string rule, name and type and an id
        if (typeof def["rule"] !== 'string' || typeof def["name"] !== 'string' || typeof def["type"] !== 'string' || typeof def["id"] === 'undefined')
            return false;
        return true;
    ***REMOVED***;

    /**
     * Tests if a definition most likely describes an enum.
     * @param ***REMOVED***!Object***REMOVED*** def
     * @returns ***REMOVED***boolean***REMOVED***
     * @expose
     */
    Builder.isEnum = function(def) ***REMOVED***
        // Enums require a string name
        if (typeof def["name"] !== 'string')
            return false;
        // Enums require at least one value
        if (typeof def["values"] === 'undefined' || !Array.isArray(def["values"]) || def["values"].length === 0)
            return false;
        return true;
    ***REMOVED***;

    /**
     * Tests if a definition most likely describes a service.
     * @param ***REMOVED***!Object***REMOVED*** def
     * @returns ***REMOVED***boolean***REMOVED***
     * @expose
     */
    Builder.isService = function(def) ***REMOVED***
        // Services require a string name and an rpc object
        if (typeof def["name"] !== 'string' || typeof def["rpc"] !== 'object' || !def["rpc"])
            return false;
        return true;
    ***REMOVED***;

    /**
     * Tests if a definition most likely describes an extended message
     * @param ***REMOVED***!Object***REMOVED*** def
     * @returns ***REMOVED***boolean***REMOVED***
     * @expose
     */
    Builder.isExtend = function(def) ***REMOVED***
        // Extends rquire a string ref
        if (typeof def["ref"] !== 'string')
            return false;
        return true;
    ***REMOVED***;

    // ----- Building -----

    /**
     * Resets the pointer to the root namespace.
     * @returns ***REMOVED***!ProtoBuf.Builder***REMOVED*** this
     * @expose
     */
    BuilderPrototype.reset = function() ***REMOVED***
        this.ptr = this.ns;
        return this;
    ***REMOVED***;

    /**
     * Defines a namespace on top of the current pointer position and places the pointer on it.
     * @param ***REMOVED***string***REMOVED*** namespace
     * @return ***REMOVED***!ProtoBuf.Builder***REMOVED*** this
     * @expose
     */
    BuilderPrototype.define = function(namespace) ***REMOVED***
        if (typeof namespace !== 'string' || !Lang.TYPEREF.test(namespace))
            throw Error("illegal namespace: "+namespace);
        namespace.split(".").forEach(function(part) ***REMOVED***
            var ns = this.ptr.getChild(part);
            if (ns === null) // Keep existing
                this.ptr.addChild(ns = new Reflect.Namespace(this, this.ptr, part));
            this.ptr = ns;
        ***REMOVED***, this);
        return this;
    ***REMOVED***;

    /**
     * Creates the specified definitions at the current pointer position.
     * @param ***REMOVED***!Array.<!Object>***REMOVED*** defs Messages, enums or services to create
     * @returns ***REMOVED***!ProtoBuf.Builder***REMOVED*** this
     * @throws ***REMOVED***Error***REMOVED*** If a message definition is invalid
     * @expose
     */
    BuilderPrototype.create = function(defs) ***REMOVED***
        if (!defs)
            return this; // Nothing to create
        if (!Array.isArray(defs))
            defs = [defs];
        else ***REMOVED***
            if (defs.length === 0)
                return this;
            defs = defs.slice();
        ***REMOVED***

        // It's quite hard to keep track of scopes and memory here, so let's do this iteratively.
        var stack = [defs];
        while (stack.length > 0) ***REMOVED***
            defs = stack.pop();

            if (!Array.isArray(defs)) // Stack always contains entire namespaces
                throw Error("not a valid namespace: "+JSON.stringify(defs));

            while (defs.length > 0) ***REMOVED***
                var def = defs.shift(); // Namespaces always contain an array of messages, enums and services

                if (Builder.isMessage(def)) ***REMOVED***
                    var obj = new Reflect.Message(this, this.ptr, def["name"], def["options"], def["isGroup"], def["syntax"]);

                    // Create OneOfs
                    var oneofs = ***REMOVED******REMOVED***;
                    if (def["oneofs"])
                        Object.keys(def["oneofs"]).forEach(function(name) ***REMOVED***
                            obj.addChild(oneofs[name] = new Reflect.Message.OneOf(this, obj, name));
                        ***REMOVED***, this);

                    // Create fields
                    if (def["fields"])
                        def["fields"].forEach(function(fld) ***REMOVED***
                            if (obj.getChild(fld["id"]|0) !== null)
                                throw Error("duplicate or invalid field id in "+obj.name+": "+fld['id']);
                            if (fld["options"] && typeof fld["options"] !== 'object')
                                throw Error("illegal field options in "+obj.name+"#"+fld["name"]);
                            var oneof = null;
                            if (typeof fld["oneof"] === 'string' && !(oneof = oneofs[fld["oneof"]]))
                                throw Error("illegal oneof in "+obj.name+"#"+fld["name"]+": "+fld["oneof"]);
                            fld = new Reflect.Message.Field(this, obj, fld["rule"], fld["keytype"], fld["type"], fld["name"], fld["id"], fld["options"], oneof, def["syntax"]);
                            if (oneof)
                                oneof.fields.push(fld);
                            obj.addChild(fld);
                        ***REMOVED***, this);

                    // Push children to stack
                    var subObj = [];
                    if (def["enums"])
                        def["enums"].forEach(function(enm) ***REMOVED***
                            subObj.push(enm);
                        ***REMOVED***);
                    if (def["messages"])
                        def["messages"].forEach(function(msg) ***REMOVED***
                            subObj.push(msg);
                        ***REMOVED***);
                    if (def["services"])
                        def["services"].forEach(function(svc) ***REMOVED***
                            subObj.push(svc);
                        ***REMOVED***);

                    // Set extension ranges
                    if (def["extensions"]) ***REMOVED***
                        if (typeof def["extensions"][0] === 'number') // pre 5.0.1
                            obj.extensions = [ def["extensions"] ];
                        else
                            obj.extensions = def["extensions"];
                    ***REMOVED***

                    // Create on top of current namespace
                    this.ptr.addChild(obj);
                    if (subObj.length > 0) ***REMOVED***
                        stack.push(defs); // Push the current level back
                        defs = subObj; // Continue processing sub level
                        subObj = null;
                        this.ptr = obj; // And move the pointer to this namespace
                        obj = null;
                        continue;
                    ***REMOVED***
                    subObj = null;

                ***REMOVED*** else if (Builder.isEnum(def)) ***REMOVED***

                    obj = new Reflect.Enum(this, this.ptr, def["name"], def["options"], def["syntax"]);
                    def["values"].forEach(function(val) ***REMOVED***
                        obj.addChild(new Reflect.Enum.Value(this, obj, val["name"], val["id"]));
                    ***REMOVED***, this);
                    this.ptr.addChild(obj);

                ***REMOVED*** else if (Builder.isService(def)) ***REMOVED***

                    obj = new Reflect.Service(this, this.ptr, def["name"], def["options"]);
                    Object.keys(def["rpc"]).forEach(function(name) ***REMOVED***
                        var mtd = def["rpc"][name];
                        obj.addChild(new Reflect.Service.RPCMethod(this, obj, name, mtd["request"], mtd["response"], !!mtd["request_stream"], !!mtd["response_stream"], mtd["options"]));
                    ***REMOVED***, this);
                    this.ptr.addChild(obj);

                ***REMOVED*** else if (Builder.isExtend(def)) ***REMOVED***

                    obj = this.ptr.resolve(def["ref"], true);
                    if (obj) ***REMOVED***
                        def["fields"].forEach(function(fld) ***REMOVED***
                            if (obj.getChild(fld['id']|0) !== null)
                                throw Error("duplicate extended field id in "+obj.name+": "+fld['id']);
                            // Check if field id is allowed to be extended
                            if (obj.extensions) ***REMOVED***
                                var valid = false;
                                obj.extensions.forEach(function(range) ***REMOVED***
                                    if (fld["id"] >= range[0] && fld["id"] <= range[1])
                                        valid = true;
                                ***REMOVED***);
                                if (!valid)
                                    throw Error("illegal extended field id in "+obj.name+": "+fld['id']+" (not within valid ranges)");
                            ***REMOVED***
                            // Convert extension field names to camel case notation if the override is set
                            var name = fld["name"];
                            if (this.options['convertFieldsToCamelCase'])
                                name = ProtoBuf.Util.toCamelCase(name);
                            // see #161: Extensions use their fully qualified name as their runtime key and...
                            var field = new Reflect.Message.ExtensionField(this, obj, fld["rule"], fld["type"], this.ptr.fqn()+'.'+name, fld["id"], fld["options"]);
                            // ...are added on top of the current namespace as an extension which is used for
                            // resolving their type later on (the extension always keeps the original name to
                            // prevent naming collisions)
                            var ext = new Reflect.Extension(this, this.ptr, fld["name"], field);
                            field.extension = ext;
                            this.ptr.addChild(ext);
                            obj.addChild(field);
                        ***REMOVED***, this);

                    ***REMOVED*** else if (!/\.?google\.protobuf\./.test(def["ref"])) // Silently skip internal extensions
                        throw Error("extended message "+def["ref"]+" is not defined");

                ***REMOVED*** else
                    throw Error("not a valid definition: "+JSON.stringify(def));

                def = null;
                obj = null;
            ***REMOVED***
            // Break goes here
            defs = null;
            this.ptr = this.ptr.parent; // Namespace done, continue at parent
        ***REMOVED***
        this.resolved = false; // Require re-resolve
        this.result = null; // Require re-build
        return this;
    ***REMOVED***;

    /**
     * Propagates syntax to all children.
     * @param ***REMOVED***!Object***REMOVED*** parent
     * @inner
     */
    function propagateSyntax(parent) ***REMOVED***
        if (parent['messages']) ***REMOVED***
            parent['messages'].forEach(function(child) ***REMOVED***
                child["syntax"] = parent["syntax"];
                propagateSyntax(child);
            ***REMOVED***);
        ***REMOVED***
        if (parent['enums']) ***REMOVED***
            parent['enums'].forEach(function(child) ***REMOVED***
                child["syntax"] = parent["syntax"];
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***

    /**
     * Imports another definition into this builder.
     * @param ***REMOVED***Object.<string,*>***REMOVED*** json Parsed import
     * @param ***REMOVED***(string|***REMOVED***root: string, file: string***REMOVED***)=***REMOVED*** filename Imported file name
     * @returns ***REMOVED***!ProtoBuf.Builder***REMOVED*** this
     * @throws ***REMOVED***Error***REMOVED*** If the definition or file cannot be imported
     * @expose
     */
    BuilderPrototype["import"] = function(json, filename) ***REMOVED***
        var delim = '/';

        // Make sure to skip duplicate imports

        if (typeof filename === 'string') ***REMOVED***

            if (ProtoBuf.Util.IS_NODE)
                filename = require("path")['resolve'](filename);
            if (this.files[filename] === true)
                return this.reset();
            this.files[filename] = true;

        ***REMOVED*** else if (typeof filename === 'object') ***REMOVED*** // Object with root, file.

            var root = filename.root;
            if (ProtoBuf.Util.IS_NODE)
                root = require("path")['resolve'](root);
            if (root.indexOf("\\") >= 0 || filename.file.indexOf("\\") >= 0)
                delim = '\\';
            var fname;
            if (ProtoBuf.Util.IS_NODE)
                fname = require("path")['join'](root, filename.file);
            else
                fname = root + delim + filename.file;
            if (this.files[fname] === true)
                return this.reset();
            this.files[fname] = true;
        ***REMOVED***

        // Import imports

        if (json['imports'] && json['imports'].length > 0) ***REMOVED***
            var importRoot,
                resetRoot = false;

            if (typeof filename === 'object') ***REMOVED*** // If an import root is specified, override

                this.importRoot = filename["root"]; resetRoot = true; // ... and reset afterwards
                importRoot = this.importRoot;
                filename = filename["file"];
                if (importRoot.indexOf("\\") >= 0 || filename.indexOf("\\") >= 0)
                    delim = '\\';

            ***REMOVED*** else if (typeof filename === 'string') ***REMOVED***

                if (this.importRoot) // If import root is overridden, use it
                    importRoot = this.importRoot;
                else ***REMOVED*** // Otherwise compute from filename
                    if (filename.indexOf("/") >= 0) ***REMOVED*** // Unix
                        importRoot = filename.replace(/\/[^\/]*$/, "");
                        if (/* /file.proto */ importRoot === "")
                            importRoot = "/";
                    ***REMOVED*** else if (filename.indexOf("\\") >= 0) ***REMOVED*** // Windows
                        importRoot = filename.replace(/\\[^\\]*$/, "");
                        delim = '\\';
                    ***REMOVED*** else
                        importRoot = ".";
                ***REMOVED***

            ***REMOVED*** else
                importRoot = null;

            for (var i=0; i<json['imports'].length; i++) ***REMOVED***
                if (typeof json['imports'][i] === 'string') ***REMOVED*** // Import file
                    if (!importRoot)
                        throw Error("cannot determine import root");
                    var importFilename = json['imports'][i];
                    if (importFilename === "google/protobuf/descriptor.proto")
                        continue; // Not needed and therefore not used
                    if (ProtoBuf.Util.IS_NODE)
                        importFilename = require("path")['join'](importRoot, importFilename);
                    else
                        importFilename = importRoot + delim + importFilename;
                    if (this.files[importFilename] === true)
                        continue; // Already imported
                    if (/\.proto$/i.test(importFilename) && !ProtoBuf.DotProto)       // If this is a light build
                        importFilename = importFilename.replace(/\.proto$/, ".json"); // always load the JSON file
                    var contents = ProtoBuf.Util.fetch(importFilename);
                    if (contents === null)
                        throw Error("failed to import '"+importFilename+"' in '"+filename+"': file not found");
                    if (/\.json$/i.test(importFilename)) // Always possible
                        this["import"](JSON.parse(contents+""), importFilename); // May throw
                    else
                        this["import"](ProtoBuf.DotProto.Parser.parse(contents), importFilename); // May throw
                ***REMOVED*** else // Import structure
                    if (!filename)
                        this["import"](json['imports'][i]);
                    else if (/\.(\w+)$/.test(filename)) // With extension: Append _importN to the name portion to make it unique
                        this["import"](json['imports'][i], filename.replace(/^(.+)\.(\w+)$/, function($0, $1, $2) ***REMOVED*** return $1+"_import"+i+"."+$2; ***REMOVED***));
                    else // Without extension: Append _importN to make it unique
                        this["import"](json['imports'][i], filename+"_import"+i);
            ***REMOVED***
            if (resetRoot) // Reset import root override when all imports are done
                this.importRoot = null;
        ***REMOVED***

        // Import structures

        if (json['package'])
            this.define(json['package']);
        if (json['syntax'])
            propagateSyntax(json);
        var base = this.ptr;
        if (json['options'])
            Object.keys(json['options']).forEach(function(key) ***REMOVED***
                base.options[key] = json['options'][key];
            ***REMOVED***);
        if (json['messages'])
            this.create(json['messages']),
            this.ptr = base;
        if (json['enums'])
            this.create(json['enums']),
            this.ptr = base;
        if (json['services'])
            this.create(json['services']),
            this.ptr = base;
        if (json['extends'])
            this.create(json['extends']);

        return this.reset();
    ***REMOVED***;

    /**
     * Resolves all namespace objects.
     * @throws ***REMOVED***Error***REMOVED*** If a type cannot be resolved
     * @returns ***REMOVED***!ProtoBuf.Builder***REMOVED*** this
     * @expose
     */
    BuilderPrototype.resolveAll = function() ***REMOVED***
        // Resolve all reflected objects
        var res;
        if (this.ptr == null || typeof this.ptr.type === 'object')
            return this; // Done (already resolved)

        if (this.ptr instanceof Reflect.Namespace) ***REMOVED*** // Resolve children

            this.ptr.children.forEach(function(child) ***REMOVED***
                this.ptr = child;
                this.resolveAll();
            ***REMOVED***, this);

        ***REMOVED*** else if (this.ptr instanceof Reflect.Message.Field) ***REMOVED*** // Resolve type

            if (!Lang.TYPE.test(this.ptr.type)) ***REMOVED***
                if (!Lang.TYPEREF.test(this.ptr.type))
                    throw Error("illegal type reference in "+this.ptr.toString(true)+": "+this.ptr.type);
                res = (this.ptr instanceof Reflect.Message.ExtensionField ? this.ptr.extension.parent : this.ptr.parent).resolve(this.ptr.type, true);
                if (!res)
                    throw Error("unresolvable type reference in "+this.ptr.toString(true)+": "+this.ptr.type);
                this.ptr.resolvedType = res;
                if (res instanceof Reflect.Enum) ***REMOVED***
                    this.ptr.type = ProtoBuf.TYPES["enum"];
                    if (this.ptr.syntax === 'proto3' && res.syntax !== 'proto3')
                        throw Error("proto3 message cannot reference proto2 enum");
                ***REMOVED***
                else if (res instanceof Reflect.Message)
                    this.ptr.type = res.isGroup ? ProtoBuf.TYPES["group"] : ProtoBuf.TYPES["message"];
                else
                    throw Error("illegal type reference in "+this.ptr.toString(true)+": "+this.ptr.type);
            ***REMOVED*** else
                this.ptr.type = ProtoBuf.TYPES[this.ptr.type];

            // If it's a map field, also resolve the key type. The key type can be only a numeric, string, or bool type
            // (i.e., no enums or messages), so we don't need to resolve against the current namespace.
            if (this.ptr.map) ***REMOVED***
                if (!Lang.TYPE.test(this.ptr.keyType))
                    throw Error("illegal key type for map field in "+this.ptr.toString(true)+": "+this.ptr.keyType);
                this.ptr.keyType = ProtoBuf.TYPES[this.ptr.keyType];
            ***REMOVED***

            // If it's a repeated and packable field then proto3 mandates it should be packed by
            // default
            if (
              this.ptr.syntax === 'proto3' &&
              this.ptr.repeated && this.ptr.options.packed === undefined &&
              ProtoBuf.PACKABLE_WIRE_TYPES.indexOf(this.ptr.type.wireType) !== -1
            ) ***REMOVED***
              this.ptr.options.packed = true;
            ***REMOVED***

        ***REMOVED*** else if (this.ptr instanceof ProtoBuf.Reflect.Service.Method) ***REMOVED***

            if (this.ptr instanceof ProtoBuf.Reflect.Service.RPCMethod) ***REMOVED***
                res = this.ptr.parent.resolve(this.ptr.requestName, true);
                if (!res || !(res instanceof ProtoBuf.Reflect.Message))
                    throw Error("Illegal type reference in "+this.ptr.toString(true)+": "+this.ptr.requestName);
                this.ptr.resolvedRequestType = res;
                res = this.ptr.parent.resolve(this.ptr.responseName, true);
                if (!res || !(res instanceof ProtoBuf.Reflect.Message))
                    throw Error("Illegal type reference in "+this.ptr.toString(true)+": "+this.ptr.responseName);
                this.ptr.resolvedResponseType = res;
            ***REMOVED*** else // Should not happen as nothing else is implemented
                throw Error("illegal service type in "+this.ptr.toString(true));

        ***REMOVED*** else if (
            !(this.ptr instanceof ProtoBuf.Reflect.Message.OneOf) && // Not built
            !(this.ptr instanceof ProtoBuf.Reflect.Extension) && // Not built
            !(this.ptr instanceof ProtoBuf.Reflect.Enum.Value) // Built in enum
        )
            throw Error("illegal object in namespace: "+typeof(this.ptr)+": "+this.ptr);

        return this.reset();
    ***REMOVED***;

    /**
     * Builds the protocol. This will first try to resolve all definitions and, if this has been successful,
     * return the built package.
     * @param ***REMOVED***(string|Array.<string>)=***REMOVED*** path Specifies what to return. If omitted, the entire namespace will be returned.
     * @returns ***REMOVED***!ProtoBuf.Builder.Message|!Object.<string,*>***REMOVED***
     * @throws ***REMOVED***Error***REMOVED*** If a type could not be resolved
     * @expose
     */
    BuilderPrototype.build = function(path) ***REMOVED***
        this.reset();
        if (!this.resolved)
            this.resolveAll(),
            this.resolved = true,
            this.result = null; // Require re-build
        if (this.result === null) // (Re-)Build
            this.result = this.ns.build();
        if (!path)
            return this.result;
        var part = typeof path === 'string' ? path.split(".") : path,
            ptr = this.result; // Build namespace pointer (no hasChild etc.)
        for (var i=0; i<part.length; i++)
            if (ptr[part[i]])
                ptr = ptr[part[i]];
            else ***REMOVED***
                ptr = null;
                break;
            ***REMOVED***
        return ptr;
    ***REMOVED***;

    /**
     * Similar to ***REMOVED***@link ProtoBuf.Builder#build***REMOVED***, but looks up the internal reflection descriptor.
     * @param ***REMOVED***string=***REMOVED*** path Specifies what to return. If omitted, the entire namespace wiil be returned.
     * @param ***REMOVED***boolean=***REMOVED*** excludeNonNamespace Excludes non-namespace types like fields, defaults to `false`
     * @returns ***REMOVED***?ProtoBuf.Reflect.T***REMOVED*** Reflection descriptor or `null` if not found
     */
    BuilderPrototype.lookup = function(path, excludeNonNamespace) ***REMOVED***
        return path ? this.ns.resolve(path, excludeNonNamespace) : this.ns;
    ***REMOVED***;

    /**
     * Returns a string representation of this object.
     * @return ***REMOVED***string***REMOVED*** String representation as of "Builder"
     * @expose
     */
    BuilderPrototype.toString = function() ***REMOVED***
        return "Builder";
    ***REMOVED***;

    // ----- Base classes -----
    // Exist for the sole purpose of being able to "... instanceof ProtoBuf.Builder.Message" etc.

    /**
     * @alias ProtoBuf.Builder.Message
     */
    Builder.Message = function() ***REMOVED******REMOVED***;

    /**
     * @alias ProtoBuf.Builder.Enum
     */
    Builder.Enum = function() ***REMOVED******REMOVED***;

    /**
     * @alias ProtoBuf.Builder.Message
     */
    Builder.Service = function() ***REMOVED******REMOVED***;

    return Builder;

***REMOVED***)(ProtoBuf, ProtoBuf.Lang, ProtoBuf.Reflect);
