// @protobufjs/eventemitter/index.js
"use strict";
module.exports = EventEmitter;

/**
 * Constructs a new event emitter instance.
 * @classdesc A minimal event emitter.
 * @memberof util
 * @constructor
 */
function EventEmitter() ***REMOVED***

    /**
     * Registered listeners.
     * @type ***REMOVED***Object.<string,*>***REMOVED***
     * @private
     */
    this._listeners = ***REMOVED******REMOVED***;
***REMOVED***

/**
 * Registers an event listener.
 * @param ***REMOVED***string***REMOVED*** evt Event name
 * @param ***REMOVED***function***REMOVED*** fn Listener
 * @param ***REMOVED*******REMOVED*** [ctx] Listener context
 * @returns ***REMOVED***util.EventEmitter***REMOVED*** `this`
 */
EventEmitter.prototype.on = function on(evt, fn, ctx) ***REMOVED***
    (this._listeners[evt] || (this._listeners[evt] = [])).push(***REMOVED***
        fn  : fn,
        ctx : ctx || this
    ***REMOVED***);
    return this;
***REMOVED***;

/**
 * Removes an event listener or any matching listeners if arguments are omitted.
 * @param ***REMOVED***string***REMOVED*** [evt] Event name. Removes all listeners if omitted.
 * @param ***REMOVED***function***REMOVED*** [fn] Listener to remove. Removes all listeners of `evt` if omitted.
 * @returns ***REMOVED***util.EventEmitter***REMOVED*** `this`
 */
EventEmitter.prototype.off = function off(evt, fn) ***REMOVED***
    if (evt === undefined)
        this._listeners = ***REMOVED******REMOVED***;
    else ***REMOVED***
        if (fn === undefined)
            this._listeners[evt] = [];
        else ***REMOVED***
            var listeners = this._listeners[evt];
            for (var i = 0; i < listeners.length;)
                if (listeners[i].fn === fn)
                    listeners.splice(i, 1);
                else
                    ++i;
        ***REMOVED***
    ***REMOVED***
    return this;
***REMOVED***;

/**
 * Emits an event by calling its listeners with the specified arguments.
 * @param ***REMOVED***string***REMOVED*** evt Event name
 * @param ***REMOVED***...****REMOVED*** args Arguments
 * @returns ***REMOVED***util.EventEmitter***REMOVED*** `this`
 */
EventEmitter.prototype.emit = function emit(evt) ***REMOVED***
    var listeners = this._listeners[evt];
    if (listeners) ***REMOVED***
        var args = [],
            i = 1;
        for (; i < arguments.length;)
            args.push(arguments[i++]);
        for (i = 0; i < listeners.length;)
            listeners[i].fn.apply(listeners[i++].ctx, args);
    ***REMOVED***
    return this;
***REMOVED***;
