// gauge/template-item.js
'use strict'
var stringWidth = require('string-width')

module.exports = TemplateItem

function isPercent (num) ***REMOVED***
  if (typeof num !== 'string') return false
  return num.slice(-1) === '%'
***REMOVED***

function percent (num) ***REMOVED***
  return Number(num.slice(0, -1)) / 100
***REMOVED***

function TemplateItem (values, outputLength) ***REMOVED***
  this.overallOutputLength = outputLength
  this.finished = false
  this.type = null
  this.value = null
  this.length = null
  this.maxLength = null
  this.minLength = null
  this.kerning = null
  this.align = 'left'
  this.padLeft = 0
  this.padRight = 0
  this.index = null
  this.first = null
  this.last = null
  if (typeof values === 'string') ***REMOVED***
    this.value = values
  ***REMOVED*** else ***REMOVED***
    for (var prop in values) this[prop] = values[prop]
  ***REMOVED***
  // Realize percents
  if (isPercent(this.length)) ***REMOVED***
    this.length = Math.round(this.overallOutputLength * percent(this.length))
  ***REMOVED***
  if (isPercent(this.minLength)) ***REMOVED***
    this.minLength = Math.round(this.overallOutputLength * percent(this.minLength))
  ***REMOVED***
  if (isPercent(this.maxLength)) ***REMOVED***
    this.maxLength = Math.round(this.overallOutputLength * percent(this.maxLength))
  ***REMOVED***
  return this
***REMOVED***

TemplateItem.prototype = ***REMOVED******REMOVED***

TemplateItem.prototype.getBaseLength = function () ***REMOVED***
  var length = this.length
  if (length == null && typeof this.value === 'string' && this.maxLength == null && this.minLength == null) ***REMOVED***
    length = stringWidth(this.value)
  ***REMOVED***
  return length
***REMOVED***

TemplateItem.prototype.getLength = function () ***REMOVED***
  var length = this.getBaseLength()
  if (length == null) return null
  return length + this.padLeft + this.padRight
***REMOVED***

TemplateItem.prototype.getMaxLength = function () ***REMOVED***
  if (this.maxLength == null) return null
  return this.maxLength + this.padLeft + this.padRight
***REMOVED***

TemplateItem.prototype.getMinLength = function () ***REMOVED***
  if (this.minLength == null) return null
  return this.minLength + this.padLeft + this.padRight
***REMOVED***
