// @firebase/database/dist/cjs/src/core/operation/Merge.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Operation_1 = require("./Operation");
var Overwrite_1 = require("./Overwrite");
var Path_1 = require("../util/Path");
var util_1 = require("@firebase/util");
/**
 * @param ***REMOVED***!OperationSource***REMOVED*** source
 * @param ***REMOVED***!Path***REMOVED*** path
 * @param ***REMOVED***!ImmutableTree.<!Node>***REMOVED*** children
 * @constructor
 * @implements ***REMOVED***Operation***REMOVED***
 */
var Merge = /** @class */ (function () ***REMOVED***
    function Merge(
        /**@inheritDoc */ source, 
        /**@inheritDoc */ path, 
        /**@inheritDoc */ children) ***REMOVED***
        this.source = source;
        this.path = path;
        this.children = children;
        /** @inheritDoc */
        this.type = Operation_1.OperationType.MERGE;
    ***REMOVED***
    /**
     * @inheritDoc
     */
    Merge.prototype.operationForChild = function (childName) ***REMOVED***
        if (this.path.isEmpty()) ***REMOVED***
            var childTree = this.children.subtree(new Path_1.Path(childName));
            if (childTree.isEmpty()) ***REMOVED***
                // This child is unaffected
                return null;
            ***REMOVED***
            else if (childTree.value) ***REMOVED***
                // We have a snapshot for the child in question.  This becomes an overwrite of the child.
                return new Overwrite_1.Overwrite(this.source, Path_1.Path.Empty, childTree.value);
            ***REMOVED***
            else ***REMOVED***
                // This is a merge at a deeper level
                return new Merge(this.source, Path_1.Path.Empty, childTree);
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            util_1.assert(this.path.getFront() === childName, "Can't get a merge for a child not on the path of the operation");
            return new Merge(this.source, this.path.popFront(), this.children);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    Merge.prototype.toString = function () ***REMOVED***
        return ('Operation(' +
            this.path +
            ': ' +
            this.source.toString() +
            ' merge: ' +
            this.children.toString() +
            ')');
    ***REMOVED***;
    return Merge;
***REMOVED***());
exports.Merge = Merge;
