// crypto-browserify/example/bundle.js
var require = function (file, cwd) ***REMOVED***
    var resolved = require.resolve(file, cwd || '/');
    var mod = require.modules[resolved];
    if (!mod) throw new Error(
        'Failed to resolve module ' + file + ', tried ' + resolved
    );
    var res = mod._cached ? mod._cached : mod();
    return res;
***REMOVED***

require.paths = [];
require.modules = ***REMOVED******REMOVED***;
require.extensions = [".js",".coffee"];

require._core = ***REMOVED***
    'assert': true,
    'events': true,
    'fs': true,
    'path': true,
    'vm': true
***REMOVED***;

require.resolve = (function () ***REMOVED***
    return function (x, cwd) ***REMOVED***
        if (!cwd) cwd = '/';
        
        if (require._core[x]) return x;
        var path = require.modules.path();
        cwd = path.resolve('/', cwd);
        var y = cwd || '/';
        
        if (x.match(/^(?:\.\.?\/|\/)/)) ***REMOVED***
            var m = loadAsFileSync(path.resolve(y, x))
                || loadAsDirectorySync(path.resolve(y, x));
            if (m) return m;
        ***REMOVED***
        
        var n = loadNodeModulesSync(x, y);
        if (n) return n;
        
        throw new Error("Cannot find module '" + x + "'");
        
        function loadAsFileSync (x) ***REMOVED***
            if (require.modules[x]) ***REMOVED***
                return x;
            ***REMOVED***
            
            for (var i = 0; i < require.extensions.length; i++) ***REMOVED***
                var ext = require.extensions[i];
                if (require.modules[x + ext]) return x + ext;
            ***REMOVED***
        ***REMOVED***
        
        function loadAsDirectorySync (x) ***REMOVED***
            x = x.replace(/\/+$/, '');
            var pkgfile = x + '/package.json';
            if (require.modules[pkgfile]) ***REMOVED***
                var pkg = require.modules[pkgfile]();
                var b = pkg.browserify;
                if (typeof b === 'object' && b.main) ***REMOVED***
                    var m = loadAsFileSync(path.resolve(x, b.main));
                    if (m) return m;
                ***REMOVED***
                else if (typeof b === 'string') ***REMOVED***
                    var m = loadAsFileSync(path.resolve(x, b));
                    if (m) return m;
                ***REMOVED***
                else if (pkg.main) ***REMOVED***
                    var m = loadAsFileSync(path.resolve(x, pkg.main));
                    if (m) return m;
                ***REMOVED***
            ***REMOVED***
            
            return loadAsFileSync(x + '/index');
        ***REMOVED***
        
        function loadNodeModulesSync (x, start) ***REMOVED***
            var dirs = nodeModulesPathsSync(start);
            for (var i = 0; i < dirs.length; i++) ***REMOVED***
                var dir = dirs[i];
                var m = loadAsFileSync(dir + '/' + x);
                if (m) return m;
                var n = loadAsDirectorySync(dir + '/' + x);
                if (n) return n;
            ***REMOVED***
            
            var m = loadAsFileSync(x);
            if (m) return m;
        ***REMOVED***
        
        function nodeModulesPathsSync (start) ***REMOVED***
            var parts;
            if (start === '/') parts = [ '' ];
            else parts = path.normalize(start).split('/');
            
            var dirs = [];
            for (var i = parts.length - 1; i >= 0; i--) ***REMOVED***
                if (parts[i] === 'node_modules') continue;
                var dir = parts.slice(0, i + 1).join('/') + '/node_modules';
                dirs.push(dir);
            ***REMOVED***
            
            return dirs;
        ***REMOVED***
    ***REMOVED***;
***REMOVED***)();

require.alias = function (from, to) ***REMOVED***
    var path = require.modules.path();
    var res = null;
    try ***REMOVED***
        res = require.resolve(from + '/package.json', '/');
    ***REMOVED***
    catch (err) ***REMOVED***
        res = require.resolve(from, '/');
    ***REMOVED***
    var basedir = path.dirname(res);
    
    var keys = (Object.keys || function (obj) ***REMOVED***
        var res = [];
        for (var key in obj) res.push(key)
        return res;
    ***REMOVED***)(require.modules);
    
    for (var i = 0; i < keys.length; i++) ***REMOVED***
        var key = keys[i];
        if (key.slice(0, basedir.length + 1) === basedir + '/') ***REMOVED***
            var f = key.slice(basedir.length);
            require.modules[to + f] = require.modules[basedir + f];
        ***REMOVED***
        else if (key === basedir) ***REMOVED***
            require.modules[to] = require.modules[basedir];
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

require.define = function (filename, fn) ***REMOVED***
    var dirname = require._core[filename]
        ? ''
        : require.modules.path().dirname(filename)
    ;
    
    var require_ = function (file) ***REMOVED***
        return require(file, dirname)
    ***REMOVED***;
    require_.resolve = function (name) ***REMOVED***
        return require.resolve(name, dirname);
    ***REMOVED***;
    require_.modules = require.modules;
    require_.define = require.define;
    var module_ = ***REMOVED*** exports : ***REMOVED******REMOVED*** ***REMOVED***;
    
    require.modules[filename] = function () ***REMOVED***
        require.modules[filename]._cached = module_.exports;
        fn.call(
            module_.exports,
            require_,
            module_,
            module_.exports,
            dirname,
            filename
        );
        require.modules[filename]._cached = module_.exports;
        return module_.exports;
    ***REMOVED***;
***REMOVED***;

if (typeof process === 'undefined') process = ***REMOVED******REMOVED***;

if (!process.nextTick) process.nextTick = (function () ***REMOVED***
    var queue = [];
    var canPost = typeof window !== 'undefined'
        && window.postMessage && window.addEventListener
    ;
    
    if (canPost) ***REMOVED***
        window.addEventListener('message', function (ev) ***REMOVED***
            if (ev.source === window && ev.data === 'browserify-tick') ***REMOVED***
                ev.stopPropagation();
                if (queue.length > 0) ***REMOVED***
                    var fn = queue.shift();
                    fn();
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***, true);
    ***REMOVED***
    
    return function (fn) ***REMOVED***
        if (canPost) ***REMOVED***
            queue.push(fn);
            window.postMessage('browserify-tick', '*');
        ***REMOVED***
        else setTimeout(fn, 0);
    ***REMOVED***;
***REMOVED***)();

if (!process.title) process.title = 'browser';

if (!process.binding) process.binding = function (name) ***REMOVED***
    if (name === 'evals') return require('vm')
    else throw new Error('No such module')
***REMOVED***;

if (!process.cwd) process.cwd = function () ***REMOVED*** return '.' ***REMOVED***;

if (!process.env) process.env = ***REMOVED******REMOVED***;
if (!process.argv) process.argv = [];

require.define("path", function (require, module, exports, __dirname, __filename) ***REMOVED***
function filter (xs, fn) ***REMOVED***
    var res = [];
    for (var i = 0; i < xs.length; i++) ***REMOVED***
        if (fn(xs[i], i, xs)) res.push(xs[i]);
    ***REMOVED***
    return res;
***REMOVED***

// resolves . and .. elements in a path array with directory names there
// must be no slashes, empty elements, or device names (c:\) in the array
// (so also no leading and trailing slashes - it does not distinguish
// relative and absolute paths)
function normalizeArray(parts, allowAboveRoot) ***REMOVED***
  // if the path tries to go above the root, `up` ends up > 0
  var up = 0;
  for (var i = parts.length; i >= 0; i--) ***REMOVED***
    var last = parts[i];
    if (last == '.') ***REMOVED***
      parts.splice(i, 1);
    ***REMOVED*** else if (last === '..') ***REMOVED***
      parts.splice(i, 1);
      up++;
    ***REMOVED*** else if (up) ***REMOVED***
      parts.splice(i, 1);
      up--;
    ***REMOVED***
  ***REMOVED***

  // if the path is allowed to go above the root, restore leading ..s
  if (allowAboveRoot) ***REMOVED***
    for (; up--; up) ***REMOVED***
      parts.unshift('..');
    ***REMOVED***
  ***REMOVED***

  return parts;
***REMOVED***

// Regex to split a filename into [*, dir, basename, ext]
// posix version
var splitPathRe = /^(.+\/(?!$)|\/)?((?:.+?)?(\.[^.]*)?)$/;

// path.resolve([from ...], to)
// posix version
exports.resolve = function() ***REMOVED***
var resolvedPath = '',
    resolvedAbsolute = false;

for (var i = arguments.length; i >= -1 && !resolvedAbsolute; i--) ***REMOVED***
  var path = (i >= 0)
      ? arguments[i]
      : process.cwd();

  // Skip empty and invalid entries
  if (typeof path !== 'string' || !path) ***REMOVED***
    continue;
  ***REMOVED***

  resolvedPath = path + '/' + resolvedPath;
  resolvedAbsolute = path.charAt(0) === '/';
***REMOVED***

// At this point the path should be resolved to a full absolute path, but
// handle relative paths to be safe (might happen when process.cwd() fails)

// Normalize the path
resolvedPath = normalizeArray(filter(resolvedPath.split('/'), function(p) ***REMOVED***
    return !!p;
  ***REMOVED***), !resolvedAbsolute).join('/');

  return ((resolvedAbsolute ? '/' : '') + resolvedPath) || '.';
***REMOVED***;

// path.normalize(path)
// posix version
exports.normalize = function(path) ***REMOVED***
var isAbsolute = path.charAt(0) === '/',
    trailingSlash = path.slice(-1) === '/';

// Normalize the path
path = normalizeArray(filter(path.split('/'), function(p) ***REMOVED***
    return !!p;
  ***REMOVED***), !isAbsolute).join('/');

  if (!path && !isAbsolute) ***REMOVED***
    path = '.';
  ***REMOVED***
  if (path && trailingSlash) ***REMOVED***
    path += '/';
  ***REMOVED***
  
  return (isAbsolute ? '/' : '') + path;
***REMOVED***;


// posix version
exports.join = function() ***REMOVED***
  var paths = Array.prototype.slice.call(arguments, 0);
  return exports.normalize(filter(paths, function(p, index) ***REMOVED***
    return p && typeof p === 'string';
  ***REMOVED***).join('/'));
***REMOVED***;


exports.dirname = function(path) ***REMOVED***
  var dir = splitPathRe.exec(path)[1] || '';
  var isWindows = false;
  if (!dir) ***REMOVED***
    // No dirname
    return '.';
  ***REMOVED*** else if (dir.length === 1 ||
      (isWindows && dir.length <= 3 && dir.charAt(1) === ':')) ***REMOVED***
    // It is just a slash or a drive letter with a slash
    return dir;
  ***REMOVED*** else ***REMOVED***
    // It is a full dirname, strip trailing slash
    return dir.substring(0, dir.length - 1);
  ***REMOVED***
***REMOVED***;


exports.basename = function(path, ext) ***REMOVED***
  var f = splitPathRe.exec(path)[2] || '';
  // TODO: make this comparison case-insensitive on windows?
  if (ext && f.substr(-1 * ext.length) === ext) ***REMOVED***
    f = f.substr(0, f.length - ext.length);
  ***REMOVED***
  return f;
***REMOVED***;


exports.extname = function(path) ***REMOVED***
  return splitPathRe.exec(path)[3] || '';
***REMOVED***;

***REMOVED***);

require.define("crypto", function (require, module, exports, __dirname, __filename) ***REMOVED***
module.exports = require("crypto-browserify")
***REMOVED***);

require.define("/node_modules/crypto-browserify/package.json", function (require, module, exports, __dirname, __filename) ***REMOVED***
module.exports = ***REMOVED******REMOVED***
***REMOVED***);

require.define("/node_modules/crypto-browserify/index.js", function (require, module, exports, __dirname, __filename) ***REMOVED***
var sha = require('./sha')

var algorithms = ***REMOVED***
  sha1: ***REMOVED***
    hex: sha.hex_sha1,
    binary: sha.b64_sha1,
    ascii: sha.str_sha1
  ***REMOVED***
***REMOVED***

function error () ***REMOVED***
  var m = [].slice.call(arguments).join(' ')
  throw new Error([
    m,
    'we accept pull requests',
    'http://github.com/dominictarr/crypto-browserify'
    ].join('\n'))
***REMOVED***

exports.createHash = function (alg) ***REMOVED***
  alg = alg || 'sha1'
  if(!algorithms[alg])
    error('algorithm:', alg, 'is not yet supported')
  var s = ''
  _alg = algorithms[alg]
  return ***REMOVED***
    update: function (data) ***REMOVED***
      s += data
      return this
    ***REMOVED***,
    digest: function (enc) ***REMOVED***
      enc = enc || 'binary'
      var fn 
      if(!(fn = _alg[enc]))
        error('encoding:', enc , 'is not yet supported for algorithm', alg)
      var r = fn(s)
      s = null //not meant to use the hash after you've called digest.
      return r
    ***REMOVED***
  ***REMOVED***
***REMOVED***
// the least I can do is make error messages for the rest of the node.js/crypto api.
;['createCredentials'
, 'createHmac'
, 'createCypher'
, 'createCypheriv'
, 'createDecipher'
, 'createDecipheriv'
, 'createSign'
, 'createVerify'
, 'createDeffieHellman',
, 'pbkdf2',
, 'randomBytes' ].forEach(function (name) ***REMOVED***
  exports[name] = function () ***REMOVED***
    error('sorry,', name, 'is not implemented yet')
  ***REMOVED***
***REMOVED***)

***REMOVED***);

require.define("/node_modules/crypto-browserify/sha.js", function (require, module, exports, __dirname, __filename) ***REMOVED***
/*
 * A JavaScript implementation of the Secure Hash Algorithm, SHA-1, as defined
 * in FIPS PUB 180-1
 * Version 2.1a Copyright Paul Johnston 2000 - 2002.
 * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
 * Distributed under the BSD License
 * See http://pajhome.org.uk/crypt/md5 for details.
 */

exports.hex_sha1 = hex_sha1;
exports.b64_sha1 = b64_sha1;
exports.str_sha1 = str_sha1;
exports.hex_hmac_sha1 = hex_hmac_sha1;
exports.b64_hmac_sha1 = b64_hmac_sha1;
exports.str_hmac_sha1 = str_hmac_sha1;

/*
 * Configurable variables. You may need to tweak these to be compatible with
 * the server-side, but the defaults work in most cases.
 */
var hexcase = 0;  /* hex output format. 0 - lowercase; 1 - uppercase        */
var b64pad  = ""; /* base-64 pad character. "=" for strict RFC compliance   */
var chrsz   = 8;  /* bits per input character. 8 - ASCII; 16 - Unicode      */

/*
 * These are the functions you'll usually want to call
 * They take string arguments and return either hex or base-64 encoded strings
 */
function hex_sha1(s)***REMOVED***return binb2hex(core_sha1(str2binb(s),s.length * chrsz));***REMOVED***
function b64_sha1(s)***REMOVED***return binb2b64(core_sha1(str2binb(s),s.length * chrsz));***REMOVED***
function str_sha1(s)***REMOVED***return binb2str(core_sha1(str2binb(s),s.length * chrsz));***REMOVED***
function hex_hmac_sha1(key, data)***REMOVED*** return binb2hex(core_hmac_sha1(key, data));***REMOVED***
function b64_hmac_sha1(key, data)***REMOVED*** return binb2b64(core_hmac_sha1(key, data));***REMOVED***
function str_hmac_sha1(key, data)***REMOVED*** return binb2str(core_hmac_sha1(key, data));***REMOVED***

/*
 * Perform a simple self-test to see if the VM is working
 */
function sha1_vm_test()
***REMOVED***
  return hex_sha1("abc") == "a9993e364706816aba3e25717850c26c9cd0d89d";
***REMOVED***

/*
 * Calculate the SHA-1 of an array of big-endian words, and a bit length
 */
function core_sha1(x, len)
***REMOVED***
  /* append padding */
  x[len >> 5] |= 0x80 << (24 - len % 32);
  x[((len + 64 >> 9) << 4) + 15] = len;

  var w = Array(80);
  var a =  1732584193;
  var b = -271733879;
  var c = -1732584194;
  var d =  271733878;
  var e = -1009589776;

  for(var i = 0; i < x.length; i += 16)
  ***REMOVED***
    var olda = a;
    var oldb = b;
    var oldc = c;
    var oldd = d;
    var olde = e;

    for(var j = 0; j < 80; j++)
    ***REMOVED***
      if(j < 16) w[j] = x[i + j];
      else w[j] = rol(w[j-3] ^ w[j-8] ^ w[j-14] ^ w[j-16], 1);
      var t = safe_add(safe_add(rol(a, 5), sha1_ft(j, b, c, d)),
                       safe_add(safe_add(e, w[j]), sha1_kt(j)));
      e = d;
      d = c;
      c = rol(b, 30);
      b = a;
      a = t;
    ***REMOVED***

    a = safe_add(a, olda);
    b = safe_add(b, oldb);
    c = safe_add(c, oldc);
    d = safe_add(d, oldd);
    e = safe_add(e, olde);
  ***REMOVED***
  return Array(a, b, c, d, e);

***REMOVED***

/*
 * Perform the appropriate triplet combination function for the current
 * iteration
 */
function sha1_ft(t, b, c, d)
***REMOVED***
  if(t < 20) return (b & c) | ((~b) & d);
  if(t < 40) return b ^ c ^ d;
  if(t < 60) return (b & c) | (b & d) | (c & d);
  return b ^ c ^ d;
***REMOVED***

/*
 * Determine the appropriate additive constant for the current iteration
 */
function sha1_kt(t)
***REMOVED***
  return (t < 20) ?  1518500249 : (t < 40) ?  1859775393 :
         (t < 60) ? -1894007588 : -899497514;
***REMOVED***

/*
 * Calculate the HMAC-SHA1 of a key and some data
 */
function core_hmac_sha1(key, data)
***REMOVED***
  var bkey = str2binb(key);
  if(bkey.length > 16) bkey = core_sha1(bkey, key.length * chrsz);

  var ipad = Array(16), opad = Array(16);
  for(var i = 0; i < 16; i++)
  ***REMOVED***
    ipad[i] = bkey[i] ^ 0x36363636;
    opad[i] = bkey[i] ^ 0x5C5C5C5C;
  ***REMOVED***

  var hash = core_sha1(ipad.concat(str2binb(data)), 512 + data.length * chrsz);
  return core_sha1(opad.concat(hash), 512 + 160);
***REMOVED***

/*
 * Add integers, wrapping at 2^32. This uses 16-bit operations internally
 * to work around bugs in some JS interpreters.
 */
function safe_add(x, y)
***REMOVED***
  var lsw = (x & 0xFFFF) + (y & 0xFFFF);
  var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
  return (msw << 16) | (lsw & 0xFFFF);
***REMOVED***

/*
 * Bitwise rotate a 32-bit number to the left.
 */
function rol(num, cnt)
***REMOVED***
  return (num << cnt) | (num >>> (32 - cnt));
***REMOVED***

/*
 * Convert an 8-bit or 16-bit string to an array of big-endian words
 * In 8-bit function, characters >255 have their hi-byte silently ignored.
 */
function str2binb(str)
***REMOVED***
  var bin = Array();
  var mask = (1 << chrsz) - 1;
  for(var i = 0; i < str.length * chrsz; i += chrsz)
    bin[i>>5] |= (str.charCodeAt(i / chrsz) & mask) << (32 - chrsz - i%32);
  return bin;
***REMOVED***

/*
 * Convert an array of big-endian words to a string
 */
function binb2str(bin)
***REMOVED***
  var str = "";
  var mask = (1 << chrsz) - 1;
  for(var i = 0; i < bin.length * 32; i += chrsz)
    str += String.fromCharCode((bin[i>>5] >>> (32 - chrsz - i%32)) & mask);
  return str;
***REMOVED***

/*
 * Convert an array of big-endian words to a hex string.
 */
function binb2hex(binarray)
***REMOVED***
  var hex_tab = hexcase ? "0123456789ABCDEF" : "0123456789abcdef";
  var str = "";
  for(var i = 0; i < binarray.length * 4; i++)
  ***REMOVED***
    str += hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8+4)) & 0xF) +
           hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8  )) & 0xF);
  ***REMOVED***
  return str;
***REMOVED***

/*
 * Convert an array of big-endian words to a base-64 string
 */
function binb2b64(binarray)
***REMOVED***
  var tab = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  var str = "";
  for(var i = 0; i < binarray.length * 4; i += 3)
  ***REMOVED***
    var triplet = (((binarray[i   >> 2] >> 8 * (3 -  i   %4)) & 0xFF) << 16)
                | (((binarray[i+1 >> 2] >> 8 * (3 - (i+1)%4)) & 0xFF) << 8 )
                |  ((binarray[i+2 >> 2] >> 8 * (3 - (i+2)%4)) & 0xFF);
    for(var j = 0; j < 4; j++)
    ***REMOVED***
      if(i * 8 + j * 6 > binarray.length * 32) str += b64pad;
      else str += tab.charAt((triplet >> 6*(3-j)) & 0x3F);
    ***REMOVED***
  ***REMOVED***
  return str;
***REMOVED***


***REMOVED***);

require.define("/test.js", function (require, module, exports, __dirname, __filename) ***REMOVED***
    var crypto = require('crypto')
var abc = crypto.createHash('sha1').update('abc').digest('hex')
console.log(abc)
//require('hello').inlineCall().call2()

***REMOVED***);
require("/test.js");
