// events/tests/common.js
var assert = require('assert');

var mustCallChecks = [];

function runCallChecks() ***REMOVED***
  var failed_count = 0;
  for (var i=0 ; i< mustCallChecks.length; ++i) ***REMOVED***
    var context = mustCallChecks[i];
    if (context.actual === context.expected) ***REMOVED***
      continue;
    ***REMOVED***

    failed_count++;
    console.log('Mismatched %s function calls. Expected %d, actual %d.',
                context.name,
                context.expected,
                context.actual);
    console.log(context.stack.split('\n').slice(2).join('\n'));
  ***REMOVED***

  assert(failed_count === 0);
***REMOVED***

after(runCallChecks);

exports.mustCall = function(fn, expected) ***REMOVED***
  if (typeof expected !== 'number') expected = 1;

  var context = ***REMOVED***
    expected: expected,
    actual: 0,
    stack: (new Error).stack,
    name: fn.name || '<anonymous>'
  ***REMOVED***;

  mustCallChecks.push(context);

  return function() ***REMOVED***
    context.actual++;
    return fn.apply(this, arguments);
  ***REMOVED***;
***REMOVED***;
