// @firebase/firestore/dist/esm/src/util/input_validation.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** fail ***REMOVED*** from './assert';
import ***REMOVED*** Code, FirestoreError ***REMOVED*** from './error';
import * as obj from './obj';
/**
 * Validates the invocation of functionName has the exact number of arguments.
 *
 * Forward the magic "arguments" variable as second parameter on which the
 * parameter validation is performed:
 * validateExactNumberOfArgs('myFunction', arguments, 2);
 */
export function validateExactNumberOfArgs(functionName, args, numberOfArgs) ***REMOVED***
    if (args.length !== numberOfArgs) ***REMOVED***
        throw new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires " +
            formatPlural(numberOfArgs, 'argument') +
            ', but was called with ' +
            formatPlural(args.length, 'argument') +
            '.');
    ***REMOVED***
***REMOVED***
/**
 * Validates the invocation of functionName has at least the provided number of
 * arguments (but can have many more).
 *
 * Forward the magic "arguments" variable as second parameter on which the
 * parameter validation is performed:
 * validateAtLeastNumberOfArgs('myFunction', arguments, 2);
 */
export function validateAtLeastNumberOfArgs(functionName, args, minNumberOfArgs) ***REMOVED***
    if (args.length < minNumberOfArgs) ***REMOVED***
        throw new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires at least " +
            formatPlural(minNumberOfArgs, 'argument') +
            ', but was called with ' +
            formatPlural(args.length, 'argument') +
            '.');
    ***REMOVED***
***REMOVED***
/**
 * Validates the invocation of functionName has number of arguments between
 * the values provided.
 *
 * Forward the magic "arguments" variable as second parameter on which the
 * parameter validation is performed:
 * validateBetweenNumberOfArgs('myFunction', arguments, 2, 3);
 */
export function validateBetweenNumberOfArgs(functionName, args, minNumberOfArgs, maxNumberOfArgs) ***REMOVED***
    if (args.length < minNumberOfArgs || args.length > maxNumberOfArgs) ***REMOVED***
        throw new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires between " + minNumberOfArgs + " and " +
            (maxNumberOfArgs + " arguments, but was called with ") +
            formatPlural(args.length, 'argument') +
            '.');
    ***REMOVED***
***REMOVED***
/**
 * Validates the provided argument is an array and has as least the expected
 * number of elements.
 */
export function validateNamedArrayAtLeastNumberOfElements(functionName, value, name, minNumberOfElements) ***REMOVED***
    if (!(value instanceof Array) || value.length < minNumberOfElements) ***REMOVED***
        throw new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires its " + name + " argument to be an " +
            'array with at least ' +
            (formatPlural(minNumberOfElements, 'element') + "."));
    ***REMOVED***
***REMOVED***
/**
 * Validates the provided positional argument has the native JavaScript type
 * using typeof checks.
 */
export function validateArgType(functionName, type, position, argument) ***REMOVED***
    validateType(functionName, type, ordinal(position) + " argument", argument);
***REMOVED***
/**
 * Validates the provided argument has the native JavaScript type using
 * typeof checks or is undefined.
 */
export function validateOptionalArgType(functionName, type, position, argument) ***REMOVED***
    if (argument !== undefined) ***REMOVED***
        validateArgType(functionName, type, position, argument);
    ***REMOVED***
***REMOVED***
/**
 * Validates the provided named option has the native JavaScript type using
 * typeof checks.
 */
export function validateNamedType(functionName, type, optionName, argument) ***REMOVED***
    validateType(functionName, type, optionName + " option", argument);
***REMOVED***
/**
 * Validates the provided named option has the native JavaScript type using
 * typeof checks or is undefined.
 */
export function validateNamedOptionalType(functionName, type, optionName, argument) ***REMOVED***
    if (argument !== undefined) ***REMOVED***
        validateNamedType(functionName, type, optionName, argument);
    ***REMOVED***
***REMOVED***
/**
 * Validates that the provided named option equals one of the expected values.
 */
export function validateNamedPropertyEquals(functionName, inputName, optionName, input, expected) ***REMOVED***
    var expectedDescription = [];
    for (var _i = 0, expected_1 = expected; _i < expected_1.length; _i++) ***REMOVED***
        var val = expected_1[_i];
        if (val === input) ***REMOVED***
            return;
        ***REMOVED***
        expectedDescription.push(valueDescription(val));
    ***REMOVED***
    var actualDescription = valueDescription(input);
    throw new FirestoreError(Code.INVALID_ARGUMENT, "Invalid value " + actualDescription + " provided to function " + functionName + "() for option \"" + optionName + "\". Acceptable values: " + expectedDescription.join(', '));
***REMOVED***
/**
 * Validates that the provided named option equals one of the expected values or
 * is undefined.
 */
export function validateNamedOptionalPropertyEquals(functionName, inputName, optionName, input, expected) ***REMOVED***
    if (input !== undefined) ***REMOVED***
        validateNamedPropertyEquals(functionName, inputName, optionName, input, expected);
    ***REMOVED***
***REMOVED***
/** Helper to validate the type of a provided input. */
function validateType(functionName, type, inputName, input) ***REMOVED***
    if (typeof input !== type || (type === 'object' && !isPlainObject(input))) ***REMOVED***
        var description = valueDescription(input);
        throw new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires its " + inputName + " " +
            ("to be of type " + type + ", but it was: " + description));
    ***REMOVED***
***REMOVED***
/**
 * Returns true iff it's a non-null object without a custom prototype
 * (i.e. excludes Array, Date, etc.).
 */
export function isPlainObject(input) ***REMOVED***
    return (typeof input === 'object' &&
        input !== null &&
        Object.getPrototypeOf(input) === Object.prototype);
***REMOVED***
/** Returns a string describing the type / value of the provided input. */
export function valueDescription(input) ***REMOVED***
    if (input === undefined) ***REMOVED***
        return 'undefined';
    ***REMOVED***
    else if (input === null) ***REMOVED***
        return 'null';
    ***REMOVED***
    else if (typeof input === 'string') ***REMOVED***
        if (input.length > 20) ***REMOVED***
            input = input.substring(0, 20) + "...";
        ***REMOVED***
        return JSON.stringify(input);
    ***REMOVED***
    else if (typeof input === 'number' || typeof input === 'boolean') ***REMOVED***
        return '' + input;
    ***REMOVED***
    else if (typeof input === 'object') ***REMOVED***
        if (input instanceof Array) ***REMOVED***
            return 'an array';
        ***REMOVED***
        else ***REMOVED***
            var customObjectName = tryGetCustomObjectType(input);
            if (customObjectName) ***REMOVED***
                return "a custom " + customObjectName + " object";
            ***REMOVED***
            else ***REMOVED***
                return 'an object';
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    else if (typeof input === 'function') ***REMOVED***
        return 'a function';
    ***REMOVED***
    else ***REMOVED***
        return fail('Unknown wrong type: ' + typeof input);
    ***REMOVED***
***REMOVED***
/** Hacky method to try to get the constructor name for an object. */
export function tryGetCustomObjectType(input) ***REMOVED***
    if (input.constructor) ***REMOVED***
        var funcNameRegex = /function\s+([^\s(]+)\s*\(/;
        var results = funcNameRegex.exec(input.constructor.toString());
        if (results && results.length > 1) ***REMOVED***
            return results[1];
        ***REMOVED***
    ***REMOVED***
    return null;
***REMOVED***
/** Validates the provided argument is defined. */
export function validateDefined(functionName, position, argument) ***REMOVED***
    if (argument === undefined) ***REMOVED***
        throw new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires a valid " + ordinal(position) + " " +
            "argument, but it was undefined.");
    ***REMOVED***
***REMOVED***
/**
 * Validates the provided positional argument is an object, and its keys and
 * values match the expected keys and types provided in optionTypes.
 */
export function validateOptionNames(functionName, options, optionNames) ***REMOVED***
    obj.forEach(options, function (key, _) ***REMOVED***
        if (optionNames.indexOf(key) < 0) ***REMOVED***
            throw new FirestoreError(Code.INVALID_ARGUMENT, "Unknown option '" + key + "' passed to function " + functionName + "(). " +
                'Available options: ' +
                optionNames.join(', '));
        ***REMOVED***
    ***REMOVED***);
***REMOVED***
/**
 * Helper method to throw an error that the provided argument did not pass
 * an instanceof check.
 */
export function invalidClassError(functionName, type, position, argument) ***REMOVED***
    var description = valueDescription(argument);
    return new FirestoreError(Code.INVALID_ARGUMENT, "Function " + functionName + "() requires its " + ordinal(position) + " " +
        ("argument to be a " + type + ", but it was: " + description));
***REMOVED***
/** Converts a number to its english word representation */
function ordinal(num) ***REMOVED***
    switch (num) ***REMOVED***
        case 1:
            return 'first';
        case 2:
            return 'second';
        case 3:
            return 'third';
        default:
            return num + 'th';
    ***REMOVED***
***REMOVED***
/**
 * Formats the given word as plural conditionally given the preceding number.
 */
function formatPlural(num, str) ***REMOVED***
    return num + " " + str + (num === 1 ? '' : 's');
***REMOVED***
