// protobufjs/src/ProtoBuf/Reflect/Message.js
/**
 * Constructs a new Message.
 * @exports ProtoBuf.Reflect.Message
 * @param ***REMOVED***!ProtoBuf.Builder***REMOVED*** builder Builder reference
 * @param ***REMOVED***!ProtoBuf.Reflect.Namespace***REMOVED*** parent Parent message or namespace
 * @param ***REMOVED***string***REMOVED*** name Message name
 * @param ***REMOVED***Object.<string,*>=***REMOVED*** options Message options
 * @param ***REMOVED***boolean=***REMOVED*** isGroup `true` if this is a legacy group
 * @param ***REMOVED***string?***REMOVED*** syntax The syntax level of this definition (e.g., proto3)
 * @constructor
 * @extends ProtoBuf.Reflect.Namespace
 */
var Message = function(builder, parent, name, options, isGroup, syntax) ***REMOVED***
    Namespace.call(this, builder, parent, name, options, syntax);

    /**
     * @override
     */
    this.className = "Message";

    /**
     * Extensions range.
     * @type ***REMOVED***!Array.<number>|undefined***REMOVED***
     * @expose
     */
    this.extensions = undefined;

    /**
     * Runtime message class.
     * @type ***REMOVED***?function(new:ProtoBuf.Builder.Message)***REMOVED***
     * @expose
     */
    this.clazz = null;

    /**
     * Whether this is a legacy group or not.
     * @type ***REMOVED***boolean***REMOVED***
     * @expose
     */
    this.isGroup = !!isGroup;

    // The following cached collections are used to efficiently iterate over or look up fields when decoding.

    /**
     * Cached fields.
     * @type ***REMOVED***?Array.<!ProtoBuf.Reflect.Message.Field>***REMOVED***
     * @private
     */
    this._fields = null;

    /**
     * Cached fields by id.
     * @type ***REMOVED***?Object.<number,!ProtoBuf.Reflect.Message.Field>***REMOVED***
     * @private
     */
    this._fieldsById = null;

    /**
     * Cached fields by name.
     * @type ***REMOVED***?Object.<string,!ProtoBuf.Reflect.Message.Field>***REMOVED***
     * @private
     */
    this._fieldsByName = null;
***REMOVED***;

/**
 * @alias ProtoBuf.Reflect.Message.prototype
 * @inner
 */
var MessagePrototype = Message.prototype = Object.create(Namespace.prototype);

/**
 * Builds the message and returns the runtime counterpart, which is a fully functional class.
 * @see ProtoBuf.Builder.Message
 * @param ***REMOVED***boolean=***REMOVED*** rebuild Whether to rebuild or not, defaults to false
 * @return ***REMOVED***ProtoBuf.Reflect.Message***REMOVED*** Message class
 * @throws ***REMOVED***Error***REMOVED*** If the message cannot be built
 * @expose
 */
MessagePrototype.build = function(rebuild) ***REMOVED***
    if (this.clazz && !rebuild)
        return this.clazz;

    // Create the runtime Message class in its own scope
    var clazz = (function(ProtoBuf, T) ***REMOVED***

        //? include("../Builder/Message.js");

        return Message;

    ***REMOVED***)(ProtoBuf, this);

    // Static enums and prototyped sub-messages / cached collections
    this._fields = [];
    this._fieldsById = ***REMOVED******REMOVED***;
    this._fieldsByName = ***REMOVED******REMOVED***;
    this._oneofsByName = ***REMOVED******REMOVED***;
    for (var i=0, k=this.children.length, child; i<k; i++) ***REMOVED***
        child = this.children[i];
        if (child instanceof Enum || child instanceof Message || child instanceof Service) ***REMOVED***
            if (clazz.hasOwnProperty(child.name))
                throw Error("Illegal reflect child of "+this.toString(true)+": "+child.toString(true)+" cannot override static property '"+child.name+"'");
            clazz[child.name] = child.build();
        ***REMOVED*** else if (child instanceof Message.Field)
            child.build(),
            this._fields.push(child),
            this._fieldsById[child.id] = child,
            this._fieldsByName[child.name] = child;
        else if (child instanceof Message.OneOf) ***REMOVED***
            this._oneofsByName[child.name] = child;
        ***REMOVED***
        else if (!(child instanceof Message.OneOf) && !(child instanceof Extension)) // Not built
            throw Error("Illegal reflect child of "+this.toString(true)+": "+this.children[i].toString(true));
    ***REMOVED***

    return this.clazz = clazz;
***REMOVED***;

/**
 * Encodes a runtime message's contents to the specified buffer.
 * @param ***REMOVED***!ProtoBuf.Builder.Message***REMOVED*** message Runtime message to encode
 * @param ***REMOVED***ByteBuffer***REMOVED*** buffer ByteBuffer to write to
 * @param ***REMOVED***boolean=***REMOVED*** noVerify Whether to not verify field values, defaults to `false`
 * @return ***REMOVED***ByteBuffer***REMOVED*** The ByteBuffer for chaining
 * @throws ***REMOVED***Error***REMOVED*** If required fields are missing or the message cannot be encoded for another reason
 * @expose
 */
MessagePrototype.encode = function(message, buffer, noVerify) ***REMOVED***
    var fieldMissing = null,
        field;
    for (var i=0, k=this._fields.length, val; i<k; ++i) ***REMOVED***
        field = this._fields[i];
        val = message[field.name];
        if (field.required && val === null) ***REMOVED***
            if (fieldMissing === null)
                fieldMissing = field;
        ***REMOVED*** else
            field.encode(noVerify ? val : field.verifyValue(val), buffer, message);
    ***REMOVED***
    if (fieldMissing !== null) ***REMOVED***
        var err = Error("Missing at least one required field for "+this.toString(true)+": "+fieldMissing);
        err["encoded"] = buffer; // Still expose what we got
        throw(err);
    ***REMOVED***
    return buffer;
***REMOVED***;

/**
 * Calculates a runtime message's byte length.
 * @param ***REMOVED***!ProtoBuf.Builder.Message***REMOVED*** message Runtime message to encode
 * @returns ***REMOVED***number***REMOVED*** Byte length
 * @throws ***REMOVED***Error***REMOVED*** If required fields are missing or the message cannot be calculated for another reason
 * @expose
 */
MessagePrototype.calculate = function(message) ***REMOVED***
    for (var n=0, i=0, k=this._fields.length, field, val; i<k; ++i) ***REMOVED***
        field = this._fields[i];
        val = message[field.name];
        if (field.required && val === null)
           throw Error("Missing at least one required field for "+this.toString(true)+": "+field);
        else
            n += field.calculate(val, message);
    ***REMOVED***
    return n;
***REMOVED***;

/**
 * Skips all data until the end of the specified group has been reached.
 * @param ***REMOVED***number***REMOVED*** expectedId Expected GROUPEND id
 * @param ***REMOVED***!ByteBuffer***REMOVED*** buf ByteBuffer
 * @returns ***REMOVED***boolean***REMOVED*** `true` if a value as been skipped, `false` if the end has been reached
 * @throws ***REMOVED***Error***REMOVED*** If it wasn't possible to find the end of the group (buffer overrun or end tag mismatch)
 * @inner
 */
function skipTillGroupEnd(expectedId, buf) ***REMOVED***
    var tag = buf.readVarint32(), // Throws on OOB
        wireType = tag & 0x07,
        id = tag >>> 3;
    switch (wireType) ***REMOVED***
        case ProtoBuf.WIRE_TYPES.VARINT:
            do tag = buf.readUint8();
            while ((tag & 0x80) === 0x80);
            break;
        case ProtoBuf.WIRE_TYPES.BITS64:
            buf.offset += 8;
            break;
        case ProtoBuf.WIRE_TYPES.LDELIM:
            tag = buf.readVarint32(); // reads the varint
            buf.offset += tag;        // skips n bytes
            break;
        case ProtoBuf.WIRE_TYPES.STARTGROUP:
            skipTillGroupEnd(id, buf);
            break;
        case ProtoBuf.WIRE_TYPES.ENDGROUP:
            if (id === expectedId)
                return false;
            else
                throw Error("Illegal GROUPEND after unknown group: "+id+" ("+expectedId+" expected)");
        case ProtoBuf.WIRE_TYPES.BITS32:
            buf.offset += 4;
            break;
        default:
            throw Error("Illegal wire type in unknown group "+expectedId+": "+wireType);
    ***REMOVED***
    return true;
***REMOVED***

/**
 * Decodes an encoded message and returns the decoded message.
 * @param ***REMOVED***ByteBuffer***REMOVED*** buffer ByteBuffer to decode from
 * @param ***REMOVED***number=***REMOVED*** length Message length. Defaults to decode all remaining data.
 * @param ***REMOVED***number=***REMOVED*** expectedGroupEndId Expected GROUPEND id if this is a legacy group
 * @return ***REMOVED***ProtoBuf.Builder.Message***REMOVED*** Decoded message
 * @throws ***REMOVED***Error***REMOVED*** If the message cannot be decoded
 * @expose
 */
MessagePrototype.decode = function(buffer, length, expectedGroupEndId) ***REMOVED***
    if (typeof length !== 'number')
        length = -1;
    var start = buffer.offset,
        msg = new (this.clazz)(),
        tag, wireType, id, field;
    while (buffer.offset < start+length || (length === -1 && buffer.remaining() > 0)) ***REMOVED***
        tag = buffer.readVarint32();
        wireType = tag & 0x07;
        id = tag >>> 3;
        if (wireType === ProtoBuf.WIRE_TYPES.ENDGROUP) ***REMOVED***
            if (id !== expectedGroupEndId)
                throw Error("Illegal group end indicator for "+this.toString(true)+": "+id+" ("+(expectedGroupEndId ? expectedGroupEndId+" expected" : "not a group")+")");
            break;
        ***REMOVED***
        if (!(field = this._fieldsById[id])) ***REMOVED***
            // "messages created by your new code can be parsed by your old code: old binaries simply ignore the new field when parsing."
            switch (wireType) ***REMOVED***
                case ProtoBuf.WIRE_TYPES.VARINT:
                    buffer.readVarint32();
                    break;
                case ProtoBuf.WIRE_TYPES.BITS32:
                    buffer.offset += 4;
                    break;
                case ProtoBuf.WIRE_TYPES.BITS64:
                    buffer.offset += 8;
                    break;
                case ProtoBuf.WIRE_TYPES.LDELIM:
                    var len = buffer.readVarint32();
                    buffer.offset += len;
                    break;
                case ProtoBuf.WIRE_TYPES.STARTGROUP:
                    while (skipTillGroupEnd(id, buffer)) ***REMOVED******REMOVED***
                    break;
                default:
                    throw Error("Illegal wire type for unknown field "+id+" in "+this.toString(true)+"#decode: "+wireType);
            ***REMOVED***
            continue;
        ***REMOVED***
        if (field.repeated && !field.options["packed"]) ***REMOVED***
            msg[field.name].push(field.decode(wireType, buffer));
        ***REMOVED*** else if (field.map) ***REMOVED***
            var keyval = field.decode(wireType, buffer);
            msg[field.name].set(keyval[0], keyval[1]);
        ***REMOVED*** else ***REMOVED***
            msg[field.name] = field.decode(wireType, buffer);
            if (field.oneof) ***REMOVED*** // Field is part of an OneOf (not a virtual OneOf field)
                var currentField = msg[field.oneof.name]; // Virtual field references currently set field
                if (currentField !== null && currentField !== field.name)
                    msg[currentField] = null; // Clear currently set field
                msg[field.oneof.name] = field.name; // Point virtual field at this field
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // Check if all required fields are present and set default values for optional fields that are not
    for (var i=0, k=this._fields.length; i<k; ++i) ***REMOVED***
        field = this._fields[i];
        if (msg[field.name] === null) ***REMOVED***
            if (this.syntax === "proto3") ***REMOVED*** // Proto3 sets default values by specification
                msg[field.name] = field.defaultValue;
            ***REMOVED*** else if (field.required) ***REMOVED***
                var err = Error("Missing at least one required field for " + this.toString(true) + ": " + field.name);
                err["decoded"] = msg; // Still expose what we got
                throw(err);
            ***REMOVED*** else if (ProtoBuf.populateDefaults && field.defaultValue !== null)
                msg[field.name] = field.defaultValue;
        ***REMOVED***
    ***REMOVED***
    return msg;
***REMOVED***;
