// path-browserify/index.js
// Copyright Joyent, Inc. and other Node contributors.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.

// resolves . and .. elements in a path array with directory names there
// must be no slashes, empty elements, or device names (c:\) in the array
// (so also no leading and trailing slashes - it does not distinguish
// relative and absolute paths)
function normalizeArray(parts, allowAboveRoot) ***REMOVED***
  // if the path tries to go above the root, `up` ends up > 0
  var up = 0;
  for (var i = parts.length - 1; i >= 0; i--) ***REMOVED***
    var last = parts[i];
    if (last === '.') ***REMOVED***
      parts.splice(i, 1);
    ***REMOVED*** else if (last === '..') ***REMOVED***
      parts.splice(i, 1);
      up++;
    ***REMOVED*** else if (up) ***REMOVED***
      parts.splice(i, 1);
      up--;
    ***REMOVED***
  ***REMOVED***

  // if the path is allowed to go above the root, restore leading ..s
  if (allowAboveRoot) ***REMOVED***
    for (; up--; up) ***REMOVED***
      parts.unshift('..');
    ***REMOVED***
  ***REMOVED***

  return parts;
***REMOVED***

// Split a filename into [root, dir, basename, ext], unix version
// 'root' is just a slash, or nothing.
var splitPathRe =
    /^(\/?|)([\s\S]*?)((?:\.***REMOVED***1,2***REMOVED***|[^\/]+?|)(\.[^.\/]*|))(?:[\/]*)$/;
var splitPath = function(filename) ***REMOVED***
  return splitPathRe.exec(filename).slice(1);
***REMOVED***;

// path.resolve([from ...], to)
// posix version
exports.resolve = function() ***REMOVED***
  var resolvedPath = '',
      resolvedAbsolute = false;

  for (var i = arguments.length - 1; i >= -1 && !resolvedAbsolute; i--) ***REMOVED***
    var path = (i >= 0) ? arguments[i] : process.cwd();

    // Skip empty and invalid entries
    if (typeof path !== 'string') ***REMOVED***
      throw new TypeError('Arguments to path.resolve must be strings');
    ***REMOVED*** else if (!path) ***REMOVED***
      continue;
    ***REMOVED***

    resolvedPath = path + '/' + resolvedPath;
    resolvedAbsolute = path.charAt(0) === '/';
  ***REMOVED***

  // At this point the path should be resolved to a full absolute path, but
  // handle relative paths to be safe (might happen when process.cwd() fails)

  // Normalize the path
  resolvedPath = normalizeArray(filter(resolvedPath.split('/'), function(p) ***REMOVED***
    return !!p;
  ***REMOVED***), !resolvedAbsolute).join('/');

  return ((resolvedAbsolute ? '/' : '') + resolvedPath) || '.';
***REMOVED***;

// path.normalize(path)
// posix version
exports.normalize = function(path) ***REMOVED***
  var isAbsolute = exports.isAbsolute(path),
      trailingSlash = substr(path, -1) === '/';

  // Normalize the path
  path = normalizeArray(filter(path.split('/'), function(p) ***REMOVED***
    return !!p;
  ***REMOVED***), !isAbsolute).join('/');

  if (!path && !isAbsolute) ***REMOVED***
    path = '.';
  ***REMOVED***
  if (path && trailingSlash) ***REMOVED***
    path += '/';
  ***REMOVED***

  return (isAbsolute ? '/' : '') + path;
***REMOVED***;

// posix version
exports.isAbsolute = function(path) ***REMOVED***
  return path.charAt(0) === '/';
***REMOVED***;

// posix version
exports.join = function() ***REMOVED***
  var paths = Array.prototype.slice.call(arguments, 0);
  return exports.normalize(filter(paths, function(p, index) ***REMOVED***
    if (typeof p !== 'string') ***REMOVED***
      throw new TypeError('Arguments to path.join must be strings');
    ***REMOVED***
    return p;
  ***REMOVED***).join('/'));
***REMOVED***;


// path.relative(from, to)
// posix version
exports.relative = function(from, to) ***REMOVED***
  from = exports.resolve(from).substr(1);
  to = exports.resolve(to).substr(1);

  function trim(arr) ***REMOVED***
    var start = 0;
    for (; start < arr.length; start++) ***REMOVED***
      if (arr[start] !== '') break;
    ***REMOVED***

    var end = arr.length - 1;
    for (; end >= 0; end--) ***REMOVED***
      if (arr[end] !== '') break;
    ***REMOVED***

    if (start > end) return [];
    return arr.slice(start, end - start + 1);
  ***REMOVED***

  var fromParts = trim(from.split('/'));
  var toParts = trim(to.split('/'));

  var length = Math.min(fromParts.length, toParts.length);
  var samePartsLength = length;
  for (var i = 0; i < length; i++) ***REMOVED***
    if (fromParts[i] !== toParts[i]) ***REMOVED***
      samePartsLength = i;
      break;
    ***REMOVED***
  ***REMOVED***

  var outputParts = [];
  for (var i = samePartsLength; i < fromParts.length; i++) ***REMOVED***
    outputParts.push('..');
  ***REMOVED***

  outputParts = outputParts.concat(toParts.slice(samePartsLength));

  return outputParts.join('/');
***REMOVED***;

exports.sep = '/';
exports.delimiter = ':';

exports.dirname = function(path) ***REMOVED***
  var result = splitPath(path),
      root = result[0],
      dir = result[1];

  if (!root && !dir) ***REMOVED***
    // No dirname whatsoever
    return '.';
  ***REMOVED***

  if (dir) ***REMOVED***
    // It has a dirname, strip trailing slash
    dir = dir.substr(0, dir.length - 1);
  ***REMOVED***

  return root + dir;
***REMOVED***;


exports.basename = function(path, ext) ***REMOVED***
  var f = splitPath(path)[2];
  // TODO: make this comparison case-insensitive on windows?
  if (ext && f.substr(-1 * ext.length) === ext) ***REMOVED***
    f = f.substr(0, f.length - ext.length);
  ***REMOVED***
  return f;
***REMOVED***;


exports.extname = function(path) ***REMOVED***
  return splitPath(path)[3];
***REMOVED***;

function filter (xs, f) ***REMOVED***
    if (xs.filter) return xs.filter(f);
    var res = [];
    for (var i = 0; i < xs.length; i++) ***REMOVED***
        if (f(xs[i], i, xs)) res.push(xs[i]);
    ***REMOVED***
    return res;
***REMOVED***

// String.prototype.substr - negative index don't work in IE8
var substr = 'ab'.substr(-1) === 'b'
    ? function (str, start, len) ***REMOVED*** return str.substr(start, len) ***REMOVED***
    : function (str, start, len) ***REMOVED***
        if (start < 0) start = str.length + start;
        return str.substr(start, len);
    ***REMOVED***
;
