// faye-websocket/lib/faye/eventsource.js
var Stream      = require('stream').Stream,
    util        = require('util'),
    driver      = require('websocket-driver'),
    Headers     = require('websocket-driver/lib/websocket/driver/headers'),
    API         = require('./websocket/api'),
    EventTarget = require('./websocket/api/event_target'),
    Event       = require('./websocket/api/event');

var EventSource = function(request, response, options) ***REMOVED***
  this.writable = true;
  options = options || ***REMOVED******REMOVED***;

  this._stream = response.socket;
  this._ping   = options.ping  || this.DEFAULT_PING;
  this._retry  = options.retry || this.DEFAULT_RETRY;

  var scheme       = driver.isSecureRequest(request) ? 'https:' : 'http:';
  this.url         = scheme + '//' + request.headers.host + request.url;
  this.lastEventId = request.headers['last-event-id'] || '';
  this.readyState  = API.CONNECTING;

  var headers = new Headers(),
      self    = this;

  if (options.headers) ***REMOVED***
    for (var key in options.headers) headers.set(key, options.headers[key]);
  ***REMOVED***

  if (!this._stream || !this._stream.writable) return;
  process.nextTick(function() ***REMOVED*** self._open() ***REMOVED***);

  this._stream.setTimeout(0);
  this._stream.setNoDelay(true);

  var handshake = 'HTTP/1.1 200 OK\r\n' +
                  'Content-Type: text/event-stream\r\n' +
                  'Cache-Control: no-cache, no-store\r\n' +
                  'Connection: close\r\n' +
                  headers.toString() +
                  '\r\n' +
                  'retry: ' + Math.floor(this._retry * 1000) + '\r\n\r\n';

  this._write(handshake);

  this._stream.on('drain', function() ***REMOVED*** self.emit('drain') ***REMOVED***);

  if (this._ping)
    this._pingTimer = setInterval(function() ***REMOVED*** self.ping() ***REMOVED***, this._ping * 1000);

  ['error', 'end'].forEach(function(event) ***REMOVED***
    self._stream.on(event, function() ***REMOVED*** self.close() ***REMOVED***);
  ***REMOVED***);
***REMOVED***;
util.inherits(EventSource, Stream);

EventSource.isEventSource = function(request) ***REMOVED***
  if (request.method !== 'GET') return false;
  var accept = (request.headers.accept || '').split(/\s*,\s*/);
  return accept.indexOf('text/event-stream') >= 0;
***REMOVED***;

var instance = ***REMOVED***
  DEFAULT_PING:   10,
  DEFAULT_RETRY:  5,

  _write: function(chunk) ***REMOVED***
    if (!this.writable) return false;
    try ***REMOVED***
      return this._stream.write(chunk, 'utf8');
    ***REMOVED*** catch (e) ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***,

  _open: function() ***REMOVED***
    if (this.readyState !== API.CONNECTING) return;

    this.readyState = API.OPEN;

    var event = new Event('open');
    event.initEvent('open', false, false);
    this.dispatchEvent(event);
  ***REMOVED***,

  write: function(message) ***REMOVED***
    return this.send(message);
  ***REMOVED***,

  end: function(message) ***REMOVED***
    if (message !== undefined) this.write(message);
    this.close();
  ***REMOVED***,

  send: function(message, options) ***REMOVED***
    if (this.readyState > API.OPEN) return false;

    message = String(message).replace(/(\r\n|\r|\n)/g, '$1data: ');
    options = options || ***REMOVED******REMOVED***;

    var frame = '';
    if (options.event) frame += 'event: ' + options.event + '\r\n';
    if (options.id)    frame += 'id: '    + options.id    + '\r\n';
    frame += 'data: ' + message + '\r\n\r\n';

    return this._write(frame);
  ***REMOVED***,

  ping: function() ***REMOVED***
    return this._write(':\r\n\r\n');
  ***REMOVED***,

  close: function() ***REMOVED***
    if (this.readyState > API.OPEN) return false;

    this.readyState = API.CLOSED;
    this.writable = false;
    if (this._pingTimer) clearInterval(this._pingTimer);
    if (this._stream) this._stream.end();

    var event = new Event('close');
    event.initEvent('close', false, false);
    this.dispatchEvent(event);

    return true;
  ***REMOVED***
***REMOVED***;

for (var method in instance) EventSource.prototype[method] = instance[method];
for (var key in EventTarget) EventSource.prototype[key] = EventTarget[key];

module.exports = EventSource;
