// @google-cloud/firestore/src/convert.js
/*!
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const Buffer = require('safe-buffer').Buffer;
const is = require('is');

const validate = require('./validate')();

/*!
 * @module firestore/convert
 * @private
 *
 * This module contains utility functions to convert
 * `firestore.v1beta1.Documents` from Proto3 JSON to their equivalent
 * representation in Protobuf JS. Protobuf JS is the only encoding supported by
 * this client, and dependencies that use Proto3 JSON (such as the Google Cloud
 * Functions SDK) are supported through this conversion and its usage in
 * ***REMOVED***@see Firestore#snapshot_***REMOVED***.
 */

/**
 * Converts an ISO 8601 or google.protobuf.Timestamp proto into Protobuf JS.
 *
 * @private
 * @param ***REMOVED****=***REMOVED*** timestampValue - The value to convert.
 * @param ***REMOVED***string=***REMOVED*** argumentName - The argument name to use in the error message
 * if the conversion fails. If omitted, 'timestampValue' is used.
 * @return ***REMOVED******REMOVED***nanos,seconds***REMOVED***|undefined***REMOVED*** The value as expected by Protobuf JS or
 * undefined if no input was provided.
 */
function convertTimestamp(timestampValue, argumentName) ***REMOVED***
  let timestampProto = undefined;

  if (is.string(timestampValue)) ***REMOVED***
    let date = new Date(timestampValue);
    let seconds = Math.floor(date.getTime() / 1000);
    let nanos = 0;

    if (timestampValue.length > 20) ***REMOVED***
      const nanoString = timestampValue.substring(
        20,
        timestampValue.length - 1
      );
      const trailingZeroes = 9 - nanoString.length;
      nanos = parseInt(nanoString, 10) * Math.pow(10, trailingZeroes);
    ***REMOVED***

    if (isNaN(seconds) || isNaN(nanos)) ***REMOVED***
      argumentName = argumentName || 'timestampValue';
      throw new Error(
        `Specify a valid ISO 8601 timestamp for "$***REMOVED***argumentName***REMOVED***".`
      );
    ***REMOVED***

    timestampProto = ***REMOVED***
      seconds: seconds || undefined,
      nanos: nanos || undefined,
    ***REMOVED***;
  ***REMOVED*** else if (is.defined(timestampValue)) ***REMOVED***
    validate.isObject('timestampValue', timestampValue);
    timestampProto = ***REMOVED***
      seconds: timestampValue.seconds || undefined,
      nanos: timestampValue.nanos || undefined,
    ***REMOVED***;
  ***REMOVED***

  return timestampProto;
***REMOVED***

/**
 * Converts a Proto3 JSON 'bytesValue' field into Protobuf JS.
 *
 * @private
 * @param ***REMOVED*******REMOVED*** bytesValue - The value to convert.
 * @return ***REMOVED***Buffer***REMOVED*** The value as expected by Protobuf JS.
 */
function convertBytes(bytesValue) ***REMOVED***
  if (typeof bytesValue === 'string') ***REMOVED***
    return Buffer.from(bytesValue, 'base64');
  ***REMOVED*** else ***REMOVED***
    return bytesValue;
  ***REMOVED***
***REMOVED***

/**
 * Detects 'valueType' from a Proto3 JSON `firestore.v1beta1.Value` proto.
 *
 * @private
 * @param ***REMOVED***object***REMOVED*** proto - The `firestore.v1beta1.Value` proto.
 * @return ***REMOVED***string***REMOVED*** - The string value for 'valueType'.
 */
function detectValueType(proto) ***REMOVED***
  if (proto.valueType) ***REMOVED***
    return proto.valueType;
  ***REMOVED***

  let detectedValues = [];

  if (is.defined(proto.stringValue)) ***REMOVED***
    detectedValues.push('stringValue');
  ***REMOVED***
  if (is.defined(proto.booleanValue)) ***REMOVED***
    detectedValues.push('booleanValue');
  ***REMOVED***
  if (is.defined(proto.integerValue)) ***REMOVED***
    detectedValues.push('integerValue');
  ***REMOVED***
  if (is.defined(proto.doubleValue)) ***REMOVED***
    detectedValues.push('doubleValue');
  ***REMOVED***
  if (is.defined(proto.timestampValue)) ***REMOVED***
    detectedValues.push('timestampValue');
  ***REMOVED***
  if (is.defined(proto.referenceValue)) ***REMOVED***
    detectedValues.push('referenceValue');
  ***REMOVED***
  if (is.defined(proto.arrayValue)) ***REMOVED***
    detectedValues.push('arrayValue');
  ***REMOVED***
  if (is.defined(proto.nullValue)) ***REMOVED***
    detectedValues.push('nullValue');
  ***REMOVED***
  if (is.defined(proto.mapValue)) ***REMOVED***
    detectedValues.push('mapValue');
  ***REMOVED***
  if (is.defined(proto.geoPointValue)) ***REMOVED***
    detectedValues.push('geoPointValue');
  ***REMOVED***
  if (is.defined(proto.bytesValue)) ***REMOVED***
    detectedValues.push('bytesValue');
  ***REMOVED***

  if (detectedValues.length !== 1) ***REMOVED***
    throw new Error(
      `Unable to infer type value fom '$***REMOVED***JSON.stringify(proto)***REMOVED***'.`
    );
  ***REMOVED***

  return detectedValues[0];
***REMOVED***

/**
 * Converts a `firestore.v1beta1.Value` in Proto3 JSON encoding into the
 * Protobuf JS format expected by this client.
 *
 * @private
 * @param ***REMOVED***object***REMOVED*** fieldValue - The `firestore.v1beta1.Value` in Proto3 JSON
 * format.
 * @return ***REMOVED***object***REMOVED*** The `firestore.v1beta1.Value` in Protobuf JS format.
 */
function convertValue(fieldValue) ***REMOVED***
  let valueType = detectValueType(fieldValue);

  switch (valueType) ***REMOVED***
    case 'timestampValue':
      return ***REMOVED***
        valueType: 'timestampValue',
        timestampValue: convertTimestamp(fieldValue.timestampValue),
      ***REMOVED***;
    case 'bytesValue':
      return ***REMOVED***
        valueType: 'bytesValue',
        bytesValue: convertBytes(fieldValue.bytesValue),
      ***REMOVED***;
    case 'arrayValue': ***REMOVED***
      let arrayValue = [];
      if (is.array(fieldValue.arrayValue.values)) ***REMOVED***
        for (let value of fieldValue.arrayValue.values) ***REMOVED***
          arrayValue.push(convertValue(value));
        ***REMOVED***
      ***REMOVED***
      return ***REMOVED***
        valueType: 'arrayValue',
        arrayValue: ***REMOVED***
          values: arrayValue,
        ***REMOVED***,
      ***REMOVED***;
    ***REMOVED***
    case 'mapValue': ***REMOVED***
      let mapValue = ***REMOVED******REMOVED***;
      for (let prop in fieldValue.mapValue.fields) ***REMOVED***
        if (fieldValue.mapValue.fields.hasOwnProperty(prop)) ***REMOVED***
          mapValue[prop] = convertValue(fieldValue.mapValue.fields[prop]);
        ***REMOVED***
      ***REMOVED***
      return ***REMOVED***
        valueType: 'mapValue',
        mapValue: ***REMOVED***
          fields: mapValue,
        ***REMOVED***,
      ***REMOVED***;
    ***REMOVED***
    default:
      return Object.assign(***REMOVED***valueType***REMOVED***, fieldValue);
  ***REMOVED***
***REMOVED***

/**
 * Converts a `firestore.v1beta1.Document` in Proto3 JSON encoding into the
 * Protobuf JS format expected by this client. This conversion creates a copy of
 * the underlying document.
 *
 * @private
 * @param ***REMOVED***object***REMOVED*** document - The `firestore.v1beta1.Document` in Proto3 JSON
 * format.
 * @return ***REMOVED***object***REMOVED*** The `firestore.v1beta1.Document` in Protobuf JS format.
 */
function convertDocument(document) ***REMOVED***
  let result = ***REMOVED******REMOVED***;

  for (let prop in document) ***REMOVED***
    if (document.hasOwnProperty(prop)) ***REMOVED***
      result[prop] = convertValue(document[prop]);
    ***REMOVED***
  ***REMOVED***

  return result;
***REMOVED***

module.exports = ***REMOVED***
  documentFromJson: convertDocument,
  timestampFromJson: convertTimestamp,
  valueFromJson: convertValue,
***REMOVED***;
