// nodemon/lib/config/index.js
/**
 * Manages the internal config of nodemon, checking for the state of support
 * with fs.watch, how nodemon can watch files (using find or fs methods).
 *
 * This is *not* the user's config.
 */
var debug = require('debug')('nodemon');
var load = require('./load');
var rules = require('../rules');
var utils = require('../utils');
var pinVersion = require('../version').pin;
var command = require('./command');
var rulesToMonitor = require('../monitor/match').rulesToMonitor;
var bus = utils.bus;

function reset() ***REMOVED***
  rules.reset();

  config.dirs = [];
  config.options = ***REMOVED*** ignore: [], watch: [] ***REMOVED***;
  config.lastStarted = 0;
  config.loaded = [];
***REMOVED***

var config = ***REMOVED***
  run: false,
  system: ***REMOVED***
    cwd: process.cwd(),
  ***REMOVED***,
  required: false,
  dirs: [],
  timeout: 1000,
  options: ***REMOVED******REMOVED***,
***REMOVED***;

/**
 * Take user defined settings, then detect the local machine capability, then
 * look for local and global nodemon.json files and merge together the final
 * settings with the config for nodemon.
 *
 * @param  ***REMOVED***Object***REMOVED*** settings user defined settings for nodemon (typically on
 *  the cli)
 * @param  ***REMOVED***Function***REMOVED*** ready callback fired once the config is loaded
 */
config.load = function (settings, ready) ***REMOVED***
  reset();
  var config = this;
  load(settings, config.options, config, function (options) ***REMOVED***
    config.options = options;

    if (options.watch.length === 0) ***REMOVED***
      // this is to catch when the watch is left blank
      options.watch.push('*.*');
    ***REMOVED***

    if (options['watch_interval']) ***REMOVED*** // jshint ignore:line
      options.watchInterval = options['watch_interval']; // jshint ignore:line
    ***REMOVED***

    config.watchInterval = options.watchInterval || null;
    if (options.signal) ***REMOVED***
      config.signal = options.signal;
    ***REMOVED***

    var cmd = command(config.options);
    config.command = ***REMOVED***
      raw: cmd,
      string: utils.stringify(cmd.executable, cmd.args),
    ***REMOVED***;

    // now run automatic checks on system adding to the config object
    options.monitor = rulesToMonitor(options.watch, options.ignore, config);

    var cwd = process.cwd();
    debug('config: dirs', config.dirs);
    if (config.dirs.length === 0) ***REMOVED***
      config.dirs.unshift(cwd);
    ***REMOVED***

    bus.emit('config:update', config);
    pinVersion().then(function () ***REMOVED***
      ready(config);
    ***REMOVED***).catch(e => ***REMOVED***
      // this doesn't help testing, but does give exposure on syntax errors
      console.error(e.stack);
      throw e;
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***;

config.reset = reset;

module.exports = config;
