// @firebase/firestore/dist/cjs/src/core/snapshot_version.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var timestamp_1 = require("./timestamp");
/**
 * A version of a document in Firestore. This corresponds to the version
 * timestamp, such as update_time or read_time.
 */
var SnapshotVersion = /** @class */ (function () ***REMOVED***
    function SnapshotVersion(timestamp) ***REMOVED***
        this.timestamp = timestamp;
    ***REMOVED***
    // TODO(b/34176344): Once we no longer need to use the old alpha protos,
    // delete this constructor and use a timestamp-backed version everywhere.
    SnapshotVersion.fromMicroseconds = function (value) ***REMOVED***
        var seconds = Math.floor(value / 1e6);
        var nanos = (value % 1e6) * 1e3;
        return new SnapshotVersion(new timestamp_1.Timestamp(seconds, nanos));
    ***REMOVED***;
    SnapshotVersion.fromTimestamp = function (value) ***REMOVED***
        return new SnapshotVersion(value);
    ***REMOVED***;
    SnapshotVersion.forDeletedDoc = function () ***REMOVED***
        return SnapshotVersion.MIN;
    ***REMOVED***;
    SnapshotVersion.prototype.compareTo = function (other) ***REMOVED***
        return this.timestamp.compareTo(other.timestamp);
    ***REMOVED***;
    SnapshotVersion.prototype.isEqual = function (other) ***REMOVED***
        return this.timestamp.isEqual(other.timestamp);
    ***REMOVED***;
    /** Returns a number representation of the version for use in spec tests. */
    SnapshotVersion.prototype.toMicroseconds = function () ***REMOVED***
        // Convert to microseconds.
        return this.timestamp.seconds * 1e6 + this.timestamp.nanos / 1000;
    ***REMOVED***;
    SnapshotVersion.prototype.toString = function () ***REMOVED***
        return 'SnapshotVersion(' + this.timestamp.toString() + ')';
    ***REMOVED***;
    SnapshotVersion.prototype.toTimestamp = function () ***REMOVED***
        return this.timestamp;
    ***REMOVED***;
    SnapshotVersion.MIN = new SnapshotVersion(new timestamp_1.Timestamp(0, 0));
    return SnapshotVersion;
***REMOVED***());
exports.SnapshotVersion = SnapshotVersion;
