// @firebase/storage/dist/cjs/src/service.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var args = require("./implementation/args");
var authwrapper_1 = require("./implementation/authwrapper");
var location_1 = require("./implementation/location");
var fbsPromiseImpl = require("./implementation/promise_external");
var RequestExports = require("./implementation/request");
var reference_1 = require("./reference");
/**
 * A service that provides firebaseStorage.Reference instances.
 * @param opt_url gs:// url to a custom Storage Bucket
 *
 * @struct
 */
var Service = /** @class */ (function () ***REMOVED***
    function Service(app, pool, url) ***REMOVED***
        this.bucket_ = null;
        function maker(authWrapper, loc) ***REMOVED***
            return new reference_1.Reference(authWrapper, loc);
        ***REMOVED***
        this.authWrapper_ = new authwrapper_1.AuthWrapper(app, maker, RequestExports.makeRequest, this, pool);
        this.app_ = app;
        if (url != null) ***REMOVED***
            this.bucket_ = location_1.Location.makeFromBucketSpec(url);
        ***REMOVED***
        else ***REMOVED***
            var authWrapperBucket = this.authWrapper_.bucket();
            if (authWrapperBucket != null) ***REMOVED***
                this.bucket_ = new location_1.Location(authWrapperBucket, '');
            ***REMOVED***
        ***REMOVED***
        this.internals_ = new ServiceInternals(this);
    ***REMOVED***
    /**
     * Returns a firebaseStorage.Reference for the given path in the default
     * bucket.
     */
    Service.prototype.ref = function (path) ***REMOVED***
        function validator(path) ***REMOVED***
            if (/^[A-Za-z]+:\/\//.test(path)) ***REMOVED***
                throw 'Expected child path but got a URL, use refFromURL instead.';
            ***REMOVED***
        ***REMOVED***
        args.validate('ref', [args.stringSpec(validator, true)], arguments);
        if (this.bucket_ == null) ***REMOVED***
            throw new Error('No Storage Bucket defined in Firebase Options.');
        ***REMOVED***
        var ref = new reference_1.Reference(this.authWrapper_, this.bucket_);
        if (path != null) ***REMOVED***
            return ref.child(path);
        ***REMOVED***
        else ***REMOVED***
            return ref;
        ***REMOVED***
    ***REMOVED***;
    /**
     * Returns a firebaseStorage.Reference object for the given absolute URL,
     * which must be a gs:// or http[s]:// URL.
     */
    Service.prototype.refFromURL = function (url) ***REMOVED***
        function validator(p) ***REMOVED***
            if (!/^[A-Za-z]+:\/\//.test(p)) ***REMOVED***
                throw 'Expected full URL but got a child path, use ref instead.';
            ***REMOVED***
            try ***REMOVED***
                location_1.Location.makeFromUrl(p);
            ***REMOVED***
            catch (e) ***REMOVED***
                throw 'Expected valid full URL but got an invalid one.';
            ***REMOVED***
        ***REMOVED***
        args.validate('refFromURL', [args.stringSpec(validator, false)], arguments);
        return new reference_1.Reference(this.authWrapper_, url);
    ***REMOVED***;
    Object.defineProperty(Service.prototype, "maxUploadRetryTime", ***REMOVED***
        get: function () ***REMOVED***
            return this.authWrapper_.maxUploadRetryTime();
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Service.prototype.setMaxUploadRetryTime = function (time) ***REMOVED***
        args.validate('setMaxUploadRetryTime', [args.nonNegativeNumberSpec()], arguments);
        this.authWrapper_.setMaxUploadRetryTime(time);
    ***REMOVED***;
    Object.defineProperty(Service.prototype, "maxOperationRetryTime", ***REMOVED***
        get: function () ***REMOVED***
            return this.authWrapper_.maxOperationRetryTime();
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Service.prototype.setMaxOperationRetryTime = function (time) ***REMOVED***
        args.validate('setMaxOperationRetryTime', [args.nonNegativeNumberSpec()], arguments);
        this.authWrapper_.setMaxOperationRetryTime(time);
    ***REMOVED***;
    Object.defineProperty(Service.prototype, "app", ***REMOVED***
        get: function () ***REMOVED***
            return this.app_;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(Service.prototype, "INTERNAL", ***REMOVED***
        get: function () ***REMOVED***
            return this.internals_;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    return Service;
***REMOVED***());
exports.Service = Service;
/**
 * @struct
 */
var ServiceInternals = /** @class */ (function () ***REMOVED***
    function ServiceInternals(service) ***REMOVED***
        this.service_ = service;
    ***REMOVED***
    /**
     * Called when the associated app is deleted.
     * @see ***REMOVED***!fbs.AuthWrapper.prototype.deleteApp***REMOVED***
     */
    ServiceInternals.prototype.delete = function () ***REMOVED***
        this.service_.authWrapper_.deleteApp();
        return fbsPromiseImpl.resolve(undefined);
    ***REMOVED***;
    return ServiceInternals;
***REMOVED***());
exports.ServiceInternals = ServiceInternals;
