// @firebase/database/dist/esm/src/api/Database.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** fatal ***REMOVED*** from '../core/util/util';
import ***REMOVED*** parseRepoInfo ***REMOVED*** from '../core/util/libs/parser';
import ***REMOVED*** Path ***REMOVED*** from '../core/util/Path';
import ***REMOVED*** Reference ***REMOVED*** from './Reference';
import ***REMOVED*** Repo ***REMOVED*** from '../core/Repo';
import ***REMOVED*** RepoManager ***REMOVED*** from '../core/RepoManager';
import ***REMOVED*** validateArgCount ***REMOVED*** from '@firebase/util';
import ***REMOVED*** validateUrl ***REMOVED*** from '../core/util/validation';
/**
 * Class representing a firebase database.
 * @implements ***REMOVED***FirebaseService***REMOVED***
 */
var Database = /** @class */ (function () ***REMOVED***
    /**
     * The constructor should not be called by users of our public API.
     * @param ***REMOVED***!Repo***REMOVED*** repo_
     */
    function Database(repo_) ***REMOVED***
        this.repo_ = repo_;
        if (!(repo_ instanceof Repo)) ***REMOVED***
            fatal("Don't call new Database() directly - please use firebase.database().");
        ***REMOVED***
        /** @type ***REMOVED***Reference***REMOVED*** */
        this.root_ = new Reference(repo_, Path.Empty);
        this.INTERNAL = new DatabaseInternals(this);
    ***REMOVED***
    Object.defineProperty(Database.prototype, "app", ***REMOVED***
        get: function () ***REMOVED***
            return this.repo_.app;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    /**
     * Returns a reference to the root or the path specified in opt_pathString.
     * @param ***REMOVED***string=***REMOVED*** pathString
     * @return ***REMOVED***!Reference***REMOVED*** Firebase reference.
     */
    Database.prototype.ref = function (pathString) ***REMOVED***
        this.checkDeleted_('ref');
        validateArgCount('database.ref', 0, 1, arguments.length);
        return pathString !== undefined ? this.root_.child(pathString) : this.root_;
    ***REMOVED***;
    /**
     * Returns a reference to the root or the path specified in url.
     * We throw a exception if the url is not in the same domain as the
     * current repo.
     * @param ***REMOVED***string***REMOVED*** url
     * @return ***REMOVED***!Reference***REMOVED*** Firebase reference.
     */
    Database.prototype.refFromURL = function (url) ***REMOVED***
        /** @const ***REMOVED***string***REMOVED*** */
        var apiName = 'database.refFromURL';
        this.checkDeleted_(apiName);
        validateArgCount(apiName, 1, 1, arguments.length);
        var parsedURL = parseRepoInfo(url);
        validateUrl(apiName, 1, parsedURL);
        var repoInfo = parsedURL.repoInfo;
        if (repoInfo.host !== this.repo_.repoInfo_.host) ***REMOVED***
            fatal(apiName +
                ': Host name does not match the current database: ' +
                '(found ' +
                repoInfo.host +
                ' but expected ' +
                this.repo_.repoInfo_.host +
                ')');
        ***REMOVED***
        return this.ref(parsedURL.path.toString());
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** apiName
     */
    Database.prototype.checkDeleted_ = function (apiName) ***REMOVED***
        if (this.repo_ === null) ***REMOVED***
            fatal('Cannot call ' + apiName + ' on a deleted database.');
        ***REMOVED***
    ***REMOVED***;
    // Make individual repo go offline.
    Database.prototype.goOffline = function () ***REMOVED***
        validateArgCount('database.goOffline', 0, 0, arguments.length);
        this.checkDeleted_('goOffline');
        this.repo_.interrupt();
    ***REMOVED***;
    Database.prototype.goOnline = function () ***REMOVED***
        validateArgCount('database.goOnline', 0, 0, arguments.length);
        this.checkDeleted_('goOnline');
        this.repo_.resume();
    ***REMOVED***;
    Database.ServerValue = ***REMOVED***
        TIMESTAMP: ***REMOVED***
            '.sv': 'timestamp'
        ***REMOVED***
    ***REMOVED***;
    return Database;
***REMOVED***());
export ***REMOVED*** Database ***REMOVED***;
var DatabaseInternals = /** @class */ (function () ***REMOVED***
    /** @param ***REMOVED***!Database***REMOVED*** database */
    function DatabaseInternals(database) ***REMOVED***
        this.database = database;
    ***REMOVED***
    /** @return ***REMOVED***Promise<void>***REMOVED*** */
    DatabaseInternals.prototype.delete = function () ***REMOVED***
        this.database.checkDeleted_('delete');
        RepoManager.getInstance().deleteRepo(this.database.repo_);
        this.database.repo_ = null;
        this.database.root_ = null;
        this.database.INTERNAL = null;
        this.database = null;
        return Promise.resolve();
    ***REMOVED***;
    return DatabaseInternals;
***REMOVED***());
export ***REMOVED*** DatabaseInternals ***REMOVED***;
