// @firebase/database/dist/esm/src/api/internal.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** WebSocketConnection ***REMOVED*** from '../realtime/WebSocketConnection';
import ***REMOVED*** BrowserPollConnection ***REMOVED*** from '../realtime/BrowserPollConnection';
/**
 * INTERNAL methods for internal-use only (tests, etc.).
 *
 * Customers shouldn't use these or else should be aware that they could break at any time.
 *
 * @const
 */
export var forceLongPolling = function () ***REMOVED***
    WebSocketConnection.forceDisallow();
    BrowserPollConnection.forceAllow();
***REMOVED***;
export var forceWebSockets = function () ***REMOVED***
    BrowserPollConnection.forceDisallow();
***REMOVED***;
/* Used by App Manager */
export var isWebSocketsAvailable = function () ***REMOVED***
    return WebSocketConnection['isAvailable']();
***REMOVED***;
export var setSecurityDebugCallback = function (ref, callback) ***REMOVED***
    ref.repo.persistentConnection_.securityDebugCallback_ = callback;
***REMOVED***;
export var stats = function (ref, showDelta) ***REMOVED***
    ref.repo.stats(showDelta);
***REMOVED***;
export var statsIncrementCounter = function (ref, metric) ***REMOVED***
    ref.repo.statsIncrementCounter(metric);
***REMOVED***;
export var dataUpdateCount = function (ref) ***REMOVED***
    return ref.repo.dataUpdateCount;
***REMOVED***;
export var interceptServerData = function (ref, callback) ***REMOVED***
    return ref.repo.interceptServerData_(callback);
***REMOVED***;
