// @firebase/database/dist/esm/src/api/onDisconnect.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** validateArgCount, validateCallback ***REMOVED*** from '@firebase/util';
import ***REMOVED*** validateWritablePath, validateFirebaseDataArg, validatePriority, validateFirebaseMergeDataArg ***REMOVED*** from '../core/util/validation';
import ***REMOVED*** warn ***REMOVED*** from '../core/util/util';
import ***REMOVED*** Deferred ***REMOVED*** from '@firebase/util';
/**
 * @constructor
 */
var OnDisconnect = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***!Repo***REMOVED*** repo_
     * @param ***REMOVED***!Path***REMOVED*** path_
     */
    function OnDisconnect(repo_, path_) ***REMOVED***
        this.repo_ = repo_;
        this.path_ = path_;
    ***REMOVED***
    /**
     * @param ***REMOVED***function(?Error)=***REMOVED*** onComplete
     * @return ***REMOVED***!firebase.Promise***REMOVED***
     */
    OnDisconnect.prototype.cancel = function (onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.cancel', 0, 1, arguments.length);
        validateCallback('OnDisconnect.cancel', 1, onComplete, true);
        var deferred = new Deferred();
        this.repo_.onDisconnectCancel(this.path_, deferred.wrapCallback(onComplete));
        return deferred.promise;
    ***REMOVED***;
    /**
     * @param ***REMOVED***function(?Error)=***REMOVED*** onComplete
     * @return ***REMOVED***!firebase.Promise***REMOVED***
     */
    OnDisconnect.prototype.remove = function (onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.remove', 0, 1, arguments.length);
        validateWritablePath('OnDisconnect.remove', this.path_);
        validateCallback('OnDisconnect.remove', 1, onComplete, true);
        var deferred = new Deferred();
        this.repo_.onDisconnectSet(this.path_, null, deferred.wrapCallback(onComplete));
        return deferred.promise;
    ***REMOVED***;
    /**
     * @param ***REMOVED*******REMOVED*** value
     * @param ***REMOVED***function(?Error)=***REMOVED*** onComplete
     * @return ***REMOVED***!firebase.Promise***REMOVED***
     */
    OnDisconnect.prototype.set = function (value, onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.set', 1, 2, arguments.length);
        validateWritablePath('OnDisconnect.set', this.path_);
        validateFirebaseDataArg('OnDisconnect.set', 1, value, this.path_, false);
        validateCallback('OnDisconnect.set', 2, onComplete, true);
        var deferred = new Deferred();
        this.repo_.onDisconnectSet(this.path_, value, deferred.wrapCallback(onComplete));
        return deferred.promise;
    ***REMOVED***;
    /**
     * @param ***REMOVED*******REMOVED*** value
     * @param ***REMOVED***number|string|null***REMOVED*** priority
     * @param ***REMOVED***function(?Error)=***REMOVED*** onComplete
     * @return ***REMOVED***!firebase.Promise***REMOVED***
     */
    OnDisconnect.prototype.setWithPriority = function (value, priority, onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.setWithPriority', 2, 3, arguments.length);
        validateWritablePath('OnDisconnect.setWithPriority', this.path_);
        validateFirebaseDataArg('OnDisconnect.setWithPriority', 1, value, this.path_, false);
        validatePriority('OnDisconnect.setWithPriority', 2, priority, false);
        validateCallback('OnDisconnect.setWithPriority', 3, onComplete, true);
        var deferred = new Deferred();
        this.repo_.onDisconnectSetWithPriority(this.path_, value, priority, deferred.wrapCallback(onComplete));
        return deferred.promise;
    ***REMOVED***;
    /**
     * @param ***REMOVED***!Object***REMOVED*** objectToMerge
     * @param ***REMOVED***function(?Error)=***REMOVED*** onComplete
     * @return ***REMOVED***!firebase.Promise***REMOVED***
     */
    OnDisconnect.prototype.update = function (objectToMerge, onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.update', 1, 2, arguments.length);
        validateWritablePath('OnDisconnect.update', this.path_);
        if (Array.isArray(objectToMerge)) ***REMOVED***
            var newObjectToMerge = ***REMOVED******REMOVED***;
            for (var i = 0; i < objectToMerge.length; ++i) ***REMOVED***
                newObjectToMerge['' + i] = objectToMerge[i];
            ***REMOVED***
            objectToMerge = newObjectToMerge;
            warn('Passing an Array to firebase.database.onDisconnect().update() is deprecated. Use set() if you want to overwrite the ' +
                'existing data, or an Object with integer keys if you really do want to only update some of the children.');
        ***REMOVED***
        validateFirebaseMergeDataArg('OnDisconnect.update', 1, objectToMerge, this.path_, false);
        validateCallback('OnDisconnect.update', 2, onComplete, true);
        var deferred = new Deferred();
        this.repo_.onDisconnectUpdate(this.path_, objectToMerge, deferred.wrapCallback(onComplete));
        return deferred.promise;
    ***REMOVED***;
    return OnDisconnect;
***REMOVED***());
export ***REMOVED*** OnDisconnect ***REMOVED***;
