// @firebase/database/dist/esm/src/api/test_access.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** RepoInfo ***REMOVED*** from '../core/RepoInfo';
import ***REMOVED*** PersistentConnection ***REMOVED*** from '../core/PersistentConnection';
import ***REMOVED*** RepoManager ***REMOVED*** from '../core/RepoManager';
import ***REMOVED*** Connection ***REMOVED*** from '../realtime/Connection';
export var DataConnection = PersistentConnection;
/**
 * @param ***REMOVED***!string***REMOVED*** pathString
 * @param ***REMOVED***function(*)***REMOVED*** onComplete
 */
PersistentConnection.prototype.simpleListen = function (pathString, onComplete) ***REMOVED***
    this.sendRequest('q', ***REMOVED*** p: pathString ***REMOVED***, onComplete);
***REMOVED***;
/**
 * @param ***REMOVED*******REMOVED*** data
 * @param ***REMOVED***function(*)***REMOVED*** onEcho
 */
PersistentConnection.prototype.echo = function (data, onEcho) ***REMOVED***
    this.sendRequest('echo', ***REMOVED*** d: data ***REMOVED***, onEcho);
***REMOVED***;
// RealTimeConnection properties that we use in tests.
export var RealTimeConnection = Connection;
/**
 * @param ***REMOVED***function(): string***REMOVED*** newHash
 * @return ***REMOVED***function()***REMOVED***
 */
export var hijackHash = function (newHash) ***REMOVED***
    var oldPut = PersistentConnection.prototype.put;
    PersistentConnection.prototype.put = function (pathString, data, opt_onComplete, opt_hash) ***REMOVED***
        if (opt_hash !== undefined) ***REMOVED***
            opt_hash = newHash();
        ***REMOVED***
        oldPut.call(this, pathString, data, opt_onComplete, opt_hash);
    ***REMOVED***;
    return function () ***REMOVED***
        PersistentConnection.prototype.put = oldPut;
    ***REMOVED***;
***REMOVED***;
/**
 * @type ***REMOVED***function(new:RepoInfo, !string, boolean, !string, boolean): undefined***REMOVED***
 */
export var ConnectionTarget = RepoInfo;
/**
 * @param ***REMOVED***!Query***REMOVED*** query
 * @return ***REMOVED***!string***REMOVED***
 */
export var queryIdentifier = function (query) ***REMOVED***
    return query.queryIdentifier();
***REMOVED***;
/**
 * @param ***REMOVED***!Query***REMOVED*** firebaseRef
 * @return ***REMOVED***!Object***REMOVED***
 */
export var listens = function (firebaseRef) ***REMOVED***
    return firebaseRef.repo.persistentConnection_.listens_;
***REMOVED***;
/**
 * Forces the RepoManager to create Repos that use ReadonlyRestClient instead of PersistentConnection.
 *
 * @param ***REMOVED***boolean***REMOVED*** forceRestClient
 */
export var forceRestClient = function (forceRestClient) ***REMOVED***
    RepoManager.getInstance().forceRestClient(forceRestClient);
***REMOVED***;
