// stringifier/index.js
/**
 * stringifier
 * 
 * https://github.com/twada/stringifier
 *
 * Copyright (c) 2014-2015 Takuto Wada
 * Licensed under the MIT license.
 *   http://twada.mit-license.org/2014-2015
 */
'use strict';

var traverse = require('traverse');
var typeName = require('type-name');
var assign = require('core-js/library/fn/object/assign');
var s = require('./strategies');

function defaultHandlers () ***REMOVED***
    return ***REMOVED***
        'null': s.always('null'),
        'undefined': s.always('undefined'),
        'function': s.prune(),
        'string': s.json(),
        'boolean': s.json(),
        'number': s.number(),
        'symbol': s.toStr(),
        'RegExp': s.toStr(),
        'String': s.newLike(),
        'Boolean': s.newLike(),
        'Number': s.newLike(),
        'Date': s.newLike(),
        'Array': s.array(),
        'Object': s.object(),
        '@default': s.object()
    ***REMOVED***;
***REMOVED***

function defaultOptions () ***REMOVED***
    return ***REMOVED***
        maxDepth: null,
        indent: null,
        anonymous: '@Anonymous',
        circular: '#@Circular#',
        snip: '..(snip)',
        lineSeparator: '\n',
        typeFun: typeName
    ***REMOVED***;
***REMOVED***

function createStringifier (customOptions) ***REMOVED***
    var options = assign(***REMOVED******REMOVED***, defaultOptions(), customOptions);
    var handlers = assign(***REMOVED******REMOVED***, defaultHandlers(), options.handlers);
    return function stringifyAny (push, x) ***REMOVED***
        var context = this;
        var handler = handlerFor(context.node, options, handlers);
        var currentPath = '/' + context.path.join('/');
        var customization = handlers[currentPath];
        var acc = ***REMOVED***
            context: context,
            options: options,
            handlers: handlers,
            push: push
        ***REMOVED***;
        if (typeName(customization) === 'function') ***REMOVED***
            handler = customization;
        ***REMOVED*** else if (typeName(customization) === 'number') ***REMOVED***
            handler = s.flow.compose(s.filters.truncate(customization),handler);
        ***REMOVED*** else if (context.parent && typeName(context.parent.node) === 'Array' && !(context.key in context.parent.node)) ***REMOVED***
            // sparse arrays
            handler = s.always('');
        ***REMOVED***
        handler(acc, x);
        return push;
    ***REMOVED***;
***REMOVED***

function handlerFor (val, options, handlers) ***REMOVED***
    var tname = options.typeFun(val);
    if (typeName(handlers[tname]) === 'function') ***REMOVED***
        return handlers[tname];
    ***REMOVED***
    return handlers['@default'];
***REMOVED***

function walk (val, reducer) ***REMOVED***
    var buffer = [];
    var push = function (str) ***REMOVED***
        buffer.push(str);
    ***REMOVED***;
    traverse(val).reduce(reducer, push);
    return buffer.join('');
***REMOVED***

function stringify (val, options) ***REMOVED***
    return walk(val, createStringifier(options));
***REMOVED***

function stringifier (options) ***REMOVED***
    return function (val) ***REMOVED***
        return walk(val, createStringifier(options));
    ***REMOVED***;
***REMOVED***

stringifier.stringify = stringify;
stringifier.strategies = s;
stringifier.defaultOptions = defaultOptions;
stringifier.defaultHandlers = defaultHandlers;
module.exports = stringifier;
