// protobufjs/src/ProtoBuf/Reflect/Service/RPCMethod.js
/**
 * RPC service method.
 * @exports ProtoBuf.Reflect.Service.RPCMethod
 * @param ***REMOVED***!ProtoBuf.Builder***REMOVED*** builder Builder reference
 * @param ***REMOVED***!ProtoBuf.Reflect.Service***REMOVED*** svc Service
 * @param ***REMOVED***string***REMOVED*** name Method name
 * @param ***REMOVED***string***REMOVED*** request Request message name
 * @param ***REMOVED***string***REMOVED*** response Response message name
 * @param ***REMOVED***boolean***REMOVED*** request_stream Whether requests are streamed
 * @param ***REMOVED***boolean***REMOVED*** response_stream Whether responses are streamed
 * @param ***REMOVED***Object.<string,*>=***REMOVED*** options Options
 * @constructor
 * @extends ProtoBuf.Reflect.Service.Method
 */
var RPCMethod = function(builder, svc, name, request, response, request_stream, response_stream, options) ***REMOVED***
    Method.call(this, builder, svc, name, options);

    /**
     * @override
     */
    this.className = "Service.RPCMethod";

    /**
     * Request message name.
     * @type ***REMOVED***string***REMOVED***
     * @expose
     */
    this.requestName = request;

    /**
     * Response message name.
     * @type ***REMOVED***string***REMOVED***
     * @expose
     */
    this.responseName = response;

    /**
     * Whether requests are streamed
     * @type ***REMOVED***bool***REMOVED***
     * @expose
     */
    this.requestStream = request_stream;

    /**
     * Whether responses are streamed
     * @type ***REMOVED***bool***REMOVED***
     * @expose
     */
    this.responseStream = response_stream;

    /**
     * Resolved request message type.
     * @type ***REMOVED***ProtoBuf.Reflect.Message***REMOVED***
     * @expose
     */
    this.resolvedRequestType = null;

    /**
     * Resolved response message type.
     * @type ***REMOVED***ProtoBuf.Reflect.Message***REMOVED***
     * @expose
     */
    this.resolvedResponseType = null;
***REMOVED***;

// Extends Method
RPCMethod.prototype = Object.create(Method.prototype);
