// @firebase/database/dist/cjs/src/core/snap/indexes/PriorityIndex.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = Object.setPrototypeOf ||
        (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
        function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Index_1 = require("./Index");
var util_1 = require("../../util/util");
var Node_1 = require("../Node");
var LeafNode_1 = require("../LeafNode");
var nodeFromJSON;
var MAX_NODE;
function setNodeFromJSON(val) ***REMOVED***
    nodeFromJSON = val;
***REMOVED***
exports.setNodeFromJSON = setNodeFromJSON;
function setMaxNode(val) ***REMOVED***
    MAX_NODE = val;
***REMOVED***
exports.setMaxNode = setMaxNode;
/**
 * @constructor
 * @extends ***REMOVED***Index***REMOVED***
 * @private
 */
var PriorityIndex = /** @class */ (function (_super) ***REMOVED***
    __extends(PriorityIndex, _super);
    function PriorityIndex() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * @inheritDoc
     */
    PriorityIndex.prototype.compare = function (a, b) ***REMOVED***
        var aPriority = a.node.getPriority();
        var bPriority = b.node.getPriority();
        var indexCmp = aPriority.compareTo(bPriority);
        if (indexCmp === 0) ***REMOVED***
            return util_1.nameCompare(a.name, b.name);
        ***REMOVED***
        else ***REMOVED***
            return indexCmp;
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PriorityIndex.prototype.isDefinedOn = function (node) ***REMOVED***
        return !node.getPriority().isEmpty();
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PriorityIndex.prototype.indexedValueChanged = function (oldNode, newNode) ***REMOVED***
        return !oldNode.getPriority().equals(newNode.getPriority());
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PriorityIndex.prototype.minPost = function () ***REMOVED***
        return Node_1.NamedNode.MIN;
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    PriorityIndex.prototype.maxPost = function () ***REMOVED***
        return new Node_1.NamedNode(util_1.MAX_NAME, new LeafNode_1.LeafNode('[PRIORITY-POST]', MAX_NODE));
    ***REMOVED***;
    /**
     * @param ***REMOVED*******REMOVED*** indexValue
     * @param ***REMOVED***string***REMOVED*** name
     * @return ***REMOVED***!NamedNode***REMOVED***
     */
    PriorityIndex.prototype.makePost = function (indexValue, name) ***REMOVED***
        var priorityNode = nodeFromJSON(indexValue);
        return new Node_1.NamedNode(name, new LeafNode_1.LeafNode('[PRIORITY-POST]', priorityNode));
    ***REMOVED***;
    /**
     * @return ***REMOVED***!string***REMOVED*** String representation for inclusion in a query spec
     */
    PriorityIndex.prototype.toString = function () ***REMOVED***
        return '.priority';
    ***REMOVED***;
    return PriorityIndex;
***REMOVED***(Index_1.Index));
exports.PriorityIndex = PriorityIndex;
exports.PRIORITY_INDEX = new PriorityIndex();
