// @firebase/database/dist/cjs/src/core/snap/indexes/ValueIndex.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = Object.setPrototypeOf ||
        (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
        function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Index_1 = require("./Index");
var Node_1 = require("../Node");
var util_1 = require("../../util/util");
var nodeFromJSON_1 = require("../nodeFromJSON");
/**
 * @constructor
 * @extends ***REMOVED***Index***REMOVED***
 * @private
 */
var ValueIndex = /** @class */ (function (_super) ***REMOVED***
    __extends(ValueIndex, _super);
    function ValueIndex() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * @inheritDoc
     */
    ValueIndex.prototype.compare = function (a, b) ***REMOVED***
        var indexCmp = a.node.compareTo(b.node);
        if (indexCmp === 0) ***REMOVED***
            return util_1.nameCompare(a.name, b.name);
        ***REMOVED***
        else ***REMOVED***
            return indexCmp;
        ***REMOVED***
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    ValueIndex.prototype.isDefinedOn = function (node) ***REMOVED***
        return true;
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    ValueIndex.prototype.indexedValueChanged = function (oldNode, newNode) ***REMOVED***
        return !oldNode.equals(newNode);
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    ValueIndex.prototype.minPost = function () ***REMOVED***
        return Node_1.NamedNode.MIN;
    ***REMOVED***;
    /**
     * @inheritDoc
     */
    ValueIndex.prototype.maxPost = function () ***REMOVED***
        return Node_1.NamedNode.MAX;
    ***REMOVED***;
    /**
     * @param ***REMOVED*******REMOVED*** indexValue
     * @param ***REMOVED***string***REMOVED*** name
     * @return ***REMOVED***!NamedNode***REMOVED***
     */
    ValueIndex.prototype.makePost = function (indexValue, name) ***REMOVED***
        var valueNode = nodeFromJSON_1.nodeFromJSON(indexValue);
        return new Node_1.NamedNode(name, valueNode);
    ***REMOVED***;
    /**
     * @return ***REMOVED***!string***REMOVED*** String representation for inclusion in a query spec
     */
    ValueIndex.prototype.toString = function () ***REMOVED***
        return '.value';
    ***REMOVED***;
    return ValueIndex;
***REMOVED***(Index_1.Index));
exports.ValueIndex = ValueIndex;
exports.VALUE_INDEX = new ValueIndex();
