// ansi-align/index.js
'use strict'

const stringWidth = require('string-width')

function ansiAlign (text, opts) ***REMOVED***
  if (!text) return text

  opts = opts || ***REMOVED******REMOVED***
  const align = opts.align || 'center'

  // short-circuit `align: 'left'` as no-op
  if (align === 'left') return text

  const split = opts.split || '\n'
  const pad = opts.pad || ' '
  const widthDiffFn = align !== 'right' ? halfDiff : fullDiff

  let returnString = false
  if (!Array.isArray(text)) ***REMOVED***
    returnString = true
    text = String(text).split(split)
  ***REMOVED***

  let width
  let maxWidth = 0
  text = text.map(function (str) ***REMOVED***
    str = String(str)
    width = stringWidth(str)
    maxWidth = Math.max(width, maxWidth)
    return ***REMOVED***
      str,
      width
    ***REMOVED***
  ***REMOVED***).map(function (obj) ***REMOVED***
    return new Array(widthDiffFn(maxWidth, obj.width) + 1).join(pad) + obj.str
  ***REMOVED***)

  return returnString ? text.join(split) : text
***REMOVED***

ansiAlign.left = function left (text) ***REMOVED***
  return ansiAlign(text, ***REMOVED*** align: 'left' ***REMOVED***)
***REMOVED***

ansiAlign.center = function center (text) ***REMOVED***
  return ansiAlign(text, ***REMOVED*** align: 'center' ***REMOVED***)
***REMOVED***

ansiAlign.right = function right (text) ***REMOVED***
  return ansiAlign(text, ***REMOVED*** align: 'right' ***REMOVED***)
***REMOVED***

module.exports = ansiAlign

function halfDiff (maxWidth, curWidth) ***REMOVED***
  return Math.floor((maxWidth - curWidth) / 2)
***REMOVED***

function fullDiff (maxWidth, curWidth) ***REMOVED***
  return maxWidth - curWidth
***REMOVED***
