// string_decoder/lib/string_decoder.js
'use strict';

var Buffer = require('safe-buffer').Buffer;

var isEncoding = Buffer.isEncoding || function (encoding) ***REMOVED***
  encoding = '' + encoding;
  switch (encoding && encoding.toLowerCase()) ***REMOVED***
    case 'hex':case 'utf8':case 'utf-8':case 'ascii':case 'binary':case 'base64':case 'ucs2':case 'ucs-2':case 'utf16le':case 'utf-16le':case 'raw':
      return true;
    default:
      return false;
  ***REMOVED***
***REMOVED***;

function _normalizeEncoding(enc) ***REMOVED***
  if (!enc) return 'utf8';
  var retried;
  while (true) ***REMOVED***
    switch (enc) ***REMOVED***
      case 'utf8':
      case 'utf-8':
        return 'utf8';
      case 'ucs2':
      case 'ucs-2':
      case 'utf16le':
      case 'utf-16le':
        return 'utf16le';
      case 'latin1':
      case 'binary':
        return 'latin1';
      case 'base64':
      case 'ascii':
      case 'hex':
        return enc;
      default:
        if (retried) return; // undefined
        enc = ('' + enc).toLowerCase();
        retried = true;
    ***REMOVED***
  ***REMOVED***
***REMOVED***;

// Do not cache `Buffer.isEncoding` when checking encoding names as some
// modules monkey-patch it to support additional encodings
function normalizeEncoding(enc) ***REMOVED***
  var nenc = _normalizeEncoding(enc);
  if (typeof nenc !== 'string' && (Buffer.isEncoding === isEncoding || !isEncoding(enc))) throw new Error('Unknown encoding: ' + enc);
  return nenc || enc;
***REMOVED***

// StringDecoder provides an interface for efficiently splitting a series of
// buffers into a series of JS strings without breaking apart multi-byte
// characters.
exports.StringDecoder = StringDecoder;
function StringDecoder(encoding) ***REMOVED***
  this.encoding = normalizeEncoding(encoding);
  var nb;
  switch (this.encoding) ***REMOVED***
    case 'utf16le':
      this.text = utf16Text;
      this.end = utf16End;
      nb = 4;
      break;
    case 'utf8':
      this.fillLast = utf8FillLast;
      nb = 4;
      break;
    case 'base64':
      this.text = base64Text;
      this.end = base64End;
      nb = 3;
      break;
    default:
      this.write = simpleWrite;
      this.end = simpleEnd;
      return;
  ***REMOVED***
  this.lastNeed = 0;
  this.lastTotal = 0;
  this.lastChar = Buffer.allocUnsafe(nb);
***REMOVED***

StringDecoder.prototype.write = function (buf) ***REMOVED***
  if (buf.length === 0) return '';
  var r;
  var i;
  if (this.lastNeed) ***REMOVED***
    r = this.fillLast(buf);
    if (r === undefined) return '';
    i = this.lastNeed;
    this.lastNeed = 0;
  ***REMOVED*** else ***REMOVED***
    i = 0;
  ***REMOVED***
  if (i < buf.length) return r ? r + this.text(buf, i) : this.text(buf, i);
  return r || '';
***REMOVED***;

StringDecoder.prototype.end = utf8End;

// Returns only complete characters in a Buffer
StringDecoder.prototype.text = utf8Text;

// Attempts to complete a partial non-UTF-8 character using bytes from a Buffer
StringDecoder.prototype.fillLast = function (buf) ***REMOVED***
  if (this.lastNeed <= buf.length) ***REMOVED***
    buf.copy(this.lastChar, this.lastTotal - this.lastNeed, 0, this.lastNeed);
    return this.lastChar.toString(this.encoding, 0, this.lastTotal);
  ***REMOVED***
  buf.copy(this.lastChar, this.lastTotal - this.lastNeed, 0, buf.length);
  this.lastNeed -= buf.length;
***REMOVED***;

// Checks the type of a UTF-8 byte, whether it's ASCII, a leading byte, or a
// continuation byte.
function utf8CheckByte(byte) ***REMOVED***
  if (byte <= 0x7F) return 0;else if (byte >> 5 === 0x06) return 2;else if (byte >> 4 === 0x0E) return 3;else if (byte >> 3 === 0x1E) return 4;
  return -1;
***REMOVED***

// Checks at most 3 bytes at the end of a Buffer in order to detect an
// incomplete multi-byte UTF-8 character. The total number of bytes (2, 3, or 4)
// needed to complete the UTF-8 character (if applicable) are returned.
function utf8CheckIncomplete(self, buf, i) ***REMOVED***
  var j = buf.length - 1;
  if (j < i) return 0;
  var nb = utf8CheckByte(buf[j]);
  if (nb >= 0) ***REMOVED***
    if (nb > 0) self.lastNeed = nb - 1;
    return nb;
  ***REMOVED***
  if (--j < i) return 0;
  nb = utf8CheckByte(buf[j]);
  if (nb >= 0) ***REMOVED***
    if (nb > 0) self.lastNeed = nb - 2;
    return nb;
  ***REMOVED***
  if (--j < i) return 0;
  nb = utf8CheckByte(buf[j]);
  if (nb >= 0) ***REMOVED***
    if (nb > 0) ***REMOVED***
      if (nb === 2) nb = 0;else self.lastNeed = nb - 3;
    ***REMOVED***
    return nb;
  ***REMOVED***
  return 0;
***REMOVED***

// Validates as many continuation bytes for a multi-byte UTF-8 character as
// needed or are available. If we see a non-continuation byte where we expect
// one, we "replace" the validated continuation bytes we've seen so far with
// UTF-8 replacement characters ('\ufffd'), to match v8's UTF-8 decoding
// behavior. The continuation byte check is included three times in the case
// where all of the continuation bytes for a character exist in the same buffer.
// It is also done this way as a slight performance increase instead of using a
// loop.
function utf8CheckExtraBytes(self, buf, p) ***REMOVED***
  if ((buf[0] & 0xC0) !== 0x80) ***REMOVED***
    self.lastNeed = 0;
    return '\ufffd'.repeat(p);
  ***REMOVED***
  if (self.lastNeed > 1 && buf.length > 1) ***REMOVED***
    if ((buf[1] & 0xC0) !== 0x80) ***REMOVED***
      self.lastNeed = 1;
      return '\ufffd'.repeat(p + 1);
    ***REMOVED***
    if (self.lastNeed > 2 && buf.length > 2) ***REMOVED***
      if ((buf[2] & 0xC0) !== 0x80) ***REMOVED***
        self.lastNeed = 2;
        return '\ufffd'.repeat(p + 2);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

// Attempts to complete a multi-byte UTF-8 character using bytes from a Buffer.
function utf8FillLast(buf) ***REMOVED***
  var p = this.lastTotal - this.lastNeed;
  var r = utf8CheckExtraBytes(this, buf, p);
  if (r !== undefined) return r;
  if (this.lastNeed <= buf.length) ***REMOVED***
    buf.copy(this.lastChar, p, 0, this.lastNeed);
    return this.lastChar.toString(this.encoding, 0, this.lastTotal);
  ***REMOVED***
  buf.copy(this.lastChar, p, 0, buf.length);
  this.lastNeed -= buf.length;
***REMOVED***

// Returns all complete UTF-8 characters in a Buffer. If the Buffer ended on a
// partial character, the character's bytes are buffered until the required
// number of bytes are available.
function utf8Text(buf, i) ***REMOVED***
  var total = utf8CheckIncomplete(this, buf, i);
  if (!this.lastNeed) return buf.toString('utf8', i);
  this.lastTotal = total;
  var end = buf.length - (total - this.lastNeed);
  buf.copy(this.lastChar, 0, end);
  return buf.toString('utf8', i, end);
***REMOVED***

// For UTF-8, a replacement character for each buffered byte of a (partial)
// character needs to be added to the output.
function utf8End(buf) ***REMOVED***
  var r = buf && buf.length ? this.write(buf) : '';
  if (this.lastNeed) return r + '\ufffd'.repeat(this.lastTotal - this.lastNeed);
  return r;
***REMOVED***

// UTF-16LE typically needs two bytes per character, but even if we have an even
// number of bytes available, we need to check if we end on a leading/high
// surrogate. In that case, we need to wait for the next two bytes in order to
// decode the last character properly.
function utf16Text(buf, i) ***REMOVED***
  if ((buf.length - i) % 2 === 0) ***REMOVED***
    var r = buf.toString('utf16le', i);
    if (r) ***REMOVED***
      var c = r.charCodeAt(r.length - 1);
      if (c >= 0xD800 && c <= 0xDBFF) ***REMOVED***
        this.lastNeed = 2;
        this.lastTotal = 4;
        this.lastChar[0] = buf[buf.length - 2];
        this.lastChar[1] = buf[buf.length - 1];
        return r.slice(0, -1);
      ***REMOVED***
    ***REMOVED***
    return r;
  ***REMOVED***
  this.lastNeed = 1;
  this.lastTotal = 2;
  this.lastChar[0] = buf[buf.length - 1];
  return buf.toString('utf16le', i, buf.length - 1);
***REMOVED***

// For UTF-16LE we do not explicitly append special replacement characters if we
// end on a partial character, we simply let v8 handle that.
function utf16End(buf) ***REMOVED***
  var r = buf && buf.length ? this.write(buf) : '';
  if (this.lastNeed) ***REMOVED***
    var end = this.lastTotal - this.lastNeed;
    return r + this.lastChar.toString('utf16le', 0, end);
  ***REMOVED***
  return r;
***REMOVED***

function base64Text(buf, i) ***REMOVED***
  var n = (buf.length - i) % 3;
  if (n === 0) return buf.toString('base64', i);
  this.lastNeed = 3 - n;
  this.lastTotal = 3;
  if (n === 1) ***REMOVED***
    this.lastChar[0] = buf[buf.length - 1];
  ***REMOVED*** else ***REMOVED***
    this.lastChar[0] = buf[buf.length - 2];
    this.lastChar[1] = buf[buf.length - 1];
  ***REMOVED***
  return buf.toString('base64', i, buf.length - n);
***REMOVED***

function base64End(buf) ***REMOVED***
  var r = buf && buf.length ? this.write(buf) : '';
  if (this.lastNeed) return r + this.lastChar.toString('base64', 0, 3 - this.lastNeed);
  return r;
***REMOVED***

// Pass bytes on through for single-byte encodings (e.g. ascii, latin1, hex)
function simpleWrite(buf) ***REMOVED***
  return buf.toString(this.encoding);
***REMOVED***

function simpleEnd(buf) ***REMOVED***
  return buf && buf.length ? this.write(buf) : '';
***REMOVED***
