// stringifier/build/stringifier.js
/**
 * Modules in this bundle
 * @license
 * 
 * stringifier:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   author: Takuto Wada <takuto.wada@gmail.com>
 *   homepage: https://github.com/twada/stringifier
 *   version: 1.3.0
 * 
 * core-js:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   maintainers: zloirock <zloirock@zloirock.ru>
 *   homepage: https://github.com/zloirock/core-js#readme
 *   version: 2.4.0
 * 
 * traverse:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   author: James Halliday <mail@substack.net>
 *   maintainers: substack <mail@substack.net>
 *   homepage: https://github.com/substack/js-traverse
 *   version: 0.6.6
 * 
 * type-name:
 *   license: MIT (http://opensource.org/licenses/MIT)
 *   author: Takuto Wada <takuto.wada@gmail.com>
 *   maintainers: twada <takuto.wada@gmail.com>
 *   contributors: azu, Yosuke Furukawa
 *   homepage: https://github.com/twada/type-name
 *   version: 2.0.1
 * 
 * This header is generated by licensify (https://github.com/twada/licensify)
 */
(function(f)***REMOVED***if(typeof exports==="object"&&typeof module!=="undefined")***REMOVED***module.exports=f()***REMOVED***else if(typeof define==="function"&&define.amd)***REMOVED***define([],f)***REMOVED***else***REMOVED***var g;if(typeof window!=="undefined")***REMOVED***g=window***REMOVED***else if(typeof global!=="undefined")***REMOVED***g=global***REMOVED***else if(typeof self!=="undefined")***REMOVED***g=self***REMOVED***else***REMOVED***g=this***REMOVED***g.stringifier = f()***REMOVED******REMOVED***)(function()***REMOVED***var define,module,exports;return (function e(t,n,r)***REMOVED***function s(o,u)***REMOVED***if(!n[o])***REMOVED***if(!t[o])***REMOVED***var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw (f.code="MODULE_NOT_FOUND", f)***REMOVED***var l=n[o]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;t[o][0].call(l.exports,function(e)***REMOVED***var n=t[o][1][e];return s(n?n:e)***REMOVED***,l,l.exports,e,t,n,r)***REMOVED***return n[o].exports***REMOVED***var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s***REMOVED***)(***REMOVED***1:[function(_dereq_,module,exports)***REMOVED***
/**
 * stringifier
 * 
 * https://github.com/twada/stringifier
 *
 * Copyright (c) 2014-2015 Takuto Wada
 * Licensed under the MIT license.
 *   http://twada.mit-license.org/2014-2015
 */
'use strict';

var traverse = _dereq_('traverse');
var typeName = _dereq_('type-name');
var assign = _dereq_('core-js/library/fn/object/assign');
var s = _dereq_('./strategies');

function defaultHandlers () ***REMOVED***
    return ***REMOVED***
        'null': s.always('null'),
        'undefined': s.always('undefined'),
        'function': s.prune(),
        'string': s.json(),
        'boolean': s.json(),
        'number': s.number(),
        'symbol': s.toStr(),
        'RegExp': s.toStr(),
        'String': s.newLike(),
        'Boolean': s.newLike(),
        'Number': s.newLike(),
        'Date': s.newLike(),
        'Array': s.array(),
        'Object': s.object(),
        '@default': s.object()
    ***REMOVED***;
***REMOVED***

function defaultOptions () ***REMOVED***
    return ***REMOVED***
        maxDepth: null,
        indent: null,
        anonymous: '@Anonymous',
        circular: '#@Circular#',
        snip: '..(snip)',
        lineSeparator: '\n',
        typeFun: typeName
    ***REMOVED***;
***REMOVED***

function createStringifier (customOptions) ***REMOVED***
    var options = assign(***REMOVED******REMOVED***, defaultOptions(), customOptions);
    var handlers = assign(***REMOVED******REMOVED***, defaultHandlers(), options.handlers);
    return function stringifyAny (push, x) ***REMOVED***
        var context = this;
        var handler = handlerFor(context.node, options, handlers);
        var currentPath = '/' + context.path.join('/');
        var customization = handlers[currentPath];
        var acc = ***REMOVED***
            context: context,
            options: options,
            handlers: handlers,
            push: push
        ***REMOVED***;
        if (typeName(customization) === 'function') ***REMOVED***
            handler = customization;
        ***REMOVED*** else if (typeName(customization) === 'number') ***REMOVED***
            handler = s.flow.compose(s.filters.truncate(customization),handler);
        ***REMOVED*** else if (context.parent && typeName(context.parent.node) === 'Array' && !(context.key in context.parent.node)) ***REMOVED***
            // sparse arrays
            handler = s.always('');
        ***REMOVED***
        handler(acc, x);
        return push;
    ***REMOVED***;
***REMOVED***

function handlerFor (val, options, handlers) ***REMOVED***
    var tname = options.typeFun(val);
    if (typeName(handlers[tname]) === 'function') ***REMOVED***
        return handlers[tname];
    ***REMOVED***
    return handlers['@default'];
***REMOVED***

function walk (val, reducer) ***REMOVED***
    var buffer = [];
    var push = function (str) ***REMOVED***
        buffer.push(str);
    ***REMOVED***;
    traverse(val).reduce(reducer, push);
    return buffer.join('');
***REMOVED***

function stringify (val, options) ***REMOVED***
    return walk(val, createStringifier(options));
***REMOVED***

function stringifier (options) ***REMOVED***
    return function (val) ***REMOVED***
        return walk(val, createStringifier(options));
    ***REMOVED***;
***REMOVED***

stringifier.stringify = stringify;
stringifier.strategies = s;
stringifier.defaultOptions = defaultOptions;
stringifier.defaultHandlers = defaultHandlers;
module.exports = stringifier;

***REMOVED***,***REMOVED***"./strategies":55,"core-js/library/fn/object/assign":6,"traverse":53,"type-name":54***REMOVED***],2:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.filter');
module.exports = _dereq_('../../modules/_core').Array.filter;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.filter":48***REMOVED***],3:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.for-each');
module.exports = _dereq_('../../modules/_core').Array.forEach;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.for-each":49***REMOVED***],4:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.index-of');
module.exports = _dereq_('../../modules/_core').Array.indexOf;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.index-of":50***REMOVED***],5:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.array.reduce-right');
module.exports = _dereq_('../../modules/_core').Array.reduceRight;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.array.reduce-right":51***REMOVED***],6:[function(_dereq_,module,exports)***REMOVED***
_dereq_('../../modules/es6.object.assign');
module.exports = _dereq_('../../modules/_core').Object.assign;
***REMOVED***,***REMOVED***"../../modules/_core":15,"../../modules/es6.object.assign":52***REMOVED***],7:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(it)***REMOVED***
  if(typeof it != 'function')throw TypeError(it + ' is not a function!');
  return it;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],8:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object');
module.exports = function(it)***REMOVED***
  if(!isObject(it))throw TypeError(it + ' is not an object!');
  return it;
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-object":29***REMOVED***],9:[function(_dereq_,module,exports)***REMOVED***
// false -> Array#indexOf
// true  -> Array#includes
var toIObject = _dereq_('./_to-iobject')
  , toLength  = _dereq_('./_to-length')
  , toIndex   = _dereq_('./_to-index');
module.exports = function(IS_INCLUDES)***REMOVED***
  return function($this, el, fromIndex)***REMOVED***
    var O      = toIObject($this)
      , length = toLength(O.length)
      , index  = toIndex(fromIndex, length)
      , value;
    // Array#includes uses SameValueZero equality algorithm
    if(IS_INCLUDES && el != el)while(length > index)***REMOVED***
      value = O[index++];
      if(value != value)return true;
    // Array#toIndex ignores holes, Array#includes - not
    ***REMOVED*** else for(;length > index; index++)if(IS_INCLUDES || index in O)***REMOVED***
      if(O[index] === el)return IS_INCLUDES || index || 0;
    ***REMOVED*** return !IS_INCLUDES && -1;
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-index":40,"./_to-iobject":42,"./_to-length":43***REMOVED***],10:[function(_dereq_,module,exports)***REMOVED***
// 0 -> Array#forEach
// 1 -> Array#map
// 2 -> Array#filter
// 3 -> Array#some
// 4 -> Array#every
// 5 -> Array#find
// 6 -> Array#findIndex
var ctx      = _dereq_('./_ctx')
  , IObject  = _dereq_('./_iobject')
  , toObject = _dereq_('./_to-object')
  , toLength = _dereq_('./_to-length')
  , asc      = _dereq_('./_array-species-create');
module.exports = function(TYPE, $create)***REMOVED***
  var IS_MAP        = TYPE == 1
    , IS_FILTER     = TYPE == 2
    , IS_SOME       = TYPE == 3
    , IS_EVERY      = TYPE == 4
    , IS_FIND_INDEX = TYPE == 6
    , NO_HOLES      = TYPE == 5 || IS_FIND_INDEX
    , create        = $create || asc;
  return function($this, callbackfn, that)***REMOVED***
    var O      = toObject($this)
      , self   = IObject(O)
      , f      = ctx(callbackfn, that, 3)
      , length = toLength(self.length)
      , index  = 0
      , result = IS_MAP ? create($this, length) : IS_FILTER ? create($this, 0) : undefined
      , val, res;
    for(;length > index; index++)if(NO_HOLES || index in self)***REMOVED***
      val = self[index];
      res = f(val, index, O);
      if(TYPE)***REMOVED***
        if(IS_MAP)result[index] = res;            // map
        else if(res)switch(TYPE)***REMOVED***
          case 3: return true;                    // some
          case 5: return val;                     // find
          case 6: return index;                   // findIndex
          case 2: result.push(val);               // filter
        ***REMOVED*** else if(IS_EVERY)return false;          // every
      ***REMOVED***
    ***REMOVED***
    return IS_FIND_INDEX ? -1 : IS_SOME || IS_EVERY ? IS_EVERY : result;
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-species-create":13,"./_ctx":16,"./_iobject":27,"./_to-length":43,"./_to-object":44***REMOVED***],11:[function(_dereq_,module,exports)***REMOVED***
var aFunction = _dereq_('./_a-function')
  , toObject  = _dereq_('./_to-object')
  , IObject   = _dereq_('./_iobject')
  , toLength  = _dereq_('./_to-length');

module.exports = function(that, callbackfn, aLen, memo, isRight)***REMOVED***
  aFunction(callbackfn);
  var O      = toObject(that)
    , self   = IObject(O)
    , length = toLength(O.length)
    , index  = isRight ? length - 1 : 0
    , i      = isRight ? -1 : 1;
  if(aLen < 2)for(;;)***REMOVED***
    if(index in self)***REMOVED***
      memo = self[index];
      index += i;
      break;
    ***REMOVED***
    index += i;
    if(isRight ? index < 0 : length <= index)***REMOVED***
      throw TypeError('Reduce of empty array with no initial value');
    ***REMOVED***
  ***REMOVED***
  for(;isRight ? index >= 0 : length > index; index += i)if(index in self)***REMOVED***
    memo = callbackfn(memo, self[index], index, O);
  ***REMOVED***
  return memo;
***REMOVED***;
***REMOVED***,***REMOVED***"./_a-function":7,"./_iobject":27,"./_to-length":43,"./_to-object":44***REMOVED***],12:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object')
  , isArray  = _dereq_('./_is-array')
  , SPECIES  = _dereq_('./_wks')('species');

module.exports = function(original)***REMOVED***
  var C;
  if(isArray(original))***REMOVED***
    C = original.constructor;
    // cross-realm fallback
    if(typeof C == 'function' && (C === Array || isArray(C.prototype)))C = undefined;
    if(isObject(C))***REMOVED***
      C = C[SPECIES];
      if(C === null)C = undefined;
    ***REMOVED***
  ***REMOVED*** return C === undefined ? Array : C;
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-array":28,"./_is-object":29,"./_wks":47***REMOVED***],13:[function(_dereq_,module,exports)***REMOVED***
// 9.4.2.3 ArraySpeciesCreate(originalArray, length)
var speciesConstructor = _dereq_('./_array-species-constructor');

module.exports = function(original, length)***REMOVED***
  return new (speciesConstructor(original))(length);
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-species-constructor":12***REMOVED***],14:[function(_dereq_,module,exports)***REMOVED***
var toString = ***REMOVED******REMOVED***.toString;

module.exports = function(it)***REMOVED***
  return toString.call(it).slice(8, -1);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],15:[function(_dereq_,module,exports)***REMOVED***
var core = module.exports = ***REMOVED***version: '2.4.0'***REMOVED***;
if(typeof __e == 'number')__e = core; // eslint-disable-line no-undef
***REMOVED***,***REMOVED******REMOVED***],16:[function(_dereq_,module,exports)***REMOVED***
// optional / simple context binding
var aFunction = _dereq_('./_a-function');
module.exports = function(fn, that, length)***REMOVED***
  aFunction(fn);
  if(that === undefined)return fn;
  switch(length)***REMOVED***
    case 1: return function(a)***REMOVED***
      return fn.call(that, a);
    ***REMOVED***;
    case 2: return function(a, b)***REMOVED***
      return fn.call(that, a, b);
    ***REMOVED***;
    case 3: return function(a, b, c)***REMOVED***
      return fn.call(that, a, b, c);
    ***REMOVED***;
  ***REMOVED***
  return function(/* ...args */)***REMOVED***
    return fn.apply(that, arguments);
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_a-function":7***REMOVED***],17:[function(_dereq_,module,exports)***REMOVED***
// 7.2.1 RequireObjectCoercible(argument)
module.exports = function(it)***REMOVED***
  if(it == undefined)throw TypeError("Can't call method on  " + it);
  return it;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],18:[function(_dereq_,module,exports)***REMOVED***
// Thank's IE8 for his funny defineProperty
module.exports = !_dereq_('./_fails')(function()***REMOVED***
  return Object.defineProperty(***REMOVED******REMOVED***, 'a', ***REMOVED***get: function()***REMOVED*** return 7; ***REMOVED******REMOVED***).a != 7;
***REMOVED***);
***REMOVED***,***REMOVED***"./_fails":22***REMOVED***],19:[function(_dereq_,module,exports)***REMOVED***
var isObject = _dereq_('./_is-object')
  , document = _dereq_('./_global').document
  // in old IE typeof document.createElement is 'object'
  , is = isObject(document) && isObject(document.createElement);
module.exports = function(it)***REMOVED***
  return is ? document.createElement(it) : ***REMOVED******REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED***"./_global":23,"./_is-object":29***REMOVED***],20:[function(_dereq_,module,exports)***REMOVED***
// IE 8- don't enum bug keys
module.exports = (
  'constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf'
).split(',');
***REMOVED***,***REMOVED******REMOVED***],21:[function(_dereq_,module,exports)***REMOVED***
var global    = _dereq_('./_global')
  , core      = _dereq_('./_core')
  , ctx       = _dereq_('./_ctx')
  , hide      = _dereq_('./_hide')
  , PROTOTYPE = 'prototype';

var $export = function(type, name, source)***REMOVED***
  var IS_FORCED = type & $export.F
    , IS_GLOBAL = type & $export.G
    , IS_STATIC = type & $export.S
    , IS_PROTO  = type & $export.P
    , IS_BIND   = type & $export.B
    , IS_WRAP   = type & $export.W
    , exports   = IS_GLOBAL ? core : core[name] || (core[name] = ***REMOVED******REMOVED***)
    , expProto  = exports[PROTOTYPE]
    , target    = IS_GLOBAL ? global : IS_STATIC ? global[name] : (global[name] || ***REMOVED******REMOVED***)[PROTOTYPE]
    , key, own, out;
  if(IS_GLOBAL)source = name;
  for(key in source)***REMOVED***
    // contains in native
    own = !IS_FORCED && target && target[key] !== undefined;
    if(own && key in exports)continue;
    // export native or passed
    out = own ? target[key] : source[key];
    // prevent global pollution for namespaces
    exports[key] = IS_GLOBAL && typeof target[key] != 'function' ? source[key]
    // bind timers to global for call from export context
    : IS_BIND && own ? ctx(out, global)
    // wrap global constructors for prevent change them in library
    : IS_WRAP && target[key] == out ? (function(C)***REMOVED***
      var F = function(a, b, c)***REMOVED***
        if(this instanceof C)***REMOVED***
          switch(arguments.length)***REMOVED***
            case 0: return new C;
            case 1: return new C(a);
            case 2: return new C(a, b);
          ***REMOVED*** return new C(a, b, c);
        ***REMOVED*** return C.apply(this, arguments);
      ***REMOVED***;
      F[PROTOTYPE] = C[PROTOTYPE];
      return F;
    // make static versions for prototype methods
    ***REMOVED***)(out) : IS_PROTO && typeof out == 'function' ? ctx(Function.call, out) : out;
    // export proto methods to core.%CONSTRUCTOR%.methods.%NAME%
    if(IS_PROTO)***REMOVED***
      (exports.virtual || (exports.virtual = ***REMOVED******REMOVED***))[key] = out;
      // export proto methods to core.%CONSTRUCTOR%.prototype.%NAME%
      if(type & $export.R && expProto && !expProto[key])hide(expProto, key, out);
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
// type bitmap
$export.F = 1;   // forced
$export.G = 2;   // global
$export.S = 4;   // static
$export.P = 8;   // proto
$export.B = 16;  // bind
$export.W = 32;  // wrap
$export.U = 64;  // safe
$export.R = 128; // real proto method for `library` 
module.exports = $export;
***REMOVED***,***REMOVED***"./_core":15,"./_ctx":16,"./_global":23,"./_hide":25***REMOVED***],22:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(exec)***REMOVED***
  try ***REMOVED***
    return !!exec();
  ***REMOVED*** catch(e)***REMOVED***
    return true;
  ***REMOVED***
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],23:[function(_dereq_,module,exports)***REMOVED***
// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
var global = module.exports = typeof window != 'undefined' && window.Math == Math
  ? window : typeof self != 'undefined' && self.Math == Math ? self : Function('return this')();
if(typeof __g == 'number')__g = global; // eslint-disable-line no-undef
***REMOVED***,***REMOVED******REMOVED***],24:[function(_dereq_,module,exports)***REMOVED***
var hasOwnProperty = ***REMOVED******REMOVED***.hasOwnProperty;
module.exports = function(it, key)***REMOVED***
  return hasOwnProperty.call(it, key);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],25:[function(_dereq_,module,exports)***REMOVED***
var dP         = _dereq_('./_object-dp')
  , createDesc = _dereq_('./_property-desc');
module.exports = _dereq_('./_descriptors') ? function(object, key, value)***REMOVED***
  return dP.f(object, key, createDesc(1, value));
***REMOVED*** : function(object, key, value)***REMOVED***
  object[key] = value;
  return object;
***REMOVED***;
***REMOVED***,***REMOVED***"./_descriptors":18,"./_object-dp":31,"./_property-desc":36***REMOVED***],26:[function(_dereq_,module,exports)***REMOVED***
module.exports = !_dereq_('./_descriptors') && !_dereq_('./_fails')(function()***REMOVED***
  return Object.defineProperty(_dereq_('./_dom-create')('div'), 'a', ***REMOVED***get: function()***REMOVED*** return 7; ***REMOVED******REMOVED***).a != 7;
***REMOVED***);
***REMOVED***,***REMOVED***"./_descriptors":18,"./_dom-create":19,"./_fails":22***REMOVED***],27:[function(_dereq_,module,exports)***REMOVED***
// fallback for non-array-like ES3 and non-enumerable old V8 strings
var cof = _dereq_('./_cof');
module.exports = Object('z').propertyIsEnumerable(0) ? Object : function(it)***REMOVED***
  return cof(it) == 'String' ? it.split('') : Object(it);
***REMOVED***;
***REMOVED***,***REMOVED***"./_cof":14***REMOVED***],28:[function(_dereq_,module,exports)***REMOVED***
// 7.2.2 IsArray(argument)
var cof = _dereq_('./_cof');
module.exports = Array.isArray || function isArray(arg)***REMOVED***
  return cof(arg) == 'Array';
***REMOVED***;
***REMOVED***,***REMOVED***"./_cof":14***REMOVED***],29:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(it)***REMOVED***
  return typeof it === 'object' ? it !== null : typeof it === 'function';
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],30:[function(_dereq_,module,exports)***REMOVED***
'use strict';
// 19.1.2.1 Object.assign(target, source, ...)
var getKeys  = _dereq_('./_object-keys')
  , gOPS     = _dereq_('./_object-gops')
  , pIE      = _dereq_('./_object-pie')
  , toObject = _dereq_('./_to-object')
  , IObject  = _dereq_('./_iobject')
  , $assign  = Object.assign;

// should work with symbols and should have deterministic property order (V8 bug)
module.exports = !$assign || _dereq_('./_fails')(function()***REMOVED***
  var A = ***REMOVED******REMOVED***
    , B = ***REMOVED******REMOVED***
    , S = Symbol()
    , K = 'abcdefghijklmnopqrst';
  A[S] = 7;
  K.split('').forEach(function(k)***REMOVED*** B[k] = k; ***REMOVED***);
  return $assign(***REMOVED******REMOVED***, A)[S] != 7 || Object.keys($assign(***REMOVED******REMOVED***, B)).join('') != K;
***REMOVED***) ? function assign(target, source)***REMOVED*** // eslint-disable-line no-unused-vars
  var T     = toObject(target)
    , aLen  = arguments.length
    , index = 1
    , getSymbols = gOPS.f
    , isEnum     = pIE.f;
  while(aLen > index)***REMOVED***
    var S      = IObject(arguments[index++])
      , keys   = getSymbols ? getKeys(S).concat(getSymbols(S)) : getKeys(S)
      , length = keys.length
      , j      = 0
      , key;
    while(length > j)if(isEnum.call(S, key = keys[j++]))T[key] = S[key];
  ***REMOVED*** return T;
***REMOVED*** : $assign;
***REMOVED***,***REMOVED***"./_fails":22,"./_iobject":27,"./_object-gops":32,"./_object-keys":34,"./_object-pie":35,"./_to-object":44***REMOVED***],31:[function(_dereq_,module,exports)***REMOVED***
var anObject       = _dereq_('./_an-object')
  , IE8_DOM_DEFINE = _dereq_('./_ie8-dom-define')
  , toPrimitive    = _dereq_('./_to-primitive')
  , dP             = Object.defineProperty;

exports.f = _dereq_('./_descriptors') ? Object.defineProperty : function defineProperty(O, P, Attributes)***REMOVED***
  anObject(O);
  P = toPrimitive(P, true);
  anObject(Attributes);
  if(IE8_DOM_DEFINE)try ***REMOVED***
    return dP(O, P, Attributes);
  ***REMOVED*** catch(e)***REMOVED*** /* empty */ ***REMOVED***
  if('get' in Attributes || 'set' in Attributes)throw TypeError('Accessors not supported!');
  if('value' in Attributes)O[P] = Attributes.value;
  return O;
***REMOVED***;
***REMOVED***,***REMOVED***"./_an-object":8,"./_descriptors":18,"./_ie8-dom-define":26,"./_to-primitive":45***REMOVED***],32:[function(_dereq_,module,exports)***REMOVED***
exports.f = Object.getOwnPropertySymbols;
***REMOVED***,***REMOVED******REMOVED***],33:[function(_dereq_,module,exports)***REMOVED***
var has          = _dereq_('./_has')
  , toIObject    = _dereq_('./_to-iobject')
  , arrayIndexOf = _dereq_('./_array-includes')(false)
  , IE_PROTO     = _dereq_('./_shared-key')('IE_PROTO');

module.exports = function(object, names)***REMOVED***
  var O      = toIObject(object)
    , i      = 0
    , result = []
    , key;
  for(key in O)if(key != IE_PROTO)has(O, key) && result.push(key);
  // Don't enum bug & hidden keys
  while(names.length > i)if(has(O, key = names[i++]))***REMOVED***
    ~arrayIndexOf(result, key) || result.push(key);
  ***REMOVED***
  return result;
***REMOVED***;
***REMOVED***,***REMOVED***"./_array-includes":9,"./_has":24,"./_shared-key":37,"./_to-iobject":42***REMOVED***],34:[function(_dereq_,module,exports)***REMOVED***
// 19.1.2.14 / 15.2.3.14 Object.keys(O)
var $keys       = _dereq_('./_object-keys-internal')
  , enumBugKeys = _dereq_('./_enum-bug-keys');

module.exports = Object.keys || function keys(O)***REMOVED***
  return $keys(O, enumBugKeys);
***REMOVED***;
***REMOVED***,***REMOVED***"./_enum-bug-keys":20,"./_object-keys-internal":33***REMOVED***],35:[function(_dereq_,module,exports)***REMOVED***
exports.f = ***REMOVED******REMOVED***.propertyIsEnumerable;
***REMOVED***,***REMOVED******REMOVED***],36:[function(_dereq_,module,exports)***REMOVED***
module.exports = function(bitmap, value)***REMOVED***
  return ***REMOVED***
    enumerable  : !(bitmap & 1),
    configurable: !(bitmap & 2),
    writable    : !(bitmap & 4),
    value       : value
  ***REMOVED***;
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],37:[function(_dereq_,module,exports)***REMOVED***
var shared = _dereq_('./_shared')('keys')
  , uid    = _dereq_('./_uid');
module.exports = function(key)***REMOVED***
  return shared[key] || (shared[key] = uid(key));
***REMOVED***;
***REMOVED***,***REMOVED***"./_shared":38,"./_uid":46***REMOVED***],38:[function(_dereq_,module,exports)***REMOVED***
var global = _dereq_('./_global')
  , SHARED = '__core-js_shared__'
  , store  = global[SHARED] || (global[SHARED] = ***REMOVED******REMOVED***);
module.exports = function(key)***REMOVED***
  return store[key] || (store[key] = ***REMOVED******REMOVED***);
***REMOVED***;
***REMOVED***,***REMOVED***"./_global":23***REMOVED***],39:[function(_dereq_,module,exports)***REMOVED***
var fails = _dereq_('./_fails');

module.exports = function(method, arg)***REMOVED***
  return !!method && fails(function()***REMOVED***
    arg ? method.call(null, function()***REMOVED******REMOVED***, 1) : method.call(null);
  ***REMOVED***);
***REMOVED***;
***REMOVED***,***REMOVED***"./_fails":22***REMOVED***],40:[function(_dereq_,module,exports)***REMOVED***
var toInteger = _dereq_('./_to-integer')
  , max       = Math.max
  , min       = Math.min;
module.exports = function(index, length)***REMOVED***
  index = toInteger(index);
  return index < 0 ? max(index + length, 0) : min(index, length);
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-integer":41***REMOVED***],41:[function(_dereq_,module,exports)***REMOVED***
// 7.1.4 ToInteger
var ceil  = Math.ceil
  , floor = Math.floor;
module.exports = function(it)***REMOVED***
  return isNaN(it = +it) ? 0 : (it > 0 ? floor : ceil)(it);
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],42:[function(_dereq_,module,exports)***REMOVED***
// to indexed object, toObject with fallback for non-array-like ES3 strings
var IObject = _dereq_('./_iobject')
  , defined = _dereq_('./_defined');
module.exports = function(it)***REMOVED***
  return IObject(defined(it));
***REMOVED***;
***REMOVED***,***REMOVED***"./_defined":17,"./_iobject":27***REMOVED***],43:[function(_dereq_,module,exports)***REMOVED***
// 7.1.15 ToLength
var toInteger = _dereq_('./_to-integer')
  , min       = Math.min;
module.exports = function(it)***REMOVED***
  return it > 0 ? min(toInteger(it), 0x1fffffffffffff) : 0; // pow(2, 53) - 1 == 9007199254740991
***REMOVED***;
***REMOVED***,***REMOVED***"./_to-integer":41***REMOVED***],44:[function(_dereq_,module,exports)***REMOVED***
// 7.1.13 ToObject(argument)
var defined = _dereq_('./_defined');
module.exports = function(it)***REMOVED***
  return Object(defined(it));
***REMOVED***;
***REMOVED***,***REMOVED***"./_defined":17***REMOVED***],45:[function(_dereq_,module,exports)***REMOVED***
// 7.1.1 ToPrimitive(input [, PreferredType])
var isObject = _dereq_('./_is-object');
// instead of the ES6 spec version, we didn't implement @@toPrimitive case
// and the second argument - flag - preferred type is a string
module.exports = function(it, S)***REMOVED***
  if(!isObject(it))return it;
  var fn, val;
  if(S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it)))return val;
  if(typeof (fn = it.valueOf) == 'function' && !isObject(val = fn.call(it)))return val;
  if(!S && typeof (fn = it.toString) == 'function' && !isObject(val = fn.call(it)))return val;
  throw TypeError("Can't convert object to primitive value");
***REMOVED***;
***REMOVED***,***REMOVED***"./_is-object":29***REMOVED***],46:[function(_dereq_,module,exports)***REMOVED***
var id = 0
  , px = Math.random();
module.exports = function(key)***REMOVED***
  return 'Symbol('.concat(key === undefined ? '' : key, ')_', (++id + px).toString(36));
***REMOVED***;
***REMOVED***,***REMOVED******REMOVED***],47:[function(_dereq_,module,exports)***REMOVED***
var store      = _dereq_('./_shared')('wks')
  , uid        = _dereq_('./_uid')
  , Symbol     = _dereq_('./_global').Symbol
  , USE_SYMBOL = typeof Symbol == 'function';

var $exports = module.exports = function(name)***REMOVED***
  return store[name] || (store[name] =
    USE_SYMBOL && Symbol[name] || (USE_SYMBOL ? Symbol : uid)('Symbol.' + name));
***REMOVED***;

$exports.store = store;
***REMOVED***,***REMOVED***"./_global":23,"./_shared":38,"./_uid":46***REMOVED***],48:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export = _dereq_('./_export')
  , $filter = _dereq_('./_array-methods')(2);

$export($export.P + $export.F * !_dereq_('./_strict-method')([].filter, true), 'Array', ***REMOVED***
  // 22.1.3.7 / 15.4.4.20 Array.prototype.filter(callbackfn [, thisArg])
  filter: function filter(callbackfn /* , thisArg */)***REMOVED***
    return $filter(this, callbackfn, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-methods":10,"./_export":21,"./_strict-method":39***REMOVED***],49:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export  = _dereq_('./_export')
  , $forEach = _dereq_('./_array-methods')(0)
  , STRICT   = _dereq_('./_strict-method')([].forEach, true);

$export($export.P + $export.F * !STRICT, 'Array', ***REMOVED***
  // 22.1.3.10 / 15.4.4.18 Array.prototype.forEach(callbackfn [, thisArg])
  forEach: function forEach(callbackfn /* , thisArg */)***REMOVED***
    return $forEach(this, callbackfn, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-methods":10,"./_export":21,"./_strict-method":39***REMOVED***],50:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export       = _dereq_('./_export')
  , $indexOf      = _dereq_('./_array-includes')(false)
  , $native       = [].indexOf
  , NEGATIVE_ZERO = !!$native && 1 / [1].indexOf(1, -0) < 0;

$export($export.P + $export.F * (NEGATIVE_ZERO || !_dereq_('./_strict-method')($native)), 'Array', ***REMOVED***
  // 22.1.3.11 / 15.4.4.14 Array.prototype.indexOf(searchElement [, fromIndex])
  indexOf: function indexOf(searchElement /*, fromIndex = 0 */)***REMOVED***
    return NEGATIVE_ZERO
      // convert -0 to +0
      ? $native.apply(this, arguments) || 0
      : $indexOf(this, searchElement, arguments[1]);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-includes":9,"./_export":21,"./_strict-method":39***REMOVED***],51:[function(_dereq_,module,exports)***REMOVED***
'use strict';
var $export = _dereq_('./_export')
  , $reduce = _dereq_('./_array-reduce');

$export($export.P + $export.F * !_dereq_('./_strict-method')([].reduceRight, true), 'Array', ***REMOVED***
  // 22.1.3.19 / 15.4.4.22 Array.prototype.reduceRight(callbackfn [, initialValue])
  reduceRight: function reduceRight(callbackfn /* , initialValue */)***REMOVED***
    return $reduce(this, callbackfn, arguments.length, arguments[1], true);
  ***REMOVED***
***REMOVED***);
***REMOVED***,***REMOVED***"./_array-reduce":11,"./_export":21,"./_strict-method":39***REMOVED***],52:[function(_dereq_,module,exports)***REMOVED***
// 19.1.3.1 Object.assign(target, source)
var $export = _dereq_('./_export');

$export($export.S + $export.F, 'Object', ***REMOVED***assign: _dereq_('./_object-assign')***REMOVED***);
***REMOVED***,***REMOVED***"./_export":21,"./_object-assign":30***REMOVED***],53:[function(_dereq_,module,exports)***REMOVED***
var traverse = module.exports = function (obj) ***REMOVED***
    return new Traverse(obj);
***REMOVED***;

function Traverse (obj) ***REMOVED***
    this.value = obj;
***REMOVED***

Traverse.prototype.get = function (ps) ***REMOVED***
    var node = this.value;
    for (var i = 0; i < ps.length; i ++) ***REMOVED***
        var key = ps[i];
        if (!node || !hasOwnProperty.call(node, key)) ***REMOVED***
            node = undefined;
            break;
        ***REMOVED***
        node = node[key];
    ***REMOVED***
    return node;
***REMOVED***;

Traverse.prototype.has = function (ps) ***REMOVED***
    var node = this.value;
    for (var i = 0; i < ps.length; i ++) ***REMOVED***
        var key = ps[i];
        if (!node || !hasOwnProperty.call(node, key)) ***REMOVED***
            return false;
        ***REMOVED***
        node = node[key];
    ***REMOVED***
    return true;
***REMOVED***;

Traverse.prototype.set = function (ps, value) ***REMOVED***
    var node = this.value;
    for (var i = 0; i < ps.length - 1; i ++) ***REMOVED***
        var key = ps[i];
        if (!hasOwnProperty.call(node, key)) node[key] = ***REMOVED******REMOVED***;
        node = node[key];
    ***REMOVED***
    node[ps[i]] = value;
    return value;
***REMOVED***;

Traverse.prototype.map = function (cb) ***REMOVED***
    return walk(this.value, cb, true);
***REMOVED***;

Traverse.prototype.forEach = function (cb) ***REMOVED***
    this.value = walk(this.value, cb, false);
    return this.value;
***REMOVED***;

Traverse.prototype.reduce = function (cb, init) ***REMOVED***
    var skip = arguments.length === 1;
    var acc = skip ? this.value : init;
    this.forEach(function (x) ***REMOVED***
        if (!this.isRoot || !skip) ***REMOVED***
            acc = cb.call(this, acc, x);
        ***REMOVED***
    ***REMOVED***);
    return acc;
***REMOVED***;

Traverse.prototype.paths = function () ***REMOVED***
    var acc = [];
    this.forEach(function (x) ***REMOVED***
        acc.push(this.path); 
    ***REMOVED***);
    return acc;
***REMOVED***;

Traverse.prototype.nodes = function () ***REMOVED***
    var acc = [];
    this.forEach(function (x) ***REMOVED***
        acc.push(this.node);
    ***REMOVED***);
    return acc;
***REMOVED***;

Traverse.prototype.clone = function () ***REMOVED***
    var parents = [], nodes = [];
    
    return (function clone (src) ***REMOVED***
        for (var i = 0; i < parents.length; i++) ***REMOVED***
            if (parents[i] === src) ***REMOVED***
                return nodes[i];
            ***REMOVED***
        ***REMOVED***
        
        if (typeof src === 'object' && src !== null) ***REMOVED***
            var dst = copy(src);
            
            parents.push(src);
            nodes.push(dst);
            
            forEach(objectKeys(src), function (key) ***REMOVED***
                dst[key] = clone(src[key]);
            ***REMOVED***);
            
            parents.pop();
            nodes.pop();
            return dst;
        ***REMOVED***
        else ***REMOVED***
            return src;
        ***REMOVED***
    ***REMOVED***)(this.value);
***REMOVED***;

function walk (root, cb, immutable) ***REMOVED***
    var path = [];
    var parents = [];
    var alive = true;
    
    return (function walker (node_) ***REMOVED***
        var node = immutable ? copy(node_) : node_;
        var modifiers = ***REMOVED******REMOVED***;
        
        var keepGoing = true;
        
        var state = ***REMOVED***
            node : node,
            node_ : node_,
            path : [].concat(path),
            parent : parents[parents.length - 1],
            parents : parents,
            key : path.slice(-1)[0],
            isRoot : path.length === 0,
            level : path.length,
            circular : null,
            update : function (x, stopHere) ***REMOVED***
                if (!state.isRoot) ***REMOVED***
                    state.parent.node[state.key] = x;
                ***REMOVED***
                state.node = x;
                if (stopHere) keepGoing = false;
            ***REMOVED***,
            'delete' : function (stopHere) ***REMOVED***
                delete state.parent.node[state.key];
                if (stopHere) keepGoing = false;
            ***REMOVED***,
            remove : function (stopHere) ***REMOVED***
                if (isArray(state.parent.node)) ***REMOVED***
                    state.parent.node.splice(state.key, 1);
                ***REMOVED***
                else ***REMOVED***
                    delete state.parent.node[state.key];
                ***REMOVED***
                if (stopHere) keepGoing = false;
            ***REMOVED***,
            keys : null,
            before : function (f) ***REMOVED*** modifiers.before = f ***REMOVED***,
            after : function (f) ***REMOVED*** modifiers.after = f ***REMOVED***,
            pre : function (f) ***REMOVED*** modifiers.pre = f ***REMOVED***,
            post : function (f) ***REMOVED*** modifiers.post = f ***REMOVED***,
            stop : function () ***REMOVED*** alive = false ***REMOVED***,
            block : function () ***REMOVED*** keepGoing = false ***REMOVED***
        ***REMOVED***;
        
        if (!alive) return state;
        
        function updateState() ***REMOVED***
            if (typeof state.node === 'object' && state.node !== null) ***REMOVED***
                if (!state.keys || state.node_ !== state.node) ***REMOVED***
                    state.keys = objectKeys(state.node)
                ***REMOVED***
                
                state.isLeaf = state.keys.length == 0;
                
                for (var i = 0; i < parents.length; i++) ***REMOVED***
                    if (parents[i].node_ === node_) ***REMOVED***
                        state.circular = parents[i];
                        break;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            else ***REMOVED***
                state.isLeaf = true;
                state.keys = null;
            ***REMOVED***
            
            state.notLeaf = !state.isLeaf;
            state.notRoot = !state.isRoot;
        ***REMOVED***
        
        updateState();
        
        // use return values to update if defined
        var ret = cb.call(state, state.node);
        if (ret !== undefined && state.update) state.update(ret);
        
        if (modifiers.before) modifiers.before.call(state, state.node);
        
        if (!keepGoing) return state;
        
        if (typeof state.node == 'object'
        && state.node !== null && !state.circular) ***REMOVED***
            parents.push(state);
            
            updateState();
            
            forEach(state.keys, function (key, i) ***REMOVED***
                path.push(key);
                
                if (modifiers.pre) modifiers.pre.call(state, state.node[key], key);
                
                var child = walker(state.node[key]);
                if (immutable && hasOwnProperty.call(state.node, key)) ***REMOVED***
                    state.node[key] = child.node;
                ***REMOVED***
                
                child.isLast = i == state.keys.length - 1;
                child.isFirst = i == 0;
                
                if (modifiers.post) modifiers.post.call(state, child);
                
                path.pop();
            ***REMOVED***);
            parents.pop();
        ***REMOVED***
        
        if (modifiers.after) modifiers.after.call(state, state.node);
        
        return state;
    ***REMOVED***)(root).node;
***REMOVED***

function copy (src) ***REMOVED***
    if (typeof src === 'object' && src !== null) ***REMOVED***
        var dst;
        
        if (isArray(src)) ***REMOVED***
            dst = [];
        ***REMOVED***
        else if (isDate(src)) ***REMOVED***
            dst = new Date(src.getTime ? src.getTime() : src);
        ***REMOVED***
        else if (isRegExp(src)) ***REMOVED***
            dst = new RegExp(src);
        ***REMOVED***
        else if (isError(src)) ***REMOVED***
            dst = ***REMOVED*** message: src.message ***REMOVED***;
        ***REMOVED***
        else if (isBoolean(src)) ***REMOVED***
            dst = new Boolean(src);
        ***REMOVED***
        else if (isNumber(src)) ***REMOVED***
            dst = new Number(src);
        ***REMOVED***
        else if (isString(src)) ***REMOVED***
            dst = new String(src);
        ***REMOVED***
        else if (Object.create && Object.getPrototypeOf) ***REMOVED***
            dst = Object.create(Object.getPrototypeOf(src));
        ***REMOVED***
        else if (src.constructor === Object) ***REMOVED***
            dst = ***REMOVED******REMOVED***;
        ***REMOVED***
        else ***REMOVED***
            var proto =
                (src.constructor && src.constructor.prototype)
                || src.__proto__
                || ***REMOVED******REMOVED***
            ;
            var T = function () ***REMOVED******REMOVED***;
            T.prototype = proto;
            dst = new T;
        ***REMOVED***
        
        forEach(objectKeys(src), function (key) ***REMOVED***
            dst[key] = src[key];
        ***REMOVED***);
        return dst;
    ***REMOVED***
    else return src;
***REMOVED***

var objectKeys = Object.keys || function keys (obj) ***REMOVED***
    var res = [];
    for (var key in obj) res.push(key)
    return res;
***REMOVED***;

function toS (obj) ***REMOVED*** return Object.prototype.toString.call(obj) ***REMOVED***
function isDate (obj) ***REMOVED*** return toS(obj) === '[object Date]' ***REMOVED***
function isRegExp (obj) ***REMOVED*** return toS(obj) === '[object RegExp]' ***REMOVED***
function isError (obj) ***REMOVED*** return toS(obj) === '[object Error]' ***REMOVED***
function isBoolean (obj) ***REMOVED*** return toS(obj) === '[object Boolean]' ***REMOVED***
function isNumber (obj) ***REMOVED*** return toS(obj) === '[object Number]' ***REMOVED***
function isString (obj) ***REMOVED*** return toS(obj) === '[object String]' ***REMOVED***

var isArray = Array.isArray || function isArray (xs) ***REMOVED***
    return Object.prototype.toString.call(xs) === '[object Array]';
***REMOVED***;

var forEach = function (xs, fn) ***REMOVED***
    if (xs.forEach) return xs.forEach(fn)
    else for (var i = 0; i < xs.length; i++) ***REMOVED***
        fn(xs[i], i, xs);
    ***REMOVED***
***REMOVED***;

forEach(objectKeys(Traverse.prototype), function (key) ***REMOVED***
    traverse[key] = function (obj) ***REMOVED***
        var args = [].slice.call(arguments, 1);
        var t = new Traverse(obj);
        return t[key].apply(t, args);
    ***REMOVED***;
***REMOVED***);

var hasOwnProperty = Object.hasOwnProperty || function (obj, key) ***REMOVED***
    return key in obj;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],54:[function(_dereq_,module,exports)***REMOVED***
/**
 * type-name - Just a reasonable typeof
 * 
 * https://github.com/twada/type-name
 *
 * Copyright (c) 2014-2016 Takuto Wada
 * Licensed under the MIT license.
 *   http://twada.mit-license.org/2014-2016
 */
'use strict';

var toStr = Object.prototype.toString;

function funcName (f) ***REMOVED***
    return f.name ? f.name : /^\s*function\s*([^\(]*)/im.exec(f.toString())[1];
***REMOVED***

function ctorName (obj) ***REMOVED***
    var strName = toStr.call(obj).slice(8, -1);
    if ((strName === 'Object' || strName === 'Error') && obj.constructor) ***REMOVED***
        return funcName(obj.constructor);
    ***REMOVED***
    return strName;
***REMOVED***

function typeName (val) ***REMOVED***
    var type;
    if (val === null) ***REMOVED***
        return 'null';
    ***REMOVED***
    type = typeof(val);
    if (type === 'object') ***REMOVED***
        return ctorName(val);
    ***REMOVED***
    return type;
***REMOVED***

module.exports = typeName;

***REMOVED***,***REMOVED******REMOVED***],55:[function(_dereq_,module,exports)***REMOVED***
'use strict';

var typeName = _dereq_('type-name');
var forEach = _dereq_('core-js/library/fn/array/for-each');
var arrayFilter = _dereq_('core-js/library/fn/array/filter');
var reduceRight = _dereq_('core-js/library/fn/array/reduce-right');
var indexOf = _dereq_('core-js/library/fn/array/index-of');
var slice = Array.prototype.slice;
var END = ***REMOVED******REMOVED***;
var ITERATE = ***REMOVED******REMOVED***;

// arguments should end with end or iterate
function compose () ***REMOVED***
    var filters = slice.apply(arguments);
    return reduceRight(filters, function(right, left) ***REMOVED***
        return left(right);
    ***REMOVED***);
***REMOVED***

// skip children
function end () ***REMOVED***
    return function (acc, x) ***REMOVED***
        acc.context.keys = [];
        return END;
    ***REMOVED***;
***REMOVED***

// iterate children
function iterate () ***REMOVED***
    return function (acc, x) ***REMOVED***
        return ITERATE;
    ***REMOVED***;
***REMOVED***

function filter (predicate) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            var toBeIterated;
            var isIteratingArray = (typeName(x) === 'Array');
            if (typeName(predicate) === 'function') ***REMOVED***
                toBeIterated = [];
                forEach(acc.context.keys, function (key) ***REMOVED***
                    var indexOrKey = isIteratingArray ? parseInt(key, 10) : key;
                    var kvp = ***REMOVED***
                        key: indexOrKey,
                        value: x[key]
                    ***REMOVED***;
                    var decision = predicate(kvp);
                    if (decision) ***REMOVED***
                        toBeIterated.push(key);
                    ***REMOVED***
                    if (typeName(decision) === 'number') ***REMOVED***
                        truncateByKey(decision, key, acc);
                    ***REMOVED***
                    if (typeName(decision) === 'function') ***REMOVED***
                        customizeStrategyForKey(decision, key, acc);
                    ***REMOVED***
                ***REMOVED***);
                acc.context.keys = toBeIterated;
            ***REMOVED***
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function customizeStrategyForKey (strategy, key, acc) ***REMOVED***
    acc.handlers[currentPath(key, acc)] = strategy;
***REMOVED***

function truncateByKey (size, key, acc) ***REMOVED***
    acc.handlers[currentPath(key, acc)] = size;
***REMOVED***

function currentPath (key, acc) ***REMOVED***
    var pathToCurrentNode = [''].concat(acc.context.path);
    if (typeName(key) !== 'undefined') ***REMOVED***
        pathToCurrentNode.push(key);
    ***REMOVED***
    return pathToCurrentNode.join('/');
***REMOVED***

function allowedKeys (orderedWhiteList) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            var isIteratingArray = (typeName(x) === 'Array');
            if (!isIteratingArray && typeName(orderedWhiteList) === 'Array') ***REMOVED***
                acc.context.keys = arrayFilter(orderedWhiteList, function (propKey) ***REMOVED***
                    return indexOf(acc.context.keys, propKey) !== -1;
                ***REMOVED***);
            ***REMOVED***
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function safeKeys () ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            if (typeName(x) !== 'Array') ***REMOVED***
                acc.context.keys = arrayFilter(acc.context.keys, function (propKey) ***REMOVED***
                    // Error handling for unsafe property access.
                    // For example, on PhantomJS,
                    // accessing HTMLInputElement.selectionEnd causes TypeError
                    try ***REMOVED***
                        var val = x[propKey];
                        return true;
                    ***REMOVED*** catch (e) ***REMOVED***
                        // skip unsafe key
                        return false;
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function arrayIndicesToKeys () ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            if (typeName(x) === 'Array' && 0 < x.length) ***REMOVED***
                var indices = Array(x.length);
                for(var i = 0; i < x.length; i += 1) ***REMOVED***
                    indices[i] = String(i); // traverse uses strings as keys
                ***REMOVED***
                acc.context.keys = indices;
            ***REMOVED***
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function when (guard, then) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            var kvp = ***REMOVED***
                key: acc.context.key,
                value: x
            ***REMOVED***;
            if (guard(kvp, acc)) ***REMOVED***
                return then(acc, x);
            ***REMOVED***
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function truncate (size) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            var orig = acc.push;
            var ret;
            acc.push = function (str) ***REMOVED***
                var savings = str.length - size;
                var truncated;
                if (savings <= size) ***REMOVED***
                    orig.call(acc, str);
                ***REMOVED*** else ***REMOVED***
                    truncated = str.substring(0, size);
                    orig.call(acc, truncated + acc.options.snip);
                ***REMOVED***
            ***REMOVED***;
            ret = next(acc, x);
            acc.push = orig;
            return ret;
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function constructorName () ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            var name = acc.options.typeFun(x);
            if (name === '') ***REMOVED***
                name = acc.options.anonymous;
            ***REMOVED***
            acc.push(name);
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function always (str) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            acc.push(str);
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function optionValue (key) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            acc.push(acc.options[key]);
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function json (replacer) ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            acc.push(JSON.stringify(x, replacer));
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function toStr () ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            acc.push(x.toString());
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function decorateArray () ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            acc.context.before(function (node) ***REMOVED***
                acc.push('[');
            ***REMOVED***);
            acc.context.after(function (node) ***REMOVED***
                afterAllChildren(this, acc.push, acc.options);
                acc.push(']');
            ***REMOVED***);
            acc.context.pre(function (val, key) ***REMOVED***
                beforeEachChild(this, acc.push, acc.options);
            ***REMOVED***);
            acc.context.post(function (childContext) ***REMOVED***
                afterEachChild(childContext, acc.push);
            ***REMOVED***);
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function decorateObject () ***REMOVED***
    return function (next) ***REMOVED***
        return function (acc, x) ***REMOVED***
            acc.context.before(function (node) ***REMOVED***
                acc.push('***REMOVED***');
            ***REMOVED***);
            acc.context.after(function (node) ***REMOVED***
                afterAllChildren(this, acc.push, acc.options);
                acc.push('***REMOVED***');
            ***REMOVED***);
            acc.context.pre(function (val, key) ***REMOVED***
                beforeEachChild(this, acc.push, acc.options);
                acc.push(sanitizeKey(key) + (acc.options.indent ? ': ' : ':'));
            ***REMOVED***);
            acc.context.post(function (childContext) ***REMOVED***
                afterEachChild(childContext, acc.push);
            ***REMOVED***);
            return next(acc, x);
        ***REMOVED***;
    ***REMOVED***;
***REMOVED***

function sanitizeKey (key) ***REMOVED***
    return /^[A-Za-z_]+$/.test(key) ? key : JSON.stringify(key);
***REMOVED***

function afterAllChildren (context, push, options) ***REMOVED***
    if (options.indent && 0 < context.keys.length) ***REMOVED***
        push(options.lineSeparator);
        for(var i = 0; i < context.level; i += 1) ***REMOVED*** // indent level - 1
            push(options.indent);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function beforeEachChild (context, push, options) ***REMOVED***
    if (options.indent) ***REMOVED***
        push(options.lineSeparator);
        for(var i = 0; i <= context.level; i += 1) ***REMOVED***
            push(options.indent);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function afterEachChild (childContext, push) ***REMOVED***
    if (!childContext.isLast) ***REMOVED***
        push(',');
    ***REMOVED***
***REMOVED***

function nan (kvp, acc) ***REMOVED***
    return kvp.value !== kvp.value;
***REMOVED***

function positiveInfinity (kvp, acc) ***REMOVED***
    return !isFinite(kvp.value) && kvp.value === Infinity;
***REMOVED***

function negativeInfinity (kvp, acc) ***REMOVED***
    return !isFinite(kvp.value) && kvp.value !== Infinity;
***REMOVED***

function circular (kvp, acc) ***REMOVED***
    return acc.context.circular;
***REMOVED***

function maxDepth (kvp, acc) ***REMOVED***
    return (acc.options.maxDepth && acc.options.maxDepth <= acc.context.level);
***REMOVED***

var prune = compose(
    always('#'),
    constructorName(),
    always('#'),
    end()
);
var omitNaN = when(nan, compose(
    always('NaN'),
    end()
));
var omitPositiveInfinity = when(positiveInfinity, compose(
    always('Infinity'),
    end()
));
var omitNegativeInfinity = when(negativeInfinity, compose(
    always('-Infinity'),
    end()
));
var omitCircular = when(circular, compose(
    optionValue('circular'),
    end()
));
var omitMaxDepth = when(maxDepth, prune);

module.exports = ***REMOVED***
    filters: ***REMOVED***
        always: always,
        optionValue: optionValue,
        constructorName: constructorName,
        json: json,
        toStr: toStr,
        prune: prune,
        truncate: truncate,
        decorateArray: decorateArray,
        decorateObject: decorateObject
    ***REMOVED***,
    flow: ***REMOVED***
        compose: compose,
        when: when,
        allowedKeys: allowedKeys,
        safeKeys: safeKeys,
        arrayIndicesToKeys: arrayIndicesToKeys,
        filter: filter,
        iterate: iterate,
        end: end
    ***REMOVED***,
    symbols: ***REMOVED***
        END: END,
        ITERATE: ITERATE
    ***REMOVED***,
    always: function (str) ***REMOVED***
        return compose(always(str), end());
    ***REMOVED***,
    json: function () ***REMOVED***
        return compose(json(), end());
    ***REMOVED***,
    toStr: function () ***REMOVED***
        return compose(toStr(), end());
    ***REMOVED***,
    prune: function () ***REMOVED***
        return prune;
    ***REMOVED***,
    number: function () ***REMOVED***
        return compose(
            omitNaN,
            omitPositiveInfinity,
            omitNegativeInfinity,
            json(),
            end()
        );
    ***REMOVED***,
    newLike: function () ***REMOVED***
        return compose(
            always('new '),
            constructorName(),
            always('('),
            json(),
            always(')'),
            end()
        );
    ***REMOVED***,
    array: function (predicate) ***REMOVED***
        return compose(
            omitCircular,
            omitMaxDepth,
            decorateArray(),
            arrayIndicesToKeys(),
            filter(predicate),
            iterate()
        );
    ***REMOVED***,
    object: function (predicate, orderedWhiteList) ***REMOVED***
        return compose(
            omitCircular,
            omitMaxDepth,
            constructorName(),
            decorateObject(),
            allowedKeys(orderedWhiteList),
            safeKeys(),
            filter(predicate),
            iterate()
        );
    ***REMOVED***
***REMOVED***;

***REMOVED***,***REMOVED***"core-js/library/fn/array/filter":2,"core-js/library/fn/array/for-each":3,"core-js/library/fn/array/index-of":4,"core-js/library/fn/array/reduce-right":5,"type-name":54***REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[1])(1)
***REMOVED***);
