// is/test/index.js
/* globals window, document, HTMLElement */

'use strict';

var test = require('tape');
var is = require('../index.js');

var forEach = require('foreach');
var toStr = Object.prototype.toString;

var genFn = require('make-generator-function');

test('is.type', function (t) ***REMOVED***
  var booleans = [true, false];
  forEach(booleans, function (boolean) ***REMOVED***
    t.ok(is.type(boolean, 'boolean'), '"' + boolean + '" is a boolean');
  ***REMOVED***);

  var numbers = [1, 0 / 1, 0 / -1, NaN, Infinity, -Infinity];
  forEach(numbers, function (number) ***REMOVED***
    t.ok(is.type(number, 'number'), '"' + number + '" is a number');
  ***REMOVED***);

  var objects = [***REMOVED******REMOVED***, null, new Date()];
  forEach(objects, function (object) ***REMOVED***
    t.ok(is.type(object, 'object'), '"' + object + '" is an object');
  ***REMOVED***);

  var strings = ['', 'abc'];
  forEach(strings, function (string) ***REMOVED***
    t.ok(is.type(string, 'string'), '"' + string + '" is a string');
  ***REMOVED***);

  t.ok(is.type(undefined, 'undefined'), 'undefined is undefined');

  t.end();
***REMOVED***);

test('is.undef', function (t) ***REMOVED***
  t.ok(is.undef(), 'absent undefined is undefined');
  t.ok(is.undef(undefined), 'literal undefined is undefined');
  t.notOk(is.undef(null), 'null is not undefined');
  t.notOk(is.undef(***REMOVED******REMOVED***), 'object is not undefined');
  t.end();
***REMOVED***);

test('is.defined', function (t) ***REMOVED***
  t.notOk(is.defined(), 'undefined is not defined');
  t.ok(is.defined(null), 'null is defined');
  t.ok(is.defined(***REMOVED******REMOVED***), 'object is defined');
  t.end();
***REMOVED***);

test('is.empty', function (t) ***REMOVED***
  t.ok(is.empty(''), 'empty string is empty');
  t.ok(is.empty(Object('')), 'empty String object is empty');
  t.ok(is.empty([]), 'empty array is empty');
  t.ok(is.empty(***REMOVED******REMOVED***), 'empty object is empty');
  t.ok(is.empty(null), 'null is empty');
  t.ok(is.empty(), 'undefined is empty');
  t.ok(is.empty(undefined), 'undefined is empty');
  t.ok(is.empty(false), 'false is empty');
  t.ok(is.empty(0), '0 is empty');
  t.ok(is.empty(NaN), 'nan is empty');
  (function () ***REMOVED***
    t.ok(is.empty(arguments), 'empty arguments is empty');
  ***REMOVED***());
  t.notOk(is.empty(***REMOVED*** a: 1 ***REMOVED***), 'nonempty object is not empty');
  t.notOk(is.empty(true), 'true is not empty');
  t.notOk(is.empty(/a/g), 'regex is not empty');
  t.notOk(is.empty(new Date()), 'date is not empty');
  t.end();
***REMOVED***);

test('is.equal', function (t) ***REMOVED***
  t.test('primitives', function (pt) ***REMOVED***
    var primitives = [true, false, undefined, null, '', 'foo', 0, Infinity, -Infinity];
    pt.plan(primitives.length);
    for (var i = 0; i < primitives.length; ++i) ***REMOVED***
      pt.ok(is.equal(primitives[i], primitives[i]), 'primitives are equal to themselves: ' + primitives[i]);
    ***REMOVED***
    pt.end();
  ***REMOVED***);

  t.test('arrays', function (at) ***REMOVED***
    at.ok(is.equal([1, 2, 3], [1, 2, 3]), 'arrays are shallowly equal');
    at.ok(is.equal([1, 2, [3, 4]], [1, 2, [3, 4]]), 'arrays are deep equal');
    at.notOk(is.equal([1, 2, 3], [5, 2, 3]), 'inequal arrays are not equal');
    at.notOk(is.equal([1, 2], [2, 3]), 'inequal arrays are not equal');
    at.notOk(is.equal([1, 2, 3], [2, 3]), 'inequal length arrays are not equal');
    at.ok(is.equal([], []), 'empty arrays are equal');

    var arr = [1, 2];
    at.ok(is.equal(arr, arr), 'array is equal to itself');

    at.end();
  ***REMOVED***);

  t.test('dates', function (dt) ***REMOVED***
    dt.plan(2);
    var now = new Date();
    dt.ok(is.equal(now, new Date(now.getTime())), 'two equal date objects are equal');
    setTimeout(function () ***REMOVED***
      dt.notOk(is.equal(now, new Date()), 'two inequal date objects are not equal');
      dt.end();
    ***REMOVED***, 10);
  ***REMOVED***);

  t.test('plain objects', function (ot) ***REMOVED***
    ot.ok(is.equal(***REMOVED*** a: 1, b: 2, c: 3 ***REMOVED***, ***REMOVED*** a: 1, b: 2, c: 3 ***REMOVED***), 'objects are shallowly equal');
    ot.ok(is.equal(***REMOVED*** a: ***REMOVED*** b: 1 ***REMOVED*** ***REMOVED***, ***REMOVED*** a: ***REMOVED*** b: 1 ***REMOVED*** ***REMOVED***), 'objects are deep equal');
    ot.notOk(is.equal(***REMOVED*** a: 1 ***REMOVED***, ***REMOVED*** a: 2 ***REMOVED***), 'inequal objects are not equal');
    ot.end();
  ***REMOVED***);

  t.test('object instances', function (ot) ***REMOVED***
    var F = function F() ***REMOVED***
      this.foo = 'bar';
    ***REMOVED***;
    F.prototype = ***REMOVED******REMOVED***;
    var G = function G() ***REMOVED***
      this.foo = 'bar';
    ***REMOVED***;
    var f = new F();
    var g = new G();

    ot.ok(is.equal(f, f), 'the same object instances are equal');
    ot.ok(is.equal(f, new F()), 'two object instances are equal when the prototype and props are the same');
    ot.ok(is.equal(f, new G()), 'two object instances are equal when the prototype is not the same, but props are');

    g.bar = 'baz';
    ot.notOk(is.equal(f, g), 'object instances are not equal when the prototype and props are not the same');
    ot.notOk(is.equal(g, f), 'object instances are not equal when the prototype and props are not the same');
    ot.end();
  ***REMOVED***);

  t.test('functions', function (ft) ***REMOVED***
    var F = function () ***REMOVED******REMOVED***;
    F.prototype = ***REMOVED******REMOVED***;
    var G = function () ***REMOVED******REMOVED***;
    G.prototype = new Date();

    ft.notEqual(F.prototype, G.prototype, 'F and G have different prototypes');
    ft.notOk(is.equal(F, G), 'two functions are not equal when the prototype is not the same');

    var H = function () ***REMOVED******REMOVED***;
    H.prototype = F.prototype;

    ft.equal(F.prototype, H.prototype, 'F and H have the same prototype');
    ft.ok(is.equal(F, H), 'two functions are equal when the prototype is the same');
    ft.end();
  ***REMOVED***);

  t.end();
***REMOVED***);

test('is.hosted', function (t) ***REMOVED***
  t.ok(is.hosted('a', ***REMOVED*** a: ***REMOVED******REMOVED*** ***REMOVED***), 'object is hosted');
  t.ok(is.hosted('a', ***REMOVED*** a: [] ***REMOVED***), 'array is hosted');
  t.ok(is.hosted('a', ***REMOVED*** a: function () ***REMOVED******REMOVED*** ***REMOVED***), 'function is hosted');
  t.notOk(is.hosted('a', ***REMOVED*** a: true ***REMOVED***), 'boolean value is not hosted');
  t.notOk(is.hosted('a', ***REMOVED*** a: false ***REMOVED***), 'boolean value is not hosted');
  t.notOk(is.hosted('a', ***REMOVED*** a: 3 ***REMOVED***), 'number value is not hosted');
  t.notOk(is.hosted('a', ***REMOVED*** a: undefined ***REMOVED***), 'undefined value is not hosted');
  t.notOk(is.hosted('a', ***REMOVED*** a: 'abc' ***REMOVED***), 'string value is not hosted');
  t.notOk(is.hosted('a', ***REMOVED*** a: null ***REMOVED***), 'null value is not hosted');
  t.end();
***REMOVED***);

test('is.instance', function (t) ***REMOVED***
  t.ok(is.instance(new Date(), Date), 'new Date is instanceof Date');
  var F = function () ***REMOVED******REMOVED***;
  t.ok(is.instance(new F(), F), 'new constructor is instanceof constructor');
  t.end();
***REMOVED***);

test('is.nil', function (t) ***REMOVED***
  var isNull = is.nil;
  t.equal(isNull, is['null'], 'is.nil is the same as is.null');
  t.ok(isNull(null), 'null is null');
  t.notOk(isNull(undefined), 'undefined is not null');
  t.notOk(isNull(***REMOVED******REMOVED***), 'object is not null');
  t.end();
***REMOVED***);

test('is.args', function (t) ***REMOVED***
  t.notOk(is.args([]), 'array is not arguments');
  (function () ***REMOVED***
    t.ok(is.args(arguments), 'arguments is arguments');
  ***REMOVED***());
  (function () ***REMOVED***
    t.notOk(is.args(Array.prototype.slice.call(arguments)), 'sliced arguments is not arguments');
  ***REMOVED***());
  var fakeOldArguments = ***REMOVED***
    callee: function () ***REMOVED******REMOVED***,
    length: 3
  ***REMOVED***;
  t.ok(is.args(fakeOldArguments), 'old-style arguments object is arguments');
  t.end();
***REMOVED***);

test('is.args.empty', function (t) ***REMOVED***
  t.notOk(is.args.empty([]), 'empty array is not empty arguments');
  (function () ***REMOVED***
    t.ok(is.args.empty(arguments), 'empty arguments is empty arguments');
  ***REMOVED***());
  (function () ***REMOVED***
    t.notOk(is.args.empty(Array.prototype.slice.call(arguments)), 'empty sliced arguments is not empty arguments');
  ***REMOVED***());
  t.end();
***REMOVED***);

test('is.array', function (t) ***REMOVED***
  t.ok(is.array([]), 'array is array');
  (function () ***REMOVED***
    t.ok(is.array(Array.prototype.slice.call(arguments)), 'sliced arguments is array');
  ***REMOVED***());
  t.end();
***REMOVED***);

test('is.array.empty', function (t) ***REMOVED***
  t.ok(is.array.empty([]), 'empty array is empty array');
  (function () ***REMOVED***
    t.notOk(is.array.empty(arguments), 'empty arguments is not empty array');
  ***REMOVED***());
  (function () ***REMOVED***
    t.ok(is.array.empty(Array.prototype.slice.call(arguments)), 'empty sliced arguments is empty array');
  ***REMOVED***());
  t.end();
***REMOVED***);

test('is.isarraylike', function (t) ***REMOVED***
  t.notOk(is.arraylike(), 'undefined is not array-like');
  t.notOk(is.arraylike(null), 'null is not array-like');
  t.notOk(is.arraylike(false), 'false is not array-like');
  t.notOk(is.arraylike(true), 'true is not array-like');
  t.ok(is.arraylike(***REMOVED*** length: 0 ***REMOVED***), 'object with zero length is array-like');
  t.ok(is.arraylike(***REMOVED*** length: 1 ***REMOVED***), 'object with positive length is array-like');
  t.notOk(is.arraylike(***REMOVED*** length: -1 ***REMOVED***), 'object with negative length is not array-like');
  t.notOk(is.arraylike(***REMOVED*** length: NaN ***REMOVED***), 'object with NaN length is not array-like');
  t.notOk(is.arraylike(***REMOVED*** length: 'foo' ***REMOVED***), 'object with string length is not array-like');
  t.notOk(is.arraylike(***REMOVED*** length: '' ***REMOVED***), 'object with empty string length is not array-like');
  t.ok(is.arraylike([]), 'array is array-like');
  (function () ***REMOVED***
    t.ok(is.arraylike(arguments), 'empty arguments is array-like');
  ***REMOVED***());
  (function () ***REMOVED***
    t.ok(is.arraylike(arguments), 'nonempty arguments is array-like');
  ***REMOVED***(1, 2, 3));
  t.end();
***REMOVED***);

test('is.bool', function (t) ***REMOVED***
  t.ok(is.bool(true), 'literal true is a boolean');
  t.ok(is.bool(false), 'literal false is a boolean');
  t.ok(is.bool(Object(true)), 'object true is a boolean');
  t.ok(is.bool(Object(false)), 'object false is a boolean');
  t.notOk(is.bool(), 'undefined is not a boolean');
  t.notOk(is.bool(null), 'null is not a boolean');
  t.end();
***REMOVED***);

test('is.false', function (t) ***REMOVED***
  var isFalse = is['false'];
  t.ok(isFalse(false), 'false is false');
  t.ok(isFalse(Object(false)), 'object false is false');
  t.notOk(isFalse(true), 'true is not false');
  t.notOk(isFalse(), 'undefined is not false');
  t.notOk(isFalse(null), 'null is not false');
  t.notOk(isFalse(''), 'empty string is not false');
  t.end();
***REMOVED***);

test('is.true', function (t) ***REMOVED***
  var isTrue = is['true'];
  t.ok(isTrue(true), 'true is true');
  t.ok(isTrue(Object(true)), 'object true is true');
  t.notOk(isTrue(false), 'false is not true');
  t.notOk(isTrue(), 'undefined is not true');
  t.notOk(isTrue(null), 'null is not true');
  t.notOk(isTrue(''), 'empty string is not true');
  t.end();
***REMOVED***);

test('is.date', function (t) ***REMOVED***
  t.ok(is.date(new Date()), 'new Date is date');
  t.notOk(is.date(), 'undefined is not date');
  t.notOk(is.date(null), 'null is not date');
  t.notOk(is.date(''), 'empty string is not date');
  var nowTS = (new Date()).getTime();
  t.notOk(is.date(nowTS), 'timestamp is not date');
  var F = function () ***REMOVED******REMOVED***;
  F.prototype = new Date();
  t.notOk(is.date(new F()), 'Date subtype is not date');
  t.end();
***REMOVED***);

test('is.date.valid', function (t) ***REMOVED***
  t.ok(is.date.valid(new Date()), 'new Date() is a valid date');
  t.notOk(is.date.valid(new Date('')), 'new Date("") is not a valid date');
  t.end();
***REMOVED***);

test('is.element', function (t) ***REMOVED***
  t.notOk(is.element(), 'undefined is not element');

  t.test('when HTMLElement exists', ***REMOVED*** skip: typeof HTMLElement === 'undefined' ***REMOVED***, function (st) ***REMOVED***
    var element = document.createElement('div');
    st.ok(is.element(element), 'HTMLElement is element');
    st.notOk(is.element(***REMOVED*** nodeType: 1 ***REMOVED***), 'object with nodeType is not element');
    st.end();
  ***REMOVED***);

  t.end();
***REMOVED***);

test('is.error', function (t) ***REMOVED***
  var err = new Error('foo');
  t.ok(is.error(err), 'Error is error');
  t.notOk(is.error(***REMOVED******REMOVED***), 'object is not error');
  var objWithErrorToString = ***REMOVED***
    toString: function () ***REMOVED***
      return '[object Error]';
    ***REMOVED***
  ***REMOVED***;
  t.equal(String(objWithErrorToString), toStr.call(new Error()), 'obj has Error\'s toString');
  t.notOk(is.error(objWithErrorToString), 'object with Error\'s toString is not error');
  t.end();
***REMOVED***);

test('is.fn', function (t) ***REMOVED***
  t.equal(is['function'], is.fn, 'alias works');
  t.ok(is.fn(function () ***REMOVED******REMOVED***), 'function is function');
  if (typeof window !== 'undefined') ***REMOVED***
    // in IE7/8, typeof alert === 'object'
    t.ok(is.fn(window.alert), 'window.alert is function');
  ***REMOVED***
  t.notOk(is.fn(***REMOVED******REMOVED***), 'object is not function');
  t.notOk(is.fn(null), 'null is not function');

  t.test('generator functions', ***REMOVED*** skip: !genFn ***REMOVED***, function (st) ***REMOVED***
    t.ok(is.fn(genFn), 'generator function is function');
    st.end();
  ***REMOVED***);

  t.end();
***REMOVED***);

test('is.number', function (t) ***REMOVED***
  t.ok(is.number(0), 'positive zero is number');
  t.ok(is.number(0 / -1), 'negative zero is number');
  t.ok(is.number(3), 'three is number');
  t.ok(is.number(NaN), 'NaN is number');
  t.ok(is.number(Infinity), 'infinity is number');
  t.ok(is.number(-Infinity), 'negative infinity is number');
  t.ok(is.number(Object(42)), 'object number is number');
  t.notOk(is.number(), 'undefined is not number');
  t.notOk(is.number(null), 'null is not number');
  t.notOk(is.number(true), 'true is not number');
  t.end();
***REMOVED***);

test('is.infinite', function (t) ***REMOVED***
  t.ok(is.infinite(Infinity), 'positive infinity is infinite');
  t.ok(is.infinite(-Infinity), 'negative infinity is infinite');
  t.notOk(is.infinite(NaN), 'NaN is not infinite');
  t.notOk(is.infinite(0), 'a number is not infinite');
  t.end();
***REMOVED***);

test('is.decimal', function (t) ***REMOVED***
  t.ok(is.decimal(1.1), 'decimal is decimal');
  t.notOk(is.decimal(0), 'zero is not decimal');
  t.notOk(is.decimal(1), 'integer is not decimal');
  t.notOk(is.decimal(NaN), 'NaN is not decimal');
  t.notOk(is.decimal(Infinity), 'Infinity is not decimal');
  t.end();
***REMOVED***);

test('is.divisibleBy', function (t) ***REMOVED***
  t.ok(is.divisibleBy(4, 2), '4 is divisible by 2');
  t.ok(is.divisibleBy(4, 2), '4 is divisible by 2');
  t.ok(is.divisibleBy(0, 1), '0 is divisible by 1');
  t.ok(is.divisibleBy(Infinity, 1), 'infinity is divisible by anything');
  t.ok(is.divisibleBy(1, Infinity), 'anything is divisible by infinity');
  t.ok(is.divisibleBy(Infinity, Infinity), 'infinity is divisible by infinity');
  t.notOk(is.divisibleBy(1, 0), '1 is not divisible by 0');
  t.notOk(is.divisibleBy(NaN, 1), 'NaN is not divisible by 1');
  t.notOk(is.divisibleBy(1, NaN), '1 is not divisible by NaN');
  t.notOk(is.divisibleBy(NaN, NaN), 'NaN is not divisible by NaN');
  t.notOk(is.divisibleBy(1, 3), '1 is not divisible by 3');
  t.end();
***REMOVED***);

test('is.integer', function (t) ***REMOVED***
  t.ok(is.integer(0), '0 is integer');
  t.ok(is.integer(3), '3 is integer');
  t.notOk(is.integer(1.1), '1.1 is not integer');
  t.notOk(is.integer(NaN), 'NaN is not integer');
  t.notOk(is.integer(Infinity), 'infinity is not integer');
  t.notOk(is.integer(null), 'null is not integer');
  t.notOk(is.integer(), 'undefined is not integer');
  t.end();
***REMOVED***);

test('is.maximum', function (t) ***REMOVED***
  t.ok(is.maximum(3, [3, 2, 1]), '3 is maximum of [3,2,1]');
  t.ok(is.maximum(3, [1, 2, 3]), '3 is maximum of [1,2,3]');
  t.ok(is.maximum(4, [1, 2, 3]), '4 is maximum of [1,2,3]');
  t.ok(is.maximum('c', ['a', 'b', 'c']), 'c is maximum of [a,b,c]');
  t.notOk(is.maximum(2, [1, 2, 3]), '2 is not maximum of [1,2,3]');

  var nanError = new TypeError('NaN is not a valid value');
  t['throws'](function () ***REMOVED*** return is.maximum(NaN); ***REMOVED***, nanError, 'throws when first value is NaN');

  var error = new TypeError('second argument must be array-like');
  t['throws'](function () ***REMOVED*** return is.maximum(2, null); ***REMOVED***, error, 'throws when second value is not array-like');
  t['throws'](function () ***REMOVED*** return is.maximum(2, ***REMOVED******REMOVED***); ***REMOVED***, error, 'throws when second value is not array-like');
  t.end();
***REMOVED***);

test('is.minimum', function (t) ***REMOVED***
  t.ok(is.minimum(1, [1, 2, 3]), '1 is minimum of [1,2,3]');
  t.ok(is.minimum(0, [1, 2, 3]), '0 is minimum of [1,2,3]');
  t.ok(is.minimum('a', ['a', 'b', 'c']), 'a is minimum of [a,b,c]');
  t.notOk(is.minimum(2, [1, 2, 3]), '2 is not minimum of [1,2,3]');

  var nanError = new TypeError('NaN is not a valid value');
  t['throws'](function () ***REMOVED*** return is.minimum(NaN); ***REMOVED***, nanError, 'throws when first value is NaN');

  var error = new TypeError('second argument must be array-like');
  t['throws'](function () ***REMOVED*** return is.minimum(2, null); ***REMOVED***, error, 'throws when second value is not array-like');
  t['throws'](function () ***REMOVED*** return is.minimum(2, ***REMOVED******REMOVED***); ***REMOVED***, error, 'throws when second value is not array-like');
  t.end();
***REMOVED***);

test('is.nan', function (t) ***REMOVED***
  t.ok(is.nan(NaN), 'NaN is not a number');
  t.ok(is.nan('abc'), 'string is not a number');
  t.ok(is.nan(true), 'boolean is not a number');
  t.ok(is.nan(***REMOVED******REMOVED***), 'object is not a number');
  t.ok(is.nan([]), 'array is not a number');
  t.ok(is.nan(function () ***REMOVED******REMOVED***), 'function is not a number');
  t.notOk(is.nan(0), 'zero is a number');
  t.notOk(is.nan(3), 'three is a number');
  t.notOk(is.nan(1.1), '1.1 is a number');
  t.notOk(is.nan(Infinity), 'infinity is a number');
  t.end();
***REMOVED***);

test('is.even', function (t) ***REMOVED***
  t.ok(is.even(0), 'zero is even');
  t.ok(is.even(2), 'two is even');
  t.ok(is.even(Infinity), 'infinity is even');
  t.notOk(is.even(1), '1 is not even');
  t.notOk(is.even(), 'undefined is not even');
  t.notOk(is.even(null), 'null is not even');
  t.notOk(is.even(NaN), 'NaN is not even');
  t.end();
***REMOVED***);

test('is.odd', function (t) ***REMOVED***
  t.ok(is.odd(1), 'zero is odd');
  t.ok(is.odd(3), 'two is odd');
  t.ok(is.odd(Infinity), 'infinity is odd');
  t.notOk(is.odd(0), '0 is not odd');
  t.notOk(is.odd(2), '2 is not odd');
  t.notOk(is.odd(), 'undefined is not odd');
  t.notOk(is.odd(null), 'null is not odd');
  t.notOk(is.odd(NaN), 'NaN is not odd');
  t.end();
***REMOVED***);

test('is.ge', function (t) ***REMOVED***
  t.ok(is.ge(3, 2), '3 is greater than 2');
  t.notOk(is.ge(2, 3), '2 is not greater than 3');
  t.ok(is.ge(3, 3), '3 is greater than or equal to 3');
  t.ok(is.ge('abc', 'a'), 'abc is greater than a');
  t.ok(is.ge('abc', 'abc'), 'abc is greater than or equal to abc');
  t.notOk(is.ge('a', 'abc'), 'a is not greater than abc');
  t.notOk(is.ge(Infinity, 0), 'infinity is not greater than anything');
  t.notOk(is.ge(0, Infinity), 'anything is not greater than infinity');
  var error = new TypeError('NaN is not a valid value');
  t['throws'](function () ***REMOVED*** return is.ge(NaN, 2); ***REMOVED***, error, 'throws when first value is NaN');
  t['throws'](function () ***REMOVED*** return is.ge(2, NaN); ***REMOVED***, error, 'throws when second value is NaN');
  t.end();
***REMOVED***);

test('is.gt', function (t) ***REMOVED***
  t.ok(is.gt(3, 2), '3 is greater than 2');
  t.notOk(is.gt(2, 3), '2 is not greater than 3');
  t.notOk(is.gt(3, 3), '3 is not greater than 3');
  t.ok(is.gt('abc', 'a'), 'abc is greater than a');
  t.notOk(is.gt('abc', 'abc'), 'abc is not greater than abc');
  t.notOk(is.gt('a', 'abc'), 'a is not greater than abc');
  t.notOk(is.gt(Infinity, 0), 'infinity is not greater than anything');
  t.notOk(is.gt(0, Infinity), 'anything is not greater than infinity');
  var error = new TypeError('NaN is not a valid value');
  t['throws'](function () ***REMOVED*** return is.gt(NaN, 2); ***REMOVED***, error, 'throws when first value is NaN');
  t['throws'](function () ***REMOVED*** return is.gt(2, NaN); ***REMOVED***, error, 'throws when second value is NaN');
  t.end();
***REMOVED***);

test('is.le', function (t) ***REMOVED***
  t.ok(is.le(2, 3), '2 is lesser than or equal to 3');
  t.notOk(is.le(3, 2), '3 is not lesser than or equal to 2');
  t.ok(is.le(3, 3), '3 is lesser than or equal to 3');
  t.ok(is.le('a', 'abc'), 'a is lesser than or equal to abc');
  t.ok(is.le('abc', 'abc'), 'abc is lesser than or equal to abc');
  t.notOk(is.le('abc', 'a'), 'abc is not lesser than or equal to a');
  t.notOk(is.le(Infinity, 0), 'infinity is not lesser than or equal to anything');
  t.notOk(is.le(0, Infinity), 'anything is not lesser than or equal to infinity');
  var error = new TypeError('NaN is not a valid value');
  t['throws'](function () ***REMOVED*** return is.le(NaN, 2); ***REMOVED***, error, 'throws when first value is NaN');
  t['throws'](function () ***REMOVED*** return is.le(2, NaN); ***REMOVED***, error, 'throws when second value is NaN');
  t.end();
***REMOVED***);

test('is.lt', function (t) ***REMOVED***
  t.ok(is.lt(2, 3), '2 is lesser than 3');
  t.notOk(is.lt(3, 2), '3 is not lesser than 2');
  t.notOk(is.lt(3, 3), '3 is not lesser than 3');
  t.ok(is.lt('a', 'abc'), 'a is lesser than abc');
  t.notOk(is.lt('abc', 'abc'), 'abc is not lesser than abc');
  t.notOk(is.lt('abc', 'a'), 'abc is not lesser than a');
  t.notOk(is.lt(Infinity, 0), 'infinity is not lesser than anything');
  t.notOk(is.lt(0, Infinity), 'anything is not lesser than infinity');
  var error = new TypeError('NaN is not a valid value');
  t['throws'](function () ***REMOVED*** return is.lt(NaN, 2); ***REMOVED***, error, 'throws when first value is NaN');
  t['throws'](function () ***REMOVED*** return is.lt(2, NaN); ***REMOVED***, error, 'throws when second value is NaN');
  t.end();
***REMOVED***);

test('is.within', function (t) ***REMOVED***
  t.test('throws on NaN', function (st) ***REMOVED***
    var nanError = new TypeError('NaN is not a valid value');
    st['throws'](function () ***REMOVED*** return is.within(NaN, 0, 0); ***REMOVED***, nanError, 'throws when first value is NaN');
    st['throws'](function () ***REMOVED*** return is.within(0, NaN, 0); ***REMOVED***, nanError, 'throws when second value is NaN');
    st['throws'](function () ***REMOVED*** return is.within(0, 0, NaN); ***REMOVED***, nanError, 'throws when third value is NaN');
    st.end();
  ***REMOVED***);

  t.test('throws on non-number', function (st) ***REMOVED***
    var error = new TypeError('all arguments must be numbers');
    st['throws'](function () ***REMOVED*** return is.within('', 0, 0); ***REMOVED***, error, 'throws when first value is string');
    st['throws'](function () ***REMOVED*** return is.within(0, '', 0); ***REMOVED***, error, 'throws when second value is string');
    st['throws'](function () ***REMOVED*** return is.within(0, 0, ''); ***REMOVED***, error, 'throws when third value is string');
    st['throws'](function () ***REMOVED*** return is.within(***REMOVED******REMOVED***, 0, 0); ***REMOVED***, error, 'throws when first value is object');
    st['throws'](function () ***REMOVED*** return is.within(0, ***REMOVED******REMOVED***, 0); ***REMOVED***, error, 'throws when second value is object');
    st['throws'](function () ***REMOVED*** return is.within(0, 0, ***REMOVED******REMOVED***); ***REMOVED***, error, 'throws when third value is object');
    st['throws'](function () ***REMOVED*** return is.within(null, 0, 0); ***REMOVED***, error, 'throws when first value is null');
    st['throws'](function () ***REMOVED*** return is.within(0, null, 0); ***REMOVED***, error, 'throws when second value is null');
    st['throws'](function () ***REMOVED*** return is.within(0, 0, null); ***REMOVED***, error, 'throws when third value is null');
    st['throws'](function () ***REMOVED*** return is.within(undefined, 0, 0); ***REMOVED***, error, 'throws when first value is undefined');
    st['throws'](function () ***REMOVED*** return is.within(0, undefined, 0); ***REMOVED***, error, 'throws when second value is undefined');
    st['throws'](function () ***REMOVED*** return is.within(0, 0, undefined); ***REMOVED***, error, 'throws when third value is undefined');
    st.end();
  ***REMOVED***);

  t.ok(is.within(2, 1, 3), '2 is between 1 and 3');
  t.ok(is.within(0, -1, 1), '0 is between -1 and 1');
  t.ok(is.within(2, 0, Infinity), 'infinity always returns true');
  t.ok(is.within(2, Infinity, 2), 'infinity always returns true');
  t.ok(is.within(Infinity, 0, 1), 'infinity always returns true');
  t.notOk(is.within(2, -1, -1), '2 is not between -1 and 1');
  t.end();
***REMOVED***);

test('is.object', function (t) ***REMOVED***
  t.ok(is.object(***REMOVED******REMOVED***), 'object literal is object');
  t.notOk(is.object(), 'undefined is not an object');
  t.notOk(is.object(null), 'null is not an object');
  t.notOk(is.object(true), 'true is not an object');
  t.notOk(is.object(''), 'string is not an object');
  t.notOk(is.object(NaN), 'NaN is not an object');
  t.notOk(is.object(Object), 'object constructor is not an object');
  t.notOk(is.object(function () ***REMOVED******REMOVED***), 'function is not an object');

  t.test('Symbols', ***REMOVED*** skip: typeof Symbol !== 'function' ***REMOVED***, function (st) ***REMOVED***
    st.notOk(is.object(Symbol('foo')), 'symbol is not an object');
    st.end();
  ***REMOVED***);

  t.end();
***REMOVED***);

test('is.primitive', function (t) ***REMOVED***
  t.notOk(is.primitive(***REMOVED******REMOVED***), 'object literal is not a primitive');
  t.notOk(is.primitive([]), 'array is not a primitive');
  t.ok(is.primitive(), 'undefined is a primitive');
  t.ok(is.primitive(null), 'null is a primitive');
  t.ok(is.primitive(true), 'true is a primitive');
  t.ok(is.primitive(''), 'string is a primitive');
  t.ok(is.primitive(NaN), 'NaN is a primitive');
  t.notOk(is.primitive(Object), 'object constructor is not a primitive');
  t.notOk(is.primitive(function () ***REMOVED******REMOVED***), 'function is not a primitive');

  t.test('Symbols', ***REMOVED*** skip: typeof Symbol !== 'function' ***REMOVED***, function (st) ***REMOVED***
    st.ok(is.primitive(Symbol('foo')), 'symbol is a primitive');
    st.end();
  ***REMOVED***);

  t.end();
***REMOVED***);

test('is.hash', function (t) ***REMOVED***
  t.ok(is.hash(***REMOVED******REMOVED***), 'empty object literal is hash');
  t.ok(is.hash(***REMOVED*** 1: 2, a: 'b' ***REMOVED***), 'object literal is hash');
  t.notOk(is.hash(), 'undefined is not a hash');
  t.notOk(is.hash(null), 'null is not a hash');
  t.notOk(is.hash(new Date()), 'date is not a hash');
  t.notOk(is.hash(Object('')), 'string object is not a hash');
  t.notOk(is.hash(''), 'string literal is not a hash');
  t.notOk(is.hash(Object(0)), 'number object is not a hash');
  t.notOk(is.hash(1), 'number literal is not a hash');
  t.notOk(is.hash(true), 'true is not a hash');
  t.notOk(is.hash(false), 'false is not a hash');
  t.notOk(is.hash(Object(false)), 'boolean obj is not hash');
  t.notOk(is.hash(false), 'literal false is not hash');
  t.notOk(is.hash(true), 'literal true is not hash');

  t.test('commonJS environment', ***REMOVED*** skip: typeof module === 'undefined' ***REMOVED***, function (st) ***REMOVED***
    st.ok(is.hash(module.exports), 'module.exports is a hash');
    st.end();
  ***REMOVED***);

  t.test('browser stuff', ***REMOVED*** skip: typeof window === 'undefined' ***REMOVED***, function (st) ***REMOVED***
    st.notOk(is.hash(window), 'window is not a hash');
    st.notOk(is.hash(document.createElement('div')), 'element is not a hash');
    st.end();
  ***REMOVED***);

  t.test('node stuff', ***REMOVED*** skip: typeof process === 'undefined' ***REMOVED***, function (st) ***REMOVED***
    st.notOk(is.hash(global), 'global is not a hash');
    st.notOk(is.hash(process), 'process is not a hash');
    st.end();
  ***REMOVED***);

  t.end();
***REMOVED***);

test('is.regexp', function (t) ***REMOVED***
  t.ok(is.regexp(/a/g), 'regex literal is regex');
  t.ok(is.regexp(new RegExp('a', 'g')), 'regex object is regex');
  t.notOk(is.regexp(), 'undefined is not regex');
  t.notOk(is.regexp(function () ***REMOVED******REMOVED***), 'function is not regex');
  t.notOk(is.regexp('/a/g'), 'string regex is not regex');
  t.end();
***REMOVED***);

test('is.string', function (t) ***REMOVED***
  t.ok(is.string('foo'), 'string literal is string');
  t.ok(is.string(Object('foo')), 'string object is string');
  t.notOk(is.string(), 'undefined is not string');
  t.notOk(is.string(String), 'string constructor is not string');
  var F = function () ***REMOVED******REMOVED***;
  F.prototype = Object('');
  t.notOk(is.string(F), 'string subtype is not string');
  t.end();
***REMOVED***);

test('is.base64', function (t) ***REMOVED***
  t.ok(is.base64('wxyzWXYZ/+=='), 'string is base64 encoded');
  t.ok(is.base64(''), 'zero length string is base64 encoded');
  t.notOk(is.base64('wxyzWXYZ123/+=='), 'string length not a multiple of four is not base64 encoded');
  t.notOk(is.base64('wxyzWXYZ1234|]=='), 'string with invalid characters is not base64 encoded');
  t.notOk(is.base64('wxyzWXYZ1234==/+'), 'string with = not at end is not base64 encoded');
  t.notOk(is.base64('wxyzWXYZ1234/==='), 'string ending with === is not base64 encoded');
  t.end();
***REMOVED***);

test('is.hex', function (t) ***REMOVED***
  t.ok(is.hex('abcdABCD1234'), 'string is hex encoded');
  t.ok(is.hex(''), 'zero length string is hex encoded');
  t.notOk(is.hex('wxyzWXYZ1234/+=='), 'string with invalid characters is not hex encoded');
  t.end();
***REMOVED***);

test('is.symbol', function (t) ***REMOVED***
  t.test('not symbols', function (st) ***REMOVED***
    var notSymbols = [true, false, null, undefined, ***REMOVED******REMOVED***, [], function () ***REMOVED******REMOVED***, 42, NaN, Infinity, /a/g, '', 0, -0, new Error('error')];
    forEach(notSymbols, function (notSymbol) ***REMOVED***
      st.notOk(is.symbol(notSymbol), notSymbol + ' is not symbol');
    ***REMOVED***);

    st.end();
  ***REMOVED***);

  t.test('symbols', ***REMOVED*** skip: typeof Symbol !== 'function' ***REMOVED***, function (st) ***REMOVED***
    st.ok(is.symbol(Symbol('foo')), 'Symbol("foo") is symbol');

    var notKnownSymbols = ['length', 'name', 'arguments', 'caller', 'prototype', 'for', 'keyFor'];
    var symbolKeys = Object.getOwnPropertyNames(Symbol).filter(function (name) ***REMOVED***
      return notKnownSymbols.indexOf(name) < 0;
    ***REMOVED***);
    forEach(symbolKeys, function (symbolKey) ***REMOVED***
      st.ok(is.symbol(Symbol[symbolKey]), symbolKey + ' is symbol');
    ***REMOVED***);

    st.end();
  ***REMOVED***);

  t.end();
***REMOVED***);
