// y18n/index.js
var fs = require('fs')
var path = require('path')
var util = require('util')

function Y18N (opts) ***REMOVED***
  // configurable options.
  opts = opts || ***REMOVED******REMOVED***
  this.directory = opts.directory || './locales'
  this.updateFiles = typeof opts.updateFiles === 'boolean' ? opts.updateFiles : true
  this.locale = opts.locale || 'en'
  this.fallbackToLanguage = typeof opts.fallbackToLanguage === 'boolean' ? opts.fallbackToLanguage : true

  // internal stuff.
  this.cache = ***REMOVED******REMOVED***
  this.writeQueue = []
***REMOVED***

Y18N.prototype.__ = function () ***REMOVED***
  var args = Array.prototype.slice.call(arguments)
  var str = args.shift()
  var cb = function () ***REMOVED******REMOVED*** // start with noop.

  if (typeof args[args.length - 1] === 'function') cb = args.pop()
  cb = cb || function () ***REMOVED******REMOVED*** // noop.

  if (!this.cache[this.locale]) this._readLocaleFile()

  // we've observed a new string, update the language file.
  if (!this.cache[this.locale][str] && this.updateFiles) ***REMOVED***
    this.cache[this.locale][str] = str

    // include the current directory and locale,
    // since these values could change before the
    // write is performed.
    this._enqueueWrite([this.directory, this.locale, cb])
  ***REMOVED*** else ***REMOVED***
    cb()
  ***REMOVED***

  return util.format.apply(util, [this.cache[this.locale][str] || str].concat(args))
***REMOVED***

Y18N.prototype._enqueueWrite = function (work) ***REMOVED***
  this.writeQueue.push(work)
  if (this.writeQueue.length === 1) this._processWriteQueue()
***REMOVED***

Y18N.prototype._processWriteQueue = function () ***REMOVED***
  var _this = this
  var work = this.writeQueue[0]

  // destructure the enqueued work.
  var directory = work[0]
  var locale = work[1]
  var cb = work[2]

  var languageFile = this._resolveLocaleFile(directory, locale)
  var serializedLocale = JSON.stringify(this.cache[locale], null, 2)

  fs.writeFile(languageFile, serializedLocale, 'utf-8', function (err) ***REMOVED***
    _this.writeQueue.shift()
    if (_this.writeQueue.length > 0) _this._processWriteQueue()
    cb(err)
  ***REMOVED***)
***REMOVED***

Y18N.prototype._readLocaleFile = function () ***REMOVED***
  var localeLookup = ***REMOVED******REMOVED***
  var languageFile = this._resolveLocaleFile(this.directory, this.locale)

  try ***REMOVED***
    localeLookup = JSON.parse(fs.readFileSync(languageFile, 'utf-8'))
  ***REMOVED*** catch (err) ***REMOVED***
    if (err instanceof SyntaxError) ***REMOVED***
      err.message = 'syntax error in ' + languageFile
    ***REMOVED***

    if (err.code === 'ENOENT') localeLookup = ***REMOVED******REMOVED***
    else throw err
  ***REMOVED***

  this.cache[this.locale] = localeLookup
***REMOVED***

Y18N.prototype._resolveLocaleFile = function (directory, locale) ***REMOVED***
  var file = path.resolve(directory, './', locale + '.json')
  if (this.fallbackToLanguage && !this._fileExistsSync(file) && ~locale.lastIndexOf('_')) ***REMOVED***
    // attempt fallback to language only
    var languageFile = path.resolve(directory, './', locale.split('_')[0] + '.json')
    if (this._fileExistsSync(languageFile)) file = languageFile
  ***REMOVED***
  return file
***REMOVED***

// this only exists because fs.existsSync() "will be deprecated"
// see https://nodejs.org/api/fs.html#fs_fs_existssync_path
Y18N.prototype._fileExistsSync = function (file) ***REMOVED***
  try ***REMOVED***
    return fs.statSync(file).isFile()
  ***REMOVED*** catch (err) ***REMOVED***
    return false
  ***REMOVED***
***REMOVED***

Y18N.prototype.__n = function () ***REMOVED***
  var args = Array.prototype.slice.call(arguments)
  var singular = args.shift()
  var plural = args.shift()
  var quantity = args.shift()

  var cb = function () ***REMOVED******REMOVED*** // start with noop.
  if (typeof args[args.length - 1] === 'function') cb = args.pop()

  if (!this.cache[this.locale]) this._readLocaleFile()

  var str = quantity === 1 ? singular : plural
  if (this.cache[this.locale][singular]) ***REMOVED***
    str = this.cache[this.locale][singular][quantity === 1 ? 'one' : 'other']
  ***REMOVED***

  // we've observed a new string, update the language file.
  if (!this.cache[this.locale][singular] && this.updateFiles) ***REMOVED***
    this.cache[this.locale][singular] = ***REMOVED***
      one: singular,
      other: plural
    ***REMOVED***

    // include the current directory and locale,
    // since these values could change before the
    // write is performed.
    this._enqueueWrite([this.directory, this.locale, cb])
  ***REMOVED*** else ***REMOVED***
    cb()
  ***REMOVED***

  // if a %d placeholder is provided, add quantity
  // to the arguments expanded by util.format.
  var values = [str]
  if (~str.indexOf('%d')) values.push(quantity)

  return util.format.apply(util, values.concat(args))
***REMOVED***

Y18N.prototype.setLocale = function (locale) ***REMOVED***
  this.locale = locale
***REMOVED***

Y18N.prototype.getLocale = function () ***REMOVED***
  return this.locale
***REMOVED***

Y18N.prototype.updateLocale = function (obj) ***REMOVED***
  if (!this.cache[this.locale]) this._readLocaleFile()

  for (var key in obj) ***REMOVED***
    this.cache[this.locale][key] = obj[key]
  ***REMOVED***
***REMOVED***

module.exports = function (opts) ***REMOVED***
  var y18n = new Y18N(opts)

  // bind all functions to y18n, so that
  // they can be used in isolation.
  for (var key in y18n) ***REMOVED***
    if (typeof y18n[key] === 'function') ***REMOVED***
      y18n[key] = y18n[key].bind(y18n)
    ***REMOVED***
  ***REMOVED***

  return y18n
***REMOVED***
