// @firebase/database/dist/esm/src/core/util/EventEmitter.js
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import ***REMOVED*** assert ***REMOVED*** from '@firebase/util';
/**
 * Base class to be used if you want to emit events. Call the constructor with
 * the set of allowed event names.
 */
var EventEmitter = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***!Array.<string>***REMOVED*** allowedEvents_
     */
    function EventEmitter(allowedEvents_) ***REMOVED***
        this.allowedEvents_ = allowedEvents_;
        this.listeners_ = ***REMOVED******REMOVED***;
        assert(Array.isArray(allowedEvents_) && allowedEvents_.length > 0, 'Requires a non-empty array');
    ***REMOVED***
    /**
     * To be called by derived classes to trigger events.
     * @param ***REMOVED***!string***REMOVED*** eventType
     * @param ***REMOVED***...****REMOVED*** var_args
     */
    EventEmitter.prototype.trigger = function (eventType) ***REMOVED***
        var var_args = [];
        for (var _i = 1; _i < arguments.length; _i++) ***REMOVED***
            var_args[_i - 1] = arguments[_i];
        ***REMOVED***
        if (Array.isArray(this.listeners_[eventType])) ***REMOVED***
            // Clone the list, since callbacks could add/remove listeners.
            var listeners = this.listeners_[eventType].slice();
            for (var i = 0; i < listeners.length; i++) ***REMOVED***
                listeners[i].callback.apply(listeners[i].context, var_args);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    EventEmitter.prototype.on = function (eventType, callback, context) ***REMOVED***
        this.validateEventType_(eventType);
        this.listeners_[eventType] = this.listeners_[eventType] || [];
        this.listeners_[eventType].push(***REMOVED*** callback: callback, context: context ***REMOVED***);
        var eventData = this.getInitialEvent(eventType);
        if (eventData) ***REMOVED***
            callback.apply(context, eventData);
        ***REMOVED***
    ***REMOVED***;
    EventEmitter.prototype.off = function (eventType, callback, context) ***REMOVED***
        this.validateEventType_(eventType);
        var listeners = this.listeners_[eventType] || [];
        for (var i = 0; i < listeners.length; i++) ***REMOVED***
            if (listeners[i].callback === callback &&
                (!context || context === listeners[i].context)) ***REMOVED***
                listeners.splice(i, 1);
                return;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    EventEmitter.prototype.validateEventType_ = function (eventType) ***REMOVED***
        assert(this.allowedEvents_.find(function (et) ***REMOVED***
            return et === eventType;
        ***REMOVED***), 'Unknown event: ' + eventType);
    ***REMOVED***;
    return EventEmitter;
***REMOVED***());
export ***REMOVED*** EventEmitter ***REMOVED***;
