// @firebase/database/dist/cjs/src/realtime/TransportManager.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var BrowserPollConnection_1 = require("./BrowserPollConnection");
var WebSocketConnection_1 = require("./WebSocketConnection");
var util_1 = require("../core/util/util");
/**
 * Currently simplistic, this class manages what transport a Connection should use at various stages of its
 * lifecycle.
 *
 * It starts with longpolling in a browser, and httppolling on node. It then upgrades to websockets if
 * they are available.
 * @constructor
 */
var TransportManager = /** @class */ (function () ***REMOVED***
    /**
     * @param ***REMOVED***!RepoInfo***REMOVED*** repoInfo Metadata around the namespace we're connecting to
     */
    function TransportManager(repoInfo) ***REMOVED***
        this.initTransports_(repoInfo);
    ***REMOVED***
    Object.defineProperty(TransportManager, "ALL_TRANSPORTS", ***REMOVED***
        /**
         * @const
         * @type ***REMOVED***!Array.<function(new:Transport, string, RepoInfo, string=)>***REMOVED***
         */
        get: function () ***REMOVED***
            return [BrowserPollConnection_1.BrowserPollConnection, WebSocketConnection_1.WebSocketConnection];
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    /**
     * @param ***REMOVED***!RepoInfo***REMOVED*** repoInfo
     * @private
     */
    TransportManager.prototype.initTransports_ = function (repoInfo) ***REMOVED***
        var isWebSocketsAvailable = WebSocketConnection_1.WebSocketConnection && WebSocketConnection_1.WebSocketConnection['isAvailable']();
        var isSkipPollConnection = isWebSocketsAvailable && !WebSocketConnection_1.WebSocketConnection.previouslyFailed();
        if (repoInfo.webSocketOnly) ***REMOVED***
            if (!isWebSocketsAvailable)
                util_1.warn("wss:// URL used, but browser isn't known to support websockets.  Trying anyway.");
            isSkipPollConnection = true;
        ***REMOVED***
        if (isSkipPollConnection) ***REMOVED***
            this.transports_ = [WebSocketConnection_1.WebSocketConnection];
        ***REMOVED***
        else ***REMOVED***
            var transports_1 = (this.transports_ = []);
            util_1.each(TransportManager.ALL_TRANSPORTS, function (i, transport) ***REMOVED***
                if (transport && transport['isAvailable']()) ***REMOVED***
                    transports_1.push(transport);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
    ***REMOVED***;
    /**
     * @return ***REMOVED***function(new:Transport, !string, !RepoInfo, string=, string=)***REMOVED*** The constructor for the
     * initial transport to use
     */
    TransportManager.prototype.initialTransport = function () ***REMOVED***
        if (this.transports_.length > 0) ***REMOVED***
            return this.transports_[0];
        ***REMOVED***
        else ***REMOVED***
            throw new Error('No transports available');
        ***REMOVED***
    ***REMOVED***;
    /**
     * @return ***REMOVED***?function(new:Transport, function(),function(), string=)***REMOVED*** The constructor for the next
     * transport, or null
     */
    TransportManager.prototype.upgradeTransport = function () ***REMOVED***
        if (this.transports_.length > 1) ***REMOVED***
            return this.transports_[1];
        ***REMOVED***
        else ***REMOVED***
            return null;
        ***REMOVED***
    ***REMOVED***;
    return TransportManager;
***REMOVED***());
exports.TransportManager = TransportManager;
