// @firebase/database/dist/cjs/src/core/view/Change.js
"use strict";
/**
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
/**
 * @constructor
 * @struct
 * @param ***REMOVED***!string***REMOVED*** type The event type
 * @param ***REMOVED***!Node***REMOVED*** snapshotNode The data
 * @param ***REMOVED***string=***REMOVED*** childName The name for this child, if it's a child event
 * @param ***REMOVED***Node=***REMOVED*** oldSnap Used for intermediate processing of child changed events
 * @param ***REMOVED***string=***REMOVED*** prevName The name for the previous child, if applicable
 */
var Change = /** @class */ (function () ***REMOVED***
    function Change(type, snapshotNode, childName, oldSnap, prevName) ***REMOVED***
        this.type = type;
        this.snapshotNode = snapshotNode;
        this.childName = childName;
        this.oldSnap = oldSnap;
        this.prevName = prevName;
    ***REMOVED***
    /**
     * @param ***REMOVED***!Node***REMOVED*** snapshot
     * @return ***REMOVED***!Change***REMOVED***
     */
    Change.valueChange = function (snapshot) ***REMOVED***
        return new Change(Change.VALUE, snapshot);
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** childKey
     * @param ***REMOVED***!Node***REMOVED*** snapshot
     * @return ***REMOVED***!Change***REMOVED***
     */
    Change.childAddedChange = function (childKey, snapshot) ***REMOVED***
        return new Change(Change.CHILD_ADDED, snapshot, childKey);
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** childKey
     * @param ***REMOVED***!Node***REMOVED*** snapshot
     * @return ***REMOVED***!Change***REMOVED***
     */
    Change.childRemovedChange = function (childKey, snapshot) ***REMOVED***
        return new Change(Change.CHILD_REMOVED, snapshot, childKey);
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** childKey
     * @param ***REMOVED***!Node***REMOVED*** newSnapshot
     * @param ***REMOVED***!Node***REMOVED*** oldSnapshot
     * @return ***REMOVED***!Change***REMOVED***
     */
    Change.childChangedChange = function (childKey, newSnapshot, oldSnapshot) ***REMOVED***
        return new Change(Change.CHILD_CHANGED, newSnapshot, childKey, oldSnapshot);
    ***REMOVED***;
    /**
     * @param ***REMOVED***string***REMOVED*** childKey
     * @param ***REMOVED***!Node***REMOVED*** snapshot
     * @return ***REMOVED***!Change***REMOVED***
     */
    Change.childMovedChange = function (childKey, snapshot) ***REMOVED***
        return new Change(Change.CHILD_MOVED, snapshot, childKey);
    ***REMOVED***;
    //event types
    /** Event type for a child added */
    Change.CHILD_ADDED = 'child_added';
    /** Event type for a child removed */
    Change.CHILD_REMOVED = 'child_removed';
    /** Event type for a child changed */
    Change.CHILD_CHANGED = 'child_changed';
    /** Event type for a child moved */
    Change.CHILD_MOVED = 'child_moved';
    /** Event type for a value change */
    Change.VALUE = 'value';
    return Change;
***REMOVED***());
exports.Change = Change;
