// firebase-admin/lib/auth/user-record.js
/*! firebase-admin v5.8.2 */
"use strict";
/*!
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var deep_copy_1 = require("../utils/deep-copy");
var utils = require("../utils");
var error_1 = require("../utils/error");
/**
 * Parses a time stamp string or number and returns the corresponding date if valid.
 *
 * @param ***REMOVED***any***REMOVED*** time The unix timestamp string or number in milliseconds.
 * @return ***REMOVED***string***REMOVED*** The corresponding date as a UTC string, if valid.
 */
function parseDate(time) ***REMOVED***
    try ***REMOVED***
        var date = new Date(parseInt(time, 10));
        if (!isNaN(date.getTime())) ***REMOVED***
            return date.toUTCString();
        ***REMOVED***
    ***REMOVED***
    catch (e) ***REMOVED***
        // Do nothing. null will be returned.
    ***REMOVED***
    return null;
***REMOVED***
/**
 * User metadata class that provides metadata information like user account creation
 * and last sign in time.
 *
 * @param ***REMOVED***Object***REMOVED*** response The server side response returned from the getAccountInfo
 *     endpoint.
 * @constructor
 */
var UserMetadata = /** @class */ (function () ***REMOVED***
    function UserMetadata(response) ***REMOVED***
        // Creation date should always be available but due to some backend bugs there
        // were cases in the past where users did not have creation date properly set.
        // This included legacy Firebase migrating project users and some anonymous users.
        // These bugs have already been addressed since then.
        utils.addReadonlyGetter(this, 'creationTime', parseDate(response.createdAt));
        utils.addReadonlyGetter(this, 'lastSignInTime', parseDate(response.lastLoginAt));
    ***REMOVED***
    /** @return ***REMOVED***Object***REMOVED*** The plain object representation of the user's metadata. */
    UserMetadata.prototype.toJSON = function () ***REMOVED***
        return ***REMOVED***
            lastSignInTime: this.lastSignInTime,
            creationTime: this.creationTime,
        ***REMOVED***;
    ***REMOVED***;
    return UserMetadata;
***REMOVED***());
exports.UserMetadata = UserMetadata;
/**
 * User info class that provides provider user information for different
 * Firebase providers like google.com, facebook.com, password, etc.
 *
 * @param ***REMOVED***Object***REMOVED*** response The server side response returned from the getAccountInfo
 *     endpoint.
 * @constructor
 */
var UserInfo = /** @class */ (function () ***REMOVED***
    function UserInfo(response) ***REMOVED***
        // Provider user id and provider id are required.
        if (!response.rawId || !response.providerId) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INTERNAL_ERROR, 'INTERNAL ASSERT FAILED: Invalid user info response');
        ***REMOVED***
        utils.addReadonlyGetter(this, 'uid', response.rawId);
        utils.addReadonlyGetter(this, 'displayName', response.displayName);
        utils.addReadonlyGetter(this, 'email', response.email);
        utils.addReadonlyGetter(this, 'photoURL', response.photoUrl);
        utils.addReadonlyGetter(this, 'providerId', response.providerId);
        utils.addReadonlyGetter(this, 'phoneNumber', response.phoneNumber);
    ***REMOVED***
    /** @return ***REMOVED***Object***REMOVED*** The plain object representation of the current provider data. */
    UserInfo.prototype.toJSON = function () ***REMOVED***
        return ***REMOVED***
            uid: this.uid,
            displayName: this.displayName,
            email: this.email,
            photoURL: this.photoURL,
            providerId: this.providerId,
            phoneNumber: this.phoneNumber,
        ***REMOVED***;
    ***REMOVED***;
    return UserInfo;
***REMOVED***());
exports.UserInfo = UserInfo;
/**
 * User record class that defines the Firebase user object populated from
 * the Firebase Auth getAccountInfo response.
 *
 * @param ***REMOVED***Object***REMOVED*** response The server side response returned from the getAccountInfo
 *     endpoint.
 * @constructor
 */
var UserRecord = /** @class */ (function () ***REMOVED***
    function UserRecord(response) ***REMOVED***
        // The Firebase user id is required.
        if (!response.localId) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INTERNAL_ERROR, 'INTERNAL ASSERT FAILED: Invalid user response');
        ***REMOVED***
        utils.addReadonlyGetter(this, 'uid', response.localId);
        utils.addReadonlyGetter(this, 'email', response.email);
        utils.addReadonlyGetter(this, 'emailVerified', !!response.emailVerified);
        utils.addReadonlyGetter(this, 'displayName', response.displayName);
        utils.addReadonlyGetter(this, 'photoURL', response.photoUrl);
        utils.addReadonlyGetter(this, 'phoneNumber', response.phoneNumber);
        // If disabled is not provided, the account is enabled by default.
        utils.addReadonlyGetter(this, 'disabled', response.disabled || false);
        utils.addReadonlyGetter(this, 'metadata', new UserMetadata(response));
        var providerData = [];
        for (var _i = 0, _a = (response.providerUserInfo || []); _i < _a.length; _i++) ***REMOVED***
            var entry = _a[_i];
            providerData.push(new UserInfo(entry));
        ***REMOVED***
        utils.addReadonlyGetter(this, 'providerData', providerData);
        utils.addReadonlyGetter(this, 'passwordHash', response.passwordHash);
        utils.addReadonlyGetter(this, 'passwordSalt', response.salt);
        try ***REMOVED***
            utils.addReadonlyGetter(this, 'customClaims', JSON.parse(response.customAttributes));
        ***REMOVED***
        catch (e) ***REMOVED***
            // Ignore error.
            utils.addReadonlyGetter(this, 'customClaims', undefined);
        ***REMOVED***
        var validAfterTime = null;
        // Convert validSince first to UTC milliseconds and then to UTC date string.
        if (typeof response.validSince !== 'undefined') ***REMOVED***
            validAfterTime = parseDate(response.validSince * 1000);
        ***REMOVED***
        utils.addReadonlyGetter(this, 'tokensValidAfterTime', validAfterTime);
    ***REMOVED***
    /** @return ***REMOVED***Object***REMOVED*** The plain object representation of the user record. */
    UserRecord.prototype.toJSON = function () ***REMOVED***
        var json = ***REMOVED***
            uid: this.uid,
            email: this.email,
            emailVerified: this.emailVerified,
            displayName: this.displayName,
            photoURL: this.photoURL,
            phoneNumber: this.phoneNumber,
            disabled: this.disabled,
            // Convert metadata to json.
            metadata: this.metadata.toJSON(),
            passwordHash: this.passwordHash,
            passwordSalt: this.passwordSalt,
            customClaims: deep_copy_1.deepCopy(this.customClaims),
            tokensValidAfterTime: this.tokensValidAfterTime,
        ***REMOVED***;
        json.providerData = [];
        for (var _i = 0, _a = this.providerData; _i < _a.length; _i++) ***REMOVED***
            var entry = _a[_i];
            // Convert each provider data to json.
            json.providerData.push(entry.toJSON());
        ***REMOVED***
        return json;
    ***REMOVED***;
    return UserRecord;
***REMOVED***());
exports.UserRecord = UserRecord;
