// websocket-driver/lib/websocket/driver/hybi.js
'use strict';

var crypto     = require('crypto'),
    util       = require('util'),
    Extensions = require('websocket-extensions'),
    Base       = require('./base'),
    Frame      = require('./hybi/frame'),
    Message    = require('./hybi/message');

var Hybi = function(request, url, options) ***REMOVED***
  Base.apply(this, arguments);

  this._extensions     = new Extensions();
  this._stage          = 0;
  this._masking        = this._options.masking;
  this._protocols      = this._options.protocols || [];
  this._requireMasking = this._options.requireMasking;
  this._pingCallbacks  = ***REMOVED******REMOVED***;

  if (typeof this._protocols === 'string')
    this._protocols = this._protocols.split(/ *, */);

  if (!this._request) return;

  var secKey    = this._request.headers['sec-websocket-key'],
      protos    = this._request.headers['sec-websocket-protocol'],
      version   = this._request.headers['sec-websocket-version'],
      supported = this._protocols;

  this._headers.set('Upgrade', 'websocket');
  this._headers.set('Connection', 'Upgrade');
  this._headers.set('Sec-WebSocket-Accept', Hybi.generateAccept(secKey));

  if (protos !== undefined) ***REMOVED***
    if (typeof protos === 'string') protos = protos.split(/ *, */);
    this.protocol = protos.filter(function(p) ***REMOVED*** return supported.indexOf(p) >= 0 ***REMOVED***)[0];
    if (this.protocol) this._headers.set('Sec-WebSocket-Protocol', this.protocol);
  ***REMOVED***

  this.version = 'hybi-' + version;
***REMOVED***;
util.inherits(Hybi, Base);

Hybi.mask = function(payload, mask, offset) ***REMOVED***
  if (!mask || mask.length === 0) return payload;
  offset = offset || 0;

  for (var i = 0, n = payload.length - offset; i < n; i++) ***REMOVED***
    payload[offset + i] = payload[offset + i] ^ mask[i % 4];
  ***REMOVED***
  return payload;
***REMOVED***;

Hybi.generateAccept = function(key) ***REMOVED***
  var sha1 = crypto.createHash('sha1');
  sha1.update(key + Hybi.GUID);
  return sha1.digest('base64');
***REMOVED***;

Hybi.GUID = '258EAFA5-E914-47DA-95CA-C5AB0DC85B11';

var instance = ***REMOVED***
  FIN:    0x80,
  MASK:   0x80,
  RSV1:   0x40,
  RSV2:   0x20,
  RSV3:   0x10,
  OPCODE: 0x0F,
  LENGTH: 0x7F,

  OPCODES: ***REMOVED***
    continuation: 0,
    text:         1,
    binary:       2,
    close:        8,
    ping:         9,
    pong:         10
  ***REMOVED***,

  OPCODE_CODES:    [0, 1, 2, 8, 9, 10],
  MESSAGE_OPCODES: [0, 1, 2],
  OPENING_OPCODES: [1, 2],

  ERRORS: ***REMOVED***
    normal_closure:       1000,
    going_away:           1001,
    protocol_error:       1002,
    unacceptable:         1003,
    encoding_error:       1007,
    policy_violation:     1008,
    too_large:            1009,
    extension_error:      1010,
    unexpected_condition: 1011
  ***REMOVED***,

  ERROR_CODES:        [1000, 1001, 1002, 1003, 1007, 1008, 1009, 1010, 1011],
  DEFAULT_ERROR_CODE: 1000,
  MIN_RESERVED_ERROR: 3000,
  MAX_RESERVED_ERROR: 4999,

  // http://www.w3.org/International/questions/qa-forms-utf-8.en.php
  UTF8_MATCH: /^([\x00-\x7F]|[\xC2-\xDF][\x80-\xBF]|\xE0[\xA0-\xBF][\x80-\xBF]|[\xE1-\xEC\xEE\xEF][\x80-\xBF]***REMOVED***2***REMOVED***|\xED[\x80-\x9F][\x80-\xBF]|\xF0[\x90-\xBF][\x80-\xBF]***REMOVED***2***REMOVED***|[\xF1-\xF3][\x80-\xBF]***REMOVED***3***REMOVED***|\xF4[\x80-\x8F][\x80-\xBF]***REMOVED***2***REMOVED***)*$/,

  addExtension: function(extension) ***REMOVED***
    this._extensions.add(extension);
    return true;
  ***REMOVED***,

  parse: function(chunk) ***REMOVED***
    this._reader.put(chunk);
    var buffer = true;
    while (buffer) ***REMOVED***
      switch (this._stage) ***REMOVED***
        case 0:
          buffer = this._reader.read(1);
          if (buffer) this._parseOpcode(buffer[0]);
          break;

        case 1:
          buffer = this._reader.read(1);
          if (buffer) this._parseLength(buffer[0]);
          break;

        case 2:
          buffer = this._reader.read(this._frame.lengthBytes);
          if (buffer) this._parseExtendedLength(buffer);
          break;

        case 3:
          buffer = this._reader.read(4);
          if (buffer) ***REMOVED***
            this._stage = 4;
            this._frame.maskingKey = buffer;
          ***REMOVED***
          break;

        case 4:
          buffer = this._reader.read(this._frame.length);
          if (buffer) ***REMOVED***
            this._stage = 0;
            this._emitFrame(buffer);
          ***REMOVED***
          break;

        default:
          buffer = null;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***,

  text: function(message) ***REMOVED***
    if (this.readyState > 1) return false;
    return this.frame(message, 'text');
  ***REMOVED***,

  binary: function(message) ***REMOVED***
    if (this.readyState > 1) return false;
    return this.frame(message, 'binary');
  ***REMOVED***,

  ping: function(message, callback) ***REMOVED***
    if (this.readyState > 1) return false;
    message = message || '';
    if (callback) this._pingCallbacks[message] = callback;
    return this.frame(message, 'ping');
  ***REMOVED***,

  pong: function(message) ***REMOVED***
      if (this.readyState > 1) return false;
      message = message ||'';
      return this.frame(message, 'pong');
  ***REMOVED***,

  close: function(reason, code) ***REMOVED***
    reason = reason || '';
    code   = code   || this.ERRORS.normal_closure;

    if (this.readyState <= 0) ***REMOVED***
      this.readyState = 3;
      this.emit('close', new Base.CloseEvent(code, reason));
      return true;
    ***REMOVED*** else if (this.readyState === 1) ***REMOVED***
      this.readyState = 2;
      this._extensions.close(function() ***REMOVED*** this.frame(reason, 'close', code) ***REMOVED***, this);
      return true;
    ***REMOVED*** else ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***,

  frame: function(buffer, type, code) ***REMOVED***
    if (this.readyState <= 0) return this._queue([buffer, type, code]);
    if (this.readyState > 2) return false;

    if (buffer instanceof Array)    buffer = new Buffer(buffer);
    if (typeof buffer === 'number') buffer = buffer.toString();

    var message = new Message(),
        isText  = (typeof buffer === 'string'),
        payload, copy;

    message.rsv1   = message.rsv2 = message.rsv3 = false;
    message.opcode = this.OPCODES[type || (isText ? 'text' : 'binary')];

    payload = isText ? new Buffer(buffer, 'utf8') : buffer;

    if (code) ***REMOVED***
      copy = payload;
      payload = new Buffer(2 + copy.length);
      payload.writeUInt16BE(code, 0);
      copy.copy(payload, 2);
    ***REMOVED***
    message.data = payload;

    var onMessageReady = function(message) ***REMOVED***
      var frame = new Frame();

      frame.final   = true;
      frame.rsv1    = message.rsv1;
      frame.rsv2    = message.rsv2;
      frame.rsv3    = message.rsv3;
      frame.opcode  = message.opcode;
      frame.masked  = !!this._masking;
      frame.length  = message.data.length;
      frame.payload = message.data;

      if (frame.masked) frame.maskingKey = crypto.randomBytes(4);

      this._sendFrame(frame);
    ***REMOVED***;

    if (this.MESSAGE_OPCODES.indexOf(message.opcode) >= 0)
      this._extensions.processOutgoingMessage(message, function(error, message) ***REMOVED***
        if (error) return this._fail('extension_error', error.message);
        onMessageReady.call(this, message);
      ***REMOVED***, this);
    else
      onMessageReady.call(this, message);

    return true;
  ***REMOVED***,

  _sendFrame: function(frame) ***REMOVED***
    var length = frame.length,
        header = (length <= 125) ? 2 : (length <= 65535 ? 4 : 10),
        offset = header + (frame.masked ? 4 : 0),
        buffer = new Buffer(offset + length),
        masked = frame.masked ? this.MASK : 0;

    buffer[0] = (frame.final ? this.FIN : 0) |
                (frame.rsv1 ? this.RSV1 : 0) |
                (frame.rsv2 ? this.RSV2 : 0) |
                (frame.rsv3 ? this.RSV3 : 0) |
                frame.opcode;

    if (length <= 125) ***REMOVED***
      buffer[1] = masked | length;
    ***REMOVED*** else if (length <= 65535) ***REMOVED***
      buffer[1] = masked | 126;
      buffer.writeUInt16BE(length, 2);
    ***REMOVED*** else ***REMOVED***
      buffer[1] = masked | 127;
      buffer.writeUInt32BE(Math.floor(length / 0x100000000), 2);
      buffer.writeUInt32BE(length % 0x100000000, 6);
    ***REMOVED***

    frame.payload.copy(buffer, offset);

    if (frame.masked) ***REMOVED***
      frame.maskingKey.copy(buffer, header);
      Hybi.mask(buffer, frame.maskingKey, offset);
    ***REMOVED***

    this._write(buffer);
  ***REMOVED***,

  _handshakeResponse: function() ***REMOVED***
    try ***REMOVED***
      var extensions = this._extensions.generateResponse(this._request.headers['sec-websocket-extensions']);
    ***REMOVED*** catch (e) ***REMOVED***
      return this._fail('protocol_error', e.message);
    ***REMOVED***

    if (extensions) this._headers.set('Sec-WebSocket-Extensions', extensions);

    var start   = 'HTTP/1.1 101 Switching Protocols',
        headers = [start, this._headers.toString(), ''];

    return new Buffer(headers.join('\r\n'), 'utf8');
  ***REMOVED***,

  _shutdown: function(code, reason, error) ***REMOVED***
    delete this._frame;
    delete this._message;
    this._stage = 5;

    var sendCloseFrame = (this.readyState === 1);
    this.readyState = 2;

    this._extensions.close(function() ***REMOVED***
      if (sendCloseFrame) this.frame(reason, 'close', code);
      this.readyState = 3;
      if (error) this.emit('error', new Error(reason));
      this.emit('close', new Base.CloseEvent(code, reason));
    ***REMOVED***, this);
  ***REMOVED***,

  _fail: function(type, message) ***REMOVED***
    if (this.readyState > 1) return;
    this._shutdown(this.ERRORS[type], message, true);
  ***REMOVED***,

  _parseOpcode: function(octet) ***REMOVED***
    var rsvs = [this.RSV1, this.RSV2, this.RSV3].map(function(rsv) ***REMOVED***
      return (octet & rsv) === rsv;
    ***REMOVED***);

    var frame = this._frame = new Frame();

    frame.final  = (octet & this.FIN) === this.FIN;
    frame.rsv1   = rsvs[0];
    frame.rsv2   = rsvs[1];
    frame.rsv3   = rsvs[2];
    frame.opcode = (octet & this.OPCODE);

    this._stage = 1;

    if (!this._extensions.validFrameRsv(frame))
      return this._fail('protocol_error',
          'One or more reserved bits are on: reserved1 = ' + (frame.rsv1 ? 1 : 0) +
          ', reserved2 = ' + (frame.rsv2 ? 1 : 0) +
          ', reserved3 = ' + (frame.rsv3 ? 1 : 0));

    if (this.OPCODE_CODES.indexOf(frame.opcode) < 0)
      return this._fail('protocol_error', 'Unrecognized frame opcode: ' + frame.opcode);

    if (this.MESSAGE_OPCODES.indexOf(frame.opcode) < 0 && !frame.final)
      return this._fail('protocol_error', 'Received fragmented control frame: opcode = ' + frame.opcode);

    if (this._message && this.OPENING_OPCODES.indexOf(frame.opcode) >= 0)
      return this._fail('protocol_error', 'Received new data frame but previous continuous frame is unfinished');
  ***REMOVED***,

  _parseLength: function(octet) ***REMOVED***
    var frame = this._frame;
    frame.masked = (octet & this.MASK) === this.MASK;
    frame.length = (octet & this.LENGTH);

    if (frame.length >= 0 && frame.length <= 125) ***REMOVED***
      this._stage = frame.masked ? 3 : 4;
      if (!this._checkFrameLength()) return;
    ***REMOVED*** else ***REMOVED***
      this._stage = 2;
      frame.lengthBytes = (frame.length === 126 ? 2 : 8);
    ***REMOVED***

    if (this._requireMasking && !frame.masked)
      return this._fail('unacceptable', 'Received unmasked frame but masking is required');
  ***REMOVED***,

  _parseExtendedLength: function(buffer) ***REMOVED***
    var frame = this._frame;
    frame.length = this._readUInt(buffer);

    this._stage = frame.masked ? 3 : 4;

    if (this.MESSAGE_OPCODES.indexOf(frame.opcode) < 0 && frame.length > 125)
      return this._fail('protocol_error', 'Received control frame having too long payload: ' + frame.length);

    if (!this._checkFrameLength()) return;
  ***REMOVED***,

  _checkFrameLength: function() ***REMOVED***
    var length = this._message ? this._message.length : 0;

    if (length + this._frame.length > this._maxLength) ***REMOVED***
      this._fail('too_large', 'WebSocket frame length too large');
      return false;
    ***REMOVED*** else ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***,

  _emitFrame: function(buffer) ***REMOVED***
    var frame   = this._frame,
        payload = frame.payload = Hybi.mask(buffer, frame.maskingKey),
        opcode  = frame.opcode,
        message,
        code, reason,
        callbacks, callback;

    delete this._frame;

    if (opcode === this.OPCODES.continuation) ***REMOVED***
      if (!this._message) return this._fail('protocol_error', 'Received unexpected continuation frame');
      this._message.pushFrame(frame);
    ***REMOVED***

    if (opcode === this.OPCODES.text || opcode === this.OPCODES.binary) ***REMOVED***
      this._message = new Message();
      this._message.pushFrame(frame);
    ***REMOVED***

    if (frame.final && this.MESSAGE_OPCODES.indexOf(opcode) >= 0)
      return this._emitMessage(this._message);

    if (opcode === this.OPCODES.close) ***REMOVED***
      code   = (payload.length >= 2) ? payload.readUInt16BE(0) : null;
      reason = (payload.length > 2) ? this._encode(payload.slice(2)) : null;

      if (!(payload.length === 0) &&
          !(code !== null && code >= this.MIN_RESERVED_ERROR && code <= this.MAX_RESERVED_ERROR) &&
          this.ERROR_CODES.indexOf(code) < 0)
        code = this.ERRORS.protocol_error;

      if (payload.length > 125 || (payload.length > 2 && !reason))
        code = this.ERRORS.protocol_error;

      this._shutdown(code || this.DEFAULT_ERROR_CODE, reason || '');
    ***REMOVED***

    if (opcode === this.OPCODES.ping) ***REMOVED***
      this.frame(payload, 'pong');
      this.emit('ping', new Base.PingEvent(payload.toString()))
    ***REMOVED***

    if (opcode === this.OPCODES.pong) ***REMOVED***
      callbacks = this._pingCallbacks;
      message   = this._encode(payload);
      callback  = callbacks[message];

      delete callbacks[message];
      if (callback) callback()

      this.emit('pong', new Base.PongEvent(payload.toString()))
    ***REMOVED***
  ***REMOVED***,

  _emitMessage: function(message) ***REMOVED***
    var message = this._message;
    message.read();

    delete this._message;

    this._extensions.processIncomingMessage(message, function(error, message) ***REMOVED***
      if (error) return this._fail('extension_error', error.message);

      var payload = message.data;
      if (message.opcode === this.OPCODES.text) payload = this._encode(payload);

      if (payload === null)
        return this._fail('encoding_error', 'Could not decode a text frame as UTF-8');
      else
        this.emit('message', new Base.MessageEvent(payload));
    ***REMOVED***, this);
  ***REMOVED***,

  _encode: function(buffer) ***REMOVED***
    try ***REMOVED***
      var string = buffer.toString('binary', 0, buffer.length);
      if (!this.UTF8_MATCH.test(string)) return null;
    ***REMOVED*** catch (e) ***REMOVED******REMOVED***
    return buffer.toString('utf8', 0, buffer.length);
  ***REMOVED***,

  _readUInt: function(buffer) ***REMOVED***
    if (buffer.length === 2) return buffer.readUInt16BE(0);

    return buffer.readUInt32BE(0) * 0x100000000 +
           buffer.readUInt32BE(4);
  ***REMOVED***
***REMOVED***;

for (var key in instance)
  Hybi.prototype[key] = instance[key];

module.exports = Hybi;
