// websocket-driver/lib/websocket/driver/server.js
'use strict';

var util       = require('util'),
    HttpParser = require('../http_parser'),
    Base       = require('./base'),
    Draft75    = require('./draft75'),
    Draft76    = require('./draft76'),
    Hybi       = require('./hybi');

var Server = function(options) ***REMOVED***
  Base.call(this, null, null, options);
  this._http = new HttpParser('request');
***REMOVED***;
util.inherits(Server, Base);

var instance = ***REMOVED***
  EVENTS: ['open', 'message', 'error', 'close'],

  _bindEventListeners: function() ***REMOVED***
    this.messages.on('error', function() ***REMOVED******REMOVED***);
    this.on('error', function() ***REMOVED******REMOVED***);
  ***REMOVED***,

  parse: function(chunk) ***REMOVED***
    if (this._delegate) return this._delegate.parse(chunk);

    this._http.parse(chunk);
    if (!this._http.isComplete()) return;

    this.method  = this._http.method;
    this.url     = this._http.url;
    this.headers = this._http.headers;
    this.body    = this._http.body;

    var self = this;
    this._delegate = Server.http(this, this._options);
    this._delegate.messages = this.messages;
    this._delegate.io = this.io;
    this._open();

    this.EVENTS.forEach(function(event) ***REMOVED***
      this._delegate.on(event, function(e) ***REMOVED*** self.emit(event, e) ***REMOVED***);
    ***REMOVED***, this);

    this.protocol = this._delegate.protocol;
    this.version  = this._delegate.version;

    this.parse(this._http.body);
    this.emit('connect', new Base.ConnectEvent());
  ***REMOVED***,

  _open: function() ***REMOVED***
    this.__queue.forEach(function(msg) ***REMOVED***
      this._delegate[msg[0]].apply(this._delegate, msg[1]);
    ***REMOVED***, this);
    this.__queue = [];
  ***REMOVED***
***REMOVED***;

['addExtension', 'setHeader', 'start', 'frame', 'text', 'binary', 'ping', 'close'].forEach(function(method) ***REMOVED***
  instance[method] = function() ***REMOVED***
    if (this._delegate) ***REMOVED***
      return this._delegate[method].apply(this._delegate, arguments);
    ***REMOVED*** else ***REMOVED***
      this.__queue.push([method, arguments]);
      return true;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***);

for (var key in instance)
  Server.prototype[key] = instance[key];

Server.isSecureRequest = function(request) ***REMOVED***
  if (request.connection && request.connection.authorized !== undefined) return true;
  if (request.socket && request.socket.secure) return true;

  var headers = request.headers;
  if (!headers) return false;
  if (headers['https'] === 'on') return true;
  if (headers['x-forwarded-ssl'] === 'on') return true;
  if (headers['x-forwarded-scheme'] === 'https') return true;
  if (headers['x-forwarded-proto'] === 'https') return true;

  return false;
***REMOVED***;

Server.determineUrl = function(request) ***REMOVED***
  var scheme = this.isSecureRequest(request) ? 'wss:' : 'ws:';
  return scheme + '//' + request.headers.host + request.url;
***REMOVED***;

Server.http = function(request, options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;
  if (options.requireMasking === undefined) options.requireMasking = true;

  var headers = request.headers,
      url     = this.determineUrl(request);

  if (headers['sec-websocket-version'])
    return new Hybi(request, url, options);
  else if (headers['sec-websocket-key1'])
    return new Draft76(request, url, options);
  else
    return new Draft75(request, url, options);
***REMOVED***;

module.exports = Server;
