// big.js/big.js
/* big.js v3.1.3 https://github.com/MikeMcl/big.js/LICENCE */
;(function (global) ***REMOVED***
    'use strict';

/*
  big.js v3.1.3
  A small, fast, easy-to-use library for arbitrary-precision decimal arithmetic.
  https://github.com/MikeMcl/big.js/
  Copyright (c) 2014 Michael Mclaughlin <M8ch88l@gmail.com>
  MIT Expat Licence
*/

/***************************** EDITABLE DEFAULTS ******************************/

    // The default values below must be integers within the stated ranges.

    /*
     * The maximum number of decimal places of the results of operations
     * involving division: div and sqrt, and pow with negative exponents.
     */
    var DP = 20,                           // 0 to MAX_DP

        /*
         * The rounding mode used when rounding to the above decimal places.
         *
         * 0 Towards zero (i.e. truncate, no rounding).       (ROUND_DOWN)
         * 1 To nearest neighbour. If equidistant, round up.  (ROUND_HALF_UP)
         * 2 To nearest neighbour. If equidistant, to even.   (ROUND_HALF_EVEN)
         * 3 Away from zero.                                  (ROUND_UP)
         */
        RM = 1,                            // 0, 1, 2 or 3

        // The maximum value of DP and Big.DP.
        MAX_DP = 1E6,                      // 0 to 1000000

        // The maximum magnitude of the exponent argument to the pow method.
        MAX_POWER = 1E6,                   // 1 to 1000000

        /*
         * The exponent value at and beneath which toString returns exponential
         * notation.
         * JavaScript's Number type: -7
         * -1000000 is the minimum recommended exponent value of a Big.
         */
        E_NEG = -7,                   // 0 to -1000000

        /*
         * The exponent value at and above which toString returns exponential
         * notation.
         * JavaScript's Number type: 21
         * 1000000 is the maximum recommended exponent value of a Big.
         * (This limit is not enforced or checked.)
         */
        E_POS = 21,                   // 0 to 1000000

/******************************************************************************/

        // The shared prototype object.
        P = ***REMOVED******REMOVED***,
        isValid = /^-?(\d+(\.\d*)?|\.\d+)(e[+-]?\d+)?$/i,
        Big;


    /*
     * Create and return a Big constructor.
     *
     */
    function bigFactory() ***REMOVED***

        /*
         * The Big constructor and exported function.
         * Create and return a new instance of a Big number object.
         *
         * n ***REMOVED***number|string|Big***REMOVED*** A numeric value.
         */
        function Big(n) ***REMOVED***
            var x = this;

            // Enable constructor usage without new.
            if (!(x instanceof Big)) ***REMOVED***
                return n === void 0 ? bigFactory() : new Big(n);
            ***REMOVED***

            // Duplicate.
            if (n instanceof Big) ***REMOVED***
                x.s = n.s;
                x.e = n.e;
                x.c = n.c.slice();
            ***REMOVED*** else ***REMOVED***
                parse(x, n);
            ***REMOVED***

            /*
             * Retain a reference to this Big constructor, and shadow
             * Big.prototype.constructor which points to Object.
             */
            x.constructor = Big;
        ***REMOVED***

        Big.prototype = P;
        Big.DP = DP;
        Big.RM = RM;
        Big.E_NEG = E_NEG;
        Big.E_POS = E_POS;

        return Big;
    ***REMOVED***


    // Private functions


    /*
     * Return a string representing the value of Big x in normal or exponential
     * notation to dp fixed decimal places or significant digits.
     *
     * x ***REMOVED***Big***REMOVED*** The Big to format.
     * dp ***REMOVED***number***REMOVED*** Integer, 0 to MAX_DP inclusive.
     * toE ***REMOVED***number***REMOVED*** 1 (toExponential), 2 (toPrecision) or undefined (toFixed).
     */
    function format(x, dp, toE) ***REMOVED***
        var Big = x.constructor,

            // The index (normal notation) of the digit that may be rounded up.
            i = dp - (x = new Big(x)).e,
            c = x.c;

        // Round?
        if (c.length > ++dp) ***REMOVED***
            rnd(x, i, Big.RM);
        ***REMOVED***

        if (!c[0]) ***REMOVED***
            ++i;
        ***REMOVED*** else if (toE) ***REMOVED***
            i = dp;

        // toFixed
        ***REMOVED*** else ***REMOVED***
            c = x.c;

            // Recalculate i as x.e may have changed if value rounded up.
            i = x.e + i + 1;
        ***REMOVED***

        // Append zeros?
        for (; c.length < i; c.push(0)) ***REMOVED***
        ***REMOVED***
        i = x.e;

        /*
         * toPrecision returns exponential notation if the number of
         * significant digits specified is less than the number of digits
         * necessary to represent the integer part of the value in normal
         * notation.
         */
        return toE === 1 || toE && (dp <= i || i <= Big.E_NEG) ?

          // Exponential notation.
          (x.s < 0 && c[0] ? '-' : '') +
            (c.length > 1 ? c[0] + '.' + c.join('').slice(1) : c[0]) +
              (i < 0 ? 'e' : 'e+') + i

          // Normal notation.
          : x.toString();
    ***REMOVED***


    /*
     * Parse the number or string value passed to a Big constructor.
     *
     * x ***REMOVED***Big***REMOVED*** A Big number instance.
     * n ***REMOVED***number|string***REMOVED*** A numeric value.
     */
    function parse(x, n) ***REMOVED***
        var e, i, nL;

        // Minus zero?
        if (n === 0 && 1 / n < 0) ***REMOVED***
            n = '-0';

        // Ensure n is string and check validity.
        ***REMOVED*** else if (!isValid.test(n += '')) ***REMOVED***
            throwErr(NaN);
        ***REMOVED***

        // Determine sign.
        x.s = n.charAt(0) == '-' ? (n = n.slice(1), -1) : 1;

        // Decimal point?
        if ((e = n.indexOf('.')) > -1) ***REMOVED***
            n = n.replace('.', '');
        ***REMOVED***

        // Exponential form?
        if ((i = n.search(/e/i)) > 0) ***REMOVED***

            // Determine exponent.
            if (e < 0) ***REMOVED***
                e = i;
            ***REMOVED***
            e += +n.slice(i + 1);
            n = n.substring(0, i);

        ***REMOVED*** else if (e < 0) ***REMOVED***

            // Integer.
            e = n.length;
        ***REMOVED***

        nL = n.length;

        // Determine leading zeros.
        for (i = 0; i < nL && n.charAt(i) == '0'; i++) ***REMOVED***
        ***REMOVED***

        if (i == nL) ***REMOVED***

            // Zero.
            x.c = [ x.e = 0 ];
        ***REMOVED*** else ***REMOVED***

            // Determine trailing zeros.
            for (; nL > 0 && n.charAt(--nL) == '0';) ***REMOVED***
            ***REMOVED***

            x.e = e - i - 1;
            x.c = [];

            // Convert string to array of digits without leading/trailing zeros.
            //for (e = 0; i <= nL; x.c[e++] = +n.charAt(i++)) ***REMOVED***
            for (; i <= nL; x.c.push(+n.charAt(i++))) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return x;
    ***REMOVED***


    /*
     * Round Big x to a maximum of dp decimal places using rounding mode rm.
     * Called by div, sqrt and round.
     *
     * x ***REMOVED***Big***REMOVED*** The Big to round.
     * dp ***REMOVED***number***REMOVED*** Integer, 0 to MAX_DP inclusive.
     * rm ***REMOVED***number***REMOVED*** 0, 1, 2 or 3 (DOWN, HALF_UP, HALF_EVEN, UP)
     * [more] ***REMOVED***boolean***REMOVED*** Whether the result of division was truncated.
     */
    function rnd(x, dp, rm, more) ***REMOVED***
        var u,
            xc = x.c,
            i = x.e + dp + 1;

        if (rm === 1) ***REMOVED***

            // xc[i] is the digit after the digit that may be rounded up.
            more = xc[i] >= 5;
        ***REMOVED*** else if (rm === 2) ***REMOVED***
            more = xc[i] > 5 || xc[i] == 5 &&
              (more || i < 0 || xc[i + 1] !== u || xc[i - 1] & 1);
        ***REMOVED*** else if (rm === 3) ***REMOVED***
            more = more || xc[i] !== u || i < 0;
        ***REMOVED*** else ***REMOVED***
            more = false;

            if (rm !== 0) ***REMOVED***
                throwErr('!Big.RM!');
            ***REMOVED***
        ***REMOVED***

        if (i < 1 || !xc[0]) ***REMOVED***

            if (more) ***REMOVED***

                // 1, 0.1, 0.01, 0.001, 0.0001 etc.
                x.e = -dp;
                x.c = [1];
            ***REMOVED*** else ***REMOVED***

                // Zero.
                x.c = [x.e = 0];
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***

            // Remove any digits after the required decimal places.
            xc.length = i--;

            // Round up?
            if (more) ***REMOVED***

                // Rounding up may mean the previous digit has to be rounded up.
                for (; ++xc[i] > 9;) ***REMOVED***
                    xc[i] = 0;

                    if (!i--) ***REMOVED***
                        ++x.e;
                        xc.unshift(1);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***

            // Remove trailing zeros.
            for (i = xc.length; !xc[--i]; xc.pop()) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return x;
    ***REMOVED***


    /*
     * Throw a BigError.
     *
     * message ***REMOVED***string***REMOVED*** The error message.
     */
    function throwErr(message) ***REMOVED***
        var err = new Error(message);
        err.name = 'BigError';

        throw err;
    ***REMOVED***


    // Prototype/instance methods


    /*
     * Return a new Big whose value is the absolute value of this Big.
     */
    P.abs = function () ***REMOVED***
        var x = new this.constructor(this);
        x.s = 1;

        return x;
    ***REMOVED***;


    /*
     * Return
     * 1 if the value of this Big is greater than the value of Big y,
     * -1 if the value of this Big is less than the value of Big y, or
     * 0 if they have the same value.
    */
    P.cmp = function (y) ***REMOVED***
        var xNeg,
            x = this,
            xc = x.c,
            yc = (y = new x.constructor(y)).c,
            i = x.s,
            j = y.s,
            k = x.e,
            l = y.e;

        // Either zero?
        if (!xc[0] || !yc[0]) ***REMOVED***
            return !xc[0] ? !yc[0] ? 0 : -j : i;
        ***REMOVED***

        // Signs differ?
        if (i != j) ***REMOVED***
            return i;
        ***REMOVED***
        xNeg = i < 0;

        // Compare exponents.
        if (k != l) ***REMOVED***
            return k > l ^ xNeg ? 1 : -1;
        ***REMOVED***

        i = -1;
        j = (k = xc.length) < (l = yc.length) ? k : l;

        // Compare digit by digit.
        for (; ++i < j;) ***REMOVED***

            if (xc[i] != yc[i]) ***REMOVED***
                return xc[i] > yc[i] ^ xNeg ? 1 : -1;
            ***REMOVED***
        ***REMOVED***

        // Compare lengths.
        return k == l ? 0 : k > l ^ xNeg ? 1 : -1;
    ***REMOVED***;


    /*
     * Return a new Big whose value is the value of this Big divided by the
     * value of Big y, rounded, if necessary, to a maximum of Big.DP decimal
     * places using rounding mode Big.RM.
     */
    P.div = function (y) ***REMOVED***
        var x = this,
            Big = x.constructor,
            // dividend
            dvd = x.c,
            //divisor
            dvs = (y = new Big(y)).c,
            s = x.s == y.s ? 1 : -1,
            dp = Big.DP;

        if (dp !== ~~dp || dp < 0 || dp > MAX_DP) ***REMOVED***
            throwErr('!Big.DP!');
        ***REMOVED***

        // Either 0?
        if (!dvd[0] || !dvs[0]) ***REMOVED***

            // If both are 0, throw NaN
            if (dvd[0] == dvs[0]) ***REMOVED***
                throwErr(NaN);
            ***REMOVED***

            // If dvs is 0, throw +-Infinity.
            if (!dvs[0]) ***REMOVED***
                throwErr(s / 0);
            ***REMOVED***

            // dvd is 0, return +-0.
            return new Big(s * 0);
        ***REMOVED***

        var dvsL, dvsT, next, cmp, remI, u,
            dvsZ = dvs.slice(),
            dvdI = dvsL = dvs.length,
            dvdL = dvd.length,
            // remainder
            rem = dvd.slice(0, dvsL),
            remL = rem.length,
            // quotient
            q = y,
            qc = q.c = [],
            qi = 0,
            digits = dp + (q.e = x.e - y.e) + 1;

        q.s = s;
        s = digits < 0 ? 0 : digits;

        // Create version of divisor with leading zero.
        dvsZ.unshift(0);

        // Add zeros to make remainder as long as divisor.
        for (; remL++ < dvsL; rem.push(0)) ***REMOVED***
        ***REMOVED***

        do ***REMOVED***

            // 'next' is how many times the divisor goes into current remainder.
            for (next = 0; next < 10; next++) ***REMOVED***

                // Compare divisor and remainder.
                if (dvsL != (remL = rem.length)) ***REMOVED***
                    cmp = dvsL > remL ? 1 : -1;
                ***REMOVED*** else ***REMOVED***

                    for (remI = -1, cmp = 0; ++remI < dvsL;) ***REMOVED***

                        if (dvs[remI] != rem[remI]) ***REMOVED***
                            cmp = dvs[remI] > rem[remI] ? 1 : -1;
                            break;
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***

                // If divisor < remainder, subtract divisor from remainder.
                if (cmp < 0) ***REMOVED***

                    // Remainder can't be more than 1 digit longer than divisor.
                    // Equalise lengths using divisor with extra leading zero?
                    for (dvsT = remL == dvsL ? dvs : dvsZ; remL;) ***REMOVED***

                        if (rem[--remL] < dvsT[remL]) ***REMOVED***
                            remI = remL;

                            for (; remI && !rem[--remI]; rem[remI] = 9) ***REMOVED***
                            ***REMOVED***
                            --rem[remI];
                            rem[remL] += 10;
                        ***REMOVED***
                        rem[remL] -= dvsT[remL];
                    ***REMOVED***
                    for (; !rem[0]; rem.shift()) ***REMOVED***
                    ***REMOVED***
                ***REMOVED*** else ***REMOVED***
                    break;
                ***REMOVED***
            ***REMOVED***

            // Add the 'next' digit to the result array.
            qc[qi++] = cmp ? next : ++next;

            // Update the remainder.
            if (rem[0] && cmp) ***REMOVED***
                rem[remL] = dvd[dvdI] || 0;
            ***REMOVED*** else ***REMOVED***
                rem = [ dvd[dvdI] ];
            ***REMOVED***

        ***REMOVED*** while ((dvdI++ < dvdL || rem[0] !== u) && s--);

        // Leading zero? Do not remove if result is simply zero (qi == 1).
        if (!qc[0] && qi != 1) ***REMOVED***

            // There can't be more than one zero.
            qc.shift();
            q.e--;
        ***REMOVED***

        // Round?
        if (qi > digits) ***REMOVED***
            rnd(q, dp, Big.RM, rem[0] !== u);
        ***REMOVED***

        return q;
    ***REMOVED***;


    /*
     * Return true if the value of this Big is equal to the value of Big y,
     * otherwise returns false.
     */
    P.eq = function (y) ***REMOVED***
        return !this.cmp(y);
    ***REMOVED***;


    /*
     * Return true if the value of this Big is greater than the value of Big y,
     * otherwise returns false.
     */
    P.gt = function (y) ***REMOVED***
        return this.cmp(y) > 0;
    ***REMOVED***;


    /*
     * Return true if the value of this Big is greater than or equal to the
     * value of Big y, otherwise returns false.
     */
    P.gte = function (y) ***REMOVED***
        return this.cmp(y) > -1;
    ***REMOVED***;


    /*
     * Return true if the value of this Big is less than the value of Big y,
     * otherwise returns false.
     */
    P.lt = function (y) ***REMOVED***
        return this.cmp(y) < 0;
    ***REMOVED***;


    /*
     * Return true if the value of this Big is less than or equal to the value
     * of Big y, otherwise returns false.
     */
    P.lte = function (y) ***REMOVED***
         return this.cmp(y) < 1;
    ***REMOVED***;


    /*
     * Return a new Big whose value is the value of this Big minus the value
     * of Big y.
     */
    P.sub = P.minus = function (y) ***REMOVED***
        var i, j, t, xLTy,
            x = this,
            Big = x.constructor,
            a = x.s,
            b = (y = new Big(y)).s;

        // Signs differ?
        if (a != b) ***REMOVED***
            y.s = -b;
            return x.plus(y);
        ***REMOVED***

        var xc = x.c.slice(),
            xe = x.e,
            yc = y.c,
            ye = y.e;

        // Either zero?
        if (!xc[0] || !yc[0]) ***REMOVED***

            // y is non-zero? x is non-zero? Or both are zero.
            return yc[0] ? (y.s = -b, y) : new Big(xc[0] ? x : 0);
        ***REMOVED***

        // Determine which is the bigger number.
        // Prepend zeros to equalise exponents.
        if (a = xe - ye) ***REMOVED***

            if (xLTy = a < 0) ***REMOVED***
                a = -a;
                t = xc;
            ***REMOVED*** else ***REMOVED***
                ye = xe;
                t = yc;
            ***REMOVED***

            t.reverse();
            for (b = a; b--; t.push(0)) ***REMOVED***
            ***REMOVED***
            t.reverse();
        ***REMOVED*** else ***REMOVED***

            // Exponents equal. Check digit by digit.
            j = ((xLTy = xc.length < yc.length) ? xc : yc).length;

            for (a = b = 0; b < j; b++) ***REMOVED***

                if (xc[b] != yc[b]) ***REMOVED***
                    xLTy = xc[b] < yc[b];
                    break;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        // x < y? Point xc to the array of the bigger number.
        if (xLTy) ***REMOVED***
            t = xc;
            xc = yc;
            yc = t;
            y.s = -y.s;
        ***REMOVED***

        /*
         * Append zeros to xc if shorter. No need to add zeros to yc if shorter
         * as subtraction only needs to start at yc.length.
         */
        if (( b = (j = yc.length) - (i = xc.length) ) > 0) ***REMOVED***

            for (; b--; xc[i++] = 0) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        // Subtract yc from xc.
        for (b = i; j > a;)***REMOVED***

            if (xc[--j] < yc[j]) ***REMOVED***

                for (i = j; i && !xc[--i]; xc[i] = 9) ***REMOVED***
                ***REMOVED***
                --xc[i];
                xc[j] += 10;
            ***REMOVED***
            xc[j] -= yc[j];
        ***REMOVED***

        // Remove trailing zeros.
        for (; xc[--b] === 0; xc.pop()) ***REMOVED***
        ***REMOVED***

        // Remove leading zeros and adjust exponent accordingly.
        for (; xc[0] === 0;) ***REMOVED***
            xc.shift();
            --ye;
        ***REMOVED***

        if (!xc[0]) ***REMOVED***

            // n - n = +0
            y.s = 1;

            // Result must be zero.
            xc = [ye = 0];
        ***REMOVED***

        y.c = xc;
        y.e = ye;

        return y;
    ***REMOVED***;


    /*
     * Return a new Big whose value is the value of this Big modulo the
     * value of Big y.
     */
    P.mod = function (y) ***REMOVED***
        var yGTx,
            x = this,
            Big = x.constructor,
            a = x.s,
            b = (y = new Big(y)).s;

        if (!y.c[0]) ***REMOVED***
            throwErr(NaN);
        ***REMOVED***

        x.s = y.s = 1;
        yGTx = y.cmp(x) == 1;
        x.s = a;
        y.s = b;

        if (yGTx) ***REMOVED***
            return new Big(x);
        ***REMOVED***

        a = Big.DP;
        b = Big.RM;
        Big.DP = Big.RM = 0;
        x = x.div(y);
        Big.DP = a;
        Big.RM = b;

        return this.minus( x.times(y) );
    ***REMOVED***;


    /*
     * Return a new Big whose value is the value of this Big plus the value
     * of Big y.
     */
    P.add = P.plus = function (y) ***REMOVED***
        var t,
            x = this,
            Big = x.constructor,
            a = x.s,
            b = (y = new Big(y)).s;

        // Signs differ?
        if (a != b) ***REMOVED***
            y.s = -b;
            return x.minus(y);
        ***REMOVED***

        var xe = x.e,
            xc = x.c,
            ye = y.e,
            yc = y.c;

        // Either zero?
        if (!xc[0] || !yc[0]) ***REMOVED***

            // y is non-zero? x is non-zero? Or both are zero.
            return yc[0] ? y : new Big(xc[0] ? x : a * 0);
        ***REMOVED***
        xc = xc.slice();

        // Prepend zeros to equalise exponents.
        // Note: Faster to use reverse then do unshifts.
        if (a = xe - ye) ***REMOVED***

            if (a > 0) ***REMOVED***
                ye = xe;
                t = yc;
            ***REMOVED*** else ***REMOVED***
                a = -a;
                t = xc;
            ***REMOVED***

            t.reverse();
            for (; a--; t.push(0)) ***REMOVED***
            ***REMOVED***
            t.reverse();
        ***REMOVED***

        // Point xc to the longer array.
        if (xc.length - yc.length < 0) ***REMOVED***
            t = yc;
            yc = xc;
            xc = t;
        ***REMOVED***
        a = yc.length;

        /*
         * Only start adding at yc.length - 1 as the further digits of xc can be
         * left as they are.
         */
        for (b = 0; a;) ***REMOVED***
            b = (xc[--a] = xc[a] + yc[a] + b) / 10 | 0;
            xc[a] %= 10;
        ***REMOVED***

        // No need to check for zero, as +x + +y != 0 && -x + -y != 0

        if (b) ***REMOVED***
            xc.unshift(b);
            ++ye;
        ***REMOVED***

         // Remove trailing zeros.
        for (a = xc.length; xc[--a] === 0; xc.pop()) ***REMOVED***
        ***REMOVED***

        y.c = xc;
        y.e = ye;

        return y;
    ***REMOVED***;


    /*
     * Return a Big whose value is the value of this Big raised to the power n.
     * If n is negative, round, if necessary, to a maximum of Big.DP decimal
     * places using rounding mode Big.RM.
     *
     * n ***REMOVED***number***REMOVED*** Integer, -MAX_POWER to MAX_POWER inclusive.
     */
    P.pow = function (n) ***REMOVED***
        var x = this,
            one = new x.constructor(1),
            y = one,
            isNeg = n < 0;

        if (n !== ~~n || n < -MAX_POWER || n > MAX_POWER) ***REMOVED***
            throwErr('!pow!');
        ***REMOVED***

        n = isNeg ? -n : n;

        for (;;) ***REMOVED***

            if (n & 1) ***REMOVED***
                y = y.times(x);
            ***REMOVED***
            n >>= 1;

            if (!n) ***REMOVED***
                break;
            ***REMOVED***
            x = x.times(x);
        ***REMOVED***

        return isNeg ? one.div(y) : y;
    ***REMOVED***;


    /*
     * Return a new Big whose value is the value of this Big rounded to a
     * maximum of dp decimal places using rounding mode rm.
     * If dp is not specified, round to 0 decimal places.
     * If rm is not specified, use Big.RM.
     *
     * [dp] ***REMOVED***number***REMOVED*** Integer, 0 to MAX_DP inclusive.
     * [rm] 0, 1, 2 or 3 (ROUND_DOWN, ROUND_HALF_UP, ROUND_HALF_EVEN, ROUND_UP)
     */
    P.round = function (dp, rm) ***REMOVED***
        var x = this,
            Big = x.constructor;

        if (dp == null) ***REMOVED***
            dp = 0;
        ***REMOVED*** else if (dp !== ~~dp || dp < 0 || dp > MAX_DP) ***REMOVED***
            throwErr('!round!');
        ***REMOVED***
        rnd(x = new Big(x), dp, rm == null ? Big.RM : rm);

        return x;
    ***REMOVED***;


    /*
     * Return a new Big whose value is the square root of the value of this Big,
     * rounded, if necessary, to a maximum of Big.DP decimal places using
     * rounding mode Big.RM.
     */
    P.sqrt = function () ***REMOVED***
        var estimate, r, approx,
            x = this,
            Big = x.constructor,
            xc = x.c,
            i = x.s,
            e = x.e,
            half = new Big('0.5');

        // Zero?
        if (!xc[0]) ***REMOVED***
            return new Big(x);
        ***REMOVED***

        // If negative, throw NaN.
        if (i < 0) ***REMOVED***
            throwErr(NaN);
        ***REMOVED***

        // Estimate.
        i = Math.sqrt(x.toString());

        // Math.sqrt underflow/overflow?
        // Pass x to Math.sqrt as integer, then adjust the result exponent.
        if (i === 0 || i === 1 / 0) ***REMOVED***
            estimate = xc.join('');

            if (!(estimate.length + e & 1)) ***REMOVED***
                estimate += '0';
            ***REMOVED***

            r = new Big( Math.sqrt(estimate).toString() );
            r.e = ((e + 1) / 2 | 0) - (e < 0 || e & 1);
        ***REMOVED*** else ***REMOVED***
            r = new Big(i.toString());
        ***REMOVED***

        i = r.e + (Big.DP += 4);

        // Newton-Raphson iteration.
        do ***REMOVED***
            approx = r;
            r = half.times( approx.plus( x.div(approx) ) );
        ***REMOVED*** while ( approx.c.slice(0, i).join('') !==
                       r.c.slice(0, i).join('') );

        rnd(r, Big.DP -= 4, Big.RM);

        return r;
    ***REMOVED***;


    /*
     * Return a new Big whose value is the value of this Big times the value of
     * Big y.
     */
    P.mul = P.times = function (y) ***REMOVED***
        var c,
            x = this,
            Big = x.constructor,
            xc = x.c,
            yc = (y = new Big(y)).c,
            a = xc.length,
            b = yc.length,
            i = x.e,
            j = y.e;

        // Determine sign of result.
        y.s = x.s == y.s ? 1 : -1;

        // Return signed 0 if either 0.
        if (!xc[0] || !yc[0]) ***REMOVED***
            return new Big(y.s * 0);
        ***REMOVED***

        // Initialise exponent of result as x.e + y.e.
        y.e = i + j;

        // If array xc has fewer digits than yc, swap xc and yc, and lengths.
        if (a < b) ***REMOVED***
            c = xc;
            xc = yc;
            yc = c;
            j = a;
            a = b;
            b = j;
        ***REMOVED***

        // Initialise coefficient array of result with zeros.
        for (c = new Array(j = a + b); j--; c[j] = 0) ***REMOVED***
        ***REMOVED***

        // Multiply.

        // i is initially xc.length.
        for (i = b; i--;) ***REMOVED***
            b = 0;

            // a is yc.length.
            for (j = a + i; j > i;) ***REMOVED***

                // Current sum of products at this digit position, plus carry.
                b = c[j] + yc[i] * xc[j - i - 1] + b;
                c[j--] = b % 10;

                // carry
                b = b / 10 | 0;
            ***REMOVED***
            c[j] = (c[j] + b) % 10;
        ***REMOVED***

        // Increment result exponent if there is a final carry.
        if (b) ***REMOVED***
            ++y.e;
        ***REMOVED***

        // Remove any leading zero.
        if (!c[0]) ***REMOVED***
            c.shift();
        ***REMOVED***

        // Remove trailing zeros.
        for (i = c.length; !c[--i]; c.pop()) ***REMOVED***
        ***REMOVED***
        y.c = c;

        return y;
    ***REMOVED***;


    /*
     * Return a string representing the value of this Big.
     * Return exponential notation if this Big has a positive exponent equal to
     * or greater than Big.E_POS, or a negative exponent equal to or less than
     * Big.E_NEG.
     */
    P.toString = P.valueOf = P.toJSON = function () ***REMOVED***
        var x = this,
            Big = x.constructor,
            e = x.e,
            str = x.c.join(''),
            strL = str.length;

        // Exponential notation?
        if (e <= Big.E_NEG || e >= Big.E_POS) ***REMOVED***
            str = str.charAt(0) + (strL > 1 ? '.' + str.slice(1) : '') +
              (e < 0 ? 'e' : 'e+') + e;

        // Negative exponent?
        ***REMOVED*** else if (e < 0) ***REMOVED***

            // Prepend zeros.
            for (; ++e; str = '0' + str) ***REMOVED***
            ***REMOVED***
            str = '0.' + str;

        // Positive exponent?
        ***REMOVED*** else if (e > 0) ***REMOVED***

            if (++e > strL) ***REMOVED***

                // Append zeros.
                for (e -= strL; e-- ; str += '0') ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else if (e < strL) ***REMOVED***
                str = str.slice(0, e) + '.' + str.slice(e);
            ***REMOVED***

        // Exponent zero.
        ***REMOVED*** else if (strL > 1) ***REMOVED***
            str = str.charAt(0) + '.' + str.slice(1);
        ***REMOVED***

        // Avoid '-0'
        return x.s < 0 && x.c[0] ? '-' + str : str;
    ***REMOVED***;


    /*
     ***************************************************************************
     * If toExponential, toFixed, toPrecision and format are not required they
     * can safely be commented-out or deleted. No redundant code will be left.
     * format is used only by toExponential, toFixed and toPrecision.
     ***************************************************************************
     */


    /*
     * Return a string representing the value of this Big in exponential
     * notation to dp fixed decimal places and rounded, if necessary, using
     * Big.RM.
     *
     * [dp] ***REMOVED***number***REMOVED*** Integer, 0 to MAX_DP inclusive.
     */
    P.toExponential = function (dp) ***REMOVED***

        if (dp == null) ***REMOVED***
            dp = this.c.length - 1;
        ***REMOVED*** else if (dp !== ~~dp || dp < 0 || dp > MAX_DP) ***REMOVED***
            throwErr('!toExp!');
        ***REMOVED***

        return format(this, dp, 1);
    ***REMOVED***;


    /*
     * Return a string representing the value of this Big in normal notation
     * to dp fixed decimal places and rounded, if necessary, using Big.RM.
     *
     * [dp] ***REMOVED***number***REMOVED*** Integer, 0 to MAX_DP inclusive.
     */
    P.toFixed = function (dp) ***REMOVED***
        var str,
            x = this,
            Big = x.constructor,
            neg = Big.E_NEG,
            pos = Big.E_POS;

        // Prevent the possibility of exponential notation.
        Big.E_NEG = -(Big.E_POS = 1 / 0);

        if (dp == null) ***REMOVED***
            str = x.toString();
        ***REMOVED*** else if (dp === ~~dp && dp >= 0 && dp <= MAX_DP) ***REMOVED***
            str = format(x, x.e + dp);

            // (-0).toFixed() is '0', but (-0.1).toFixed() is '-0'.
            // (-0).toFixed(1) is '0.0', but (-0.01).toFixed(1) is '-0.0'.
            if (x.s < 0 && x.c[0] && str.indexOf('-') < 0) ***REMOVED***
        //E.g. -0.5 if rounded to -0 will cause toString to omit the minus sign.
                str = '-' + str;
            ***REMOVED***
        ***REMOVED***
        Big.E_NEG = neg;
        Big.E_POS = pos;

        if (!str) ***REMOVED***
            throwErr('!toFix!');
        ***REMOVED***

        return str;
    ***REMOVED***;


    /*
     * Return a string representing the value of this Big rounded to sd
     * significant digits using Big.RM. Use exponential notation if sd is less
     * than the number of digits necessary to represent the integer part of the
     * value in normal notation.
     *
     * sd ***REMOVED***number***REMOVED*** Integer, 1 to MAX_DP inclusive.
     */
    P.toPrecision = function (sd) ***REMOVED***

        if (sd == null) ***REMOVED***
            return this.toString();
        ***REMOVED*** else if (sd !== ~~sd || sd < 1 || sd > MAX_DP) ***REMOVED***
            throwErr('!toPre!');
        ***REMOVED***

        return format(this, sd - 1, 2);
    ***REMOVED***;


    // Export


    Big = bigFactory();

    //AMD.
    if (typeof define === 'function' && define.amd) ***REMOVED***
        define(function () ***REMOVED***
            return Big;
        ***REMOVED***);

    // Node and other CommonJS-like environments that support module.exports.
    ***REMOVED*** else if (typeof module !== 'undefined' && module.exports) ***REMOVED***
        module.exports = Big;
        module.exports.Big = Big;

    //Browser.
    ***REMOVED*** else ***REMOVED***
        global.Big = Big;
    ***REMOVED***
***REMOVED***)(this);
