// pseudomap/pseudomap.js
var hasOwnProperty = Object.prototype.hasOwnProperty

module.exports = PseudoMap

function PseudoMap (set) ***REMOVED***
  if (!(this instanceof PseudoMap)) // whyyyyyyy
    throw new TypeError("Constructor PseudoMap requires 'new'")

  this.clear()

  if (set) ***REMOVED***
    if ((set instanceof PseudoMap) ||
        (typeof Map === 'function' && set instanceof Map))
      set.forEach(function (value, key) ***REMOVED***
        this.set(key, value)
      ***REMOVED***, this)
    else if (Array.isArray(set))
      set.forEach(function (kv) ***REMOVED***
        this.set(kv[0], kv[1])
      ***REMOVED***, this)
    else
      throw new TypeError('invalid argument')
  ***REMOVED***
***REMOVED***

PseudoMap.prototype.forEach = function (fn, thisp) ***REMOVED***
  thisp = thisp || this
  Object.keys(this._data).forEach(function (k) ***REMOVED***
    if (k !== 'size')
      fn.call(thisp, this._data[k].value, this._data[k].key)
  ***REMOVED***, this)
***REMOVED***

PseudoMap.prototype.has = function (k) ***REMOVED***
  return !!find(this._data, k)
***REMOVED***

PseudoMap.prototype.get = function (k) ***REMOVED***
  var res = find(this._data, k)
  return res && res.value
***REMOVED***

PseudoMap.prototype.set = function (k, v) ***REMOVED***
  set(this._data, k, v)
***REMOVED***

PseudoMap.prototype.delete = function (k) ***REMOVED***
  var res = find(this._data, k)
  if (res) ***REMOVED***
    delete this._data[res._index]
    this._data.size--
  ***REMOVED***
***REMOVED***

PseudoMap.prototype.clear = function () ***REMOVED***
  var data = Object.create(null)
  data.size = 0

  Object.defineProperty(this, '_data', ***REMOVED***
    value: data,
    enumerable: false,
    configurable: true,
    writable: false
  ***REMOVED***)
***REMOVED***

Object.defineProperty(PseudoMap.prototype, 'size', ***REMOVED***
  get: function () ***REMOVED***
    return this._data.size
  ***REMOVED***,
  set: function (n) ***REMOVED******REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***)

PseudoMap.prototype.values =
PseudoMap.prototype.keys =
PseudoMap.prototype.entries = function () ***REMOVED***
  throw new Error('iterators are not implemented in this version')
***REMOVED***

// Either identical, or both NaN
function same (a, b) ***REMOVED***
  return a === b || a !== a && b !== b
***REMOVED***

function Entry (k, v, i) ***REMOVED***
  this.key = k
  this.value = v
  this._index = i
***REMOVED***

function find (data, k) ***REMOVED***
  for (var i = 0, s = '_' + k, key = s;
       hasOwnProperty.call(data, key);
       key = s + i++) ***REMOVED***
    if (same(data[key].key, k))
      return data[key]
  ***REMOVED***
***REMOVED***

function set (data, k, v) ***REMOVED***
  for (var i = 0, s = '_' + k, key = s;
       hasOwnProperty.call(data, key);
       key = s + i++) ***REMOVED***
    if (same(data[key].key, k)) ***REMOVED***
      data[key].value = v
      return
    ***REMOVED***
  ***REMOVED***
  data.size++
  data[key] = new Entry(k, v, key)
***REMOVED***
