// bn.js/lib/bn.js
(function (module, exports) ***REMOVED***
  'use strict';

  // Utils
  function assert (val, msg) ***REMOVED***
    if (!val) throw new Error(msg || 'Assertion failed');
  ***REMOVED***

  // Could use `inherits` module, but don't want to move from single file
  // architecture yet.
  function inherits (ctor, superCtor) ***REMOVED***
    ctor.super_ = superCtor;
    var TempCtor = function () ***REMOVED******REMOVED***;
    TempCtor.prototype = superCtor.prototype;
    ctor.prototype = new TempCtor();
    ctor.prototype.constructor = ctor;
  ***REMOVED***

  // BN

  function BN (number, base, endian) ***REMOVED***
    if (BN.isBN(number)) ***REMOVED***
      return number;
    ***REMOVED***

    this.negative = 0;
    this.words = null;
    this.length = 0;

    // Reduction context
    this.red = null;

    if (number !== null) ***REMOVED***
      if (base === 'le' || base === 'be') ***REMOVED***
        endian = base;
        base = 10;
      ***REMOVED***

      this._init(number || 0, base || 10, endian || 'be');
    ***REMOVED***
  ***REMOVED***
  if (typeof module === 'object') ***REMOVED***
    module.exports = BN;
  ***REMOVED*** else ***REMOVED***
    exports.BN = BN;
  ***REMOVED***

  BN.BN = BN;
  BN.wordSize = 26;

  var Buffer;
  try ***REMOVED***
    Buffer = require('buffer').Buffer;
  ***REMOVED*** catch (e) ***REMOVED***
  ***REMOVED***

  BN.isBN = function isBN (num) ***REMOVED***
    if (num instanceof BN) ***REMOVED***
      return true;
    ***REMOVED***

    return num !== null && typeof num === 'object' &&
      num.constructor.wordSize === BN.wordSize && Array.isArray(num.words);
  ***REMOVED***;

  BN.max = function max (left, right) ***REMOVED***
    if (left.cmp(right) > 0) return left;
    return right;
  ***REMOVED***;

  BN.min = function min (left, right) ***REMOVED***
    if (left.cmp(right) < 0) return left;
    return right;
  ***REMOVED***;

  BN.prototype._init = function init (number, base, endian) ***REMOVED***
    if (typeof number === 'number') ***REMOVED***
      return this._initNumber(number, base, endian);
    ***REMOVED***

    if (typeof number === 'object') ***REMOVED***
      return this._initArray(number, base, endian);
    ***REMOVED***

    if (base === 'hex') ***REMOVED***
      base = 16;
    ***REMOVED***
    assert(base === (base | 0) && base >= 2 && base <= 36);

    number = number.toString().replace(/\s+/g, '');
    var start = 0;
    if (number[0] === '-') ***REMOVED***
      start++;
    ***REMOVED***

    if (base === 16) ***REMOVED***
      this._parseHex(number, start);
    ***REMOVED*** else ***REMOVED***
      this._parseBase(number, base, start);
    ***REMOVED***

    if (number[0] === '-') ***REMOVED***
      this.negative = 1;
    ***REMOVED***

    this.strip();

    if (endian !== 'le') return;

    this._initArray(this.toArray(), base, endian);
  ***REMOVED***;

  BN.prototype._initNumber = function _initNumber (number, base, endian) ***REMOVED***
    if (number < 0) ***REMOVED***
      this.negative = 1;
      number = -number;
    ***REMOVED***
    if (number < 0x4000000) ***REMOVED***
      this.words = [ number & 0x3ffffff ];
      this.length = 1;
    ***REMOVED*** else if (number < 0x10000000000000) ***REMOVED***
      this.words = [
        number & 0x3ffffff,
        (number / 0x4000000) & 0x3ffffff
      ];
      this.length = 2;
    ***REMOVED*** else ***REMOVED***
      assert(number < 0x20000000000000); // 2 ^ 53 (unsafe)
      this.words = [
        number & 0x3ffffff,
        (number / 0x4000000) & 0x3ffffff,
        1
      ];
      this.length = 3;
    ***REMOVED***

    if (endian !== 'le') return;

    // Reverse the bytes
    this._initArray(this.toArray(), base, endian);
  ***REMOVED***;

  BN.prototype._initArray = function _initArray (number, base, endian) ***REMOVED***
    // Perhaps a Uint8Array
    assert(typeof number.length === 'number');
    if (number.length <= 0) ***REMOVED***
      this.words = [ 0 ];
      this.length = 1;
      return this;
    ***REMOVED***

    this.length = Math.ceil(number.length / 3);
    this.words = new Array(this.length);
    for (var i = 0; i < this.length; i++) ***REMOVED***
      this.words[i] = 0;
    ***REMOVED***

    var j, w;
    var off = 0;
    if (endian === 'be') ***REMOVED***
      for (i = number.length - 1, j = 0; i >= 0; i -= 3) ***REMOVED***
        w = number[i] | (number[i - 1] << 8) | (number[i - 2] << 16);
        this.words[j] |= (w << off) & 0x3ffffff;
        this.words[j + 1] = (w >>> (26 - off)) & 0x3ffffff;
        off += 24;
        if (off >= 26) ***REMOVED***
          off -= 26;
          j++;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (endian === 'le') ***REMOVED***
      for (i = 0, j = 0; i < number.length; i += 3) ***REMOVED***
        w = number[i] | (number[i + 1] << 8) | (number[i + 2] << 16);
        this.words[j] |= (w << off) & 0x3ffffff;
        this.words[j + 1] = (w >>> (26 - off)) & 0x3ffffff;
        off += 24;
        if (off >= 26) ***REMOVED***
          off -= 26;
          j++;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
    return this.strip();
  ***REMOVED***;

  function parseHex (str, start, end) ***REMOVED***
    var r = 0;
    var len = Math.min(str.length, end);
    for (var i = start; i < len; i++) ***REMOVED***
      var c = str.charCodeAt(i) - 48;

      r <<= 4;

      // 'a' - 'f'
      if (c >= 49 && c <= 54) ***REMOVED***
        r |= c - 49 + 0xa;

      // 'A' - 'F'
      ***REMOVED*** else if (c >= 17 && c <= 22) ***REMOVED***
        r |= c - 17 + 0xa;

      // '0' - '9'
      ***REMOVED*** else ***REMOVED***
        r |= c & 0xf;
      ***REMOVED***
    ***REMOVED***
    return r;
  ***REMOVED***

  BN.prototype._parseHex = function _parseHex (number, start) ***REMOVED***
    // Create possibly bigger array to ensure that it fits the number
    this.length = Math.ceil((number.length - start) / 6);
    this.words = new Array(this.length);
    for (var i = 0; i < this.length; i++) ***REMOVED***
      this.words[i] = 0;
    ***REMOVED***

    var j, w;
    // Scan 24-bit chunks and add them to the number
    var off = 0;
    for (i = number.length - 6, j = 0; i >= start; i -= 6) ***REMOVED***
      w = parseHex(number, i, i + 6);
      this.words[j] |= (w << off) & 0x3ffffff;
      // NOTE: `0x3fffff` is intentional here, 26bits max shift + 24bit hex limb
      this.words[j + 1] |= w >>> (26 - off) & 0x3fffff;
      off += 24;
      if (off >= 26) ***REMOVED***
        off -= 26;
        j++;
      ***REMOVED***
    ***REMOVED***
    if (i + 6 !== start) ***REMOVED***
      w = parseHex(number, start, i + 6);
      this.words[j] |= (w << off) & 0x3ffffff;
      this.words[j + 1] |= w >>> (26 - off) & 0x3fffff;
    ***REMOVED***
    this.strip();
  ***REMOVED***;

  function parseBase (str, start, end, mul) ***REMOVED***
    var r = 0;
    var len = Math.min(str.length, end);
    for (var i = start; i < len; i++) ***REMOVED***
      var c = str.charCodeAt(i) - 48;

      r *= mul;

      // 'a'
      if (c >= 49) ***REMOVED***
        r += c - 49 + 0xa;

      // 'A'
      ***REMOVED*** else if (c >= 17) ***REMOVED***
        r += c - 17 + 0xa;

      // '0' - '9'
      ***REMOVED*** else ***REMOVED***
        r += c;
      ***REMOVED***
    ***REMOVED***
    return r;
  ***REMOVED***

  BN.prototype._parseBase = function _parseBase (number, base, start) ***REMOVED***
    // Initialize as zero
    this.words = [ 0 ];
    this.length = 1;

    // Find length of limb in base
    for (var limbLen = 0, limbPow = 1; limbPow <= 0x3ffffff; limbPow *= base) ***REMOVED***
      limbLen++;
    ***REMOVED***
    limbLen--;
    limbPow = (limbPow / base) | 0;

    var total = number.length - start;
    var mod = total % limbLen;
    var end = Math.min(total, total - mod) + start;

    var word = 0;
    for (var i = start; i < end; i += limbLen) ***REMOVED***
      word = parseBase(number, i, i + limbLen, base);

      this.imuln(limbPow);
      if (this.words[0] + word < 0x4000000) ***REMOVED***
        this.words[0] += word;
      ***REMOVED*** else ***REMOVED***
        this._iaddn(word);
      ***REMOVED***
    ***REMOVED***

    if (mod !== 0) ***REMOVED***
      var pow = 1;
      word = parseBase(number, i, number.length, base);

      for (i = 0; i < mod; i++) ***REMOVED***
        pow *= base;
      ***REMOVED***

      this.imuln(pow);
      if (this.words[0] + word < 0x4000000) ***REMOVED***
        this.words[0] += word;
      ***REMOVED*** else ***REMOVED***
        this._iaddn(word);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;

  BN.prototype.copy = function copy (dest) ***REMOVED***
    dest.words = new Array(this.length);
    for (var i = 0; i < this.length; i++) ***REMOVED***
      dest.words[i] = this.words[i];
    ***REMOVED***
    dest.length = this.length;
    dest.negative = this.negative;
    dest.red = this.red;
  ***REMOVED***;

  BN.prototype.clone = function clone () ***REMOVED***
    var r = new BN(null);
    this.copy(r);
    return r;
  ***REMOVED***;

  BN.prototype._expand = function _expand (size) ***REMOVED***
    while (this.length < size) ***REMOVED***
      this.words[this.length++] = 0;
    ***REMOVED***
    return this;
  ***REMOVED***;

  // Remove leading `0` from `this`
  BN.prototype.strip = function strip () ***REMOVED***
    while (this.length > 1 && this.words[this.length - 1] === 0) ***REMOVED***
      this.length--;
    ***REMOVED***
    return this._normSign();
  ***REMOVED***;

  BN.prototype._normSign = function _normSign () ***REMOVED***
    // -0 = 0
    if (this.length === 1 && this.words[0] === 0) ***REMOVED***
      this.negative = 0;
    ***REMOVED***
    return this;
  ***REMOVED***;

  BN.prototype.inspect = function inspect () ***REMOVED***
    return (this.red ? '<BN-R: ' : '<BN: ') + this.toString(16) + '>';
  ***REMOVED***;

  /*

  var zeros = [];
  var groupSizes = [];
  var groupBases = [];

  var s = '';
  var i = -1;
  while (++i < BN.wordSize) ***REMOVED***
    zeros[i] = s;
    s += '0';
  ***REMOVED***
  groupSizes[0] = 0;
  groupSizes[1] = 0;
  groupBases[0] = 0;
  groupBases[1] = 0;
  var base = 2 - 1;
  while (++base < 36 + 1) ***REMOVED***
    var groupSize = 0;
    var groupBase = 1;
    while (groupBase < (1 << BN.wordSize) / base) ***REMOVED***
      groupBase *= base;
      groupSize += 1;
    ***REMOVED***
    groupSizes[base] = groupSize;
    groupBases[base] = groupBase;
  ***REMOVED***

  */

  var zeros = [
    '',
    '0',
    '00',
    '000',
    '0000',
    '00000',
    '000000',
    '0000000',
    '00000000',
    '000000000',
    '0000000000',
    '00000000000',
    '000000000000',
    '0000000000000',
    '00000000000000',
    '000000000000000',
    '0000000000000000',
    '00000000000000000',
    '000000000000000000',
    '0000000000000000000',
    '00000000000000000000',
    '000000000000000000000',
    '0000000000000000000000',
    '00000000000000000000000',
    '000000000000000000000000',
    '0000000000000000000000000'
  ];

  var groupSizes = [
    0, 0,
    25, 16, 12, 11, 10, 9, 8,
    8, 7, 7, 7, 7, 6, 6,
    6, 6, 6, 6, 6, 5, 5,
    5, 5, 5, 5, 5, 5, 5,
    5, 5, 5, 5, 5, 5, 5
  ];

  var groupBases = [
    0, 0,
    33554432, 43046721, 16777216, 48828125, 60466176, 40353607, 16777216,
    43046721, 10000000, 19487171, 35831808, 62748517, 7529536, 11390625,
    16777216, 24137569, 34012224, 47045881, 64000000, 4084101, 5153632,
    6436343, 7962624, 9765625, 11881376, 14348907, 17210368, 20511149,
    24300000, 28629151, 33554432, 39135393, 45435424, 52521875, 60466176
  ];

  BN.prototype.toString = function toString (base, padding) ***REMOVED***
    base = base || 10;
    padding = padding | 0 || 1;

    var out;
    if (base === 16 || base === 'hex') ***REMOVED***
      out = '';
      var off = 0;
      var carry = 0;
      for (var i = 0; i < this.length; i++) ***REMOVED***
        var w = this.words[i];
        var word = (((w << off) | carry) & 0xffffff).toString(16);
        carry = (w >>> (24 - off)) & 0xffffff;
        if (carry !== 0 || i !== this.length - 1) ***REMOVED***
          out = zeros[6 - word.length] + word + out;
        ***REMOVED*** else ***REMOVED***
          out = word + out;
        ***REMOVED***
        off += 2;
        if (off >= 26) ***REMOVED***
          off -= 26;
          i--;
        ***REMOVED***
      ***REMOVED***
      if (carry !== 0) ***REMOVED***
        out = carry.toString(16) + out;
      ***REMOVED***
      while (out.length % padding !== 0) ***REMOVED***
        out = '0' + out;
      ***REMOVED***
      if (this.negative !== 0) ***REMOVED***
        out = '-' + out;
      ***REMOVED***
      return out;
    ***REMOVED***

    if (base === (base | 0) && base >= 2 && base <= 36) ***REMOVED***
      // var groupSize = Math.floor(BN.wordSize * Math.LN2 / Math.log(base));
      var groupSize = groupSizes[base];
      // var groupBase = Math.pow(base, groupSize);
      var groupBase = groupBases[base];
      out = '';
      var c = this.clone();
      c.negative = 0;
      while (!c.isZero()) ***REMOVED***
        var r = c.modn(groupBase).toString(base);
        c = c.idivn(groupBase);

        if (!c.isZero()) ***REMOVED***
          out = zeros[groupSize - r.length] + r + out;
        ***REMOVED*** else ***REMOVED***
          out = r + out;
        ***REMOVED***
      ***REMOVED***
      if (this.isZero()) ***REMOVED***
        out = '0' + out;
      ***REMOVED***
      while (out.length % padding !== 0) ***REMOVED***
        out = '0' + out;
      ***REMOVED***
      if (this.negative !== 0) ***REMOVED***
        out = '-' + out;
      ***REMOVED***
      return out;
    ***REMOVED***

    assert(false, 'Base should be between 2 and 36');
  ***REMOVED***;

  BN.prototype.toNumber = function toNumber () ***REMOVED***
    var ret = this.words[0];
    if (this.length === 2) ***REMOVED***
      ret += this.words[1] * 0x4000000;
    ***REMOVED*** else if (this.length === 3 && this.words[2] === 0x01) ***REMOVED***
      // NOTE: at this stage it is known that the top bit is set
      ret += 0x10000000000000 + (this.words[1] * 0x4000000);
    ***REMOVED*** else if (this.length > 2) ***REMOVED***
      assert(false, 'Number can only safely store up to 53 bits');
    ***REMOVED***
    return (this.negative !== 0) ? -ret : ret;
  ***REMOVED***;

  BN.prototype.toJSON = function toJSON () ***REMOVED***
    return this.toString(16);
  ***REMOVED***;

  BN.prototype.toBuffer = function toBuffer (endian, length) ***REMOVED***
    assert(typeof Buffer !== 'undefined');
    return this.toArrayLike(Buffer, endian, length);
  ***REMOVED***;

  BN.prototype.toArray = function toArray (endian, length) ***REMOVED***
    return this.toArrayLike(Array, endian, length);
  ***REMOVED***;

  BN.prototype.toArrayLike = function toArrayLike (ArrayType, endian, length) ***REMOVED***
    var byteLength = this.byteLength();
    var reqLength = length || Math.max(1, byteLength);
    assert(byteLength <= reqLength, 'byte array longer than desired length');
    assert(reqLength > 0, 'Requested array length <= 0');

    this.strip();
    var littleEndian = endian === 'le';
    var res = new ArrayType(reqLength);

    var b, i;
    var q = this.clone();
    if (!littleEndian) ***REMOVED***
      // Assume big-endian
      for (i = 0; i < reqLength - byteLength; i++) ***REMOVED***
        res[i] = 0;
      ***REMOVED***

      for (i = 0; !q.isZero(); i++) ***REMOVED***
        b = q.andln(0xff);
        q.iushrn(8);

        res[reqLength - i - 1] = b;
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      for (i = 0; !q.isZero(); i++) ***REMOVED***
        b = q.andln(0xff);
        q.iushrn(8);

        res[i] = b;
      ***REMOVED***

      for (; i < reqLength; i++) ***REMOVED***
        res[i] = 0;
      ***REMOVED***
    ***REMOVED***

    return res;
  ***REMOVED***;

  if (Math.clz32) ***REMOVED***
    BN.prototype._countBits = function _countBits (w) ***REMOVED***
      return 32 - Math.clz32(w);
    ***REMOVED***;
  ***REMOVED*** else ***REMOVED***
    BN.prototype._countBits = function _countBits (w) ***REMOVED***
      var t = w;
      var r = 0;
      if (t >= 0x1000) ***REMOVED***
        r += 13;
        t >>>= 13;
      ***REMOVED***
      if (t >= 0x40) ***REMOVED***
        r += 7;
        t >>>= 7;
      ***REMOVED***
      if (t >= 0x8) ***REMOVED***
        r += 4;
        t >>>= 4;
      ***REMOVED***
      if (t >= 0x02) ***REMOVED***
        r += 2;
        t >>>= 2;
      ***REMOVED***
      return r + t;
    ***REMOVED***;
  ***REMOVED***

  BN.prototype._zeroBits = function _zeroBits (w) ***REMOVED***
    // Short-cut
    if (w === 0) return 26;

    var t = w;
    var r = 0;
    if ((t & 0x1fff) === 0) ***REMOVED***
      r += 13;
      t >>>= 13;
    ***REMOVED***
    if ((t & 0x7f) === 0) ***REMOVED***
      r += 7;
      t >>>= 7;
    ***REMOVED***
    if ((t & 0xf) === 0) ***REMOVED***
      r += 4;
      t >>>= 4;
    ***REMOVED***
    if ((t & 0x3) === 0) ***REMOVED***
      r += 2;
      t >>>= 2;
    ***REMOVED***
    if ((t & 0x1) === 0) ***REMOVED***
      r++;
    ***REMOVED***
    return r;
  ***REMOVED***;

  // Return number of used bits in a BN
  BN.prototype.bitLength = function bitLength () ***REMOVED***
    var w = this.words[this.length - 1];
    var hi = this._countBits(w);
    return (this.length - 1) * 26 + hi;
  ***REMOVED***;

  function toBitArray (num) ***REMOVED***
    var w = new Array(num.bitLength());

    for (var bit = 0; bit < w.length; bit++) ***REMOVED***
      var off = (bit / 26) | 0;
      var wbit = bit % 26;

      w[bit] = (num.words[off] & (1 << wbit)) >>> wbit;
    ***REMOVED***

    return w;
  ***REMOVED***

  // Number of trailing zero bits
  BN.prototype.zeroBits = function zeroBits () ***REMOVED***
    if (this.isZero()) return 0;

    var r = 0;
    for (var i = 0; i < this.length; i++) ***REMOVED***
      var b = this._zeroBits(this.words[i]);
      r += b;
      if (b !== 26) break;
    ***REMOVED***
    return r;
  ***REMOVED***;

  BN.prototype.byteLength = function byteLength () ***REMOVED***
    return Math.ceil(this.bitLength() / 8);
  ***REMOVED***;

  BN.prototype.toTwos = function toTwos (width) ***REMOVED***
    if (this.negative !== 0) ***REMOVED***
      return this.abs().inotn(width).iaddn(1);
    ***REMOVED***
    return this.clone();
  ***REMOVED***;

  BN.prototype.fromTwos = function fromTwos (width) ***REMOVED***
    if (this.testn(width - 1)) ***REMOVED***
      return this.notn(width).iaddn(1).ineg();
    ***REMOVED***
    return this.clone();
  ***REMOVED***;

  BN.prototype.isNeg = function isNeg () ***REMOVED***
    return this.negative !== 0;
  ***REMOVED***;

  // Return negative clone of `this`
  BN.prototype.neg = function neg () ***REMOVED***
    return this.clone().ineg();
  ***REMOVED***;

  BN.prototype.ineg = function ineg () ***REMOVED***
    if (!this.isZero()) ***REMOVED***
      this.negative ^= 1;
    ***REMOVED***

    return this;
  ***REMOVED***;

  // Or `num` with `this` in-place
  BN.prototype.iuor = function iuor (num) ***REMOVED***
    while (this.length < num.length) ***REMOVED***
      this.words[this.length++] = 0;
    ***REMOVED***

    for (var i = 0; i < num.length; i++) ***REMOVED***
      this.words[i] = this.words[i] | num.words[i];
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  BN.prototype.ior = function ior (num) ***REMOVED***
    assert((this.negative | num.negative) === 0);
    return this.iuor(num);
  ***REMOVED***;

  // Or `num` with `this`
  BN.prototype.or = function or (num) ***REMOVED***
    if (this.length > num.length) return this.clone().ior(num);
    return num.clone().ior(this);
  ***REMOVED***;

  BN.prototype.uor = function uor (num) ***REMOVED***
    if (this.length > num.length) return this.clone().iuor(num);
    return num.clone().iuor(this);
  ***REMOVED***;

  // And `num` with `this` in-place
  BN.prototype.iuand = function iuand (num) ***REMOVED***
    // b = min-length(num, this)
    var b;
    if (this.length > num.length) ***REMOVED***
      b = num;
    ***REMOVED*** else ***REMOVED***
      b = this;
    ***REMOVED***

    for (var i = 0; i < b.length; i++) ***REMOVED***
      this.words[i] = this.words[i] & num.words[i];
    ***REMOVED***

    this.length = b.length;

    return this.strip();
  ***REMOVED***;

  BN.prototype.iand = function iand (num) ***REMOVED***
    assert((this.negative | num.negative) === 0);
    return this.iuand(num);
  ***REMOVED***;

  // And `num` with `this`
  BN.prototype.and = function and (num) ***REMOVED***
    if (this.length > num.length) return this.clone().iand(num);
    return num.clone().iand(this);
  ***REMOVED***;

  BN.prototype.uand = function uand (num) ***REMOVED***
    if (this.length > num.length) return this.clone().iuand(num);
    return num.clone().iuand(this);
  ***REMOVED***;

  // Xor `num` with `this` in-place
  BN.prototype.iuxor = function iuxor (num) ***REMOVED***
    // a.length > b.length
    var a;
    var b;
    if (this.length > num.length) ***REMOVED***
      a = this;
      b = num;
    ***REMOVED*** else ***REMOVED***
      a = num;
      b = this;
    ***REMOVED***

    for (var i = 0; i < b.length; i++) ***REMOVED***
      this.words[i] = a.words[i] ^ b.words[i];
    ***REMOVED***

    if (this !== a) ***REMOVED***
      for (; i < a.length; i++) ***REMOVED***
        this.words[i] = a.words[i];
      ***REMOVED***
    ***REMOVED***

    this.length = a.length;

    return this.strip();
  ***REMOVED***;

  BN.prototype.ixor = function ixor (num) ***REMOVED***
    assert((this.negative | num.negative) === 0);
    return this.iuxor(num);
  ***REMOVED***;

  // Xor `num` with `this`
  BN.prototype.xor = function xor (num) ***REMOVED***
    if (this.length > num.length) return this.clone().ixor(num);
    return num.clone().ixor(this);
  ***REMOVED***;

  BN.prototype.uxor = function uxor (num) ***REMOVED***
    if (this.length > num.length) return this.clone().iuxor(num);
    return num.clone().iuxor(this);
  ***REMOVED***;

  // Not ``this`` with ``width`` bitwidth
  BN.prototype.inotn = function inotn (width) ***REMOVED***
    assert(typeof width === 'number' && width >= 0);

    var bytesNeeded = Math.ceil(width / 26) | 0;
    var bitsLeft = width % 26;

    // Extend the buffer with leading zeroes
    this._expand(bytesNeeded);

    if (bitsLeft > 0) ***REMOVED***
      bytesNeeded--;
    ***REMOVED***

    // Handle complete words
    for (var i = 0; i < bytesNeeded; i++) ***REMOVED***
      this.words[i] = ~this.words[i] & 0x3ffffff;
    ***REMOVED***

    // Handle the residue
    if (bitsLeft > 0) ***REMOVED***
      this.words[i] = ~this.words[i] & (0x3ffffff >> (26 - bitsLeft));
    ***REMOVED***

    // And remove leading zeroes
    return this.strip();
  ***REMOVED***;

  BN.prototype.notn = function notn (width) ***REMOVED***
    return this.clone().inotn(width);
  ***REMOVED***;

  // Set `bit` of `this`
  BN.prototype.setn = function setn (bit, val) ***REMOVED***
    assert(typeof bit === 'number' && bit >= 0);

    var off = (bit / 26) | 0;
    var wbit = bit % 26;

    this._expand(off + 1);

    if (val) ***REMOVED***
      this.words[off] = this.words[off] | (1 << wbit);
    ***REMOVED*** else ***REMOVED***
      this.words[off] = this.words[off] & ~(1 << wbit);
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  // Add `num` to `this` in-place
  BN.prototype.iadd = function iadd (num) ***REMOVED***
    var r;

    // negative + positive
    if (this.negative !== 0 && num.negative === 0) ***REMOVED***
      this.negative = 0;
      r = this.isub(num);
      this.negative ^= 1;
      return this._normSign();

    // positive + negative
    ***REMOVED*** else if (this.negative === 0 && num.negative !== 0) ***REMOVED***
      num.negative = 0;
      r = this.isub(num);
      num.negative = 1;
      return r._normSign();
    ***REMOVED***

    // a.length > b.length
    var a, b;
    if (this.length > num.length) ***REMOVED***
      a = this;
      b = num;
    ***REMOVED*** else ***REMOVED***
      a = num;
      b = this;
    ***REMOVED***

    var carry = 0;
    for (var i = 0; i < b.length; i++) ***REMOVED***
      r = (a.words[i] | 0) + (b.words[i] | 0) + carry;
      this.words[i] = r & 0x3ffffff;
      carry = r >>> 26;
    ***REMOVED***
    for (; carry !== 0 && i < a.length; i++) ***REMOVED***
      r = (a.words[i] | 0) + carry;
      this.words[i] = r & 0x3ffffff;
      carry = r >>> 26;
    ***REMOVED***

    this.length = a.length;
    if (carry !== 0) ***REMOVED***
      this.words[this.length] = carry;
      this.length++;
    // Copy the rest of the words
    ***REMOVED*** else if (a !== this) ***REMOVED***
      for (; i < a.length; i++) ***REMOVED***
        this.words[i] = a.words[i];
      ***REMOVED***
    ***REMOVED***

    return this;
  ***REMOVED***;

  // Add `num` to `this`
  BN.prototype.add = function add (num) ***REMOVED***
    var res;
    if (num.negative !== 0 && this.negative === 0) ***REMOVED***
      num.negative = 0;
      res = this.sub(num);
      num.negative ^= 1;
      return res;
    ***REMOVED*** else if (num.negative === 0 && this.negative !== 0) ***REMOVED***
      this.negative = 0;
      res = num.sub(this);
      this.negative = 1;
      return res;
    ***REMOVED***

    if (this.length > num.length) return this.clone().iadd(num);

    return num.clone().iadd(this);
  ***REMOVED***;

  // Subtract `num` from `this` in-place
  BN.prototype.isub = function isub (num) ***REMOVED***
    // this - (-num) = this + num
    if (num.negative !== 0) ***REMOVED***
      num.negative = 0;
      var r = this.iadd(num);
      num.negative = 1;
      return r._normSign();

    // -this - num = -(this + num)
    ***REMOVED*** else if (this.negative !== 0) ***REMOVED***
      this.negative = 0;
      this.iadd(num);
      this.negative = 1;
      return this._normSign();
    ***REMOVED***

    // At this point both numbers are positive
    var cmp = this.cmp(num);

    // Optimization - zeroify
    if (cmp === 0) ***REMOVED***
      this.negative = 0;
      this.length = 1;
      this.words[0] = 0;
      return this;
    ***REMOVED***

    // a > b
    var a, b;
    if (cmp > 0) ***REMOVED***
      a = this;
      b = num;
    ***REMOVED*** else ***REMOVED***
      a = num;
      b = this;
    ***REMOVED***

    var carry = 0;
    for (var i = 0; i < b.length; i++) ***REMOVED***
      r = (a.words[i] | 0) - (b.words[i] | 0) + carry;
      carry = r >> 26;
      this.words[i] = r & 0x3ffffff;
    ***REMOVED***
    for (; carry !== 0 && i < a.length; i++) ***REMOVED***
      r = (a.words[i] | 0) + carry;
      carry = r >> 26;
      this.words[i] = r & 0x3ffffff;
    ***REMOVED***

    // Copy rest of the words
    if (carry === 0 && i < a.length && a !== this) ***REMOVED***
      for (; i < a.length; i++) ***REMOVED***
        this.words[i] = a.words[i];
      ***REMOVED***
    ***REMOVED***

    this.length = Math.max(this.length, i);

    if (a !== this) ***REMOVED***
      this.negative = 1;
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  // Subtract `num` from `this`
  BN.prototype.sub = function sub (num) ***REMOVED***
    return this.clone().isub(num);
  ***REMOVED***;

  function smallMulTo (self, num, out) ***REMOVED***
    out.negative = num.negative ^ self.negative;
    var len = (self.length + num.length) | 0;
    out.length = len;
    len = (len - 1) | 0;

    // Peel one iteration (compiler can't do it, because of code complexity)
    var a = self.words[0] | 0;
    var b = num.words[0] | 0;
    var r = a * b;

    var lo = r & 0x3ffffff;
    var carry = (r / 0x4000000) | 0;
    out.words[0] = lo;

    for (var k = 1; k < len; k++) ***REMOVED***
      // Sum all words with the same `i + j = k` and accumulate `ncarry`,
      // note that ncarry could be >= 0x3ffffff
      var ncarry = carry >>> 26;
      var rword = carry & 0x3ffffff;
      var maxJ = Math.min(k, num.length - 1);
      for (var j = Math.max(0, k - self.length + 1); j <= maxJ; j++) ***REMOVED***
        var i = (k - j) | 0;
        a = self.words[i] | 0;
        b = num.words[j] | 0;
        r = a * b + rword;
        ncarry += (r / 0x4000000) | 0;
        rword = r & 0x3ffffff;
      ***REMOVED***
      out.words[k] = rword | 0;
      carry = ncarry | 0;
    ***REMOVED***
    if (carry !== 0) ***REMOVED***
      out.words[k] = carry | 0;
    ***REMOVED*** else ***REMOVED***
      out.length--;
    ***REMOVED***

    return out.strip();
  ***REMOVED***

  // TODO(indutny): it may be reasonable to omit it for users who don't need
  // to work with 256-bit numbers, otherwise it gives 20% improvement for 256-bit
  // multiplication (like elliptic secp256k1).
  var comb10MulTo = function comb10MulTo (self, num, out) ***REMOVED***
    var a = self.words;
    var b = num.words;
    var o = out.words;
    var c = 0;
    var lo;
    var mid;
    var hi;
    var a0 = a[0] | 0;
    var al0 = a0 & 0x1fff;
    var ah0 = a0 >>> 13;
    var a1 = a[1] | 0;
    var al1 = a1 & 0x1fff;
    var ah1 = a1 >>> 13;
    var a2 = a[2] | 0;
    var al2 = a2 & 0x1fff;
    var ah2 = a2 >>> 13;
    var a3 = a[3] | 0;
    var al3 = a3 & 0x1fff;
    var ah3 = a3 >>> 13;
    var a4 = a[4] | 0;
    var al4 = a4 & 0x1fff;
    var ah4 = a4 >>> 13;
    var a5 = a[5] | 0;
    var al5 = a5 & 0x1fff;
    var ah5 = a5 >>> 13;
    var a6 = a[6] | 0;
    var al6 = a6 & 0x1fff;
    var ah6 = a6 >>> 13;
    var a7 = a[7] | 0;
    var al7 = a7 & 0x1fff;
    var ah7 = a7 >>> 13;
    var a8 = a[8] | 0;
    var al8 = a8 & 0x1fff;
    var ah8 = a8 >>> 13;
    var a9 = a[9] | 0;
    var al9 = a9 & 0x1fff;
    var ah9 = a9 >>> 13;
    var b0 = b[0] | 0;
    var bl0 = b0 & 0x1fff;
    var bh0 = b0 >>> 13;
    var b1 = b[1] | 0;
    var bl1 = b1 & 0x1fff;
    var bh1 = b1 >>> 13;
    var b2 = b[2] | 0;
    var bl2 = b2 & 0x1fff;
    var bh2 = b2 >>> 13;
    var b3 = b[3] | 0;
    var bl3 = b3 & 0x1fff;
    var bh3 = b3 >>> 13;
    var b4 = b[4] | 0;
    var bl4 = b4 & 0x1fff;
    var bh4 = b4 >>> 13;
    var b5 = b[5] | 0;
    var bl5 = b5 & 0x1fff;
    var bh5 = b5 >>> 13;
    var b6 = b[6] | 0;
    var bl6 = b6 & 0x1fff;
    var bh6 = b6 >>> 13;
    var b7 = b[7] | 0;
    var bl7 = b7 & 0x1fff;
    var bh7 = b7 >>> 13;
    var b8 = b[8] | 0;
    var bl8 = b8 & 0x1fff;
    var bh8 = b8 >>> 13;
    var b9 = b[9] | 0;
    var bl9 = b9 & 0x1fff;
    var bh9 = b9 >>> 13;

    out.negative = self.negative ^ num.negative;
    out.length = 19;
    /* k = 0 */
    lo = Math.imul(al0, bl0);
    mid = Math.imul(al0, bh0);
    mid = (mid + Math.imul(ah0, bl0)) | 0;
    hi = Math.imul(ah0, bh0);
    var w0 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w0 >>> 26)) | 0;
    w0 &= 0x3ffffff;
    /* k = 1 */
    lo = Math.imul(al1, bl0);
    mid = Math.imul(al1, bh0);
    mid = (mid + Math.imul(ah1, bl0)) | 0;
    hi = Math.imul(ah1, bh0);
    lo = (lo + Math.imul(al0, bl1)) | 0;
    mid = (mid + Math.imul(al0, bh1)) | 0;
    mid = (mid + Math.imul(ah0, bl1)) | 0;
    hi = (hi + Math.imul(ah0, bh1)) | 0;
    var w1 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w1 >>> 26)) | 0;
    w1 &= 0x3ffffff;
    /* k = 2 */
    lo = Math.imul(al2, bl0);
    mid = Math.imul(al2, bh0);
    mid = (mid + Math.imul(ah2, bl0)) | 0;
    hi = Math.imul(ah2, bh0);
    lo = (lo + Math.imul(al1, bl1)) | 0;
    mid = (mid + Math.imul(al1, bh1)) | 0;
    mid = (mid + Math.imul(ah1, bl1)) | 0;
    hi = (hi + Math.imul(ah1, bh1)) | 0;
    lo = (lo + Math.imul(al0, bl2)) | 0;
    mid = (mid + Math.imul(al0, bh2)) | 0;
    mid = (mid + Math.imul(ah0, bl2)) | 0;
    hi = (hi + Math.imul(ah0, bh2)) | 0;
    var w2 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w2 >>> 26)) | 0;
    w2 &= 0x3ffffff;
    /* k = 3 */
    lo = Math.imul(al3, bl0);
    mid = Math.imul(al3, bh0);
    mid = (mid + Math.imul(ah3, bl0)) | 0;
    hi = Math.imul(ah3, bh0);
    lo = (lo + Math.imul(al2, bl1)) | 0;
    mid = (mid + Math.imul(al2, bh1)) | 0;
    mid = (mid + Math.imul(ah2, bl1)) | 0;
    hi = (hi + Math.imul(ah2, bh1)) | 0;
    lo = (lo + Math.imul(al1, bl2)) | 0;
    mid = (mid + Math.imul(al1, bh2)) | 0;
    mid = (mid + Math.imul(ah1, bl2)) | 0;
    hi = (hi + Math.imul(ah1, bh2)) | 0;
    lo = (lo + Math.imul(al0, bl3)) | 0;
    mid = (mid + Math.imul(al0, bh3)) | 0;
    mid = (mid + Math.imul(ah0, bl3)) | 0;
    hi = (hi + Math.imul(ah0, bh3)) | 0;
    var w3 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w3 >>> 26)) | 0;
    w3 &= 0x3ffffff;
    /* k = 4 */
    lo = Math.imul(al4, bl0);
    mid = Math.imul(al4, bh0);
    mid = (mid + Math.imul(ah4, bl0)) | 0;
    hi = Math.imul(ah4, bh0);
    lo = (lo + Math.imul(al3, bl1)) | 0;
    mid = (mid + Math.imul(al3, bh1)) | 0;
    mid = (mid + Math.imul(ah3, bl1)) | 0;
    hi = (hi + Math.imul(ah3, bh1)) | 0;
    lo = (lo + Math.imul(al2, bl2)) | 0;
    mid = (mid + Math.imul(al2, bh2)) | 0;
    mid = (mid + Math.imul(ah2, bl2)) | 0;
    hi = (hi + Math.imul(ah2, bh2)) | 0;
    lo = (lo + Math.imul(al1, bl3)) | 0;
    mid = (mid + Math.imul(al1, bh3)) | 0;
    mid = (mid + Math.imul(ah1, bl3)) | 0;
    hi = (hi + Math.imul(ah1, bh3)) | 0;
    lo = (lo + Math.imul(al0, bl4)) | 0;
    mid = (mid + Math.imul(al0, bh4)) | 0;
    mid = (mid + Math.imul(ah0, bl4)) | 0;
    hi = (hi + Math.imul(ah0, bh4)) | 0;
    var w4 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w4 >>> 26)) | 0;
    w4 &= 0x3ffffff;
    /* k = 5 */
    lo = Math.imul(al5, bl0);
    mid = Math.imul(al5, bh0);
    mid = (mid + Math.imul(ah5, bl0)) | 0;
    hi = Math.imul(ah5, bh0);
    lo = (lo + Math.imul(al4, bl1)) | 0;
    mid = (mid + Math.imul(al4, bh1)) | 0;
    mid = (mid + Math.imul(ah4, bl1)) | 0;
    hi = (hi + Math.imul(ah4, bh1)) | 0;
    lo = (lo + Math.imul(al3, bl2)) | 0;
    mid = (mid + Math.imul(al3, bh2)) | 0;
    mid = (mid + Math.imul(ah3, bl2)) | 0;
    hi = (hi + Math.imul(ah3, bh2)) | 0;
    lo = (lo + Math.imul(al2, bl3)) | 0;
    mid = (mid + Math.imul(al2, bh3)) | 0;
    mid = (mid + Math.imul(ah2, bl3)) | 0;
    hi = (hi + Math.imul(ah2, bh3)) | 0;
    lo = (lo + Math.imul(al1, bl4)) | 0;
    mid = (mid + Math.imul(al1, bh4)) | 0;
    mid = (mid + Math.imul(ah1, bl4)) | 0;
    hi = (hi + Math.imul(ah1, bh4)) | 0;
    lo = (lo + Math.imul(al0, bl5)) | 0;
    mid = (mid + Math.imul(al0, bh5)) | 0;
    mid = (mid + Math.imul(ah0, bl5)) | 0;
    hi = (hi + Math.imul(ah0, bh5)) | 0;
    var w5 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w5 >>> 26)) | 0;
    w5 &= 0x3ffffff;
    /* k = 6 */
    lo = Math.imul(al6, bl0);
    mid = Math.imul(al6, bh0);
    mid = (mid + Math.imul(ah6, bl0)) | 0;
    hi = Math.imul(ah6, bh0);
    lo = (lo + Math.imul(al5, bl1)) | 0;
    mid = (mid + Math.imul(al5, bh1)) | 0;
    mid = (mid + Math.imul(ah5, bl1)) | 0;
    hi = (hi + Math.imul(ah5, bh1)) | 0;
    lo = (lo + Math.imul(al4, bl2)) | 0;
    mid = (mid + Math.imul(al4, bh2)) | 0;
    mid = (mid + Math.imul(ah4, bl2)) | 0;
    hi = (hi + Math.imul(ah4, bh2)) | 0;
    lo = (lo + Math.imul(al3, bl3)) | 0;
    mid = (mid + Math.imul(al3, bh3)) | 0;
    mid = (mid + Math.imul(ah3, bl3)) | 0;
    hi = (hi + Math.imul(ah3, bh3)) | 0;
    lo = (lo + Math.imul(al2, bl4)) | 0;
    mid = (mid + Math.imul(al2, bh4)) | 0;
    mid = (mid + Math.imul(ah2, bl4)) | 0;
    hi = (hi + Math.imul(ah2, bh4)) | 0;
    lo = (lo + Math.imul(al1, bl5)) | 0;
    mid = (mid + Math.imul(al1, bh5)) | 0;
    mid = (mid + Math.imul(ah1, bl5)) | 0;
    hi = (hi + Math.imul(ah1, bh5)) | 0;
    lo = (lo + Math.imul(al0, bl6)) | 0;
    mid = (mid + Math.imul(al0, bh6)) | 0;
    mid = (mid + Math.imul(ah0, bl6)) | 0;
    hi = (hi + Math.imul(ah0, bh6)) | 0;
    var w6 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w6 >>> 26)) | 0;
    w6 &= 0x3ffffff;
    /* k = 7 */
    lo = Math.imul(al7, bl0);
    mid = Math.imul(al7, bh0);
    mid = (mid + Math.imul(ah7, bl0)) | 0;
    hi = Math.imul(ah7, bh0);
    lo = (lo + Math.imul(al6, bl1)) | 0;
    mid = (mid + Math.imul(al6, bh1)) | 0;
    mid = (mid + Math.imul(ah6, bl1)) | 0;
    hi = (hi + Math.imul(ah6, bh1)) | 0;
    lo = (lo + Math.imul(al5, bl2)) | 0;
    mid = (mid + Math.imul(al5, bh2)) | 0;
    mid = (mid + Math.imul(ah5, bl2)) | 0;
    hi = (hi + Math.imul(ah5, bh2)) | 0;
    lo = (lo + Math.imul(al4, bl3)) | 0;
    mid = (mid + Math.imul(al4, bh3)) | 0;
    mid = (mid + Math.imul(ah4, bl3)) | 0;
    hi = (hi + Math.imul(ah4, bh3)) | 0;
    lo = (lo + Math.imul(al3, bl4)) | 0;
    mid = (mid + Math.imul(al3, bh4)) | 0;
    mid = (mid + Math.imul(ah3, bl4)) | 0;
    hi = (hi + Math.imul(ah3, bh4)) | 0;
    lo = (lo + Math.imul(al2, bl5)) | 0;
    mid = (mid + Math.imul(al2, bh5)) | 0;
    mid = (mid + Math.imul(ah2, bl5)) | 0;
    hi = (hi + Math.imul(ah2, bh5)) | 0;
    lo = (lo + Math.imul(al1, bl6)) | 0;
    mid = (mid + Math.imul(al1, bh6)) | 0;
    mid = (mid + Math.imul(ah1, bl6)) | 0;
    hi = (hi + Math.imul(ah1, bh6)) | 0;
    lo = (lo + Math.imul(al0, bl7)) | 0;
    mid = (mid + Math.imul(al0, bh7)) | 0;
    mid = (mid + Math.imul(ah0, bl7)) | 0;
    hi = (hi + Math.imul(ah0, bh7)) | 0;
    var w7 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w7 >>> 26)) | 0;
    w7 &= 0x3ffffff;
    /* k = 8 */
    lo = Math.imul(al8, bl0);
    mid = Math.imul(al8, bh0);
    mid = (mid + Math.imul(ah8, bl0)) | 0;
    hi = Math.imul(ah8, bh0);
    lo = (lo + Math.imul(al7, bl1)) | 0;
    mid = (mid + Math.imul(al7, bh1)) | 0;
    mid = (mid + Math.imul(ah7, bl1)) | 0;
    hi = (hi + Math.imul(ah7, bh1)) | 0;
    lo = (lo + Math.imul(al6, bl2)) | 0;
    mid = (mid + Math.imul(al6, bh2)) | 0;
    mid = (mid + Math.imul(ah6, bl2)) | 0;
    hi = (hi + Math.imul(ah6, bh2)) | 0;
    lo = (lo + Math.imul(al5, bl3)) | 0;
    mid = (mid + Math.imul(al5, bh3)) | 0;
    mid = (mid + Math.imul(ah5, bl3)) | 0;
    hi = (hi + Math.imul(ah5, bh3)) | 0;
    lo = (lo + Math.imul(al4, bl4)) | 0;
    mid = (mid + Math.imul(al4, bh4)) | 0;
    mid = (mid + Math.imul(ah4, bl4)) | 0;
    hi = (hi + Math.imul(ah4, bh4)) | 0;
    lo = (lo + Math.imul(al3, bl5)) | 0;
    mid = (mid + Math.imul(al3, bh5)) | 0;
    mid = (mid + Math.imul(ah3, bl5)) | 0;
    hi = (hi + Math.imul(ah3, bh5)) | 0;
    lo = (lo + Math.imul(al2, bl6)) | 0;
    mid = (mid + Math.imul(al2, bh6)) | 0;
    mid = (mid + Math.imul(ah2, bl6)) | 0;
    hi = (hi + Math.imul(ah2, bh6)) | 0;
    lo = (lo + Math.imul(al1, bl7)) | 0;
    mid = (mid + Math.imul(al1, bh7)) | 0;
    mid = (mid + Math.imul(ah1, bl7)) | 0;
    hi = (hi + Math.imul(ah1, bh7)) | 0;
    lo = (lo + Math.imul(al0, bl8)) | 0;
    mid = (mid + Math.imul(al0, bh8)) | 0;
    mid = (mid + Math.imul(ah0, bl8)) | 0;
    hi = (hi + Math.imul(ah0, bh8)) | 0;
    var w8 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w8 >>> 26)) | 0;
    w8 &= 0x3ffffff;
    /* k = 9 */
    lo = Math.imul(al9, bl0);
    mid = Math.imul(al9, bh0);
    mid = (mid + Math.imul(ah9, bl0)) | 0;
    hi = Math.imul(ah9, bh0);
    lo = (lo + Math.imul(al8, bl1)) | 0;
    mid = (mid + Math.imul(al8, bh1)) | 0;
    mid = (mid + Math.imul(ah8, bl1)) | 0;
    hi = (hi + Math.imul(ah8, bh1)) | 0;
    lo = (lo + Math.imul(al7, bl2)) | 0;
    mid = (mid + Math.imul(al7, bh2)) | 0;
    mid = (mid + Math.imul(ah7, bl2)) | 0;
    hi = (hi + Math.imul(ah7, bh2)) | 0;
    lo = (lo + Math.imul(al6, bl3)) | 0;
    mid = (mid + Math.imul(al6, bh3)) | 0;
    mid = (mid + Math.imul(ah6, bl3)) | 0;
    hi = (hi + Math.imul(ah6, bh3)) | 0;
    lo = (lo + Math.imul(al5, bl4)) | 0;
    mid = (mid + Math.imul(al5, bh4)) | 0;
    mid = (mid + Math.imul(ah5, bl4)) | 0;
    hi = (hi + Math.imul(ah5, bh4)) | 0;
    lo = (lo + Math.imul(al4, bl5)) | 0;
    mid = (mid + Math.imul(al4, bh5)) | 0;
    mid = (mid + Math.imul(ah4, bl5)) | 0;
    hi = (hi + Math.imul(ah4, bh5)) | 0;
    lo = (lo + Math.imul(al3, bl6)) | 0;
    mid = (mid + Math.imul(al3, bh6)) | 0;
    mid = (mid + Math.imul(ah3, bl6)) | 0;
    hi = (hi + Math.imul(ah3, bh6)) | 0;
    lo = (lo + Math.imul(al2, bl7)) | 0;
    mid = (mid + Math.imul(al2, bh7)) | 0;
    mid = (mid + Math.imul(ah2, bl7)) | 0;
    hi = (hi + Math.imul(ah2, bh7)) | 0;
    lo = (lo + Math.imul(al1, bl8)) | 0;
    mid = (mid + Math.imul(al1, bh8)) | 0;
    mid = (mid + Math.imul(ah1, bl8)) | 0;
    hi = (hi + Math.imul(ah1, bh8)) | 0;
    lo = (lo + Math.imul(al0, bl9)) | 0;
    mid = (mid + Math.imul(al0, bh9)) | 0;
    mid = (mid + Math.imul(ah0, bl9)) | 0;
    hi = (hi + Math.imul(ah0, bh9)) | 0;
    var w9 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w9 >>> 26)) | 0;
    w9 &= 0x3ffffff;
    /* k = 10 */
    lo = Math.imul(al9, bl1);
    mid = Math.imul(al9, bh1);
    mid = (mid + Math.imul(ah9, bl1)) | 0;
    hi = Math.imul(ah9, bh1);
    lo = (lo + Math.imul(al8, bl2)) | 0;
    mid = (mid + Math.imul(al8, bh2)) | 0;
    mid = (mid + Math.imul(ah8, bl2)) | 0;
    hi = (hi + Math.imul(ah8, bh2)) | 0;
    lo = (lo + Math.imul(al7, bl3)) | 0;
    mid = (mid + Math.imul(al7, bh3)) | 0;
    mid = (mid + Math.imul(ah7, bl3)) | 0;
    hi = (hi + Math.imul(ah7, bh3)) | 0;
    lo = (lo + Math.imul(al6, bl4)) | 0;
    mid = (mid + Math.imul(al6, bh4)) | 0;
    mid = (mid + Math.imul(ah6, bl4)) | 0;
    hi = (hi + Math.imul(ah6, bh4)) | 0;
    lo = (lo + Math.imul(al5, bl5)) | 0;
    mid = (mid + Math.imul(al5, bh5)) | 0;
    mid = (mid + Math.imul(ah5, bl5)) | 0;
    hi = (hi + Math.imul(ah5, bh5)) | 0;
    lo = (lo + Math.imul(al4, bl6)) | 0;
    mid = (mid + Math.imul(al4, bh6)) | 0;
    mid = (mid + Math.imul(ah4, bl6)) | 0;
    hi = (hi + Math.imul(ah4, bh6)) | 0;
    lo = (lo + Math.imul(al3, bl7)) | 0;
    mid = (mid + Math.imul(al3, bh7)) | 0;
    mid = (mid + Math.imul(ah3, bl7)) | 0;
    hi = (hi + Math.imul(ah3, bh7)) | 0;
    lo = (lo + Math.imul(al2, bl8)) | 0;
    mid = (mid + Math.imul(al2, bh8)) | 0;
    mid = (mid + Math.imul(ah2, bl8)) | 0;
    hi = (hi + Math.imul(ah2, bh8)) | 0;
    lo = (lo + Math.imul(al1, bl9)) | 0;
    mid = (mid + Math.imul(al1, bh9)) | 0;
    mid = (mid + Math.imul(ah1, bl9)) | 0;
    hi = (hi + Math.imul(ah1, bh9)) | 0;
    var w10 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w10 >>> 26)) | 0;
    w10 &= 0x3ffffff;
    /* k = 11 */
    lo = Math.imul(al9, bl2);
    mid = Math.imul(al9, bh2);
    mid = (mid + Math.imul(ah9, bl2)) | 0;
    hi = Math.imul(ah9, bh2);
    lo = (lo + Math.imul(al8, bl3)) | 0;
    mid = (mid + Math.imul(al8, bh3)) | 0;
    mid = (mid + Math.imul(ah8, bl3)) | 0;
    hi = (hi + Math.imul(ah8, bh3)) | 0;
    lo = (lo + Math.imul(al7, bl4)) | 0;
    mid = (mid + Math.imul(al7, bh4)) | 0;
    mid = (mid + Math.imul(ah7, bl4)) | 0;
    hi = (hi + Math.imul(ah7, bh4)) | 0;
    lo = (lo + Math.imul(al6, bl5)) | 0;
    mid = (mid + Math.imul(al6, bh5)) | 0;
    mid = (mid + Math.imul(ah6, bl5)) | 0;
    hi = (hi + Math.imul(ah6, bh5)) | 0;
    lo = (lo + Math.imul(al5, bl6)) | 0;
    mid = (mid + Math.imul(al5, bh6)) | 0;
    mid = (mid + Math.imul(ah5, bl6)) | 0;
    hi = (hi + Math.imul(ah5, bh6)) | 0;
    lo = (lo + Math.imul(al4, bl7)) | 0;
    mid = (mid + Math.imul(al4, bh7)) | 0;
    mid = (mid + Math.imul(ah4, bl7)) | 0;
    hi = (hi + Math.imul(ah4, bh7)) | 0;
    lo = (lo + Math.imul(al3, bl8)) | 0;
    mid = (mid + Math.imul(al3, bh8)) | 0;
    mid = (mid + Math.imul(ah3, bl8)) | 0;
    hi = (hi + Math.imul(ah3, bh8)) | 0;
    lo = (lo + Math.imul(al2, bl9)) | 0;
    mid = (mid + Math.imul(al2, bh9)) | 0;
    mid = (mid + Math.imul(ah2, bl9)) | 0;
    hi = (hi + Math.imul(ah2, bh9)) | 0;
    var w11 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w11 >>> 26)) | 0;
    w11 &= 0x3ffffff;
    /* k = 12 */
    lo = Math.imul(al9, bl3);
    mid = Math.imul(al9, bh3);
    mid = (mid + Math.imul(ah9, bl3)) | 0;
    hi = Math.imul(ah9, bh3);
    lo = (lo + Math.imul(al8, bl4)) | 0;
    mid = (mid + Math.imul(al8, bh4)) | 0;
    mid = (mid + Math.imul(ah8, bl4)) | 0;
    hi = (hi + Math.imul(ah8, bh4)) | 0;
    lo = (lo + Math.imul(al7, bl5)) | 0;
    mid = (mid + Math.imul(al7, bh5)) | 0;
    mid = (mid + Math.imul(ah7, bl5)) | 0;
    hi = (hi + Math.imul(ah7, bh5)) | 0;
    lo = (lo + Math.imul(al6, bl6)) | 0;
    mid = (mid + Math.imul(al6, bh6)) | 0;
    mid = (mid + Math.imul(ah6, bl6)) | 0;
    hi = (hi + Math.imul(ah6, bh6)) | 0;
    lo = (lo + Math.imul(al5, bl7)) | 0;
    mid = (mid + Math.imul(al5, bh7)) | 0;
    mid = (mid + Math.imul(ah5, bl7)) | 0;
    hi = (hi + Math.imul(ah5, bh7)) | 0;
    lo = (lo + Math.imul(al4, bl8)) | 0;
    mid = (mid + Math.imul(al4, bh8)) | 0;
    mid = (mid + Math.imul(ah4, bl8)) | 0;
    hi = (hi + Math.imul(ah4, bh8)) | 0;
    lo = (lo + Math.imul(al3, bl9)) | 0;
    mid = (mid + Math.imul(al3, bh9)) | 0;
    mid = (mid + Math.imul(ah3, bl9)) | 0;
    hi = (hi + Math.imul(ah3, bh9)) | 0;
    var w12 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w12 >>> 26)) | 0;
    w12 &= 0x3ffffff;
    /* k = 13 */
    lo = Math.imul(al9, bl4);
    mid = Math.imul(al9, bh4);
    mid = (mid + Math.imul(ah9, bl4)) | 0;
    hi = Math.imul(ah9, bh4);
    lo = (lo + Math.imul(al8, bl5)) | 0;
    mid = (mid + Math.imul(al8, bh5)) | 0;
    mid = (mid + Math.imul(ah8, bl5)) | 0;
    hi = (hi + Math.imul(ah8, bh5)) | 0;
    lo = (lo + Math.imul(al7, bl6)) | 0;
    mid = (mid + Math.imul(al7, bh6)) | 0;
    mid = (mid + Math.imul(ah7, bl6)) | 0;
    hi = (hi + Math.imul(ah7, bh6)) | 0;
    lo = (lo + Math.imul(al6, bl7)) | 0;
    mid = (mid + Math.imul(al6, bh7)) | 0;
    mid = (mid + Math.imul(ah6, bl7)) | 0;
    hi = (hi + Math.imul(ah6, bh7)) | 0;
    lo = (lo + Math.imul(al5, bl8)) | 0;
    mid = (mid + Math.imul(al5, bh8)) | 0;
    mid = (mid + Math.imul(ah5, bl8)) | 0;
    hi = (hi + Math.imul(ah5, bh8)) | 0;
    lo = (lo + Math.imul(al4, bl9)) | 0;
    mid = (mid + Math.imul(al4, bh9)) | 0;
    mid = (mid + Math.imul(ah4, bl9)) | 0;
    hi = (hi + Math.imul(ah4, bh9)) | 0;
    var w13 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w13 >>> 26)) | 0;
    w13 &= 0x3ffffff;
    /* k = 14 */
    lo = Math.imul(al9, bl5);
    mid = Math.imul(al9, bh5);
    mid = (mid + Math.imul(ah9, bl5)) | 0;
    hi = Math.imul(ah9, bh5);
    lo = (lo + Math.imul(al8, bl6)) | 0;
    mid = (mid + Math.imul(al8, bh6)) | 0;
    mid = (mid + Math.imul(ah8, bl6)) | 0;
    hi = (hi + Math.imul(ah8, bh6)) | 0;
    lo = (lo + Math.imul(al7, bl7)) | 0;
    mid = (mid + Math.imul(al7, bh7)) | 0;
    mid = (mid + Math.imul(ah7, bl7)) | 0;
    hi = (hi + Math.imul(ah7, bh7)) | 0;
    lo = (lo + Math.imul(al6, bl8)) | 0;
    mid = (mid + Math.imul(al6, bh8)) | 0;
    mid = (mid + Math.imul(ah6, bl8)) | 0;
    hi = (hi + Math.imul(ah6, bh8)) | 0;
    lo = (lo + Math.imul(al5, bl9)) | 0;
    mid = (mid + Math.imul(al5, bh9)) | 0;
    mid = (mid + Math.imul(ah5, bl9)) | 0;
    hi = (hi + Math.imul(ah5, bh9)) | 0;
    var w14 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w14 >>> 26)) | 0;
    w14 &= 0x3ffffff;
    /* k = 15 */
    lo = Math.imul(al9, bl6);
    mid = Math.imul(al9, bh6);
    mid = (mid + Math.imul(ah9, bl6)) | 0;
    hi = Math.imul(ah9, bh6);
    lo = (lo + Math.imul(al8, bl7)) | 0;
    mid = (mid + Math.imul(al8, bh7)) | 0;
    mid = (mid + Math.imul(ah8, bl7)) | 0;
    hi = (hi + Math.imul(ah8, bh7)) | 0;
    lo = (lo + Math.imul(al7, bl8)) | 0;
    mid = (mid + Math.imul(al7, bh8)) | 0;
    mid = (mid + Math.imul(ah7, bl8)) | 0;
    hi = (hi + Math.imul(ah7, bh8)) | 0;
    lo = (lo + Math.imul(al6, bl9)) | 0;
    mid = (mid + Math.imul(al6, bh9)) | 0;
    mid = (mid + Math.imul(ah6, bl9)) | 0;
    hi = (hi + Math.imul(ah6, bh9)) | 0;
    var w15 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w15 >>> 26)) | 0;
    w15 &= 0x3ffffff;
    /* k = 16 */
    lo = Math.imul(al9, bl7);
    mid = Math.imul(al9, bh7);
    mid = (mid + Math.imul(ah9, bl7)) | 0;
    hi = Math.imul(ah9, bh7);
    lo = (lo + Math.imul(al8, bl8)) | 0;
    mid = (mid + Math.imul(al8, bh8)) | 0;
    mid = (mid + Math.imul(ah8, bl8)) | 0;
    hi = (hi + Math.imul(ah8, bh8)) | 0;
    lo = (lo + Math.imul(al7, bl9)) | 0;
    mid = (mid + Math.imul(al7, bh9)) | 0;
    mid = (mid + Math.imul(ah7, bl9)) | 0;
    hi = (hi + Math.imul(ah7, bh9)) | 0;
    var w16 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w16 >>> 26)) | 0;
    w16 &= 0x3ffffff;
    /* k = 17 */
    lo = Math.imul(al9, bl8);
    mid = Math.imul(al9, bh8);
    mid = (mid + Math.imul(ah9, bl8)) | 0;
    hi = Math.imul(ah9, bh8);
    lo = (lo + Math.imul(al8, bl9)) | 0;
    mid = (mid + Math.imul(al8, bh9)) | 0;
    mid = (mid + Math.imul(ah8, bl9)) | 0;
    hi = (hi + Math.imul(ah8, bh9)) | 0;
    var w17 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w17 >>> 26)) | 0;
    w17 &= 0x3ffffff;
    /* k = 18 */
    lo = Math.imul(al9, bl9);
    mid = Math.imul(al9, bh9);
    mid = (mid + Math.imul(ah9, bl9)) | 0;
    hi = Math.imul(ah9, bh9);
    var w18 = (((c + lo) | 0) + ((mid & 0x1fff) << 13)) | 0;
    c = (((hi + (mid >>> 13)) | 0) + (w18 >>> 26)) | 0;
    w18 &= 0x3ffffff;
    o[0] = w0;
    o[1] = w1;
    o[2] = w2;
    o[3] = w3;
    o[4] = w4;
    o[5] = w5;
    o[6] = w6;
    o[7] = w7;
    o[8] = w8;
    o[9] = w9;
    o[10] = w10;
    o[11] = w11;
    o[12] = w12;
    o[13] = w13;
    o[14] = w14;
    o[15] = w15;
    o[16] = w16;
    o[17] = w17;
    o[18] = w18;
    if (c !== 0) ***REMOVED***
      o[19] = c;
      out.length++;
    ***REMOVED***
    return out;
  ***REMOVED***;

  // Polyfill comb
  if (!Math.imul) ***REMOVED***
    comb10MulTo = smallMulTo;
  ***REMOVED***

  function bigMulTo (self, num, out) ***REMOVED***
    out.negative = num.negative ^ self.negative;
    out.length = self.length + num.length;

    var carry = 0;
    var hncarry = 0;
    for (var k = 0; k < out.length - 1; k++) ***REMOVED***
      // Sum all words with the same `i + j = k` and accumulate `ncarry`,
      // note that ncarry could be >= 0x3ffffff
      var ncarry = hncarry;
      hncarry = 0;
      var rword = carry & 0x3ffffff;
      var maxJ = Math.min(k, num.length - 1);
      for (var j = Math.max(0, k - self.length + 1); j <= maxJ; j++) ***REMOVED***
        var i = k - j;
        var a = self.words[i] | 0;
        var b = num.words[j] | 0;
        var r = a * b;

        var lo = r & 0x3ffffff;
        ncarry = (ncarry + ((r / 0x4000000) | 0)) | 0;
        lo = (lo + rword) | 0;
        rword = lo & 0x3ffffff;
        ncarry = (ncarry + (lo >>> 26)) | 0;

        hncarry += ncarry >>> 26;
        ncarry &= 0x3ffffff;
      ***REMOVED***
      out.words[k] = rword;
      carry = ncarry;
      ncarry = hncarry;
    ***REMOVED***
    if (carry !== 0) ***REMOVED***
      out.words[k] = carry;
    ***REMOVED*** else ***REMOVED***
      out.length--;
    ***REMOVED***

    return out.strip();
  ***REMOVED***

  function jumboMulTo (self, num, out) ***REMOVED***
    var fftm = new FFTM();
    return fftm.mulp(self, num, out);
  ***REMOVED***

  BN.prototype.mulTo = function mulTo (num, out) ***REMOVED***
    var res;
    var len = this.length + num.length;
    if (this.length === 10 && num.length === 10) ***REMOVED***
      res = comb10MulTo(this, num, out);
    ***REMOVED*** else if (len < 63) ***REMOVED***
      res = smallMulTo(this, num, out);
    ***REMOVED*** else if (len < 1024) ***REMOVED***
      res = bigMulTo(this, num, out);
    ***REMOVED*** else ***REMOVED***
      res = jumboMulTo(this, num, out);
    ***REMOVED***

    return res;
  ***REMOVED***;

  // Cooley-Tukey algorithm for FFT
  // slightly revisited to rely on looping instead of recursion

  function FFTM (x, y) ***REMOVED***
    this.x = x;
    this.y = y;
  ***REMOVED***

  FFTM.prototype.makeRBT = function makeRBT (N) ***REMOVED***
    var t = new Array(N);
    var l = BN.prototype._countBits(N) - 1;
    for (var i = 0; i < N; i++) ***REMOVED***
      t[i] = this.revBin(i, l, N);
    ***REMOVED***

    return t;
  ***REMOVED***;

  // Returns binary-reversed representation of `x`
  FFTM.prototype.revBin = function revBin (x, l, N) ***REMOVED***
    if (x === 0 || x === N - 1) return x;

    var rb = 0;
    for (var i = 0; i < l; i++) ***REMOVED***
      rb |= (x & 1) << (l - i - 1);
      x >>= 1;
    ***REMOVED***

    return rb;
  ***REMOVED***;

  // Performs "tweedling" phase, therefore 'emulating'
  // behaviour of the recursive algorithm
  FFTM.prototype.permute = function permute (rbt, rws, iws, rtws, itws, N) ***REMOVED***
    for (var i = 0; i < N; i++) ***REMOVED***
      rtws[i] = rws[rbt[i]];
      itws[i] = iws[rbt[i]];
    ***REMOVED***
  ***REMOVED***;

  FFTM.prototype.transform = function transform (rws, iws, rtws, itws, N, rbt) ***REMOVED***
    this.permute(rbt, rws, iws, rtws, itws, N);

    for (var s = 1; s < N; s <<= 1) ***REMOVED***
      var l = s << 1;

      var rtwdf = Math.cos(2 * Math.PI / l);
      var itwdf = Math.sin(2 * Math.PI / l);

      for (var p = 0; p < N; p += l) ***REMOVED***
        var rtwdf_ = rtwdf;
        var itwdf_ = itwdf;

        for (var j = 0; j < s; j++) ***REMOVED***
          var re = rtws[p + j];
          var ie = itws[p + j];

          var ro = rtws[p + j + s];
          var io = itws[p + j + s];

          var rx = rtwdf_ * ro - itwdf_ * io;

          io = rtwdf_ * io + itwdf_ * ro;
          ro = rx;

          rtws[p + j] = re + ro;
          itws[p + j] = ie + io;

          rtws[p + j + s] = re - ro;
          itws[p + j + s] = ie - io;

          /* jshint maxdepth : false */
          if (j !== l) ***REMOVED***
            rx = rtwdf * rtwdf_ - itwdf * itwdf_;

            itwdf_ = rtwdf * itwdf_ + itwdf * rtwdf_;
            rtwdf_ = rx;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;

  FFTM.prototype.guessLen13b = function guessLen13b (n, m) ***REMOVED***
    var N = Math.max(m, n) | 1;
    var odd = N & 1;
    var i = 0;
    for (N = N / 2 | 0; N; N = N >>> 1) ***REMOVED***
      i++;
    ***REMOVED***

    return 1 << i + 1 + odd;
  ***REMOVED***;

  FFTM.prototype.conjugate = function conjugate (rws, iws, N) ***REMOVED***
    if (N <= 1) return;

    for (var i = 0; i < N / 2; i++) ***REMOVED***
      var t = rws[i];

      rws[i] = rws[N - i - 1];
      rws[N - i - 1] = t;

      t = iws[i];

      iws[i] = -iws[N - i - 1];
      iws[N - i - 1] = -t;
    ***REMOVED***
  ***REMOVED***;

  FFTM.prototype.normalize13b = function normalize13b (ws, N) ***REMOVED***
    var carry = 0;
    for (var i = 0; i < N / 2; i++) ***REMOVED***
      var w = Math.round(ws[2 * i + 1] / N) * 0x2000 +
        Math.round(ws[2 * i] / N) +
        carry;

      ws[i] = w & 0x3ffffff;

      if (w < 0x4000000) ***REMOVED***
        carry = 0;
      ***REMOVED*** else ***REMOVED***
        carry = w / 0x4000000 | 0;
      ***REMOVED***
    ***REMOVED***

    return ws;
  ***REMOVED***;

  FFTM.prototype.convert13b = function convert13b (ws, len, rws, N) ***REMOVED***
    var carry = 0;
    for (var i = 0; i < len; i++) ***REMOVED***
      carry = carry + (ws[i] | 0);

      rws[2 * i] = carry & 0x1fff; carry = carry >>> 13;
      rws[2 * i + 1] = carry & 0x1fff; carry = carry >>> 13;
    ***REMOVED***

    // Pad with zeroes
    for (i = 2 * len; i < N; ++i) ***REMOVED***
      rws[i] = 0;
    ***REMOVED***

    assert(carry === 0);
    assert((carry & ~0x1fff) === 0);
  ***REMOVED***;

  FFTM.prototype.stub = function stub (N) ***REMOVED***
    var ph = new Array(N);
    for (var i = 0; i < N; i++) ***REMOVED***
      ph[i] = 0;
    ***REMOVED***

    return ph;
  ***REMOVED***;

  FFTM.prototype.mulp = function mulp (x, y, out) ***REMOVED***
    var N = 2 * this.guessLen13b(x.length, y.length);

    var rbt = this.makeRBT(N);

    var _ = this.stub(N);

    var rws = new Array(N);
    var rwst = new Array(N);
    var iwst = new Array(N);

    var nrws = new Array(N);
    var nrwst = new Array(N);
    var niwst = new Array(N);

    var rmws = out.words;
    rmws.length = N;

    this.convert13b(x.words, x.length, rws, N);
    this.convert13b(y.words, y.length, nrws, N);

    this.transform(rws, _, rwst, iwst, N, rbt);
    this.transform(nrws, _, nrwst, niwst, N, rbt);

    for (var i = 0; i < N; i++) ***REMOVED***
      var rx = rwst[i] * nrwst[i] - iwst[i] * niwst[i];
      iwst[i] = rwst[i] * niwst[i] + iwst[i] * nrwst[i];
      rwst[i] = rx;
    ***REMOVED***

    this.conjugate(rwst, iwst, N);
    this.transform(rwst, iwst, rmws, _, N, rbt);
    this.conjugate(rmws, _, N);
    this.normalize13b(rmws, N);

    out.negative = x.negative ^ y.negative;
    out.length = x.length + y.length;
    return out.strip();
  ***REMOVED***;

  // Multiply `this` by `num`
  BN.prototype.mul = function mul (num) ***REMOVED***
    var out = new BN(null);
    out.words = new Array(this.length + num.length);
    return this.mulTo(num, out);
  ***REMOVED***;

  // Multiply employing FFT
  BN.prototype.mulf = function mulf (num) ***REMOVED***
    var out = new BN(null);
    out.words = new Array(this.length + num.length);
    return jumboMulTo(this, num, out);
  ***REMOVED***;

  // In-place Multiplication
  BN.prototype.imul = function imul (num) ***REMOVED***
    return this.clone().mulTo(num, this);
  ***REMOVED***;

  BN.prototype.imuln = function imuln (num) ***REMOVED***
    assert(typeof num === 'number');
    assert(num < 0x4000000);

    // Carry
    var carry = 0;
    for (var i = 0; i < this.length; i++) ***REMOVED***
      var w = (this.words[i] | 0) * num;
      var lo = (w & 0x3ffffff) + (carry & 0x3ffffff);
      carry >>= 26;
      carry += (w / 0x4000000) | 0;
      // NOTE: lo is 27bit maximum
      carry += lo >>> 26;
      this.words[i] = lo & 0x3ffffff;
    ***REMOVED***

    if (carry !== 0) ***REMOVED***
      this.words[i] = carry;
      this.length++;
    ***REMOVED***

    return this;
  ***REMOVED***;

  BN.prototype.muln = function muln (num) ***REMOVED***
    return this.clone().imuln(num);
  ***REMOVED***;

  // `this` * `this`
  BN.prototype.sqr = function sqr () ***REMOVED***
    return this.mul(this);
  ***REMOVED***;

  // `this` * `this` in-place
  BN.prototype.isqr = function isqr () ***REMOVED***
    return this.imul(this.clone());
  ***REMOVED***;

  // Math.pow(`this`, `num`)
  BN.prototype.pow = function pow (num) ***REMOVED***
    var w = toBitArray(num);
    if (w.length === 0) return new BN(1);

    // Skip leading zeroes
    var res = this;
    for (var i = 0; i < w.length; i++, res = res.sqr()) ***REMOVED***
      if (w[i] !== 0) break;
    ***REMOVED***

    if (++i < w.length) ***REMOVED***
      for (var q = res.sqr(); i < w.length; i++, q = q.sqr()) ***REMOVED***
        if (w[i] === 0) continue;

        res = res.mul(q);
      ***REMOVED***
    ***REMOVED***

    return res;
  ***REMOVED***;

  // Shift-left in-place
  BN.prototype.iushln = function iushln (bits) ***REMOVED***
    assert(typeof bits === 'number' && bits >= 0);
    var r = bits % 26;
    var s = (bits - r) / 26;
    var carryMask = (0x3ffffff >>> (26 - r)) << (26 - r);
    var i;

    if (r !== 0) ***REMOVED***
      var carry = 0;

      for (i = 0; i < this.length; i++) ***REMOVED***
        var newCarry = this.words[i] & carryMask;
        var c = ((this.words[i] | 0) - newCarry) << r;
        this.words[i] = c | carry;
        carry = newCarry >>> (26 - r);
      ***REMOVED***

      if (carry) ***REMOVED***
        this.words[i] = carry;
        this.length++;
      ***REMOVED***
    ***REMOVED***

    if (s !== 0) ***REMOVED***
      for (i = this.length - 1; i >= 0; i--) ***REMOVED***
        this.words[i + s] = this.words[i];
      ***REMOVED***

      for (i = 0; i < s; i++) ***REMOVED***
        this.words[i] = 0;
      ***REMOVED***

      this.length += s;
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  BN.prototype.ishln = function ishln (bits) ***REMOVED***
    // TODO(indutny): implement me
    assert(this.negative === 0);
    return this.iushln(bits);
  ***REMOVED***;

  // Shift-right in-place
  // NOTE: `hint` is a lowest bit before trailing zeroes
  // NOTE: if `extended` is present - it will be filled with destroyed bits
  BN.prototype.iushrn = function iushrn (bits, hint, extended) ***REMOVED***
    assert(typeof bits === 'number' && bits >= 0);
    var h;
    if (hint) ***REMOVED***
      h = (hint - (hint % 26)) / 26;
    ***REMOVED*** else ***REMOVED***
      h = 0;
    ***REMOVED***

    var r = bits % 26;
    var s = Math.min((bits - r) / 26, this.length);
    var mask = 0x3ffffff ^ ((0x3ffffff >>> r) << r);
    var maskedWords = extended;

    h -= s;
    h = Math.max(0, h);

    // Extended mode, copy masked part
    if (maskedWords) ***REMOVED***
      for (var i = 0; i < s; i++) ***REMOVED***
        maskedWords.words[i] = this.words[i];
      ***REMOVED***
      maskedWords.length = s;
    ***REMOVED***

    if (s === 0) ***REMOVED***
      // No-op, we should not move anything at all
    ***REMOVED*** else if (this.length > s) ***REMOVED***
      this.length -= s;
      for (i = 0; i < this.length; i++) ***REMOVED***
        this.words[i] = this.words[i + s];
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      this.words[0] = 0;
      this.length = 1;
    ***REMOVED***

    var carry = 0;
    for (i = this.length - 1; i >= 0 && (carry !== 0 || i >= h); i--) ***REMOVED***
      var word = this.words[i] | 0;
      this.words[i] = (carry << (26 - r)) | (word >>> r);
      carry = word & mask;
    ***REMOVED***

    // Push carried bits as a mask
    if (maskedWords && carry !== 0) ***REMOVED***
      maskedWords.words[maskedWords.length++] = carry;
    ***REMOVED***

    if (this.length === 0) ***REMOVED***
      this.words[0] = 0;
      this.length = 1;
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  BN.prototype.ishrn = function ishrn (bits, hint, extended) ***REMOVED***
    // TODO(indutny): implement me
    assert(this.negative === 0);
    return this.iushrn(bits, hint, extended);
  ***REMOVED***;

  // Shift-left
  BN.prototype.shln = function shln (bits) ***REMOVED***
    return this.clone().ishln(bits);
  ***REMOVED***;

  BN.prototype.ushln = function ushln (bits) ***REMOVED***
    return this.clone().iushln(bits);
  ***REMOVED***;

  // Shift-right
  BN.prototype.shrn = function shrn (bits) ***REMOVED***
    return this.clone().ishrn(bits);
  ***REMOVED***;

  BN.prototype.ushrn = function ushrn (bits) ***REMOVED***
    return this.clone().iushrn(bits);
  ***REMOVED***;

  // Test if n bit is set
  BN.prototype.testn = function testn (bit) ***REMOVED***
    assert(typeof bit === 'number' && bit >= 0);
    var r = bit % 26;
    var s = (bit - r) / 26;
    var q = 1 << r;

    // Fast case: bit is much higher than all existing words
    if (this.length <= s) return false;

    // Check bit and return
    var w = this.words[s];

    return !!(w & q);
  ***REMOVED***;

  // Return only lowers bits of number (in-place)
  BN.prototype.imaskn = function imaskn (bits) ***REMOVED***
    assert(typeof bits === 'number' && bits >= 0);
    var r = bits % 26;
    var s = (bits - r) / 26;

    assert(this.negative === 0, 'imaskn works only with positive numbers');

    if (this.length <= s) ***REMOVED***
      return this;
    ***REMOVED***

    if (r !== 0) ***REMOVED***
      s++;
    ***REMOVED***
    this.length = Math.min(s, this.length);

    if (r !== 0) ***REMOVED***
      var mask = 0x3ffffff ^ ((0x3ffffff >>> r) << r);
      this.words[this.length - 1] &= mask;
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  // Return only lowers bits of number
  BN.prototype.maskn = function maskn (bits) ***REMOVED***
    return this.clone().imaskn(bits);
  ***REMOVED***;

  // Add plain number `num` to `this`
  BN.prototype.iaddn = function iaddn (num) ***REMOVED***
    assert(typeof num === 'number');
    assert(num < 0x4000000);
    if (num < 0) return this.isubn(-num);

    // Possible sign change
    if (this.negative !== 0) ***REMOVED***
      if (this.length === 1 && (this.words[0] | 0) < num) ***REMOVED***
        this.words[0] = num - (this.words[0] | 0);
        this.negative = 0;
        return this;
      ***REMOVED***

      this.negative = 0;
      this.isubn(num);
      this.negative = 1;
      return this;
    ***REMOVED***

    // Add without checks
    return this._iaddn(num);
  ***REMOVED***;

  BN.prototype._iaddn = function _iaddn (num) ***REMOVED***
    this.words[0] += num;

    // Carry
    for (var i = 0; i < this.length && this.words[i] >= 0x4000000; i++) ***REMOVED***
      this.words[i] -= 0x4000000;
      if (i === this.length - 1) ***REMOVED***
        this.words[i + 1] = 1;
      ***REMOVED*** else ***REMOVED***
        this.words[i + 1]++;
      ***REMOVED***
    ***REMOVED***
    this.length = Math.max(this.length, i + 1);

    return this;
  ***REMOVED***;

  // Subtract plain number `num` from `this`
  BN.prototype.isubn = function isubn (num) ***REMOVED***
    assert(typeof num === 'number');
    assert(num < 0x4000000);
    if (num < 0) return this.iaddn(-num);

    if (this.negative !== 0) ***REMOVED***
      this.negative = 0;
      this.iaddn(num);
      this.negative = 1;
      return this;
    ***REMOVED***

    this.words[0] -= num;

    if (this.length === 1 && this.words[0] < 0) ***REMOVED***
      this.words[0] = -this.words[0];
      this.negative = 1;
    ***REMOVED*** else ***REMOVED***
      // Carry
      for (var i = 0; i < this.length && this.words[i] < 0; i++) ***REMOVED***
        this.words[i] += 0x4000000;
        this.words[i + 1] -= 1;
      ***REMOVED***
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  BN.prototype.addn = function addn (num) ***REMOVED***
    return this.clone().iaddn(num);
  ***REMOVED***;

  BN.prototype.subn = function subn (num) ***REMOVED***
    return this.clone().isubn(num);
  ***REMOVED***;

  BN.prototype.iabs = function iabs () ***REMOVED***
    this.negative = 0;

    return this;
  ***REMOVED***;

  BN.prototype.abs = function abs () ***REMOVED***
    return this.clone().iabs();
  ***REMOVED***;

  BN.prototype._ishlnsubmul = function _ishlnsubmul (num, mul, shift) ***REMOVED***
    var len = num.length + shift;
    var i;

    this._expand(len);

    var w;
    var carry = 0;
    for (i = 0; i < num.length; i++) ***REMOVED***
      w = (this.words[i + shift] | 0) + carry;
      var right = (num.words[i] | 0) * mul;
      w -= right & 0x3ffffff;
      carry = (w >> 26) - ((right / 0x4000000) | 0);
      this.words[i + shift] = w & 0x3ffffff;
    ***REMOVED***
    for (; i < this.length - shift; i++) ***REMOVED***
      w = (this.words[i + shift] | 0) + carry;
      carry = w >> 26;
      this.words[i + shift] = w & 0x3ffffff;
    ***REMOVED***

    if (carry === 0) return this.strip();

    // Subtraction overflow
    assert(carry === -1);
    carry = 0;
    for (i = 0; i < this.length; i++) ***REMOVED***
      w = -(this.words[i] | 0) + carry;
      carry = w >> 26;
      this.words[i] = w & 0x3ffffff;
    ***REMOVED***
    this.negative = 1;

    return this.strip();
  ***REMOVED***;

  BN.prototype._wordDiv = function _wordDiv (num, mode) ***REMOVED***
    var shift = this.length - num.length;

    var a = this.clone();
    var b = num;

    // Normalize
    var bhi = b.words[b.length - 1] | 0;
    var bhiBits = this._countBits(bhi);
    shift = 26 - bhiBits;
    if (shift !== 0) ***REMOVED***
      b = b.ushln(shift);
      a.iushln(shift);
      bhi = b.words[b.length - 1] | 0;
    ***REMOVED***

    // Initialize quotient
    var m = a.length - b.length;
    var q;

    if (mode !== 'mod') ***REMOVED***
      q = new BN(null);
      q.length = m + 1;
      q.words = new Array(q.length);
      for (var i = 0; i < q.length; i++) ***REMOVED***
        q.words[i] = 0;
      ***REMOVED***
    ***REMOVED***

    var diff = a.clone()._ishlnsubmul(b, 1, m);
    if (diff.negative === 0) ***REMOVED***
      a = diff;
      if (q) ***REMOVED***
        q.words[m] = 1;
      ***REMOVED***
    ***REMOVED***

    for (var j = m - 1; j >= 0; j--) ***REMOVED***
      var qj = (a.words[b.length + j] | 0) * 0x4000000 +
        (a.words[b.length + j - 1] | 0);

      // NOTE: (qj / bhi) is (0x3ffffff * 0x4000000 + 0x3ffffff) / 0x2000000 max
      // (0x7ffffff)
      qj = Math.min((qj / bhi) | 0, 0x3ffffff);

      a._ishlnsubmul(b, qj, j);
      while (a.negative !== 0) ***REMOVED***
        qj--;
        a.negative = 0;
        a._ishlnsubmul(b, 1, j);
        if (!a.isZero()) ***REMOVED***
          a.negative ^= 1;
        ***REMOVED***
      ***REMOVED***
      if (q) ***REMOVED***
        q.words[j] = qj;
      ***REMOVED***
    ***REMOVED***
    if (q) ***REMOVED***
      q.strip();
    ***REMOVED***
    a.strip();

    // Denormalize
    if (mode !== 'div' && shift !== 0) ***REMOVED***
      a.iushrn(shift);
    ***REMOVED***

    return ***REMOVED***
      div: q || null,
      mod: a
    ***REMOVED***;
  ***REMOVED***;

  // NOTE: 1) `mode` can be set to `mod` to request mod only,
  //       to `div` to request div only, or be absent to
  //       request both div & mod
  //       2) `positive` is true if unsigned mod is requested
  BN.prototype.divmod = function divmod (num, mode, positive) ***REMOVED***
    assert(!num.isZero());

    if (this.isZero()) ***REMOVED***
      return ***REMOVED***
        div: new BN(0),
        mod: new BN(0)
      ***REMOVED***;
    ***REMOVED***

    var div, mod, res;
    if (this.negative !== 0 && num.negative === 0) ***REMOVED***
      res = this.neg().divmod(num, mode);

      if (mode !== 'mod') ***REMOVED***
        div = res.div.neg();
      ***REMOVED***

      if (mode !== 'div') ***REMOVED***
        mod = res.mod.neg();
        if (positive && mod.negative !== 0) ***REMOVED***
          mod.iadd(num);
        ***REMOVED***
      ***REMOVED***

      return ***REMOVED***
        div: div,
        mod: mod
      ***REMOVED***;
    ***REMOVED***

    if (this.negative === 0 && num.negative !== 0) ***REMOVED***
      res = this.divmod(num.neg(), mode);

      if (mode !== 'mod') ***REMOVED***
        div = res.div.neg();
      ***REMOVED***

      return ***REMOVED***
        div: div,
        mod: res.mod
      ***REMOVED***;
    ***REMOVED***

    if ((this.negative & num.negative) !== 0) ***REMOVED***
      res = this.neg().divmod(num.neg(), mode);

      if (mode !== 'div') ***REMOVED***
        mod = res.mod.neg();
        if (positive && mod.negative !== 0) ***REMOVED***
          mod.isub(num);
        ***REMOVED***
      ***REMOVED***

      return ***REMOVED***
        div: res.div,
        mod: mod
      ***REMOVED***;
    ***REMOVED***

    // Both numbers are positive at this point

    // Strip both numbers to approximate shift value
    if (num.length > this.length || this.cmp(num) < 0) ***REMOVED***
      return ***REMOVED***
        div: new BN(0),
        mod: this
      ***REMOVED***;
    ***REMOVED***

    // Very short reduction
    if (num.length === 1) ***REMOVED***
      if (mode === 'div') ***REMOVED***
        return ***REMOVED***
          div: this.divn(num.words[0]),
          mod: null
        ***REMOVED***;
      ***REMOVED***

      if (mode === 'mod') ***REMOVED***
        return ***REMOVED***
          div: null,
          mod: new BN(this.modn(num.words[0]))
        ***REMOVED***;
      ***REMOVED***

      return ***REMOVED***
        div: this.divn(num.words[0]),
        mod: new BN(this.modn(num.words[0]))
      ***REMOVED***;
    ***REMOVED***

    return this._wordDiv(num, mode);
  ***REMOVED***;

  // Find `this` / `num`
  BN.prototype.div = function div (num) ***REMOVED***
    return this.divmod(num, 'div', false).div;
  ***REMOVED***;

  // Find `this` % `num`
  BN.prototype.mod = function mod (num) ***REMOVED***
    return this.divmod(num, 'mod', false).mod;
  ***REMOVED***;

  BN.prototype.umod = function umod (num) ***REMOVED***
    return this.divmod(num, 'mod', true).mod;
  ***REMOVED***;

  // Find Round(`this` / `num`)
  BN.prototype.divRound = function divRound (num) ***REMOVED***
    var dm = this.divmod(num);

    // Fast case - exact division
    if (dm.mod.isZero()) return dm.div;

    var mod = dm.div.negative !== 0 ? dm.mod.isub(num) : dm.mod;

    var half = num.ushrn(1);
    var r2 = num.andln(1);
    var cmp = mod.cmp(half);

    // Round down
    if (cmp < 0 || r2 === 1 && cmp === 0) return dm.div;

    // Round up
    return dm.div.negative !== 0 ? dm.div.isubn(1) : dm.div.iaddn(1);
  ***REMOVED***;

  BN.prototype.modn = function modn (num) ***REMOVED***
    assert(num <= 0x3ffffff);
    var p = (1 << 26) % num;

    var acc = 0;
    for (var i = this.length - 1; i >= 0; i--) ***REMOVED***
      acc = (p * acc + (this.words[i] | 0)) % num;
    ***REMOVED***

    return acc;
  ***REMOVED***;

  // In-place division by number
  BN.prototype.idivn = function idivn (num) ***REMOVED***
    assert(num <= 0x3ffffff);

    var carry = 0;
    for (var i = this.length - 1; i >= 0; i--) ***REMOVED***
      var w = (this.words[i] | 0) + carry * 0x4000000;
      this.words[i] = (w / num) | 0;
      carry = w % num;
    ***REMOVED***

    return this.strip();
  ***REMOVED***;

  BN.prototype.divn = function divn (num) ***REMOVED***
    return this.clone().idivn(num);
  ***REMOVED***;

  BN.prototype.egcd = function egcd (p) ***REMOVED***
    assert(p.negative === 0);
    assert(!p.isZero());

    var x = this;
    var y = p.clone();

    if (x.negative !== 0) ***REMOVED***
      x = x.umod(p);
    ***REMOVED*** else ***REMOVED***
      x = x.clone();
    ***REMOVED***

    // A * x + B * y = x
    var A = new BN(1);
    var B = new BN(0);

    // C * x + D * y = y
    var C = new BN(0);
    var D = new BN(1);

    var g = 0;

    while (x.isEven() && y.isEven()) ***REMOVED***
      x.iushrn(1);
      y.iushrn(1);
      ++g;
    ***REMOVED***

    var yp = y.clone();
    var xp = x.clone();

    while (!x.isZero()) ***REMOVED***
      for (var i = 0, im = 1; (x.words[0] & im) === 0 && i < 26; ++i, im <<= 1);
      if (i > 0) ***REMOVED***
        x.iushrn(i);
        while (i-- > 0) ***REMOVED***
          if (A.isOdd() || B.isOdd()) ***REMOVED***
            A.iadd(yp);
            B.isub(xp);
          ***REMOVED***

          A.iushrn(1);
          B.iushrn(1);
        ***REMOVED***
      ***REMOVED***

      for (var j = 0, jm = 1; (y.words[0] & jm) === 0 && j < 26; ++j, jm <<= 1);
      if (j > 0) ***REMOVED***
        y.iushrn(j);
        while (j-- > 0) ***REMOVED***
          if (C.isOdd() || D.isOdd()) ***REMOVED***
            C.iadd(yp);
            D.isub(xp);
          ***REMOVED***

          C.iushrn(1);
          D.iushrn(1);
        ***REMOVED***
      ***REMOVED***

      if (x.cmp(y) >= 0) ***REMOVED***
        x.isub(y);
        A.isub(C);
        B.isub(D);
      ***REMOVED*** else ***REMOVED***
        y.isub(x);
        C.isub(A);
        D.isub(B);
      ***REMOVED***
    ***REMOVED***

    return ***REMOVED***
      a: C,
      b: D,
      gcd: y.iushln(g)
    ***REMOVED***;
  ***REMOVED***;

  // This is reduced incarnation of the binary EEA
  // above, designated to invert members of the
  // _prime_ fields F(p) at a maximal speed
  BN.prototype._invmp = function _invmp (p) ***REMOVED***
    assert(p.negative === 0);
    assert(!p.isZero());

    var a = this;
    var b = p.clone();

    if (a.negative !== 0) ***REMOVED***
      a = a.umod(p);
    ***REMOVED*** else ***REMOVED***
      a = a.clone();
    ***REMOVED***

    var x1 = new BN(1);
    var x2 = new BN(0);

    var delta = b.clone();

    while (a.cmpn(1) > 0 && b.cmpn(1) > 0) ***REMOVED***
      for (var i = 0, im = 1; (a.words[0] & im) === 0 && i < 26; ++i, im <<= 1);
      if (i > 0) ***REMOVED***
        a.iushrn(i);
        while (i-- > 0) ***REMOVED***
          if (x1.isOdd()) ***REMOVED***
            x1.iadd(delta);
          ***REMOVED***

          x1.iushrn(1);
        ***REMOVED***
      ***REMOVED***

      for (var j = 0, jm = 1; (b.words[0] & jm) === 0 && j < 26; ++j, jm <<= 1);
      if (j > 0) ***REMOVED***
        b.iushrn(j);
        while (j-- > 0) ***REMOVED***
          if (x2.isOdd()) ***REMOVED***
            x2.iadd(delta);
          ***REMOVED***

          x2.iushrn(1);
        ***REMOVED***
      ***REMOVED***

      if (a.cmp(b) >= 0) ***REMOVED***
        a.isub(b);
        x1.isub(x2);
      ***REMOVED*** else ***REMOVED***
        b.isub(a);
        x2.isub(x1);
      ***REMOVED***
    ***REMOVED***

    var res;
    if (a.cmpn(1) === 0) ***REMOVED***
      res = x1;
    ***REMOVED*** else ***REMOVED***
      res = x2;
    ***REMOVED***

    if (res.cmpn(0) < 0) ***REMOVED***
      res.iadd(p);
    ***REMOVED***

    return res;
  ***REMOVED***;

  BN.prototype.gcd = function gcd (num) ***REMOVED***
    if (this.isZero()) return num.abs();
    if (num.isZero()) return this.abs();

    var a = this.clone();
    var b = num.clone();
    a.negative = 0;
    b.negative = 0;

    // Remove common factor of two
    for (var shift = 0; a.isEven() && b.isEven(); shift++) ***REMOVED***
      a.iushrn(1);
      b.iushrn(1);
    ***REMOVED***

    do ***REMOVED***
      while (a.isEven()) ***REMOVED***
        a.iushrn(1);
      ***REMOVED***
      while (b.isEven()) ***REMOVED***
        b.iushrn(1);
      ***REMOVED***

      var r = a.cmp(b);
      if (r < 0) ***REMOVED***
        // Swap `a` and `b` to make `a` always bigger than `b`
        var t = a;
        a = b;
        b = t;
      ***REMOVED*** else if (r === 0 || b.cmpn(1) === 0) ***REMOVED***
        break;
      ***REMOVED***

      a.isub(b);
    ***REMOVED*** while (true);

    return b.iushln(shift);
  ***REMOVED***;

  // Invert number in the field F(num)
  BN.prototype.invm = function invm (num) ***REMOVED***
    return this.egcd(num).a.umod(num);
  ***REMOVED***;

  BN.prototype.isEven = function isEven () ***REMOVED***
    return (this.words[0] & 1) === 0;
  ***REMOVED***;

  BN.prototype.isOdd = function isOdd () ***REMOVED***
    return (this.words[0] & 1) === 1;
  ***REMOVED***;

  // And first word and num
  BN.prototype.andln = function andln (num) ***REMOVED***
    return this.words[0] & num;
  ***REMOVED***;

  // Increment at the bit position in-line
  BN.prototype.bincn = function bincn (bit) ***REMOVED***
    assert(typeof bit === 'number');
    var r = bit % 26;
    var s = (bit - r) / 26;
    var q = 1 << r;

    // Fast case: bit is much higher than all existing words
    if (this.length <= s) ***REMOVED***
      this._expand(s + 1);
      this.words[s] |= q;
      return this;
    ***REMOVED***

    // Add bit and propagate, if needed
    var carry = q;
    for (var i = s; carry !== 0 && i < this.length; i++) ***REMOVED***
      var w = this.words[i] | 0;
      w += carry;
      carry = w >>> 26;
      w &= 0x3ffffff;
      this.words[i] = w;
    ***REMOVED***
    if (carry !== 0) ***REMOVED***
      this.words[i] = carry;
      this.length++;
    ***REMOVED***
    return this;
  ***REMOVED***;

  BN.prototype.isZero = function isZero () ***REMOVED***
    return this.length === 1 && this.words[0] === 0;
  ***REMOVED***;

  BN.prototype.cmpn = function cmpn (num) ***REMOVED***
    var negative = num < 0;

    if (this.negative !== 0 && !negative) return -1;
    if (this.negative === 0 && negative) return 1;

    this.strip();

    var res;
    if (this.length > 1) ***REMOVED***
      res = 1;
    ***REMOVED*** else ***REMOVED***
      if (negative) ***REMOVED***
        num = -num;
      ***REMOVED***

      assert(num <= 0x3ffffff, 'Number is too big');

      var w = this.words[0] | 0;
      res = w === num ? 0 : w < num ? -1 : 1;
    ***REMOVED***
    if (this.negative !== 0) return -res | 0;
    return res;
  ***REMOVED***;

  // Compare two numbers and return:
  // 1 - if `this` > `num`
  // 0 - if `this` == `num`
  // -1 - if `this` < `num`
  BN.prototype.cmp = function cmp (num) ***REMOVED***
    if (this.negative !== 0 && num.negative === 0) return -1;
    if (this.negative === 0 && num.negative !== 0) return 1;

    var res = this.ucmp(num);
    if (this.negative !== 0) return -res | 0;
    return res;
  ***REMOVED***;

  // Unsigned comparison
  BN.prototype.ucmp = function ucmp (num) ***REMOVED***
    // At this point both numbers have the same sign
    if (this.length > num.length) return 1;
    if (this.length < num.length) return -1;

    var res = 0;
    for (var i = this.length - 1; i >= 0; i--) ***REMOVED***
      var a = this.words[i] | 0;
      var b = num.words[i] | 0;

      if (a === b) continue;
      if (a < b) ***REMOVED***
        res = -1;
      ***REMOVED*** else if (a > b) ***REMOVED***
        res = 1;
      ***REMOVED***
      break;
    ***REMOVED***
    return res;
  ***REMOVED***;

  BN.prototype.gtn = function gtn (num) ***REMOVED***
    return this.cmpn(num) === 1;
  ***REMOVED***;

  BN.prototype.gt = function gt (num) ***REMOVED***
    return this.cmp(num) === 1;
  ***REMOVED***;

  BN.prototype.gten = function gten (num) ***REMOVED***
    return this.cmpn(num) >= 0;
  ***REMOVED***;

  BN.prototype.gte = function gte (num) ***REMOVED***
    return this.cmp(num) >= 0;
  ***REMOVED***;

  BN.prototype.ltn = function ltn (num) ***REMOVED***
    return this.cmpn(num) === -1;
  ***REMOVED***;

  BN.prototype.lt = function lt (num) ***REMOVED***
    return this.cmp(num) === -1;
  ***REMOVED***;

  BN.prototype.lten = function lten (num) ***REMOVED***
    return this.cmpn(num) <= 0;
  ***REMOVED***;

  BN.prototype.lte = function lte (num) ***REMOVED***
    return this.cmp(num) <= 0;
  ***REMOVED***;

  BN.prototype.eqn = function eqn (num) ***REMOVED***
    return this.cmpn(num) === 0;
  ***REMOVED***;

  BN.prototype.eq = function eq (num) ***REMOVED***
    return this.cmp(num) === 0;
  ***REMOVED***;

  //
  // A reduce context, could be using montgomery or something better, depending
  // on the `m` itself.
  //
  BN.red = function red (num) ***REMOVED***
    return new Red(num);
  ***REMOVED***;

  BN.prototype.toRed = function toRed (ctx) ***REMOVED***
    assert(!this.red, 'Already a number in reduction context');
    assert(this.negative === 0, 'red works only with positives');
    return ctx.convertTo(this)._forceRed(ctx);
  ***REMOVED***;

  BN.prototype.fromRed = function fromRed () ***REMOVED***
    assert(this.red, 'fromRed works only with numbers in reduction context');
    return this.red.convertFrom(this);
  ***REMOVED***;

  BN.prototype._forceRed = function _forceRed (ctx) ***REMOVED***
    this.red = ctx;
    return this;
  ***REMOVED***;

  BN.prototype.forceRed = function forceRed (ctx) ***REMOVED***
    assert(!this.red, 'Already a number in reduction context');
    return this._forceRed(ctx);
  ***REMOVED***;

  BN.prototype.redAdd = function redAdd (num) ***REMOVED***
    assert(this.red, 'redAdd works only with red numbers');
    return this.red.add(this, num);
  ***REMOVED***;

  BN.prototype.redIAdd = function redIAdd (num) ***REMOVED***
    assert(this.red, 'redIAdd works only with red numbers');
    return this.red.iadd(this, num);
  ***REMOVED***;

  BN.prototype.redSub = function redSub (num) ***REMOVED***
    assert(this.red, 'redSub works only with red numbers');
    return this.red.sub(this, num);
  ***REMOVED***;

  BN.prototype.redISub = function redISub (num) ***REMOVED***
    assert(this.red, 'redISub works only with red numbers');
    return this.red.isub(this, num);
  ***REMOVED***;

  BN.prototype.redShl = function redShl (num) ***REMOVED***
    assert(this.red, 'redShl works only with red numbers');
    return this.red.shl(this, num);
  ***REMOVED***;

  BN.prototype.redMul = function redMul (num) ***REMOVED***
    assert(this.red, 'redMul works only with red numbers');
    this.red._verify2(this, num);
    return this.red.mul(this, num);
  ***REMOVED***;

  BN.prototype.redIMul = function redIMul (num) ***REMOVED***
    assert(this.red, 'redMul works only with red numbers');
    this.red._verify2(this, num);
    return this.red.imul(this, num);
  ***REMOVED***;

  BN.prototype.redSqr = function redSqr () ***REMOVED***
    assert(this.red, 'redSqr works only with red numbers');
    this.red._verify1(this);
    return this.red.sqr(this);
  ***REMOVED***;

  BN.prototype.redISqr = function redISqr () ***REMOVED***
    assert(this.red, 'redISqr works only with red numbers');
    this.red._verify1(this);
    return this.red.isqr(this);
  ***REMOVED***;

  // Square root over p
  BN.prototype.redSqrt = function redSqrt () ***REMOVED***
    assert(this.red, 'redSqrt works only with red numbers');
    this.red._verify1(this);
    return this.red.sqrt(this);
  ***REMOVED***;

  BN.prototype.redInvm = function redInvm () ***REMOVED***
    assert(this.red, 'redInvm works only with red numbers');
    this.red._verify1(this);
    return this.red.invm(this);
  ***REMOVED***;

  // Return negative clone of `this` % `red modulo`
  BN.prototype.redNeg = function redNeg () ***REMOVED***
    assert(this.red, 'redNeg works only with red numbers');
    this.red._verify1(this);
    return this.red.neg(this);
  ***REMOVED***;

  BN.prototype.redPow = function redPow (num) ***REMOVED***
    assert(this.red && !num.red, 'redPow(normalNum)');
    this.red._verify1(this);
    return this.red.pow(this, num);
  ***REMOVED***;

  // Prime numbers with efficient reduction
  var primes = ***REMOVED***
    k256: null,
    p224: null,
    p192: null,
    p25519: null
  ***REMOVED***;

  // Pseudo-Mersenne prime
  function MPrime (name, p) ***REMOVED***
    // P = 2 ^ N - K
    this.name = name;
    this.p = new BN(p, 16);
    this.n = this.p.bitLength();
    this.k = new BN(1).iushln(this.n).isub(this.p);

    this.tmp = this._tmp();
  ***REMOVED***

  MPrime.prototype._tmp = function _tmp () ***REMOVED***
    var tmp = new BN(null);
    tmp.words = new Array(Math.ceil(this.n / 13));
    return tmp;
  ***REMOVED***;

  MPrime.prototype.ireduce = function ireduce (num) ***REMOVED***
    // Assumes that `num` is less than `P^2`
    // num = HI * (2 ^ N - K) + HI * K + LO = HI * K + LO (mod P)
    var r = num;
    var rlen;

    do ***REMOVED***
      this.split(r, this.tmp);
      r = this.imulK(r);
      r = r.iadd(this.tmp);
      rlen = r.bitLength();
    ***REMOVED*** while (rlen > this.n);

    var cmp = rlen < this.n ? -1 : r.ucmp(this.p);
    if (cmp === 0) ***REMOVED***
      r.words[0] = 0;
      r.length = 1;
    ***REMOVED*** else if (cmp > 0) ***REMOVED***
      r.isub(this.p);
    ***REMOVED*** else ***REMOVED***
      r.strip();
    ***REMOVED***

    return r;
  ***REMOVED***;

  MPrime.prototype.split = function split (input, out) ***REMOVED***
    input.iushrn(this.n, 0, out);
  ***REMOVED***;

  MPrime.prototype.imulK = function imulK (num) ***REMOVED***
    return num.imul(this.k);
  ***REMOVED***;

  function K256 () ***REMOVED***
    MPrime.call(
      this,
      'k256',
      'ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff fffffffe fffffc2f');
  ***REMOVED***
  inherits(K256, MPrime);

  K256.prototype.split = function split (input, output) ***REMOVED***
    // 256 = 9 * 26 + 22
    var mask = 0x3fffff;

    var outLen = Math.min(input.length, 9);
    for (var i = 0; i < outLen; i++) ***REMOVED***
      output.words[i] = input.words[i];
    ***REMOVED***
    output.length = outLen;

    if (input.length <= 9) ***REMOVED***
      input.words[0] = 0;
      input.length = 1;
      return;
    ***REMOVED***

    // Shift by 9 limbs
    var prev = input.words[9];
    output.words[output.length++] = prev & mask;

    for (i = 10; i < input.length; i++) ***REMOVED***
      var next = input.words[i] | 0;
      input.words[i - 10] = ((next & mask) << 4) | (prev >>> 22);
      prev = next;
    ***REMOVED***
    prev >>>= 22;
    input.words[i - 10] = prev;
    if (prev === 0 && input.length > 10) ***REMOVED***
      input.length -= 10;
    ***REMOVED*** else ***REMOVED***
      input.length -= 9;
    ***REMOVED***
  ***REMOVED***;

  K256.prototype.imulK = function imulK (num) ***REMOVED***
    // K = 0x1000003d1 = [ 0x40, 0x3d1 ]
    num.words[num.length] = 0;
    num.words[num.length + 1] = 0;
    num.length += 2;

    // bounded at: 0x40 * 0x3ffffff + 0x3d0 = 0x100000390
    var lo = 0;
    for (var i = 0; i < num.length; i++) ***REMOVED***
      var w = num.words[i] | 0;
      lo += w * 0x3d1;
      num.words[i] = lo & 0x3ffffff;
      lo = w * 0x40 + ((lo / 0x4000000) | 0);
    ***REMOVED***

    // Fast length reduction
    if (num.words[num.length - 1] === 0) ***REMOVED***
      num.length--;
      if (num.words[num.length - 1] === 0) ***REMOVED***
        num.length--;
      ***REMOVED***
    ***REMOVED***
    return num;
  ***REMOVED***;

  function P224 () ***REMOVED***
    MPrime.call(
      this,
      'p224',
      'ffffffff ffffffff ffffffff ffffffff 00000000 00000000 00000001');
  ***REMOVED***
  inherits(P224, MPrime);

  function P192 () ***REMOVED***
    MPrime.call(
      this,
      'p192',
      'ffffffff ffffffff ffffffff fffffffe ffffffff ffffffff');
  ***REMOVED***
  inherits(P192, MPrime);

  function P25519 () ***REMOVED***
    // 2 ^ 255 - 19
    MPrime.call(
      this,
      '25519',
      '7fffffffffffffff ffffffffffffffff ffffffffffffffff ffffffffffffffed');
  ***REMOVED***
  inherits(P25519, MPrime);

  P25519.prototype.imulK = function imulK (num) ***REMOVED***
    // K = 0x13
    var carry = 0;
    for (var i = 0; i < num.length; i++) ***REMOVED***
      var hi = (num.words[i] | 0) * 0x13 + carry;
      var lo = hi & 0x3ffffff;
      hi >>>= 26;

      num.words[i] = lo;
      carry = hi;
    ***REMOVED***
    if (carry !== 0) ***REMOVED***
      num.words[num.length++] = carry;
    ***REMOVED***
    return num;
  ***REMOVED***;

  // Exported mostly for testing purposes, use plain name instead
  BN._prime = function prime (name) ***REMOVED***
    // Cached version of prime
    if (primes[name]) return primes[name];

    var prime;
    if (name === 'k256') ***REMOVED***
      prime = new K256();
    ***REMOVED*** else if (name === 'p224') ***REMOVED***
      prime = new P224();
    ***REMOVED*** else if (name === 'p192') ***REMOVED***
      prime = new P192();
    ***REMOVED*** else if (name === 'p25519') ***REMOVED***
      prime = new P25519();
    ***REMOVED*** else ***REMOVED***
      throw new Error('Unknown prime ' + name);
    ***REMOVED***
    primes[name] = prime;

    return prime;
  ***REMOVED***;

  //
  // Base reduction engine
  //
  function Red (m) ***REMOVED***
    if (typeof m === 'string') ***REMOVED***
      var prime = BN._prime(m);
      this.m = prime.p;
      this.prime = prime;
    ***REMOVED*** else ***REMOVED***
      assert(m.gtn(1), 'modulus must be greater than 1');
      this.m = m;
      this.prime = null;
    ***REMOVED***
  ***REMOVED***

  Red.prototype._verify1 = function _verify1 (a) ***REMOVED***
    assert(a.negative === 0, 'red works only with positives');
    assert(a.red, 'red works only with red numbers');
  ***REMOVED***;

  Red.prototype._verify2 = function _verify2 (a, b) ***REMOVED***
    assert((a.negative | b.negative) === 0, 'red works only with positives');
    assert(a.red && a.red === b.red,
      'red works only with red numbers');
  ***REMOVED***;

  Red.prototype.imod = function imod (a) ***REMOVED***
    if (this.prime) return this.prime.ireduce(a)._forceRed(this);
    return a.umod(this.m)._forceRed(this);
  ***REMOVED***;

  Red.prototype.neg = function neg (a) ***REMOVED***
    if (a.isZero()) ***REMOVED***
      return a.clone();
    ***REMOVED***

    return this.m.sub(a)._forceRed(this);
  ***REMOVED***;

  Red.prototype.add = function add (a, b) ***REMOVED***
    this._verify2(a, b);

    var res = a.add(b);
    if (res.cmp(this.m) >= 0) ***REMOVED***
      res.isub(this.m);
    ***REMOVED***
    return res._forceRed(this);
  ***REMOVED***;

  Red.prototype.iadd = function iadd (a, b) ***REMOVED***
    this._verify2(a, b);

    var res = a.iadd(b);
    if (res.cmp(this.m) >= 0) ***REMOVED***
      res.isub(this.m);
    ***REMOVED***
    return res;
  ***REMOVED***;

  Red.prototype.sub = function sub (a, b) ***REMOVED***
    this._verify2(a, b);

    var res = a.sub(b);
    if (res.cmpn(0) < 0) ***REMOVED***
      res.iadd(this.m);
    ***REMOVED***
    return res._forceRed(this);
  ***REMOVED***;

  Red.prototype.isub = function isub (a, b) ***REMOVED***
    this._verify2(a, b);

    var res = a.isub(b);
    if (res.cmpn(0) < 0) ***REMOVED***
      res.iadd(this.m);
    ***REMOVED***
    return res;
  ***REMOVED***;

  Red.prototype.shl = function shl (a, num) ***REMOVED***
    this._verify1(a);
    return this.imod(a.ushln(num));
  ***REMOVED***;

  Red.prototype.imul = function imul (a, b) ***REMOVED***
    this._verify2(a, b);
    return this.imod(a.imul(b));
  ***REMOVED***;

  Red.prototype.mul = function mul (a, b) ***REMOVED***
    this._verify2(a, b);
    return this.imod(a.mul(b));
  ***REMOVED***;

  Red.prototype.isqr = function isqr (a) ***REMOVED***
    return this.imul(a, a.clone());
  ***REMOVED***;

  Red.prototype.sqr = function sqr (a) ***REMOVED***
    return this.mul(a, a);
  ***REMOVED***;

  Red.prototype.sqrt = function sqrt (a) ***REMOVED***
    if (a.isZero()) return a.clone();

    var mod3 = this.m.andln(3);
    assert(mod3 % 2 === 1);

    // Fast case
    if (mod3 === 3) ***REMOVED***
      var pow = this.m.add(new BN(1)).iushrn(2);
      return this.pow(a, pow);
    ***REMOVED***

    // Tonelli-Shanks algorithm (Totally unoptimized and slow)
    //
    // Find Q and S, that Q * 2 ^ S = (P - 1)
    var q = this.m.subn(1);
    var s = 0;
    while (!q.isZero() && q.andln(1) === 0) ***REMOVED***
      s++;
      q.iushrn(1);
    ***REMOVED***
    assert(!q.isZero());

    var one = new BN(1).toRed(this);
    var nOne = one.redNeg();

    // Find quadratic non-residue
    // NOTE: Max is such because of generalized Riemann hypothesis.
    var lpow = this.m.subn(1).iushrn(1);
    var z = this.m.bitLength();
    z = new BN(2 * z * z).toRed(this);

    while (this.pow(z, lpow).cmp(nOne) !== 0) ***REMOVED***
      z.redIAdd(nOne);
    ***REMOVED***

    var c = this.pow(z, q);
    var r = this.pow(a, q.addn(1).iushrn(1));
    var t = this.pow(a, q);
    var m = s;
    while (t.cmp(one) !== 0) ***REMOVED***
      var tmp = t;
      for (var i = 0; tmp.cmp(one) !== 0; i++) ***REMOVED***
        tmp = tmp.redSqr();
      ***REMOVED***
      assert(i < m);
      var b = this.pow(c, new BN(1).iushln(m - i - 1));

      r = r.redMul(b);
      c = b.redSqr();
      t = t.redMul(c);
      m = i;
    ***REMOVED***

    return r;
  ***REMOVED***;

  Red.prototype.invm = function invm (a) ***REMOVED***
    var inv = a._invmp(this.m);
    if (inv.negative !== 0) ***REMOVED***
      inv.negative = 0;
      return this.imod(inv).redNeg();
    ***REMOVED*** else ***REMOVED***
      return this.imod(inv);
    ***REMOVED***
  ***REMOVED***;

  Red.prototype.pow = function pow (a, num) ***REMOVED***
    if (num.isZero()) return new BN(1).toRed(this);
    if (num.cmpn(1) === 0) return a.clone();

    var windowSize = 4;
    var wnd = new Array(1 << windowSize);
    wnd[0] = new BN(1).toRed(this);
    wnd[1] = a;
    for (var i = 2; i < wnd.length; i++) ***REMOVED***
      wnd[i] = this.mul(wnd[i - 1], a);
    ***REMOVED***

    var res = wnd[0];
    var current = 0;
    var currentLen = 0;
    var start = num.bitLength() % 26;
    if (start === 0) ***REMOVED***
      start = 26;
    ***REMOVED***

    for (i = num.length - 1; i >= 0; i--) ***REMOVED***
      var word = num.words[i];
      for (var j = start - 1; j >= 0; j--) ***REMOVED***
        var bit = (word >> j) & 1;
        if (res !== wnd[0]) ***REMOVED***
          res = this.sqr(res);
        ***REMOVED***

        if (bit === 0 && current === 0) ***REMOVED***
          currentLen = 0;
          continue;
        ***REMOVED***

        current <<= 1;
        current |= bit;
        currentLen++;
        if (currentLen !== windowSize && (i !== 0 || j !== 0)) continue;

        res = this.mul(res, wnd[current]);
        currentLen = 0;
        current = 0;
      ***REMOVED***
      start = 26;
    ***REMOVED***

    return res;
  ***REMOVED***;

  Red.prototype.convertTo = function convertTo (num) ***REMOVED***
    var r = num.umod(this.m);

    return r === num ? r.clone() : r;
  ***REMOVED***;

  Red.prototype.convertFrom = function convertFrom (num) ***REMOVED***
    var res = num.clone();
    res.red = null;
    return res;
  ***REMOVED***;

  //
  // Montgomery method engine
  //

  BN.mont = function mont (num) ***REMOVED***
    return new Mont(num);
  ***REMOVED***;

  function Mont (m) ***REMOVED***
    Red.call(this, m);

    this.shift = this.m.bitLength();
    if (this.shift % 26 !== 0) ***REMOVED***
      this.shift += 26 - (this.shift % 26);
    ***REMOVED***

    this.r = new BN(1).iushln(this.shift);
    this.r2 = this.imod(this.r.sqr());
    this.rinv = this.r._invmp(this.m);

    this.minv = this.rinv.mul(this.r).isubn(1).div(this.m);
    this.minv = this.minv.umod(this.r);
    this.minv = this.r.sub(this.minv);
  ***REMOVED***
  inherits(Mont, Red);

  Mont.prototype.convertTo = function convertTo (num) ***REMOVED***
    return this.imod(num.ushln(this.shift));
  ***REMOVED***;

  Mont.prototype.convertFrom = function convertFrom (num) ***REMOVED***
    var r = this.imod(num.mul(this.rinv));
    r.red = null;
    return r;
  ***REMOVED***;

  Mont.prototype.imul = function imul (a, b) ***REMOVED***
    if (a.isZero() || b.isZero()) ***REMOVED***
      a.words[0] = 0;
      a.length = 1;
      return a;
    ***REMOVED***

    var t = a.imul(b);
    var c = t.maskn(this.shift).mul(this.minv).imaskn(this.shift).mul(this.m);
    var u = t.isub(c).iushrn(this.shift);
    var res = u;

    if (u.cmp(this.m) >= 0) ***REMOVED***
      res = u.isub(this.m);
    ***REMOVED*** else if (u.cmpn(0) < 0) ***REMOVED***
      res = u.iadd(this.m);
    ***REMOVED***

    return res._forceRed(this);
  ***REMOVED***;

  Mont.prototype.mul = function mul (a, b) ***REMOVED***
    if (a.isZero() || b.isZero()) return new BN(0)._forceRed(this);

    var t = a.mul(b);
    var c = t.maskn(this.shift).mul(this.minv).imaskn(this.shift).mul(this.m);
    var u = t.isub(c).iushrn(this.shift);
    var res = u;
    if (u.cmp(this.m) >= 0) ***REMOVED***
      res = u.isub(this.m);
    ***REMOVED*** else if (u.cmpn(0) < 0) ***REMOVED***
      res = u.iadd(this.m);
    ***REMOVED***

    return res._forceRed(this);
  ***REMOVED***;

  Mont.prototype.invm = function invm (a) ***REMOVED***
    // (AR)^-1 * R^2 = (A^-1 * R^-1) * R^2 = A^-1 * R
    var res = this.imod(a._invmp(this.m).mul(this.r2));
    return res._forceRed(this);
  ***REMOVED***;
***REMOVED***)(typeof module === 'undefined' || module, this);
