// es6-map/primitive/index.js
'use strict';

var clear          = require('es5-ext/object/clear')
  , setPrototypeOf = require('es5-ext/object/set-prototype-of')
  , validValue     = require('es5-ext/object/valid-value')
  , callable       = require('es5-ext/object/valid-callable')
  , d              = require('d')
  , iterator       = require('es6-iterator/valid-iterable')
  , forOf          = require('es6-iterator/for-of')
  , isNative       = require('../is-native-implemented')
  , MapPolyfill    = require('../polyfill')
  , Iterator       = require('../lib/primitive-iterator')

  , call = Function.prototype.call
  , create = Object.create, defineProperty = Object.defineProperty
  , defineProperties = Object.defineProperties, getPrototypeOf = Object.getPrototypeOf
  , hasOwnProperty = Object.prototype.hasOwnProperty
  , PrimitiveMap;

module.exports = PrimitiveMap = function (/*iterable, serialize*/) ***REMOVED***
	var iterable = arguments[0], serialize = arguments[1], self;
	if (!(this instanceof PrimitiveMap)) throw new TypeError('Constructor requires \'new\'');
	if (isNative && setPrototypeOf && (Map !== MapPolyfill)) ***REMOVED***
		self = setPrototypeOf(new Map(), getPrototypeOf(this));
	***REMOVED*** else ***REMOVED***
		self = this;
	***REMOVED***
	if (iterable != null) iterator(iterable);
	if (serialize !== undefined) ***REMOVED***
		callable(serialize);
		defineProperty(self, '_serialize', d('', serialize));
	***REMOVED***
	defineProperties(self, ***REMOVED***
		__mapKeysData__: d('c', create(null)),
		__mapValuesData__: d('c', create(null)),
		__size__: d('w', 0)
	***REMOVED***);
	if (!iterable) return self;
	forOf(iterable, function (value) ***REMOVED***
		var key = validValue(value)[0], sKey = self._serialize(key);
		if (sKey == null) throw new TypeError(key + " cannot be serialized");
		value = value[1];
		if (hasOwnProperty.call(self.__mapKeysData__, sKey)) ***REMOVED***
			if (self.__mapValuesData__[sKey] === value) return;
		***REMOVED*** else ***REMOVED***
			++self.__size__;
		***REMOVED***
		self.__mapKeysData__[sKey] = key;
		self.__mapValuesData__[sKey] = value;
	***REMOVED***);
	return self;
***REMOVED***;
if (setPrototypeOf) setPrototypeOf(PrimitiveMap, MapPolyfill);

PrimitiveMap.prototype = create(MapPolyfill.prototype, ***REMOVED***
	constructor: d(PrimitiveMap),
	_serialize: d(function (value) ***REMOVED***
		if (value && (typeof value.toString !== 'function')) return null;
		return String(value);
	***REMOVED***),
	clear: d(function () ***REMOVED***
		if (!this.__size__) return;
		clear(this.__mapKeysData__);
		clear(this.__mapValuesData__);
		this.__size__ = 0;
		this.emit('_clear');
	***REMOVED***),
	delete: d(function (key) ***REMOVED***
		var sKey = this._serialize(key);
		if (sKey == null) return false;
		if (!hasOwnProperty.call(this.__mapKeysData__, sKey)) return false;
		delete this.__mapKeysData__[sKey];
		delete this.__mapValuesData__[sKey];
		--this.__size__;
		this.emit('_delete', sKey);
		return true;
	***REMOVED***),
	entries: d(function () ***REMOVED*** return new Iterator(this, 'key+value'); ***REMOVED***),
	forEach: d(function (cb/*, thisArg*/) ***REMOVED***
		var thisArg = arguments[1], iterator, result, sKey;
		callable(cb);
		iterator = this.entries();
		result = iterator._next();
		while (result !== undefined) ***REMOVED***
			sKey = iterator.__list__[result];
			call.call(cb, thisArg, this.__mapValuesData__[sKey],
				this.__mapKeysData__[sKey], this);
			result = iterator._next();
		***REMOVED***
	***REMOVED***),
	get: d(function (key) ***REMOVED***
		var sKey = this._serialize(key);
		if (sKey == null) return;
		return this.__mapValuesData__[sKey];
	***REMOVED***),
	has: d(function (key) ***REMOVED***
		var sKey = this._serialize(key);
		if (sKey == null) return false;
		return hasOwnProperty.call(this.__mapKeysData__, sKey);
	***REMOVED***),
	keys: d(function () ***REMOVED*** return new Iterator(this, 'key'); ***REMOVED***),
	size: d.gs(function () ***REMOVED*** return this.__size__; ***REMOVED***),
	set: d(function (key, value) ***REMOVED***
		var sKey = this._serialize(key);
		if (sKey == null) throw new TypeError(key + " cannot be serialized");
		if (hasOwnProperty.call(this.__mapKeysData__, sKey)) ***REMOVED***
			if (this.__mapValuesData__[sKey] === value) return this;
		***REMOVED*** else ***REMOVED***
			++this.__size__;
		***REMOVED***
		this.__mapKeysData__[sKey] = key;
		this.__mapValuesData__[sKey] = value;
		this.emit('_add', sKey);
		return this;
	***REMOVED***),
	values: d(function () ***REMOVED*** return new Iterator(this, 'value'); ***REMOVED***)
***REMOVED***);
