// power-assert-formatter/lib/create.js
'use strict';

var createFormatter = require('power-assert-context-formatter');
var appendAst = require('power-assert-context-reducer-ast');
var FileRenderer = require('power-assert-renderer-file');
var AssertionRenderer = require('power-assert-renderer-assertion');
var DiagramRenderer = require('power-assert-renderer-diagram');
var ComparisonRenderer = require('power-assert-renderer-comparison');
var defaultOptions = require('./default-options');
var assign = require('core-js/library/fn/object/assign');
var map = require('core-js/library/fn/array/map');

// "Browserify can only analyze static requires. It is not in the scope of browserify to handle dynamic requires."
// https://github.com/substack/node-browserify/issues/377
var defaultRendererClasses = ***REMOVED***
    './built-in/file': FileRenderer,
    './built-in/assertion': AssertionRenderer,
    './built-in/diagram': DiagramRenderer,
    './built-in/binary-expression': ComparisonRenderer
***REMOVED***;

function toRendererClass (rendererName) ***REMOVED***
    var RendererClass;
    if (typeof rendererName === 'function') ***REMOVED***
        RendererClass = rendererName;
    ***REMOVED*** else if (typeof rendererName === 'string') ***REMOVED***
        if (defaultRendererClasses[rendererName]) ***REMOVED***
            RendererClass = defaultRendererClasses[rendererName];
        ***REMOVED*** else ***REMOVED***
            RendererClass = require(rendererName);
        ***REMOVED***
    ***REMOVED***
    return RendererClass;
***REMOVED***

function create (options) ***REMOVED***
    var config = assign(defaultOptions(), options);
    var rendererClasses = map(config.renderers, toRendererClass);
    var renderers = map(rendererClasses, function (clazz) ***REMOVED***
        return ***REMOVED*** ctor: clazz, options: config ***REMOVED***;
    ***REMOVED***);
    return createFormatter(assign(***REMOVED******REMOVED***, config, ***REMOVED***
        reducers: [
            appendAst
        ],
        renderers: renderers,
        legacy: true
    ***REMOVED***));
***REMOVED***

create.renderers = ***REMOVED***
    AssertionRenderer: AssertionRenderer,
    FileRenderer: FileRenderer,
    DiagramRenderer: DiagramRenderer,
    BinaryExpressionRenderer: ComparisonRenderer
***REMOVED***;
create.defaultOptions = defaultOptions;
module.exports = create;
