// console-control-strings/index.js
'use strict'

// These tables borrowed from `ansi`

var prefix = '\x1b['

exports.up = function up (num) ***REMOVED***
  return prefix + (num || '') + 'A'
***REMOVED***

exports.down = function down (num) ***REMOVED***
  return prefix + (num || '') + 'B'
***REMOVED***

exports.forward = function forward (num) ***REMOVED***
  return prefix + (num || '') + 'C'
***REMOVED***

exports.back = function back (num) ***REMOVED***
  return prefix + (num || '') + 'D'
***REMOVED***

exports.nextLine = function nextLine (num) ***REMOVED***
  return prefix + (num || '') + 'E'
***REMOVED***

exports.previousLine = function previousLine (num) ***REMOVED***
  return prefix + (num || '') + 'F'
***REMOVED***

exports.horizontalAbsolute = function horizontalAbsolute (num) ***REMOVED***
  if (num == null) throw new Error('horizontalAboslute requires a column to position to')
  return prefix + num + 'G'
***REMOVED***

exports.eraseData = function eraseData () ***REMOVED***
  return prefix + 'J'
***REMOVED***

exports.eraseLine = function eraseLine () ***REMOVED***
  return prefix + 'K'
***REMOVED***

exports.goto = function (x, y) ***REMOVED***
  return prefix + y + ';' + x + 'H'
***REMOVED***

exports.gotoSOL = function () ***REMOVED***
  return '\r'
***REMOVED***

exports.beep = function () ***REMOVED***
  return '\x07'
***REMOVED***

exports.hideCursor = function hideCursor () ***REMOVED***
  return prefix + '?25l'
***REMOVED***

exports.showCursor = function showCursor () ***REMOVED***
  return prefix + '?25h'
***REMOVED***

var colors = ***REMOVED***
  reset: 0,
// styles
  bold: 1,
  italic: 3,
  underline: 4,
  inverse: 7,
// resets
  stopBold: 22,
  stopItalic: 23,
  stopUnderline: 24,
  stopInverse: 27,
// colors
  white: 37,
  black: 30,
  blue: 34,
  cyan: 36,
  green: 32,
  magenta: 35,
  red: 31,
  yellow: 33,
  bgWhite: 47,
  bgBlack: 40,
  bgBlue: 44,
  bgCyan: 46,
  bgGreen: 42,
  bgMagenta: 45,
  bgRed: 41,
  bgYellow: 43,

  grey: 90,
  brightBlack: 90,
  brightRed: 91,
  brightGreen: 92,
  brightYellow: 93,
  brightBlue: 94,
  brightMagenta: 95,
  brightCyan: 96,
  brightWhite: 97,

  bgGrey: 100,
  bgBrightBlack: 100,
  bgBrightRed: 101,
  bgBrightGreen: 102,
  bgBrightYellow: 103,
  bgBrightBlue: 104,
  bgBrightMagenta: 105,
  bgBrightCyan: 106,
  bgBrightWhite: 107
***REMOVED***

exports.color = function color (colorWith) ***REMOVED***
  if (arguments.length !== 1 || !Array.isArray(colorWith)) ***REMOVED***
    colorWith = Array.prototype.slice.call(arguments)
  ***REMOVED***
  return prefix + colorWith.map(colorNameToCode).join(';') + 'm'
***REMOVED***

function colorNameToCode (color) ***REMOVED***
  if (colors[color] != null) return colors[color]
  throw new Error('Unknown color or style name: ' + color)
***REMOVED***
